// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/instructions/closeExecute.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type CloseExecuteInstructionAccounts = {
  admin: Signer;
  config: PublicKey | Pda;
  executeHash: PublicKey | Pda;
};

// Data.
export type CloseExecuteInstructionData = {
  discriminator: Uint8Array;
  digestHash: Uint8Array;
};

export type CloseExecuteInstructionDataArgs = { digestHash: Uint8Array };

export function getCloseExecuteInstructionDataSerializer(): Serializer<
  CloseExecuteInstructionDataArgs,
  CloseExecuteInstructionData
> {
  return mapSerializer<
    CloseExecuteInstructionDataArgs,
    any,
    CloseExecuteInstructionData
  >(
    struct<CloseExecuteInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['digestHash', bytes({ size: 32 })],
      ],
      { description: 'CloseExecuteInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([253, 220, 23, 127, 15, 171, 189, 226]),
    })
  ) as Serializer<CloseExecuteInstructionDataArgs, CloseExecuteInstructionData>;
}

// Args.
export type CloseExecuteInstructionArgs = CloseExecuteInstructionDataArgs;

// Instruction.
export function closeExecute(
  context: Pick<Context, 'programs'>,
  input: CloseExecuteInstructionAccounts & CloseExecuteInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'dvn',
    'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false as boolean,
      value: input.admin ?? null,
    },
    config: {
      index: 1,
      isWritable: true as boolean,
      value: input.config ?? null,
    },
    executeHash: {
      index: 2,
      isWritable: true as boolean,
      value: input.executeHash ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: CloseExecuteInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getCloseExecuteInstructionDataSerializer().serialize(
    resolvedArgs as CloseExecuteInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
