// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/instructions/extendDvnConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type ExtendDvnConfigInstructionAccounts = {
  admin: Signer;
  config: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
};

// Data.
export type ExtendDvnConfigInstructionData = { discriminator: Uint8Array };

export type ExtendDvnConfigInstructionDataArgs = {};

export function getExtendDvnConfigInstructionDataSerializer(): Serializer<
  ExtendDvnConfigInstructionDataArgs,
  ExtendDvnConfigInstructionData
> {
  return mapSerializer<
    ExtendDvnConfigInstructionDataArgs,
    any,
    ExtendDvnConfigInstructionData
  >(
    struct<ExtendDvnConfigInstructionData>(
      [['discriminator', bytes({ size: 8 })]],
      { description: 'ExtendDvnConfigInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([183, 248, 24, 49, 110, 229, 169, 66]),
    })
  ) as Serializer<
    ExtendDvnConfigInstructionDataArgs,
    ExtendDvnConfigInstructionData
  >;
}

// Instruction.
export function extendDvnConfig(
  context: Pick<Context, 'programs'>,
  input: ExtendDvnConfigInstructionAccounts
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'dvn',
    'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: true as boolean,
      value: input.admin ?? null,
    },
    config: {
      index: 1,
      isWritable: true as boolean,
      value: input.config ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Default values.
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getExtendDvnConfigInstructionDataSerializer().serialize({});

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
