// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/instructions/initDvn.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type InitDvnInstructionAccounts = {
  payer?: Signer;
  config: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
};

// Data.
export type InitDvnInstructionData = {
  discriminator: Uint8Array;
  vid: number;
  msglibs: Array<PublicKey>;
  priceFeed: PublicKey;
  signers: Array<Uint8Array>;
  quorum: number;
  admins: Array<PublicKey>;
};

export type InitDvnInstructionDataArgs = {
  vid: number;
  msglibs: Array<PublicKey>;
  priceFeed: PublicKey;
  signers: Array<Uint8Array>;
  quorum: number;
  admins: Array<PublicKey>;
};

export function getInitDvnInstructionDataSerializer(): Serializer<
  InitDvnInstructionDataArgs,
  InitDvnInstructionData
> {
  return mapSerializer<InitDvnInstructionDataArgs, any, InitDvnInstructionData>(
    struct<InitDvnInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['vid', u32()],
        ['msglibs', array(publicKeySerializer())],
        ['priceFeed', publicKeySerializer()],
        ['signers', array(bytes({ size: 64 }))],
        ['quorum', u8()],
        ['admins', array(publicKeySerializer())],
      ],
      { description: 'InitDvnInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([58, 108, 132, 12, 20, 10, 215, 210]),
    })
  ) as Serializer<InitDvnInstructionDataArgs, InitDvnInstructionData>;
}

// Args.
export type InitDvnInstructionArgs = InitDvnInstructionDataArgs;

// Instruction.
export function initDvn(
  context: Pick<Context, 'payer' | 'programs'>,
  input: InitDvnInstructionAccounts & InitDvnInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'dvn',
    'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
  );

  // Accounts.
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true as boolean,
      value: input.payer ?? null,
    },
    config: {
      index: 1,
      isWritable: true as boolean,
      value: input.config ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: InitDvnInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getInitDvnInstructionDataSerializer().serialize(
    resolvedArgs as InitDvnInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
