// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/instructions/setConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import {
  AdminConfig,
  AdminConfigArgs,
  getAdminConfigSerializer,
} from '../types';

// Accounts.
export type SetConfigInstructionAccounts = {
  admin: Signer;
  config: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type SetConfigInstructionData = {
  discriminator: Uint8Array;
  config: AdminConfig;
};

export type SetConfigInstructionDataArgs = { config: AdminConfigArgs };

export function getSetConfigInstructionDataSerializer(): Serializer<
  SetConfigInstructionDataArgs,
  SetConfigInstructionData
> {
  return mapSerializer<
    SetConfigInstructionDataArgs,
    any,
    SetConfigInstructionData
  >(
    struct<SetConfigInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['config', getAdminConfigSerializer()],
      ],
      { description: 'SetConfigInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([108, 158, 154, 175, 212, 98, 52, 66]),
    })
  ) as Serializer<SetConfigInstructionDataArgs, SetConfigInstructionData>;
}

// Args.
export type SetConfigInstructionArgs = SetConfigInstructionDataArgs;

// Instruction.
export function setConfig(
  context: Pick<Context, 'programs'>,
  accounts: SetConfigInstructionAccounts,
  args: SetConfigInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'dvn',
    'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.admin ?? null,
    },
    config: {
      index: 1,
      isWritable: true as boolean,
      value: accounts.config ?? null,
    },
    eventAuthority: {
      index: 2,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 3,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetConfigInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSetConfigInstructionDataSerializer().serialize(
    resolvedArgs as SetConfigInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
