// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/instructions/verifiable.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type VerifiableInstructionAccounts = {
  nonce: PublicKey | Pda;
  payloadHash: PublicKey | Pda;
  receiveConfig: PublicKey | Pda;
  defaultReceiveConfig: PublicKey | Pda;
};

// Data.
export type VerifiableInstructionData = {
  discriminator: Uint8Array;
  packetHeader: Array<number>;
  payloadHash: Uint8Array;
};

export type VerifiableInstructionDataArgs = {
  packetHeader: Array<number>;
  payloadHash: Uint8Array;
};

export function getVerifiableInstructionDataSerializer(): Serializer<
  VerifiableInstructionDataArgs,
  VerifiableInstructionData
> {
  return mapSerializer<
    VerifiableInstructionDataArgs,
    any,
    VerifiableInstructionData
  >(
    struct<VerifiableInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['packetHeader', array(u8(), { size: 81 })],
        ['payloadHash', bytes({ size: 32 })],
      ],
      { description: 'VerifiableInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([249, 50, 227, 157, 238, 249, 211, 90]),
    })
  ) as Serializer<VerifiableInstructionDataArgs, VerifiableInstructionData>;
}

// Args.
export type VerifiableInstructionArgs = VerifiableInstructionDataArgs;

// Instruction.
export function verifiable(
  context: Pick<Context, 'programs'>,
  accounts: VerifiableInstructionAccounts,
  args: VerifiableInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'dvn',
    'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
  );

  // Accounts.
  const resolvedAccounts = {
    nonce: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.nonce ?? null,
    },
    payloadHash: {
      index: 1,
      isWritable: false as boolean,
      value: accounts.payloadHash ?? null,
    },
    receiveConfig: {
      index: 2,
      isWritable: false as boolean,
      value: accounts.receiveConfig ?? null,
    },
    defaultReceiveConfig: {
      index: 3,
      isWritable: false as boolean,
      value: accounts.defaultReceiveConfig ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: VerifiableInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getVerifiableInstructionDataSerializer().serialize(
    resolvedArgs as VerifiableInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
