// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/instructions/commitVerification.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type CommitVerificationInstructionAccounts = {
  /** The custom receive config account may be uninitialized, so deserialize it only if it's initialized */
  receiveConfig: PublicKey | Pda;
  defaultReceiveConfig: PublicKey | Pda;
  uln: PublicKey | Pda;
};

// Data.
export type CommitVerificationInstructionData = {
  discriminator: Uint8Array;
  packetHeader: Array<number>;
  payloadHash: Uint8Array;
};

export type CommitVerificationInstructionDataArgs = {
  packetHeader: Array<number>;
  payloadHash: Uint8Array;
};

export function getCommitVerificationInstructionDataSerializer(): Serializer<
  CommitVerificationInstructionDataArgs,
  CommitVerificationInstructionData
> {
  return mapSerializer<
    CommitVerificationInstructionDataArgs,
    any,
    CommitVerificationInstructionData
  >(
    struct<CommitVerificationInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['packetHeader', array(u8(), { size: 81 })],
        ['payloadHash', bytes({ size: 32 })],
      ],
      { description: 'CommitVerificationInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([194, 120, 44, 252, 140, 215, 139, 103]),
    })
  ) as Serializer<
    CommitVerificationInstructionDataArgs,
    CommitVerificationInstructionData
  >;
}

// Args.
export type CommitVerificationInstructionArgs =
  CommitVerificationInstructionDataArgs;

// Instruction.
export function commitVerification(
  context: Pick<Context, 'programs'>,
  input: CommitVerificationInstructionAccounts &
    CommitVerificationInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );

  // Accounts.
  const resolvedAccounts = {
    receiveConfig: {
      index: 0,
      isWritable: false as boolean,
      value: input.receiveConfig ?? null,
    },
    defaultReceiveConfig: {
      index: 1,
      isWritable: false as boolean,
      value: input.defaultReceiveConfig ?? null,
    },
    uln: { index: 2, isWritable: false as boolean, value: input.uln ?? null },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: CommitVerificationInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getCommitVerificationInstructionDataSerializer().serialize(
    resolvedArgs as CommitVerificationInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
