// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/instructions/setTreasury.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Option,
  OptionOrNullable,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  option,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import { Treasury, TreasuryArgs, getTreasurySerializer } from '../types';

// Accounts.
export type SetTreasuryInstructionAccounts = {
  /** The admin or treasury admin */
  signer: Signer;
  uln: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type SetTreasuryInstructionData = {
  discriminator: Uint8Array;
  treasury: Option<Treasury>;
};

export type SetTreasuryInstructionDataArgs = {
  treasury: OptionOrNullable<TreasuryArgs>;
};

export function getSetTreasuryInstructionDataSerializer(): Serializer<
  SetTreasuryInstructionDataArgs,
  SetTreasuryInstructionData
> {
  return mapSerializer<
    SetTreasuryInstructionDataArgs,
    any,
    SetTreasuryInstructionData
  >(
    struct<SetTreasuryInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['treasury', option(getTreasurySerializer())],
      ],
      { description: 'SetTreasuryInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([57, 97, 196, 95, 195, 206, 106, 136]),
    })
  ) as Serializer<SetTreasuryInstructionDataArgs, SetTreasuryInstructionData>;
}

// Args.
export type SetTreasuryInstructionArgs = SetTreasuryInstructionDataArgs;

// Instruction.
export function setTreasury(
  context: Pick<Context, 'programs'>,
  input: SetTreasuryInstructionAccounts & SetTreasuryInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );

  // Accounts.
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false as boolean,
      value: input.signer ?? null,
    },
    uln: { index: 1, isWritable: true as boolean, value: input.uln ?? null },
    eventAuthority: {
      index: 2,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 3,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetTreasuryInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSetTreasuryInstructionDataSerializer().serialize(
    resolvedArgs as SetTreasuryInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
