// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/instructions/verify.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  struct,
  u64,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type VerifyInstructionAccounts = {
  dvn: Signer;
  confirmations: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type VerifyInstructionData = {
  discriminator: Uint8Array;
  packetHeader: Array<number>;
  payloadHash: Uint8Array;
  confirmations: bigint;
};

export type VerifyInstructionDataArgs = {
  packetHeader: Array<number>;
  payloadHash: Uint8Array;
  confirmations: number | bigint;
};

export function getVerifyInstructionDataSerializer(): Serializer<
  VerifyInstructionDataArgs,
  VerifyInstructionData
> {
  return mapSerializer<VerifyInstructionDataArgs, any, VerifyInstructionData>(
    struct<VerifyInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['packetHeader', array(u8(), { size: 81 })],
        ['payloadHash', bytes({ size: 32 })],
        ['confirmations', u64()],
      ],
      { description: 'VerifyInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([133, 161, 141, 48, 120, 198, 88, 150]),
    })
  ) as Serializer<VerifyInstructionDataArgs, VerifyInstructionData>;
}

// Args.
export type VerifyInstructionArgs = VerifyInstructionDataArgs;

// Instruction.
export function verify(
  context: Pick<Context, 'programs'>,
  accounts: VerifyInstructionAccounts,
  args: VerifyInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );

  // Accounts.
  const resolvedAccounts = {
    dvn: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.dvn ?? null,
    },
    confirmations: {
      index: 1,
      isWritable: true as boolean,
      value: accounts.confirmations ?? null,
    },
    eventAuthority: {
      index: 2,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 3,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: VerifyInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getVerifyInstructionDataSerializer().serialize(
    resolvedArgs as VerifyInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
