// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/instructions/version.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Data.
export type VersionInstructionData = { discriminator: Uint8Array };

export type VersionInstructionDataArgs = {};

export function getVersionInstructionDataSerializer(): Serializer<
  VersionInstructionDataArgs,
  VersionInstructionData
> {
  return mapSerializer<VersionInstructionDataArgs, any, VersionInstructionData>(
    struct<VersionInstructionData>([['discriminator', bytes({ size: 8 })]], {
      description: 'VersionInstructionData',
    }),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([118, 65, 195, 198, 129, 216, 252, 192]),
    })
  ) as Serializer<VersionInstructionDataArgs, VersionInstructionData>;
}

// Instruction.
export function version(
  context: Pick<Context, 'programs'>
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );

  // Accounts.
  const resolvedAccounts = {} satisfies ResolvedAccountsWithIndices;

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts as ResolvedAccountsWithIndices
  );

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getVersionInstructionDataSerializer().serialize({});

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
