// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/errors/index.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * InvalidSendLibrary: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidSendLibraryError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'InvalidSendLibrary'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidSendLibraryError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new InvalidSendLibraryError())
createErrorFromNameLookup.set(
  'InvalidSendLibrary',
  () => new InvalidSendLibraryError()
)

/**
 * InvalidReceiveLibrary: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidReceiveLibraryError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'InvalidReceiveLibrary'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidReceiveLibraryError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new InvalidReceiveLibraryError())
createErrorFromNameLookup.set(
  'InvalidReceiveLibrary',
  () => new InvalidReceiveLibraryError()
)

/**
 * SameValue: ''
 *
 * @category Errors
 * @category generated
 */
export class SameValueError extends Error {
  readonly code: number = 0x1772
  readonly name: string = 'SameValue'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, SameValueError)
    }
  }
}

createErrorFromCodeLookup.set(0x1772, () => new SameValueError())
createErrorFromNameLookup.set('SameValue', () => new SameValueError())

/**
 * AccountNotFound: ''
 *
 * @category Errors
 * @category generated
 */
export class AccountNotFoundError extends Error {
  readonly code: number = 0x1773
  readonly name: string = 'AccountNotFound'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, AccountNotFoundError)
    }
  }
}

createErrorFromCodeLookup.set(0x1773, () => new AccountNotFoundError())
createErrorFromNameLookup.set(
  'AccountNotFound',
  () => new AccountNotFoundError()
)

/**
 * OnlySendLib: ''
 *
 * @category Errors
 * @category generated
 */
export class OnlySendLibError extends Error {
  readonly code: number = 0x1774
  readonly name: string = 'OnlySendLib'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, OnlySendLibError)
    }
  }
}

createErrorFromCodeLookup.set(0x1774, () => new OnlySendLibError())
createErrorFromNameLookup.set('OnlySendLib', () => new OnlySendLibError())

/**
 * OnlyReceiveLib: ''
 *
 * @category Errors
 * @category generated
 */
export class OnlyReceiveLibError extends Error {
  readonly code: number = 0x1775
  readonly name: string = 'OnlyReceiveLib'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, OnlyReceiveLibError)
    }
  }
}

createErrorFromCodeLookup.set(0x1775, () => new OnlyReceiveLibError())
createErrorFromNameLookup.set('OnlyReceiveLib', () => new OnlyReceiveLibError())

/**
 * InvalidExpiry: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidExpiryError extends Error {
  readonly code: number = 0x1776
  readonly name: string = 'InvalidExpiry'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidExpiryError)
    }
  }
}

createErrorFromCodeLookup.set(0x1776, () => new InvalidExpiryError())
createErrorFromNameLookup.set('InvalidExpiry', () => new InvalidExpiryError())

/**
 * OnlyNonDefaultLib: ''
 *
 * @category Errors
 * @category generated
 */
export class OnlyNonDefaultLibError extends Error {
  readonly code: number = 0x1777
  readonly name: string = 'OnlyNonDefaultLib'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, OnlyNonDefaultLibError)
    }
  }
}

createErrorFromCodeLookup.set(0x1777, () => new OnlyNonDefaultLibError())
createErrorFromNameLookup.set(
  'OnlyNonDefaultLib',
  () => new OnlyNonDefaultLibError()
)

/**
 * InvalidAmount: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidAmountError extends Error {
  readonly code: number = 0x1778
  readonly name: string = 'InvalidAmount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidAmountError)
    }
  }
}

createErrorFromCodeLookup.set(0x1778, () => new InvalidAmountError())
createErrorFromNameLookup.set('InvalidAmount', () => new InvalidAmountError())

/**
 * InvalidNonce: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidNonceError extends Error {
  readonly code: number = 0x1779
  readonly name: string = 'InvalidNonce'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidNonceError)
    }
  }
}

createErrorFromCodeLookup.set(0x1779, () => new InvalidNonceError())
createErrorFromNameLookup.set('InvalidNonce', () => new InvalidNonceError())

/**
 * Unauthorized: ''
 *
 * @category Errors
 * @category generated
 */
export class UnauthorizedError extends Error {
  readonly code: number = 0x177a
  readonly name: string = 'Unauthorized'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, UnauthorizedError)
    }
  }
}

createErrorFromCodeLookup.set(0x177a, () => new UnauthorizedError())
createErrorFromNameLookup.set('Unauthorized', () => new UnauthorizedError())

/**
 * PayloadHashNotFound: ''
 *
 * @category Errors
 * @category generated
 */
export class PayloadHashNotFoundError extends Error {
  readonly code: number = 0x177b
  readonly name: string = 'PayloadHashNotFound'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, PayloadHashNotFoundError)
    }
  }
}

createErrorFromCodeLookup.set(0x177b, () => new PayloadHashNotFoundError())
createErrorFromNameLookup.set(
  'PayloadHashNotFound',
  () => new PayloadHashNotFoundError()
)

/**
 * ComposeNotFound: ''
 *
 * @category Errors
 * @category generated
 */
export class ComposeNotFoundError extends Error {
  readonly code: number = 0x177c
  readonly name: string = 'ComposeNotFound'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ComposeNotFoundError)
    }
  }
}

createErrorFromCodeLookup.set(0x177c, () => new ComposeNotFoundError())
createErrorFromNameLookup.set(
  'ComposeNotFound',
  () => new ComposeNotFoundError()
)

/**
 * InvalidPayloadHash: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidPayloadHashError extends Error {
  readonly code: number = 0x177d
  readonly name: string = 'InvalidPayloadHash'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidPayloadHashError)
    }
  }
}

createErrorFromCodeLookup.set(0x177d, () => new InvalidPayloadHashError())
createErrorFromNameLookup.set(
  'InvalidPayloadHash',
  () => new InvalidPayloadHashError()
)

/**
 * LzTokenUnavailable: ''
 *
 * @category Errors
 * @category generated
 */
export class LzTokenUnavailableError extends Error {
  readonly code: number = 0x177e
  readonly name: string = 'LzTokenUnavailable'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, LzTokenUnavailableError)
    }
  }
}

createErrorFromCodeLookup.set(0x177e, () => new LzTokenUnavailableError())
createErrorFromNameLookup.set(
  'LzTokenUnavailable',
  () => new LzTokenUnavailableError()
)

/**
 * ReadOnlyAccount: ''
 *
 * @category Errors
 * @category generated
 */
export class ReadOnlyAccountError extends Error {
  readonly code: number = 0x177f
  readonly name: string = 'ReadOnlyAccount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ReadOnlyAccountError)
    }
  }
}

createErrorFromCodeLookup.set(0x177f, () => new ReadOnlyAccountError())
createErrorFromNameLookup.set(
  'ReadOnlyAccount',
  () => new ReadOnlyAccountError()
)

/**
 * InvalidMessageLib: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidMessageLibError extends Error {
  readonly code: number = 0x1780
  readonly name: string = 'InvalidMessageLib'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidMessageLibError)
    }
  }
}

createErrorFromCodeLookup.set(0x1780, () => new InvalidMessageLibError())
createErrorFromNameLookup.set(
  'InvalidMessageLib',
  () => new InvalidMessageLibError()
)

/**
 * WritableAccountNotAllowed: ''
 *
 * @category Errors
 * @category generated
 */
export class WritableAccountNotAllowedError extends Error {
  readonly code: number = 0x1781
  readonly name: string = 'WritableAccountNotAllowed'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, WritableAccountNotAllowedError)
    }
  }
}

createErrorFromCodeLookup.set(
  0x1781,
  () => new WritableAccountNotAllowedError()
)
createErrorFromNameLookup.set(
  'WritableAccountNotAllowed',
  () => new WritableAccountNotAllowedError()
)

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
