// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/types/Config.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import { UlnConfig, ulnConfigBeet } from './UlnConfig'
import { ExecutorConfig, executorConfigBeet } from './ExecutorConfig'
/**
 * This type is used to derive the {@link Config} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link Config} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type ConfigRecord = {
  SendUln: { fields: [UlnConfig] }
  ReceiveUln: { fields: [UlnConfig] }
  Executor: { fields: [ExecutorConfig] }
}

/**
 * Union type respresenting the Config data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isConfig*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type Config = beet.DataEnumKeyAsKind<ConfigRecord>

export const isConfigSendUln = (
  x: Config
): x is Config & { __kind: 'SendUln' } => x.__kind === 'SendUln'
export const isConfigReceiveUln = (
  x: Config
): x is Config & { __kind: 'ReceiveUln' } => x.__kind === 'ReceiveUln'
export const isConfigExecutor = (
  x: Config
): x is Config & { __kind: 'Executor' } => x.__kind === 'Executor'

/**
 * @category userTypes
 * @category generated
 */
export const configBeet = beet.dataEnum<ConfigRecord>([
  [
    'SendUln',
    new beet.FixableBeetArgsStruct<ConfigRecord['SendUln']>(
      [['fields', beet.tuple([ulnConfigBeet])]],
      'ConfigRecord["SendUln"]'
    ),
  ],
  [
    'ReceiveUln',
    new beet.FixableBeetArgsStruct<ConfigRecord['ReceiveUln']>(
      [['fields', beet.tuple([ulnConfigBeet])]],
      'ConfigRecord["ReceiveUln"]'
    ),
  ],
  [
    'Executor',
    new beet.BeetArgsStruct<ConfigRecord['Executor']>(
      [['fields', beet.fixedSizeTuple([executorConfigBeet])]],
      'ConfigRecord["Executor"]'
    ),
  ],
]) as beet.FixableBeet<Config, Config>
