// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/types/AdminConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { DstConfig, dstConfigBeet } from './DstConfig'
/**
 * This type is used to derive the {@link AdminConfig} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link AdminConfig} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type AdminConfigRecord = {
  Admins: { fields: [web3.PublicKey[]] }
  DefaultMultiplierBps: { fields: [number] }
  DstConfigs: { fields: [DstConfig[]] }
  PriceFeed: { fields: [web3.PublicKey] }
  RemoveDstConfigs: { fields: [number[]] }
}

/**
 * Union type respresenting the AdminConfig data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isAdminConfig*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type AdminConfig = beet.DataEnumKeyAsKind<AdminConfigRecord>

export const isAdminConfigAdmins = (
  x: AdminConfig
): x is AdminConfig & { __kind: 'Admins' } => x.__kind === 'Admins'
export const isAdminConfigDefaultMultiplierBps = (
  x: AdminConfig
): x is AdminConfig & { __kind: 'DefaultMultiplierBps' } =>
  x.__kind === 'DefaultMultiplierBps'
export const isAdminConfigDstConfigs = (
  x: AdminConfig
): x is AdminConfig & { __kind: 'DstConfigs' } => x.__kind === 'DstConfigs'
export const isAdminConfigPriceFeed = (
  x: AdminConfig
): x is AdminConfig & { __kind: 'PriceFeed' } => x.__kind === 'PriceFeed'
export const isAdminConfigRemoveDstConfigs = (
  x: AdminConfig
): x is AdminConfig & { __kind: 'RemoveDstConfigs' } =>
  x.__kind === 'RemoveDstConfigs'

/**
 * @category userTypes
 * @category generated
 */
export const adminConfigBeet = beet.dataEnum<AdminConfigRecord>([
  [
    'Admins',
    new beet.FixableBeetArgsStruct<AdminConfigRecord['Admins']>(
      [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
      'AdminConfigRecord["Admins"]'
    ),
  ],
  [
    'DefaultMultiplierBps',
    new beet.BeetArgsStruct<AdminConfigRecord['DefaultMultiplierBps']>(
      [['fields', beet.fixedSizeTuple([beet.u16])]],
      'AdminConfigRecord["DefaultMultiplierBps"]'
    ),
  ],
  [
    'DstConfigs',
    new beet.FixableBeetArgsStruct<AdminConfigRecord['DstConfigs']>(
      [['fields', beet.tuple([beet.array(dstConfigBeet)])]],
      'AdminConfigRecord["DstConfigs"]'
    ),
  ],
  [
    'PriceFeed',
    new beet.BeetArgsStruct<AdminConfigRecord['PriceFeed']>(
      [['fields', beet.fixedSizeTuple([beetSolana.publicKey])]],
      'AdminConfigRecord["PriceFeed"]'
    ),
  ],
  [
    'RemoveDstConfigs',
    new beet.FixableBeetArgsStruct<AdminConfigRecord['RemoveDstConfigs']>(
      [['fields', beet.tuple([beet.array(beet.u32)])]],
      'AdminConfigRecord["RemoveDstConfigs"]'
    ),
  ],
]) as beet.FixableBeet<AdminConfig, AdminConfig>
