// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/types/ExecuteTransactionDigest.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import {
  TransactionAccount,
  transactionAccountBeet,
} from './TransactionAccount'
export type ExecuteTransactionDigest = {
  vid: number
  programId: web3.PublicKey
  accounts: TransactionAccount[]
  data: Uint8Array
  expiration: beet.bignum
}

/**
 * @category userTypes
 * @category generated
 */
export const executeTransactionDigestBeet =
  new beet.FixableBeetArgsStruct<ExecuteTransactionDigest>(
    [
      ['vid', beet.u32],
      ['programId', beetSolana.publicKey],
      ['accounts', beet.array(transactionAccountBeet)],
      ['data', beet.bytes],
      ['expiration', beet.i64],
    ],
    'ExecuteTransactionDigest'
  )
