// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/types/MultisigConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
/**
 * This type is used to derive the {@link MultisigConfig} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link MultisigConfig} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type MultisigConfigRecord = {
  Admins: { fields: [web3.PublicKey[]] }
  Allowlist: { fields: [web3.PublicKey[]] }
  Denylist: { fields: [web3.PublicKey[]] }
  Msglibs: { fields: [web3.PublicKey[]] }
  Paused: { fields: [boolean] }
  Quorum: { fields: [number] }
  Signers: { fields: [number[] /* size: 64 */[]] }
}

/**
 * Union type respresenting the MultisigConfig data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isMultisigConfig*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type MultisigConfig = beet.DataEnumKeyAsKind<MultisigConfigRecord>

export const isMultisigConfigAdmins = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Admins' } => x.__kind === 'Admins'
export const isMultisigConfigAllowlist = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Allowlist' } => x.__kind === 'Allowlist'
export const isMultisigConfigDenylist = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Denylist' } => x.__kind === 'Denylist'
export const isMultisigConfigMsglibs = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Msglibs' } => x.__kind === 'Msglibs'
export const isMultisigConfigPaused = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Paused' } => x.__kind === 'Paused'
export const isMultisigConfigQuorum = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Quorum' } => x.__kind === 'Quorum'
export const isMultisigConfigSigners = (
  x: MultisigConfig
): x is MultisigConfig & { __kind: 'Signers' } => x.__kind === 'Signers'

/**
 * @category userTypes
 * @category generated
 */
export const multisigConfigBeet = beet.dataEnum<MultisigConfigRecord>([
  [
    'Admins',
    new beet.FixableBeetArgsStruct<MultisigConfigRecord['Admins']>(
      [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
      'MultisigConfigRecord["Admins"]'
    ),
  ],
  [
    'Allowlist',
    new beet.FixableBeetArgsStruct<MultisigConfigRecord['Allowlist']>(
      [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
      'MultisigConfigRecord["Allowlist"]'
    ),
  ],
  [
    'Denylist',
    new beet.FixableBeetArgsStruct<MultisigConfigRecord['Denylist']>(
      [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
      'MultisigConfigRecord["Denylist"]'
    ),
  ],
  [
    'Msglibs',
    new beet.FixableBeetArgsStruct<MultisigConfigRecord['Msglibs']>(
      [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
      'MultisigConfigRecord["Msglibs"]'
    ),
  ],
  [
    'Paused',
    new beet.BeetArgsStruct<MultisigConfigRecord['Paused']>(
      [['fields', beet.fixedSizeTuple([beet.bool])]],
      'MultisigConfigRecord["Paused"]'
    ),
  ],
  [
    'Quorum',
    new beet.BeetArgsStruct<MultisigConfigRecord['Quorum']>(
      [['fields', beet.fixedSizeTuple([beet.u8])]],
      'MultisigConfigRecord["Quorum"]'
    ),
  ],
  [
    'Signers',
    new beet.FixableBeetArgsStruct<MultisigConfigRecord['Signers']>(
      [
        [
          'fields',
          beet.tuple([beet.array(beet.uniformFixedSizeArray(beet.u8, 64))]),
        ],
      ],
      'MultisigConfigRecord["Signers"]'
    ),
  ],
]) as beet.FixableBeet<MultisigConfig, MultisigConfig>
