// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/accounts/PriceFeed.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { Price, priceBeet } from '../types/Price'

/**
 * Arguments used to create {@link PriceFeed}
 * @category Accounts
 * @category generated
 */
export type PriceFeedArgs = {
  admin: web3.PublicKey
  updaters: web3.PublicKey[]
  priceRatioDenominator: beet.bignum
  arbitrumCompressionPercent: beet.bignum
  nativeTokenPriceUsd: beet.COption<beet.bignum>
  prices: Price[]
  bump: number
}

export const priceFeedDiscriminator = [189, 103, 252, 23, 152, 35, 243, 156]
/**
 * Holds the data for the {@link PriceFeed} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class PriceFeed implements PriceFeedArgs {
  private constructor(
    readonly admin: web3.PublicKey,
    readonly updaters: web3.PublicKey[],
    readonly priceRatioDenominator: beet.bignum,
    readonly arbitrumCompressionPercent: beet.bignum,
    readonly nativeTokenPriceUsd: beet.COption<beet.bignum>,
    readonly prices: Price[],
    readonly bump: number
  ) {}

  /**
   * Creates a {@link PriceFeed} instance from the provided args.
   */
  static fromArgs(args: PriceFeedArgs) {
    return new PriceFeed(
      args.admin,
      args.updaters,
      args.priceRatioDenominator,
      args.arbitrumCompressionPercent,
      args.nativeTokenPriceUsd,
      args.prices,
      args.bump
    )
  }

  /**
   * Deserializes the {@link PriceFeed} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [PriceFeed, number] {
    return PriceFeed.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link PriceFeed} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<PriceFeed> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find PriceFeed account at ${address}`)
    }
    return PriceFeed.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '8ahPGPjEbpgGaZx2NV1iG5Shj7TDwvsjkEDcGWjt94TP'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, priceFeedBeet)
  }

  /**
   * Deserializes the {@link PriceFeed} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [PriceFeed, number] {
    return priceFeedBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link PriceFeed} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return priceFeedBeet.serialize({
      accountDiscriminator: priceFeedDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link PriceFeed} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: PriceFeedArgs) {
    const instance = PriceFeed.fromArgs(args)
    return priceFeedBeet.toFixedFromValue({
      accountDiscriminator: priceFeedDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link PriceFeed} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: PriceFeedArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      PriceFeed.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link PriceFeed} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      admin: this.admin.toBase58(),
      updaters: this.updaters,
      priceRatioDenominator: (() => {
        const x = <{ toNumber: () => number }>this.priceRatioDenominator
        if (typeof x.toNumber === 'function') {
          try {
            return x.toNumber()
          } catch (_) {
            return x
          }
        }
        return x
      })(),
      arbitrumCompressionPercent: (() => {
        const x = <{ toNumber: () => number }>this.arbitrumCompressionPercent
        if (typeof x.toNumber === 'function') {
          try {
            return x.toNumber()
          } catch (_) {
            return x
          }
        }
        return x
      })(),
      nativeTokenPriceUsd: this.nativeTokenPriceUsd,
      prices: this.prices,
      bump: this.bump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const priceFeedBeet = new beet.FixableBeetStruct<
  PriceFeed,
  PriceFeedArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['admin', beetSolana.publicKey],
    ['updaters', beet.array(beetSolana.publicKey)],
    ['priceRatioDenominator', beet.u128],
    ['arbitrumCompressionPercent', beet.u128],
    ['nativeTokenPriceUsd', beet.coption(beet.u128)],
    ['prices', beet.array(priceBeet)],
    ['bump', beet.u8],
  ],
  PriceFeed.fromArgs,
  'PriceFeed'
)
