// @layerzerolabs/lz-foundation/dist/index.mjs
import { blake2b as blake2b$1 } from '@noble/hashes/blake2b';
import { sha256 } from '@noble/hashes/sha256';
import { keccak_256 as keccak_256$1, sha3_256 as sha3_256$1 } from '@noble/hashes/sha3';
import * as ed25519 from '@noble/ed25519';
import * as secp256k1 from '@noble/secp256k1';
import { base58, createBase58check, base64 } from '@scure/base';
import { arrayify, isHex, hexlify, padify } from '@layerzerolabs/lz-utilities';
import { bech32 } from 'bech32';
import { chainToChainType, ChainType, Chain } from '@layerzerolabs/lz-definitions';

var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
function keccak_256(message) {
  return keccak_256$1(message);
}
function blake2b(message) {
  return blake2b$1(message);
}
function sha3_256(message) {
  return sha3_256$1(message);
}
function sha2_256(message) {
  return sha256(message);
}

// src/sign/constant.ts
var SignAlgorithm = /* @__PURE__ */ ((SignAlgorithm2) => {
  SignAlgorithm2[SignAlgorithm2["NATIVE"] = 0] = "NATIVE";
  SignAlgorithm2[SignAlgorithm2["SECP256K1"] = 1] = "SECP256K1";
  SignAlgorithm2[SignAlgorithm2["ED25519"] = 2] = "ED25519";
  return SignAlgorithm2;
})(SignAlgorithm || {});

// src/sign/ed25519.ts
var ed25519_exports = {};
__export(ed25519_exports, {
  getPublicKey: () => getPublicKey2,
  signHash: () => signHash
});

// src/sign/internal.ts
function normalizePrivateKey(privateKey) {
  if (typeof privateKey === "string") {
    return privateKey.replace(/^0x/, "");
  } else if (typeof privateKey === "bigint") {
    return privateKey.toString(16).padStart(64, "0");
  } else {
    return privateKey;
  }
}

// src/sign/ed25519.ts
async function signHash(hash, privateKey) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  privateKey = normalizePrivateKey(privateKey);
  const pubKey = await ed25519.getPublicKey(privateKey);
  const signature = await ed25519.sign(hash, privateKey);
  const isValid = await ed25519.verify(signature, hash, pubKey);
  if (!isValid) {
    throw new Error("Invalid signature");
  }
  return signature;
}
async function getPublicKey2(privateKey) {
  privateKey = normalizePrivateKey(privateKey);
  return ed25519.getPublicKey(privateKey);
}

// src/sign/secp256k1.ts
var secp256k1_exports = {};
__export(secp256k1_exports, {
  getPublicKey: () => getPublicKey4,
  signHash: () => signHash2
});
async function signHash2(hash, privateKey) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  privateKey = normalizePrivateKey(privateKey);
  const pubKey = secp256k1.getPublicKey(privateKey, false);
  const [signature, recoveryId] = await secp256k1.sign(hash, privateKey, {
    // recovered - true indicates the recovered bit should be included in the result
    recovered: true,
    // canonical - true makes signatures compatible with libsecp256k1, false makes signatures compatible with openssl
    canonical: true,
    // der - whether the returned signature should be in DER format. If false, it would be in Compact format (32-byte r + 32-byte s)
    der: false,
    // extraEntropy - When true, it would automatically be filled with 32 bytes of cryptographically secure entropy. Strongly recommended to pass true to improve security:
    extraEntropy: void 0
  });
  const isValid = secp256k1.verify(signature, hash, pubKey, { strict: true });
  if (!isValid) {
    throw new Error("Invalid signature");
  }
  return Uint8Array.from([...signature, recoveryId]);
}
async function getPublicKey4(privateKey) {
  privateKey = normalizePrivateKey(privateKey);
  return Promise.resolve(secp256k1.getPublicKey(privateKey, false));
}

// src/sign/index.ts
async function signHash3(hash, privateKey, algorithm) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  const privKey = typeof privateKey === "string" ? privateKey.replace(/^0x/, "") : Buffer.from(privateKey).toString("hex");
  switch (algorithm) {
    case 1 /* SECP256K1 */:
      return signHash2(hash, privKey);
    case 2 /* ED25519 */:
      return signHash(hash, privKey);
    default:
      throw new Error(`Unsupported algorithm: ${algorithm}`);
  }
}

// src/signer.ts
var Secp256k1Signer = class _Secp256k1Signer {
  constructor(privateKey) {
    this.privateKey = privateKey;
  }
  /**
   * Creates an instance of Secp256k1Signer from a private key.
   * @param {Uint8Array} privateKey - The private key.
   * @returns {Secp256k1Signer} The created Secp256k1Signer instance.
   */
  static from(privatekey) {
    return new _Secp256k1Signer(privatekey);
  }
  /**
   * Returns the public key of the signer.
   * @returns {Promise<Uint8Array>} A promise that resolves to the public key.
   */
  async getPublicKey() {
    return secp256k1_exports.getPublicKey(this.privateKey);
  }
  /**
   * Signs a hash.
   * @param {Uint8Array} hash - The hash to sign, must be 32 bytes long.
   * @returns {Promise<Uint8Array>} A promise that resolves to the signature.
   */
  async signHash(hash) {
    return secp256k1_exports.signHash(hash, this.privateKey);
  }
};
var Ed25519Signer = class _Ed25519Signer {
  constructor(privateKey) {
    this.privateKey = privateKey;
  }
  /**
   * Creates an instance of Ed25519Signer from a private key.
   * @param {Uint8Array} privateKey - The private key.
   * @returns {Ed25519Signer} The created Ed25519Signer instance.
   */
  static from(privateKey) {
    return new _Ed25519Signer(privateKey);
  }
  /**
   * Signs a hash.
   * @param {Uint8Array} hash - The hash to sign, must be 32 bytes long.
   * @returns {Promise<Uint8Array>} A promise that resolves to the signature.
   */
  async signHash(hash) {
    return ed25519_exports.signHash(hash, this.privateKey);
  }
  /**
   * Returns the public key of the signer.
   * @returns {Promise<Uint8Array>} A promise that resolves to the public key.
   */
  async getPublicKey() {
    return ed25519_exports.getPublicKey(this.privateKey);
  }
};

// src/base/base58.ts
var base58_exports = {};
__export(base58_exports, {
  decode: () => decode,
  encode: () => encode
});
function encode(value) {
  if (typeof value === "string") {
    return base58.encode(arrayify(value));
  } else {
    return base58.encode(value);
  }
}
function decode(value) {
  return base58.decode(value);
}

// src/base/base58check.ts
var base58check_exports = {};
__export(base58check_exports, {
  decode: () => decode2,
  encode: () => encode2
});
function encode2(value) {
  const base58check = createBase58check(sha256);
  if (typeof value === "string") {
    return base58check.encode(arrayify(value));
  } else {
    return base58check.encode(value);
  }
}
function decode2(value) {
  const base58check = createBase58check(sha256);
  return base58check.decode(value);
}

// src/base/base64.ts
var base64_exports = {};
__export(base64_exports, {
  decode: () => decode3,
  encode: () => encode3
});
function encode3(value) {
  if (typeof value === "string") {
    return base64.encode(arrayify(value));
  } else {
    return base64.encode(value);
  }
}
function decode3(value) {
  return base64.decode(value);
}

// src/address/aptos.ts
var aptos_exports = {};
__export(aptos_exports, {
  Address: () => Address
});
var Address = class _Address {
  /**
   * Creates an instance of Address.
   * @param {Uint8Array | string} value - The address value.
   * @throws {Error} If the address is invalid.
   */
  constructor(value) {
    let data = void 0;
    if (value instanceof Uint8Array) {
      data = value;
    } else if (typeof value === "string") {
      if (isHex(value)) {
        data = arrayify(value);
      }
    }
    if (data === void 0) {
      throw new Error("Invalid address");
    }
    if (data.length > 32) {
      throw new Error("Invalid address length. Expected less or equal than 32 bytes.");
    }
    this.data = data;
  }
  /**
   * Creates an Address instance from a value.
   * @param {Uint8Array | string} value - The address value.
   * @returns {Address} The Address instance.
   */
  static from(value) {
    return new _Address(value);
  }
  /**
   * Inspects the address.
   * @returns {string} The address as a string.
   */
  inspect() {
    return this.toString();
  }
  /**
   * Converts the address to a string.
   * @returns {string} The address as a string.
   */
  toString() {
    return this.hr();
  }
  /**
   * Gets the raw address data.
   * @returns {Uint8Array} The raw address data.
   */
  raw() {
    return this.data;
  }
  /**
   * Gets the human-readable address.
   * @returns {string} The human-readable address.
   */
  hr() {
    return hexlify(this.data);
  }
  /**
   * Gets the address as a hexadecimal string.
   * @returns {Hex} The address as a hexadecimal string.
   */
  hex() {
    return hexlify(this.data);
  }
};

// src/address/evm.ts
var evm_exports = {};
__export(evm_exports, {
  Address: () => Address2,
  toChecksumAddress: () => toChecksumAddress
});
var Address2 = class _Address {
  /**
   * Creates an instance of Address.
   * @param {Uint8Array | string} value - The address value.
   * @throws {Error} If the address is invalid.
   */
  constructor(value) {
    let data = void 0;
    if (value instanceof Uint8Array) {
      data = value;
    } else if (typeof value === "string" && isHex(value)) {
      data = arrayify(value);
    }
    if (data === void 0) {
      throw new Error("Invalid address");
    }
    if (data.length > 20) {
      const prefix = data.slice(0, data.length - 20);
      const isAllZero = prefix.every((byte) => byte === 0);
      if (!isAllZero) {
        throw new Error("Invalid address length. Expected less or equal than 20 bytes.");
      }
    }
    this.data = data;
  }
  /**
   * Creates an Address instance from a value.
   * @param {Uint8Array | string} value - The address value.
   * @returns {Address} The Address instance.
   */
  static from(value) {
    return new _Address(value);
  }
  /**
   * Inspects the address.
   * @returns {string} The address as a string.
   */
  inspect() {
    return this.toString();
  }
  /**
   * Converts the address to a string.
   * @returns {string} The address as a string.
   */
  toString() {
    return this.hr();
  }
  /**
   * Gets the raw address data.
   * @returns {Uint8Array} The raw address data.
   */
  raw() {
    return this.data;
  }
  /**
   * Gets the human-readable address.
   * @returns {string} The human-readable address.
   */
  hr() {
    const value = hexlify(padify(this.data, { size: 20 }));
    return toChecksumAddress(value);
  }
  /**
   * Gets the address as a hexadecimal string.
   * @returns {Hex} The address as a hexadecimal string.
   */
  hex() {
    return hexlify(padify(this.data, { size: 20 }));
  }
};
function toChecksumAddress(address) {
  const lowercasedAddress = address.toLowerCase().replace("0x", "");
  const hash = Buffer.from(keccak_256(lowercasedAddress)).toString("hex");
  let checksumAddress = "0x";
  for (let i = 0; i < lowercasedAddress.length; i++) {
    if (parseInt(hash[i], 16) > 7) {
      checksumAddress += lowercasedAddress[i].toUpperCase();
    } else {
      checksumAddress += lowercasedAddress[i];
    }
  }
  return checksumAddress;
}

// src/address/initia.ts
var initia_exports = {};
__export(initia_exports, {
  Address: () => Address3
});
var Address3 = class _Address {
  /**
   * Creates an instance of Address.
   * @param {Uint8Array | string} value - The address value.
   * @throws {Error} If the address is invalid.
   */
  constructor(value) {
    let data = void 0;
    if (value instanceof Uint8Array) {
      data = value;
    } else {
      if (isHex(value)) {
        data = arrayify(value);
      } else if (this.validateInitiaAddress(value)) {
        const vals = bech32.decode(value);
        data = new Uint8Array(Buffer.from(bech32.fromWords(vals.words)));
      }
    }
    if (data === void 0) {
      throw new Error("Invalid address");
    }
    if (data.length > 32) {
      throw new Error("Invalid address length. Expected less or equal than 32 bytes.");
    }
    this.data = data;
  }
  /**
   * Creates an Address instance from a value.
   * @param {Uint8Array | string} value - The address value.
   * @returns {Address} The Address instance.
   */
  static from(value) {
    return new _Address(value);
  }
  /**
   * Inspects the address.
   * @returns {string} The address as a string.
   */
  inspect() {
    return this.toString();
  }
  /**
   * Converts the address to a string.
   * @returns {string} The address as a string.
   */
  toString() {
    return this.hr();
  }
  /**
   * Gets the raw address data.
   * @returns {Uint8Array} The raw address data.
   */
  raw() {
    return this.data;
  }
  /**
   * Gets the human-readable address.
   * @returns {string} The human-readable address.
   */
  hr() {
    return this.fromHex(this.hex());
  }
  /**
   * Gets the address as a hexadecimal string.
   * @returns {Hex} The address as a hexadecimal string.
   */
  hex() {
    return hexlify(this.data);
  }
  checkPrefixAndLength(prefix, data, length) {
    try {
      const vals = bech32.decode(data);
      return vals.prefix === prefix && data.length == length;
    } catch (e) {
      return false;
    }
  }
  /**
   * Checks if a string is a valid Initia account address.
   * @param data string to check
   */
  validateInitiaAddress(data) {
    return this.checkPrefixAndLength("init", data, 43) || this.checkPrefixAndLength("init", data, 63);
  }
  /**
   * Converts a hex address into an account address
   * @param hexAddress hex address
   */
  fromHex(hexAddress) {
    const hex = hexAddress.replace(/^0x0+|^0x|^0+(?!x)/, "");
    if (hex.length <= 40) {
      return bech32.encode("init", bech32.toWords(Buffer.from(hex.padStart(40, "0"), "hex")));
    } else {
      return bech32.encode("init", bech32.toWords(Buffer.from(hex.padStart(64, "0"), "hex")));
    }
  }
};

// src/address/solana.ts
var solana_exports = {};
__export(solana_exports, {
  Address: () => Address4
});
var Address4 = class _Address {
  /**
   * Creates an instance of Address.
   * @param {Uint8Array | string} value - The address value.
   * @throws {Error} If the address is invalid.
   */
  constructor(value) {
    let data = void 0;
    if (value instanceof Uint8Array) {
      data = value;
    } else if (typeof value === "string") {
      if (isHex(value)) {
        data = arrayify(value);
      } else {
        data = base58_exports.decode(value);
      }
    }
    if (data === void 0) {
      throw new Error("Invalid address");
    }
    if (data.length > 32) {
      throw new Error("Invalid address length. Expected less or equal than 32 bytes.");
    }
    this.data = data;
  }
  /**
   * Creates an Address instance from a value.
   * @param {Uint8Array | string} value - The address value.
   * @returns {Address} The Address instance.
   */
  static from(value) {
    return new _Address(value);
  }
  /**
   * Inspects the address.
   * @returns {string} The address as a string.
   */
  inspect() {
    return this.toString();
  }
  /**
   * Converts the address to a string.
   * @returns {string} The address as a string.
   */
  toString() {
    return this.hr();
  }
  /**
   * Gets the raw address data.
   * @returns {Uint8Array} The raw address data.
   */
  raw() {
    return this.data;
  }
  /**
   * Gets the human-readable address.
   * @returns {string} The human-readable address.
   */
  hr() {
    return base58_exports.encode(this.data);
  }
  /**
   * Gets the address as a hexadecimal string.
   * @returns {Hex} The address as a hexadecimal string.
   */
  hex() {
    return hexlify(this.data);
  }
};

// src/address/ton.ts
var ton_exports = {};
__export(ton_exports, {
  Address: () => Address5
});
var bounceable_tag = 17;
var non_bounceable_tag = 81;
var test_flag = 128;
var Address5 = class _Address {
  /**
   * Creates an instance of Address.
   * @param {Uint8Array | string} value - The address value.
   * @throws {Error} If the address is invalid.
   */
  constructor(value) {
    this.workchain = 0;
    this.testOnly = false;
    this.bounceable = true;
    let data = void 0;
    if (value instanceof Uint8Array) {
      data = value;
    } else if (typeof value === "string") {
      console.log("value", value);
      if (this.isFriendly(value)) {
        const { workchain, hashPart, isTestOnly, isBounceable } = this.parseFriendlyAddress(
          value.replace(/-/g, "+").replace(/_/g, "/")
        );
        this.workchain = workchain;
        data = hashPart;
        this.testOnly = isTestOnly;
        this.bounceable = isBounceable;
      } else if (this.isRaw(value)) {
        const { workchain, hashPart } = this.parseRaw(value);
        this.workchain = workchain;
        data = hashPart;
      } else if (isHex(value)) {
        data = arrayify(value);
      }
    }
    if (data === void 0) {
      throw new Error("Invalid address");
    }
    if (data.length != 32) {
      throw new Error("Invalid hash part length. Expected equal to 32 bytes.");
    }
    this.data = data;
  }
  /**
   * Creates an Address instance from a value.
   * @param {Uint8Array | string} value - The address value.
   * @returns {Address} The Address instance.
   */
  static from(value) {
    return new _Address(value);
  }
  /**
   * Inspects the address.
   * @returns {string} The address as a string.
   */
  inspect() {
    return this.toString();
  }
  /**
   * Converts the address to a string.
   * @returns {string} The address as a string.
   */
  toString() {
    return this.hr();
  }
  /**
   * Gets the raw address data.
   * @returns {Uint8Array} The raw address data.
   */
  raw() {
    return this.data;
  }
  /**
   * Gets the human-readable address.
   * @returns {string} The human-readable address.
   */
  hr() {
    const buffer = this.toStringBuffer();
    return buffer.toString("base64");
  }
  /**
   * Gets the address as a hexadecimal string.
   * @returns {Hex} The address as a hexadecimal string.
   */
  hex() {
    return hexlify(this.data);
  }
  isFriendly(source) {
    if (source.length !== 48) {
      return false;
    }
    if (!/[A-Za-z0-9+/_-]+/.test(source)) {
      return false;
    }
    return true;
  }
  isRaw(source) {
    if (!source.includes(":")) {
      return false;
    }
    const [wc, hash] = source.split(":");
    if (!Number.isInteger(parseFloat(wc))) {
      return false;
    }
    if (!/[a-f0-9]+/.test(hash.toLowerCase())) {
      return false;
    }
    if (hash.length !== 64) {
      return false;
    }
    return true;
  }
  parseRaw(source) {
    return {
      workchain: parseInt(source.split(":")[0]),
      hashPart: new Uint8Array(Buffer.from(source.split(":")[1], "hex"))
    };
  }
  parseFriendlyAddress(src) {
    if (!this.isFriendly(src)) {
      throw new Error("Unknown address type");
    }
    const data = Buffer.isBuffer(src) ? src : Buffer.from(src, "base64");
    if (data.length !== 36) {
      throw new Error("Unknown address type: byte length is not equal to 36");
    }
    const addr = data.subarray(0, 34);
    const crc = data.subarray(34, 36);
    const calcedCrc = this.crc16(addr);
    if (!(calcedCrc[0] === crc[0] && calcedCrc[1] === crc[1])) {
      throw new Error(`Invalid checksum: ${src}`);
    }
    let tag = addr[0];
    let isTestOnly = false;
    let isBounceable = false;
    if (tag & test_flag) {
      isTestOnly = true;
      tag = tag ^ test_flag;
    }
    if (tag !== bounceable_tag && tag !== non_bounceable_tag) throw new Error("Unknown address tag");
    isBounceable = tag === bounceable_tag;
    let workchain = null;
    if (addr[1] === 255) {
      workchain = -1;
    } else {
      workchain = addr[1];
    }
    const hashPart = addr.subarray(2, 34);
    return { isTestOnly, isBounceable, workchain, hashPart };
  }
  crc16(data) {
    const poly = 4129;
    let reg = 0;
    const message = Buffer.alloc(data.length + 2);
    message.set(data);
    for (const byte of message) {
      let mask = 128;
      while (mask > 0) {
        reg <<= 1;
        if (byte & mask) {
          reg += 1;
        }
        mask >>= 1;
        if (reg > 65535) {
          reg &= 65535;
          reg ^= poly;
        }
      }
    }
    return Buffer.from([Math.floor(reg / 256), reg % 256]);
  }
  toStringBuffer() {
    let tag = this.bounceable ? bounceable_tag : non_bounceable_tag;
    if (this.testOnly) {
      tag |= test_flag;
    }
    const addr = Buffer.alloc(34);
    addr[0] = tag;
    addr[1] = this.workchain;
    addr.set(this.data, 2);
    const addressWithChecksum = Buffer.alloc(36);
    addressWithChecksum.set(addr);
    addressWithChecksum.set(this.crc16(addr), 34);
    return addressWithChecksum;
  }
};

// src/address/tron.ts
var tron_exports = {};
__export(tron_exports, {
  Address: () => Address6
});
var Address6 = class _Address {
  /**
   * Creates an instance of Address.
   * @param {Uint8Array | string} value - The address value.
   * @throws {Error} If the address is invalid.
   */
  constructor(value) {
    let data = void 0;
    if (value instanceof Uint8Array) {
      data = value;
    } else if (typeof value === "string") {
      if (isHex(value)) {
        data = arrayify(value);
      } else {
        data = base58check_exports.decode(value);
      }
    }
    if (data === void 0) {
      throw new Error("Invalid address");
    }
    if (data.length != 20) {
      throw new Error("Invalid address length. Expected 20 bytes.");
    }
    this.data = data;
  }
  /**
   * Creates an Address instance from a value.
   * @param {Uint8Array | string} value - The address value.
   * @returns {Address} The Address instance.
   */
  static from(value) {
    return new _Address(value);
  }
  /**
   * Inspects the address.
   * @returns {string} The address as a string.
   */
  inspect() {
    return this.toString();
  }
  /**
   * Converts the address to a string.
   * @returns {string} The address as a string.
   */
  toString() {
    return this.hr();
  }
  /**
   * Gets the raw address data.
   * @returns {Uint8Array} The raw address data.
   */
  raw() {
    return this.data;
  }
  /**
   * Gets the human-readable address.
   * @returns {string} The human-readable address.
   */
  hr() {
    return base58check_exports.encode(this.data);
  }
  /**
   * Gets the address as a hexadecimal string.
   * @returns {Hex} The address as a hexadecimal string.
   */
  hex() {
    return hexlify(this.data);
  }
};
var Address7 = {
  from: (address, chain) => {
    const chainType = chainToChainType(chain);
    switch (chainType) {
      case ChainType.EVM:
        if (chain === Chain.TRON) {
          return Address6.from(address);
        }
        return Address2.from(address);
      case ChainType.APTOS:
      case ChainType.SUI:
        return Address.from(address);
      case ChainType.INITIA:
        return Address3.from(address);
      case ChainType.SOLANA:
        return Address4.from(address);
      case ChainType.TON:
        return Address5.from(address);
      default:
        throw new Error(`Unsupported chain type: ${chainType}`);
    }
  }
};

// src/address/address.ts
var Address8 = Address7;

export { Address8 as Address, Ed25519Signer, Secp256k1Signer, SignAlgorithm, aptos_exports as aptos, base58_exports as base58, base58check_exports as base58check, base64_exports as base64, blake2b, ed25519_exports as ed25519, evm_exports as evm, initia_exports as initia, keccak_256, secp256k1_exports as secp256k1, sha2_256, sha3_256, signHash3 as signHash, solana_exports as solana, ton_exports as ton, tron_exports as tron };
