// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/types/AdminSetConfigParams.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { DstConfig, dstConfigBeet } from './DstConfig'
/**
 * This type is used to derive the {@link AdminSetConfigParams} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link AdminSetConfigParams} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type AdminSetConfigParamsRecord = {
  PriceFeed: { fields: [web3.PublicKey] }
  DefaultMultiplierBps: { fields: [number] }
  DstConfigs: { fields: [DstConfig[]] }
}

/**
 * Union type respresenting the AdminSetConfigParams data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isAdminSetConfigParams*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type AdminSetConfigParams =
  beet.DataEnumKeyAsKind<AdminSetConfigParamsRecord>

export const isAdminSetConfigParamsPriceFeed = (
  x: AdminSetConfigParams
): x is AdminSetConfigParams & { __kind: 'PriceFeed' } =>
  x.__kind === 'PriceFeed'
export const isAdminSetConfigParamsDefaultMultiplierBps = (
  x: AdminSetConfigParams
): x is AdminSetConfigParams & { __kind: 'DefaultMultiplierBps' } =>
  x.__kind === 'DefaultMultiplierBps'
export const isAdminSetConfigParamsDstConfigs = (
  x: AdminSetConfigParams
): x is AdminSetConfigParams & { __kind: 'DstConfigs' } =>
  x.__kind === 'DstConfigs'

/**
 * @category userTypes
 * @category generated
 */
export const adminSetConfigParamsBeet =
  beet.dataEnum<AdminSetConfigParamsRecord>([
    [
      'PriceFeed',
      new beet.BeetArgsStruct<AdminSetConfigParamsRecord['PriceFeed']>(
        [['fields', beet.fixedSizeTuple([beetSolana.publicKey])]],
        'AdminSetConfigParamsRecord["PriceFeed"]'
      ),
    ],
    [
      'DefaultMultiplierBps',
      new beet.BeetArgsStruct<
        AdminSetConfigParamsRecord['DefaultMultiplierBps']
      >(
        [['fields', beet.fixedSizeTuple([beet.u16])]],
        'AdminSetConfigParamsRecord["DefaultMultiplierBps"]'
      ),
    ],
    [
      'DstConfigs',
      new beet.FixableBeetArgsStruct<AdminSetConfigParamsRecord['DstConfigs']>(
        [['fields', beet.tuple([beet.array(dstConfigBeet)])]],
        'AdminSetConfigParamsRecord["DstConfigs"]'
      ),
    ],
  ]) as beet.FixableBeet<AdminSetConfigParams, AdminSetConfigParams>
