// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/types/OwnerSetConfigParams.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
/**
 * This type is used to derive the {@link OwnerSetConfigParams} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link OwnerSetConfigParams} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type OwnerSetConfigParamsRecord = {
  Admins: { fields: [web3.PublicKey[]] }
  Executors: { fields: [web3.PublicKey[]] }
  Msglibs: { fields: [web3.PublicKey[]] }
  Owner: { fields: [web3.PublicKey] }
  Paused: { fields: [boolean] }
  Allowlist: { fields: [web3.PublicKey[]] }
  Denylist: { fields: [web3.PublicKey[]] }
}

/**
 * Union type respresenting the OwnerSetConfigParams data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isOwnerSetConfigParams*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type OwnerSetConfigParams =
  beet.DataEnumKeyAsKind<OwnerSetConfigParamsRecord>

export const isOwnerSetConfigParamsAdmins = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Admins' } => x.__kind === 'Admins'
export const isOwnerSetConfigParamsExecutors = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Executors' } =>
  x.__kind === 'Executors'
export const isOwnerSetConfigParamsMsglibs = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Msglibs' } => x.__kind === 'Msglibs'
export const isOwnerSetConfigParamsOwner = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Owner' } => x.__kind === 'Owner'
export const isOwnerSetConfigParamsPaused = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Paused' } => x.__kind === 'Paused'
export const isOwnerSetConfigParamsAllowlist = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Allowlist' } =>
  x.__kind === 'Allowlist'
export const isOwnerSetConfigParamsDenylist = (
  x: OwnerSetConfigParams
): x is OwnerSetConfigParams & { __kind: 'Denylist' } => x.__kind === 'Denylist'

/**
 * @category userTypes
 * @category generated
 */
export const ownerSetConfigParamsBeet =
  beet.dataEnum<OwnerSetConfigParamsRecord>([
    [
      'Admins',
      new beet.FixableBeetArgsStruct<OwnerSetConfigParamsRecord['Admins']>(
        [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
        'OwnerSetConfigParamsRecord["Admins"]'
      ),
    ],
    [
      'Executors',
      new beet.FixableBeetArgsStruct<OwnerSetConfigParamsRecord['Executors']>(
        [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
        'OwnerSetConfigParamsRecord["Executors"]'
      ),
    ],
    [
      'Msglibs',
      new beet.FixableBeetArgsStruct<OwnerSetConfigParamsRecord['Msglibs']>(
        [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
        'OwnerSetConfigParamsRecord["Msglibs"]'
      ),
    ],
    [
      'Owner',
      new beet.BeetArgsStruct<OwnerSetConfigParamsRecord['Owner']>(
        [['fields', beet.fixedSizeTuple([beetSolana.publicKey])]],
        'OwnerSetConfigParamsRecord["Owner"]'
      ),
    ],
    [
      'Paused',
      new beet.BeetArgsStruct<OwnerSetConfigParamsRecord['Paused']>(
        [['fields', beet.fixedSizeTuple([beet.bool])]],
        'OwnerSetConfigParamsRecord["Paused"]'
      ),
    ],
    [
      'Allowlist',
      new beet.FixableBeetArgsStruct<OwnerSetConfigParamsRecord['Allowlist']>(
        [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
        'OwnerSetConfigParamsRecord["Allowlist"]'
      ),
    ],
    [
      'Denylist',
      new beet.FixableBeetArgsStruct<OwnerSetConfigParamsRecord['Denylist']>(
        [['fields', beet.tuple([beet.array(beetSolana.publicKey)])]],
        'OwnerSetConfigParamsRecord["Denylist"]'
      ),
    ],
  ]) as beet.FixableBeet<OwnerSetConfigParams, OwnerSetConfigParams>
