// @layerzerolabs/lz-serdes/dist/index.cjs
'use strict';

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = __getOwnPropDesc(target, key) ;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (decorator(target, key, result) ) || result;
  if (result) __defProp(target, key, result);
  return result;
};

// src/consts.ts
var MAX_U8_NUMBER = 255;
var MAX_U16_NUMBER = 65535;
var MAX_U32_NUMBER = 4294967295;
var MAX_U64_BIG_INT = 18446744073709551615n;
var MAX_U128_BIG_INT = 340282366920938463463374607431768211455n;
var MAX_U256_BIG_INT = 115792089237316195423570985008687907853269984665640564039457584007913129639935n;

// src/serializer.ts
var Serializer = class {
  /**
   * Creates an instance of Serializer.
   *
   * @param {boolean} [littleEndian=true] - Whether to use little-endian byte order.
   */
  constructor(littleEndian = true) {
    this.littleEndian = littleEndian;
    this.buffer = new ArrayBuffer(64);
    this.offset = 0;
  }
  /**
   * Ensures that the buffer has enough space to handle the specified number of bytes.
   * If the buffer is not large enough, it will be resized to accommodate the additional bytes.
   *
   * @param {number} bytes - The number of bytes to ensure space for.
   */
  ensureBufferWillHandleSize(bytes) {
    while (this.buffer.byteLength < this.offset + bytes) {
      const newBuffer = new ArrayBuffer(this.buffer.byteLength * 2);
      new Uint8Array(newBuffer).set(new Uint8Array(this.buffer));
      this.buffer = newBuffer;
    }
  }
  /**
   * Serializes the given byte values into the buffer.
   *
   * @param {Bytes} values - The byte values to serialize.
   */
  serialize(values) {
    this.ensureBufferWillHandleSize(values.length);
    new Uint8Array(this.buffer, this.offset).set(values);
    this.offset += values.length;
  }
  /**
   * Serializes a value using the specified DataView function.
   *
   * @param {(byteOffset: number, value: number, littleEndian?: boolean) => void} fn - The DataView function to use for serialization.
   * @param {number} bytesLength - The length of the bytes to serialize.
   * @param {number} value - The value to serialize.
   */
  serializeWithFunction(fn, bytesLength, value) {
    this.ensureBufferWillHandleSize(bytesLength);
    const dv = new DataView(this.buffer, this.offset);
    fn.apply(dv, [0, value, this.littleEndian]);
    this.offset += bytesLength;
  }
  /**
   * Serializes a string. UTF8 string is supported. Serializes the string's bytes length "l" first,
   * and then serializes "l" bytes of the string content.
   *
   * layout for "string": string_length | string_content. string_length is the bytes length of
   * the string that is uleb128/ubeb128 encoded. string_length is a u32 integer.
   *
   * @example
   * ```ts
   * const serializer = new Serializer();
   * serializer.serializeStr("çå∞≠¢õß∂ƒ∫");
   * assert(serializer.getBytes() === new Uint8Array([24, 0xc3, 0xa7, 0xc3, 0xa5, 0xe2, 0x88, 0x9e,
   * 0xe2, 0x89, 0xa0, 0xc2, 0xa2, 0xc3, 0xb5, 0xc3, 0x9f, 0xe2, 0x88, 0x82, 0xc6, 0x92, 0xe2, 0x88, 0xab]));
   * ```
   */
  serializeStr(value) {
    const textEncoder = new TextEncoder();
    this.serializeBytes(textEncoder.encode(value));
    return this;
  }
  /**
   * Serializes an array of bytes.
   *
   * layout for "bytes": bytes_length | bytes. bytes_length is the length of the bytes array that is
   * uleb128/ubeb128 encoded. bytes_length is a u32 integer.
   */
  serializeBytes(value) {
    if (this.littleEndian) {
      this.serializeU32AsUleb128(value.length);
    } else {
      this.serializeU32AsUbeb128(value.length);
    }
    this.serialize(value);
    return this;
  }
  /**
   * Serializes an array of bytes with known length. Therefore length doesn't need to be
   * serialized to help deserialization.  When deserializing, the number of
   * bytes to deserialize needs to be passed in.
   */
  serializeFixedBytes(value) {
    this.serialize(value);
    return this;
  }
  /**
   * Serializes a boolean value.
   *
   * layout for "boolean": One byte. "0x01" for True and "0x00" for False.
   */
  serializeBool(value) {
    if (typeof value !== "boolean") {
      throw new Error("Value needs to be a boolean");
    }
    const byteValue = value ? 1 : 0;
    this.serialize(new Uint8Array([byteValue]));
    return this;
  }
  serializeU8(value) {
    this.serialize(new Uint8Array([value]));
    return this;
  }
  serializeU16(value) {
    this.serializeWithFunction(DataView.prototype.setUint16, 2, value);
    return this;
  }
  serializeU32(value) {
    this.serializeWithFunction(DataView.prototype.setUint32, 4, value);
    return this;
  }
  serializeU64(value) {
    const low = BigInt(value.toString()) & BigInt(MAX_U32_NUMBER);
    const high = BigInt(value.toString()) >> BigInt(32);
    const array = this.littleEndian ? [low, high] : [high, low];
    array.forEach((x) => {
      this.serializeU32(Number(x));
    });
    return this;
  }
  serializeU128(value) {
    const low = BigInt(value.toString()) & MAX_U64_BIG_INT;
    const high = BigInt(value.toString()) >> BigInt(64);
    const array = this.littleEndian ? [low, high] : [high, low];
    array.forEach((x) => {
      this.serializeU64(x);
    });
    return this;
  }
  serializeU256(value) {
    const low = BigInt(value.toString()) & MAX_U128_BIG_INT;
    const high = BigInt(value.toString()) >> BigInt(128);
    const array = this.littleEndian ? [low, high] : [high, low];
    array.forEach((x) => {
      this.serializeU128(x);
    });
    return this;
  }
  serializeU32AsUleb128(val) {
    let value = val;
    const valueArray = [];
    while (value >>> 7 !== 0) {
      valueArray.push(value & 127 | 128);
      value >>>= 7;
    }
    valueArray.push(value);
    this.serialize(new Uint8Array(valueArray));
    return this;
  }
  serializeU32AsUbeb128(val) {
    let value = val;
    const valueArray = [];
    valueArray.unshift(value & 127);
    while (value >>> 7 !== 0) {
      value >>>= 7;
      const byte = value & 127 | 128;
      valueArray.unshift(byte);
    }
    this.serialize(new Uint8Array(valueArray));
    return this;
  }
  /**
   * Returns the buffered bytes
   *
   * @returns {Bytes} The buffered bytes.
   */
  getBytes() {
    return new Uint8Array(this.buffer).slice(0, this.offset);
  }
};
__decorateClass([
  checkNumberRange(0, MAX_U8_NUMBER)
], Serializer.prototype, "serializeU8");
__decorateClass([
  checkNumberRange(0, MAX_U16_NUMBER)
], Serializer.prototype, "serializeU16");
__decorateClass([
  checkNumberRange(0, MAX_U32_NUMBER)
], Serializer.prototype, "serializeU32");
__decorateClass([
  checkNumberRange(BigInt(0), MAX_U64_BIG_INT)
], Serializer.prototype, "serializeU64");
__decorateClass([
  checkNumberRange(BigInt(0), MAX_U128_BIG_INT)
], Serializer.prototype, "serializeU128");
__decorateClass([
  checkNumberRange(BigInt(0), MAX_U256_BIG_INT)
], Serializer.prototype, "serializeU256");
__decorateClass([
  checkNumberRange(0, MAX_U32_NUMBER)
], Serializer.prototype, "serializeU32AsUleb128");
__decorateClass([
  checkNumberRange(0, MAX_U32_NUMBER)
], Serializer.prototype, "serializeU32AsUbeb128");
function checkNumberRange(minValue, maxValue, message) {
  return (target, propertyKey, descriptor) => {
    const childFunction = descriptor.value;
    descriptor.value = function deco(value) {
      const valueBigInt = BigInt(value.toString());
      if (valueBigInt > BigInt(maxValue.toString()) || valueBigInt < BigInt(minValue.toString())) {
        throw new Error("Value is out of range");
      }
      return childFunction.apply(this, [value]);
    };
    return descriptor;
  };
}

// src/deserializer.ts
var Deserializer = class {
  /**
   * Creates an instance of Deserializer.
   *
   * @param {Bytes} data - The data to deserialize.
   * @param {boolean} [littleEndian=true] - Whether to use little-endian byte order.
   */
  constructor(data, littleEndian = true) {
    this.littleEndian = littleEndian;
    this.buffer = new ArrayBuffer(data.length);
    new Uint8Array(this.buffer).set(data, 0);
    this.offset = 0;
  }
  /**
   * Reads a specified number of bytes from the buffer.
   *
   * @param {number} length - The number of bytes to read.
   * @returns {ArrayBuffer} The read bytes.
   * @throws {Error} If the end of the buffer is reached.
   */
  read(length) {
    if (this.offset + length > this.buffer.byteLength) {
      throw new Error("Reached to the end of buffer");
    }
    const bytes = this.buffer.slice(this.offset, this.offset + length);
    this.offset += length;
    return bytes;
  }
  /**
   * Returns the number of remaining bytes in the buffer.
   *
   * @returns {number} The number of remaining bytes.
   */
  remaining() {
    return this.buffer.byteLength - this.offset;
  }
  /**
   * Rewinds the buffer by a specified number of bytes.
   *
   * @param {number} length - The number of bytes to rewind.
   */
  rewind(length) {
    this.offset -= length;
  }
  /**
   * Deserializes a string. UTF8 string is supported. Reads the string's bytes length "l" first,
   * and then reads "l" bytes of content. Decodes the byte array into a string.
   *
   * layout for "string": string_length | string_content. string_length is the bytes length of
   * the string that is uleb128/ubeb128 encoded. string_length is a u32 integer.
   *
   * @example
   * ```ts
   * const deserializer = new Deserializer(new Uint8Array([24, 0xc3, 0xa7, 0xc3, 0xa5, 0xe2, 0x88, 0x9e,
   * 0xe2, 0x89, 0xa0, 0xc2, 0xa2, 0xc3, 0xb5, 0xc3, 0x9f, 0xe2, 0x88, 0x82, 0xc6, 0x92, 0xe2, 0x88, 0xab]), true);
   * assert(deserializer.deserializeStr() === "çå∞≠¢õß∂ƒ∫");
   * ```
   * @returns {string} The deserialized string.
   */
  deserializeStr() {
    const value = this.deserializeBytes();
    const textDecoder = new TextDecoder();
    return textDecoder.decode(value);
  }
  /**
   * Deserializes an array of bytes.
   *
   * layout for "bytes": bytes_length | bytes. bytes_length is the length of the bytes array that is
   * uleb128/ubeb128 encoded. bytes_length is a u32 integer.
   *
   * @returns {Bytes} The deserialized bytes.
   */
  deserializeBytes() {
    const len = this.littleEndian ? this.deserializeUleb128AsU32() : this.deserializeUbeb128AsU32();
    return new Uint8Array(this.read(len));
  }
  /**
   * Deserializes an array of bytes. The number of bytes to read is already known.
   *
   * @param {number} len - The number of bytes to read.
   * @returns {Bytes} The deserialized bytes.
   */
  deserializeFixedBytes(len) {
    return new Uint8Array(this.read(len));
  }
  /**
   * Deserializes a boolean value.
   *
   * layout for "boolean": One byte. "0x01" for True and "0x00" for False.
   *
   * @returns {boolean} The deserialized boolean value.
   * @throws {Error} If the boolean value is invalid.
   */
  deserializeBool() {
    const bool = new Uint8Array(this.read(1))[0];
    if (bool !== 1 && bool !== 0) {
      throw new Error("Invalid boolean value");
    }
    return bool === 1;
  }
  /**
   * Deserializes a uint8 number.
   *
   * layout for "uint8": One byte. Binary format in little-endian representation.
   *
   * @returns {Uint8} The deserialized uint8 number.
   */
  deserializeU8() {
    return new DataView(this.read(1)).getUint8(0);
  }
  /**
   * Deserializes a uint16 number.
   *
   * layout for "uint16": Two bytes. Binary format in little-endian representation.
   * @example
   * ```ts
   * const deserializer = new Deserializer(new Uint8Array([0x34, 0x12]), true);
   * assert(deserializer.deserializeU16() === 4660);
   * const deserializer = new Deserializer(new Uint8Array([0x12, 0x34]), false);
   * assert(deserializer.deserializeU16() === 4660);
   * ```
   * @returns {Uint16} The deserialized uint16 number.
   */
  deserializeU16() {
    return new DataView(this.read(2)).getUint16(0, this.littleEndian);
  }
  /**
   * Deserializes a uint32 number.
   *
   * layout for "uint32": Four bytes. Binary format in little-endian representation.
   * @example
   * ```ts
   * const deserializer = new Deserializer(new Uint8Array([0x78, 0x56, 0x34, 0x12]), true);
   * assert(deserializer.deserializeU32() === 305419896);
   * const deserializer = new Deserializer(new Uint8Array([0x12, 0x34, 0x56, 0x78]), false);
   * assert(deserializer.deserializeU32() === 305419896);
   * ```
   * @returns {Uint32} The deserialized uint32 number.
   */
  deserializeU32() {
    return new DataView(this.read(4)).getUint32(0, this.littleEndian);
  }
  /**
   * Deserializes a uint64 number.
   *
   * layout for "uint64": Eight bytes. Binary format in little-endian representation.
   * @example
   * ```ts
   * const deserializer = new Deserializer(new Uint8Array([0x00, 0xEF, 0xCD, 0xAB, 0x78, 0x56, 0x34, 0x12]), true);
   * assert(deserializer.deserializeU64() === 1311768467750121216);
   * const deserializer = new Deserializer(new Uint8Array([0x12, 0x34, 0x56, 0x78, 0xAB, 0xCD, 0xEF, 0x00]), false);
   * assert(deserializer.deserializeU64() === 1311768467750121216);
   * ```
   * @returns {Uint64} The deserialized uint64 number.
   */
  deserializeU64() {
    const low = this.deserializeU32();
    const high = this.deserializeU32();
    const values = this.littleEndian ? [low, high] : [high, low];
    return BigInt(BigInt(values[1]) << BigInt(32) | BigInt(values[0]));
  }
  /**
   * Deserializes a uint128 number.
   *
   * layout for "uint128": Sixteen bytes. Binary format in little-endian representation.
   *
   * @returns {Uint128} The deserialized uint128 number.
   */
  deserializeU128() {
    const low = this.deserializeU64();
    const high = this.deserializeU64();
    const values = this.littleEndian ? [low, high] : [high, low];
    return BigInt(values[1] << BigInt(64) | values[0]);
  }
  /**
   * Deserializes a uint256 number.
   *
   * layout for "uint256": Thirty-two bytes. Binary format in little-endian representation.
   *
   * @returns {Uint256} The deserialized uint256 number.
   */
  deserializeU256() {
    const low = this.deserializeU128();
    const high = this.deserializeU128();
    const values = this.littleEndian ? [low, high] : [high, low];
    return BigInt(values[1] << BigInt(128) | values[0]);
  }
  /**
   * Deserializes a uleb128 encoded uint32 number.
   *
   * use uleb128 encoding in two cases: (1) lengths of variable-length sequences and (2) tags of enum values
   *
   * @returns {Uint32} The deserialized uleb128 encoded uint32 number.
   * @throws {Error} If there is an overflow while parsing the value.
   */
  deserializeUleb128AsU32() {
    let value = BigInt(0);
    let shift = 0;
    while (value < MAX_U32_NUMBER) {
      const byte = this.deserializeU8();
      value |= BigInt(byte & 127) << BigInt(shift);
      if ((byte & 128) === 0) {
        break;
      }
      shift += 7;
    }
    if (value > MAX_U32_NUMBER) {
      throw new Error("Overflow while parsing uleb128-encoded uint32 value");
    }
    return Number(value);
  }
  /**
   * Deserializes a unsigned Variable-Length Big-Endian Encoding encoded uint32 number.
   *
   * use ubeb128 encoding in two cases: (1) lengths of variable-length sequences and (2) tags of enum values
   *
   * @returns {Uint32} The deserialized ubeb128 encoded uint32 number.
   * @throws {Error} If there is an overflow while parsing the value.
   */
  deserializeUbeb128AsU32() {
    let value = BigInt(0);
    while (value < MAX_U32_NUMBER) {
      const byte = this.deserializeU8();
      value |= BigInt(byte & 127);
      if ((byte & 128) === 0) {
        break;
      }
      value = value << BigInt(7);
    }
    if (value > MAX_U32_NUMBER) {
      throw new Error("Overflow while parsing ubeb128-encoded uint32 value");
    }
    return Number(value);
  }
};

// src/helper.ts
function serializeVector(value, serializer) {
  serializer.serializeU32AsUleb128(value.length);
  value.forEach((item) => {
    item.serialize(serializer);
  });
}
function serializeVectorWithFunc(value, func) {
  const serializer = new Serializer();
  serializer.serializeU32AsUleb128(value.length);
  const f = serializer[func];
  if (typeof f !== "function") {
    throw new Error(`The function '${func}' does not exist on the serializer.`);
  }
  value.forEach((item) => {
    f.call(serializer, item);
  });
  return serializer.getBytes();
}
function deserializeVector(deserializer, cls) {
  const length = deserializer.deserializeUleb128AsU32();
  const list = [];
  for (let i = 0; i < length; i += 1) {
    list.push(cls.deserialize(deserializer));
  }
  return list;
}
function bcsToBytes(value) {
  const serializer = new Serializer();
  value.serialize(serializer);
  return serializer.getBytes();
}
function bcsSerializeUint64(value) {
  const serializer = new Serializer();
  serializer.serializeU64(value);
  return serializer.getBytes();
}
function bcsSerializeU8(value) {
  const serializer = new Serializer();
  serializer.serializeU8(value);
  return serializer.getBytes();
}
function bcsSerializeU16(value) {
  const serializer = new Serializer();
  serializer.serializeU16(value);
  return serializer.getBytes();
}
function bcsSerializeU32(value) {
  const serializer = new Serializer();
  serializer.serializeU32(value);
  return serializer.getBytes();
}
function bcsSerializeU128(value) {
  const serializer = new Serializer();
  serializer.serializeU128(value);
  return serializer.getBytes();
}
function bcsSerializeU256(value) {
  const serializer = new Serializer();
  serializer.serializeU256(value);
  return serializer.getBytes();
}
function bcsSerializeBool(value) {
  const serializer = new Serializer();
  serializer.serializeBool(value);
  return serializer.getBytes();
}
function bcsSerializeStr(value) {
  const serializer = new Serializer();
  serializer.serializeStr(value);
  return serializer.getBytes();
}
function bcsSerializeBytes(value) {
  const serializer = new Serializer();
  serializer.serializeBytes(value);
  return serializer.getBytes();
}
function bcsSerializeFixedBytes(value) {
  const serializer = new Serializer();
  serializer.serializeFixedBytes(value);
  return serializer.getBytes();
}

exports.Deserializer = Deserializer;
exports.Serializer = Serializer;
exports.bcsSerializeBool = bcsSerializeBool;
exports.bcsSerializeBytes = bcsSerializeBytes;
exports.bcsSerializeFixedBytes = bcsSerializeFixedBytes;
exports.bcsSerializeStr = bcsSerializeStr;
exports.bcsSerializeU128 = bcsSerializeU128;
exports.bcsSerializeU16 = bcsSerializeU16;
exports.bcsSerializeU256 = bcsSerializeU256;
exports.bcsSerializeU32 = bcsSerializeU32;
exports.bcsSerializeU8 = bcsSerializeU8;
exports.bcsSerializeUint64 = bcsSerializeUint64;
exports.bcsToBytes = bcsToBytes;
exports.deserializeVector = deserializeVector;
exports.serializeVector = serializeVector;
exports.serializeVectorWithFunc = serializeVectorWithFunc;
