// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/instructions/initMessageLib.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitMessageLibParams,
  initMessageLibParamsBeet,
} from '../types/InitMessageLibParams'

/**
 * @category Instructions
 * @category InitMessageLib
 * @category generated
 */
export type InitMessageLibInstructionArgs = {
  params: InitMessageLibParams
}
/**
 * @category Instructions
 * @category InitMessageLib
 * @category generated
 */
export const initMessageLibStruct = new beet.BeetArgsStruct<
  InitMessageLibInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initMessageLibParamsBeet],
  ],
  'InitMessageLibInstructionArgs'
)
/**
 * Accounts required by the _initMessageLib_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [_writable_] messageLib
 * @category Instructions
 * @category InitMessageLib
 * @category generated
 */
export type InitMessageLibInstructionAccounts = {
  payer: web3.PublicKey
  messageLib: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initMessageLibInstructionDiscriminator = [
  184, 93, 59, 206, 98, 238, 225, 54,
]

/**
 * Creates a _InitMessageLib_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitMessageLib
 * @category generated
 */
export function createInitMessageLibInstruction(
  accounts: InitMessageLibInstructionAccounts,
  args: InitMessageLibInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initMessageLibStruct.serialize({
    instructionDiscriminator: initMessageLibInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitMessageLib_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitMessageLib
 * @category generated
 */
export function createInitMessageLibInstructionAccounts(
  accounts: InitMessageLibInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
