// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/instructions/initConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitConfigParams,
  initConfigParamsBeet,
} from '../types/InitConfigParams'

/**
 * @category Instructions
 * @category InitConfig
 * @category generated
 */
export type InitConfigInstructionArgs = {
  params: InitConfigParams
}
/**
 * @category Instructions
 * @category InitConfig
 * @category generated
 */
export const initConfigStruct = new beet.BeetArgsStruct<
  InitConfigInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initConfigParamsBeet],
  ],
  'InitConfigInstructionArgs'
)
/**
 * Accounts required by the _initConfig_ instruction
 *
 * @property [**signer**] endpoint
 * @property [_writable_, **signer**] payer
 * @property [] messageLib
 * @property [_writable_] sendConfig
 * @property [_writable_] receiveConfig
 * @category Instructions
 * @category InitConfig
 * @category generated
 */
export type InitConfigInstructionAccounts = {
  endpoint: web3.PublicKey
  payer: web3.PublicKey
  messageLib: web3.PublicKey
  sendConfig: web3.PublicKey
  receiveConfig: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initConfigInstructionDiscriminator = [
  23, 235, 115, 232, 168, 96, 1, 231,
]

/**
 * Creates a _InitConfig_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitConfig
 * @category generated
 */
export function createInitConfigInstruction(
  accounts: InitConfigInstructionAccounts,
  args: InitConfigInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initConfigStruct.serialize({
    instructionDiscriminator: initConfigInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitConfig_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitConfig
 * @category generated
 */
export function createInitConfigInstructionAccounts(
  accounts: InitConfigInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
