// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/instructions/setFee.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { SetFeeParams, setFeeParamsBeet } from '../types/SetFeeParams'

/**
 * @category Instructions
 * @category SetFee
 * @category generated
 */
export type SetFeeInstructionArgs = {
  params: SetFeeParams
}
/**
 * @category Instructions
 * @category SetFee
 * @category generated
 */
export const setFeeStruct = new beet.BeetArgsStruct<
  SetFeeInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setFeeParamsBeet],
  ],
  'SetFeeInstructionArgs'
)
/**
 * Accounts required by the _setFee_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] messageLib
 * @category Instructions
 * @category SetFee
 * @category generated
 */
export type SetFeeInstructionAccounts = {
  admin: web3.PublicKey
  messageLib: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setFeeInstructionDiscriminator = [
  18, 154, 24, 18, 237, 214, 19, 80,
]

/**
 * Creates a _SetFee_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetFee
 * @category generated
 */
export function createSetFeeInstruction(
  accounts: SetFeeInstructionAccounts,
  args: SetFeeInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setFeeStruct.serialize({
    instructionDiscriminator: setFeeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetFee_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetFee
 * @category generated
 */
export function createSetFeeInstructionAccounts(
  accounts: SetFeeInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
