// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/instructions/setWlCaller.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetWlCallerParams,
  setWlCallerParamsBeet,
} from '../types/SetWlCallerParams'

/**
 * @category Instructions
 * @category SetWlCaller
 * @category generated
 */
export type SetWlCallerInstructionArgs = {
  params: SetWlCallerParams
}
/**
 * @category Instructions
 * @category SetWlCaller
 * @category generated
 */
export const setWlCallerStruct = new beet.BeetArgsStruct<
  SetWlCallerInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setWlCallerParamsBeet],
  ],
  'SetWlCallerInstructionArgs'
)
/**
 * Accounts required by the _setWlCaller_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] messageLib
 * @category Instructions
 * @category SetWlCaller
 * @category generated
 */
export type SetWlCallerInstructionAccounts = {
  admin: web3.PublicKey
  messageLib: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setWlCallerInstructionDiscriminator = [
  153, 41, 240, 37, 126, 107, 14, 253,
]

/**
 * Creates a _SetWlCaller_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetWlCaller
 * @category generated
 */
export function createSetWlCallerInstruction(
  accounts: SetWlCallerInstructionAccounts,
  args: SetWlCallerInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setWlCallerStruct.serialize({
    instructionDiscriminator: setWlCallerInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetWlCaller_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetWlCaller
 * @category generated
 */
export function createSetWlCallerInstructionAccounts(
  accounts: SetWlCallerInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
