// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/instructions/validatePacket.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  ValidatePacketParams,
  validatePacketParamsBeet,
} from '../types/ValidatePacketParams'

/**
 * @category Instructions
 * @category ValidatePacket
 * @category generated
 */
export type ValidatePacketInstructionArgs = {
  params: ValidatePacketParams
}
/**
 * @category Instructions
 * @category ValidatePacket
 * @category generated
 */
export const validatePacketStruct = new beet.FixableBeetArgsStruct<
  ValidatePacketInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', validatePacketParamsBeet],
  ],
  'ValidatePacketInstructionArgs'
)
/**
 * Accounts required by the _validatePacket_ instruction
 *
 * @property [**signer**] payer
 * @property [] receiveLibrary
 * @category Instructions
 * @category ValidatePacket
 * @category generated
 */
export type ValidatePacketInstructionAccounts = {
  payer: web3.PublicKey
  receiveLibrary: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const validatePacketInstructionDiscriminator = [
  34, 146, 107, 76, 157, 98, 105, 211,
]

/**
 * Creates a _ValidatePacket_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category ValidatePacket
 * @category generated
 */
export function createValidatePacketInstruction(
  accounts: ValidatePacketInstructionAccounts,
  args: ValidatePacketInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = validatePacketStruct.serialize({
    instructionDiscriminator: validatePacketInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.receiveLibrary,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _ValidatePacket_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category ValidatePacket
 * @category generated
 */
export function createValidatePacketInstructionAccounts(
  accounts: ValidatePacketInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.receiveLibrary,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
