// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/instructions/withdrawFees.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  WithdrawFeesParams,
  withdrawFeesParamsBeet,
} from '../types/WithdrawFeesParams'

/**
 * @category Instructions
 * @category WithdrawFees
 * @category generated
 */
export type WithdrawFeesInstructionArgs = {
  params: WithdrawFeesParams
}
/**
 * @category Instructions
 * @category WithdrawFees
 * @category generated
 */
export const withdrawFeesStruct = new beet.BeetArgsStruct<
  WithdrawFeesInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', withdrawFeesParamsBeet],
  ],
  'WithdrawFeesInstructionArgs'
)
/**
 * Accounts required by the _withdrawFees_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] messageLib
 * @property [_writable_] receiver
 * @category Instructions
 * @category WithdrawFees
 * @category generated
 */
export type WithdrawFeesInstructionAccounts = {
  admin: web3.PublicKey
  messageLib: web3.PublicKey
  receiver: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const withdrawFeesInstructionDiscriminator = [
  198, 212, 171, 109, 144, 215, 174, 89,
]

/**
 * Creates a _WithdrawFees_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category WithdrawFees
 * @category generated
 */
export function createWithdrawFeesInstruction(
  accounts: WithdrawFeesInstructionAccounts,
  args: WithdrawFeesInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = withdrawFeesStruct.serialize({
    instructionDiscriminator: withdrawFeesInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiver,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _WithdrawFees_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category WithdrawFees
 * @category generated
 */
export function createWithdrawFeesInstructionAccounts(
  accounts: WithdrawFeesInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.messageLib,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiver,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
