// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/errors/index.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * InvalidSignatureLen: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidSignatureLenError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'InvalidSignatureLen'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidSignatureLenError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new InvalidSignatureLenError())
createErrorFromNameLookup.set(
  'InvalidSignatureLen',
  () => new InvalidSignatureLenError()
)

/**
 * NotAdmin: ''
 *
 * @category Errors
 * @category generated
 */
export class NotAdminError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'NotAdmin'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, NotAdminError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new NotAdminError())
createErrorFromNameLookup.set('NotAdmin', () => new NotAdminError())

/**
 * MsgLibNotAllowed: ''
 *
 * @category Errors
 * @category generated
 */
export class MsgLibNotAllowedError extends Error {
  readonly code: number = 0x1772
  readonly name: string = 'MsgLibNotAllowed'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, MsgLibNotAllowedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1772, () => new MsgLibNotAllowedError())
createErrorFromNameLookup.set(
  'MsgLibNotAllowed',
  () => new MsgLibNotAllowedError()
)

/**
 * InvalidQuorum: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidQuorumError extends Error {
  readonly code: number = 0x1773
  readonly name: string = 'InvalidQuorum'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidQuorumError)
    }
  }
}

createErrorFromCodeLookup.set(0x1773, () => new InvalidQuorumError())
createErrorFromNameLookup.set('InvalidQuorum', () => new InvalidQuorumError())

/**
 * InvalidSignersLen: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidSignersLenError extends Error {
  readonly code: number = 0x1774
  readonly name: string = 'InvalidSignersLen'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidSignersLenError)
    }
  }
}

createErrorFromCodeLookup.set(0x1774, () => new InvalidSignersLenError())
createErrorFromNameLookup.set(
  'InvalidSignersLen',
  () => new InvalidSignersLenError()
)

/**
 * UniqueOwners: ''
 *
 * @category Errors
 * @category generated
 */
export class UniqueOwnersError extends Error {
  readonly code: number = 0x1775
  readonly name: string = 'UniqueOwners'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, UniqueOwnersError)
    }
  }
}

createErrorFromCodeLookup.set(0x1775, () => new UniqueOwnersError())
createErrorFromNameLookup.set('UniqueOwners', () => new UniqueOwnersError())

/**
 * SignatureError: ''
 *
 * @category Errors
 * @category generated
 */
export class SignatureErrorError extends Error {
  readonly code: number = 0x1776
  readonly name: string = 'SignatureError'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, SignatureErrorError)
    }
  }
}

createErrorFromCodeLookup.set(0x1776, () => new SignatureErrorError())
createErrorFromNameLookup.set('SignatureError', () => new SignatureErrorError())

/**
 * SignerNotInCommittee: ''
 *
 * @category Errors
 * @category generated
 */
export class SignerNotInCommitteeError extends Error {
  readonly code: number = 0x1777
  readonly name: string = 'SignerNotInCommittee'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, SignerNotInCommitteeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1777, () => new SignerNotInCommitteeError())
createErrorFromNameLookup.set(
  'SignerNotInCommittee',
  () => new SignerNotInCommitteeError()
)

/**
 * TooManyAdmins: ''
 *
 * @category Errors
 * @category generated
 */
export class TooManyAdminsError extends Error {
  readonly code: number = 0x1778
  readonly name: string = 'TooManyAdmins'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, TooManyAdminsError)
    }
  }
}

createErrorFromCodeLookup.set(0x1778, () => new TooManyAdminsError())
createErrorFromNameLookup.set('TooManyAdmins', () => new TooManyAdminsError())

/**
 * TooManyOptionTypes: ''
 *
 * @category Errors
 * @category generated
 */
export class TooManyOptionTypesError extends Error {
  readonly code: number = 0x1779
  readonly name: string = 'TooManyOptionTypes'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, TooManyOptionTypesError)
    }
  }
}

createErrorFromCodeLookup.set(0x1779, () => new TooManyOptionTypesError())
createErrorFromNameLookup.set(
  'TooManyOptionTypes',
  () => new TooManyOptionTypesError()
)

/**
 * DuplicateSignature: ''
 *
 * @category Errors
 * @category generated
 */
export class DuplicateSignatureError extends Error {
  readonly code: number = 0x177a
  readonly name: string = 'DuplicateSignature'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, DuplicateSignatureError)
    }
  }
}

createErrorFromCodeLookup.set(0x177a, () => new DuplicateSignatureError())
createErrorFromNameLookup.set(
  'DuplicateSignature',
  () => new DuplicateSignatureError()
)

/**
 * Expired: ''
 *
 * @category Errors
 * @category generated
 */
export class ExpiredError extends Error {
  readonly code: number = 0x177b
  readonly name: string = 'Expired'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ExpiredError)
    }
  }
}

createErrorFromCodeLookup.set(0x177b, () => new ExpiredError())
createErrorFromNameLookup.set('Expired', () => new ExpiredError())

/**
 * InvalidVid: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidVidError extends Error {
  readonly code: number = 0x177c
  readonly name: string = 'InvalidVid'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidVidError)
    }
  }
}

createErrorFromCodeLookup.set(0x177c, () => new InvalidVidError())
createErrorFromNameLookup.set('InvalidVid', () => new InvalidVidError())

/**
 * Paused: ''
 *
 * @category Errors
 * @category generated
 */
export class PausedError extends Error {
  readonly code: number = 0x177d
  readonly name: string = 'Paused'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, PausedError)
    }
  }
}

createErrorFromCodeLookup.set(0x177d, () => new PausedError())
createErrorFromNameLookup.set('Paused', () => new PausedError())

/**
 * UnexpiredExecuteHash: ''
 *
 * @category Errors
 * @category generated
 */
export class UnexpiredExecuteHashError extends Error {
  readonly code: number = 0x177e
  readonly name: string = 'UnexpiredExecuteHash'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, UnexpiredExecuteHashError)
    }
  }
}

createErrorFromCodeLookup.set(0x177e, () => new UnexpiredExecuteHashError())
createErrorFromNameLookup.set(
  'UnexpiredExecuteHash',
  () => new UnexpiredExecuteHashError()
)

/**
 * InvalidAmount: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidAmountError extends Error {
  readonly code: number = 0x177f
  readonly name: string = 'InvalidAmount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidAmountError)
    }
  }
}

createErrorFromCodeLookup.set(0x177f, () => new InvalidAmountError())
createErrorFromNameLookup.set('InvalidAmount', () => new InvalidAmountError())

/**
 * EidNotSupported: ''
 *
 * @category Errors
 * @category generated
 */
export class EidNotSupportedError extends Error {
  readonly code: number = 0x1780
  readonly name: string = 'EidNotSupported'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, EidNotSupportedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1780, () => new EidNotSupportedError())
createErrorFromNameLookup.set(
  'EidNotSupported',
  () => new EidNotSupportedError()
)

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
