// @layerzerolabs/lz-solana-sdk-v2/src/solita/index.ts
export * from './pda-deriver'
export * as EndpointProgram from './endpoint'
export * as SimpleMessageLibProgram from './simple-message-lib'
export * as UlnProgram from './uln'
export * as ExecutorProgram from './executor'
export * as PriceFeedProgram from './pricefeed'
export * as DVNProgram from './dvn'

export * from './base-oapp'
export * from './utility'
export * from './config'
export * from './recevie'
export * from './types'
export * from './deployment'

export * as BlockedMessageLibProgram from '../generated/solita/blocked_messagelib'
export * as DvnProgram from '../generated/solita/dvn'
export * from './send-helper'
export * as UMI from '..'

import { AccountMeta, Commitment, Connection, PublicKey } from '@solana/web3.js'

import { PacketPath } from '@layerzerolabs/lz-v2-utilities'

export type SolanaPacketPath = Pick<PacketPath, 'sender' | 'dstEid' | 'receiver'>

/**
 * Enum representing different configuration types.
 */
export enum SetConfigType {
    /**
     * Executor configuration type.
     */
    EXECUTOR = 1,

    /**
     * Send ULN configuration type.
     */
    SEND_ULN = 2,

    /**
     * Receive ULN configuration type.
     */
    RECEIVE_ULN = 3,
}

/**
 * Interface representing the Message Library.
 */
export interface MessageLibInterface {
    /**
     * The program public key.
     */
    program: PublicKey

    /**
     * Gets the account metadata for a quote instruction.
     *
     * @param {Connection} connection - The Solana connection.
     * @param {PublicKey} payer - The payer public key.
     * @param {PacketPath} path - The packet path.
     * @param {Commitment} [commitment] - The commitment level.
     * @returns {Promise<AccountMeta[]>} A promise that resolves to an array of account metadata.
     */
    getQuoteIXAccountMetaForCPI(
        connection: Connection,
        payer: PublicKey,
        path: SolanaPacketPath,
        commitment?: Commitment
    ): Promise<AccountMeta[]>

    /**
     * Gets the account metadata for a send instruction.
     *
     * @param {Connection} connection - The Solana connection.
     * @param {PublicKey} payer - The payer public key.
     * @param {PacketPath} path - The packet path.
     * @returns {Promise<AccountMeta[]>} A promise that resolves to an array of account metadata.
     */
    getSendIXAccountMetaForCPI(connection: Connection, payer: PublicKey, path: SolanaPacketPath): Promise<AccountMeta[]>

    /**
     * Gets the account metadata for an init config instruction.
     *
     * @param {PublicKey} payer - The payer public key.
     * @param {PublicKey} oappID - The OAPP ID.
     * @param {number} eid - The endpoint ID.
     * @returns {AccountMeta[]} An array of account metadata.
     */
    getInitConfigIXAccountMetaForCPI(payer: PublicKey, oappID: PublicKey, eid: number): AccountMeta[]

    /**
     * Gets the account metadata for a set config instruction.
     *
     * @param {PublicKey} endpointProgram - The endpoint program public key.
     * @param {PublicKey} oappID - The OAPP ID.
     * @param {number} eid - The endpoint ID.
     * @returns {Promise<AccountMeta[]>} A promise that resolves to an array of account metadata.
     */
    getSetConfigIXAccountMetaForCPI(endpointProgram: PublicKey, oappID: PublicKey, eid: number): Promise<AccountMeta[]>
}
