// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/errors/endpoint.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Program, ProgramError } from '@metaplex-foundation/umi';

type ProgramErrorConstructor = new (
  program: Program,
  cause?: Error
) => ProgramError;
const codeToErrorMap: Map<number, ProgramErrorConstructor> = new Map();
const nameToErrorMap: Map<string, ProgramErrorConstructor> = new Map();

/** InvalidSendLibrary */
export class InvalidSendLibraryError extends ProgramError {
  override readonly name: string = 'InvalidSendLibrary';

  readonly code: number = 0x1770; // 6000

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1770, InvalidSendLibraryError);
nameToErrorMap.set('InvalidSendLibrary', InvalidSendLibraryError);

/** InvalidReceiveLibrary */
export class InvalidReceiveLibraryError extends ProgramError {
  override readonly name: string = 'InvalidReceiveLibrary';

  readonly code: number = 0x1771; // 6001

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1771, InvalidReceiveLibraryError);
nameToErrorMap.set('InvalidReceiveLibrary', InvalidReceiveLibraryError);

/** SameValue */
export class SameValueError extends ProgramError {
  override readonly name: string = 'SameValue';

  readonly code: number = 0x1772; // 6002

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1772, SameValueError);
nameToErrorMap.set('SameValue', SameValueError);

/** AccountNotFound */
export class AccountNotFoundError extends ProgramError {
  override readonly name: string = 'AccountNotFound';

  readonly code: number = 0x1773; // 6003

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1773, AccountNotFoundError);
nameToErrorMap.set('AccountNotFound', AccountNotFoundError);

/** OnlySendLib */
export class OnlySendLibError extends ProgramError {
  override readonly name: string = 'OnlySendLib';

  readonly code: number = 0x1774; // 6004

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1774, OnlySendLibError);
nameToErrorMap.set('OnlySendLib', OnlySendLibError);

/** OnlyReceiveLib */
export class OnlyReceiveLibError extends ProgramError {
  override readonly name: string = 'OnlyReceiveLib';

  readonly code: number = 0x1775; // 6005

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1775, OnlyReceiveLibError);
nameToErrorMap.set('OnlyReceiveLib', OnlyReceiveLibError);

/** InvalidExpiry */
export class InvalidExpiryError extends ProgramError {
  override readonly name: string = 'InvalidExpiry';

  readonly code: number = 0x1776; // 6006

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1776, InvalidExpiryError);
nameToErrorMap.set('InvalidExpiry', InvalidExpiryError);

/** OnlyNonDefaultLib */
export class OnlyNonDefaultLibError extends ProgramError {
  override readonly name: string = 'OnlyNonDefaultLib';

  readonly code: number = 0x1777; // 6007

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1777, OnlyNonDefaultLibError);
nameToErrorMap.set('OnlyNonDefaultLib', OnlyNonDefaultLibError);

/** InvalidAmount */
export class InvalidAmountError extends ProgramError {
  override readonly name: string = 'InvalidAmount';

  readonly code: number = 0x1778; // 6008

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1778, InvalidAmountError);
nameToErrorMap.set('InvalidAmount', InvalidAmountError);

/** InvalidNonce */
export class InvalidNonceError extends ProgramError {
  override readonly name: string = 'InvalidNonce';

  readonly code: number = 0x1779; // 6009

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1779, InvalidNonceError);
nameToErrorMap.set('InvalidNonce', InvalidNonceError);

/** Unauthorized */
export class UnauthorizedError extends ProgramError {
  override readonly name: string = 'Unauthorized';

  readonly code: number = 0x177a; // 6010

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177a, UnauthorizedError);
nameToErrorMap.set('Unauthorized', UnauthorizedError);

/** PayloadHashNotFound */
export class PayloadHashNotFoundError extends ProgramError {
  override readonly name: string = 'PayloadHashNotFound';

  readonly code: number = 0x177b; // 6011

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177b, PayloadHashNotFoundError);
nameToErrorMap.set('PayloadHashNotFound', PayloadHashNotFoundError);

/** ComposeNotFound */
export class ComposeNotFoundError extends ProgramError {
  override readonly name: string = 'ComposeNotFound';

  readonly code: number = 0x177c; // 6012

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177c, ComposeNotFoundError);
nameToErrorMap.set('ComposeNotFound', ComposeNotFoundError);

/** InvalidPayloadHash */
export class InvalidPayloadHashError extends ProgramError {
  override readonly name: string = 'InvalidPayloadHash';

  readonly code: number = 0x177d; // 6013

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177d, InvalidPayloadHashError);
nameToErrorMap.set('InvalidPayloadHash', InvalidPayloadHashError);

/** LzTokenUnavailable */
export class LzTokenUnavailableError extends ProgramError {
  override readonly name: string = 'LzTokenUnavailable';

  readonly code: number = 0x177e; // 6014

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177e, LzTokenUnavailableError);
nameToErrorMap.set('LzTokenUnavailable', LzTokenUnavailableError);

/** ReadOnlyAccount */
export class ReadOnlyAccountError extends ProgramError {
  override readonly name: string = 'ReadOnlyAccount';

  readonly code: number = 0x177f; // 6015

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177f, ReadOnlyAccountError);
nameToErrorMap.set('ReadOnlyAccount', ReadOnlyAccountError);

/** InvalidMessageLib */
export class InvalidMessageLibError extends ProgramError {
  override readonly name: string = 'InvalidMessageLib';

  readonly code: number = 0x1780; // 6016

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1780, InvalidMessageLibError);
nameToErrorMap.set('InvalidMessageLib', InvalidMessageLibError);

/** WritableAccountNotAllowed */
export class WritableAccountNotAllowedError extends ProgramError {
  override readonly name: string = 'WritableAccountNotAllowed';

  readonly code: number = 0x1781; // 6017

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1781, WritableAccountNotAllowedError);
nameToErrorMap.set('WritableAccountNotAllowed', WritableAccountNotAllowedError);

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
export function getEndpointErrorFromCode(
  code: number,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
export function getEndpointErrorFromName(
  name: string,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}
