// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/events/defaultConfigSetEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable } from '@metaplex-foundation/umi';
import {
  Serializer,
  option,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ExecutorConfig,
  ExecutorConfigArgs,
  UlnConfig,
  UlnConfigArgs,
  getExecutorConfigSerializer,
  getUlnConfigSerializer,
} from '../types';

export type DefaultConfigSetEvent = {
  eid: number;
  sendUlnConfig: Option<UlnConfig>;
  receiveUlnConfig: Option<UlnConfig>;
  executorConfig: Option<ExecutorConfig>;
};

export type DefaultConfigSetEventArgs = {
  eid: number;
  sendUlnConfig: OptionOrNullable<UlnConfigArgs>;
  receiveUlnConfig: OptionOrNullable<UlnConfigArgs>;
  executorConfig: OptionOrNullable<ExecutorConfigArgs>;
};

export function getDefaultConfigSetEventSerializer(): Serializer<
  DefaultConfigSetEventArgs,
  DefaultConfigSetEvent
> {
  return struct<DefaultConfigSetEvent>(
    [
      ['eid', u32()],
      ['sendUlnConfig', option(getUlnConfigSerializer())],
      ['receiveUlnConfig', option(getUlnConfigSerializer())],
      ['executorConfig', option(getExecutorConfigSerializer())],
    ],
    { description: 'DefaultConfigSetEvent' }
  ) as Serializer<DefaultConfigSetEventArgs, DefaultConfigSetEvent>;
}
