// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/events/feesPaidEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable } from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  option,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  TreasuryFee,
  TreasuryFeeArgs,
  WorkerFee,
  WorkerFeeArgs,
  getTreasuryFeeSerializer,
  getWorkerFeeSerializer,
} from '../types';

export type FeesPaidEvent = {
  executor: WorkerFee;
  dvns: Array<WorkerFee>;
  treasury: Option<TreasuryFee>;
};

export type FeesPaidEventArgs = {
  executor: WorkerFeeArgs;
  dvns: Array<WorkerFeeArgs>;
  treasury: OptionOrNullable<TreasuryFeeArgs>;
};

export function getFeesPaidEventSerializer(): Serializer<
  FeesPaidEventArgs,
  FeesPaidEvent
> {
  return struct<FeesPaidEvent>(
    [
      ['executor', getWorkerFeeSerializer()],
      ['dvns', array(getWorkerFeeSerializer())],
      ['treasury', option(getTreasuryFeeSerializer())],
    ],
    { description: 'FeesPaidEvent' }
  ) as Serializer<FeesPaidEventArgs, FeesPaidEvent>;
}
