// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/events/payloadVerifiedEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u64,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type PayloadVerifiedEvent = {
  dvn: PublicKey;
  header: Array<number>;
  confirmations: bigint;
  proofHash: Uint8Array;
};

export type PayloadVerifiedEventArgs = {
  dvn: PublicKey;
  header: Array<number>;
  confirmations: number | bigint;
  proofHash: Uint8Array;
};

export function getPayloadVerifiedEventSerializer(): Serializer<
  PayloadVerifiedEventArgs,
  PayloadVerifiedEvent
> {
  return struct<PayloadVerifiedEvent>(
    [
      ['dvn', publicKeySerializer()],
      ['header', array(u8(), { size: 81 })],
      ['confirmations', u64()],
      ['proofHash', bytes({ size: 32 })],
    ],
    { description: 'PayloadVerifiedEvent' }
  ) as Serializer<PayloadVerifiedEventArgs, PayloadVerifiedEvent>;
}
