// vite-plugin-cjs-interop/dist/index.js
// src/index.ts
import oxc from "oxc-parser";
import MagicString from "magic-string";
import { minimatch } from "minimatch";
var walker = import("estree-walker");
var CSS_LANGS_RE = /\.(css|less|sass|scss|styl|stylus|pcss|postcss|sss)(?:$|\?)/;
function cjsInterop(options) {
  const virtualModuleId = "virtual:cjs-dyn-import";
  const dependencies = Array.from(new Set(options.dependencies));
  const { client = false, apply = "both" } = options;
  let sourcemaps = false;
  const matchedDependencies = {};
  const matchesDependencies = (value) => {
    if (!(value in matchedDependencies)) {
      matchedDependencies[value] = dependencies.some(
        (dependency) => minimatch(value, dependency)
      );
    }
    return matchedDependencies[value];
  };
  return {
    name: "cjs-interop",
    enforce: "post",
    apply: apply === "both" ? void 0 : apply,
    configResolved(config) {
      sourcemaps = !!config.build.sourcemap;
    },
    async transform(code, id, options2) {
      if (!client && !options2?.ssr) return;
      if (CSS_LANGS_RE.test(id)) return;
      const { program: ast } = await oxc.parseAsync(id, code);
      const toBeFixed = [];
      const preambles = [];
      let hasDynamicImportsToFix = false;
      const { walk } = await walker;
      walk(ast, {
        enter(node) {
          switch (node.type) {
            case "ImportDeclaration":
            case "ExportNamedDeclaration":
            case "ImportExpression":
              if (node.source && node.source.type === "Literal" && matchesDependencies(node.source.value)) {
                toBeFixed.push(node);
                if (node.type === "ImportExpression") {
                  hasDynamicImportsToFix = true;
                }
              }
              break;
            default:
          }
        }
      });
      if (toBeFixed.length === 0) {
        return;
      }
      const bottomUpToBeFixed = toBeFixed.reverse();
      const ms = sourcemaps ? new MagicString(code) : null;
      let counter = 1;
      let specifierCounter = 1;
      let isNamespaceImport = false;
      for (const node of bottomUpToBeFixed) {
        if (node.type === "ImportExpression") {
          const insertion = ".then(__cjs_dyn_import__)";
          if (sourcemaps) {
            ms.appendRight(node.end, insertion);
          } else {
            code = code.slice(0, node.end) + insertion + code.slice(node.end);
          }
          continue;
        }
        if (node.type === "ExportNamedDeclaration") {
          const importDestructurings = [];
          const exportDestructurings = [];
          const name2 = `__cjsInterop${counter++}__`;
          let changed2 = false;
          let defaultExportSpecifier = null;
          for (const specifier of node.specifiers) {
            if (specifier.type === "ExportSpecifier") {
              if (specifier.local.name === "default") {
                defaultExportSpecifier = specifier;
                continue;
              }
              changed2 = true;
              const specifierName = `__cjsInteropSpecifier${specifierCounter++}__`;
              importDestructurings.push(
                `${specifier.local.name}: ${specifierName}`
              );
              exportDestructurings.push(
                `${specifierName} as ${specifier.exported.name}`
              );
            } else {
              throw new Error(
                `Unknown ExportNamedDeclaration type specifier: ${specifier.type}`
              );
            }
          }
          if (!changed2) {
            continue;
          }
          if (defaultExportSpecifier) {
            exportDestructurings.push(
              `${name2} as ${defaultExportSpecifier.exported.name}}`
            );
          }
          preambles.push(
            `const { ${importDestructurings.join(
              ", "
            )} } = ${name2}?.default?.__esModule ? ${name2}.default : ${name2};`
          );
          const replacementNamedImports = `import ${name2} from ${JSON.stringify(
            node.source.value
          )};`;
          const replacementNamedExports = `export { ${exportDestructurings.join(", ")} };`;
          const replacement2 = [
            replacementNamedImports,
            replacementNamedExports
          ].filter(Boolean).join("\n");
          if (ms) {
            ms.overwrite(node.start, node.end, replacement2);
          } else {
            code = code.slice(0, node.start) + replacement2 + code.slice(node.end);
          }
          continue;
        }
        const destructurings = [];
        const name = `__cjsInterop${counter++}__`;
        let changed = false;
        for (const specifier of node.specifiers || []) {
          if (specifier.type === "ImportDefaultSpecifier") {
            changed = true;
            destructurings.push(
              `default: ${specifier.local.name} = ${name}`
            );
          } else if (specifier.type === "ImportSpecifier") {
            changed = true;
            if (specifier.imported.name === specifier.local.name) {
              destructurings.push(specifier.local.name);
            } else {
              destructurings.push(
                `${specifier.imported.name}: ${specifier.local.name}`
              );
            }
          } else if (specifier.type === "ImportNamespaceSpecifier") {
            changed = true;
            isNamespaceImport = true;
            destructurings.push(specifier.local.name);
          }
        }
        if (!changed) {
          continue;
        }
        if (!isNamespaceImport)
          preambles.push(
            `const { ${destructurings.join(
              ", "
            )} } = ${name}?.default?.__esModule ? ${name}.default : ${name};`
          );
        else
          preambles.push(
            `const ${destructurings[0]} = ${name}?.default?.__esModule ? ${name}.default : ${name};`
          );
        const replacement = `import ${name} from ${JSON.stringify(
          node.source.value
        )};`;
        if (sourcemaps) {
          ms.overwrite(node.start, node.end, replacement);
        } else {
          code = code.slice(0, node.start) + replacement + code.slice(node.end);
        }
      }
      if (hasDynamicImportsToFix) {
        const importCompat = `import { __cjs_dyn_import__ } from "${virtualModuleId}";
`;
        if (sourcemaps) {
          ms.prepend(importCompat);
        } else {
          code = importCompat + code;
        }
      }
      const preamble = preambles.reverse().join("\n") + "\n";
      if (sourcemaps) {
        ms.prepend(preamble);
        return {
          code: ms.toString(),
          map: ms.generateMap({ hires: true })
        };
      } else {
        code = preamble + code;
        return {
          code
        };
      }
    },
    resolveId(id) {
      if (id === virtualModuleId) {
        return id;
      }
    },
    load(id) {
      if (id === virtualModuleId) {
        return `
// This file is generated by vite-plugin-cjs-interop
// The cache allows the same object to be returned for the same dynamic import
// however static imports are not affected and will therefore return a different
// object
const modCache = new WeakMap();

export function __cjs_dyn_import__(rawImport) {
  if (rawImport?.default?.__esModule) {
    return rawImport.default;
  }
  if (modCache.has(rawImport)) {
    return modCache.get(rawImport);
  }
  const source = rawImport?.default;
  if (source) {
    const mod = Object.create(rawImport);
    modCache.set(rawImport, mod);
    Object.keys(source)
      .filter(key => !Object.hasOwn(rawImport, key))
      .forEach(key => {
        Object.defineProperty(mod, key, { enumerable: true, get: () => source[key] })
      });
    return mod;
  }
  return rawImport;
}`;
      }
    }
  };
}
export {
  cjsInterop
};
