// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/types/adminSetConfigParams.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  GetDataEnumKind,
  GetDataEnumKindContent,
  Serializer,
  array,
  dataEnum,
  publicKey as publicKeySerializer,
  struct,
  tuple,
  u16,
} from '@metaplex-foundation/umi/serializers';
import { DstConfig, DstConfigArgs, getDstConfigSerializer } from '.';

export type AdminSetConfigParams =
  | { __kind: 'PriceFeed'; fields: [PublicKey] }
  | { __kind: 'DefaultMultiplierBps'; fields: [number] }
  | { __kind: 'DstConfigs'; fields: [Array<DstConfig>] };

export type AdminSetConfigParamsArgs =
  | { __kind: 'PriceFeed'; fields: [PublicKey] }
  | { __kind: 'DefaultMultiplierBps'; fields: [number] }
  | { __kind: 'DstConfigs'; fields: [Array<DstConfigArgs>] };

export function getAdminSetConfigParamsSerializer(): Serializer<
  AdminSetConfigParamsArgs,
  AdminSetConfigParams
> {
  return dataEnum<AdminSetConfigParams>(
    [
      [
        'PriceFeed',
        struct<GetDataEnumKindContent<AdminSetConfigParams, 'PriceFeed'>>([
          ['fields', tuple([publicKeySerializer()])],
        ]),
      ],
      [
        'DefaultMultiplierBps',
        struct<
          GetDataEnumKindContent<AdminSetConfigParams, 'DefaultMultiplierBps'>
        >([['fields', tuple([u16()])]]),
      ],
      [
        'DstConfigs',
        struct<GetDataEnumKindContent<AdminSetConfigParams, 'DstConfigs'>>([
          ['fields', tuple([array(getDstConfigSerializer())])],
        ]),
      ],
    ],
    { description: 'AdminSetConfigParams' }
  ) as Serializer<AdminSetConfigParamsArgs, AdminSetConfigParams>;
}

// Data Enum Helpers.
export function adminSetConfigParams(
  kind: 'PriceFeed',
  data: GetDataEnumKindContent<AdminSetConfigParamsArgs, 'PriceFeed'>['fields']
): GetDataEnumKind<AdminSetConfigParamsArgs, 'PriceFeed'>;
export function adminSetConfigParams(
  kind: 'DefaultMultiplierBps',
  data: GetDataEnumKindContent<
    AdminSetConfigParamsArgs,
    'DefaultMultiplierBps'
  >['fields']
): GetDataEnumKind<AdminSetConfigParamsArgs, 'DefaultMultiplierBps'>;
export function adminSetConfigParams(
  kind: 'DstConfigs',
  data: GetDataEnumKindContent<AdminSetConfigParamsArgs, 'DstConfigs'>['fields']
): GetDataEnumKind<AdminSetConfigParamsArgs, 'DstConfigs'>;
export function adminSetConfigParams<
  K extends AdminSetConfigParamsArgs['__kind'],
>(kind: K, data?: any): Extract<AdminSetConfigParamsArgs, { __kind: K }> {
  return Array.isArray(data)
    ? { __kind: kind, fields: data }
    : { __kind: kind, ...(data ?? {}) };
}
export function isAdminSetConfigParams<
  K extends AdminSetConfigParams['__kind'],
>(
  kind: K,
  value: AdminSetConfigParams
): value is AdminSetConfigParams & { __kind: K } {
  return value.__kind === kind;
}
