// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/types/ownerSetConfigParams.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  GetDataEnumKind,
  GetDataEnumKindContent,
  Serializer,
  array,
  bool,
  dataEnum,
  publicKey as publicKeySerializer,
  struct,
  tuple,
} from '@metaplex-foundation/umi/serializers';

export type OwnerSetConfigParams =
  | { __kind: 'Admins'; fields: [Array<PublicKey>] }
  | { __kind: 'Executors'; fields: [Array<PublicKey>] }
  | { __kind: 'Msglibs'; fields: [Array<PublicKey>] }
  | { __kind: 'Owner'; fields: [PublicKey] }
  | { __kind: 'Paused'; fields: [boolean] }
  | { __kind: 'Allowlist'; fields: [Array<PublicKey>] }
  | { __kind: 'Denylist'; fields: [Array<PublicKey>] };

export type OwnerSetConfigParamsArgs = OwnerSetConfigParams;

export function getOwnerSetConfigParamsSerializer(): Serializer<
  OwnerSetConfigParamsArgs,
  OwnerSetConfigParams
> {
  return dataEnum<OwnerSetConfigParams>(
    [
      [
        'Admins',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Admins'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Executors',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Executors'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Msglibs',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Msglibs'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Owner',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Owner'>>([
          ['fields', tuple([publicKeySerializer()])],
        ]),
      ],
      [
        'Paused',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Paused'>>([
          ['fields', tuple([bool()])],
        ]),
      ],
      [
        'Allowlist',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Allowlist'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Denylist',
        struct<GetDataEnumKindContent<OwnerSetConfigParams, 'Denylist'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
    ],
    { description: 'OwnerSetConfigParams' }
  ) as Serializer<OwnerSetConfigParamsArgs, OwnerSetConfigParams>;
}

// Data Enum Helpers.
export function ownerSetConfigParams(
  kind: 'Admins',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Admins'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Admins'>;
export function ownerSetConfigParams(
  kind: 'Executors',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Executors'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Executors'>;
export function ownerSetConfigParams(
  kind: 'Msglibs',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Msglibs'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Msglibs'>;
export function ownerSetConfigParams(
  kind: 'Owner',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Owner'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Owner'>;
export function ownerSetConfigParams(
  kind: 'Paused',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Paused'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Paused'>;
export function ownerSetConfigParams(
  kind: 'Allowlist',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Allowlist'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Allowlist'>;
export function ownerSetConfigParams(
  kind: 'Denylist',
  data: GetDataEnumKindContent<OwnerSetConfigParamsArgs, 'Denylist'>['fields']
): GetDataEnumKind<OwnerSetConfigParamsArgs, 'Denylist'>;
export function ownerSetConfigParams<
  K extends OwnerSetConfigParamsArgs['__kind'],
>(kind: K, data?: any): Extract<OwnerSetConfigParamsArgs, { __kind: K }> {
  return Array.isArray(data)
    ? { __kind: kind, fields: data }
    : { __kind: kind, ...(data ?? {}) };
}
export function isOwnerSetConfigParams<
  K extends OwnerSetConfigParams['__kind'],
>(
  kind: K,
  value: OwnerSetConfigParams
): value is OwnerSetConfigParams & { __kind: K } {
  return value.__kind === kind;
}
