// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/accounts/ExecutorConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { Acl, aclBeet } from '../types/Acl'
import { DstConfig, dstConfigBeet } from '../types/DstConfig'

/**
 * Arguments used to create {@link ExecutorConfig}
 * @category Accounts
 * @category generated
 */
export type ExecutorConfigArgs = {
  bump: number
  owner: web3.PublicKey
  acl: Acl
  admins: web3.PublicKey[]
  executors: web3.PublicKey[]
  msglibs: web3.PublicKey[]
  paused: boolean
  defaultMultiplierBps: number
  priceFeed: web3.PublicKey
  dstConfigs: DstConfig[]
}

export const executorConfigDiscriminator = [134, 17, 226, 24, 10, 173, 157, 78]
/**
 * Holds the data for the {@link ExecutorConfig} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class ExecutorConfig implements ExecutorConfigArgs {
  private constructor(
    readonly bump: number,
    readonly owner: web3.PublicKey,
    readonly acl: Acl,
    readonly admins: web3.PublicKey[],
    readonly executors: web3.PublicKey[],
    readonly msglibs: web3.PublicKey[],
    readonly paused: boolean,
    readonly defaultMultiplierBps: number,
    readonly priceFeed: web3.PublicKey,
    readonly dstConfigs: DstConfig[]
  ) {}

  /**
   * Creates a {@link ExecutorConfig} instance from the provided args.
   */
  static fromArgs(args: ExecutorConfigArgs) {
    return new ExecutorConfig(
      args.bump,
      args.owner,
      args.acl,
      args.admins,
      args.executors,
      args.msglibs,
      args.paused,
      args.defaultMultiplierBps,
      args.priceFeed,
      args.dstConfigs
    )
  }

  /**
   * Deserializes the {@link ExecutorConfig} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [ExecutorConfig, number] {
    return ExecutorConfig.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link ExecutorConfig} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<ExecutorConfig> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find ExecutorConfig account at ${address}`)
    }
    return ExecutorConfig.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, executorConfigBeet)
  }

  /**
   * Deserializes the {@link ExecutorConfig} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [ExecutorConfig, number] {
    return executorConfigBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link ExecutorConfig} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return executorConfigBeet.serialize({
      accountDiscriminator: executorConfigDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link ExecutorConfig} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: ExecutorConfigArgs) {
    const instance = ExecutorConfig.fromArgs(args)
    return executorConfigBeet.toFixedFromValue({
      accountDiscriminator: executorConfigDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link ExecutorConfig} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: ExecutorConfigArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      ExecutorConfig.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link ExecutorConfig} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      bump: this.bump,
      owner: this.owner.toBase58(),
      acl: this.acl,
      admins: this.admins,
      executors: this.executors,
      msglibs: this.msglibs,
      paused: this.paused,
      defaultMultiplierBps: this.defaultMultiplierBps,
      priceFeed: this.priceFeed.toBase58(),
      dstConfigs: this.dstConfigs,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const executorConfigBeet = new beet.FixableBeetStruct<
  ExecutorConfig,
  ExecutorConfigArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['bump', beet.u8],
    ['owner', beetSolana.publicKey],
    ['acl', aclBeet],
    ['admins', beet.array(beetSolana.publicKey)],
    ['executors', beet.array(beetSolana.publicKey)],
    ['msglibs', beet.array(beetSolana.publicKey)],
    ['paused', beet.bool],
    ['defaultMultiplierBps', beet.u16],
    ['priceFeed', beetSolana.publicKey],
    ['dstConfigs', beet.array(dstConfigBeet)],
  ],
  ExecutorConfig.fromArgs,
  'ExecutorConfig'
)
