// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/burn.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type BurnInstructionAccounts = {
  /** The PDA of the OApp or delegate */
  signer: Signer;
  oappRegistry: PublicKey | Pda;
  nonce: PublicKey | Pda;
  /** close the account and return the lamports to endpoint settings account */
  payloadHash: PublicKey | Pda;
  endpoint: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type BurnInstructionData = {
  discriminator: Uint8Array;
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: bigint;
  payloadHash: Uint8Array;
};

export type BurnInstructionDataArgs = {
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: number | bigint;
  payloadHash: Uint8Array;
};

export function getBurnInstructionDataSerializer(): Serializer<
  BurnInstructionDataArgs,
  BurnInstructionData
> {
  return mapSerializer<BurnInstructionDataArgs, any, BurnInstructionData>(
    struct<BurnInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['receiver', publicKeySerializer()],
        ['srcEid', u32()],
        ['sender', bytes({ size: 32 })],
        ['nonce', u64()],
        ['payloadHash', bytes({ size: 32 })],
      ],
      { description: 'BurnInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([116, 110, 29, 56, 107, 219, 42, 93]),
    })
  ) as Serializer<BurnInstructionDataArgs, BurnInstructionData>;
}

// Args.
export type BurnInstructionArgs = BurnInstructionDataArgs;

// Instruction.
export function burn(
  context: Pick<Context, 'programs'>,
  accounts: BurnInstructionAccounts,
  args: BurnInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.signer ?? null,
    },
    oappRegistry: {
      index: 1,
      isWritable: false as boolean,
      value: accounts.oappRegistry ?? null,
    },
    nonce: {
      index: 2,
      isWritable: false as boolean,
      value: accounts.nonce ?? null,
    },
    payloadHash: {
      index: 3,
      isWritable: true as boolean,
      value: accounts.payloadHash ?? null,
    },
    endpoint: {
      index: 4,
      isWritable: true as boolean,
      value: accounts.endpoint ?? null,
    },
    eventAuthority: {
      index: 5,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 6,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: BurnInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getBurnInstructionDataSerializer().serialize(
    resolvedArgs as BurnInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
