// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/initEndpoint.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type InitEndpointInstructionAccounts = {
  payer?: Signer;
  endpoint: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
};

// Data.
export type InitEndpointInstructionData = {
  discriminator: Uint8Array;
  eid: number;
  admin: PublicKey;
};

export type InitEndpointInstructionDataArgs = { eid: number; admin: PublicKey };

export function getInitEndpointInstructionDataSerializer(): Serializer<
  InitEndpointInstructionDataArgs,
  InitEndpointInstructionData
> {
  return mapSerializer<
    InitEndpointInstructionDataArgs,
    any,
    InitEndpointInstructionData
  >(
    struct<InitEndpointInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['eid', u32()],
        ['admin', publicKeySerializer()],
      ],
      { description: 'InitEndpointInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([178, 30, 29, 207, 120, 225, 246, 134]),
    })
  ) as Serializer<InitEndpointInstructionDataArgs, InitEndpointInstructionData>;
}

// Args.
export type InitEndpointInstructionArgs = InitEndpointInstructionDataArgs;

// Instruction.
export function initEndpoint(
  context: Pick<Context, 'payer' | 'programs'>,
  input: InitEndpointInstructionAccounts & InitEndpointInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true as boolean,
      value: input.payer ?? null,
    },
    endpoint: {
      index: 1,
      isWritable: true as boolean,
      value: input.endpoint ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: InitEndpointInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getInitEndpointInstructionDataSerializer().serialize(
    resolvedArgs as InitEndpointInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
