// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/initConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type InitConfigInstructionAccounts = {
  /** only the delegate can initialize the config accounts */
  delegate: Signer;
  oappRegistry: PublicKey | Pda;
  /** The PDA signer to the message lib when the endpoint calls the message lib program. */
  messageLibInfo: PublicKey | Pda;
  /** the pda of the message_lib_program */
  messageLib: PublicKey | Pda;
  messageLibProgram: PublicKey | Pda;
};

// Data.
export type InitConfigInstructionData = {
  discriminator: Uint8Array;
  oapp: PublicKey;
  eid: number;
};

export type InitConfigInstructionDataArgs = { oapp: PublicKey; eid: number };

export function getInitConfigInstructionDataSerializer(): Serializer<
  InitConfigInstructionDataArgs,
  InitConfigInstructionData
> {
  return mapSerializer<
    InitConfigInstructionDataArgs,
    any,
    InitConfigInstructionData
  >(
    struct<InitConfigInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['oapp', publicKeySerializer()],
        ['eid', u32()],
      ],
      { description: 'InitConfigInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([23, 235, 115, 232, 168, 96, 1, 231]),
    })
  ) as Serializer<InitConfigInstructionDataArgs, InitConfigInstructionData>;
}

// Args.
export type InitConfigInstructionArgs = InitConfigInstructionDataArgs;

// Instruction.
export function initConfig(
  context: Pick<Context, 'programs'>,
  input: InitConfigInstructionAccounts & InitConfigInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    delegate: {
      index: 0,
      isWritable: false as boolean,
      value: input.delegate ?? null,
    },
    oappRegistry: {
      index: 1,
      isWritable: false as boolean,
      value: input.oappRegistry ?? null,
    },
    messageLibInfo: {
      index: 2,
      isWritable: false as boolean,
      value: input.messageLibInfo ?? null,
    },
    messageLib: {
      index: 3,
      isWritable: false as boolean,
      value: input.messageLib ?? null,
    },
    messageLibProgram: {
      index: 4,
      isWritable: false as boolean,
      value: input.messageLibProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: InitConfigInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getInitConfigInstructionDataSerializer().serialize(
    resolvedArgs as InitConfigInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
