// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/initDefaultSendLibrary.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type InitDefaultSendLibraryInstructionAccounts = {
  admin: Signer;
  endpoint: PublicKey | Pda;
  defaultSendLibraryConfig: PublicKey | Pda;
  messageLibInfo: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type InitDefaultSendLibraryInstructionData = {
  discriminator: Uint8Array;
  eid: number;
  newLib: PublicKey;
};

export type InitDefaultSendLibraryInstructionDataArgs = {
  eid: number;
  newLib: PublicKey;
};

export function getInitDefaultSendLibraryInstructionDataSerializer(): Serializer<
  InitDefaultSendLibraryInstructionDataArgs,
  InitDefaultSendLibraryInstructionData
> {
  return mapSerializer<
    InitDefaultSendLibraryInstructionDataArgs,
    any,
    InitDefaultSendLibraryInstructionData
  >(
    struct<InitDefaultSendLibraryInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['eid', u32()],
        ['newLib', publicKeySerializer()],
      ],
      { description: 'InitDefaultSendLibraryInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([120, 187, 15, 31, 174, 97, 138, 58]),
    })
  ) as Serializer<
    InitDefaultSendLibraryInstructionDataArgs,
    InitDefaultSendLibraryInstructionData
  >;
}

// Args.
export type InitDefaultSendLibraryInstructionArgs =
  InitDefaultSendLibraryInstructionDataArgs;

// Instruction.
export function initDefaultSendLibrary(
  context: Pick<Context, 'programs'>,
  input: InitDefaultSendLibraryInstructionAccounts &
    InitDefaultSendLibraryInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: true as boolean,
      value: input.admin ?? null,
    },
    endpoint: {
      index: 1,
      isWritable: false as boolean,
      value: input.endpoint ?? null,
    },
    defaultSendLibraryConfig: {
      index: 2,
      isWritable: true as boolean,
      value: input.defaultSendLibraryConfig ?? null,
    },
    messageLibInfo: {
      index: 3,
      isWritable: false as boolean,
      value: input.messageLibInfo ?? null,
    },
    systemProgram: {
      index: 4,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
    eventAuthority: {
      index: 5,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 6,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: InitDefaultSendLibraryInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getInitDefaultSendLibraryInstructionDataSerializer().serialize(
    resolvedArgs as InitDefaultSendLibraryInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
