// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/lzComposeAlert.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u16,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type LzComposeAlertInstructionAccounts = {
  executor: Signer;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type LzComposeAlertInstructionData = {
  discriminator: Uint8Array;
  from: PublicKey;
  to: PublicKey;
  guid: Uint8Array;
  index: number;
  computeUnits: bigint;
  value: bigint;
  message: Uint8Array;
  extraData: Uint8Array;
  reason: Uint8Array;
};

export type LzComposeAlertInstructionDataArgs = {
  from: PublicKey;
  to: PublicKey;
  guid: Uint8Array;
  index: number;
  computeUnits: number | bigint;
  value: number | bigint;
  message: Uint8Array;
  extraData: Uint8Array;
  reason: Uint8Array;
};

export function getLzComposeAlertInstructionDataSerializer(): Serializer<
  LzComposeAlertInstructionDataArgs,
  LzComposeAlertInstructionData
> {
  return mapSerializer<
    LzComposeAlertInstructionDataArgs,
    any,
    LzComposeAlertInstructionData
  >(
    struct<LzComposeAlertInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['from', publicKeySerializer()],
        ['to', publicKeySerializer()],
        ['guid', bytes({ size: 32 })],
        ['index', u16()],
        ['computeUnits', u64()],
        ['value', u64()],
        ['message', bytes({ size: u32() })],
        ['extraData', bytes({ size: u32() })],
        ['reason', bytes({ size: u32() })],
      ],
      { description: 'LzComposeAlertInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([27, 91, 198, 77, 66, 92, 122, 167]),
    })
  ) as Serializer<
    LzComposeAlertInstructionDataArgs,
    LzComposeAlertInstructionData
  >;
}

// Args.
export type LzComposeAlertInstructionArgs = LzComposeAlertInstructionDataArgs;

// Instruction.
export function lzComposeAlert(
  context: Pick<Context, 'programs'>,
  input: LzComposeAlertInstructionAccounts & LzComposeAlertInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    executor: {
      index: 0,
      isWritable: false as boolean,
      value: input.executor ?? null,
    },
    eventAuthority: {
      index: 1,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 2,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: LzComposeAlertInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getLzComposeAlertInstructionDataSerializer().serialize(
    resolvedArgs as LzComposeAlertInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
