// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/nilify.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type NilifyInstructionAccounts = {
  /** The PDA of the OApp or delegate */
  signer: Signer;
  oappRegistry: PublicKey | Pda;
  nonce: PublicKey | Pda;
  pendingInboundNonce: PublicKey | Pda;
  payloadHash: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type NilifyInstructionData = {
  discriminator: Uint8Array;
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: bigint;
  payloadHash: Uint8Array;
};

export type NilifyInstructionDataArgs = {
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: number | bigint;
  payloadHash: Uint8Array;
};

export function getNilifyInstructionDataSerializer(): Serializer<
  NilifyInstructionDataArgs,
  NilifyInstructionData
> {
  return mapSerializer<NilifyInstructionDataArgs, any, NilifyInstructionData>(
    struct<NilifyInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['receiver', publicKeySerializer()],
        ['srcEid', u32()],
        ['sender', bytes({ size: 32 })],
        ['nonce', u64()],
        ['payloadHash', bytes({ size: 32 })],
      ],
      { description: 'NilifyInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([143, 136, 129, 199, 36, 35, 160, 85]),
    })
  ) as Serializer<NilifyInstructionDataArgs, NilifyInstructionData>;
}

// Args.
export type NilifyInstructionArgs = NilifyInstructionDataArgs;

// Instruction.
export function nilify(
  context: Pick<Context, 'programs'>,
  accounts: NilifyInstructionAccounts,
  args: NilifyInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.signer ?? null,
    },
    oappRegistry: {
      index: 1,
      isWritable: false as boolean,
      value: accounts.oappRegistry ?? null,
    },
    nonce: {
      index: 2,
      isWritable: true as boolean,
      value: accounts.nonce ?? null,
    },
    pendingInboundNonce: {
      index: 3,
      isWritable: true as boolean,
      value: accounts.pendingInboundNonce ?? null,
    },
    payloadHash: {
      index: 4,
      isWritable: true as boolean,
      value: accounts.payloadHash ?? null,
    },
    eventAuthority: {
      index: 5,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 6,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: NilifyInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getNilifyInstructionDataSerializer().serialize(
    resolvedArgs as NilifyInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
