// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/registerOapp.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type RegisterOappInstructionAccounts = {
  payer?: Signer;
  /** The PDA of the OApp */
  oapp: Signer;
  oappRegistry: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type RegisterOappInstructionData = {
  discriminator: Uint8Array;
  delegate: PublicKey;
};

export type RegisterOappInstructionDataArgs = { delegate: PublicKey };

export function getRegisterOappInstructionDataSerializer(): Serializer<
  RegisterOappInstructionDataArgs,
  RegisterOappInstructionData
> {
  return mapSerializer<
    RegisterOappInstructionDataArgs,
    any,
    RegisterOappInstructionData
  >(
    struct<RegisterOappInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['delegate', publicKeySerializer()],
      ],
      { description: 'RegisterOappInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([129, 89, 71, 68, 11, 82, 210, 125]),
    })
  ) as Serializer<RegisterOappInstructionDataArgs, RegisterOappInstructionData>;
}

// Args.
export type RegisterOappInstructionArgs = RegisterOappInstructionDataArgs;

// Instruction.
export function registerOapp(
  context: Pick<Context, 'payer' | 'programs'>,
  input: RegisterOappInstructionAccounts & RegisterOappInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true as boolean,
      value: input.payer ?? null,
    },
    oapp: { index: 1, isWritable: false as boolean, value: input.oapp ?? null },
    oappRegistry: {
      index: 2,
      isWritable: true as boolean,
      value: input.oappRegistry ?? null,
    },
    systemProgram: {
      index: 3,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
    eventAuthority: {
      index: 4,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 5,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: RegisterOappInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getRegisterOappInstructionDataSerializer().serialize(
    resolvedArgs as RegisterOappInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
