// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/registerLibrary.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import {
  MessageLibType,
  MessageLibTypeArgs,
  getMessageLibTypeSerializer,
} from '../types';

// Accounts.
export type RegisterLibraryInstructionAccounts = {
  admin: Signer;
  endpoint: PublicKey | Pda;
  messageLibInfo: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type RegisterLibraryInstructionData = {
  discriminator: Uint8Array;
  libProgram: PublicKey;
  libType: MessageLibType;
};

export type RegisterLibraryInstructionDataArgs = {
  libProgram: PublicKey;
  libType: MessageLibTypeArgs;
};

export function getRegisterLibraryInstructionDataSerializer(): Serializer<
  RegisterLibraryInstructionDataArgs,
  RegisterLibraryInstructionData
> {
  return mapSerializer<
    RegisterLibraryInstructionDataArgs,
    any,
    RegisterLibraryInstructionData
  >(
    struct<RegisterLibraryInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['libProgram', publicKeySerializer()],
        ['libType', getMessageLibTypeSerializer()],
      ],
      { description: 'RegisterLibraryInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([23, 171, 28, 116, 111, 193, 238, 142]),
    })
  ) as Serializer<
    RegisterLibraryInstructionDataArgs,
    RegisterLibraryInstructionData
  >;
}

// Args.
export type RegisterLibraryInstructionArgs = RegisterLibraryInstructionDataArgs;

// Instruction.
export function registerLibrary(
  context: Pick<Context, 'programs'>,
  input: RegisterLibraryInstructionAccounts & RegisterLibraryInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: true as boolean,
      value: input.admin ?? null,
    },
    endpoint: {
      index: 1,
      isWritable: false as boolean,
      value: input.endpoint ?? null,
    },
    messageLibInfo: {
      index: 2,
      isWritable: true as boolean,
      value: input.messageLibInfo ?? null,
    },
    systemProgram: {
      index: 3,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
    eventAuthority: {
      index: 4,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 5,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: RegisterLibraryInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getRegisterLibraryInstructionDataSerializer().serialize(
    resolvedArgs as RegisterLibraryInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
