// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/setDefaultReceiveLibraryTimeout.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type SetDefaultReceiveLibraryTimeoutInstructionAccounts = {
  admin: Signer;
  endpoint: PublicKey | Pda;
  defaultReceiveLibraryConfig: PublicKey | Pda;
  messageLibInfo: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type SetDefaultReceiveLibraryTimeoutInstructionData = {
  discriminator: Uint8Array;
  eid: number;
  lib: PublicKey;
  expiry: bigint;
};

export type SetDefaultReceiveLibraryTimeoutInstructionDataArgs = {
  eid: number;
  lib: PublicKey;
  expiry: number | bigint;
};

export function getSetDefaultReceiveLibraryTimeoutInstructionDataSerializer(): Serializer<
  SetDefaultReceiveLibraryTimeoutInstructionDataArgs,
  SetDefaultReceiveLibraryTimeoutInstructionData
> {
  return mapSerializer<
    SetDefaultReceiveLibraryTimeoutInstructionDataArgs,
    any,
    SetDefaultReceiveLibraryTimeoutInstructionData
  >(
    struct<SetDefaultReceiveLibraryTimeoutInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['eid', u32()],
        ['lib', publicKeySerializer()],
        ['expiry', u64()],
      ],
      { description: 'SetDefaultReceiveLibraryTimeoutInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([118, 32, 94, 51, 25, 247, 16, 232]),
    })
  ) as Serializer<
    SetDefaultReceiveLibraryTimeoutInstructionDataArgs,
    SetDefaultReceiveLibraryTimeoutInstructionData
  >;
}

// Args.
export type SetDefaultReceiveLibraryTimeoutInstructionArgs =
  SetDefaultReceiveLibraryTimeoutInstructionDataArgs;

// Instruction.
export function setDefaultReceiveLibraryTimeout(
  context: Pick<Context, 'programs'>,
  input: SetDefaultReceiveLibraryTimeoutInstructionAccounts &
    SetDefaultReceiveLibraryTimeoutInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false as boolean,
      value: input.admin ?? null,
    },
    endpoint: {
      index: 1,
      isWritable: false as boolean,
      value: input.endpoint ?? null,
    },
    defaultReceiveLibraryConfig: {
      index: 2,
      isWritable: true as boolean,
      value: input.defaultReceiveLibraryConfig ?? null,
    },
    messageLibInfo: {
      index: 3,
      isWritable: false as boolean,
      value: input.messageLibInfo ?? null,
    },
    eventAuthority: {
      index: 4,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 5,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetDefaultReceiveLibraryTimeoutInstructionArgs = {
    ...input,
  };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data =
    getSetDefaultReceiveLibraryTimeoutInstructionDataSerializer().serialize(
      resolvedArgs as SetDefaultReceiveLibraryTimeoutInstructionDataArgs
    );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
