// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/setReceiveLibraryTimeout.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type SetReceiveLibraryTimeoutInstructionAccounts = {
  /** The PDA of the OApp or delegate */
  signer: Signer;
  oappRegistry: PublicKey | Pda;
  receiveLibraryConfig: PublicKey | Pda;
  messageLibInfo: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type SetReceiveLibraryTimeoutInstructionData = {
  discriminator: Uint8Array;
  receiver: PublicKey;
  eid: number;
  lib: PublicKey;
  expiry: bigint;
};

export type SetReceiveLibraryTimeoutInstructionDataArgs = {
  receiver: PublicKey;
  eid: number;
  lib: PublicKey;
  expiry: number | bigint;
};

export function getSetReceiveLibraryTimeoutInstructionDataSerializer(): Serializer<
  SetReceiveLibraryTimeoutInstructionDataArgs,
  SetReceiveLibraryTimeoutInstructionData
> {
  return mapSerializer<
    SetReceiveLibraryTimeoutInstructionDataArgs,
    any,
    SetReceiveLibraryTimeoutInstructionData
  >(
    struct<SetReceiveLibraryTimeoutInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['receiver', publicKeySerializer()],
        ['eid', u32()],
        ['lib', publicKeySerializer()],
        ['expiry', u64()],
      ],
      { description: 'SetReceiveLibraryTimeoutInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([33, 159, 29, 53, 218, 156, 41, 192]),
    })
  ) as Serializer<
    SetReceiveLibraryTimeoutInstructionDataArgs,
    SetReceiveLibraryTimeoutInstructionData
  >;
}

// Args.
export type SetReceiveLibraryTimeoutInstructionArgs =
  SetReceiveLibraryTimeoutInstructionDataArgs;

// Instruction.
export function setReceiveLibraryTimeout(
  context: Pick<Context, 'programs'>,
  input: SetReceiveLibraryTimeoutInstructionAccounts &
    SetReceiveLibraryTimeoutInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false as boolean,
      value: input.signer ?? null,
    },
    oappRegistry: {
      index: 1,
      isWritable: false as boolean,
      value: input.oappRegistry ?? null,
    },
    receiveLibraryConfig: {
      index: 2,
      isWritable: true as boolean,
      value: input.receiveLibraryConfig ?? null,
    },
    messageLibInfo: {
      index: 3,
      isWritable: false as boolean,
      value: input.messageLibInfo ?? null,
    },
    eventAuthority: {
      index: 4,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 5,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetReceiveLibraryTimeoutInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSetReceiveLibraryTimeoutInstructionDataSerializer().serialize(
    resolvedArgs as SetReceiveLibraryTimeoutInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
