// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/skip.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type SkipInstructionAccounts = {
  /** The PDA of the OApp or delegate */
  signer: Signer;
  oappRegistry: PublicKey | Pda;
  nonce: PublicKey | Pda;
  pendingInboundNonce: PublicKey | Pda;
  /**
   * the payload hash needs to be initialized before it can be skipped and closed, in order to prevent someone
   * from skipping a payload hash that has been initialized and can be re-verified and executed after skipping
   */

  payloadHash: PublicKey | Pda;
  endpoint: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type SkipInstructionData = {
  discriminator: Uint8Array;
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: bigint;
};

export type SkipInstructionDataArgs = {
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: number | bigint;
};

export function getSkipInstructionDataSerializer(): Serializer<
  SkipInstructionDataArgs,
  SkipInstructionData
> {
  return mapSerializer<SkipInstructionDataArgs, any, SkipInstructionData>(
    struct<SkipInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['receiver', publicKeySerializer()],
        ['srcEid', u32()],
        ['sender', bytes({ size: 32 })],
        ['nonce', u64()],
      ],
      { description: 'SkipInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([154, 63, 181, 53, 19, 26, 117, 45]),
    })
  ) as Serializer<SkipInstructionDataArgs, SkipInstructionData>;
}

// Args.
export type SkipInstructionArgs = SkipInstructionDataArgs;

// Instruction.
export function skip(
  context: Pick<Context, 'programs'>,
  accounts: SkipInstructionAccounts,
  args: SkipInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.signer ?? null,
    },
    oappRegistry: {
      index: 1,
      isWritable: false as boolean,
      value: accounts.oappRegistry ?? null,
    },
    nonce: {
      index: 2,
      isWritable: true as boolean,
      value: accounts.nonce ?? null,
    },
    pendingInboundNonce: {
      index: 3,
      isWritable: true as boolean,
      value: accounts.pendingInboundNonce ?? null,
    },
    payloadHash: {
      index: 4,
      isWritable: true as boolean,
      value: accounts.payloadHash ?? null,
    },
    endpoint: {
      index: 5,
      isWritable: true as boolean,
      value: accounts.endpoint ?? null,
    },
    eventAuthority: {
      index: 6,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 7,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SkipInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSkipInstructionDataSerializer().serialize(
    resolvedArgs as SkipInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
