// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/transferAdmin.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type TransferAdminInstructionAccounts = {
  admin: Signer;
  endpoint: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type TransferAdminInstructionData = {
  discriminator: Uint8Array;
  admin: PublicKey;
};

export type TransferAdminInstructionDataArgs = { admin: PublicKey };

export function getTransferAdminInstructionDataSerializer(): Serializer<
  TransferAdminInstructionDataArgs,
  TransferAdminInstructionData
> {
  return mapSerializer<
    TransferAdminInstructionDataArgs,
    any,
    TransferAdminInstructionData
  >(
    struct<TransferAdminInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['admin', publicKeySerializer()],
      ],
      { description: 'TransferAdminInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([42, 242, 66, 106, 228, 10, 111, 156]),
    })
  ) as Serializer<
    TransferAdminInstructionDataArgs,
    TransferAdminInstructionData
  >;
}

// Args.
export type TransferAdminInstructionArgs = TransferAdminInstructionDataArgs;

// Instruction.
export function transferAdmin(
  context: Pick<Context, 'programs'>,
  accounts: TransferAdminInstructionAccounts,
  args: TransferAdminInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.admin ?? null,
    },
    endpoint: {
      index: 1,
      isWritable: true as boolean,
      value: accounts.endpoint ?? null,
    },
    eventAuthority: {
      index: 2,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 3,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: TransferAdminInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getTransferAdminInstructionDataSerializer().serialize(
    resolvedArgs as TransferAdminInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
