// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/instructions/verify.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type VerifyInstructionAccounts = {
  /** The PDA of the receive library. */
  receiveLibrary: Signer;
  receiveLibraryConfig: PublicKey | Pda;
  defaultReceiveLibraryConfig: PublicKey | Pda;
  nonce: PublicKey | Pda;
  pendingInboundNonce: PublicKey | Pda;
  payloadHash: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type VerifyInstructionData = {
  discriminator: Uint8Array;
  srcEid: number;
  sender: Uint8Array;
  receiver: PublicKey;
  nonce: bigint;
  payloadHash: Uint8Array;
};

export type VerifyInstructionDataArgs = {
  srcEid: number;
  sender: Uint8Array;
  receiver: PublicKey;
  nonce: number | bigint;
  payloadHash: Uint8Array;
};

export function getVerifyInstructionDataSerializer(): Serializer<
  VerifyInstructionDataArgs,
  VerifyInstructionData
> {
  return mapSerializer<VerifyInstructionDataArgs, any, VerifyInstructionData>(
    struct<VerifyInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['srcEid', u32()],
        ['sender', bytes({ size: 32 })],
        ['receiver', publicKeySerializer()],
        ['nonce', u64()],
        ['payloadHash', bytes({ size: 32 })],
      ],
      { description: 'VerifyInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([133, 161, 141, 48, 120, 198, 88, 150]),
    })
  ) as Serializer<VerifyInstructionDataArgs, VerifyInstructionData>;
}

// Args.
export type VerifyInstructionArgs = VerifyInstructionDataArgs;

// Instruction.
export function verify(
  context: Pick<Context, 'programs'>,
  accounts: VerifyInstructionAccounts,
  args: VerifyInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );

  // Accounts.
  const resolvedAccounts = {
    receiveLibrary: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.receiveLibrary ?? null,
    },
    receiveLibraryConfig: {
      index: 1,
      isWritable: false as boolean,
      value: accounts.receiveLibraryConfig ?? null,
    },
    defaultReceiveLibraryConfig: {
      index: 2,
      isWritable: false as boolean,
      value: accounts.defaultReceiveLibraryConfig ?? null,
    },
    nonce: {
      index: 3,
      isWritable: true as boolean,
      value: accounts.nonce ?? null,
    },
    pendingInboundNonce: {
      index: 4,
      isWritable: true as boolean,
      value: accounts.pendingInboundNonce ?? null,
    },
    payloadHash: {
      index: 5,
      isWritable: true as boolean,
      value: accounts.payloadHash ?? null,
    },
    eventAuthority: {
      index: 6,
      isWritable: false as boolean,
      value: accounts.eventAuthority ?? null,
    },
    program: {
      index: 7,
      isWritable: false as boolean,
      value: accounts.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: VerifyInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getVerifyInstructionDataSerializer().serialize(
    resolvedArgs as VerifyInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
