// @orderly.network/ui-notification/dist/index.js
'use strict';

var react = require('react');
var i18n = require('@orderly.network/i18n');
var ui = require('@orderly.network/ui');
var types = require('@orderly.network/types');
var jsxRuntime = require('react/jsx-runtime');
var hooks = require('@orderly.network/hooks');
var reactApp = require('@orderly.network/react-app');
var utc = require('@date-fns/utc');
var dateFns = require('date-fns');

// src/components/announcementCenter/announcementCenter.ui.tsx
var BattleIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Icon, { size: 18, viewBox: "0 0 18 18", ...props, children: [
    /* @__PURE__ */ jsxRuntime.jsx("defs", { children: /* @__PURE__ */ jsxRuntime.jsxs(
      "linearGradient",
      {
        id: "paint0_linear_555_7374",
        x1: "17.0157",
        y1: "9.26513",
        x2: "0.969849",
        y2: "9.26513",
        gradientUnits: "userSpaceOnUse",
        children: [
          /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
          /* @__PURE__ */ jsxRuntime.jsx("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
        ]
      }
    ) }),
    /* @__PURE__ */ jsxRuntime.jsx(
      "path",
      {
        fill: "url(#paint0_linear_555_7374)",
        fillOpacity: 1,
        d: "m1.5 1.5.75 3 4.843 4.56-1.808 1.919-.63-.633-1.06 1.06.665.665-1.88 1.996-.35-.346-1.06 1.06 2.25 2.25 1.06-1.06-.348-.349 1.998-1.878.665.662 1.06-1.06-.63-.63L9 10.852l1.969 1.854-.639.639 1.06 1.06.672-.67 1.998 1.877-.358.358 1.064 1.06 2.25-2.25-1.06-1.06-.34.34-1.881-1.998.656-.657-1.06-1.06-.625.624-1.799-1.91.003-.003L9 7.03 4.5 2.25zm15 0-3 .75-3.636 3.861 2.022 2.025L15.75 4.5z"
      }
    )
  ] });
};
var CampaignIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Icon, { size: 18, viewBox: "0 0 18 18", ...props, children: /* @__PURE__ */ jsxRuntime.jsx(
    "path",
    {
      fill: "currentcolor",
      fillOpacity: 1,
      d: "m1.5 1.5.75 3 4.843 4.56-1.808 1.919-.63-.633-1.06 1.06.665.665-1.88 1.996-.35-.346-1.06 1.06 2.25 2.25 1.06-1.06-.348-.349 1.998-1.878.665.662 1.06-1.06-.63-.63L9 10.852l1.969 1.854-.639.639 1.06 1.06.672-.67 1.998 1.877-.358.358 1.064 1.06 2.25-2.25-1.06-1.06-.34.34-1.881-1.998.656-.657-1.06-1.06-.625.624-1.799-1.91.003-.003L9 7.03 4.5 2.25zm15 0-3 .75-3.636 3.861 2.022 2.025L15.75 4.5z"
    }
  ) });
};
var ArrowRightShortIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Icon, { size: 18, viewBox: "0 0 18 18", ...props, children: [
    /* @__PURE__ */ jsxRuntime.jsx("defs", { children: /* @__PURE__ */ jsxRuntime.jsxs(
      "linearGradient",
      {
        id: "paint0_linear_555_7374",
        x1: "17.0157",
        y1: "9.26513",
        x2: "0.969849",
        y2: "9.26513",
        gradientUnits: "userSpaceOnUse",
        children: [
          /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
          /* @__PURE__ */ jsxRuntime.jsx("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
        ]
      }
    ) }),
    /* @__PURE__ */ jsxRuntime.jsx(
      "path",
      {
        fill: "url(#paint0_linear_555_7374)",
        fillOpacity: 1,
        d: "M4.509 8.995a.75.75 0 0 1 .75-.75h5.666L8.7 5.998l1.054-1.054 3.535 3.512c.147.146.22.343.22.54a.76.76 0 0 1-.22.537l-3.535 3.512L8.7 11.992l2.225-2.248H5.258a.75.75 0 0 1-.749-.75"
      }
    )
  ] });
};
var FundIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Icon, { size: 18, viewBox: "0 0 18 18", ...props, children: /* @__PURE__ */ jsxRuntime.jsx(
    "path",
    {
      fill: "currentcolor",
      fillOpacity: 0.8,
      d: "M15.694 7.087c-.788-.393-1.857-.618-2.925-.618-1.125 0-2.138.225-2.925.618-.056 0-.113.057-.113.057V3.769c0-.9-.731-1.575-1.462-1.913-.788-.394-1.857-.619-2.925-.619-1.069 0-2.25.225-3.038.62C1.575 2.193.844 2.868.844 3.768v10.519c0 .9.731 1.575 1.462 1.912.788.394 1.856.619 2.925.619s2.138-.225 2.925-.619c.281-.169.563-.337.788-.562.225.225.506.393.787.562.788.394 1.857.619 2.925.619 1.125 0 2.138-.225 2.925-.619.732-.394 1.463-1.012 1.463-1.912V9c.112-.9-.619-1.519-1.35-1.913m-7.369 7.144c0 .113-.112.45-.731.732-.563.28-1.407.506-2.363.506s-1.8-.169-2.362-.506c-.619-.282-.731-.62-.731-.732V13.5c.056 0 .112 0 .168.056.788.394 1.856.619 2.925.619s2.138-.225 2.925-.619c.056 0 .113-.056.113-.056v.731zm0-2.587c0 .112-.112.45-.731.731-.563.281-1.407.506-2.363.506s-1.8-.168-2.362-.506c-.619-.281-.731-.619-.731-.731v-.788c.056 0 .112.056.112.056.788.394 1.856.62 2.925.62s2.138-.226 2.925-.62c.056 0 .113-.056.113-.056v.788zm0-2.644c0 .112-.112.45-.731.731-.563.281-1.407.507-2.363.507s-1.8-.17-2.362-.507C2.25 9.45 2.138 9.112 2.138 9v-.788c.056 0 .112.057.112.057.788.393 1.856.618 2.925.618S7.313 8.662 8.1 8.27c.056 0 .113-.057.113-.057V9zm0-2.644c0 .113-.112.45-.731.731-.563.282-1.407.507-2.363.507s-1.8-.169-2.362-.507c-.619-.28-.731-.618-.731-.73v-.732c.056 0 .112 0 .168.056.788.394 1.856.619 2.925.619s2.138-.225 2.925-.619c.056 0 .113-.056.113-.056v.731zM7.594 4.5c-.563.281-1.407.506-2.363.506S3.488 4.781 2.925 4.5c-.619-.338-.731-.619-.731-.731 0-.113.112-.45.731-.732.563-.28 1.406-.506 2.363-.506.956 0 1.8.169 2.362.506.619.282.731.62.731.732-.056.112-.168.393-.787.731m8.212 9.731c0 .113-.112.45-.731.732-.562.28-1.406.506-2.362.506s-1.8-.169-2.363-.506c-.619-.282-.731-.62-.731-.732V13.5c.056 0 .112.056.112.056.788.394 1.857.619 2.925.619 1.125 0 2.138-.225 2.925-.619.056 0 .113-.056.113-.056v.731zm0-2.587c0 .112-.112.45-.731.731-.562.281-1.406.506-2.362.506s-1.8-.168-2.363-.506c-.619-.281-.731-.619-.731-.731v-.788c.056 0 .112.056.112.056.788.394 1.857.62 2.925.62 1.125 0 2.138-.226 2.925-.62.056 0 .113-.056.113-.056v.788zm-.675-1.913c-.562.281-1.406.507-2.362.507s-1.8-.17-2.363-.507c-.619-.281-.731-.619-.731-.731 0-.113.112-.45.731-.731.563-.282 1.406-.507 2.363-.507.956 0 1.8.17 2.362.507.619.281.732.618.732.731-.057.112-.17.45-.732.731"
    }
  ) });
};
var AnnouncementIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Icon, { size: 18, viewBox: "0 0 18 18", ...props, children: /* @__PURE__ */ jsxRuntime.jsx(
    "path",
    {
      fill: "currentcolor",
      fillOpacity: 0.8,
      d: "M11.53 2.255a.745.745 0 0 0-.82.165c-.525.524-1.188.959-1.947 1.284-.77.33-2.353.757-3.508.75-2.248-.015-3.756 1.259-3.756 3.726 0 2.163 1.228 3.42 2.994 3.703l.006 1.565a2.25 2.25 0 0 0 2.25 2.247c1.142 0 2.083-.883 2.228-1.988.007-.053.022-.26.022-1.01.674.315 1.232.812 1.711 1.29.473.472 1.29.13 1.29-.538 0-.678-.004-1.906-.004-3.13.884-.327 1.503-1.165 1.503-2.115s-.585-1.81-1.496-2.129c0-1.225-.004-2.439-.004-3.117a.76.76 0 0 0-.469-.703m1.805 1.335a.73.73 0 0 0-.445.351.75.75 0 0 0 .258 1.03 3.74 3.74 0 0 1 1.851 3.233 3.74 3.74 0 0 1-1.851 3.231.76.76 0 0 0-.282 1.03.754.754 0 0 0 1.032.258 5.24 5.24 0 0 0 2.601-4.52 5.24 5.24 0 0 0-2.601-4.519.74.74 0 0 0-.563-.094m-2.832.884c0 .694-.004 1.417-.004 2.231v2.997c0 .815.003 1.508.003 2.202-1.52-.896-3.085-1.294-4.501-1.421 0-1.144-.004-3.406-.004-4.584.147-.007.318-.026.565-.06a11.2 11.2 0 0 0 2.812-.773c.43-.184.75-.35 1.129-.592m-6.005 1.55-.001 4.34c-.976-.27-1.498-.959-1.498-2.184 0-1.21.47-1.928 1.499-2.155M6 11.998c.238.008 1.098.161 1.493.261l.007 1.19a.75.75 0 0 1-1.5 0z"
    }
  ) });
};
var SecurityIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Icon, { size: 18, viewBox: "0 0 18 18", ...props, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      "path",
      {
        d: "M6.563 7.678V6.603c0-1.153 1.09-2.088 2.437-2.088s2.438.935 2.438 2.088v1.075c.448 0 .812.311.812.696v3.479c0 .384-.364.696-.812.696H6.563c-.449 0-.813-.312-.813-.696v-3.48c0-.384.364-.695.813-.695m1.187-.006 1.431.006h1.07V6.564c0-.64-.504-1.104-1.25-1.104-.748 0-1.25.465-1.25 1.104zm1.956 2.14c0-.335-.317-.605-.706-.605-.39 0-.706.27-.706.604v.77c0 .335.316.605.706.605s.706-.27.706-.604z",
        fill: "currentcolor",
        fillOpacity: ".8"
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      "path",
      {
        d: "M2 3.607c.003 5.606.146 11.651 6.879 13.378.078.02.164.02.242 0C15.854 15.258 15.997 9.213 16 3.607c0-.217-.14-.41-.345-.477L9.31 1.05a1 1 0 0 0-.62 0L2.345 3.13A.5.5 0 0 0 2 3.607m2.326 7.63c-.707-1.954-.81-4.33-.824-6.902L9 2.532l5.498 1.803c-.014 2.573-.117 4.948-.824 6.901-.697 1.93-1.987 3.463-4.674 4.222-2.687-.759-3.977-2.293-4.674-4.222",
        fill: "currentcolor",
        fillOpacity: ".8"
      }
    )
  ] });
};
var AnnouncementItem = (props) => {
  const { t } = i18n.useTranslation();
  const Icon2 = react.useMemo(() => {
    switch (props.type) {
      case types.AnnouncementType.Campaign:
        return CampaignIcon;
      case types.AnnouncementType.Listing:
        return FundIcon;
      case types.AnnouncementType.Maintenance:
        return SecurityIcon;
      case types.AnnouncementType.Delisting:
      default:
        return AnnouncementIcon;
    }
  }, [props.type]);
  const title = react.useMemo(() => {
    switch (props.type) {
      case types.AnnouncementType.Campaign:
        return t("notification.campaign");
      case types.AnnouncementType.Delisting:
        return t("notification.delisting");
      case types.AnnouncementType.Listing:
        return t("notification.listing");
      case types.AnnouncementType.Maintenance:
        return t("notification.maintenance");
      case types.AnnouncementType.Vote:
        return t("notification.vote");
      default:
        return t("notification.general");
    }
  }, [props.type, t]);
  const action = react.useMemo(() => {
    if ((props.type === types.AnnouncementType.Campaign || props.type === types.AnnouncementType.Vote) && typeof props.url === "string" && props.url !== "" && typeof props.onItemClick === "function") {
      return /* @__PURE__ */ jsxRuntime.jsxs(
        ui.Flex,
        {
          gap: 1,
          itemAlign: "center",
          className: "oui-cursor-pointer",
          onClick: (event) => {
            event.stopPropagation();
            props.onItemClick(props.url);
          },
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text,
              {
                size: "xs",
                color: "buy",
                className: "oui-bg-clip-text oui-text-transparent oui-gradient-brand",
                children: t("notification.joinNow")
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(ArrowRightShortIcon, { size: 18 })
          ]
        }
      );
    }
    return null;
  }, [props.type, props.url, t]);
  const updateTime = react.useMemo(() => {
    if (props.type === types.AnnouncementType.Maintenance) {
      return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 36, children: t("notification.recentlyUpdated") });
    }
    return /* @__PURE__ */ jsxRuntime.jsx(
      ui.Text.formatted,
      {
        size: "2xs",
        intensity: 36,
        rule: "date",
        formatString: "yyyy-MM-dd HH:mm:ss",
        children: props.updatedTime
      }
    );
  }, [props.updatedTime, props.type, t]);
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      gap: 2,
      itemAlign: "start",
      className: ui.cn(
        "oui-px-2 oui-py-[6px] oui-text-base-contrast-80",
        !props.showDivider && "oui-rounded-md hover:oui-bg-base-6",
        !props.showDivider && props.expanded && "oui-bg-base-6",
        props.className
      ),
      onClick: () => {
        props.onExpandToggle?.();
      },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(Icon2, { color: "white", className: "oui-mt-3 oui-shrink-0" }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: "start", grow: true, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "xs", intensity: 80, weight: "bold", children: title }),
          updateTime,
          /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: "oui-grid oui-transition-all oui-duration-300 oui-ease-in-out",
              style: {
                gridTemplateRows: props.expanded ? "1fr" : "0fr"
              },
              children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-2 oui-overflow-hidden", children: [
                /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 80, as: "div", className: "oui-pt-2", children: props.message }),
                action
              ] })
            }
          )
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-pt-3", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.ChevronDownIcon,
          {
            color: "white",
            size: 18,
            className: `oui-transition-transform oui-duration-300 oui-ease-in-out ${props.expanded ? "oui-rotate-180" : "oui-rotate-0"}`
          }
        ) })
      ]
    }
  );
};
var AnnouncementContent = (props) => {
  const { dataSource, current, onExpandToggle, onItemClick } = props;
  const { t } = i18n.useTranslation();
  if (!Array.isArray(dataSource) || dataSource.length === 0) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-h-[160px] oui-items-center oui-justify-center", children: /* @__PURE__ */ jsxRuntime.jsx(
      ui.ExtensionSlot,
      {
        position: ui.ExtensionPositionEnum.EmptyDataIdentifier,
        title: t("notification.empty")
      }
    ) });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: ui.cn(
        "oui-flex oui-flex-col oui-space-y-1",
        props.showDivider && "[&>*:not(:first-child)]:oui-border-t [&>*:not(:first-child)]:oui-border-line-12  [&>*:not(:first-child)]:oui-pt-1"
      ),
      children: dataSource.map((item) => /* @__PURE__ */ jsxRuntime.jsx(
        AnnouncementItem,
        {
          url: item.url,
          onItemClick,
          message: item.message,
          updatedTime: item.updated_time ?? 0,
          expanded: current === item.announcement_id,
          type: item.type,
          showDivider: props.showDivider,
          onExpandToggle: () => {
            if (current === item.announcement_id) {
              onExpandToggle(null);
            } else {
              onExpandToggle(item.announcement_id);
            }
          }
        },
        item.announcement_id
      ))
    }
  );
};
var AnnouncementCenterUI = (props) => {
  const { t } = i18n.useTranslation();
  const [expanded, setExpanded] = react.useState(null);
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-px-5 oui-pt-4", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 80, weight: "bold", children: t("notification.title") }) }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.ScrollArea, { className: "oui-flex oui-h-[300px] oui-flex-col oui-space-y-1 oui-p-3", children: /* @__PURE__ */ jsxRuntime.jsx(
      AnnouncementContent,
      {
        dataSource: props.dataSource,
        current: expanded,
        onExpandToggle: setExpanded,
        onItemClick: props.onItemClick
      }
    ) })
  ] });
};
var CampaignContentCard = ({ message, coverImage, url, onItemClick }) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-5", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "bold", children: message }),
    /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: "oui-rounded-xl oui-bg-base-9 oui-bg-cover oui-bg-center oui-bg-no-repeat",
        style: {
          backgroundImage: `url(${coverImage})`,
          height: "100px"
        }
      }
    ),
    typeof url === "string" && url !== "" && typeof onItemClick === "function" && /* @__PURE__ */ jsxRuntime.jsxs(
      "button",
      {
        className: "oui-flex oui-items-center oui-gap-1",
        onClick: () => onItemClick(url),
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              size: "xs",
              color: "buy",
              className: "oui-bg-clip-text oui-text-transparent oui-gradient-brand",
              children: t("notification.joinNow")
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(ArrowRightShortIcon, { size: 18, color: "success" })
        ]
      }
    )
  ] });
};
var MaintenanceContentCard = ({ message, startTime, endTime }) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "xs", intensity: 54, children: t("notification.recentlyUpdated") }),
    message && /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 80, as: "div", className: "oui-mt-2", children: message })
  ] });
};
var DelistingContentCard = ({ message, updateTime }) => {
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Text.formatted,
      {
        rule: "date",
        intensity: 54,
        formatString: "yyyy-MM-dd HH:mm:ss",
        size: "xs",
        children: updateTime
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "bold", children: message })
  ] });
};
var ListingContentCard = ({ message, updateTime }) => {
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Text.formatted,
      {
        rule: "date",
        intensity: 54,
        formatString: "yyyy-MM-dd HH:mm:ss",
        size: "xs",
        children: updateTime
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "bold", children: message })
  ] });
};
var CommunityVoteContentCard = ({ message, url, onItemClick }) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-2", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "bold", children: message }),
    typeof url === "string" && url !== "" && typeof onItemClick === "function" && /* @__PURE__ */ jsxRuntime.jsxs(
      "button",
      {
        className: "oui-flex oui-items-center oui-gap-1",
        onClick: () => onItemClick(url),
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              size: "xs",
              color: "buy",
              className: "oui-bg-clip-text oui-text-transparent oui-gradient-brand",
              children: t("notification.joinNow")
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(ArrowRightShortIcon, { size: 18, color: "success" })
        ]
      }
    )
  ] });
};
var NotificationHeader = (props) => {
  const { t } = i18n.useTranslation();
  const { expanded } = props;
  const { type } = props.dataSource[props.current];
  const title = react.useMemo(() => {
    switch (type) {
      case types.AnnouncementType.Campaign:
        return /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text,
          {
            size: "sm",
            className: "oui-text-transparent oui-bg-clip-text oui-gradient-brand",
            children: t("notification.campaign")
          }
        );
      case types.AnnouncementType.Delisting:
        return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", children: t("notification.delistingTitle") });
      case types.AnnouncementType.Listing:
        return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", color: "buy", children: t("notification.listing") });
      case types.AnnouncementType.Maintenance:
        return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", color: "warning", children: t("notification.maintenanceTitle") });
      case types.AnnouncementType.Vote:
        return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", color: "inherit", children: t("notification.vote") });
      default:
        return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", color: "inherit", children: t("notification.generalTitle") });
    }
  }, [type, t]);
  const icon = react.useMemo(() => {
    switch (type) {
      case types.AnnouncementType.Campaign:
        return /* @__PURE__ */ jsxRuntime.jsx(BattleIcon, { color: "white" });
      case types.AnnouncementType.Listing:
        return /* @__PURE__ */ jsxRuntime.jsx(FundIcon, { color: "success" });
      case types.AnnouncementType.Maintenance:
        return /* @__PURE__ */ jsxRuntime.jsx(SecurityIcon, { color: "warning" });
      case types.AnnouncementType.Delisting:
      default:
        return /* @__PURE__ */ jsxRuntime.jsx(AnnouncementIcon, { color: "white" });
    }
  }, [type]);
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "center", justify: "between", className: "oui-px-4 oui-py-3", children: [
    /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "orderly-notification-header oui-flex oui-items-center oui-gap-2", children: [
      icon,
      title
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(
      "button",
      {
        onClick: props.onExpandToggle,
        className: "oui-transition-transform oui-duration-300",
        style: {
          transform: !expanded ? "rotate(180deg)" : "rotate(0deg)"
        },
        children: /* @__PURE__ */ jsxRuntime.jsx(ui.ChevronDownIcon, { size: 18, color: "white" })
      }
    )
  ] });
};
var NotificationFooter = (props) => {
  const { total, current, onCloseAll, onPrev, onNext } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      className: "orderly-notification-footer oui-px-4 oui-py-2",
      itemAlign: "center",
      justify: "between",
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            "button",
            {
              disabled: current - 1 < 0,
              onClick: onPrev,
              className: "oui-flex oui-size-[18px] oui-items-center oui-justify-center oui-rounded-full oui-bg-base-6 hover:oui-bg-base-5 disabled:oui-opacity-50",
              children: /* @__PURE__ */ jsxRuntime.jsx(ui.ChevronLeftIcon, { size: 14, color: "white" })
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { intensity: 54, size: "sm", children: [
            current + 1,
            "/",
            total
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(
            "button",
            {
              disabled: current + 1 >= total,
              onClick: onNext,
              className: "oui-flex oui-size-[18px] oui-items-center oui-justify-center oui-rounded-full oui-bg-base-6 hover:oui-bg-base-5 disabled:oui-opacity-50",
              children: /* @__PURE__ */ jsxRuntime.jsx(ui.ChevronRightIcon, { size: 14, color: "white" })
            }
          )
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx("button", { onClick: onCloseAll, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "xs", color: "primary", children: t("notification.closeAll", { total }) }) })
      ]
    }
  );
};
var NotificationContent = (props) => {
  const elements = react.useMemo(() => {
    return props.dataSource.map((message) => {
      const { type } = message;
      switch (type) {
        case types.AnnouncementType.Campaign:
          return /* @__PURE__ */ jsxRuntime.jsx(
            CampaignContentCard,
            {
              message: message.message,
              coverImage: message.coverImage ?? "",
              updateTime: message.updated_time ?? 0,
              url: message.url ?? "",
              onItemClick: props.onItemClick
            }
          );
        case types.AnnouncementType.Maintenance:
          return /* @__PURE__ */ jsxRuntime.jsx(
            MaintenanceContentCard,
            {
              message: message.message,
              startTime: message.startTime ?? 0,
              endTime: message.endTime ?? 0
            }
          );
        case types.AnnouncementType.Delisting:
          return /* @__PURE__ */ jsxRuntime.jsx(
            DelistingContentCard,
            {
              message: message.message,
              updateTime: message.updated_time ?? 0
            }
          );
        case types.AnnouncementType.Vote:
          return /* @__PURE__ */ jsxRuntime.jsx(
            CommunityVoteContentCard,
            {
              message: message.message,
              url: message.url ?? "",
              onItemClick: props.onItemClick
            }
          );
        default:
          return /* @__PURE__ */ jsxRuntime.jsx(
            ListingContentCard,
            {
              message: message.message,
              updateTime: message.updated_time ?? 0
            }
          );
      }
    });
  }, [props.dataSource]);
  return /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, { children: elements[props.current] });
};
var NotificationUI = (props) => {
  const [expanded, setExpanded] = react.useState(true);
  const [current, setCurrent] = react.useState(0);
  const len = react.useMemo(() => props.dataSource?.length ?? 0, [props.dataSource]);
  if (len === 0) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "orderly-notification oui-w-full ", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      NotificationHeader,
      {
        dataSource: props.dataSource ?? [],
        current,
        expanded,
        onExpandToggle: () => {
          setExpanded(!expanded);
        }
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { className: "oui-mx-4" }),
    /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: "oui-grid oui-transition-all oui-duration-300 oui-ease-in-out",
        style: {
          gridTemplateRows: expanded ? "1fr" : "0fr"
        },
        children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-min-h-0 oui-overflow-hidden ", children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-px-4 oui-py-3", children: /* @__PURE__ */ jsxRuntime.jsx(
          NotificationContent,
          {
            dataSource: props.dataSource ?? [],
            current,
            onItemClick: props.onItemClick
          }
        ) }) })
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      NotificationFooter,
      {
        total: len,
        current,
        onCloseAll: props.onClose ?? (() => {
        }),
        onPrev: () => {
          if (current - 1 < 0) {
            return;
          }
          setCurrent(current - 1);
        },
        onNext: () => {
          if (current + 1 >= len) {
            return;
          }
          setCurrent(current + 1);
        }
      }
    )
  ] });
};
var getTimeString = (timestamp) => {
  const date = dateFns.format(new utc.UTCDateMini(timestamp), "MMM dd");
  const time = dateFns.format(new utc.UTCDateMini(timestamp), "h:mm aa");
  return `${time} (UTC) on ${date}`;
};
var sortDataByUpdatedTime = (list) => {
  return list.sort((a, b) => {
    if (a.updated_time && b.updated_time) {
      return b.updated_time - a.updated_time;
    }
    return 0;
  });
};
var filterDuplicateArrayById = (list) => {
  const seenIds = /* @__PURE__ */ new Set();
  const newList = [];
  list.forEach((item) => {
    if (!seenIds.has(item.announcement_id)) {
      seenIds.add(item.announcement_id);
      newList.push(item);
    }
  });
  return newList;
};
var extractSymbolsFromMessage = (message) => {
  if (!message) return [];
  const matches = message.matchAll(/\$([A-Z0-9]+)/gi);
  const tickers = [...matches].map((m) => m[1].toUpperCase());
  return [...new Set(tickers)];
};
var isSymbolInMarketList = (symbol, marketList) => {
  const prefix = `PERP_${symbol}_`;
  return marketList.some((m) => m.symbol?.startsWith(prefix));
};
var shouldShowListingDelistingAnnouncement = (item, marketList) => {
  if (item.type !== types.AnnouncementType.Listing && item.type !== types.AnnouncementType.Delisting) {
    return true;
  }
  const symbols = extractSymbolsFromMessage(item.message ?? "");
  if (symbols.length === 0) {
    return true;
  }
  return symbols.every((s) => isSymbolInMarketList(s, marketList));
};

// src/hooks/useAnnouncement.ts
var maintentanceId = "-1";
var ORDERLY_ANNOUNCEMENT_KEY = "orderly_announcement";
var useAnnouncementData = () => {
  const { t } = i18n.useTranslation();
  const ws = hooks.useWS();
  const { dataAdapter } = hooks.useOrderlyContext();
  const marketList = hooks.useMarketList();
  const [announcementStore, setAnnouncementStore] = hooks.useLocalStorage(ORDERLY_ANNOUNCEMENT_KEY, {});
  const [tips, setTips] = react.useState([]);
  const [maintenances, setMaintenances] = react.useState([]);
  const [maintenanceDialogInfo, setMaintenanceDialogInfo] = react.useState();
  const { startTime, endTime, status, brokerName } = hooks.useMaintenanceStatus();
  const { data: announcements } = hooks.useQuery(
    `/v2/public/announcement`,
    {
      revalidateOnFocus: false,
      refreshInterval: 60 * 60 * 1e3,
      // refresh every 1 hour
      formatter: (data) => data
    }
  );
  const getMaintentTipsContent = (brokerName2, startDate, endDate) => t("maintenance.tips.description", { brokerName: brokerName2, startDate, endDate });
  const getMaintentDialogContent = (brokerName2, endDate) => t("maintenance.dialog.description", { brokerName: brokerName2, endDate });
  react.useEffect(() => {
    const startDate = startTime ? getTimeString(startTime) : "-";
    const endDate = endTime ? getTimeString(endTime) : "-";
    const dialogContent = status === hooks.MaintenanceStatus.Maintenance ? getMaintentDialogContent(brokerName, endDate) : void 0;
    setMaintenanceDialogInfo(dialogContent);
    if (startTime && endTime) {
      setMaintenances([
        {
          announcement_id: maintentanceId,
          type: types.AnnouncementType.Maintenance,
          startTime,
          endTime,
          updated_time: startTime,
          message: getMaintentTipsContent(brokerName, startDate, endDate)
        }
      ]);
    } else {
      setMaintenances([]);
    }
  }, [startTime, endTime, status, brokerName, t]);
  react.useEffect(() => {
    let list = [...maintenances ?? [], ...announcements?.rows ?? []];
    if (typeof dataAdapter?.announcementList === "function") {
      list = dataAdapter.announcementList(list);
    }
    list = list.filter(
      (item) => shouldShowListingDelistingAnnouncement(item, marketList)
    );
    const removedDuplicateList = filterDuplicateArrayById(list);
    const sortedList = sortDataByUpdatedTime(removedDuplicateList);
    setTips(sortedList);
  }, [maintenances, announcements, dataAdapter?.announcementList, marketList]);
  react.useEffect(() => {
    if (!announcements?.rows) {
      return;
    }
    const lastUpdateTime = announcements.last_updated_time ?? 0;
    const firstTipTime = tips[0]?.updated_time ?? 0;
    const updatedTime = Math.max(lastUpdateTime, firstTipTime);
    const closedTime = announcementStore.lastUpdateTime ?? 0;
    if (closedTime < updatedTime) {
      setAnnouncementStore({ show: true, lastUpdateTime: updatedTime });
    }
  }, [announcements, tips]);
  react.useEffect(() => {
    const unsubscribe = ws.subscribe("announcement", {
      onMessage(message) {
        if (message) {
          const newTip = {
            announcement_id: message.announcement_id,
            message: message.message,
            url: message.url,
            i18n: message.i18n,
            type: message.type,
            updated_time: message.updated_time
          };
          if (!shouldShowListingDelistingAnnouncement(newTip, marketList)) {
            return;
          }
          setTips((prev) => {
            const list = prev.filter(
              (item) => item.announcement_id !== message.announcement_id
            );
            return [...list, newTip];
          });
          setAnnouncementStore((prev) => ({ ...prev, show: true }));
        }
      }
    });
    return () => {
      unsubscribe?.();
    };
  }, [ws, marketList]);
  return {
    tips,
    maintenanceDialogInfo,
    announcementStore,
    setAnnouncementStore
  };
};
var useAnnouncement = (options) => {
  const { showAnnouncement, setShowAnnouncement } = reactApp.useAppContext();
  const {
    tips,
    maintenanceDialogInfo,
    announcementStore,
    setAnnouncementStore
  } = useAnnouncementData();
  const closeTips = () => {
    setAnnouncementStore((prev) => ({ ...prev, show: false }));
  };
  react.useEffect(() => {
    const len = tips.length;
    setShowAnnouncement(
      Boolean(len) && announcementStore.show && !options?.hideTips
    );
  }, [tips, announcementStore.show, options?.hideTips, setShowAnnouncement]);
  return {
    maintenanceDialogInfo,
    tips,
    closeTips,
    showAnnouncement
  };
};

// src/pages/announcementCenter.script.tsx
var useAnnouncementCenterScript = () => {
  const { tips } = useAnnouncement();
  const [current, setCurrent] = react.useState(null);
  return {
    dataSource: tips,
    current,
    setCurrent
  };
};
var AnnouncementCenterUI2 = (props) => {
  const { dataSource, current, setCurrent, onItemClick } = props;
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-m-1 oui-rounded-xl oui-bg-base-9 oui-p-2", children: /* @__PURE__ */ jsxRuntime.jsx(
    AnnouncementContent,
    {
      dataSource,
      current,
      onExpandToggle: setCurrent,
      onItemClick,
      showDivider: true
    }
  ) });
};
var AnnouncementCenterWidget = (props) => {
  const { dataSource, current, setCurrent } = useAnnouncementCenterScript();
  const onItemClick = react.useCallback(
    (url) => {
      if (!url) return;
      props.onRouteChange(url);
    },
    [props.onRouteChange]
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    AnnouncementCenterUI2,
    {
      dataSource,
      current,
      setCurrent,
      onItemClick
    }
  );
};
var AnnouncementCenterPage = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    AnnouncementCenterWidget,
    {
      onRouteChange: (url) => {
        if (!url) return;
        props.routerAdapter?.onRouteChange({
          href: url,
          name: url,
          target: "_blank"
        });
      }
    }
  );
};

exports.AnnouncementCenterPage = AnnouncementCenterPage;
exports.AnnouncementCenterUI = AnnouncementCenterUI;
exports.AnnouncementItem = AnnouncementItem;
exports.NotificationUI = NotificationUI;
exports.useAnnouncement = useAnnouncement;
