// @kpi/di/lib/index.js
const values = {};
export class Di {
    get(name, newInstance = false) {
        if (!this.has(name)) {
            throw new Error(`Di error: ${name} not found!`);
        }
        const value = values[name];
        if (value.func) {
            if (value.instance && !newInstance) {
                return value.instance;
            }
            const Func = value.func;
            value.instance = new (Func)();
            return value.instance;
        }
        else if (value.factory) {
            if (value.instance && !newInstance) {
                return value.instance;
            }
            value.instance = value.factory.call(this, [this]);
            return value.instance;
        }
        return value.value;
    }
    bindValue(name, value) {
        values[name] = {
            value
        };
        return this;
    }
    bindFactory(name, factory, lazy = true) {
        values[name] = {
            factory,
            instance: null
        };
        if (!lazy) {
            this.get(name);
        }
        return this;
    }
    bind(name, func, lazy = true) {
        values[name] = {
            func,
            instance: null
        };
        if (!lazy) {
            this.get(name);
        }
        return this;
    }
    remove(name) {
        delete values[name];
        return this;
    }
    has(name) {
        return !!values[name];
    }
    getAll() {
        return values;
    }
}
export default new Di();
