// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/errors/uln.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Program, ProgramError } from '@metaplex-foundation/umi';

type ProgramErrorConstructor = new (
  program: Program,
  cause?: Error
) => ProgramError;
const codeToErrorMap: Map<number, ProgramErrorConstructor> = new Map();
const nameToErrorMap: Map<string, ProgramErrorConstructor> = new Map();

/** Unauthorized */
export class UnauthorizedError extends ProgramError {
  override readonly name: string = 'Unauthorized';

  readonly code: number = 0x1770; // 6000

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1770, UnauthorizedError);
nameToErrorMap.set('Unauthorized', UnauthorizedError);

/** InvalidAmount */
export class InvalidAmountError extends ProgramError {
  override readonly name: string = 'InvalidAmount';

  readonly code: number = 0x1771; // 6001

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1771, InvalidAmountError);
nameToErrorMap.set('InvalidAmount', InvalidAmountError);

/** InvalidExecutor */
export class InvalidExecutorError extends ProgramError {
  override readonly name: string = 'InvalidExecutor';

  readonly code: number = 0x1772; // 6002

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1772, InvalidExecutorError);
nameToErrorMap.set('InvalidExecutor', InvalidExecutorError);

/** ZeroMessageSize */
export class ZeroMessageSizeError extends ProgramError {
  override readonly name: string = 'ZeroMessageSize';

  readonly code: number = 0x1773; // 6003

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1773, ZeroMessageSizeError);
nameToErrorMap.set('ZeroMessageSize', ZeroMessageSizeError);

/** InvalidRequiredDVNCount */
export class InvalidRequiredDVNCountError extends ProgramError {
  override readonly name: string = 'InvalidRequiredDVNCount';

  readonly code: number = 0x1774; // 6004

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1774, InvalidRequiredDVNCountError);
nameToErrorMap.set('InvalidRequiredDVNCount', InvalidRequiredDVNCountError);

/** InvalidOptionalDVNCount */
export class InvalidOptionalDVNCountError extends ProgramError {
  override readonly name: string = 'InvalidOptionalDVNCount';

  readonly code: number = 0x1775; // 6005

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1775, InvalidOptionalDVNCountError);
nameToErrorMap.set('InvalidOptionalDVNCount', InvalidOptionalDVNCountError);

/** InvalidOptionalDVNThreshold */
export class InvalidOptionalDVNThresholdError extends ProgramError {
  override readonly name: string = 'InvalidOptionalDVNThreshold';

  readonly code: number = 0x1776; // 6006

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1776, InvalidOptionalDVNThresholdError);
nameToErrorMap.set(
  'InvalidOptionalDVNThreshold',
  InvalidOptionalDVNThresholdError
);

/** InvalidConfirmations */
export class InvalidConfirmationsError extends ProgramError {
  override readonly name: string = 'InvalidConfirmations';

  readonly code: number = 0x1777; // 6007

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1777, InvalidConfirmationsError);
nameToErrorMap.set('InvalidConfirmations', InvalidConfirmationsError);

/** AtLeastOneDVN */
export class AtLeastOneDVNError extends ProgramError {
  override readonly name: string = 'AtLeastOneDVN';

  readonly code: number = 0x1778; // 6008

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1778, AtLeastOneDVNError);
nameToErrorMap.set('AtLeastOneDVN', AtLeastOneDVNError);

/** Unsorted */
export class UnsortedError extends ProgramError {
  override readonly name: string = 'Unsorted';

  readonly code: number = 0x1779; // 6009

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1779, UnsortedError);
nameToErrorMap.set('Unsorted', UnsortedError);

/** InvalidTreasuryFeeCap */
export class InvalidTreasuryFeeCapError extends ProgramError {
  override readonly name: string = 'InvalidTreasuryFeeCap';

  readonly code: number = 0x177a; // 6010

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177a, InvalidTreasuryFeeCapError);
nameToErrorMap.set('InvalidTreasuryFeeCap', InvalidTreasuryFeeCapError);

/** InvalidPacketVersion */
export class InvalidPacketVersionError extends ProgramError {
  override readonly name: string = 'InvalidPacketVersion';

  readonly code: number = 0x177b; // 6011

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177b, InvalidPacketVersionError);
nameToErrorMap.set('InvalidPacketVersion', InvalidPacketVersionError);

/** InvalidEid */
export class InvalidEidError extends ProgramError {
  override readonly name: string = 'InvalidEid';

  readonly code: number = 0x177c; // 6012

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177c, InvalidEidError);
nameToErrorMap.set('InvalidEid', InvalidEidError);

/** Verifying */
export class VerifyingError extends ProgramError {
  override readonly name: string = 'Verifying';

  readonly code: number = 0x177d; // 6013

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177d, VerifyingError);
nameToErrorMap.set('Verifying', VerifyingError);

/** InvalidWorkerId */
export class InvalidWorkerIdError extends ProgramError {
  override readonly name: string = 'InvalidWorkerId';

  readonly code: number = 0x177e; // 6014

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177e, InvalidWorkerIdError);
nameToErrorMap.set('InvalidWorkerId', InvalidWorkerIdError);

/** InvalidOptionType */
export class InvalidOptionTypeError extends ProgramError {
  override readonly name: string = 'InvalidOptionType';

  readonly code: number = 0x177f; // 6015

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177f, InvalidOptionTypeError);
nameToErrorMap.set('InvalidOptionType', InvalidOptionTypeError);

/** InvalidBps */
export class InvalidBpsError extends ProgramError {
  override readonly name: string = 'InvalidBps';

  readonly code: number = 0x1780; // 6016

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1780, InvalidBpsError);
nameToErrorMap.set('InvalidBps', InvalidBpsError);

/** ExceededMaxMessageSize */
export class ExceededMaxMessageSizeError extends ProgramError {
  override readonly name: string = 'ExceededMaxMessageSize';

  readonly code: number = 0x1781; // 6017

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1781, ExceededMaxMessageSizeError);
nameToErrorMap.set('ExceededMaxMessageSize', ExceededMaxMessageSizeError);

/** InvalidExecutorProgram */
export class InvalidExecutorProgramError extends ProgramError {
  override readonly name: string = 'InvalidExecutorProgram';

  readonly code: number = 0x1782; // 6018

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1782, InvalidExecutorProgramError);
nameToErrorMap.set('InvalidExecutorProgram', InvalidExecutorProgramError);

/** InvalidAccountLength */
export class InvalidAccountLengthError extends ProgramError {
  override readonly name: string = 'InvalidAccountLength';

  readonly code: number = 0x1783; // 6019

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1783, InvalidAccountLengthError);
nameToErrorMap.set('InvalidAccountLength', InvalidAccountLengthError);

/** InvalidDvnProgram */
export class InvalidDvnProgramError extends ProgramError {
  override readonly name: string = 'InvalidDvnProgram';

  readonly code: number = 0x1784; // 6020

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1784, InvalidDvnProgramError);
nameToErrorMap.set('InvalidDvnProgram', InvalidDvnProgramError);

/** InvalidDvn */
export class InvalidDvnError extends ProgramError {
  override readonly name: string = 'InvalidDvn';

  readonly code: number = 0x1785; // 6021

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1785, InvalidDvnError);
nameToErrorMap.set('InvalidDvn', InvalidDvnError);

/** LzTokenUnavailable */
export class LzTokenUnavailableError extends ProgramError {
  override readonly name: string = 'LzTokenUnavailable';

  readonly code: number = 0x1786; // 6022

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1786, LzTokenUnavailableError);
nameToErrorMap.set('LzTokenUnavailable', LzTokenUnavailableError);

/** InsufficientFee */
export class InsufficientFeeError extends ProgramError {
  override readonly name: string = 'InsufficientFee';

  readonly code: number = 0x1787; // 6023

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1787, InsufficientFeeError);
nameToErrorMap.set('InsufficientFee', InsufficientFeeError);

/** InvalidTreasury */
export class InvalidTreasuryError extends ProgramError {
  override readonly name: string = 'InvalidTreasury';

  readonly code: number = 0x1788; // 6024

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1788, InvalidTreasuryError);
nameToErrorMap.set('InvalidTreasury', InvalidTreasuryError);

/** InvalidLzTokenMint */
export class InvalidLzTokenMintError extends ProgramError {
  override readonly name: string = 'InvalidLzTokenMint';

  readonly code: number = 0x1789; // 6025

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1789, InvalidLzTokenMintError);
nameToErrorMap.set('InvalidLzTokenMint', InvalidLzTokenMintError);

/** InvalidConfigType */
export class InvalidConfigTypeError extends ProgramError {
  override readonly name: string = 'InvalidConfigType';

  readonly code: number = 0x178a; // 6026

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x178a, InvalidConfigTypeError);
nameToErrorMap.set('InvalidConfigType', InvalidConfigTypeError);

/** InvalidConfirmation */
export class InvalidConfirmationError extends ProgramError {
  override readonly name: string = 'InvalidConfirmation';

  readonly code: number = 0x178b; // 6027

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x178b, InvalidConfirmationError);
nameToErrorMap.set('InvalidConfirmation', InvalidConfirmationError);

/** InvalidType1Size */
export class InvalidType1SizeError extends ProgramError {
  override readonly name: string = 'InvalidType1Size';

  readonly code: number = 0x178c; // 6028

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x178c, InvalidType1SizeError);
nameToErrorMap.set('InvalidType1Size', InvalidType1SizeError);

/** InvalidType2Size */
export class InvalidType2SizeError extends ProgramError {
  override readonly name: string = 'InvalidType2Size';

  readonly code: number = 0x178d; // 6029

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x178d, InvalidType2SizeError);
nameToErrorMap.set('InvalidType2Size', InvalidType2SizeError);

/** ExceededU128 */
export class ExceededU128Error extends ProgramError {
  override readonly name: string = 'ExceededU128';

  readonly code: number = 0x178e; // 6030

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x178e, ExceededU128Error);
nameToErrorMap.set('ExceededU128', ExceededU128Error);

/** NonSigner */
export class NonSignerError extends ProgramError {
  override readonly name: string = 'NonSigner';

  readonly code: number = 0x178f; // 6031

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x178f, NonSignerError);
nameToErrorMap.set('NonSigner', NonSignerError);

/** InvalidPayer */
export class InvalidPayerError extends ProgramError {
  override readonly name: string = 'InvalidPayer';

  readonly code: number = 0x1790; // 6032

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1790, InvalidPayerError);
nameToErrorMap.set('InvalidPayer', InvalidPayerError);

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
export function getUlnErrorFromCode(
  code: number,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
export function getUlnErrorFromName(
  name: string,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}
