// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/accounts/Confirmations.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'

/**
 * Arguments used to create {@link Confirmations}
 * @category Accounts
 * @category generated
 */
export type ConfirmationsArgs = {
  value: beet.COption<beet.bignum>
  bump: number
}

export const confirmationsDiscriminator = [206, 57, 50, 8, 124, 133, 138, 112]
/**
 * Holds the data for the {@link Confirmations} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class Confirmations implements ConfirmationsArgs {
  private constructor(
    readonly value: beet.COption<beet.bignum>,
    readonly bump: number
  ) {}

  /**
   * Creates a {@link Confirmations} instance from the provided args.
   */
  static fromArgs(args: ConfirmationsArgs) {
    return new Confirmations(args.value, args.bump)
  }

  /**
   * Deserializes the {@link Confirmations} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [Confirmations, number] {
    return Confirmations.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link Confirmations} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<Confirmations> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find Confirmations account at ${address}`)
    }
    return Confirmations.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, confirmationsBeet)
  }

  /**
   * Deserializes the {@link Confirmations} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [Confirmations, number] {
    return confirmationsBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link Confirmations} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return confirmationsBeet.serialize({
      accountDiscriminator: confirmationsDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link Confirmations} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: ConfirmationsArgs) {
    const instance = Confirmations.fromArgs(args)
    return confirmationsBeet.toFixedFromValue({
      accountDiscriminator: confirmationsDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link Confirmations} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: ConfirmationsArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      Confirmations.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link Confirmations} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      value: this.value,
      bump: this.bump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const confirmationsBeet = new beet.FixableBeetStruct<
  Confirmations,
  ConfirmationsArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['value', beet.coption(beet.u64)],
    ['bump', beet.u8],
  ],
  Confirmations.fromArgs,
  'Confirmations'
)
