// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/accounts/ReceiveConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { UlnConfig, ulnConfigBeet } from '../types/UlnConfig'

/**
 * Arguments used to create {@link ReceiveConfig}
 * @category Accounts
 * @category generated
 */
export type ReceiveConfigArgs = {
  bump: number
  uln: UlnConfig
}

export const receiveConfigDiscriminator = [162, 159, 153, 188, 56, 65, 245, 58]
/**
 * Holds the data for the {@link ReceiveConfig} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class ReceiveConfig implements ReceiveConfigArgs {
  private constructor(readonly bump: number, readonly uln: UlnConfig) {}

  /**
   * Creates a {@link ReceiveConfig} instance from the provided args.
   */
  static fromArgs(args: ReceiveConfigArgs) {
    return new ReceiveConfig(args.bump, args.uln)
  }

  /**
   * Deserializes the {@link ReceiveConfig} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [ReceiveConfig, number] {
    return ReceiveConfig.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link ReceiveConfig} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<ReceiveConfig> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find ReceiveConfig account at ${address}`)
    }
    return ReceiveConfig.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, receiveConfigBeet)
  }

  /**
   * Deserializes the {@link ReceiveConfig} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [ReceiveConfig, number] {
    return receiveConfigBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link ReceiveConfig} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return receiveConfigBeet.serialize({
      accountDiscriminator: receiveConfigDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link ReceiveConfig} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: ReceiveConfigArgs) {
    const instance = ReceiveConfig.fromArgs(args)
    return receiveConfigBeet.toFixedFromValue({
      accountDiscriminator: receiveConfigDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link ReceiveConfig} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: ReceiveConfigArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      ReceiveConfig.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link ReceiveConfig} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      bump: this.bump,
      uln: this.uln,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const receiveConfigBeet = new beet.FixableBeetStruct<
  ReceiveConfig,
  ReceiveConfigArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['bump', beet.u8],
    ['uln', ulnConfigBeet],
  ],
  ReceiveConfig.fromArgs,
  'ReceiveConfig'
)
