// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/accounts/SendConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { UlnConfig, ulnConfigBeet } from '../types/UlnConfig'
import { ExecutorConfig, executorConfigBeet } from '../types/ExecutorConfig'

/**
 * Arguments used to create {@link SendConfig}
 * @category Accounts
 * @category generated
 */
export type SendConfigArgs = {
  bump: number
  uln: UlnConfig
  executor: ExecutorConfig
}

export const sendConfigDiscriminator = [91, 221, 135, 137, 131, 199, 174, 25]
/**
 * Holds the data for the {@link SendConfig} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class SendConfig implements SendConfigArgs {
  private constructor(
    readonly bump: number,
    readonly uln: UlnConfig,
    readonly executor: ExecutorConfig
  ) {}

  /**
   * Creates a {@link SendConfig} instance from the provided args.
   */
  static fromArgs(args: SendConfigArgs) {
    return new SendConfig(args.bump, args.uln, args.executor)
  }

  /**
   * Deserializes the {@link SendConfig} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [SendConfig, number] {
    return SendConfig.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link SendConfig} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<SendConfig> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find SendConfig account at ${address}`)
    }
    return SendConfig.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, sendConfigBeet)
  }

  /**
   * Deserializes the {@link SendConfig} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [SendConfig, number] {
    return sendConfigBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link SendConfig} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return sendConfigBeet.serialize({
      accountDiscriminator: sendConfigDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link SendConfig} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: SendConfigArgs) {
    const instance = SendConfig.fromArgs(args)
    return sendConfigBeet.toFixedFromValue({
      accountDiscriminator: sendConfigDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link SendConfig} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: SendConfigArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      SendConfig.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link SendConfig} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      bump: this.bump,
      uln: this.uln,
      executor: this.executor,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const sendConfigBeet = new beet.FixableBeetStruct<
  SendConfig,
  SendConfigArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['bump', beet.u8],
    ['uln', ulnConfigBeet],
    ['executor', executorConfigBeet],
  ],
  SendConfig.fromArgs,
  'SendConfig'
)
