// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/accounts/UlnSettings.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { Treasury, treasuryBeet } from '../types/Treasury'

/**
 * Arguments used to create {@link UlnSettings}
 * @category Accounts
 * @category generated
 */
export type UlnSettingsArgs = {
  eid: number
  endpoint: web3.PublicKey
  endpointProgram: web3.PublicKey
  bump: number
  admin: web3.PublicKey
  treasury: beet.COption<Treasury>
}

export const ulnSettingsDiscriminator = [226, 199, 100, 253, 38, 115, 167, 154]
/**
 * Holds the data for the {@link UlnSettings} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class UlnSettings implements UlnSettingsArgs {
  private constructor(
    readonly eid: number,
    readonly endpoint: web3.PublicKey,
    readonly endpointProgram: web3.PublicKey,
    readonly bump: number,
    readonly admin: web3.PublicKey,
    readonly treasury: beet.COption<Treasury>
  ) {}

  /**
   * Creates a {@link UlnSettings} instance from the provided args.
   */
  static fromArgs(args: UlnSettingsArgs) {
    return new UlnSettings(
      args.eid,
      args.endpoint,
      args.endpointProgram,
      args.bump,
      args.admin,
      args.treasury
    )
  }

  /**
   * Deserializes the {@link UlnSettings} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [UlnSettings, number] {
    return UlnSettings.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link UlnSettings} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<UlnSettings> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find UlnSettings account at ${address}`)
    }
    return UlnSettings.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, ulnSettingsBeet)
  }

  /**
   * Deserializes the {@link UlnSettings} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [UlnSettings, number] {
    return ulnSettingsBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link UlnSettings} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return ulnSettingsBeet.serialize({
      accountDiscriminator: ulnSettingsDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link UlnSettings} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: UlnSettingsArgs) {
    const instance = UlnSettings.fromArgs(args)
    return ulnSettingsBeet.toFixedFromValue({
      accountDiscriminator: ulnSettingsDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link UlnSettings} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: UlnSettingsArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      UlnSettings.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link UlnSettings} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      eid: this.eid,
      endpoint: this.endpoint.toBase58(),
      endpointProgram: this.endpointProgram.toBase58(),
      bump: this.bump,
      admin: this.admin.toBase58(),
      treasury: this.treasury,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const ulnSettingsBeet = new beet.FixableBeetStruct<
  UlnSettings,
  UlnSettingsArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['eid', beet.u32],
    ['endpoint', beetSolana.publicKey],
    ['endpointProgram', beetSolana.publicKey],
    ['bump', beet.u8],
    ['admin', beetSolana.publicKey],
    ['treasury', beet.coption(treasuryBeet)],
  ],
  UlnSettings.fromArgs,
  'UlnSettings'
)
