// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/commitVerification.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  CommitVerificationParams,
  commitVerificationParamsBeet,
} from '../types/CommitVerificationParams'

/**
 * @category Instructions
 * @category CommitVerification
 * @category generated
 */
export type CommitVerificationInstructionArgs = {
  params: CommitVerificationParams
}
/**
 * @category Instructions
 * @category CommitVerification
 * @category generated
 */
export const commitVerificationStruct = new beet.BeetArgsStruct<
  CommitVerificationInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', commitVerificationParamsBeet],
  ],
  'CommitVerificationInstructionArgs'
)
/**
 * Accounts required by the _commitVerification_ instruction
 *
 * @property [] receiveConfig
 * @property [] defaultReceiveConfig
 * @property [] uln
 * @category Instructions
 * @category CommitVerification
 * @category generated
 */
export type CommitVerificationInstructionAccounts = {
  receiveConfig: web3.PublicKey
  defaultReceiveConfig: web3.PublicKey
  uln: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const commitVerificationInstructionDiscriminator = [
  194, 120, 44, 252, 140, 215, 139, 103,
]

/**
 * Creates a _CommitVerification_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category CommitVerification
 * @category generated
 */
export function createCommitVerificationInstruction(
  accounts: CommitVerificationInstructionAccounts,
  args: CommitVerificationInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = commitVerificationStruct.serialize({
    instructionDiscriminator: commitVerificationInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.receiveConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _CommitVerification_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category CommitVerification
 * @category generated
 */
export function createCommitVerificationInstructionAccounts(
  accounts: CommitVerificationInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.receiveConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
