// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/initDefaultConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitDefaultConfigParams,
  initDefaultConfigParamsBeet,
} from '../types/InitDefaultConfigParams'

/**
 * @category Instructions
 * @category InitDefaultConfig
 * @category generated
 */
export type InitDefaultConfigInstructionArgs = {
  params: InitDefaultConfigParams
}
/**
 * @category Instructions
 * @category InitDefaultConfig
 * @category generated
 */
export const initDefaultConfigStruct = new beet.FixableBeetArgsStruct<
  InitDefaultConfigInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initDefaultConfigParamsBeet],
  ],
  'InitDefaultConfigInstructionArgs'
)
/**
 * Accounts required by the _initDefaultConfig_ instruction
 *
 * @property [_writable_, **signer**] admin
 * @property [] uln
 * @property [_writable_] sendConfig
 * @property [_writable_] receiveConfig
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category InitDefaultConfig
 * @category generated
 */
export type InitDefaultConfigInstructionAccounts = {
  admin: web3.PublicKey
  uln: web3.PublicKey
  sendConfig: web3.PublicKey
  receiveConfig: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initDefaultConfigInstructionDiscriminator = [
  98, 218, 197, 194, 173, 179, 112, 21,
]

/**
 * Creates a _InitDefaultConfig_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitDefaultConfig
 * @category generated
 */
export function createInitDefaultConfigInstruction(
  accounts: InitDefaultConfigInstructionAccounts,
  args: InitDefaultConfigInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initDefaultConfigStruct.serialize({
    instructionDiscriminator: initDefaultConfigInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitDefaultConfig_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitDefaultConfig
 * @category generated
 */
export function createInitDefaultConfigInstructionAccounts(
  accounts: InitDefaultConfigInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
