// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/initUln.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { InitUlnParams, initUlnParamsBeet } from '../types/InitUlnParams'

/**
 * @category Instructions
 * @category InitUln
 * @category generated
 */
export type InitUlnInstructionArgs = {
  params: InitUlnParams
}
/**
 * @category Instructions
 * @category InitUln
 * @category generated
 */
export const initUlnStruct = new beet.BeetArgsStruct<
  InitUlnInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initUlnParamsBeet],
  ],
  'InitUlnInstructionArgs'
)
/**
 * Accounts required by the _initUln_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [_writable_] uln
 * @category Instructions
 * @category InitUln
 * @category generated
 */
export type InitUlnInstructionAccounts = {
  payer: web3.PublicKey
  uln: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initUlnInstructionDiscriminator = [
  19, 215, 207, 92, 197, 112, 119, 240,
]

/**
 * Creates a _InitUln_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitUln
 * @category generated
 */
export function createInitUlnInstruction(
  accounts: InitUlnInstructionAccounts,
  args: InitUlnInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initUlnStruct.serialize({
    instructionDiscriminator: initUlnInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitUln_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitUln
 * @category generated
 */
export function createInitUlnInstructionAccounts(
  accounts: InitUlnInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
