// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/quote.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { QuoteParams, quoteParamsBeet } from '../types/QuoteParams'

/**
 * @category Instructions
 * @category Quote
 * @category generated
 */
export type QuoteInstructionArgs = {
  params: QuoteParams
}
/**
 * @category Instructions
 * @category Quote
 * @category generated
 */
export const quoteStruct = new beet.FixableBeetArgsStruct<
  QuoteInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', quoteParamsBeet],
  ],
  'QuoteInstructionArgs'
)
/**
 * Accounts required by the _quote_ instruction
 *
 * @property [**signer**] endpoint
 * @property [] uln
 * @property [] sendConfig
 * @property [] defaultSendConfig
 * @category Instructions
 * @category Quote
 * @category generated
 */
export type QuoteInstructionAccounts = {
  endpoint: web3.PublicKey
  uln: web3.PublicKey
  sendConfig: web3.PublicKey
  defaultSendConfig: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const quoteInstructionDiscriminator = [
  149, 42, 109, 247, 134, 146, 213, 123,
]

/**
 * Creates a _Quote_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Quote
 * @category generated
 */
export function createQuoteInstruction(
  accounts: QuoteInstructionAccounts,
  args: QuoteInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = quoteStruct.serialize({
    instructionDiscriminator: quoteInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultSendConfig,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Quote_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Quote
 * @category generated
 */
export function createQuoteInstructionAccounts(
  accounts: QuoteInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultSendConfig,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
