// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/setDefaultConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetDefaultConfigParams,
  setDefaultConfigParamsBeet,
} from '../types/SetDefaultConfigParams'

/**
 * @category Instructions
 * @category SetDefaultConfig
 * @category generated
 */
export type SetDefaultConfigInstructionArgs = {
  params: SetDefaultConfigParams
}
/**
 * @category Instructions
 * @category SetDefaultConfig
 * @category generated
 */
export const setDefaultConfigStruct = new beet.FixableBeetArgsStruct<
  SetDefaultConfigInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setDefaultConfigParamsBeet],
  ],
  'SetDefaultConfigInstructionArgs'
)
/**
 * Accounts required by the _setDefaultConfig_ instruction
 *
 * @property [**signer**] admin
 * @property [] uln
 * @property [_writable_] sendConfig
 * @property [_writable_] receiveConfig
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetDefaultConfig
 * @category generated
 */
export type SetDefaultConfigInstructionAccounts = {
  admin: web3.PublicKey
  uln: web3.PublicKey
  sendConfig: web3.PublicKey
  receiveConfig: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setDefaultConfigInstructionDiscriminator = [
  111, 228, 17, 75, 5, 76, 213, 169,
]

/**
 * Creates a _SetDefaultConfig_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetDefaultConfig
 * @category generated
 */
export function createSetDefaultConfigInstruction(
  accounts: SetDefaultConfigInstructionAccounts,
  args: SetDefaultConfigInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setDefaultConfigStruct.serialize({
    instructionDiscriminator: setDefaultConfigInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetDefaultConfig_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetDefaultConfig
 * @category generated
 */
export function createSetDefaultConfigInstructionAccounts(
  accounts: SetDefaultConfigInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
