// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/setTreasury.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetTreasuryParams,
  setTreasuryParamsBeet,
} from '../types/SetTreasuryParams'

/**
 * @category Instructions
 * @category SetTreasury
 * @category generated
 */
export type SetTreasuryInstructionArgs = {
  params: SetTreasuryParams
}
/**
 * @category Instructions
 * @category SetTreasury
 * @category generated
 */
export const setTreasuryStruct = new beet.FixableBeetArgsStruct<
  SetTreasuryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setTreasuryParamsBeet],
  ],
  'SetTreasuryInstructionArgs'
)
/**
 * Accounts required by the _setTreasury_ instruction
 *
 * @property [**signer**] signer
 * @property [_writable_] uln
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetTreasury
 * @category generated
 */
export type SetTreasuryInstructionAccounts = {
  signer: web3.PublicKey
  uln: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setTreasuryInstructionDiscriminator = [
  57, 97, 196, 95, 195, 206, 106, 136,
]

/**
 * Creates a _SetTreasury_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetTreasury
 * @category generated
 */
export function createSetTreasuryInstruction(
  accounts: SetTreasuryInstructionAccounts,
  args: SetTreasuryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setTreasuryStruct.serialize({
    instructionDiscriminator: setTreasuryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetTreasury_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetTreasury
 * @category generated
 */
export function createSetTreasuryInstructionAccounts(
  accounts: SetTreasuryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
