// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/instructions/withdrawRent.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  WithdrawRentParams,
  withdrawRentParamsBeet,
} from '../types/WithdrawRentParams'

/**
 * @category Instructions
 * @category WithdrawRent
 * @category generated
 */
export type WithdrawRentInstructionArgs = {
  params: WithdrawRentParams
}
/**
 * @category Instructions
 * @category WithdrawRent
 * @category generated
 */
export const withdrawRentStruct = new beet.BeetArgsStruct<
  WithdrawRentInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', withdrawRentParamsBeet],
  ],
  'WithdrawRentInstructionArgs'
)
/**
 * Accounts required by the _withdrawRent_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] uln
 * @property [_writable_] receiver
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category WithdrawRent
 * @category generated
 */
export type WithdrawRentInstructionAccounts = {
  admin: web3.PublicKey
  uln: web3.PublicKey
  receiver: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const withdrawRentInstructionDiscriminator = [
  226, 7, 41, 158, 173, 111, 192, 107,
]

/**
 * Creates a _WithdrawRent_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category WithdrawRent
 * @category generated
 */
export function createWithdrawRentInstruction(
  accounts: WithdrawRentInstructionAccounts,
  args: WithdrawRentInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = withdrawRentStruct.serialize({
    instructionDiscriminator: withdrawRentInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiver,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _WithdrawRent_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category WithdrawRent
 * @category generated
 */
export function createWithdrawRentInstructionAccounts(
  accounts: WithdrawRentInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.uln,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiver,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
