// @orderly.network/utils/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Decimal: () => decimal_default,
  camelCaseToUnderscoreCase: () => camelCaseToUnderscoreCase,
  capitalizeString: () => capitalizeString,
  checkIsNaN: () => checkIsNaN,
  commify: () => commify,
  commifyOptional: () => commifyOptional,
  cutNumber: () => cutNumber,
  dayjs: () => import_dayjs.default,
  findLongestCommonSubString: () => findLongestCommonSubString,
  formatNum: () => formatNumWithNamespace,
  formatSymbol: () => formatSymbol,
  formatWithPrecision: () => formatWithPrecision,
  getBBOType: () => getBBOType,
  getGlobalObject: () => getGlobalObject,
  getPrecisionByNumber: () => getPrecisionByNumber,
  getTPSLDirection: () => getTPSLDirection,
  getTimestamp: () => getTimestamp,
  getTrailingStopPrice: () => getTrailingStopPrice,
  hex2int: () => hex2int,
  int2hex: () => int2hex,
  isSolana: () => isSolana,
  isTestnet: () => isTestnet,
  numberToHumanStyle: () => numberToHumanStyle,
  optimizeSymbolDisplay: () => optimizeSymbolDisplay,
  parseChainIdToNumber: () => parseChainIdToNumber,
  parseNumStr: () => parseNumStr,
  praseChainId: () => praseChainId,
  praseChainIdToNumber: () => praseChainIdToNumber,
  removeTrailingZeros: () => removeTrailingZeros,
  subtractDaysFromCurrentDate: () => subtractDaysFromCurrentDate,
  timeConvertString: () => timeConvertString,
  timestampToString: () => timestampToString,
  toNonExponential: () => toNonExponential,
  todpIfNeed: () => todpIfNeed,
  transSymbolformString: () => transSymbolformString,
  windowGuard: () => windowGuard,
  zero: () => zero
});
module.exports = __toCommonJS(index_exports);

// src/decimal.ts
var import_decimal = __toESM(require("decimal.js-light"));
import_decimal.default.set({ rounding: import_decimal.default.ROUND_DOWN });
var decimal_default = import_decimal.default;
var cutNumber = (num, lenght) => {
};
var zero = new import_decimal.default(0);
var commifyOptional = (num, options) => {
  if (typeof num === "string" && Number.isNaN(Number(num))) {
    return (options == null ? void 0 : options.fallback) || "--";
  }
  const prefix = (options == null ? void 0 : options.prefix) || "";
  if (typeof num === "undefined") {
    return prefix + ((options == null ? void 0 : options.fallback) || "--");
  }
  const value = commify(num, options == null ? void 0 : options.fix);
  if (options && options.padEnd && options.fix) {
    const fillString = (options == null ? void 0 : options.fillString) || "0";
    const hasDecimal = value.includes(".");
    const list = value.split(".");
    if (hasDecimal) {
      return prefix + list[0] + "." + list[1].padEnd(options.fix, fillString);
    }
    return prefix + list[0] + "." + "".padEnd(options.fix, fillString);
  }
  return prefix + value;
};
var THOUSANDS_REGEXP = /\B(?=(\d{3})+(?!\d))/g;
var commify = (num, fix) => {
  const str = `${num}`;
  const parts = str.split(".");
  const numberPart = parts[0];
  const decimalPart = parts[1];
  const endsWithPoint = str.endsWith(".") && str.length > 1;
  const result = numberPart.replace(THOUSANDS_REGEXP, ",") + (decimalPart ? "." + decimalPart.substring(0, fix || decimalPart.length) : endsWithPoint ? "." : "");
  if (fix === 0 && result.includes(".")) {
    return result.substring(0, result.indexOf("."));
  }
  return result;
};
var SCIENTIFICNOTATION_REGEX = /\d(?:\.(\d*))?e([+-]\d+)/;
var toNonExponential = (num) => {
  const m = num.toExponential().match(SCIENTIFICNOTATION_REGEX);
  if (!Array.isArray(m)) {
    return num;
  }
  return num.toFixed(
    Math.max(0, (m[1] || "").length - m[2])
  );
};
var getPrecisionByNumber = (num) => {
  num = toNonExponential(Number(num));
  const parts = num.toString().split(".");
  return parts[1] ? parts[1].length : 0;
};
function numberToHumanStyle(number, decimalPlaces = 2, options) {
  const { padding } = options || {};
  const abbreviations = ["", "K", "M", "B", "T"];
  let index = 0;
  while (number >= 1e3 && index < abbreviations.length - 1) {
    number /= 1e3;
    index++;
  }
  const roundedNumber = new import_decimal.default(number).toFixed(decimalPlaces, import_decimal.default.ROUND_DOWN).toString().replace(/\.0+$/, "");
  return `${roundedNumber}${abbreviations[index]}`;
}
function parseNumStr(str) {
  const value = str.toString();
  const cleanedStr = value.replace(/,/g, "");
  const numberPart = new import_decimal.default(cleanedStr);
  const unitPart = cleanedStr.slice(-1);
  if (Number.isNaN(numberPart.toNumber())) {
    return void 0;
  }
  let result;
  switch (unitPart) {
    case "k":
    case "K":
      result = numberPart.mul(1e3);
      break;
    case "m":
    case "M":
      result = numberPart.mul(1e6);
      break;
    case "b":
    case "B":
      result = numberPart.mul(1e9);
      break;
    case "t":
    case "T":
      result = numberPart.mul(1e12);
      break;
    default:
      result = numberPart;
  }
  return result;
}
var SCIENTIFICNOTATIONPATTERN = /^[-+]?[0-9]+(\.[0-9]+)?[eE][-+]?[0-9]+$/;
var TRAILINGZERODECIMAL_REGEX = /(\.[0-9]*[1-9])0+$/;
var removeTrailingZeros = (value, fixedCount = 16) => {
  const text = `${value}`;
  const isScientific = SCIENTIFICNOTATIONPATTERN.test(text);
  if (!value.toString().includes(".") && !isScientific) {
    return `${value}`;
  }
  const formattedNumber = new import_decimal.default(value).toFixed(fixedCount).replace(TRAILINGZERODECIMAL_REGEX, "$1");
  return formattedNumber;
};
var todpIfNeed = (value, dp) => {
  if (value === void 0 || value === "") {
    return value;
  }
  if (typeof value === "number") {
    value = value.toString();
  }
  if (value.endsWith(".")) {
    return value;
  }
  const numbers = value.split(".");
  if (numbers.length === 1) {
    return value;
  }
  if (numbers[1].length <= dp || !numbers[1]) {
    return value;
  }
  return `${numbers[0]}.${numbers[1].substring(0, dp)}`;
};
var checkIsNaN = (value) => {
  if (value === void 0 || value === "" || value === null) {
    return true;
  }
  return Number.isNaN(Number(value));
};
var formatWithPrecision = (num, precision = 2) => {
  const numDecimal = new import_decimal.default(num || 0);
  if (numDecimal.gte(1)) {
    return numDecimal.todp(precision, import_decimal.default.ROUND_DOWN).toNumber();
  }
  return numDecimal.toSignificantDigits(precision, import_decimal.default.ROUND_DOWN).toFixed();
};

// src/dateTime.ts
var timeConvertString = (time) => {
  time /= 1e3;
  const h = Math.floor(time / 3600);
  const m = Math.floor(time / 60 % 60);
  const s = Math.floor(time % 60);
  return [h, m, s];
};
var timestampToString = (timestamp) => {
  const date = new Date(timestamp);
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");
  const seconds = String(date.getSeconds()).padStart(2, "0");
  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
};
function subtractDaysFromCurrentDate(days, startDate) {
  const currentDate = startDate || /* @__PURE__ */ new Date();
  const resultDate = new Date(currentDate);
  resultDate.setDate(currentDate.getDate() - days);
  return resultDate;
}

// src/chain.ts
var import_types = require("@orderly.network/types");
var hex2int = (chainId) => parseInt(chainId);
var int2hex = (chainId) => `0x${chainId.toString(16)}`;
var praseChainId = (chainId) => {
  if (typeof chainId === "string") return hex2int(chainId);
  return chainId;
};
var praseChainIdToNumber = (chainId) => {
  if (typeof chainId === "string" && chainId.startsWith("0x") && /^[a-f0-9]+$/iu.test(chainId.slice(2)))
    return hex2int(chainId);
  return chainId;
};
var parseChainIdToNumber = praseChainIdToNumber;
var isTestnet = (chainId) => {
  const testnetIds = [
    import_types.ARBITRUM_TESTNET_CHAINID,
    import_types.SOLANA_TESTNET_CHAINID,
    import_types.STORY_TESTNET_CHAINID,
    import_types.MONAD_TESTNET_CHAINID,
    import_types.ABSTRACT_TESTNET_CHAINID,
    import_types.BSC_TESTNET_CHAINID
  ];
  return testnetIds.includes(chainId);
};
var isSolana = (chainId) => {
  return [import_types.SOLANA_TESTNET_CHAINID, import_types.SOLANA_MAINNET_CHAINID].includes(chainId);
};

// src/string.ts
function capitalizeString(str) {
  const lowercaseStr = str.toLowerCase();
  const capitalizedStr = lowercaseStr.charAt(0).toUpperCase() + lowercaseStr.slice(1);
  return capitalizedStr;
}
function transSymbolformString(input) {
  const parts = input.split("_");
  if (parts.length !== 3) {
    throw new Error("Invalid string format");
  }
  const [first, second, third] = parts;
  if (!first.startsWith("PERP")) {
    throw new Error("Invalid string format");
  }
  const result = `${second}-${first}`;
  return result;
}
function camelCaseToUnderscoreCase(str) {
  return str.replace(/([a-z])([A-Z])/g, "$1_$2").toLowerCase();
}
var findLongestCommonSubString = (str1, str2) => {
  for (let index = 0; index < str1.length; index++) {
    const ele1 = str1.at(index);
    const ele2 = str2.at(index);
    if (ele1 === ele2) {
      continue;
    }
    return index;
  }
  return -1;
};

// src/window.ts
var windowGuard = (cb) => {
  if (typeof window !== "undefined") {
    cb();
  }
};
var getGlobalObject = () => {
  if (typeof globalThis !== "undefined") {
    return globalThis;
  }
  if (typeof self !== "undefined") {
    return self;
  }
  if (typeof window !== "undefined") {
    return window;
  }
  if (typeof global !== "undefined") {
    return global;
  }
  throw new Error("cannot find the global object");
};
var getTimestamp = () => {
  var _a;
  if (typeof window !== "undefined") {
    const timeOffset = (_a = getGlobalObject()) == null ? void 0 : _a.__ORDERLY_timestamp_offset;
    if (typeof timeOffset === "number") {
      return Date.now() + (timeOffset || 0);
    }
  }
  return Date.now();
};

// src/index.ts
var import_dayjs = __toESM(require("dayjs"));

// src/symbol.ts
function formatSymbol(symbol, formatString = "base-type") {
  if (!symbol) {
    return "";
  }
  const arr = symbol.split("_");
  const type = arr[0];
  const base = arr[1];
  const quote = arr[2];
  return formatString.replace("type", type).replace("base", base).replace("quote", quote);
}
function optimizeSymbolDisplay(symbol, decimalPlaces = 0) {
  if (!symbol) {
    return "";
  }
  const numberRegex = /^(\d+(?:\.\d+)?)/;
  const match = symbol.match(numberRegex);
  if (!match) {
    return symbol;
  }
  const numberPart = match[1];
  const textPart = symbol.slice(numberPart.length);
  const numericValue = parseFloat(numberPart);
  if (numericValue < 1e3) {
    return symbol;
  }
  const abbreviatedNumber = numberToHumanStyle(numericValue, decimalPlaces);
  return abbreviatedNumber + textPart;
}

// src/order.ts
var import_types2 = require("@orderly.network/types");
function getTrailingStopPrice(order) {
  const { side, extreme_price, callback_value, callback_rate } = order;
  const isBuy = side === import_types2.OrderSide.BUY;
  if (!extreme_price) {
    return 0;
  }
  if (callback_value) {
    return isBuy ? new decimal_default(extreme_price).plus(callback_value).toNumber() : new decimal_default(extreme_price).minus(callback_value).toNumber();
  }
  if (callback_rate) {
    return isBuy ? new decimal_default(extreme_price).mul(new decimal_default(1).plus(callback_rate)).toNumber() : new decimal_default(extreme_price).mul(new decimal_default(1).minus(callback_rate)).toNumber();
  }
  return 0;
}
function getBBOType(options) {
  const { type, side, level } = options;
  if (type === import_types2.OrderType.ASK) {
    if (level === import_types2.OrderLevel.ONE) {
      return side === import_types2.OrderSide.BUY ? import_types2.BBOOrderType.COUNTERPARTY1 : import_types2.BBOOrderType.QUEUE1;
    }
    if (level === import_types2.OrderLevel.FIVE) {
      return side === import_types2.OrderSide.BUY ? import_types2.BBOOrderType.COUNTERPARTY5 : import_types2.BBOOrderType.QUEUE5;
    }
  }
  if (type === import_types2.OrderType.BID) {
    if (level === import_types2.OrderLevel.ONE) {
      return side === import_types2.OrderSide.BUY ? import_types2.BBOOrderType.QUEUE1 : import_types2.BBOOrderType.COUNTERPARTY1;
    }
    if (level === import_types2.OrderLevel.FIVE) {
      return side === import_types2.OrderSide.BUY ? import_types2.BBOOrderType.QUEUE5 : import_types2.BBOOrderType.COUNTERPARTY5;
    }
  }
}
function getTPSLDirection(inputs) {
  const { side, type, closePrice, orderPrice } = inputs;
  let direction = 1;
  if (side === import_types2.OrderSide.BUY) {
    if (type === "tp") {
      direction = closePrice >= orderPrice ? 1 : -1;
    } else {
      direction = closePrice < orderPrice ? -1 : 1;
    }
  }
  if (side === import_types2.OrderSide.SELL) {
    if (type === "tp") {
      direction = closePrice <= orderPrice ? 1 : -1;
    } else {
      direction = closePrice > orderPrice ? -1 : 1;
    }
  }
  return direction;
}

// src/formatNum.ts
function formatNum(type, dp = 2, num, rm) {
  const decimalNum = parseToDecimal(num);
  if (!decimalNum) {
    return void 0;
  }
  const isMoreThanZero = decimalNum.greaterThan(0);
  switch (type) {
    case 0 /* pnl */:
    case 2 /* roi */:
      const innerRm = rm != null ? rm : isMoreThanZero ? decimal_default.ROUND_DOWN : decimal_default.ROUND_UP;
      return format(decimalNum, dp, innerRm);
    case 1 /* notional */:
      return format(decimalNum, dp, rm != null ? rm : decimal_default.ROUND_DOWN);
    case 3 /* assetValue */:
      return format(decimalNum, dp, rm != null ? rm : decimal_default.ROUND_DOWN);
    case 4 /* collateral */:
      return format(decimalNum, dp, rm != null ? rm : decimal_default.ROUND_DOWN);
  }
}
function format(num, dp, rm) {
  return num.toDecimalPlaces(dp, rm);
}
function parseToDecimal(num) {
  try {
    if (num instanceof decimal_default) {
      return num;
    }
    if (!num) {
      return void 0;
    }
    if (typeof num === "number") {
      return new decimal_default(num);
    }
    if (typeof num === "string") {
      if (!num.trim()) {
        return void 0;
      }
      return new decimal_default(num);
    }
    return num;
  } catch (error) {
    return void 0;
  }
}
var formatNumWithNamespace = formatNum;
formatNumWithNamespace.pnl = (num) => {
  return formatNum(0 /* pnl */, 2, num);
};
formatNumWithNamespace.notional = (num) => {
  return formatNum(1 /* notional */, 2, num);
};
formatNumWithNamespace.roi = (num, dp) => {
  return formatNum(2 /* roi */, dp != null ? dp : 4, num);
};
formatNumWithNamespace.assetValue = (num) => {
  return formatNum(3 /* assetValue */, 2, num);
};
formatNumWithNamespace.collateral = (num) => {
  return formatNum(4 /* collateral */, 2, num);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Decimal,
  camelCaseToUnderscoreCase,
  capitalizeString,
  checkIsNaN,
  commify,
  commifyOptional,
  cutNumber,
  dayjs,
  findLongestCommonSubString,
  formatNum,
  formatSymbol,
  formatWithPrecision,
  getBBOType,
  getGlobalObject,
  getPrecisionByNumber,
  getTPSLDirection,
  getTimestamp,
  getTrailingStopPrice,
  hex2int,
  int2hex,
  isSolana,
  isTestnet,
  numberToHumanStyle,
  optimizeSymbolDisplay,
  parseChainIdToNumber,
  parseNumStr,
  praseChainId,
  praseChainIdToNumber,
  removeTrailingZeros,
  subtractDaysFromCurrentDate,
  timeConvertString,
  timestampToString,
  toNonExponential,
  todpIfNeed,
  transSymbolformString,
  windowGuard,
  zero
});
