// @layerzerolabs/oft-v2-solana-sdk/dist/index.mjs
import { hexlify } from '@ethersproject/bytes';
import { deserializeAccount, publicKey as publicKey$1, assertAccountExists, gpaBuilder, ProgramError, isPda, isSigner, transactionBuilder, publicKeyBytes, createNullRpc, some, none, defaultPublicKey } from '@metaplex-foundation/umi';
import { createDefaultProgramRepository } from '@metaplex-foundation/umi-program-repository';
import { fromWeb3JsPublicKey, toWeb3JsInstruction, toWeb3JsPublicKey } from '@metaplex-foundation/umi-web3js-adapters';
import { PublicKey, ComputeBudgetProgram } from '@solana/web3.js';
import { EndpointProgram, UlnProgram, SimpleMessageLibProgram, EventPDA, simulateWeb3JsTransaction, EndpointPDA, UlnPDA } from '@layerzerolabs/lz-solana-sdk-v2/umi';
import { mapSerializer, struct, bytes, publicKey, u32, bool, u64, string, scalarEnum, dataEnum, tuple, option, u16, array, u8, Endian } from '@metaplex-foundation/umi/serializers';
import { createWeb3JsEddsa } from '@metaplex-foundation/umi-eddsa-web3js';
import { LZ_RECEIVE_TYPES_SEED } from '@layerzerolabs/lz-solana-sdk-v2';

var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/oft302.ts
var oft302_exports = {};
__export(oft302_exports, {
  accounts: () => accounts_exports,
  createOFTProgramRepo: () => createOFTProgramRepo,
  errors: () => errors_exports,
  getDelegate: () => getDelegate,
  getEndpointConfig: () => getEndpointConfig,
  getEnforcedOptions: () => getEnforcedOptions,
  getPeerAddress: () => getPeerAddress,
  initConfig: () => initConfig,
  initOAppNonce: () => initOAppNonce,
  initOft: () => initOft2,
  initReceiveLibrary: () => initReceiveLibrary,
  initSendLibrary: () => initSendLibrary,
  instructions: () => instructions_exports,
  programs: () => programs_exports,
  quote: () => quote,
  quoteOft: () => quoteOft2,
  send: () => send2,
  setConfig: () => setConfig,
  setOFTConfig: () => setOFTConfig,
  setPeerConfig: () => setPeerConfig2,
  setReceiveLibrary: () => setReceiveLibrary,
  setSendLibrary: () => setSendLibrary,
  shared: () => shared_exports,
  types: () => types_exports,
  withdrawFee: () => withdrawFee2
});
var TOKEN_PROGRAM_ID = new PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
new PublicKey("TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb");
new PublicKey("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
new PublicKey("So11111111111111111111111111111111111111112");
new PublicKey("9pan9bMn5HatX4EJdBwg9VgCa7Uz5HL8N1m5D3NdXejP");

// src/consts.ts
var OFT_DECIMALS = 6;
var OFT_SEED = "OFT";
var PEER_SEED = "Peer";

// src/generated/oft302/accounts/index.ts
var accounts_exports = {};
__export(accounts_exports, {
  deserializeLzReceiveTypesAccounts: () => deserializeLzReceiveTypesAccounts,
  deserializeOFTStore: () => deserializeOFTStore,
  deserializePeerConfig: () => deserializePeerConfig,
  fetchAllLzReceiveTypesAccounts: () => fetchAllLzReceiveTypesAccounts,
  fetchAllOFTStore: () => fetchAllOFTStore,
  fetchAllPeerConfig: () => fetchAllPeerConfig,
  fetchLzReceiveTypesAccounts: () => fetchLzReceiveTypesAccounts,
  fetchOFTStore: () => fetchOFTStore,
  fetchPeerConfig: () => fetchPeerConfig,
  getLzReceiveTypesAccountsAccountDataSerializer: () => getLzReceiveTypesAccountsAccountDataSerializer,
  getLzReceiveTypesAccountsGpaBuilder: () => getLzReceiveTypesAccountsGpaBuilder,
  getLzReceiveTypesAccountsSize: () => getLzReceiveTypesAccountsSize,
  getOFTStoreAccountDataSerializer: () => getOFTStoreAccountDataSerializer,
  getOFTStoreGpaBuilder: () => getOFTStoreGpaBuilder,
  getPeerConfigAccountDataSerializer: () => getPeerConfigAccountDataSerializer,
  getPeerConfigGpaBuilder: () => getPeerConfigGpaBuilder,
  safeFetchAllLzReceiveTypesAccounts: () => safeFetchAllLzReceiveTypesAccounts,
  safeFetchAllOFTStore: () => safeFetchAllOFTStore,
  safeFetchAllPeerConfig: () => safeFetchAllPeerConfig,
  safeFetchLzReceiveTypesAccounts: () => safeFetchLzReceiveTypesAccounts,
  safeFetchOFTStore: () => safeFetchOFTStore,
  safeFetchPeerConfig: () => safeFetchPeerConfig
});
function getLzReceiveTypesAccountsAccountDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["oftStore", publicKey()],
        ["tokenMint", publicKey()]
      ],
      { description: "LzReceiveTypesAccountsAccountData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([248, 87, 167, 117, 5, 251, 21, 126])
    })
  );
}
function deserializeLzReceiveTypesAccounts(rawAccount) {
  return deserializeAccount(
    rawAccount,
    getLzReceiveTypesAccountsAccountDataSerializer()
  );
}
async function fetchLzReceiveTypesAccounts(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  assertAccountExists(maybeAccount, "LzReceiveTypesAccounts");
  return deserializeLzReceiveTypesAccounts(maybeAccount);
}
async function safeFetchLzReceiveTypesAccounts(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  return maybeAccount.exists ? deserializeLzReceiveTypesAccounts(maybeAccount) : null;
}
async function fetchAllLzReceiveTypesAccounts(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, "LzReceiveTypesAccounts");
    return deserializeLzReceiveTypesAccounts(maybeAccount);
  });
}
async function safeFetchAllLzReceiveTypesAccounts(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.filter((maybeAccount) => maybeAccount.exists).map(
    (maybeAccount) => deserializeLzReceiveTypesAccounts(maybeAccount)
  );
}
function getLzReceiveTypesAccountsGpaBuilder(context) {
  const programId = context.programs.getPublicKey("oft", "");
  return gpaBuilder(context, programId).registerFields({
    discriminator: [0, bytes({ size: 8 })],
    oftStore: [8, publicKey()],
    tokenMint: [40, publicKey()]
  }).deserializeUsing(
    (account) => deserializeLzReceiveTypesAccounts(account)
  ).whereField(
    "discriminator",
    new Uint8Array([248, 87, 167, 117, 5, 251, 21, 126])
  );
}
function getLzReceiveTypesAccountsSize() {
  return 72;
}

// src/generated/oft302/types/index.ts
var types_exports = {};
__export(types_exports, {
  OFTType: () => OFTType,
  getEnforcedOptionsSerializer: () => getEnforcedOptionsSerializer,
  getLzAccountSerializer: () => getLzAccountSerializer,
  getLzReceiveParamsSerializer: () => getLzReceiveParamsSerializer,
  getMessagingFeeSerializer: () => getMessagingFeeSerializer,
  getOFTFeeDetailSerializer: () => getOFTFeeDetailSerializer,
  getOFTLimitsSerializer: () => getOFTLimitsSerializer,
  getOFTReceiptSerializer: () => getOFTReceiptSerializer,
  getOFTTypeSerializer: () => getOFTTypeSerializer,
  getPeerConfigParamSerializer: () => getPeerConfigParamSerializer,
  getQuoteOFTResultSerializer: () => getQuoteOFTResultSerializer,
  getRateLimitParamsSerializer: () => getRateLimitParamsSerializer,
  getRateLimiterSerializer: () => getRateLimiterSerializer,
  getSetOFTConfigParamsSerializer: () => getSetOFTConfigParamsSerializer,
  getVersionSerializer: () => getVersionSerializer,
  isPeerConfigParam: () => isPeerConfigParam,
  isSetOFTConfigParams: () => isSetOFTConfigParams,
  peerConfigParam: () => peerConfigParam,
  setOFTConfigParams: () => setOFTConfigParams
});
function getEnforcedOptionsSerializer() {
  return struct(
    [
      ["send", bytes({ size: u32() })],
      ["sendAndCall", bytes({ size: u32() })]
    ],
    { description: "EnforcedOptions" }
  );
}
function getLzAccountSerializer() {
  return struct(
    [
      ["pubkey", publicKey()],
      ["isSigner", bool()],
      ["isWritable", bool()]
    ],
    { description: "LzAccount" }
  );
}
function getLzReceiveParamsSerializer() {
  return struct(
    [
      ["srcEid", u32()],
      ["sender", bytes({ size: 32 })],
      ["nonce", u64()],
      ["guid", bytes({ size: 32 })],
      ["message", bytes({ size: u32() })],
      ["extraData", bytes({ size: u32() })]
    ],
    { description: "LzReceiveParams" }
  );
}
function getMessagingFeeSerializer() {
  return struct(
    [
      ["nativeFee", u64()],
      ["lzTokenFee", u64()]
    ],
    { description: "MessagingFee" }
  );
}
function getOFTFeeDetailSerializer() {
  return struct(
    [
      ["feeAmountLd", u64()],
      ["description", string()]
    ],
    { description: "OFTFeeDetail" }
  );
}
function getOFTLimitsSerializer() {
  return struct(
    [
      ["minAmountLd", u64()],
      ["maxAmountLd", u64()]
    ],
    { description: "OFTLimits" }
  );
}
function getOFTReceiptSerializer() {
  return struct(
    [
      ["amountSentLd", u64()],
      ["amountReceivedLd", u64()]
    ],
    { description: "OFTReceipt" }
  );
}
var OFTType = /* @__PURE__ */ ((OFTType7) => {
  OFTType7[OFTType7["Native"] = 0] = "Native";
  OFTType7[OFTType7["Adapter"] = 1] = "Adapter";
  return OFTType7;
})(OFTType || {});
function getOFTTypeSerializer() {
  return scalarEnum(OFTType, { description: "OFTType" });
}
function getPeerConfigParamSerializer() {
  return dataEnum(
    [
      [
        "PeerAddress",
        struct([
          ["fields", tuple([bytes({ size: 32 })])]
        ])
      ],
      [
        "FeeBps",
        struct([
          ["fields", tuple([option(u16())])]
        ])
      ],
      [
        "EnforcedOptions",
        struct([
          ["send", bytes({ size: u32() })],
          ["sendAndCall", bytes({ size: u32() })]
        ])
      ],
      [
        "OutboundRateLimit",
        struct([
          ["fields", tuple([option(getRateLimitParamsSerializer())])]
        ])
      ],
      [
        "InboundRateLimit",
        struct([
          ["fields", tuple([option(getRateLimitParamsSerializer())])]
        ])
      ]
    ],
    { description: "PeerConfigParam" }
  );
}
function peerConfigParam(kind, data) {
  return Array.isArray(data) ? { __kind: kind, fields: data } : { __kind: kind, ...data ?? {} };
}
function isPeerConfigParam(kind, value) {
  return value.__kind === kind;
}
function getQuoteOFTResultSerializer() {
  return struct(
    [
      ["oftLimits", getOFTLimitsSerializer()],
      ["oftFeeDetails", array(getOFTFeeDetailSerializer())],
      ["oftReceipt", getOFTReceiptSerializer()]
    ],
    { description: "QuoteOFTResult" }
  );
}
function getRateLimiterSerializer() {
  return struct(
    [
      ["capacity", u64()],
      ["tokens", u64()],
      ["refillPerSecond", u64()],
      ["lastRefillTime", u64()]
    ],
    { description: "RateLimiter" }
  );
}
function getRateLimitParamsSerializer() {
  return struct(
    [
      ["refillPerSecond", option(u64())],
      ["capacity", option(u64())]
    ],
    { description: "RateLimitParams" }
  );
}
function getSetOFTConfigParamsSerializer() {
  return dataEnum(
    [
      [
        "Admin",
        struct([
          ["fields", tuple([publicKey()])]
        ])
      ],
      [
        "Delegate",
        struct([
          ["fields", tuple([publicKey()])]
        ])
      ],
      [
        "DefaultFee",
        struct([
          ["fields", tuple([u16()])]
        ])
      ],
      [
        "Paused",
        struct([
          ["fields", tuple([bool()])]
        ])
      ],
      [
        "Pauser",
        struct([
          ["fields", tuple([option(publicKey())])]
        ])
      ],
      [
        "Unpauser",
        struct([
          ["fields", tuple([option(publicKey())])]
        ])
      ]
    ],
    { description: "SetOFTConfigParams" }
  );
}
function setOFTConfigParams(kind, data) {
  return Array.isArray(data) ? { __kind: kind, fields: data } : { __kind: kind, ...data ?? {} };
}
function isSetOFTConfigParams(kind, value) {
  return value.__kind === kind;
}
function getVersionSerializer() {
  return struct(
    [
      ["interface", u64()],
      ["message", u64()]
    ],
    { description: "Version" }
  );
}

// src/generated/oft302/accounts/oFTStore.ts
function getOFTStoreAccountDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["oftType", getOFTTypeSerializer()],
        ["ld2sdRate", u64()],
        ["tokenMint", publicKey()],
        ["tokenEscrow", publicKey()],
        ["endpointProgram", publicKey()],
        ["bump", u8()],
        ["tvlLd", u64()],
        ["admin", publicKey()],
        ["defaultFeeBps", u16()],
        ["paused", bool()],
        ["pauser", option(publicKey())],
        ["unpauser", option(publicKey())]
      ],
      { description: "OFTStoreAccountData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([195, 215, 104, 134, 185, 195, 240, 114])
    })
  );
}
function deserializeOFTStore(rawAccount) {
  return deserializeAccount(rawAccount, getOFTStoreAccountDataSerializer());
}
async function fetchOFTStore(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  assertAccountExists(maybeAccount, "OFTStore");
  return deserializeOFTStore(maybeAccount);
}
async function safeFetchOFTStore(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  return maybeAccount.exists ? deserializeOFTStore(maybeAccount) : null;
}
async function fetchAllOFTStore(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, "OFTStore");
    return deserializeOFTStore(maybeAccount);
  });
}
async function safeFetchAllOFTStore(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.filter((maybeAccount) => maybeAccount.exists).map((maybeAccount) => deserializeOFTStore(maybeAccount));
}
function getOFTStoreGpaBuilder(context) {
  const programId = context.programs.getPublicKey("oft", "");
  return gpaBuilder(context, programId).registerFields({
    discriminator: [0, bytes({ size: 8 })],
    oftType: [8, getOFTTypeSerializer()],
    ld2sdRate: [9, u64()],
    tokenMint: [17, publicKey()],
    tokenEscrow: [49, publicKey()],
    endpointProgram: [81, publicKey()],
    bump: [113, u8()],
    tvlLd: [114, u64()],
    admin: [122, publicKey()],
    defaultFeeBps: [154, u16()],
    paused: [156, bool()],
    pauser: [157, option(publicKey())],
    unpauser: [null, option(publicKey())]
  }).deserializeUsing((account) => deserializeOFTStore(account)).whereField(
    "discriminator",
    new Uint8Array([195, 215, 104, 134, 185, 195, 240, 114])
  );
}
function getPeerConfigAccountDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["peerAddress", bytes({ size: 32 })],
        ["enforcedOptions", getEnforcedOptionsSerializer()],
        ["outboundRateLimiter", option(getRateLimiterSerializer())],
        ["inboundRateLimiter", option(getRateLimiterSerializer())],
        ["feeBps", option(u16())],
        ["bump", u8()]
      ],
      { description: "PeerConfigAccountData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([181, 157, 86, 198, 33, 193, 94, 203])
    })
  );
}
function deserializePeerConfig(rawAccount) {
  return deserializeAccount(rawAccount, getPeerConfigAccountDataSerializer());
}
async function fetchPeerConfig(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  assertAccountExists(maybeAccount, "PeerConfig");
  return deserializePeerConfig(maybeAccount);
}
async function safeFetchPeerConfig(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  return maybeAccount.exists ? deserializePeerConfig(maybeAccount) : null;
}
async function fetchAllPeerConfig(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, "PeerConfig");
    return deserializePeerConfig(maybeAccount);
  });
}
async function safeFetchAllPeerConfig(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.filter((maybeAccount) => maybeAccount.exists).map((maybeAccount) => deserializePeerConfig(maybeAccount));
}
function getPeerConfigGpaBuilder(context) {
  const programId = context.programs.getPublicKey("oft", "");
  return gpaBuilder(context, programId).registerFields({
    discriminator: [0, bytes({ size: 8 })],
    peerAddress: [8, bytes({ size: 32 })],
    enforcedOptions: [40, getEnforcedOptionsSerializer()],
    outboundRateLimiter: [null, option(getRateLimiterSerializer())],
    inboundRateLimiter: [null, option(getRateLimiterSerializer())],
    feeBps: [null, option(u16())],
    bump: [null, u8()]
  }).deserializeUsing((account) => deserializePeerConfig(account)).whereField(
    "discriminator",
    new Uint8Array([181, 157, 86, 198, 33, 193, 94, 203])
  );
}

// src/generated/oft302/errors/index.ts
var errors_exports = {};
__export(errors_exports, {
  InvalidDecimalsError: () => InvalidDecimalsError,
  InvalidFeeError: () => InvalidFeeError,
  InvalidMintAuthorityError: () => InvalidMintAuthorityError,
  InvalidSenderError: () => InvalidSenderError,
  InvalidTokenDestError: () => InvalidTokenDestError,
  PausedError: () => PausedError,
  RateLimitExceededError: () => RateLimitExceededError,
  SlippageExceededError: () => SlippageExceededError,
  UnauthorizedError: () => UnauthorizedError,
  getOftErrorFromCode: () => getOftErrorFromCode,
  getOftErrorFromName: () => getOftErrorFromName
});
var codeToErrorMap = /* @__PURE__ */ new Map();
var nameToErrorMap = /* @__PURE__ */ new Map();
var UnauthorizedError = class extends ProgramError {
  // 6000
  constructor(program, cause) {
    super("", program, cause);
    this.name = "Unauthorized";
    this.code = 6e3;
  }
};
codeToErrorMap.set(6e3, UnauthorizedError);
nameToErrorMap.set("Unauthorized", UnauthorizedError);
var InvalidSenderError = class extends ProgramError {
  // 6001
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidSender";
    this.code = 6001;
  }
};
codeToErrorMap.set(6001, InvalidSenderError);
nameToErrorMap.set("InvalidSender", InvalidSenderError);
var InvalidDecimalsError = class extends ProgramError {
  // 6002
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidDecimals";
    this.code = 6002;
  }
};
codeToErrorMap.set(6002, InvalidDecimalsError);
nameToErrorMap.set("InvalidDecimals", InvalidDecimalsError);
var SlippageExceededError = class extends ProgramError {
  // 6003
  constructor(program, cause) {
    super("", program, cause);
    this.name = "SlippageExceeded";
    this.code = 6003;
  }
};
codeToErrorMap.set(6003, SlippageExceededError);
nameToErrorMap.set("SlippageExceeded", SlippageExceededError);
var InvalidTokenDestError = class extends ProgramError {
  // 6004
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidTokenDest";
    this.code = 6004;
  }
};
codeToErrorMap.set(6004, InvalidTokenDestError);
nameToErrorMap.set("InvalidTokenDest", InvalidTokenDestError);
var RateLimitExceededError = class extends ProgramError {
  // 6005
  constructor(program, cause) {
    super("", program, cause);
    this.name = "RateLimitExceeded";
    this.code = 6005;
  }
};
codeToErrorMap.set(6005, RateLimitExceededError);
nameToErrorMap.set("RateLimitExceeded", RateLimitExceededError);
var InvalidFeeError = class extends ProgramError {
  // 6006
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidFee";
    this.code = 6006;
  }
};
codeToErrorMap.set(6006, InvalidFeeError);
nameToErrorMap.set("InvalidFee", InvalidFeeError);
var InvalidMintAuthorityError = class extends ProgramError {
  // 6007
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidMintAuthority";
    this.code = 6007;
  }
};
codeToErrorMap.set(6007, InvalidMintAuthorityError);
nameToErrorMap.set("InvalidMintAuthority", InvalidMintAuthorityError);
var PausedError = class extends ProgramError {
  // 6008
  constructor(program, cause) {
    super("", program, cause);
    this.name = "Paused";
    this.code = 6008;
  }
};
codeToErrorMap.set(6008, PausedError);
nameToErrorMap.set("Paused", PausedError);
function getOftErrorFromCode(code, program, cause) {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}
function getOftErrorFromName(name, program, cause) {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}

// src/generated/oft302/instructions/index.ts
var instructions_exports = {};
__export(instructions_exports, {
  getInitOftInstructionDataSerializer: () => getInitOftInstructionDataSerializer,
  getLzReceiveInstructionDataSerializer: () => getLzReceiveInstructionDataSerializer,
  getLzReceiveTypesInstructionDataSerializer: () => getLzReceiveTypesInstructionDataSerializer,
  getOftVersionInstructionDataSerializer: () => getOftVersionInstructionDataSerializer,
  getQuoteOftInstructionDataSerializer: () => getQuoteOftInstructionDataSerializer,
  getQuoteSendInstructionDataSerializer: () => getQuoteSendInstructionDataSerializer,
  getSendInstructionDataSerializer: () => getSendInstructionDataSerializer,
  getSetOftConfigInstructionDataSerializer: () => getSetOftConfigInstructionDataSerializer,
  getSetPauseInstructionDataSerializer: () => getSetPauseInstructionDataSerializer,
  getSetPeerConfigInstructionDataSerializer: () => getSetPeerConfigInstructionDataSerializer,
  getWithdrawFeeInstructionDataSerializer: () => getWithdrawFeeInstructionDataSerializer,
  initOft: () => initOft,
  lzReceive: () => lzReceive,
  lzReceiveTypes: () => lzReceiveTypes,
  oftVersion: () => oftVersion,
  quoteOft: () => quoteOft,
  quoteSend: () => quoteSend,
  send: () => send,
  setOftConfig: () => setOftConfig,
  setPause: () => setPause,
  setPeerConfig: () => setPeerConfig,
  withdrawFee: () => withdrawFee
});

// src/generated/oft302/shared/index.ts
var shared_exports = {};
__export(shared_exports, {
  expectPda: () => expectPda,
  expectPublicKey: () => expectPublicKey,
  expectSome: () => expectSome,
  getAccountMetasAndSigners: () => getAccountMetasAndSigners
});
function expectSome(value) {
  if (value == null) {
    throw new Error("Expected a value but received null or undefined.");
  }
  return value;
}
function expectPublicKey(value) {
  if (!value) {
    throw new Error("Expected a PublicKey.");
  }
  return publicKey$1(value, false);
}
function expectPda(value) {
  if (!value || !Array.isArray(value) || !isPda(value)) {
    throw new Error("Expected a PDA.");
  }
  return value;
}
function getAccountMetasAndSigners(accounts, optionalAccountStrategy, programId) {
  const keys = [];
  const signers = [];
  accounts.forEach((account) => {
    if (!account.value) {
      if (optionalAccountStrategy === "omitted") return;
      keys.push({ pubkey: programId, isSigner: false, isWritable: false });
      return;
    }
    if (isSigner(account.value)) {
      signers.push(account.value);
    }
    keys.push({
      pubkey: publicKey$1(account.value, false),
      isSigner: isSigner(account.value),
      isWritable: account.isWritable
    });
  });
  return [keys, signers];
}

// src/generated/oft302/instructions/initOft.ts
function getInitOftInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["oftType", getOFTTypeSerializer()],
        ["admin", publicKey()],
        ["sharedDecimals", u8()],
        ["endpointProgram", option(publicKey())]
      ],
      { description: "InitOftInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([182, 169, 147, 16, 201, 45, 76, 23])
    })
  );
}
function initOft(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true,
      value: input.payer ?? null
    },
    oftStore: {
      index: 1,
      isWritable: true,
      value: input.oftStore ?? null
    },
    lzReceiveTypesAccounts: {
      index: 2,
      isWritable: true,
      value: input.lzReceiveTypesAccounts ?? null
    },
    tokenMint: {
      index: 3,
      isWritable: false,
      value: input.tokenMint ?? null
    },
    tokenEscrow: {
      index: 4,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    tokenProgram: {
      index: 5,
      isWritable: false,
      value: input.tokenProgram ?? null
    },
    systemProgram: {
      index: 6,
      isWritable: false,
      value: input.systemProgram ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      "splSystem",
      "11111111111111111111111111111111"
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getInitOftInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getLzReceiveInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["params", getLzReceiveParamsSerializer()]
      ],
      { description: "LzReceiveInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([8, 179, 120, 109, 33, 118, 189, 80])
    })
  );
}
function lzReceive(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true,
      value: input.payer ?? null
    },
    peer: { index: 1, isWritable: true, value: input.peer ?? null },
    oftStore: {
      index: 2,
      isWritable: true,
      value: input.oftStore ?? null
    },
    tokenEscrow: {
      index: 3,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    toAddress: {
      index: 4,
      isWritable: false,
      value: input.toAddress ?? null
    },
    tokenDest: {
      index: 5,
      isWritable: true,
      value: input.tokenDest ?? null
    },
    tokenMint: {
      index: 6,
      isWritable: true,
      value: input.tokenMint ?? null
    },
    mintAuthority: {
      index: 7,
      isWritable: false,
      value: input.mintAuthority ?? null
    },
    tokenProgram: {
      index: 8,
      isWritable: false,
      value: input.tokenProgram ?? null
    },
    associatedTokenProgram: {
      index: 9,
      isWritable: false,
      value: input.associatedTokenProgram ?? null
    },
    systemProgram: {
      index: 10,
      isWritable: false,
      value: input.systemProgram ?? null
    },
    eventAuthority: {
      index: 11,
      isWritable: false,
      value: input.eventAuthority ?? null
    },
    program: {
      index: 12,
      isWritable: false,
      value: input.program ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      "splSystem",
      "11111111111111111111111111111111"
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getLzReceiveInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getLzReceiveTypesInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["params", getLzReceiveParamsSerializer()]
      ],
      { description: "LzReceiveTypesInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([221, 17, 246, 159, 248, 128, 31, 96])
    })
  );
}
function lzReceiveTypes(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    oftStore: {
      index: 0,
      isWritable: false,
      value: input.oftStore ?? null
    },
    tokenMint: {
      index: 1,
      isWritable: false,
      value: input.tokenMint ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getLzReceiveTypesInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getOftVersionInstructionDataSerializer() {
  return mapSerializer(
    struct([["discriminator", bytes({ size: 8 })]], {
      description: "OftVersionInstructionData"
    }),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([179, 115, 253, 187, 88, 22, 80, 64])
    })
  );
}
function oftVersion(context) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {};
  const orderedAccounts = Object.values(
    resolvedAccounts
  );
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getOftVersionInstructionDataSerializer().serialize({});
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getQuoteOftInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["dstEid", u32()],
        ["to", bytes({ size: 32 })],
        ["amountLd", u64()],
        ["minAmountLd", u64()],
        ["options", bytes({ size: u32() })],
        ["composeMsg", option(bytes({ size: u32() }))],
        ["payInLzToken", bool()]
      ],
      { description: "QuoteOftInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([179, 255, 92, 202, 251, 82, 82, 118])
    })
  );
}
function quoteOft(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    oftStore: {
      index: 0,
      isWritable: false,
      value: input.oftStore ?? null
    },
    peer: { index: 1, isWritable: false, value: input.peer ?? null },
    tokenMint: {
      index: 2,
      isWritable: false,
      value: input.tokenMint ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getQuoteOftInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getQuoteSendInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["dstEid", u32()],
        ["to", bytes({ size: 32 })],
        ["amountLd", u64()],
        ["minAmountLd", u64()],
        ["options", bytes({ size: u32() })],
        ["composeMsg", option(bytes({ size: u32() }))],
        ["payInLzToken", bool()]
      ],
      { description: "QuoteSendInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([207, 0, 49, 214, 160, 211, 76, 211])
    })
  );
}
function quoteSend(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    oftStore: {
      index: 0,
      isWritable: false,
      value: input.oftStore ?? null
    },
    peer: { index: 1, isWritable: false, value: input.peer ?? null },
    tokenMint: {
      index: 2,
      isWritable: false,
      value: input.tokenMint ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getQuoteSendInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSendInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["dstEid", u32()],
        ["to", bytes({ size: 32 })],
        ["amountLd", u64()],
        ["minAmountLd", u64()],
        ["options", bytes({ size: u32() })],
        ["composeMsg", option(bytes({ size: u32() }))],
        ["nativeFee", u64()],
        ["lzTokenFee", u64()]
      ],
      { description: "SendInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([102, 251, 20, 187, 65, 75, 12, 69])
    })
  );
}
function send(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false,
      value: input.signer ?? null
    },
    peer: { index: 1, isWritable: true, value: input.peer ?? null },
    oftStore: {
      index: 2,
      isWritable: true,
      value: input.oftStore ?? null
    },
    tokenSource: {
      index: 3,
      isWritable: true,
      value: input.tokenSource ?? null
    },
    tokenEscrow: {
      index: 4,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    tokenMint: {
      index: 5,
      isWritable: true,
      value: input.tokenMint ?? null
    },
    tokenProgram: {
      index: 6,
      isWritable: false,
      value: input.tokenProgram ?? null
    },
    eventAuthority: {
      index: 7,
      isWritable: false,
      value: input.eventAuthority ?? null
    },
    program: {
      index: 8,
      isWritable: false,
      value: input.program ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSendInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSetOftConfigInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["params", getSetOFTConfigParamsSerializer()]
      ],
      { description: "SetOftConfigInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([55, 126, 87, 217, 159, 66, 24, 194])
    })
  );
}
function setOftConfig(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false,
      value: input.admin ?? null
    },
    oftStore: {
      index: 1,
      isWritable: true,
      value: input.oftStore ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSetOftConfigInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSetPauseInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["paused", bool()]
      ],
      { description: "SetPauseInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([63, 32, 154, 2, 56, 103, 79, 45])
    })
  );
}
function setPause(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false,
      value: input.signer ?? null
    },
    oftStore: {
      index: 1,
      isWritable: true,
      value: input.oftStore ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSetPauseInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSetPeerConfigInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["remoteEid", u32()],
        ["config", getPeerConfigParamSerializer()]
      ],
      { description: "SetPeerConfigInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([79, 187, 168, 57, 139, 140, 93, 47])
    })
  );
}
function setPeerConfig(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: true,
      value: input.admin ?? null
    },
    peer: { index: 1, isWritable: true, value: input.peer ?? null },
    oftStore: {
      index: 2,
      isWritable: false,
      value: input.oftStore ?? null
    },
    systemProgram: {
      index: 3,
      isWritable: false,
      value: input.systemProgram ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      "splSystem",
      "11111111111111111111111111111111"
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSetPeerConfigInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getWithdrawFeeInstructionDataSerializer() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["feeLd", u64()]
      ],
      { description: "WithdrawFeeInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([14, 122, 231, 218, 31, 238, 223, 150])
    })
  );
}
function withdrawFee(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false,
      value: input.admin ?? null
    },
    oftStore: {
      index: 1,
      isWritable: false,
      value: input.oftStore ?? null
    },
    tokenMint: {
      index: 2,
      isWritable: false,
      value: input.tokenMint ?? null
    },
    tokenEscrow: {
      index: 3,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    tokenDest: {
      index: 4,
      isWritable: true,
      value: input.tokenDest ?? null
    },
    tokenProgram: {
      index: 5,
      isWritable: false,
      value: input.tokenProgram ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getWithdrawFeeInstructionDataSerializer().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
var eddsa = createWeb3JsEddsa();
var OftPDA = class {
  constructor(program) {
    this.program = program;
  }
  config() {
    return eddsa.findPda(this.program, [Buffer.from("OftConfig", "utf8")]);
  }
  peer(oftStore, eid) {
    return eddsa.findPda(this.program, [
      Buffer.from(PEER_SEED, "utf8"),
      publicKeyBytes(oftStore),
      u32({ endian: Endian.Big }).serialize(eid)
      // new BN(eid).toArrayLike(Buffer, 'be', 4),
    ]);
  }
  oftStore(escrow) {
    return eddsa.findPda(this.program, [Buffer.from(OFT_SEED, "utf8"), publicKeyBytes(escrow)]);
  }
  lzReceiveTypesAccounts(oftConfig) {
    return eddsa.findPda(this.program, [Buffer.from(LZ_RECEIVE_TYPES_SEED, "utf8"), publicKeyBytes(oftConfig)]);
  }
};

// src/generated/oft302/programs/index.ts
var programs_exports = {};
__export(programs_exports, {
  OFT_PROGRAM_ID: () => OFT_PROGRAM_ID,
  createOftProgram: () => createOftProgram,
  getOftProgram: () => getOftProgram,
  getOftProgramId: () => getOftProgramId
});

// src/generated/oft302/programs/oft.ts
var OFT_PROGRAM_ID = "";
function createOftProgram() {
  return {
    name: "oft",
    publicKey: OFT_PROGRAM_ID,
    getErrorFromCode(code, cause) {
      return getOftErrorFromCode(code, this, cause);
    },
    getErrorFromName(name, cause) {
      return getOftErrorFromName(name, this, cause);
    },
    isOnCluster() {
      return true;
    }
  };
}
function getOftProgram(context, clusterFilter) {
  return context.programs.get("oft", clusterFilter);
}
function getOftProgramId(context, clusterFilter) {
  return context.programs.getPublicKey("oft", OFT_PROGRAM_ID, clusterFilter);
}

// src/oft302.ts
var ENDPOINT_PROGRAM_ID = EndpointProgram.ENDPOINT_PROGRAM_ID;
var ULN_PROGRAM_ID = UlnProgram.ULN_PROGRAM_ID;
function createOFTProgramRepo(oftProgram, rpc) {
  if (rpc === void 0) {
    rpc = createNullRpc();
    rpc.getCluster = () => "custom";
  }
  return createDefaultProgramRepository({ rpc }, [
    {
      name: "oft",
      publicKey: oftProgram,
      getErrorFromCode(code, cause) {
        return getOftErrorFromCode(code, this, cause);
      },
      getErrorFromName(name, cause) {
        return getOftErrorFromName(name, this, cause);
      },
      isOnCluster() {
        return true;
      }
    }
  ]);
}
function initOft2(accounts, oftType, sharedDecimals = OFT_DECIMALS, programs) {
  const programsRepo = typeof programs.oft === "string" ? createOFTProgramRepo(programs.oft) : programs.oft;
  const deriver = new OftPDA(programsRepo.getPublicKey("oft"));
  const endpoint = new EndpointProgram.Endpoint(programs.endpoint ?? ENDPOINT_PROGRAM_ID);
  const { payer, admin, mint, escrow } = accounts;
  const [oftStore] = deriver.oftStore(escrow.publicKey);
  const [lzReceiveTypes3] = deriver.lzReceiveTypesAccounts(oftStore);
  const txBuilder = initOft(
    {
      payer,
      programs: programsRepo
    },
    {
      // accounts
      oftStore,
      lzReceiveTypesAccounts: lzReceiveTypes3,
      tokenMint: mint,
      tokenEscrow: escrow,
      tokenProgram: programs.token ?? fromWeb3JsPublicKey(TOKEN_PROGRAM_ID),
      // params
      oftType,
      admin,
      sharedDecimals,
      endpointProgram: endpoint.programId
    }
  );
  const retval = txBuilder.addRemainingAccounts(
    endpoint.getRegisterOappIxAccountMetaForCPI(payer.publicKey, oftStore).map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  ).items[0];
  retval.signers = [payer, escrow];
  return retval;
}
function setOFTConfig(accounts, params, programs) {
  let actualParams;
  const { oftStore, admin } = accounts;
  const remainingAccounts = [];
  if (params.__kind === "Admin") {
    if (params.admin === void 0) {
      throw new Error("Admin is required");
    }
    actualParams = {
      __kind: "Admin",
      fields: [params.admin]
    };
  } else if (params.__kind === "Delegate") {
    if (params.delegate === void 0) {
      throw new Error("Delegate is required");
    }
    actualParams = {
      __kind: "Delegate",
      fields: [params.delegate]
    };
    const endpointProgram = programs.endpoint ?? ENDPOINT_PROGRAM_ID;
    const endpointSDK = new EndpointProgram.Endpoint(endpointProgram);
    const keys = endpointSDK.getSetDelegateIxAccountMetaForCPI(oftStore);
    for (const acc of keys) {
      acc.isSigner = false;
    }
    remainingAccounts.push(...keys);
  } else if (params.__kind === "DefaultFee") {
    if (params.defaultFee === void 0) {
      throw new Error("DefaultFee is required");
    }
    actualParams = {
      __kind: "DefaultFee",
      fields: [params.defaultFee]
    };
  } else if (params.__kind === "Paused") {
    if (params.paused === void 0) {
      throw new Error("Paused is required");
    }
    actualParams = {
      __kind: "Paused",
      fields: [params.paused]
    };
  } else if (params.__kind === "Pauser") {
    actualParams = {
      __kind: "Pauser",
      fields: [params.pauser ? some(params.pauser) : none()]
    };
  } else {
    actualParams = {
      __kind: "Unpauser",
      fields: [params.unpauser ? some(params.unpauser) : none()]
    };
  }
  const txBuilder = setOftConfig(
    { programs: createOFTProgramRepo(programs.oft) },
    {
      admin,
      oftStore,
      params: actualParams
    }
  );
  return txBuilder.addRemainingAccounts(
    remainingAccounts.map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  ).items[0];
}
function setPeerConfig2(accounts, param, oftProgramId) {
  const programsRepo = typeof oftProgramId === "string" ? createOFTProgramRepo(oftProgramId) : oftProgramId;
  const { remote: remoteId } = param;
  if (remoteId % 3e4 == 0) {
    throw new Error("Invalid remote ID");
  }
  const { admin, oftStore } = accounts;
  const [peerPda] = new OftPDA(programsRepo.getPublicKey("oft")).peer(oftStore, remoteId);
  let config;
  if (param.__kind === "PeerAddress") {
    if (param.peer.length !== 32) {
      throw new Error("Peer must be 32 bytes (left-padded with zeroes)");
    }
    config = peerConfigParam("PeerAddress", [param.peer]);
  } else if (param.__kind === "FeeBps") {
    config = { __kind: "FeeBps", fields: [some(param.feeBps)] };
  } else if (param.__kind === "EnforcedOptions") {
    config = {
      __kind: "EnforcedOptions",
      send: param.send,
      sendAndCall: param.sendAndCall
    };
  } else if (param.__kind === "OutboundRateLimit" || param.__kind === "InboundRateLimit") {
    config = {
      __kind: param.__kind,
      fields: [
        param.rateLimit ? some({
          refillPerSecond: some(param.rateLimit.refillPerSecond),
          capacity: some(param.rateLimit.capacity)
        }) : null
      ]
    };
  } else {
    throw new Error("Invalid peer config");
  }
  return setPeerConfig(
    { programs: programsRepo },
    {
      admin,
      peer: peerPda,
      oftStore,
      // params
      remoteEid: remoteId,
      config
    }
  ).items[0];
}
function initSendLibrary(accounts, remoteEid, endpointProgram = ENDPOINT_PROGRAM_ID) {
  const { admin, oftStore } = accounts;
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  return endpoint.initOAppSendLibrary(admin, { sender: oftStore, remote: remoteEid });
}
function initReceiveLibrary(accounts, remoteEid, endpointProgram = ENDPOINT_PROGRAM_ID) {
  const { admin, oftStore } = accounts;
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  return endpoint.initOAppReceiveLibrary(admin, { receiver: oftStore, remote: remoteEid });
}
function setSendLibrary(accounts, params, endpointProgram = ENDPOINT_PROGRAM_ID) {
  const { sendLibraryProgram, remoteEid } = params;
  const { admin, oftStore } = accounts;
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  return endpoint.setOAppSendLibrary(admin, {
    sender: oftStore,
    remote: remoteEid,
    msgLibProgram: sendLibraryProgram
  });
}
function setReceiveLibrary(accounts, params, endpointProgram = ENDPOINT_PROGRAM_ID) {
  const { receiveLibraryProgram, remoteEid, gracePeriod } = params;
  const { admin, oftStore } = accounts;
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  return endpoint.setOAppReceiveLibrary(admin, {
    receiver: oftStore,
    remote: remoteEid,
    msgLibProgram: receiveLibraryProgram,
    gracePeriod
  });
}
function initConfig(accounts, remoteEid, programs) {
  const { admin, oftStore, payer } = accounts;
  let msgLibProgram, endpointProgram;
  if (programs === void 0) {
    msgLibProgram = ULN_PROGRAM_ID;
    endpointProgram = ENDPOINT_PROGRAM_ID;
  } else {
    msgLibProgram = programs.msgLib ?? ULN_PROGRAM_ID;
    endpointProgram = programs.endpoint ?? ENDPOINT_PROGRAM_ID;
  }
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  let msgLib;
  if (msgLibProgram === SimpleMessageLibProgram.SIMPLE_MESSAGELIB_PROGRAM_ID) {
    msgLib = new SimpleMessageLibProgram.SimpleMessageLib(SimpleMessageLibProgram.SIMPLE_MESSAGELIB_PROGRAM_ID);
  } else {
    msgLib = new UlnProgram.Uln(msgLibProgram);
  }
  return endpoint.initOAppConfig(
    {
      delegate: admin,
      payer: payer.publicKey
    },
    {
      msgLibSDK: msgLib,
      oapp: oftStore,
      remote: remoteEid
    }
  );
}
async function setConfig(rpc, accounts, params, programs) {
  const { signer, oftStore } = accounts;
  const { remoteEid, configType, config } = params;
  let msgLibProgram, endpointProgram;
  if (programs === void 0) {
    msgLibProgram = ULN_PROGRAM_ID;
    endpointProgram = ENDPOINT_PROGRAM_ID;
  } else {
    msgLibProgram = programs.msgLib ?? ULN_PROGRAM_ID;
    endpointProgram = programs.endpoint ?? ENDPOINT_PROGRAM_ID;
  }
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  return endpoint.setOAppConfig(rpc, signer, {
    oapp: oftStore,
    eid: remoteEid,
    config: {
      configType,
      value: config
    },
    msgLibProgram
  });
}
function withdrawFee2(accounts, amount, programs) {
  const { admin, mint, escrow, dest } = accounts;
  const programsRepo = typeof programs.oft === "string" ? createOFTProgramRepo(programs.oft) : programs.oft;
  const [oftStore] = new OftPDA(programsRepo.getPublicKey("oft")).oftStore(escrow);
  return withdrawFee(
    { programs: programsRepo },
    {
      admin,
      tokenEscrow: escrow,
      tokenDest: dest,
      tokenProgram: programs.token ?? fromWeb3JsPublicKey(TOKEN_PROGRAM_ID),
      oftStore,
      tokenMint: mint,
      // params
      feeLd: amount
    }
  ).items[0];
}
async function send2(rpc, accounts, sendParams, programs, remainingAccounts) {
  const { payer, tokenMint, tokenEscrow, tokenSource } = accounts;
  const { dstEid, to, amountLd, minAmountLd, options, composeMsg, nativeFee, lzTokenFee } = sendParams;
  const deriver = new OftPDA(programs.oft);
  const [oftStore] = deriver.oftStore(tokenEscrow);
  const [peer] = deriver.peer(oftStore, dstEid);
  if (remainingAccounts === void 0 || remainingAccounts.length === 0) {
    const peerAddr = accounts.peerAddr ?? await fetchPeerConfig({ rpc }, peer).then((peerInfo) => peerInfo.peerAddress);
    const endpoint = new EndpointProgram.Endpoint(programs.endpoint ?? ENDPOINT_PROGRAM_ID);
    const msgLibProgram = await getSendLibraryProgram(rpc, endpoint, payer.publicKey, oftStore, dstEid);
    const packetPath = {
      dstEid,
      sender: oftStore,
      receiver: peerAddr
    };
    remainingAccounts = await endpoint.getSendIXAccountMetaForCPI(rpc, payer.publicKey, {
      path: packetPath,
      msgLibProgram
    });
  }
  const [eventAuthorityPDA] = new EventPDA(programs.oft).eventAuthority();
  const tokenProgram = programs.token ?? fromWeb3JsPublicKey(TOKEN_PROGRAM_ID);
  const txBuilder = send(
    { programs: createOFTProgramRepo(programs.oft) },
    {
      signer: payer,
      peer,
      oftStore,
      tokenSource,
      tokenEscrow,
      tokenMint,
      tokenProgram,
      eventAuthority: eventAuthorityPDA,
      program: programs.oft,
      // params
      dstEid,
      to,
      amountLd,
      minAmountLd,
      options: options ?? new Uint8Array(),
      composeMsg: composeMsg ?? null,
      nativeFee,
      lzTokenFee: lzTokenFee ?? 0
    }
  );
  return txBuilder.addRemainingAccounts(
    remainingAccounts.map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  ).items[0];
}
async function quote(rpc, accounts, quoteParams, programs, remainingAccounts, addressLookupTables) {
  const { dstEid, to, amountLd, minAmountLd, options, payInLzToken, composeMsg } = quoteParams;
  const { payer, tokenMint, tokenEscrow } = accounts;
  const deriver = new OftPDA(programs.oft);
  const [oftStore] = deriver.oftStore(tokenEscrow);
  const [peer] = deriver.peer(oftStore, dstEid);
  if (remainingAccounts === void 0 || remainingAccounts.length === 0) {
    const peerAddr = accounts.peerAddr ?? await fetchPeerConfig({ rpc }, peer).then((peerInfo) => peerInfo.peerAddress);
    const endpoint = new EndpointProgram.Endpoint(programs.endpoint ?? ENDPOINT_PROGRAM_ID);
    const messageLib = await getSendLibraryProgram(rpc, endpoint, payer, oftStore, dstEid);
    remainingAccounts = await endpoint.getQuoteIXAccountMetaForCPI(rpc, payer, {
      path: { sender: oftStore, dstEid, receiver: peerAddr },
      msgLibProgram: messageLib
    });
  }
  let txBuilder = quoteSend(
    { programs: createOFTProgramRepo(programs.oft) },
    {
      oftStore,
      peer,
      tokenMint,
      // params
      dstEid,
      to,
      amountLd,
      minAmountLd,
      options: options ?? new Uint8Array(),
      payInLzToken: payInLzToken ?? false,
      composeMsg: composeMsg ?? null
    }
  );
  txBuilder = txBuilder.addRemainingAccounts(
    // Get remaining accounts from msgLib(simple_msgLib or uln)
    remainingAccounts.map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  );
  const web3Ix = toWeb3JsInstruction(txBuilder.getInstructions()[0]);
  const modifyComputeUnits = ComputeBudgetProgram.setComputeUnitLimit({
    units: 1e6
  });
  return simulateWeb3JsTransaction(
    rpc,
    [modifyComputeUnits, web3Ix],
    web3Ix.programId,
    toWeb3JsPublicKey(payer),
    EndpointProgram.types.getMessagingFeeSerializer(),
    "confirmed",
    void 0,
    addressLookupTables !== void 0 ? Array.isArray(addressLookupTables) ? addressLookupTables.map(toWeb3JsPublicKey) : toWeb3JsPublicKey(addressLookupTables) : void 0
  );
}
async function quoteOft2(rpc, accounts, quoteParams, oftProgram) {
  const { payer, tokenMint, tokenEscrow } = accounts;
  const { dstEid, to, amountLd, minAmountLd, options, payInLzToken, composeMsg } = quoteParams;
  const deriver = new OftPDA(oftProgram);
  const [oftStore] = deriver.oftStore(tokenEscrow);
  const [peer] = deriver.peer(oftStore, dstEid);
  const ix = quoteOft(
    { programs: createOFTProgramRepo(oftProgram) },
    {
      oftStore,
      peer,
      tokenMint,
      // params
      dstEid,
      to,
      amountLd,
      minAmountLd,
      options: options ?? new Uint8Array(),
      payInLzToken: payInLzToken ?? false,
      composeMsg: composeMsg ?? null
    }
  ).getInstructions()[0];
  const web3Ix = toWeb3JsInstruction(ix);
  return simulateWeb3JsTransaction(
    rpc.getEndpoint(),
    [web3Ix],
    web3Ix.programId,
    toWeb3JsPublicKey(payer),
    getQuoteOFTResultSerializer(),
    "confirmed"
  );
}
function initOAppNonce(accounts, remoteEid, remoteOappAddr, endpointProgram = ENDPOINT_PROGRAM_ID) {
  const { admin, oftStore } = accounts;
  const endpoint = new EndpointProgram.Endpoint(endpointProgram);
  return endpoint.initOAppNonce(admin, {
    localOApp: oftStore,
    remote: remoteEid,
    remoteOApp: remoteOappAddr
  });
}
async function getEndpointConfig(rpc, oftStore, endpointId, programs) {
  let msgLibProgram, endpointProgram;
  if (programs === void 0) {
    msgLibProgram = ULN_PROGRAM_ID;
    endpointProgram = ENDPOINT_PROGRAM_ID;
  } else {
    msgLibProgram = programs.msgLib ?? ULN_PROGRAM_ID;
    endpointProgram = programs.endpoint ?? ENDPOINT_PROGRAM_ID;
  }
  const endpointDeriver = new EndpointPDA(endpointProgram);
  const ulnDeriver = new UlnPDA(msgLibProgram);
  const [sendLib] = endpointDeriver.sendLibraryConfig(oftStore, endpointId);
  const [defaultSendLib] = endpointDeriver.defaultSendLibraryConfig(endpointId);
  const [receiveLib] = endpointDeriver.receiveLibraryConfig(oftStore, endpointId);
  const [defaultReceiveLib] = endpointDeriver.defaultReceiveLibraryConfig(endpointId);
  const [msgLib] = ulnDeriver.messageLib();
  let sendLibraryConfig = await EndpointProgram.accounts.fetchSendLibraryConfig({ rpc }, sendLib);
  let receiveLibraryConfig = await EndpointProgram.accounts.fetchReceiveLibraryConfig({ rpc }, receiveLib);
  const defaultSendLibraryConfig = await EndpointProgram.accounts.fetchSendLibraryConfig({ rpc }, defaultSendLib);
  const defaultReceiveLibraryConfig = await EndpointProgram.accounts.fetchReceiveLibraryConfig({ rpc }, defaultReceiveLib);
  const nil64 = 18446744073709551615n;
  const nil8 = 255;
  if (sendLibraryConfig.messageLib === defaultPublicKey()) {
    sendLibraryConfig = defaultSendLibraryConfig;
  }
  if (receiveLibraryConfig.messageLib === defaultPublicKey()) {
    receiveLibraryConfig = defaultReceiveLibraryConfig;
  }
  if (sendLibraryConfig.messageLib === msgLib) {
    const [ulnDefaultSendConfigPDA] = ulnDeriver.defaultSendConfig(endpointId);
    const [ulnSendConfigPDA] = ulnDeriver.sendConfig(endpointId, oftStore);
    const ulnSendConfig = await UlnProgram.accounts.fetchSendConfig({ rpc }, ulnSendConfigPDA);
    const ulnDefaultSendConfig = await UlnProgram.accounts.fetchSendConfig({ rpc }, ulnDefaultSendConfigPDA);
    if (nil64 === ulnSendConfig.uln.confirmations) {
      ulnSendConfig.uln.confirmations = 0n;
    } else if (ulnSendConfig.uln.confirmations == 0n) {
      ulnSendConfig.uln.confirmations = ulnDefaultSendConfig.uln.confirmations;
    }
    if (ulnSendConfig.uln.requiredDvnCount == nil8) {
      ulnSendConfig.uln.requiredDvnCount = 0;
    } else if (ulnSendConfig.uln.requiredDvnCount == 0) {
      ulnSendConfig.uln.requiredDvnCount = ulnDefaultSendConfig.uln.requiredDvnCount;
      ulnSendConfig.uln.requiredDvns = ulnDefaultSendConfig.uln.requiredDvns;
    }
    if (ulnSendConfig.uln.optionalDvnCount == nil8) {
      ulnSendConfig.uln.optionalDvnCount = 0;
      ulnSendConfig.uln.optionalDvnThreshold = 0;
    } else if (ulnSendConfig.uln.optionalDvnCount == 0) {
      ulnSendConfig.uln.optionalDvnCount = ulnDefaultSendConfig.uln.optionalDvnCount;
      ulnSendConfig.uln.optionalDvnThreshold = ulnDefaultSendConfig.uln.optionalDvnThreshold;
      ulnSendConfig.uln.optionalDvns = ulnDefaultSendConfig.uln.optionalDvns;
    }
    sendLibraryConfig.ulnSendConfig = ulnSendConfig;
  }
  if (receiveLibraryConfig.messageLib === msgLib) {
    const [ulnDefaultReceiveConfigPDA] = ulnDeriver.defaultReceiveConfig(endpointId);
    const [ulnReceiveConfigPDA] = ulnDeriver.receiveConfig(endpointId, oftStore);
    const ulnReceiveConfig = await UlnProgram.accounts.fetchReceiveConfig({ rpc }, ulnReceiveConfigPDA);
    const ulnDefaultReceiveConfig = await UlnProgram.accounts.fetchReceiveConfig(
      { rpc },
      ulnDefaultReceiveConfigPDA
    );
    if (nil64 === ulnReceiveConfig.uln.confirmations) {
      ulnReceiveConfig.uln.confirmations = 0n;
    } else if (ulnReceiveConfig.uln.confirmations == 0n) {
      ulnReceiveConfig.uln.confirmations = ulnDefaultReceiveConfig.uln.confirmations;
    }
    if (ulnReceiveConfig.uln.requiredDvnCount == nil8) {
      ulnReceiveConfig.uln.requiredDvnCount = 0;
    } else if (ulnReceiveConfig.uln.requiredDvnCount == 0) {
      ulnReceiveConfig.uln.requiredDvnCount = ulnDefaultReceiveConfig.uln.requiredDvnCount;
      ulnReceiveConfig.uln.requiredDvns = ulnDefaultReceiveConfig.uln.requiredDvns;
    }
    if (ulnReceiveConfig.uln.optionalDvnCount == nil8) {
      ulnReceiveConfig.uln.optionalDvnCount = 0;
      ulnReceiveConfig.uln.optionalDvnThreshold = 0;
    } else if (ulnReceiveConfig.uln.optionalDvnCount == 0) {
      ulnReceiveConfig.uln.optionalDvnCount = ulnDefaultReceiveConfig.uln.optionalDvnCount;
      ulnReceiveConfig.uln.optionalDvnThreshold = ulnDefaultReceiveConfig.uln.optionalDvnThreshold;
      ulnReceiveConfig.uln.optionalDvns = ulnDefaultReceiveConfig.uln.optionalDvns;
    }
    receiveLibraryConfig.ulnReceiveConfig = ulnReceiveConfig;
  }
  return {
    sendLibraryConfig,
    receiveLibraryConfig
  };
}
async function getPeerAddress(rpc, oftInstance, remoteEid, oftProgramId) {
  const [peer] = new OftPDA(oftProgramId).peer(oftInstance, remoteEid);
  const peerInfo = await fetchPeerConfig({ rpc }, peer);
  return hexlify(peerInfo.peerAddress);
}
async function getDelegate(rpc, oftInstance, endpointProgram = ENDPOINT_PROGRAM_ID) {
  const [oAppRegistry] = new EndpointPDA(endpointProgram).oappRegistry(oftInstance);
  const oAppRegistryInfo = await EndpointProgram.accounts.fetchOAppRegistry({ rpc }, oAppRegistry);
  return oAppRegistryInfo.delegate;
}
async function getEnforcedOptions(rpc, oftInstance, remoteEid, oftProgramId) {
  const [peer] = new OftPDA(oftProgramId).peer(oftInstance, remoteEid);
  const peerInfo = await fetchPeerConfig({ rpc }, peer);
  return peerInfo.enforcedOptions;
}
async function getSendLibraryProgram(rpc, endpoint, payer, oftStore, remoteEid) {
  const sendLibInfo = await endpoint.getSendLibrary(rpc, oftStore, remoteEid);
  if (!sendLibInfo.programId) {
    throw new Error("Send library not initialized or blocked message library");
  }
  const { programId: msgLibProgram } = sendLibInfo;
  const msgLibVersion = await endpoint.getMessageLibVersion(rpc, payer, msgLibProgram);
  if (msgLibVersion.major.toString() === "0" && msgLibVersion.minor == 0 && msgLibVersion.endpointVersion == 2) {
    return new SimpleMessageLibProgram.SimpleMessageLib(msgLibProgram);
  } else if (msgLibVersion.major.toString() === "3" && msgLibVersion.minor == 0 && msgLibVersion.endpointVersion == 2) {
    return new UlnProgram.Uln(msgLibProgram);
  }
  throw new Error(`Unsupported message library version: ${JSON.stringify(msgLibVersion, null, 2)}`);
}

// src/oft202.ts
var oft202_exports = {};
__export(oft202_exports, {
  accounts: () => accounts_exports2,
  createOFTProgramRepo: () => createOFTProgramRepo2,
  errors: () => errors_exports2,
  getDelegate: () => getDelegate,
  getEndpointConfig: () => getEndpointConfig,
  getEnforcedOptions: () => getEnforcedOptions,
  getPeerAddress: () => getPeerAddress,
  initConfig: () => initConfig,
  initOAppNonce: () => initOAppNonce,
  initOft: () => initOft4,
  initReceiveLibrary: () => initReceiveLibrary,
  initSendLibrary: () => initSendLibrary,
  instructions: () => instructions_exports2,
  programs: () => programs_exports2,
  quote: () => quote2,
  quoteOft: () => quoteOft4,
  send: () => send4,
  setConfig: () => setConfig,
  setOFTConfig: () => setOFTConfig2,
  setPeerConfig: () => setPeerConfig4,
  setReceiveLibrary: () => setReceiveLibrary,
  setSendLibrary: () => setSendLibrary,
  shared: () => shared_exports2,
  types: () => types_exports2,
  withdrawFee: () => withdrawFee4
});

// src/generated/oft202/accounts/index.ts
var accounts_exports2 = {};
__export(accounts_exports2, {
  deserializeLzReceiveTypesAccounts: () => deserializeLzReceiveTypesAccounts2,
  deserializeOFTStore: () => deserializeOFTStore2,
  deserializePeerConfig: () => deserializePeerConfig2,
  fetchAllLzReceiveTypesAccounts: () => fetchAllLzReceiveTypesAccounts2,
  fetchAllOFTStore: () => fetchAllOFTStore2,
  fetchAllPeerConfig: () => fetchAllPeerConfig2,
  fetchLzReceiveTypesAccounts: () => fetchLzReceiveTypesAccounts2,
  fetchOFTStore: () => fetchOFTStore2,
  fetchPeerConfig: () => fetchPeerConfig2,
  getLzReceiveTypesAccountsAccountDataSerializer: () => getLzReceiveTypesAccountsAccountDataSerializer2,
  getLzReceiveTypesAccountsGpaBuilder: () => getLzReceiveTypesAccountsGpaBuilder2,
  getLzReceiveTypesAccountsSize: () => getLzReceiveTypesAccountsSize2,
  getOFTStoreAccountDataSerializer: () => getOFTStoreAccountDataSerializer2,
  getOFTStoreGpaBuilder: () => getOFTStoreGpaBuilder2,
  getPeerConfigAccountDataSerializer: () => getPeerConfigAccountDataSerializer2,
  getPeerConfigGpaBuilder: () => getPeerConfigGpaBuilder2,
  safeFetchAllLzReceiveTypesAccounts: () => safeFetchAllLzReceiveTypesAccounts2,
  safeFetchAllOFTStore: () => safeFetchAllOFTStore2,
  safeFetchAllPeerConfig: () => safeFetchAllPeerConfig2,
  safeFetchLzReceiveTypesAccounts: () => safeFetchLzReceiveTypesAccounts2,
  safeFetchOFTStore: () => safeFetchOFTStore2,
  safeFetchPeerConfig: () => safeFetchPeerConfig2
});
function getLzReceiveTypesAccountsAccountDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["oftStore", publicKey()],
        ["tokenMint", publicKey()]
      ],
      { description: "LzReceiveTypesAccountsAccountData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([248, 87, 167, 117, 5, 251, 21, 126])
    })
  );
}
function deserializeLzReceiveTypesAccounts2(rawAccount) {
  return deserializeAccount(
    rawAccount,
    getLzReceiveTypesAccountsAccountDataSerializer2()
  );
}
async function fetchLzReceiveTypesAccounts2(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  assertAccountExists(maybeAccount, "LzReceiveTypesAccounts");
  return deserializeLzReceiveTypesAccounts2(maybeAccount);
}
async function safeFetchLzReceiveTypesAccounts2(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  return maybeAccount.exists ? deserializeLzReceiveTypesAccounts2(maybeAccount) : null;
}
async function fetchAllLzReceiveTypesAccounts2(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, "LzReceiveTypesAccounts");
    return deserializeLzReceiveTypesAccounts2(maybeAccount);
  });
}
async function safeFetchAllLzReceiveTypesAccounts2(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.filter((maybeAccount) => maybeAccount.exists).map(
    (maybeAccount) => deserializeLzReceiveTypesAccounts2(maybeAccount)
  );
}
function getLzReceiveTypesAccountsGpaBuilder2(context) {
  const programId = context.programs.getPublicKey("oft", "");
  return gpaBuilder(context, programId).registerFields({
    discriminator: [0, bytes({ size: 8 })],
    oftStore: [8, publicKey()],
    tokenMint: [40, publicKey()]
  }).deserializeUsing(
    (account) => deserializeLzReceiveTypesAccounts2(account)
  ).whereField(
    "discriminator",
    new Uint8Array([248, 87, 167, 117, 5, 251, 21, 126])
  );
}
function getLzReceiveTypesAccountsSize2() {
  return 72;
}

// src/generated/oft202/types/index.ts
var types_exports2 = {};
__export(types_exports2, {
  OFTType: () => OFTType4,
  getComposeParamsSerializer: () => getComposeParamsSerializer,
  getEnforcedOptionsSerializer: () => getEnforcedOptionsSerializer2,
  getLzAccountSerializer: () => getLzAccountSerializer2,
  getLzReceiveParamsSerializer: () => getLzReceiveParamsSerializer2,
  getMessagingFeeSerializer: () => getMessagingFeeSerializer2,
  getOFTFeeDetailSerializer: () => getOFTFeeDetailSerializer2,
  getOFTLimitsSerializer: () => getOFTLimitsSerializer2,
  getOFTReceiptSerializer: () => getOFTReceiptSerializer2,
  getOFTTypeSerializer: () => getOFTTypeSerializer2,
  getPeerConfigParamSerializer: () => getPeerConfigParamSerializer2,
  getQuoteOFTResultSerializer: () => getQuoteOFTResultSerializer2,
  getRateLimitParamsSerializer: () => getRateLimitParamsSerializer2,
  getRateLimiterSerializer: () => getRateLimiterSerializer2,
  getSetOFTConfigParamsSerializer: () => getSetOFTConfigParamsSerializer2,
  isPeerConfigParam: () => isPeerConfigParam2,
  isSetOFTConfigParams: () => isSetOFTConfigParams2,
  peerConfigParam: () => peerConfigParam2,
  setOFTConfigParams: () => setOFTConfigParams2
});
function getComposeParamsSerializer() {
  return struct(
    [
      ["composeGas", u64()],
      ["composeMsg", bytes({ size: u32() })]
    ],
    { description: "ComposeParams" }
  );
}
function getEnforcedOptionsSerializer2() {
  return struct(
    [
      ["send", bytes({ size: u32() })],
      ["sendAndCall", bytes({ size: u32() })]
    ],
    { description: "EnforcedOptions" }
  );
}
function getLzAccountSerializer2() {
  return struct(
    [
      ["pubkey", publicKey()],
      ["isSigner", bool()],
      ["isWritable", bool()]
    ],
    { description: "LzAccount" }
  );
}
function getLzReceiveParamsSerializer2() {
  return struct(
    [
      ["srcEid", u32()],
      ["sender", bytes({ size: 32 })],
      ["nonce", u64()],
      ["guid", bytes({ size: 32 })],
      ["message", bytes({ size: u32() })],
      ["extraData", bytes({ size: u32() })]
    ],
    { description: "LzReceiveParams" }
  );
}
function getMessagingFeeSerializer2() {
  return struct(
    [
      ["nativeFee", u64()],
      ["lzTokenFee", u64()]
    ],
    { description: "MessagingFee" }
  );
}
function getOFTFeeDetailSerializer2() {
  return struct(
    [
      ["feeAmountLd", u64()],
      ["description", string()]
    ],
    { description: "OFTFeeDetail" }
  );
}
function getOFTLimitsSerializer2() {
  return struct(
    [
      ["minAmountLd", u64()],
      ["maxAmountLd", u64()]
    ],
    { description: "OFTLimits" }
  );
}
function getOFTReceiptSerializer2() {
  return struct(
    [
      ["amountSentLd", u64()],
      ["amountReceivedLd", u64()]
    ],
    { description: "OFTReceipt" }
  );
}
var OFTType4 = /* @__PURE__ */ ((OFTType7) => {
  OFTType7[OFTType7["Native"] = 0] = "Native";
  OFTType7[OFTType7["Adapter"] = 1] = "Adapter";
  return OFTType7;
})(OFTType4 || {});
function getOFTTypeSerializer2() {
  return scalarEnum(OFTType4, { description: "OFTType" });
}
function getPeerConfigParamSerializer2() {
  return dataEnum(
    [
      [
        "PeerAddress",
        struct([
          ["fields", tuple([bytes({ size: 32 })])]
        ])
      ],
      [
        "FeeBps",
        struct([
          ["fields", tuple([option(u16())])]
        ])
      ],
      [
        "EnforcedOptions",
        struct([
          ["send", bytes({ size: u32() })],
          ["sendAndCall", bytes({ size: u32() })]
        ])
      ],
      [
        "OutboundRateLimit",
        struct([
          ["fields", tuple([option(getRateLimitParamsSerializer2())])]
        ])
      ],
      [
        "InboundRateLimit",
        struct([
          ["fields", tuple([option(getRateLimitParamsSerializer2())])]
        ])
      ],
      [
        "IsEndpointV1",
        struct([
          ["fields", tuple([bool()])]
        ])
      ]
    ],
    { description: "PeerConfigParam" }
  );
}
function peerConfigParam2(kind, data) {
  return Array.isArray(data) ? { __kind: kind, fields: data } : { __kind: kind, ...data ?? {} };
}
function isPeerConfigParam2(kind, value) {
  return value.__kind === kind;
}
function getQuoteOFTResultSerializer2() {
  return struct(
    [
      ["oftLimits", getOFTLimitsSerializer2()],
      ["oftFeeDetails", array(getOFTFeeDetailSerializer2())],
      ["oftReceipt", getOFTReceiptSerializer2()]
    ],
    { description: "QuoteOFTResult" }
  );
}
function getRateLimiterSerializer2() {
  return struct(
    [
      ["capacity", u64()],
      ["tokens", u64()],
      ["refillPerSecond", u64()],
      ["lastRefillTime", u64()]
    ],
    { description: "RateLimiter" }
  );
}
function getRateLimitParamsSerializer2() {
  return struct(
    [
      ["refillPerSecond", option(u64())],
      ["capacity", option(u64())]
    ],
    { description: "RateLimitParams" }
  );
}
function getSetOFTConfigParamsSerializer2() {
  return dataEnum(
    [
      [
        "Admin",
        struct([
          ["fields", tuple([publicKey()])]
        ])
      ],
      [
        "Delegate",
        struct([
          ["fields", tuple([publicKey()])]
        ])
      ],
      [
        "DefaultFee",
        struct([
          ["fields", tuple([u16()])]
        ])
      ],
      [
        "Paused",
        struct([
          ["fields", tuple([bool()])]
        ])
      ],
      [
        "Pauser",
        struct([
          ["fields", tuple([option(publicKey())])]
        ])
      ],
      [
        "Unpauser",
        struct([
          ["fields", tuple([option(publicKey())])]
        ])
      ]
    ],
    { description: "SetOFTConfigParams" }
  );
}
function setOFTConfigParams2(kind, data) {
  return Array.isArray(data) ? { __kind: kind, fields: data } : { __kind: kind, ...data ?? {} };
}
function isSetOFTConfigParams2(kind, value) {
  return value.__kind === kind;
}

// src/generated/oft202/accounts/oFTStore.ts
function getOFTStoreAccountDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["oftType", getOFTTypeSerializer2()],
        ["ld2sdRate", u64()],
        ["tokenMint", publicKey()],
        ["tokenEscrow", publicKey()],
        ["endpointProgram", publicKey()],
        ["bump", u8()],
        ["tvlLd", u64()],
        ["admin", publicKey()],
        ["defaultFeeBps", u16()],
        ["paused", bool()],
        ["pauser", option(publicKey())],
        ["unpauser", option(publicKey())]
      ],
      { description: "OFTStoreAccountData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([195, 215, 104, 134, 185, 195, 240, 114])
    })
  );
}
function deserializeOFTStore2(rawAccount) {
  return deserializeAccount(rawAccount, getOFTStoreAccountDataSerializer2());
}
async function fetchOFTStore2(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  assertAccountExists(maybeAccount, "OFTStore");
  return deserializeOFTStore2(maybeAccount);
}
async function safeFetchOFTStore2(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  return maybeAccount.exists ? deserializeOFTStore2(maybeAccount) : null;
}
async function fetchAllOFTStore2(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, "OFTStore");
    return deserializeOFTStore2(maybeAccount);
  });
}
async function safeFetchAllOFTStore2(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.filter((maybeAccount) => maybeAccount.exists).map((maybeAccount) => deserializeOFTStore2(maybeAccount));
}
function getOFTStoreGpaBuilder2(context) {
  const programId = context.programs.getPublicKey("oft", "");
  return gpaBuilder(context, programId).registerFields({
    discriminator: [0, bytes({ size: 8 })],
    oftType: [8, getOFTTypeSerializer2()],
    ld2sdRate: [9, u64()],
    tokenMint: [17, publicKey()],
    tokenEscrow: [49, publicKey()],
    endpointProgram: [81, publicKey()],
    bump: [113, u8()],
    tvlLd: [114, u64()],
    admin: [122, publicKey()],
    defaultFeeBps: [154, u16()],
    paused: [156, bool()],
    pauser: [157, option(publicKey())],
    unpauser: [null, option(publicKey())]
  }).deserializeUsing((account) => deserializeOFTStore2(account)).whereField(
    "discriminator",
    new Uint8Array([195, 215, 104, 134, 185, 195, 240, 114])
  );
}
function getPeerConfigAccountDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["peerAddress", bytes({ size: 32 })],
        ["enforcedOptions", getEnforcedOptionsSerializer2()],
        ["outboundRateLimiter", option(getRateLimiterSerializer2())],
        ["inboundRateLimiter", option(getRateLimiterSerializer2())],
        ["feeBps", option(u16())],
        ["bump", u8()],
        ["isEndpointV1", bool()]
      ],
      { description: "PeerConfigAccountData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([181, 157, 86, 198, 33, 193, 94, 203])
    })
  );
}
function deserializePeerConfig2(rawAccount) {
  return deserializeAccount(rawAccount, getPeerConfigAccountDataSerializer2());
}
async function fetchPeerConfig2(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  assertAccountExists(maybeAccount, "PeerConfig");
  return deserializePeerConfig2(maybeAccount);
}
async function safeFetchPeerConfig2(context, publicKey3, options) {
  const maybeAccount = await context.rpc.getAccount(
    publicKey$1(publicKey3, false),
    options
  );
  return maybeAccount.exists ? deserializePeerConfig2(maybeAccount) : null;
}
async function fetchAllPeerConfig2(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, "PeerConfig");
    return deserializePeerConfig2(maybeAccount);
  });
}
async function safeFetchAllPeerConfig2(context, publicKeys, options) {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => publicKey$1(key, false)),
    options
  );
  return maybeAccounts.filter((maybeAccount) => maybeAccount.exists).map((maybeAccount) => deserializePeerConfig2(maybeAccount));
}
function getPeerConfigGpaBuilder2(context) {
  const programId = context.programs.getPublicKey("oft", "");
  return gpaBuilder(context, programId).registerFields({
    discriminator: [0, bytes({ size: 8 })],
    peerAddress: [8, bytes({ size: 32 })],
    enforcedOptions: [40, getEnforcedOptionsSerializer2()],
    outboundRateLimiter: [null, option(getRateLimiterSerializer2())],
    inboundRateLimiter: [null, option(getRateLimiterSerializer2())],
    feeBps: [null, option(u16())],
    bump: [null, u8()],
    isEndpointV1: [null, bool()]
  }).deserializeUsing((account) => deserializePeerConfig2(account)).whereField(
    "discriminator",
    new Uint8Array([181, 157, 86, 198, 33, 193, 94, 203])
  );
}

// src/generated/oft202/errors/index.ts
var errors_exports2 = {};
__export(errors_exports2, {
  InvalidDecimalsError: () => InvalidDecimalsError2,
  InvalidFeeError: () => InvalidFeeError2,
  InvalidMintAuthorityError: () => InvalidMintAuthorityError2,
  InvalidSenderError: () => InvalidSenderError2,
  InvalidTokenDestError: () => InvalidTokenDestError2,
  PausedError: () => PausedError2,
  RateLimitExceededError: () => RateLimitExceededError2,
  SlippageExceededError: () => SlippageExceededError2,
  UnauthorizedError: () => UnauthorizedError2,
  getOftErrorFromCode: () => getOftErrorFromCode2,
  getOftErrorFromName: () => getOftErrorFromName2
});
var codeToErrorMap2 = /* @__PURE__ */ new Map();
var nameToErrorMap2 = /* @__PURE__ */ new Map();
var UnauthorizedError2 = class extends ProgramError {
  // 6000
  constructor(program, cause) {
    super("", program, cause);
    this.name = "Unauthorized";
    this.code = 6e3;
  }
};
codeToErrorMap2.set(6e3, UnauthorizedError2);
nameToErrorMap2.set("Unauthorized", UnauthorizedError2);
var InvalidSenderError2 = class extends ProgramError {
  // 6001
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidSender";
    this.code = 6001;
  }
};
codeToErrorMap2.set(6001, InvalidSenderError2);
nameToErrorMap2.set("InvalidSender", InvalidSenderError2);
var InvalidDecimalsError2 = class extends ProgramError {
  // 6002
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidDecimals";
    this.code = 6002;
  }
};
codeToErrorMap2.set(6002, InvalidDecimalsError2);
nameToErrorMap2.set("InvalidDecimals", InvalidDecimalsError2);
var SlippageExceededError2 = class extends ProgramError {
  // 6003
  constructor(program, cause) {
    super("", program, cause);
    this.name = "SlippageExceeded";
    this.code = 6003;
  }
};
codeToErrorMap2.set(6003, SlippageExceededError2);
nameToErrorMap2.set("SlippageExceeded", SlippageExceededError2);
var InvalidTokenDestError2 = class extends ProgramError {
  // 6004
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidTokenDest";
    this.code = 6004;
  }
};
codeToErrorMap2.set(6004, InvalidTokenDestError2);
nameToErrorMap2.set("InvalidTokenDest", InvalidTokenDestError2);
var RateLimitExceededError2 = class extends ProgramError {
  // 6005
  constructor(program, cause) {
    super("", program, cause);
    this.name = "RateLimitExceeded";
    this.code = 6005;
  }
};
codeToErrorMap2.set(6005, RateLimitExceededError2);
nameToErrorMap2.set("RateLimitExceeded", RateLimitExceededError2);
var InvalidFeeError2 = class extends ProgramError {
  // 6006
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidFee";
    this.code = 6006;
  }
};
codeToErrorMap2.set(6006, InvalidFeeError2);
nameToErrorMap2.set("InvalidFee", InvalidFeeError2);
var InvalidMintAuthorityError2 = class extends ProgramError {
  // 6007
  constructor(program, cause) {
    super("", program, cause);
    this.name = "InvalidMintAuthority";
    this.code = 6007;
  }
};
codeToErrorMap2.set(6007, InvalidMintAuthorityError2);
nameToErrorMap2.set("InvalidMintAuthority", InvalidMintAuthorityError2);
var PausedError2 = class extends ProgramError {
  // 6008
  constructor(program, cause) {
    super("", program, cause);
    this.name = "Paused";
    this.code = 6008;
  }
};
codeToErrorMap2.set(6008, PausedError2);
nameToErrorMap2.set("Paused", PausedError2);
function getOftErrorFromCode2(code, program, cause) {
  const constructor = codeToErrorMap2.get(code);
  return constructor ? new constructor(program, cause) : null;
}
function getOftErrorFromName2(name, program, cause) {
  const constructor = nameToErrorMap2.get(name);
  return constructor ? new constructor(program, cause) : null;
}

// src/generated/oft202/instructions/index.ts
var instructions_exports2 = {};
__export(instructions_exports2, {
  getInitOftInstructionDataSerializer: () => getInitOftInstructionDataSerializer2,
  getLzReceiveInstructionDataSerializer: () => getLzReceiveInstructionDataSerializer2,
  getLzReceiveTypesInstructionDataSerializer: () => getLzReceiveTypesInstructionDataSerializer2,
  getQuoteOftInstructionDataSerializer: () => getQuoteOftInstructionDataSerializer2,
  getQuoteSendInstructionDataSerializer: () => getQuoteSendInstructionDataSerializer2,
  getSendInstructionDataSerializer: () => getSendInstructionDataSerializer2,
  getSetOftConfigInstructionDataSerializer: () => getSetOftConfigInstructionDataSerializer2,
  getSetPauseInstructionDataSerializer: () => getSetPauseInstructionDataSerializer2,
  getSetPeerConfigInstructionDataSerializer: () => getSetPeerConfigInstructionDataSerializer2,
  getWithdrawFeeInstructionDataSerializer: () => getWithdrawFeeInstructionDataSerializer2,
  initOft: () => initOft3,
  lzReceive: () => lzReceive2,
  lzReceiveTypes: () => lzReceiveTypes2,
  quoteOft: () => quoteOft3,
  quoteSend: () => quoteSend2,
  send: () => send3,
  setOftConfig: () => setOftConfig2,
  setPause: () => setPause2,
  setPeerConfig: () => setPeerConfig3,
  withdrawFee: () => withdrawFee3
});

// src/generated/oft202/shared/index.ts
var shared_exports2 = {};
__export(shared_exports2, {
  expectPda: () => expectPda2,
  expectPublicKey: () => expectPublicKey2,
  expectSome: () => expectSome2,
  getAccountMetasAndSigners: () => getAccountMetasAndSigners2
});
function expectSome2(value) {
  if (value == null) {
    throw new Error("Expected a value but received null or undefined.");
  }
  return value;
}
function expectPublicKey2(value) {
  if (!value) {
    throw new Error("Expected a PublicKey.");
  }
  return publicKey$1(value, false);
}
function expectPda2(value) {
  if (!value || !Array.isArray(value) || !isPda(value)) {
    throw new Error("Expected a PDA.");
  }
  return value;
}
function getAccountMetasAndSigners2(accounts, optionalAccountStrategy, programId) {
  const keys = [];
  const signers = [];
  accounts.forEach((account) => {
    if (!account.value) {
      if (optionalAccountStrategy === "omitted") return;
      keys.push({ pubkey: programId, isSigner: false, isWritable: false });
      return;
    }
    if (isSigner(account.value)) {
      signers.push(account.value);
    }
    keys.push({
      pubkey: publicKey$1(account.value, false),
      isSigner: isSigner(account.value),
      isWritable: account.isWritable
    });
  });
  return [keys, signers];
}

// src/generated/oft202/instructions/initOft.ts
function getInitOftInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["oftType", getOFTTypeSerializer2()],
        ["admin", publicKey()],
        ["sharedDecimals", u8()],
        ["endpointProgram", option(publicKey())]
      ],
      { description: "InitOftInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([182, 169, 147, 16, 201, 45, 76, 23])
    })
  );
}
function initOft3(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true,
      value: input.payer ?? null
    },
    oftStore: {
      index: 1,
      isWritable: true,
      value: input.oftStore ?? null
    },
    lzReceiveTypesAccounts: {
      index: 2,
      isWritable: true,
      value: input.lzReceiveTypesAccounts ?? null
    },
    tokenMint: {
      index: 3,
      isWritable: false,
      value: input.tokenMint ?? null
    },
    tokenEscrow: {
      index: 4,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    tokenProgram: {
      index: 5,
      isWritable: false,
      value: input.tokenProgram ?? null
    },
    systemProgram: {
      index: 6,
      isWritable: false,
      value: input.systemProgram ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      "splSystem",
      "11111111111111111111111111111111"
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getInitOftInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getLzReceiveInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["params", getLzReceiveParamsSerializer2()]
      ],
      { description: "LzReceiveInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([8, 179, 120, 109, 33, 118, 189, 80])
    })
  );
}
function lzReceive2(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true,
      value: input.payer ?? null
    },
    peer: { index: 1, isWritable: true, value: input.peer ?? null },
    oftStore: {
      index: 2,
      isWritable: true,
      value: input.oftStore ?? null
    },
    tokenEscrow: {
      index: 3,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    toAddress: {
      index: 4,
      isWritable: false,
      value: input.toAddress ?? null
    },
    tokenDest: {
      index: 5,
      isWritable: true,
      value: input.tokenDest ?? null
    },
    tokenMint: {
      index: 6,
      isWritable: true,
      value: input.tokenMint ?? null
    },
    mintAuthority: {
      index: 7,
      isWritable: false,
      value: input.mintAuthority ?? null
    },
    tokenProgram: {
      index: 8,
      isWritable: false,
      value: input.tokenProgram ?? null
    },
    associatedTokenProgram: {
      index: 9,
      isWritable: false,
      value: input.associatedTokenProgram ?? null
    },
    systemProgram: {
      index: 10,
      isWritable: false,
      value: input.systemProgram ?? null
    },
    eventAuthority: {
      index: 11,
      isWritable: false,
      value: input.eventAuthority ?? null
    },
    program: {
      index: 12,
      isWritable: false,
      value: input.program ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      "splSystem",
      "11111111111111111111111111111111"
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getLzReceiveInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getLzReceiveTypesInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["params", getLzReceiveParamsSerializer2()]
      ],
      { description: "LzReceiveTypesInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([221, 17, 246, 159, 248, 128, 31, 96])
    })
  );
}
function lzReceiveTypes2(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    oftStore: {
      index: 0,
      isWritable: false,
      value: input.oftStore ?? null
    },
    tokenMint: {
      index: 1,
      isWritable: false,
      value: input.tokenMint ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getLzReceiveTypesInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getQuoteOftInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["dstEid", u32()],
        ["to", bytes({ size: 32 })],
        ["amountLd", u64()],
        ["minAmountLd", u64()],
        ["options", bytes({ size: u32() })],
        ["composeMsg", option(bytes({ size: u32() }))],
        ["payInLzToken", bool()]
      ],
      { description: "QuoteOftInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([179, 255, 92, 202, 251, 82, 82, 118])
    })
  );
}
function quoteOft3(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    oftStore: {
      index: 0,
      isWritable: false,
      value: input.oftStore ?? null
    },
    peer: { index: 1, isWritable: false, value: input.peer ?? null },
    tokenMint: {
      index: 2,
      isWritable: false,
      value: input.tokenMint ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getQuoteOftInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getQuoteSendInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["dstEid", u32()],
        ["to", bytes({ size: 32 })],
        ["amountLd", u64()],
        ["minAmountLd", u64()],
        ["options", bytes({ size: u32() })],
        ["payInLzToken", bool()],
        ["composeParams", option(getComposeParamsSerializer())]
      ],
      { description: "QuoteSendInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([207, 0, 49, 214, 160, 211, 76, 211])
    })
  );
}
function quoteSend2(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    oftStore: {
      index: 0,
      isWritable: false,
      value: input.oftStore ?? null
    },
    peer: { index: 1, isWritable: false, value: input.peer ?? null },
    tokenMint: {
      index: 2,
      isWritable: false,
      value: input.tokenMint ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getQuoteSendInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSendInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["dstEid", u32()],
        ["to", bytes({ size: 32 })],
        ["amountLd", u64()],
        ["minAmountLd", u64()],
        ["options", bytes({ size: u32() })],
        ["nativeFee", u64()],
        ["lzTokenFee", u64()],
        ["composeParams", option(getComposeParamsSerializer())]
      ],
      { description: "SendInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([102, 251, 20, 187, 65, 75, 12, 69])
    })
  );
}
function send3(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false,
      value: input.signer ?? null
    },
    peer: { index: 1, isWritable: true, value: input.peer ?? null },
    oftStore: {
      index: 2,
      isWritable: true,
      value: input.oftStore ?? null
    },
    tokenSource: {
      index: 3,
      isWritable: true,
      value: input.tokenSource ?? null
    },
    tokenEscrow: {
      index: 4,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    tokenMint: {
      index: 5,
      isWritable: true,
      value: input.tokenMint ?? null
    },
    tokenProgram: {
      index: 6,
      isWritable: false,
      value: input.tokenProgram ?? null
    },
    eventAuthority: {
      index: 7,
      isWritable: false,
      value: input.eventAuthority ?? null
    },
    program: {
      index: 8,
      isWritable: false,
      value: input.program ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSendInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSetOftConfigInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["params", getSetOFTConfigParamsSerializer2()]
      ],
      { description: "SetOftConfigInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([55, 126, 87, 217, 159, 66, 24, 194])
    })
  );
}
function setOftConfig2(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false,
      value: input.admin ?? null
    },
    oftStore: {
      index: 1,
      isWritable: true,
      value: input.oftStore ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSetOftConfigInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSetPauseInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["paused", bool()]
      ],
      { description: "SetPauseInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([63, 32, 154, 2, 56, 103, 79, 45])
    })
  );
}
function setPause2(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    signer: {
      index: 0,
      isWritable: false,
      value: input.signer ?? null
    },
    oftStore: {
      index: 1,
      isWritable: true,
      value: input.oftStore ?? null
    }
  };
  const resolvedArgs = { ...input };
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSetPauseInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getSetPeerConfigInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["remoteEid", u32()],
        ["config", getPeerConfigParamSerializer2()]
      ],
      { description: "SetPeerConfigInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([79, 187, 168, 57, 139, 140, 93, 47])
    })
  );
}
function setPeerConfig3(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: true,
      value: input.admin ?? null
    },
    peer: { index: 1, isWritable: true, value: input.peer ?? null },
    oftStore: {
      index: 2,
      isWritable: false,
      value: input.oftStore ?? null
    },
    systemProgram: {
      index: 3,
      isWritable: false,
      value: input.systemProgram ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      "splSystem",
      "11111111111111111111111111111111"
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getSetPeerConfigInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}
function getWithdrawFeeInstructionDataSerializer2() {
  return mapSerializer(
    struct(
      [
        ["discriminator", bytes({ size: 8 })],
        ["feeLd", u64()]
      ],
      { description: "WithdrawFeeInstructionData" }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([14, 122, 231, 218, 31, 238, 223, 150])
    })
  );
}
function withdrawFee3(context, input) {
  const programId = context.programs.getPublicKey("oft", "");
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false,
      value: input.admin ?? null
    },
    oftStore: {
      index: 1,
      isWritable: false,
      value: input.oftStore ?? null
    },
    tokenMint: {
      index: 2,
      isWritable: false,
      value: input.tokenMint ?? null
    },
    tokenEscrow: {
      index: 3,
      isWritable: true,
      value: input.tokenEscrow ?? null
    },
    tokenDest: {
      index: 4,
      isWritable: true,
      value: input.tokenDest ?? null
    },
    tokenProgram: {
      index: 5,
      isWritable: false,
      value: input.tokenProgram ?? null
    }
  };
  const resolvedArgs = { ...input };
  if (!resolvedAccounts.tokenProgram.value) {
    resolvedAccounts.tokenProgram.value = context.programs.getPublicKey(
      "splToken",
      "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    );
    resolvedAccounts.tokenProgram.isWritable = false;
  }
  const orderedAccounts = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);
  const [keys, signers] = getAccountMetasAndSigners2(
    orderedAccounts,
    "programId",
    programId
  );
  const data = getWithdrawFeeInstructionDataSerializer2().serialize(
    resolvedArgs
  );
  const bytesCreatedOnChain = 0;
  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain }
  ]);
}

// src/generated/oft202/programs/index.ts
var programs_exports2 = {};
__export(programs_exports2, {
  OFT_PROGRAM_ID: () => OFT_PROGRAM_ID2,
  createOftProgram: () => createOftProgram2,
  getOftProgram: () => getOftProgram2,
  getOftProgramId: () => getOftProgramId2
});

// src/generated/oft202/programs/oft.ts
var OFT_PROGRAM_ID2 = "";
function createOftProgram2() {
  return {
    name: "oft",
    publicKey: OFT_PROGRAM_ID2,
    getErrorFromCode(code, cause) {
      return getOftErrorFromCode2(code, this, cause);
    },
    getErrorFromName(name, cause) {
      return getOftErrorFromName2(name, this, cause);
    },
    isOnCluster() {
      return true;
    }
  };
}
function getOftProgram2(context, clusterFilter) {
  return context.programs.get("oft", clusterFilter);
}
function getOftProgramId2(context, clusterFilter) {
  return context.programs.getPublicKey("oft", OFT_PROGRAM_ID2, clusterFilter);
}

// src/oft202.ts
var { ENDPOINT_PROGRAM_ID: ENDPOINT_PROGRAM_ID2 } = EndpointProgram;
var PROGRAM_NAME = "oft";
function createOFTProgramRepo2(oftProgram, rpc) {
  if (rpc === void 0) {
    rpc = createNullRpc();
    rpc.getCluster = () => "custom";
  }
  return createDefaultProgramRepository({ rpc }, [
    {
      name: PROGRAM_NAME,
      publicKey: oftProgram,
      getErrorFromCode(code, cause) {
        return getOftErrorFromCode2(code, this, cause);
      },
      getErrorFromName(name, cause) {
        return getOftErrorFromName2(name, this, cause);
      },
      isOnCluster() {
        return true;
      }
    }
  ]);
}
function initOft4(accounts, oftType, sharedDecimals = OFT_DECIMALS, programs) {
  const programsRepo = createOFTProgramRepo2(programs.oft);
  const deriver = new OftPDA(programsRepo.getPublicKey(PROGRAM_NAME));
  const endpoint = new EndpointProgram.Endpoint(programs.endpoint ?? ENDPOINT_PROGRAM_ID2);
  const { payer, admin, mint, escrow } = accounts;
  const [oftStore] = deriver.oftStore(escrow.publicKey);
  const [lzReceiveTypes3] = deriver.lzReceiveTypesAccounts(oftStore);
  const txBuilder = initOft3(
    {
      payer,
      programs: programsRepo
    },
    {
      // accounts
      oftStore,
      lzReceiveTypesAccounts: lzReceiveTypes3,
      tokenMint: mint,
      tokenEscrow: escrow,
      tokenProgram: programs.token ?? fromWeb3JsPublicKey(TOKEN_PROGRAM_ID),
      // params
      oftType,
      admin,
      sharedDecimals,
      endpointProgram: endpoint.programId
    }
  );
  const retval = txBuilder.addRemainingAccounts(
    endpoint.getRegisterOappIxAccountMetaForCPI(payer.publicKey, oftStore).map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  ).items[0];
  retval.signers = [payer, escrow];
  return retval;
}
function setOFTConfig2(accounts, params, programs) {
  let actualParams;
  const { oftStore, admin } = accounts;
  const remainingAccounts = [];
  if (params.__kind === "Admin") {
    if (params.admin === void 0) {
      throw new Error("Admin is required");
    }
    actualParams = {
      __kind: "Admin",
      fields: [params.admin]
    };
  } else if (params.__kind === "Delegate") {
    if (params.delegate === void 0) {
      throw new Error("Delegate is required");
    }
    actualParams = {
      __kind: "Delegate",
      fields: [params.delegate]
    };
    const endpointProgram = programs.endpoint ?? ENDPOINT_PROGRAM_ID2;
    const endpointSDK = new EndpointProgram.Endpoint(endpointProgram);
    const keys = endpointSDK.getSetDelegateIxAccountMetaForCPI(oftStore);
    for (const acc of keys) {
      acc.isSigner = false;
    }
    remainingAccounts.push(
      {
        pubkey: endpointProgram,
        isSigner: false,
        isWritable: false
      },
      ...keys
    );
  } else if (params.__kind === "DefaultFee") {
    if (params.defaultFee === void 0) {
      throw new Error("DefaultFee is required");
    }
    actualParams = {
      __kind: "DefaultFee",
      fields: [params.defaultFee]
    };
  } else if (params.__kind === "Paused") {
    if (params.paused === void 0) {
      throw new Error("Paused is required");
    }
    actualParams = {
      __kind: "Paused",
      fields: [params.paused]
    };
  } else if (params.__kind === "Pauser") {
    actualParams = {
      __kind: "Pauser",
      fields: [params.pauser ? some(params.pauser) : none()]
    };
  } else {
    actualParams = {
      __kind: "Unpauser",
      fields: [params.unpauser ? some(params.unpauser) : none()]
    };
  }
  const txBuilder = setOftConfig2(
    { programs: createOFTProgramRepo2(programs.oft) },
    {
      admin,
      oftStore,
      params: actualParams
    }
  );
  return txBuilder.addRemainingAccounts(
    remainingAccounts.map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  ).items[0];
}
function setPeerConfig4(accounts, param, oftProgramId) {
  const programsRepo = typeof oftProgramId === "string" ? createOFTProgramRepo2(oftProgramId) : oftProgramId;
  const { remote: remoteId } = param;
  if (remoteId % 3e4 == 0) {
    throw new Error("Invalid remote ID");
  }
  const { admin, oftStore } = accounts;
  const [peerPda] = new OftPDA(programsRepo.getPublicKey(PROGRAM_NAME)).peer(oftStore, remoteId);
  let config;
  if (param.__kind === "PeerAddress") {
    if (param.peer.length !== 32) {
      throw new Error("Peer must be 32 bytes (left-padded with zeroes)");
    }
    config = peerConfigParam2("PeerAddress", [param.peer]);
  } else if (param.__kind === "FeeBps") {
    config = { __kind: "FeeBps", fields: [some(param.feeBps)] };
  } else if (param.__kind === "EnforcedOptions") {
    config = {
      __kind: "EnforcedOptions",
      send: param.send,
      sendAndCall: param.sendAndCall
    };
  } else if (param.__kind === "OutboundRateLimit" || param.__kind === "InboundRateLimit") {
    config = {
      __kind: param.__kind,
      fields: [
        param.rateLimit ? some({
          refillPerSecond: some(param.rateLimit.refillPerSecond),
          capacity: some(param.rateLimit.capacity)
        }) : null
      ]
    };
  } else if (param.__kind === "IsEndpointV1") {
    config = { __kind: "IsEndpointV1", fields: [param.isEndpointV1] };
  } else {
    throw new Error("Invalid peer config");
  }
  return setPeerConfig3(
    { programs: programsRepo },
    {
      admin,
      peer: peerPda,
      oftStore,
      // params
      remoteEid: remoteId,
      config
    }
  ).items[0];
}
function withdrawFee4(accounts, amount, programs) {
  const { admin, mint, escrow, dest } = accounts;
  const programsRepo = typeof programs.oft === "string" ? createOFTProgramRepo2(programs.oft) : programs.oft;
  const [oftStore] = new OftPDA(programsRepo.getPublicKey("oft")).oftStore(escrow);
  return withdrawFee3(
    { programs: programsRepo },
    {
      admin,
      tokenEscrow: escrow,
      tokenDest: dest,
      tokenProgram: programs.token ?? fromWeb3JsPublicKey(TOKEN_PROGRAM_ID),
      oftStore,
      tokenMint: mint,
      // params
      feeLd: amount
    }
  ).items[0];
}
async function send4(rpc, accounts, sendParams, programs, remainingAccounts) {
  const { payer, tokenMint, tokenEscrow, tokenSource } = accounts;
  const { dstEid, to, amountLd, minAmountLd, options, composeParams, nativeFee, lzTokenFee } = sendParams;
  const programsRepo = typeof programs.oft === "string" ? createOFTProgramRepo2(programs.oft) : programs.oft;
  const oftProgramId = programsRepo.getPublicKey(PROGRAM_NAME);
  const deriver = new OftPDA(oftProgramId);
  const [oftStore] = deriver.oftStore(tokenEscrow);
  const [peer] = deriver.peer(oftStore, dstEid);
  if (remainingAccounts === void 0 || remainingAccounts.length === 0) {
    const peerAddr = accounts.peerAddr ?? await fetchPeerConfig2({ rpc }, peer).then((peerInfo) => peerInfo.peerAddress);
    const endpoint = new EndpointProgram.Endpoint(programs.endpoint ?? ENDPOINT_PROGRAM_ID2);
    const msgLibProgram = await getSendLibraryProgram2(rpc, endpoint, payer.publicKey, oftStore, dstEid);
    const packetPath = {
      dstEid,
      sender: oftStore,
      receiver: peerAddr
    };
    remainingAccounts = await endpoint.getSendIXAccountMetaForCPI(rpc, payer.publicKey, {
      path: packetPath,
      msgLibProgram
    });
  }
  const [eventAuthorityPDA] = new EventPDA(oftProgramId).eventAuthority();
  const tokenProgram = programs.token ?? fromWeb3JsPublicKey(TOKEN_PROGRAM_ID);
  const txBuilder = send3(
    { programs: programsRepo },
    {
      signer: payer,
      peer,
      oftStore,
      tokenSource,
      tokenEscrow,
      tokenMint,
      tokenProgram,
      eventAuthority: eventAuthorityPDA,
      program: oftProgramId,
      // params
      dstEid,
      to,
      amountLd,
      minAmountLd,
      options: options ?? new Uint8Array(),
      nativeFee,
      lzTokenFee: lzTokenFee ?? 0,
      composeParams: composeParams ?? null
    }
  );
  return txBuilder.addRemainingAccounts(
    remainingAccounts.map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  ).items[0];
}
async function quote2(rpc, accounts, quoteParams, programs, remainingAccounts, addressLookupTable) {
  const { dstEid, to, amountLd, minAmountLd, options, payInLzToken, composeParams } = quoteParams;
  const { payer, tokenMint, tokenEscrow } = accounts;
  const programsRepo = typeof programs.oft === "string" ? createOFTProgramRepo2(programs.oft) : programs.oft;
  const deriver = new OftPDA(programsRepo.getPublicKey(PROGRAM_NAME));
  const [oftStore] = deriver.oftStore(tokenEscrow);
  const [peer] = deriver.peer(oftStore, dstEid);
  if (remainingAccounts === void 0 || remainingAccounts.length === 0) {
    const peerAddr = accounts.peerAddr ?? await fetchPeerConfig2({ rpc }, peer).then((peerInfo) => peerInfo.peerAddress);
    const endpoint = new EndpointProgram.Endpoint(programs.endpoint ?? ENDPOINT_PROGRAM_ID2);
    const messageLib = await getSendLibraryProgram2(rpc, endpoint, payer, oftStore, dstEid);
    remainingAccounts = await endpoint.getQuoteIXAccountMetaForCPI(rpc, payer, {
      path: { sender: oftStore, dstEid, receiver: peerAddr },
      msgLibProgram: messageLib
    });
  }
  let txBuilder = quoteSend2(
    { programs: programsRepo },
    {
      oftStore,
      peer,
      tokenMint,
      // params
      dstEid,
      to,
      amountLd,
      minAmountLd,
      options: options ?? new Uint8Array(),
      payInLzToken: payInLzToken ?? false,
      composeParams: composeParams ?? null
    }
  );
  txBuilder = txBuilder.addRemainingAccounts(
    // Get remaining accounts from msgLib(simple_msgLib or uln)
    remainingAccounts.map((acc) => {
      return {
        pubkey: acc.pubkey,
        isSigner: acc.isSigner,
        isWritable: acc.isWritable
      };
    })
  );
  const web3Ix = toWeb3JsInstruction(txBuilder.getInstructions()[0]);
  const modifyComputeUnits = ComputeBudgetProgram.setComputeUnitLimit({
    units: 1e6
  });
  return simulateWeb3JsTransaction(
    rpc,
    [modifyComputeUnits, web3Ix],
    web3Ix.programId,
    toWeb3JsPublicKey(payer),
    EndpointProgram.types.getMessagingFeeSerializer(),
    "confirmed",
    void 0,
    addressLookupTable === void 0 ? void 0 : toWeb3JsPublicKey(addressLookupTable)
  );
}
async function quoteOft4(rpc, accounts, quoteParams, oftProgram) {
  const { payer, tokenMint, tokenEscrow } = accounts;
  const { dstEid, to, amountLd, minAmountLd, options, payInLzToken, composeMsg } = quoteParams;
  const programsRepo = typeof oftProgram === "string" ? createOFTProgramRepo2(oftProgram) : oftProgram;
  const deriver = new OftPDA(programsRepo.getPublicKey(PROGRAM_NAME));
  const [oftStore] = deriver.oftStore(tokenEscrow);
  const [peer] = deriver.peer(oftStore, dstEid);
  const ix = quoteOft3(
    { programs: programsRepo },
    {
      oftStore,
      peer,
      tokenMint,
      // params
      dstEid,
      to,
      amountLd,
      minAmountLd,
      options: options ?? new Uint8Array(),
      payInLzToken: payInLzToken ?? false,
      composeMsg: composeMsg ?? null
    }
  ).getInstructions()[0];
  const web3Ix = toWeb3JsInstruction(ix);
  return simulateWeb3JsTransaction(
    rpc.getEndpoint(),
    [web3Ix],
    web3Ix.programId,
    toWeb3JsPublicKey(payer),
    getQuoteOFTResultSerializer2(),
    "confirmed"
  );
}
async function getSendLibraryProgram2(rpc, endpoint, payer, oftStore, remoteEid) {
  const sendLibInfo = await endpoint.getSendLibrary(rpc, oftStore, remoteEid);
  if (!sendLibInfo.programId) {
    throw new Error("Send library not initialized or blocked message library");
  }
  const { programId: msgLibProgram } = sendLibInfo;
  const msgLibVersion = await endpoint.getMessageLibVersion(rpc, payer, msgLibProgram);
  if (msgLibVersion.major.toString() === "0" && msgLibVersion.minor == 0 && msgLibVersion.endpointVersion == 2) {
    return new SimpleMessageLibProgram.SimpleMessageLib(msgLibProgram);
  } else if (msgLibVersion.major.toString() === "3" && msgLibVersion.minor == 0 && msgLibVersion.endpointVersion == 2) {
    return new UlnProgram.Uln(msgLibProgram);
  }
  throw new Error(`Unsupported message library version: ${JSON.stringify(msgLibVersion, null, 2)}`);
}

export { OFT_DECIMALS, OFT_SEED, OftPDA, PEER_SEED, oft302_exports as oft, oft202_exports as oft202, oft302_exports as oft302 };
