// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/errors/index.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * OnlyWhitelistedCaller: ''
 *
 * @category Errors
 * @category generated
 */
export class OnlyWhitelistedCallerError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'OnlyWhitelistedCaller'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, OnlyWhitelistedCallerError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new OnlyWhitelistedCallerError())
createErrorFromNameLookup.set(
  'OnlyWhitelistedCaller',
  () => new OnlyWhitelistedCallerError()
)

/**
 * InsufficientFee: ''
 *
 * @category Errors
 * @category generated
 */
export class InsufficientFeeError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'InsufficientFee'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InsufficientFeeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new InsufficientFeeError())
createErrorFromNameLookup.set(
  'InsufficientFee',
  () => new InsufficientFeeError()
)

/**
 * InvalidAmount: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidAmountError extends Error {
  readonly code: number = 0x1772
  readonly name: string = 'InvalidAmount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidAmountError)
    }
  }
}

createErrorFromCodeLookup.set(0x1772, () => new InvalidAmountError())
createErrorFromNameLookup.set('InvalidAmount', () => new InvalidAmountError())

/**
 * InvalidConfigType: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidConfigTypeError extends Error {
  readonly code: number = 0x1773
  readonly name: string = 'InvalidConfigType'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidConfigTypeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1773, () => new InvalidConfigTypeError())
createErrorFromNameLookup.set(
  'InvalidConfigType',
  () => new InvalidConfigTypeError()
)

/**
 * InvalidLzTokenMint: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidLzTokenMintError extends Error {
  readonly code: number = 0x1774
  readonly name: string = 'InvalidLzTokenMint'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidLzTokenMintError)
    }
  }
}

createErrorFromCodeLookup.set(0x1774, () => new InvalidLzTokenMintError())
createErrorFromNameLookup.set(
  'InvalidLzTokenMint',
  () => new InvalidLzTokenMintError()
)

/**
 * LzTokenUnavailable: ''
 *
 * @category Errors
 * @category generated
 */
export class LzTokenUnavailableError extends Error {
  readonly code: number = 0x1775
  readonly name: string = 'LzTokenUnavailable'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, LzTokenUnavailableError)
    }
  }
}

createErrorFromCodeLookup.set(0x1775, () => new LzTokenUnavailableError())
createErrorFromNameLookup.set(
  'LzTokenUnavailable',
  () => new LzTokenUnavailableError()
)

/**
 * SendReentrancy: ''
 *
 * @category Errors
 * @category generated
 */
export class SendReentrancyError extends Error {
  readonly code: number = 0x1776
  readonly name: string = 'SendReentrancy'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, SendReentrancyError)
    }
  }
}

createErrorFromCodeLookup.set(0x1776, () => new SendReentrancyError())
createErrorFromNameLookup.set('SendReentrancy', () => new SendReentrancyError())

/**
 * OnlyRevert: ''
 *
 * @category Errors
 * @category generated
 */
export class OnlyRevertError extends Error {
  readonly code: number = 0x1777
  readonly name: string = 'OnlyRevert'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, OnlyRevertError)
    }
  }
}

createErrorFromCodeLookup.set(0x1777, () => new OnlyRevertError())
createErrorFromNameLookup.set('OnlyRevert', () => new OnlyRevertError())

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
