// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/accounts/composeMessageState.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bool,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type ComposeMessageState = Account<ComposeMessageStateAccountData>;

export type ComposeMessageStateAccountData = {
  discriminator: Uint8Array;
  received: boolean;
  bump: number;
};

export type ComposeMessageStateAccountDataArgs = {
  received: boolean;
  bump: number;
};

export function getComposeMessageStateAccountDataSerializer(): Serializer<
  ComposeMessageStateAccountDataArgs,
  ComposeMessageStateAccountData
> {
  return mapSerializer<
    ComposeMessageStateAccountDataArgs,
    any,
    ComposeMessageStateAccountData
  >(
    struct<ComposeMessageStateAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['received', bool()],
        ['bump', u8()],
      ],
      { description: 'ComposeMessageStateAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([55, 107, 79, 254, 243, 22, 172, 240]),
    })
  ) as Serializer<
    ComposeMessageStateAccountDataArgs,
    ComposeMessageStateAccountData
  >;
}

export function deserializeComposeMessageState(
  rawAccount: RpcAccount
): ComposeMessageState {
  return deserializeAccount(
    rawAccount,
    getComposeMessageStateAccountDataSerializer()
  );
}

export async function fetchComposeMessageState(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<ComposeMessageState> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'ComposeMessageState');
  return deserializeComposeMessageState(maybeAccount);
}

export async function safeFetchComposeMessageState(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<ComposeMessageState | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists
    ? deserializeComposeMessageState(maybeAccount)
    : null;
}

export async function fetchAllComposeMessageState(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<ComposeMessageState[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'ComposeMessageState');
    return deserializeComposeMessageState(maybeAccount);
  });
}

export async function safeFetchAllComposeMessageState(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<ComposeMessageState[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) =>
      deserializeComposeMessageState(maybeAccount as RpcAccount)
    );
}

export function getComposeMessageStateGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      received: boolean;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      received: [8, bool()],
      bump: [9, u8()],
    })
    .deserializeUsing<ComposeMessageState>((account) =>
      deserializeComposeMessageState(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([55, 107, 79, 254, 243, 22, 172, 240])
    );
}

export function getComposeMessageStateSize(): number {
  return 10;
}
