// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/accounts/messageLibInfo.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  MessageLibType,
  MessageLibTypeArgs,
  getMessageLibTypeSerializer,
} from '../types';

export type MessageLibInfo = Account<MessageLibInfoAccountData>;

export type MessageLibInfoAccountData = {
  discriminator: Uint8Array;
  messageLibType: MessageLibType;
  bump: number;
  messageLibBump: number;
};

export type MessageLibInfoAccountDataArgs = {
  messageLibType: MessageLibTypeArgs;
  bump: number;
  messageLibBump: number;
};

export function getMessageLibInfoAccountDataSerializer(): Serializer<
  MessageLibInfoAccountDataArgs,
  MessageLibInfoAccountData
> {
  return mapSerializer<
    MessageLibInfoAccountDataArgs,
    any,
    MessageLibInfoAccountData
  >(
    struct<MessageLibInfoAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['messageLibType', getMessageLibTypeSerializer()],
        ['bump', u8()],
        ['messageLibBump', u8()],
      ],
      { description: 'MessageLibInfoAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([103, 102, 218, 28, 204, 135, 71, 14]),
    })
  ) as Serializer<MessageLibInfoAccountDataArgs, MessageLibInfoAccountData>;
}

export function deserializeMessageLibInfo(
  rawAccount: RpcAccount
): MessageLibInfo {
  return deserializeAccount(
    rawAccount,
    getMessageLibInfoAccountDataSerializer()
  );
}

export async function fetchMessageLibInfo(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<MessageLibInfo> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'MessageLibInfo');
  return deserializeMessageLibInfo(maybeAccount);
}

export async function safeFetchMessageLibInfo(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<MessageLibInfo | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeMessageLibInfo(maybeAccount) : null;
}

export async function fetchAllMessageLibInfo(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<MessageLibInfo[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'MessageLibInfo');
    return deserializeMessageLibInfo(maybeAccount);
  });
}

export async function safeFetchAllMessageLibInfo(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<MessageLibInfo[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) =>
      deserializeMessageLibInfo(maybeAccount as RpcAccount)
    );
}

export function getMessageLibInfoGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      messageLibType: MessageLibTypeArgs;
      bump: number;
      messageLibBump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      messageLibType: [8, getMessageLibTypeSerializer()],
      bump: [9, u8()],
      messageLibBump: [10, u8()],
    })
    .deserializeUsing<MessageLibInfo>((account) =>
      deserializeMessageLibInfo(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([103, 102, 218, 28, 204, 135, 71, 14])
    );
}

export function getMessageLibInfoSize(): number {
  return 11;
}
