// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/accounts/oAppRegistry.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type OAppRegistry = Account<OAppRegistryAccountData>;

export type OAppRegistryAccountData = {
  discriminator: Uint8Array;
  delegate: PublicKey;
  bump: number;
};

export type OAppRegistryAccountDataArgs = { delegate: PublicKey; bump: number };

export function getOAppRegistryAccountDataSerializer(): Serializer<
  OAppRegistryAccountDataArgs,
  OAppRegistryAccountData
> {
  return mapSerializer<
    OAppRegistryAccountDataArgs,
    any,
    OAppRegistryAccountData
  >(
    struct<OAppRegistryAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['delegate', publicKeySerializer()],
        ['bump', u8()],
      ],
      { description: 'OAppRegistryAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([6, 152, 199, 30, 217, 50, 69, 149]),
    })
  ) as Serializer<OAppRegistryAccountDataArgs, OAppRegistryAccountData>;
}

export function deserializeOAppRegistry(rawAccount: RpcAccount): OAppRegistry {
  return deserializeAccount(rawAccount, getOAppRegistryAccountDataSerializer());
}

export async function fetchOAppRegistry(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<OAppRegistry> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'OAppRegistry');
  return deserializeOAppRegistry(maybeAccount);
}

export async function safeFetchOAppRegistry(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<OAppRegistry | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeOAppRegistry(maybeAccount) : null;
}

export async function fetchAllOAppRegistry(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<OAppRegistry[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'OAppRegistry');
    return deserializeOAppRegistry(maybeAccount);
  });
}

export async function safeFetchAllOAppRegistry(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<OAppRegistry[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) => deserializeOAppRegistry(maybeAccount as RpcAccount));
}

export function getOAppRegistryGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      delegate: PublicKey;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      delegate: [8, publicKeySerializer()],
      bump: [40, u8()],
    })
    .deserializeUsing<OAppRegistry>((account) =>
      deserializeOAppRegistry(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([6, 152, 199, 30, 217, 50, 69, 149])
    );
}

export function getOAppRegistrySize(): number {
  return 41;
}
