// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/accounts/payloadHash.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type PayloadHash = Account<PayloadHashAccountData>;

export type PayloadHashAccountData = {
  discriminator: Uint8Array;
  hash: Uint8Array;
  bump: number;
};

export type PayloadHashAccountDataArgs = { hash: Uint8Array; bump: number };

export function getPayloadHashAccountDataSerializer(): Serializer<
  PayloadHashAccountDataArgs,
  PayloadHashAccountData
> {
  return mapSerializer<PayloadHashAccountDataArgs, any, PayloadHashAccountData>(
    struct<PayloadHashAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['hash', bytes({ size: 32 })],
        ['bump', u8()],
      ],
      { description: 'PayloadHashAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([96, 28, 106, 145, 103, 32, 186, 70]),
    })
  ) as Serializer<PayloadHashAccountDataArgs, PayloadHashAccountData>;
}

export function deserializePayloadHash(rawAccount: RpcAccount): PayloadHash {
  return deserializeAccount(rawAccount, getPayloadHashAccountDataSerializer());
}

export async function fetchPayloadHash(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<PayloadHash> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'PayloadHash');
  return deserializePayloadHash(maybeAccount);
}

export async function safeFetchPayloadHash(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<PayloadHash | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializePayloadHash(maybeAccount) : null;
}

export async function fetchAllPayloadHash(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<PayloadHash[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'PayloadHash');
    return deserializePayloadHash(maybeAccount);
  });
}

export async function safeFetchAllPayloadHash(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<PayloadHash[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) => deserializePayloadHash(maybeAccount as RpcAccount));
}

export function getPayloadHashGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      hash: Uint8Array;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      hash: [8, bytes({ size: 32 })],
      bump: [40, u8()],
    })
    .deserializeUsing<PayloadHash>((account) => deserializePayloadHash(account))
    .whereField(
      'discriminator',
      new Uint8Array([96, 28, 106, 145, 103, 32, 186, 70])
    );
}

export function getPayloadHashSize(): number {
  return 41;
}
