// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/accounts/sendLibraryConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';

/** the reason for not using Option::None to indicate default is to respect the spec on evm */
export type SendLibraryConfig = Account<SendLibraryConfigAccountData>;

export type SendLibraryConfigAccountData = {
  discriminator: Uint8Array;
  messageLib: PublicKey;
  bump: number;
};

export type SendLibraryConfigAccountDataArgs = {
  messageLib: PublicKey;
  bump: number;
};

export function getSendLibraryConfigAccountDataSerializer(): Serializer<
  SendLibraryConfigAccountDataArgs,
  SendLibraryConfigAccountData
> {
  return mapSerializer<
    SendLibraryConfigAccountDataArgs,
    any,
    SendLibraryConfigAccountData
  >(
    struct<SendLibraryConfigAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['messageLib', publicKeySerializer()],
        ['bump', u8()],
      ],
      { description: 'SendLibraryConfigAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([61, 238, 31, 72, 251, 117, 66, 176]),
    })
  ) as Serializer<
    SendLibraryConfigAccountDataArgs,
    SendLibraryConfigAccountData
  >;
}

export function deserializeSendLibraryConfig(
  rawAccount: RpcAccount
): SendLibraryConfig {
  return deserializeAccount(
    rawAccount,
    getSendLibraryConfigAccountDataSerializer()
  );
}

export async function fetchSendLibraryConfig(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<SendLibraryConfig> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'SendLibraryConfig');
  return deserializeSendLibraryConfig(maybeAccount);
}

export async function safeFetchSendLibraryConfig(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<SendLibraryConfig | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists
    ? deserializeSendLibraryConfig(maybeAccount)
    : null;
}

export async function fetchAllSendLibraryConfig(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<SendLibraryConfig[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'SendLibraryConfig');
    return deserializeSendLibraryConfig(maybeAccount);
  });
}

export async function safeFetchAllSendLibraryConfig(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<SendLibraryConfig[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) =>
      deserializeSendLibraryConfig(maybeAccount as RpcAccount)
    );
}

export function getSendLibraryConfigGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'endpoint',
    '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      messageLib: PublicKey;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      messageLib: [8, publicKeySerializer()],
      bump: [40, u8()],
    })
    .deserializeUsing<SendLibraryConfig>((account) =>
      deserializeSendLibraryConfig(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([61, 238, 31, 72, 251, 117, 66, 176])
    );
}

export function getSendLibraryConfigSize(): number {
  return 41;
}
