// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/pricefeed/accounts/priceFeed.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Option,
  OptionOrNullable,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  option,
  publicKey as publicKeySerializer,
  struct,
  u128,
  u8,
} from '@metaplex-foundation/umi/serializers';
import { Price, PriceArgs, getPriceSerializer } from '../types';

export type PriceFeed = Account<PriceFeedAccountData>;

export type PriceFeedAccountData = {
  discriminator: Uint8Array;
  admin: PublicKey;
  updaters: Array<PublicKey>;
  priceRatioDenominator: bigint;
  arbitrumCompressionPercent: bigint;
  nativeTokenPriceUsd: Option<bigint>;
  prices: Array<Price>;
  bump: number;
};

export type PriceFeedAccountDataArgs = {
  admin: PublicKey;
  updaters: Array<PublicKey>;
  priceRatioDenominator: number | bigint;
  arbitrumCompressionPercent: number | bigint;
  nativeTokenPriceUsd: OptionOrNullable<number | bigint>;
  prices: Array<PriceArgs>;
  bump: number;
};

export function getPriceFeedAccountDataSerializer(): Serializer<
  PriceFeedAccountDataArgs,
  PriceFeedAccountData
> {
  return mapSerializer<PriceFeedAccountDataArgs, any, PriceFeedAccountData>(
    struct<PriceFeedAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['admin', publicKeySerializer()],
        ['updaters', array(publicKeySerializer())],
        ['priceRatioDenominator', u128()],
        ['arbitrumCompressionPercent', u128()],
        ['nativeTokenPriceUsd', option(u128())],
        ['prices', array(getPriceSerializer())],
        ['bump', u8()],
      ],
      { description: 'PriceFeedAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([189, 103, 252, 23, 152, 35, 243, 156]),
    })
  ) as Serializer<PriceFeedAccountDataArgs, PriceFeedAccountData>;
}

export function deserializePriceFeed(rawAccount: RpcAccount): PriceFeed {
  return deserializeAccount(rawAccount, getPriceFeedAccountDataSerializer());
}

export async function fetchPriceFeed(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<PriceFeed> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'PriceFeed');
  return deserializePriceFeed(maybeAccount);
}

export async function safeFetchPriceFeed(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<PriceFeed | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializePriceFeed(maybeAccount) : null;
}

export async function fetchAllPriceFeed(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<PriceFeed[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'PriceFeed');
    return deserializePriceFeed(maybeAccount);
  });
}

export async function safeFetchAllPriceFeed(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<PriceFeed[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) => deserializePriceFeed(maybeAccount as RpcAccount));
}

export function getPriceFeedGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'pricefeed',
    '8ahPGPjEbpgGaZx2NV1iG5Shj7TDwvsjkEDcGWjt94TP'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      admin: PublicKey;
      updaters: Array<PublicKey>;
      priceRatioDenominator: number | bigint;
      arbitrumCompressionPercent: number | bigint;
      nativeTokenPriceUsd: OptionOrNullable<number | bigint>;
      prices: Array<PriceArgs>;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      admin: [8, publicKeySerializer()],
      updaters: [40, array(publicKeySerializer())],
      priceRatioDenominator: [null, u128()],
      arbitrumCompressionPercent: [null, u128()],
      nativeTokenPriceUsd: [null, option(u128())],
      prices: [null, array(getPriceSerializer())],
      bump: [null, u8()],
    })
    .deserializeUsing<PriceFeed>((account) => deserializePriceFeed(account))
    .whereField(
      'discriminator',
      new Uint8Array([189, 103, 252, 23, 152, 35, 243, 156])
    );
}
