// @orderly.network/ui-chain-selector/dist/index.js
'use strict';

var ui = require('@orderly.network/ui');
var react = require('react');
var hooks = require('@orderly.network/hooks');
var reactApp = require('@orderly.network/react-app');
var i18n = require('@orderly.network/i18n');
var jsxRuntime = require('react/jsx-runtime');

// src/chainSelector.widget.tsx
var KEY = "orderly_selected_chains";
var MAX_RECENT_CHAINS = 6;
var useChainSelectorScript = (options) => {
  const { networkId, bridgeLessOnly } = options || {};
  const { setStorageChain } = hooks.useStorageChain();
  const config = hooks.useConfig();
  const [_chains, { checkChainSupport }] = hooks.useChains();
  const { setChain, connectedChain } = hooks.useWalletConnector();
  const { onChainChanged, currentChainId, setCurrentChainId, wrongNetwork } = reactApp.useAppContext();
  const [selectChainId, setSelectChainId] = react.useState(
    currentChainId
  );
  const { getComponentTheme } = ui.useOrderlyTheme();
  const chains = react.useMemo(() => {
    const bridgeLessChains = bridgeLessOnly ? _chains.mainnet.filter((chain) => chain.network_infos.bridgeless) : _chains.mainnet;
    return {
      mainnet: bridgeLessChains.map((chain) => ({
        name: chain.network_infos.name,
        id: chain.network_infos.chain_id,
        lowestFee: chain.network_infos.bridgeless,
        isTestnet: false
      })),
      testnet: _chains.testnet.map((chain) => ({
        name: chain.network_infos.name,
        id: chain.network_infos.chain_id,
        lowestFee: chain.network_infos.bridgeless,
        isTestnet: true
      }))
    };
  }, [_chains, bridgeLessOnly]);
  const showTestnet = react.useMemo(() => {
    const chainSelectorOverrides = getComponentTheme("chainSelector", {
      showTestnet: true
    });
    if (chainSelectorOverrides.showTestnet === false || !chains.testnet.length) {
      return false;
    }
    return true;
  }, [chains.testnet]);
  const { recentChains, saveRecentChain } = useRecentChains(chains);
  const onChainChange = async (chain) => {
    if (connectedChain) {
      const result = await setChain({ chainId: chain.id });
      if (!result) return result;
      return {
        result,
        wrongNetwork: !checkChainSupport(chain.id, config.get("networkId")),
        chainId: chain.id
      };
    }
    setStorageChain(chain.id);
    setCurrentChainId(chain.id);
    return {
      result: true,
      wrongNetwork: false,
      chainId: chain.id
    };
  };
  const changedCallback = (chain, isWalletConnected) => {
    const params = {
      isTestnet: chain.isTestnet,
      isWalletConnected
    };
    options.onChainChangeAfter?.(chain.id, params);
    onChainChanged?.(chain.id, params);
  };
  const onChainClick = async (chain) => {
    setSelectChainId(chain.id);
    options.onChainChangeBefore?.(chain.id, { isTestnet: chain.isTestnet });
    try {
      const complete = await onChainChange?.(chain);
      if (complete) {
        options.resolve?.(complete);
        options.close?.();
        saveRecentChain(chain);
        changedCallback(chain, true);
      } else {
        setSelectChainId(void 0);
        changedCallback(chain, false);
      }
    } catch (err) {
      changedCallback(chain, false);
    }
  };
  const { selectedTab, onTabChange } = useChainTab(
    chains,
    currentChainId,
    wrongNetwork,
    showTestnet
  );
  return {
    recentChains,
    chains,
    selectChainId,
    onChainClick,
    selectedTab,
    onTabChange,
    showTestnet
  };
};
function useChainTab(chains, currentChainId, wrongNetwork, showTestnet) {
  const [selectedTab, setSelectedTab] = react.useState("Mainnet" /* Mainnet */);
  const onTabChange = (tab) => {
    setSelectedTab(tab);
  };
  react.useEffect(() => {
    if (!showTestnet) {
      setSelectedTab("Mainnet" /* Mainnet */);
      return;
    }
    if (currentChainId) {
      const isMainnet = chains.mainnet?.some(
        (chain) => chain.id === currentChainId
      );
      if (isMainnet) {
        setSelectedTab(wrongNetwork ? "Testnet" /* Testnet */ : "Mainnet" /* Mainnet */);
        return;
      }
      const isTestnet = chains.testnet?.some(
        (chain) => chain.id === currentChainId
      );
      if (isTestnet) {
        setSelectedTab(wrongNetwork ? "Mainnet" /* Mainnet */ : "Testnet" /* Testnet */);
        return;
      }
    }
  }, [currentChainId, chains, wrongNetwork, showTestnet]);
  return { selectedTab, onTabChange };
}
function useRecentChains(chains) {
  const [recentChainsIds, setRecentChainsIds] = hooks.useLocalStorage(
    KEY,
    []
  );
  const recentChains = react.useMemo(() => {
    return recentChainsIds?.map(
      (id) => chains.mainnet?.find((item) => item.id === parseInt(id))
    ).filter((chains2) => !!chains2);
  }, [chains, recentChainsIds]);
  const saveRecentChain = react.useCallback(
    (chain) => {
      if (chain.isTestnet) {
        return;
      }
      let ids = recentChainsIds?.filter((id) => id !== chain.id);
      ids = [chain.id, ...ids].slice(0, MAX_RECENT_CHAINS);
      setRecentChainsIds(ids);
    },
    [recentChainsIds]
  );
  return { recentChains, saveRecentChain };
}
var chainSelectorVariants = ui.tv({
  slots: {
    icon: "",
    list: "oui-grid oui-grid-cols-1 oui-gap-1",
    mainnetList: '"',
    testnetList: '"',
    recentList: "",
    item: "oui-w-full oui-rounded-md",
    tip: "oui-text-center"
  },
  variants: {
    variant: {
      compact: {
        icon: "oui-w-6 oui-h-6",
        list: "oui-bg-base-9 oui-rounded-lg oui-p-1",
        mainnetList: "oui-grid-cols-2 oui-mt-4",
        testnetList: "oui-grid-cols-1 oui-mt-4",
        recentList: "oui-mt-4",
        item: "oui-bg-base-6 hover:oui-bg-base-7",
        tip: "oui-pt-6"
      },
      wide: {
        icon: "oui-w-[18px] oui-h-[18px]",
        mainnetList: "oui-grid-cols-3 oui-mt-3",
        testnetList: "oui-grid-cols-2 oui-mt-3",
        recentList: "oui-mt-3",
        item: "oui-bg-base-5 hover:oui-bg-base-6",
        tip: "oui-pt-8"
      }
    },
    selected: {
      true: {
        item: ""
      },
      false: { item: "oui-bg-transparent" }
    }
  },
  compoundVariants: [
    {
      variant: "compact",
      selected: true,
      className: {
        item: "hover:oui-bg-base-6"
      }
    },
    {
      variant: "wide",
      selected: true,
      className: {
        item: "hover:oui-bg-base-5"
      }
    }
  ],
  defaultVariants: {
    variant: "wide",
    selected: false
  }
});
var ChainSelector = (props) => {
  const { isWrongNetwork, variant = "wide" } = props;
  const { t } = i18n.useTranslation();
  const { list, recentList, mainnetList, testnetList, icon, item, tip } = chainSelectorVariants({ variant });
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { className: ui.cn("oui-font-semibold", props.className), children: [
    /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Tabs,
      {
        value: props.selectedTab,
        variant: "contained",
        size: variant === "wide" ? "md" : "lg",
        onValueChange: (e) => props.onTabChange(e),
        children: [
          /* @__PURE__ */ jsxRuntime.jsxs(ui.TabPanel, { value: "Mainnet" /* Mainnet */, title: t("connector.mainnet"), children: [
            !!props.recentChains?.length && /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { gap: 2, className: recentList(), children: props.recentChains?.map((item2) => {
              return /* @__PURE__ */ jsxRuntime.jsx(
                RecentChainItem,
                {
                  item: item2,
                  onClick: () => props.onChainClick(item2),
                  iconClassName: icon()
                },
                item2.id
              );
            }) }),
            /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { r: "2xl", className: ui.cn(list(), mainnetList()), children: props.chains.mainnet?.map((chain) => {
              const selected = props.selectChainId === chain.id;
              return /* @__PURE__ */ jsxRuntime.jsx(
                ChainItem,
                {
                  selected,
                  item: chain,
                  onClick: () => props.onChainClick(chain),
                  className: item({ selected })
                },
                chain.id
              );
            }) })
          ] }),
          props.showTestnet && /* @__PURE__ */ jsxRuntime.jsx(ui.TabPanel, { value: "Testnet" /* Testnet */, title: t("connector.testnet"), children: /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { r: "2xl", className: ui.cn(list(), testnetList()), children: props.chains.testnet?.map((chain) => {
            const selected = props.selectChainId === chain.id;
            return /* @__PURE__ */ jsxRuntime.jsx(
              ChainItem,
              {
                selected,
                item: chain,
                onClick: () => props.onChainClick(chain),
                className: item({ selected })
              },
              chain.id
            );
          }) }) })
        ]
      }
    ),
    isWrongNetwork && /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { className: tip(), children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "warning", size: "xs", children: t("connector.wrongNetwork.tooltip") }) })
  ] });
};
var ChainItem = (props) => {
  const { item } = props;
  return /* @__PURE__ */ jsxRuntime.jsx("button", { className: props.className, onClick: props.onClick, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", className: "oui-py-2.5", px: 3, children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "center", width: "100%", className: "oui-gap-x-[6px]", children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.ChainIcon, { chainId: item.id, size: "xs" }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", children: item.name })
    ] }),
    props.selected && /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { gradient: "brand", r: "full", width: 4, height: 4 })
  ] }) });
};
var RecentChainItem = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    "button",
    {
      className: "oui-border oui-border-line-12 oui-rounded-lg hover:oui-border-primary-light",
      onClick: props.onClick,
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { itemAlign: "center", p: 2, children: /* @__PURE__ */ jsxRuntime.jsx(ui.ChainIcon, { chainId: props.item.id, className: props.iconClassName }) })
    }
  );
};
var ChainSelectorWidget = (props) => {
  const state = useChainSelectorScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(
    ChainSelector,
    {
      ...state,
      variant: props.variant,
      isWrongNetwork: props.isWrongNetwork
    }
  );
};
var ChainSelectorDialogId = "ChainSelectorDialogId";
var ChainSelectorSheetId = "ChainSelectorSheetId";
ui.registerSimpleDialog(ChainSelectorDialogId, ChainSelectorWidget, {
  size: "lg",
  title: () => i18n.i18n.t("connector.switchNetwork"),
  variant: "wide",
  isWrongNetwork: true
});
ui.registerSimpleSheet(ChainSelectorSheetId, ChainSelectorWidget, {
  title: () => i18n.i18n.t("connector.switchNetwork"),
  classNames: {
    content: "!oui-bg-base-8",
    body: "!oui-bg-base-8"
  },
  variant: "compact",
  isWrongNetwork: true
});

exports.ChainSelectorDialogId = ChainSelectorDialogId;
exports.ChainSelectorSheetId = ChainSelectorSheetId;
exports.ChainSelectorWidget = ChainSelectorWidget;
