// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/worker_interface/instructions/quoteDvn.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import { LzOption, LzOptionArgs, getLzOptionSerializer } from '../types';

// Accounts.
export type QuoteDvnInstructionAccounts = {
  workerConfig: PublicKey | Pda;
  priceFeedProgram: PublicKey | Pda;
  priceFeedConfig: PublicKey | Pda;
};

// Data.
export type QuoteDvnInstructionData = {
  discriminator: Uint8Array;
  msglib: PublicKey;
  dstEid: number;
  sender: PublicKey;
  packetHeader: Uint8Array;
  payloadHash: Uint8Array;
  confirmations: bigint;
  options: Array<LzOption>;
};

export type QuoteDvnInstructionDataArgs = {
  msglib: PublicKey;
  dstEid: number;
  sender: PublicKey;
  packetHeader: Uint8Array;
  payloadHash: Uint8Array;
  confirmations: number | bigint;
  options: Array<LzOptionArgs>;
};

export function getQuoteDvnInstructionDataSerializer(): Serializer<
  QuoteDvnInstructionDataArgs,
  QuoteDvnInstructionData
> {
  return mapSerializer<
    QuoteDvnInstructionDataArgs,
    any,
    QuoteDvnInstructionData
  >(
    struct<QuoteDvnInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['msglib', publicKeySerializer()],
        ['dstEid', u32()],
        ['sender', publicKeySerializer()],
        ['packetHeader', bytes({ size: u32() })],
        ['payloadHash', bytes({ size: 32 })],
        ['confirmations', u64()],
        ['options', array(getLzOptionSerializer())],
      ],
      { description: 'QuoteDvnInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([223, 98, 51, 200, 123, 133, 132, 194]),
    })
  ) as Serializer<QuoteDvnInstructionDataArgs, QuoteDvnInstructionData>;
}

// Args.
export type QuoteDvnInstructionArgs = QuoteDvnInstructionDataArgs;

// Instruction.
export function quoteDvn(
  context: Pick<Context, 'programs'>,
  input: QuoteDvnInstructionAccounts & QuoteDvnInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'workerInterface',
    '2iENutMTfTfxdFEjEQSbFCBstZqakXFZQCLZLLk4Ti58'
  );

  // Accounts.
  const resolvedAccounts = {
    workerConfig: {
      index: 0,
      isWritable: false as boolean,
      value: input.workerConfig ?? null,
    },
    priceFeedProgram: {
      index: 1,
      isWritable: false as boolean,
      value: input.priceFeedProgram ?? null,
    },
    priceFeedConfig: {
      index: 2,
      isWritable: false as boolean,
      value: input.priceFeedConfig ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: QuoteDvnInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getQuoteDvnInstructionDataSerializer().serialize(
    resolvedArgs as QuoteDvnInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
