// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/events/LzReceiveAlertEvent.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
export type LzReceiveAlertEvent = {
  receiver: web3.PublicKey
  executor: web3.PublicKey
  srcEid: number
  sender: number[] /* size: 32 */
  nonce: beet.bignum
  guid: number[] /* size: 32 */
  computeUnits: beet.bignum
  value: beet.bignum
  message: Uint8Array
  extraData: Uint8Array
  reason: Uint8Array
}

/**
 * @category userTypes
 * @category generated
 */
export const lzReceiveAlertEventBeet =
  new beet.FixableBeetArgsStruct<LzReceiveAlertEvent>(
    [
      ['receiver', beetSolana.publicKey],
      ['executor', beetSolana.publicKey],
      ['srcEid', beet.u32],
      ['sender', beet.uniformFixedSizeArray(beet.u8, 32)],
      ['nonce', beet.u64],
      ['guid', beet.uniformFixedSizeArray(beet.u8, 32)],
      ['computeUnits', beet.u64],
      ['value', beet.u64],
      ['message', beet.bytes],
      ['extraData', beet.bytes],
      ['reason', beet.bytes],
    ],
    'LzReceiveAlertEvent'
  )
