// @layerzerolabs/lz-corekit-solana/dist/index.cjs
'use strict';

var umi = require('@metaplex-foundation/umi');
var serializers = require('@metaplex-foundation/umi/serializers');
var umiProgramRepository = require('@metaplex-foundation/umi-program-repository');
var umiRpcWeb3js = require('@metaplex-foundation/umi-rpc-web3js');
var umiTransactionFactoryWeb3js = require('@metaplex-foundation/umi-transaction-factory-web3js');
var web3_js = require('@solana/web3.js');
var lzCore = require('@layerzerolabs/lz-core');
var umiEddsaWeb3js = require('@metaplex-foundation/umi-eddsa-web3js');
var bip39 = require('bip39');
var ed25519HdKey = require('ed25519-hd-key');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var bip39__namespace = /*#__PURE__*/_interopNamespace(bip39);

// src/providers/solana.ts
var SolanaProvider = class _SolanaProvider {
  /**
   * Creates an instance of SolanaProvider.
   *
   * @param {string | Connection} url - The URL of the Solana node.
   */
  constructor(url) {
    if (url instanceof web3_js.Connection) {
      const umi$1 = {
        transactions: umiTransactionFactoryWeb3js.createWeb3JsTransactionFactory(),
        programs: umiProgramRepository.createDefaultProgramRepository({ rpc: umi.createNullRpc() })
      };
      this.nativeProvider = umiRpcWeb3js.createWeb3JsRpc(umi$1, url);
      this.url = url.rpcEndpoint;
    } else {
      const connection = new web3_js.Connection(url, "confirmed");
      const umi$1 = {
        transactions: umiTransactionFactoryWeb3js.createWeb3JsTransactionFactory(),
        programs: umiProgramRepository.createDefaultProgramRepository({ rpc: umi.createNullRpc() })
      };
      this.nativeProvider = umiRpcWeb3js.createWeb3JsRpc(umi$1, connection);
      this.url = url;
    }
  }
  /**
   * Creates an instance of SolanaProvider from the given URL or Connection.
   *
   * @param {string | Connection} source - The URL of the Solana node.
   * @returns {SolanaProvider} The created SolanaProvider instance.
   * @throws {Error} If the URL parameter is invalid.
   */
  static from(source) {
    if (typeof source === "string") {
      return new _SolanaProvider(source);
    } else if (source instanceof web3_js.Connection) {
      return new _SolanaProvider(source);
    } else {
      throw new Error("Invalid parameters");
    }
  }
  get native() {
    return this.nativeProvider;
  }
  /**
   * Gets the balance of the specified address.
   *
   * @param {string} address - The address to get the balance of.
   * @returns {Promise<string>} A promise that resolves to the balance of the address.
   */
  async getBalance(address) {
    return this.nativeProvider.connection.getBalance(new web3_js.PublicKey(address)).then((balance) => balance.toString());
  }
  /**
   * Gets the block specified by slot.
   *
   * @param {BlockTag} slot - The slot to specify the block.
   * @returns {Promise<Block>} A promise that resolves to the block.
   */
  async getBlock(slot) {
    const response = await this.nativeProvider.connection.getParsedBlock(parseInt(slot.toString()), {
      commitment: "confirmed"
    });
    return lzCore.Block.from(response);
  }
  /**
   * Gets the current block number.
   *
   * @returns {Promise<number>} A promise that resolves to the current block number.
   */
  async getBlockNumber() {
    return this.nativeProvider.connection.getBlockHeight({ commitment: "confirmed" });
  }
  /**
   * Gets the current slot number for commitment.
   *
   * @param {Finality} [finality] - The commitment level (optional).
   * @returns {Promise<number>} A promise that resolves to the current slot number.
   */
  async getSlot(finality) {
    return this.nativeProvider.getSlot({ commitment: finality });
  }
  /**
   * Gets the UNIX timestamp for the block identified by slot.
   *
   * @param {BlockTag} slot - The slot to specify the block.
   * @returns {Promise<number>} A promise that resolves to the UNIX timestamp of the block.
   * @throws {Error} If the block timestamp is not available.
   */
  async getBlockTimestamp(slot) {
    const t = await this.nativeProvider.getBlockTime(parseInt(slot.toString()));
    if (t === null) {
      throw new Error("Block timestamp not available");
    }
    return Number(t.toString());
  }
  /**
   * Gets information about a transaction.
   *
   * @param {string} txHash - The transaction hash is TransactionSignature in Solana
   * @returns {Promise<TransactionResponse>} A promise that resolves to the transaction response
   */
  async getTransaction(txHash) {
    const signature = typeof txHash === "string" ? serializers.base58.serialize(txHash) : txHash;
    const response = await this.nativeProvider.getTransaction(signature, {
      commitment: "confirmed"
    });
    if (response === null) {
      throw new Error("Transaction not found");
    }
    return lzCore.TransactionResponse.from(response);
  }
  /**
   * Gets the receipt of a transaction.
   *
   * @param {string | Uint8Array} txHash - The transaction hash is TransactionSignature in Solana
   * @returns {Promise<TransactionReceipt>} A promise that resolves to the transaction receipt.
   * @throws {Error} If the transaction is not found.
   */
  async getTransactionReceipt(txHash) {
    const signature = typeof txHash === "string" ? serializers.base58.serialize(txHash) : txHash;
    const response = await this.nativeProvider.getTransaction(signature, {
      commitment: "confirmed"
    });
    if (response === null) {
      throw new Error("Transaction not found");
    }
    return lzCore.TransactionReceipt.from(response);
  }
  /**
   * Gets the number of transactions sent from the specified address.
   *
   * @param {string | Promise<string>} _addressOrName - The address to get the number of transactions from.
   * @param {BlockTag | Promise<BlockTag>} [_blockTag] - The block tag to get the number of transactions from (optional).
   * @returns {Promise<number>} A promise that resolves to the number of transactions sent from the address.
   * @throws {Error} Method not implemented.
   */
  // eslint-disable-next-line @typescript-eslint/require-await
  async getTransactionCount(_addressOrName, _blockTag) {
    await Promise.resolve();
    throw new Error("Method not implemented.");
  }
  /**
   * Gets the block with transactions specified by blockTag.
   *
   * @param {BlockTag} blockTag - The block tag to specify the block.
   * @returns {Promise<BlockWithTransactions>} A promise that resolves to the block with transactions.
   * @throws {Error} If the block is not found.
   */
  async getBlockWithTransactions(blockTag) {
    if (typeof blockTag === "string") {
      blockTag = parseInt(blockTag);
    }
    const response = await this.nativeProvider.connection.getBlock(blockTag, { transactionDetails: "full" });
    if (response === null) {
      throw new Error("Block not found");
    }
    return lzCore.BlockWithTransactions.from(response);
  }
  /**
   * Sends a signed transaction to the blockchain.
   *
   * @param {SignedTransaction} transaction - The signed transaction to send.
   * @param {RpcSendTransactionOptions} [options] - Optional parameters for sending the transaction.
   * @returns {Promise<TransactionPending>} A promise that resolves to the pending transaction.
   */
  async sendTransaction(transaction, options) {
    const signedTransaction = transaction.signed;
    const signature = await this.nativeProvider.sendTransaction(signedTransaction, options);
    const txHash = serializers.base58.deserialize(signature)[0];
    return lzCore.TransactionPending.from(txHash);
  }
  /**
   * Confirms a pending transaction.
   *
   * @param {TransactionPending} pending - The pending transaction to confirm.
   * @param {TransactionConfirmationStrategy} [opts] - Optional parameters for the confirmation.
   * @returns {Promise<TransactionReceipt>} A promise that resolves to the transaction receipt.
   * @throws {Error} If the transaction fails.
   */
  async confirmTransaction(pending, options) {
    const signature = pending.pending;
    let strategy;
    if (options?.strategy) {
      strategy = options.strategy;
    } else {
      const blockhash = await this.native.getLatestBlockhash();
      strategy = options?.strategy ?? { type: "blockhash", ...blockhash };
    }
    const result = await this.native.confirmTransaction(serializers.base58.serialize(signature), {
      ...options,
      strategy
    });
    const status = result.value;
    if (status.err !== null) {
      throw new Error(`transaction ${signature} failed (${JSON.stringify(status)})`);
    }
    return this.getTransactionReceipt(signature);
  }
  /**
   * Sends a signed transaction and waits for confirmation.
   *
   * @param {SignedTransaction} transaction - The signed transaction to send.
   * @param {object} [opts] - Optional parameters for sending and confirming the transaction.
   * @returns {Promise<TransactionReceipt>} A promise that resolves to the transaction receipt.
   */
  async sendAndConfirm(transaction, options) {
    const pending = await this.sendTransaction(transaction, options?.send);
    return this.confirmTransaction(pending, options?.confirm);
  }
};
var SolanaSigner = class {
  /**
   * Creates an instance of SolanaSigner.
   *
   * @param {KeypairSigner} keypair - The Solana keypair to use as the signer.
   */
  constructor(keypair) {
    this.keypair = keypair;
  }
  /**
   * Creates an instance of SolanaSigner from the given source.
   *
   * @param {KeypairSigner | string} source - The source to create the signer from.
   * @param {string} [path] - The derivation path (optional).
   * @returns {SolanaSigner} The created SolanaSigner instance.
   * @throws {Error} If the parameters are invalid.
   */
  static from(source, path) {
    if (typeof source === "object") {
      return new this(source);
    } else if (typeof source === "string") {
      const keypair = getKeypair(source, path);
      return new this(keypair);
    } else {
      throw new Error("Invalid parameters");
    }
  }
  /**
   * Gets the native Solana keypair instance.
   *
   * @returns {KeypairSigner} The native Solana keypair instance.
   */
  get native() {
    return this.keypair;
  }
  /**
   * Connects the signer to a provider.
   *
   * @param {Provider | Connection} provider - The provider to connect to.
   * @returns {this} The connected signer.
   * @throws {Error} If the provider is not an instance of Connection.
   */
  connect(provider) {
    if (provider instanceof web3_js.Connection) {
      this.provider = SolanaProvider.from(provider);
    } else if (provider instanceof SolanaProvider) {
      this.provider = provider;
    } else {
      throw new Error("Invalid parameters");
    }
    return this;
  }
  /**
   * Gets the address of the signer.
   *
   * @returns {Promise<string>} A promise that resolves to the address of the signer.
   */
  async getAddress() {
    return Promise.resolve(this.keypair.publicKey);
  }
  /**
   * Gets the address of the signer.
   *
   * @returns {string} The address of the signer.
   */
  get address() {
    return this.publicKey;
  }
  /**
   * Gets the public key of the signer.
   *
   * @returns {PublicKey} The public key of the signer.
   */
  get publicKey() {
    return this.keypair.publicKey;
  }
  /**
   * Gets the secret key of the signer.
   *
   * @returns {Uint8Array} The secret key of the signer.
   */
  get secretKey() {
    return this.keypair.secretKey;
  }
  /**
   * Sends a signed transaction and waits for confirmation.
   *
   * @param {SignedTransaction} transaction - The signed transaction to send.
   * @param {TransactionBuilderSendAndConfirmOptions} [opts] - Optional parameters for sending and confirming the transaction.
   * @returns {Promise<TransactionReceipt>} A promise that resolves to the transaction receipt.
   */
  async sendAndConfirm(transaction, opts) {
    const provider = this.assertAndGetProvider();
    return provider.sendAndConfirm(transaction, opts);
  }
  /**
   * Sends a signed transaction.
   *
   * @param {SignedTransaction} transaction - The signed transaction to send.
   * @param {RpcSendTransactionOptions} [opts] - Optional parameters for sending the transaction.
   * @returns {Promise<TransactionPending>} A promise that resolves to the pending transaction.
   */
  async sendTransaction(transaction, opts) {
    const provider = this.assertAndGetProvider();
    return provider.sendTransaction(transaction, opts);
  }
  /**
   * Signs a transaction.
   *
   * @param {TransactionRequest} transaction - The transaction request to sign.
   * @param {Commitment | GetLatestBlockhashConfig} [opts] - Optional parameters for signing the transaction.
   * @returns {Promise<SignedTransaction>} A promise that resolves to the signed transaction.
   */
  async signTransaction(transaction) {
    const tx = transaction.request;
    const signedTransaction = await this.keypair.signTransaction(tx);
    return lzCore.SignedTransaction.from(signedTransaction);
  }
  /**
   * Signs a buffer (e.g., a message hash) using the native Solana signer.
   *
   * @param {Uint8Array} _buffer - The buffer to sign.
   * @returns {Promise<Uint8Array>} A promise that resolves to the signed buffer as a Uint8Array.
   * @throws {Error} Method not implemented.
   */
  async signBuffer(_buffer) {
    await Promise.resolve();
    throw new Error("Method not implemented.");
  }
  assertAndGetProvider() {
    if (!this.provider) {
      throw new Error("Signer was not connected to a provider, did you forgot to connect to provider?");
    }
    return this.provider;
  }
};
function getKeypair(mnemonic, path = "m/44'/501'/0'/0'") {
  const eddsa = umiEddsaWeb3js.createWeb3JsEddsa();
  const seed = bip39__namespace.mnemonicToSeedSync(mnemonic, "");
  const privateKey = ed25519HdKey.derivePath(path, seed.toString("hex")).key;
  return umi.createSignerFromKeypair({ eddsa }, eddsa.createKeypairFromSeed(privateKey));
}

exports.SolanaProvider = SolanaProvider;
exports.SolanaSigner = SolanaSigner;
