// @orderly.network/ui-leverage/dist/index.js
'use strict';

var i18n = require('@orderly.network/i18n');
var ui = require('@orderly.network/ui');
var React = require('react');
var jsxRuntime = require('react/jsx-runtime');
var hooks = require('@orderly.network/hooks');
var perp = require('@orderly.network/perp');
var types = require('@orderly.network/types');
var utils = require('@orderly.network/utils');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

// src/index.ts
var IconButton = (props) => {
  const { Icon, onClick, disabled } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(
    Icon,
    {
      onClick: disabled ? void 0 : onClick,
      className: ui.cn(
        "oui-m-2 oui-text-white oui-transition-all",
        disabled ? "oui-cursor-not-allowed oui-opacity-20" : "oui-cursor-pointer oui-opacity-100"
      )
    }
  );
};
var LeverageInput = (props) => {
  const formatters = React__default.default.useMemo(
    () => [ui.inputFormatter.numberFormatter, ui.inputFormatter.dpFormatter(0)],
    []
  );
  const id = React.useId();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "label",
    {
      htmlFor: id,
      className: ui.cn(
        "oui-w-full",
        "oui-rounded",
        "oui-bg-base-6",
        "oui-flex",
        "oui-items-center",
        "oui-justify-between",
        "oui-outline",
        "oui-outline-offset-0",
        "oui-outline-1",
        "oui-outline-transparent",
        "focus-within:oui-outline-primary-light",
        "oui-input-root"
      ),
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          IconButton,
          {
            Icon: ui.ReduceIcon,
            onClick: props.onLeverageReduce,
            disabled: props.isReduceDisabled
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "center", justify: "center", className: "oui-mr-4", children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Input,
            {
              value: props.value,
              id,
              autoComplete: "off",
              classNames: {
                input: ui.cn("oui-text-right oui-text-[24px]"),
                root: ui.cn(
                  "oui-w-12",
                  "oui-px-0",
                  "oui-outline",
                  "oui-outline-offset-0",
                  "oui-outline-1",
                  "oui-outline-transparent",
                  "focus-within:oui-outline-primary-none"
                )
              },
              formatters,
              onChange: props.onInputChange,
              onBlur: props.onInputBlur
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: ui.cn(
                "oui-ml-1 oui-mt-1 oui-select-none",
                "oui-text-base oui-text-base-contrast-36"
              ),
              children: "x"
            }
          )
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(
          IconButton,
          {
            Icon: ui.PlusIcon,
            onClick: props.onLeverageIncrease,
            disabled: props.isIncreaseDisabled
          }
        )
      ]
    }
  );
};
var Leverage = (props) => {
  const { currentLeverage } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "start", direction: "column", mb: 0, children: [
    /* @__PURE__ */ jsxRuntime.jsx(LeverageHeader, { currentLeverage }),
    /* @__PURE__ */ jsxRuntime.jsx(LeverageInput, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(LeverageSelector, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(LeverageSlider, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(LeverageFooter, { ...props })
  ] });
};
var LeverageFooter = (props) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "row", gap: 2, width: "100%", mt: 0, pt: 5, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        variant: "contained",
        color: "gray",
        fullWidth: true,
        onClick: props.onCancel,
        "data-testid": "oui-testid-leverage-cancel-btn",
        size: props.isMobile ? "md" : "lg",
        children: t("common.cancel")
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        fullWidth: true,
        loading: props.isLoading,
        onClick: props.onSave,
        "data-testid": "oui-testid-leverage-save-btn",
        disabled: props.disabled,
        size: props.isMobile ? "md" : "lg",
        children: t("common.save")
      }
    )
  ] });
};
var LeverageHeader = (props) => {
  const { t } = i18n.useTranslation();
  const { currentLeverage } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { justify: "center", width: "100%", mb: 2, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, children: [
    `${t("common.current")}:`,
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { unit: "x", size: "sm", intensity: 80, dp: 0, children: currentLeverage ?? "--" })
  ] }) });
};
var LeverageSelector = (props) => {
  const { value, onLeverageChange } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Flex,
    {
      itemAlign: "center",
      justify: "between",
      width: "100%",
      mt: 4,
      className: "oui-text-base-contrast-80",
      children: props.toggles.map((option) => /* @__PURE__ */ jsxRuntime.jsx(
        ui.Flex,
        {
          itemAlign: "center",
          justify: "center",
          className: ui.cn(
            `oui-box-border oui-cursor-pointer oui-rounded-md oui-border oui-border-solid oui-bg-clip-padding oui-px-3 oui-py-2.5 oui-transition-all`,
            value === option ? "oui-border-primary oui-bg-base-6" : "oui-border-line-12"
          ),
          onClick: () => onLeverageChange?.(option),
          children: /* @__PURE__ */ jsxRuntime.jsxs(
            ui.Flex,
            {
              itemAlign: "center",
              justify: "center",
              className: ui.cn(`oui-h-3 oui-w-9 oui-select-none`),
              children: [
                option,
                "x"
              ]
            }
          )
        },
        option
      ))
    }
  );
};
var getMarkPosition = (item, index, max, total) => {
  const min = 1;
  const maxSteps = max - min;
  const percentPerStep = 100 / maxSteps;
  const position = percentPerStep * (item - min);
  if (index === 0) return Math.min(position + 2, 100);
  if (index === total - 1) return Math.max(position - 3, 0);
  return position;
};
var LeverageSlider = (props) => {
  const {
    leverageLevers,
    maxLeverage = 0,
    className,
    value,
    showSliderTip,
    marks
  } = props;
  const sliderMax = leverageLevers.length > 0 ? Math.max(...leverageLevers) : maxLeverage;
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { pt: 4, pb: 7, width: "100%", className, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Slider,
      {
        step: 1,
        max: maxLeverage,
        min: 1,
        marks,
        value: [value],
        onValueChange: (e) => {
          props.onLeverageChange(e[0]);
          props.setShowSliderTip(true);
        },
        color: "primary",
        onValueCommit: (e) => {
          props.onValueCommit?.(e);
          props.setShowSliderTip(false);
        },
        showTip: showSliderTip,
        tipFormatter: (value2) => {
          return `${value2}x`;
        }
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-w-full oui-pt-3", children: leverageLevers?.map((item, index) => {
      const position = getMarkPosition(
        item,
        index,
        sliderMax,
        leverageLevers.length
      );
      return /* @__PURE__ */ jsxRuntime.jsx(
        "button",
        {
          onClick: () => {
            props.onLeverageChange(item);
            props.onValueCommit?.([item]);
          },
          className: ui.cn(
            "oui-absolute oui-pb-3 oui-text-2xs oui-transform oui--translate-x-1/2",
            Number(props.value) >= Number(item) ? "oui-text-primary-light" : "oui-text-base-contrast-54"
          ),
          style: {
            left: `${position}%`
          },
          "data-testid": `oui-testid-leverage-${item}-btn`,
          children: `${item}x`
        },
        item
      );
    }) })
  ] });
};
var useLeverageScript = (options) => {
  const [showSliderTip, setShowSliderTip] = React.useState(false);
  const { t } = i18n.useTranslation();
  const { curLeverage, maxLeverage, isLoading, leverageLevers, update } = hooks.useLeverage();
  const marks = React.useMemo(() => {
    return leverageLevers.map((e) => ({
      label: `${e}x`,
      value: e
    }));
  }, [leverageLevers]);
  const [leverage, setLeverage] = React.useState(curLeverage ?? 0);
  const step = 100 / ((marks?.length || 0) - 1);
  const onLeverageChange = (leverage2) => {
    setLeverage(leverage2);
  };
  const onLeverageIncrease = () => {
    setLeverage((prev) => prev + 1);
  };
  const onLeverageReduce = () => {
    setLeverage((prev) => prev - 1);
  };
  const onInputChange = React.useCallback(
    (e) => {
      const parsed = Number.parseInt(e.target.value);
      const value = Number.isNaN(parsed) ? "" : parsed;
      setLeverage(value);
    },
    [maxLeverage]
  );
  const onSave = async () => {
    try {
      update({ leverage }).then(
        () => {
          options?.close?.();
          ui.toast.success(t("leverage.updated"));
        },
        (err) => {
          ui.toast.error(err.message);
        }
      );
    } catch (err) {
    }
  };
  const isReduceDisabled = leverage <= 1;
  const isIncreaseDisabled = leverage >= maxLeverage;
  const disabled = !leverage || leverage < 1 || leverage > maxLeverage;
  const toggles = React.useMemo(() => {
    return [5, 10, 20, 50, 100].filter((e) => e <= maxLeverage);
  }, [maxLeverage]);
  return {
    leverageLevers,
    currentLeverage: curLeverage,
    value: leverage,
    marks,
    onLeverageChange,
    onLeverageIncrease,
    onLeverageReduce,
    onInputChange,
    isReduceDisabled,
    isIncreaseDisabled,
    disabled,
    step,
    onCancel: options?.close,
    onSave,
    isLoading,
    showSliderTip,
    setShowSliderTip,
    maxLeverage,
    toggles
  };
};
var LeverageEditor = (props) => {
  const state = useLeverageScript({ close: props.close });
  return /* @__PURE__ */ jsxRuntime.jsx(Leverage, { ...state });
};
var useSymbolLeverageScript = (options) => {
  const { curLeverage = 1, symbol, side } = options;
  const [showSliderTip, setShowSliderTip] = React.useState(false);
  const [leverage, setLeverage] = React.useState(curLeverage);
  const { t } = i18n.useTranslation();
  const { isMobile } = ui.useScreen();
  const {
    maxLeverage: originalMaxLeverage,
    update,
    isLoading
  } = hooks.useSymbolLeverage(symbol);
  const maxLeverage = originalMaxLeverage;
  const {
    position,
    maxPositionNotional,
    maxPositionLeverage,
    overMaxPositionLeverage,
    overRequiredMargin
  } = useCalc({ symbol, leverage, maxLeverage });
  const formattedLeverageLevers = React.useMemo(() => {
    return generateLeverageLeversForSelector(maxLeverage);
  }, [maxLeverage]);
  const leverageLevers = React.useMemo(() => {
    return generateLeverageLevers(maxLeverage);
  }, [maxLeverage]);
  const marks = React.useMemo(() => {
    return leverageLevers.map((e) => ({
      label: `${e}x`,
      value: e
    })) || [];
  }, [leverageLevers]);
  const step = React.useMemo(() => {
    return 100 / ((marks?.length || 0) - 1);
  }, [marks]);
  const onLeverageChange = (leverage2) => {
    setLeverage(leverage2);
  };
  const onLeverageIncrease = () => {
    setLeverage((prev) => prev + 1);
  };
  const onLeverageReduce = () => {
    setLeverage((prev) => prev - 1);
  };
  const onInputChange = React.useCallback(
    (e) => {
      const parsed = Number.parseInt(e.target.value);
      const value = Number.isNaN(parsed) ? "" : parsed;
      setLeverage(value);
    },
    []
  );
  const onInputBlur = React.useCallback(
    (e) => {
      const inputValue = e.target.value.trim();
      if (inputValue === "") {
        setLeverage(curLeverage);
      }
    },
    [curLeverage]
  );
  const onConfirmSave = async () => {
    try {
      update?.({ leverage, symbol }).then(
        (res) => {
          if (res.success) {
            options?.close?.();
            ui.toast.success(t("leverage.updated"));
          } else {
            ui.toast.error(res.message);
          }
        },
        (err) => {
          ui.toast.error(err.message);
        }
      );
    } catch (err) {
    }
  };
  const onSave = async () => {
    ui.modal.confirm({
      title: t("leverage.confirm"),
      content: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("leverage.confirm.content") }),
      onOk: onConfirmSave,
      onCancel: () => {
        return Promise.resolve();
      }
    });
  };
  const isReduceDisabled = leverage <= 1;
  const isIncreaseDisabled = leverage >= maxLeverage;
  const isBuy = side ? side === types.OrderSide.BUY : position?.position_qty && position.position_qty > 0;
  const disabled = !leverage || leverage < 1 || leverage > maxLeverage || overRequiredMargin || overMaxPositionLeverage;
  return {
    leverageLevers,
    currentLeverage: curLeverage,
    // Keep the displayed leverage fixed until the user confirms the change.
    value: leverage,
    // Input and slider reflect the temporary value being edited.
    marks,
    onLeverageChange,
    onLeverageIncrease,
    onLeverageReduce,
    onInputChange,
    onInputBlur,
    isReduceDisabled,
    isIncreaseDisabled,
    disabled,
    step,
    onCancel: options?.close,
    onSave,
    isLoading,
    showSliderTip,
    setShowSliderTip,
    maxLeverage,
    toggles: formattedLeverageLevers,
    symbol,
    maxPositionNotional,
    maxPositionLeverage,
    overMaxPositionLeverage,
    overRequiredMargin,
    isBuy,
    isMobile
  };
};
var generateLeverageLeversForSelector = (max) => {
  if (max === 10) {
    return [1, 3, 5, 8, 10];
  } else if (max === 50) {
    return [1, 10, 20, 35, 50];
  }
  const min = 1;
  const parts = 5;
  const step = (max - min) / (parts - 1);
  const result = [];
  for (let i = 0; i < parts; i++) {
    result.push(Math.floor(min + step * i));
  }
  return result;
};
var generateEvenlyDistributedMarks = (max) => {
  const result = [];
  if (max % 5 === 0) {
    const step = max / 5;
    for (let i = 0; i < 6; i++) {
      const value = step * i;
      result.push(value === 0 ? 1 : value);
    }
  } else {
    result.push(1);
    const quarter = max * 0.25;
    const half = max * 0.5;
    const threeQuarter = max * 0.75;
    const quarterRounded = Math.round(quarter);
    const halfRounded = Math.round(half);
    const threeQuarterRounded = Math.round(threeQuarter);
    if (quarterRounded > 1 && quarterRounded !== halfRounded) {
      result.push(quarterRounded);
    }
    if (halfRounded > 1) {
      result.push(halfRounded);
    }
    if (threeQuarterRounded > halfRounded && threeQuarterRounded < max) {
      result.push(threeQuarterRounded);
    }
    if (max > 1) {
      result.push(max);
    }
  }
  return result;
};
var generateLeverageLevers = (max) => {
  switch (max) {
    case 10:
      return [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    case 20:
      return [1, 5, 10, 15, 20];
    case 50:
      return [1, 10, 20, 30, 40, 50];
    case 100:
      return [1, 20, 40, 60, 80, 100];
  }
  const result = [];
  if (max < 10) {
    for (let i = 1; i <= max; i++) {
      result.push(i);
    }
  } else {
    result.push(...generateEvenlyDistributedMarks(max));
  }
  return result;
};
function useCalc(inputs) {
  const { symbol, leverage, maxLeverage } = inputs;
  const symbolsInfo = hooks.useSymbolsInfo();
  const { data: accountInfo } = hooks.useAccountInfo();
  const { data: markPrices } = hooks.useMarkPricesStream();
  const { totalCollateral } = hooks.usePortfolio();
  const [unPnlPriceBasis, setUnPnlPriceBasic] = hooks.useLocalStorage(
    "unPnlPriceBasis",
    "markPrice"
  );
  const [positions] = hooks.usePositionStream("all", {
    calcMode: unPnlPriceBasis
  });
  const position = React.useMemo(() => {
    if (symbol && positions?.rows?.length) {
      return positions.rows.find((item) => item.symbol === symbol);
    }
  }, [positions, symbol]);
  const maxPositionLeverage = React.useMemo(() => {
    const IMRFactor = accountInfo?.imr_factor?.[symbol];
    const notional = position?.notional;
    if (IMRFactor && notional) {
      const maxPositionLeverage2 = perp.positions.maxPositionLeverage({
        IMRFactor,
        notional
      });
      return Math.min(maxPositionLeverage2, maxLeverage);
    }
    return maxLeverage;
  }, [position, maxLeverage, symbol]);
  const maxPositionNotional = React.useMemo(() => {
    const IMRFactor = accountInfo?.imr_factor?.[symbol];
    if (leverage && IMRFactor) {
      return perp.positions.maxPositionNotional({
        leverage,
        IMRFactor
      });
    }
  }, [leverage, symbol]);
  const overMaxPositionLeverage = React.useMemo(() => {
    return leverage > maxPositionLeverage;
  }, [leverage, maxPositionLeverage]);
  const freeCollateral = React.useMemo(() => {
    if (!accountInfo || !markPrices || !symbolsInfo) {
      return utils.zero;
    }
    const positionList = leverage ? positions?.rows.map((item) => {
      if (item.symbol === symbol) {
        return {
          ...item,
          leverage
        };
      }
      return item;
    }) : positions?.rows;
    const totalInitialMarginWithOrders = perp.account.totalInitialMarginWithQty({
      positions: positionList,
      markPrices,
      IMR_Factors: accountInfo.imr_factor,
      // not used
      maxLeverage: accountInfo.max_leverage,
      symbolInfo: symbolsInfo
    });
    const freeCollateral2 = perp.account.freeCollateral({
      totalCollateral,
      totalInitialMarginWithOrders
    });
    return freeCollateral2;
  }, [
    positions,
    symbolsInfo,
    accountInfo,
    markPrices,
    totalCollateral,
    leverage,
    symbol
  ]);
  const overRequiredMargin = React.useMemo(() => {
    return freeCollateral.eq(0) || freeCollateral.isNegative();
  }, [freeCollateral]);
  return {
    position,
    freeCollateral,
    maxPositionNotional,
    maxPositionLeverage,
    overMaxPositionLeverage,
    overRequiredMargin
  };
}
var SymbolLeverage = (props) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-3 lg:oui-gap-4", children: [
    /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-2", children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.TokenIcon, { symbol: props.symbol, className: "oui-size-5" }),
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.formatted,
        {
          rule: "symbol",
          formatString: "base-type",
          size: props.isMobile ? "xs" : "base",
          weight: "semibold",
          intensity: 98,
          children: props.symbol
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          className: ui.cn(["oui-ml-auto oui-flex oui-items-center oui-gap-1"]),
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: props.isBuy ? "success" : "danger", size: "xs", children: props.isBuy ? t("common.long") : t("common.short") }),
            /* @__PURE__ */ jsxRuntime.jsx(LeverageBadge, { leverage: props.currentLeverage })
          ]
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, {}),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "start", direction: "column", mb: 0, children: [
      /* @__PURE__ */ jsxRuntime.jsx(LeverageHeader, { currentLeverage: props.currentLeverage }),
      /* @__PURE__ */ jsxRuntime.jsx(LeverageInput, { ...props }),
      /* @__PURE__ */ jsxRuntime.jsx(LeverageSelector, { ...props }),
      /* @__PURE__ */ jsxRuntime.jsx(LeverageSlider, { ...props }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { className: "oui-mb-3 oui-w-full" }),
      /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1 oui-pb-4 oui-text-xs oui-font-normal oui-text-base-contrast-54", children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(
          i18n.Trans,
          {
            i18nKey: "leverage.maxAvailableLeverage.tips",
            values: { leverage: props.maxPositionLeverage },
            components: [
              // @ts-ignore
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.numeral,
                {
                  dp: 0,
                  suffix: "x",
                  as: "span",
                  className: "oui-text-base-contrast"
                },
                "0"
              )
            ]
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { children: t("leverage.actualPositionLeverage.tips") })
      ] }),
      /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          className: ui.cn([
            "-oui-mb-2",
            props.overRequiredMargin || props.overMaxPositionLeverage ? "oui-block oui-text-xs oui-font-normal" : "oui-hidden"
          ]),
          children: [
            props.overRequiredMargin && /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "warning", children: t("leverage.overRequiredMargin.tips") }) }),
            props.overMaxPositionLeverage && /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "warning", children: /* @__PURE__ */ jsxRuntime.jsx(
              i18n.Trans,
              {
                i18nKey: "leverage.overMaxPositionLeverage.tips",
                values: { leverage: props.maxPositionLeverage },
                components: [
                  // @ts-ignore
                  /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 0, suffix: "X", as: "span" }, "0")
                ]
              }
            ) }) })
          ]
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(LeverageFooter, { ...props })
    ] })
  ] });
};
var LeverageBadge = ({ leverage }) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "div",
    {
      className: ui.cn(
        "oui-flex oui-h-[18px] oui-items-center oui-gap-1",
        "oui-cursor-pointer oui-rounded oui-bg-line-6 oui-px-2",
        "oui-text-2xs oui-font-semibold oui-text-base-contrast-36"
      ),
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: "Cross" }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 0, size: "2xs", unit: "X", children: leverage })
      ]
    }
  );
};
var SymbolLeverageWidget = (props) => {
  const state = useSymbolLeverageScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(SymbolLeverage, { ...state });
};

// src/symbolLeverage/index.ts
var SymbolLeverageSheetId = "SymbolLeverageSheetId";
var SymbolLeverageDialogId = "SymbolLeverageDialogId";
ui.registerSimpleSheet(SymbolLeverageSheetId, SymbolLeverageWidget, {
  title: () => i18n.i18n.t("leverage.adjustedLeverage"),
  classNames: {
    // content: "oui-p-5",
  }
});
ui.registerSimpleDialog(SymbolLeverageDialogId, SymbolLeverageWidget, {
  title: () => i18n.i18n.t("leverage.adjustedLeverage"),
  classNames: {
    content: "oui-w-[420px]"
  }
});

// src/index.ts
var LeverageWidgetWithDialogId = "LeverageWidgetWithDialog";
var LeverageWidgetWithSheetId = "LeverageWidgetWithSheet";
ui.registerSimpleDialog(LeverageWidgetWithDialogId, LeverageEditor, {
  title: () => i18n.i18n.t("leverage.maxAccountLeverage"),
  size: "md"
});
ui.registerSimpleSheet(LeverageWidgetWithSheetId, LeverageEditor, {
  title: () => i18n.i18n.t("leverage.maxAccountLeverage")
});

exports.Leverage = Leverage;
exports.LeverageEditor = LeverageEditor;
exports.LeverageHeader = LeverageHeader;
exports.LeverageSlider = LeverageSlider;
exports.LeverageWidgetWithDialogId = LeverageWidgetWithDialogId;
exports.LeverageWidgetWithSheetId = LeverageWidgetWithSheetId;
exports.SymbolLeverageDialogId = SymbolLeverageDialogId;
exports.SymbolLeverageSheetId = SymbolLeverageSheetId;
exports.SymbolLeverageWidget = SymbolLeverageWidget;
exports.useLeverageScript = useLeverageScript;
