// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/accounts/executeHash.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  i64,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type ExecuteHash = Account<ExecuteHashAccountData>;

export type ExecuteHashAccountData = {
  discriminator: Uint8Array;
  expiration: bigint;
  bump: number;
};

export type ExecuteHashAccountDataArgs = {
  expiration: number | bigint;
  bump: number;
};

export function getExecuteHashAccountDataSerializer(): Serializer<
  ExecuteHashAccountDataArgs,
  ExecuteHashAccountData
> {
  return mapSerializer<ExecuteHashAccountDataArgs, any, ExecuteHashAccountData>(
    struct<ExecuteHashAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['expiration', i64()],
        ['bump', u8()],
      ],
      { description: 'ExecuteHashAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([34, 17, 137, 123, 62, 25, 205, 155]),
    })
  ) as Serializer<ExecuteHashAccountDataArgs, ExecuteHashAccountData>;
}

export function deserializeExecuteHash(rawAccount: RpcAccount): ExecuteHash {
  return deserializeAccount(rawAccount, getExecuteHashAccountDataSerializer());
}

export async function fetchExecuteHash(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<ExecuteHash> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'ExecuteHash');
  return deserializeExecuteHash(maybeAccount);
}

export async function safeFetchExecuteHash(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<ExecuteHash | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeExecuteHash(maybeAccount) : null;
}

export async function fetchAllExecuteHash(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<ExecuteHash[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'ExecuteHash');
    return deserializeExecuteHash(maybeAccount);
  });
}

export async function safeFetchAllExecuteHash(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<ExecuteHash[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) => deserializeExecuteHash(maybeAccount as RpcAccount));
}

export function getExecuteHashGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'dvn',
    'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      expiration: number | bigint;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      expiration: [8, i64()],
      bump: [16, u8()],
    })
    .deserializeUsing<ExecuteHash>((account) => deserializeExecuteHash(account))
    .whereField(
      'discriminator',
      new Uint8Array([34, 17, 137, 123, 62, 25, 205, 155])
    );
}

export function getExecuteHashSize(): number {
  return 17;
}
