// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/instructions/getFee.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { GetFeeParams, getFeeParamsBeet } from '../types/GetFeeParams'

/**
 * @category Instructions
 * @category GetFee
 * @category generated
 */
export type GetFeeInstructionArgs = {
  params: GetFeeParams
}
/**
 * @category Instructions
 * @category GetFee
 * @category generated
 */
export const getFeeStruct = new beet.BeetArgsStruct<
  GetFeeInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', getFeeParamsBeet],
  ],
  'GetFeeInstructionArgs'
)
/**
 * Accounts required by the _getFee_ instruction
 *
 * @property [] priceFeed
 * @category Instructions
 * @category GetFee
 * @category generated
 */
export type GetFeeInstructionAccounts = {
  priceFeed: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const getFeeInstructionDiscriminator = [
  115, 195, 235, 161, 25, 219, 60, 29,
]

/**
 * Creates a _GetFee_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category GetFee
 * @category generated
 */
export function createGetFeeInstruction(
  accounts: GetFeeInstructionAccounts,
  args: GetFeeInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = getFeeStruct.serialize({
    instructionDiscriminator: getFeeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.priceFeed,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _GetFee_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category GetFee
 * @category generated
 */
export function createGetFeeInstructionAccounts(
  accounts: GetFeeInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.priceFeed,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
