// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/instructions/initPriceFeed.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitPriceFeedParams,
  initPriceFeedParamsBeet,
} from '../types/InitPriceFeedParams'

/**
 * @category Instructions
 * @category InitPriceFeed
 * @category generated
 */
export type InitPriceFeedInstructionArgs = {
  params: InitPriceFeedParams
}
/**
 * @category Instructions
 * @category InitPriceFeed
 * @category generated
 */
export const initPriceFeedStruct = new beet.FixableBeetArgsStruct<
  InitPriceFeedInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initPriceFeedParamsBeet],
  ],
  'InitPriceFeedInstructionArgs'
)
/**
 * Accounts required by the _initPriceFeed_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [_writable_] priceFeed
 * @category Instructions
 * @category InitPriceFeed
 * @category generated
 */
export type InitPriceFeedInstructionAccounts = {
  payer: web3.PublicKey
  priceFeed: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initPriceFeedInstructionDiscriminator = [
  27, 209, 184, 5, 152, 116, 136, 16,
]

/**
 * Creates a _InitPriceFeed_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitPriceFeed
 * @category generated
 */
export function createInitPriceFeedInstruction(
  accounts: InitPriceFeedInstructionAccounts,
  args: InitPriceFeedInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initPriceFeedStruct.serialize({
    instructionDiscriminator: initPriceFeedInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.priceFeed,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitPriceFeed_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitPriceFeed
 * @category generated
 */
export function createInitPriceFeedInstructionAccounts(
  accounts: InitPriceFeedInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.priceFeed,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
