// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/instructions/setSolPrice.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetSolPriceParams,
  setSolPriceParamsBeet,
} from '../types/SetSolPriceParams'

/**
 * @category Instructions
 * @category SetSolPrice
 * @category generated
 */
export type SetSolPriceInstructionArgs = {
  params: SetSolPriceParams
}
/**
 * @category Instructions
 * @category SetSolPrice
 * @category generated
 */
export const setSolPriceStruct = new beet.FixableBeetArgsStruct<
  SetSolPriceInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setSolPriceParamsBeet],
  ],
  'SetSolPriceInstructionArgs'
)
/**
 * Accounts required by the _setSolPrice_ instruction
 *
 * @property [**signer**] updater
 * @property [_writable_] priceFeed
 * @category Instructions
 * @category SetSolPrice
 * @category generated
 */
export type SetSolPriceInstructionAccounts = {
  updater: web3.PublicKey
  priceFeed: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setSolPriceInstructionDiscriminator = [
  151, 237, 70, 191, 232, 81, 138, 214,
]

/**
 * Creates a _SetSolPrice_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetSolPrice
 * @category generated
 */
export function createSetSolPriceInstruction(
  accounts: SetSolPriceInstructionAccounts,
  args: SetSolPriceInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setSolPriceStruct.serialize({
    instructionDiscriminator: setSolPriceInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.updater,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.priceFeed,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetSolPrice_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetSolPrice
 * @category generated
 */
export function createSetSolPriceInstructionAccounts(
  accounts: SetSolPriceInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.updater,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.priceFeed,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
