// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/instructions/transferAdmin.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  TransferAdminParams,
  transferAdminParamsBeet,
} from '../types/TransferAdminParams'

/**
 * @category Instructions
 * @category TransferAdmin
 * @category generated
 */
export type TransferAdminInstructionArgs = {
  params: TransferAdminParams
}
/**
 * @category Instructions
 * @category TransferAdmin
 * @category generated
 */
export const transferAdminStruct = new beet.BeetArgsStruct<
  TransferAdminInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', transferAdminParamsBeet],
  ],
  'TransferAdminInstructionArgs'
)
/**
 * Accounts required by the _transferAdmin_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] priceFeed
 * @category Instructions
 * @category TransferAdmin
 * @category generated
 */
export type TransferAdminInstructionAccounts = {
  admin: web3.PublicKey
  priceFeed: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const transferAdminInstructionDiscriminator = [
  42, 242, 66, 106, 228, 10, 111, 156,
]

/**
 * Creates a _TransferAdmin_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category TransferAdmin
 * @category generated
 */
export function createTransferAdminInstruction(
  accounts: TransferAdminInstructionAccounts,
  args: TransferAdminInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = transferAdminStruct.serialize({
    instructionDiscriminator: transferAdminInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.priceFeed,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _TransferAdmin_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category TransferAdmin
 * @category generated
 */
export function createTransferAdminInstructionAccounts(
  accounts: TransferAdminInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.priceFeed,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
