// @orderly.network/react-app/dist/index.mjs
import { useTrack, OrderlyConfigProvider, useAccount, ERROR_MSG_CODES, useTrackingInstance, useRestrictedInfo, useEventEmitter, useSymbolsInfo, useOrderlyContext, useLocalStorage, useAudioPlayer, useDebouncedCallback, useStorageChain, useChains, useConfig, useWalletConnector, useKeyStore, useSessionStorage, useWalletSubscription, useSettleSubscription, useWS, useStorageLedgerAddress, parseJSON } from '@orderly.network/hooks';
import { OrderlyThemeProvider, LocaleProvider, TooltipProvider, ModalProvider, Toaster, Flex, cn, Text, Button, useScreen, toast, modal, parseNumber } from '@orderly.network/ui';
import { createContext, useContext, useCallback, Component, useMemo, useEffect, useState, useRef } from 'react';
import { AccountStatusEnum, SDKError, ABSTRACT_CHAIN_ID_MAP, ChainNamespace, TrackerEventName, OrderStatus, AlgoOrderRootType } from '@orderly.network/types';
import { useTranslation, i18n, useLocaleCode, LocaleEnum } from '@orderly.network/i18n';
import { parseChainIdToNumber, windowGuard, praseChainIdToNumber, capitalizeString, getTimestamp, transSymbolformString } from '@orderly.network/utils';
import { jsx, jsxs } from 'react/jsx-runtime';
import { zhCN, enUS } from 'date-fns/locale';

// src/provider/orderlyAppProvider.tsx
var useBootstrap = () => {
  useEffect(() => {
    const searchParams = new URLSearchParams(window.location.search);
    const refCode = searchParams.get("ref");
    if (refCode) {
      localStorage.setItem("referral_code", refCode);
    }
  }, []);
};
function getDisplaySide(side) {
  if (side === "BUY") {
    return i18n.t("common.buy");
  } else if (side === "SELL") {
    return i18n.t("common.sell");
  }
  return capitalizeString(side);
}
function getOrderExecutionReportMsg(data, symbolsInfo) {
  const { symbol, side, quantity, client_order_id, fieldChanges } = data;
  const total_executed_quantity = "total_executed_quantity" in data ? data.total_executed_quantity : 0;
  const status = "status" in data ? data.status : data.algo_status;
  const getSymbolInfo = symbolsInfo[symbol];
  const base_dp = getSymbolInfo("base_dp");
  getSymbolInfo("quote_dp");
  const displaySide = getDisplaySide(side);
  const displaySymbol = transSymbolformString(symbol);
  const displayQuantity = "algo_type" in data && data.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL ? i18n.t("tpsl.entirePosition") : base_dp === void 0 ? quantity : parseNumber(quantity, { dp: base_dp });
  let title = "";
  let msg = "";
  switch (status) {
    case OrderStatus.NEW:
      const isScaledOrder = client_order_id?.startsWith("scaled_");
      if (isScaledOrder) {
        title = i18n.t("orders.status.scaledSubOrderOpened.toast.title");
        msg = `${displaySide} ${displaySymbol} ${displayQuantity}`;
      } else {
        title = i18n.t("orders.status.opened.toast.title");
        msg = `${displaySide} ${displaySymbol} ${displayQuantity}`;
      }
      break;
    case OrderStatus.FILLED:
    case OrderStatus.PARTIAL_FILLED:
      const displayTotalExecutedQuantity = base_dp === void 0 ? total_executed_quantity : parseNumber(total_executed_quantity, { dp: base_dp });
      title = i18n.t("orders.status.filled.toast.title");
      msg = `${displaySide} ${displaySymbol} ${displayTotalExecutedQuantity} / ${displayQuantity}`;
      break;
    case OrderStatus.CANCELLED:
      title = i18n.t("orders.status.canceled.toast.title");
      msg = `${displaySide} ${displaySymbol} ${displayQuantity}`;
      break;
    case OrderStatus.REJECTED:
      title = i18n.t("orders.status.rejected.toast.title");
      msg = `${displaySide} ${displaySymbol} ${displayQuantity}`;
      break;
    case OrderStatus.REPLACED:
      const { algo_type, activated_price } = data;
      if (algo_type === AlgoOrderRootType.TRAILING_STOP) {
        const fieldChange = fieldChanges?.[AlgoOrderRootType.TRAILING_STOP] || {};
        if (fieldChange.is_activated && fieldChange.extreme_price && activated_price) {
          title = i18n.t("orders.trailingStop.activated");
          msg = `${displaySymbol} @${activated_price}`;
        } else if (fieldChange.extreme_price) {
          title = "";
          msg = "";
        }
      } else {
        title = i18n.t("orders.status.replaced.toast.title");
        msg = `${side} ${displaySymbol} ${total_executed_quantity} / ${displayQuantity}`;
      }
      break;
  }
  return {
    title,
    msg,
    status
  };
}
var ORDERLY_ORDER_SOUND_ALERT_KEY = "orderly_order_sound_alert";
var useExecutionReport = () => {
  const ee = useEventEmitter();
  const symbolsInfo = useSymbolsInfo();
  const symbolsInfoRef = useRef({});
  const { notification } = useOrderlyContext();
  useEffect(() => {
    symbolsInfoRef.current = symbolsInfo;
  }, [symbolsInfo]);
  const src = notification?.orderFilled?.media ?? "";
  const [soundAutoPlay] = useLocalStorage(
    ORDERLY_ORDER_SOUND_ALERT_KEY,
    notification?.orderFilled?.defaultOpen ?? false
  );
  const [audioElement] = useAudioPlayer(src, {
    autoPlay: soundAutoPlay,
    volume: 1
  });
  const handler = useDebouncedCallback((data) => {
    const showToast = (data2) => {
      const { title, msg, status } = getOrderExecutionReportMsg(
        data2,
        symbolsInfoRef.current
      );
      const isFilled = status === OrderStatus.FILLED || status === OrderStatus.PARTIAL_FILLED;
      const orderType = data2.algo_type || data2.type;
      if (title && msg) {
        toast.success(
          /* @__PURE__ */ jsxs("div", { children: [
            title,
            /* @__PURE__ */ jsx("br", {}),
            /* @__PURE__ */ jsx("div", { className: "orderly-text-white/[0.54] orderly-text-xs", children: msg }),
            isFilled && audioElement
          ] }),
          { id: orderType }
        );
      }
    };
    showToast(data);
  }, 100);
  useEffect(() => {
    ee.on("orders:changed", handler);
    return () => {
      ee.off("orders:changed", handler);
      handler.cancel();
    };
  }, [ee, handler]);
};
function useUILocale() {
  const { t } = useTranslation();
  const localeCode = useLocaleCode();
  return useMemo(() => {
    const calendarLocale = {
      [LocaleEnum.en]: enUS,
      [LocaleEnum.zh]: zhCN
      // [LocaleEnum.ja]: ja,
      // [LocaleEnum.es]: es,
      // [LocaleEnum.ko]: ko,
      // [LocaleEnum.vi]: vi,
      // [LocaleEnum.de]: de,
      // [LocaleEnum.fr]: fr,
      // [LocaleEnum.nl]: nl,
    };
    return {
      locale: localeCode,
      dialog: {
        ok: t("common.ok"),
        cancel: t("common.cancel")
      },
      modal: {
        confirm: t("common.confirm"),
        cancel: t("common.cancel")
      },
      pagination: {
        morePages: t("ui.pagination.morePages"),
        rowsPerPage: t("ui.pagination.rowsPerPage")
      },
      picker: {
        selectDate: t("ui.picker.selectDate"),
        dayPicker: calendarLocale[localeCode]
      },
      empty: {
        description: t("ui.empty.description")
      }
    };
  }, [t, localeCode]);
}
var AppConfigContext = createContext({});
var useAppConfig = () => {
  return useContext(AppConfigContext);
};
var AppConfigProvider = (props) => {
  return /* @__PURE__ */ jsx(AppConfigContext.Provider, { value: props, children: props.children });
};
var useAssetconvertEvent = () => {
  const ws = useWS();
  const { t } = useTranslation();
  useEffect(() => {
    const unsubscribe = ws.privateSubscribe(
      {
        id: "assetconvert",
        event: "subscribe",
        topic: "assetconvert",
        ts: getTimestamp()
      },
      {
        onMessage(data) {
          if (data.convertId) {
            toast.success(t("transfer.convert.completed"));
          }
        }
      }
    );
    return () => unsubscribe();
  }, []);
};
function useCurrentChainId(defaultChain) {
  const { storageChain, setStorageChain } = useStorageChain();
  const [currentChainId, setCurrentChainId] = useState();
  const [chains] = useChains();
  const networkId = useConfig("networkId");
  const { connectedChain } = useWalletConnector();
  useEffect(() => {
    if (connectedChain) {
      setCurrentChainId?.(
        typeof connectedChain.id === "number" ? connectedChain.id : parseInt(connectedChain.id)
      );
    } else {
      if (!!currentChainId) return;
      let fallbackChain;
      const firstChain = networkId === "mainnet" ? chains.mainnet?.[0] : chains.testnet?.[0];
      if (typeof defaultChain === "function") {
        fallbackChain = defaultChain(networkId, chains);
      } else if (typeof defaultChain === "object") {
        fallbackChain = networkId === "mainnet" ? defaultChain?.mainnet : defaultChain?.testnet;
      }
      const chainId = fallbackChain?.id || firstChain?.network_infos?.chain_id;
      if (!chainId) return;
      if (storageChain) {
        setCurrentChainId?.(storageChain.chainId);
      } else {
        setStorageChain(chainId);
        setCurrentChainId?.(chainId);
      }
    }
  }, [
    connectedChain,
    chains,
    currentChainId,
    networkId,
    setCurrentChainId,
    defaultChain
  ]);
  return [currentChainId, setCurrentChainId];
}
var WALLET_KEY = "orderly:wallet-info";
function useLinkDevice() {
  const { connectedChain, disconnect } = useWalletConnector();
  const [_, setLinkDeviceStorage] = useLocalStorage(
    "orderly_link_device",
    {}
  );
  const { account } = useAccount();
  const { isMobile } = useScreen();
  const configStore = useConfig();
  const onDisconnect = async (label) => {
    localStorage.removeItem(WALLET_KEY);
    await account.disconnect();
    await disconnect({ label });
  };
  useEffect(() => {
    const linkData = getLinkDeviceData();
    const walletInfo = JSON.parse(localStorage.getItem(WALLET_KEY) ?? "{}");
    if (linkData && walletInfo) {
      onDisconnect(walletInfo.label);
    }
  }, []);
  const linkDevice = async () => {
    const linkData = getLinkDeviceData();
    if (!linkData) return;
    const { address, secretKey, chainId, chainNamespace } = linkData;
    const isSuccess = await account.importOrderlyKey({
      address,
      secretKey,
      chainNamespace
    });
    if (!isSuccess) return;
    setLinkDeviceStorage({
      chainId,
      chainNamespace
    });
    const url = new URL(window.location.href);
    url.searchParams.delete("link");
    url.searchParams.set("utm_medium", "qrcode");
    const decodedUrl = decodeURIComponent(url.toString());
    history.replaceState(null, "", decodedUrl);
  };
  useEffect(() => {
    if (isMobile && !connectedChain) {
      linkDevice();
    }
  }, [account, connectedChain, isMobile]);
  const autoLinkDevice = async () => {
    const { chainId, chainNamespace } = getLinkDeviceStorage() || {};
    if (isMobile && !connectedChain && chainId && chainNamespace) {
      const address = account.keyStore.getAddress();
      const orderlyKey = account.keyStore.getOrderlyKey();
      const accountId = account.keyStore.getAccountId(address);
      const res = await account.checkOrderlyKey(
        address,
        orderlyKey,
        accountId
      );
      if (res) {
        configStore.set("chainNamespace", chainNamespace);
      }
    }
  };
  useEffect(() => {
    autoLinkDevice();
  }, [account, isMobile, connectedChain]);
  return { linkDevice };
}
function getLinkDeviceStorage() {
  try {
    const linkDeviceStorage = localStorage.getItem("orderly_link_device");
    const json = linkDeviceStorage ? parseJSON(linkDeviceStorage) : null;
    return json;
  } catch (err) {
  }
}
function getLinkDeviceData() {
  const url = new URL(window.location.href);
  const link = url.searchParams.get("link");
  if (!link) return;
  const {
    a: address,
    k: secretKey,
    i: chainId,
    n: chainNamespace
  } = decodeBase64(link) || {};
  if (address && secretKey && chainId && chainNamespace) {
    return {
      address,
      secretKey,
      chainId,
      chainNamespace
    };
  }
}
function decodeBase64(base64) {
  try {
    const data = JSON.parse(window.atob(base64));
    const currentTime = Math.floor(Date.now() / 1e3);
    const expiredTime = data.t;
    if (!expiredTime || currentTime > expiredTime) {
      return;
    }
    return data;
  } catch (error) {
  }
}
function useSettleEvent() {
  const { t } = useTranslation();
  useSettleSubscription({
    onMessage: (data) => {
      const { status } = data;
      switch (status) {
        case "COMPLETED":
          toast.success(t("settle.settlement.completed"));
          break;
        case "FAILED":
          toast.error(t("settle.settlement.failed"));
          break;
      }
    }
  });
}
function useWalletConnectError() {
  const { t } = useTranslation();
  const ee = useEventEmitter();
  const { setLedgerAddress } = useStorageLedgerAddress();
  useEffect(() => {
    const handleConnectError = (data) => {
      toast.error(data.message);
    };
    const handleLedgerError = (data) => {
      window.setTimeout(() => {
        modal.confirm({
          title: t("connector.ledger.signMessageFailed"),
          content: t("connector.ledger.signMessageFailed.description"),
          size: "sm",
          onOk: async () => {
            setLedgerAddress(data.userAddress);
            return Promise.resolve();
          },
          okLabel: t("common.ok"),
          onCancel: async () => {
            toast.error(data.message);
            return Promise.resolve();
          },
          cancelLabel: t("common.no")
        }).then((res) => {
        });
      });
    };
    ee.on("wallet:connect-error", handleConnectError);
    ee.on("wallet:sign-message-with-ledger-error", handleLedgerError);
    return () => {
      ee.off("wallet:connect-error", handleConnectError);
      ee.off("wallet:sign-message-with-ledger-error", handleLedgerError);
    };
  }, [ee, t]);
  return {};
}
function useWalletEvent() {
  const { t } = useTranslation();
  const ee = useEventEmitter();
  const recordRef = useRef({});
  const [record, setRecord] = useSessionStorage(
    "orderly_wallet_change_id",
    {}
  );
  recordRef.current = record;
  useWalletSubscription({
    onMessage: (data) => {
      const { id, side, transStatus } = data;
      let showToast = true;
      if (["DEPOSIT", "WITHDRAW"].includes(side) && ["COMPLETED", "FAILED"].includes(transStatus)) {
        const isPushed = !!recordRef.current[id];
        if (!isPushed) {
          recordRef.current[id] = true;
          setRecord((prev) => ({
            ...prev,
            [id]: true
          }));
        }
        showToast = !isPushed;
      }
      if (transStatus === "COMPLETED" && showToast) {
        let msg = `${capitalizeString(side)} completed`;
        if (side === "DEPOSIT") {
          msg = t("transfer.deposit.completed");
        } else if (side === "WITHDRAW") {
          msg = t("transfer.withdraw.completed");
        }
        toast.success(msg);
      } else if (transStatus === "FAILED" && showToast) {
        let msg = `${capitalizeString(side)} failed`;
        if (side === "DEPOSIT") {
          msg = t("transfer.deposit.failed");
        } else if (side === "WITHDRAW") {
          msg = t("transfer.withdraw.failed");
        }
        toast.error(msg);
      }
      ee.emit("wallet:changed", data);
    }
  });
}
var WALLET_KEY2 = "orderly:wallet-info";
var useWalletStateHandle = (options) => {
  const {
    wallet: connectedWallet,
    connect,
    connectedChain,
    disconnect,
    namespace
  } = useWalletConnector();
  if (typeof connect !== "function") {
    throw new SDKError("Please provide a wallet connector provider");
  }
  useEventEmitter();
  const isManualConnect = useRef(false);
  useConfig("brokerId");
  const { account, state: accountState } = useAccount();
  const keyStore = useKeyStore();
  const networkId = useConfig("networkId");
  const [chains, { checkChainSupport }] = useChains();
  const { storageChain, setStorageChain } = useStorageChain();
  const [unsupported, setUnsupported] = useState(false);
  const { track, setTrackUserId } = useTrack();
  const currentWalletAddress = useMemo(() => {
    return connectedWallet?.accounts?.[0]?.address;
  }, [connectedWallet]);
  const currentChain = useMemo(() => {
    const id = connectedWallet?.chains?.[0]?.id;
    const namespace2 = connectedWallet?.chains?.[0]?.namespace;
    if (typeof id === "undefined") return void 0;
    return {
      id: parseChainIdToNumber(id),
      namespace: namespace2
    };
  }, [connectedWallet]);
  useEffect(() => {
    if (accountState.status >= AccountStatusEnum.EnableTrading && account.accountId) {
      setTrackUserId(account.accountId);
    }
  }, [account, accountState]);
  useEffect(() => {
    if (!connectedChain) {
      setUnsupported(false);
      return;
    }
    let isSupported = checkChainSupport(
      connectedChain.id,
      networkId
      // networkId === "testnet" ? chains.testnet : chains.mainnet
    );
    if (ABSTRACT_CHAIN_ID_MAP.has(parseInt(connectedChain.id)) && connectedWallet?.label !== "AGW") {
      isSupported = false;
    }
    setUnsupported(!isSupported);
  }, [connectedChain, chains, checkChainSupport, networkId, connectedWallet]);
  useEffect(() => {
    windowGuard(() => {
      const localAddress = keyStore.getAddress();
      const walletInfo = JSON.parse(localStorage.getItem(WALLET_KEY2) ?? "{}");
      if (connectedChain?.namespace === ChainNamespace.solana) {
        return;
      }
      if (storageChain?.namespace === ChainNamespace.solana) {
        return;
      }
      if (localAddress && account.address !== localAddress && walletInfo.label) {
        connect({
          autoSelect: {
            label: walletInfo.label,
            disableModals: true
          }
        }).then(
          (res) => {
          },
          (error) => void 0
        );
      }
    });
  }, [connectedWallet, account.address, storageChain?.namespace]);
  useEffect(() => {
    if (connectedWallet === null && accountState.status > AccountStatusEnum.NotConnected && !accountState.validating) {
      account.disconnect();
      return;
    }
    if (unsupported || !connectedChain) return;
    if (isManualConnect.current) return;
    const linkData = getLinkDeviceData();
    if (!!currentWalletAddress && currentWalletAddress !== account.address && !linkData) {
      account.setAddress(currentWalletAddress, {
        provider: connectedWallet?.provider,
        chain: {
          id: praseChainIdToNumber(currentChain.id),
          namespace: currentChain.namespace.toUpperCase()
        },
        wallet: {
          name: connectedWallet?.label ?? ""
        },
        additionalInfo: connectedWallet?.additionalInfo ?? {}
      });
      track(TrackerEventName.walletConnect, {
        wallet: connectedWallet?.label ?? "",
        network: currentChain.namespace.toUpperCase()
      });
      windowGuard(() => {
        localStorage.setItem(
          WALLET_KEY2,
          JSON.stringify({
            label: connectedWallet?.label ?? ""
          })
        );
        setStorageChain(currentChain.id);
      });
    }
    if (currentChain?.id !== account.chainId) {
      account.switchChainId(currentChain?.id);
    }
  }, [
    connectedWallet,
    connectedChain,
    currentWalletAddress,
    currentChain,
    account.address,
    accountState,
    account.chainId,
    unsupported
  ]);
  const connectWallet = async () => {
    isManualConnect.current = true;
    return connect({ chainId: options.currentChainId }).then(async (walletState) => {
      if (Array.isArray(walletState) && walletState.length > 0 && walletState[0] && walletState[0].accounts.length > 0) {
        const wallet = walletState[0];
        const chainId = praseChainIdToNumber(wallet.chains[0].id);
        if (!checkChainSupport(chainId, networkId)) {
          return {
            wrongNetwork: true
          };
        }
        if (!account) {
          throw new Error("account is not initialized");
        }
        if (accountState.status === AccountStatusEnum.EnableTradingWithoutConnected) {
          localStorage.removeItem("orderly_link_device");
          await account.disconnect();
        }
        const status = await account.setAddress(wallet.accounts[0].address, {
          provider: wallet.provider,
          chain: {
            id: praseChainIdToNumber(wallet.chains[0].id),
            namespace: wallet.chains[0].namespace.toUpperCase()
          },
          wallet: {
            name: wallet.label
          }
          // label: ,
        });
        track(TrackerEventName.walletConnect, {
          wallet: wallet.label,
          network: wallet.chains[0].namespace.toUpperCase()
        });
        return { wallet, status, wrongNetwork: false };
      }
      return null;
    }).finally(() => {
      isManualConnect.current = false;
    });
  };
  return {
    connectWallet,
    wrongNetwork: unsupported
  };
};
var AppStateContext = createContext({
  setCurrentChainId: (chainId) => {
  },
  restrictedInfo: {},
  setShowAnnouncement: (show) => {
  }
});
var useAppContext = () => {
  return useContext(AppStateContext);
};
var AppStateProvider = (props) => {
  const [showAnnouncement, setShowAnnouncement] = useState(false);
  const [initialized, setInitialized] = useState(false);
  const [currentChainId, setCurrentChainId] = useCurrentChainId(
    props.defaultChain
  );
  useLinkDevice();
  useTrackingInstance();
  const { connectWallet, wrongNetwork } = useWalletStateHandle({
    // onChainChanged: props.onChainChanged,
    currentChainId
  });
  useWalletEvent();
  useSettleEvent();
  useAssetconvertEvent();
  useWalletConnectError();
  const restrictedInfo = useRestrictedInfo(props.restrictedInfo);
  const disabledConnect = restrictedInfo.restrictedOpen;
  useEffect(() => {
    setInitialized(true);
  }, []);
  const memoizedValue = useMemo(
    () => ({
      connectWallet,
      wrongNetwork,
      currentChainId,
      setCurrentChainId,
      onChainChanged: props.onChainChanged,
      disabledConnect,
      restrictedInfo,
      showAnnouncement,
      setShowAnnouncement,
      onRouteChange: props.onRouteChange,
      widgetConfigs: props.widgetConfigs,
      initialized
    }),
    [
      connectWallet,
      currentChainId,
      disabledConnect,
      props.onChainChanged,
      restrictedInfo,
      setCurrentChainId,
      showAnnouncement,
      wrongNetwork,
      props.onRouteChange,
      props.widgetConfigs,
      initialized
    ]
  );
  return /* @__PURE__ */ jsx(AppStateContext.Provider, { value: memoizedValue, children: props.children });
};
var ExecutionReportListener = () => {
  useExecutionReport();
  return null;
};
var OrderlyAppProvider = (props) => {
  const {
    // dateFormatting,
    components,
    appIcons,
    onChainChanged,
    defaultChain,
    widgetConfigs,
    ...configProps
  } = props;
  useTrack();
  useBootstrap();
  const uiLocale = useUILocale();
  return /* @__PURE__ */ jsx(AppConfigProvider, { appIcons, brokerName: props.brokerName, children: /* @__PURE__ */ jsx(
    OrderlyThemeProvider,
    {
      components,
      overrides: props.overrides,
      children: /* @__PURE__ */ jsxs(OrderlyConfigProvider, { ...configProps, children: [
        /* @__PURE__ */ jsx(ExecutionReportListener, {}),
        /* @__PURE__ */ jsx(
          AppStateProvider,
          {
            onChainChanged,
            defaultChain,
            restrictedInfo: props.restrictedInfo,
            onRouteChange: props.onRouteChange,
            widgetConfigs,
            children: /* @__PURE__ */ jsx(LocaleProvider, { locale: uiLocale, children: /* @__PURE__ */ jsx(TooltipProvider, { delayDuration: 300, children: /* @__PURE__ */ jsx(ModalProvider, { children: props.children }) }) })
          }
        ),
        /* @__PURE__ */ jsx(Toaster, {})
      ] })
    }
  ) });
};
if (process.env.NODE_ENV !== "production") {
  OrderlyAppProvider.displayName = "OrderlyAppProvider";
}
var useDataTap = (data, options) => {
  const { wrongNetwork, disabledConnect } = useAppContext();
  const { state } = useAccount();
  if (options?.skip) {
    return data;
  }
  if (wrongNetwork || disabledConnect) {
    return typeof options?.fallbackData !== "undefined" ? options.fallbackData : null;
  }
  if (typeof options?.accountStatus !== "undefined") {
    if (state.status < options.accountStatus) {
      return typeof options?.fallbackData !== "undefined" ? options.fallbackData : null;
    }
  }
  return data;
};
function useOrderEntryFormErrorMsg(errors) {
  const { t } = useTranslation();
  const getMessage = (key, type, params) => {
    const { value, min, max } = params || {};
    const map = {
      quantity: {
        required: t("orderEntry.orderQuantity.error.required"),
        min: t("orderEntry.orderQuantity.error.min", { value }),
        max: t("orderEntry.orderQuantity.error.max", { value })
      },
      order_quantity: {
        required: t("orderEntry.orderQuantity.error.required"),
        min: t("orderEntry.orderQuantity.error.min", { value }),
        max: t("orderEntry.orderQuantity.error.max", { value })
      },
      order_price: {
        required: t("orderEntry.orderPrice.error.required"),
        min: t("orderEntry.orderPrice.error.min", { value }),
        max: t("orderEntry.orderPrice.error.max", { value })
      },
      trigger_price: {
        required: t("orderEntry.triggerPrice.error.required"),
        min: t("orderEntry.triggerPrice.error.min", { value }),
        max: t("orderEntry.triggerPrice.error.max", { value })
      },
      tp_trigger_price: {
        required: t("tpsl.validate.tpTriggerPrice.error.required"),
        min: t("orderEntry.tpTriggerPrice.error.min", { value }),
        max: t("orderEntry.tpTriggerPrice.error.max", { value }),
        priceErrorMin: t("tpsl.validate.tpTriggerPrice.error.priceErrorMin"),
        priceErrorMax: t("tpsl.validate.tpTriggerPrice.error.priceErrorMax")
      },
      sl_trigger_price: {
        required: t("tpsl.validate.slTriggerPrice.error.required"),
        min: t("orderEntry.slTriggerPrice.error.min", { value }),
        max: t("orderEntry.slTriggerPrice.error.max", { value }),
        priceErrorMin: t("tpsl.validate.slTriggerPrice.error.priceErrorMin"),
        priceErrorMax: t("tpsl.validate.slTriggerPrice.error.priceErrorMax"),
        [ERROR_MSG_CODES.SL_PRICE_WARNING]: t(
          "tpsl.validate.slTriggerPrice.warning.closeToLiqPrice"
        ),
        [ERROR_MSG_CODES.SL_PRICE_ERROR]: t(
          "tpsl.validate.slTriggerPrice.error.crossLiqPrice"
        )
      },
      tp_order_price: {
        required: t("tpsl.validate.tpOrderPrice.error.required"),
        min: t("tpsl.validate.tpOrderPrice.error.min", { value }),
        max: t("tpsl.validate.tpOrderPrice.error.max", { value })
      },
      sl_order_price: {
        required: t("tpsl.validate.slOrderPrice.error.required"),
        min: t("tpsl.validate.slOrderPrice.error.min", { value }),
        max: t("tpsl.validate.slOrderPrice.error.max", { value })
      },
      total: {
        min: t("orderEntry.total.error.min", { value })
      },
      // not show form input tooltip
      // slippage: {
      //   max: t("orderEntry.slippage.error.max"),
      // },
      start_price: {
        required: t("orderEntry.startPrice.error.required"),
        min: t("orderEntry.startPrice.error.min", { value }),
        max: t("orderEntry.startPrice.error.max", { value })
      },
      end_price: {
        required: t("orderEntry.endPrice.error.required"),
        min: t("orderEntry.endPrice.error.min", { value }),
        max: t("orderEntry.endPrice.error.max", { value })
      },
      total_orders: {
        required: t("orderEntry.totalOrders.error.required"),
        range: t("orderEntry.totalOrders.error.range")
      },
      skew: {
        required: t("orderEntry.skew.error.required"),
        min: t("orderEntry.skew.error.min", { value }),
        max: t("orderEntry.skew.error.max", { value })
      },
      activated_price: {
        min: t("orderEntry.triggerPrice.error.min", { value }),
        max: t("orderEntry.triggerPrice.error.max", { value })
      },
      callback_value: {
        required: t("orderEntry.callbackValue.error.required"),
        min: t("orderEntry.callbackValue.error.min", { value }),
        range: t("orderEntry.callbackValue.error.range", {
          min,
          max
        })
      },
      callback_rate: {
        required: t("orderEntry.callbackRate.error.required"),
        range: t("orderEntry.callbackRate.error.range", {
          min,
          max
        })
      }
    };
    return map[key]?.[type] || "";
  };
  const getErrorMsg = useCallback(
    (key, customValue) => {
      const { type, value, min, max } = errors?.[key] || {};
      if (type) {
        return getMessage(key, type, { value: customValue || value, min, max });
      }
      return "";
    },
    [errors]
  );
  return {
    getErrorMsg
  };
}
var ErrorBoundary = class extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }
  static getDerivedStateFromError(error) {
    return { hasError: true, error };
  }
  componentDidCatch(error, errorInfo) {
    this.setState({
      error,
      errorInfo
    });
    if (process.env.NODE_ENV === "development") ;
    this.props.onError?.(error, errorInfo);
  }
  render() {
    if (this.state.hasError) {
      return /* @__PURE__ */ jsxs(
        Flex,
        {
          direction: "column",
          width: "100%",
          itemAlign: "center",
          justify: "center",
          className: cn("oui-h-screen oui-px-[10%]", this.props.className),
          children: [
            this.props.icon || /* @__PURE__ */ jsx(Icon, {}),
            /* @__PURE__ */ jsx(
              Text,
              {
                className: cn(
                  "oui-mt-10",
                  "oui-text-[32px] oui-leading-[40px]",
                  "md:oui-text-[48px] md:oui-leading-[56px]"
                ),
                intensity: 80,
                weight: "bold",
                children: this.props.title || i18n.t("errorBoundary.title")
              }
            ),
            /* @__PURE__ */ jsx(
              Text,
              {
                className: cn(
                  "oui-mt-2",
                  "oui-text-[16px] oui-leading-[24px]",
                  "md:oui-text-[20px] md:oui-leading-[28px]",
                  "oui-text-center"
                ),
                weight: "semibold",
                intensity: 36,
                children: this.props.description || i18n.t("errorBoundary.description")
              }
            ),
            /* @__PURE__ */ jsxs(
              Button,
              {
                onClick: () => {
                  if (typeof this.props.onRefresh === "function") {
                    this.props.onRefresh();
                  } else {
                    window.location.reload();
                  }
                },
                className: "oui-mt-4 oui-font-semibold",
                variant: "gradient",
                size: "md",
                children: [
                  /* @__PURE__ */ jsx(RefreshIcon, { className: "oui-mr-1" }),
                  this.props.refreshButtonText || i18n.t("errorBoundary.refresh")
                ]
              }
            )
          ]
        }
      );
    }
    return this.props.children;
  }
};
var Icon = (props) => {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      width: "156",
      height: "165",
      viewBox: "0 0 156 165",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      className: props.className,
      children: [
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M77.958 11.2581C117.408 11.2581 149.224 43.7623 149.225 83.635C149.225 123.508 117.408 156.013 77.958 156.013C38.5076 156.013 6.69141 123.508 6.69141 83.635C6.69161 43.7623 38.5077 11.2581 77.958 11.2581Z",
            stroke: "url(#paint0_linear_32953_90176)",
            strokeWidth: "12"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M74.1236 18.8531C74.8967 18.2272 76.0309 18.3465 76.6568 19.1196L99.9666 47.9101C100.593 48.6832 100.473 49.8174 99.7001 50.4433C98.927 51.0692 97.7928 50.9499 97.1669 50.1768L73.8571 21.3863C73.2312 20.6132 73.3505 19.479 74.1236 18.8531Z",
            fill: "#394155"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M67.4775 153.345C66.5687 153.75 65.5041 153.341 65.0996 152.432L47.5699 113.05C47.1654 112.142 47.5742 111.077 48.483 110.672C49.3918 110.268 50.4564 110.677 50.8609 111.585L68.3906 150.967C68.7951 151.876 68.3863 152.941 67.4775 153.345Z",
            fill: "#394155"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M87.1937 2.50608C88.9032 1.12201 91.4111 1.38582 92.7951 3.09532L117.65 33.7939C119.034 35.5033 118.77 38.0112 117.06 39.3952L106.349 48.0675C104.64 49.4515 102.132 49.1877 100.748 47.4782L75.8932 16.7797C74.5092 15.0702 74.773 12.5624 76.4825 11.1783L87.1937 2.50608Z",
            fill: "url(#paint1_linear_32953_90176)"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M54.8689 163.161C52.8595 164.055 50.5054 163.152 49.6109 161.142L30.369 117.913C29.4745 115.904 30.3784 113.55 32.3879 112.655L38.8256 109.79C40.8351 108.895 43.1892 109.799 44.0836 111.809L63.3256 155.037C64.22 157.047 63.3161 159.401 61.3067 160.295L54.8689 163.161Z",
            fill: "url(#paint2_linear_32953_90176)"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M21.0497 137.635C19.2748 138.425 17.1955 137.627 16.4054 135.852C15.6154 134.077 16.4138 131.998 18.1887 131.208L29.4767 126.183C31.2516 125.393 33.331 126.191 34.121 127.966C34.911 129.741 34.1127 131.821 32.3377 132.611L21.0497 137.635Z",
            fill: "url(#paint3_linear_32953_90176)"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M31.0067 160.657C29.2318 161.447 27.1525 160.649 26.3625 158.874C25.5724 157.099 26.3708 155.02 28.1457 154.23L39.4338 149.205C41.2087 148.415 43.288 149.213 44.078 150.988C44.8681 152.763 44.0697 154.843 42.2948 155.633L31.0067 160.657Z",
            fill: "url(#paint4_linear_32953_90176)"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M94.2979 17.9219C94.2979 19.7549 92.812 21.2408 90.979 21.2408C89.1461 21.2408 87.6602 19.7549 87.6602 17.9219C87.6602 16.0889 89.1461 14.603 90.979 14.603C92.812 14.603 94.2979 16.0889 94.2979 17.9219Z",
            fill: "#07080A"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M106.243 32.5249C106.243 34.3579 104.757 35.8438 102.924 35.8438C101.091 35.8438 99.6055 34.3579 99.6055 32.5249C99.6055 30.692 101.091 29.2061 102.924 29.2061C104.757 29.2061 106.243 30.692 106.243 32.5249Z",
            fill: "#07080A"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M143.238 87.3294C143.283 86.3294 143.306 85.3235 143.306 84.3123C143.306 80.0077 142.89 75.8003 142.096 71.728C137.49 69.544 132.251 67.6307 127.118 66.8126C122.674 66.1042 117.636 66.1072 113.053 68.1318C108.166 70.2911 104.601 74.3958 102.851 80.1999C101.677 84.091 100.713 87.7506 99.857 90.9972L99.8569 90.9974C99.5483 92.1682 99.2539 93.2853 98.969 94.3402C97.8574 98.455 96.882 101.656 95.6795 104.319C94.5159 106.895 93.2041 108.812 91.5007 110.366C89.7881 111.929 87.3882 113.389 83.7037 114.64L87.9295 127.083C92.9507 125.378 97.0389 123.101 100.357 120.074C103.684 117.039 105.944 113.516 107.655 109.727C109.327 106.025 110.528 101.937 111.655 97.7671C111.977 96.5735 112.294 95.3715 112.616 94.1496L112.616 94.1488L112.616 94.1487C113.456 90.9644 114.33 87.646 115.432 83.9941C116.178 81.5196 117.297 80.6226 118.363 80.1517C119.734 79.5462 121.92 79.2905 125.05 79.7894C131.168 80.7644 138.314 84.1982 143.238 87.3294ZM13.9805 70.6827C14.9192 66.2565 16.3058 61.9957 18.0906 57.9501C23.709 58.5554 30.4365 58.6747 37.3538 57.7247C48.6584 56.1721 59.8141 51.8684 67.7619 42.9131L77.5901 51.6356C66.8783 63.7052 52.3551 68.9284 39.1417 70.7431C29.9766 72.0018 21.1264 71.6711 13.9805 70.6827Z",
            fill: "#394155"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M68.7641 23.3532C70.4736 21.9692 72.9814 22.233 74.3655 23.9425L94.9298 49.3421C96.3139 51.0515 96.0501 53.5594 94.3406 54.9434L86.1984 61.5356C84.4889 62.9197 81.9811 62.6559 80.597 60.9464L60.0326 35.5468C58.6486 33.8373 58.9124 31.3295 60.6219 29.9454L68.7641 23.3532Z",
            fill: "#535E7B"
          }
        ),
        /* @__PURE__ */ jsx(
          "path",
          {
            d: "M74.0447 150.372C72.0352 151.267 69.6811 150.363 68.7867 148.353L53.0318 112.959C52.1373 110.949 53.0412 108.595 55.0507 107.7L72.5225 99.9234C74.532 99.029 76.8861 99.9329 77.7806 101.942L93.5355 137.337C94.4299 139.347 93.526 141.701 91.5165 142.595L74.0447 150.372Z",
            fill: "#535E7B"
          }
        ),
        /* @__PURE__ */ jsxs("defs", { children: [
          /* @__PURE__ */ jsxs(
            "linearGradient",
            {
              id: "paint0_linear_32953_90176",
              x1: "71.7422",
              y1: "138.236",
              x2: "124.758",
              y2: "41.3888",
              gradientUnits: "userSpaceOnUse",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "#1B1D22" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "#26292E" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "linearGradient",
            {
              id: "paint1_linear_32953_90176",
              x1: "111.705",
              y1: "43.7314",
              x2: "81.8381",
              y2: "6.84218",
              gradientUnits: "userSpaceOnUse",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "linearGradient",
            {
              id: "paint2_linear_32953_90176",
              x1: "35.6067",
              y1: "111.223",
              x2: "58.0878",
              y2: "161.728",
              gradientUnits: "userSpaceOnUse",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "linearGradient",
            {
              id: "paint3_linear_32953_90176",
              x1: "23.8327",
              y1: "128.695",
              x2: "26.6937",
              y2: "135.123",
              gradientUnits: "userSpaceOnUse",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            "linearGradient",
            {
              id: "paint4_linear_32953_90176",
              x1: "33.7897",
              y1: "151.717",
              x2: "36.6508",
              y2: "158.145",
              gradientUnits: "userSpaceOnUse",
              children: [
                /* @__PURE__ */ jsx("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
                /* @__PURE__ */ jsx("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
              ]
            }
          )
        ] })
      ]
    }
  );
};
var RefreshIcon = (props) => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      width: "15",
      height: "15",
      viewBox: "0 0 15 15",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      className: props.className,
      children: /* @__PURE__ */ jsx(
        "path",
        {
          d: "M7.5026 2.52515C6.30444 2.52515 5.15994 2.92706 4.23944 3.65565C3.98686 3.85515 3.94835 4.22322 4.14844 4.47581C4.34852 4.72839 4.71602 4.76689 4.96861 4.56681C5.68436 4.00039 6.56985 3.69181 7.5026 3.69181C9.75777 3.69181 11.5859 5.51998 11.5859 7.77514H10.4193L12.1693 10.1085L13.9193 7.77514H12.7526C12.7526 4.87539 10.4024 2.52515 7.5026 2.52515ZM2.83594 5.44181L1.08594 7.77514H2.2526C2.2526 10.6749 4.60285 13.0251 7.5026 13.0251C8.70135 13.0251 9.84527 12.6238 10.7658 11.8946C11.0184 11.6951 11.0569 11.3271 10.8568 11.0745C10.6567 10.8219 10.2892 10.7834 10.0366 10.9835C9.32027 11.5505 8.43594 11.8585 7.5026 11.8585C5.24744 11.8585 3.41927 10.0303 3.41927 7.77514H4.58594L2.83594 5.44181Z",
          fill: "black"
        }
      )
    }
  );
};
function useCanTrade() {
  const { state } = useAccount();
  const { wrongNetwork, disabledConnect } = useAppContext();
  const canTrade = useMemo(() => {
    return !wrongNetwork && !disabledConnect && (state.status === AccountStatusEnum.EnableTrading || state.status === AccountStatusEnum.EnableTradingWithoutConnected);
  }, [state.status, wrongNetwork, disabledConnect]);
  return canTrade;
}

export { ErrorBoundary, OrderlyAppProvider, useAppConfig, useAppContext, useCanTrade, useDataTap, useOrderEntryFormErrorMsg };
