// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/worker_interface/types/QuoteDvnParams.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { LzOption, lzOptionBeet } from './LzOption'
export type QuoteDvnParams = {
  msglib: web3.PublicKey
  dstEid: number
  sender: web3.PublicKey
  packetHeader: Uint8Array
  payloadHash: number[] /* size: 32 */
  confirmations: beet.bignum
  options: LzOption[]
}

/**
 * @category userTypes
 * @category generated
 */
export const quoteDvnParamsBeet =
  new beet.FixableBeetArgsStruct<QuoteDvnParams>(
    [
      ['msglib', beetSolana.publicKey],
      ['dstEid', beet.u32],
      ['sender', beetSolana.publicKey],
      ['packetHeader', beet.bytes],
      ['payloadHash', beet.uniformFixedSizeArray(beet.u8, 32)],
      ['confirmations', beet.u64],
      ['options', beet.array(lzOptionBeet)],
    ],
    'QuoteDvnParams'
  )
