// @orderly.network/default-solana-adapter/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.ts
var index_exports = {};
__export(index_exports, {
  DefaultSolanaWalletAdapter: () => DefaultSolanaWalletAdapter,
  version: () => version_default
});
module.exports = __toCommonJS(index_exports);

// src/version.ts
if (typeof window !== "undefined") {
  window.__ORDERLY_VERSION__ = window.__ORDERLY_VERSION__ || {};
  window.__ORDERLY_VERSION__["@orderly.network/default-solana-adapter"] = "2.10.1";
}
var version_default = "2.10.1";

// src/walletAdapter.ts
var ed = __toESM(require("@noble/ed25519"));
var import_spl_token2 = require("@solana/spl-token");
var import_wallet_adapter_base = require("@solana/wallet-adapter-base");
var import_web34 = require("@solana/web3.js");
var import_bs582 = require("bs58");
var import_utils2 = require("ethereum-cryptography/utils");
var import_core2 = require("@orderly.network/core");
var import_types2 = require("@orderly.network/types");

// src/helper.ts
var import_anchor = require("@coral-xyz/anchor");
var import_web33 = require("@solana/web3.js");
var import_bs58 = require("bs58");
var import_keccak = require("ethereum-cryptography/keccak");
var import_utils = require("ethereum-cryptography/utils");
var import_ethers = require("ethers");
var import_core = require("@orderly.network/core");
var import_types = require("@orderly.network/types");

// src/constant.ts
var import_lz_v2_utilities = require("@layerzerolabs/lz-v2-utilities");
var import_web3 = require("@solana/web3.js");
var ENDPOINT_PROGRAM_ID = new import_web3.PublicKey(
  "76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6"
);
var SEND_LIB_PROGRAM_ID = new import_web3.PublicKey(
  "7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH"
);
var EXECUTOR_PROGRAM_ID = new import_web3.PublicKey(
  "6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn"
);
var PRICE_FEED_PROGRAM_ID = new import_web3.PublicKey(
  "8ahPGPjEbpgGaZx2NV1iG5Shj7TDwvsjkEDcGWjt94TP"
);
var RECEIVE_LIB_PROGRAM_ID = SEND_LIB_PROGRAM_ID;
var TREASURY_PROGRAM_ID = SEND_LIB_PROGRAM_ID;
var DVN_PROGRAM_ID = new import_web3.PublicKey(
  "HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW"
);
var VAULT_AUTHORITY_SEED = "VaultAuthority";
var BROKER_SEED = "Broker";
var TOKEN_SEED = "Token";
var SOL_VAULT_SEED = "SolVault";
var DEV_PEER_ADDRESS = (0, import_lz_v2_utilities.addressToBytes32)(
  "0x9Dc724b24146BeDD2dA28b8C4B74126169B8f312"
);
var QA_PEER_ADDRESS = (0, import_lz_v2_utilities.addressToBytes32)(
  "0x45b6C6266A7A2170617d8A27A50C642fd68b91c4"
);
var STAGING_PEER_ADDRESS = (0, import_lz_v2_utilities.addressToBytes32)(
  "0x5Bf771A65d057e778C5f0Ed52A0003316f94322D"
);
var MAINNET_PEER_ADDRESS = (0, import_lz_v2_utilities.addressToBytes32)(
  "0xCecAe061aa078e13b5e70D5F9eCee90a3F2B6AeA"
);
var DEV_DST_EID = 40200;
var MAIN_DST_EID = 30213;
var DEV_LOOKUP_TABLE_ADDRESS = new import_web3.PublicKey(
  "BWp8HaYYhiNHekt3zgQhqoCrRftneGxxfgKmCZ6svHN"
);
var QA_LOOKUP_TABLE_ADDRESS = new import_web3.PublicKey(
  "BswrQQoPKAFojTuJutZcBMtigAgTghEH4M8ofn3EG2X2"
);
var STAGING_LOOKUP_TABLE_ADDRESS = new import_web3.PublicKey(
  "BbGKfxuPwDmu58BjPpd7PMG69TqnZjSpKaLDMgf9E9Dr"
);
var MAINNET_LOOKUP_TABLE_ADDRESS = new import_web3.PublicKey(
  "8iq7xCQt3bLdRRn4A46d5GuaXYinBoiAhbe2sUmZVzwg"
);
var DEV_OAPP_PROGRAM_ID = new import_web3.PublicKey(
  "EYJq9eU4GMRUriUJBgGoZ8YLQBXcWaciXuSsEXE7ieQS"
);
var QA_OAPP_PROGRAM_ID = new import_web3.PublicKey(
  "5zBjLor7vEraAt4zp2H82sy9MSqFoDnNa1Lx6EYKTYRZ"
);
var STAGING_OAPP_PROGRAM_ID = new import_web3.PublicKey(
  "9shwxWDUNhtwkHocsUAmrNAQfBH2DHh4njdAEdHZZkF2"
);
var MAINNET_OAPP_PROGRAM_ID = new import_web3.PublicKey(
  "ErBmAD61mGFKvrFNaTJuxoPwqrS8GgtwtqJTJVjFWx9Q"
);

// src/idl/solana_vault.ts
var IDL = {
  version: "0.1.0",
  name: "solana_vault",
  instructions: [
    {
      name: "setVault",
      accounts: [
        {
          name: "admin",
          isMut: true,
          isSigner: true
        },
        {
          name: "vaultAuthority",
          isMut: true,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetVaultParams"
          }
        }
      ]
    },
    {
      name: "deposit",
      accounts: [
        {
          name: "user",
          isMut: true,
          isSigner: true
        },
        {
          name: "userTokenAccount",
          isMut: true,
          isSigner: false
        },
        {
          name: "vaultAuthority",
          isMut: true,
          isSigner: false
        },
        {
          name: "vaultTokenAccount",
          isMut: true,
          isSigner: false
        },
        {
          name: "depositToken",
          isMut: false,
          isSigner: false
        },
        {
          name: "peer",
          isMut: false,
          isSigner: false
        },
        {
          name: "enforcedOptions",
          isMut: false,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "allowedBroker",
          isMut: false,
          isSigner: false
        },
        {
          name: "allowedToken",
          isMut: false,
          isSigner: false
        },
        {
          name: "tokenProgram",
          isMut: false,
          isSigner: false
        },
        {
          name: "associatedTokenProgram",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "depositParams",
          type: {
            defined: "DepositParams"
          }
        },
        {
          name: "oappParams",
          type: {
            defined: "OAppSendParams"
          }
        }
      ],
      returns: {
        defined: "MessagingReceipt"
      }
    },
    {
      name: "depositSol",
      accounts: [
        {
          name: "user",
          isMut: true,
          isSigner: true
        },
        {
          name: "vaultAuthority",
          isMut: true,
          isSigner: false
        },
        {
          name: "solVault",
          isMut: true,
          isSigner: false,
          docs: ["CHECKED: sol_vault is used for SOL deposit"]
        },
        {
          name: "peer",
          isMut: false,
          isSigner: false
        },
        {
          name: "enforcedOptions",
          isMut: false,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "allowedBroker",
          isMut: false,
          isSigner: false
        },
        {
          name: "allowedToken",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "depositParams",
          type: {
            defined: "DepositParams"
          }
        },
        {
          name: "oappParams",
          type: {
            defined: "OAppSendParams"
          }
        }
      ],
      returns: {
        defined: "MessagingReceipt"
      }
    },
    {
      name: "initOapp",
      accounts: [
        {
          name: "payer",
          isMut: true,
          isSigner: true
        },
        {
          name: "oappConfig",
          isMut: true,
          isSigner: false
        },
        {
          name: "lzReceiveTypes",
          isMut: true,
          isSigner: false
        },
        {
          name: "accountList",
          isMut: true,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "InitOAppParams"
          }
        }
      ]
    },
    {
      name: "setAccountList",
      accounts: [
        {
          name: "admin",
          isMut: true,
          isSigner: true
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "lzReceiveTypes",
          isMut: true,
          isSigner: false
        },
        {
          name: "accountsList",
          isMut: true,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetAccountListParams"
          }
        }
      ]
    },
    {
      name: "setManagerRole",
      accounts: [
        {
          name: "owner",
          isMut: true,
          isSigner: true
        },
        {
          name: "vaultAuthority",
          isMut: true,
          isSigner: false
        },
        {
          name: "managerRole",
          isMut: true,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetManagerRoleParams"
          }
        }
      ]
    },
    {
      name: "setBroker",
      accounts: [
        {
          name: "brokerManager",
          isMut: true,
          isSigner: true
        },
        {
          name: "allowedBroker",
          isMut: true,
          isSigner: false
        },
        {
          name: "managerRole",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetBrokerParams"
          }
        }
      ]
    },
    {
      name: "setWithdrawBroker",
      accounts: [
        {
          name: "brokerManager",
          isMut: true,
          isSigner: true
        },
        {
          name: "withdrawBroker",
          isMut: true,
          isSigner: false
        },
        {
          name: "managerRole",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetWithdrawBrokerParams"
          }
        }
      ]
    },
    {
      name: "setToken",
      accounts: [
        {
          name: "tokenManager",
          isMut: true,
          isSigner: true
        },
        {
          name: "allowedToken",
          isMut: true,
          isSigner: false
        },
        {
          name: "managerRole",
          isMut: false,
          isSigner: false
        },
        {
          name: "mintAccount",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetTokenParams"
          }
        }
      ]
    },
    {
      name: "setWithdrawToken",
      accounts: [
        {
          name: "tokenManager",
          isMut: true,
          isSigner: true
        },
        {
          name: "withdrawToken",
          isMut: true,
          isSigner: false
        },
        {
          name: "managerRole",
          isMut: false,
          isSigner: false
        },
        {
          name: "mintAccount",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetWithdrawTokenParams"
          }
        }
      ]
    },
    {
      name: "setOrderDelivery",
      accounts: [
        {
          name: "owner",
          isMut: true,
          isSigner: true
        },
        {
          name: "vaultAuthority",
          isMut: true,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetOrderDeliveryParams"
          }
        }
      ]
    },
    {
      name: "oappQuote",
      accounts: [
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "peer",
          isMut: false,
          isSigner: false
        },
        {
          name: "enforcedOptions",
          isMut: false,
          isSigner: false
        },
        {
          name: "vaultAuthority",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "DepositParams"
          }
        }
      ],
      returns: {
        defined: "MessagingFee"
      }
    },
    {
      name: "lzReceive",
      accounts: [
        {
          name: "payer",
          isMut: true,
          isSigner: true
        },
        {
          name: "peer",
          isMut: false,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "withdrawBrokerPda",
          isMut: false,
          isSigner: false
        },
        {
          name: "withdrawTokenPda",
          isMut: false,
          isSigner: false,
          docs: ["CHECK"]
        },
        {
          name: "tokenMint",
          isMut: false,
          isSigner: false,
          docs: ["CHECK"]
        },
        {
          name: "receiver",
          isMut: true,
          isSigner: false,
          docs: ["CHECK"]
        },
        {
          name: "receiverTokenAccount",
          isMut: true,
          isSigner: false,
          docs: ["UNCHECKED"]
        },
        {
          name: "vaultAuthority",
          isMut: true,
          isSigner: false
        },
        {
          name: "vaultTokenAccount",
          isMut: true,
          isSigner: false
        },
        {
          name: "tokenProgram",
          isMut: false,
          isSigner: false,
          docs: ["CHECKED: sol_vault is used for SOL withdrawal"]
        },
        {
          name: "associatedTokenProgram",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        },
        {
          name: "eventAuthority",
          isMut: false,
          isSigner: false
        },
        {
          name: "program",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "OAppLzReceiveParams"
          }
        }
      ]
    },
    {
      name: "lzReceiveTypes",
      accounts: [
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "accountList",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "OAppLzReceiveParams"
          }
        }
      ],
      returns: {
        vec: {
          defined: "oapp::endpoint_cpi::LzAccount"
        }
      }
    },
    {
      name: "setRateLimit",
      accounts: [
        {
          name: "admin",
          isMut: false,
          isSigner: true
        },
        {
          name: "peer",
          isMut: true,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetRateLimitParams"
          }
        }
      ]
    },
    {
      name: "setDelegate",
      accounts: [
        {
          name: "admin",
          isMut: false,
          isSigner: true
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetDelegateParams"
          }
        }
      ]
    },
    {
      name: "transferAdmin",
      accounts: [
        {
          name: "admin",
          isMut: false,
          isSigner: true
        },
        {
          name: "oappConfig",
          isMut: true,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "TransferAdminParams"
          }
        }
      ]
    },
    {
      name: "setPeer",
      accounts: [
        {
          name: "admin",
          isMut: true,
          isSigner: true
        },
        {
          name: "peer",
          isMut: true,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetPeerParams"
          }
        }
      ]
    },
    {
      name: "setEnforcedOptions",
      accounts: [
        {
          name: "admin",
          isMut: true,
          isSigner: true
        },
        {
          name: "enforcedOptions",
          isMut: true,
          isSigner: false
        },
        {
          name: "oappConfig",
          isMut: false,
          isSigner: false
        },
        {
          name: "systemProgram",
          isMut: false,
          isSigner: false
        }
      ],
      args: [
        {
          name: "params",
          type: {
            defined: "SetEnforcedOptionsParams"
          }
        }
      ]
    }
  ],
  accounts: [
    {
      name: "enforcedOptions",
      type: {
        kind: "struct",
        fields: [
          {
            name: "send",
            type: "bytes"
          },
          {
            name: "sendAndCall",
            type: "bytes"
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    },
    {
      name: "oAppConfig",
      type: {
        kind: "struct",
        fields: [
          {
            name: "endpointProgram",
            type: "publicKey"
          },
          {
            name: "bump",
            type: "u8"
          },
          {
            name: "admin",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "oAppLzReceiveTypesAccounts",
      type: {
        kind: "struct",
        fields: [
          {
            name: "oappConfig",
            type: "publicKey"
          },
          {
            name: "accountList",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "accountList",
      type: {
        kind: "struct",
        fields: [
          {
            name: "bump",
            type: "u8"
          },
          {
            name: "woofiProPda",
            type: "publicKey"
          },
          {
            name: "withdrawUsdcPda",
            type: "publicKey"
          },
          {
            name: "usdcMint",
            type: "publicKey"
          },
          {
            name: "withdrawUsdtPda",
            type: "publicKey"
          },
          {
            name: "usdtMint",
            type: "publicKey"
          },
          {
            name: "withdrawWsolPda",
            type: "publicKey"
          },
          {
            name: "wsolMint",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "peer",
      type: {
        kind: "struct",
        fields: [
          {
            name: "address",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "rateLimiter",
            type: {
              option: {
                defined: "RateLimiter"
              }
            }
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    },
    {
      name: "allowedBroker",
      type: {
        kind: "struct",
        fields: [
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "allowed",
            type: "bool"
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    },
    {
      name: "allowedToken",
      type: {
        kind: "struct",
        fields: [
          {
            name: "mintAccount",
            type: "publicKey"
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenDecimals",
            type: "u8"
          },
          {
            name: "allowed",
            type: "bool"
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    },
    {
      name: "managerRole",
      type: {
        kind: "struct",
        fields: [
          {
            name: "roleHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "allowed",
            type: "bool"
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    },
    {
      name: "vaultAuthority",
      type: {
        kind: "struct",
        fields: [
          {
            name: "bump",
            docs: ["Bump seed for the vault authority PDA"],
            type: "u8"
          },
          {
            name: "owner",
            type: "publicKey"
          },
          {
            name: "depositNonce",
            type: "u64"
          },
          {
            name: "orderDelivery",
            type: "bool"
          },
          {
            name: "inboundNonce",
            type: "u64"
          },
          {
            name: "dstEid",
            type: "u32"
          },
          {
            name: "solChainId",
            type: "u128"
          }
        ]
      }
    },
    {
      name: "withdrawBroker",
      type: {
        kind: "struct",
        fields: [
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerIndex",
            type: "u16"
          },
          {
            name: "allowed",
            type: "bool"
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    },
    {
      name: "withdrawToken",
      type: {
        kind: "struct",
        fields: [
          {
            name: "mintAccount",
            type: "publicKey"
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenDecimals",
            type: "u8"
          },
          {
            name: "tokenIndex",
            type: "u8"
          },
          {
            name: "allowed",
            type: "bool"
          },
          {
            name: "bump",
            type: "u8"
          }
        ]
      }
    }
  ],
  types: [
    {
      name: "VaultDepositParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "accountId",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "userAddress",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "srcChainId",
            type: "u128"
          },
          {
            name: "tokenAmount",
            type: "u128"
          },
          {
            name: "srcChainDepositNonce",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "LzMessage",
      type: {
        kind: "struct",
        fields: [
          {
            name: "msgType",
            type: "u8"
          },
          {
            name: "payload",
            type: "bytes"
          }
        ]
      }
    },
    {
      name: "AccountWithdrawSol",
      type: {
        kind: "struct",
        fields: [
          {
            name: "sender",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "receiver",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerIndex",
            type: "u16"
          },
          {
            name: "tokenIndex",
            type: "u8"
          },
          {
            name: "tokenAmount",
            type: "u64"
          },
          {
            name: "fee",
            type: "u64"
          },
          {
            name: "chainId",
            type: "u64"
          },
          {
            name: "withdrawNonce",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "VaultWithdrawParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "accountId",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "sender",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "receiver",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenAmount",
            type: "u64"
          },
          {
            name: "fee",
            type: "u128"
          },
          {
            name: "chainId",
            type: "u128"
          },
          {
            name: "withdrawNonce",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "InitOAppParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "admin",
            type: "publicKey"
          },
          {
            name: "accountList",
            type: "publicKey"
          },
          {
            name: "endpointProgram",
            type: {
              option: "publicKey"
            }
          }
        ]
      }
    },
    {
      name: "OAppLzReceiveParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "srcEid",
            type: "u32"
          },
          {
            name: "sender",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "nonce",
            type: "u64"
          },
          {
            name: "guid",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "message",
            type: "bytes"
          },
          {
            name: "extraData",
            type: "bytes"
          }
        ]
      }
    },
    {
      name: "MessagingFee",
      type: {
        kind: "struct",
        fields: [
          {
            name: "nativeFee",
            type: "u64"
          },
          {
            name: "lzTokenFee",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "SetAccountListParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "woofiProPda",
            type: "publicKey"
          },
          {
            name: "withdrawUsdcPda",
            type: "publicKey"
          },
          {
            name: "usdcMint",
            type: "publicKey"
          },
          {
            name: "withdrawUsdtPda",
            type: "publicKey"
          },
          {
            name: "usdtMint",
            type: "publicKey"
          },
          {
            name: "withdrawWsolPda",
            type: "publicKey"
          },
          {
            name: "wsolMint",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "SetDelegateParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "delegate",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "SetEnforcedOptionsParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "dstEid",
            type: "u32"
          },
          {
            name: "send",
            type: "bytes"
          },
          {
            name: "sendAndCall",
            type: "bytes"
          }
        ]
      }
    },
    {
      name: "SetPeerParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "dstEid",
            type: "u32"
          },
          {
            name: "peer",
            type: {
              array: ["u8", 32]
            }
          }
        ]
      }
    },
    {
      name: "SetRateLimitParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "dstEid",
            type: "u32"
          },
          {
            name: "refillPerSecond",
            type: {
              option: "u64"
            }
          },
          {
            name: "capacity",
            type: {
              option: "u64"
            }
          },
          {
            name: "enabled",
            type: "bool"
          }
        ]
      }
    },
    {
      name: "TransferAdminParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "admin",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "DepositParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "accountId",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "userAddress",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenAmount",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "OAppSendParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "nativeFee",
            type: "u64"
          },
          {
            name: "lzTokenFee",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "SetBrokerParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "brokerManagerRole",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "allowed",
            type: "bool"
          }
        ]
      }
    },
    {
      name: "SetManagerRoleParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "roleHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "managerAddress",
            type: "publicKey"
          },
          {
            name: "allowed",
            type: "bool"
          }
        ]
      }
    },
    {
      name: "SetOrderDeliveryParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "orderDelivery",
            type: "bool"
          },
          {
            name: "nonce",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "SetTokenParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "tokenManagerRole",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "mintAccount",
            type: "publicKey"
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "allowed",
            type: "bool"
          }
        ]
      }
    },
    {
      name: "SetVaultParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "owner",
            type: "publicKey"
          },
          {
            name: "depositNonce",
            type: "u64"
          },
          {
            name: "orderDelivery",
            type: "bool"
          },
          {
            name: "inboundNonce",
            type: "u64"
          },
          {
            name: "dstEid",
            type: "u32"
          },
          {
            name: "solChainId",
            type: "u128"
          }
        ]
      }
    },
    {
      name: "SetWithdrawBrokerParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "brokerManagerRole",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "brokerIndex",
            type: "u16"
          },
          {
            name: "allowed",
            type: "bool"
          }
        ]
      }
    },
    {
      name: "SetWithdrawTokenParams",
      type: {
        kind: "struct",
        fields: [
          {
            name: "tokenManagerRole",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenHash",
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenIndex",
            type: "u8"
          },
          {
            name: "allowed",
            type: "bool"
          }
        ]
      }
    },
    {
      name: "RateLimiter",
      type: {
        kind: "struct",
        fields: [
          {
            name: "capacity",
            type: "u64"
          },
          {
            name: "tokens",
            type: "u64"
          },
          {
            name: "refillPerSecond",
            type: "u64"
          },
          {
            name: "lastRefillTime",
            type: "u64"
          }
        ]
      }
    },
    {
      name: "OAppError",
      type: {
        kind: "enum",
        variants: [
          {
            name: "Unauthorized"
          },
          {
            name: "InvalidSender"
          },
          {
            name: "InvalidReceiver"
          },
          {
            name: "InvalidOptions"
          },
          {
            name: "InvalidEndpointProgram"
          },
          {
            name: "RateLimitExceeded"
          },
          {
            name: "WithdrawFailed"
          },
          {
            name: "InvalidInboundNonce"
          },
          {
            name: "InvalidReceiverTokenAccount"
          },
          {
            name: "InvalidAdminTokenAccount"
          },
          {
            name: "InvalidMessageType"
          },
          {
            name: "InvalidTokenIndex"
          }
        ]
      }
    },
    {
      name: "MsgType",
      type: {
        kind: "enum",
        variants: [
          {
            name: "Deposit"
          },
          {
            name: "Withdraw"
          },
          {
            name: "RebalanceBurn"
          },
          {
            name: "RebalanceMint"
          }
        ]
      }
    }
  ],
  events: [
    {
      name: "SetAllowedBroker",
      fields: [
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        }
      ]
    },
    {
      name: "ResetAllowedBroker",
      fields: [
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        }
      ]
    },
    {
      name: "SetAllowedToken",
      fields: [
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "mintAccount",
          type: "publicKey",
          index: false
        }
      ]
    },
    {
      name: "ResetAllowedToken",
      fields: [
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "mintAccount",
          type: "publicKey",
          index: false
        }
      ]
    },
    {
      name: "SetWithdrawTokenIndex",
      fields: [
        {
          name: "tokenIndex",
          type: "u8",
          index: false
        },
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "mintAccount",
          type: "publicKey",
          index: false
        }
      ]
    },
    {
      name: "ResetWithdrawTokenIndex",
      fields: [
        {
          name: "tokenIndex",
          type: "u8",
          index: false
        },
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "mintAccount",
          type: "publicKey",
          index: false
        }
      ]
    },
    {
      name: "SetWithdrawBrokerEvent",
      fields: [
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "brokerIndex",
          type: "u16",
          index: false
        }
      ]
    },
    {
      name: "ResetWithdrawBrokerEvent",
      fields: [
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "brokerIndex",
          type: "u16",
          index: false
        }
      ]
    },
    {
      name: "SetManager",
      fields: [
        {
          name: "roleHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "managerAddress",
          type: "publicKey",
          index: false
        },
        {
          name: "allowed",
          type: "bool",
          index: false
        }
      ]
    },
    {
      name: "Deposited",
      fields: [
        {
          name: "user",
          type: "publicKey",
          index: false
        },
        {
          name: "amount",
          type: "u64",
          index: false
        }
      ]
    },
    {
      name: "Withdrawn",
      fields: [
        {
          name: "user",
          type: "publicKey",
          index: false
        },
        {
          name: "amount",
          type: "u64",
          index: false
        }
      ]
    },
    {
      name: "CreatedATA",
      fields: [
        {
          name: "accountId",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "receiver",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "receiverTokenAccount",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "withdrawNonce",
          type: "u64",
          index: false
        }
      ]
    },
    {
      name: "VaultDeposited",
      fields: [
        {
          name: "accountId",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "userAddress",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "srcChainId",
          type: "u128",
          index: false
        },
        {
          name: "tokenAmount",
          type: "u64",
          index: false
        },
        {
          name: "srcChainDepositNonce",
          type: "u64",
          index: false
        }
      ]
    },
    {
      name: "VaultWithdrawn",
      fields: [
        {
          name: "accountId",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "sender",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "receiver",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "tokenAmount",
          type: "u64",
          index: false
        },
        {
          name: "fee",
          type: "u128",
          index: false
        },
        {
          name: "chainId",
          type: "u128",
          index: false
        },
        {
          name: "withdrawNonce",
          type: "u64",
          index: false
        }
      ]
    },
    {
      name: "FrozenWithdrawn",
      fields: [
        {
          name: "accountId",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "sender",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "receiver",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "brokerHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "tokenHash",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "tokenAmount",
          type: "u64",
          index: false
        },
        {
          name: "fee",
          type: "u128",
          index: false
        },
        {
          name: "chainId",
          type: "u128",
          index: false
        },
        {
          name: "withdrawNonce",
          type: "u64",
          index: false
        }
      ]
    },
    {
      name: "OAppSent",
      fields: [
        {
          name: "guid",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "dstEid",
          type: "u32",
          index: false
        }
      ]
    },
    {
      name: "OAppReceived",
      fields: [
        {
          name: "guid",
          type: {
            array: ["u8", 32]
          },
          index: false
        },
        {
          name: "srcEid",
          type: "u32",
          index: false
        }
      ]
    }
  ],
  errors: [
    {
      code: 6e3,
      name: "InsufficientFunds",
      msg: "Deposited funds are insufficient for withdrawal"
    },
    {
      code: 6001,
      name: "UserInfoBelongsToAnotherUser",
      msg: "User info pda belongs to another user"
    },
    {
      code: 6002,
      name: "BrokerNotAllowed",
      msg: "Broker is not allowed"
    },
    {
      code: 6003,
      name: "TokenNotAllowed",
      msg: "Token is not allowed"
    },
    {
      code: 6004,
      name: "InvalidAccountId",
      msg: "AccountId is invalid"
    },
    {
      code: 6005,
      name: "InvalidVaultOwner",
      msg: "Vault owner is not the same as the payer"
    },
    {
      code: 6006,
      name: "ManagerRoleNotAllowed",
      msg: "Manager role is not allowed"
    }
  ]
};

// src/solana.util.ts
var import_lz_solana_sdk_v2 = require("@layerzerolabs/lz-solana-sdk-v2");
var import_spl_token = require("@solana/spl-token");
var import_web32 = require("@solana/web3.js");
var getTokenAccounts = (token, owner) => {
  const tokenAccount = (0, import_spl_token.getAssociatedTokenAddressSync)(token, owner, true);
  return tokenAccount;
};
function getVaultAuthorityPda(VAULT_PROGRAM_ID) {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(VAULT_AUTHORITY_SEED, "utf8")],
    VAULT_PROGRAM_ID
  )[0];
}
function getSolVaultPda(VAULT_PROGRAM_ID) {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(SOL_VAULT_SEED, "utf8")],
    VAULT_PROGRAM_ID
  )[0];
}
function getBrokerPDA(programId, brokerHash) {
  const hash = Array.from(Buffer.from(brokerHash.slice(2), "hex"));
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(BROKER_SEED, "utf8"), Buffer.from(hash)],
    programId
  )[0];
}
function getTokenPDA(programId, tokenHash) {
  const hash = Array.from(Buffer.from(tokenHash.slice(2), "hex"));
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(TOKEN_SEED, "utf8"), Buffer.from(hash)],
    programId
  )[0];
}
function getOAppConfigPda(programId) {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.OAPP_SEED, "utf8")],
    programId
  )[0];
}
function getPeerPda(OAPP_PROGRAM_ID, oappConfigPda, dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.PEER_SEED, "utf8"), oappConfigPda.toBuffer(), bufferDstEid],
    OAPP_PROGRAM_ID
  )[0];
}
function getEndorcedOptionsPda(OAPP_PROGRAM_ID, oappConfigPda, dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [
      Buffer.from(import_lz_solana_sdk_v2.ENFORCED_OPTIONS_SEED, "utf8"),
      oappConfigPda.toBuffer(),
      bufferDstEid
    ],
    OAPP_PROGRAM_ID
  )[0];
}
function getSendLibPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.MESSAGE_LIB_SEED, "utf8")],
    SEND_LIB_PROGRAM_ID
  )[0];
}
function getSendLibConfigPda(oappConfigPda, dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [
      Buffer.from(import_lz_solana_sdk_v2.SEND_LIBRARY_CONFIG_SEED, "utf8"),
      oappConfigPda.toBuffer(),
      bufferDstEid
    ],
    ENDPOINT_PROGRAM_ID
  )[0];
}
function getDefaultSendLibConfigPda(dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.SEND_LIBRARY_CONFIG_SEED, "utf8"), bufferDstEid],
    ENDPOINT_PROGRAM_ID
  )[0];
}
function getSendLibInfoPda(sendLibPda) {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.MESSAGE_LIB_SEED, "utf8"), sendLibPda.toBuffer()],
    ENDPOINT_PROGRAM_ID
  )[0];
}
function getEndpointSettingPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.ENDPOINT_SEED, "utf8")],
    ENDPOINT_PROGRAM_ID
  )[0];
}
function getPeerAddress(OAPP_PROGRAM_ID) {
  let peer_address = DEV_PEER_ADDRESS;
  if (OAPP_PROGRAM_ID.toBase58() === DEV_OAPP_PROGRAM_ID.toBase58()) {
    peer_address = DEV_PEER_ADDRESS;
  }
  if (OAPP_PROGRAM_ID.toBase58() === QA_OAPP_PROGRAM_ID.toBase58()) {
    peer_address = QA_PEER_ADDRESS;
  }
  if (OAPP_PROGRAM_ID.toBase58() === STAGING_OAPP_PROGRAM_ID.toBase58()) {
    peer_address = STAGING_PEER_ADDRESS;
  }
  if (OAPP_PROGRAM_ID.toBase58() === MAINNET_OAPP_PROGRAM_ID.toBase58()) {
    peer_address = MAINNET_PEER_ADDRESS;
  }
  return peer_address;
}
function getNoncePda(OAPP_PROGRAM_ID, oappConfigPda, dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  const peer_address = getPeerAddress(OAPP_PROGRAM_ID);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [
      Buffer.from(import_lz_solana_sdk_v2.NONCE_SEED, "utf8"),
      oappConfigPda.toBuffer(),
      bufferDstEid,
      peer_address
    ],
    ENDPOINT_PROGRAM_ID
  )[0];
}
function getEventAuthorityPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.EVENT_SEED, "utf8")],
    ENDPOINT_PROGRAM_ID
  )[0];
}
function getUlnSettingPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.ULN_SEED, "utf8")],
    SEND_LIB_PROGRAM_ID
  )[0];
}
function getSendConfigPda(oappConfigPda, dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [
      Buffer.from(import_lz_solana_sdk_v2.SEND_CONFIG_SEED, "utf8"),
      bufferDstEid,
      oappConfigPda.toBuffer()
    ],
    SEND_LIB_PROGRAM_ID
  )[0];
}
function getDefaultSendConfigPda(dstEid) {
  const bufferDstEid = Buffer.alloc(4);
  bufferDstEid.writeUInt32BE(dstEid);
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.SEND_CONFIG_SEED, "utf8"), bufferDstEid],
    RECEIVE_LIB_PROGRAM_ID
  )[0];
}
function getUlnEventAuthorityPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.EVENT_SEED, "utf8")],
    SEND_LIB_PROGRAM_ID
  )[0];
}
function getExecutorConfigPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.EXECUTOR_CONFIG_SEED, "utf8")],
    EXECUTOR_PROGRAM_ID
  )[0];
}
function getPriceFeedPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.PRICE_FEED_SEED, "utf8")],
    PRICE_FEED_PROGRAM_ID
  )[0];
}
function getDvnConfigPda() {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.DVN_CONFIG_SEED, "utf8")],
    DVN_PROGRAM_ID
  )[0];
}
function getDstEID(OAPP_PROGRAM_ID) {
  if (OAPP_PROGRAM_ID.toBase58() === MAINNET_OAPP_PROGRAM_ID.toBase58()) {
    return MAIN_DST_EID;
  }
  return DEV_DST_EID;
}
function getLookupTableAddress(OAPP_PROGRAM_ID) {
  if (OAPP_PROGRAM_ID.toBase58() === DEV_OAPP_PROGRAM_ID.toBase58()) {
    console.log(
      "DEV_LOOKUP_TABLE_ADDRESS: ",
      DEV_LOOKUP_TABLE_ADDRESS.toBase58()
    );
    return DEV_LOOKUP_TABLE_ADDRESS;
  }
  if (OAPP_PROGRAM_ID.toBase58() === QA_OAPP_PROGRAM_ID.toBase58()) {
    return QA_LOOKUP_TABLE_ADDRESS;
  }
  if (OAPP_PROGRAM_ID.toBase58() === STAGING_OAPP_PROGRAM_ID.toBase58()) {
    return STAGING_LOOKUP_TABLE_ADDRESS;
  }
  if (OAPP_PROGRAM_ID.toBase58() === MAINNET_OAPP_PROGRAM_ID.toBase58()) {
    return MAINNET_LOOKUP_TABLE_ADDRESS;
  }
  return DEV_LOOKUP_TABLE_ADDRESS;
}
function getMessageLibPda(programId) {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.MESSAGE_LIB_SEED, "utf8")],
    programId ? programId : SEND_LIB_PROGRAM_ID
  )[0];
}
function getMessageLibInfoPda(msgLibPda, programId) {
  return import_web32.PublicKey.findProgramAddressSync(
    [Buffer.from(import_lz_solana_sdk_v2.MESSAGE_LIB_SEED, "utf8"), msgLibPda.toBytes()],
    programId ? programId : ENDPOINT_PROGRAM_ID
  )[0];
}
function getLookupTableAccount(connection, lookupTableAddress) {
  return __async(this, null, function* () {
    const lookupTableAccount = (yield connection.getAddressLookupTable(lookupTableAddress)).value;
    return lookupTableAccount;
  });
}

// src/helper.ts
function addOrderlyKeyMessage(inputs) {
  const {
    publicKey,
    brokerId,
    expiration = 365,
    timestamp = Date.now(),
    scope,
    tag,
    chainId,
    subAccountId
  } = inputs;
  const message = __spreadValues(__spreadValues({
    brokerId,
    chainType: "SOL",
    orderlyKey: publicKey,
    scope: scope || import_types.DEFAUL_ORDERLY_KEY_SCOPE,
    chainId,
    timestamp,
    expiration: timestamp + 1e3 * 60 * 60 * 24 * expiration
  }, typeof tag !== "undefined" ? { tag } : {}), typeof subAccountId !== "undefined" ? { subAccountId } : {});
  const brokerIdHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [message.brokerId]);
  const orderlyKeyHash = (0, import_ethers.solidityPackedKeccak256)(
    ["string"],
    [message.orderlyKey]
  );
  const scopeHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [message.scope]);
  const abicoder = import_ethers.AbiCoder.defaultAbiCoder();
  const msgToSign = (0, import_keccak.keccak256)(
    (0, import_utils.hexToBytes)(
      abicoder.encode(
        ["bytes32", "bytes32", "bytes32", "uint256", "uint256", "uint256"],
        [
          brokerIdHash,
          orderlyKeyHash,
          scopeHash,
          message.chainId,
          message.timestamp,
          message.expiration
        ]
      )
    )
  );
  const msgToSignHex = (0, import_utils.bytesToHex)(msgToSign);
  const msgToSignTextEncoded = new TextEncoder().encode(
    msgToSignHex
  );
  return [message, msgToSignTextEncoded];
}
function registerAccountMessage(inputs) {
  const { chainId, registrationNonce, brokerId, timestamp } = inputs;
  const message = {
    brokerId,
    chainId,
    timestamp,
    registrationNonce
  };
  const brokerIdHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [message.brokerId]);
  const abicoder = import_ethers.AbiCoder.defaultAbiCoder();
  const msgToSign = (0, import_keccak.keccak256)(
    (0, import_utils.hexToBytes)(
      abicoder.encode(
        ["bytes32", "uint256", "uint256", "uint256"],
        [
          brokerIdHash,
          message.chainId,
          message.timestamp,
          message.registrationNonce
        ]
      )
    )
  );
  const msgToSignHex = (0, import_utils.bytesToHex)(msgToSign);
  const msgToSignTextEncoded = new TextEncoder().encode(
    msgToSignHex
  );
  return [message, msgToSignTextEncoded];
}
function internalTransferMessage(inputs) {
  const { chainId, receiver, token, amount, nonce } = inputs;
  const message = {
    chainId,
    receiver,
    token,
    amount,
    transferNonce: nonce,
    chainType: "SOL"
  };
  const tokenSymbolHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [message.token]);
  const abicoder = import_ethers.AbiCoder.defaultAbiCoder();
  const msgToSign = (0, import_keccak.keccak256)(
    (0, import_utils.hexToBytes)(
      abicoder.encode(
        ["bytes32", "bytes32", "uint256", "uint64", "uint256"],
        [
          message.receiver,
          tokenSymbolHash,
          message.amount,
          message.transferNonce,
          chainId
        ]
      )
    )
  );
  const msgToSignHex = (0, import_utils.bytesToHex)(msgToSign);
  const msgToSignTextEncoded = new TextEncoder().encode(
    msgToSignHex
  );
  return [message, msgToSignTextEncoded];
}
function dexRequestMessage(inputs) {
  return __async(this, null, function* () {
    const {
      payloadType,
      nonce,
      receiver,
      amount,
      vaultId,
      token,
      dexBrokerId,
      chainId
    } = inputs;
    const message = {
      payloadType,
      nonce,
      receiver,
      amount,
      vaultId,
      token,
      dexBrokerId,
      chainId
    };
    let receiverBytes;
    receiverBytes = (0, import_bs58.decode)(receiver);
    const receiverBytes32 = new Uint8Array(32);
    receiverBytes32.set(receiverBytes);
    const vaultIdHex = vaultId;
    const vaultIdBytes = (0, import_utils.hexToBytes)(vaultIdHex);
    const tokenHash = (0, import_keccak.keccak256)(new TextEncoder().encode(token));
    const dexBrokerIdHash = (0, import_keccak.keccak256)(new TextEncoder().encode(dexBrokerId));
    const abicoder = import_ethers.AbiCoder.defaultAbiCoder();
    const msgToSign = (0, import_keccak.keccak256)(
      (0, import_utils.hexToBytes)(
        abicoder.encode(
          [
            "uint8",
            // payloadType
            "uint256",
            // nonce
            "bytes32",
            // receiver (Base58 decoded)
            "uint256",
            // amount
            "bytes32",
            // vaultId
            "bytes32",
            // token hash
            "bytes32",
            // dexBrokerId hash
            "uint256"
            // chainId
          ],
          [
            payloadType,
            nonce,
            receiverBytes,
            amount,
            vaultIdBytes,
            tokenHash,
            dexBrokerIdHash,
            chainId
          ]
        )
      )
    );
    const msgToSignHex = (0, import_utils.bytesToHex)(msgToSign);
    const msgToSignTextEncoded = new TextEncoder().encode(
      msgToSignHex
    );
    return [message, msgToSignTextEncoded];
  });
}
function withdrawMessage(inputs) {
  const { chainId, receiver, token, amount, nonce, brokerId } = inputs;
  const timestamp = Date.now();
  const message = {
    brokerId,
    chainId,
    receiver,
    token,
    amount,
    withdrawNonce: nonce,
    timestamp,
    chainType: "SOL"
  };
  const brokerIdHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [message.brokerId]);
  const tokenSymbolHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [message.token]);
  const salt = (0, import_keccak.keccak256)(Buffer.from("Orderly Network"));
  const abicoder = import_ethers.AbiCoder.defaultAbiCoder();
  const msgToSign = (0, import_keccak.keccak256)(
    (0, import_utils.hexToBytes)(
      abicoder.encode(
        [
          "bytes32",
          "bytes32",
          "uint256",
          "bytes32",
          "uint256",
          "uint64",
          "uint64",
          "bytes32"
        ],
        [
          brokerIdHash,
          tokenSymbolHash,
          chainId,
          (0, import_bs58.decode)(message.receiver),
          message.amount,
          message.withdrawNonce,
          timestamp,
          salt
        ]
      )
    )
  );
  const msgToSignHex = (0, import_utils.bytesToHex)(msgToSign);
  const msgToSignTextEncoded = new TextEncoder().encode(
    msgToSignHex
  );
  return [message, msgToSignTextEncoded];
}
function settleMessage(inputs) {
  const { settlePnlNonce, brokerId, chainId, timestamp } = inputs;
  const message = {
    brokerId,
    chainId,
    timestamp,
    chainType: "SOL",
    settleNonce: settlePnlNonce
  };
  const brokerIdHash = (0, import_ethers.solidityPackedKeccak256)(["string"], [brokerId]);
  const abicoder = import_ethers.AbiCoder.defaultAbiCoder();
  const msgToSign = (0, import_keccak.keccak256)(
    (0, import_utils.hexToBytes)(
      abicoder.encode(
        ["bytes32", "uint256", "uint64", "uint64"],
        [brokerIdHash, message.chainId, message.settleNonce, message.timestamp]
      )
    )
  );
  const msgToSignHex = (0, import_utils.bytesToHex)(msgToSign);
  const msgToSignTextEncoded = new TextEncoder().encode(
    msgToSignHex
  );
  return [message, msgToSignTextEncoded];
}
function getDepositQuoteFee(_0) {
  return __async(this, arguments, function* ({
    vaultAddress,
    userAddress,
    connection,
    depositData
  }) {
    console.log("-- vaultAddress", vaultAddress);
    const appProgramId = new import_web33.PublicKey(vaultAddress);
    const DST_EID = getDstEID(appProgramId);
    const program = new import_anchor.Program(IDL, appProgramId, {
      connection
    });
    const userPublicKey = new import_web33.PublicKey(userAddress);
    const oappConfigPDA = getOAppConfigPda(appProgramId);
    const peerPDA = getPeerPda(appProgramId, oappConfigPDA, DST_EID);
    const endorcedPDA = getEndorcedOptionsPda(
      appProgramId,
      oappConfigPDA,
      DST_EID
    );
    const sendLibConfigPDA = getSendLibConfigPda(oappConfigPDA, DST_EID);
    const defaultSendLibPDA = getDefaultSendLibConfigPda(DST_EID);
    const endpointSettingPDA = getEndpointSettingPda();
    const noncePDA = getNoncePda(appProgramId, oappConfigPDA, DST_EID);
    const sendConfigPDA = getSendConfigPda(oappConfigPDA, DST_EID);
    const defaultSendConfigPDA = getDefaultSendConfigPda(DST_EID);
    const executorConfigPDA = getExecutorConfigPda();
    const priceFeedPDA = getPriceFeedPda();
    const dvnConfigPDA = getDvnConfigPda();
    const messageLibPDA = getMessageLibPda(SEND_LIB_PROGRAM_ID);
    const messageLibInfoPDA = getMessageLibInfoPda(messageLibPDA);
    const vaultAuthorityPDA = getVaultAuthorityPda(appProgramId);
    const depositParams = getDepositParams(userAddress, depositData);
    const quoteFee = yield program.methods.oappQuote(depositParams).accounts({
      oappConfig: oappConfigPDA,
      peer: peerPDA,
      enforcedOptions: endorcedPDA,
      vaultAuthority: vaultAuthorityPDA
    }).remainingAccounts([
      // ENDPOINT solana/programs/programs/uln/src/instructions/endpoint/send.rs
      {
        pubkey: ENDPOINT_PROGRAM_ID,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: SEND_LIB_PROGRAM_ID,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: sendLibConfigPDA,
        // send_library_config
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: defaultSendLibPDA,
        // default_send_library_config
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: messageLibInfoPDA,
        // send_library_info
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: endpointSettingPDA,
        // endpoint settings
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: noncePDA,
        // nonce
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: messageLibPDA,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: sendConfigPDA,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: defaultSendConfigPDA,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: EXECUTOR_PROGRAM_ID,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: executorConfigPDA,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: PRICE_FEED_PROGRAM_ID,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: priceFeedPDA,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: DVN_PROGRAM_ID,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: dvnConfigPDA,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: PRICE_FEED_PROGRAM_ID,
        isWritable: false,
        isSigner: false
      },
      {
        pubkey: priceFeedPDA,
        isWritable: false,
        isSigner: false
      }
    ]).instruction();
    const lastBlockHash = yield connection.getLatestBlockhash();
    const lookupTableAddress = getLookupTableAddress(appProgramId);
    const lookupTableAccount = yield getLookupTableAccount(
      connection,
      lookupTableAddress
    );
    if (!lookupTableAccount) {
      console.log("-- lookup table account error");
      throw new Error("-- lookup table account error");
    }
    const feeMsg = new import_web33.TransactionMessage({
      payerKey: userPublicKey,
      recentBlockhash: lastBlockHash.blockhash,
      instructions: [quoteFee]
    }).compileToV0Message([lookupTableAccount]);
    const feeTx = new import_web33.VersionedTransaction(feeMsg);
    const feeRes = yield connection.simulateTransaction(feeTx);
    console.log("-- feeRes", feeRes);
    if (feeRes.value.err) {
      const errorInfo = typeof feeRes.value.err === "object" ? JSON.stringify(feeRes.value.err) : feeRes.value.err;
      if (errorInfo.toString().includes("AccountNotFound")) {
        throw new Error("Error: Account gas is insufficient.");
      }
      throw new Error(`Error: ${errorInfo}`);
    }
    const returnPrefix = `Program return: ${program.programId} `;
    const returnLogEntry = feeRes.value.logs.find(
      (log) => log.startsWith(returnPrefix)
    );
    if (!returnLogEntry) {
      throw new Error("Error: get deposit fee error");
    }
    const encodedReturnData = returnLogEntry.slice(returnPrefix.length);
    const decodedBuffer = Buffer.from(encodedReturnData, "base64");
    return decodedBuffer.readBigUInt64LE(0);
  });
}
var getDepositParams = (userAddress, depositData) => {
  const brokerHash = depositData.brokerHash;
  const codedBrokerHash = Array.from(Buffer.from(brokerHash.slice(2), "hex"));
  const tokenHash = depositData.tokenHash;
  const codedTokenHash = Array.from(Buffer.from(tokenHash.slice(2), "hex"));
  const solAccountId = depositData.accountId;
  const codedAccountId = Array.from(Buffer.from(solAccountId.slice(2), "hex"));
  const userPublicKey = new import_web33.PublicKey(userAddress);
  return {
    accountId: codedAccountId,
    brokerHash: codedBrokerHash,
    tokenHash: codedTokenHash,
    userAddress: Array.from(userPublicKey.toBuffer()),
    tokenAmount: new import_anchor.BN(depositData.tokenAmount)
  };
};
function deposit(_0) {
  return __async(this, arguments, function* ({
    vaultAddress,
    userAddress,
    connection,
    sendTransaction,
    depositData
  }) {
    const brokerHash = depositData.brokerHash;
    const tokenHash = depositData.tokenHash;
    const SOL_HASH = import_core.utils.parseTokenHash("SOL");
    const isSolDeposit = tokenHash.toLowerCase() === SOL_HASH.toLowerCase();
    console.log("-- vault address", vaultAddress);
    const appProgramId = new import_web33.PublicKey(vaultAddress);
    const program = new import_anchor.Program(IDL, appProgramId, {
      connection
    });
    const token = new import_web33.PublicKey(depositData.tokenAddress);
    const userPublicKey = new import_web33.PublicKey(userAddress);
    const userTokenAccount = getTokenAccounts(token, userPublicKey);
    const vaultAuthorityPda = getVaultAuthorityPda(appProgramId);
    const vaultTokenAccount = getTokenAccounts(token, vaultAuthorityPda);
    const allowedBrokerPDA = getBrokerPDA(appProgramId, brokerHash);
    const allowedTokenPDA = getTokenPDA(appProgramId, tokenHash);
    const oappConfigPDA = getOAppConfigPda(appProgramId);
    console.log("-- oappconfig pda", oappConfigPDA.toBase58());
    const DST_EID = getDstEID(appProgramId);
    const peerPDA = getPeerPda(appProgramId, oappConfigPDA, DST_EID);
    const endorcedPDA = getEndorcedOptionsPda(
      appProgramId,
      oappConfigPDA,
      DST_EID
    );
    const sendLibPDA = getSendLibPda();
    const sendLibConfigPDA = getSendLibConfigPda(oappConfigPDA, DST_EID);
    const defaultSendLibPDA = getDefaultSendLibConfigPda(DST_EID);
    const sendLibInfoPDA = getSendLibInfoPda(sendLibPDA);
    const endpointSettingPDA = getEndpointSettingPda();
    const noncePDA = getNoncePda(appProgramId, oappConfigPDA, DST_EID);
    const eventAuthorityPDA = getEventAuthorityPda();
    const ulnSettingPDA = getUlnSettingPda();
    const sendConfigPDA = getSendConfigPda(oappConfigPDA, DST_EID);
    const defaultSendConfigPDA = getDefaultSendConfigPda(DST_EID);
    const ulnEventAuthorityPDA = getUlnEventAuthorityPda();
    const executorConfigPDA = getExecutorConfigPda();
    const priceFeedPDA = getPriceFeedPda();
    const dvnConfigPDA = getDvnConfigPda();
    const vaultDepositParams = getDepositParams(userAddress, depositData);
    const buildSendRemainingAccounts = () => [
      // ENDPOINT solana/programs/programs/uln/src/instructions/endpoint/send.rs
      {
        isSigner: false,
        isWritable: false,
        pubkey: ENDPOINT_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        // 0
        pubkey: oappConfigPDA
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: SEND_LIB_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        // 7
        pubkey: sendLibConfigPDA
      },
      {
        isSigner: false,
        isWritable: false,
        // 9
        pubkey: defaultSendLibPDA
      },
      {
        isSigner: false,
        isWritable: false,
        // 8
        pubkey: sendLibInfoPDA
      },
      {
        isSigner: false,
        isWritable: false,
        // 14
        pubkey: endpointSettingPDA
      },
      {
        isSigner: false,
        isWritable: true,
        // 15
        pubkey: noncePDA
      },
      {
        isSigner: false,
        isWritable: false,
        // 3
        pubkey: eventAuthorityPDA
      },
      // ULN solana/programs/programs/uln/src/instructions/endpoint/send.rs
      {
        isSigner: false,
        isWritable: false,
        pubkey: ENDPOINT_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        // 13
        pubkey: ulnSettingPDA
      },
      {
        isSigner: false,
        isWritable: false,
        // 10
        pubkey: sendConfigPDA
      },
      {
        isSigner: false,
        isWritable: false,
        // 11
        pubkey: defaultSendConfigPDA
      },
      {
        isSigner: true,
        isWritable: false,
        pubkey: userPublicKey
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: TREASURY_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: import_web33.SystemProgram.programId
      },
      {
        isSigner: false,
        isWritable: false,
        // 12
        pubkey: ulnEventAuthorityPDA
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: SEND_LIB_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: EXECUTOR_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: true,
        // 16
        pubkey: executorConfigPDA
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: PRICE_FEED_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        // 17
        pubkey: priceFeedPDA
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: DVN_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: true,
        // 18
        pubkey: dvnConfigPDA
      },
      {
        isSigner: false,
        isWritable: false,
        pubkey: PRICE_FEED_PROGRAM_ID
      },
      {
        isSigner: false,
        isWritable: false,
        // 17
        pubkey: priceFeedPDA
      }
    ];
    const fee = yield getDepositQuoteFee({
      vaultAddress,
      userAddress,
      connection,
      depositData
    });
    const sendParam = {
      nativeFee: new import_anchor.BN(fee.toString()),
      lzTokenFee: new import_anchor.BN(0)
    };
    console.log("--- value params", {
      vaultDepositParams,
      sendParam
    });
    const ixDepositEntry = isSolDeposit ? yield program.methods.depositSol(vaultDepositParams, sendParam).accounts({
      solVault: getSolVaultPda(appProgramId),
      vaultAuthority: vaultAuthorityPda,
      user: userPublicKey,
      peer: peerPDA,
      enforcedOptions: endorcedPDA,
      oappConfig: oappConfigPDA,
      allowedBroker: allowedBrokerPDA,
      allowedToken: allowedTokenPDA
    }).remainingAccounts(buildSendRemainingAccounts()).instruction() : yield program.methods.deposit(vaultDepositParams, sendParam).accounts({
      userTokenAccount,
      vaultAuthority: vaultAuthorityPda,
      vaultTokenAccount,
      depositToken: token,
      user: userPublicKey,
      peer: peerPDA,
      enforcedOptions: endorcedPDA,
      oappConfig: oappConfigPDA,
      allowedBroker: allowedBrokerPDA,
      allowedToken: allowedTokenPDA
    }).remainingAccounts(buildSendRemainingAccounts()).instruction();
    const lookupTableAddress = getLookupTableAddress(appProgramId);
    const lookupTableAccount = yield getLookupTableAccount(
      connection,
      lookupTableAddress
    );
    if (!lookupTableAccount) {
      console.log("-- lookup table account error");
      return;
    }
    const ixAddComputeBudget = import_web33.ComputeBudgetProgram.setComputeUnitLimit({
      units: 4e5
    });
    const lastBlockHash = yield connection.getLatestBlockhash();
    const msg = new import_web33.TransactionMessage({
      payerKey: userPublicKey,
      recentBlockhash: lastBlockHash.blockhash,
      instructions: [ixDepositEntry, ixAddComputeBudget]
    }).compileToV0Message([lookupTableAccount]);
    const tx = new import_web33.VersionedTransaction(msg);
    const res = yield sendTransaction(tx, connection);
    console.log("res", res);
    return res;
  });
}
function checkIsLedgerWallet(userAddress) {
  const info = window.localStorage.getItem(import_types.LedgerWalletKey);
  if (!info) {
    return false;
  }
  const addressArr = JSON.parse(info != null ? info : "[]");
  console.log("-- addressArr", addressArr);
  if (addressArr.includes(userAddress)) {
    return true;
  }
  return false;
}

// src/walletAdapter.ts
var DefaultSolanaWalletAdapter = class extends import_core2.BaseWalletAdapter {
  constructor() {
    super();
    __publicField(this, "chainNamespace", import_types2.ChainNamespace.solana);
    __publicField(this, "_address");
    __publicField(this, "_chainId");
    __publicField(this, "_provider");
    __publicField(this, "_connection");
  }
  get address() {
    return this._address;
  }
  get chainId() {
    return this._chainId;
  }
  set chainId(chainId) {
    this._chainId = chainId;
  }
  get connection() {
    if (this._connection) {
      return this._connection;
    }
    if (this._provider.rpcUrl) {
      this._connection = new import_web34.Connection(this._provider.rpcUrl, {
        commitment: "confirmed"
      });
      return this._connection;
    }
    if (this._provider.network === import_wallet_adapter_base.WalletAdapterNetwork.Devnet) {
      this._connection = new import_web34.Connection((0, import_web34.clusterApiUrl)(this._provider.network), {
        commitment: "confirmed"
      });
      return this._connection;
    }
    const account = import_core2.SimpleDI.get("account");
    const url = "/v1/solana-rpc-proxy";
    this._connection = new import_web34.Connection(`${account.apiBaseUrl}${url}`, {
      commitment: "confirmed",
      fetchMiddleware: (info, init, fetch) => __async(this, null, function* () {
        const payload = {
          url,
          method: init == null ? void 0 : init.method,
          data: JSON.parse(init == null ? void 0 : init.body)
        };
        const signature = yield this.signMessageByOrderlyKey(payload);
        for (const key of Object.keys(signature)) {
          (init == null ? void 0 : init.headers)[key] = signature[key];
        }
        return fetch(info, init);
      })
    });
    return this._connection;
  }
  setConfig(config) {
    this._address = config.address;
    this._chainId = config.chain.id;
    if (config.provider) {
      this._provider = config.provider;
    }
  }
  lifecycleName(name, data) {
    console.log("lifecycle", name, data);
  }
  active(config) {
    this.setConfig(config);
    this.lifecycleName("active", config);
  }
  deactivate() {
    this.lifecycleName("deactivate", {});
  }
  update(config) {
    this.lifecycleName("update", config);
    this.setConfig(config);
  }
  generateSecretKey() {
    let privKey, secretKey;
    do {
      privKey = ed.utils.randomPrivateKey();
      secretKey = (0, import_bs582.encode)(privKey);
    } while (secretKey.length !== 44);
    return secretKey;
  }
  uint8ArrayToHexString(uint8Array) {
    return Array.from(uint8Array).map((byte) => byte.toString(16).padStart(2, "0")).join("");
  }
  signMessage(message) {
    return __async(this, null, function* () {
      const isLedger = checkIsLedgerWallet(this._address);
      if (isLedger) {
        const transaction = new import_web34.Transaction();
        transaction.add(
          new import_web34.TransactionInstruction({
            keys: [],
            programId: new import_web34.PublicKey(
              "ComputeBudget111111111111111111111111111111"
            ),
            data: new Uint8Array([3, 0, 0, 0, 0, 0, 0, 0, 0])
          })
        );
        transaction.add(
          new import_web34.TransactionInstruction({
            keys: [],
            programId: new import_web34.PublicKey(
              "ComputeBudget111111111111111111111111111111"
            ),
            data: new Uint8Array([2, 0, 0, 0, 0])
          })
        );
        transaction.add(
          new import_web34.TransactionInstruction({
            keys: [],
            programId: new import_web34.PublicKey(
              "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr"
            ),
            data: message
          })
        );
        const userPublicKey = new import_web34.PublicKey(this.address);
        transaction.feePayer = userPublicKey;
        const zeroHash = new Uint8Array(32).fill(0);
        transaction.recentBlockhash = new import_web34.PublicKey(zeroHash).toString();
        const signedTransaction = yield this._provider.signTransaction(transaction);
        const signature = signedTransaction.signatures[0].signature;
        if (signature) {
          return this.uint8ArrayToHexString(signature);
        } else {
          console.log("-- sign message error", signature);
          throw new Error("Unsupported signature");
        }
      }
      const signRes = yield this._provider.signMessage(message);
      return "0x" + (0, import_utils2.bytesToHex)(signRes);
    });
  }
  generateRegisterAccountMessage(inputs) {
    return __async(this, null, function* () {
      const [message, toSignatureMessage] = registerAccountMessage(__spreadProps(__spreadValues({}, inputs), {
        chainId: this.chainId
      }));
      const signature = yield this.signMessage(toSignatureMessage);
      return {
        message: __spreadProps(__spreadValues({}, message), {
          chainType: "SOL"
        }),
        signatured: signature
      };
    });
  }
  generateWithdrawMessage(inputs) {
    return __async(this, null, function* () {
      const [message, toSignatureMessage] = withdrawMessage(__spreadProps(__spreadValues({}, inputs), {
        chainId: this.chainId
      }));
      const signature = yield this.signMessage(toSignatureMessage);
      return {
        message: __spreadProps(__spreadValues({}, message), {
          chainType: "SOL"
        }),
        domain: {
          name: "",
          version: "",
          chainId: this.chainId,
          verifyingContract: inputs.verifyContract
        },
        signatured: signature
      };
    });
  }
  generateInternalTransferMessage(inputs) {
    return __async(this, null, function* () {
      const [message, toSignatureMessage] = internalTransferMessage(__spreadProps(__spreadValues({}, inputs), {
        chainId: this.chainId
      }));
      const signature = yield this.signMessage(toSignatureMessage);
      return {
        message: __spreadProps(__spreadValues({}, message), {
          chainType: "SOL"
        }),
        domain: {
          name: "",
          version: "",
          chainId: this.chainId,
          verifyingContract: inputs.verifyContract
        },
        signatured: signature
      };
    });
  }
  generateAddOrderlyKeyMessage(inputs) {
    return __async(this, null, function* () {
      const [message, toSignatureMessage] = addOrderlyKeyMessage(__spreadProps(__spreadValues({}, inputs), {
        chainId: this.chainId
      }));
      const signature = yield this.signMessage(toSignatureMessage);
      return {
        message: __spreadProps(__spreadValues({}, message), {
          chainType: "SOL"
        }),
        signatured: signature
      };
    });
  }
  generateSettleMessage(inputs) {
    return __async(this, null, function* () {
      const [message, toSignatureMessage] = settleMessage(__spreadProps(__spreadValues({}, inputs), {
        chainId: this.chainId
      }));
      const signature = yield this.signMessage(toSignatureMessage);
      return {
        message: __spreadProps(__spreadValues({}, message), {
          chainType: "SOL"
        }),
        domain: {
          name: "",
          version: "",
          chainId: this.chainId,
          verifyingContract: inputs.verifyContract
        },
        signatured: signature
      };
    });
  }
  generateDexRequestMessage(inputs) {
    return __async(this, null, function* () {
      const [message, toSignatureMessage] = yield dexRequestMessage(__spreadProps(__spreadValues({}, inputs), {
        chainId: this.chainId
      }));
      const signature = yield this.signMessage(toSignatureMessage);
      return {
        message: __spreadProps(__spreadValues({}, inputs), {
          chainType: "SOL"
        }),
        signatured: signature,
        domain: inputs.domain
      };
    });
  }
  getBalance() {
    return __async(this, null, function* () {
      const publicKey = new import_web34.PublicKey(this.address);
      const lamports = yield this.connection.getBalance(publicKey);
      return BigInt(lamports);
    });
  }
  getBalances(addresses) {
    return __async(this, null, function* () {
      const userPublicKey = new import_web34.PublicKey(this._address);
      const connection = this.connection;
      const isNativeToken = addresses.map(
        (address) => (0, import_types2.isNativeTokenChecker)(address)
      );
      const hasNativeSOL = addresses.some(
        (address) => (0, import_types2.isNativeTokenChecker)(address)
      );
      let nativeSolBalance = null;
      if (hasNativeSOL) {
        nativeSolBalance = yield this.getBalance();
      }
      const splTokenAddresses = addresses.filter(
        (address) => !(0, import_types2.isNativeTokenChecker)(address)
      );
      let splTokenBalances = [];
      if (splTokenAddresses.length > 0) {
        const tokenPublicKeys = splTokenAddresses.map(
          (address) => new import_web34.PublicKey(address)
        );
        const userTokenAccounts = tokenPublicKeys.map(
          (tokenPublicKey) => getTokenAccounts(tokenPublicKey, userPublicKey)
        );
        const tokenAmount = yield (0, import_spl_token2.getMultipleAccounts)(
          connection,
          userTokenAccounts,
          "confirmed"
        );
        splTokenBalances = tokenAmount.map((item) => item.amount);
      }
      const results = [];
      let splIndex = 0;
      for (let i = 0; i < addresses.length; i++) {
        if (isNativeToken[i]) {
          results.push(nativeSolBalance);
        } else {
          results.push(splTokenBalances[splIndex]);
          splIndex++;
        }
      }
      return results;
    });
  }
  getBalanceByAddress(address) {
    return __async(this, null, function* () {
      const tokenPublicKey = new import_web34.PublicKey(address);
      const userPublicKey = new import_web34.PublicKey(this._address);
      const userTokenAccount = getTokenAccounts(tokenPublicKey, userPublicKey);
      const connection = this.connection;
      const tokenAmount = yield (0, import_spl_token2.getAccount)(
        connection,
        userTokenAccount,
        "confirmed"
      );
      return tokenAmount.amount;
    });
  }
  call(address, method, params, options) {
    return __async(this, null, function* () {
      if (method === "balanceOf") {
        return this.getBalanceByAddress(address);
      }
      if (method === "allowance") {
        return import_types2.MaxUint256;
      }
      return BigInt(0);
    });
  }
  sendTransaction(contractAddress, method, payload, options) {
    return __async(this, null, function* () {
      console.log("-- solanan sendTransaction", {
        contractAddress,
        method,
        payload,
        options
      });
      if (method === "deposit") {
        return deposit({
          vaultAddress: contractAddress,
          userAddress: this._address,
          connection: this.connection,
          depositData: payload.data[0],
          sendTransaction: this._provider.sendTransaction
        });
      }
    });
  }
  callOnChain(chain, address, method, params, options) {
    return __async(this, null, function* () {
      console.log("-- params ", {
        chain,
        address,
        method,
        params
      });
      if (method === "getDepositFee") {
        return getDepositQuoteFee({
          vaultAddress: address,
          userAddress: this._address,
          connection: this.connection,
          depositData: params[1]
        });
      }
      return 0;
    });
  }
  estimateGasFee(contractAddress, method, payload) {
    return __async(this, null, function* () {
      return BigInt(0);
    });
  }
  pollTransactionReceiptWithBackoff(txHash, baseInterval, maxInterval, maxRetries) {
    return __async(this, null, function* () {
      return Promise.resolve({ status: 1 });
    });
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DefaultSolanaWalletAdapter,
  version
});
