// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/errors/index.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * TooManyUpdaters: ''
 *
 * @category Errors
 * @category generated
 */
export class TooManyUpdatersError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'TooManyUpdaters'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, TooManyUpdatersError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new TooManyUpdatersError())
createErrorFromNameLookup.set(
  'TooManyUpdaters',
  () => new TooManyUpdatersError()
)

/**
 * InvalidUpdater: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidUpdaterError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'InvalidUpdater'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidUpdaterError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new InvalidUpdaterError())
createErrorFromNameLookup.set('InvalidUpdater', () => new InvalidUpdaterError())

/**
 * NotFound: ''
 *
 * @category Errors
 * @category generated
 */
export class NotFoundError extends Error {
  readonly code: number = 0x1772
  readonly name: string = 'NotFound'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, NotFoundError)
    }
  }
}

createErrorFromCodeLookup.set(0x1772, () => new NotFoundError())
createErrorFromNameLookup.set('NotFound', () => new NotFoundError())

/**
 * InvalidSize: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidSizeError extends Error {
  readonly code: number = 0x1773
  readonly name: string = 'InvalidSize'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidSizeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1773, () => new InvalidSizeError())
createErrorFromNameLookup.set('InvalidSize', () => new InvalidSizeError())

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
