// @orderly.network/ui-connector/dist/index.js
'use strict';

var react = require('react');
var hooks = require('@orderly.network/hooks');
var i18n = require('@orderly.network/i18n');
var types = require('@orderly.network/types');
var ui = require('@orderly.network/ui');
var jsxRuntime = require('react/jsx-runtime');
var reactApp = require('@orderly.network/react-app');
var uiChainSelector = require('@orderly.network/ui-chain-selector');

// src/component/walletConnectorContent.tsx
var StepItem = (props) => {
  const { title, description, showDivider } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { position: "relative", className: "oui-pl-8", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { as: "div", intensity: 98, size: "sm", children: title }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { as: "div", intensity: 54, size: "2xs", children: description })
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-absolute oui-left-0 oui-top-1 oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(Identifier, { ...props }) }),
    showDivider && /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { position: "absolute", left: 12, top: 23, bottom: -21, zIndex: 0, children: /* @__PURE__ */ jsxRuntime.jsx(
      ui.Divider,
      {
        lineStyle: "dashed",
        direction: "vertical",
        intensity: 16,
        className: "oui-h-full"
      }
    ) })
  ] });
};
var Identifier = (props) => {
  const { active, isLoading, isCompleted } = props;
  if (isLoading) {
    return /* @__PURE__ */ jsxRuntime.jsx(ui.Spinner, { size: "sm", className: "oui-ml-1" });
  }
  if (isCompleted) {
    return /* @__PURE__ */ jsxRuntime.jsx(ui.CheckedCircleFillIcon, { opacity: 1, className: "oui-text-primary" });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(Dot, { active: !!active });
};
var Dot = ({
  active,
  className
}) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: ui.cn(
        "oui-w-[8.3px] oui-h-[8.3px] oui-rounded-full oui-ml-2 oui-mt-1",
        className,
        active ? "oui-bg-primary-light" : "oui-bg-base-2"
      )
    }
  );
};
var WalletConnectContent = (props) => {
  const { initAccountState = types.AccountStatusEnum.NotConnected } = props;
  const [remember, setRemember] = react.useState(true);
  const ee = hooks.useEventEmitter();
  const { t } = i18n.useTranslation();
  const { disconnect, namespace } = hooks.useWalletConnector();
  const { state: accountState, account } = hooks.useAccount();
  const [state, setState] = react.useState(initAccountState);
  const [activeStep, setActiveStep] = react.useState(0);
  const [loading, setLoading] = react.useState(false);
  const [showLedgerButton, setShowLedgerButton] = react.useState(false);
  const { ledgerWallet, setLedgerAddress } = hooks.useStorageLedgerAddress();
  const handleRef = react.useRef(0);
  const [firstShowDialog] = hooks.useLocalStorage(
    "orderly-first-show-wallet-connector-dialog",
    void 0
  );
  react.useEffect(() => {
    return () => {
      localStorage.setItem("orderly-first-show-wallet-connector-dialog", "1");
    };
  }, []);
  react.useEffect(() => {
    setState(accountState.status);
  }, [accountState]);
  const steps = react.useMemo(() => {
    const steps2 = [];
    if (initAccountState < types.AccountStatusEnum.SignedIn) {
      steps2.push({
        key: "signIn",
        title: t("connector.createAccount"),
        description: t("connector.createAccount.description")
      });
    }
    if (initAccountState < types.AccountStatusEnum.EnableTrading) {
      steps2.push({
        key: "enableTrading",
        title: t("connector.enableTrading"),
        description: t("connector.enableTrading.description")
      });
    }
    return steps2;
  }, [initAccountState, t]);
  react.useEffect(() => {
    if (namespace != types.ChainNamespace.solana) {
      setShowLedgerButton(false);
      return;
    }
    if (!ledgerWallet) {
      setShowLedgerButton(true);
      return;
    }
    if (ledgerWallet && account.address) {
      if (!ledgerWallet.includes(account.address)) {
        setShowLedgerButton(true);
        return;
      }
    }
    setShowLedgerButton(false);
  }, [namespace, account.address, ledgerWallet]);
  const onEnableTrading = () => {
    setLoading(true);
    return props.enableTrading(remember).then(
      async (res) => {
        handleRef.current++;
        setLoading(false);
        setActiveStep((step) => step + 1);
        try {
          await props.enableTradingComplted?.();
        } catch (e) {
        }
        if (typeof props.onCompleted === "function") {
          props.onCompleted();
        } else if (typeof props.close === "function") {
          props.close();
        }
      },
      (reject) => {
        setLoading(false);
        if (reject === -1) {
          return;
        }
        if (reject.message.indexOf(
          "Signing off chain messages with Ledger is not yet supported"
        ) !== -1) {
          ee.emit("wallet:sign-message-with-ledger-error", {
            message: reject.message,
            userAddress: account.address
          });
          return;
        }
        ui.toast.error(paseErrorMsg(reject));
      }
    ).catch((e) => {
      setLoading(false);
    });
  };
  const onDisconnect = async () => {
    localStorage.removeItem("orderly_link_device");
    disconnect({
      label: state.connectWallet?.name
    }).then(() => {
      account.disconnect();
      if (typeof props.close === "function") {
        props.close();
      }
    });
  };
  const onSignIn = () => {
    setLoading(true);
    return props.signIn().then(
      (res) => {
        setActiveStep((step) => step + 1);
        onEnableTrading();
      },
      (reject) => {
        setLoading(false);
        if (reject === -1) {
          return;
        }
        if (reject.message.indexOf(
          "Signing off chain messages with Ledger is not yet supported"
        ) !== -1) {
          ee.emit("wallet:sign-message-with-ledger-error", {
            message: reject.message,
            userAddress: account.address
          });
          return;
        }
        ui.toast.error(paseErrorMsg(reject));
      }
    ).catch((e) => {
      setLoading(false);
    });
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { id: "oui-wallet-connect-dialog-content", className: "oui-font-semibold", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, size: "xs", children: t("connector.expired") }),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Box,
      {
        p: 4,
        my: 6,
        intensity: 600,
        r: "lg",
        className: "oui-space-y-5",
        position: "relative",
        children: steps.map((step, index) => {
          const isLast = index === steps.length - 1;
          return /* @__PURE__ */ jsxRuntime.jsx(
            StepItem,
            {
              title: step.title,
              description: step.description,
              isCompleted: activeStep > index,
              active: activeStep === index,
              isLoading: loading && activeStep === index,
              showDivider: !isLast
            },
            step.key
          );
        })
      }
    ),
    props.showRefCodeInput && steps.length == 2 && /* @__PURE__ */ jsxRuntime.jsx(ReferralCode, { ...props }),
    firstShowDialog && /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", itemAlign: "center", children: [
      /* @__PURE__ */ jsxRuntime.jsx(RememberMe, {}),
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Switch,
        {
          color: "primary",
          checked: remember,
          onCheckedChange: setRemember,
          disabled: loading,
          className: "data-[state=checked]:oui-bg-primary-darken"
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { justify: "center", mt: 8, className: "oui-w-full", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { className: "oui-w-full", children: /* @__PURE__ */ jsxRuntime.jsx(
      ActionButton,
      {
        state,
        signIn: onSignIn,
        enableTrading: onEnableTrading,
        loading,
        disabled: state >= types.AccountStatusEnum.EnableTrading,
        showLedgerButton
      }
    ) }) }),
    state > types.AccountStatusEnum.NotConnected && /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Flex,
      {
        justify: "center",
        mt: 4,
        gap: 1,
        className: "oui-w-full oui-cursor-pointer",
        onClick: onDisconnect,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(DisconnectIcon, {}),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { className: "oui-text-base-contrast-80 oui-text-sm", children: t("connector.disconnectWallet") })
        ]
      }
    )
  ] });
};
var DisconnectIcon = () => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "18",
      height: "18",
      viewBox: "0 0 18 18",
      fill: "none",
      children: /* @__PURE__ */ jsxRuntime.jsx(
        "path",
        {
          d: "M2.24219 5.24316C2.24219 3.58641 3.58536 2.24316 5.24219 2.24316H8.24219C9.89894 2.24316 11.2422 3.58641 11.2422 5.24316C11.2422 5.65716 10.9062 5.99316 10.4922 5.99316C10.0782 5.99316 9.74219 5.65716 9.74219 5.24316C9.74219 4.41441 9.07094 3.74316 8.24219 3.74316H5.24219C4.41374 3.74316 3.74219 4.41441 3.74219 5.24316V12.7432C3.74219 13.5719 4.41374 14.2432 5.24219 14.2432H8.24219C9.07094 14.2432 9.74219 13.5719 9.74219 12.7432C9.74219 12.3292 10.0782 11.9932 10.4922 11.9932C10.9062 11.9932 11.2422 12.3292 11.2422 12.7432C11.2422 14.3999 9.89894 15.7432 8.24219 15.7432H5.24219C3.58536 15.7432 2.24219 14.3999 2.24219 12.7432V5.24316ZM7.49219 8.99316C7.49219 8.57916 7.82819 8.24316 8.24219 8.24316H13.9144L12.4377 6.74316L13.4922 5.68866L16.2814 8.45391C16.5739 8.74716 16.5739 9.23915 16.2814 9.5324L13.4922 12.2977L12.4377 11.2432L13.9144 9.74316H8.24219C7.82819 9.74316 7.49219 9.40716 7.49219 8.99316Z",
          fill: "white",
          fillOpacity: "0.8"
        }
      )
    }
  );
};
var ActionButton = ({
  state,
  signIn,
  enableTrading,
  loading,
  disabled,
  showLedgerButton
}) => {
  const { t } = i18n.useTranslation();
  if (state <= types.AccountStatusEnum.NotSignedIn) {
    return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", gap: 3, className: "oui-w-full", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Button,
        {
          fullWidth: true,
          onClick: () => signIn(),
          loading,
          disabled,
          children: t("connector.createAccount")
        }
      ),
      showLedgerButton && /* @__PURE__ */ jsxRuntime.jsx(
        WithLedgerButton,
        {
          onClick: () => signIn(),
          content: t("connector.createAccountWithLedger")
        }
      )
    ] });
  }
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", gap: 3, className: "oui-w-full", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        fullWidth: true,
        onClick: () => enableTrading(),
        loading,
        disabled,
        children: t("connector.enableTrading")
      }
    ),
    showLedgerButton && /* @__PURE__ */ jsxRuntime.jsx(
      WithLedgerButton,
      {
        onClick: () => enableTrading(),
        disabled,
        content: t("connector.enableTradingWithLedger")
      }
    )
  ] });
};
var WithLedgerButton = ({
  onClick,
  disabled,
  content
}) => {
  const { t } = i18n.useTranslation();
  const { state } = hooks.useAccount();
  const address = state.address;
  const { setLedgerAddress } = hooks.useStorageLedgerAddress();
  if (!address) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Button,
    {
      variant: "outlined",
      color: "primary",
      fullWidth: true,
      onClick: () => {
        setLedgerAddress(address);
        onClick();
      },
      disabled,
      className: "oui-w-full",
      children: content
    }
  );
};
var ReferralCode = (props) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.TextField,
    {
      placeholder: t("connector.referralCode.placeholder"),
      fullWidth: true,
      label: "",
      value: props.refCode,
      onChange: (e) => {
        const _value = e.target.value.toUpperCase().replace(/[^A-Z0-9]/g, "");
        props.setRefCode(_value);
      },
      classNames: {
        label: "oui-text-base-contrast-54 oui-text-xs",
        input: "placeholder:oui-text-base-contrast-20 placeholder:oui-text-sm"
      },
      formatters: [
        ui.inputFormatter.createRegexInputFormatter((value) => {
          return String(value).replace(
            /[a-z]/g,
            (char) => char.toUpperCase()
          );
        }),
        ui.inputFormatter.createRegexInputFormatter(/[^A-Z0-9]/g)
      ],
      onClear: () => {
        props.setRefCode("");
      },
      autoComplete: "off",
      helpText: props.helpText,
      className: "oui-mb-4",
      color: props.helpText ? "danger" : void 0
    }
  );
};
var RememberMe = () => {
  const { t } = i18n.useTranslation();
  const showRememberHint = () => {
    if (window.innerWidth > 768) {
      return;
    }
    ui.modal.alert({
      title: t("connector.rememberMe"),
      message: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-text-2xs oui-text-base-contrast/60", children: t("connector.rememberMe.description") })
    });
  };
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Tooltip,
    {
      content: t("connector.rememberMe.description"),
      className: "oui-max-w-[300px]",
      children: /* @__PURE__ */ jsxRuntime.jsx("button", { onClick: showRememberHint, children: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text,
        {
          intensity: 54,
          size: "xs",
          className: "oui-underline oui-underline-offset-4 oui-decoration-dashed oui-decoration-base-contrast-36",
          children: t("connector.rememberMe")
        }
      ) })
    }
  );
};
function paseErrorMsg(reject) {
  Object.keys(reject).forEach((key) => {
  });
  let msg = i18n.i18n.t("connector.somethingWentWrong");
  if (reject.toString().includes("rejected")) {
    msg = i18n.i18n.t("connector.userRejected");
  }
  return ui.capitalizeFirstLetter(msg) ?? msg;
}
var useWalletConnectorBuilder = () => {
  const { state, createOrderlyKey, createAccount } = hooks.useAccount();
  const [refCode, setRefCode] = react.useState("");
  const [helpText, setHelpText] = react.useState("");
  const { t } = i18n.useTranslation();
  const { trigger: verifyRefCode } = hooks.useLazyQuery(
    `/v1/public/referral/verify_ref_code?referral_code=${refCode}`
  );
  react.useEffect(() => {
    const refCode2 = localStorage.getItem("referral_code");
    if (refCode2 != null) {
      setRefCode(refCode2);
    }
  }, []);
  const { referral_code, isLoading } = hooks.useGetReferralCode(state.accountId);
  const [bindRefCode, { error: updateOrderError, isMutating: updateMutating }] = hooks.useMutation("/v1/referral/bind", "POST");
  react.useEffect(() => {
    if (refCode.length === 0) {
      setHelpText("");
    }
  }, [refCode]);
  const enableTradingComplted = () => {
    ui.toast.success(t("connector.walletConnected"));
    if (refCode.length >= 4 && refCode.length <= 10)
      bindRefCode({ referral_code: refCode }).finally(() => {
        localStorage.removeItem("referral_code");
      });
  };
  const checkRefCode = async () => {
    if (refCode.length === 0) {
      return Promise.resolve(void 0);
    }
    if (refCode.length > 0 && (refCode.length < 4 || refCode.length > 10)) {
      return Promise.resolve(t("connector.referralCode.invalid"));
    }
    const { exist } = await verifyRefCode();
    if (exist === false) {
      return Promise.resolve(t("connector.referralCode.notExist"));
    }
    return Promise.resolve(void 0);
  };
  const showRefCodeInput = (referral_code?.length || 0) === 0 && !isLoading;
  const signIn = async () => {
    if (showRefCodeInput) {
      const info = await checkRefCode();
      if (typeof info !== "undefined") {
        setHelpText(info);
        return Promise.reject(-1);
      }
    }
    setHelpText("");
    return createAccount();
  };
  const enableTrading = async (remember) => {
    if (showRefCodeInput) {
      const info = await checkRefCode();
      if (typeof info !== "undefined") {
        setHelpText(info);
        return Promise.reject(-1);
      }
    }
    setHelpText("");
    return createOrderlyKey(remember);
  };
  return {
    enableTrading,
    initAccountState: state.status,
    signIn,
    enableTradingComplted,
    refCode,
    setRefCode,
    helpText,
    showRefCodeInput
  };
};
var WalletConnectorModalId = "walletConnector";
var WalletConnectorSheetId = "walletConnectorSheet";
var WalletConnectorWidget = (props) => {
  const state = useWalletConnectorBuilder();
  return /* @__PURE__ */ jsxRuntime.jsx(WalletConnectContent, { ...state, ...props });
};
ui.registerSimpleDialog(WalletConnectorModalId, WalletConnectorWidget, {
  size: "sm",
  title: () => i18n.i18n.t("connector.connectWallet")
});
ui.registerSimpleSheet(WalletConnectorSheetId, WalletConnectorWidget, {
  title: () => i18n.i18n.t("connector.connectWallet")
});
var useAuthGuard = (status) => {
  const { state } = hooks.useAccount();
  const { wrongNetwork, disabledConnect } = reactApp.useAppContext();
  const _status = react.useMemo(() => {
    if (status === void 0) {
      return state.status === types.AccountStatusEnum.EnableTradingWithoutConnected ? types.AccountStatusEnum.EnableTradingWithoutConnected : types.AccountStatusEnum.EnableTrading;
    }
    return status;
  }, [status, state.status]);
  return react.useMemo(() => {
    return state.status >= _status && !wrongNetwork && !disabledConnect;
  }, [state.status, _status, wrongNetwork, disabledConnect]);
};
var AuthGuard = (props) => {
  const {
    status,
    buttonProps,
    fallback,
    descriptions,
    classNames,
    networkId,
    id,
    bridgeLessOnly
    // ...rest
  } = props;
  const { t } = i18n.useTranslation();
  const { state } = hooks.useAccount();
  const { wrongNetwork, disabledConnect } = reactApp.useAppContext();
  const _status = react.useMemo(() => {
    if (status === void 0) {
      return state.status === types.AccountStatusEnum.EnableTradingWithoutConnected ? types.AccountStatusEnum.EnableTradingWithoutConnected : types.AccountStatusEnum.EnableTrading;
    }
    return status;
  }, [status, state.status]);
  const labels = {
    connectWallet: t("connector.connectWallet"),
    switchChain: t("connector.wrongNetwork"),
    enableTrading: t("connector.enableTrading"),
    signin: t("connector.createAccount"),
    ...props.labels
  };
  const Left = react.useMemo(() => {
    if (typeof fallback !== "undefined") {
      return fallback({
        validating: state.validating,
        status: state.status,
        wrongNetwork
      });
    }
    if (state.validating && !disabledConnect) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        StatusInfo,
        {
          angle: 45,
          disabled: true,
          loading: true,
          description: descriptions?.connectWallet,
          id,
          type: "button",
          ...buttonProps,
          children: labels.connectWallet
        }
      );
    }
    return /* @__PURE__ */ jsxRuntime.jsx(
      DefaultFallback,
      {
        bridgeLessOnly,
        status: state.status,
        buttonProps: { ...buttonProps, id, type: "button" },
        wrongNetwork,
        networkId: props.networkId,
        labels,
        descriptions,
        disabledConnect
      }
    );
  }, [
    state.status,
    state.validating,
    buttonProps,
    wrongNetwork,
    labels,
    descriptions
  ]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Either,
    {
      value: state.status >= _status && !wrongNetwork && !disabledConnect,
      left: Left,
      children: props.children
    }
  );
};
var ModalTitle = () => {
  const { t } = i18n.useTranslation();
  const { state } = hooks.useAccount();
  if (state.status < types.AccountStatusEnum.SignedIn) {
    return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("connector.createAccount") });
  }
  if (state.status < types.AccountStatusEnum.EnableTrading) {
    return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("connector.enableTrading") });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("connector.connectWallet") });
};
var DefaultFallback = (props) => {
  const { buttonProps, labels, descriptions } = props;
  const { t } = i18n.useTranslation();
  const { connectWallet } = reactApp.useAppContext();
  const { account } = hooks.useAccount();
  const { isMobile } = ui.useScreen();
  const matches = hooks.useMediaQuery(types.MEDIA_TABLET);
  const onConnectOrderly = () => {
    ui.modal.show(matches ? WalletConnectorSheetId : WalletConnectorModalId, {
      title: /* @__PURE__ */ jsxRuntime.jsx(ModalTitle, {})
    }).then(
      (r) => void 0,
      (error) => void 0
    );
  };
  const onConnectWallet = async () => {
    const res = await connectWallet();
    if (!res) {
      return;
    }
    if (res.wrongNetwork) {
      switchChain();
    } else {
      if ((res?.status ?? types.AccountStatusEnum.NotConnected) < types.AccountStatusEnum.EnableTrading) {
        onConnectOrderly();
      }
    }
  };
  const switchChain = () => {
    account.once("validate:end", (status) => {
      if (status < types.AccountStatusEnum.EnableTrading) {
        onConnectOrderly();
      } else {
        ui.toast.success(t("connector.walletConnected"));
      }
    });
    ui.modal.show(
      isMobile ? uiChainSelector.ChainSelectorSheetId : uiChainSelector.ChainSelectorDialogId,
      {
        networkId: props.networkId,
        bridgeLessOnly: props.bridgeLessOnly
      }
    ).then(
      (r) => {
        if (!r.wrongNetwork) {
          if (props.status >= types.AccountStatusEnum.Connected) {
            if (props.status < types.AccountStatusEnum.EnableTrading) {
              onConnectOrderly();
            } else {
              ui.toast.success(t("connector.walletConnected"));
            }
          }
        }
      },
      (error) => void 0
    );
  };
  if (props.wrongNetwork && !props.disabledConnect) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      StatusInfo,
      {
        color: "warning",
        onClick: () => {
          switchChain();
        },
        description: descriptions?.switchChain,
        ...buttonProps,
        children: labels.switchChain
      }
    );
  }
  if (props.status <= types.AccountStatusEnum.NotConnected || props.disabledConnect) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      StatusInfo,
      {
        size: "lg",
        onClick: () => {
          onConnectWallet();
        },
        variant: props.disabledConnect ? void 0 : "gradient",
        angle: 45,
        description: descriptions?.connectWallet,
        disabled: props.disabledConnect,
        ...buttonProps,
        children: labels.connectWallet
      }
    );
  }
  if (props.status <= types.AccountStatusEnum.NotSignedIn) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      StatusInfo,
      {
        size: "lg",
        onClick: () => {
          onConnectOrderly();
        },
        angle: 45,
        description: descriptions?.signin,
        ...buttonProps,
        children: labels.signin
      }
    );
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    StatusInfo,
    {
      size: "lg",
      description: descriptions?.enableTrading,
      ...buttonProps,
      onClick: () => onConnectOrderly(),
      children: labels.enableTrading
    }
  );
};
AuthGuard.displayName = "AuthGuard";
var StatusInfo = (props) => {
  const { description, ...buttonProps } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { ...buttonProps }),
    !!description && /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { mt: 4, className: "oui-leading-none", style: { lineHeight: 0 }, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 36, children: description }) })
  ] });
};
var AuthGuardDataTable = (props) => {
  const {
    status,
    // message,
    labels,
    description,
    dataSource,
    children,
    ...rest
  } = props;
  const { state } = hooks.useAccount();
  const { wrongNetwork, disabledConnect } = reactApp.useAppContext();
  const _status = react.useMemo(() => {
    if (status === void 0) {
      return state.status === types.AccountStatusEnum.EnableTradingWithoutConnected ? types.AccountStatusEnum.EnableTradingWithoutConnected : types.AccountStatusEnum.EnableTrading;
    }
    return status;
  }, [status, state.status]);
  const data = reactApp.useDataTap(dataSource, {
    accountStatus: _status
  });
  const ignoreLoadingCheck = wrongNetwork || disabledConnect || state.status < _status || props.ignoreLoadingCheck;
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.DataTable,
    {
      dataSource: data,
      ignoreLoadingCheck,
      emptyView: /* @__PURE__ */ jsxRuntime.jsx(
        GuardView,
        {
          status: _status,
          description,
          labels,
          className: props.classNames?.authGuardDescription,
          visible: !state.validating
        }
      ),
      manualPagination: true,
      ...rest,
      children
    }
  );
};
var GuardView = (props) => {
  const { t } = i18n.useTranslation();
  const DESCRIPTIONS = {
    connectWallet: t("connector.trade.connectWallet.tooltip"),
    switchChain: t("connector.wrongNetwork.tooltip"),
    enableTrading: t("connector.trade.enableTrading.tooltip"),
    signin: t("connector.trade.createAccount.tooltip")
  };
  const descriptions = { ...DESCRIPTIONS, ...props.description };
  if (!props.visible) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { py: 8, children: /* @__PURE__ */ jsxRuntime.jsx(
    AuthGuard,
    {
      status: props.status,
      labels: props.labels,
      descriptions,
      buttonProps: {
        size: "md"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.ExtensionSlot, { position: ui.ExtensionPositionEnum.EmptyDataIdentifier })
    }
  ) });
};
var AuthGuardEmpty = (props) => {
  const { t } = i18n.useTranslation();
  const {
    hint = {
      connectWallet: t("connector.trade.connectWallet.tooltip"),
      signIn: t("connector.trade.createAccount.tooltip"),
      enableTrading: t("connector.trade.enableTrading.tooltip"),
      wrongNetwork: t("connector.wrongNetwork.tooltip")
    },
    status
  } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { my: 8, children: /* @__PURE__ */ jsxRuntime.jsx(
    AuthGuard,
    {
      status,
      descriptions: { ...hint, switchChain: hint.wrongNetwork },
      buttonProps: {
        size: "md"
      },
      children: props.children || /* @__PURE__ */ jsxRuntime.jsx(ui.EmptyDataState, {})
    }
  ) });
};
AuthGuardEmpty.displayName = "AuthGuardEmpty";
var AuthGuardTooltip = (props) => {
  const { t } = i18n.useTranslation();
  const {
    opactiy = 90,
    tooltip = {
      connectWallet: t("connector.setUp.connectWallet.tooltip"),
      signIn: t("connector.setUp.createAccount.tooltip"),
      enableTrading: t("connector.setUp.enableTrading.tooltip"),
      wrongNetwork: t("connector.wrongNetwork.tooltip")
    }
  } = props;
  const [open, setOpen] = react.useState(false);
  const { state } = hooks.useAccount();
  const isSupport = true;
  const { wrongNetwork } = reactApp.useAppContext();
  const hint = react.useMemo(() => {
    if (wrongNetwork) {
      return tooltip?.wrongNetwork;
    }
    switch (state.status) {
      case types.AccountStatusEnum.NotConnected:
        return tooltip?.connectWallet;
      case types.AccountStatusEnum.NotSignedIn:
        return tooltip?.signIn;
      case types.AccountStatusEnum.DisabledTrading:
        return tooltip?.enableTrading;
      case types.AccountStatusEnum.EnableTrading: {
        return "";
      }
      default:
        return props.content;
    }
  }, [props.content, state, isSupport, tooltip]);
  const newOpacity = react.useMemo(() => {
    switch (state.status) {
      case types.AccountStatusEnum.NotConnected:
      case types.AccountStatusEnum.NotSignedIn:
        return opactiy;
      case types.AccountStatusEnum.EnableTrading: {
        return void 0;
      }
      default:
        return void 0;
    }
  }, [props.opactiy, state, isSupport]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Tooltip,
    {
      open: hint ? open : false,
      onOpenChange: setOpen,
      content: hint,
      className: "oui-text-2xs",
      align: props.align,
      alignOffset: props.alignOffset,
      side: props.side,
      sideOffset: props.sideOffset,
      children: /* @__PURE__ */ jsxRuntime.jsx("div", { style: { opacity: newOpacity }, children: props.children })
    }
  );
};
AuthGuardTooltip.displayName = "AuthGuardTooltip";
var AuthStatusEnum = /* @__PURE__ */ ((AuthStatusEnum2) => {
  AuthStatusEnum2[AuthStatusEnum2["WrongNetwork"] = 0] = "WrongNetwork";
  AuthStatusEnum2[AuthStatusEnum2["ConnectWallet"] = 1] = "ConnectWallet";
  AuthStatusEnum2[AuthStatusEnum2["CreateAccount"] = 2] = "CreateAccount";
  AuthStatusEnum2[AuthStatusEnum2["EnableTrading"] = 3] = "EnableTrading";
  return AuthStatusEnum2;
})(AuthStatusEnum || {});
var useAuthStatus = () => {
  const { state } = hooks.useAccount();
  const { wrongNetwork, disabledConnect } = reactApp.useAppContext();
  return react.useMemo(() => {
    if (wrongNetwork && !disabledConnect) {
      return 0 /* WrongNetwork */;
    }
    if (state.status === types.AccountStatusEnum.EnableTradingWithoutConnected) {
      return 3 /* EnableTrading */;
    }
    if (state.status <= types.AccountStatusEnum.NotConnected || disabledConnect) {
      return 1 /* ConnectWallet */;
    }
    if (state.status <= types.AccountStatusEnum.NotSignedIn) {
      return 2 /* CreateAccount */;
    }
    return 3 /* EnableTrading */;
  }, [state.status, wrongNetwork, disabledConnect]);
};

exports.AuthGuard = AuthGuard;
exports.AuthGuardDataTable = AuthGuardDataTable;
exports.AuthGuardEmpty = AuthGuardEmpty;
exports.AuthGuardTooltip = AuthGuardTooltip;
exports.AuthStatusEnum = AuthStatusEnum;
exports.WalletConnectContent = WalletConnectContent;
exports.WalletConnectorModalId = WalletConnectorModalId;
exports.WalletConnectorSheetId = WalletConnectorSheetId;
exports.WalletConnectorWidget = WalletConnectorWidget;
exports.useAuthGuard = useAuthGuard;
exports.useAuthStatus = useAuthStatus;
exports.useWalletConnectorBuilder = useWalletConnectorBuilder;
