// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/programs/executor.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  ClusterFilter,
  Context,
  Program,
  PublicKey,
} from '@metaplex-foundation/umi';
import { getExecutorErrorFromCode, getExecutorErrorFromName } from '../errors';

export const EXECUTOR_PROGRAM_ID =
  '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn' as PublicKey<'6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'>;

export function createExecutorProgram(): Program {
  return {
    name: 'executor',
    publicKey: EXECUTOR_PROGRAM_ID,
    getErrorFromCode(code: number, cause?: Error) {
      return getExecutorErrorFromCode(code, this, cause);
    },
    getErrorFromName(name: string, cause?: Error) {
      return getExecutorErrorFromName(name, this, cause);
    },
    isOnCluster() {
      return true;
    },
  };
}

export function getExecutorProgram<T extends Program = Program>(
  context: Pick<Context, 'programs'>,
  clusterFilter?: ClusterFilter
): T {
  return context.programs.get<T>('executor', clusterFilter);
}

export function getExecutorProgramId(
  context: Pick<Context, 'programs'>,
  clusterFilter?: ClusterFilter
): PublicKey {
  return context.programs.getPublicKey(
    'executor',
    EXECUTOR_PROGRAM_ID,
    clusterFilter
  );
}
