// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/simple_messagelib/instructions/initMessageLib.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type InitMessageLibInstructionAccounts = {
  payer?: Signer;
  messageLib: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
};

// Data.
export type InitMessageLibInstructionData = {
  discriminator: Uint8Array;
  eid: number;
  endpoint: PublicKey;
  endpointProgram: PublicKey;
  admin: PublicKey;
  fee: bigint;
  lzTokenFee: bigint;
};

export type InitMessageLibInstructionDataArgs = {
  eid: number;
  endpoint: PublicKey;
  endpointProgram: PublicKey;
  admin: PublicKey;
  fee: number | bigint;
  lzTokenFee: number | bigint;
};

export function getInitMessageLibInstructionDataSerializer(): Serializer<
  InitMessageLibInstructionDataArgs,
  InitMessageLibInstructionData
> {
  return mapSerializer<
    InitMessageLibInstructionDataArgs,
    any,
    InitMessageLibInstructionData
  >(
    struct<InitMessageLibInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['eid', u32()],
        ['endpoint', publicKeySerializer()],
        ['endpointProgram', publicKeySerializer()],
        ['admin', publicKeySerializer()],
        ['fee', u64()],
        ['lzTokenFee', u64()],
      ],
      { description: 'InitMessageLibInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([184, 93, 59, 206, 98, 238, 225, 54]),
    })
  ) as Serializer<
    InitMessageLibInstructionDataArgs,
    InitMessageLibInstructionData
  >;
}

// Args.
export type InitMessageLibInstructionArgs = InitMessageLibInstructionDataArgs;

// Instruction.
export function initMessageLib(
  context: Pick<Context, 'payer' | 'programs'>,
  input: InitMessageLibInstructionAccounts & InitMessageLibInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'simpleMessagelib',
    '6GsmxMTHAAiFKfemuM4zBjumTjNSX5CAiw4xSSXM2Toy'
  );

  // Accounts.
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true as boolean,
      value: input.payer ?? null,
    },
    messageLib: {
      index: 1,
      isWritable: true as boolean,
      value: input.messageLib ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: InitMessageLibInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getInitMessageLibInstructionDataSerializer().serialize(
    resolvedArgs as InitMessageLibInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
