// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/simple_messagelib/instructions/quote.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bool,
  bytes,
  mapSerializer,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import { Packet, PacketArgs, getPacketSerializer } from '../types';

// Accounts.
export type QuoteInstructionAccounts = {
  /** The message lib authority of the endpoint */
  endpoint: Signer;
  messageLib: PublicKey | Pda;
};

// Data.
export type QuoteInstructionData = {
  discriminator: Uint8Array;
  packet: Packet;
  options: Uint8Array;
  payInLzToken: boolean;
};

export type QuoteInstructionDataArgs = {
  packet: PacketArgs;
  options: Uint8Array;
  payInLzToken: boolean;
};

export function getQuoteInstructionDataSerializer(): Serializer<
  QuoteInstructionDataArgs,
  QuoteInstructionData
> {
  return mapSerializer<QuoteInstructionDataArgs, any, QuoteInstructionData>(
    struct<QuoteInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['packet', getPacketSerializer()],
        ['options', bytes({ size: u32() })],
        ['payInLzToken', bool()],
      ],
      { description: 'QuoteInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([149, 42, 109, 247, 134, 146, 213, 123]),
    })
  ) as Serializer<QuoteInstructionDataArgs, QuoteInstructionData>;
}

// Args.
export type QuoteInstructionArgs = QuoteInstructionDataArgs;

// Instruction.
export function quote(
  context: Pick<Context, 'programs'>,
  input: QuoteInstructionAccounts & QuoteInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'simpleMessagelib',
    '6GsmxMTHAAiFKfemuM4zBjumTjNSX5CAiw4xSSXM2Toy'
  );

  // Accounts.
  const resolvedAccounts = {
    endpoint: {
      index: 0,
      isWritable: false as boolean,
      value: input.endpoint ?? null,
    },
    messageLib: {
      index: 1,
      isWritable: false as boolean,
      value: input.messageLib ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: QuoteInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getQuoteInstructionDataSerializer().serialize(
    resolvedArgs as QuoteInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
