// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/simple_messagelib/instructions/setFee.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type SetFeeInstructionAccounts = {
  admin: Signer;
  messageLib: PublicKey | Pda;
};

// Data.
export type SetFeeInstructionData = {
  discriminator: Uint8Array;
  fee: bigint;
  lzTokenFee: bigint;
};

export type SetFeeInstructionDataArgs = {
  fee: number | bigint;
  lzTokenFee: number | bigint;
};

export function getSetFeeInstructionDataSerializer(): Serializer<
  SetFeeInstructionDataArgs,
  SetFeeInstructionData
> {
  return mapSerializer<SetFeeInstructionDataArgs, any, SetFeeInstructionData>(
    struct<SetFeeInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['fee', u64()],
        ['lzTokenFee', u64()],
      ],
      { description: 'SetFeeInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([18, 154, 24, 18, 237, 214, 19, 80]),
    })
  ) as Serializer<SetFeeInstructionDataArgs, SetFeeInstructionData>;
}

// Args.
export type SetFeeInstructionArgs = SetFeeInstructionDataArgs;

// Instruction.
export function setFee(
  context: Pick<Context, 'programs'>,
  input: SetFeeInstructionAccounts & SetFeeInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'simpleMessagelib',
    '6GsmxMTHAAiFKfemuM4zBjumTjNSX5CAiw4xSSXM2Toy'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false as boolean,
      value: input.admin ?? null,
    },
    messageLib: {
      index: 1,
      isWritable: true as boolean,
      value: input.messageLib ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetFeeInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSetFeeInstructionDataSerializer().serialize(
    resolvedArgs as SetFeeInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
