// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/simple_messagelib/instructions/setWlCaller.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type SetWlCallerInstructionAccounts = {
  admin: Signer;
  messageLib: PublicKey | Pda;
};

// Data.
export type SetWlCallerInstructionData = {
  discriminator: Uint8Array;
  newCaller: PublicKey;
};

export type SetWlCallerInstructionDataArgs = { newCaller: PublicKey };

export function getSetWlCallerInstructionDataSerializer(): Serializer<
  SetWlCallerInstructionDataArgs,
  SetWlCallerInstructionData
> {
  return mapSerializer<
    SetWlCallerInstructionDataArgs,
    any,
    SetWlCallerInstructionData
  >(
    struct<SetWlCallerInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['newCaller', publicKeySerializer()],
      ],
      { description: 'SetWlCallerInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([153, 41, 240, 37, 126, 107, 14, 253]),
    })
  ) as Serializer<SetWlCallerInstructionDataArgs, SetWlCallerInstructionData>;
}

// Args.
export type SetWlCallerInstructionArgs = SetWlCallerInstructionDataArgs;

// Instruction.
export function setWlCaller(
  context: Pick<Context, 'programs'>,
  input: SetWlCallerInstructionAccounts & SetWlCallerInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'simpleMessagelib',
    '6GsmxMTHAAiFKfemuM4zBjumTjNSX5CAiw4xSSXM2Toy'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: false as boolean,
      value: input.admin ?? null,
    },
    messageLib: {
      index: 1,
      isWritable: true as boolean,
      value: input.messageLib ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetWlCallerInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSetWlCallerInstructionDataSerializer().serialize(
    resolvedArgs as SetWlCallerInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
