// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/accounts/ExecuteHash.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'

/**
 * Arguments used to create {@link ExecuteHash}
 * @category Accounts
 * @category generated
 */
export type ExecuteHashArgs = {
  expiration: beet.bignum
  bump: number
}

export const executeHashDiscriminator = [34, 17, 137, 123, 62, 25, 205, 155]
/**
 * Holds the data for the {@link ExecuteHash} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class ExecuteHash implements ExecuteHashArgs {
  private constructor(
    readonly expiration: beet.bignum,
    readonly bump: number
  ) {}

  /**
   * Creates a {@link ExecuteHash} instance from the provided args.
   */
  static fromArgs(args: ExecuteHashArgs) {
    return new ExecuteHash(args.expiration, args.bump)
  }

  /**
   * Deserializes the {@link ExecuteHash} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [ExecuteHash, number] {
    return ExecuteHash.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link ExecuteHash} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<ExecuteHash> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find ExecuteHash account at ${address}`)
    }
    return ExecuteHash.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      'HtEYV4xB4wvsj5fgTkcfuChYpvGYzgzwvNhgDZQNh7wW'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, executeHashBeet)
  }

  /**
   * Deserializes the {@link ExecuteHash} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [ExecuteHash, number] {
    return executeHashBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link ExecuteHash} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return executeHashBeet.serialize({
      accountDiscriminator: executeHashDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link ExecuteHash}
   */
  static get byteSize() {
    return executeHashBeet.byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link ExecuteHash} data from rent
   *
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      ExecuteHash.byteSize,
      commitment
    )
  }

  /**
   * Determines if the provided {@link Buffer} has the correct byte size to
   * hold {@link ExecuteHash} data.
   */
  static hasCorrectByteSize(buf: Buffer, offset = 0) {
    return buf.byteLength - offset === ExecuteHash.byteSize
  }

  /**
   * Returns a readable version of {@link ExecuteHash} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      expiration: (() => {
        const x = <{ toNumber: () => number }>this.expiration
        if (typeof x.toNumber === 'function') {
          try {
            return x.toNumber()
          } catch (_) {
            return x
          }
        }
        return x
      })(),
      bump: this.bump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const executeHashBeet = new beet.BeetStruct<
  ExecuteHash,
  ExecuteHashArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['expiration', beet.i64],
    ['bump', beet.u8],
  ],
  ExecuteHash.fromArgs,
  'ExecuteHash'
)
