// @orderly.network/wallet-connector-privy/dist/index.mjs
import * as viemExport from 'viem';
import { defineChain } from 'viem';
import * as WagmiExport from 'wagmi';
import { createConfig, injected, createStorage, http, WagmiProvider, useConnect, useDisconnect, useAccount as useAccount$1, useSwitchChain } from 'wagmi';
import * as WagmiConnectorsExport from 'wagmi/connectors';
import React19, { createContext, useState, useEffect, useMemo, useContext, useRef, useCallback } from 'react';
import { mainnet, abstract, abstractTestnet } from 'viem/chains';
import { useWalletConnector, useAccount, useMainnetChainsStore, useTestnetChainsStore, WalletConnectorContext, useTrack, useEventEmitter, useStorageLedgerAddress, useLocalStorage, useStorageChain } from '@orderly.network/hooks';
import { SOLANA_TESTNET_CHAINID, SOLANA_MAINNET_CHAINID, ChainNamespace, ABSTRACT_CHAIN_ID_MAP, AccountStatusEnum, ABSTRACT_TESTNET_CHAINID, ABSTRACT_MAINNET_CHAINID, SolanaChains as SolanaChains$1, AbstractChains, ArbitrumSepoliaChainInfo, SolanaDevnetChainInfo, EMPTY_OBJECT, defaultMainnetChains, defaultTestnetChains, TrackerEventName, ConnectorKey } from '@orderly.network/types';
import { modal, useModal, SimpleSheet, SheetHeader, Text, Flex, Divider, cn, installExtension, ExtensionPositionEnum, useScreen, formatAddress, Button, SimpleDialog, TooltipProvider, CloseSquareFillIcon, ScrollArea, ExclamationFillIcon, Tooltip, CopyIcon, Checkbox, toast, DropdownMenuRoot, DropdownMenuTrigger, DropdownMenuPortal, DropdownMenuContent, DropdownMenuItem, ChevronDownIcon, ChevronUpIcon, Grid, ChainIcon, Popover } from '@orderly.network/ui';
import { useTranslation, Trans } from '@orderly.network/i18n';
import { AbstractWalletProvider as AbstractWalletProvider$1, useLoginWithAbstract, useAbstractClient, useGlobalWalletSignerAccount } from '@abstract-foundation/agw-react';
import { windowGuard } from '@orderly.network/utils';
import { WalletAdapterNetwork } from '@solana/wallet-adapter-base';
import { PrivyProvider, usePrivy, useWallets, useSolanaWallets } from '@privy-io/react-auth';
import { WalletProvider, useWallet } from '@solana/wallet-adapter-react';
import { create } from 'zustand';
import { AuthGuard } from '@orderly.network/ui-connector';
import { PhantomWalletAdapter } from '@solana/wallet-adapter-wallets';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

// src/index.ts
var Network = /* @__PURE__ */ ((Network3) => {
  Network3["mainnet"] = "mainnet";
  Network3["testnet"] = "testnet";
  return Network3;
})(Network || {});
var WalletType = /* @__PURE__ */ ((WalletType4) => {
  WalletType4["EVM"] = "EVM";
  WalletType4["SOL"] = "SOL";
  WalletType4["ABSTRACT"] = "Abstract";
  return WalletType4;
})(WalletType || {});
var WalletConnectType = /* @__PURE__ */ ((WalletConnectType2) => {
  WalletConnectType2["EVM"] = "EVM";
  WalletConnectType2["SOL"] = "SOL";
  WalletConnectType2["PRIVY"] = "privy";
  WalletConnectType2["ABSTRACT"] = "Abstract";
  return WalletConnectType2;
})(WalletConnectType || {});
var SolanaChains = /* @__PURE__ */ new Map([
  [WalletAdapterNetwork.Devnet, SOLANA_TESTNET_CHAINID],
  [WalletAdapterNetwork.Mainnet, SOLANA_MAINNET_CHAINID]
]);
var SolanaChainsMap = /* @__PURE__ */ new Map([
  [WalletAdapterNetwork.Devnet, SOLANA_TESTNET_CHAINID],
  ["testnet" /* testnet */, SOLANA_TESTNET_CHAINID],
  ["mainnet" /* mainnet */, SOLANA_MAINNET_CHAINID],
  [WalletAdapterNetwork.Mainnet, SOLANA_MAINNET_CHAINID]
]);
var AbstractChainsMap = /* @__PURE__ */ new Map([
  ["testnet" /* testnet */, ABSTRACT_TESTNET_CHAINID],
  ["mainnet" /* mainnet */, ABSTRACT_MAINNET_CHAINID]
]);
var WalletChainTypeEnum = /* @__PURE__ */ ((WalletChainTypeEnum2) => {
  WalletChainTypeEnum2["onlyEVM"] = "onlyEVM";
  WalletChainTypeEnum2["onlySOL"] = "onlySOL";
  WalletChainTypeEnum2["EVM_SOL"] = "EVM_SOL";
  return WalletChainTypeEnum2;
})(WalletChainTypeEnum || {});

// src/providers/abstractWallet/abstractWalletProvider.tsx
var AbstractWalletContext = createContext(
  null
);
var AbstractWalletProvider = (props) => {
  const { network } = useWalletConnectorPrivy();
  const { login, logout } = useLoginWithAbstract();
  const [wallet, setWallet] = useState(null);
  const { data: client, isLoading: isClientLoading } = useAbstractClient();
  const { connector } = useAccount$1();
  const { address, status } = useGlobalWalletSignerAccount();
  const connect = () => {
    return login();
  };
  const disconnect = () => {
    return logout();
  };
  const isConnected = useMemo(() => {
    return !!(client && connector);
  }, [client, connector]);
  const connectedChain = useMemo(() => {
    if (!client || !connector) {
      return;
    }
    return {
      id: AbstractChainsMap.get(network),
      namespace: ChainNamespace.evm
    };
  }, [client, connector]);
  const value = useMemo(
    () => ({
      isConnected,
      connect,
      disconnect,
      wallet,
      connectedChain
    }),
    [connect, disconnect, isConnected, wallet, connectedChain]
  );
  useEffect(() => {
    if (!client || !connector || !address) {
      setWallet(null);
      return;
    }
    connector?.getProvider().then((provider) => {
      const tempWallet = {
        label: "AGW",
        icon: "",
        provider: {
          ...provider,
          agwWallet: true,
          sendTransaction: async (params) => {
            return client.sendTransaction(params);
          },
          writeContract: async (params) => {
            return client.writeContract(params);
          }
        },
        accounts: [
          {
            address
          }
        ],
        chains: [
          {
            id: AbstractChainsMap.get(network),
            namespace: ChainNamespace.evm
          }
        ],
        chain: connectedChain,
        additionalInfo: {
          AGWAddress: client.account.address
        }
      };
      setWallet(tempWallet);
    });
  }, [client, connectedChain, connector, address, isClientLoading]);
  useEffect(() => {
    windowGuard(() => {
      const connection = localStorage.getItem(
        "privy-caw:cm04asygd041fmry9zmcyn5o5:connection"
      );
      if (connection) {
        login();
      }
    });
  }, []);
  return /* @__PURE__ */ React19.createElement(AbstractWalletContext.Provider, { value }, props.children);
};
function useAbstractWallet() {
  const context = useContext(AbstractWalletContext);
  if (!context) {
    throw new Error(
      "useAbstractWallet must be used within a AbstractWalletProvider"
    );
  }
  return context;
}
var getPrivyEmbeddedWalletChainId = (chainId) => {
  if (!chainId) {
    return null;
  }
  return parseInt(chainId.split("eip155:")[1]);
};
var defaultUseSolanaWallets = {
  ready: false,
  wallets: [],
  createWallet: () => Promise.resolve(),
  exportWallet: () => Promise.resolve()
};
var defaultUseWallets = { wallets: [] };
var PrivyWalletContext = createContext(null);
var PrivyWalletProvider = ({
  children
}) => {
  const { network, solanaInfo, setSolanaInfo, connectorWalletType } = useWalletConnectorPrivy();
  const {
    login,
    logout,
    ready,
    authenticated,
    user,
    exportWallet: exportEvmWallet,
    createWallet: createEvmWallet
  } = usePrivy();
  const { wallets: walletsEVM } = connectorWalletType.disablePrivy ? defaultUseWallets : useWallets();
  const connectedRef = useRef(false);
  const {
    ready: solanaReady,
    wallets: walletsSOL,
    createWallet: createSolanaWallet,
    exportWallet: exportSolanaWallet
  } = connectorWalletType.disablePrivy ? defaultUseSolanaWallets : useSolanaWallets();
  const [walletEVM, setWalletEVM] = useState(null);
  const [walletSOL, setWalletSOL] = useState(null);
  const { track } = useTrack();
  const linkedAccount = useMemo(() => {
    if (user && user.linkedAccounts) {
      const account = user.linkedAccounts.filter((item) => item.type !== "wallet").sort(
        (a, b) => (b.latestVerifiedAt?.getTime() ?? 0) - (a.latestVerifiedAt?.getTime() ?? 0)
      )[0];
      let address = null;
      if (account.type === "email") {
        address = account.address;
      } else if (account.type === "twitter_oauth") {
        address = `@${account.username}`;
      } else if (account.type === "google_oauth") {
        address = `@${account.name}`;
      } else if (account.type === "telegram") {
        address = `@${account.username}`;
      }
      return {
        type: account.type,
        address
      };
    }
    return null;
  }, [user]);
  const switchChain = (chainId) => {
    const wallet = walletsEVM[0];
    if (wallet) {
      return wallet.switchChain(chainId);
    }
    return Promise.reject("no wallet");
  };
  const connect = () => {
    login();
  };
  const disconnect = () => {
    return logout();
  };
  const exportWallet = (namespace) => {
    if (namespace === ChainNamespace.evm) {
      track(TrackerEventName.clickExportPrivateKey, {
        type: "evm"
      });
      return exportEvmWallet();
    } else if (namespace === ChainNamespace.solana) {
      track(TrackerEventName.clickExportPrivateKey, {
        type: "solana"
      });
      return exportSolanaWallet();
    }
    return Promise.reject("no namespace");
  };
  const isConnected = useMemo(() => {
    if (ready && authenticated) {
      return true;
    }
    return false;
  }, [ready, authenticated]);
  useEffect(() => {
    if (!authenticated || !walletsEVM || !walletsEVM[0]) {
      setWalletEVM(null);
      return;
    }
    const wallet = walletsEVM[0];
    wallet.getEthereumProvider().then((provider) => {
      setWalletEVM({
        label: "privy",
        icon: "",
        provider,
        accounts: [
          {
            address: wallet.address
          }
        ],
        chains: [
          {
            id: getPrivyEmbeddedWalletChainId(wallet.chainId) ?? 1,
            namespace: ChainNamespace.evm
          }
        ],
        chain: {
          id: getPrivyEmbeddedWalletChainId(wallet.chainId) ?? 1,
          namespace: ChainNamespace.evm
        }
      });
    });
  }, [walletsEVM, authenticated]);
  useEffect(() => {
    if (!authenticated) {
      setWalletSOL(null);
      return;
    }
    if (!solanaReady) {
      return;
    }
    if (!user) {
      return;
    }
    const embededSolanaWallet = (user?.linkedAccounts).find(
      (item) => item.chainType === "solana" && item.connectorType === "embedded"
    );
    if (!embededSolanaWallet) {
      createSolanaWallet().then();
      return;
    }
    if (!walletsSOL || !walletsSOL[0]) {
      return;
    }
    const wallet = walletsSOL.find((w) => w.connectorType === "embedded");
    if (wallet) {
      if (walletSOL && wallet.address === walletSOL.accounts[0].address) {
        if (walletSOL.chain.id === SolanaChainsMap.get(network)) {
          return;
        }
      }
      setWalletSOL({
        label: "privy",
        icon: "",
        provider: {
          signMessage: wallet.signMessage,
          signTransaction: wallet.signTransaction,
          sendTransaction: wallet.sendTransaction,
          network: solanaInfo?.network ?? WalletAdapterNetwork.Devnet,
          rpcUrl: solanaInfo?.rpcUrl ?? void 0
        },
        accounts: [
          {
            address: wallet.address
          }
        ],
        chains: [
          {
            id: SolanaChainsMap.get(network),
            namespace: ChainNamespace.solana
          }
        ],
        chain: {
          id: SolanaChainsMap.get(network),
          namespace: ChainNamespace.solana
        }
      });
    }
  }, [
    walletsSOL,
    authenticated,
    createSolanaWallet,
    solanaReady,
    user,
    walletSOL,
    network,
    solanaInfo
  ]);
  useEffect(() => {
    if (isConnected && linkedAccount) {
      if (connectedRef.current) {
        return;
      }
      connectedRef.current = true;
      track(TrackerEventName.socialLoginSuccess, {
        type: linkedAccount.type,
        address: linkedAccount.address
      });
    }
  }, [isConnected, linkedAccount, connectedRef]);
  const value = useMemo(
    () => ({
      connect,
      walletEVM,
      walletSOL,
      isConnected,
      disconnect,
      switchChain,
      linkedAccount,
      exportWallet,
      createEvmWallet,
      createSolanaWallet
    }),
    [
      connect,
      walletEVM,
      walletSOL,
      isConnected,
      disconnect,
      switchChain,
      linkedAccount,
      exportWallet,
      createEvmWallet,
      createSolanaWallet
    ]
  );
  return /* @__PURE__ */ React19.createElement(PrivyWalletContext.Provider, { value }, children);
};
function usePrivyWallet() {
  const context = useContext(PrivyWalletContext);
  if (!context) {
    throw new Error("usePrivyWallet must be used within a PrivyWalletProvider");
  }
  return context;
}
var useSolanaWalletStore = create((set, get) => ({
  wallet: null,
  isConnecting: false,
  error: null,
  isManual: false,
  walletMethods: null,
  pendingWalletName: null,
  connect: async (walletName) => {
    const { setIsManual, walletMethods } = get();
    if (!walletMethods) {
      throw new Error("Wallet methods not initialized");
    }
    setIsManual(true);
    set({ pendingWalletName: walletName });
    try {
      set({ isConnecting: true, error: null });
      const { walletSolana, select } = walletMethods;
      select(walletName);
      await new Promise((resolve) => {
        const checkWallet = () => {
          const currentWallet = get().walletMethods?.walletSolana;
          if (currentWallet) {
            resolve();
          } else {
            setTimeout(checkWallet, 500);
          }
        };
        checkWallet();
      });
      const updatedWalletMethods = get().walletMethods;
      if (!updatedWalletMethods) {
        throw new Error("Wallet methods not initialized");
      }
      if (!updatedWalletMethods.publicKey) {
        await updatedWalletMethods.connectSolana();
      }
      const lastestWalletMethods = get().walletMethods;
      if (!lastestWalletMethods) {
        throw new Error("Wallet methods not initialized");
      }
      if (!lastestWalletMethods.publicKey) {
        throw new Error("Wallet not connected");
      }
      const wallet = {
        label: updatedWalletMethods.walletSolana.adapter.name,
        icon: "",
        provider: {
          rpcUrl: lastestWalletMethods.solanaInfo?.rpcUrl ?? null,
          network: lastestWalletMethods.solanaInfo?.network ?? WalletAdapterNetwork.Devnet,
          signMessage: lastestWalletMethods.signMessage,
          signTransaction: lastestWalletMethods.signTransaction,
          sendTransaction: lastestWalletMethods.sendTransaction
        },
        accounts: [
          {
            address: lastestWalletMethods.publicKey.toBase58()
          }
        ],
        chains: [
          {
            id: SolanaChainsMap.get(lastestWalletMethods.network),
            namespace: ChainNamespace.solana
          }
        ],
        chain: {
          id: SolanaChainsMap.get(lastestWalletMethods.network),
          namespace: ChainNamespace.solana
        }
      };
      set({
        wallet,
        isConnecting: false,
        isManual: false,
        pendingWalletName: null
      });
      return wallet;
    } catch (error) {
      set({
        error,
        isConnecting: false,
        isManual: false,
        pendingWalletName: null
      });
      throw error;
    }
  },
  disconnect: async () => {
    const { walletMethods } = get();
    if (!walletMethods) {
      throw new Error("Wallet methods not initialized");
    }
    try {
      await walletMethods.disconnectSolana();
      set({ wallet: null, error: null });
    } catch (error) {
      set({ error });
      throw error;
    }
  },
  setWallet: (wallet) => set({ wallet }),
  setError: (error) => set({ error }),
  setIsManual: (isManual) => set({ isManual }),
  setWalletMethods: (methods) => set({ walletMethods: methods }),
  setPendingWalletName: (name) => set({ pendingWalletName: name })
}));

// src/providers/solana/solanaWalletProvider.tsx
var defaultUseSolanaWallet = {
  wallets: [],
  select: () => Promise.resolve(),
  connect: () => Promise.resolve(),
  wallet: null,
  publicKey: null,
  connecting: false,
  signMessage: () => Promise.resolve(),
  signTransaction: () => Promise.resolve(),
  sendTransaction: () => Promise.resolve(),
  disconnect: () => Promise.resolve()
};
var SolanaWalletContext = createContext(
  null
);
var SolanaWalletProvider = ({
  children
}) => {
  const { setLedgerAddress } = useStorageLedgerAddress();
  const { network, solanaInfo, connectorWalletType } = useWalletConnectorPrivy();
  const {
    wallets,
    select,
    connect: connectSolana,
    wallet: walletSolana,
    publicKey,
    connecting,
    signMessage,
    signTransaction,
    sendTransaction,
    disconnect: disconnectSolana
  } = connectorWalletType.disableSolana ? defaultUseSolanaWallet : useWallet();
  const {
    wallet,
    connect,
    disconnect: disconnectWallet,
    setWallet,
    setWalletMethods,
    isManual
  } = useSolanaWalletStore();
  useEffect(() => {
    if (!walletSolana || isManual) return;
    if (!publicKey) {
      return;
    }
    const newWallet = {
      label: walletSolana.adapter.name,
      icon: "",
      provider: {
        rpcUrl: solanaInfo?.rpcUrl ?? null,
        network: solanaInfo?.network ?? WalletAdapterNetwork.Devnet,
        signMessage,
        signTransaction,
        sendTransaction
      },
      accounts: [
        {
          address: publicKey.toBase58()
        }
      ],
      chains: [
        {
          id: SolanaChainsMap.get(network),
          namespace: ChainNamespace.solana
        }
      ],
      chain: {
        id: SolanaChainsMap.get(network),
        namespace: ChainNamespace.solana
      }
    };
    if (walletSolana.adapter.name === "Ledger") {
      setLedgerAddress(publicKey.toBase58());
    }
    setWallet(newWallet);
  }, [
    publicKey,
    walletSolana,
    signMessage,
    signTransaction,
    sendTransaction,
    solanaInfo,
    isManual
  ]);
  useEffect(() => {
    setWalletMethods({
      select,
      connectSolana,
      walletSolana,
      publicKey,
      connecting,
      signMessage,
      signTransaction,
      sendTransaction,
      disconnectSolana,
      network,
      solanaInfo
    });
  }, [
    select,
    connectSolana,
    walletSolana,
    publicKey,
    connecting,
    signMessage,
    signTransaction,
    sendTransaction,
    disconnectSolana,
    network,
    solanaInfo
  ]);
  const dedupedWallets = useMemo(() => {
    const seen = /* @__PURE__ */ new Set();
    return wallets.filter((w) => {
      const name = w?.adapter?.name?.toLowerCase?.();
      if (!name || seen.has(name)) return !name;
      seen.add(name);
      return true;
    });
  }, [wallets]);
  const value = useMemo(
    () => ({
      wallets: dedupedWallets,
      connectedChain: publicKey ? {
        id: SolanaChainsMap.get(network),
        namespace: ChainNamespace.solana
      } : null,
      connect,
      wallet,
      disconnect: disconnectWallet,
      isConnected: !!publicKey
    }),
    [dedupedWallets, publicKey, wallet, network]
  );
  return /* @__PURE__ */ React19.createElement(SolanaWalletContext.Provider, { value }, children);
};
function useSolanaWallet() {
  const context = useContext(SolanaWalletContext);
  if (!context) {
    throw new Error(
      "useSolanaWallet must be used within a SolanaWalletProvider"
    );
  }
  return context;
}
var WagmiWalletContext = createContext(null);
var WagmiWalletProvider = ({
  children
}) => {
  const { connectorWalletType } = useWalletConnectorPrivy();
  const [wallet, setWallet] = useState(void 0);
  const { connect, connectors: wagmiConnectors2 } = connectorWalletType.disableWagmi ? { connect: () => Promise.resolve(), connectors: [] } : useConnect();
  const { disconnect } = connectorWalletType.disableWagmi ? { disconnect: () => Promise.resolve() } : useDisconnect();
  const { connector, isConnected, address, chainId } = connectorWalletType.disableWagmi ? { connector: null, isConnected: false, address: null, chainId: null } : useAccount$1();
  const { switchChain } = connectorWalletType.disableWagmi ? { switchChain: () => Promise.resolve() } : useSwitchChain();
  const connectedChain = useMemo(() => {
    if (chainId) {
      return {
        id: chainId,
        namespace: ChainNamespace.evm
      };
    }
    return null;
  }, [chainId]);
  const setChain = useCallback(
    (chainId2) => {
      return new Promise((resolve, reject) => {
        switchChain(
          { chainId: chainId2 },
          {
            onSuccess: () => resolve(true),
            onError: (e) => {
              return reject(e);
            }
          }
        );
      });
    },
    [switchChain]
  );
  useEffect(() => {
    if (!connector || !isConnected) {
      setWallet(void 0);
      return;
    }
    connector.getProvider?.().then((provider) => {
      setWallet({
        label: connector.name,
        icon: "",
        provider,
        accounts: [
          {
            address
          }
        ],
        chains: [
          {
            id: chainId,
            namespace: ChainNamespace.evm
          }
        ],
        chain: connectedChain
      });
    });
  }, [connector, chainId, isConnected, address, connectedChain]);
  const connectors = useMemo(() => {
    return wagmiConnectors2.filter((connector2) => connector2.id !== "injected").sort(
      (a, b) => a.type === "injected" ? -1 : 1
    );
  }, [wagmiConnectors2]);
  const value = useMemo(
    () => ({
      connectors,
      connect,
      wallet,
      connectedChain,
      setChain,
      disconnect,
      isConnected
    }),
    [
      connectors,
      connect,
      wallet,
      connectedChain,
      setChain,
      disconnect,
      isConnected
    ]
  );
  return /* @__PURE__ */ React19.createElement(WagmiWalletContext.Provider, { value }, children);
};
function useWagmiWallet() {
  const context = useContext(WagmiWalletContext);
  if (!context) {
    throw new Error("useWagmiWallet must be used within a WagmiWalletProvider");
  }
  return context;
}
var MoreIcon = (props) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: "16",
    height: "16",
    viewBox: "0 0 16 16",
    fill: "currentColor",
    ...props
  },
  /* @__PURE__ */ React19.createElement("path", { d: "M8.00521 6.66797C8.74161 6.66797 9.33854 7.26464 9.33854 8.0013C9.33854 8.73797 8.74161 9.33464 8.00521 9.33464C7.26881 9.33464 6.67188 8.73797 6.67188 8.0013C6.67188 7.26464 7.26881 6.66797 8.00521 6.66797Z" }),
  /* @__PURE__ */ React19.createElement("path", { d: "M3.33333 6.66797C4.06973 6.66797 4.66667 7.26464 4.66667 8.0013C4.66667 8.73797 4.06973 9.33464 3.33333 9.33464C2.59693 9.33464 2 8.73797 2 8.0013C2 7.26464 2.59693 6.66797 3.33333 6.66797Z" }),
  /* @__PURE__ */ React19.createElement("path", { d: "M12.6666 6.66797C13.403 6.66797 14 7.26464 14 8.0013C14 8.73797 13.403 9.33464 12.6666 9.33464C11.9302 9.33464 11.3333 8.73797 11.3333 8.0013C11.3333 7.26464 11.9302 6.66797 12.6666 6.66797Z" })
);
var DisconnectIcon = (props) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    width: "16",
    height: "16",
    viewBox: "0 0 16 16",
    xmlns: "http://www.w3.org/2000/svg",
    fill: "currentColor",
    ...props
  },
  /* @__PURE__ */ React19.createElement("g", { id: "box-arrow-right" }, /* @__PURE__ */ React19.createElement(
    "path",
    {
      id: "Vector",
      d: "M1.99365 4.66056C1.99365 3.1879 3.18759 1.9939 4.66032 1.9939H7.32698C8.79965 1.9939 9.99365 3.1879 9.99365 4.66056C9.99365 5.02856 9.69498 5.32723 9.32698 5.32723C8.95898 5.32723 8.66032 5.02856 8.66032 4.66056C8.66032 3.92389 8.06365 3.32723 7.32698 3.32723H4.66032C3.92392 3.32723 3.32699 3.92389 3.32699 4.66056V11.3272C3.32699 12.0639 3.92392 12.6606 4.66032 12.6606H7.32698C8.06365 12.6606 8.66032 12.0639 8.66032 11.3272C8.66032 10.9592 8.95898 10.6606 9.32698 10.6606C9.69498 10.6606 9.99365 10.9592 9.99365 11.3272C9.99365 12.7999 8.79965 13.9939 7.32698 13.9939H4.66032C3.18759 13.9939 1.99365 12.7999 1.99365 11.3272V4.66056ZM6.66032 7.99389C6.66032 7.62589 6.95898 7.32722 7.32698 7.32722H12.369L11.0563 5.99389L11.9937 5.05656L14.473 7.51456C14.733 7.77523 14.733 8.21255 14.473 8.47322L11.9937 10.9312L11.0563 9.99389L12.369 8.66056H7.32698C6.95898 8.66056 6.66032 8.36189 6.66032 7.99389Z"
    }
  ))
);
var ArrowRightIcon = ({ size, ...props }) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    width: size,
    height: size,
    viewBox: "0 0 14 14",
    fill: "currentColor",
    xmlns: "http://www.w3.org/2000/svg",
    ...props
  },
  /* @__PURE__ */ React19.createElement("path", { d: "M3.50696 6.9958C3.50696 6.6738 3.76788 6.41305 4.08971 6.41305C4.36842 6.41305 7.62762 6.41305 8.49679 6.41305L6.76662 4.6648L7.58621 3.84521L10.336 6.57697C10.4504 6.69072 10.507 6.84355 10.507 6.99639C10.507 7.14864 10.4498 7.30089 10.336 7.41464L7.58621 10.1464L6.76662 9.3268L8.49679 7.57854C7.62762 7.57854 4.36842 7.57854 4.08971 7.57854C3.76788 7.57854 3.50696 7.3178 3.50696 6.9958Z" })
);
var ArrowRightLinearGradientIcon = ({
  size,
  ...props
}) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: size,
    height: size,
    viewBox: "0 0 20 20",
    fill: "none"
  },
  /* @__PURE__ */ React19.createElement(
    "path",
    {
      d: "M5.01001 9.994C5.01001 9.534 5.38276 9.1615 5.84251 9.1615C6.24068 9.1615 10.8967 9.1615 12.1383 9.1615L9.66667 6.664L10.8375 5.49316L14.7658 9.39568C14.9292 9.55818 15.01 9.7765 15.01 9.99484C15.01 10.2123 14.9283 10.4298 14.7658 10.5923L10.8375 14.4948L9.66667 13.324L12.1383 10.8265C10.8967 10.8265 6.24068 10.8265 5.84251 10.8265C5.38276 10.8265 5.01001 10.454 5.01001 9.994Z",
      fill: "url(#paint0_linear_2269_12077)"
    }
  ),
  /* @__PURE__ */ React19.createElement("defs", null, /* @__PURE__ */ React19.createElement(
    "linearGradient",
    {
      id: "paint0_linear_2269_12077",
      x1: "15.01",
      y1: "9.994",
      x2: "5.01001",
      y2: "9.994",
      gradientUnits: "userSpaceOnUse"
    },
    /* @__PURE__ */ React19.createElement("stop", { stopColor: "#59B0FE" }),
    /* @__PURE__ */ React19.createElement("stop", { offset: "1", stopColor: "#26FEFE" })
  ))
);
var PwaDialog = modal.create((props) => {
  const { visible, hide, resolve, reject, onOpenChange } = useModal();
  const [domain, setDomain] = useState("");
  const { t } = useTranslation();
  useEffect(() => {
    setDomain(window.location.hostname);
  }, []);
  return /* @__PURE__ */ React19.createElement(SimpleSheet, { open: visible, onOpenChange }, /* @__PURE__ */ React19.createElement(SheetHeader, null, /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-base-contrast-80 oui-text " }, t("connector.privy.pwa.sheetTitle"))), /* @__PURE__ */ React19.createElement(Flex, { direction: "column", gap: 3, className: "oui-w-full" }, /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-base-contrast-80 oui-text " }, t("connector.privy.pwa.description")), /* @__PURE__ */ React19.createElement(Flex, { direction: "column", gap: 4, className: "oui-w-full" }, /* @__PURE__ */ React19.createElement(
    StepItem,
    {
      index: 1,
      title: t("connector.privy.pwa.step1"),
      content: /* @__PURE__ */ React19.createElement(Flex, { className: "oui-w-full oui-bg-dark-8" }, /* @__PURE__ */ React19.createElement(
        Flex,
        {
          className: "oui-bg-black/[0.08] oui-p-2 oui-h-[52px] oui-rounded-[44px] oui-w-full oui-px-4 oui-relative",
          itemAlign: "center",
          justify: "center"
        },
        /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-black/[0.88] oui-text-[18px] oui-leading-[36px] oui-flex-1 oui-text-center oui-whitespace-nowrap" }, domain),
        /* @__PURE__ */ React19.createElement(ShareIcon, { className: "oui-absolute oui-right-4 oui-top-1/2 oui-translate-y-[-50%]" })
      ))
    }
  ), /* @__PURE__ */ React19.createElement(
    StepItem,
    {
      index: 2,
      title: t("connector.privy.pwa.step2"),
      classNames: {
        content: "oui-bg-[#e9e9eb]"
      },
      content: /* @__PURE__ */ React19.createElement(
        Flex,
        {
          className: "oui-w-full oui-bg-base-contrast oui-rounded-[12px]",
          direction: "column"
        },
        /* @__PURE__ */ React19.createElement(
          Flex,
          {
            className: "oui-w-full oui-px-4 oui-py-3",
            justify: "start",
            itemAlign: "center"
          },
          /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-black/[0.2] oui-text-[18px]" }, t("connector.privy.pwa.findOnPage"))
        ),
        /* @__PURE__ */ React19.createElement(Divider, { className: "oui-bg-black/[0.15] oui-w-full" }),
        /* @__PURE__ */ React19.createElement(
          Flex,
          {
            className: "oui-w-full oui-px-4 oui-py-3",
            justify: "between",
            itemAlign: "center"
          },
          /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-black/[0.88] oui-text-[18px] oui-col-span-2 oui-whitespace-nowrap" }, t("connector.privy.pwa.addToHomeScreen")),
          /* @__PURE__ */ React19.createElement(AddIcon, null)
        )
      )
    }
  ))));
});
var ShareIcon = (props) => {
  return /* @__PURE__ */ React19.createElement(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "28",
      height: "28",
      viewBox: "0 0 28 28",
      fill: "none",
      className: props.className
    },
    /* @__PURE__ */ React19.createElement(
      "path",
      {
        d: "M5 23.3333V13.6667C5 12.1939 6.15127 11 7.57143 11H10.6744V12.7778H7.57143C7.09804 12.7778 6.71429 13.1757 6.71429 13.6667V23.3333C6.71429 23.8243 7.09804 24.2222 7.57143 24.2222H20.4286C20.902 24.2222 21.2857 23.8243 21.2857 23.3333V13.6667C21.2857 13.1757 20.902 12.7778 20.4286 12.7778H17.555V11H20.4286C21.8487 11 23 12.1939 23 13.6667V23.3333C23 24.8061 21.8487 26 20.4286 26H7.57143C6.15127 26 5 24.8061 5 23.3333Z",
        fill: "black",
        fillOpacity: "0.48"
      }
    ),
    /* @__PURE__ */ React19.createElement(
      "path",
      {
        d: "M13.3759 2.25933C13.7206 1.91356 14.2794 1.91356 14.6241 2.25933L20.2416 7.89432C20.586 8.24011 20.5862 8.80066 20.2416 9.14634C19.897 9.49203 19.3382 9.49185 18.9935 9.14634L14.8827 5.02278V19H13.1173V5.02278L9.00654 9.14634C8.66182 9.49185 8.10301 9.49203 7.7584 9.14634C7.41378 8.80066 7.41396 8.24011 7.7584 7.89432L13.3759 2.25933Z",
        fill: "black",
        fillOpacity: "0.48"
      }
    )
  );
};
var AddIcon = () => {
  return /* @__PURE__ */ React19.createElement(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "24",
      height: "24",
      viewBox: "0 0 24 24",
      fill: "none"
    },
    /* @__PURE__ */ React19.createElement(
      "path",
      {
        d: "M7.00598 3.92969H17.006C18.6627 3.92969 20.006 5.27297 20.006 6.92969V16.9297C20.006 18.5864 18.6627 19.9297 17.006 19.9297H7.00598C5.34927 19.9297 4.00598 18.5864 4.00598 16.9297V6.92969C4.00598 5.27297 5.34927 3.92969 7.00598 3.92969Z",
        stroke: "black",
        strokeOpacity: "0.88",
        strokeWidth: "2"
      }
    ),
    /* @__PURE__ */ React19.createElement(
      "path",
      {
        d: "M8 12L16 12",
        stroke: "black",
        strokeOpacity: "0.88",
        strokeWidth: "2",
        strokeLinecap: "round"
      }
    ),
    /* @__PURE__ */ React19.createElement(
      "path",
      {
        d: "M12 8L12 16",
        stroke: "black",
        strokeOpacity: "0.88",
        strokeWidth: "2",
        strokeLinecap: "round"
      }
    )
  );
};
var StepItem = (props) => {
  return /* @__PURE__ */ React19.createElement(
    Flex,
    {
      className: " oui-rounded-[12px] oui-w-full oui-overflow-hidden",
      direction: "column"
    },
    /* @__PURE__ */ React19.createElement(Flex, { className: "oui-bg-base-4 oui-w-full oui-px-3 oui-py-3 oui-gap-[6px]" }, /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-base-contrast-80 oui-text-xs oui-w-[18px] oui-h-[18px] oui-rounded-[100%] oui-bg-[linear-gradient(270deg,#59B0FE_0%,#26FEFE_100%)] oui-text-center oui-text-black " }, props.index), /* @__PURE__ */ React19.createElement(Text, { className: "oui-text-base-contrast-80 oui-text-xs" }, props.title)),
    /* @__PURE__ */ React19.createElement(
      Flex,
      {
        className: cn(
          "oui-bg-base-contrast oui-h-[120px] oui-w-full oui-px-6",
          props.classNames?.content
        ),
        justify: "center",
        itemAlign: "center"
      },
      props.content
    )
  );
};
var OrderlyOSS = "https://oss.orderly.network";
var PrivyConnectorImagePath = `${OrderlyOSS}/static/sdk/privy`;
var WALLET_ICONS = {
  metamask: "https://oss.orderly.network/static/sdk/evm_wallets/metamask.png",
  walletconnect: "https://oss.orderly.network/static/sdk/evm_wallets/walletConnect.png",
  binance: "https://oss.orderly.network/static/sdk/evm_wallets/binance.png",
  ledger: "https://oss.orderly.network/static/sdk/evm_wallets/ledger.png",
  coinbase: "https://oss.orderly.network/static/sdk/evm_wallets/coinbase.png",
  magic: "https://oss.orderly.network/static/sdk/evm_wallets/magic.png",
  trezor: "https://oss.orderly.network/static/sdk/evm_wallets/trezor.png",
  exodus: "https://oss.orderly.network/static/sdk/evm_wallets/exodus.png",
  frame: "https://oss.orderly.network/static/sdk/evm_wallets/frame.png",
  rabby: "https://oss.orderly.network/static/sdk/evm_wallets/rabby.png"
};
var getWalletIcon = (type) => {
  return WALLET_ICONS[type.toLowerCase()];
};
var getChainType = (chainId) => {
  if (AbstractChains.has(chainId)) {
    return "Abstract" /* ABSTRACT */;
  }
  if (SolanaChains$1.has(chainId)) {
    return "SOL" /* SOL */;
  }
  return "EVM" /* EVM */;
};

// src/hooks/useWallet.tsx
function useWallet2() {
  const { track } = useTrack();
  const ee = useEventEmitter();
  const { walletChainTypeConfig, initChains, network } = useWalletConnectorPrivy();
  const [connectorKey, setConnectorKey] = useLocalStorage(ConnectorKey, "");
  const {
    disconnect: disconnectEVM,
    connect: connectEVM,
    wallet: walletEVM,
    connectedChain: connectedChainEvm,
    setChain: setChainEvm,
    isConnected: isConnectedEVM
  } = useWagmiWallet();
  const {
    disconnect: disconnectSOL,
    connect: connectSOL,
    wallet: walletSOL,
    isConnected: isConnectedSOL,
    connectedChain: connectedChainSOL
  } = useSolanaWallet();
  const {
    disconnect: disconnectPrivy,
    connect: connectPrivy,
    walletSOL: privyWalletSOL,
    walletEVM: privyWalletEVM,
    isConnected: isConnectedPrivy,
    switchChain: setChainPrivy
  } = usePrivyWallet();
  const {
    connect: connectAbstract,
    wallet: walletAbstract,
    connectedChain: connectedChainAbstract,
    isConnected: isConnectedAbstract,
    disconnect: disconnectAbstract
  } = useAbstractWallet();
  const [wallet, setWallet] = useState(null);
  const [namespace, setNamespace] = useState(null);
  const [walletType, setWalletType] = useState(null);
  const { storageChain, setStorageChain } = useStorageChain();
  const { setOpenConnectDrawer, targetWalletType, setTargetWalletType } = useWalletConnectorPrivy();
  const supportedEvmChainIds = useMemo(() => {
    const ids = initChains?.map((c) => c.id).filter((id) => !SolanaChains$1.has(id) && !AbstractChains.has(id));
    return new Set(ids ?? []);
  }, [initChains]);
  const preferredEvmChainId = useMemo(() => {
    const preferredOrder = network === "mainnet" ? defaultMainnetChains : defaultTestnetChains;
    return preferredOrder.map((c) => c.id).find((id) => supportedEvmChainIds.has(id));
  }, [network, supportedEvmChainIds]);
  const defaultEvmChainId = useMemo(() => {
    for (const chain of initChains ?? []) {
      if (!SolanaChains$1.has(chain.id) && !AbstractChains.has(chain.id)) {
        return chain.id;
      }
    }
    return void 0;
  }, [initChains]);
  const fallbackEvmChainId = preferredEvmChainId ?? defaultEvmChainId;
  const isManual = useRef(false);
  const connect = (params) => {
    setTargetWalletType(void 0);
    try {
      if (params.walletType === "EVM" /* EVM */) {
        setConnectorKey("EVM" /* EVM */);
        connectEVM({ connector: params.connector });
      }
      if (params.walletType === "SOL" /* SOL */) {
        setConnectorKey("SOL" /* SOL */);
        connectSOL(params.walletAdapter.name).catch((err) => {
          const message = String(err?.message ?? "").toLowerCase();
          if (message === "wallet not connected") {
            return;
          }
          ee.emit("wallet:connect-error", {
            message: err?.message || "Please switch to a wallet with Solana address."
          });
        });
      }
      if (params.walletType === "privy" /* PRIVY */) {
        setConnectorKey("privy" /* PRIVY */);
        connectPrivy();
      }
      if (params.walletType === "Abstract" /* ABSTRACT */) {
        setConnectorKey("Abstract" /* ABSTRACT */);
        connectAbstract();
      }
    } catch (e) {
    }
  };
  const isPrivy = useMemo(() => {
    if (connectorKey === "privy" /* PRIVY */) {
      return true;
    }
    return false;
  }, [connectorKey]);
  const [connectedChain, setConnectedChain] = useState();
  const setChain = async (chain) => {
    const chainType = getChainType(parseInt(chain.chainId));
    if (isPrivy) {
      if (chainType === "EVM" /* EVM */) {
        if (!walletChainTypeConfig.hasEvm) {
          return Promise.reject(new Error("No evm wallet found"));
        }
        isManual.current = true;
        return setChainPrivy(parseInt(chain.chainId)).then((res) => {
          track(TrackerEventName.switchNetworkSuccess, {
            from_chain: storageChain?.chainId,
            to_chain: chain.chainId
          });
          setStorageChain(parseInt(chain.chainId));
          return Promise.resolve(true);
        }).catch((e) => {
          return Promise.reject(e);
        });
      }
      if (chainType === "Abstract" /* ABSTRACT */) {
        setOpenConnectDrawer(true);
        setTargetWalletType("Abstract" /* ABSTRACT */);
      }
      if (chainType === "SOL" /* SOL */) {
        isManual.current = true;
        if (!walletChainTypeConfig.hasSol) {
          return Promise.reject(new Error("No solana wallet found"));
        }
        if (privyWalletSOL) {
          setStorageChain(parseInt(chain.chainId));
          return Promise.resolve(true);
        } else {
          setOpenConnectDrawer(true);
          setTargetWalletType("SOL" /* SOL */);
          return Promise.reject(new Error("No solana wallet found"));
        }
      }
    } else {
      if (chainType === "EVM" /* EVM */) {
        if (isConnectedEVM && walletEVM) {
          await setChainEvm(parseInt(chain.chainId));
          setStorageChain(parseInt(chain.chainId));
          track(TrackerEventName.switchNetworkSuccess, {
            from_chain: storageChain?.chainId,
            to_chain: chain.chainId
          });
        } else {
          setOpenConnectDrawer(true);
          setTargetWalletType("EVM" /* EVM */);
          return Promise.resolve(true);
        }
        return Promise.resolve(true);
      }
      if (chainType === "SOL" /* SOL */) {
        if (isConnectedSOL && walletSOL) {
          setStorageChain(parseInt(chain.chainId));
          return Promise.resolve(true);
        } else {
          setOpenConnectDrawer(true);
          setTargetWalletType("SOL" /* SOL */);
          return Promise.resolve(true);
        }
      }
      if (chainType === "Abstract" /* ABSTRACT */) {
        if (isConnectedAbstract && walletAbstract) {
          setStorageChain(parseInt(chain.chainId));
          return Promise.resolve(true);
        } else {
          setOpenConnectDrawer(true);
          setTargetWalletType("Abstract" /* ABSTRACT */);
          return Promise.reject(new Error("No abstract wallet found"));
        }
      }
    }
  };
  const switchWallet = (walletType2) => {
    const fromWallet = wallet?.accounts[0].address;
    let toWallet;
    if (isPrivy) {
      switch (walletType2) {
        case "EVM" /* EVM */:
          if (privyWalletEVM) {
            const desired = privyWalletEVM.chain.id;
            const nextChainId = supportedEvmChainIds.has(desired) ? desired : fallbackEvmChainId;
            if (typeof nextChainId === "number") {
              setStorageChain(nextChainId);
            }
            toWallet = privyWalletEVM.accounts[0].address;
          }
          break;
        case "SOL" /* SOL */:
          if (privyWalletSOL) {
            setStorageChain(privyWalletSOL.chain.id);
            toWallet = privyWalletSOL.accounts[0].address;
          }
          break;
      }
    } else {
      switch (walletType2) {
        case "EVM" /* EVM */:
          if (walletEVM) {
            setStorageChain(walletEVM.chain.id);
            toWallet = walletEVM.accounts[0].address;
          }
          break;
        case "SOL" /* SOL */:
          if (walletSOL) {
            setStorageChain(walletSOL.chain.id);
            toWallet = walletSOL.accounts[0].address;
          }
          break;
        case "Abstract" /* ABSTRACT */:
          if (walletAbstract) {
            setStorageChain(walletAbstract.chain.id);
            toWallet = walletAbstract.accounts[0].address;
          }
          break;
      }
    }
    track(TrackerEventName.clickSwitchWallet, {
      fromWallet,
      toWallet
    });
  };
  const disconnect = async (walletType2) => {
    switch (walletType2) {
      case "privy" /* PRIVY */:
        return await disconnectPrivy();
      case "EVM" /* EVM */:
        return disconnectEVM();
      case "SOL" /* SOL */:
        return disconnectSOL();
      case "Abstract" /* ABSTRACT */:
        return disconnectAbstract();
    }
  };
  const onDisconnect = (parmas) => {
    if (!walletType) {
      throw new Error("No wallet type found");
    }
    return new Promise((resolve, reject) => {
      disconnect(walletType).then(() => {
        resolve(true);
      }).catch((e) => {
        reject(e);
      });
    });
  };
  const setNullWalletStatus = () => {
    setWallet(null);
    setWalletType(null);
    setConnectedChain(null);
    setNamespace(null);
  };
  useEffect(() => {
    if (connectorKey !== "privy" /* PRIVY */) {
      return;
    }
    if (storageChain?.namespace === ChainNamespace.evm) {
      if (AbstractChains.has(storageChain.chainId)) {
        return setNullWalletStatus();
      }
      if (privyWalletEVM) {
        setWallet(privyWalletEVM);
        setWalletType("privy" /* PRIVY */);
        setConnectedChain(privyWalletEVM.chain);
        setNamespace(ChainNamespace.evm);
      } else {
        setNullWalletStatus();
      }
    }
    if (storageChain?.namespace === ChainNamespace.solana) {
      if (privyWalletSOL) {
        setWallet(privyWalletSOL);
        setWalletType("privy" /* PRIVY */);
        setConnectedChain(privyWalletSOL.chain);
        setNamespace(ChainNamespace.solana);
      } else {
        setNullWalletStatus();
      }
    }
  }, [connectorKey, privyWalletEVM, privyWalletSOL, storageChain]);
  useEffect(() => {
    if (connectorKey === "privy" /* PRIVY */ && isConnectedPrivy && AbstractChains.has(storageChain.chainId)) {
      setOpenConnectDrawer(true);
    }
  }, [connectorKey, isConnectedPrivy, storageChain]);
  useEffect(() => {
    if (connectorKey === "privy" /* PRIVY */) {
      return;
    }
    if (storageChain?.namespace === ChainNamespace.evm) {
      if (AbstractChains.has(storageChain.chainId)) {
        if (isConnectedAbstract && walletAbstract) {
          setWallet(walletAbstract);
          setWalletType("Abstract" /* ABSTRACT */);
          setConnectedChain(connectedChainAbstract);
          setNamespace(ChainNamespace.evm);
        } else {
          setNullWalletStatus();
        }
      } else {
        if (isConnectedEVM && walletEVM) {
          setWallet(walletEVM);
          setWalletType("EVM" /* EVM */);
          setConnectedChain(connectedChainEvm);
          setNamespace(ChainNamespace.evm);
        } else {
          setNullWalletStatus();
        }
      }
    }
    if (storageChain?.namespace === ChainNamespace.solana) {
      if (isConnectedSOL && walletSOL) {
        setWallet(walletSOL);
        setWalletType("SOL" /* SOL */);
        setConnectedChain(connectedChainSOL);
        setNamespace(ChainNamespace.solana);
      } else {
        setNullWalletStatus();
      }
    }
  }, [
    connectorKey,
    storageChain,
    walletEVM,
    walletSOL,
    isConnectedEVM,
    isConnectedSOL,
    isConnectedAbstract,
    walletAbstract,
    connectedChainAbstract
  ]);
  return {
    connect,
    wallet,
    connectedChain,
    setChain,
    namespace,
    switchWallet,
    disconnect,
    onDisconnect
  };
}
function AbstractConnectArea({ connect }) {
  const { isMobile } = useScreen();
  return /* @__PURE__ */ React19.createElement("div", null, /* @__PURE__ */ React19.createElement("div", { className: "oui-mb-2 oui-text-sm oui-font-semibold oui-text-base-contrast-80" }, "Abstract"), /* @__PURE__ */ React19.createElement("div", { className: "oui-grid oui-grid-cols-2 oui-gap-2" }, /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        " oui-flex oui-flex-1 oui-cursor-pointer oui-items-center oui-justify-start oui-gap-1 oui-rounded-[6px]  oui-px-2 oui-py-[11px]",
        isMobile ? "oui-bg-base-5" : "oui-bg-[#07080A]"
      ),
      onClick: () => connect()
    },
    /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-size-[18px] oui-items-center oui-justify-center" }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        className: cn("oui-size-[12px]"),
        src: `${PrivyConnectorImagePath}/abstract.png`,
        alt: "abstract wallet"
      }
    )),
    /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, "Abstract")
  )));
}
var RenderLoginMethodsDom = ({
  connect,
  loginMethods
}) => {
  const { t } = useTranslation();
  const { isDesktop } = useScreen();
  const arr = [];
  if (loginMethods?.includes("email")) {
    arr.push(
      /* @__PURE__ */ React19.createElement(
        "div",
        {
          className: cn(
            "oui-flex oui-cursor-pointer oui-items-center oui-justify-center oui-gap-1 oui-rounded-[6px]  oui-border oui-border-base-contrast-12 oui-px-2 oui-py-[11px]",
            "oui-border-none oui-bg-[#333948]"
          ),
          onClick: () => connect({ walletType: "privy", extraType: "email" })
        },
        /* @__PURE__ */ React19.createElement(
          "img",
          {
            src: "https://oss.orderly.network/static/sdk/privy/email.svg",
            className: "oui-size-[18px]"
          }
        ),
        /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, t("connector.privy.email"))
      )
    );
  }
  if (loginMethods?.includes("google")) {
    arr.push(
      /* @__PURE__ */ React19.createElement(
        "div",
        {
          className: cn(
            "oui-flex oui-cursor-pointer oui-items-center oui-justify-center oui-gap-1 oui-rounded-[6px]  oui-border oui-border-base-contrast-12 oui-px-2 oui-py-[11px]",
            "oui-border-none oui-bg-[#335FFC]"
          ),
          onClick: () => connect({ walletType: "privy", extraType: "google" })
        },
        /* @__PURE__ */ React19.createElement(
          "img",
          {
            src: "https://oss.orderly.network/static/sdk/privy/google.svg",
            className: "oui-size-[18px]"
          }
        ),
        /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, t("connector.privy.google"))
      )
    );
  }
  if (loginMethods?.includes("twitter")) {
    arr.push(
      /* @__PURE__ */ React19.createElement(
        "div",
        {
          className: cn(
            "oui-flex oui-cursor-pointer oui-items-center oui-justify-center oui-gap-1 oui-rounded-[6px]  oui-border oui-border-base-contrast-12 oui-px-2 oui-py-[11px]",
            "oui-border-none oui-bg-[#07080A]"
          ),
          onClick: () => connect({ walletType: "privy", extraType: "twitter" })
        },
        /* @__PURE__ */ React19.createElement(
          "img",
          {
            src: "https://oss.orderly.network/static/sdk/privy/twitter.svg",
            className: "oui-size-[18px]"
          }
        ),
        /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, t("connector.privy.twitter"))
      )
    );
  }
  if (loginMethods?.includes("telegram")) {
    arr.push(
      /* @__PURE__ */ React19.createElement(
        "div",
        {
          className: cn(
            "oui-flex oui-cursor-pointer oui-items-center oui-justify-center oui-gap-1 oui-rounded-[6px]  oui-border oui-border-base-contrast-12 oui-px-2 oui-py-[11px]",
            "oui-border-none oui-bg-[#3D79A9]"
          ),
          onClick: () => connect({ walletType: "privy", extraType: "telegram" })
        },
        /* @__PURE__ */ React19.createElement(
          "svg",
          {
            width: "18",
            height: "18",
            viewBox: "0 0 18 18",
            fill: "none",
            xmlns: "http://www.w3.org/2000/svg"
          },
          /* @__PURE__ */ React19.createElement(
            "path",
            {
              d: "M2.652 8.12916L15.0683 4.04929C15.6468 3.87358 16.1499 4.16929 15.9587 4.91069L13.8458 13.3962C13.6898 13.9961 13.2672 14.1461 12.6836 13.859L9.46384 11.8362L7.90929 13.109C7.73824 13.2547 7.59234 13.379 7.2603 13.379L7.4867 10.5891L13.4584 5.99923C13.72 5.80638 13.398 5.69495 13.0559 5.8878L5.68061 9.8434L2.50107 8.99914C1.81687 8.81058 1.80178 8.41201 2.652 8.12916Z",
              fill: "white"
            }
          )
        ),
        /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, t("connector.privy.telegram"))
      )
    );
  }
  return arr.map((item, index) => {
    return /* @__PURE__ */ React19.createElement("div", { key: index }, item);
  });
};
var PrivyConnectAreaMobile = ({
  connect
}) => {
  const { t } = useTranslation();
  const { isMobile, isDesktop } = useScreen();
  const { connectorWalletType, privyConfig } = useWalletConnectorPrivy();
  const loginMethods = privyConfig.loginMethods;
  return /* @__PURE__ */ React19.createElement(Flex, { direction: "column", className: "oui-w-full" }, /* @__PURE__ */ React19.createElement(Grid, { cols: 2, rows: 2, className: "oui-w-full", gap: 2 }, /* @__PURE__ */ React19.createElement(RenderLoginMethodsDom, { connect, loginMethods })), (!connectorWalletType.disableWagmi || !connectorWalletType.disableSolana) && /* @__PURE__ */ React19.createElement("div", { className: "oui-mt-4 oui-h-px oui-w-full oui-bg-line md:oui-mt-5" }));
};
var PrivyConnectAreaDesktop = ({
  connect
}) => {
  const { t } = useTranslation();
  const { connectorWalletType, privyConfig } = useWalletConnectorPrivy();
  const loginMethods = privyConfig.loginMethods;
  return /* @__PURE__ */ React19.createElement(Flex, { direction: "column", itemAlign: "start", className: "oui-w-full" }, /* @__PURE__ */ React19.createElement(Grid, { cols: 1, gap: 2, className: "oui-w-full" }, /* @__PURE__ */ React19.createElement(RenderLoginMethodsDom, { connect, loginMethods })), (!connectorWalletType.disableWagmi || !connectorWalletType.disableSolana) && /* @__PURE__ */ React19.createElement("div", { className: "oui-mt-4 oui-h-px oui-w-full oui-bg-line md:oui-mt-5" }));
};
function PrivyConnectArea({
  connect
}) {
  const { isDesktop } = useScreen();
  if (isDesktop) {
    return /* @__PURE__ */ React19.createElement(PrivyConnectAreaDesktop, { connect });
  }
  return /* @__PURE__ */ React19.createElement(PrivyConnectAreaMobile, { connect });
}
function RenderPrivyTypeIcon({
  type,
  size,
  black
}) {
  if (type === "email") {
    return /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: `https://oss.orderly.network/static/sdk/privy/email${black ? "-black" : ""}.svg`,
        width: size
      }
    );
  }
  if (type === "google_oauth") {
    return /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: "https://oss.orderly.network/static/sdk/privy/google.svg",
        width: size
      }
    );
  }
  if (type === "twitter_oauth") {
    return /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: `https://oss.orderly.network/static/sdk/privy/twitter${black ? "-black" : ""}.svg`,
        width: size
      }
    );
  }
  if (type === "telegram") {
    return /* @__PURE__ */ React19.createElement(
      "svg",
      {
        width: size,
        height: size,
        viewBox: "0 0 18 18",
        fill: "none",
        xmlns: "http://www.w3.org/2000/svg"
      },
      /* @__PURE__ */ React19.createElement(
        "path",
        {
          d: "M2.652 8.12916L15.0683 4.04929C15.6468 3.87358 16.1499 4.16929 15.9587 4.91069L13.8458 13.3962C13.6898 13.9961 13.2672 14.1461 12.6836 13.859L9.46384 11.8362L7.90929 13.109C7.73824 13.2547 7.59234 13.379 7.2603 13.379L7.4867 10.5891L13.4584 5.99923C13.72 5.80638 13.398 5.69495 13.0559 5.8878L5.68061 9.8434L2.50107 8.99914C1.81687 8.81058 1.80178 8.41201 2.652 8.12916Z",
          fill: black ? "black" : "white"
        }
      )
    );
  }
  return /* @__PURE__ */ React19.createElement(
    "img",
    {
      src: `https://oss.orderly.network/static/sdk/privy/email${black ? "-black" : ""}.svg`,
      width: size
    }
  );
}
function RenderWalletIcon({
  connector
}) {
  const icon = connector.icon ? connector.icon : getWalletIcon(connector.type);
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-w-[18px] oui-h-[18px] oui-flex oui-items-center oui-justify-center" }, /* @__PURE__ */ React19.createElement(
    "img",
    {
      className: cn(
        connector.icon ? "oui-w-[12px] oui-h-[12px]" : "oui-w-[18px] oui-h-[18px]"
      ),
      src: icon,
      alt: connector.name
    }
  ));
}
function EVMChainPopover({ children }) {
  const { getChainsByNetwork } = useWalletConnectorPrivy();
  const [chains] = useState(getChainsByNetwork("mainnet"));
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  const [open, setOpen] = useState(false);
  if (isMobile) {
    return /* @__PURE__ */ React19.createElement(React19.Fragment, null, /* @__PURE__ */ React19.createElement(
      SimpleDialog,
      {
        open,
        size: "xs",
        onOpenChange: setOpen,
        classNames: {
          content: "oui-p-4 oui-z-[99]",
          overlay: "oui-z-[99]"
        },
        title: t("connector.privy.supportedEvmChain")
      },
      /* @__PURE__ */ React19.createElement("div", { className: "oui-p-1 oui-grid oui-grid-cols-2 oui-gap-x-2 oui-gap-y-3 oui-text-2xs oui-text-base-contrast-54" }, chains.map((item, key) => /* @__PURE__ */ React19.createElement(
        "div",
        {
          key,
          className: "oui-flex oui-items-center oui-justify-start oui-gap-1"
        },
        /* @__PURE__ */ React19.createElement(ChainIcon, { chainId: item.id, size: "2xs" }),
        /* @__PURE__ */ React19.createElement("div", null, item.name)
      )))
    ), /* @__PURE__ */ React19.createElement("button", { onClick: () => setOpen(true) }, children));
  }
  return /* @__PURE__ */ React19.createElement(
    Popover,
    {
      content: /* @__PURE__ */ React19.createElement("div", null, /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast" }, t("connector.privy.supportedEvmChain")), /* @__PURE__ */ React19.createElement("div", { className: "oui-mt-3 oui-grid oui-grid-cols-3 oui-gap-x-2 oui-gap-y-3 oui-text-2xs oui-text-base-contrast-54" }, chains.map((item, key) => /* @__PURE__ */ React19.createElement(
        "div",
        {
          key,
          className: "oui-flex oui-items-center oui-justify-start oui-gap-1"
        },
        /* @__PURE__ */ React19.createElement(ChainIcon, { chainId: item.id, size: "2xs" }),
        /* @__PURE__ */ React19.createElement("div", null, item.name)
      )))),
      arrow: true,
      contentProps: {
        side: "bottom",
        align: "center",
        className: "oui-p-2 oui-z-[65]"
      }
    },
    /* @__PURE__ */ React19.createElement("button", null, children)
  );
}

// src/components/renderConnector/solanaConnector.tsx
function SOLConnectArea({
  connect
}) {
  const { wallets } = useSolanaWallet();
  const { isMobile } = useScreen();
  return /* @__PURE__ */ React19.createElement("div", null, /* @__PURE__ */ React19.createElement("div", { className: "oui-mb-2 oui-text-sm oui-font-semibold oui-text-base-contrast-80" }, "Solana"), /* @__PURE__ */ React19.createElement("div", { className: "oui-grid oui-grid-cols-2 oui-gap-2" }, wallets.map((item, key) => /* @__PURE__ */ React19.createElement(
    "div",
    {
      key,
      className: cn(
        " oui-flex oui-flex-1 oui-cursor-pointer oui-items-center oui-justify-start oui-gap-1 oui-rounded-[6px]  oui-px-2 oui-py-[11px]",
        isMobile ? "oui-bg-base-5" : "oui-bg-[#07080A]"
      ),
      onClick: () => connect(item.adapter)
    },
    /* @__PURE__ */ React19.createElement(RenderWalletIcon, { connector: item.adapter }),
    /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, item.adapter.name)
  ))));
}
function EVMConnectArea({
  connect
}) {
  const { connectors } = useWagmiWallet();
  const { setOpenConnectDrawer } = useWalletConnectorPrivy();
  const { isMobile } = useScreen();
  const onConnect = (item) => {
    if (isMobile) {
      if (item.id === "walletConnect") {
        setOpenConnectDrawer(false);
      }
    }
    connect(item);
  };
  return /* @__PURE__ */ React19.createElement("div", { className: "" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-mb-2 oui-text-sm oui-font-semibold oui-text-base-contrast-80" }, "EVM"), /* @__PURE__ */ React19.createElement("div", { className: "oui-grid oui-grid-cols-2 oui-gap-2" }, connectors.map((item, key) => /* @__PURE__ */ React19.createElement(
    "div",
    {
      key,
      className: cn(
        " oui-flex oui-flex-1 oui-cursor-pointer oui-items-center oui-justify-start oui-gap-1 oui-rounded-[6px]  oui-px-2 oui-py-[11px]",
        isMobile ? "oui-bg-base-5" : "oui-bg-[#07080A]"
      ),
      onClick: () => onConnect(item)
    },
    /* @__PURE__ */ React19.createElement(RenderWalletIcon, { connector: item }),
    /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, item.name)
  ))));
}

// src/components/renderConnector/index.tsx
function RenderConnector() {
  const { connect } = useWallet2();
  const {
    setOpenConnectDrawer,
    connectorWalletType,
    walletChainTypeConfig,
    targetWalletType
  } = useWalletConnectorPrivy();
  const { storageChain } = useStorageChain();
  const selectedWalletType = (() => {
    if (targetWalletType) return targetWalletType;
    if (!storageChain?.chainId) return void 0;
    try {
      return getChainType(parseInt(storageChain.chainId));
    } catch {
      return void 0;
    }
  })();
  const handleConnect = (params) => {
    connect(params);
    if (params.walletType === "privy" /* PRIVY */) {
      setOpenConnectDrawer(false);
    }
  };
  const renderPrivyConnectArea = () => {
    if (connectorWalletType.disablePrivy) {
      return null;
    }
    return /* @__PURE__ */ React19.createElement(
      PrivyConnectArea,
      {
        connect: (type) => handleConnect({
          walletType: "privy" /* PRIVY */,
          extraType: type
        })
      }
    );
  };
  const renderWagmiConnectArea = () => {
    if (connectorWalletType.disableWagmi) {
      return null;
    }
    if (!walletChainTypeConfig.hasEvm) {
      return null;
    }
    return /* @__PURE__ */ React19.createElement(
      EVMConnectArea,
      {
        connect: (connector) => handleConnect({
          walletType: "EVM" /* EVM */,
          connector
        })
      }
    );
  };
  const renderSolanaConnectArea = () => {
    if (connectorWalletType.disableSolana) {
      return null;
    }
    if (!walletChainTypeConfig.hasSol) {
      return null;
    }
    return /* @__PURE__ */ React19.createElement(
      SOLConnectArea,
      {
        connect: (walletAdapter) => handleConnect({
          walletType: "SOL" /* SOL */,
          walletAdapter
        })
      }
    );
  };
  const renderAbstractConnectArea = () => {
    if (connectorWalletType.disableAGW) {
      return null;
    }
    if (!walletChainTypeConfig.hasAbstract) {
      return null;
    }
    return /* @__PURE__ */ React19.createElement(
      AbstractConnectArea,
      {
        connect: () => handleConnect({ walletType: "Abstract" /* ABSTRACT */ })
      }
    );
  };
  const walletOrder = ["evm", "sol", "abstract"];
  const typeToKey = {
    ["EVM" /* EVM */]: "evm",
    ["SOL" /* SOL */]: "sol",
    ["Abstract" /* ABSTRACT */]: "abstract"
  };
  const prioritizedKey = selectedWalletType ? typeToKey[selectedWalletType] : void 0;
  const orderedWalletKeys = prioritizedKey ? [
    prioritizedKey,
    ...walletOrder.filter((k) => k !== prioritizedKey)
  ] : walletOrder;
  const renderByKey = (key) => {
    switch (key) {
      case "evm":
        return renderWagmiConnectArea();
      case "sol":
        return renderSolanaConnectArea();
      case "abstract":
        return renderAbstractConnectArea();
      default:
        return null;
    }
  };
  return /* @__PURE__ */ React19.createElement(ScrollArea, { className: "oui-flex oui-grow oui-shrik oui-basis-auto oui-custom-scrollbar" }, /* @__PURE__ */ React19.createElement("div", { className: cn("oui-flex oui-flex-col oui-gap-4", "md:oui-gap-5") }, renderPrivyConnectArea(), orderedWalletKeys.map((key) => /* @__PURE__ */ React19.createElement(React19.Fragment, { key }, renderByKey(key)))));
}
function SwitchNetworkTips({
  tipsContent
}) {
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  const { network, setOpenConnectDrawer } = useWalletConnectorPrivy();
  if (!tipsContent) {
    return null;
  }
  const onSwitchNetwork = () => {
    setOpenConnectDrawer(false);
    modal.show(isMobile ? "ChainSelectorSheetId" : "ChainSelectorDialogId", {
      networkId: network,
      bridgeLessOnly: false
    }).then(
      (r) => {
      },
      (error) => void 0
    );
  };
  return /* @__PURE__ */ React19.createElement(
    "div",
    {
      onClick: onSwitchNetwork,
      className: "oui-mb-3 oui-flex oui-cursor-pointer oui-items-center oui-justify-between  oui-gap-1 oui-rounded-[8px] oui-bg-[rgba(255,125,0,0.1)] oui-px-2 oui-py-[6px] "
    },
    /* @__PURE__ */ React19.createElement(Flex, { gap: 1 }, /* @__PURE__ */ React19.createElement(
      ExclamationFillIcon,
      {
        size: 14,
        className: " oui-shrink-0 oui-text-warning-darken"
      }
    ), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-warning-darken" }, t("connector.privy.switchNetwork.tips", {
      chainName: tipsContent
    }))),
    /* @__PURE__ */ React19.createElement(ArrowRightIcon, { size: 14, className: "oui-text-[#FF7D00]" })
  );
}
var StorageChainNotCurrentWalletType = ({
  currentWalletType
}) => {
  const { storageChain } = useStorageChain();
  const tipsContent = useMemo(() => {
    if (!currentWalletType || currentWalletType.size === 0 || !storageChain) {
      return null;
    }
    let text = null;
    const isSolana = SolanaChains$1.has(parseInt(storageChain?.chainId));
    const isAbstract = AbstractChains.has(parseInt(storageChain?.chainId));
    const isEvm = !isSolana && !isAbstract;
    if (isSolana && currentWalletType.has("SOL" /* SOL */)) {
      return null;
    } else if (isAbstract && currentWalletType.has("Abstract" /* ABSTRACT */)) {
      return null;
    } else if (isEvm && currentWalletType.has("EVM" /* EVM */)) {
      return null;
    }
    text = Array.from(currentWalletType).map((item) => {
      if (item === "EVM" /* EVM */) {
        return "Evm";
      } else if (item === "SOL" /* SOL */) {
        return "Solana";
      } else if (item === "Abstract" /* ABSTRACT */) {
        return "Abstract";
      }
    }).join("/");
    return text;
  }, [storageChain, currentWalletType]);
  return /* @__PURE__ */ React19.createElement(SwitchNetworkTips, { tipsContent });
};
var getCardBgClassName = (type = "EVM" /* EVM */) => {
  const cardBgColorMap = {
    ["EVM" /* EVM */]: "oui-bg-[#283BEE]",
    ["Abstract" /* ABSTRACT */]: "oui-bg-[#00A858]",
    ["SOL" /* SOL */]: "oui-bg-[#630EAD]"
  };
  return cardBgColorMap[type];
};
var getCardActiveClassName = (isActive, isMulti, type = "EVM" /* EVM */) => {
  const cardActiveColorMap = {
    ["EVM" /* EVM */]: "oui-border-[2px] oui-border-[#B9D1FF]",
    ["Abstract" /* ABSTRACT */]: "oui-border-[2px] oui-border-[#B9D1FF]",
    ["SOL" /* SOL */]: "oui-border-[2px] oui-border-[#faedff]"
  };
  return isActive && isMulti && cardActiveColorMap[type];
};
function WalletCard(props) {
  const { t } = useTranslation();
  const copyAddress = async (address) => {
    await navigator.clipboard.writeText(address);
    toast.success(t("common.copy.copied"));
  };
  return /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        "oui-relative oui-h-[110px] oui-overflow-hidden oui-rounded-2xl  oui-p-4",
        getCardBgClassName(props.type),
        getCardActiveClassName(
          props.isActive,
          props.isMulti || false,
          props.type
        )
      )
    },
    /* @__PURE__ */ React19.createElement(
      "div",
      {
        style: {
          position: "absolute",
          top: 0,
          right: -20,
          background: "url('https://oss.orderly.network/static/sdk/wallet-card-bg.png')",
          width: "110px",
          height: "110px",
          backgroundSize: "contain",
          backgroundRepeat: "no-repeat",
          backgroundPosition: "center",
          zIndex: 0
        }
      }
    ),
    /* @__PURE__ */ React19.createElement("div", { className: "oui-relative oui-z-10 oui-flex oui-h-full oui-flex-col oui-justify-between" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-between" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-text-sm oui-font-semibold oui-text-base-contrast" }, formatAddress(props.address)), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-2" }, /* @__PURE__ */ React19.createElement(Tooltip, { content: t("common.copy"), className: "oui-z-[65]" }, /* @__PURE__ */ React19.createElement(
      CopyIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-cursor-pointer oui-text-base-contrast-80 hover:oui-text-base-contrast",
        onClick: () => copyAddress(props.address)
      }
    )), props.isPrivy ? /* @__PURE__ */ React19.createElement(
      PrivyWalletHandleOption,
      {
        address: props.address,
        type: props.type
      }
    ) : /* @__PURE__ */ React19.createElement(NonPrivyWalletHandleOption, { walletType: props.type }))), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-between" }, /* @__PURE__ */ React19.createElement(RenderWalletType, { walletType: props.type }), props.isMulti && /* @__PURE__ */ React19.createElement("div", null, /* @__PURE__ */ React19.createElement(
      Checkbox,
      {
        checked: props.isActive,
        onCheckedChange: props.onActiveChange
      }
    ))))
  );
}
function NonPrivyWalletHandleOption({
  walletType
}) {
  const { disconnect } = useWallet2();
  const disconnectWallet = () => {
    let walletConnectType = "EVM" /* EVM */;
    switch (walletType) {
      case "EVM" /* EVM */:
        walletConnectType = "EVM" /* EVM */;
        break;
      case "SOL" /* SOL */:
        walletConnectType = "SOL" /* SOL */;
        break;
      case "Abstract" /* ABSTRACT */:
        walletConnectType = "Abstract" /* ABSTRACT */;
        break;
    }
    disconnect(walletConnectType);
  };
  return /* @__PURE__ */ React19.createElement("div", { onClick: () => disconnectWallet() }, /* @__PURE__ */ React19.createElement(DisconnectIcon, { className: "oui-size-4 oui-cursor-pointer oui-text-base-contrast-80 hover:oui-text-base-contrast" }));
}
function PrivyWalletHandleOption({
  address,
  type
}) {
  const { t } = useTranslation();
  const { exportWallet } = usePrivyWallet();
  return /* @__PURE__ */ React19.createElement(DropdownMenuRoot, null, /* @__PURE__ */ React19.createElement(DropdownMenuTrigger, { asChild: true }, /* @__PURE__ */ React19.createElement("button", null, /* @__PURE__ */ React19.createElement(MoreIcon, { className: "oui-cursor-pointer oui-text-base-contrast-80 hover:oui-text-base-contrast" }))), /* @__PURE__ */ React19.createElement(DropdownMenuPortal, null, /* @__PURE__ */ React19.createElement(
    DropdownMenuContent,
    {
      size: "xl",
      align: "end",
      side: "top",
      style: { width: "100px" },
      className: "oui-z-[65] oui-rounded oui-p-1 oui-font-semibold",
      sideOffset: 0
    },
    /* @__PURE__ */ React19.createElement(
      DropdownMenuItem,
      {
        className: "oui-cursor-pointer oui-px-2 oui-py-1 oui-text-2xs oui-text-base-contrast-54 hover:oui-text-base-contrast",
        onClick: () => {
          exportWallet(type);
        }
      },
      /* @__PURE__ */ React19.createElement("div", null, t("common.export"))
    )
  )));
}
var RenderWalletType = ({ walletType }) => {
  if (walletType === "SOL" /* SOL */) {
    return /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-start oui-gap-1" }, /* @__PURE__ */ React19.createElement("div", { className: "" }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: "https://oss.orderly.network/static/sdk/solana-logo.png",
        className: "oui-w-4"
      }
    )), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast" }, "Solana"));
  }
  if (walletType === "Abstract" /* ABSTRACT */) {
    return /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-start oui-gap-[6px]" }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: `${PrivyConnectorImagePath}/abstract-transparent.png`,
        className: "oui-w-4"
      }
    ), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast" }, "Abstract"));
  }
  if (walletType === "EVM" /* EVM */) {
    return /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-relative oui-flex oui-items-center oui-justify-center" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-h-[18px] oui-items-center oui-justify-center " }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: "https://oss.orderly.network/static/sdk/chains.png",
        className: "oui-relative oui-z-0 oui-h-[18px] oui-w-[49px]"
      }
    )), /* @__PURE__ */ React19.createElement("div", { className: "oui-relative oui-left-[-9px] oui-flex oui-items-center oui-justify-center oui-gap-1" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-size-[18px] oui-items-center oui-justify-center oui-rounded-full oui-bg-[#282e3a]" }, /* @__PURE__ */ React19.createElement(EVMChainPopover, null, /* @__PURE__ */ React19.createElement(
      MoreIcon,
      {
        className: "oui-relative oui-z-10 oui-size-3 oui-text-base-contrast-54 hover:oui-text-base-contrast",
        style: { zIndex: 1 }
      }
    ))), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast" }, "Evm"))));
  }
  return /* @__PURE__ */ React19.createElement(React19.Fragment, null);
};
function AddAbstractWallet() {
  const [visible, setVisible] = useState(false);
  const onToggleVisibility = () => {
    setVisible(!visible);
  };
  const { connect } = useWallet2();
  const { targetWalletType } = useWalletConnectorPrivy();
  const [open, setOpen] = useState(false);
  const { t } = useTranslation();
  useEffect(() => {
    let timer = 0;
    if (targetWalletType === "Abstract" /* ABSTRACT */) {
      timer = window.setTimeout(() => {
        setOpen(true);
      }, 200);
    }
    return () => {
      if (timer) {
        window.clearTimeout(timer);
      }
    };
  }, [targetWalletType]);
  useEffect(() => {
    if (open === false) {
      return;
    }
    const timeId = window.setTimeout(() => {
      setOpen(false);
    }, 5e3);
    return () => {
      if (timeId) {
        window.clearTimeout(timeId);
      }
    };
  }, [open]);
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-rounded-[8px] oui-bg-[#07080A] oui-px-2 oui-py-[11px]" }, /* @__PURE__ */ React19.createElement(
    Tooltip,
    {
      className: "oui-z-[65] oui-max-w-[200px] oui-text-warning-darken",
      open,
      content: t("connector.privy.addAbstractWallet.tips")
    },
    /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1 " }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: `${PrivyConnectorImagePath}/abstract-transparent.png`,
        className: "oui-size-[15px]"
      }
    ), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast-80" }, t("connector.privy.addAbstractWallet")), /* @__PURE__ */ React19.createElement("button", { onClick: onToggleVisibility }, visible ? /* @__PURE__ */ React19.createElement(
      ChevronDownIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-text-base-contrast-36"
      }
    ) : /* @__PURE__ */ React19.createElement(
      ChevronUpIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-text-base-contrast"
      }
    )))
  ), /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        "oui-transition-height oui-grid oui-grid-cols-2 oui-gap-2 oui-overflow-hidden oui-duration-150",
        visible ? "oui-mt-0 oui-max-h-0" : "oui-mt-3 oui-max-h-[400px]"
      )
    },
    /* @__PURE__ */ React19.createElement(
      "div",
      {
        className: "oui-flex oui-cursor-pointer oui-items-center oui-justify-start  oui-gap-1 oui-bg-[#131519] oui-px-2 oui-py-[11px]",
        onClick: () => connect({
          walletType: "Abstract" /* ABSTRACT */
        })
      },
      /* @__PURE__ */ React19.createElement(
        "img",
        {
          className: cn("oui-size-[12px]"),
          src: `${PrivyConnectorImagePath}/abstract.png`,
          alt: "abstract wallet"
        }
      ),
      /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, "Abstract")
    )
  ));
}
function AddEvmWallet() {
  const { t } = useTranslation();
  const [visible, setVisible] = useState(false);
  const { connect } = useWallet2();
  const [open, setOpen] = useState(false);
  const { connectors } = useWagmiWallet();
  const { targetWalletType } = useWalletConnectorPrivy();
  useEffect(() => {
    let timer = 0;
    if (targetWalletType === "EVM" /* EVM */) {
      timer = window.setTimeout(() => {
        setOpen(true);
      }, 200);
    }
    return () => {
      if (timer) {
        window.clearTimeout(timer);
      }
    };
  }, [targetWalletType]);
  useEffect(() => {
    if (open === false) {
      return;
    }
    const timeId = window.setTimeout(() => {
      setOpen(false);
    }, 5e3);
    return () => {
      if (timeId) {
        window.clearTimeout(timeId);
      }
    };
  }, [open]);
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-rounded-[8px] oui-bg-[#07080A] oui-px-2 oui-py-[11px]" }, /* @__PURE__ */ React19.createElement(
    Tooltip,
    {
      className: "oui-z-[65] oui-max-w-[200px] oui-text-warning-darken",
      open,
      content: t("connector.privy.addEvmWallet.tips")
    },
    /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1 " }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-relative oui-flex oui-w-[55px] oui-items-center oui-justify-start" }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: "https://oss.orderly.network/static/sdk/chains.png",
        className: "oui-relative oui-z-0 oui-h-[18px]"
      }
    ), /* @__PURE__ */ React19.createElement("div", { className: "oui-absolute oui-right-0 oui-flex oui-size-[18px] oui-items-center oui-justify-center oui-rounded-full oui-bg-[#282e3a]" }, /* @__PURE__ */ React19.createElement(EVMChainPopover, null, /* @__PURE__ */ React19.createElement(
      MoreIcon,
      {
        className: "oui-relative oui-z-10 oui-size-3 oui-text-base-contrast-54 hover:oui-text-base-contrast",
        style: { zIndex: 1 }
      }
    ))))), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast-80" }, t("connector.privy.addEvmWallet")), /* @__PURE__ */ React19.createElement("button", { onClick: () => setVisible(!visible) }, visible ? /* @__PURE__ */ React19.createElement(
      ChevronDownIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-text-base-contrast-36"
      }
    ) : /* @__PURE__ */ React19.createElement(
      ChevronUpIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-text-base-contrast"
      }
    )))
  ), /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        "oui-transition-height oui-grid oui-grid-cols-2 oui-gap-2 oui-overflow-hidden oui-duration-150",
        visible ? "oui-mt-0 oui-max-h-0" : "oui-mt-3 oui-max-h-[400px]"
      )
    },
    connectors.map((item, index) => /* @__PURE__ */ React19.createElement(
      "div",
      {
        key: index,
        className: "oui-flex oui-cursor-pointer oui-items-center oui-justify-center  oui-gap-1 oui-bg-[#131519] oui-px-2 oui-py-[11px] oui-text-2xs",
        onClick: () => connect({ walletType: "EVM" /* EVM */, connector: item })
      },
      /* @__PURE__ */ React19.createElement(RenderWalletIcon, { connector: item }),
      /* @__PURE__ */ React19.createElement("div", { className: "oui-text-base-contrast" }, item.name)
    ))
  ));
}
function AddSolanaWallet() {
  const { t } = useTranslation();
  const { wallets } = useSolanaWallet();
  const { connect } = useWallet2();
  const [visible, setVisible] = useState(false);
  const onToggleVisibility = () => {
    setVisible(!visible);
  };
  const { targetWalletType } = useWalletConnectorPrivy();
  const [open, setOpen] = useState(false);
  useEffect(() => {
    let timer = 0;
    if (targetWalletType === "SOL" /* SOL */) {
      timer = window.setTimeout(() => {
        setOpen(true);
      }, 200);
    }
    return () => {
      if (timer) {
        window.clearTimeout(timer);
      }
    };
  }, [targetWalletType]);
  useEffect(() => {
    if (open === false) {
      return;
    }
    const timeId = window.setTimeout(() => {
      setOpen(false);
    }, 5e3);
    return () => {
      if (timeId) {
        window.clearTimeout(timeId);
      }
    };
  }, [open]);
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-rounded-[8px] oui-bg-[#07080A] oui-px-2 oui-py-[11px]" }, /* @__PURE__ */ React19.createElement(
    Tooltip,
    {
      className: "oui-z-[65] oui-max-w-[200px] oui-text-warning-darken",
      open,
      content: t("connector.privy.addSolanaWallet.tips")
    },
    /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1 " }, /* @__PURE__ */ React19.createElement(
      "img",
      {
        src: "https://oss.orderly.network/static/sdk/solana-logo.png",
        className: "oui-size-[15px]"
      }
    ), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast-80" }, t("connector.privy.addSolanaWallet")), /* @__PURE__ */ React19.createElement("button", { onClick: onToggleVisibility }, visible ? /* @__PURE__ */ React19.createElement(
      ChevronDownIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-text-base-contrast-36"
      }
    ) : /* @__PURE__ */ React19.createElement(
      ChevronUpIcon,
      {
        size: 16,
        opacity: 1,
        className: "oui-text-base-contrast"
      }
    )))
  ), /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        "oui-transition-height oui-grid oui-grid-cols-2 oui-gap-2 oui-overflow-hidden oui-duration-150",
        visible ? "oui-mt-0 oui-max-h-0" : "oui-mt-3 oui-max-h-[400px]"
      )
    },
    wallets.map((item, index) => /* @__PURE__ */ React19.createElement(
      "div",
      {
        key: index,
        className: "oui-flex oui-cursor-pointer oui-items-center oui-justify-start  oui-gap-1 oui-bg-[#131519] oui-px-2 oui-py-[11px]",
        onClick: () => connect({
          walletType: "SOL" /* SOL */,
          walletAdapter: item.adapter
        })
      },
      /* @__PURE__ */ React19.createElement(RenderWalletIcon, { connector: item.adapter }),
      /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-base-contrast" }, item.adapter.name)
    ))
  ));
}

// src/components/renderNonPrivyWallet/index.tsx
function RenderNonPrivyWallet() {
  const [walletList, setWalletList] = useState([]);
  const [addWalletList, setAddWalletList] = useState([]);
  const { storageChain } = useStorageChain();
  const { connectorWalletType, walletChainTypeConfig } = useWalletConnectorPrivy();
  const { wallet: walletInWagmi, isConnected: isConnectedEvm } = useWagmiWallet();
  const { wallet: walletInSolana, isConnected: isConnectedSolana } = useSolanaWallet();
  const { wallet: walletInAbstract, isConnected: isConnectedAbstract } = useAbstractWallet();
  const { namespace, switchWallet } = useWallet2();
  const isActive = (walletType) => {
    if (namespace === ChainNamespace.evm) {
      if (walletType === "Abstract" /* ABSTRACT */) {
        return AbstractChains.has(storageChain?.chainId);
      }
      if (walletType === "EVM" /* EVM */) {
        return !AbstractChains.has(storageChain?.chainId);
      }
      return false;
    }
    if (namespace === ChainNamespace.solana) {
      return walletType === "SOL" /* SOL */;
    }
    return false;
  };
  const currentConnectWalletType = useMemo(() => {
    const temp = /* @__PURE__ */ new Set();
    walletList.forEach((wallet) => {
      temp.add(wallet.type);
    });
    return temp;
  }, [walletList]);
  useEffect(() => {
    const tempWalletList = [];
    const tempAddWallet = [];
    if (!connectorWalletType.disableWagmi && walletChainTypeConfig.hasEvm) {
      if (isConnectedEvm) {
        tempWalletList.push({
          type: "EVM" /* EVM */,
          address: walletInWagmi?.accounts[0].address
        });
      } else {
        tempAddWallet.push("EVM" /* EVM */);
      }
    }
    if (!connectorWalletType.disableSolana && walletChainTypeConfig.hasSol) {
      if (isConnectedSolana) {
        tempWalletList.push({
          type: "SOL" /* SOL */,
          address: walletInSolana?.accounts[0].address
        });
      } else {
        tempAddWallet.push("SOL" /* SOL */);
      }
    }
    if (!connectorWalletType.disableAGW && walletChainTypeConfig.hasAbstract) {
      if (isConnectedAbstract) {
        tempWalletList.push({
          type: "Abstract" /* ABSTRACT */,
          address: walletInAbstract?.additionalInfo?.AGWAddress
        });
      } else {
        tempAddWallet.push("Abstract" /* ABSTRACT */);
      }
    }
    setWalletList(tempWalletList);
    setAddWalletList(tempAddWallet);
  }, [
    isConnectedEvm,
    isConnectedSolana,
    isConnectedAbstract,
    walletInWagmi,
    walletInSolana,
    walletInAbstract,
    walletChainTypeConfig,
    connectorWalletType
  ]);
  return /* @__PURE__ */ React19.createElement(React19.Fragment, null, walletList.length && /* @__PURE__ */ React19.createElement(
    StorageChainNotCurrentWalletType,
    {
      currentWalletType: currentConnectWalletType
    }
  ), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-flex-col oui-gap-5" }, walletList.map((wallet) => /* @__PURE__ */ React19.createElement(
    WalletCard,
    {
      key: wallet.type,
      type: wallet.type,
      address: wallet.address,
      isActive: isActive(wallet.type),
      onActiveChange: () => {
        switchWallet(wallet.type);
      },
      isPrivy: false,
      isMulti: walletList.length > 1
    }
  ))), /* @__PURE__ */ React19.createElement(RenderAddWallet, { addWalletList }));
}
function RenderAddWallet({ addWalletList }) {
  if (addWalletList.length === 0) {
    return null;
  }
  return /* @__PURE__ */ React19.createElement(React19.Fragment, null, /* @__PURE__ */ React19.createElement("div", { className: "oui-my-5 oui-h-px oui-bg-line" }), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-flex-col oui-gap-5" }, addWalletList.map((wallet, index) => {
    if (wallet === "EVM" /* EVM */) {
      return /* @__PURE__ */ React19.createElement(AddEvmWallet, { key: index });
    }
    if (wallet === "SOL" /* SOL */) {
      return /* @__PURE__ */ React19.createElement(AddSolanaWallet, { key: index });
    }
    if (wallet === "Abstract" /* ABSTRACT */) {
      return /* @__PURE__ */ React19.createElement(AddAbstractWallet, { key: index });
    }
  })));
}
function NoWallet() {
  const { walletChainType } = useWalletConnectorPrivy();
  const { t } = useTranslation();
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-mt-5 oui-flex oui-w-full oui-flex-col oui-items-center oui-justify-center" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-w-full oui-flex-col oui-items-center oui-justify-center oui-gap-3" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-w-full oui-flex-col oui-items-center oui-justify-center oui-gap-1 oui-rounded-[8px] oui-border  oui-border-line-12 oui-px-2 oui-py-[13px] " }, /* @__PURE__ */ React19.createElement(
    "img",
    {
      src: "https://oss.orderly.network/static/sdk/privy/no-wallet.png",
      className: "oui-size-[64px]"
    }
  ), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast-36" }, t("connector.privy.noWallet"))), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-start oui-gap-1 oui-rounded-[4px] oui-bg-[rgba(232,136,0,0.15)] oui-px-2 oui-py-[6px] " }, /* @__PURE__ */ React19.createElement(
    ExclamationFillIcon,
    {
      size: 10,
      className: "oui-mt-1 oui-size-[10px] oui-shrink-0 oui-text-warning-darken"
    }
  ), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-text-warning-darken" }, t("connector.privy.noWallet.description")))), /* @__PURE__ */ React19.createElement("div", { className: "oui-my-5 oui-h-px oui-w-full oui-bg-line" }), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-w-full oui-flex-col oui-gap-2" }, walletChainType === "EVM_SOL" /* EVM_SOL */ && /* @__PURE__ */ React19.createElement(React19.Fragment, null, /* @__PURE__ */ React19.createElement(CreateEVMWallet, null), /* @__PURE__ */ React19.createElement(CreateSOLWallet, null)), walletChainType === "onlyEVM" /* onlyEVM */ && /* @__PURE__ */ React19.createElement(CreateEVMWallet, null), walletChainType === "onlySOL" /* onlySOL */ && /* @__PURE__ */ React19.createElement(CreateSOLWallet, null)));
}
function CreateEVMWallet() {
  const { t } = useTranslation();
  const { createEvmWallet } = usePrivyWallet();
  const [loading, setLoading] = useState(false);
  const doCreate = () => {
    if (loading) {
      return;
    }
    setLoading(true);
    createEvmWallet().then(() => {
      setLoading(false);
    }).catch((e) => {
      setLoading(false);
    });
  };
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-w-full oui-rounded-[8px] oui-bg-[#07080A] oui-px-2 oui-py-[11px]" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1 " }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-relative oui-flex oui-w-[55px] oui-items-center oui-justify-start" }, /* @__PURE__ */ React19.createElement(
    "img",
    {
      src: "https://oss.orderly.network/static/sdk/chains.png",
      className: "oui-relative oui-z-0 oui-h-[18px]"
    }
  ), /* @__PURE__ */ React19.createElement("div", { className: "oui-absolute oui-right-0 oui-flex oui-size-[18px] oui-items-center oui-justify-center oui-rounded-full oui-bg-[#282e3a]" }, /* @__PURE__ */ React19.createElement(EVMChainPopover, null, /* @__PURE__ */ React19.createElement(
    MoreIcon,
    {
      className: "oui-relative oui-z-10 oui-size-3 oui-text-base-contrast-54 hover:oui-text-base-contrast",
      style: { zIndex: 1 }
    }
  ))))), /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        "oui-cursor-pointer oui-text-2xs oui-font-semibold oui-text-base-contrast-80",
        loading && "oui-opacity-50"
      ),
      onClick: doCreate
    },
    t("connector.privy.createEvmWallet")
  )));
}
function CreateSOLWallet() {
  const { t } = useTranslation();
  const { createSolanaWallet } = usePrivyWallet();
  const [loading, setLoading] = useState(false);
  const doCreate = () => {
    if (loading) {
      return;
    }
    setLoading(true);
    createSolanaWallet().then(() => {
      setLoading(false);
    }).catch((e) => {
      setLoading(false);
    });
  };
  return /* @__PURE__ */ React19.createElement("div", { className: "oui-w-full oui-rounded-[8px] oui-bg-[#07080A] oui-px-2 oui-py-[11px]" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-center oui-gap-1 " }, /* @__PURE__ */ React19.createElement(
    "img",
    {
      src: "https://oss.orderly.network/static/sdk/solana-logo.png",
      className: "oui-size-[15px]"
    }
  ), /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: cn(
        "oui-cursor-pointer oui-text-2xs oui-font-semibold oui-text-base-contrast-80",
        loading && "oui-opacity-50"
      ),
      onClick: doCreate
    },
    t("connector.privy.createSolanaWallet")
  )));
}
function RenderPrivyWallet() {
  const { t } = useTranslation();
  const {
    targetWalletType,
    setTargetWalletType,
    walletChainTypeConfig,
    connectorWalletType
  } = useWalletConnectorPrivy();
  const { walletEVM, walletSOL, linkedAccount } = usePrivyWallet();
  const { switchWallet, disconnect } = useWallet2();
  const { storageChain } = useStorageChain();
  const [walletList, setWalletList] = useState([]);
  const [addWallet, setAddWallet] = useState([]);
  const [loading, setLoading] = useState(true);
  const isActive = useCallback(
    (walletType) => {
      if (storageChain?.namespace === ChainNamespace.evm) {
        if (walletType === "EVM" /* EVM */) {
          return !AbstractChains.has(storageChain?.chainId);
        }
        return false;
      }
      if (storageChain?.namespace === ChainNamespace.solana) {
        return walletType === "SOL" /* SOL */;
      }
      return false;
    },
    [storageChain]
  );
  const isHaveEvmWallet = useMemo(() => {
    return walletEVM && walletEVM.accounts.length;
  }, [walletEVM]);
  const isHaveSolWallet = useMemo(() => {
    return walletSOL && walletSOL.accounts.length;
  }, [walletSOL]);
  const renderWarning = useCallback(() => {
    let showWarning = false;
    if (AbstractChains.has(storageChain?.chainId)) {
      showWarning = true;
    }
    if (targetWalletType === "Abstract" /* ABSTRACT */) {
      showWarning = true;
    }
    if (!showWarning) {
      return;
    }
    return /* @__PURE__ */ React19.createElement("div", { className: "oui-mt-5 oui-border-b oui-border-line oui-pb-5" }, /* @__PURE__ */ React19.createElement(
      "div",
      {
        className: cn(
          "oui-flex oui-items-start oui-justify-center oui-gap-1",
          "oui-w-full  oui-rounded-[8px] oui-px-2 oui-py-[13px]",
          "oui-bg-warning-darken/10"
        )
      },
      /* @__PURE__ */ React19.createElement(
        ExclamationFillIcon,
        {
          size: 14,
          className: "oui-mt-[2px] oui-size-[14px] oui-shrink-0 oui-text-warning-darken"
        }
      ),
      /* @__PURE__ */ React19.createElement("div", { className: "oui-text-2xs oui-leading-[18px] oui-text-warning-darken" }, "Abstract Chain access requires Abstract Global Wallet. Privy connection is currently unsupported.")
    ));
  }, [targetWalletType]);
  const renderWallet = useCallback(() => {
    if (loading) {
      return;
    }
    if (!walletList.length) {
      return /* @__PURE__ */ React19.createElement(NoWallet, null);
    }
    return /* @__PURE__ */ React19.createElement("div", { className: "oui-mt-5 oui-flex oui-flex-col oui-gap-5" }, walletList.map((wallet) => /* @__PURE__ */ React19.createElement(
      WalletCard,
      {
        key: wallet.type,
        type: wallet.type,
        address: wallet.address,
        isActive: isActive(wallet.type),
        isPrivy: true,
        isMulti: walletList.length > 1,
        onActiveChange: () => {
          switchWallet(wallet.type);
        }
      }
    )), addWallet.map((node, index) => /* @__PURE__ */ React19.createElement("div", { key: index }, /* @__PURE__ */ React19.createElement("div", { className: "oui-my-5 oui-h-px oui-w-full oui-bg-line" }), /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-w-full oui-flex-col oui-gap-2" }, node))));
  }, [walletList, addWallet, isActive, switchWallet, t, loading]);
  useEffect(() => {
    new Promise(
      (resolve) => setTimeout(() => {
        setLoading(false);
        resolve(true);
      }, 200)
    );
  }, []);
  useEffect(() => {
    const tempWalletList = [];
    const tempAddWallet = [];
    if (!connectorWalletType.disableWagmi && walletChainTypeConfig.hasEvm) {
      if (isHaveEvmWallet) {
        tempWalletList.push({
          type: "EVM" /* EVM */,
          address: walletEVM.accounts[0].address
        });
      } else {
        tempAddWallet.push(/* @__PURE__ */ React19.createElement(CreateEVMWallet, null));
      }
    }
    if (!connectorWalletType.disableSolana && walletChainTypeConfig.hasSol) {
      if (isHaveSolWallet) {
        tempWalletList.push({
          type: "SOL" /* SOL */,
          address: walletSOL.accounts[0].address
        });
      } else {
        tempAddWallet.push(/* @__PURE__ */ React19.createElement(CreateSOLWallet, null));
      }
    }
    setWalletList(tempWalletList);
    setAddWallet(tempAddWallet);
  }, [
    connectorWalletType,
    walletChainTypeConfig,
    walletEVM,
    walletSOL,
    isHaveEvmWallet,
    isHaveSolWallet
  ]);
  useEffect(() => {
    if (targetWalletType === "Abstract" /* ABSTRACT */) {
      windowGuard(() => {
        setTimeout(() => {
          setTargetWalletType(void 0);
        }, 5e3);
      });
    }
  }, [targetWalletType, setTargetWalletType]);
  return /* @__PURE__ */ React19.createElement("div", null, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-between" }, linkedAccount && /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-items-center oui-justify-start oui-gap-2 oui-text-base-contrast" }, /* @__PURE__ */ React19.createElement("div", null, /* @__PURE__ */ React19.createElement(RenderPrivyTypeIcon, { type: linkedAccount.type, size: 24 })), /* @__PURE__ */ React19.createElement("div", { className: "oui-text-xs" }, linkedAccount.address)), /* @__PURE__ */ React19.createElement(
    "div",
    {
      className: "oui-cursor-pointer oui-text-2xs oui-font-semibold oui-text-primary",
      onClick: () => disconnect("privy" /* PRIVY */)
    },
    t("connector.privy.logout")
  )), renderWarning(), renderWallet());
}

// src/components/connectDrawer.tsx
function MyWallet() {
  const [connectorKey, setConnectorKey] = useLocalStorage(ConnectorKey, "");
  return /* @__PURE__ */ React19.createElement("div", null, connectorKey === "privy" && /* @__PURE__ */ React19.createElement(RenderPrivyWallet, null), connectorKey !== "privy" && /* @__PURE__ */ React19.createElement(RenderNonPrivyWallet, null));
}
function ConnectDrawer(props) {
  const { t } = useTranslation();
  const { isConnected: isConnectedPrivy } = usePrivyWallet();
  const { isConnected: isConnectedEvm } = useWagmiWallet();
  const { isConnected: isConnectedSolana } = useSolanaWallet();
  const { isConnected: isConnectedAbstract } = useAbstractWallet();
  const { termsOfUse } = useWalletConnectorPrivy();
  const [connectorKey] = useLocalStorage(ConnectorKey, "");
  const isConnected = useMemo(() => {
    if (connectorKey === "privy" /* PRIVY */ && isConnectedPrivy) {
      return true;
    }
    if (connectorKey !== "privy" /* PRIVY */) {
      if (isConnectedEvm) {
        return true;
      }
      if (isConnectedSolana) {
        return true;
      }
      if (isConnectedAbstract) {
        return true;
      }
    }
    return false;
  }, [
    isConnectedPrivy,
    isConnectedEvm,
    isConnectedSolana,
    isConnectedAbstract,
    connectorKey
  ]);
  const { isMobile } = useScreen();
  const showPwaDialog = () => {
    modal.show(PwaDialog);
  };
  const renderHeader = useCallback(() => {
    return /* @__PURE__ */ React19.createElement(
      "div",
      {
        className: cn(
          "oui-font-semibold oui-text-base-contrast-80 ",
          "oui-pb-2 oui-text-[20px]",
          "md:oui-py-0 md:oui-text-base"
        )
      },
      isConnected ? t("connector.privy.myWallet") : t("connector.connectWallet")
    );
  }, [isConnected]);
  return /* @__PURE__ */ React19.createElement(
    SimpleSheet,
    {
      open: props.open,
      onOpenChange: props.onChangeOpen,
      classNames: {
        body: "oui-h-full oui-py-0  oui-border-none oui-relative",
        // overlay: "!oui-bg-base-10/60",
        content: cn(
          "!oui-p-4  !oui-bg-[#131519] !oui-border !oui-border-solid !oui-border-line-12",
          isMobile ? "oui-inset-y-0 oui-right-0 oui-w-[280px] oui-rounded-none !oui-bg-[#181C23]" : "!oui-bottom-[30px] oui-right-3 oui-top-[48px] !oui-h-auto oui-w-[300px] oui-overflow-hidden oui-rounded-[16px] !oui-bg-[#131519] "
        )
      },
      contentProps: { side: "right", closeable: false }
    },
    !isMobile && /* @__PURE__ */ React19.createElement(
      "div",
      {
        className: "oui-absolute oui-inset-x-[50px] -oui-top-[calc(100vh/2)] oui-z-0 oui-h-screen",
        style: {
          background: "conic-gradient(from -41deg at 40.63% 50.41%, rgba(242, 98, 181, 0.00) 125.17920970916748deg, rgba(95, 197, 255, 0.20) 193.4119462966919deg, rgba(255, 172, 137, 0.20) 216.0206937789917deg, rgba(129, 85, 255, 0.20) 236.0708713531494deg, rgba(120, 157, 255, 0.20) 259.95326042175293deg, rgba(159, 115, 241, 0.00) 311.0780096054077deg)",
          filter: "blur(50px)"
        }
      }
    ),
    /* @__PURE__ */ React19.createElement("div", { className: "oui-relative oui-z-10 oui-flex oui-h-full oui-flex-col oui-gap-4 md:oui-gap-5" }, /* @__PURE__ */ React19.createElement("div", { className: "oui-flex oui-flex-none oui-items-center oui-justify-between" }, renderHeader(), /* @__PURE__ */ React19.createElement(
      CloseSquareFillIcon,
      {
        className: "oui-size-5 oui-cursor-pointer oui-text-base-contrast-20 hover:oui-text-base-contrast-80",
        onClick: () => props.onChangeOpen(false)
      }
    )), isConnected ? /* @__PURE__ */ React19.createElement(MyWallet, null) : /* @__PURE__ */ React19.createElement(RenderConnector, null), !isConnected && /* @__PURE__ */ React19.createElement(Flex, { gap: 4, direction: "column" }, isMobile && /* @__PURE__ */ React19.createElement(
      Flex,
      {
        itemAlign: "center",
        justify: "between",
        gap: 4,
        className: "oui-rounded-[8px] oui-px-3 oui-py-2 oui-text-2xs oui-bg-[linear-gradient(270deg,rgba(89,176,254,0.10)_0%,rgba(38,254,254,0.10)_100%)] oui-w-full",
        onClick: () => showPwaDialog()
      },
      /* @__PURE__ */ React19.createElement(Text, { className: "oui-bg-[linear-gradient(270deg,#59B0FE_0%,#26FEFE_100%)] oui-bg-clip-text oui-text-transparent" }, t("connector.privy.pwa.title")),
      /* @__PURE__ */ React19.createElement(
        ArrowRightLinearGradientIcon,
        {
          size: 20,
          className: "oui-flex-shrink-0"
        }
      )
    ), termsOfUse && /* @__PURE__ */ React19.createElement("div", { className: "oui-flex-none oui-text-center oui-text-2xs oui-font-semibold  oui-text-base-contrast-80" }, /* @__PURE__ */ React19.createElement(
      Trans,
      {
        i18nKey: "connector.privy.termsOfUse",
        components: [
          /* @__PURE__ */ React19.createElement(
            "a",
            {
              key: "termsOfUse",
              href: termsOfUse,
              className: "oui-cursor-pointer oui-text-primary oui-underline",
              target: "_blank",
              rel: "noreferrer"
            }
          )
        ]
      }
    ))))
  );
}
var LinkDeviceMobile = (props) => {
  const [open, setOpen] = useState(false);
  const { account } = useAccount();
  const { t } = useTranslation();
  const onDisconnect = async () => {
    localStorage.removeItem("orderly_link_device");
    await account.disconnect();
  };
  const showDialog = () => {
    setOpen(true);
  };
  const hideDialog = () => {
    setOpen(false);
  };
  return /* @__PURE__ */ React19.createElement(React19.Fragment, null, /* @__PURE__ */ React19.createElement(
    SimpleDialog,
    {
      open,
      onOpenChange: setOpen,
      title: t("common.tips"),
      size: "xs",
      actions: {
        secondary: {
          label: t("common.cancel"),
          onClick: hideDialog,
          size: "md",
          fullWidth: true
        },
        primary: {
          label: t("connector.disconnect"),
          onClick: async () => {
            await onDisconnect();
            hideDialog();
          },
          size: "md",
          variant: "outlined",
          color: "danger"
        }
      }
    },
    /* @__PURE__ */ React19.createElement(Text, { intensity: 54, size: "sm" }, t("linkDevice.scanQRCode.connected.description"))
  ), /* @__PURE__ */ React19.createElement(Flex, { className: "oui-gap-[6px]", itemAlign: "center", onClick: showDialog }, /* @__PURE__ */ React19.createElement(
    Flex,
    {
      className: "oui-text-base-contrast",
      intensity: 500,
      height: 28,
      r: "md"
    },
    /* @__PURE__ */ React19.createElement(DesktopIcon, null),
    /* @__PURE__ */ React19.createElement(Dot, null),
    /* @__PURE__ */ React19.createElement(LinkIcon, null),
    /* @__PURE__ */ React19.createElement(Dot, null),
    /* @__PURE__ */ React19.createElement(MobileIcon, null)
  ), props.children));
};
var Dot = () => {
  return /* @__PURE__ */ React19.createElement(Flex, { className: "oui-gap-x-[1px] oui-px-[1px]" }, /* @__PURE__ */ React19.createElement(DotIcon, null), /* @__PURE__ */ React19.createElement(DotIcon, null), /* @__PURE__ */ React19.createElement(DotIcon, null));
};
var DesktopIcon = (props) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    width: "18",
    height: "18",
    viewBox: "0 0 18 18",
    fill: "currentColor",
    xmlns: "http://www.w3.org/2000/svg",
    ...props
  },
  /* @__PURE__ */ React19.createElement("path", { d: "M4.5 2.93a2.25 2.25 0 0 0-2.25 2.25v5.25a.75.75 0 0 0-.75.75v1.5c0 1.272.93 2.25 2.25 2.25h10.5c1.318 0 2.25-.978 2.25-2.25v-1.5a.75.75 0 0 0-.75-.75V5.18a2.25 2.25 0 0 0-2.25-2.25zm0 1.5h9a.75.75 0 0 1 .75.75v5.25H3.75V5.18a.75.75 0 0 1 .75-.75M3 11.93h12v.75c0 .46-.277.75-.75.75H3.75c-.474 0-.75-.29-.75-.75z" })
);
var MobileIcon = (props) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    width: "18",
    height: "18",
    viewBox: "0 0 18 18",
    fill: "currentColor",
    xmlns: "http://www.w3.org/2000/svg",
    ...props
  },
  /* @__PURE__ */ React19.createElement("path", { d: "M13.498 3.715a2.25 2.25 0 0 0-2.25-2.25h-4.5a2.25 2.25 0 0 0-2.25 2.25v10.5a2.25 2.25 0 0 0 2.25 2.25h4.5a2.25 2.25 0 0 0 2.25-2.25zm-1.5 0v9.75h-6v-9.75a.75.75 0 0 1 .75-.75h4.5a.75.75 0 0 1 .75.75m-2.25 11.25a.75.75 0 1 1-1.5 0 .75.75 0 0 1 1.5 0" })
);
var LinkIcon = (props) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  },
  /* @__PURE__ */ React19.createElement(
    "path",
    {
      fillRule: "evenodd",
      clipRule: "evenodd",
      d: "M1.007 6a5 5 0 1 1 10 0 5 5 0 0 1-10 0m7.486-2.344A.6.6 0 0 1 8.91 3.5c.15 0 .305.048.418.156a.55.55 0 0 1 0 .798L5.254 8.337a.62.62 0 0 1-.837 0L2.67 6.673a.55.55 0 0 1 0-.798.62.62 0 0 1 .837 0l1.329 1.266z",
      fill: "url(#a)"
    }
  ),
  /* @__PURE__ */ React19.createElement("defs", null, /* @__PURE__ */ React19.createElement(
    "linearGradient",
    {
      id: "a",
      x1: "11.007",
      y1: "5.999",
      x2: "1.007",
      y2: "5.999",
      gradientUnits: "userSpaceOnUse"
    },
    /* @__PURE__ */ React19.createElement("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
    /* @__PURE__ */ React19.createElement("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
  ))
);
var DotIcon = (props) => /* @__PURE__ */ React19.createElement(
  "svg",
  {
    width: "2",
    height: "2",
    viewBox: "0 0 2 2",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  },
  /* @__PURE__ */ React19.createElement(
    "path",
    {
      d: "M.667.334a.667.667 0 1 1 0 1.333.667.667 0 0 1 0-1.333",
      fill: "url(#a)"
    }
  ),
  /* @__PURE__ */ React19.createElement("defs", null, /* @__PURE__ */ React19.createElement(
    "linearGradient",
    {
      id: "a",
      x1: "1.333",
      y1: "1.001",
      x2: "0",
      y2: "1.001",
      gradientUnits: "userSpaceOnUse"
    },
    /* @__PURE__ */ React19.createElement("stop", { stopColor: "rgb(var(--oui-gradient-brand-end))" }),
    /* @__PURE__ */ React19.createElement("stop", { offset: "1", stopColor: "rgb(var(--oui-gradient-brand-start))" })
  ))
);

// src/components/userCenter.tsx
function UserCenter(props) {
  const { accountState: state } = props;
  return /* @__PURE__ */ React19.createElement(RenderUserCenter, { state, disabledConnect: props.disabledConnect });
}
var MwebUserCenter = (props) => {
  const { state } = props;
  return /* @__PURE__ */ React19.createElement(RenderUserCenter, { state, disabledConnect: props.disabledConnect });
};
var RenderUserCenter = (props) => {
  const { state } = props;
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  const { connect, wallet } = useWalletConnector();
  const { linkedAccount } = usePrivyWallet();
  const { state: accountState, account } = useAccount();
  const { connectedChain } = useWalletConnector();
  const disabled = state.validating || props.disabledConnect;
  const userAddress = useMemo(() => {
    if (connectedChain?.id && ABSTRACT_CHAIN_ID_MAP.has(parseInt(connectedChain?.id))) {
      return account.getAdditionalInfo()?.AGWAddress;
    }
    return account.address;
  }, [account, connectedChain, accountState]);
  if (state.status === AccountStatusEnum.EnableTradingWithoutConnected) {
    return /* @__PURE__ */ React19.createElement(Flex, { className: "oui-bg-base-5 oui-px-[7px] oui-rounded-[6px] oui-gap-[6px]" }, /* @__PURE__ */ React19.createElement(LinkDeviceMobile, null, /* @__PURE__ */ React19.createElement(
      Text.formatted,
      {
        rule: "address",
        className: "oui-text-base-contrast oui-text-xs"
      },
      formatAddress(userAddress)
    )));
  }
  if (state.status <= AccountStatusEnum.NotConnected || disabled) {
    return /* @__PURE__ */ React19.createElement(
      Button,
      {
        "data-testid": "oui-testid-nav-bar-connectWallet-btn",
        size: "md",
        variant: disabled ? void 0 : "gradient",
        angle: 45,
        className: cn(
          "wallet-connect-button",
          isMobile && "oui-font-semibold oui-px-2"
        ),
        loading: state.validating,
        disabled,
        onClick: () => {
          connect().then((r) => {
          }).catch((e) => void 0);
        }
      },
      isMobile ? t("connector.connect") : t("connector.connectWallet")
    );
  }
  if (!wallet) {
    return;
  }
  if (isMobile) {
    return /* @__PURE__ */ React19.createElement(
      AuthGuard,
      {
        buttonProps: {
          size: "md"
        }
      },
      /* @__PURE__ */ React19.createElement("div", { onClick: () => connect() }, /* @__PURE__ */ React19.createElement(
        Button,
        {
          size: "md",
          variant: "gradient",
          angle: 45,
          "data-testid": "oui-testid-nav-bar-address-btn",
          className: "oui-px-2 oui-flex oui-items-center oui-justify-center oui-gap-1"
        },
        linkedAccount && /* @__PURE__ */ React19.createElement(
          RenderPrivyTypeIcon,
          {
            type: linkedAccount.type,
            size: 14,
            black: true
          }
        ),
        /* @__PURE__ */ React19.createElement(
          Text.formatted,
          {
            rule: "address",
            className: "oui-text-[rgba(0,0,0,.88)] oui-font-semibold"
          },
          formatAddress(userAddress)
        )
      ))
    );
  }
  return /* @__PURE__ */ React19.createElement("div", { onClick: () => connect() }, /* @__PURE__ */ React19.createElement(
    Button,
    {
      size: "md",
      variant: "gradient",
      angle: 45,
      "data-testid": "oui-testid-nav-bar-address-btn",
      className: "oui-flex oui-items-center oui-justify-center oui-gap-2"
    },
    linkedAccount && /* @__PURE__ */ React19.createElement(
      RenderPrivyTypeIcon,
      {
        type: linkedAccount.type,
        size: 18,
        black: true
      }
    ),
    /* @__PURE__ */ React19.createElement(Text.formatted, { rule: "address", className: "oui-text-[rgba(0,0,0,.88)]" }, formatAddress(userAddress))
  ));
};
installExtension({
  name: "account-menu-privy",
  scope: ["*"],
  positions: [ExtensionPositionEnum.AccountMenu]
})((props) => {
  return /* @__PURE__ */ React19.createElement(UserCenter, { ...props });
});
installExtension({
  name: "mobile-account-menu-privy",
  scope: ["*"],
  positions: [ExtensionPositionEnum.MobileAccountMenu]
})((props) => {
  return /* @__PURE__ */ React19.createElement(MwebUserCenter, { ...props });
});

// src/main.tsx
var Main = (props) => {
  const { headerProps, children } = props;
  const { wallet, connectedChain, setChain, namespace, onDisconnect } = useWallet2();
  const { openConnectDrawer, setOpenConnectDrawer, setTargetWalletType } = useWalletConnectorPrivy();
  const connect = (options) => {
    if (options && options.autoSelect) {
      return Promise.resolve([]);
    }
    setTargetWalletType(void 0);
    return new Promise((resolve) => {
      setOpenConnectDrawer(true);
      resolve([]);
    });
  };
  const memoizedValue = useMemo(
    () => ({
      connect,
      disconnect: onDisconnect,
      connecting: false,
      wallet,
      setChain,
      connectedChain,
      namespace,
      chains: [],
      settingChain: false
    }),
    [connect, setChain, onDisconnect, connectedChain, wallet, namespace]
  );
  return /* @__PURE__ */ React19.createElement(WalletConnectorContext.Provider, { value: memoizedValue }, /* @__PURE__ */ React19.createElement(
    ConnectDrawer,
    {
      open: openConnectDrawer,
      onChangeOpen: setOpenConnectDrawer,
      headerProps
    }
  ), children);
};
var InitAbstractProvider = (props) => {
  const { network } = useWalletConnectorPrivy();
  const chain = useMemo(() => {
    if (network === "mainnet" /* mainnet */) {
      return abstract;
    }
    return abstractTestnet;
  }, [network]);
  return /* @__PURE__ */ React19.createElement(AbstractWalletProvider$1, { chain }, props.children);
};

// src/providers/abstractWallet/index.tsx
function AbstractWallet({ children }) {
  return /* @__PURE__ */ React19.createElement(InitAbstractProvider, null, /* @__PURE__ */ React19.createElement(AbstractWalletProvider, null, children));
}
function InitPrivyProvider({
  privyConfig,
  initChains,
  children
}) {
  if (!privyConfig) {
    return children;
  }
  const { network } = useWalletConnectorPrivy();
  const config = useMemo(() => {
    const chains = initChains;
    const preferredDefaultChainIds = (network === "mainnet" ? defaultMainnetChains : defaultTestnetChains).map((c) => c.id);
    const preferredDefaultChain = preferredDefaultChainIds.map((id) => chains.find((c) => c.id === id)).find((c) => !!c);
    const firstEvmChain = chains.find(
      (chain) => !SolanaChains$1.has(chain.id) && !AbstractChains.has(chain.id)
    );
    const defaultEvmChain = preferredDefaultChain ?? firstEvmChain ?? chains[0];
    return {
      loginMethods: privyConfig.config?.loginMethods || [
        "email",
        "google",
        "twitter"
      ],
      appearance: {
        ...privyConfig.config?.appearance,
        walletChainType: "ethereum-and-solana"
      },
      embeddedWallets: {
        ethereum: {
          createOnLogin: "all-users"
        },
        solana: {
          createOnLogin: "all-users"
        }
      },
      externalWallets: {
        walletConnect: {
          enabled: false
        }
      },
      defaultChain: defaultEvmChain,
      supportedChains: chains
    };
  }, [initChains, privyConfig, network]);
  if (!initChains.length) {
    return;
  }
  return /* @__PURE__ */ React19.createElement(PrivyProvider, { appId: privyConfig.appid, config }, children);
}

// src/providers/privy/index.tsx
function PrivyWallet(props) {
  return /* @__PURE__ */ React19.createElement(
    InitPrivyProvider,
    {
      privyConfig: props.privyConfig,
      initChains: props.initChains
    },
    /* @__PURE__ */ React19.createElement(PrivyWalletProvider, null, props.children)
  );
}
function InitSolanaProvider({
  mainnetRpc,
  devnetRpc,
  wallets: walletsProp,
  onError,
  children
}) {
  const ee = useEventEmitter();
  const { network, setSolanaInfo, connectorWalletType } = useWalletConnectorPrivy();
  if (connectorWalletType.disableSolana) {
    return children;
  }
  const wallets = useMemo(() => {
    return walletsProp ?? [new PhantomWalletAdapter()];
  }, [walletsProp]);
  useEffect(() => {
    let rpcUrl = null;
    if (network === "mainnet") {
      rpcUrl = mainnetRpc ?? null;
    } else {
      rpcUrl = devnetRpc ?? null;
    }
    setSolanaInfo({
      rpcUrl,
      network: network === "mainnet" ? WalletAdapterNetwork.Mainnet : WalletAdapterNetwork.Devnet
    });
  }, [network, mainnetRpc, devnetRpc, setSolanaInfo]);
  const handleOnError = useCallback(
    (error, adapter) => {
      if (error.name === "WalletAccountError") {
        ee.emit("wallet:connect-error", {
          message: "Please switch to a wallet with Solana address."
        });
        return;
      }
      onError?.(error, adapter);
    },
    [ee, onError]
  );
  return /* @__PURE__ */ React19.createElement(WalletProvider, { wallets, onError: handleOnError, autoConnect: true }, children);
}

// src/providers/solana/index.tsx
function SolanaWallet(props) {
  return /* @__PURE__ */ React19.createElement(
    InitSolanaProvider,
    {
      ...props.solanaConfig ?? { wallets: [], onError: () => {
      } }
    },
    /* @__PURE__ */ React19.createElement(SolanaWalletProvider, null, props.children)
  );
}
function InitWagmiProvider({
  children,
  initialState,
  initChains,
  wagmiConfig
}) {
  const { connectorWalletType } = useWalletConnectorPrivy();
  if (connectorWalletType.disableWagmi) {
    return children;
  }
  const [queryClient] = useState(() => new QueryClient());
  const [config] = useState(
    () => createConfig({
      chains: initChains && initChains.length ? initChains : [mainnet],
      multiInjectedProviderDiscovery: true,
      storage: wagmiConfig.storage ? wagmiConfig.storage : createStorage({
        storage: localStorage,
        key: "wagmi"
      }),
      connectors: wagmiConfig.connectors ? wagmiConfig.connectors : [injected()],
      transports: Object.fromEntries(
        initChains.map((chain) => [chain.id, http()])
      )
    })
  );
  return /* @__PURE__ */ React19.createElement(WagmiProvider, { config, initialState }, /* @__PURE__ */ React19.createElement(QueryClientProvider, { client: queryClient }, children));
}

// src/providers/wagmi/index.tsx
var WagmiWallet = (props) => {
  return /* @__PURE__ */ React19.createElement(
    InitWagmiProvider,
    {
      initChains: props.initChains,
      wagmiConfig: props.wagmiConfig ?? EMPTY_OBJECT
    },
    /* @__PURE__ */ React19.createElement(WagmiWalletProvider, null, props.children)
  );
};

// src/provider.tsx
var testnetChainFallback = [ArbitrumSepoliaChainInfo, SolanaDevnetChainInfo];
var processChainInfo = (chainInfo) => chainInfo.map(
  (row) => defineChain({
    id: Number(row.chain_id),
    name: row.name,
    nativeCurrency: {
      decimals: row.currency_decimal,
      name: "Ether",
      symbol: row.currency_symbol
    },
    rpcUrls: {
      default: {
        http: [row.public_rpc_url]
      }
    },
    blockExplorers: {
      default: { name: "Explorer", url: row.explorer_base_url }
    }
  })
) || [];
var WalletConnectorPrivyContext = createContext({
  initChains: [mainnet],
  mainnetChains: [],
  testnetChains: [],
  getChainsByNetwork: () => [],
  openConnectDrawer: false,
  setOpenConnectDrawer: () => {
  },
  targetWalletType: void 0,
  setTargetWalletType: () => {
  },
  network: "mainnet" /* mainnet */,
  setNetwork: () => {
  },
  solanaInfo: null,
  setSolanaInfo: () => {
  },
  termsOfUse: "",
  walletChainType: "EVM_SOL" /* EVM_SOL */,
  walletChainTypeConfig: {
    hasEvm: true,
    hasSol: true,
    hasAbstract: false
  },
  connectorWalletType: {
    disableWagmi: false,
    disablePrivy: false,
    disableSolana: false,
    disableAGW: false
  },
  privyConfig: {
    loginMethods: []
  }
});
var useWalletConnectorPrivy = () => useContext(WalletConnectorPrivyContext);
var defaultPrivyLoginMethod = [
  "email",
  "google",
  "twitter"
];
function WalletConnectorPrivyProvider(props) {
  const [walletChainType, setWalletChainType] = useState(
    "EVM_SOL" /* EVM_SOL */
  );
  const [termsOfUse] = useState(props.termsOfUse || "");
  const [network, setNetwork] = useState(props.network);
  const [initChains, setInitChains] = useState([]);
  const [mainnetChains, setMainnetChains] = useState([]);
  const [testnetChains, setTestnetChains] = useState([]);
  const [mainnetChainInfos, setMainnetChainInfos] = useState(null);
  const [testChainInfos, setTestChainInfos] = useState(
    null
  );
  const fetchMainChains = useMainnetChainsStore((state) => state.fetchData);
  const fetchTestChains = useTestnetChainsStore((state) => state.fetchData);
  const mainnetChainsHydrated = useMainnetChainsStore(
    (state) => state.hydrated
  );
  const testChainsHydrated = useTestnetChainsStore((state) => state.hydrated);
  const mainnetChainInfosFromStore = useMainnetChainsStore(
    (state) => state.data
  );
  const testChainInfosFromStore = useTestnetChainsStore((state) => state.data);
  const hasCustomChains = Array.isArray(props.customChains) && props.customChains.length > 0;
  const initRef = useRef(hasCustomChains);
  const [openConnectDrawer, setOpenConnectDrawer] = useState(false);
  const [targetWalletType, setTargetWalletType] = useState();
  const [privyConfig, setPrivyConfig] = useState({
    loginMethods: props.privyConfig ? props.privyConfig?.config?.loginMethods ?? defaultPrivyLoginMethod : defaultPrivyLoginMethod
  });
  const [solanaInfo, setSolanaInfo] = useState(null);
  const connectorWalletType = useMemo(() => {
    const type = {
      disableWagmi: false,
      disablePrivy: false,
      disableSolana: false,
      disableAGW: false
    };
    if (!props.privyConfig) {
      type.disablePrivy = true;
    }
    if (!props.wagmiConfig) {
      type.disableWagmi = true;
    }
    if (!props.solanaConfig) {
      type.disableSolana = true;
    }
    if (!props.abstractConfig) {
      type.disableAGW = true;
    }
    return type;
  }, [props.privyConfig, props.wagmiConfig, props.solanaConfig]);
  const walletChainTypeConfig = useMemo(() => {
    const chainTypeObj = {
      hasEvm: false,
      hasSol: false,
      hasAbstract: false
    };
    initChains.forEach((chain) => {
      if (SolanaChains$1.has(chain.id)) {
        chainTypeObj.hasSol = true;
      } else if (AbstractChains.has(chain.id)) {
        chainTypeObj.hasAbstract = true;
      } else {
        chainTypeObj.hasEvm = true;
      }
    });
    return chainTypeObj;
  }, [initChains]);
  useEffect(() => {
    if (!mainnetChainsHydrated || !testChainsHydrated) return;
    if (Array.isArray(useMainnetChainsStore.getState().data) && useTestnetChainsStore.getState().data) {
      return;
    }
    fetchMainChains().then((data) => {
      setMainnetChainInfos(data);
    });
    fetchTestChains().then((data) => {
      setTestChainInfos(data);
    });
  }, [mainnetChainsHydrated, testChainsHydrated]);
  const handleCustomerChains = () => {
    const testChains = processChainInfo(
      props.customChains.testnet?.map((item) => item.network_infos)
    );
    const mainnetChains2 = processChainInfo(
      props.customChains.mainnet?.map((item) => item.network_infos)
    );
    setTestnetChains(testChains);
    setMainnetChains(mainnetChains2);
    setInitChains([...testChains, ...mainnetChains2]);
    const chainTypeObj = {
      hasEvm: false,
      hasSol: false,
      hasAbstract: false
    };
    [...testChains, ...mainnetChains2].forEach((chain) => {
      if (SolanaChains$1.has(chain.id)) {
        chainTypeObj.hasSol = true;
      } else if (AbstractChains.has(chain.id)) {
        chainTypeObj.hasAbstract = true;
      } else {
        chainTypeObj.hasEvm = true;
      }
    });
    initRef.current = true;
    if (chainTypeObj.hasEvm && chainTypeObj.hasSol) {
      setWalletChainType("EVM_SOL" /* EVM_SOL */);
    } else if (chainTypeObj.hasEvm) {
      setWalletChainType("onlyEVM" /* onlyEVM */);
    } else if (chainTypeObj.hasSol) {
      setWalletChainType("onlySOL" /* onlySOL */);
    }
  };
  if (connectorWalletType.disablePrivy && connectorWalletType.disableWagmi && connectorWalletType.disableSolana && connectorWalletType.disableAGW) {
    throw new Error(
      "Privy, Wagmi, Solana, Abstract are all disabled. Please enable at least one of them."
    );
  }
  const getChainsByNetwork = (network2) => {
    return network2 === "mainnet" ? mainnetChains : testnetChains;
  };
  const value = useMemo(
    () => ({
      initChains,
      mainnetChains,
      testnetChains,
      getChainsByNetwork,
      openConnectDrawer,
      setOpenConnectDrawer,
      targetWalletType,
      setTargetWalletType,
      network,
      setNetwork,
      solanaInfo,
      setSolanaInfo,
      termsOfUse,
      walletChainType,
      connectorWalletType,
      walletChainTypeConfig,
      privyConfig
    }),
    [
      initChains,
      mainnetChains,
      testnetChains,
      getChainsByNetwork,
      openConnectDrawer,
      setOpenConnectDrawer,
      targetWalletType,
      setTargetWalletType,
      network,
      setNetwork,
      solanaInfo,
      setSolanaInfo,
      termsOfUse,
      walletChainType,
      connectorWalletType,
      walletChainTypeConfig,
      privyConfig
    ]
  );
  useEffect(() => {
    if (initRef.current) return;
    if (hasCustomChains) {
      return;
    }
    const hasStoreData = mainnetChainInfosFromStore && testChainInfosFromStore;
    const hasApiData = mainnetChainInfos && testChainInfos;
    if (!hasStoreData && !hasApiData) {
      return;
    }
    let testChainsList = [];
    let mainnetChainsList = [];
    try {
      if (hasStoreData) {
        testChainsList = testChainInfosFromStore;
        mainnetChainsList = mainnetChainInfosFromStore;
      } else {
        testChainsList = testChainInfos || testnetChainFallback;
        mainnetChainsList = mainnetChainInfos || [];
      }
      const testChains = processChainInfo(testChainsList);
      const mainnetChains2 = processChainInfo(mainnetChainsList);
      const chains = [...testChains, ...mainnetChains2];
      setTestnetChains(testChains);
      setMainnetChains(mainnetChains2);
      setInitChains(chains);
    } catch (error) {
      testChainsList = [ArbitrumSepoliaChainInfo, SolanaDevnetChainInfo];
      mainnetChainsList = [];
    }
    initRef.current = true;
  }, [
    props.customChains,
    mainnetChainInfos,
    testChainInfos,
    mainnetChainInfosFromStore,
    testChainInfosFromStore
  ]);
  useEffect(() => {
    if (props.customChains) {
      handleCustomerChains();
    }
  }, [props.customChains]);
  if (!initRef.current) {
    return null;
  }
  return /* @__PURE__ */ React19.createElement(WalletConnectorPrivyContext.Provider, { value }, /* @__PURE__ */ React19.createElement(TooltipProvider, { delayDuration: 300 }, /* @__PURE__ */ React19.createElement(PrivyWallet, { privyConfig: props.privyConfig, initChains }, /* @__PURE__ */ React19.createElement(WagmiWallet, { wagmiConfig: props.wagmiConfig, initChains }, /* @__PURE__ */ React19.createElement(SolanaWallet, { solanaConfig: props.solanaConfig }, /* @__PURE__ */ React19.createElement(AbstractWallet, null, /* @__PURE__ */ React19.createElement(Main, { headerProps: props.headerProps }, props.children)))))));
}

// src/index.ts
var viem = viemExport;
var wagmiConnectors = WagmiConnectorsExport;
var wagmi = WagmiExport;

export { AbstractChainsMap, MwebUserCenter, Network, SolanaChains, SolanaChainsMap, UserCenter, WalletChainTypeEnum, WalletConnectType, WalletConnectorPrivyProvider, WalletType, useWalletConnectorPrivy, viem, wagmi, wagmiConnectors };
