// @orderly.network/wallet-connector/dist/index.mjs
import React, { createContext, useMemo, useState, useEffect, useRef, useCallback, useContext } from 'react';
import { WalletAdapterNetwork, WalletNotReadyError, WalletReadyState } from '@solana/wallet-adapter-base';
import { EMPTY_OBJECT, ChainNamespace } from '@orderly.network/types';
import { SolanaMobileWalletAdapter, createDefaultAuthorizationResultCache, createDefaultAddressSelector } from '@solana-mobile/wallet-adapter-mobile';
import { WalletProvider, useWallet } from '@solana/wallet-adapter-react';
import { WalletModalProvider, useWalletModal } from '@solana/wallet-adapter-react-ui';
import { PhantomWalletAdapter } from '@solana/wallet-adapter-wallets';
import { useEventEmitter, useSimpleDI, WalletConnectorContext, useStorageLedgerAddress } from '@orderly.network/hooks';
import { useScreen } from '@orderly.network/ui';
import { getGlobalObject, hex2int } from '@orderly.network/utils';
import '@solana/wallet-adapter-react-ui/styles.css';
import injectedModule from '@web3-onboard/injected-wallets';
import { init, useConnectWallet, useSetChain } from '@web3-onboard/react';
import binanceModule from '@binance/w3w-blocknative-connector';
import { merge } from 'lodash';

// src/provider.tsx
var SolanaContext = createContext({
  network: WalletAdapterNetwork.Devnet,
  endpoint: ""
});
var useSolanaContext = () => {
  const context = useContext(SolanaContext);
  if (context === void 0) {
    throw new Error("useSolanaContext must be used within a SolanaProvider");
  }
  return context;
};
function SolanaProvider({ children, ...props }) {
  const { isMobile } = useScreen();
  const network = useMemo(
    () => props.network ?? WalletAdapterNetwork.Devnet,
    [props.network]
  );
  const endpoint = useMemo(() => {
    if (network === WalletAdapterNetwork.Mainnet) {
      return props.mainnetRpc ?? "";
    }
    if (network === WalletAdapterNetwork.Devnet) {
      return props.devnetRpc ?? "";
    }
    return "";
  }, [network, props.mainnetRpc, props.devnetRpc]);
  const mobileWalletNotFoundHanlder = (adapter) => {
    console.log("-- mobile wallet adapter", adapter);
    return Promise.reject(new WalletNotReadyError("wallet not ready"));
  };
  const ee = useEventEmitter();
  const handleSolanaError = (error, adapter) => {
    console.log("-- solanan error", error);
    console.log("-- solana adapter", adapter);
    if (error instanceof WalletNotReadyError) {
      if (isMobile) {
        ee.emit("wallet:connect-error", {
          message: "Please open the wallet app and use the in-app browser."
        });
      } else {
        window.open(adapter?.url, "_blank");
      }
    } else {
      ee.emit("wallet:connect-error", {
        message: error?.message || "Please switch to a wallet with Solana address."
      });
    }
  };
  const wallets = useMemo(() => {
    let uri = "";
    if (typeof window !== "undefined") {
      const location = getGlobalObject().location;
      uri = `${location.protocol}//${location.host}`;
    }
    return props.wallets ?? [
      new PhantomWalletAdapter(),
      new SolanaMobileWalletAdapter({
        addressSelector: createDefaultAddressSelector(),
        appIdentity: {
          uri
        },
        authorizationResultCache: createDefaultAuthorizationResultCache(),
        chain: network,
        onWalletNotFound: mobileWalletNotFoundHanlder
      })
    ];
  }, [props.wallets, network]);
  const contextValue = useMemo(
    () => ({
      network,
      endpoint
    }),
    [network, endpoint]
  );
  return /* @__PURE__ */ React.createElement(SolanaContext.Provider, { value: contextValue }, /* @__PURE__ */ React.createElement(
    WalletProvider,
    {
      wallets,
      onError: props.onError ?? handleSolanaError,
      autoConnect: true
    },
    /* @__PURE__ */ React.createElement(WalletModalProvider, { className: "oui-pointer-events-auto" }, children)
  ));
}
var binance = binanceModule({ options: { lng: "en" } });
var initConfig = (apiKey, options) => {
  const defaultOptions = {
    apiKey,
    connect: {
      // autoConnectAllPreviousWallet: true,
      autoConnectLastWallet: true
    },
    wallets: [injectedModule(), binance],
    chains: [],
    appMetadata: {
      name: "Orderly",
      // icon: blocknativeIcon,
      description: "Orderly",
      recommendedInjectedWallets: [
        { name: "Coinbase", url: "https://wallet.coinbase.com/" },
        { name: "MetaMask", url: "https://metamask.io" },
        { name: "Trezor", url: "https://trezor.io/" },
        { name: "Walletconnect", url: "https://walletconnect.com/" },
        { name: "Ledger", url: "https://www.ledger.com/" }
      ],
      agreement: {
        version: "1.0.0",
        termsUrl: "https://www.blocknative.com/terms-conditions",
        privacyUrl: "https://www.blocknative.com/privacy-policy"
      },
      gettingStartedGuide: "https://blocknative.com",
      explore: "https://blocknative.com"
    },
    accountCenter: {
      desktop: {
        enabled: false
      },
      mobile: {
        enabled: false
      }
    },
    theme: "dark"
  };
  const mergedOptions = merge(defaultOptions, options);
  return init(mergedOptions);
};
var SolanaChains = /* @__PURE__ */ new Map([[WalletAdapterNetwork.Devnet, 901901901], [WalletAdapterNetwork.Mainnet, 900900900]]);
function InitEvm(props) {
  const [initialized, setInitialized] = useState(!!props.skipInit);
  const { get, register } = useSimpleDI();
  useEffect(() => {
    document.body.style.setProperty("--onboard-modal-z-index", "88");
  }, []);
  useEffect(() => {
    if (props.skipInit) {
      return;
    }
    let onboardAPI = get("onboardAPI");
    if (onboardAPI) {
      console.log("[Orderly SDK]:onboardAPI already initialized");
      setInitialized(true);
      return;
    }
    Promise.all([
      fetchChainInfo("https://testnet-api.orderly.org/v1/public/chain_info"),
      fetchChainInfo("https://api.orderly.org/v1/public/chain_info")
    ]).then(([testChainInfo, mainnetChainInfo]) => {
      const testChains = processChainInfo(testChainInfo);
      const mainnetChains = processChainInfo(mainnetChainInfo);
      let options = props.options || {};
      options = merge({ chains: [...testChains, ...mainnetChains] }, options);
      onboardAPI = initConfig(props.apiKey, options);
      register("onboardAPI", onboardAPI);
      setInitialized(true);
    }).catch((error) => {
      console.error("Error fetching data:", error);
    });
  }, []);
  if (!initialized) return null;
  return props.children;
}
var fetchChainInfo = async (url) => {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch data from ${url}`);
  }
  return response.json();
};
var processChainInfo = (chainInfo) => chainInfo?.data?.rows?.map((row) => ({
  id: Number(row.chain_id),
  token: row.currency_symbol,
  label: row.name,
  rpcUrl: row.public_rpc_url,
  blockExplorerUrl: row.explorer_base_url
})) || [];
function useEvm() {
  const [
    {
      wallet,
      // the wallet that has been connected or null if not yet connected
      connecting
      // boolean indicating if connection is in progress
    },
    connect,
    // function to call to initiate user to connect wallet
    evmDisconnect,
    // function to call with wallet<DisconnectOptions> to disconnect wallet
    updateBalances,
    // function to be called with an optional array of wallet addresses connected through Onboard to update balance or empty/no params to update all connected wallets
    setWalletModules,
    // function to be called with an array of wallet modules to conditionally allow connection of wallet types i.e. setWalletModules([ledger, trezor, injected])
    setPrimaryWallet
    // function that can set the primary wallet and/or primary account within that wallet. The wallet that is set needs to be passed in for the first parameter and if you would like to set the primary account, the address of that account also needs to be passed in
  ] = useConnectWallet();
  const [
    {
      chains,
      // the list of chains that web3-onboard was initialized with
      connectedChain: evmConnectChain,
      // the current chain the user's wallet is connected to
      settingChain
      // boolean indicating if the chain is in the process of being set
    },
    setChain
    // function to call to initiate user to switch chains in their wallet
  ] = useSetChain();
  const connected = useMemo(() => {
    return !!(wallet && wallet.accounts && wallet.accounts[0] && wallet.accounts[0].address);
  }, [wallet]);
  const disconnect = async () => {
    console.log("-- disconnect evm xxxxxxxxxxxx");
    if (!wallet) {
      return;
    }
    return evmDisconnect({
      label: wallet.label
    });
  };
  const connectedChain = useMemo(() => {
    return evmConnectChain ? { ...evmConnectChain, id: parseInt(evmConnectChain.id) } : null;
  }, [evmConnectChain]);
  const changeChain = (chain) => {
    return setChain(chain);
  };
  return {
    connect,
    connected,
    disconnect,
    connecting,
    wallet,
    connectedChain,
    changeChain
  };
}
function useSOL() {
  const [wallet, setWallet] = useState(null);
  const { isMobile } = useScreen();
  const { endpoint, network } = useSolanaContext();
  const { setVisible: setModalVisible, visible } = useWalletModal();
  const { setLedgerAddress } = useStorageLedgerAddress();
  const {
    signMessage,
    signTransaction,
    sendTransaction,
    publicKey,
    wallet: solanaWallet,
    connect: solanaConnect,
    disconnect: solanaDisconnect,
    connecting
  } = useWallet();
  const selectModalVisibleRef = useRef(false);
  const [connected, setConnected] = useState(false);
  const isManual = useRef(false);
  const solanaPromiseRef = useRef({
    walletSelect: null,
    connect: null,
    walletSelectResolve: () => {
    },
    walletSelectReject: () => {
    },
    connectReject: () => {
    },
    connectResolve: () => {
    }
  });
  const ee = useEventEmitter();
  const initPromiseRef = () => {
    console.log("-- init solana promise");
    solanaPromiseRef.current.walletSelectResolve = () => {
    };
    solanaPromiseRef.current.walletSelectReject = () => {
    };
    solanaPromiseRef.current.connectReject = () => {
    };
    solanaPromiseRef.current.connectReject = () => {
    };
    solanaPromiseRef.current.connect = null;
    solanaPromiseRef.current.walletSelect = null;
    solanaPromiseRef.current.walletSelect = new Promise((resolve, reject) => {
      solanaPromiseRef.current.walletSelectResolve = resolve;
      solanaPromiseRef.current.walletSelectReject = reject;
    });
    solanaPromiseRef.current.connect = new Promise((resolve, reject) => {
      solanaPromiseRef.current.connectResolve = resolve;
      solanaPromiseRef.current.connectReject = reject;
    });
  };
  const handleSolanaError = useCallback(
    (e) => {
      console.log("solan connect error", e);
      if (e instanceof WalletNotReadyError) {
        if (isMobile) {
          console.log("-- need toast wallet not ready", e);
          ee.emit("wallet:connect-error", {
            message: "Please open the wallet app and use the in-app browser."
          });
        }
      }
      solanaDisconnect().then();
    },
    [isMobile, ee]
  );
  const connect = async () => {
    initPromiseRef();
    isManual.current = true;
    if (!solanaPromiseRef.current) {
      return;
    }
    if (!solanaWallet) {
      setModalVisible(true);
      selectModalVisibleRef.current = true;
    } else {
      solanaPromiseRef.current.walletSelectResolve(solanaWallet);
      if (!publicKey) {
        try {
          await solanaConnect();
        } catch (e) {
          solanaPromiseRef.current.connectReject(e);
        }
      } else {
        solanaPromiseRef.current.connectResolve({
          userAddress: publicKey.toBase58(),
          signMessage,
          signTransaction,
          sendTransaction
        });
      }
    }
    console.log("-- connect fn", solanaWallet, publicKey);
    return Promise.all([
      solanaPromiseRef.current.walletSelect,
      solanaPromiseRef.current.connect
    ]).then(
      ([
        wallet2,
        { userAddress, signMessage: signMessage2, signTransaction: signTransaction2, sendTransaction: sendTransaction2 }
      ]) => {
        const tempWallet = {
          label: wallet2.adapter.name,
          icon: "",
          provider: {
            rpcUrl: endpoint,
            network,
            signMessage: signMessage2,
            signTransaction: signTransaction2,
            sendTransaction: sendTransaction2,
            publicKey
          },
          accounts: [
            {
              address: userAddress
            }
          ],
          chains: [
            {
              id: SolanaChains.get(network),
              namespace: ChainNamespace.solana
            }
          ]
        };
        if (wallet2.adapter.name === "Ledger") {
          setLedgerAddress(userAddress);
        }
        setWallet(tempWallet);
        setConnected(true);
        return [tempWallet];
      }
    ).catch((e) => {
      console.log("connect solana error", e);
      handleSolanaError(e);
      return Promise.reject(e);
    }).finally(() => {
      isManual.current = false;
    });
  };
  const disconnect = async () => {
    console.log("--- discconnect sol");
    await solanaDisconnect();
    setWallet(null);
    setConnected(false);
    return [];
  };
  const connectedChain = useMemo(() => {
    if (!publicKey) {
      return null;
    }
    return {
      id: network === WalletAdapterNetwork.Mainnet ? 900900900 /* MAINNET */ : 901901901 /* DEVNET */,
      namespace: ChainNamespace.solana
    };
  }, [publicKey]);
  useEffect(() => {
    if (selectModalVisibleRef.current) {
      if (!visible && !solanaWallet && solanaPromiseRef.current) {
        console.log(
          "-- select modal visible ref",
          selectModalVisibleRef.current
        );
        console.log("-- use reject solana select modal");
        solanaPromiseRef.current.walletSelectReject("user reject");
        selectModalVisibleRef.current = false;
      } else if (solanaWallet) {
        selectModalVisibleRef.current = false;
      }
    }
  }, [
    visible,
    solanaWallet,
    solanaPromiseRef.current,
    selectModalVisibleRef.current
  ]);
  useEffect(() => {
    if (!solanaWallet || !publicKey) {
      console.log("--- not connect sol", solanaWallet, publicKey);
      setConnected(false);
      return;
    }
    console.log("-- publick", {
      publicKey: publicKey.toBase58(),
      isManual: isManual.current
    });
    if (isManual.current) {
      if (solanaPromiseRef.current) {
        solanaPromiseRef.current.connectResolve({
          userAddress: publicKey?.toBase58(),
          signMessage,
          signTransaction,
          sendTransaction
        });
      }
      return;
    }
    const userAddress = publicKey.toBase58();
    setWallet({
      label: solanaWallet.adapter.name,
      icon: "",
      provider: {
        signMessage,
        signTransaction,
        sendTransaction,
        rpcUrl: endpoint,
        network,
        publicKey
      },
      accounts: [
        {
          address: userAddress
        }
      ],
      chains: [
        {
          id: SolanaChains.get(network),
          namespace: ChainNamespace.solana
        }
      ]
    });
    if (solanaWallet.adapter.name === "Ledger") {
      setLedgerAddress(userAddress);
    }
    setConnected(true);
  }, [
    publicKey,
    solanaWallet,
    signMessage,
    signTransaction,
    isManual,
    sendTransaction,
    endpoint,
    network
  ]);
  useEffect(() => {
    if (!solanaWallet) {
      return;
    }
    console.log("-- public key", publicKey, { isMobile });
    if (isMobile && solanaWallet.readyState === WalletReadyState.Loadable && !isManual.current) {
      solanaDisconnect().then();
      return;
    }
    console.log("-- solana refresh auto connect", solanaWallet);
    if (solanaPromiseRef.current) {
      solanaPromiseRef.current.walletSelectResolve(solanaWallet);
    }
  }, [solanaWallet, publicKey, handleSolanaError, isMobile]);
  return {
    connected,
    connect,
    disconnect,
    connecting,
    wallet,
    connectedChain
  };
}

// src/main.tsx
var Main = (props) => {
  const sol = useSOL();
  const evm = useEvm();
  const [namespace, setNamespace] = useState(null);
  const newNamespace = useRef();
  const connect = async (options) => {
    if (Array.from(SolanaChains.values()).includes(options.chainId)) {
      newNamespace.current = ChainNamespace.solana;
      return sol.connect().then((res) => {
        if (res) {
          return res;
        }
      });
    }
    newNamespace.current = ChainNamespace.evm;
    const evmOption = options.autoSelect ? {
      autoSelect: options.autoSelect
    } : void 0;
    return evm.connect(evmOption).then((res) => {
      if (!res.length) {
        return Promise.reject({ message: "user reject" });
      }
      return res;
    }).catch((e) => {
      return Promise.reject(e);
    });
  };
  const disconnect = async () => {
    if (namespace === ChainNamespace.evm) {
      return evm.disconnect();
    }
    if (namespace === ChainNamespace.solana) {
      return sol.disconnect();
    }
  };
  const connecting = newNamespace.current == ChainNamespace.solana ? sol.connecting : evm.connecting;
  const wallet = namespace === ChainNamespace.solana && sol.connected ? sol.wallet : namespace === ChainNamespace.evm && evm.connected ? evm.wallet : null;
  const connectedChain = namespace === ChainNamespace.solana ? sol.connectedChain : evm.connectedChain;
  const setChain = (chain) => {
    const chainId = typeof chain.chainId === "number" ? chain.chainId : hex2int(chain.chainId);
    let tempNamespace = ChainNamespace.evm;
    if (Array.from(SolanaChains.values()).includes(chainId)) {
      tempNamespace = ChainNamespace.solana;
    }
    if (namespace === tempNamespace && namespace === ChainNamespace.evm) {
      return evm.changeChain(chain);
    }
    if (namespace !== tempNamespace) {
      return connect({ chainId }).then();
    }
  };
  useEffect(() => {
    if (sol.connected && evm.connected) {
      if (newNamespace.current === ChainNamespace.solana) {
        evm.disconnect().then();
        setNamespace(ChainNamespace.solana);
        return;
      } else {
        setNamespace(ChainNamespace.evm);
        sol.disconnect().then();
        return;
      }
    }
    if (sol.connected) {
      setNamespace(ChainNamespace.solana);
      return;
    }
    if (evm.connected) {
      setNamespace(ChainNamespace.evm);
      return;
    }
  }, [newNamespace.current, sol.connected, evm.connected]);
  const memoizedValue = useMemo(() => {
    return {
      connect,
      disconnect,
      connecting,
      wallet,
      setChain,
      connectedChain,
      namespace,
      chains: [],
      settingChain: false
    };
  }, [
    connect,
    disconnect,
    connecting,
    wallet,
    setChain,
    connectedChain,
    namespace
  ]);
  return /* @__PURE__ */ React.createElement(WalletConnectorContext.Provider, { value: memoizedValue }, props.children);
};

// src/provider.tsx
var WalletConnectorProvider = (props) => {
  return /* @__PURE__ */ React.createElement(SolanaProvider, { ...props.solanaInitial ?? EMPTY_OBJECT }, /* @__PURE__ */ React.createElement(InitEvm, { ...props.evmInitial ?? EMPTY_OBJECT }, /* @__PURE__ */ React.createElement(
    Main,
    {
      solanaNetwork: props.solanaInitial?.network ?? WalletAdapterNetwork.Devnet
    },
    props.children
  )));
};

export { WalletConnectorProvider };
