// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/accounts/confirmations.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Option,
  OptionOrNullable,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  option,
  struct,
  u64,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type Confirmations = Account<ConfirmationsAccountData>;

export type ConfirmationsAccountData = {
  discriminator: Uint8Array;
  value: Option<bigint>;
  bump: number;
};

export type ConfirmationsAccountDataArgs = {
  value: OptionOrNullable<number | bigint>;
  bump: number;
};

export function getConfirmationsAccountDataSerializer(): Serializer<
  ConfirmationsAccountDataArgs,
  ConfirmationsAccountData
> {
  return mapSerializer<
    ConfirmationsAccountDataArgs,
    any,
    ConfirmationsAccountData
  >(
    struct<ConfirmationsAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['value', option(u64())],
        ['bump', u8()],
      ],
      { description: 'ConfirmationsAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([206, 57, 50, 8, 124, 133, 138, 112]),
    })
  ) as Serializer<ConfirmationsAccountDataArgs, ConfirmationsAccountData>;
}

export function deserializeConfirmations(
  rawAccount: RpcAccount
): Confirmations {
  return deserializeAccount(
    rawAccount,
    getConfirmationsAccountDataSerializer()
  );
}

export async function fetchConfirmations(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<Confirmations> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'Confirmations');
  return deserializeConfirmations(maybeAccount);
}

export async function safeFetchConfirmations(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<Confirmations | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeConfirmations(maybeAccount) : null;
}

export async function fetchAllConfirmations(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<Confirmations[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'Confirmations');
    return deserializeConfirmations(maybeAccount);
  });
}

export async function safeFetchAllConfirmations(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<Confirmations[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) =>
      deserializeConfirmations(maybeAccount as RpcAccount)
    );
}

export function getConfirmationsGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      value: OptionOrNullable<number | bigint>;
      bump: number;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      value: [8, option(u64())],
      bump: [null, u8()],
    })
    .deserializeUsing<Confirmations>((account) =>
      deserializeConfirmations(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([206, 57, 50, 8, 124, 133, 138, 112])
    );
}
