// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/accounts/receiveConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';
import { UlnConfig, UlnConfigArgs, getUlnConfigSerializer } from '../types';

export type ReceiveConfig = Account<ReceiveConfigAccountData>;

export type ReceiveConfigAccountData = {
  discriminator: Uint8Array;
  bump: number;
  uln: UlnConfig;
};

export type ReceiveConfigAccountDataArgs = { bump: number; uln: UlnConfigArgs };

export function getReceiveConfigAccountDataSerializer(): Serializer<
  ReceiveConfigAccountDataArgs,
  ReceiveConfigAccountData
> {
  return mapSerializer<
    ReceiveConfigAccountDataArgs,
    any,
    ReceiveConfigAccountData
  >(
    struct<ReceiveConfigAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['bump', u8()],
        ['uln', getUlnConfigSerializer()],
      ],
      { description: 'ReceiveConfigAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([162, 159, 153, 188, 56, 65, 245, 58]),
    })
  ) as Serializer<ReceiveConfigAccountDataArgs, ReceiveConfigAccountData>;
}

export function deserializeReceiveConfig(
  rawAccount: RpcAccount
): ReceiveConfig {
  return deserializeAccount(
    rawAccount,
    getReceiveConfigAccountDataSerializer()
  );
}

export async function fetchReceiveConfig(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<ReceiveConfig> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'ReceiveConfig');
  return deserializeReceiveConfig(maybeAccount);
}

export async function safeFetchReceiveConfig(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<ReceiveConfig | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeReceiveConfig(maybeAccount) : null;
}

export async function fetchAllReceiveConfig(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<ReceiveConfig[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'ReceiveConfig');
    return deserializeReceiveConfig(maybeAccount);
  });
}

export async function safeFetchAllReceiveConfig(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<ReceiveConfig[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) =>
      deserializeReceiveConfig(maybeAccount as RpcAccount)
    );
}

export function getReceiveConfigGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      bump: number;
      uln: UlnConfigArgs;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      bump: [8, u8()],
      uln: [9, getUlnConfigSerializer()],
    })
    .deserializeUsing<ReceiveConfig>((account) =>
      deserializeReceiveConfig(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([162, 159, 153, 188, 56, 65, 245, 58])
    );
}
