// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/accounts/sendConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ExecutorConfig,
  ExecutorConfigArgs,
  UlnConfig,
  UlnConfigArgs,
  getExecutorConfigSerializer,
  getUlnConfigSerializer,
} from '../types';

export type SendConfig = Account<SendConfigAccountData>;

export type SendConfigAccountData = {
  discriminator: Uint8Array;
  bump: number;
  uln: UlnConfig;
  executor: ExecutorConfig;
};

export type SendConfigAccountDataArgs = {
  bump: number;
  uln: UlnConfigArgs;
  executor: ExecutorConfigArgs;
};

export function getSendConfigAccountDataSerializer(): Serializer<
  SendConfigAccountDataArgs,
  SendConfigAccountData
> {
  return mapSerializer<SendConfigAccountDataArgs, any, SendConfigAccountData>(
    struct<SendConfigAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['bump', u8()],
        ['uln', getUlnConfigSerializer()],
        ['executor', getExecutorConfigSerializer()],
      ],
      { description: 'SendConfigAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([91, 221, 135, 137, 131, 199, 174, 25]),
    })
  ) as Serializer<SendConfigAccountDataArgs, SendConfigAccountData>;
}

export function deserializeSendConfig(rawAccount: RpcAccount): SendConfig {
  return deserializeAccount(rawAccount, getSendConfigAccountDataSerializer());
}

export async function fetchSendConfig(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<SendConfig> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'SendConfig');
  return deserializeSendConfig(maybeAccount);
}

export async function safeFetchSendConfig(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<SendConfig | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeSendConfig(maybeAccount) : null;
}

export async function fetchAllSendConfig(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<SendConfig[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'SendConfig');
    return deserializeSendConfig(maybeAccount);
  });
}

export async function safeFetchAllSendConfig(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<SendConfig[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) => deserializeSendConfig(maybeAccount as RpcAccount));
}

export function getSendConfigGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      bump: number;
      uln: UlnConfigArgs;
      executor: ExecutorConfigArgs;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      bump: [8, u8()],
      uln: [9, getUlnConfigSerializer()],
      executor: [null, getExecutorConfigSerializer()],
    })
    .deserializeUsing<SendConfig>((account) => deserializeSendConfig(account))
    .whereField(
      'discriminator',
      new Uint8Array([91, 221, 135, 137, 131, 199, 174, 25])
    );
}
