// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/accounts/ulnSettings.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Option,
  OptionOrNullable,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  option,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u8,
} from '@metaplex-foundation/umi/serializers';
import { Treasury, TreasuryArgs, getTreasurySerializer } from '../types';

export type UlnSettings = Account<UlnSettingsAccountData>;

export type UlnSettingsAccountData = {
  discriminator: Uint8Array;
  eid: number;
  endpoint: PublicKey;
  endpointProgram: PublicKey;
  bump: number;
  admin: PublicKey;
  treasury: Option<Treasury>;
};

export type UlnSettingsAccountDataArgs = {
  eid: number;
  endpoint: PublicKey;
  endpointProgram: PublicKey;
  bump: number;
  admin: PublicKey;
  treasury: OptionOrNullable<TreasuryArgs>;
};

export function getUlnSettingsAccountDataSerializer(): Serializer<
  UlnSettingsAccountDataArgs,
  UlnSettingsAccountData
> {
  return mapSerializer<UlnSettingsAccountDataArgs, any, UlnSettingsAccountData>(
    struct<UlnSettingsAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['eid', u32()],
        ['endpoint', publicKeySerializer()],
        ['endpointProgram', publicKeySerializer()],
        ['bump', u8()],
        ['admin', publicKeySerializer()],
        ['treasury', option(getTreasurySerializer())],
      ],
      { description: 'UlnSettingsAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([226, 199, 100, 253, 38, 115, 167, 154]),
    })
  ) as Serializer<UlnSettingsAccountDataArgs, UlnSettingsAccountData>;
}

export function deserializeUlnSettings(rawAccount: RpcAccount): UlnSettings {
  return deserializeAccount(rawAccount, getUlnSettingsAccountDataSerializer());
}

export async function fetchUlnSettings(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<UlnSettings> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'UlnSettings');
  return deserializeUlnSettings(maybeAccount);
}

export async function safeFetchUlnSettings(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<UlnSettings | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeUlnSettings(maybeAccount) : null;
}

export async function fetchAllUlnSettings(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<UlnSettings[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'UlnSettings');
    return deserializeUlnSettings(maybeAccount);
  });
}

export async function safeFetchAllUlnSettings(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<UlnSettings[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) => deserializeUlnSettings(maybeAccount as RpcAccount));
}

export function getUlnSettingsGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'uln',
    '7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      eid: number;
      endpoint: PublicKey;
      endpointProgram: PublicKey;
      bump: number;
      admin: PublicKey;
      treasury: OptionOrNullable<TreasuryArgs>;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      eid: [8, u32()],
      endpoint: [12, publicKeySerializer()],
      endpointProgram: [44, publicKeySerializer()],
      bump: [76, u8()],
      admin: [77, publicKeySerializer()],
      treasury: [109, option(getTreasurySerializer())],
    })
    .deserializeUsing<UlnSettings>((account) => deserializeUlnSettings(account))
    .whereField(
      'discriminator',
      new Uint8Array([226, 199, 100, 253, 38, 115, 167, 154])
    );
}
