// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/pricefeed/types/ModelType.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
/**
 * This type is used to derive the {@link ModelType} type as well as the de/serializer.
 * However don't refer to it in your code but use the {@link ModelType} type instead.
 *
 * @category userTypes
 * @category enums
 * @category generated
 * @private
 */
export type ModelTypeRecord = {
  Arbitrum: { gasPerL2Tx: beet.bignum; gasPerL1CalldataByte: number }
  Optimism: { l1Eid: number }
}

/**
 * Union type respresenting the ModelType data enum defined in Rust.
 *
 * NOTE: that it includes a `__kind` property which allows to narrow types in
 * switch/if statements.
 * Additionally `isModelType*` type guards are exposed below to narrow to a specific variant.
 *
 * @category userTypes
 * @category enums
 * @category generated
 */
export type ModelType = beet.DataEnumKeyAsKind<ModelTypeRecord>

export const isModelTypeArbitrum = (
  x: ModelType
): x is ModelType & { __kind: 'Arbitrum' } => x.__kind === 'Arbitrum'
export const isModelTypeOptimism = (
  x: ModelType
): x is ModelType & { __kind: 'Optimism' } => x.__kind === 'Optimism'

/**
 * @category userTypes
 * @category generated
 */
export const modelTypeBeet = beet.dataEnum<ModelTypeRecord>([
  [
    'Arbitrum',
    new beet.BeetArgsStruct<ModelTypeRecord['Arbitrum']>(
      [
        ['gasPerL2Tx', beet.u64],
        ['gasPerL1CalldataByte', beet.u32],
      ],
      'ModelTypeRecord["Arbitrum"]'
    ),
  ],

  [
    'Optimism',
    new beet.BeetArgsStruct<ModelTypeRecord['Optimism']>(
      [['l1Eid', beet.u32]],
      'ModelTypeRecord["Optimism"]'
    ),
  ],
]) as beet.FixableBeet<ModelType, ModelType>
