// @layerzerolabs/lz-utilities/dist/index.cjs
'use strict';

var http = require('http');
var path = require('path');
var initia_js = require('@initia/initia.js');
var ed25519 = require('@mysten/sui/keypairs/ed25519');
var web3_js = require('@solana/web3.js');
var crypto = require('@ton/crypto');
var ton = require('@ton/ton');
var aptos = require('aptos');
var bip39 = require('bip39');
var ed25519HdKey = require('ed25519-hd-key');
var ethers = require('ethers');
var lzDefinitions = require('@layerzerolabs/lz-definitions');
var dayjs = require('dayjs');
var pc = require('picocolors');
var pino = require('pino');
var module2 = require('module');
var fs = require('fs');
var memoizee = require('memoizee');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var http__default = /*#__PURE__*/_interopDefault(http);
var path__namespace = /*#__PURE__*/_interopNamespace(path);
var aptos__namespace = /*#__PURE__*/_interopNamespace(aptos);
var bip39__namespace = /*#__PURE__*/_interopNamespace(bip39);
var ed25519HdKey__namespace = /*#__PURE__*/_interopNamespace(ed25519HdKey);
var dayjs__default = /*#__PURE__*/_interopDefault(dayjs);
var pc__default = /*#__PURE__*/_interopDefault(pc);
var module2__default = /*#__PURE__*/_interopDefault(module2);
var fs__namespace = /*#__PURE__*/_interopNamespace(fs);
var memoizee__default = /*#__PURE__*/_interopDefault(memoizee);

// src/index.ts

// src/pad.ts
var SizeExceedsPaddingSizeError = class extends Error {
  constructor({ size, targetSize, type }) {
    super(
      `${type.charAt(0).toUpperCase()}${type.slice(1).toLowerCase()} size (${size}) exceeds padding size (${targetSize})`
    );
    this.name = "SizeExceedsPaddingSizeError";
  }
};
function padify(hexOrBytes, { dir, size = 32 } = {}) {
  if (typeof hexOrBytes === "string") {
    return padHex(hexOrBytes, { dir, size });
  }
  return padBytes(hexOrBytes, { dir, size });
}
function padHex(hex, { dir, size = 32 } = {}) {
  if (size === null) return hex;
  const value = hex.replace("0x", "");
  if (value.length > size * 2)
    throw new SizeExceedsPaddingSizeError({
      size: Math.ceil(value.length / 2),
      targetSize: size,
      type: "hex"
    });
  return `0x${value[dir === "right" ? "padEnd" : "padStart"](size * 2, "0")}`;
}
function padBytes(bytes, { dir, size = 32 } = {}) {
  if (size === null) return bytes;
  if (bytes.length > size)
    throw new SizeExceedsPaddingSizeError({
      size: bytes.length,
      targetSize: size,
      type: "bytes"
    });
  const paddedBytes = new Uint8Array(size);
  for (let i = 0; i < size; i++) {
    const padEnd = dir === "right";
    paddedBytes[padEnd ? i : size - i - 1] = bytes[padEnd ? i : bytes.length - i - 1];
  }
  return paddedBytes;
}

// src/types.ts
function isHex(value) {
  return /^(0x)?[0-9A-F]+$/i.test(value);
}
function isHash(value) {
  return /^(0x)?[0-9A-F]+/i.test(value);
}

// src/format.ts
function bytesToHex(bytes) {
  return trim0x(hexlify(bytes)).replace(/^0x/i, "");
}
function hexToBytes(hex) {
  return arrayify(hex);
}
function trim0x(hex) {
  return hex.replace(/^0x/i, "");
}
function ensure0x(hex) {
  if (!isHex(hex)) {
    throw new Error("invalid hex string");
  }
  const value = trim0x(hex);
  const retval = `0x${value}`;
  return retval;
}
function isHexString(value) {
  return isHex(value);
}
function _arrayify(value) {
  if (value instanceof Uint8Array) {
    return value;
  }
  if (typeof value === "string") {
    if (value.match(/^(0x)?[0-9A-F]*$/i)) {
      const hex = value.replace(/^0x/i, "");
      const len = hex.length + 1 - (hex.length + 1) % 2;
      return Uint8Array.from(Buffer.from(hex.padStart(len, "0"), "hex"));
    }
    throw new Error("Invalid hex string");
  }
  if (typeof value === "number") {
    if (value < 0) {
      throw new Error("Number must be non-negative");
    }
    const byteArray = [];
    while (value > 0) {
      byteArray.push(value & 255);
      value >>= 8;
    }
    return new Uint8Array(byteArray.reverse());
  }
  if (typeof Buffer !== "undefined" && Buffer.isBuffer(value)) {
    return new Uint8Array(value);
  }
  if (typeof value === "bigint") {
    const hex = value.toString(16);
    return _arrayify(hex);
  }
  throw new Error("unsupported type");
}
function arrayify(value, size) {
  const bytes = _arrayify(value);
  if (size === void 0) {
    return bytes;
  }
  return padify(bytes, { size });
}
function hexlify(value) {
  if (typeof value === "string" && /^(0x)?[0-9A-F]*$/i.test(value)) {
    const retval = "0x" + trim0x(value);
    return retval;
  }
  const bytes = arrayify(value);
  const hex = Buffer.from(bytes).toString("hex");
  return ensure0x(hex);
}

// src/account.ts
function getBIP044Path(chainType, account, change, index) {
  switch (chainType) {
    case lzDefinitions.ChainType.EVM:
      return `m/44'/60'/${account}'/${change}/${index}`;
    case lzDefinitions.ChainType.APTOS:
      return `m/44'/637'/${account}'/${change}'/${index}'`;
    case lzDefinitions.ChainType.SOLANA:
      return `m/44'/501'/${account}'/${change}'`;
    case lzDefinitions.ChainType.TON:
      return `m/44'/607'/${account}'/${change}/${index}`;
    default:
      throw new Error(`Unsupported chain: ${chainType}`);
  }
}
function getEvmAccountFromMnemonic(mnemonic, path3 = "m/44'/60'/0'/0/0") {
  const wallet = ethers.ethers.Wallet.fromMnemonic(mnemonic, path3);
  return {
    mnemonic,
    path: path3,
    privateKey: wallet.privateKey,
    address: wallet.address
  };
}
function getAptosAccountFromMnemonic(mnemonic, path3 = "m/44'/637'/0'/0'/0'") {
  if (!aptos__namespace.AptosAccount.isValidPath(path3)) {
    throw new Error(`Invalid derivation path: ${path3}`);
  }
  const normalizeMnemonics = mnemonic.trim().split(/\s+/).map((part) => part.toLowerCase()).join(" ");
  {
    const { key } = aptos__namespace.derivePath(path3, trim0x(hexlify(bip39__namespace.mnemonicToSeedSync(normalizeMnemonics))));
    const account = new aptos__namespace.AptosAccount(new Uint8Array(key)).toPrivateKeyObject();
    return {
      mnemonic,
      path: path3,
      privateKey: account.privateKeyHex,
      address: account.address
    };
  }
}
function getInitiaAccountFromMnemonic(mnemonic, path3 = "m/44'/118'/0'/0/0") {
  const [_, coinType, account, __, index] = path3.match(/\d+/g)?.map(Number) ?? [44, 118, 0, 0, 0];
  const normalizeMnemonics = mnemonic.trim().split(/\s+/).map((part) => part.toLowerCase()).join(" ");
  const key = new initia_js.MnemonicKey({
    mnemonic: normalizeMnemonics,
    coinType,
    account,
    index,
    eth: coinType === initia_js.INIT_COIN_TYPE
  });
  {
    return {
      mnemonic,
      path: path3,
      privateKey: key.privateKey.toString("hex"),
      address: key.accAddress
    };
  }
}
function getSolanaAccountFromMnemonic(mnemonic, path3 = "m/44'/501'/0'/0'") {
  const seed = bip39__namespace.mnemonicToSeedSync(mnemonic, "");
  const keyPair = web3_js.Keypair.fromSeed(ed25519HdKey__namespace.derivePath(path3, seed.toString("hex")).key);
  return {
    mnemonic,
    path: path3,
    privateKey: ethers.ethers.utils.hexlify(keyPair.secretKey),
    address: keyPair.publicKey.toBase58()
  };
}
function getPlaceholderAccountFromMnemonic(mnemonic, path3 = "m/44'/784'/0'/0'/0'") {
  const signer = ed25519.Ed25519Keypair.deriveKeypair(mnemonic, path3);
  return {
    mnemonic,
    path: path3,
    privateKey: signer.getSecretKey(),
    address: signer.toSuiAddress()
  };
}
async function getTonAccountFromMnemonic(mnemonic, path3 = "m/44'/607'/0'/0'/0'", workchain = 0) {
  const { wallet, keyPair } = await getTonWalletFromMnemonic(mnemonic, path3, workchain);
  return {
    mnemonic,
    path: path3,
    privateKey: ethers.ethers.utils.hexlify(keyPair.secretKey),
    address: wallet.address.toString({ bounceable: false, urlSafe: true })
  };
}
async function getKeypairFromMnemonic(chainType, mnemonic, path3) {
  switch (chainType) {
    case lzDefinitions.ChainType.EVM:
      return getEvmAccountFromMnemonic(mnemonic, path3);
    case lzDefinitions.ChainType.APTOS:
      return getAptosAccountFromMnemonic(mnemonic, path3);
    case lzDefinitions.ChainType.INITIA:
      return getInitiaAccountFromMnemonic(mnemonic, path3);
    case lzDefinitions.ChainType.SOLANA:
      return getSolanaAccountFromMnemonic(mnemonic, path3);
    case lzDefinitions.ChainType.TON:
      return getTonAccountFromMnemonic(mnemonic, path3);
    case lzDefinitions.ChainType.SUI:
      return getPlaceholderAccountFromMnemonic(mnemonic, path3);
    default:
      throw new Error(`Unsupported chain: ${chainType}`);
  }
}
async function getTonWalletFromMnemonic(mnemonic, path3 = "m/44'/607'/0'/0'/0'", workchain = 0) {
  const seed = await crypto.mnemonicToHDSeed(mnemonic.split(" "));
  const indices = toPathArray(path3);
  const derivedSeed = await crypto.deriveEd25519Path(seed, indices);
  const keyPair = crypto.keyPairFromSeed(derivedSeed);
  return {
    wallet: ton.WalletContractV4.create({ publicKey: keyPair.publicKey, workchain }),
    keyPair
  };
}
function toPathArray(path3) {
  if (!/^[mM]'?/.test(path3)) {
    throw new Error('Path must start with "m" or "M"');
  }
  const parts = path3.replace(/^[mM]'?\//, "").split("/");
  const ret = Array(parts.length);
  for (let i = 0; i < parts.length; i++) {
    const tmp = /(\d+)[hH']/.exec(parts[i]);
    if (tmp === null) {
      throw new Error("Invalid input");
    }
    ret[i] = parseInt(tmp[1], 10);
    if (ret[i] >= 2147483648) {
      throw new Error("Invalid child index");
    }
  }
  return ret;
}
var logger = void 0;
function getStackTrace() {
  const oldLimit = Error.stackTraceLimit;
  Error.stackTraceLimit = Infinity;
  const retval = new Error().stack;
  Error.stackTraceLimit = oldLimit;
  return retval;
}
function getCircularReplacer() {
  const ancestors = [];
  return function(_key, value) {
    if (typeof value !== "object" || value === null) {
      return value;
    }
    while (ancestors.length > 0 && ancestors.at(-1) !== this) {
      ancestors.pop();
    }
    if (ancestors.includes(value)) {
      return "[Circular]";
    }
    ancestors.push(value);
    return value;
  };
}
function extractCallerInfo(line) {
  for (const pattern of [/\((.*?:\d+:\d+)\)$/, /at (.*?:\d+:\d+)$/]) {
    const m = line.match(pattern);
    if (m === null) {
      continue;
    }
    const [, fileInfoLine] = m;
    return fileInfoLine;
  }
  return "<unknown>";
}
function getCaller(pattern) {
  const stack = getStackTrace() ?? "";
  const stackLines = stack.split("\n");
  const index = stackLines.findIndex((line) => {
    return line.match(pattern);
  }) + 1;
  let fileInfo = "<unknown>";
  if (stackLines.length > index) {
    const line = stackLines[index];
    fileInfo = extractCallerInfo(line);
  }
  return fileInfo;
}
function initLogger(level) {
  if (!logger) {
    logger = createLogger(level);
  }
}
var levelColors = {
  TRACE: pc__default.default.reset,
  DEBUG: pc__default.default.reset,
  INFO: pc__default.default.green,
  WARN: pc__default.default.yellow,
  ERROR: pc__default.default.red,
  FATAL: pc__default.default.red
};
var customStream = {
  write: (msg) => {
    try {
      const customLogger = (output) => {
        if (typeof process !== "undefined") {
          process.stdout.write(output + "\n");
        } else {
          console.log(output);
        }
      };
      const logObj = JSON.parse(msg);
      const { level, msg: message, time } = logObj;
      const newMsgs = message.split("#caller#");
      const formattedTime = dayjs__default.default(time).format("YYYY-MM-DD HH:mm:ss");
      const colorFn = levelColors[level] || pc__default.default.reset;
      newMsgs.forEach((newMsg) => {
        customLogger(colorFn(`[${formattedTime}] [${level}] ${newMsg}`));
      });
    } catch (error) {
      console.error("Failed to process log:", msg);
    }
  }
};
function createLogger(level) {
  const logger3 = pino.pino(
    {
      level,
      base: void 0,
      // customLevels: customLevels.levels,
      // timestamp: () => `,"time":"${formatTime(Date.now())}"`,
      formatters: {
        level(levelLabel, levelNumber) {
          return { level: levelLabel.toUpperCase(), levelNumber };
        }
      },
      hooks: {
        logMethod(inputArgs, method, level2) {
          const [msg, ...args] = inputArgs;
          let newMsg = msg;
          let caller = "";
          if (level2 >= 50) {
            caller = getCaller(/lib\/tools.js/);
            newMsg = `${caller}#caller#${msg} `;
          }
          method.apply(this, [newMsg, ...args, caller]);
        }
      }
    },
    customStream
  );
  return logger3;
}
function getLogger() {
  initLogger(process.env.LZ_LOG ?? "info");
  if (!logger) {
    throw new Error("Logger is not initialized");
  }
  return logger;
}
if (process.env.NODE_ENV === "test") ;
function findDeployment(deployments, nameOrAddress, options) {
  const retval = tryFindDeployment(deployments, nameOrAddress, options);
  if (retval === void 0) {
    throw new Error(`Deployment not found: ${nameOrAddress} options:${JSON.stringify(options)}`);
  }
  return retval;
}
function tryFindDeployment(deployments, nameOrAddress, options) {
  return deployments.find((deployment) => {
    let hasMatchingNameOrAddress = deployment.name === nameOrAddress;
    if (!hasMatchingNameOrAddress && ethers.ethers.utils.isAddress(nameOrAddress)) {
      if (ethers.ethers.utils.isAddress(deployment.address)) {
        hasMatchingNameOrAddress = ethers.ethers.utils.getAddress(deployment.address) === ethers.ethers.utils.getAddress(nameOrAddress);
      } else {
        hasMatchingNameOrAddress = deployment.address === ethers.ethers.utils.getAddress(nameOrAddress);
      }
    }
    const hasMatchingChain = options.chain == null || options.chain === lzDefinitions.networkToChain(deployment.network);
    const hasMatchingNetwork = options.network === void 0 || options.network === deployment.network;
    const hasMatchingSource = options.source === void 0 || options.source === deployment.source;
    let hasMatchingEndpoint = true;
    if (options.endpointId != null) {
      const compatibleEndpoints = deployment.compatibleVersions.map((v) => {
        if (lzDefinitions.isNetworkEndpointIdSupported(deployment.network, v)) {
          return lzDefinitions.networkToEndpointId(deployment.network, v);
        }
        return void 0;
      });
      hasMatchingEndpoint = compatibleEndpoints.includes(options.endpointId);
    }
    return hasMatchingNameOrAddress && hasMatchingChain && hasMatchingNetwork && hasMatchingEndpoint && hasMatchingSource;
  });
}
var contractCache = {};
function deploymentToEvmContract(deployment, provider) {
  const key = `${deployment.network}-${deployment.address}`;
  if (!(key in contractCache)) {
    if (deployment.abi === void 0 || deployment.bytecode === void 0) {
      throw new Error("Deployment does not have ABI or bytecode");
    }
    const Contract = new ethers.ethers.ContractFactory(deployment.abi, deployment.bytecode);
    contractCache[key] = Contract.attach(deployment.address);
  }
  if (!provider) {
    return contractCache[key];
  }
  return contractCache[key].connect(provider);
}
function findContract(provider, deployments, nameOrAddress, options) {
  const deployment = findDeployment(deployments, nameOrAddress, options);
  const retval = deploymentToEvmContract(deployment, provider);
  return retval;
}
function dirname(path3) {
  const match = path3.match(/(.*)([\\/][^\\/]+)[/\\]?$/);
  const [, basePath] = match ?? [];
  const dirname3 = typeof basePath !== "undefined" ? basePath : path3;
  return dirname3;
}
function getStackTrace2(stackTraceLimit = Infinity) {
  const oldLimit = Error.stackTraceLimit;
  Error.stackTraceLimit = stackTraceLimit;
  const retval = new Error().stack;
  Error.stackTraceLimit = oldLimit;
  return retval;
}
function getCaller2() {
  const lines = (getStackTrace2(10) ?? "").split("\n");
  if (lines.length > 1 + 1 + 1 + 1) {
    const line = lines[4];
    const m = line.match(/^.*\(([^:]*)[^)]*\)/);
    if (m) {
      return m[1];
    }
  }
  return void 0;
}
function pkgroot(packageName, relativeToPath) {
  if (relativeToPath === void 0) {
    relativeToPath = getCaller2();
    if (relativeToPath === void 0) {
      relativeToPath = __filename;
    }
  }
  const filepath = module2__default.default.createRequire(relativeToPath).resolve(`${packageName}/package.json`);
  const packagePath = dirname(filepath);
  return packagePath.replace(/.yarn\/([^/]*\/)?(__virtual__|\$\$virtual)\/[^/]*\/\d*\//, "");
}

// src/assert.ts
function assert(condition, message) {
  if (!condition) {
    throw new Error(`Assertion Error:  ${message ?? "condition is false"}`);
  }
}
function assertType(value, typeOrGuard, message) {
  if (typeof typeOrGuard === "function") {
    if (typeOrGuard === String || typeOrGuard === Number || typeOrGuard === Boolean) {
      if (typeof value !== typeOrGuard.name.toLowerCase()) {
        throw new Error(`Expected value to be ${message ?? "of correct type"}`);
      }
    } else if ("prototype" in typeOrGuard && typeOrGuard.prototype !== void 0) {
      if (!(value instanceof typeOrGuard)) {
        throw new Error(`Expected value to be ${message ?? "of correct type"}`);
      }
    } else {
      if (!typeOrGuard(value)) {
        throw new Error(`Expected value to be ${message ?? "of correct type"}`);
      }
    }
  }
}
function assertDefined(value, message) {
  if (value === void 0 || value === null) {
    throw new Error(message ?? "Value is undefined or null");
  }
}
function asType(value, fn, message) {
  if (!fn(value)) {
    throw new Error(`Expected value to be ${message ?? "of correct type"}`);
  }
  return value;
}
function assumeType(value) {
}

// src/promise.ts
var sequence = async (tasks) => {
  const collector = [];
  for (const task of tasks) {
    collector.push(await task());
  }
  return collector;
};
var parallel = async (tasks) => Promise.all(tasks.map(async (task) => task()));
var first = async (tasks) => {
  assert(tasks.length !== 0, `Must have at least one task for first()`);
  let lastError;
  for (const task of tasks) {
    try {
      return await task();
    } catch (error) {
      lastError = error;
    }
  }
  throw lastError;
};
var firstFactory = (...factories) => async (...input) => first(factories.map((factory) => async () => factory(...input)));

// src/array.ts
function safeMap(elements, callbackfn) {
  const result = [];
  try {
    for (let i = 0; i < elements.length; i++) {
      const rv = callbackfn(elements[i], i);
      result.push(rv);
    }
    return [result, void 0];
  } catch (e) {
    return [result, e];
  }
}

// src/enum.ts
function asEnum(enumType, value) {
  const enumValues = Object.values(enumType);
  if (enumValues.includes(value)) {
    return value;
  } else {
    throw new Error(`Invalid enum value: ${value}`);
  }
}

// src/generic.ts
function hasRequiredProperties(obj, paths) {
  return paths.every((path3) => {
    const keys = path3.split(".");
    let current = obj;
    for (const key of keys) {
      if (current !== void 0 && key in current) {
        current = current[key];
      } else {
        return false;
      }
    }
    return true;
  });
}
function findUp(expectations, options = {}) {
  const expectationSet = new Set(Array.isArray(expectations) ? expectations : [expectations]);
  if (expectationSet.size === 0) {
    return Array.isArray(expectations) ? [] : void 0;
  }
  const cwd = options.cwd ?? process.cwd();
  let currentDir = cwd;
  while (currentDir !== "/") {
    const dirFiles = fs__namespace.readdirSync(currentDir);
    const foundFiles = dirFiles.filter((file) => expectationSet.has(file));
    if (foundFiles.length > 0) {
      if (!Array.isArray(expectations)) {
        return path__namespace.join(currentDir, foundFiles[0]);
      }
      return foundFiles.map((file) => path__namespace.join(currentDir, file));
    }
    currentDir = path__namespace.dirname(currentDir);
  }
  return Array.isArray(expectations) ? [] : void 0;
}
function enableTS(relativeToPath) {
  const moduleName = "ts-node";
  const require2 = module2__default.default.createRequire(relativeToPath);
  const modulePath = require2.resolve(moduleName);
  const tsnode = require2(modulePath);
  tsnode.register({
    transpileOnly: true,
    typeCheck: false
  });
}
async function loadJSorTS(fileName, relativeToPath) {
  const require2 = module2__default.default.createRequire(relativeToPath);
  const modulePath = require2.resolve(fileName);
  if (fileName.endsWith(".ts")) {
    enableTS(relativeToPath);
    return require2(modulePath);
  } else if (fileName.endsWith(".mjs")) {
    return import(modulePath);
  } else if (fileName.endsWith(".cjs")) {
    return Promise.resolve(require2(modulePath));
  } else if (fileName.endsWith(".js")) {
    try {
      return await Promise.resolve(require2(modulePath));
    } catch (requireError) {
      try {
        return await import(modulePath);
      } catch (importError) {
        throw new Error(
          `Failed to load module: ${fileName}. Require error: ${requireError}. Import error: ${importError}`
        );
      }
    }
  } else {
    throw new Error(`Unsupported file extension: ${fileName}`);
  }
}
function Memoizee(options) {
  return function(target, propertyKey, descriptor) {
    const originalMethod = descriptor.value;
    if (typeof originalMethod !== "function") {
      throw new Error(`@Memoizee decorator can only be applied to methods.`);
    }
    const isAsyncFunction = originalMethod.constructor.name === "AsyncFunction";
    const memoizeOptions = {
      ...options,
      promise: options?.promise ?? isAsyncFunction
    };
    if (!isAsyncFunction) {
      const originalMethodStr = originalMethod.toString();
      if (originalMethodStr.includes("callback") || originalMethodStr.includes("cb")) {
        memoizeOptions.async = true;
      }
    }
    const memoizedFn = memoizee__default.default(originalMethod, memoizeOptions);
    const fn = memoizedFn;
    fn.clear = () => {
      memoizedFn.clear();
    };
    descriptor.value = fn;
  };
}

// src/combinatorial.ts
function cartesianProduct(...inputs) {
  if (inputs.length === 0) {
    return [];
  }
  const normalizedArrays = inputs.map((input) => Array.isArray(input) ? input : [input]);
  return normalizedArrays.reduce(
    (combinations, currentArray) => combinations.flatMap((combination) => currentArray.map((element) => [...combination, element])),
    [[]]
  );
}

// src/index.ts
var logger2 = getLogger();
async function sleep(timeout) {
  await new Promise((resolve) => setTimeout(resolve, timeout));
}
function getProjectPackageManager(cwd) {
  const yarn = findUp("yarn.lock", { cwd });
  if (yarn !== void 0) return "yarn";
  const npm = findUp("package-lock.json", { cwd });
  if (npm !== void 0) return "npm";
  const pnpm = findUp("pnpm-lock.yaml", { cwd });
  if (pnpm !== void 0) return "pnpm";
  throw new Error("Cannot find package.json or yarn.lock");
}
function getProjectRootDir(cwd) {
  const yarn = findUp("yarn.lock", { cwd });
  if (yarn !== void 0) return path__namespace.default.dirname(yarn);
  const npm = findUp("package-lock.json", { cwd });
  if (npm !== void 0) return path__namespace.default.dirname(npm);
  const pnpm = findUp("pnpm-lock.yaml", { cwd });
  if (pnpm !== void 0) return path__namespace.default.dirname(pnpm);
  throw new Error("Cannot find yarn.lock or package-lock.json or pnpm-lock.yaml");
}
async function isHttpServiceReachable(host, port, timeout, path3) {
  return new Promise((resolve, _reject) => {
    const options = {
      host,
      port,
      timeout,
      path: path3,
      method: "HEAD"
    };
    const request = http__default.default.request(options, (_response) => {
      resolve(true);
    });
    request.on("error", (_err) => {
      resolve(false);
    });
    request.end();
  });
}
function extractUrlInfo(url) {
  const m = url.match(/(?<schema>http|https):\/\/(?<host>.*):(?<port>\d+)/);
  if (m?.groups?.host === void 0) {
    throw new Error(`Invalid url ${url}`);
  }
  return {
    schema: m.groups.schema,
    host: m.groups.host,
    port: m.groups.port
  };
}

exports.Memoizee = Memoizee;
exports.SizeExceedsPaddingSizeError = SizeExceedsPaddingSizeError;
exports.arrayify = arrayify;
exports.asEnum = asEnum;
exports.asType = asType;
exports.assert = assert;
exports.assertDefined = assertDefined;
exports.assertType = assertType;
exports.assumeType = assumeType;
exports.bytesToHex = bytesToHex;
exports.cartesianProduct = cartesianProduct;
exports.createLogger = createLogger;
exports.deploymentToEvmContract = deploymentToEvmContract;
exports.dirname = dirname;
exports.enableTS = enableTS;
exports.ensure0x = ensure0x;
exports.extractUrlInfo = extractUrlInfo;
exports.findContract = findContract;
exports.findDeployment = findDeployment;
exports.findUp = findUp;
exports.first = first;
exports.firstFactory = firstFactory;
exports.getAptosAccountFromMnemonic = getAptosAccountFromMnemonic;
exports.getBIP044Path = getBIP044Path;
exports.getCircularReplacer = getCircularReplacer;
exports.getEvmAccountFromMnemonic = getEvmAccountFromMnemonic;
exports.getInitiaAccountFromMnemonic = getInitiaAccountFromMnemonic;
exports.getKeypairFromMnemonic = getKeypairFromMnemonic;
exports.getLogger = getLogger;
exports.getPlaceholderAccountFromMnemonic = getPlaceholderAccountFromMnemonic;
exports.getProjectPackageManager = getProjectPackageManager;
exports.getProjectRootDir = getProjectRootDir;
exports.getSolanaAccountFromMnemonic = getSolanaAccountFromMnemonic;
exports.getTonAccountFromMnemonic = getTonAccountFromMnemonic;
exports.getTonWalletFromMnemonic = getTonWalletFromMnemonic;
exports.hasRequiredProperties = hasRequiredProperties;
exports.hexToBytes = hexToBytes;
exports.hexlify = hexlify;
exports.initLogger = initLogger;
exports.isHash = isHash;
exports.isHex = isHex;
exports.isHexString = isHexString;
exports.isHttpServiceReachable = isHttpServiceReachable;
exports.loadJSorTS = loadJSorTS;
exports.logger = logger2;
exports.padify = padify;
exports.parallel = parallel;
exports.pkgroot = pkgroot;
exports.safeMap = safeMap;
exports.sequence = sequence;
exports.sleep = sleep;
exports.trim0x = trim0x;
exports.tryFindDeployment = tryFindDeployment;
