// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/events/nativeDropAppliedEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bool,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  NativeDropRequest,
  NativeDropRequestArgs,
  getNativeDropRequestSerializer,
} from '../types';

export type NativeDropAppliedEvent = {
  srcEid: number;
  sender: Uint8Array;
  nonce: bigint;
  dstEid: number;
  oapp: PublicKey;
  nativeDropRequests: Array<NativeDropRequest>;
  successes: Array<boolean>;
};

export type NativeDropAppliedEventArgs = {
  srcEid: number;
  sender: Uint8Array;
  nonce: number | bigint;
  dstEid: number;
  oapp: PublicKey;
  nativeDropRequests: Array<NativeDropRequestArgs>;
  successes: Array<boolean>;
};

export function getNativeDropAppliedEventSerializer(): Serializer<
  NativeDropAppliedEventArgs,
  NativeDropAppliedEvent
> {
  return struct<NativeDropAppliedEvent>(
    [
      ['srcEid', u32()],
      ['sender', bytes({ size: 32 })],
      ['nonce', u64()],
      ['dstEid', u32()],
      ['oapp', publicKeySerializer()],
      ['nativeDropRequests', array(getNativeDropRequestSerializer())],
      ['successes', array(bool())],
    ],
    { description: 'NativeDropAppliedEvent' }
  ) as Serializer<NativeDropAppliedEventArgs, NativeDropAppliedEvent>;
}
