// @orderly.network/ui-orders/dist/index.mjs
import React2, { forwardRef, useMemo, useEffect, useRef, useImperativeHandle, useCallback, useState, createContext, memo, useContext } from 'react';
import { useSymbolsInfo, useOrderStream, useLocalStorage, utils, useThrottledCallback, useMutation, useSWRConfig, useAccount, unstable_serialize, useMemoizedFn, getMinNotional, findTPSLFromOrder, findTPSLOrderPriceFromOrder, useMaxQty, useConfig, useTpslPriceChecker, useOrderEntity, useReferralInfo, useLeverageBySymbol } from '@orderly.network/hooks';
import { useTranslation, Trans, i18n } from '@orderly.network/i18n';
import { OrderStatus, AlgoOrderRootType, AlgoOrderType, OrderSide, OrderType, EMPTY_LIST } from '@orderly.network/types';
import { formatSymbol, Decimal, commify, commifyOptional, getTrailingStopPrice, formatNum } from '@orderly.network/utils';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { usePagination, modal, Text, Button, ThrottledButton, CloseIcon, cn, Flex, Input, inputFormatter, Slider, DataFilter, TableFeatures, Grid, Picker, ListView, Tabs, TabPanel, Badge, Tooltip, EditIcon, toast as toast$1, Popover, PopoverRoot, PopoverTrigger, PopoverContent, Divider, ShareIcon, CheckIcon, Statistic, SimpleDialog, useModal, Checkbox } from '@orderly.network/ui';
import { useDataTap, useOrderEntryFormErrorMsg } from '@orderly.network/react-app';
import { PositionTPSLPopover, CloseToLiqPriceIcon, PositionTPSLSheet } from '@orderly.network/ui-tpsl';
import { format, subDays, differenceInDays } from 'date-fns';
import { SharePnLDialogId } from '@orderly.network/ui-share';
import { AuthGuardDataTable } from '@orderly.network/ui-connector';

var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var useOrdersScript;
var init_orders_script = __esm({
  "src/components/orders.script.ts"() {
    useOrdersScript = (props) => {
      const { current, pnlNotionalDecimalPrecision, sharePnLConfig } = props;
      const orderListRef = useRef(null);
      useImperativeHandle(props.ref, () => ({
        download: () => {
          orderListRef.current?.download?.();
        }
      }));
      return {
        current,
        pnlNotionalDecimalPrecision,
        orderListRef,
        sharePnLConfig
      };
    };
  }
});
function parseBadgesFor(record) {
  const orderType = record.type;
  const algoType = record.algo_type;
  if (typeof orderType !== "undefined") {
    const list = [];
    if (!!record.parent_algo_type) {
      if (algoType === AlgoOrderType.STOP_LOSS) {
        const types = orderType === OrderType.CLOSE_POSITION ? [i18n.t("common.position"), i18n.t("tpsl.sl")] : [i18n.t("tpsl.sl")];
        list.push(...types);
      }
      if (algoType === AlgoOrderType.TAKE_PROFIT) {
        const types = orderType === OrderType.CLOSE_POSITION ? [i18n.t("common.position"), i18n.t("tpsl.tp")] : [i18n.t("tpsl.tp")];
        list.push(...types);
      }
      return list;
    }
    const type = typeof orderType === "string" ? orderType.replace("_ORDER", "") : "";
    if ([OrderType.ASK, OrderType.BID].includes(orderType)) {
      return [i18n.t("orderEntry.orderType.limit")];
    }
    if (record.algo_order_id === void 0 || record.algo_order_id && algoType === "BRACKET") {
      const typeMap = {
        [OrderType.LIMIT]: i18n.t("orderEntry.orderType.limit"),
        [OrderType.MARKET]: i18n.t("orderEntry.orderType.market"),
        [OrderType.POST_ONLY]: i18n.t("orderEntry.orderType.postOnly"),
        [OrderType.IOC]: i18n.t("orderEntry.orderType.ioc"),
        [OrderType.FOK]: i18n.t("orderEntry.orderType.fok")
      };
      return [
        typeMap[type] || upperCaseFirstLetter(type)
      ];
    }
    if (algoType === AlgoOrderRootType.TRAILING_STOP) {
      return [i18n.t("orderEntry.orderType.trailingStop")];
    }
    if (type) {
      const typeMap = {
        [OrderType.LIMIT]: i18n.t("orderEntry.orderType.stopLimit"),
        [OrderType.MARKET]: i18n.t("orderEntry.orderType.stopMarket")
      };
      return [typeMap[type] || type];
    }
  }
  if (typeof algoType !== "undefined") {
    const list = [];
    if (algoType === AlgoOrderRootType.POSITIONAL_TP_SL) {
      list.push(i18n.t("common.position"));
    }
    const tpOrder = record?.child_orders?.find(
      (order) => order.algo_type === AlgoOrderType.TAKE_PROFIT && !!order.trigger_price
    );
    const slOrder = record?.child_orders?.find(
      (order) => order.algo_type === AlgoOrderType.STOP_LOSS && !!order.trigger_price
    );
    if (tpOrder || slOrder) {
      list.push(
        tpOrder && slOrder ? i18n.t("common.tpsl") : tpOrder ? i18n.t("tpsl.tp") : i18n.t("tpsl.sl")
      );
    }
    return list;
  }
  return void 0;
}
function grayCell(record) {
  return record.status === OrderStatus.CANCELLED || record.algo_status === OrderStatus.CANCELLED;
}
function getOrderStatus(record) {
  return record.status || record.algo_status;
}
function isGrayCell(status2) {
  return status2 === OrderStatus.CANCELLED;
}
function findBracketTPSLOrder(order) {
  if (order.algo_type !== AlgoOrderRootType.BRACKET) {
    return {
      tpOrder: void 0,
      slOrder: void 0
    };
  }
  const innerOrder = order.child_orders?.[0];
  if (!innerOrder)
    return {
      tpOrder: void 0,
      slOrder: void 0
    };
  const tpOrder = innerOrder?.child_orders?.find(
    (item) => item.algo_type === AlgoOrderType.TAKE_PROFIT
  );
  const slOrder = innerOrder?.child_orders?.find(
    (item) => item.algo_type === AlgoOrderType.STOP_LOSS
  );
  return {
    tpOrder,
    slOrder
  };
}
function calcBracketRoiAndPnL(order) {
  const defaultCallback = {
    pnl: {
      tpPnL: void 0,
      slPnL: void 0
    }
    // roi: {
    //   tpRoi: undefined,
    //   slRoi: undefined,
    // },
  };
  const { tpOrder, slOrder } = findBracketTPSLOrder(order);
  if (!tpOrder && !slOrder) return defaultCallback;
  if (typeof order.price === void 0 || !order.price) return defaultCallback;
  const quantity2 = order.side === OrderSide.BUY ? order.quantity : order.quantity * -1;
  const tpPnL = tpOrder?.trigger_price && utils.priceToPnl({
    qty: quantity2,
    price: tpOrder?.trigger_price,
    entryPrice: order.price,
    // @ts-ignore
    orderSide: order.side,
    // @ts-ignore
    orderType: tpOrder.algo_type
  });
  const slPnL = slOrder?.trigger_price && utils.priceToPnl({
    qty: quantity2,
    // trigger price
    price: slOrder?.trigger_price,
    //
    entryPrice: order.price,
    // @ts-ignore
    orderSide: order.side,
    // @ts-ignore
    orderType: slOrder.algo_type
  });
  return {
    pnl: {
      tpPnL,
      slPnL
    }
    // roi: {
    //   tpRoi,
    //   slRoi,
    // },
  };
}
function areDatesEqual(date1, date2) {
  if (!date1 || !date2) return false;
  return date1.getFullYear() === date2.getFullYear() && date1.getMonth() === date2.getMonth() && date1.getDate() === date2.getDate();
}
function isTrailingStopOrder(order) {
  return order.algo_type === OrderType.TRAILING_STOP;
}
function getNotional(order, dp = 2) {
  if (order.price && order.quantity) {
    return new Decimal(order.price).mul(order.quantity).toFixed(dp, Decimal.ROUND_DOWN);
  }
  return 0;
}
function convertApiOrderTypeToOrderEntryType(order) {
  if (order.algo_type === OrderType.TRAILING_STOP) {
    return order.algo_type;
  }
  const isAlgoOrder = order.algo_order_id !== void 0;
  if (isAlgoOrder && order.algo_type !== AlgoOrderRootType.BRACKET) {
    return `STOP_${order.type}`;
  }
  return order.type;
}
var upperCaseFirstLetter;
var init_util = __esm({
  "src/utils/util.ts"() {
    upperCaseFirstLetter = (str) => {
      if (str === void 0) return str;
      if (str.length === 0) return str;
      if (str.length === 1) return str.charAt(0).toUpperCase();
      return str.charAt(0).toUpperCase() + str.toLowerCase().slice(1);
    };
  }
});
var SymbolContext, useSymbolContext;
var init_symbolContext = __esm({
  "src/components/provider/symbolContext.tsx"() {
    SymbolContext = createContext(
      {}
    );
    useSymbolContext = () => {
      return useContext(SymbolContext);
    };
  }
});
var SymbolProvider;
var init_symbolProvider = __esm({
  "src/components/provider/symbolProvider.tsx"() {
    init_symbolContext();
    SymbolProvider = (props) => {
      const { symbol, children } = props;
      const symbolInfo = useSymbolsInfo()[symbol];
      const memoizedValue = useMemo(() => {
        return {
          symbol,
          base_dp: symbolInfo("base_dp"),
          quote_dp: symbolInfo("quote_dp"),
          base_tick: symbolInfo("base_tick"),
          quote_tick: symbolInfo("quote_tick"),
          base: symbolInfo("base"),
          quote: symbolInfo("quote"),
          origin: symbolInfo(),
          quote_max: symbolInfo("quote_max"),
          quote_min: symbolInfo("quote_min")
        };
      }, [symbol, symbolInfo]);
      return /* @__PURE__ */ jsx(SymbolContext.Provider, { value: memoizedValue, children });
    };
  }
});
var ShareButton;
var init_shareButton_ui = __esm({
  "src/components/shareButton/shareButton.ui.tsx"() {
    ShareButton = (props) => {
      if (props.sharePnLConfig == null) {
        return null;
      }
      return /* @__PURE__ */ jsx(
        "button",
        {
          type: "button",
          onClick: (e) => {
            e.stopPropagation();
            props.showModal();
          },
          children: /* @__PURE__ */ jsx(ShareIcon, { color: "white", opacity: 0.54, size: props.iconSize ?? 16 })
        }
      );
    };
  }
});
var useShareButtonScript;
var init_shareButton_script = __esm({
  "src/components/shareButton/shareButton.script.tsx"() {
    useShareButtonScript = (props) => {
      const { sharePnLConfig, order, iconSize } = props;
      const { t } = useTranslation();
      const { getFirstRefCode } = useReferralInfo();
      const refCode = useMemo(() => {
        return getFirstRefCode()?.code;
      }, [getFirstRefCode]);
      const leverage = useLeverageBySymbol(order.symbol);
      const showModal = () => {
        modal.show(props.modalId, {
          pnl: {
            entity: {
              symbol: order.symbol,
              pnl: order.realized_pnl,
              side: order.side == "BUY" ? t("share.pnl.share.long") : t("share.pnl.share.short"),
              openPrice: order.average_executed_price,
              openTime: order.updated_time,
              quantity: order.quantity
            },
            refCode,
            leverage,
            ...sharePnLConfig
          }
        });
      };
      return {
        iconSize,
        sharePnLConfig,
        showModal
      };
    };
  }
});
var ShareButtonWidget;
var init_shareButton_widget = __esm({
  "src/components/shareButton/shareButton.widget.tsx"() {
    init_shareButton_script();
    init_shareButton_ui();
    ShareButtonWidget = (props) => {
      const state = useShareButtonScript(props);
      return /* @__PURE__ */ jsx(ShareButton, { ...state });
    };
  }
});

// src/components/shareButton/index.ts
var init_shareButton = __esm({
  "src/components/shareButton/index.ts"() {
    init_shareButton_ui();
    init_shareButton_widget();
  }
});
var AvgPrice;
var init_avgPrice = __esm({
  "src/components/orderList/desktop/avgPrice.tsx"() {
    AvgPrice = memo((props) => {
      const symbolsInfo = useSymbolsInfo();
      const info = symbolsInfo[props.symbol];
      if (!props.value) {
        return "--";
      }
      return /* @__PURE__ */ jsx(
        Text.numeral,
        {
          padding: false,
          dp: info("quote_dp", 2),
          rm: Decimal.ROUND_UP,
          children: props.value
        }
      );
    });
  }
});
var BracketOrderPrice, EditBracketOrder, Price;
var init_bracketOrderPrice = __esm({
  "src/components/orderList/desktop/bracketOrderPrice.tsx"() {
    init_util();
    init_symbolContext();
    BracketOrderPrice = (props) => {
      const { order } = props;
      const { quote_dp, base_dp } = useSymbolContext();
      const { t } = useTranslation();
      const { sl_trigger_price, tp_trigger_price } = useMemo(() => {
        if (!("algo_type" in order) || !Array.isArray(order.child_orders)) {
          return {};
        }
        return utils.findTPSLFromOrder(props.order.child_orders[0]);
      }, [props.order]);
      const { pnl } = calcBracketRoiAndPnL(order);
      if (!tp_trigger_price && !sl_trigger_price) {
        return "--";
      }
      return /* @__PURE__ */ jsx(
        Tooltip,
        {
          content: /* @__PURE__ */ jsxs(Flex, { direction: "column", itemAlign: "start", gap: 1, children: [
            typeof pnl.tpPnL !== "undefined" && /* @__PURE__ */ jsx(
              Text.numeral,
              {
                prefix: /* @__PURE__ */ jsxs(Text, { intensity: 80, children: [
                  `${t("tpsl.tpPnl")}:`,
                  " \xA0"
                ] }),
                suffix: /* @__PURE__ */ jsx(Text, { intensity: 20, children: " USDC" }),
                dp: quote_dp,
                color: "buy",
                showIdentifier: true,
                children: pnl.tpPnL
              }
            ),
            typeof pnl.slPnL !== "undefined" && /* @__PURE__ */ jsx(
              Text.numeral,
              {
                prefix: /* @__PURE__ */ jsxs(Text, { intensity: 80, children: [
                  `${t("tpsl.slPnl")}:`,
                  " \xA0"
                ] }),
                suffix: /* @__PURE__ */ jsx(Text, { intensity: 20, children: " USDC" }),
                dp: quote_dp,
                color: "sell",
                children: pnl.slPnL
              }
            )
          ] }),
          className: "oui-bg-base-6",
          children: /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", justify: "start", gap: 2, children: [
            /* @__PURE__ */ jsxs(Flex, { direction: "column", justify: "start", itemAlign: "start", children: [
              /* @__PURE__ */ jsx(Price, { type: "TP", value: tp_trigger_price, quote_dp }),
              /* @__PURE__ */ jsx(Price, { type: "SL", value: sl_trigger_price, quote_dp })
            ] }),
            /* @__PURE__ */ jsx(EditBracketOrder, { order })
          ] })
        }
      );
    };
    EditBracketOrder = (props) => {
      const { order } = props;
      const onEdit = () => {
        modal.show("EditBracketOrderDialogId", {
          order
        });
      };
      return /* @__PURE__ */ jsx(
        EditIcon,
        {
          size: 16,
          className: "oui-text-base-contrast oui-cursor-pointer",
          onClick: () => {
            onEdit();
          }
        }
      );
    };
    Price = (props) => {
      const { type, value, quote_dp } = props;
      const { t } = useTranslation();
      return value ? /* @__PURE__ */ jsx(
        Text.numeral,
        {
          className: cn(
            "oui-gap-0 oui-decoration-white/20 oui-border-b oui-border-dashed oui-border-base-contrast-12",
            type === "TP" ? "oui-text-trade-profit" : "oui-text-trade-loss"
          ),
          rule: "price",
          dp: quote_dp,
          prefix: /* @__PURE__ */ jsxs("span", { className: "oui-text-base-contrast-54", children: [
            type === "TP" ? `${t("tpsl.tp")} -` : `${t("tpsl.sl")} -`,
            "\xA0"
          ] }),
          children: value
        },
        "tp"
      ) : /* @__PURE__ */ jsx(Fragment, {});
    };
  }
});
var OrderListContext, useOrderListContext;
var init_orderListContext = __esm({
  "src/components/orderList/orderListContext.tsx"() {
    OrderListContext = createContext(
      {}
    );
    useOrderListContext = () => {
      return useContext(OrderListContext);
    };
  }
});
var CancelButton;
var init_cancelBtn = __esm({
  "src/components/orderList/desktop/cancelBtn.tsx"() {
    init_orderListContext();
    CancelButton = (props) => {
      const { order } = props;
      const { t } = useTranslation();
      const { onCancelOrder } = useOrderListContext();
      const [isLoading, setIsLoading] = useState(false);
      return /* @__PURE__ */ jsx(
        ThrottledButton,
        {
          size: "sm",
          variant: "outlined",
          color: "secondary",
          onClick: (event) => {
            if (!onCancelOrder) return;
            event.preventDefault();
            event.stopPropagation();
            setIsLoading(true);
            onCancelOrder(order).then(
              (res) => res,
              (error) => {
                toast$1.error(error.message);
              }
            ).finally(() => {
              setIsLoading(false);
            });
          },
          loading: isLoading,
          children: t("common.cancel")
        }
      );
    };
  }
});

// src/type.ts
var init_type = __esm({
  "src/type.ts"() {
  }
});
var ConfirmContent;
var init_confirmContent = __esm({
  "src/components/orderList/desktop/editOrder/confirmContent.tsx"() {
    init_type();
    ConfirmContent = memo((props) => {
      const { type, base, value, cancelPopover, isSubmitting, onConfirm } = props;
      const { t } = useTranslation();
      const renderLabel = () => {
        const common = {
          values: { base, value: commify(value) },
          components: [/* @__PURE__ */ jsx("span", { className: "oui-text-warning-darken" }, "0")]
        };
        switch (type) {
          case 0 /* quantity */:
            return (
              // @ts-ignore
              /* @__PURE__ */ jsx(Trans, { i18nKey: "order.edit.confirm.quantity", ...common })
            );
          case 1 /* price */:
            return (
              // @ts-ignore
              /* @__PURE__ */ jsx(Trans, { i18nKey: "order.edit.confirm.price", ...common })
            );
          case 2 /* triggerPrice */:
            return (
              // @ts-ignore
              /* @__PURE__ */ jsx(Trans, { i18nKey: "order.edit.confirm.triggerPrice", ...common })
            );
          case 3 /* callbackValue */:
            return (
              // @ts-ignore
              /* @__PURE__ */ jsx(Trans, { i18nKey: "order.edit.confirm.callbackValue", ...common })
            );
          case 4 /* callbackRate */:
            return (
              // @ts-ignore
              /* @__PURE__ */ jsx(Trans, { i18nKey: "order.edit.confirm.callbackRate", ...common })
            );
        }
      };
      return /* @__PURE__ */ jsxs("div", { className: "oui-relative oui-pt-5", children: [
        /* @__PURE__ */ jsx("div", { className: "desktop:oui-text-sm oui-text-2xs oui-text-base-contrast-54", children: renderLabel() }),
        /* @__PURE__ */ jsxs("div", { className: "oui-mt-5 oui-grid oui-grid-cols-2 oui-gap-2", children: [
          /* @__PURE__ */ jsx(
            Button,
            {
              color: "secondary",
              size: "md",
              onClick: cancelPopover,
              disabled: isSubmitting,
              children: t("common.cancel")
            }
          ),
          /* @__PURE__ */ jsx(ThrottledButton, { size: "md", loading: isSubmitting, onClick: onConfirm, children: t("common.confirm") })
        ] }),
        /* @__PURE__ */ jsx(
          "button",
          {
            className: "oui-absolute oui-right-0 oui-top-0 oui-text-base-contrast-54",
            onClick: cancelPopover,
            children: /* @__PURE__ */ jsx(CloseIcon, { size: 16, color: "white", opacity: 1 })
          }
        )
      ] });
    });
    ConfirmContent.displayName = "ConfirmContent";
  }
});
function usePopoverState(props) {
  const { originValue, value, setValue } = props;
  const [open, setOpen] = useState(false);
  const [editing, setEditing] = useState(false);
  const containerRef = useRef(null);
  const closePopover = useCallback(() => {
    setOpen(false);
    setEditing(false);
  }, []);
  const cancelPopover = useCallback(() => {
    closePopover();
    setValue(originValue);
  }, [originValue]);
  const onClick = useCallback(
    async (event) => {
      event.stopPropagation();
      event.preventDefault();
      if (value === originValue) {
        setEditing(false);
        return;
      }
      setOpen(true);
    },
    [value, originValue]
  );
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (containerRef.current && !containerRef.current.contains(event.target) && !open) {
        cancelPopover();
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [open, cancelPopover]);
  return {
    value,
    setValue,
    open,
    setOpen,
    editing,
    setEditing,
    containerRef,
    closePopover,
    cancelPopover,
    onClick
  };
}
var init_usePopoverState = __esm({
  "src/components/orderList/desktop/hooks/usePopoverState.ts"() {
  }
});
function useValidateField(props) {
  const { order, field, originValue, value, fieldValues } = props;
  const changeFields = fieldValues || { [field]: value };
  const orderType = useMemo(
    () => convertApiOrderTypeToOrderEntryType(order),
    [order]
  );
  const { errors, validate, clearErrors } = useOrderEntity({
    symbol: order.symbol,
    order_type: orderType,
    side: order.side,
    ...changeFields
  });
  const { getErrorMsg } = useOrderEntryFormErrorMsg(errors);
  const error = useMemo(
    () => field ? getErrorMsg(field) : void 0,
    [getErrorMsg, field]
  );
  useEffect(() => {
    if (value === originValue) {
      clearErrors();
      return;
    }
    validate().then((order2) => {
    }).catch((errors2) => {
    });
  }, [value, originValue]);
  return { error, errors, getErrorMsg };
}
var init_useValidateField = __esm({
  "src/components/orderList/desktop/hooks/useValidateField.ts"() {
    init_util();
  }
});
var EditableCellInput;
var init_editableCellInput = __esm({
  "src/components/orderList/desktop/components/editableCellInput.tsx"() {
    EditableCellInput = (props) => {
      const { onClick } = props;
      const inputRef = useRef(null);
      const handleKeyDown = useCallback(
        (event) => {
          if (event.key === "Enter") {
            onClick(event);
          }
        },
        [onClick]
      );
      useEffect(() => {
        const input = inputRef.current;
        if (input) {
          const length = input.value.length;
          input.setSelectionRange(length, length);
        }
      }, [props.value]);
      return /* @__PURE__ */ jsx(
        Input.tooltip,
        {
          ref: inputRef,
          tooltip: props.error,
          autoComplete: "off",
          autoFocus: true,
          type: "text",
          size: "sm",
          placeholder: props.readonly ? "" : props.placeholder || "",
          id: props.id,
          name: props.name,
          color: props.error ? "danger" : void 0,
          value: props.readonly ? "" : props.value || "",
          onValueChange: props.onChange,
          onClick: (e) => {
            e.stopPropagation();
            e.preventDefault();
          },
          onFocus: props.onFocus,
          onBlur: props.onBlur,
          onKeyDown: props.handleKeyDown || handleKeyDown,
          formatters: props.overrideFormatters || [
            ...props.formatters ?? EMPTY_LIST,
            inputFormatter.numberFormatter,
            inputFormatter.decimalPointFormatter
          ],
          classNames: {
            root: "oui-bg-base-700 oui-px-2 oui-py-1 oui-rounded",
            input: "oui-pr-2"
          },
          readOnly: props.readonly,
          suffix: /* @__PURE__ */ jsxs(Flex, { gapX: 1, children: [
            props.suffix,
            /* @__PURE__ */ jsx("button", { onClick, disabled: !!props.error, children: /* @__PURE__ */ jsx(
              CheckIcon,
              {
                size: 18,
                color: "white",
                opacity: 1,
                className: cn(
                  "oui-opacity-50",
                  props.error ? "oui-cursor-not-allowed" : "oui-cursor-pointer hover:oui-opacity-100"
                )
              }
            ) })
          ] })
        }
      );
    };
  }
});
var PreviewCell;
var init_previewCell = __esm({
  "src/components/orderList/desktop/components/previewCell.tsx"() {
    init_util();
    PreviewCell = memo((props) => {
      const { status: status2, value, disabled } = props;
      return /* @__PURE__ */ jsx(
        "div",
        {
          className: cn(
            "oui-flex oui-items-center oui-justify-start",
            "oui-relative oui-max-w-[110px] oui-gap-1 oui-font-semibold",
            isGrayCell(status2) && "oui-text-base-contrast-20",
            props.className
          ),
          onClick: (e) => {
            if (!disabled) {
              e.stopPropagation();
              e.preventDefault();
              props.setEditing(true);
            }
          },
          children: /* @__PURE__ */ jsx(
            Flex,
            {
              r: "base",
              className: cn(
                "oui-h-[28px] oui-min-w-[70px]",
                !disabled && "oui-border oui-border-line-12 oui-bg-base-7 oui-px-2"
              ),
              children: /* @__PURE__ */ jsxs(Text, { size: "2xs", children: [
                commifyOptional(value),
                props.suffix
              ] })
            }
          )
        }
      );
    });
  }
});
var ActivedPriceCell;
var init_activedPriceCell = __esm({
  "src/components/orderList/desktop/components/activedPriceCell.tsx"() {
    init_type();
    init_util();
    init_symbolContext();
    init_orderListContext();
    init_confirmContent();
    init_usePopoverState();
    init_useValidateField();
    init_editableCellInput();
    init_previewCell();
    ActivedPriceCell = (props) => {
      const { order } = props;
      const { t } = useTranslation();
      const originValue = order.activated_price?.toString() ?? "";
      const [value, setValue] = useState(originValue);
      const [submitting, setSubmitting] = useState(false);
      const disabled = props.disabled || order.is_activated || order.is_triggered;
      const { editAlgoOrder } = useOrderListContext();
      const { base, quote_dp } = useSymbolContext();
      const {
        open,
        setOpen,
        editing,
        setEditing,
        containerRef,
        closePopover,
        cancelPopover,
        onClick
      } = usePopoverState({
        originValue,
        value,
        setValue
      });
      const { error } = useValidateField({
        order,
        field: "activated_price",
        originValue,
        value
      });
      const onConfirm = useCallback(() => {
        setSubmitting(true);
        const data = {
          algo_order_id: order.algo_order_id,
          activated_price: value
        };
        editAlgoOrder(order.algo_order_id.toString(), data).then((result) => {
          closePopover();
          setValue(value);
        }).catch((err) => {
          toast$1.error(err.message);
          cancelPopover();
        }).finally(() => setSubmitting(false));
      }, [order.algo_order_id, value, cancelPopover]);
      if (!order.activated_price && order.type === OrderType.MARKET) {
        return /* @__PURE__ */ jsx("span", { children: t("common.marketPrice") });
      }
      const renderContent = () => {
        if (!editing || disabled) {
          return /* @__PURE__ */ jsx(
            PreviewCell,
            {
              value,
              status: getOrderStatus(order),
              setEditing,
              disabled
            }
          );
        }
        return /* @__PURE__ */ jsx(
          EditableCellInput,
          {
            value,
            onChange: setValue,
            onClick,
            error: value ? error : t("orderEntry.triggerPrice.error.required"),
            formatters: [inputFormatter.dpFormatter(quote_dp)]
          }
        );
      };
      return /* @__PURE__ */ jsx(
        Popover,
        {
          open,
          onOpenChange: setOpen,
          content: /* @__PURE__ */ jsx(
            ConfirmContent,
            {
              type: 2 /* triggerPrice */,
              base,
              value,
              cancelPopover,
              isSubmitting: submitting,
              onConfirm
            }
          ),
          children: /* @__PURE__ */ jsx(
            "div",
            {
              onClick: (e) => {
                e.stopPropagation();
                e.preventDefault();
              },
              ref: containerRef,
              children: renderContent()
            }
          )
        }
      );
    };
  }
});
var PriceCell;
var init_priceCell = __esm({
  "src/components/orderList/desktop/components/priceCell.tsx"() {
    init_type();
    init_util();
    init_symbolContext();
    init_orderListContext();
    init_confirmContent();
    init_usePopoverState();
    init_useValidateField();
    init_editableCellInput();
    init_previewCell();
    PriceCell = (props) => {
      const { order } = props;
      const { t } = useTranslation();
      const originValue = useMemo(() => {
        if (order.type === OrderType.MARKET && !order.price) {
          return "Market";
        }
        return order.price?.toString() ?? "Market";
      }, [order.price, order.type]);
      const [value, setValue] = useState(originValue);
      const isAlgoOrder = order?.algo_order_id !== void 0;
      const [submitting, setSubmitting] = useState(false);
      const { editOrder, editAlgoOrder } = useOrderListContext();
      const { base, quote_dp } = useSymbolContext();
      const {
        open,
        setOpen,
        editing,
        setEditing,
        containerRef,
        closePopover,
        cancelPopover,
        onClick
      } = usePopoverState({
        originValue,
        value,
        setValue
      });
      const { error, getErrorMsg } = useValidateField({
        order,
        field: "order_price",
        originValue,
        value,
        fieldValues: {
          order_price: value,
          // add order quantity to validate total
          order_quantity: order.quantity?.toString(),
          // need to pass trigger_price to validate order_price, because order_price is depend on trigger_price when type is stop limit
          trigger_price: order.trigger_price?.toString()
        }
      });
      const onConfirm = () => {
        setSubmitting(true);
        let order_id = order.order_id;
        let data = {
          symbol: order.symbol,
          order_type: order.type,
          side: order.side,
          order_quantity: order.quantity,
          order_price: value
        };
        if (order.reduce_only !== void 0) {
          data.reduce_only = order.reduce_only;
        }
        if (order.order_tag) {
          data.order_tag = order.order_tag;
        }
        if (order.client_order_id) {
          data.client_order_id = order.client_order_id;
        }
        if (isAlgoOrder) {
          order_id = order.algo_order_id;
          data = {
            ...data,
            order_id,
            price: value,
            algo_order_id: order_id
          };
        }
        if (order.tag !== void 0) {
          data["order_tag"] = order.tag;
        }
        let future;
        if (order.algo_order_id !== void 0) {
          future = editAlgoOrder(order.algo_order_id.toString(), data);
        } else {
          future = editOrder(order.order_id.toString(), data);
        }
        future.then((result) => {
          closePopover();
          setValue(value);
        }).catch((err) => {
          toast$1.error(err.message);
          cancelPopover();
        }).finally(() => setSubmitting(false));
      };
      const isAlgoMarketOrder = order.algo_order_id && order.type == "MARKET";
      if (isAlgoMarketOrder || value === "Market") {
        return /* @__PURE__ */ jsx("span", { children: t("common.marketPrice") });
      }
      const renderContent = () => {
        if (!editing || props.disabled) {
          return /* @__PURE__ */ jsx(
            PreviewCell,
            {
              value,
              status: getOrderStatus(order),
              setEditing,
              disabled: props.disabled
            }
          );
        }
        return /* @__PURE__ */ jsx(
          EditableCellInput,
          {
            value,
            onChange: setValue,
            onClick,
            error: error || getErrorMsg("total"),
            formatters: [inputFormatter.dpFormatter(quote_dp)]
          }
        );
      };
      return /* @__PURE__ */ jsx(
        Popover,
        {
          open,
          onOpenChange: setOpen,
          content: /* @__PURE__ */ jsx(
            ConfirmContent,
            {
              type: 1 /* price */,
              base,
              value,
              cancelPopover,
              isSubmitting: submitting,
              onConfirm
            }
          ),
          children: /* @__PURE__ */ jsx(
            "div",
            {
              onClick: (e) => {
                e.stopPropagation();
                e.preventDefault();
              },
              ref: containerRef,
              children: renderContent()
            }
          )
        }
      );
    };
  }
});
function calcTPSLPnL(props) {
  const { order, position, quote_dp } = props;
  if (!position)
    return {
      sl_trigger_price: void 0,
      tp_trigger_price: void 0,
      slPnL: void 0,
      tpPnL: void 0
    };
  const isTPSLOrder = "algo_type" in order && Array.isArray(order.child_orders);
  const { sl_trigger_price, tp_trigger_price } = isTPSLOrder ? findTPSLFromOrder(order) : {
    sl_trigger_price: void 0,
    tp_trigger_price: void 0
  };
  const { sl_order_price, tp_order_price } = isTPSLOrder ? findTPSLOrderPriceFromOrder(order) : {
    sl_order_price: void 0,
    tp_order_price: void 0
  };
  let quantity2 = order.quantity;
  if (quantity2 === 0) {
    if (order.child_orders?.[0].type === "CLOSE_POSITION") {
      quantity2 = position.position_qty;
    }
  }
  const avgOpenPrice = position.average_open_price;
  const tpPnL = typeof quantity2 === "number" && typeof tp_trigger_price === "number" && typeof avgOpenPrice === "number" ? utils.priceToPnl(
    {
      qty: quantity2,
      price: tp_trigger_price,
      entryPrice: position.average_open_price,
      orderSide: order.side,
      orderType: AlgoOrderType.TAKE_PROFIT
    },
    { symbol: { quote_dp } }
  ) : void 0;
  const slPnL = typeof quantity2 === "number" && typeof sl_trigger_price === "number" && typeof avgOpenPrice === "number" ? utils.priceToPnl(
    {
      qty: quantity2,
      price: sl_trigger_price,
      entryPrice: position.average_open_price,
      orderSide: order.side,
      orderType: AlgoOrderType.STOP_LOSS
    },
    { symbol: { quote_dp } }
  ) : void 0;
  return {
    sl_trigger_price,
    tp_trigger_price,
    sl_order_price,
    tp_order_price,
    slPnL,
    tpPnL
  };
}
var TPSLOrderRowContext, useTPSLOrderRowContext, TPSLOrderRowProvider;
var init_tpslOrderRowContext = __esm({
  "src/components/orderList/tpslOrderRowContext.tsx"() {
    init_symbolContext();
    TPSLOrderRowContext = createContext(
      {}
    );
    useTPSLOrderRowContext = () => {
      return useContext(TPSLOrderRowContext);
    };
    TPSLOrderRowProvider = (props) => {
      const { order, children } = props;
      const { quote_dp } = useSymbolContext();
      const [position, setPosition] = useState();
      const [doDeleteOrder] = useMutation("/v1/algo/order", "DELETE");
      const [doUpdateOrder] = useMutation("/v1/algo/order", "PUT");
      const config = useSWRConfig();
      const { state } = useAccount();
      const positionKey = useMemo(() => {
        return unstable_serialize(() => ["/v1/positions", state.accountId]);
      }, [state.accountId]);
      const onCancelOrder = useMemoizedFn(async (order2) => {
        return doDeleteOrder(null, {
          order_id: order2.algo_order_id,
          symbol: order2.symbol
        });
      });
      const onUpdateOrder = useMemoizedFn(
        async (order2, params) => {
          return doUpdateOrder({
            order_id: order2.algo_order_id,
            child_orders: order2.child_orders.map((order3) => ({
              order_id: order3.algo_order_id,
              quantity: params.order_quantity
            }))
          });
        }
      );
      const getRelatedPosition = useMemoizedFn(
        (symbol) => {
          const positions = config.cache.get(positionKey);
          return positions?.data?.rows?.find(
            (p) => p.symbol === symbol
          );
        }
      );
      const {
        sl_trigger_price,
        tp_trigger_price,
        tpPnL,
        slPnL,
        sl_order_price,
        tp_order_price
      } = calcTPSLPnL({
        order,
        position,
        quote_dp
      });
      useEffect(() => {
        if ("algo_type" in order || (order?.reduce_only ?? false)) {
          const position2 = getRelatedPosition(order.symbol);
          if (position2) {
            setPosition(position2);
          }
        }
      }, [order.symbol]);
      const memoizedValue = useMemo(() => {
        return {
          order,
          sl_trigger_price,
          tp_trigger_price,
          sl_order_price,
          tp_order_price,
          tpPnL,
          slPnL,
          position,
          onCancelOrder,
          onUpdateOrder,
          getRelatedPosition
        };
      }, [
        order,
        sl_trigger_price,
        tp_trigger_price,
        sl_order_price,
        tp_order_price,
        tpPnL,
        slPnL,
        position,
        onCancelOrder,
        onUpdateOrder,
        getRelatedPosition
      ]);
      return /* @__PURE__ */ jsx(TPSLOrderRowContext.Provider, { value: memoizedValue, children });
    };
  }
});
var InnerInput;
var init_innerInput = __esm({
  "src/components/orderList/desktop/editOrder/innerInput.tsx"() {
    InnerInput = (props) => {
      const {
        inputRef,
        dp,
        value,
        setValue,
        setEditing,
        error,
        handleKeyDown,
        onClick,
        onClose,
        onFocus,
        onBlur,
        hintInfo
      } = props;
      useEffect(() => {
        const input = inputRef.current;
        if (input) {
          const length = input.value.length;
          input.setSelectionRange(length, length);
        }
        setEditing(true);
      }, []);
      const open = (hintInfo?.length || 0) > 0;
      return /* @__PURE__ */ jsx(Tooltip, { content: hintInfo, open, children: /* @__PURE__ */ jsx(
        Input,
        {
          ref: inputRef,
          type: "text",
          size: "sm",
          formatters: [
            inputFormatter.numberFormatter,
            inputFormatter.dpFormatter(dp),
            inputFormatter.currencyFormatter
          ],
          value,
          onValueChange: (e) => setValue(e),
          helpText: error,
          onClick: (e) => {
            e.stopPropagation();
            e.preventDefault();
          },
          autoComplete: "off",
          onFocus,
          onBlur,
          onKeyDown: handleKeyDown,
          autoFocus: true,
          color: open ? "danger" : void 0,
          classNames: {
            root: "oui-bg-base-700 oui-px-2 oui-py-1 oui-rounded",
            input: "oui-pr-2"
          },
          suffix: /* @__PURE__ */ jsx("button", { onClick, children: /* @__PURE__ */ jsx(
            CheckIcon,
            {
              size: 18,
              color: "white",
              opacity: 1,
              className: "oui-cursor-pointer oui-opacity-50 hover:oui-opacity-100"
            }
          ) })
        }
      ) });
    };
  }
});
var QuantityCell, PreviewCell2, EditState, Buttons;
var init_quantityCell = __esm({
  "src/components/orderList/desktop/components/quantityCell.tsx"() {
    init_type();
    init_util();
    init_symbolContext();
    init_orderListContext();
    init_tpslOrderRowContext();
    init_confirmContent();
    init_innerInput();
    QuantityCell = (props) => {
      const { order } = props;
      const { reduce_only } = order;
      const originValue = order.quantity.toString();
      const [value, setValue] = useState(originValue);
      const [editing, setEditing] = useState(false);
      const { t } = useTranslation();
      const [open, setOpen] = useState(false);
      const [error, setError] = useState();
      const { editOrder, editAlgoOrder, checkMinNotional } = useOrderListContext();
      const { onUpdateOrder: onUpdateTPSLOrder, position } = useTPSLOrderRowContext();
      const { base_dp, base, base_tick } = useSymbolContext();
      const setQuantity = async (qty, maxQty) => {
        setValue(qty);
        const positionQty = Math.abs(position?.position_qty || 0);
        if (position && reduce_only && Number(qty) > positionQty) {
          setError(
            t("orders.quantity.lessThanPosition", {
              quantity: positionQty
            })
          );
        } else {
          const quantity2 = Number(qty);
          if (maxQty && quantity2 > maxQty) {
            setError(
              t("orders.quantity.lessThan", {
                quantity: commifyOptional(maxQty, {
                  fix: base_dp
                })
              })
            );
          } else {
            setError(void 0);
          }
        }
        return Promise.resolve();
      };
      useEffect(() => {
        setQuantity(order.quantity.toString());
      }, [props.order.quantity]);
      const closePopover = () => {
        setOpen(false);
        setEditing(false);
      };
      const cancelPopover = () => {
        setOpen(false);
        setQuantity(order.quantity.toString());
        setEditing(false);
      };
      const [isSubmitting, setIsSubmitting] = useState(false);
      const inputRef = useRef(null);
      const clickHandler = () => {
        if (!!error) {
          return;
        }
        if (Number(value) === Number(order.quantity)) {
          setEditing(false);
          return;
        }
        const price2 = order.algo_order_id !== void 0 ? order.trigger_price : order.price;
        if (price2 !== null && order.reduce_only !== true) {
          const notionalText = checkMinNotional(order.symbol, price2, value);
          if (notionalText) {
            toast$1.error(notionalText);
            setIsSubmitting(false);
            cancelPopover();
            return;
          }
        }
        setOpen(true);
      };
      const onClick = (event) => {
        event?.stopPropagation();
        event?.preventDefault();
        clickHandler();
      };
      const handleKeyDown = (event) => {
        if (event.key === "Enter") {
          event?.stopPropagation();
          event?.preventDefault();
          clickHandler();
        }
      };
      const onConfirm = useCallback(() => {
        setIsSubmitting(true);
        let params = {
          symbol: order.symbol,
          order_type: order.type,
          side: order.side,
          order_price: order.price,
          order_quantity: value,
          // reduce_only: Boolean(order.reduce_only),
          algo_order_id: order.algo_order_id
        };
        if (typeof params.algo_order_id !== "undefined" && params.order_type === "MARKET") {
          const { order_price, ...rest } = params;
          params = rest;
        }
        if (typeof order.reduce_only !== "undefined") {
          params.reduce_only = order.reduce_only;
        }
        if (order.order_tag) {
          params.order_tag = order.order_tag;
        }
        if (order.client_order_id) {
          params.client_order_id = order.client_order_id;
        }
        if (order?.visible_quantity === 0) {
          params["visible_quantity"] = 0;
        }
        if (order?.tag !== void 0) {
          params["order_tag"] = order.tag;
        }
        let future;
        if ("algo_type" in order && order.algo_type === AlgoOrderRootType.TP_SL) {
          future = onUpdateTPSLOrder(order, params);
        } else {
          if (order.algo_order_id !== void 0) {
            future = editAlgoOrder(order.algo_order_id.toString(), params);
          } else {
            future = editOrder(order.order_id.toString(), params);
          }
        }
        future.then(
          (result) => {
            closePopover();
            setQuantity(value.toString());
          },
          (err) => {
            toast$1.error(err.message);
            setQuantity(order.quantity.toString());
            cancelPopover();
          }
        ).finally(() => setIsSubmitting(false));
      }, [value]);
      const componentRef = useRef(null);
      const quantitySliderRef = useRef(null);
      const handleClickOutside = (event) => {
        if (componentRef.current && quantitySliderRef.current && !componentRef.current.contains(event.target) && !quantitySliderRef.current.contains(event.target) && !open) {
          cancelPopover();
        }
      };
      useEffect(() => {
        document.addEventListener("mousedown", handleClickOutside);
        return () => {
          document.removeEventListener("mousedown", handleClickOutside);
        };
      }, [open, order.quantity]);
      const renderContent = () => {
        if (!editing || props.disabled) {
          return /* @__PURE__ */ jsx(
            PreviewCell2,
            {
              order,
              value,
              setEditing,
              disable: props.disabled
            }
          );
        }
        return /* @__PURE__ */ jsx(
          EditState,
          {
            inputRef,
            quantitySliderRef,
            base_dp,
            base_tick,
            quantity: value,
            setQuantity,
            editing,
            setEditing,
            handleKeyDown,
            onClick,
            onClose: cancelPopover,
            symbol: order.symbol,
            reduce_only,
            positionQty: position?.position_qty,
            error,
            confirmOpen: open,
            side: order.side,
            order,
            setError
          }
        );
      };
      return /* @__PURE__ */ jsx(
        Popover,
        {
          open,
          onOpenChange: setOpen,
          content: /* @__PURE__ */ jsx(
            ConfirmContent,
            {
              type: 0 /* quantity */,
              base,
              value,
              cancelPopover,
              isSubmitting,
              onConfirm
            }
          ),
          contentProps: {
            onOpenAutoFocus: (e) => {
            }
          },
          children: /* @__PURE__ */ jsx(
            "div",
            {
              onClick: (e) => {
                e.stopPropagation();
                e.preventDefault();
              },
              ref: componentRef,
              children: renderContent()
            }
          )
        }
      );
    };
    PreviewCell2 = (props) => {
      const { order, value } = props;
      const executed = order.total_executed_quantity;
      return /* @__PURE__ */ jsxs(
        Flex,
        {
          direction: "row",
          justify: "start",
          gap: 1,
          className: cn(
            "oui-relative oui-max-w-[110px]",
            order.side === OrderSide.BUY && "oui-text-trade-profit",
            order.side === OrderSide.SELL && "oui-text-trade-loss",
            grayCell(order) && "oui-text-base-contrast-20"
          ),
          onClick: (e) => {
            e.stopPropagation();
            e.preventDefault();
            props.setEditing(true);
          },
          children: [
            "algo_type" in order && order.algo_type === AlgoOrderRootType.TP_SL ? null : /* @__PURE__ */ jsxs(Fragment, { children: [
              /* @__PURE__ */ jsx("span", { children: executed }),
              /* @__PURE__ */ jsx("span", { children: "/" })
            ] }),
            /* @__PURE__ */ jsx(
              Flex,
              {
                r: "base",
                className: cn(
                  "oui-h-[28px] oui-min-w-[70px]",
                  !props.disable && "oui-border oui-border-line-12 oui-bg-base-7 oui-px-2"
                ),
                children: /* @__PURE__ */ jsx(Text, { size: "2xs", children: value })
              }
            )
          ]
        }
      );
    };
    EditState = (props) => {
      const {
        inputRef,
        quantitySliderRef,
        base_dp,
        base_tick,
        quantity: quantity2,
        setQuantity,
        setEditing,
        handleKeyDown,
        onClick,
        onClose,
        error,
        symbol,
        reduce_only,
        positionQty,
        confirmOpen,
        order
      } = props;
      const maxBuyQty = useMaxQty(symbol, order.side, order.reduce_only);
      const maxQty = useMemo(() => {
        if (reduce_only) {
          return Math.abs(positionQty ?? 0);
        }
        return order.quantity + Math.abs(maxBuyQty);
      }, [order.quantity, maxBuyQty, reduce_only, positionQty]);
      const [sliderValue, setSliderValue] = useState(void 0);
      useEffect(() => {
        if (sliderValue === void 0) {
          const sliderValue2 = new Decimal(quantity2).div(maxQty).abs().mul(100).toNumber();
          setSliderValue(sliderValue2);
        }
      }, [sliderValue, maxQty, quantity2]);
      const formatQuantity = async (_qty) => {
        if (base_tick > 0) {
          _qty = utils.formatNumber(_qty, base_tick) ?? _qty;
        }
        return setQuantity(`${_qty}`, maxQty);
      };
      return /* @__PURE__ */ jsxs(PopoverRoot, { open: !confirmOpen, children: [
        /* @__PURE__ */ jsx(PopoverTrigger, { children: /* @__PURE__ */ jsx(
          InnerInput,
          {
            inputRef,
            dp: base_dp,
            value: quantity2,
            setValue: (e) => {
              const quantity3 = Math.abs(Math.min(Number(e), maxQty)).toString();
              setQuantity(e, maxQty);
              if (e.endsWith(".")) return;
              if (Number(quantity3) === 0) {
                setSliderValue(0);
                return;
              }
              const sliderValue2 = new Decimal(e).div(maxQty).mul(100).toDecimalPlaces(2, Decimal.ROUND_DOWN).toNumber();
              setSliderValue(Math.min(100, sliderValue2));
            },
            setEditing,
            handleKeyDown,
            onClick,
            onClose,
            onBlur: (e) => {
              formatQuantity(e.target.value);
            },
            hintInfo: error
          }
        ) }),
        /* @__PURE__ */ jsx(
          PopoverContent,
          {
            className: "oui-w-[360px] oui-rounded-xl",
            align: "start",
            side: "bottom",
            onOpenAutoFocus: (event) => {
              event.preventDefault();
            },
            children: /* @__PURE__ */ jsxs(
              Flex,
              {
                p: 1,
                gap: 2,
                width: "100%",
                itemAlign: "start",
                ref: quantitySliderRef,
                children: [
                  /* @__PURE__ */ jsx(
                    Text.numeral,
                    {
                      size: "xs",
                      intensity: 98,
                      className: "oui-min-w-[30px] ",
                      dp: 2,
                      padding: false,
                      unit: "%",
                      rm: Decimal.ROUND_DOWN,
                      children: `${sliderValue}`
                    }
                  ),
                  /* @__PURE__ */ jsxs(
                    Flex,
                    {
                      direction: "column",
                      width: "100%",
                      gap: 2,
                      className: "oui-mt-[6px]",
                      children: [
                        /* @__PURE__ */ jsx(
                          Slider,
                          {
                            markCount: 4,
                            value: [sliderValue ?? 0],
                            onValueChange: (e) => {
                              const values = Array.from(e.values());
                              setSliderValue(values[0]);
                              const quantity3 = new Decimal(values[0]).div(100).mul(maxQty).abs().toFixed(base_dp, Decimal.ROUND_DOWN);
                              setQuantity(quantity3, maxQty);
                            },
                            onValueCommit: (values) => {
                              const quantity3 = new Decimal(values[0]).div(100).mul(maxQty).abs().toFixed(base_dp, Decimal.ROUND_DOWN);
                              formatQuantity(quantity3).finally(() => {
                                inputRef.current.focus();
                              });
                            }
                          }
                        ),
                        /* @__PURE__ */ jsx(
                          Buttons,
                          {
                            onClick: (value) => {
                              setSliderValue(value * 100);
                              let quantity3 = new Decimal(value).mul(maxQty).abs().toFixed(base_dp, Decimal.ROUND_DOWN);
                              quantity3 = utils.formatNumber(quantity3, base_tick) ?? quantity3;
                              setQuantity(quantity3, maxQty);
                              setTimeout(() => {
                                inputRef.current.focus();
                                inputRef.current.setSelectionRange(
                                  quantity3.length,
                                  quantity3.length
                                );
                              }, 100);
                            }
                          }
                        )
                      ]
                    }
                  )
                ]
              }
            )
          }
        )
      ] });
    };
    Buttons = (props) => {
      const { t } = useTranslation();
      const list = [
        {
          label: "0%",
          value: 0
        },
        {
          label: "25%",
          value: 0.25
        },
        {
          label: "50%",
          value: 0.5
        },
        {
          label: "75%",
          value: 0.75
        },
        {
          label: t("common.max"),
          value: 1
        }
      ];
      return /* @__PURE__ */ jsx(Flex, { gap: 2, width: "100%", children: list.map((item, index) => {
        return /* @__PURE__ */ jsx(
          Button,
          {
            variant: "outlined",
            color: "secondary",
            size: "xs",
            onClick: (e) => {
              e.stopPropagation();
              e.preventDefault();
              props.onClick(item.value);
            },
            className: "oui-w-1/5",
            children: item.label
          },
          index
        );
      }) });
    };
  }
});
var TrailingCallbackCell;
var init_trailingCallbackCell = __esm({
  "src/components/orderList/desktop/components/trailingCallbackCell.tsx"() {
    init_type();
    init_util();
    init_symbolContext();
    init_orderListContext();
    init_confirmContent();
    init_usePopoverState();
    init_useValidateField();
    init_editableCellInput();
    init_previewCell();
    TrailingCallbackCell = (props) => {
      const { order } = props;
      const { callback_value, callback_rate } = order;
      const isCallbackValue = !!callback_value;
      const isCallbackRate = !!callback_rate;
      const originValue = useMemo(() => {
        return callback_rate ? (callback_rate * 100).toString() : callback_value?.toString();
      }, [callback_value, callback_rate]);
      const [value, setValue] = useState(originValue);
      const [submitting, setSubmitting] = useState(false);
      const disabled = props.disabled || order.is_triggered;
      const { editAlgoOrder } = useOrderListContext();
      const { base, quote_dp } = useSymbolContext();
      const {
        open,
        setOpen,
        editing,
        setEditing,
        containerRef,
        closePopover,
        cancelPopover,
        onClick
      } = usePopoverState({
        originValue,
        value,
        setValue
      });
      const { error } = useValidateField({
        order,
        originValue,
        value,
        field: isCallbackValue ? "callback_value" : "callback_rate",
        fieldValues: {
          callback_value: isCallbackValue ? value : void 0,
          callback_rate: isCallbackRate ? value : void 0
        }
      });
      const onConfirm = useCallback(() => {
        setSubmitting(true);
        const data = {
          algo_order_id: order.algo_order_id,
          callback_value: isCallbackValue ? value : void 0,
          callback_rate: isCallbackRate ? `${Number(value) / 100}` : void 0
        };
        editAlgoOrder(order.algo_order_id.toString(), data).then((result) => {
          closePopover();
          setValue(value);
        }).catch((err) => {
          toast$1.error(err.message);
          cancelPopover();
        }).finally(() => setSubmitting(false));
      }, [order.algo_order_id, value, cancelPopover]);
      const renderContent = () => {
        const suffix = isCallbackRate ? "%" : void 0;
        if (!editing || disabled) {
          return /* @__PURE__ */ jsx(
            PreviewCell,
            {
              status: getOrderStatus(order),
              value,
              setEditing,
              disabled,
              suffix
            }
          );
        }
        const formatters = isCallbackRate ? [inputFormatter.dpFormatter(1)] : [
          inputFormatter.currencyFormatter,
          inputFormatter.dpFormatter(quote_dp)
        ];
        return /* @__PURE__ */ jsx(
          EditableCellInput,
          {
            value,
            onChange: setValue,
            onClick,
            error,
            suffix,
            formatters
          }
        );
      };
      return /* @__PURE__ */ jsx(
        Popover,
        {
          open,
          onOpenChange: setOpen,
          content: /* @__PURE__ */ jsx(
            ConfirmContent,
            {
              type: isCallbackRate ? 4 /* callbackRate */ : 3 /* callbackValue */,
              base,
              value: isCallbackRate ? `${value}%` : value,
              cancelPopover,
              isSubmitting: submitting,
              onConfirm
            }
          ),
          children: /* @__PURE__ */ jsx(
            "div",
            {
              onClick: (e) => {
                e.stopPropagation();
                e.preventDefault();
              },
              ref: containerRef,
              children: renderContent()
            }
          )
        }
      );
    };
  }
});
var TriggerPriceCell;
var init_triggerPriceCell = __esm({
  "src/components/orderList/desktop/components/triggerPriceCell.tsx"() {
    init_type();
    init_util();
    init_symbolContext();
    init_orderListContext();
    init_confirmContent();
    init_usePopoverState();
    init_useValidateField();
    init_editableCellInput();
    init_previewCell();
    TriggerPriceCell = (props) => {
      const { order } = props;
      const originValue = order.trigger_price?.toString();
      const [value, setValue] = useState(originValue);
      const [isSubmitting, setIsSubmitting] = useState(false);
      const disabled = props.disabled || order.is_triggered;
      const { editAlgoOrder } = useOrderListContext();
      const { base, quote_dp } = useSymbolContext();
      const {
        open,
        setOpen,
        editing,
        setEditing,
        containerRef,
        closePopover,
        cancelPopover,
        onClick
      } = usePopoverState({
        originValue,
        value,
        setValue
      });
      const { error } = useValidateField({
        order,
        field: "trigger_price",
        originValue,
        value
      });
      const onConfirm = useCallback(() => {
        setIsSubmitting(true);
        const data = {
          trigger_price: value,
          algo_order_id: order.algo_order_id
        };
        editAlgoOrder(`${order.algo_order_id}`, data).then((result) => {
          closePopover();
          setValue(value);
        }).catch((err) => {
          toast$1.error(err.message);
          cancelPopover();
        }).finally(() => setIsSubmitting(false));
      }, [order.algo_order_id, value, cancelPopover]);
      const renderContent = () => {
        if (!editing || disabled) {
          return /* @__PURE__ */ jsx(
            PreviewCell,
            {
              value,
              status: getOrderStatus(order),
              setEditing,
              disabled
            }
          );
        }
        return /* @__PURE__ */ jsx(
          EditableCellInput,
          {
            value,
            onChange: setValue,
            onClick,
            error,
            formatters: [inputFormatter.dpFormatter(quote_dp)]
          }
        );
      };
      return /* @__PURE__ */ jsx(
        Popover,
        {
          open,
          onOpenChange: setOpen,
          content: /* @__PURE__ */ jsx(
            ConfirmContent,
            {
              type: 2 /* triggerPrice */,
              base,
              value,
              cancelPopover,
              isSubmitting,
              onConfirm
            }
          ),
          children: /* @__PURE__ */ jsx(
            "div",
            {
              onClick: (e) => {
                e.stopPropagation();
                e.preventDefault();
              },
              ref: containerRef,
              children: renderContent()
            }
          )
        }
      );
    };
  }
});
var Renew;
var init_renew = __esm({
  "src/components/orderList/desktop/renew.tsx"() {
    Renew = (props) => {
      const { record } = props;
      const [open, setOpen] = useState(false);
      const [doCreateOrder, { data, error, reset, isMutating }] = useMutation("/v1/order");
      const { t } = useTranslation();
      const brokerId = useConfig("brokerId");
      const onSubmit = useCallback(() => {
        setOpen(false);
        const data2 = {
          symbol: record.symbol,
          order_type: record.type,
          order_price: record.price,
          order_quantity: record.quantity,
          order_amount: record.amount,
          side: record.side,
          broker_id: brokerId
        };
        if (Number(record.visible_quantity) < Number(record.quantity)) {
          data2.visible_quantity = 0;
        }
        if (typeof record.reduce_only !== "undefined") {
          data2.reduce_only = record.reduce_only;
        }
        if (record.order_tag) {
          data2.order_tag = record.order_tag;
        }
        if (record.client_order_id) {
          data2.client_order_id = record.client_order_id;
        }
        doCreateOrder(data2);
      }, []);
      return /* @__PURE__ */ jsx(
        Button,
        {
          size: "sm",
          variant: "outlined",
          color: "secondary",
          loading: isMutating,
          disabled: isMutating,
          onClick: (event) => {
            if (isMutating) return;
            event.preventDefault();
            event.stopPropagation();
            onSubmit();
          },
          children: t("orders.history.renew")
        }
      );
    };
  }
});
var TP_SLEditButton;
var init_tpslEdit = __esm({
  "src/components/orderList/desktop/tpslEdit.tsx"() {
    init_symbolContext();
    init_tpslOrderRowContext();
    TP_SLEditButton = (props) => {
      const { position, order } = useTPSLOrderRowContext();
      const { quote_dp, base_dp } = useSymbolContext();
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        PositionTPSLPopover,
        {
          quoteDP: quote_dp,
          baseDP: base_dp,
          position,
          order,
          label: t("common.edit"),
          isEditing: true
        }
      );
    };
  }
});
function useTPSLOrderPrice(order) {
  const { sl_trigger_price, tp_trigger_price } = !("algo_type" in order) || !Array.isArray(order.child_orders) ? {} : findTPSLFromOrder(order);
  const tpTriggerPrice = tp_trigger_price ? `${i18n.t("tpsl.tp")} - ${i18n.t("common.marketPrice")}` : void 0;
  const slTriggerPrice = sl_trigger_price ? `${i18n.t("tpsl.sl")} - ${i18n.t("common.marketPrice")}` : void 0;
  return { tpTriggerPrice, slTriggerPrice };
}
var TPSLOrderPrice, TPSLOrderPriceItem;
var init_tpslPrice = __esm({
  "src/components/orderList/desktop/tpslPrice.tsx"() {
    init_tpslOrderRowContext();
    TPSLOrderPrice = () => {
      const {
        sl_trigger_price,
        tp_trigger_price,
        sl_order_price,
        tp_order_price,
        order
      } = useTPSLOrderRowContext();
      const { t } = useTranslation();
      const symbolInfo = useSymbolsInfo()[order.symbol]();
      return /* @__PURE__ */ jsxs("div", { children: [
        !!tp_trigger_price ? /* @__PURE__ */ jsxs("div", { className: "oui-text-base-contrast-80 oui-td-bg-transparent", children: [
          /* @__PURE__ */ jsxs(Text, { intensity: 54, children: [
            `${t("tpsl.tp")} -`,
            "\xA0"
          ] }),
          /* @__PURE__ */ jsx(
            TPSLOrderPriceItem,
            {
              price: tp_order_price,
              quoteDP: symbolInfo.quote_dp
            }
          )
        ] }) : null,
        !!sl_trigger_price ? /* @__PURE__ */ jsxs("div", { className: "oui-text-base-contrast-80 oui-td-bg-transparent", children: [
          /* @__PURE__ */ jsxs("span", { className: "oui-text-base-contrast-54", children: [
            `${t("tpsl.sl")} -`,
            "\xA0"
          ] }),
          /* @__PURE__ */ jsx(
            TPSLOrderPriceItem,
            {
              price: sl_order_price,
              quoteDP: symbolInfo.quote_dp
            }
          )
        ] }) : null
      ] });
    };
    TPSLOrderPriceItem = ({
      price: price2,
      quoteDP
    }) => {
      const { t } = useTranslation();
      if (!price2) {
        return null;
      }
      if (price2 === OrderType.MARKET) {
        return /* @__PURE__ */ jsx("span", { children: t("common.marketPrice") });
      }
      return /* @__PURE__ */ jsx(
        Text.numeral,
        {
          className: "oui-text-base-contrast-80",
          rule: "price",
          dp: quoteDP,
          children: price2
        }
      );
    };
  }
});
var OrderTriggerPrice, TPSLTriggerPrice, TriggerPriceItem;
var init_tpslTriggerPrice = __esm({
  "src/components/orderList/desktop/tpslTriggerPrice.tsx"() {
    init_tpslOrderRowContext();
    OrderTriggerPrice = () => {
      const { sl_trigger_price, tp_trigger_price, order, position } = useTPSLOrderRowContext();
      const positionSide = position?.position_qty && position.position_qty > 0 ? OrderSide.BUY : OrderSide.SELL;
      const slPriceError = useTpslPriceChecker({
        slPrice: sl_trigger_price?.toString() ?? void 0,
        liqPrice: position?.est_liq_price ?? null,
        side: positionSide
      });
      return /* @__PURE__ */ jsx(
        TPSLTriggerPrice,
        {
          stopLossPrice: sl_trigger_price,
          takeProfitPrice: tp_trigger_price,
          direction: "column",
          order,
          position,
          slPriceError,
          tooltip: true
        }
      );
    };
    TPSLTriggerPrice = (props) => {
      const { direction = "row", order, position } = props;
      const symbolInfo = useSymbolsInfo();
      const { t } = useTranslation();
      const pnl = useMemo(() => {
        const msgs = [];
        if (!props.tooltip || !order || !position) return;
        let quantity2 = order.quantity;
        if (quantity2 === 0) {
          if (order.child_orders?.[0].type === "CLOSE_POSITION") {
            quantity2 = position.position_qty;
          }
        }
        if (!!props.takeProfitPrice) {
          msgs.push(
            /* @__PURE__ */ jsx(
              TriggerPriceItem,
              {
                qty: quantity2,
                price: props.takeProfitPrice,
                entryPrice: position.average_open_price,
                orderSide: order.side,
                orderType: AlgoOrderType.TAKE_PROFIT,
                symbolInfo: symbolInfo[order.symbol]()
              },
              "tp"
            )
          );
        }
        if (!!props.stopLossPrice) {
          msgs.push(
            /* @__PURE__ */ jsx(
              TriggerPriceItem,
              {
                qty: quantity2,
                price: props.stopLossPrice,
                entryPrice: position.average_open_price,
                orderSide: order.side,
                orderType: AlgoOrderType.STOP_LOSS,
                symbolInfo: symbolInfo[order.symbol]()
              },
              "sl"
            )
          );
        }
        return /* @__PURE__ */ jsx("div", { children: msgs });
      }, [
        props.takeProfitPrice,
        props.stopLossPrice,
        position?.average_open_price,
        order?.side,
        order?.quantity,
        order?.algo_type
      ]);
      const child = useMemo(() => {
        const children = [];
        if (!order?.symbol) return /* @__PURE__ */ jsx("span", { children: "-" });
        if (props.takeProfitPrice) {
          children.push(
            /* @__PURE__ */ jsx(
              Text.numeral,
              {
                className: cn(
                  "oui-text-trade-profit oui-gap-0  oui-decoration-white/20 oui-border-b oui-border-dashed oui-border-base-contrast-12"
                ),
                rule: "price",
                dp: symbolInfo[order.symbol]("quote_dp", 2),
                children: props.takeProfitPrice,
                prefix: !props.stopLossPrice || direction === "column" ? /* @__PURE__ */ jsxs("span", { className: "oui-text-base-contrast-54", children: [
                  `${t("tpsl.tp")} -`,
                  "\xA0"
                ] }) : ""
              },
              "tp"
            )
          );
        }
        if (props.stopLossPrice) {
          children.push(
            /* @__PURE__ */ jsx(
              Text.numeral,
              {
                className: cn(
                  "oui-text-trade-loss oui-gap-0 oui-decoration-white/20 oui-border-b oui-border-dashed oui-border-base-contrast-12",
                  "oui-flex oui-items-center"
                ),
                as: "div",
                rule: "price",
                dp: symbolInfo[order.symbol]("quote_dp", 2),
                children: props.stopLossPrice,
                prefix: !props.takeProfitPrice || direction === "column" ? /* @__PURE__ */ jsxs("span", { className: "oui-text-base-contrast-54", children: [
                  `${t("tpsl.sl")} -`,
                  "\xA0"
                ] }) : "",
                suffix: /* @__PURE__ */ jsx(
                  CloseToLiqPriceIcon,
                  {
                    slPriceError: props.slPriceError,
                    className: "oui-ml-1"
                  }
                )
              },
              "sl"
            )
          );
        }
        if (children.length === 0) return /* @__PURE__ */ jsx("span", { children: "-" });
        if (children.length === 2 && direction === "row") {
          children.splice(1, 0, /* @__PURE__ */ jsx("span", { children: "/" }, "split"));
        }
        return children;
      }, [
        props.takeProfitPrice,
        props.stopLossPrice,
        order?.symbol,
        t,
        props.slPriceError
      ]);
      const content = /* @__PURE__ */ jsx(
        "div",
        {
          className: cn(
            "oui-inline-flex oui-text-base-contrast-36",
            props.direction === "column" ? "oui-flex-col" : "oui-flex-row oui-gap-1",
            props.className
          ),
          children: child
        }
      );
      if (props.tooltip) {
        return /* @__PURE__ */ jsx(
          Tooltip,
          {
            content: pnl,
            className: "oui-bg-base-5",
            arrow: { className: "oui-fill-base-5" },
            children: content
          }
        );
      }
      return content;
    };
    TriggerPriceItem = (props) => {
      const { qty, price: price2, entryPrice, orderSide, orderType, symbolInfo } = props;
      const { t } = useTranslation();
      const pnl = utils.priceToPnl(
        {
          qty,
          price: price2,
          entryPrice,
          orderSide,
          orderType
        },
        {
          symbol: symbolInfo
        }
      );
      const type = orderType === AlgoOrderType.TAKE_PROFIT ? "TP" : "SL";
      const label = type === "TP" ? `${t("tpsl.tpPnl")}:` : `${t("tpsl.slPnl")}:`;
      return /* @__PURE__ */ jsxs("div", { className: "oui-flex oui-items-center", children: [
        /* @__PURE__ */ jsx("span", { className: "oui-text-base-contrast-54 oui-mr-1", children: label }),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            rule: "price",
            dp: symbolInfo.quote_dp,
            padding: false,
            className: pnl === 0 ? "oui-text-base-contrast-36" : pnl > 0 ? "oui-text-trade-profit oui-gap-0" : "oui-text-trade-loss oui-gap-0",
            prefix: (
              // @ts-ignore
              /* @__PURE__ */ jsx("span", { children: pnl === 0 ? "" : pnl > 0 ? "+" : "-" })
            ),
            suffix: /* @__PURE__ */ jsx("span", { className: "oui-text-base-contrast-36 oui-ml-1", children: "USDC" }),
            children: `${Math.abs(pnl)}`
          }
        )
      ] });
    };
  }
});
function instrument(option) {
  return {
    title: i18n.t("common.symbol"),
    dataIndex: "symbol",
    fixed: "left",
    // className: "oui-h-[48px]",
    width: option?.width,
    onSort: option?.enableSort ? (r1, r2) => {
      return r1.symbol.localeCompare(r2.symbol);
    } : void 0,
    renderPlantText: (value, record) => {
      const badges = parseBadgesFor(record)?.join(",");
      const displayBadges = badges?.length ? ` (${badges})` : "";
      return `${value.split("_")[1]}-PERP${displayBadges}`;
    },
    render: (value, record) => {
      const showGray = grayCell(record);
      return /* @__PURE__ */ jsxs(Flex, { gap: 2, children: [
        /* @__PURE__ */ jsx(
          "div",
          {
            className: cn(
              "oui-h-7 oui-w-1 oui-shrink-0 oui-rounded-[1px]",
              record.side === OrderSide.BUY ? "oui-bg-trade-profit" : "oui-bg-trade-loss"
            )
          }
        ),
        /* @__PURE__ */ jsxs(Flex, { direction: "column", itemAlign: "start", children: [
          /* @__PURE__ */ jsx(
            Text.formatted,
            {
              size: "xs",
              className: "oui-cursor-pointer oui-text-xs",
              onClick: (e) => {
                option?.onSymbolChange?.({ symbol: value });
                e.stopPropagation();
                e.preventDefault();
              },
              children: `${value.split("_")[1]}-PERP`
            }
          ),
          option?.showType && /* @__PURE__ */ jsx(Flex, { direction: "row", gap: 1, children: parseBadgesFor(record)?.map((e, index) => /* @__PURE__ */ jsx(
            Badge,
            {
              color: e.toLocaleLowerCase() === "position" ? showGray ? "neutral" : "primary" : "neutral",
              size: "xs",
              className: "oui-whitespace-nowrap oui-break-normal",
              children: e
            },
            index
          )) })
        ] })
      ] });
    }
  };
}
function side(option) {
  return {
    title: i18n.t("common.side"),
    dataIndex: "side",
    width: option?.width,
    // className: "oui-h-[48px]",
    onSort: option?.enableSort ? (r1, r2, sortOrder) => {
      return r2.side.localeCompare(r1.side);
    } : void 0,
    renderPlantText: (value, record) => upperCaseFirstLetter(value),
    render: (value, record) => {
      const clsName = grayCell(record) ? "oui-text-base-contrast-20" : value === OrderSide.BUY ? "oui-text-trade-profit" : "oui-text-trade-loss";
      return /* @__PURE__ */ jsx("span", { className: cn("oui-font-semibold", clsName), children: upperCaseFirstLetter(value) });
    }
  };
}
function fillAndQuantity(option) {
  return {
    title: i18n.t("orders.column.fill&Quantity"),
    dataIndex: "fill_quantity",
    className: option?.className,
    width: option?.width,
    onSort: option?.enableSort ?? false ? (a, b) => {
      const aQuantity = (a.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL ? 0 : a.quantity) ?? 0;
      const bQuantity = (b.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL ? 0 : b.quantity) ?? 0;
      return compareNumbers(aQuantity, bQuantity);
    } : void 0,
    renderPlantText: (value, record) => {
      if (record.type === OrderType.CLOSE_POSITION && record.status !== OrderStatus.FILLED) {
        return i18n.t("tpsl.entirePosition");
      }
      const executed = record.total_executed_quantity;
      const first = "algo_type" in record && record.algo_type === AlgoOrderRootType.TP_SL ? "" : `${executed} / `;
      return first + `${record.quantity}`;
    },
    render: (value, record) => {
      if (record.type === OrderType.CLOSE_POSITION && record.status !== OrderStatus.FILLED) {
        return i18n.t("tpsl.entirePosition");
      }
      return /* @__PURE__ */ jsx(QuantityCell, { order: record, disabled: option?.disableEdit });
    }
  };
}
function quantity(option) {
  return {
    title: i18n.t("common.quantity"),
    className: option?.className,
    dataIndex: "quantity",
    width: option?.width,
    onSort: option?.enableSort ?? false ? (a, b) => {
      const aQuantity = (a.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL ? 0 : a.quantity) ?? 0;
      const bQuantity = (b.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL ? 0 : b.quantity) ?? 0;
      return compareNumbers(aQuantity, bQuantity);
    } : void 0,
    renderPlantText: (value, record) => {
      if (record.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL) {
        return i18n.t("tpsl.entirePosition");
      }
      const executed = record.total_executed_quantity;
      const first = "algo_type" in record && record.algo_type === AlgoOrderRootType.TP_SL ? "" : `${executed}/`;
      return first + `${record.quantity}`;
    },
    render: (value, record) => {
      if (record.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL) {
        return i18n.t("tpsl.entirePosition");
      }
      return /* @__PURE__ */ jsx(QuantityCell, { order: record });
    }
  };
}
function price(option) {
  return {
    title: option?.title ?? i18n.t("common.price"),
    dataIndex: "price",
    className: option?.className,
    width: option?.width,
    onSort: option?.enableSort ?? false ? (a, b, type) => {
      return compareNumbers(a.price ?? 0, b.price ?? 0);
    } : void 0,
    renderPlantText: (value, record) => {
      const isTrailingStop = isTrailingStopOrder(record);
      if (isTrailingStop) {
        return option?.tabType === "orderHistory" /* orderHistory */ ? i18n.t("common.marketPrice") : getTrailingStopPrice(record) || "--";
      }
      return commifyOptional(record.price, {
        fallback: i18n.t("common.marketPrice")
      });
    },
    render: (value, record) => {
      const isTrailingStop = isTrailingStopOrder(record);
      if (isTrailingStop) {
        return option?.tabType === "orderHistory" /* orderHistory */ ? i18n.t("common.marketPrice") : getTrailingStopPrice(record) || "--";
      }
      return /* @__PURE__ */ jsx(PriceCell, { order: record, disabled: option?.disableEdit });
    }
  };
}
function tpslPrice(option) {
  return {
    title: option?.title ?? i18n.t("common.price"),
    dataIndex: "price",
    className: option?.className,
    width: option?.width,
    onSort: option?.enableSort,
    renderPlantText: (value, record) => {
      const { tpTriggerPrice, slTriggerPrice } = useTPSLOrderPrice(record);
      const callback = `${tpTriggerPrice || ""}${slTriggerPrice ? `${tpTriggerPrice ? "\n" : ""}${slTriggerPrice}` : ""}`;
      return callback.length > 0 ? callback : "--";
    },
    render: (value, record) => {
      return /* @__PURE__ */ jsx(TPSLOrderPrice, {});
    }
  };
}
function avgPrice(option) {
  return {
    title: i18n.t("common.avgPrice"),
    dataIndex: "average_executed_price",
    className: option?.className,
    width: option?.width,
    onSort: option?.enableSort,
    renderPlantText: (value, record) => {
      return commifyOptional(value);
    },
    render: (value, record) => {
      return /* @__PURE__ */ jsx(AvgPrice, { symbol: record.symbol, value });
    }
  };
}
function triggerPrice(option) {
  const { isPending } = option || {};
  const checkOrderType = (record) => {
    const isAlgoOrder = record?.algo_order_id !== void 0;
    const isBracketOrder = record?.algo_type === "BRACKET";
    return isAlgoOrder && !isBracketOrder;
  };
  return {
    title: i18n.t("common.trigger"),
    className: option?.className,
    dataIndex: "trigger_price",
    width: option?.width,
    onSort: option?.enableSort,
    renderPlantText: (value, record) => {
      if (!checkOrderType(record)) {
        return "--";
      }
      if (isTrailingStopOrder(record) && isPending) {
        return record.activated_price ? commifyOptional(record.activated_price) : "--";
      }
      return commifyOptional(value);
    },
    render: (value, record) => {
      if (!checkOrderType(record)) {
        return "--";
      }
      if (isTrailingStopOrder(record) && isPending) {
        return /* @__PURE__ */ jsx(ActivedPriceCell, { order: record, disabled: option?.disableEdit });
      }
      return /* @__PURE__ */ jsx(TriggerPriceCell, { order: record, disabled: option?.disableEdit });
    }
  };
}
function tpslTriggerPrice(option) {
  return {
    title: option?.title ?? i18n.t("common.trigger"),
    className: option?.className,
    dataIndex: "tpsl_trigger_price",
    width: option?.width,
    onSort: option?.enableSort,
    renderPlantText: (value, record) => {
      const info = option?.symbolsInfo?.[record.symbol];
      const quote_dp = info?.("quote_dp");
      const { sl_trigger_price, tp_trigger_price } = !("algo_type" in record) || !Array.isArray(record.child_orders) ? {} : utils.findTPSLFromOrder(record);
      const callback = (tp_trigger_price != null ? `${i18n.t("tpsl.tp")}: ${commifyOptional(tp_trigger_price, {
        fix: quote_dp,
        padEnd: true
      })}` : "") + (sl_trigger_price != null ? `${tp_trigger_price ? "\n" : ""}${i18n.t(
        "tpsl.sl"
      )}: ${commifyOptional(sl_trigger_price, {
        fix: quote_dp,
        padEnd: true
      })}` : "");
      return callback.length > 0 ? callback : "--";
    },
    render: (value, record) => /* @__PURE__ */ jsx(OrderTriggerPrice, {})
  };
}
function bracketOrderPrice(option) {
  return {
    title: i18n.t("common.tpsl"),
    className: option?.className,
    dataIndex: "bracketOrderPrice",
    width: option?.width,
    onSort: option?.enableSort,
    renderPlantText: (value, record) => {
      const getTPSLTriggerPrice = () => {
        if (!("algo_type" in record) || !Array.isArray(record.child_orders)) {
          return {};
        }
        return utils.findTPSLFromOrder(record.child_orders[0]);
      };
      const { sl_trigger_price, tp_trigger_price } = getTPSLTriggerPrice();
      const callback = (tp_trigger_price != null ? `${i18n.t("tpsl.tp")}: ${tp_trigger_price}` : "") + (sl_trigger_price != null ? `${tp_trigger_price ? "\n" : ""}${i18n.t(
        "tpsl.sl"
      )}: ${sl_trigger_price}` : "");
      return callback.length > 0 ? callback : "--";
    },
    render: (value, record) => /* @__PURE__ */ jsx(BracketOrderPrice, { order: record })
  };
}
function estTotal(option) {
  return {
    title: i18n.t("common.notional"),
    dataIndex: "executed",
    width: option?.width,
    className: option?.className,
    onSort: option?.enableSort ?? false ? (a, b, type) => {
      const aTotal = a.type === OrderType.CLOSE_POSITION && a.status !== OrderStatus.FILLED ? 0 : a.total_executed_quantity === 0 || Number.isNaN(a.average_executed_price) || a.average_executed_price === null ? 0 : a.total_executed_quantity * a.average_executed_price;
      const bTotal = b.type === OrderType.CLOSE_POSITION && b.status !== OrderStatus.FILLED ? 0 : b.total_executed_quantity === 0 || Number.isNaN(b.average_executed_price) || b.average_executed_price === null ? 0 : b.total_executed_quantity * b.average_executed_price;
      return compareNumbers(aTotal, bTotal);
    } : void 0,
    renderPlantText: (value, record) => {
      const estTotal2 = estTotalValue(record, option?.isPending ?? false);
      if (estTotal2 === "Entire position") {
        return i18n.t("tpsl.entirePosition");
      }
      return commifyOptional(
        estTotalValue(record, option?.isPending ?? false),
        { fix: 2 }
      );
    },
    render: (value, record) => {
      const estTotal2 = estTotalValue(record, option?.isPending ?? false);
      if (estTotal2 === "Entire position") {
        return i18n.t("tpsl.entirePosition");
      }
      return /* @__PURE__ */ jsx(Text.numeral, { rm: Decimal.ROUND_DOWN, dp: 2, children: estTotal2 });
    }
  };
}
function realizedPnL(option) {
  return {
    title: i18n.t("common.realizedPnl"),
    dataIndex: "realized_pnl",
    width: option?.width,
    className: option?.className,
    renderPlantText: (_value, record) => {
      const info = option?.symbolsInfo?.[record.symbol];
      const quote_dp = info?.("quote_dp");
      const dp = option?.pnlNotionalDecimalPrecision ?? quote_dp;
      const value = new Decimal(_value ?? 0).toDecimalPlaces(dp, Decimal.ROUND_DOWN).toNumber();
      const formatValue = commifyOptional(value);
      return value > 0 ? `+${formatValue}` : formatValue;
    },
    render: (_value, record) => {
      const { quote_dp } = useSymbolContext();
      const dp = option?.pnlNotionalDecimalPrecision ?? quote_dp;
      const value = new Decimal(_value ?? 0).toNumber();
      return /* @__PURE__ */ jsxs(Flex, { gap: 1, children: [
        /* @__PURE__ */ jsx(
          Text.pnl,
          {
            dp,
            padding: false,
            intensity: (value ?? 0) == 0 ? 80 : void 0,
            showIdentifier: (value ?? 0) > 0,
            coloring: (value ?? 0) != 0,
            children: value ?? "--"
          }
        ),
        !option?.hideShare && /* @__PURE__ */ jsx(
          ShareButtonWidget,
          {
            order: record,
            sharePnLConfig: option?.sharePnLConfig,
            modalId: SharePnLDialogId
          }
        )
      ] });
    }
  };
}
function reduceOnly(option) {
  return {
    title: i18n.t("orderEntry.reduceOnly"),
    dataIndex: "reduce_only",
    width: option?.width,
    className: option?.className,
    renderPlantText: (value, record) => {
      return value ? i18n.t("common.yes") : i18n.t("common.no");
    },
    render: (value) => {
      return /* @__PURE__ */ jsx(Text, { children: value ? i18n.t("common.yes") : i18n.t("common.no") });
    }
  };
}
function hidden(option) {
  return {
    title: i18n.t("orders.column.hidden"),
    dataIndex: "visible",
    width: option?.width,
    className: option?.className,
    renderPlantText: (value, record) => {
      return value !== 0 ? i18n.t("common.no") : i18n.t("common.yes");
    },
    render: (value, record) => {
      return /* @__PURE__ */ jsx(Text, { children: record.visible_quantity !== 0 ? i18n.t("common.no") : i18n.t("common.yes") });
    }
  };
}
function orderTime(option) {
  return {
    title: i18n.t("orders.column.orderTime"),
    dataIndex: "created_time",
    width: option?.width,
    onSort: option?.enableSort,
    className: option?.className,
    renderPlantText: (value, record) => {
      const date = new Date(value);
      const formattedDate = format(date, "yyyy-MM-dd HH:mm:ss");
      return formattedDate;
    },
    render: (value) => /* @__PURE__ */ jsx(
      Text.formatted,
      {
        rule: "date",
        formatString: option?.formatString || "yyyy-MM-dd HH:mm:ss",
        className: "oui-break-normal oui-whitespace-nowrap oui-font-semibold",
        children: value
      }
    )
  };
}
function fee(option) {
  return {
    title: i18n.t("common.fee"),
    dataIndex: "total_fee",
    width: option?.width,
    onSort: option?.enableSort,
    className: option?.className
  };
}
function tpslNotional(option) {
  return {
    title: i18n.t("common.notional"),
    dataIndex: "executed",
    width: option?.width,
    onSort: option?.enableSort,
    className: option?.className,
    renderPlantText: (value, record) => {
      if (record.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL) {
        return i18n.t("tpsl.entirePosition");
      }
      return commifyOptional(
        record.quantity === 0 ? "--" : `${formatNum.notional(new Decimal(record.mark_price).mul(record.quantity))?.toNumber()}`
      );
    },
    render: (value, record) => {
      if (record.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL) {
        return i18n.t("tpsl.entirePosition");
      }
      return /* @__PURE__ */ jsx(Text.numeral, { className: "oui-break-normal oui-whitespace-nowrap oui-font-semibold", children: record.quantity === 0 ? "--" : `${formatNum.notional(new Decimal(record.mark_price).mul(record.quantity))?.toNumber()}` });
    }
  };
}
function status(option) {
  const statusMap = {
    [OrderStatus.NEW]: i18n.t("orders.status.pending"),
    [OrderStatus.FILLED]: i18n.t("orders.status.filled"),
    [OrderStatus.PARTIAL_FILLED]: i18n.t("orders.status.partialFilled"),
    [OrderStatus.CANCELLED]: i18n.t("orders.status.canceled"),
    [OrderStatus.REJECTED]: i18n.t("orders.status.rejected"),
    [OrderStatus.INCOMPLETE]: i18n.t("orders.status.incomplete"),
    [OrderStatus.COMPLETED]: i18n.t("orders.status.completed")
  };
  return {
    title: i18n.t("common.status"),
    dataIndex: "status",
    width: option?.width,
    onSort: option?.enableSort,
    className: option?.className,
    renderPlantText: (value, record) => {
      const status2 = value || record.algo_status;
      return statusMap[status2] || upperCaseFirstLetter(status2);
    },
    render: (value, record) => {
      const status2 = value || record.algo_status;
      return statusMap[status2] || upperCaseFirstLetter(status2);
    }
  };
}
function avgOpen(option) {
  return {
    title: i18n.t("common.avgPrice"),
    dataIndex: "average_executed_price",
    width: option?.width,
    onSort: option?.enableSort ?? false ? (a, b) => {
      return compareNumbers(
        a.average_executed_price ?? 0,
        b.average_executed_price ?? 0
      );
    } : void 0,
    className: option?.className,
    render: (value, record) => {
      if (record.type === OrderType.MARKET && !value) {
        return "--";
      }
      return /* @__PURE__ */ jsx(AvgPrice, { symbol: record.symbol, value });
    },
    renderPlantText: (value, record) => {
      return value ? commifyOptional(value) : "--";
    }
  };
}
function cancelBtn(option) {
  return {
    title: "",
    type: "action",
    dataIndex: "action",
    width: option?.width,
    className: option?.className,
    align: "right",
    fixed: "right",
    render: (_, record) => {
      if (record.status === OrderStatus.CANCELLED) {
        return /* @__PURE__ */ jsx(Renew, { record });
      }
      if (record.status === OrderStatus.NEW || record.algo_status === OrderStatus.NEW) {
        return /* @__PURE__ */ jsx(CancelButton, { order: record });
      }
      return null;
    }
  };
}
function pendingTabCancelBtn(option) {
  return {
    title: "",
    type: "action",
    dataIndex: "action",
    width: option?.width,
    className: option?.className,
    align: "right",
    fixed: "right",
    render: (_, record) => {
      return /* @__PURE__ */ jsx(CancelButton, { order: record });
    }
  };
}
function tpslAction(option) {
  return {
    title: "",
    type: "action",
    dataIndex: "action",
    width: option?.width,
    className: option?.className,
    align: "right",
    fixed: "right",
    render: (_, record) => {
      return /* @__PURE__ */ jsxs(Flex, { gap: 3, children: [
        /* @__PURE__ */ jsx(TP_SLEditButton, { order: record }),
        /* @__PURE__ */ jsx(CancelButton, { order: record })
      ] });
    }
  };
}
function compareNumbers(a, b) {
  if (a > b) return 1;
  if (a < b) return -1;
  return 0;
}
function estTotalValue(record, isPending) {
  if (isPending) {
    return getNotional(record) || "--";
  }
  if (record.type === OrderType.CLOSE_POSITION && record.status !== OrderStatus.FILLED) {
    return i18n.t("tpsl.entirePosition");
  }
  return record.total_executed_quantity === 0 || Number.isNaN(record.average_executed_price) || record.average_executed_price === null ? "--" : `${record.total_executed_quantity * record.average_executed_price}`;
}
function trailingCallback(option) {
  return {
    title: option?.title ?? i18n.t("orderEntry.trailing"),
    dataIndex: "callback_value",
    className: option?.className,
    width: option?.width,
    renderPlantText: (value, record) => {
      const { callback_value, callback_rate } = record;
      const val = callback_value || (callback_rate ? `${callback_rate * 100}%` : "--");
      return val?.toString();
    },
    render: (value, record) => {
      if (isTrailingStopOrder(record)) {
        return /* @__PURE__ */ jsx(TrailingCallbackCell, { order: record, disabled: option?.disableEdit });
      }
      return "--";
    }
  };
}
var useOrderColumn;
var init_useColumn = __esm({
  "src/components/orderList/desktop/useColumn.tsx"() {
    init_util();
    init_orders_widget();
    init_symbolContext();
    init_shareButton();
    init_avgPrice();
    init_bracketOrderPrice();
    init_cancelBtn();
    init_activedPriceCell();
    init_priceCell();
    init_quantityCell();
    init_trailingCallbackCell();
    init_triggerPriceCell();
    init_renew();
    init_tpslEdit();
    init_tpslPrice();
    init_tpslTriggerPrice();
    useOrderColumn = (props) => {
      const {
        _type,
        onSymbolChange,
        pnlNotionalDecimalPrecision,
        sharePnLConfig,
        symbolsInfo
      } = props;
      const { t } = useTranslation();
      const columns = useMemo(() => {
        switch (_type) {
          case "all" /* all */:
            return [
              instrument({
                width: 130,
                showType: true,
                onSymbolChange,
                enableSort: false
              }),
              side({ width: 130 }),
              fillAndQuantity({
                width: 130,
                disableEdit: true,
                className: "oui-pl-0 oui-pr-0",
                enableSort: false
              }),
              price({
                width: 130,
                title: t("common.orderPrice"),
                disableEdit: true,
                enableSort: false
              }),
              avgOpen({ width: 130, enableSort: false}),
              tpslTriggerPrice({ width: 130, symbolsInfo: props.symbolsInfo }),
              realizedPnL({
                width: 124,
                pnlNotionalDecimalPrecision,
                sharePnLConfig,
                symbolsInfo: props.symbolsInfo,
                hideShare: true
              }),
              estTotal({ width: 130, enableSort: false }),
              fee({ width: 130 }),
              status({ width: 130 }),
              reduceOnly({ width: 130 }),
              hidden({ width: 130 }),
              cancelBtn({ width: 130 }),
              orderTime({ width: 160, enableSort: false })
            ];
          case "pending" /* pending */:
            return [
              instrument({
                width: 172,
                showType: true,
                onSymbolChange,
                enableSort: false
              }),
              // side({ width: 162 }),
              fillAndQuantity({
                width: 162,
                className: "oui-pr-0",
                enableSort: false
              }),
              price({ width: 162, className: "oui-pr-0", enableSort: false }),
              trailingCallback({ width: 162 }),
              triggerPrice({ width: 162, className: "oui-pr-0", isPending: true }),
              bracketOrderPrice({ width: 130 }),
              estTotal({ width: 162, isPending: true }),
              reduceOnly({ width: 162 }),
              hidden({ width: 162 }),
              orderTime({ width: 162, enableSort: false }),
              pendingTabCancelBtn({ width: 80 })
            ];
          case "tp_sl" /* tp_sl */:
            return [
              instrument({
                width: 176,
                showType: true,
                onSymbolChange,
                enableSort: false
              }),
              // side({ width: 176 }),
              quantity({ width: 176 }),
              tpslTriggerPrice({ width: 176, symbolsInfo }),
              tpslPrice({ width: 176}),
              tpslNotional({ width: 176 }),
              reduceOnly({ width: 176 }),
              orderTime({ width: 176, enableSort: false }),
              tpslAction({ width: 176 })
            ];
          case "filled" /* filled */:
            return [
              instrument({
                showType: true,
                width: 154,
                onSymbolChange
              }),
              // type({ width: 124 }),
              // side({ width: 124 }),
              fillAndQuantity({
                width: 124,
                disableEdit: true,
                className: "oui-pl-0 oui-pr-0"
              }),
              price({
                width: 124,
                title: t("common.orderPrice"),
                disableEdit: true
              }),
              avgPrice({ width: 124 }),
              triggerPrice({ width: 124, disableEdit: true }),
              realizedPnL({
                width: 124,
                pnlNotionalDecimalPrecision,
                sharePnLConfig,
                symbolsInfo,
                hideShare: true
              }),
              estTotal({ width: 124 }),
              fee({ width: 124 }),
              status({ width: 124 }),
              reduceOnly({ width: 124 }),
              hidden({ width: 124 }),
              orderTime({ width: 176 })
            ];
          case "cancelled" /* cancelled */:
            return [
              instrument({
                showType: true,
                width: 154,
                onSymbolChange,
                enableSort: false
              }),
              // side({ width: 124 }),
              fillAndQuantity({
                width: 124,
                disableEdit: true,
                className: "oui-pl-0 oui-pr-0",
                enableSort: false
              }),
              price({ width: 124, disableEdit: true, enableSort: false }),
              avgOpen({ width: 124, enableSort: false }),
              triggerPrice({ width: 124, disableEdit: true }),
              estTotal({ width: 124 }),
              fee({ width: 124 }),
              status({ width: 124 }),
              reduceOnly({ width: 124 }),
              hidden({ width: 124 })
            ];
          case "rejected" /* rejected */:
            return [
              instrument({
                showType: true,
                width: 154,
                onSymbolChange
              }),
              // side({ width: 124 }),
              fillAndQuantity({
                width: 124,
                disableEdit: true,
                className: "oui-pl-0 oui-pr-0"
              }),
              price({ width: 124, disableEdit: true }),
              avgOpen({ width: 124 }),
              triggerPrice({ width: 124, disableEdit: true }),
              estTotal({ width: 124 }),
              fee({ width: 124 }),
              status({ width: 124 }),
              reduceOnly({ width: 124 }),
              hidden({ width: 124 }),
              orderTime({ width: 176 })
            ];
          case "orderHistory" /* orderHistory */:
            return [
              instrument({
                showType: true,
                width: 154,
                onSymbolChange
              }),
              // side({ width: 124 }),
              fillAndQuantity({
                width: 150,
                disableEdit: true,
                className: "oui-pl-6 oui-pr-0"
              }),
              price({
                width: 124,
                disableEdit: true,
                tabType: "orderHistory" /* orderHistory */
              }),
              avgOpen({ width: 124 }),
              triggerPrice({ width: 124, disableEdit: true }),
              realizedPnL({
                width: 124,
                pnlNotionalDecimalPrecision,
                sharePnLConfig,
                symbolsInfo: props.symbolsInfo,
                hideShare: true
              }),
              estTotal({ width: 124 }),
              trailingCallback({ width: 124, disableEdit: true }),
              fee({ width: 124 }),
              status({ width: 124 }),
              reduceOnly({ width: 124 }),
              hidden({ width: 124 }),
              orderTime({ width: 150 }),
              cancelBtn({ width: 80 })
            ];
        }
      }, [_type, pnlNotionalDecimalPrecision, sharePnLConfig, t]);
      return columns;
    };
  }
});
var SymbolToken, OrderTypeView, OrderTime, OrderState, Qty, Filled, EstTotal, TriggerPrice, MarkPrice, LimitPrice, TPTrigger, SLTrigger, TPPrice, SLPrice, TPSLQuantity, AvgPrice2, OrderPrice, RealizedPnL, MobileTooltip, Fee, TrailingCallback, TrailingPrice;
var init_items = __esm({
  "src/components/orderList/mobile/items.tsx"() {
    init_util();
    init_tpslOrderRowContext();
    SymbolToken = (props) => {
      const { item } = props;
      const isBuy = item.side === OrderSide.BUY;
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Text.formatted,
        {
          intensity: 80,
          rule: "symbol",
          formatString: "base-type",
          size: "sm",
          prefix: /* @__PURE__ */ jsx(Badge, { color: isBuy ? "success" : "danger", size: "xs", children: isBuy ? t("common.buy") : t("common.sell") }),
          onClick: () => {
            props.onSymbolChange?.({ symbol: item.symbol });
          },
          children: item.symbol
        }
      );
    };
    OrderTypeView = (props) => {
      const { item } = props;
      useCallback(() => {
        const type = typeof item.type === "string" ? item.type.replace("_ORDER", "").toLowerCase() : item.type;
        const isAlgoOrder = item.algo_order_id && item.algo_type !== AlgoOrderRootType.BRACKET;
        if (isAlgoOrder) {
          return `Stop ${type}`;
        }
        return upperCaseFirstLetter(item.type);
      }, [item]);
      return /* @__PURE__ */ jsx(Flex, { direction: "row", gap: 1, children: parseBadgesFor(props.item)?.map((e, index) => /* @__PURE__ */ jsx(
        Badge,
        {
          color: e.toLocaleLowerCase() === "position" ? "primary" : "neutral",
          size: "xs",
          children: e
        },
        index
      )) });
    };
    OrderTime = (props) => {
      const { item } = props;
      return /* @__PURE__ */ jsx(
        Text.formatted,
        {
          rule: "date",
          formatString: "yyyy-MM-dd HH:mm:ss",
          intensity: 36,
          size: "2xs",
          children: item.created_time
        }
      );
    };
    OrderState = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const status2 = item.status || item.algo_status;
      const statusMap = useMemo(() => {
        return {
          [OrderStatus.NEW]: t("orders.status.pending"),
          [OrderStatus.FILLED]: t("orders.status.filled"),
          [OrderStatus.PARTIAL_FILLED]: t("orders.status.partialFilled"),
          [OrderStatus.CANCELLED]: t("orders.status.canceled"),
          [OrderStatus.REJECTED]: t("orders.status.rejected"),
          [OrderStatus.INCOMPLETE]: t("orders.status.incomplete"),
          [OrderStatus.COMPLETED]: t("orders.status.completed")
        };
      }, [t]);
      return /* @__PURE__ */ jsx(Text.formatted, { intensity: 80, size: "2xs", children: statusMap[status2] || upperCaseFirstLetter(status2) });
    };
    Qty = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const isEntirePosition = item.type === OrderType.CLOSE_POSITION && // @ts-ignore
      item?.status !== OrderStatus.FILLED;
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("common.qty"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.base_dp,
              padding: false,
              coloring: true,
              placeholder: t("tpsl.entirePosition"),
              intensity: 80,
              children: isEntirePosition ? t("tpsl.entirePosition") : item.quantity
            }
          )
        }
      );
    };
    Filled = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: /* @__PURE__ */ jsx(Text, { children: t("orders.status.filled") }),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.base_dp,
              intensity: 80,
              padding: false,
              rm: Decimal.ROUND_DOWN,
              children: item.total_executed_quantity
            }
          )
        }
      );
    };
    EstTotal = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const value = useMemo(() => getNotional(item), [item.price, item.quantity]);
      const rendrerContent = () => {
        if (value) {
          return /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              coloring: true,
              intensity: 80,
              padding: false,
              rm: Decimal.ROUND_DOWN,
              children: value
            }
          );
        }
        return "--";
      };
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          align: "end",
          label: /* @__PURE__ */ jsxs(Text, { children: [
            t("common.notional"),
            /* @__PURE__ */ jsx(Text, { intensity: 20, children: "(USDC)" })
          ] }),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: rendrerContent()
        }
      );
    };
    TriggerPrice = (props) => {
      const { item, isPending } = props;
      const { t } = useTranslation();
      const isTrailingStop = item?.algo_type === AlgoOrderRootType.TRAILING_STOP;
      const rendrerContent = () => {
        if (isTrailingStop && isPending && item.activated_price) {
          return /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              intensity: 80,
              padding: false,
              rm: Decimal.ROUND_DOWN,
              children: item.activated_price
            }
          );
        }
        if (item.trigger_price) {
          return /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              intensity: 80,
              padding: false,
              rm: Decimal.ROUND_DOWN,
              children: item.trigger_price
            }
          );
        }
        return "--";
      };
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: isTrailingStop ? t("common.trigger") : t("common.triggerPrice"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          align: props.align,
          children: rendrerContent()
        }
      );
    };
    MarkPrice = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("common.markPrice"),
          align: "end",
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              rm: Decimal.ROUND_DOWN,
              intensity: 80,
              padding: false,
              children: item.mark_price
            }
          )
        }
      );
    };
    LimitPrice = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const isAlgoOrder = item?.algo_order_id !== void 0;
      const isTrailingStop = item?.algo_type === AlgoOrderRootType.TRAILING_STOP;
      const isStopMarket = item?.type === "MARKET" && isAlgoOrder;
      const rendrerContent = () => {
        if (isTrailingStop) {
          return "--";
        }
        if (isStopMarket) {
          return /* @__PURE__ */ jsx(Text, { children: t("common.marketPrice") });
        }
        return /* @__PURE__ */ jsx(
          Text.numeral,
          {
            dp: props.quote_dp,
            rm: Decimal.ROUND_DOWN,
            intensity: 80,
            padding: false,
            children: item.price ?? "--"
          }
        );
      };
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("common.limitPrice"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: rendrerContent()
        }
      );
    };
    TPTrigger = (props) => {
      const { tp_trigger_price, tpPnL } = useTPSLOrderRowContext();
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("tpsl.tpTrigger"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            MobileTooltip,
            {
              content: tpPnL && /* @__PURE__ */ jsx(
                Text.numeral,
                {
                  size: "2xs",
                  showIdentifier: true,
                  prefix: /* @__PURE__ */ jsxs(Text, { intensity: 54, children: [
                    `${t("tpsl.tpPnl")}:`,
                    "\xA0"
                  ] }),
                  suffix: /* @__PURE__ */ jsx(Text, { intensity: 20, children: "\xA0USDC" }),
                  coloring: true,
                  children: tpPnL
                }
              ),
              classNames: {
                content: "oui-bg-base-6 oui-ml-2",
                arrow: "oui-fill-base-6"
              },
              children: /* @__PURE__ */ jsx(
                Text.numeral,
                {
                  dp: props.quote_dp,
                  rm: Decimal.ROUND_DOWN,
                  color: "buy",
                  padding: false,
                  className: tp_trigger_price ? "oui-border-b oui-border-dashed oui-border-base-contrast-12" : void 0,
                  children: tp_trigger_price ?? "--"
                }
              )
            }
          )
        }
      );
    };
    SLTrigger = (props) => {
      const { sl_trigger_price, slPnL, position: item } = useTPSLOrderRowContext();
      const { t } = useTranslation();
      const side2 = (item?.position_qty ?? 0) > 0 ? OrderSide.BUY : OrderSide.SELL;
      const slPriceError = useTpslPriceChecker({
        slPrice: sl_trigger_price?.toString() ?? void 0,
        liqPrice: item?.est_liq_price ?? null,
        side: side2
      });
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("tpsl.slTrigger"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            MobileTooltip,
            {
              content: slPnL && /* @__PURE__ */ jsx(
                Text.numeral,
                {
                  size: "2xs",
                  prefix: /* @__PURE__ */ jsxs(Text, { intensity: 54, children: [
                    `${t("tpsl.slPnl")}:`,
                    "\xA0\xA0"
                  ] }),
                  suffix: /* @__PURE__ */ jsx(Text, { intensity: 20, children: "\xA0USDC" }),
                  coloring: true,
                  children: slPnL
                }
              ),
              classNames: {
                content: "oui-bg-base-6 oui-ml-2",
                arrow: "oui-fill-base-6"
              },
              children: /* @__PURE__ */ jsx(
                Text.numeral,
                {
                  dp: props.quote_dp,
                  as: "div",
                  rm: Decimal.ROUND_DOWN,
                  color: "sell",
                  padding: false,
                  suffix: /* @__PURE__ */ jsx(CloseToLiqPriceIcon, { slPriceError }),
                  className: cn(
                    sl_trigger_price ? "oui-border-b oui-border-dashed oui-border-base-contrast-12" : void 0,
                    "oui-flex oui-items-center"
                  ),
                  children: sl_trigger_price ?? "--"
                }
              )
            }
          )
        }
      );
    };
    TPPrice = (props) => {
      const { t } = useTranslation();
      const { tp_order_price } = useTPSLOrderRowContext();
      const renderOrderPrice = useMemo(() => {
        if (tp_order_price === void 0) {
          return /* @__PURE__ */ jsx(Text, { children: "--" });
        }
        if (tp_order_price === OrderType.MARKET) {
          return /* @__PURE__ */ jsx(Text, { children: t("common.marketPrice") });
        }
        return /* @__PURE__ */ jsx(
          Text.numeral,
          {
            dp: props.quote_dp,
            rm: Decimal.ROUND_DOWN,
            intensity: 80,
            padding: false,
            children: tp_order_price
          }
        );
      }, [tp_order_price]);
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("tpsl.tpPrice"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: renderOrderPrice
        }
      );
    };
    SLPrice = (props) => {
      const { t } = useTranslation();
      const { sl_order_price } = useTPSLOrderRowContext();
      const renderOrderPrice = useMemo(() => {
        if (sl_order_price === void 0) {
          return /* @__PURE__ */ jsx(Text, { children: "--" });
        }
        if (sl_order_price === OrderType.MARKET) {
          return /* @__PURE__ */ jsx(Text, { children: t("common.marketPrice") });
        }
        return /* @__PURE__ */ jsx(
          Text.numeral,
          {
            dp: props.quote_dp,
            rm: Decimal.ROUND_DOWN,
            intensity: 80,
            padding: false,
            children: sl_order_price
          }
        );
      }, [sl_order_price]);
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("tpsl.slPrice"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: renderOrderPrice
        }
      );
    };
    TPSLQuantity = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const quantity2 = useMemo(() => {
        if (item.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL) {
          return /* @__PURE__ */ jsx("span", { className: "oui-text-base-contrast-80", children: t("tpsl.entirePosition") });
        }
        return /* @__PURE__ */ jsx(
          Text.numeral,
          {
            dp: props.base_dp,
            rm: Decimal.ROUND_DOWN,
            intensity: 80,
            padding: false,
            children: item.quantity
          }
        );
      }, [item, t]);
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("common.quantity"),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          align: "end",
          children: quantity2
        }
      );
    };
    AvgPrice2 = (props) => {
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: /* @__PURE__ */ jsxs(Text, { children: [
            t("common.avgPrice"),
            /* @__PURE__ */ jsx(Text, { intensity: 20, children: "(USDC)" })
          ] }),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              rm: Decimal.ROUND_DOWN,
              intensity: 80,
              padding: false,
              children: props.item?.average_executed_price ?? "--"
            }
          )
        }
      );
    };
    OrderPrice = (props) => {
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: /* @__PURE__ */ jsxs(Text, { children: [
            t("common.orderPrice"),
            /* @__PURE__ */ jsx(Text, { intensity: 20, children: "(USDC)" })
          ] }),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              rm: Decimal.ROUND_DOWN,
              intensity: 80,
              padding: false,
              placeholder: t("common.marketPrice"),
              children: props.item?.price ?? "--"
            }
          )
        }
      );
    };
    RealizedPnL = (props) => {
      const value = props?.item?.realized_pnl;
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: /* @__PURE__ */ jsxs(Text, { children: [
            t("common.realizedPnl"),
            /* @__PURE__ */ jsx(Text, { intensity: 20, children: "(USDC)" })
          ] }),
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          align: "end",
          children: /* @__PURE__ */ jsx(Flex, { gap: 1, children: /* @__PURE__ */ jsx(
            Text.pnl,
            {
              dp: props.quote_dp,
              padding: false,
              intensity: (value ?? 0) == 0 ? 80 : void 0,
              showIdentifier: (value ?? 0) > 0,
              coloring: (value ?? 0) != 0,
              children: value ?? "--"
            }
          ) })
        }
      );
    };
    MobileTooltip = (props) => {
      const { classNames, content } = props;
      const [open, setOpen] = useState(false);
      if (typeof content === "undefined") return props.children;
      return /* @__PURE__ */ jsx(
        Tooltip,
        {
          content,
          className: classNames?.content,
          open,
          onOpenChange: setOpen,
          arrow: { className: classNames?.arrow },
          children: /* @__PURE__ */ jsx("div", { onClick: () => setOpen((e) => !e), children: props.children })
        }
      );
    };
    Fee = (props) => {
      const { t } = useTranslation();
      return /* @__PURE__ */ jsxs(Flex, { justify: "end", gapX: 1, mt: 1, children: [
        /* @__PURE__ */ jsxs(Text, { size: "2xs", intensity: 36, children: [
          t("common.fee"),
          ":"
        ] }),
        /* @__PURE__ */ jsx(Text, { size: "2xs", intensity: 80, children: props.item?.total_fee ?? "--" })
      ] });
    };
    TrailingCallback = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const { callback_value, callback_rate } = item;
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("orderEntry.trailing"),
          align: "end",
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: callback_rate ? /* @__PURE__ */ jsx(Text, { intensity: 80, children: `${callback_rate * 100}%` }) : /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              rm: Decimal.ROUND_DOWN,
              intensity: 80,
              padding: false,
              children: callback_value ?? "--"
            }
          )
        }
      );
    };
    TrailingPrice = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      const value = useMemo(() => {
        return getTrailingStopPrice(item);
      }, [item]);
      return /* @__PURE__ */ jsx(
        Statistic,
        {
          label: t("common.price"),
          align: "start",
          classNames: {
            root: "oui-text-xs",
            label: "oui-text-2xs"
          },
          children: value ? /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: props.quote_dp,
              rm: Decimal.ROUND_DOWN,
              intensity: 80,
              padding: false,
              children: value
            }
          ) : "--"
        }
      );
    };
  }
});
var BracketOrderPrice2, Price2;
var init_bracketOrderPrice_ui = __esm({
  "src/components/orderList/mobile/bracketOrderPrice/bracketOrderPrice.ui.tsx"() {
    init_items();
    BracketOrderPrice2 = (props) => {
      const { t } = useTranslation();
      const onEdit = () => {
        modal.show("EditBracketOrderSheetId", {
          order: props.order
        });
      };
      if (!props.sl_trigger_price && !props.tp_trigger_price) return /* @__PURE__ */ jsx(Fragment, {});
      return /* @__PURE__ */ jsxs(Flex, { gap: 1, itemAlign: "center", justify: "start", children: [
        /* @__PURE__ */ jsx(
          MobileTooltip,
          {
            classNames: {
              content: "oui-bg-base-6 oui-ml-2",
              arrow: "oui-fill-base-6"
            },
            content: /* @__PURE__ */ jsxs(Flex, { direction: "column", itemAlign: "start", gap: 1, children: [
              typeof props.pnl?.tpPnL !== "undefined" && /* @__PURE__ */ jsx(
                Text.numeral,
                {
                  prefix: /* @__PURE__ */ jsxs(Text, { intensity: 80, children: [
                    `${t("tpsl.tpPnl")}:`,
                    " \xA0"
                  ] }),
                  suffix: /* @__PURE__ */ jsx(Text, { intensity: 20, children: " USDC" }),
                  dp: props.quote_dp,
                  color: "buy",
                  showIdentifier: true,
                  children: props.pnl?.tpPnL
                }
              ),
              typeof props.pnl?.slPnL !== "undefined" && /* @__PURE__ */ jsx(
                Text.numeral,
                {
                  prefix: /* @__PURE__ */ jsxs(Text, { intensity: 80, children: [
                    `${t("tpsl.slPnl")}:`,
                    " \xA0"
                  ] }),
                  suffix: /* @__PURE__ */ jsx(Text, { intensity: 20, children: " USDC" }),
                  dp: props.quote_dp,
                  color: "sell",
                  children: props.pnl?.slPnL
                }
              )
            ] }),
            children: /* @__PURE__ */ jsx(
              "button",
              {
                onClick: () => {
                  props.setOpen(!props.open);
                },
                children: /* @__PURE__ */ jsxs(Flex, { gap: 1, width: "1", children: [
                  props.tp_trigger_price && /* @__PURE__ */ jsx(
                    Price2,
                    {
                      type: "TP",
                      value: props.tp_trigger_price,
                      quote_dp: props.quote_dp
                    }
                  ),
                  props.sl_trigger_price && /* @__PURE__ */ jsx(
                    Price2,
                    {
                      type: "SL",
                      value: props.sl_trigger_price,
                      quote_dp: props.quote_dp
                    }
                  )
                ] })
              }
            )
          }
        ),
        /* @__PURE__ */ jsx(
          EditIcon,
          {
            onClick: onEdit,
            opacity: 1,
            className: "oui-cursor-pointer oui-text-base-contrast-54",
            size: 16
          }
        )
      ] });
    };
    Price2 = (props) => {
      const { type, value, quote_dp } = props;
      return value ? /* @__PURE__ */ jsx(
        Text.numeral,
        {
          size: "2xs",
          className: cn(
            "oui-border-b oui-border-dashed oui-border-base-contrast-12",
            type === "TP" ? "oui-text-trade-profit" : "oui-text-trade-loss"
          ),
          rule: "price",
          dp: quote_dp,
          padding: false,
          rm: Decimal.ROUND_DOWN,
          prefix: /* @__PURE__ */ jsxs("span", { className: "oui-text-base-contrast-36", children: [
            `${type} `,
            ":\xA0\xA0"
          ] }),
          children: value
        },
        "tp"
      ) : /* @__PURE__ */ jsx(Fragment, {});
    };
  }
});
var useBracketOrderPriceScript;
var init_bracketOrderPrice_script = __esm({
  "src/components/orderList/mobile/bracketOrderPrice/bracketOrderPrice.script.tsx"() {
    init_util();
    useBracketOrderPriceScript = (props) => {
      const { item: order, quote_dp } = props;
      const [open, setOpen] = useState(false);
      const { sl_trigger_price, tp_trigger_price } = useMemo(() => {
        if (!("algo_type" in order) || !Array.isArray(order.child_orders)) {
          return {};
        }
        return utils.findTPSLFromOrder(order.child_orders[0]);
      }, [order]);
      const { pnl } = calcBracketRoiAndPnL(order);
      return {
        sl_trigger_price,
        order,
        tp_trigger_price,
        pnl,
        quote_dp,
        open,
        setOpen
      };
    };
  }
});
var BracketOrderPriceWidget;
var init_bracketOrderPrice_widget = __esm({
  "src/components/orderList/mobile/bracketOrderPrice/bracketOrderPrice.widget.tsx"() {
    init_bracketOrderPrice_script();
    init_bracketOrderPrice_ui();
    BracketOrderPriceWidget = (props) => {
      const state = useBracketOrderPriceScript(props);
      return /* @__PURE__ */ jsx(BracketOrderPrice2, { ...state });
    };
  }
});

// src/components/orderList/mobile/bracketOrderPrice/index.ts
var init_bracketOrderPrice2 = __esm({
  "src/components/orderList/mobile/bracketOrderPrice/index.ts"() {
    init_bracketOrderPrice_ui();
    init_bracketOrderPrice_widget();
    init_bracketOrderPrice_script();
  }
});
var CancelBtn;
var init_cancelBtn_ui = __esm({
  "src/components/orderList/mobile/cancelBtn/cancelBtn.ui.tsx"() {
    CancelBtn = (props) => {
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        Button,
        {
          variant: "outlined",
          fullWidth: true,
          color: "secondary",
          size: "sm",
          className: "oui-border-base-contrast-36",
          onClick: (e) => props.onCancel(e),
          children: t("common.cancel")
        }
      );
    };
  }
});
var useCancelBtnScript;
var init_cancelBtn_script = __esm({
  "src/components/orderList/mobile/cancelBtn/cancelBtn.script.tsx"() {
    init_orderListContext();
    useCancelBtnScript = (props) => {
      const { state } = props;
      const [open, setOpen] = useState(false);
      const { onCancelOrder } = useOrderListContext();
      const [isLoading, setIsLoading] = useState(false);
      const onCancel = (event) => {
        if (!onCancelOrder) return Promise.resolve();
        event?.preventDefault();
        event?.stopPropagation();
        setIsLoading(true);
        return onCancelOrder(state.item).then(
          (res) => {
            setOpen(false);
          },
          (error) => {
            toast$1.error(error.message);
          }
        ).finally(() => {
          setIsLoading(false);
        });
      };
      const onClose = () => {
        setOpen(false);
      };
      return {
        ...state,
        open,
        setOpen,
        onCancel,
        onClose,
        isLoading
      };
    };
  }
});
var CancelBtnWidget;
var init_cancelBtn_widget = __esm({
  "src/components/orderList/mobile/cancelBtn/cancelBtn.widget.tsx"() {
    init_cancelBtn_script();
    init_cancelBtn_ui();
    CancelBtnWidget = (props) => {
      const state = useCancelBtnScript(props);
      return /* @__PURE__ */ jsx(CancelBtn, { ...state });
    };
  }
});

// src/components/orderList/mobile/cancelBtn/index.ts
var init_cancelBtn2 = __esm({
  "src/components/orderList/mobile/cancelBtn/index.ts"() {
    init_cancelBtn_ui();
    init_cancelBtn_widget();
    init_cancelBtn_script();
  }
});
var EditBtn;
var init_editBtn_ui = __esm({
  "src/components/orderList/mobile/editBtn/editBtn.ui.tsx"() {
    EditBtn = (props) => {
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(Fragment, { children: /* @__PURE__ */ jsx(
        Button,
        {
          variant: "outlined",
          fullWidth: true,
          color: "secondary",
          size: "sm",
          className: "oui-border-base-contrast-36",
          onClick: () => {
            props.onShowEditSheet();
          },
          children: t("common.edit")
        }
      ) });
    };
  }
});
function useEditSheetContext() {
  return useContext(EditSheetContext);
}
var EditSheetContext;
var init_editSheetContext = __esm({
  "src/components/orderList/mobile/editSheet/components/editSheetContext.ts"() {
    EditSheetContext = createContext({});
  }
});
var ActivitedPriceInput;
var init_activitedPriceInput = __esm({
  "src/components/orderList/mobile/editSheet/components/activitedPriceInput.tsx"() {
    init_editSheetContext();
    ActivitedPriceInput = memo((props) => {
      const { activated_price } = props;
      const { t } = useTranslation();
      const { symbolInfo, setOrderValue, getErrorMsg } = useEditSheetContext();
      const { quote, quote_dp } = symbolInfo;
      const error = activated_price ? getErrorMsg("activated_price") : t("orderEntry.triggerPrice.error.required");
      return /* @__PURE__ */ jsx(
        Input.tooltip,
        {
          prefix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: t("common.trigger") }),
          suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: quote }),
          color: error ? "danger" : void 0,
          align: "right",
          fullWidth: true,
          autoComplete: "off",
          formatters: [
            inputFormatter.numberFormatter,
            inputFormatter.dpFormatter(quote_dp)
          ],
          value: activated_price,
          onValueChange: (val) => setOrderValue("activated_price", val),
          disabled: props.disabled,
          tooltip: error,
          tooltipProps: {
            content: {
              className: "oui-bg-base-5 oui-max-w-[360px]"
            },
            arrow: {
              className: "oui-fill-base-5"
            }
          },
          classNames: {
            input: "oui-text-base-contrast",
            root: cn("oui-outline-line-12", error && "oui-outline-danger")
          }
        }
      );
    });
    ActivitedPriceInput.displayName = "ActivitedPriceInput";
  }
});
var ConfirmDialogContent;
var init_editDialogContent = __esm({
  "src/components/orderList/mobile/editSheet/components/editDialogContent.tsx"() {
    init_util();
    ConfirmDialogContent = (props) => {
      const { item } = props;
      const {
        formattedOrder,
        symbolInfo,
        isStopMarket,
        showTriggerPrice,
        isTrailingStop
      } = props;
      const { t } = useTranslation();
      const showPrice = !isTrailingStop;
      const showActivatedPrice = isTrailingStop && formattedOrder.activated_price;
      const showTrailingCallback = isTrailingStop;
      const header = /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(
          Text,
          {
            intensity: 80,
            children: `You agree to edit your ${props.base}-PERP order.`
          }
        ),
        /* @__PURE__ */ jsxs(Flex, { gap: 2, mb: 3, mt: 2, justify: "between", children: [
          /* @__PURE__ */ jsx(
            Text.formatted,
            {
              rule: "symbol",
              formatString: "base-type",
              size: "base",
              showIcon: true,
              children: item.symbol
            }
          ),
          /* @__PURE__ */ jsxs(Flex, { direction: "row", gap: 1, children: [
            parseBadgesFor(item)?.map((e, index) => /* @__PURE__ */ jsx(
              Badge,
              {
                color: e.toLocaleLowerCase() === "position" ? "primary" : "neutral",
                size: "xs",
                children: e
              },
              index
            )),
            item.side === OrderSide.BUY ? /* @__PURE__ */ jsx(Badge, { color: "success", size: "xs", children: t("common.buy") }) : /* @__PURE__ */ jsx(Badge, { color: "danger", size: "xs", children: t("common.sell") })
          ] })
        ] })
      ] });
      const triggerPriceItem = showTriggerPrice && /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", gap: 1, children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.triggerPrice") }),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            intensity: 80,
            dp: symbolInfo.quote_dp,
            padding: false,
            rm: Decimal.ROUND_DOWN,
            suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, children: " USDC" }),
            children: formattedOrder.trigger_price ?? "--"
          }
        )
      ] });
      const priceItem = showPrice && /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", gap: 1, children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.price") }),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            intensity: 80,
            dp: symbolInfo.quote_dp,
            padding: false,
            rm: Decimal.ROUND_DOWN,
            suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, children: " USDC" }),
            placeholder: isStopMarket ? t("common.marketPrice") : "--",
            children: isStopMarket ? t("common.marketPrice") : formattedOrder.order_price ?? "--"
          }
        )
      ] });
      const activatedPriceItem = showActivatedPrice && /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", gap: 1, children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.triggerPrice") }),
        formattedOrder.activated_price ? /* @__PURE__ */ jsx(
          Text.numeral,
          {
            intensity: 80,
            dp: symbolInfo.quote_dp,
            padding: false,
            rm: Decimal.ROUND_DOWN,
            suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, children: " USDC" }),
            children: formattedOrder.activated_price
          }
        ) : t("common.marketPrice")
      ] });
      const trailingCallbackItem = showTrailingCallback && /* @__PURE__ */ jsxs(Fragment, { children: [
        formattedOrder.callback_value && /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", gap: 1, children: [
          /* @__PURE__ */ jsx(Text, { children: t("orderEntry.trailingValue") }),
          /* @__PURE__ */ jsx(
            Text.numeral,
            {
              dp: symbolInfo.quote_dp,
              padding: false,
              rm: Decimal.ROUND_DOWN,
              children: formattedOrder.callback_value
            }
          )
        ] }),
        formattedOrder.callback_rate && /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", gap: 1, children: [
          /* @__PURE__ */ jsx(Text, { children: t("orderEntry.trailingRate") }),
          /* @__PURE__ */ jsxs(Text, { className: "oui-text-base-contrast", children: [
            formattedOrder.callback_rate,
            "%"
          ] })
        ] })
      ] });
      const quantityItem = /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", gap: 1, children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.qty") }),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            color: item.side === OrderSide.BUY ? "buy" : "sell",
            dp: symbolInfo.base_dp,
            padding: false,
            rm: Decimal.ROUND_DOWN,
            children: formattedOrder.order_quantity
          }
        )
      ] });
      const orderConfirmCheckbox = /* @__PURE__ */ jsxs(Flex, { className: "oui-gap-[2px]", children: [
        /* @__PURE__ */ jsx(
          Checkbox,
          {
            color: "white",
            id: "oui-checkbox-disableOrderConfirmation",
            checked: !props.orderConfirm,
            onCheckedChange: (e) => {
              props.setOrderConfirm(!e);
            }
          }
        ),
        /* @__PURE__ */ jsx(
          "label",
          {
            className: "oui-text-2xs oui-text-base-contrast-54",
            htmlFor: "oui-checkbox-disableOrderConfirmation",
            children: t("orderEntry.disableOrderConfirm")
          }
        )
      ] });
      return /* @__PURE__ */ jsxs("div", { className: "oui-pt-2", children: [
        header,
        /* @__PURE__ */ jsx(Divider, {}),
        /* @__PURE__ */ jsxs(
          Flex,
          {
            direction: "column",
            gap: 1,
            width: "100%",
            className: "oui-text-sm oui-text-base-contrast-54",
            py: 3,
            children: [
              triggerPriceItem,
              priceItem,
              quantityItem,
              activatedPriceItem,
              trailingCallbackItem
            ]
          }
        ),
        orderConfirmCheckbox
      ] });
    };
  }
});
var EditSheetHeader;
var init_editSheetHeader = __esm({
  "src/components/orderList/mobile/editSheet/components/editSheetHeader.tsx"() {
    init_util();
    EditSheetHeader = (props) => {
      const { item } = props;
      const { t } = useTranslation();
      return /* @__PURE__ */ jsxs(Flex, { width: "100%", justify: "between", children: [
        /* @__PURE__ */ jsx(Text.formatted, { rule: "symbol", showIcon: true, intensity: 80, children: item.symbol }),
        /* @__PURE__ */ jsxs(Flex, { direction: "row", gap: 1, children: [
          parseBadgesFor(props.item)?.map((e, index) => /* @__PURE__ */ jsx(
            Badge,
            {
              color: e.toLocaleLowerCase() === "position" ? "primary" : "neutral",
              size: "xs",
              children: e
            },
            index
          )),
          item.side === OrderSide.BUY ? /* @__PURE__ */ jsx(Badge, { color: "success", size: "xs", children: t("common.buy") }) : /* @__PURE__ */ jsx(Badge, { color: "danger", size: "xs", children: t("common.sell") })
        ] })
      ] });
    };
  }
});
var PriceInput;
var init_priceInput = __esm({
  "src/components/orderList/mobile/editSheet/components/priceInput.tsx"() {
    init_editSheetContext();
    PriceInput = memo((props) => {
      const [focus, setFocus] = useState(false);
      const { t } = useTranslation();
      const { symbolInfo, setOrderValue, getErrorMsg } = useEditSheetContext();
      const { quote, quote_dp } = symbolInfo;
      const error = getErrorMsg("order_price") || (focus ? getErrorMsg("total") : "");
      return /* @__PURE__ */ jsx(
        Input.tooltip,
        {
          prefix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: t("common.price") }),
          suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: quote }),
          color: error ? "danger" : void 0,
          align: "right",
          fullWidth: true,
          autoComplete: "off",
          formatters: [
            inputFormatter.numberFormatter,
            inputFormatter.dpFormatter(quote_dp)
          ],
          disabled: props.disabled,
          value: props.order_price,
          onValueChange: (val) => setOrderValue("order_price", val),
          onBlur: () => {
            setFocus(false);
          },
          onFocus: () => {
            setFocus(true);
          },
          tooltip: error,
          tooltipProps: {
            content: {
              className: "oui-bg-base-5"
            },
            arrow: {
              className: "oui-fill-base-5"
            }
          },
          classNames: {
            input: "oui-text-base-contrast",
            root: cn("oui-outline-line-12", error && "oui-outline-danger")
          }
        }
      );
    });
    PriceInput.displayName = "PriceInput";
  }
});
var QuantityInput;
var init_quantityInput = __esm({
  "src/components/orderList/mobile/editSheet/components/quantityInput.tsx"() {
    init_editSheetContext();
    QuantityInput = memo((props) => {
      const [focus, setFocus] = useState(false);
      const { t } = useTranslation();
      const { symbolInfo, setOrderValue, getErrorMsg } = useEditSheetContext();
      const { base_dp, base_tick, base } = symbolInfo;
      const error = getErrorMsg("order_quantity") || (focus ? getErrorMsg("total") : "");
      const setQuantity = (val) => {
        setOrderValue("order_quantity", val);
      };
      const onBlur = (value) => {
        if (base_tick && base_tick > 0) {
          const formatQty = utils.formatNumber(value, base_tick) ?? value;
          setQuantity(formatQty);
        }
      };
      return /* @__PURE__ */ jsx(
        Input.tooltip,
        {
          prefix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: t("common.quantity") }),
          suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: base }),
          color: error ? "danger" : void 0,
          align: "right",
          fullWidth: true,
          autoComplete: "off",
          formatters: [
            inputFormatter.numberFormatter,
            inputFormatter.dpFormatter(base_dp)
            // inputFormatter.rangeFormatter({ max: props.maxQty }),
          ],
          value: props.value,
          onValueChange: setQuantity,
          onBlur: (event) => {
            onBlur(event.target.value);
            setFocus(false);
          },
          onFocus: () => {
            setFocus(true);
          },
          tooltip: error,
          tooltipProps: {
            content: {
              className: "oui-bg-base-6"
            },
            arrow: {
              className: "oui-fill-base-6"
            }
          },
          classNames: {
            input: "oui-text-base-contrast",
            root: cn("oui-outline-line-12", error && "oui-outline-danger")
          }
        }
      );
    });
    QuantityInput.displayName = "QuantityInput";
  }
});
var QuantitySlider;
var init_quantitySlider = __esm({
  "src/components/orderList/mobile/editSheet/components/quantitySlider.tsx"() {
    init_editSheetContext();
    QuantitySlider = memo((props) => {
      const { quantity: quantity2, maxQty } = props;
      const { t } = useTranslation();
      const { symbolInfo, setOrderValue } = useEditSheetContext();
      const { base_dp, base_tick } = symbolInfo;
      const sliderValue = useMemo(() => {
        if (quantity2 && Number(quantity2) !== 0 && maxQty !== 0) {
          const value = new Decimal(quantity2).div(maxQty).mul(100).toDecimalPlaces(2, Decimal.ROUND_DOWN).toNumber();
          return value;
        }
        return 0;
      }, [quantity2, maxQty]);
      const sliderToQuantity = useThrottledCallback(
        (value) => {
          const newQty = new Decimal(value).div(100).mul(maxQty).toDecimalPlaces(base_dp, Decimal.ROUND_DOWN).toNumber();
          setOrderValue("order_quantity", utils.formatNumber(newQty, base_tick));
        },
        50,
        {}
      );
      const percentages = props.quantity && props.maxQty ? Math.min(Number(props.quantity) / props.maxQty, 1) : void 0;
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(
          Slider,
          {
            markCount: 4,
            value: [sliderValue],
            onValueChange: (e) => {
              sliderToQuantity(e[0]);
            },
            color: "primary"
          }
        ),
        /* @__PURE__ */ jsxs(Flex, { width: "100%", justify: "between", children: [
          /* @__PURE__ */ jsx(
            Text.numeral,
            {
              color: "primary",
              size: "2xs",
              dp: 2,
              padding: false,
              rule: "percentages",
              children: `${percentages ?? 0}`
            }
          ),
          /* @__PURE__ */ jsxs(
            Flex,
            {
              gap: 1,
              onClick: () => {
                setOrderValue("order_quantity", maxQty.toString());
              },
              children: [
                /* @__PURE__ */ jsx(Text, { size: "2xs", color: "primary", children: t("common.max") }),
                /* @__PURE__ */ jsx(Text.numeral, { intensity: 54, size: "2xs", dp: base_dp, children: props.maxQty })
              ]
            }
          )
        ] })
      ] });
    });
    QuantitySlider.displayName = "QuantitySlider";
  }
});
var TrailingCallbackInput;
var init_trailingCallbackInput = __esm({
  "src/components/orderList/mobile/editSheet/components/trailingCallbackInput.tsx"() {
    init_editSheetContext();
    TrailingCallbackInput = memo(
      (props) => {
        const { isCallbackValue, callback_value, callback_rate } = props;
        const { t } = useTranslation();
        const { symbolInfo, setOrderValue, getErrorMsg } = useEditSheetContext();
        const { quote, quote_dp } = symbolInfo;
        const label = isCallbackValue ? t("orderEntry.trailingValue") : t("orderEntry.trailingRate");
        const field = isCallbackValue ? "callback_value" : "callback_rate";
        const suffix = isCallbackValue ? /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: quote }) : "%";
        const error = getErrorMsg(field);
        const formatters = [
          inputFormatter.numberFormatter,
          inputFormatter.decimalPointFormatter
        ];
        return /* @__PURE__ */ jsx(
          Input.tooltip,
          {
            prefix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: label }),
            suffix,
            color: error ? "danger" : void 0,
            align: "right",
            fullWidth: true,
            autoComplete: "off",
            formatters: isCallbackValue ? [
              ...formatters,
              inputFormatter.currencyFormatter,
              inputFormatter.dpFormatter(quote_dp)
            ] : [...formatters, inputFormatter.dpFormatter(1)],
            value: isCallbackValue ? callback_value : callback_rate,
            onValueChange: (val) => setOrderValue(field, val),
            tooltip: error,
            tooltipProps: {
              content: {
                // mobile max width
                className: "oui-bg-base-5 oui-max-w-[360px]"
              },
              arrow: {
                className: "oui-fill-base-5"
              }
            },
            classNames: {
              input: "oui-text-base-contrast",
              root: cn("oui-outline-line-12", error && "oui-outline-danger")
            }
          }
        );
      }
    );
    TrailingCallbackInput.displayName = "TrailingCallbackInput";
  }
});
var TriggerPriceInput;
var init_triggerPriceInput = __esm({
  "src/components/orderList/mobile/editSheet/components/triggerPriceInput.tsx"() {
    init_editSheetContext();
    TriggerPriceInput = memo((props) => {
      const { t } = useTranslation();
      const { symbolInfo, setOrderValue, getErrorMsg } = useEditSheetContext();
      const { quote, quote_dp } = symbolInfo;
      const error = getErrorMsg("trigger_price");
      return /* @__PURE__ */ jsx(
        Input.tooltip,
        {
          prefix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: t("common.triggerPrice") }),
          suffix: /* @__PURE__ */ jsx(Text, { intensity: 54, className: "oui-px-3", children: quote }),
          color: error ? "danger" : void 0,
          align: "right",
          fullWidth: true,
          autoComplete: "off",
          formatters: [
            inputFormatter.numberFormatter,
            inputFormatter.dpFormatter(quote_dp)
          ],
          value: props.trigger_price,
          onValueChange: (val) => setOrderValue("trigger_price", val),
          disabled: props.disabled,
          tooltip: error,
          tooltipProps: {
            content: {
              className: "oui-bg-base-6 oui-text-base-contrast-80"
            },
            arrow: {
              className: "oui-fill-base-6"
            }
          },
          classNames: {
            input: "oui-text-base-contrast oui-w-full",
            root: cn("oui-outline-line-12", error && "oui-outline-danger")
          }
        }
      );
    });
    TriggerPriceInput.displayName = "TriggerPriceInput";
  }
});
var EditSheet;
var init_editSheet_ui = __esm({
  "src/components/orderList/mobile/editSheet/editSheet.ui.tsx"() {
    init_activitedPriceInput();
    init_editDialogContent();
    init_editSheetContext();
    init_editSheetHeader();
    init_priceInput();
    init_quantityInput();
    init_quantitySlider();
    init_trailingCallbackInput();
    init_triggerPriceInput();
    EditSheet = (props) => {
      const {
        item,
        symbolInfo,
        formattedOrder,
        isStopMarket,
        isTrailingStop,
        setOrderValue,
        errors
      } = props;
      const { t } = useTranslation();
      const { getErrorMsg } = useOrderEntryFormErrorMsg(errors);
      const disabled = !props.isChanged || Object.keys(errors).length > 0 || // when activated price is not empty, the activated price input should not be empty
      !!item.activated_price && !formattedOrder.activated_price;
      const header = /* @__PURE__ */ jsx(EditSheetHeader, { item });
      const lastPrice = /* @__PURE__ */ jsxs(Flex, { width: "100%", justify: "between", children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.lastPrice") }),
        /* @__PURE__ */ jsx(Text.numeral, { dp: symbolInfo.quote_dp, children: props.markPrice ?? "--" })
      ] });
      const triggerPriceInput = props.showTriggerPrice && /* @__PURE__ */ jsx(
        TriggerPriceInput,
        {
          trigger_price: formattedOrder.trigger_price,
          disabled: !!item.is_triggered
        }
      );
      const renderPriceInput = () => {
        if (isTrailingStop) {
          return item.activated_price && /* @__PURE__ */ jsx(
            ActivitedPriceInput,
            {
              activated_price: formattedOrder.activated_price,
              disabled: item.is_activated
            }
          );
        }
        return /* @__PURE__ */ jsx(
          PriceInput,
          {
            order_price: isStopMarket ? t("orderEntry.orderType.market") : formattedOrder.order_price,
            disabled: isStopMarket
          }
        );
      };
      const trailingCallbackInput = isTrailingStop && /* @__PURE__ */ jsx(
        TrailingCallbackInput,
        {
          isCallbackValue: !!item.callback_value,
          callback_value: formattedOrder.callback_value,
          callback_rate: formattedOrder.callback_rate
        }
      );
      const quantityInput = /* @__PURE__ */ jsx(QuantityInput, { value: formattedOrder.order_quantity });
      const quantitySlider = /* @__PURE__ */ jsx(
        QuantitySlider,
        {
          maxQty: props.maxQty,
          quantity: formattedOrder.order_quantity
        }
      );
      const buttons = /* @__PURE__ */ jsxs(Flex, { width: "100%", gap: 3, mt: 2, children: [
        /* @__PURE__ */ jsx(
          Button,
          {
            fullWidth: true,
            color: "secondary",
            onClick: (e) => {
              props.onClose();
            },
            children: t("common.cancel")
          }
        ),
        /* @__PURE__ */ jsx(
          ThrottledButton,
          {
            fullWidth: true,
            onClick: (e) => {
              e.stopPropagation();
              e.preventDefault();
              props.onSheetConfirm();
            },
            loading: props.submitting,
            disabled,
            children: t("common.confirm")
          }
        )
      ] });
      const ctxValue = useMemo(() => {
        return {
          symbolInfo,
          getErrorMsg,
          setOrderValue
        };
      }, [symbolInfo, getErrorMsg, setOrderValue]);
      return /* @__PURE__ */ jsxs(EditSheetContext.Provider, { value: ctxValue, children: [
        /* @__PURE__ */ jsxs(
          Flex,
          {
            direction: "column",
            gap: 3,
            width: "100%",
            itemAlign: "start",
            className: "oui-text-sm",
            children: [
              header,
              /* @__PURE__ */ jsx(Divider, { intensity: 8, className: "oui-w-full" }),
              lastPrice,
              /* @__PURE__ */ jsxs(Flex, { width: "100%", direction: "column", itemAlign: "stretch", gap: 2, children: [
                triggerPriceInput,
                renderPriceInput(),
                trailingCallbackInput,
                quantityInput,
                quantitySlider
              ] }),
              buttons
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          SimpleDialog,
          {
            open: props.dialogOpen,
            onOpenChange: props.setDialogOpen,
            title: t("orders.editOrder"),
            size: "xs",
            actions: {
              primary: {
                label: t("common.confirm"),
                onClick: props.onDialogConfirm,
                loading: props.submitting,
                fullWidth: true
              },
              secondary: {
                label: t("common.cancel"),
                onClick: props.onCloseDialog,
                fullWidth: true
              }
            },
            classNames: {
              content: "oui-pb-4",
              body: "oui-p-0",
              footer: "oui-pt-3 oui-pb-0"
            },
            children: /* @__PURE__ */ jsx(ConfirmDialogContent, { ...props })
          }
        )
      ] });
    };
  }
});
var useEditOrderEntry;
var init_useEditOrderEntry = __esm({
  "src/components/orderList/mobile/editSheet/hooks/useEditOrderEntry.ts"() {
    useEditOrderEntry = (props) => {
      const { order, orderType, maxQty } = props;
      const [formattedOrder, setFormattedOrder] = useState({
        symbol: order.symbol,
        side: order.side,
        reduce_only: order.reduce_only,
        order_type: orderType,
        // TODO: trailing stop order edit price twice, order.price will be 0
        order_price: order.price || void 0,
        order_quantity: order.quantity,
        trigger_price: order.trigger_price,
        activated_price: order.activated_price,
        callback_value: order.callback_value,
        callback_rate: order.callback_rate ? order.callback_rate * 100 : void 0
      });
      const { markPrice, errors, validate, clearErrors } = useOrderEntity(
        formattedOrder,
        {
          maxQty
        }
      );
      const setOrderValue = useCallback((key, value) => {
        setFormattedOrder((oldValue) => ({
          ...oldValue,
          [key]: value
        }));
      }, []);
      const {
        order_price,
        order_quantity,
        trigger_price,
        activated_price,
        callback_value,
        callback_rate
      } = formattedOrder;
      const isChanged = order.price && order.price != order_price || order.quantity && order.quantity != order_quantity || order.trigger_price && order.trigger_price != trigger_price || // trailing stop fields
      order.activated_price && order.activated_price != activated_price || order.callback_value && order.callback_value != callback_value || order.callback_rate && order.callback_rate * 100 != callback_rate;
      useEffect(() => {
        if (isChanged) {
          validate().then((order2) => {
          }).catch((err) => {
          });
        } else {
          clearErrors();
        }
      }, [
        isChanged,
        order_price,
        order_quantity,
        trigger_price,
        activated_price,
        callback_value,
        callback_rate
      ]);
      return {
        markPrice,
        errors,
        validate,
        setOrderValue,
        formattedOrder,
        isChanged
      };
    };
  }
});
function useEditOrderMaxQty(order, positionQty) {
  const { reduce_only } = order;
  const maxQty = useMaxQty(order.symbol, order.side, reduce_only);
  return useMemo(() => {
    if (reduce_only) {
      return Math.abs(positionQty ?? 0);
    }
    return order.quantity + Math.abs(maxQty);
  }, [order.quantity, maxQty, reduce_only, positionQty]);
}
var init_useEditOrderMaxQty = __esm({
  "src/components/orderList/mobile/editSheet/hooks/useEditOrderMaxQty.ts"() {
  }
});
var useEditSheetScript;
var init_editSheet_script = __esm({
  "src/components/orderList/mobile/editSheet/editSheet.script.tsx"() {
    init_util();
    init_useEditOrderEntry();
    init_useEditOrderMaxQty();
    useEditSheetScript = (props) => {
      const { state, editAlgoOrder, editOrder, position } = props;
      const { item: order } = state;
      const { hide: onClose } = useModal();
      const [dialogOpen, setDialogOpen] = useState(false);
      const [submitting, setSubmitting] = useState(false);
      const isAlgoOrder = order.algo_order_id !== void 0 && order.algo_type !== AlgoOrderRootType.BRACKET;
      const isStopMarket = order?.type === "MARKET" && isAlgoOrder;
      const isTrailingStop = order.algo_type === OrderType.TRAILING_STOP;
      const showTriggerPrice = isAlgoOrder && !isTrailingStop;
      const orderType = useMemo(
        () => convertApiOrderTypeToOrderEntryType(order),
        [order]
      );
      const [orderConfirm, setOrderConfirm] = useLocalStorage(
        "orderly_order_confirm",
        true
      );
      const maxQty = useEditOrderMaxQty(order, position?.position_qty);
      const {
        formattedOrder,
        setOrderValue,
        markPrice,
        errors,
        validate,
        isChanged
      } = useEditOrderEntry({
        order,
        orderType,
        maxQty
      });
      useEffect(() => {
      }, [order.price, formattedOrder.order_price]);
      const onCloseDialog = useCallback(() => {
        setDialogOpen(false);
      }, []);
      const onSubmit = useCallback(
        async (values) => {
          let future;
          const isHidden = order.visible_quantity !== void 0 ? order.visible_quantity === 0 : order.visible !== void 0 ? order.visible === 0 : false;
          if (order.order_tag) {
            values.order_tag = order.order_tag;
          }
          if (order.client_order_id) {
            values.client_order_id = order.client_order_id;
          }
          if (order.algo_order_id !== void 0) {
            if (isTrailingStop) {
              if (values.callback_rate) {
                values = {
                  ...values,
                  callback_rate: (Number(values.callback_rate) / 100).toString()
                };
              } else if (isStopMarket && "order_price" in values) {
                const { order_price, ...rest } = values;
                values = rest;
              }
            }
            future = editAlgoOrder(order.algo_order_id.toString(), {
              ...values
            });
          } else {
            future = editOrder(order.order_id.toString(), {
              ...values,
              ...isHidden ? { visible_quantity: 0 } : {}
            });
          }
          try {
            setSubmitting(true);
            const res = await future;
            onClose();
          } catch (err) {
            toast$1.error(err?.message ?? `${err}`);
          } finally {
            setSubmitting(false);
          }
        },
        [editAlgoOrder, editOrder, isTrailingStop]
      );
      const onDialogConfirm = () => {
        if (formattedOrder) {
          return onSubmit(formattedOrder);
        }
        return Promise.reject();
      };
      const onSheetConfirm = () => {
        validate().then(
          (result) => {
            if (orderConfirm) {
              setDialogOpen(true);
            } else {
              onSubmit(formattedOrder);
            }
          },
          (error) => {
            if (error?.total?.message) {
              toast$1.error(error?.total.message);
            }
          }
        ).catch((err) => {
        });
      };
      return {
        ...state,
        markPrice,
        isStopMarket,
        isTrailingStop,
        showTriggerPrice,
        formattedOrder,
        setOrderValue,
        maxQty,
        onClose,
        onSheetConfirm,
        errors,
        isChanged,
        dialogOpen,
        setDialogOpen,
        onDialogConfirm,
        onCloseDialog,
        submitting,
        orderConfirm,
        setOrderConfirm
      };
    };
  }
});
var EditSheetWidget;
var init_editSheet_widget = __esm({
  "src/components/orderList/mobile/editSheet/editSheet.widget.tsx"() {
    init_editSheet_script();
    init_editSheet_ui();
    EditSheetWidget = (props) => {
      const state = useEditSheetScript(props);
      return /* @__PURE__ */ jsx(EditSheet, { ...state });
    };
  }
});

// src/components/orderList/mobile/editSheet/index.ts
var init_editSheet = __esm({
  "src/components/orderList/mobile/editSheet/index.ts"() {
    init_editSheet_ui();
    init_editSheet_widget();
    init_editSheet_script();
  }
});
var useEditBtnScript;
var init_editBtn_script = __esm({
  "src/components/orderList/mobile/editBtn/editBtn.script.tsx"() {
    init_orders_widget();
    init_orderListContext();
    init_tpslOrderRowContext();
    init_editSheet();
    useEditBtnScript = (props) => {
      const { state } = props;
      const { t } = useTranslation();
      const symbolInfo = useSymbolsInfo()[props.state.item.symbol]();
      const { editAlgoOrder, editOrder } = useOrderListContext();
      const { position } = useTPSLOrderRowContext();
      const onShowEditSheet = useCallback(() => {
        if (props.state.type === "tp_sl" /* tp_sl */) {
          modal.sheet({
            content: /* @__PURE__ */ jsx(
              PositionTPSLSheet,
              {
                symbol: props.state.item.symbol,
                symbolInfo,
                isEditing: true,
                order: props.state.item
              }
            )
          }).catch((error) => {
          });
        } else {
          modal.sheet({
            title: t("orders.editOrder"),
            // size: "xs",
            classNames: {
              content: "oui-bg-base-8"
            },
            content: /* @__PURE__ */ jsx(
              EditSheetWidget,
              {
                state,
                position,
                editAlgoOrder,
                editOrder
              }
            )
          }).catch((error) => {
          });
        }
      }, [state]);
      return {
        ...state,
        onShowEditSheet
      };
    };
  }
});
var EditBtnWidget;
var init_editBtn_widget = __esm({
  "src/components/orderList/mobile/editBtn/editBtn.widget.tsx"() {
    init_editBtn_script();
    init_editBtn_ui();
    EditBtnWidget = (props) => {
      const state = useEditBtnScript(props);
      return /* @__PURE__ */ jsx(EditBtn, { ...state });
    };
  }
});

// src/components/orderList/mobile/editBtn/index.ts
var init_editBtn = __esm({
  "src/components/orderList/mobile/editBtn/index.ts"() {
    init_editBtn_ui();
    init_editBtn_widget();
    init_editBtn_script();
  }
});
function itemsWithType(props) {
  switch (props.type) {
    case "all" /* all */:
      return /* @__PURE__ */ jsx(Fragment, {});
    case "pending" /* pending */:
      const isTrailingStop = props.item?.algo_type === AlgoOrderRootType.TRAILING_STOP;
      if (isTrailingStop) {
        return /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(Qty, { ...props }),
          /* @__PURE__ */ jsx(Filled, { ...props }),
          /* @__PURE__ */ jsx(EstTotal, { ...props }),
          /* @__PURE__ */ jsx(TriggerPrice, { ...props, isPending: true }),
          /* @__PURE__ */ jsx(TrailingPrice, { ...props }),
          /* @__PURE__ */ jsx(TrailingCallback, { ...props })
        ] });
      }
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(Qty, { ...props }),
        /* @__PURE__ */ jsx(Filled, { ...props }),
        /* @__PURE__ */ jsx(EstTotal, { ...props }),
        /* @__PURE__ */ jsx(TriggerPrice, { ...props, isPending: true }),
        /* @__PURE__ */ jsx(LimitPrice, { ...props }),
        /* @__PURE__ */ jsx(MarkPrice, { ...props })
      ] });
    case "tp_sl" /* tp_sl */:
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(TPTrigger, { ...props }),
        /* @__PURE__ */ jsx(TPPrice, { ...props }),
        /* @__PURE__ */ jsx(TPSLQuantity, { ...props }),
        /* @__PURE__ */ jsx(SLTrigger, { ...props }),
        /* @__PURE__ */ jsx(SLPrice, { ...props })
      ] });
    case "filled" /* filled */:
      return /* @__PURE__ */ jsx(Fragment, {});
    case "cancelled" /* cancelled */:
      return /* @__PURE__ */ jsx(Fragment, {});
    case "rejected" /* rejected */:
      return /* @__PURE__ */ jsx(Fragment, {});
    case "orderHistory" /* orderHistory */:
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(Qty, { ...props }),
        /* @__PURE__ */ jsx(Filled, { ...props }),
        /* @__PURE__ */ jsx(TriggerPrice, { ...props, align: "end" }),
        /* @__PURE__ */ jsx(AvgPrice2, { ...props }),
        /* @__PURE__ */ jsx(OrderPrice, { ...props }),
        /* @__PURE__ */ jsx(RealizedPnL, { ...props }),
        /* @__PURE__ */ jsx("div", {}),
        /* @__PURE__ */ jsx("div", {}),
        /* @__PURE__ */ jsx(Fee, { ...props })
      ] });
  }
}
var OrderCell, Header, Body, Btns;
var init_orderCell_ui = __esm({
  "src/components/orderList/mobile/orderCell.ui.tsx"() {
    init_orders_widget();
    init_bracketOrderPrice2();
    init_cancelBtn2();
    init_editBtn();
    init_items();
    OrderCell = (props) => {
      return /* @__PURE__ */ jsxs(
        Flex,
        {
          direction: "column",
          width: "100%",
          gap: 2,
          itemAlign: "start",
          className: props.className,
          children: [
            /* @__PURE__ */ jsx(Header, { ...props }),
            /* @__PURE__ */ jsx(Divider, { intensity: 6, className: "oui-w-full" }),
            /* @__PURE__ */ jsx(Body, { ...props }),
            props.type === "pending" /* pending */ && /* @__PURE__ */ jsx(BracketOrderPriceWidget, { ...props }),
            props.type !== "orderHistory" /* orderHistory */ && /* @__PURE__ */ jsx(Btns, { ...props })
          ]
        }
      );
    };
    Header = (props) => {
      return /* @__PURE__ */ jsxs(Flex, { direction: "column", gap: 1, width: "100%", children: [
        /* @__PURE__ */ jsxs(Flex, { justify: "between", width: "100%", children: [
          /* @__PURE__ */ jsx(SymbolToken, { ...props }),
          /* @__PURE__ */ jsx(OrderTime, { ...props })
        ] }),
        /* @__PURE__ */ jsxs(Flex, { width: "100%", justify: "between", children: [
          /* @__PURE__ */ jsx(OrderTypeView, { ...props }),
          props.type === "orderHistory" /* orderHistory */ && /* @__PURE__ */ jsx(OrderState, { ...props })
        ] })
      ] });
    };
    Body = (props) => {
      return /* @__PURE__ */ jsx(Grid, { cols: 3, rows: 2, width: "100%", gap: 1, children: itemsWithType(props) });
    };
    Btns = (props) => {
      return /* @__PURE__ */ jsxs(Grid, { cols: 3, rows: 1, width: "100%", gap: 2, children: [
        /* @__PURE__ */ jsx("div", {}),
        /* @__PURE__ */ jsx(EditBtnWidget, { state: props }),
        /* @__PURE__ */ jsx(CancelBtnWidget, { state: props })
      ] });
    };
  }
});

// src/components/orderList/mobile/orderCell.script.tsx
var useOrderCellScript;
var init_orderCell_script = __esm({
  "src/components/orderList/mobile/orderCell.script.tsx"() {
    init_symbolContext();
    useOrderCellScript = (props) => {
      const symbolInfo = useSymbolContext();
      return {
        ...props,
        ...symbolInfo,
        // use symbolInfo replace ...symbolInfo
        symbolInfo
      };
    };
  }
});
var OrderCellWidget;
var init_orderCell_widget = __esm({
  "src/components/orderList/mobile/orderCell.widget.tsx"() {
    init_orderCell_script();
    init_orderCell_ui();
    OrderCellWidget = (props) => {
      const { className, ...rest } = props;
      const state = useOrderCellScript(rest);
      return /* @__PURE__ */ jsx(OrderCell, { ...state, className });
    };
  }
});

// src/components/orderList/mobile/index.ts
var init_mobile = __esm({
  "src/components/orderList/mobile/index.ts"() {
    init_orderCell_ui();
    init_orderCell_widget();
    init_orderCell_script();
  }
});
var OrderListProvider;
var init_orderListProvider = __esm({
  "src/components/orderList/orderListProvider.tsx"() {
    init_orderListContext();
    OrderListProvider = (props) => {
      const { cancelOrder, editOrder, cancelAlgoOrder, editAlgoOrder } = props;
      const { t } = useTranslation();
      const symbolsInfo = useSymbolsInfo();
      const onCancelOrder = useMemoizedFn(
        async (order) => {
          if (order.algo_order_id !== void 0) {
            if ("root_algo_order_id" in order && order.root_algo_order_id !== order.algo_order_id) {
              return cancelAlgoOrder(order.root_algo_order_id, order.symbol);
            }
            return cancelAlgoOrder(order.algo_order_id, order.symbol).then(
              () => {
              }
            );
          }
          return cancelOrder(order.order_id, order.symbol).then(() => {
          });
        }
      );
      const checkMinNotional = useMemoizedFn(
        (symbol, price2, qty) => {
          const { min_notional } = symbolsInfo[symbol]();
          const minNotional = getMinNotional({ price: price2, qty, min_notional });
          if (minNotional !== void 0) {
            return t("orderEntry.total.error.min", { value: minNotional });
          }
        }
      );
      const memoizedValue = useMemo(() => {
        return {
          onCancelOrder,
          editOrder,
          editAlgoOrder,
          checkMinNotional
        };
      }, [onCancelOrder, editOrder, editAlgoOrder, checkMinNotional]);
      return /* @__PURE__ */ jsx(OrderListContext.Provider, { value: memoizedValue, children: props.children });
    };
  }
});
var DesktopOrderList, MobileOrderList, CancelAll;
var init_orderList_ui = __esm({
  "src/components/orderList/orderList.ui.tsx"() {
    init_util();
    init_orders_widget();
    init_symbolProvider();
    init_useColumn();
    init_mobile();
    init_orderListProvider();
    init_tpslOrderRowContext();
    DesktopOrderList = (props) => {
      const { sharePnLConfig, ...rest } = props;
      const columns = useOrderColumn({
        _type: props.type,
        onSymbolChange: props.onSymbolChange,
        pnlNotionalDecimalPrecision: props.pnlNotionalDecimalPrecision,
        sharePnLConfig,
        symbolsInfo: props.symbolsInfo
      });
      return /* @__PURE__ */ jsx(
        OrderListProvider,
        {
          cancelOrder: props.cancelOrder,
          editOrder: props.updateOrder,
          cancelAlgoOrder: props.cancelAlgoOrder,
          editAlgoOrder: props.updateAlgoOrder,
          children: /* @__PURE__ */ jsxs(Flex, { direction: "column", width: "100%", height: "100%", itemAlign: "start", children: [
            props.filterItems.length > 0 && /* @__PURE__ */ jsx(
              DataFilter,
              {
                items: props.filterItems,
                onFilter: (value) => {
                  props.onFilter(value);
                },
                trailing: ["pending" /* pending */, "tp_sl" /* tp_sl */].includes(props.type) && /* @__PURE__ */ jsx(CancelAll, { ...props })
              }
            ),
            /* @__PURE__ */ jsx(
              AuthGuardDataTable,
              {
                columns,
                loading: props.isLoading,
                dataSource: props.dataSource,
                bordered: true,
                ignoreLoadingCheck: true,
                testIds: {
                  body: props.testIds?.tableBody
                },
                classNames: {
                  header: "oui-h-[38px]",
                  root: "oui-items-start !oui-h-[calc(100%_-_49px)]"
                },
                onRow: (record, index) => {
                  return {
                    className: cn(
                      "oui-h-[48px]",
                      grayCell(record) ? "oui-text-base-contrast-20" : "oui-text-base-contrast-80"
                    )
                  };
                },
                generatedRowKey: (record, index) => `${props.type}${index}${record.order_id || record.algo_order_id}_index${index}`,
                renderRowContainer: (record, index, children) => {
                  if (props.type === "tp_sl" /* tp_sl */ || props.type === "pending" /* pending */) {
                    children = /* @__PURE__ */ jsx(TPSLOrderRowProvider, { order: record, children });
                  }
                  return /* @__PURE__ */ jsx(SymbolProvider, { symbol: record.symbol, children });
                },
                pagination: props.pagination,
                manualPagination: props.manualPagination,
                features: [TableFeatures.DownloadFeature],
                getTableInstance: (table) => {
                  props.tableInstance.current = table;
                }
              }
            )
          ] })
        }
      );
    };
    MobileOrderList = (props) => {
      const { t } = useTranslation();
      return /* @__PURE__ */ jsx(
        OrderListProvider,
        {
          cancelOrder: props.cancelOrder,
          editOrder: props.updateOrder,
          cancelAlgoOrder: props.cancelAlgoOrder,
          editAlgoOrder: props.updateAlgoOrder,
          children: /* @__PURE__ */ jsxs(
            Grid,
            {
              cols: 1,
              rows: 2,
              className: "oui-w-full oui-grid-rows-[auto,1fr]",
              gap: 2,
              children: [
                props.showFilter ? /* @__PURE__ */ jsx(Flex, { gap: 2, p: 2, className: "oui-rounded-b-xl oui-bg-base-9", children: props.filterItems.map((item) => {
                  if (item.type !== "select") {
                    return null;
                  }
                  return /* @__PURE__ */ jsx(
                    Picker,
                    {
                      options: item.options,
                      size: "sm",
                      value: item.value,
                      className: "oui-text-2xs oui-text-base-contrast-54 ",
                      placeholder: item.name === "side" ? t("common.side.all") : item.name === "status" ? t("common.status.all") : "",
                      onValueChange: (value) => {
                        props.onFilter?.({ name: item.name, value });
                      }
                    },
                    `item-${item.name}`
                  );
                }) }) : /* @__PURE__ */ jsx("div", {}),
                /* @__PURE__ */ jsx(
                  ListView,
                  {
                    className: props.classNames?.root,
                    contentClassName: props.classNames?.content,
                    dataSource: props.dataSource,
                    loadMore: props.loadMore,
                    isLoading: props.isLoading,
                    renderItem: (item, index) => {
                      let children = /* @__PURE__ */ jsx(
                        OrderCellWidget,
                        {
                          item,
                          index,
                          className: props.classNames?.cell,
                          type: props.type,
                          onSymbolChange: props.onSymbolChange,
                          sharePnLConfig: props.sharePnLConfig
                        }
                      );
                      if (["tp_sl" /* tp_sl */, "pending" /* pending */].includes(props.type)) {
                        children = /* @__PURE__ */ jsx(TPSLOrderRowProvider, { order: item, children });
                      }
                      return /* @__PURE__ */ jsx(SymbolProvider, { symbol: item.symbol, children });
                    }
                  }
                )
              ]
            }
          )
        }
      );
    };
    CancelAll = (props) => {
      const { t } = useTranslation();
      const { symbol } = props;
      const formattedSymbol = symbol ? formatSymbol(symbol, "base") : symbol;
      return /* @__PURE__ */ jsx(
        Button,
        {
          variant: "outlined",
          color: "secondary",
          size: "xs",
          disabled: (props.dataSource?.length ?? 0) == 0,
          className: "disabled:oui-bg-transport",
          onClick: (e) => props.onCancelAll(),
          "data-testid": `oui-testid-dataList-${props.type.toLowerCase()}-cancelAll-button`,
          children: symbol ? t("orders.cancelAll.ofSymbol", { symbol: formattedSymbol }) : t("orders.cancelAll")
        }
      );
    };
  }
});
var useFormatOrderHistory;
var init_useFormatOrderHistory = __esm({
  "src/components/orderList/useFormatOrderHistory.ts"() {
    useFormatOrderHistory = (data) => {
      const formattedData = useMemo(() => {
        const _data = [];
        for (let index = 0; index < data.length; index++) {
          const element = data[index];
          if (element.algo_type === AlgoOrderRootType.POSITIONAL_TP_SL || element.algo_type === AlgoOrderRootType.TP_SL) {
            if (element.algo_status !== OrderStatus.FILLED && element.algo_status !== OrderStatus.PARTIAL_FILLED) {
              for (let j = 0; j < element.child_orders.length; j++) {
                const e = element.child_orders[j];
                if (!e.is_activated || !e.trigger_price) {
                  continue;
                }
                e.parent_algo_type = element.algo_type;
                _data.push(e);
              }
            } else {
              for (let j = 0; j < element.child_orders.length; j++) {
                const e = element.child_orders[j];
                if (!!e && (e.algo_status === OrderStatus.FILLED || e.algo_status === OrderStatus.PARTIAL_FILLED)) {
                  e.parent_algo_type = element.algo_type;
                  _data.push(e);
                }
              }
            }
          } else {
            _data.push(element);
          }
        }
        return _data;
      }, [data]);
      return formattedData;
    };
  }
});
function offsetStartOfDay(date) {
  if (date == null) return date;
  const newDate = new Date(date);
  newDate.setHours(0, 0, 0, 0);
  return newDate;
}
function offsetEndOfDay(date) {
  if (date == null) return date;
  const newDate = new Date(date);
  newDate.setHours(23, 59, 59, 999);
  return newDate;
}
var useOrderListScript, useFilter, formatDatePickerRange, getDialogInfo;
var init_orderList_script = __esm({
  "src/components/orderList/orderList.script.tsx"() {
    init_util();
    init_orders_widget();
    init_useFormatOrderHistory();
    useOrderListScript = (props) => {
      const {
        ordersStatus,
        type,
        enableLoadMore = false,
        onSymbolChange,
        filterConfig,
        pnlNotionalDecimalPrecision,
        sharePnLConfig
      } = props;
      const { t } = useTranslation();
      const symbolsInfo = useSymbolsInfo();
      const manualPagination = useMemo(() => {
        return ordersStatus !== OrderStatus.INCOMPLETE;
      }, [ordersStatus]);
      const defaultPageSize = 50;
      const { page, pageSize, setPage, pagination, parsePagination } = usePagination({
        pageSize: defaultPageSize
      });
      useEffect(() => {
        setPage(1);
      }, [props.symbol]);
      const {
        orderStatus,
        ordersSide,
        dateRange,
        filterItems,
        onFilter,
        filterDays,
        updateFilterDays
      } = useFilter(type, {
        ordersStatus,
        setPage,
        filterConfig
      });
      const includes = useMemo(() => {
        if (type === "tp_sl" /* tp_sl */) {
          return [AlgoOrderRootType.POSITIONAL_TP_SL, AlgoOrderRootType.TP_SL];
        }
        return void 0;
      }, [type]);
      const excludes = useMemo(() => {
        if (type === "pending" /* pending */) {
          return [AlgoOrderRootType.POSITIONAL_TP_SL, AlgoOrderRootType.TP_SL];
        }
        return void 0;
      }, [type]);
      const tableInstance = useRef();
      useImperativeHandle(props.ref, () => ({
        download: () => {
          const filename = `orders_${format(/* @__PURE__ */ new Date(), "yyyyMMdd_HHmmss")}`;
          tableInstance.current?.download(filename);
        }
      }));
      const [
        data,
        {
          isLoading,
          loadMore,
          cancelOrder,
          updateOrder,
          cancelAlgoOrder,
          updateAlgoOrder,
          cancelAllPendingOrders,
          cancelAllTPSLOrders,
          meta,
          refresh
        }
      ] = useOrderStream({
        symbol: props.symbol,
        status: orderStatus,
        side: ordersSide,
        page: enableLoadMore || !manualPagination ? void 0 : page,
        // pending and ts_sl list get all data
        size: manualPagination ? pageSize : 500,
        sourceTypeAll: type === "orderHistory" /* orderHistory */,
        dateRange,
        includes,
        excludes
      });
      const localPageSizeKey = `orderly_${type}_pageSize`;
      const [typePageSize, setTypePageSize] = useLocalStorage(
        localPageSizeKey,
        defaultPageSize
      );
      useEffect(() => {
        if (typePageSize !== pageSize) {
          setTypePageSize(pageSize);
        }
      }, [pageSize, typePageSize]);
      const onCancelAll = useCallback(() => {
        const { title, content } = getDialogInfo(type, t, props.symbol);
        modal.confirm({
          title,
          content: /* @__PURE__ */ jsx(Text, { size: "sm", children: content }),
          onCancel: async () => {
          },
          onOk: async () => {
            try {
              if (type === "tp_sl" /* tp_sl */) {
                await cancelAllTPSLOrders(props.symbol);
              } else {
                await cancelAllPendingOrders(props.symbol);
              }
              refresh();
              return Promise.resolve(true);
            } catch (error) {
              if (error?.message !== void 0) {
                toast.error(error.message);
              }
              return Promise.resolve(false);
            } finally {
              Promise.resolve();
            }
          }
        });
      }, [type, t, props.symbol]);
      const formattedData = useFormatOrderHistory(data ?? []);
      const filteredData = useMemo(() => {
        const rawData = type !== "tp_sl" /* tp_sl */ ? formattedData : data;
        if (!props.symbol || !rawData) {
          return rawData;
        }
        return rawData.filter(
          (order) => order.symbol === props.symbol
        );
      }, [type, formattedData, data, props.symbol]);
      const dataSource = useDataTap(filteredData, {
        fallbackData: []
      }) ?? void 0;
      const _pagination = useMemo(() => {
        if (manualPagination) {
          return parsePagination(meta);
        }
        return pagination;
      }, [meta, manualPagination, parsePagination, pagination]);
      return {
        type,
        dataSource,
        isLoading,
        loadMore,
        cancelOrder,
        updateOrder,
        cancelAlgoOrder,
        updateAlgoOrder,
        pagination: _pagination,
        manualPagination,
        pnlNotionalDecimalPrecision,
        // filter
        onFilter,
        filterItems,
        onCancelAll,
        onSymbolChange,
        sharePnLConfig,
        tableInstance,
        symbolsInfo,
        filterDays,
        updateFilterDays,
        symbol: props.symbol
      };
    };
    useFilter = (type, option) => {
      const { t } = useTranslation();
      const [orderStatus, setOrderStatus] = useState(
        option.ordersStatus ?? "all"
      );
      const [ordersSide, setOrdersSide] = useState(
        option.filterConfig?.side ?? "all"
      );
      const [filterDays, setFilterDays] = useState(90);
      const defaultRange = option.filterConfig?.range ?? (type === "all" /* all */ || type === "orderHistory" /* orderHistory */ ? formatDatePickerRange({
        to: /* @__PURE__ */ new Date(),
        from: offsetEndOfDay(subDays(/* @__PURE__ */ new Date(), 89))
      }) : {});
      const [dateRange, setDateRange] = useState(defaultRange);
      const updateFilterDays = (days) => {
        setFilterDays(days);
        setDateRange({
          from: offsetStartOfDay(subDays(/* @__PURE__ */ new Date(), days - 1)),
          to: offsetEndOfDay(/* @__PURE__ */ new Date())
        });
        option.setPage(1);
      };
      const onFilter = (filter) => {
        if (filter.name === "side") {
          setOrdersSide(filter.value);
          option.setPage(1);
        }
        if (filter.name === "status") {
          setOrderStatus(filter.value);
          option.setPage(1);
        }
        if (filter.name === "dateRange") {
          setDateRange(formatDatePickerRange(filter.value));
          option.setPage(1);
          const newDateRange = formatDatePickerRange(filter.value);
          if (newDateRange.from && newDateRange.to) {
            const diffDays = Math.abs(differenceInDays(newDateRange.from, newDateRange.to)) + 1;
            const dateRangeMap = {
              1: {
                from: offsetStartOfDay(/* @__PURE__ */ new Date()),
                to: offsetEndOfDay(/* @__PURE__ */ new Date())
              },
              7: {
                from: offsetStartOfDay(subDays(/* @__PURE__ */ new Date(), 6)),
                to: offsetEndOfDay(/* @__PURE__ */ new Date())
              },
              30: {
                from: offsetStartOfDay(subDays(/* @__PURE__ */ new Date(), 29)),
                to: offsetEndOfDay(/* @__PURE__ */ new Date())
              },
              90: {
                from: offsetStartOfDay(subDays(/* @__PURE__ */ new Date(), 89)),
                to: offsetEndOfDay(/* @__PURE__ */ new Date())
              }
            };
            const dateRange2 = dateRangeMap[diffDays];
            if (dateRange2 && areDatesEqual(dateRange2.from, newDateRange.from) && areDatesEqual(dateRange2.to, newDateRange.to)) {
              setFilterDays(diffDays);
            } else {
              setFilterDays(null);
            }
          }
        }
      };
      const filterItems = useMemo(() => {
        const sideFilter = {
          type: "select",
          name: "side",
          options: [
            {
              label: t("common.side.all"),
              value: "all"
            },
            {
              label: t("common.buy"),
              value: "BUY"
            },
            {
              label: t("common.sell"),
              value: "SELL"
            }
          ],
          value: ordersSide
        };
        const dateRangeFilter = {
          type: "range",
          name: "dateRange",
          value: dateRange
        };
        const statusFilter = {
          type: "select",
          name: "status",
          options: [
            {
              label: t("common.status.all"),
              value: "all"
            },
            {
              label: t("orders.status.pending"),
              value: OrderStatus.INCOMPLETE
            },
            {
              label: t("orders.status.filled"),
              value: OrderStatus.FILLED
            },
            {
              label: t("orders.status.partialFilled"),
              value: OrderStatus.PARTIAL_FILLED
            },
            {
              label: t("orders.status.canceled"),
              value: OrderStatus.CANCELLED
            },
            {
              label: t("orders.status.rejected"),
              value: OrderStatus.REJECTED
            }
          ],
          value: orderStatus
        };
        switch (type) {
          case "all" /* all */:
            return [sideFilter, statusFilter, dateRangeFilter];
          case "pending" /* pending */:
            return [sideFilter];
          case "tp_sl" /* tp_sl */:
            return [sideFilter];
          case "filled" /* filled */:
            return [sideFilter];
          case "cancelled" /* cancelled */:
            return [sideFilter];
          case "rejected" /* rejected */:
            return [sideFilter];
          case "orderHistory" /* orderHistory */:
            return [sideFilter, statusFilter, dateRangeFilter];
        }
      }, [type, ordersSide, orderStatus, dateRange, t]);
      return {
        filterItems,
        onFilter,
        ordersSide: ordersSide === "all" ? void 0 : ordersSide,
        dateRange,
        orderStatus: orderStatus === "all" ? void 0 : orderStatus,
        filterDays,
        updateFilterDays
      };
    };
    formatDatePickerRange = (option) => ({
      from: offsetStartOfDay(option.from),
      to: offsetEndOfDay(option.to ?? option.from)
    });
    getDialogInfo = (type, t, symbol) => {
      const formattedSymbol = symbol ? formatSymbol(symbol, "base") : symbol;
      switch (type) {
        case "pending" /* pending */:
          if (symbol !== void 0) {
            return {
              title: t("orders.pending.cancelAll.forSymbol", {
                symbol: formattedSymbol
              }),
              content: t("orders.pending.cancelAll.forSymbol.description", {
                symbol: formattedSymbol
              })
            };
          }
          return {
            title: t("orders.pending.cancelAll"),
            content: t("orders.pending.cancelAll.description")
          };
        case "tp_sl" /* tp_sl */:
          if (symbol !== void 0) {
            return {
              title: t("orders.tpsl.cancelAll.forSymbol", {
                symbol: formattedSymbol
              }),
              content: t("orders.tpsl.cancelAll.forSymbol.description", {
                symbol: formattedSymbol
              })
            };
          }
          return {
            title: t("orders.tpsl.cancelAll"),
            content: t("orders.tpsl.cancelAll.description")
          };
        default:
          return {
            title: "",
            content: ""
          };
      }
    };
  }
});
var DesktopOrderListWidget, MobileOrderListWidget;
var init_orderList_widget = __esm({
  "src/components/orderList/orderList.widget.tsx"() {
    init_orderList_script();
    init_orderList_ui();
    DesktopOrderListWidget = forwardRef((props, ref) => {
      const { testIds, ...rest } = props;
      const state = useOrderListScript({ ...rest, ref });
      return /* @__PURE__ */ jsx(DesktopOrderList, { ...state, testIds });
    });
    MobileOrderListWidget = (props) => {
      const state = useOrderListScript({ ...props, enableLoadMore: true });
      return /* @__PURE__ */ jsx(
        MobileOrderList,
        {
          ...state,
          classNames: props.classNames,
          showFilter: props.showFilter
        }
      );
    };
  }
});

// src/components/orderList/index.ts
var orderList_exports = {};
__export(orderList_exports, {
  DesktopOrderList: () => DesktopOrderList,
  DesktopOrderListWidget: () => DesktopOrderListWidget,
  MobileOrderList: () => MobileOrderList,
  MobileOrderListWidget: () => MobileOrderListWidget,
  useOrderListScript: () => useOrderListScript
});
var init_orderList = __esm({
  "src/components/orderList/index.ts"() {
    init_orderList_ui();
    init_orderList_widget();
    init_orderList_script();
  }
});
var LazyDesktopOrderListWidget, Orders;
var init_orders_ui = __esm({
  "src/components/orders.ui.tsx"() {
    init_orders_widget();
    LazyDesktopOrderListWidget = React2.lazy(
      () => Promise.resolve().then(() => (init_orderList(), orderList_exports)).then((mod) => {
        return { default: mod.DesktopOrderListWidget };
      })
    );
    Orders = (props) => {
      const { t } = useTranslation();
      const { pnlNotionalDecimalPrecision, orderListRef } = props;
      const commonProps = {
        pnlNotionalDecimalPrecision
      };
      return /* @__PURE__ */ jsxs(
        Tabs,
        {
          defaultValue: props.current || "all" /* all */,
          variant: "contained",
          className: "oui-h-full",
          classNames: {
            tabsContent: "oui-h-[calc(100%_-_28px)]"
          },
          children: [
            /* @__PURE__ */ jsx(TabPanel, { value: "all" /* all */, title: t("common.all"), children: /* @__PURE__ */ jsx(React2.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
              LazyDesktopOrderListWidget,
              {
                ref: orderListRef,
                type: "all" /* all */,
                pnlNotionalDecimalPrecision,
                sharePnLConfig: props.sharePnLConfig,
                ...commonProps
              }
            ) }) }),
            /* @__PURE__ */ jsx(TabPanel, { value: "pending" /* pending */, title: t("orders.status.pending"), children: /* @__PURE__ */ jsx(React2.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
              LazyDesktopOrderListWidget,
              {
                ref: orderListRef,
                type: "pending" /* pending */,
                ordersStatus: OrderStatus.INCOMPLETE,
                ...commonProps
              }
            ) }) }),
            /* @__PURE__ */ jsx(TabPanel, { value: "tp_sl" /* tp_sl */, title: t("common.tpsl"), children: /* @__PURE__ */ jsx(React2.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
              LazyDesktopOrderListWidget,
              {
                ref: orderListRef,
                type: "tp_sl" /* tp_sl */,
                ordersStatus: OrderStatus.INCOMPLETE,
                ...commonProps
              }
            ) }) }),
            /* @__PURE__ */ jsx(TabPanel, { value: "filled" /* filled */, title: t("orders.status.filled"), children: /* @__PURE__ */ jsx(React2.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
              LazyDesktopOrderListWidget,
              {
                ref: orderListRef,
                type: "filled" /* filled */,
                ordersStatus: OrderStatus.FILLED,
                pnlNotionalDecimalPrecision,
                sharePnLConfig: props.sharePnLConfig,
                ...commonProps
              }
            ) }) }),
            /* @__PURE__ */ jsx(TabPanel, { value: "cancelled" /* cancelled */, title: t("orders.status.canceled"), children: /* @__PURE__ */ jsx(React2.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
              LazyDesktopOrderListWidget,
              {
                ref: orderListRef,
                type: "cancelled" /* cancelled */,
                ordersStatus: OrderStatus.CANCELLED,
                ...commonProps
              }
            ) }) }),
            /* @__PURE__ */ jsx(TabPanel, { value: "rejected" /* rejected */, title: t("orders.status.rejected"), children: /* @__PURE__ */ jsx(React2.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
              LazyDesktopOrderListWidget,
              {
                ref: orderListRef,
                type: "rejected" /* rejected */,
                ordersStatus: OrderStatus.REJECTED,
                ...commonProps
              }
            ) }) })
          ]
        }
      );
    };
  }
});
var TabType, OrdersWidget;
var init_orders_widget = __esm({
  "src/components/orders.widget.tsx"() {
    init_orders_script();
    init_orders_ui();
    TabType = /* @__PURE__ */ ((TabType2) => {
      TabType2["all"] = "all";
      TabType2["pending"] = "pending";
      TabType2["tp_sl"] = "tp_sl";
      TabType2["filled"] = "filled";
      TabType2["cancelled"] = "cancelled";
      TabType2["rejected"] = "rejected";
      TabType2["orderHistory"] = "orderHistory";
      return TabType2;
    })(TabType || {});
    OrdersWidget = forwardRef(
      (props, ref) => {
        const state = useOrdersScript({ ...props, ref });
        return /* @__PURE__ */ jsx(Orders, { ...state });
      }
    );
  }
});

// src/index.ts
init_orders_widget();
init_orders_script();
init_orderList();

export { DesktopOrderList, DesktopOrderListWidget, MobileOrderList, MobileOrderListWidget, OrdersWidget, TabType, useOrderListScript, useOrdersScript };
