// @orderly.network/ui-tradingview/dist/index.mjs
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import React3, { forwardRef, useRef, useState, useMemo, useEffect, useCallback } from 'react';
import { useTranslation, Trans, useLocaleCode, i18n } from '@orderly.network/i18n';
import { cn, Flex, Divider, Box, toast, modal, DropdownMenuRoot, DropdownMenuTrigger, DropdownMenuPortal, DropdownMenuContent, Text, Switch } from '@orderly.network/ui';
import { useMediaQuery, useConfig, useAccount, useSymbolsInfo, useLocalStorage, useOrderEntry_deprecated, useWS, usePositionStream, useOrderStream, useEventEmitter } from '@orderly.network/hooks';
import { MEDIA_TABLET, OrderSide, TradingviewFullscreenKey, OrderType, AccountStatusEnum, OrderStatus } from '@orderly.network/types';
import { Decimal, commify, getTrailingStopPrice } from '@orderly.network/utils';
import { startOfSecond, startOfMinute, startOfHour, startOfDay, startOfMonth, startOfYear, startOfWeek } from 'date-fns';

var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var CaretIcon, DisplaySettingIcon, IndicatorsIcon, SettingIcon, BarIcon, CandlesIcon, HollowCandlesIcon, LineIcon, AreaIcon, BaseLineIcon, UnSelectIcon, SelectedIcon;
var init_icons = __esm({
  "src/icons/index.tsx"() {
    CaretIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 12 12",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Caret-down", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Vector",
            d: "M3.00349 3.9978C2.59149 3.9978 2.36564 4.4653 2.61289 4.7948C2.98789 5.2948 5.23784 8.29479 5.61289 8.79479C5.81289 9.06128 6.20974 9.06128 6.40974 8.79479L9.40974 4.7948C9.65694 4.4653 9.41554 3.9978 9.00349 3.9978H3.00349Z"
          }
        ) })
      }
    );
    DisplaySettingIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 18 18",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Display settings", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Union",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M15.0571 10.9856L11.6295 5.06523L8.71377 10.8968C8.38933 11.5456 7.54464 11.7223 6.98733 11.2579L5.75311 10.2294L2.83709 14.7654L1.57532 13.9543L4.49134 9.41824C4.97892 8.65978 6.0207 8.49981 6.71338 9.07705L7.58387 9.80246L10.5991 3.77192C10.9997 2.97082 12.1302 2.93624 12.579 3.71137L14.6517 7.29147C14.7714 7.03643 15.0305 6.85983 15.3309 6.85983C15.7451 6.85983 16.0809 7.19562 16.0809 7.60983C16.0809 8.02405 15.7451 8.35983 15.3309 8.35983C15.31 8.35983 15.2892 8.35897 15.2687 8.35729L16.3553 10.2341L15.0571 10.9856ZM2.58093 6.85983C2.16672 6.85983 1.83093 7.19562 1.83093 7.60983C1.83093 8.02405 2.16672 8.35983 2.58093 8.35983C2.99515 8.35983 3.33093 8.02405 3.33093 7.60983C3.33093 7.19562 2.99515 6.85983 2.58093 6.85983ZM4.83093 6.85983C4.41672 6.85983 4.08093 7.19562 4.08093 7.60983C4.08093 8.02405 4.41672 8.35983 4.83093 8.35983C5.24515 8.35983 5.58093 8.02405 5.58093 7.60983C5.58093 7.19562 5.24515 6.85983 4.83093 6.85983ZM6.33093 7.60983C6.33093 7.19562 6.66672 6.85983 7.08093 6.85983C7.49515 6.85983 7.83093 7.19562 7.83093 7.60983C7.83093 8.02405 7.49515 8.35983 7.08093 8.35983C6.66672 8.35983 6.33093 8.02405 6.33093 7.60983ZM11.5809 6.85983C11.1667 6.85983 10.8309 7.19562 10.8309 7.60983C10.8309 8.02405 11.1667 8.35983 11.5809 8.35983C11.9951 8.35983 12.3309 8.02405 12.3309 7.60983C12.3309 7.19562 11.9951 6.85983 11.5809 6.85983Z"
          }
        ) })
      }
    );
    IndicatorsIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 18 18",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Indicators", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Union",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M9.74483 2.25C8.96145 2.25 8.26885 2.76459 8.04549 3.51818L7.46202 5.48254H5.06262V7.08339H6.98652L5.00539 13.7532L4.31554 13.1112L3.22498 14.2831L4.4777 15.4489C5.08739 16.015 6.08317 15.7469 6.32134 14.945L8.65649 7.08339H11.6032V5.48254H9.13198L9.58033 3.97311C9.60167 3.90112 9.66833 3.85084 9.74483 3.85084H11.69V2.25H9.74483ZM10.6234 11.2354L10.2638 11.5172C10.0153 11.712 9.65591 11.6684 9.46115 11.4199C9.26639 11.1713 9.30998 10.812 9.55852 10.6172L10.1451 10.1575C10.5693 9.82514 11.1868 9.94011 11.4642 10.4005L11.4682 10.407L12.1304 11.5752L13.4594 10.3111C13.6882 10.0935 14.05 10.1025 14.2677 10.3313C14.4853 10.5601 14.4763 10.922 14.2475 11.1396L12.7116 12.6006L13.3897 13.797L13.6482 13.4503C13.8369 13.1972 14.1951 13.145 14.4483 13.3337C14.7014 13.5225 14.7536 13.8807 14.5649 14.1338L13.9917 14.9025L13.94 14.9457C13.5178 15.2981 12.8837 15.1886 12.6003 14.7226L12.5958 14.7151L11.8576 13.4129L10.3563 14.841C10.1275 15.0586 9.76561 15.0495 9.54798 14.8208C9.33036 14.592 9.3394 14.2301 9.56819 14.0125L11.2764 12.3875L10.6234 11.2354Z"
          }
        ) })
      }
    );
    SettingIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 18 18",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Setting", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Vector",
            d: "M6.72589 1.84154C5.66239 2.17904 4.74515 2.72954 3.93665 3.48254C3.6869 3.71504 3.6104 4.09604 3.7724 4.39604C4.37315 5.50679 3.74464 6.69554 2.38939 6.76379C2.05789 6.78029 1.76465 7.0263 1.68665 7.34955C1.5524 7.9083 1.49915 8.37629 1.49915 8.98979C1.49915 9.50504 1.55464 10.0885 1.66339 10.6075C1.73089 10.9315 2.0129 11.1648 2.3429 11.1933C3.7064 11.311 4.3814 12.3513 3.7724 13.6773C3.6374 13.972 3.6989 14.3245 3.93665 14.545C4.73315 15.2815 5.64814 15.8013 6.72589 16.1388C7.03339 16.2348 7.38064 16.1185 7.56964 15.8575C8.40364 14.7033 9.6134 14.6995 10.4054 15.8575C10.5921 16.1298 10.9341 16.261 11.2491 16.162C12.2894 15.8343 13.2584 15.277 14.0616 14.545C14.3091 14.3193 14.3744 13.954 14.2259 13.654C13.6019 12.3948 14.3197 11.239 15.6082 11.2165C15.9502 11.2105 16.2547 10.9863 16.3349 10.654C16.4647 10.1163 16.4991 9.64829 16.4991 8.98979C16.4991 8.42429 16.4324 7.86704 16.3116 7.32629C16.2351 6.98354 15.9359 6.74054 15.5849 6.73979C14.3167 6.73754 13.6056 5.49104 14.2259 4.39604C14.3984 4.09154 14.3444 3.71804 14.0849 3.48254C13.2674 2.74004 12.2706 2.15954 11.2259 1.84154C10.9049 1.74404 10.5636 1.86404 10.3821 2.14604C9.65764 3.27179 8.30465 3.29129 7.5929 2.17004C7.41065 1.88204 7.04989 1.73879 6.72589 1.84154ZM11.2844 3.43455C11.7966 3.64905 12.2009 3.8733 12.6651 4.22505C12.1206 5.9478 13.0439 7.73729 14.9564 8.17304C15.0036 8.48279 14.9991 8.67029 14.9991 8.98979C14.9991 9.37154 15.0044 9.50579 14.9609 9.78179C13.0566 10.1763 12.1146 11.9155 12.6449 13.7538C12.1889 14.0845 11.8626 14.314 11.2904 14.5353C9.94639 13.1673 8.0804 13.1073 6.7079 14.5443C6.1724 14.3088 5.76066 14.0995 5.34441 13.7478C5.86041 11.881 4.98815 10.2693 3.05165 9.7803C2.9654 9.4383 2.9999 8.4723 3.0494 8.17905C5.0519 7.69905 5.8409 5.92679 5.34365 4.21979C5.78315 3.88904 6.1784 3.6483 6.69215 3.4398C7.9859 4.75605 9.92839 4.8873 11.2844 3.43455ZM8.99915 5.98979C7.3424 5.98979 5.99915 7.33304 5.99915 8.98979C5.99915 10.6473 7.3424 11.9898 8.99915 11.9898C10.6559 11.9898 11.9991 10.6473 11.9991 8.98979C11.9991 7.33304 10.6559 5.98979 8.99915 5.98979ZM8.99915 7.48979C9.8279 7.48979 10.4991 8.16179 10.4991 8.98979C10.4991 9.81855 9.8279 10.4898 8.99915 10.4898C8.1704 10.4898 7.49915 9.81855 7.49915 8.98979C7.49915 8.16179 8.1704 7.48979 8.99915 7.48979Z"
          }
        ) })
      }
    );
    BarIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 20 20",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Bars", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Union",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M12.5 15.5C12.5 15.7761 12.7239 16 13 16H14C14.2761 16 14.5 15.7761 14.5 15.5V14H16C16.2761 14 16.5 13.7761 16.5 13.5V12.5C16.5 12.2239 16.2761 12 16 12H14.5V2.5C14.5 2.22386 14.2761 2 14 2H13C12.7239 2 12.5 2.22386 12.5 2.5V4H11C10.7239 4 10.5 4.22386 10.5 4.5V5.5C10.5 5.77614 10.7239 6 11 6H12.5V15.5ZM5.5 17.5C5.5 17.7761 5.72386 18 6 18H7C7.27614 18 7.5 17.7761 7.5 17.5V10H9C9.27614 10 9.5 9.77614 9.5 9.5V8.5C9.5 8.22386 9.27614 8 9 8H7.5V4.5C7.5 4.22386 7.27614 4 7 4H6C5.72386 4 5.5 4.22386 5.5 4.5V14H4C3.72386 14 3.5 14.2239 3.5 14.5V15.5C3.5 15.7761 3.72386 16 4 16H5.5V17.5Z"
          }
        ) })
      }
    );
    CandlesIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 20 20",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Candles", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Union",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M7.25 2.5V4H8.5C8.77614 4 9 4.22386 9 4.5V15.5C9 15.7761 8.77614 16 8.5 16H7.25V17.5C7.25 17.7761 7.02614 18 6.75 18H6.25C5.97386 18 5.75 17.7761 5.75 17.5V16H4.5C4.22386 16 4 15.7761 4 15.5V4.5C4 4.22386 4.22386 4 4.5 4H5.75V2.5C5.75 2.22386 5.97386 2 6.25 2H6.75C7.02614 2 7.25 2.22386 7.25 2.5ZM5.5 5.5V14.5H7.5V5.5H5.5ZM14.25 4.5V7H15.5C15.7761 7 16 7.22386 16 7.5V15.5C16 15.7761 15.7761 16 15.5 16H14.25V17.5C14.25 17.7761 14.0261 18 13.75 18H13.25C12.9739 18 12.75 17.7761 12.75 17.5V16H11.5C11.2239 16 11 15.7761 11 15.5V7.5C11 7.22386 11.2239 7 11.5 7H12.75V4.5C12.75 4.22386 12.9739 4 13.25 4H13.75C14.0261 4 14.25 4.22386 14.25 4.5ZM12.5 8.5V14.5H14.5V8.5H12.5Z"
          }
        ) })
      }
    );
    HollowCandlesIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 20 20",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsxs("g", { id: "Hollow Candles", children: [
          /* @__PURE__ */ jsx(
            "path",
            {
              id: "Subtract",
              fillRule: "evenodd",
              clipRule: "evenodd",
              d: "M5.5 14.5V5.5H7.5V14.5H5.5Z",
              fill: "white",
              fillOpacity: "0.12"
            }
          ),
          /* @__PURE__ */ jsx(
            "path",
            {
              id: "Union",
              fillRule: "evenodd",
              clipRule: "evenodd",
              d: "M7.25 2.5V4H8.5C8.77614 4 9 4.22386 9 4.5V15.5C9 15.7761 8.77614 16 8.5 16H7.25V17.5C7.25 17.7761 7.02614 18 6.75 18H6.25C5.97386 18 5.75 17.7761 5.75 17.5V16H4.5C4.22386 16 4 15.7761 4 15.5V4.5C4 4.22386 4.22386 4 4.5 4H5.75V2.5C5.75 2.22386 5.97386 2 6.25 2H6.75C7.02614 2 7.25 2.22386 7.25 2.5ZM5.5 5.5V14.5H7.5V5.5H5.5ZM14.25 4.5V7H15.5C15.7761 7 16 7.22386 16 7.5V15.5C16 15.7761 15.7761 16 15.5 16H14.25V17.5C14.25 17.7761 14.0261 18 13.75 18H13.25C12.9739 18 12.75 17.7761 12.75 17.5V16H11.5C11.2239 16 11 15.7761 11 15.5V7.5C11 7.22386 11.2239 7 11.5 7H12.75V4.5C12.75 4.22386 12.9739 4 13.25 4H13.75C14.0261 4 14.25 4.22386 14.25 4.5ZM12.5 8.5V14.5H14.5V8.5H12.5Z"
            }
          )
        ] })
      }
    );
    LineIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 20 20",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Line", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Vector 16 (Stroke)",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M18.3 2.72288C18.542 2.85596 18.6303 3.15999 18.4972 3.40195L13.495 12.4968C13.0776 13.2557 12.1061 13.5044 11.3754 13.0394L7.19878 10.3816L2.97738 17.558C2.83737 17.796 2.53092 17.8754 2.2929 17.7354L1.43097 17.2284C1.19295 17.0884 1.1135 16.7819 1.25351 16.5439L5.73746 8.92122C6.16792 8.18943 7.1194 7.96045 7.83567 8.41626L11.9995 11.0659L16.7448 2.43811C16.8778 2.19615 17.1819 2.10788 17.4238 2.24096L18.3 2.72288Z"
          }
        ) })
      }
    );
    AreaIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 20 20",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsxs("g", { id: "Area", children: [
          /* @__PURE__ */ jsx(
            "path",
            {
              id: "Vector 16 (Stroke)",
              fillRule: "evenodd",
              clipRule: "evenodd",
              d: "M18.3 2.72288C18.542 2.85596 18.6303 3.15999 18.4972 3.40195L13.495 12.4968C13.0776 13.2557 12.1061 13.5044 11.3754 13.0394L7.19878 10.3816L2.97738 17.558C2.83737 17.796 2.53092 17.8754 2.2929 17.7354L1.43097 17.2284C1.19295 17.0884 1.1135 16.7819 1.25351 16.5439L5.73746 8.92122C6.16792 8.18943 7.1194 7.96045 7.83567 8.41626L11.9995 11.0659L16.7448 2.43811C16.8778 2.19615 17.1819 2.10788 17.4238 2.24096L18.3 2.72288Z"
            }
          ),
          /* @__PURE__ */ jsx(
            "path",
            {
              id: "Subtract",
              fillRule: "evenodd",
              clipRule: "evenodd",
              d: "M2.7608 17.7562C2.84602 17.899 3.00106 18 3.19016 18H16.362C17.4665 18 18.362 17.1046 18.362 16V3.94663C18.362 3.86618 18.3453 3.79397 18.3162 3.73108L13.495 12.4968C13.0776 13.2557 12.1061 13.5044 11.3754 13.0394L7.19881 10.3816L2.9774 17.558C2.92465 17.6477 2.84826 17.7148 2.7608 17.7562Z",
              fill: "white",
              fillOpacity: "0.12"
            }
          )
        ] })
      }
    );
    BaseLineIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 20 20",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Baseline", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Union",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M6.05938 2.93544C6.55642 1.71198 8.27766 1.68139 8.81785 2.88642L11.4361 8.72723H9.24438L7.4717 4.77279L5.86521 8.72723H3.70647L6.05938 2.93544ZM13.9282 14.2865L13.0051 12.2272H10.8133L12.5586 16.1204C13.088 17.3014 14.7641 17.3026 15.2952 16.1224L17.048 12.2272H14.8549L13.9282 14.2865ZM2.28459 12.2272H4.44333L1.85293 18.6036L0 17.8509L2.28459 12.2272ZM16.4299 8.72723H18.623L19.3384 7.13759L17.5145 6.31686L16.4299 8.72723ZM1.42645 10.2272C1.42645 9.95109 1.65031 9.72723 1.92645 9.72723H2.92645C3.20259 9.72723 3.42645 9.95109 3.42645 10.2272V10.7272C3.42645 11.0034 3.20259 11.2272 2.92645 11.2272H1.92645C1.65031 11.2272 1.42645 11.0034 1.42645 10.7272V10.2272ZM4.42645 10.2272C4.42645 9.95109 4.65031 9.72723 4.92645 9.72723H5.92645C6.20259 9.72723 6.42645 9.95109 6.42645 10.2272V10.7272C6.42645 11.0034 6.20259 11.2272 5.92645 11.2272H4.92645C4.65031 11.2272 4.42645 11.0034 4.42645 10.7272V10.2272ZM7.92645 9.72723C7.65031 9.72723 7.42645 9.95109 7.42645 10.2272V10.7272C7.42645 11.0034 7.65031 11.2272 7.92645 11.2272H8.92645C9.20259 11.2272 9.42645 11.0034 9.42645 10.7272V10.2272C9.42645 9.95109 9.20259 9.72723 8.92645 9.72723H7.92645ZM10.4265 10.2272C10.4265 9.95109 10.6503 9.72723 10.9265 9.72723H11.9265C12.2026 9.72723 12.4265 9.95109 12.4265 10.2272V10.7272C12.4265 11.0034 12.2026 11.2272 11.9265 11.2272H10.9265C10.6503 11.2272 10.4265 11.0034 10.4265 10.7272V10.2272ZM13.9265 9.72723C13.6503 9.72723 13.4265 9.95109 13.4265 10.2272V10.7272C13.4265 11.0034 13.6503 11.2272 13.9265 11.2272H14.9265C15.2026 11.2272 15.4265 11.0034 15.4265 10.7272V10.2272C15.4265 9.95109 15.2026 9.72723 14.9265 9.72723H13.9265ZM16.4265 10.2272C16.4265 9.95109 16.6503 9.72723 16.9265 9.72723H17.9265C18.2026 9.72723 18.4265 9.95109 18.4265 10.2272V10.7272C18.4265 11.0034 18.2026 11.2272 17.9265 11.2272H16.9265C16.6503 11.2272 16.4265 11.0034 16.4265 10.7272V10.2272Z"
          }
        ) })
      }
    );
    UnSelectIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 12 12",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Unselected", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Vector",
            d: "M6.00684 0.999023C3.24544 0.999023 1.00684 3.23752 1.00684 5.99902C1.00684 8.76051 3.24544 10.999 6.00684 10.999C8.76834 10.999 11.0068 8.76051 11.0068 5.99902C11.0068 3.23752 8.76834 0.999023 6.00684 0.999023ZM6.00684 1.99902C8.21584 1.99902 10.0068 3.79002 10.0068 5.99902C10.0068 8.20801 8.21584 9.99901 6.00684 9.99901C3.79769 9.99901 2.00684 8.20801 2.00684 5.99902C2.00684 3.79002 3.79769 1.99902 6.00684 1.99902Z",
            fill: "white",
            fillOpacity: "0.2"
          }
        ) })
      }
    );
    SelectedIcon = (props) => /* @__PURE__ */ jsx(
      "svg",
      {
        viewBox: "0 0 12 12",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ...props,
        children: /* @__PURE__ */ jsx("g", { id: "Selected-Multiple Choices-fill", children: /* @__PURE__ */ jsx(
          "path",
          {
            id: "Subtract",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M1.00684 5.99902C1.00684 3.23752 3.24544 0.999023 6.00684 0.999023C8.76834 0.999023 11.0068 3.23752 11.0068 5.99902C11.0068 8.76051 8.76834 10.999 6.00684 10.999C3.24544 10.999 1.00684 8.76051 1.00684 5.99902ZM8.49243 3.65563C8.60602 3.54726 8.76156 3.49957 8.91068 3.49957C9.05981 3.49957 9.21534 3.54726 9.32893 3.65563C9.5567 3.87231 9.5567 4.23649 9.32893 4.45317L5.25362 8.33706C5.02644 8.55374 4.6443 8.55374 4.41712 8.33706L2.67014 6.67253C2.44296 6.45585 2.44296 6.09161 2.67014 5.87499C2.89791 5.65831 3.28004 5.65831 3.50722 5.87499L4.83537 7.1407L8.49243 3.65563Z",
            fill: "#608CFF"
          }
        ) })
      }
    );
  }
});

// src/components/lineType/index.tsx
var lineType_exports = {};
__export(lineType_exports, {
  default: () => lineType_default
});
var LineType, lineType_default;
var init_lineType = __esm({
  "src/components/lineType/index.tsx"() {
    init_icons();
    LineType = (props) => {
      const [open, setOpen] = useState(false);
      const { t } = useTranslation();
      const lineTypeList = useMemo(() => {
        return [
          {
            icon: /* @__PURE__ */ jsx(BarIcon, { fill: "currentColor", className: "oui-w-5 oui-h-5" }),
            label: t("tradingView.lineType.bars"),
            value: "0"
          },
          {
            icon: /* @__PURE__ */ jsx(CandlesIcon, { fill: "currentColor", className: "oui-w-5 oui-h-5" }),
            label: t("tradingView.lineType.candles"),
            value: "1"
          },
          {
            icon: /* @__PURE__ */ jsx(HollowCandlesIcon, { fill: "currentColor", className: "oui-w-5 oui-h-5" }),
            label: t("tradingView.lineType.hollowCandles"),
            value: "9"
          },
          {
            icon: /* @__PURE__ */ jsx(LineIcon, { fill: "currentColor", className: "oui-w-5 oui-h-5" }),
            label: t("tradingView.lineType.line"),
            value: "2"
          },
          {
            icon: /* @__PURE__ */ jsx(AreaIcon, { fill: "currentColor", className: "oui-w-5 oui-h-5" }),
            label: t("tradingView.lineType.area"),
            value: "3"
          },
          {
            icon: /* @__PURE__ */ jsx(BaseLineIcon, { fill: "currentColor", className: "oui-w-5 oui-h-5" }),
            label: t("tradingView.lineType.baseline"),
            value: "10"
          }
        ];
      }, [t]);
      const currentLineTypeIcon = useMemo(() => {
        const data = lineTypeList.find((item) => item.value === props.lineType);
        if (data) {
          return data.icon;
        }
        return lineTypeList[1].icon;
      }, [props.lineType]);
      return /* @__PURE__ */ jsxs(DropdownMenuRoot, { open, onOpenChange: setOpen, children: [
        /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsx(
          Box,
          {
            className: cn(
              "oui-w-[18px] oui-h-[18px] oui-cursor-pointer oui-text-base-contrast-36 hover:oui-text-base-contrast-80",
              open && "oui-text-base-contrast-80"
            ),
            children: currentLineTypeIcon
          }
        ) }),
        /* @__PURE__ */ jsx(DropdownMenuPortal, { children: /* @__PURE__ */ jsx(
          DropdownMenuContent,
          {
            onCloseAutoFocus: (e) => e.preventDefault(),
            onClick: (e) => e.stopPropagation(),
            align: "start",
            sideOffset: 20,
            className: "oui-bg-base-8",
            children: /* @__PURE__ */ jsx(
              Flex,
              {
                direction: "column",
                gap: 4,
                px: 5,
                py: 5,
                width: 240,
                justify: "start",
                itemAlign: "start",
                children: lineTypeList.map((item) => /* @__PURE__ */ jsxs(
                  Flex,
                  {
                    justify: "start",
                    itemAlign: "center",
                    gap: 2,
                    className: cn(
                      "oui-text-base-contrast-36 oui-cursor-pointer oui-w-full hover:oui-text-base-contrast",
                      props.lineType === item.value && "oui-text-base-contrast"
                    ),
                    onClick: () => props.changeLineType(item.value),
                    children: [
                      item.icon,
                      /* @__PURE__ */ jsx(Text, { className: "oui-text-sm", children: item.label })
                    ]
                  },
                  item.value
                ))
              }
            )
          }
        ) })
      ] });
    };
    lineType_default = LineType;
  }
});

// src/components/timeInterval/index.tsx
var timeInterval_exports = {};
__export(timeInterval_exports, {
  MobileTimeInterval: () => MobileTimeInterval,
  TimeInterval: () => TimeInterval
});
var useMobileTimeIntervalMoreMap, TimeInterval, DesktopTimeInterval, MobileTimeInterval, DropDownTimeInterval;
var init_timeInterval = __esm({
  "src/components/timeInterval/index.tsx"() {
    init_icons();
    useMobileTimeIntervalMoreMap = () => {
      const { t } = useTranslation();
      const mobileTimeIntervalMoreMap = useMemo(() => {
        return [
          [
            {
              value: "3",
              label: t("tradingView.timeInterval.3m")
            },
            {
              value: "5",
              label: t("tradingView.timeInterval.5m")
            },
            {
              value: "30",
              label: t("tradingView.timeInterval.30m")
            },
            {
              value: "120",
              label: t("tradingView.timeInterval.2h")
            }
          ],
          [
            {
              value: "360",
              label: t("tradingView.timeInterval.6h")
            },
            {
              value: "720",
              label: t("tradingView.timeInterval.12h")
            },
            {
              value: "3d",
              label: t("tradingView.timeInterval.3d")
            },
            {
              value: "1M",
              label: t("tradingView.timeInterval.1M")
            }
          ]
        ];
      }, [t]);
      return { mobileTimeIntervalMoreMap };
    };
    TimeInterval = (props) => {
      const isMobile = useMediaQuery(MEDIA_TABLET);
      if (isMobile) {
        return /* @__PURE__ */ jsx(MobileTimeInterval, { ...props });
      }
      return /* @__PURE__ */ jsx(DesktopTimeInterval, { ...props });
    };
    DesktopTimeInterval = (props) => {
      const { t } = useTranslation();
      const timeIntervalMap = useMemo(() => {
        return [
          {
            value: "1",
            label: t("tradingView.timeInterval.1m")
          },
          {
            value: "3",
            label: t("tradingView.timeInterval.3m")
          },
          {
            value: "5",
            label: t("tradingView.timeInterval.5m")
          },
          {
            value: "15",
            label: t("tradingView.timeInterval.15m")
          },
          {
            value: "30",
            label: t("tradingView.timeInterval.30m")
          },
          {
            value: "60",
            label: t("tradingView.timeInterval.1h")
          },
          {
            value: "240",
            label: t("tradingView.timeInterval.4h")
          },
          {
            value: "720",
            label: t("tradingView.timeInterval.12h")
          },
          {
            value: "1D",
            label: t("tradingView.timeInterval.1d")
          },
          {
            value: "1W",
            label: t("tradingView.timeInterval.1w")
          },
          {
            value: "1M",
            label: t("tradingView.timeInterval.1M")
          }
        ];
      }, [t]);
      return /* @__PURE__ */ jsx(
        "div",
        {
          className: cn(
            "oui-text-2xs oui-text-base-contrast-36 oui-flex oui-gap-[2px] oui-items-center oui-mr-3 oui-font-semibold",
            "oui-overflow-hidden"
          ),
          children: timeIntervalMap.map((item) => /* @__PURE__ */ jsx(
            "div",
            {
              className: cn(
                "oui-cursor-pointer oui-px-2",
                "hover:oui-text-base-contrast-80",
                "oui-break-normal oui-whitespace-nowrap",
                props.interval === item.value && "oui-text-base-contrast-80 oui-bg-white/[.06] oui-rounded"
              ),
              id: item.value,
              onClick: () => props.changeInterval(item.value),
              children: item.label
            },
            item.value
          ))
        }
      );
    };
    MobileTimeInterval = (props) => {
      const { t } = useTranslation();
      const mobileTimeIntervalDefaultMap = useMemo(() => {
        return [
          {
            value: "1",
            label: t("tradingView.timeInterval.1m")
          },
          {
            value: "15",
            label: t("tradingView.timeInterval.15m")
          },
          {
            value: "60",
            label: t("tradingView.timeInterval.1h")
          },
          {
            value: "240",
            label: t("tradingView.timeInterval.4h")
          },
          {
            value: "1D",
            label: t("tradingView.timeInterval.1d")
          },
          {
            value: "1W",
            label: t("tradingView.timeInterval.1w")
          }
        ];
      }, [t]);
      const { mobileTimeIntervalMoreMap } = useMobileTimeIntervalMoreMap();
      const currentIntervalIsInExpand = useMemo(() => {
        for (const row of mobileTimeIntervalMoreMap) {
          for (const item of row) {
            if (item.value === props.interval) {
              return item.label;
            }
          }
        }
        return null;
      }, [props.interval, mobileTimeIntervalMoreMap]);
      return /* @__PURE__ */ jsxs(
        Flex,
        {
          justify: "start",
          itemAlign: "center",
          gap: 3,
          className: cn(
            "oui-text-2xs oui-text-base-contrast-36",
            "oui-overflow-hidden"
          ),
          children: [
            /* @__PURE__ */ jsx("div", { className: " oui-flex oui-gap-1 oui-items-center oui-mr-3 oui-font-semibold", children: mobileTimeIntervalDefaultMap.map((item) => /* @__PURE__ */ jsx(
              "div",
              {
                className: cn(
                  "oui-px-2",
                  "oui-break-normal oui-whitespace-nowrap",
                  props.interval === item.value && "oui-text-base-contrast-80 oui-bg-white/[.06] oui-rounded"
                ),
                onClick: () => props.changeInterval(item.value),
                children: item.label
              },
              item.value
            )) }),
            /* @__PURE__ */ jsx(DropDownTimeInterval, { ...props, children: currentIntervalIsInExpand ? /* @__PURE__ */ jsx("div", { className: "oui-text-base-contrast-80", children: currentIntervalIsInExpand }) : /* @__PURE__ */ jsx(Text, { className: "oui-break-normal oui-whitespace-nowrap", children: t("tradingView.timeInterval.more") }) })
          ]
        }
      );
    };
    DropDownTimeInterval = (props) => {
      const [open, setOpen] = React3.useState(false);
      const { mobileTimeIntervalMoreMap } = useMobileTimeIntervalMoreMap();
      return /* @__PURE__ */ jsxs(DropdownMenuRoot, { open, onOpenChange: setOpen, children: [
        /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsxs("div", { className: "oui-flex oui-justify-start oui-items-center oui-gap-0.5", children: [
          props.children,
          /* @__PURE__ */ jsx(
            CaretIcon,
            {
              className: cn(
                "oui-w-3 oui-h-3",
                open && "oui-text-base-contrast-80 oui-rotate-180"
              )
            }
          )
        ] }) }),
        /* @__PURE__ */ jsx(DropdownMenuPortal, { children: /* @__PURE__ */ jsx(
          DropdownMenuContent,
          {
            onCloseAutoFocus: (e) => e.preventDefault(),
            onClick: (e) => e.stopPropagation(),
            align: "start",
            alignOffset: 0,
            sideOffset: 0,
            className: cn(
              "oui-markets-dropdown-menu-content oui-bg-base-9 oui-w-screen oui-flex oui-flex-col oui-gap-2 oui-p-3"
            ),
            children: mobileTimeIntervalMoreMap.map((row, id) => /* @__PURE__ */ jsx("div", { className: "oui-flex oui-gap-2", children: row.map((item) => /* @__PURE__ */ jsx(
              "div",
              {
                className: cn(
                  "oui-w-full  oui-text-2xs oui-flex oui-items-center oui-justify-center oui-h-6  oui-rounded",
                  item.value === props.interval ? "oui-text-base-contrast oui-bg-primary-darken" : "oui-text-base-contrast-36 oui-bg-base-5"
                ),
                onClick: () => {
                  props.changeInterval(item.value);
                },
                children: /* @__PURE__ */ jsx("div", { children: item.label })
              },
              item.value
            )) }, id))
          }
        ) })
      ] });
    };
  }
});
var DesktopDisplayControl;
var init_displayControl_desktop = __esm({
  "src/components/displayControl/displayControl.desktop.tsx"() {
    init_icons();
    DesktopDisplayControl = (props) => {
      const { displayControlState, changeDisplayControlState } = props;
      const [open, setOpen] = useState(false);
      const { t } = useTranslation();
      const displayControlMap = useMemo(() => {
        return [
          {
            label: t("common.position"),
            id: "position"
          },
          {
            label: t("tradingView.displayControl.buySell"),
            id: "buySell"
          },
          {
            label: t("tradingView.displayControl.limitOrders"),
            id: "limitOrders"
          },
          {
            label: t("tradingView.displayControl.stopOrders"),
            id: "stopOrders"
          },
          {
            label: t("common.tpsl"),
            id: "tpsl"
          },
          {
            label: t("tpsl.positionTpsl"),
            id: "positionTpsl"
          },
          {
            label: t("orderEntry.orderType.trailingStop"),
            id: "trailingStop"
          }
        ];
      }, [t]);
      return /* @__PURE__ */ jsx(Fragment, { children: /* @__PURE__ */ jsxs(DropdownMenuRoot, { open, onOpenChange: setOpen, children: [
        /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
          Flex,
          {
            justify: "start",
            itemAlign: "center",
            className: "oui-gap-[2px] oui-cursor-pointer oui-text-base-contrast-36  hover:oui-text-base-contrast-80",
            children: [
              /* @__PURE__ */ jsx(
                DisplaySettingIcon,
                {
                  className: cn(
                    "oui-w-[18px] oui-h-[18px] ",
                    open && "oui-text-base-contrast-80"
                  )
                }
              ),
              /* @__PURE__ */ jsx(
                CaretIcon,
                {
                  className: cn(
                    "oui-w-3 oui-h-3",
                    open && "oui-text-base-contrast-80 oui-rotate-180"
                  )
                }
              )
            ]
          }
        ) }),
        /* @__PURE__ */ jsx(DropdownMenuPortal, { children: /* @__PURE__ */ jsx(
          DropdownMenuContent,
          {
            onCloseAutoFocus: (e) => e.preventDefault(),
            onClick: (e) => e.stopPropagation(),
            align: "start",
            className: "oui-bg-base-8",
            children: /* @__PURE__ */ jsx(
              Flex,
              {
                direction: "column",
                gap: 4,
                px: 5,
                py: 5,
                width: 240,
                justify: "start",
                itemAlign: "start",
                children: displayControlMap.map((item) => /* @__PURE__ */ jsxs(
                  Flex,
                  {
                    justify: "between",
                    itemAlign: "center",
                    className: "oui-w-full",
                    children: [
                      /* @__PURE__ */ jsx(
                        Text,
                        {
                          className: cn(
                            "oui-text-sm oui-text-base-contrast-80",
                            !displayControlState[item.id] && "oui-text-base-contrast-36"
                          ),
                          children: item.label
                        }
                      ),
                      /* @__PURE__ */ jsx(
                        Switch,
                        {
                          className: "oui-h-4 oui-w-8",
                          checked: displayControlState[item.id],
                          onCheckedChange: (checked) => {
                            changeDisplayControlState({
                              ...displayControlState,
                              [item.id]: checked
                            });
                          }
                        }
                      )
                    ]
                  },
                  item.id
                ))
              }
            )
          }
        ) })
      ] }) });
    };
  }
});
var MobileDisplayControl;
var init_displayControl_mobile = __esm({
  "src/components/displayControl/displayControl.mobile.tsx"() {
    init_icons();
    MobileDisplayControl = (props) => {
      const [open, setOpen] = useState(false);
      const { t } = useTranslation();
      const mobileDisplayControlMap = useMemo(() => {
        return [
          [
            {
              label: t("common.position"),
              id: "position"
            },
            {
              label: t("tradingView.displayControl.limitOrders"),
              id: "limitOrders"
            }
          ],
          [
            {
              label: t("tradingView.displayControl.stopOrders"),
              id: "stopOrders"
            },
            {
              label: t("common.tpsl"),
              id: "tpsl"
            }
          ],
          [
            {
              label: t("tpsl.positionTpsl"),
              id: "positionTpsl"
            },
            {
              label: t("tradingView.displayControl.buySell"),
              id: "buySell"
            }
          ],
          [
            {
              label: t("orderEntry.orderType.trailingStop"),
              id: "trailingStop"
            },
            // placeholder
            {}
          ]
        ];
      }, [t]);
      return /* @__PURE__ */ jsxs(DropdownMenuRoot, { open, onOpenChange: setOpen, children: [
        /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsxs(
          "div",
          {
            className: cn(
              "oui-flex oui-items-center oui-justify-center oui-gap-0.5 oui-text-base-contrast-36",
              open && "oui-text-base-contrast-8"
            ),
            children: [
              /* @__PURE__ */ jsx(
                DisplaySettingIcon,
                {
                  className: cn(
                    "oui-size-[18px] ",
                    open && "oui-text-base-contrast-80"
                  )
                }
              ),
              /* @__PURE__ */ jsx(
                CaretIcon,
                {
                  className: cn(
                    "oui-size-3",
                    open && "oui-rotate-180 oui-text-base-contrast-80"
                  )
                }
              )
            ]
          }
        ) }),
        /* @__PURE__ */ jsx(DropdownMenuPortal, { children: /* @__PURE__ */ jsx(
          DropdownMenuContent,
          {
            onCloseAutoFocus: (e) => e.preventDefault(),
            onClick: (e) => e.stopPropagation(),
            align: "start",
            alignOffset: 0,
            sideOffset: 0,
            className: cn(
              "oui-tradingview-display-control-dropdown-menu-content oui-flex oui-w-screen oui-flex-col oui-gap-2 oui-bg-base-9 oui-p-3"
            ),
            children: mobileDisplayControlMap.map((row, id) => /* @__PURE__ */ jsx("div", { className: "oui-flex oui-gap-2", children: row.map((item, index) => {
              return /* @__PURE__ */ jsx(
                "div",
                {
                  className: cn(
                    "oui-flex oui-h-6 oui-w-full oui-items-center oui-justify-between",
                    "oui-rounded oui-px-2 oui-text-2xs",
                    item.id && "oui-bg-base-5",
                    props.displayControlState[item.id] ? "oui-text-base-contrast" : "oui-text-base-contrast-36"
                  ),
                  onClick: () => {
                    if (!item.id) {
                      return;
                    }
                    props.changeDisplayControlState({
                      ...props.displayControlState,
                      [item.id]: !props.displayControlState[item.id]
                    });
                  },
                  children: item.id && /* @__PURE__ */ jsxs(Fragment, { children: [
                    /* @__PURE__ */ jsx("div", { children: item.label }),
                    props.displayControlState[item.id] ? /* @__PURE__ */ jsx(SelectedIcon, { className: "oui-size-3" }) : /* @__PURE__ */ jsx(UnSelectIcon, { className: "oui-size-3" })
                  ] })
                },
                item.id || index
              );
            }) }, id))
          }
        ) })
      ] });
    };
  }
});

// src/components/displayControl/index.tsx
var displayControl_exports = {};
__export(displayControl_exports, {
  DesktopDisplayControl: () => DesktopDisplayControl,
  MobileDisplayControl: () => MobileDisplayControl
});
var init_displayControl = __esm({
  "src/components/displayControl/index.tsx"() {
    init_displayControl_desktop();
    init_displayControl_mobile();
  }
});

// src/utils/chart.util.ts
var upColor = "#008676";
var downColor = "#D92D6B";
var chartBG = "#131519";
var pnlUpColor = "#00B49E";
var pnlDownColor = "#FF447C";
var pnlZoreColor = "#333948";
var textColor = "#FFFFFF";
var qtyTextColor = "#F4F7F9";
var font = "regular 11px Manrope";
var defaultColorConfig = {
  upColor,
  downColor,
  chartBG,
  pnlUpColor,
  pnlDownColor,
  pnlZoreColor,
  textColor,
  qtyTextColor,
  font,
  volumeUpColor: "#0C3E3A",
  volumeDownColor: "#5A1E36",
  closeIcon: "rgba(255, 255, 255, 0.8)"
};
var getOveriides = (colorConfig, isMobile) => {
  const overrides = {
    "paneProperties.background": colorConfig.chartBG,
    // "paneProperties.background": "#ffff00",
    // "mainSeriesProperties.style": 1,
    "paneProperties.backgroundType": "solid",
    // "paneProperties.background": "#151822",
    "mainSeriesProperties.candleStyle.upColor": colorConfig.upColor,
    "mainSeriesProperties.candleStyle.downColor": colorConfig.downColor,
    "mainSeriesProperties.candleStyle.borderColor": colorConfig.upColor,
    "mainSeriesProperties.candleStyle.borderUpColor": colorConfig.upColor,
    "mainSeriesProperties.candleStyle.borderDownColor": colorConfig.downColor,
    "mainSeriesProperties.candleStyle.wickUpColor": colorConfig.upColor,
    "mainSeriesProperties.candleStyle.wickDownColor": colorConfig.downColor,
    "paneProperties.separatorColor": "#2B2833",
    "paneProperties.vertGridProperties.color": "#26232F",
    "paneProperties.horzGridProperties.color": "#26232F",
    "scalesProperties.fontSize": isMobile ? 8 : 12,
    "scalesProperties.textColor": "#97969B",
    "paneProperties.legendProperties.showSeriesTitle": isMobile ? false : true,
    "mainSeriesProperties.statusViewStyle.symbolTextSource": "ticker"
  };
  const studiesOverrides = {
    "volume.volume.color.0": colorConfig.volumeDownColor,
    "volume.volume.color.1": colorConfig.volumeUpColor
  };
  return {
    overrides,
    studiesOverrides
  };
};
var EXCHANGE = "Orderly";
var withoutExchangePrefix = (symbol) => symbol.includes(":") ? symbol.split(":")[1] : symbol;
var withExchangePrefix = (symbol) => symbol.startsWith(`${EXCHANGE}:`) ? symbol : `${EXCHANGE}:${symbol}`;
var getBrokerAdapter = (host, broker) => {
  let _symbolInfo;
  const getOrderCombinationType = (orderType) => {
    return [
      null,
      "limit",
      "MARKET" /* MARKET */,
      "STOP_MARKET" /* STOP_MARKET */,
      "STOP_LIMIT" /* STOP_LIMIT */
    ][orderType];
  };
  return {
    symbolInfo: async (symbol2) => {
      _symbolInfo = broker.getSymbolInfo(withoutExchangePrefix(symbol2));
      return {
        qty: {
          min: _symbolInfo?.baseMin ?? 0,
          max: _symbolInfo?.baseMax ?? 0,
          step: _symbolInfo?.baseTick ?? 0
        },
        pipValue: 0,
        pipSize: _symbolInfo?.quoteTick ?? 0,
        minTick: _symbolInfo?.quoteTick ?? 0,
        description: ""
      };
    },
    placeOrder: async (order) => {
      const side = ["SELL" /* SELL */, "BUY" /* BUY */][+(order.side > 0)];
      const quantity = order.qty.toString();
      (order.limitPrice ?? 0).toString();
      (order.stopPrice ?? 0).toString();
      const symbol2 = withoutExchangePrefix(order.symbol);
      const orderCombinationType = getOrderCombinationType(order.type);
      if (orderCombinationType === "MARKET" /* MARKET */) {
        broker.sendMarketOrder({ side, order_quantity: quantity, symbol: symbol2, order_type: OrderType.MARKET });
      }
    },
    orders: () => [],
    positions: () => [],
    executions: () => [],
    connectionStatus: () => 1,
    chartContextMenuActions: (context) => host.defaultContextMenuActions(context),
    isTradable: async () => true,
    accountManagerInfo: () => ({}),
    currentAccount: () => "1",
    accountsMetainfo: async () => [{ id: "1" }],
    remove: () => host?.silentOrdersPlacement().unsubscribe()
  };
};
var getBrokerAdapter_default = getBrokerAdapter;
function logMessage(message) {
}
function getErrorMessage(error) {
  if (error === void 0) {
    return "";
  }
  if (typeof error === "string") {
    return error;
  }
  return error.message;
}

// src/tradingviewAdapter/datafeed/history-provider.ts
var HISTORY_PATH = "tv/history";
var KLINE_HISTORY_PATH = "v1/tv/kline_history";
var _HistoryProvider = class _HistoryProvider {
  constructor(datafeedUrl, requester, limitedServerResponse) {
    this._klinePreference = /* @__PURE__ */ new Map();
    this._datafeedUrl = datafeedUrl;
    this._requester = requester;
    this._limitedServerResponse = limitedServerResponse;
  }
  /**
   * Build request parameters for history API calls
   * @param symbolInfo - Symbol information
   * @param resolution - Resolution string
   * @param periodParams - Period parameters with optional countback
   * @returns Request parameters object
   */
  _buildRequestParams(symbolInfo, resolution, periodParams) {
    const requestParams = {
      symbol: symbolInfo.ticker || "",
      resolution,
      from: periodParams.from,
      to: periodParams.to
    };
    const countBack = Math.min(periodParams.countBack ?? 0, 1e3);
    if (periodParams.countBack !== void 0) {
      requestParams.countback = countBack;
    }
    if (symbolInfo.currency_code !== void 0) {
      requestParams.currencyCode = symbolInfo.currency_code;
    }
    if (symbolInfo.unit_id !== void 0) {
      requestParams.unitId = symbolInfo.unit_id;
    }
    return requestParams;
  }
  getBars(symbolInfo, resolution, periodParams) {
    const requestParams = this._buildRequestParams(
      symbolInfo,
      resolution,
      periodParams
    );
    const preferenceKey = this._getPreferenceKey(symbolInfo, resolution);
    const prefersKline = this._klinePreference.get(preferenceKey) === true;
    const countBack = Math.min(periodParams.countBack ?? 0, 1e3);
    return new Promise(
      async (resolve, reject) => {
        try {
          let result;
          let usedHistoryResult = false;
          if (prefersKline) {
            result = await this._requestKlineHistory(
              this._buildKlineParams(requestParams, countBack)
            );
          } else {
            const initialResponse = await this._requestHistory(requestParams);
            result = this._processHistoryResponse(initialResponse);
            usedHistoryResult = true;
            const needsFallback = this._shouldFallbackToKline(
              initialResponse,
              countBack
            );
            if (needsFallback) {
              const earliestTime = this._getEarliestTime(result.bars);
              if (earliestTime !== null) {
                const remainingCountBack = Math.max(
                  0,
                  countBack - result.bars.length
                );
                const mergedResult = await this._tryKlineFallbackWithMerge(
                  requestParams,
                  remainingCountBack,
                  earliestTime,
                  result
                );
                if (mergedResult !== null) {
                  result = mergedResult;
                  usedHistoryResult = false;
                  this._klinePreference.set(preferenceKey, true);
                } else {
                  this._klinePreference.set(preferenceKey, false);
                }
              } else {
                const klineResult = await this._tryKlineFallback(
                  requestParams,
                  countBack
                );
                if (klineResult !== null) {
                  result = klineResult;
                  usedHistoryResult = false;
                  this._klinePreference.set(preferenceKey, true);
                } else {
                  this._klinePreference.set(preferenceKey, false);
                }
              }
            } else {
              this._klinePreference.set(preferenceKey, false);
            }
          }
          if (usedHistoryResult && this._limitedServerResponse) {
            await this._processTruncatedResponse(result, { ...requestParams });
          }
          resolve(result);
        } catch (e) {
          const error = e instanceof Error ? e : typeof e === "string" ? e : void 0;
          const reasonString = getErrorMessage(error);
          reject(reasonString || "Error");
        }
      }
    );
  }
  /**
   * Request kline history using KLINE_HISTORY_PATH endpoint
   * Handles rate limiting (429 errors) by waiting and retrying
   * @param requestParams - Request parameters
   * @param retryCount - Current retry attempt (internal use)
   * @returns Processed history response
   */
  async _requestKlineHistory(requestParams, retryCount = 0) {
    const maxRetries = 5;
    const baseRetryDelay = 2500;
    const maxRetryDelay = 1e4;
    const params = {
      ...requestParams,
      resolution: this._mapToKlineHistoryResolution(
        requestParams.resolution
      )
    };
    const urlPath = this._buildUrlWithParams(KLINE_HISTORY_PATH, params);
    const options = { credentials: "same-origin" };
    const requesterWithHeaders = this._requester;
    if (requesterWithHeaders._headers !== void 0) {
      options.headers = requesterWithHeaders._headers;
    }
    let response;
    try {
      response = await fetch(`${this._datafeedUrl}/${urlPath}`, options);
    } catch (error) {
      throw error;
    }
    if (response.status === 429) {
      if (retryCount >= maxRetries) {
        throw new Error(
          `Rate limit exceeded: Maximum retry attempts (${maxRetries}) reached`
        );
      }
      const retryDelay = Math.min(
        baseRetryDelay * Math.pow(2, retryCount),
        maxRetryDelay
      );
      const retryAfter = response.headers.get("Retry-After");
      const delay = retryAfter ? parseInt(retryAfter, 10) * 1e3 : retryDelay;
      await new Promise((resolve) => setTimeout(resolve, delay));
      return this._requestKlineHistory(requestParams, retryCount + 1);
    }
    if (!response.ok) {
      const errorData = await response.json().catch(() => ({
        message: response.statusText
      }));
      throw new Error(
        errorData.message || errorData.errmsg || response.statusText
      );
    }
    let data;
    try {
      data = await response.json();
    } catch {
      throw new Error("Failed to parse response JSON");
    }
    return this._processHistoryResponse(data);
  }
  /**
   * Process truncated response by making follow-up requests if needed
   * @param result - Current result with bars
   * @param requestParams - Request parameters
   */
  async _processTruncatedResponse(result, requestParams) {
    let lastResultLength = result.bars.length;
    try {
      while (this._limitedServerResponse && this._limitedServerResponse.maxResponseLength > 0 && this._limitedServerResponse.maxResponseLength === lastResultLength && requestParams.from < requestParams.to) {
        if (requestParams.countback) {
          requestParams.countback = requestParams.countback - lastResultLength;
        }
        if (this._limitedServerResponse.expectedOrder === "earliestFirst") {
          requestParams.from = Math.round(
            result.bars[result.bars.length - 1].time / 1e3
          );
        } else {
          requestParams.to = Math.round(result.bars[0].time / 1e3);
        }
        const followupResponse = await this._requester.sendRequest(
          this._datafeedUrl,
          HISTORY_PATH,
          requestParams
        );
        const followupResult = this._processHistoryResponse(followupResponse);
        lastResultLength = followupResult.bars.length;
        if (this._limitedServerResponse.expectedOrder === "earliestFirst") {
          if (followupResult.bars[0].time === result.bars[result.bars.length - 1].time) {
            followupResult.bars.shift();
          }
          result.bars.push(...followupResult.bars);
        } else {
          if (followupResult.bars[followupResult.bars.length - 1].time === result.bars[0].time) {
            followupResult.bars.pop();
          }
          result.bars.unshift(...followupResult.bars);
        }
      }
    } catch (e) {
      if (e instanceof Error || typeof e === "string") {
        getErrorMessage(e);
      }
    }
  }
  _processHistoryResponse(response) {
    if (response.s !== "ok" && response.s !== "no_data") {
      throw new Error(response.errmsg);
    }
    const bars = [];
    const meta = {
      noData: false
    };
    if (response.s === "no_data") {
      meta.noData = true;
      meta.nextTime = response.nextTime;
    } else {
      const volumePresent = response.v !== void 0;
      const ohlPresent = response.o !== void 0;
      for (let i = 0; i < response.t.length; ++i) {
        const barValue = {
          time: response.t[i] * 1e3,
          close: parseFloat(response.c[i]),
          open: parseFloat(response.c[i]),
          high: parseFloat(response.c[i]),
          low: parseFloat(response.c[i])
        };
        if (ohlPresent) {
          barValue.open = parseFloat(
            response.o[i]
          );
          barValue.high = parseFloat(
            response.h[i]
          );
          barValue.low = parseFloat(response.l[i]);
        }
        if (volumePresent) {
          const volume = parseFloat(response.v[i]);
          if (volume > 0) {
            barValue.volume = volume;
          } else {
            barValue.volume = 1e-10;
          }
        }
        bars.push(barValue);
      }
    }
    return {
      bars,
      meta
    };
  }
  /**
   * Maps TradingView resolution format to Kline History API resolution format
   * @param resolution - TradingView resolution string (e.g., "1", "60", "1D")
   * @returns Kline History API resolution string (e.g., "1m", "1h", "1d")
   */
  _mapToKlineHistoryResolution(resolution) {
    return _HistoryProvider._RESOLUTION_MAP.get(resolution) ?? resolution;
  }
  _buildUrlWithParams(path, params) {
    if (!params || Object.keys(params).length === 0) {
      return path;
    }
    const searchParams = new URLSearchParams();
    Object.keys(params).forEach((key) => {
      const value = params[key];
      if (Array.isArray(value)) {
        value.forEach((item) => searchParams.append(key, item));
      } else {
        searchParams.append(key, value.toString());
      }
    });
    const queryString = searchParams.toString();
    return queryString ? `${path}?${queryString}` : path;
  }
  _getPreferenceKey(symbolInfo, resolution) {
    return `${symbolInfo.ticker ?? ""}|${resolution}`;
  }
  _requestHistory(requestParams) {
    return this._requester.sendRequest(
      this._datafeedUrl,
      HISTORY_PATH,
      requestParams
    );
  }
  _buildKlineParams(requestParams, countBack) {
    const params = {
      ...requestParams
    };
    delete params.countback;
    if (countBack > 0) {
      params.limit = countBack;
    } else {
      delete params.limit;
    }
    return params;
  }
  /**
   * Get the earliest time from bars array
   * @param bars - Array of bars
   * @returns Earliest timestamp in seconds, or null if bars array is empty
   */
  _getEarliestTime(bars) {
    if (bars.length === 0) {
      return null;
    }
    return bars[0].time / 1e3;
  }
  /**
   * Try kline fallback with merge logic
   * Requests kline data from the earliest time in history result to the original end time,
   * then merges the kline data with history data
   * @param requestParams - Original request parameters
   * @param countBack - Count back value
   * @param earliestTime - Earliest time from history result (in seconds)
   * @param historyResult - History result to merge with
   * @returns Merged result or null if kline request fails
   */
  async _tryKlineFallbackWithMerge(requestParams, countBack, earliestTime, historyResult) {
    try {
      const klineParams = {
        ...requestParams,
        from: requestParams.from,
        to: earliestTime
      };
      const klineResult = await this._requestKlineHistory(
        this._buildKlineParams(klineParams, countBack)
      );
      if (klineResult.bars.length === 0) {
        return null;
      }
      return this._mergeBars(historyResult, klineResult);
    } catch {
      return null;
    }
  }
  /**
   * Merge history bars with kline bars
   * Optimized for the case where klineBars are earlier data than historyBars
   * Uses direct concatenation when no overlap, otherwise uses two-pointer merge
   * Assumes both arrays are sorted by time (ascending, earliest first)
   * Kline data takes precedence when timestamps match
   * @param historyResult - History result (later time range)
   * @param klineResult - Kline result (earlier time range)
   * @returns Merged result with sorted bars
   */
  _mergeBars(historyResult, klineResult) {
    const historyBars = historyResult.bars;
    const klineBars = klineResult.bars;
    if (historyBars.length === 0) {
      return klineResult;
    }
    if (klineBars.length === 0) {
      return historyResult;
    }
    const latestKlineTime = klineBars[klineBars.length - 1].time;
    const earliestHistoryTime = historyBars[0].time;
    let mergedBars;
    if (latestKlineTime < earliestHistoryTime) {
      mergedBars = [...klineBars, ...historyBars];
    } else if (latestKlineTime === earliestHistoryTime) {
      mergedBars = [...klineBars, ...historyBars.slice(1)];
    } else {
      mergedBars = [];
      let historyIdx = 0;
      let klineIdx = 0;
      while (historyIdx < historyBars.length && klineIdx < klineBars.length) {
        const historyBar = historyBars[historyIdx];
        const klineBar = klineBars[klineIdx];
        if (historyBar.time < klineBar.time) {
          mergedBars.push(historyBar);
          historyIdx++;
        } else if (historyBar.time > klineBar.time) {
          mergedBars.push(klineBar);
          klineIdx++;
        } else {
          mergedBars.push(klineBar);
          historyIdx++;
          klineIdx++;
        }
      }
      while (historyIdx < historyBars.length) {
        mergedBars.push(historyBars[historyIdx++]);
      }
      while (klineIdx < klineBars.length) {
        mergedBars.push(klineBars[klineIdx++]);
      }
    }
    const meta = {
      noData: klineResult.meta.noData && historyResult.meta.noData,
      nextTime: klineResult.meta.nextTime ?? historyResult.meta.nextTime
    };
    return {
      bars: mergedBars,
      meta
    };
  }
  async _tryKlineFallback(requestParams, countBack) {
    try {
      const result = await this._requestKlineHistory(
        this._buildKlineParams(requestParams, countBack)
      );
      return result.bars.length > 0 ? result : null;
    } catch {
      return null;
    }
  }
  _shouldFallbackToKline(response, expectedCount) {
    if (response.s !== "ok") {
      return false;
    }
    const barsCount = response.t.length;
    if (expectedCount > 0 && barsCount < expectedCount) {
      return true;
    }
    if (this._limitedServerResponse && this._limitedServerResponse.maxResponseLength > 0 && barsCount >= this._limitedServerResponse.maxResponseLength) {
      return true;
    }
    return false;
  }
};
/**
 * Static mapping table for resolution conversion
 * Maps TradingView resolution format to Kline History API resolution format
 * Key: TradingView resolution, Value: Kline History API resolution
 */
_HistoryProvider._RESOLUTION_MAP = /* @__PURE__ */ new Map([
  ["1", "1m"],
  // 1 minute
  ["3", "3m"],
  // 3 minutes
  ["5", "5m"],
  // 5 minutes
  ["15", "15m"],
  // 15 minutes
  ["30", "30m"],
  // 30 minutes
  ["60", "1h"],
  // 1 hour
  ["240", "4h"],
  // 4 hours
  ["720", "12h"],
  // 12 hours
  ["1D", "1d"],
  // 1 day
  ["1W", "1w"],
  // 1 week
  ["1M", "1mon"]
  // 1 month (mapped to 1m)
]);
var HistoryProvider = _HistoryProvider;

// src/tradingviewAdapter/datafeed/requester.ts
var Requester = class {
  constructor(headers) {
    if (headers) {
      this._headers = headers;
    }
  }
  sendRequest(datafeedUrl, urlPath, params) {
    if (params !== void 0) {
      const paramKeys = Object.keys(params);
      if (paramKeys.length !== 0) {
        urlPath += "?";
      }
      urlPath += paramKeys.map((key) => {
        return `${encodeURIComponent(key)}=${encodeURIComponent(
          params[key].toString()
        )}`;
      }).join("&");
    }
    const options = { credentials: "same-origin" };
    if (this._headers !== void 0) {
      options.headers = this._headers;
    }
    return fetch(`${datafeedUrl}/${urlPath}`, options).then((response) => response.json()).then((data) => {
      if (typeof data.success === "undefined") {
        return data;
      }
      if (!data.success) {
        throw new Error(data.message);
      }
      return data;
    });
  }
};

// src/tradingviewAdapter/datafeed/symbol-storage.ts
function extractField(data, field, arrayIndex, valueIsArray) {
  const value = data[field];
  if (Array.isArray(value) && (!valueIsArray || Array.isArray(value[0]))) {
    return value[arrayIndex];
  }
  return value;
}
function symbolKey(symbol, currency, unit) {
  return symbol + (currency !== void 0 ? "_%|#|%_" + currency : "") + (unit !== void 0 ? "_%|#|%_" + unit : "");
}
var SymbolsStorage = class {
  constructor(datafeedUrl, datafeedSupportedResolutions, requester) {
    this._exchangesList = ["Orderly"];
    this._symbolsInfo = {};
    this._symbolsList = [];
    this._datafeedUrl = datafeedUrl;
    this._datafeedSupportedResolutions = datafeedSupportedResolutions;
    this._requester = requester;
    this._readyPromise = this._init();
    this._readyPromise.catch((error) => {
    });
  }
  // BEWARE: this function does not consider symbol's exchange
  resolveSymbol(symbolName, currencyCode, unitId) {
    return this._readyPromise.then(() => {
      const symbolInfo = this._symbolsInfo[symbolKey(symbolName, currencyCode, unitId)];
      if (symbolInfo === void 0) {
        return Promise.reject("invalid symbol");
      }
      return Promise.resolve(symbolInfo);
    });
  }
  searchSymbols(searchString, exchange, symbolType, maxSearchResults) {
    return this._readyPromise.then(() => {
      const weightedResult = [];
      const queryIsEmpty = searchString.length === 0;
      searchString = searchString.toUpperCase();
      for (const symbolName of this._symbolsList) {
        const symbolInfo = this._symbolsInfo[symbolName];
        if (symbolInfo === void 0) {
          continue;
        }
        if (symbolType.length > 0 && symbolInfo.type !== symbolType) {
          continue;
        }
        if (exchange && exchange.length > 0 && symbolInfo.exchange !== exchange) {
          continue;
        }
        const positionInName = symbolInfo.name.toUpperCase().indexOf(searchString);
        const positionInDescription = symbolInfo.description.toUpperCase().indexOf(searchString);
        if (queryIsEmpty || positionInName >= 0 || positionInDescription >= 0) {
          const alreadyExists = weightedResult.some(
            (item) => item.symbolInfo === symbolInfo
          );
          if (!alreadyExists) {
            const weight = positionInName >= 0 ? positionInName : 8e3 + positionInDescription;
            weightedResult.push({ symbolInfo, weight });
          }
        }
      }
      const result = weightedResult.sort(
        (item1, item2) => item1.weight - item2.weight
      ).slice(0, maxSearchResults).map((item) => {
        const symbolInfo = item.symbolInfo;
        return {
          symbol: symbolInfo.name,
          full_name: `${symbolInfo.exchange}:${symbolInfo.name}`,
          description: symbolInfo.description,
          exchange: symbolInfo.exchange,
          params: [],
          type: symbolInfo.type,
          ticker: symbolInfo.name
        };
      });
      return Promise.resolve(result);
    });
  }
  _init() {
    const promises = [];
    const alreadyRequestedExchanges = {};
    for (const exchange of this._exchangesList) {
      if (alreadyRequestedExchanges[exchange]) {
        continue;
      }
      alreadyRequestedExchanges[exchange] = true;
      promises.push(this._requestExchangeData(exchange));
    }
    return Promise.all(promises).then(() => {
      this._symbolsList.sort();
    });
  }
  _requestExchangeData(exchange) {
    return new Promise(
      (resolve, reject) => {
        this._requester.sendRequest(
          this._datafeedUrl,
          "tv/symbol_info",
          {
            group: exchange
          }
        ).then((response) => {
          try {
            this._onExchangeDataReceived(exchange, response);
          } catch (error) {
            reject(
              error instanceof Error ? error : new Error(`SymbolsStorage: Unexpected exception ${error}`)
            );
            return;
          }
          resolve();
        }).catch((reason) => {
          logMessage(
            `SymbolsStorage: Request data for exchange '${exchange}' failed, reason=${getErrorMessage(
              reason
            )}`
          );
          resolve();
        });
      }
    );
  }
  _onExchangeDataReceived(exchange, data) {
    let symbolIndex = 0;
    try {
      const symbolsCount = data.symbol.length;
      const tickerPresent = data.ticker !== void 0;
      for (; symbolIndex < symbolsCount; ++symbolIndex) {
        const symbolName = data.symbol[symbolIndex];
        const listedExchange = extractField(
          data,
          "exchange-listed",
          symbolIndex
        );
        const tradedExchange = extractField(
          data,
          "exchange-traded",
          symbolIndex
        );
        const fullName = tradedExchange + ":" + symbolName;
        const currencyCode = extractField(data, "currency-code", symbolIndex);
        const unitId = extractField(data, "unit-id", symbolIndex);
        const ticker = tickerPresent ? extractField(data, "ticker", symbolIndex) : symbolName;
        const symbolInfo = {
          ticker,
          name: symbolName,
          base_name: [listedExchange + ":" + symbolName],
          listed_exchange: listedExchange,
          exchange: tradedExchange,
          currency_code: currencyCode,
          original_currency_code: extractField(
            data,
            "original-currency-code",
            symbolIndex
          ),
          unit_id: unitId,
          original_unit_id: extractField(data, "original-unit-id", symbolIndex),
          unit_conversion_types: extractField(
            data,
            "unit-conversion-types",
            symbolIndex,
            true
          ),
          description: extractField(data, "description", symbolIndex),
          has_intraday: definedValueOrDefault(
            extractField(data, "has-intraday", symbolIndex),
            false
          ),
          // show volume at the bottom by default: https://github.com/tradingview/charting_library/issues/8306
          visible_plots_set: definedValueOrDefault(
            extractField(data, "visible-plots-set", symbolIndex),
            "ohlcv"
          ),
          minmov: extractField(data, "minmovement", symbolIndex) || extractField(data, "minmov", symbolIndex) || 0,
          minmove2: extractField(data, "minmove2", symbolIndex) || extractField(data, "minmov2", symbolIndex),
          fractional: extractField(data, "fractional", symbolIndex),
          pricescale: extractField(data, "pricescale", symbolIndex),
          type: extractField(data, "type", symbolIndex),
          session: extractField(data, "session-regular", symbolIndex),
          session_holidays: extractField(data, "session-holidays", symbolIndex),
          corrections: extractField(data, "corrections", symbolIndex),
          timezone: extractField(data, "timezone", symbolIndex),
          supported_resolutions: definedValueOrDefault(
            extractField(data, "supported-resolutions", symbolIndex, true),
            this._datafeedSupportedResolutions
          ),
          has_daily: definedValueOrDefault(
            extractField(data, "has-daily", symbolIndex),
            true
          ),
          intraday_multipliers: definedValueOrDefault(
            extractField(data, "intraday-multipliers", symbolIndex, true),
            ["1", "5", "15", "30", "60"]
          ),
          // daily_multipliers:["1"],
          // has_weekly_and_monthly: extractField(
          //   data,
          //   "has-weekly-and-monthly",
          //   symbolIndex,
          // ),
          has_weekly_and_monthly: true,
          has_empty_bars: extractField(data, "has-empty-bars", symbolIndex),
          volume_precision: definedValueOrDefault(
            extractField(data, "volume-precision", symbolIndex),
            0
          ),
          format: "price"
        };
        this._symbolsInfo[ticker] = symbolInfo;
        this._symbolsInfo[symbolName] = symbolInfo;
        this._symbolsInfo[fullName] = symbolInfo;
        if (currencyCode !== void 0 || unitId !== void 0) {
          this._symbolsInfo[symbolKey(ticker, currencyCode, unitId)] = symbolInfo;
          this._symbolsInfo[symbolKey(symbolName, currencyCode, unitId)] = symbolInfo;
          this._symbolsInfo[symbolKey(fullName, currencyCode, unitId)] = symbolInfo;
        }
        this._symbolsList.push(symbolName);
      }
    } catch (error) {
      throw new Error(
        `SymbolsStorage: API error when processing exchange ${exchange} symbol #${symbolIndex} (${data.symbol[symbolIndex]}): ${Object(error).message}`
      );
    }
  }
};
function definedValueOrDefault(value, defaultValue) {
  return value !== void 0 ? value : defaultValue;
}

// src/tradingviewAdapter/datafeed/abstract-datafeed.ts
var AbstractDatafeed = class {
  constructor(datafeedURL) {
    this._configuration = defaultConfiguration();
    this._symbolsStorage = null;
    this._historyCursor = null;
    this._datafeedURL = datafeedURL;
    this._requester = new Requester();
    this._historyProvider = new HistoryProvider(datafeedURL, this._requester);
    this._configurationReadyPromise = this._requestConfiguration().then(
      (configuration) => {
        if (configuration === null) {
          configuration = defaultConfiguration();
        }
        this._setupWithConfiguration(configuration);
      }
    );
  }
  getBars(symbolInfo, resolution, periodParams, onResult, onError) {
    const { to, firstDataRequest } = periodParams;
    if (firstDataRequest || this._historyCursor === null) {
      this._historyCursor = to;
    }
    const resolutionInSeconds = this._resolutionToSeconds(resolution);
    const cursor = this._historyCursor ?? to;
    const maxBarsPerRequest = 1e3;
    const shouldUseOriginalRange = resolution.toLowerCase?.() === "1m" && periodParams.from !== void 0;
    const barsToLoad = shouldUseOriginalRange ? periodParams.countBack ?? maxBarsPerRequest : maxBarsPerRequest;
    const requestTo = shouldUseOriginalRange ? periodParams.to ?? to : cursor;
    const requestFrom = shouldUseOriginalRange ? periodParams.from : Math.max(requestTo - resolutionInSeconds * barsToLoad, 0);
    this._historyProvider.getBars(symbolInfo, resolution, {
      ...periodParams,
      countBack: barsToLoad,
      from: requestFrom,
      to: requestTo
    }).then(
      (result) => {
        if (result.bars.length > 0) {
          this._historyCursor = Math.floor(result.bars[0].time / 1e3) - 1;
        }
        onResult(result.bars, result.meta);
      },
      (error) => {
        onError(error);
      }
    ).catch((error) => {
      onError(error);
    });
  }
  _resolutionToSeconds(resolution) {
    switch (resolution) {
      case "1D":
        return 86400;
      case "3D":
        return 86400 * 3;
      case "1W":
        return 86400 * 7;
      case "1M":
        return 86400 * 30;
    }
    const res = parseInt(resolution);
    if (!isNaN(res)) {
      return res * 60;
    }
    return 60;
  }
  onReady(callback) {
    this._configurationReadyPromise.then(() => {
      callback(this._configuration);
    });
  }
  searchSymbols(userInput, exchange, symbolType, onResult) {
    if (this._symbolsStorage === null) {
      throw new Error("Datafeed: inconsistent configuration (symbols storage)");
    }
    this._symbolsStorage.searchSymbols(
      userInput,
      exchange,
      symbolType,
      30 /* SearchItemsLimit */
    ).then(onResult).catch(onResult.bind(null, []));
  }
  resolveSymbol(symbolName, onResolve, onError, extension) {
    const currencyCode = extension && extension.currencyCode;
    const unitId = extension && extension.unitId;
    function onResultReady(symbolInfo) {
      onResolve(symbolInfo);
    }
    if (this._symbolsStorage === null) {
      throw new Error("Datafeed: inconsistent configuration (symbols storage)");
    }
    this._symbolsStorage.resolveSymbol(symbolName, currencyCode, unitId).then(onResultReady).catch(onError);
  }
  getMarks() {
  }
  getTimescaleMarks() {
  }
  getServerTime() {
  }
  _requestConfiguration() {
    return this._send("tv/config").catch(
      (reason) => {
        logMessage(
          `Datafeed: Cannot get datafeed configuration - use default, error=${getErrorMessage(
            reason
          )}`
        );
        return null;
      }
    );
  }
  _send(urlPath, params) {
    return this._requester.sendRequest(this._datafeedURL, urlPath, params);
  }
  _setupWithConfiguration(configurationData) {
    this._configuration = configurationData;
    if (configurationData.exchanges === void 0) {
      configurationData.exchanges = [];
    }
    if (!configurationData.supports_search && !configurationData.supports_group_request) {
      throw new Error(
        "Unsupported datafeed configuration. Must either support search, or support group request"
      );
    }
    if (configurationData.supports_group_request || !configurationData.supports_search) {
      this._symbolsStorage = new SymbolsStorage(
        this._datafeedURL,
        configurationData.supported_resolutions || [],
        this._requester
      );
    }
    logMessage(
      `Datafeed: Initialized with ${JSON.stringify(configurationData)}`
    );
  }
};
function defaultConfiguration() {
  return {
    supports_search: false,
    supports_group_request: true,
    supported_resolutions: [
      "1",
      "3",
      "5",
      "15",
      "30",
      "60",
      "120",
      "240",
      "480",
      "720",
      "1D",
      "3D",
      "1W",
      "1M"
    ],
    supports_marks: false,
    supports_timescale_marks: false
  };
}

// src/tradingviewAdapter/datafeed/eventBus.ts
var MultiBroadcastEventBus = class {
  constructor() {
    this.subscribers = /* @__PURE__ */ new Map();
  }
  subscribe(eventName, callback) {
    if (!this.subscribers.has(eventName)) {
      this.subscribers.set(eventName, []);
    }
    const callbacks = this.subscribers.get(eventName);
    callbacks.push(callback);
    return () => {
      this.unsubscribe(eventName, callback);
    };
  }
  unsubscribe(eventName, callback) {
    if (this.subscribers.has(eventName)) {
      const callbacks = this.subscribers.get(eventName);
      this.subscribers.set(
        eventName,
        callbacks.filter((cb) => cb !== callback)
      );
      if (this.subscribers.get(eventName).length === 0) {
        this.subscribers.delete(eventName);
      }
    }
  }
  publish(eventName, data) {
    if (this.subscribers.has(eventName)) {
      const callbacks = this.subscribers.get(eventName);
      callbacks.forEach((callback) => {
        callback(data);
      });
    }
  }
};

// src/utils/common.util.ts
var mapResolution = (resolution) => {
  let time = "1d";
  switch (resolution) {
    case "1" /* RESOLUTION_1m */:
      time = "1m";
      break;
    case "3" /* RESOLUTION_3m */:
      time = "3m";
      break;
    case "5" /* RESOLUTION_5m */:
      time = "5m";
      break;
    case "15" /* RESOLUTION_15m */:
      time = "15m";
      break;
    case "30" /* RESOLUTION_30m */:
      time = "30m";
      break;
    case "60" /* RESOLUTION_60m */:
      time = "1h";
      break;
    case "120" /* RESOLUTION_120m */:
      time = "2h";
      break;
    case "240" /* RESOLUTION_240m */:
      time = "4h";
      break;
    case "480" /* RESOLUTION_480m */:
      time = "8h";
      break;
    case "720" /* RESOLUTION_720m */:
      time = "12h";
      break;
    case "D" /* RESOLUTION_D */:
    case "1D" /* RESOLUTION_1D */:
      time = "1d";
      break;
    case "3D" /* RESOLUTION_3D */:
      time = "3d";
      break;
    case "1W" /* RESOLUTION_1W */:
      time = "1w";
      break;
    case "1M" /* RESOLUTION_1M */:
      time = "1M ";
      break;
  }
  return time;
};
var TradingViewSDKLocalstorageKey = {
  interval: "TradingviewSDK.lastUsedTimeBasedResolution",
  lineType: "TradingviewSDK.lastUsedStyle",
  displayControlSetting: "TradingviewSDK.displaySetting"
};

// src/tradingviewAdapter/datafeed/websocket.service.ts
var getKlineKey = (symbol, type) => `${symbol}kline_${type}`;
var getSymbolTopics = (symbol) => {
  const symbolTopics = ["trade"];
  return symbolTopics.map((topic) => `${symbol}@${topic}`);
};
var _WebsocketService = class _WebsocketService {
  constructor(ws) {
    this.klineSubscribeIdMap = /* @__PURE__ */ new Map();
    this.klineOnTickCallback = /* @__PURE__ */ new Map();
    this.subscribeCachedTopics = /* @__PURE__ */ new Map();
    this.wsInstance = null;
    this.klineData = /* @__PURE__ */ new Map();
    if (!_WebsocketService._created) {
      this.wsInstance = ws;
      _WebsocketService._instance = this;
      _WebsocketService._created = true;
    }
    return _WebsocketService._instance;
  }
  subscribeKline(subscribeId, symbol, resolution, onTickCallback) {
    const time = mapResolution(resolution);
    this.klineSubscribeIdMap.set(subscribeId, { symbol, resolution });
    const klineKey = getKlineKey(symbol, time);
    if (this.klineOnTickCallback.has(klineKey)) {
      const value = this.klineOnTickCallback.get(klineKey);
      value[subscribeId] = onTickCallback;
    } else {
      this.klineOnTickCallback.set(klineKey, {
        [subscribeId]: onTickCallback
      });
      const unsub = this.wsInstance?.subscribe(
        {
          event: "subscribe",
          topic: `${symbol}@kline_${time}`,
          id: `${symbol}@kline_${time}`,
          ts: (/* @__PURE__ */ new Date()).getTime()
        },
        {
          onMessage: (data) => {
            const { open, close, high, low, volume, startTime } = data;
            const key = getKlineKey(data.symbol, data.type);
            this.updateKline(key, {
              time: startTime,
              close,
              open,
              high,
              low,
              volume
            });
          }
        }
      );
      this.subscribeCachedTopics.set(`${symbol}@kline_${time}`, unsub);
    }
  }
  unsubscribeKline(subscribeId) {
    if (!this.klineSubscribeIdMap.has(subscribeId)) {
      return;
    }
    const { symbol, resolution } = this.klineSubscribeIdMap.get(subscribeId);
    const time = mapResolution(resolution);
    const klineKey = getKlineKey(symbol, time);
    if (this.klineOnTickCallback.has(klineKey)) {
      const value = this.klineOnTickCallback.get(klineKey);
      delete value[subscribeId];
      if (Object.keys(value).length === 0) {
        this.klineOnTickCallback.delete(klineKey);
        const unsub = this.subscribeCachedTopics.get(`${symbol}@kline_${time}`);
        unsub();
      }
    }
    delete this.klineSubscribeIdMap[subscribeId];
  }
  subscribeSymbol(symbol) {
    const symbolTopics = getSymbolTopics(symbol);
    symbolTopics.forEach((topic) => {
      if (!this.subscribeCachedTopics.has(topic)) {
        const unsub = this.wsInstance?.subscribe(
          {
            event: "subscribe",
            topic,
            id: topic,
            ts: (/* @__PURE__ */ new Date()).getTime()
          },
          {
            onMessage: (data) => {
              this.updateKlineByLastPrice(data.symbol, data.price);
            }
          }
        );
        this.subscribeCachedTopics.set(topic, unsub);
      }
    });
  }
  updateKlineByLastPrice(symbol, lastPrice) {
    this.klineOnTickCallback.forEach((_, key) => {
      if (key.startsWith(symbol)) {
        const klineData = this.klineData.get(key);
        if (klineData) {
          this.updateKline(key, { ...klineData, close: lastPrice });
        }
      }
    });
  }
  updateKline(key, cbParams) {
    const onTickCbs = this.klineOnTickCallback.get(key);
    if (onTickCbs && cbParams) {
      this.klineData.set(key, cbParams);
      Object.keys(onTickCbs).forEach((key2) => {
        const onTickCb = onTickCbs[key2];
        if (onTickCb && typeof onTickCb === "function") {
          onTickCb(cbParams);
        }
      });
    }
  }
};
_WebsocketService._created = false;
_WebsocketService._instance = null;
var WebsocketService = _WebsocketService;

// src/tradingviewAdapter/datafeed/datafeed.ts
var getAutoIncrementId = /* @__PURE__ */ (() => {
  let id = 0;
  return () => id++;
})();
var Datafeed = class extends AbstractDatafeed {
  constructor(apiUrl, ws) {
    const datafeedURL = `${apiUrl}`;
    super(datafeedURL);
    this.bbosMap = /* @__PURE__ */ new Map();
    this.tickersMap = /* @__PURE__ */ new Map();
    this.eventBus = new MultiBroadcastEventBus();
    this._subscribeQuoteMap = /* @__PURE__ */ new Map();
    this._prefixId = getAutoIncrementId();
    this._publicWs = new WebsocketService(ws);
    this.bbosMap = /* @__PURE__ */ new Map();
    ws.on("tickers", (message) => {
      for (const ticker of message.data) {
        ticker.change = ticker.close - ticker.open;
        ticker.perChange = ticker.open ? +(100 * ticker.change / ticker.open).toFixed(2) : 0;
        this.tickersMap.set(ticker.symbol, ticker);
      }
      this.eventBus.publish("tickerUpdate", { message: "ticker" });
    });
    ws.subscribe(
      {
        event: "subscribe",
        topic: `bbos`
      },
      {
        formatter: (message) => message,
        onMessage: (message) => {
          for (const bbo of message.data) {
            this.bbosMap.set(bbo.symbol, {
              ask: bbo.ask,
              bid: bbo.bid,
              askSize: bbo.askSize,
              bidSize: bbo.bidSize
            });
          }
          this.eventBus.publish("tickerUpdate", { message: "bbos" });
        }
      }
    );
  }
  remove() {
    Array.from(this._subscribeQuoteMap.values()).forEach((s) => s?.());
  }
  getSubscriptionId(listenerGuid) {
    return `${this._prefixId}${listenerGuid}`;
  }
  subscribeBars(symbolInfo, resolution, onTick, listenerGuid, onResetCacheNeededCallback) {
    window.onResetCacheNeededCallback = onResetCacheNeededCallback;
    this._publicWs.subscribeKline(
      `${this._prefixId}${listenerGuid}`,
      symbolInfo.ticker,
      resolution,
      onTick
    );
  }
  unsubscribeBars(listenerGuid) {
    this._publicWs.unsubscribeKline(`${this._prefixId}${listenerGuid}`);
  }
  getQuotes(symbols, onDataCallback) {
    const subscriptionId = this.getSubscriptionId("getQuotes");
    this.unsubscribeQuotes("getQuotes");
    const unsub = this.eventBus.subscribe("tickerUpdate", (msg) => {
      const dataMap = /* @__PURE__ */ new Map();
      symbols.forEach((symbol) => {
        const bbo = this.bbosMap.get(withoutExchangePrefix(symbol));
        const ticker = this.tickersMap.get(withoutExchangePrefix(symbol));
        if (!bbo || !ticker) {
          return;
        }
        const data = {
          ...ticker,
          ask: bbo.ask,
          bid: bbo.bid
        };
        dataMap.set(withoutExchangePrefix(symbol), data);
      });
      if (!dataMap.size) {
        return;
      }
      onDataCallback(
        Array.from(dataMap.values()).map(
          (symbolData) => this._toUDFTicker(symbolData)
        )
      );
    });
    this._subscribeQuoteMap.set(subscriptionId, unsub);
  }
  subscribeQuotes(symbols, fastSymbols, onRealtimeCallback, listenerGuid) {
    const subscriptionId = `${this._prefixId}${listenerGuid}`;
    if (symbols.length > 0) {
      this.unsubscribeQuotes(subscriptionId);
      const unsub = this.eventBus.subscribe("tickerUpdate", (msg) => {
        const dataMap = /* @__PURE__ */ new Map();
        symbols.forEach((symbol) => {
          const bbo = this.bbosMap.get(withoutExchangePrefix(symbol));
          const ticker = this.tickersMap.get(withoutExchangePrefix(symbol));
          if (!bbo || !ticker) {
            return;
          }
          const data = {
            ...ticker,
            ask: bbo.ask,
            bid: bbo.bid
          };
          dataMap.set(withoutExchangePrefix(symbol), data);
        });
        if (!dataMap.size) {
          return;
        }
        onRealtimeCallback(
          Array.from(dataMap.values()).map(
            (symbolData) => this._toUDFTicker(symbolData)
          )
        );
      });
      this._subscribeQuoteMap.set(subscriptionId, unsub);
    }
  }
  unsubscribeQuotes(listenerGuid) {
    const subscriptionId = this.getSubscriptionId(listenerGuid);
    const unsub = this._subscribeQuoteMap.get(subscriptionId);
    if (unsub) {
      unsub();
      this._subscribeQuoteMap.delete(subscriptionId);
    }
  }
  _toUDFTicker(t) {
    return {
      n: withExchangePrefix(t.symbol),
      s: "ok",
      v: {
        ask: t.ask,
        bid: t.bid,
        ch: t.change,
        chp: t.perChange / 100,
        description: "",
        exchange: EXCHANGE,
        hight_price: t.high,
        low_price: t.low,
        lp: t.close,
        open_price: t.open,
        prev_close_price: 0,
        volume: t.volume
      }
    };
  }
};
var EST_TPSL_PNL_DECIMAL = 2;
var textDash = "--";
var BracketAlgoType = ["BRACKET" /* BRACKET */, "STOP_BRACKET" /* STOP_BRACKET */];
var TpslAlgoType = ["POSITIONAL_TP_SL" /* POSITIONAL_TP_SL */, "TP_SL" /* TP_SL */];
var isTpslOrder = (order) => order.root_algo_order_id !== order.algo_order_id && (order.algo_type === "TAKE_PROFIT" /* TAKE_PROFIT */ || order.algo_type === "STOP_LOSS" /* STOP_LOSS */);
var isBracketAlgoType = (algoType) => !!algoType && BracketAlgoType.includes(algoType);
var isPositionTpsl = (order) => order.type === "CLOSE_POSITION" /* CLOSE_POSITION */;
var isActivatedPositionTpsl = (order) => isPositionTpsl(order) && order.is_activated;
var isActivatedTpslOrder = (order) => isTpslOrder(order) && order.is_activated;
var isActivatedQuantityTpsl = (order) => order.root_algo_order_algo_type === "TP_SL" /* TP_SL */ || isBracketAlgoType(order.root_algo_order_algo_type) && order.is_activated;
var getTpslTag = (order, quantityTpslNoMap) => {
  const algoType = order.algo_type;
  const suffix = {
    ["TAKE_PROFIT" /* TAKE_PROFIT */]: i18n.t("tpsl.takeProfit"),
    ["STOP_LOSS" /* STOP_LOSS */]: i18n.t("tpsl.stopLoss")
  }[algoType];
  if (!suffix) {
    return null;
  }
  return suffix;
};
var buildQuantityTpslNoMap = (orders) => {
  const quantityTpslNoMap = /* @__PURE__ */ new Map();
  let idx = 1;
  [...orders].reverse().filter(isActivatedQuantityTpsl).forEach((order) => {
    if (order.root_algo_order_id && !quantityTpslNoMap.has(order.root_algo_order_id)) {
      quantityTpslNoMap.set(order.root_algo_order_id, idx++);
    }
  });
  return quantityTpslNoMap;
};
var getTpslEstPnl = (tpslOrder, position) => {
  const quantity = Math.abs(
    tpslOrder.type === "CLOSE_POSITION" /* CLOSE_POSITION */ ? position.balance : tpslOrder.quantity
  );
  const sideFlag = tpslOrder.side === "SELL" /* SELL */ ? 1 : -1;
  const openPrice = position.open.toString();
  const estPnl = new Decimal(tpslOrder.trigger_price).minus(openPrice ?? 0).times(quantity).times(sideFlag).toString();
  return { estPnl, quantity, openPrice };
};
var formatPnl = (pnl) => {
  return pnl !== void 0 && pnl !== "" ? new Decimal(pnl).todp(EST_TPSL_PNL_DECIMAL, Decimal.ROUND_FLOOR) : textDash;
};
function useCancelOrder() {
  const [
    pendingOrders,
    { cancelOrder, cancelAlgoOrder, cancelTPSLChildOrder }
  ] = useOrderStream({
    status: OrderStatus.INCOMPLETE
  });
  return useCallback(
    (order) => {
      if (order.algo_order_id) {
        if (TpslAlgoType.includes(order.root_algo_order_algo_type)) {
          const rootOrder = pendingOrders?.find(
            (item) => item.algo_order_id === order.root_algo_order_id
          );
          const isEditActivated = rootOrder.child_orders.every(
            (item) => !!item.trigger_price
          );
          if (isEditActivated) {
            return cancelTPSLChildOrder(
              order.algo_order_id,
              order.root_algo_order_id
            );
          }
          return cancelAlgoOrder(rootOrder.algo_order_id, order.symbol).then();
        }
        return cancelAlgoOrder(order.algo_order_id, order.symbol).then();
      }
      return cancelOrder(order.order_id, order.symbol).then();
    },
    [cancelOrder, pendingOrders]
  );
}
function useEditOrder(onToast) {
  useEventEmitter();
  const [, { updateOrder, cancelAlgoOrder, updateAlgoOrder, updateTPSLOrder }] = useOrderStream({
    status: OrderStatus.INCOMPLETE
  });
  return useCallback(
    (order, lineValue) => {
      if (order.algo_order_id) {
        if (TpslAlgoType.includes(order.root_algo_order_algo_type)) {
          const algoParams = [
            {
              order_id: order.algo_order_id,
              trigger_price: new Decimal(lineValue.value).toString()
            }
          ];
          return updateTPSLOrder(order.root_algo_order_id, algoParams).then((res) => {
          }).catch((e) => {
            if (onToast) {
              onToast.error(e.message);
            }
          });
        } else if (BracketAlgoType.includes(order.algo_type)) {
          return updateAlgoOrder(order.algo_order_id, {
            order_price: new Decimal(lineValue.value).toString()
          }).then((res) => {
          }).catch((e) => {
            if (onToast) {
              onToast.error(e.message);
            }
          });
        } else {
          const values2 = {
            quantity: order.quantity,
            trigger_price: order.trigger_price,
            symbol: order.symbol,
            price: order.price,
            // order_type: order.type,
            // side: order.side,
            // reduce_only: Boolean(order.reduce_only),
            algo_order_id: order.algo_order_id
          };
          if (order.order_tag) {
            values2.order_tag = order.order_tag;
          }
          if (order.client_order_id) {
            values2.client_order_id = order.client_order_id;
          }
          if (lineValue.type === "price") {
            values2.price = new Decimal(lineValue.value).toString();
          }
          if (lineValue.type === "trigger_price") {
            values2.trigger_price = new Decimal(lineValue.value).toString();
          }
          return updateAlgoOrder(order.algo_order_id, values2).then((res) => {
          }).catch((e) => {
            if (onToast) {
              onToast.error(e.message);
            }
          });
        }
      }
      const values = {
        order_price: order.price?.toString(),
        order_quantity: order.quantity.toString(),
        symbol: order.symbol,
        order_type: order.type,
        side: order.side,
        visible_quantity: 0,
        reduce_only: order.reduce_only
      };
      if (new Decimal(order.visible_quantity ?? order.visible ?? 0).eq(
        order.quantity
      )) {
        delete values.visible_quantity;
      }
      if (!Object.keys(order).includes("reduce_only")) {
        delete values.reduce_only;
      }
      if (order.order_tag) {
        values.order_tag = order.order_tag;
      }
      if (order.client_order_id) {
        values.client_order_id = order.client_order_id;
      }
      if (lineValue.type === "price") {
        values.order_price = new Decimal(lineValue.value).toString();
      }
      return updateOrder(order.order_id, values).then((res) => {
      }).catch((e) => {
        onToast.error(e.message);
      });
    },
    [updateOrder]
  );
}
function useSendOrder(symbol) {
  const { onSubmit: _sendMarketOrder } = useOrderEntry_deprecated(
    {
      symbol,
      side: OrderSide.BUY,
      order_type: OrderType.MARKET
    },
    {
      watchOrderbook: true
    }
  );
  const sendLimitOrder = () => {
  };
  const sendMarketOrder = (data) => {
    data.reduce_only = false;
    return _sendMarketOrder(data).catch((e) => {
      toast.error(e);
    });
  };
  return {
    sendLimitOrder,
    sendMarketOrder
  };
}
var createBrokerMethod = (method) => {
  return (params) => method(params);
};
var useBroker = ({
  closeConfirm,
  colorConfig,
  onToast,
  mode,
  symbol
}) => {
  const cancelOrder = useCancelOrder();
  const editOrder = useEditOrder(onToast);
  const symbolData = useSymbolsInfo();
  const closePosition = useCallback(
    (position) => closeConfirm && closeConfirm(position),
    [closeConfirm]
  );
  const { sendMarketOrder, sendLimitOrder } = useSendOrder(symbol);
  const getSymbolInfo = useCallback(
    (symbol2) => {
      if (!symbolData) {
        return;
      }
      return {
        baseMin: symbolData[symbol2]("base_min"),
        baseMax: symbolData[symbol2]("base_max"),
        baseTick: symbolData[symbol2]("base_tick"),
        quoteTick: symbolData[symbol2]("quote_tick")
      };
    },
    [symbolData]
  );
  const broker = useRef({
    cancelOrder,
    closePosition,
    editOrder,
    colorConfig,
    sendLimitOrder,
    getSymbolInfo,
    sendMarketOrder: createBrokerMethod(sendMarketOrder),
    mode
  });
  useEffect(() => {
    broker.current.getSymbolInfo = getSymbolInfo;
  }, [symbolData]);
  useEffect(() => {
    broker.current.sendLimitOrder = sendLimitOrder;
    broker.current.sendMarketOrder = sendMarketOrder;
  }, [sendLimitOrder, sendMarketOrder]);
  useEffect(() => {
    broker.current.closePosition = closePosition;
  }, [closeConfirm]);
  useEffect(() => {
    broker.current.cancelOrder = cancelOrder;
  }, [cancelOrder]);
  return broker.current;
};
var useBroker_default = useBroker;
var IntervalMap = {
  1: { startOf: "minute", period: 0 },
  3: { startOf: "hour", period: 3 * 60 * 1e3 },
  5: { startOf: "hour", period: 5 * 60 * 1e3 },
  15: { startOf: "hour", period: 15 * 60 * 1e3 },
  30: { startOf: "hour", period: 30 * 60 * 1e3 },
  60: { startOf: "hour", period: 0 },
  120: { startOf: "day", period: 2 * 60 * 60 * 1e3 },
  240: { startOf: "day", period: 4 * 60 * 60 * 1e3 },
  480: { startOf: "day", period: 8 * 60 * 60 * 1e3 },
  720: { startOf: "day", period: 12 * 60 * 60 * 1e3 },
  D: { startOf: "day", period: 0 },
  // default day interval shown as D instead of 1D
  "1D": { startOf: "day", period: 0 },
  "3D": { startOf: "year", period: 3 * 24 * 60 * 60 * 1e3 },
  "5D": { startOf: "year", period: 5 * 24 * 60 * 60 * 1e3 },
  "1W": { startOf: "week", period: 0 },
  "1M": { startOf: "month", period: 0 }
};
function collectionOrders(order, collection, interval, orderList) {
  const { startOf, period } = IntervalMap[interval];
  const time = new Date(order.updated_time).getTime();
  let base = startOfSecond(time).getTime();
  if (startOf === "minute") {
    base = startOfMinute(time).getTime();
  } else if (startOf === "hour") {
    base = startOfHour(time).getTime();
  } else if (startOf === "day") {
    base = startOfDay(time).getTime();
  } else if (startOf === "month") {
    base = startOfMonth(time).getTime();
  } else if (startOf === "year") {
    base = startOfYear(time).getTime();
  } else if (startOf === "week") {
    base = startOfWeek(time).getTime();
  } else if (startOf === "month") {
    base = startOfMonth(time).getTime();
  }
  const group = period === 0 ? base : Math.floor((time - base) / period) * period + base;
  if (!collection[group]) {
    collection[group] = { ["BUY" /* BUY */]: [], ["SELL" /* SELL */]: [] };
  }
  if (collection[group][order.side].length < 5) {
    collection[group][order.side].push(order);
    orderList.push(order);
  }
}
var limitOrdersByInterval = (orders, interval) => {
  const res = [];
  const collection = {};
  if (!IntervalMap[interval]) {
    return [];
  }
  orders.forEach((order) => {
    if (order.child_orders) {
      for (const child of order.child_orders) {
        if (child.is_activated && child.algo_status === OrderStatus.FILLED) {
          collectionOrders(child, collection, interval, res);
        }
      }
    } else {
      collectionOrders(order, collection, interval, res);
    }
  });
  return res;
};

// src/tradingviewAdapter/renderer/execution.service.ts
var ExecutionService = class _ExecutionService {
  constructor(instance, broker) {
    this.interval = "1D";
    this.changeInterval = (interval) => {
      const rerenderExecutions = () => {
        this.renderExecutions(this.filledOrders, this.basePriceDecimal);
        this.instance.activeChart().onDataLoaded().unsubscribe(null, rerenderExecutions);
      };
      this.interval = interval;
      this.instance.activeChart().onDataLoaded().subscribe(null, rerenderExecutions);
    };
    this.instance = instance;
    this.executions = [];
    this.filledOrders = [];
    this.basePriceDecimal = 0;
    this.broker = broker;
    this.subscribeIntervalChange();
  }
  async subscribeIntervalChange() {
    this.interval = this.instance.symbolInterval().interval;
    const changeInterval = this.changeInterval;
    this.instance.activeChart().onIntervalChanged().subscribe(null, changeInterval);
  }
  renderExecutions(filledOrders, basePriceDecimal) {
    this.filledOrders = filledOrders;
    this.basePriceDecimal = basePriceDecimal;
    if (!this.interval) {
      return;
    }
    this.removeAll();
    limitOrdersByInterval(filledOrders, this.interval).forEach((order) => {
      this.executions.push(this.drawExecution(order, basePriceDecimal));
    });
  }
  removeAll() {
    this.executions.forEach((execution) => execution.remove());
    this.executions = [];
  }
  static getExecutionInfo(order, basePriceDecimal) {
    const side = order.side;
    const avgExecPrice = order.average_executed_price || order.child_orders?.find((child) => !!child.average_executed_price)?.average_executed_price || 0;
    const excutedNumber = new Decimal(avgExecPrice).todp(basePriceDecimal, Decimal.ROUND_FLOOR).toString();
    return `${side === "BUY" /* BUY */ ? i18n.t("common.buy") : i18n.t("common.sell")} ${order.total_executed_quantity} @${commify(excutedNumber)}`;
  }
  drawExecution(order, basePriceDecimal) {
    const side = order.side;
    const avgExecPrice = order.average_executed_price || order.child_orders?.find((child) => !!child.average_executed_price)?.average_executed_price || 0;
    const timestamp = new Date(order.updated_time).getTime() / 1e3;
    const colorConfig = this.broker.colorConfig;
    return this.instance.activeChart().createExecutionShape().setArrowHeight(9).setTooltip(_ExecutionService.getExecutionInfo(order, basePriceDecimal)).setTime(timestamp).setPrice(avgExecPrice).setArrowColor(
      side === "BUY" /* BUY */ ? colorConfig.upColor : colorConfig.downColor
    ).setDirection(side === "BUY" /* BUY */ ? "buy" : "sell");
  }
  unsubscribeIntervalChange() {
    const changeInterval = this.changeInterval;
    try {
      this.instance.activeChart().onIntervalChanged().unsubscribe(null, changeInterval);
    } catch (e) {
      const errorString = e?.toString() || String(e);
      if (errorString.includes("tradingViewApi") || errorString.includes("Cannot read properties of null") || errorString.includes("Cannot read property") || e instanceof TypeError && errorString.includes("null")) ;
    }
  }
  destroy() {
    this.removeAll();
    this.unsubscribeIntervalChange();
  }
};

// src/tradingviewAdapter/renderer/order.util.ts
var getOrderId = (order) => {
  if (order === null || order === void 0) {
    return void 0;
  }
  return order.algo_order_id || order.order_id;
};

// src/tradingviewAdapter/renderer/tpslCal.service.ts
var TpslCalService = class {
  constructor() {
    this.quantityTpslNoMap = /* @__PURE__ */ new Map();
    this.tpslPnlMap = /* @__PURE__ */ new Map();
    this.positions = null;
  }
  getQuantityTpslNoMap() {
    return this.quantityTpslNoMap;
  }
  getTpslPnlMap() {
    return this.tpslPnlMap;
  }
  getFormattedEstPnl(tpslOrder) {
    if (this.positions === null) {
      return "";
    }
    const position = this.positions[0];
    if (!position) {
      return "";
    }
    const { estPnl } = getTpslEstPnl(tpslOrder, position);
    return formatPnl(estPnl);
  }
  prepareTpslPnlMap(newPendingOrders) {
    const changed = [];
    newPendingOrders.forEach((order) => {
      const orderId = getOrderId(order);
      if (orderId && isActivatedTpslOrder(order)) {
        const prevPnl = this.tpslPnlMap.get(orderId);
        const newPnl = this.getFormattedEstPnl(order);
        if (prevPnl !== newPnl) {
          changed.push(orderId);
          this.tpslPnlMap.set(orderId, newPnl);
        }
      }
    });
    return changed;
  }
  prepareQuantityTpslNoMap(newPendingOrders) {
    this.quantityTpslNoMap = buildQuantityTpslNoMap(newPendingOrders);
  }
  recalculatePnl(positions, pendingOrders) {
    this.positions = positions;
    return this.prepareTpslPnlMap(pendingOrders);
  }
  clear() {
    this.positions = null;
    this.quantityTpslNoMap.clear();
    this.tpslPnlMap.clear();
  }
};

// src/tradingviewAdapter/renderer/orderLine.service.ts
var _OrderLineService = class _OrderLineService {
  constructor(instance, broker) {
    this.instance = instance;
    this.pendingOrderLineMap = /* @__PURE__ */ new Map();
    this.pendingOrders = [];
    this.broker = broker;
    this.tpslCalService = new TpslCalService();
  }
  renderPendingOrders(newPendingOrders) {
    if (newPendingOrders) {
      this.pendingOrders = newPendingOrders;
    }
    this.cleanOldPendingOrders(this.pendingOrders);
    this.tpslCalService.prepareTpslPnlMap(this.pendingOrders);
    this.tpslCalService.prepareQuantityTpslNoMap(this.pendingOrders);
    this.pendingOrders.forEach((order) => this.renderPendingOrder(order));
  }
  updatePositions(positions) {
    const changed = this.tpslCalService.recalculatePnl(
      positions,
      this.pendingOrders
    );
    this.pendingOrders.filter((order) => changed.includes(getOrderId(order))).forEach((order) => this.renderPendingOrder(order));
  }
  renderPendingOrder(order) {
    const orderId = _OrderLineService.getOrderId(order);
    if (!orderId) {
      return;
    }
    const orderLine = this.drawOrderLine(orderId, order);
    if (orderLine) {
      this.pendingOrderLineMap.set(orderId, orderLine);
    }
  }
  cleanOldPendingOrders(newPendingOrders) {
    const newOrderIdSet = new Set(
      newPendingOrders.map((order) => _OrderLineService.getOrderId(order))
    );
    this.pendingOrderLineMap.forEach(
      (_, orderId) => !newOrderIdSet.has(orderId) && this.removePendingOrder(orderId)
    );
  }
  removePendingOrder(orderId) {
    if (orderId === void 0) {
      return;
    }
    const orderLine = this.pendingOrderLineMap.get(orderId);
    if (orderLine) {
      this.pendingOrderLineMap.delete(orderId);
      orderLine.remove();
    }
  }
  /**
   * Creates a base order line with default styling.
   * Returns null if the chart is not ready (e.g., during initialization, hot reload, or chart switching).
   *
   * @returns IOrderLineAdapter instance or null if chart is not available
   */
  getBaseOrderLine() {
    try {
      const activeChart = this.instance.activeChart();
      if (!activeChart) {
        return null;
      }
      const colorConfig = this.broker.colorConfig;
      const orderLine = activeChart.createOrderLine();
      if (!orderLine) {
        return null;
      }
      return orderLine.setCancelTooltip(i18n.t("orders.cancelOrder")).setQuantityTextColor(colorConfig.qtyTextColor).setQuantityBackgroundColor(colorConfig.chartBG).setBodyBackgroundColor(colorConfig.chartBG).setCancelButtonBackgroundColor(colorConfig.chartBG).setLineStyle(1).setBodyFont(colorConfig.font).setQuantityFont(colorConfig.font);
    } catch (e) {
      const errorString = e?.toString() || String(e);
      if (errorString.includes("Value is null") || errorString.includes("tradingViewApi") || errorString.includes("Cannot read properties of null") || errorString.includes("Cannot read property") || e instanceof TypeError && errorString.includes("null")) {
        return null;
      }
      throw e;
    }
  }
  static getCombinationType(order) {
    const { algo_type: algoType, type } = order;
    if ((algoType === "STOP_LOSS" /* STOP_LOSS */ || algoType === "TAKE_PROFIT" /* TAKE_PROFIT */ || algoType === "STOP" /* STOP */) && type === "LIMIT" /* LIMIT */) {
      return "STOP_LIMIT" /* STOP_LIMIT */;
    }
    if ((algoType === "STOP_LOSS" /* STOP_LOSS */ || algoType === "TAKE_PROFIT" /* TAKE_PROFIT */ || algoType === "STOP" /* STOP */) && type === "MARKET" /* MARKET */) {
      return "STOP_MARKET" /* STOP_MARKET */;
    }
    if (algoType === "BRACKET" /* BRACKET */) {
      if (type === "LIMIT" /* LIMIT */) {
        return "BRACKET_LIMIT" /* BRACKET_LIMIT */;
      }
      if (type === "MARKET" /* MARKET */) {
        return "BRACKET_MARKET" /* BRACKET_MARKET */;
      }
    }
    if (algoType === "TRAILING_STOP" /* TRAILING_STOP */) {
      return "TRAILING_STOP" /* TRAILING_STOP */;
    }
    return "LIMIT" /* LIMIT */;
  }
  static getText(pendingOrder) {
    const orderCombinationType = _OrderLineService.getCombinationType(pendingOrder);
    if (orderCombinationType === "STOP_LIMIT" /* STOP_LIMIT */ || orderCombinationType === "STOP_MARKET" /* STOP_MARKET */ || orderCombinationType === "STOP_BRACKET_LIMIT" /* STOP_BRACKET_LIMIT */ || orderCombinationType === "STOP_BRACKET_MARKET" /* STOP_BRACKET_MARKET */) {
      if (pendingOrder.type === "LIMIT" /* LIMIT */) {
        return `${i18n.t("orderEntry.orderType.stopLimit")} ${commify(
          pendingOrder.price
        )}`;
      }
      return i18n.t("orderEntry.orderType.stopMarket");
    }
    if (orderCombinationType === "TRAILING_STOP" /* TRAILING_STOP */) {
      return i18n.t("orderEntry.trailing");
    }
    return i18n.t("orderEntry.orderType.limit");
  }
  static getOrderPrice(pendingOrder) {
    if (pendingOrder.algo_type === "TRAILING_STOP" /* TRAILING_STOP */) {
      return getTrailingStopPrice(pendingOrder);
    }
    return pendingOrder.trigger_price || pendingOrder.price;
  }
  getTPSLTextWithTpsl(text, pendingOrder) {
    const orderId = getOrderId(pendingOrder);
    if (!orderId) {
      return text;
    }
    return text;
  }
  getTPSLText(pendingOrder) {
    const tpslTypeText = getTpslTag(
      pendingOrder,
      this.tpslCalService.getQuantityTpslNoMap()
    );
    if (tpslTypeText) {
      return this.getTPSLTextWithTpsl(tpslTypeText, pendingOrder);
    }
    return null;
  }
  getOrderQuantity(pendingOrder) {
    if (pendingOrder.algo_order_id) {
      if (isActivatedPositionTpsl(pendingOrder) || isPositionTpsl(pendingOrder)) {
        return "100%";
      }
      if (isActivatedQuantityTpsl(pendingOrder)) {
        const qty = new Decimal(pendingOrder.quantity).minus(
          pendingOrder.executed ?? 0
        );
        const per = qty.div(new Decimal(pendingOrder.position_qty)).mul(100).todp(2).toNumber();
        return `${Math.min(Math.abs(per), 100).toString()}%`;
      }
    }
    return commify(new Decimal(pendingOrder.quantity).toString());
  }
  drawOrderLine(orderId, pendingOrder) {
    const text = isTpslOrder(pendingOrder) ? this.getTPSLText(pendingOrder) : _OrderLineService.getText(pendingOrder);
    if (text === null) {
      return null;
    }
    const colorConfig = this.broker.colorConfig;
    const existingOrderLine = this.pendingOrderLineMap.get(orderId);
    const orderLine = existingOrderLine ?? this.getBaseOrderLine();
    if (!orderLine) {
      return null;
    }
    const color = pendingOrder.side === "BUY" /* BUY */ ? colorConfig.upColor : colorConfig.downColor;
    pendingOrder.side === "BUY" /* BUY */ ? colorConfig.pnlUpColor : colorConfig.pnlDownColor;
    const price = _OrderLineService.getOrderPrice(pendingOrder);
    const lineLength = 100;
    const quantity = this.getOrderQuantity(pendingOrder);
    const textColor2 = colorConfig.textColor;
    orderLine.setText(text).setCancelButtonIconColor(colorConfig.closeIcon).setCancelButtonBorderColor(color).setBodyTextColor(textColor2).setBodyBorderColor(color).setQuantityBorderColor(color).setQuantityTextColor(color).setLineColor(color).setLineLength(lineLength).setQuantity(quantity ?? "").setPrice(price);
    if (this.broker.mode !== 3 /* MOBILE */) {
      orderLine.onCancel(null, () => this.broker.cancelOrder(pendingOrder));
      this.applyEditOnMove(orderLine, pendingOrder);
    } else {
      orderLine.setEditable(false).setCancellable(false);
    }
    return orderLine;
  }
  static getOrderEditKey(pendingOrder) {
    const orderCombinationType = this.getCombinationType(pendingOrder);
    if (["LIMIT" /* LIMIT */, "BRACKET_LIMIT" /* BRACKET_LIMIT */].includes(
      orderCombinationType
    )) {
      return "price";
    }
    if (orderCombinationType === "STOP_LIMIT" /* STOP_LIMIT */ || orderCombinationType === "STOP_MARKET" /* STOP_MARKET */ || pendingOrder.rootAlgoOrderAlgoType === "POSITIONAL_TP_SL" /* POSITIONAL_TP_SL */ || pendingOrder.rootAlgoOrderAlgoType === "TP_SL" /* TP_SL */ || pendingOrder.rootAlgoOrderAlgoType === "BRACKET" /* BRACKET */ || pendingOrder.rootAlgoOrderAlgoType === "STOP_BRACKET" /* STOP_BRACKET */) {
      return "trigger_price";
    }
  }
  applyEditOnMove(orderLine, pendingOrder) {
    const editKey = _OrderLineService.getOrderEditKey(pendingOrder);
    if (!editKey) {
      return;
    }
    orderLine.onMove(() => {
      this.broker.editOrder(pendingOrder, {
        type: editKey,
        value: `${orderLine.getPrice()}`
      }).then((res) => {
        if (!res.success) {
          this.renderPendingOrder(pendingOrder);
        }
      }).catch(() => this.renderPendingOrder(pendingOrder));
    });
  }
  removeAll() {
    this.pendingOrderLineMap.forEach((orderLine) => orderLine.remove());
    this.pendingOrderLineMap.clear();
    this.pendingOrders = [];
    this.tpslCalService.clear();
  }
};
_OrderLineService.getOrderId = (order) => {
  if (order === null || order === void 0) {
    return void 0;
  }
  return order.algo_order_id || order.order_id;
};
var OrderLineService = _OrderLineService;
var PositionLineService = class _PositionLineService {
  constructor(instance, broker) {
    this.instance = instance;
    this.currentSymbol = "";
    this.broker = broker;
    this.positionLines = {};
    this.lastPositions = null;
  }
  renderPositions(positions) {
    if (positions === null || positions.length === 0) {
      this.removePositions();
      return;
    }
    if (this.lastPositions?.length !== positions.length) {
      this.removePositions();
    }
    if (positions[0].symbol !== this.currentSymbol) {
      this.removePositions();
      this.currentSymbol = positions[0].symbol;
    }
    positions.forEach((position, idx) => this.drawPositionLine(position, idx));
    this.lastPositions = positions;
  }
  getBasePositionLine() {
    return this.instance.activeChart().createPositionLine().setTooltip(i18n.t("positions.closePosition")).setQuantityBackgroundColor(this.broker.colorConfig.chartBG).setCloseButtonBackgroundColor(this.broker.colorConfig.chartBG).setBodyTextColor(this.broker.colorConfig.textColor).setQuantityTextColor(this.broker.colorConfig.qtyTextColor).setBodyFont(this.broker.colorConfig.font).setQuantityFont(this.broker.colorConfig.font).setLineLength(100).setLineStyle(1);
  }
  static getPositionQuantity(balance) {
    return commify(new Decimal(balance).todp(4, Decimal.ROUND_DOWN).toString());
  }
  static getPositionPnL(unrealPnl, decimal) {
    const text = i18n.t("tpsl.pnl");
    const pnl = new Decimal(unrealPnl).toFixed(decimal, Decimal.ROUND_DOWN);
    if (new Decimal(unrealPnl).eq(0)) {
      return `${text} 0`;
    }
    if (new Decimal(unrealPnl).greaterThan(0)) {
      return `${text} +${commify(pnl)}`;
    }
    return `${text} ${commify(pnl)}`;
  }
  removePositions() {
    Object.keys(this.positionLines).forEach((lineId) => {
      this.positionLines[Number(lineId)].remove();
      delete this.positionLines[Number(lineId)];
    });
  }
  drawPositionLine(position, idx) {
    const colorConfig = this.broker.colorConfig;
    const isPositiveUnrealPnl = position.unrealPnl >= 0;
    const isPositiveBalance = position.balance >= 0;
    let pnlColor = colorConfig.pnlZoreColor;
    const pnlDecimal = new Decimal(position.unrealPnl);
    if (pnlDecimal.greaterThan(0)) {
      pnlColor = colorConfig.upColor;
    } else if (pnlDecimal.lessThan(0)) {
      pnlColor = colorConfig.downColor;
    }
    isPositiveUnrealPnl ? colorConfig.pnlUpColor : colorConfig.pnlDownColor;
    const sideColor = isPositiveBalance ? colorConfig.upColor : colorConfig.downColor;
    const price = new Decimal(position.open).toNumber();
    this.positionLines[idx] = this.positionLines[idx] ?? this.getBasePositionLine();
    this.positionLines[idx].setQuantity(_PositionLineService.getPositionQuantity(position.balance)).setPrice(price).setCloseButtonIconColor(colorConfig.closeIcon).setCloseButtonBorderColor(sideColor).setBodyBackgroundColor(pnlColor).setQuantityTextColor(sideColor).setBodyBorderColor(pnlColor).setLineColor(sideColor).setQuantityBorderColor(sideColor).setText(
      _PositionLineService.getPositionPnL(
        position.unrealPnl,
        position.unrealPnlDecimal
      )
    );
    if (this.broker.mode !== 3 /* MOBILE */) {
      this.positionLines[idx].onClose(null, () => {
        this.broker.closePosition(position);
      });
    }
  }
};
var DEFAULT_THRESHOLD = 10;
var TPSLService = class {
  constructor(instance, broker) {
    this.interactiveMode = 0 /* NONE */;
    this.tpslElRemoveTimer = null;
    this.currentPosition = null;
    this.tpslOrderLine = null;
    this.tpslPnLVerticalLineEntityId = null;
    this.tpslVerticalLineTime = null;
    this.tpslStartCircleEntityId = null;
    this.tpslEndCircleEntityId = null;
    this.threshold = 10;
    this.lastArgs = null;
    this.instance = instance;
    this.broker = broker;
    this.lastPositions = null;
    this.tpslElRemoveTimer = null;
    this.currentPosition = null;
    this.bindEvent();
  }
  bindEvent() {
    this.chart.crossHairMoved().subscribe(null, (args) => {
      this.lastArgs = args;
      if (this.interactiveMode === 2 /* TP_SL_DRAGGING */) {
        this.clearTpslElRemoveTimer();
        return;
      }
      const position = this.getIntersectantPosition(args);
      if (this.currentPosition && position && this.currentPosition.symbol === position.symbol) {
        return;
      }
      if (!position) {
        this.clearTPSLElements();
        return;
      }
      if (position) {
        this.clearTpslElRemoveTimer();
        this.currentPosition = position;
        this.createTPSLTriggerButton(args);
      }
    });
  }
  clearTpslElRemoveTimer() {
    if (this.tpslElRemoveTimer) {
      clearTimeout(this.tpslElRemoveTimer);
      this.tpslElRemoveTimer = null;
    }
  }
  showTPSLDialog(params) {
    const pnl = new Decimal(params.price).minus(this.currentPosition.open).mul(this.currentPosition?.balance ?? 0);
    modal.show("TPSLSimpleDialogId", {
      title: pnl.gt(0) ? i18n.t("tpsl.TPOrderConfirm") : i18n.t("tpsl.SLOrderConfirm"),
      triggerPrice: params.price,
      type: pnl.gt(0) ? "tp" : "sl",
      symbol: this.currentPosition.symbol,
      onComplete: () => {
        this.clearTPSLElements();
        this.chart.setScrollEnabled(true);
        this.chart.setZoomEnabled(true);
        this.interactiveMode = 0 /* NONE */;
      },
      showAdvancedTPSLDialog: (options) => {
        this.showAdvancedTPSLDialog({
          type: pnl.gt(0) ? "tp" : "sl",
          triggerPrice: params.price,
          qty: options.qty
        });
      }
    }).then(
      () => {
      },
      (err) => {
      }
    ).finally(() => {
      this.clearTPSLElements();
      this.chart.setScrollEnabled(true);
      this.chart.setZoomEnabled(true);
      this.interactiveMode = 0 /* NONE */;
    });
  }
  showAdvancedTPSLDialog({
    type,
    triggerPrice,
    qty
  }) {
    modal.show("TPSLDialogId", {
      withTriggerPrice: true,
      type,
      triggerPrice,
      symbol: this.currentPosition?.symbol,
      qty,
      onComplete: () => {
        this.clearTPSLElements();
        this.chart.setScrollEnabled(true);
        this.chart.setZoomEnabled(true);
        this.interactiveMode = 0 /* NONE */;
      }
    }).then(
      () => {
      },
      (err) => {
      }
    ).finally(() => {
      this.clearTPSLElements();
      this.chart.setScrollEnabled(true);
      this.chart.setZoomEnabled(true);
      this.interactiveMode = 0 /* NONE */;
    });
  }
  updatePositions(positions) {
    this.lastPositions = positions;
    this.threshold = this.generateThreshold();
  }
  /**
   * Generates a threshold value for position intersection detection.
   * The threshold is calculated as 2% of the visible price range but is capped
   * to ensure it never exceeds the actual price range (highest - lowest price).
   *
   * @returns The calculated threshold value, or DEFAULT_THRESHOLD if calculation fails
   */
  generateThreshold() {
    const priceScale = this.chart.getPanes()[0]?.getRightPriceScales()[0];
    if (!priceScale) {
      return DEFAULT_THRESHOLD;
    }
    try {
      const priceRange = priceScale.getVisiblePriceRange();
      if (!priceRange) {
        return DEFAULT_THRESHOLD;
      }
      const priceWidth = priceRange.to - priceRange.from;
      const calculatedThreshold = priceWidth * 0.02;
      const threshold = Math.min(calculatedThreshold, priceWidth);
      return threshold;
    } catch (error) {
      return DEFAULT_THRESHOLD;
    }
  }
  drawTPSL(params) {
    const { price } = params;
    const pnl = new Decimal(price).minus(this.currentPosition.open).mul(this.currentPosition?.balance ?? 0);
    const { tpslOrderLine, verticalLine } = this.ensureTPSLElements({
      price,
      pnl
    });
    const direction = pnl.gt(0) ? i18n.t("tpsl.tp") : i18n.t("tpsl.sl");
    const color = pnl.gt(0) ? this.broker.colorConfig.upColor : this.broker.colorConfig.downColor;
    tpslOrderLine?.setText(`${direction} ${pnl.toDecimalPlaces(2).toNumber()}`).setBodyTextColor(color).setBodyBorderColor(color).setLineColor(color);
    if (this.tpslVerticalLineTime) {
      verticalLine?.setPoints([
        { price: this.currentPosition?.open, time: this.tpslVerticalLineTime },
        { time: this.tpslVerticalLineTime, price }
      ]);
    }
  }
  ensureTPSLElements(params) {
    const tpslOrderLine = this.tpslOrderLine;
    let verticalLine;
    if (!tpslOrderLine) {
      this.tpslOrderLine = this.createTPSLOrderLine();
    }
    if (this.tpslPnLVerticalLineEntityId) {
      verticalLine = this.chart.getShapeById(this.tpslPnLVerticalLineEntityId);
    }
    if (!verticalLine) {
      if (!this.currentPosition || !this.tpslVerticalLineTime) {
        return {};
      }
      this.tpslPnLVerticalLineEntityId = this.chart.createMultipointShape(
        [
          {
            price: this.currentPosition.open,
            time: this.tpslVerticalLineTime
          },
          { time: this.tpslVerticalLineTime, price: params.price }
        ],
        {
          shape: "trend_line",
          // shape: "parallel_channel",
          lock: true,
          disableSave: true,
          disableSelection: true,
          disableUndo: true,
          zOrder: "top",
          overrides: {
            linecolor: "rgba(255,255,255, 0.2)",
            linewidth: 1,
            rightEnd: 1,
            leftEnd: 1
          }
        }
      );
    }
    verticalLine?.setProperties({
      linecolor: params.pnl.gt(0) ? this.broker.colorConfig.upColor : this.broker.colorConfig.downColor,
      linewidth: 1
    });
    return {
      tpslOrderLine,
      verticalLine
    };
  }
  createTPSLTriggerButton(params) {
    if (!this.tpslOrderLine) {
      this.tpslOrderLine = this.createTPSLOrderLine();
    }
    this.tpslOrderLine.onMove(() => {
      const price = this.tpslOrderLine?.getPrice();
      this.showTPSLDialog({ price: price ?? 0 });
    });
    this.tpslOrderLine.onMoving(() => {
      this.interactiveMode = 2 /* TP_SL_DRAGGING */;
      const price = this.tpslOrderLine?.getPrice();
      this.verticalLineTime();
      this.drawTPSL({ price: price ?? 0 });
    });
  }
  createTPSLOrderLine() {
    return this.chart.createOrderLine().setCancellable(false).setExtendLeft(true).setTooltip(i18n.t("tpsl.dragToSet")).setPrice(this.currentPosition.open).setLineLength(-200, "pixel").setText(i18n.t("common.tpsl")).setQuantity("").setBodyTextColor(this.broker.colorConfig.textColor).setBodyBackgroundColor(this.broker.colorConfig.chartBG).setBodyBorderColor(this.broker.colorConfig.pnlZoreColor).setQuantityBackgroundColor(this.broker.colorConfig.chartBG).setQuantityBorderColor(this.broker.colorConfig.pnlZoreColor).setQuantityTextColor(this.broker.colorConfig.qtyTextColor).setBodyFont(this.broker.colorConfig.font).setQuantityFont(this.broker.colorConfig.font).setLineStyle(3);
  }
  verticalLineTime() {
    const range = this.chart.getVisibleRange();
    this.tpslVerticalLineTime = this.getTimeAtPercentage(
      range.from,
      range.to,
      90
    );
  }
  clearTPSLElements() {
    if (this.tpslElRemoveTimer) {
      return;
    }
    this.tpslElRemoveTimer = setTimeout(() => {
      this.currentPosition = null;
      if (this.tpslOrderLine) {
        this.tpslOrderLine.remove();
        this.tpslOrderLine = null;
      }
      if (this.tpslPnLVerticalLineEntityId) {
        this.chart.removeEntity(this.tpslPnLVerticalLineEntityId);
        this.tpslPnLVerticalLineEntityId = null;
      }
      this.tpslElRemoveTimer = null;
    }, 100);
  }
  getIntersectantPosition(params) {
    if (!Array.isArray(this.lastPositions) || this.lastPositions.length === 0) {
      return null;
    }
    const { price, time } = params;
    let intersectantPosition = null;
    for (const position of this.lastPositions) {
      if (position) {
        if (Math.abs(position.open - price) < this.threshold) {
          intersectantPosition = position;
          break;
        }
      }
    }
    return intersectantPosition;
  }
  get chart() {
    return this.instance.activeChart();
  }
  getTimeAtPercentage(startTime, endTime, percentage) {
    if (typeof startTime !== "number" || typeof endTime !== "number" || typeof percentage !== "number") {
      return null;
    }
    if (percentage < 0 || percentage > 100) {
      return null;
    }
    if (startTime > endTime) {
      return null;
    }
    const duration = endTime - startTime;
    const factor = percentage / 100;
    const offset = duration * factor;
    const targetTime = startTime + offset;
    return Math.round(targetTime);
  }
};

// src/tradingviewAdapter/renderer/renderer.ts
var Renderer = class {
  constructor(instance, host, broker) {
    this.instance = instance;
    this.positionLineService = new PositionLineService(instance, broker);
    this.orderLineService = new OrderLineService(instance, broker);
    this.executionService = new ExecutionService(instance, broker);
    this.tpslService = new TPSLService(instance, broker);
  }
  async renderPositions(positions) {
    await this.chartReady();
    await this.onDataLoaded();
    this.positionLineService.renderPositions(positions);
    this.orderLineService.updatePositions(positions);
    this.tpslService.updatePositions(positions);
  }
  async renderPendingOrders(pendingOrders) {
    await this.chartReady();
    this.orderLineService.renderPendingOrders(pendingOrders);
  }
  async renderFilledOrders(filledOrders, basePriceDecimal) {
    await this.chartReady();
    await this.onDataLoaded();
    this.executionService.renderExecutions(filledOrders, basePriceDecimal);
  }
  remove() {
    this.orderLineService.removeAll();
    this.positionLineService.removePositions();
    this.executionService.destroy();
  }
  onDataLoaded() {
    if (this.instance.activeChart().symbolExt()) {
      return Promise.resolve();
    }
    return new Promise(
      (resolve) => this.instance.activeChart().onDataLoaded().subscribe(
        null,
        () => {
          resolve();
        },
        true
      )
    );
  }
  chartReady() {
    return new Promise(
      (resolve) => this.instance.onChartReady(() => {
        try {
          this.instance.activeChart().dataReady(() => resolve());
        } catch (err) {
          if (err.toString().includes("tradingViewApi")) ;
        }
      })
    );
  }
};

// src/tradingviewAdapter/hooks/useCreateRenderer.ts
function useCreateRenderer(symbol, displayControlSetting) {
  const [renderer, setRenderer] = useState();
  const rendererRef = useRef();
  const { state } = useAccount();
  const [unPnlPriceBasis] = useLocalStorage("unPnlPriceBasis", "markPrice");
  const [{ rows: positions }, positionsInfo] = usePositionStream(symbol, {
    calcMode: unPnlPriceBasis
  });
  const [pendingOrders] = useOrderStream({
    status: OrderStatus.INCOMPLETE,
    symbol
  });
  const config = useSymbolsInfo();
  const symbolInfo = config?.[symbol];
  const quote_dp = symbolInfo("quote_dp");
  const [fillOrders] = useOrderStream({
    symbol,
    status: OrderStatus.FILLED,
    size: 500
  });
  const createRenderer = useRef(
    (instance, host, broker, container) => {
      if (rendererRef.current) {
        rendererRef.current.remove();
      }
      rendererRef.current = new Renderer(instance, host, broker);
      setRenderer(rendererRef.current);
    }
  );
  const removeRenderer = useRef(() => {
    rendererRef.current?.remove();
    rendererRef.current = void 0;
  });
  useEffect(() => {
    if (state.status < AccountStatusEnum.EnableTrading && state.status !== AccountStatusEnum.EnableTradingWithoutConnected) {
      renderer?.renderPositions([]);
      return;
    }
    if (!displayControlSetting || !displayControlSetting.position) {
      renderer?.renderPositions([]);
      return;
    }
    const positionList = (positions ?? []).filter((_) => _.symbol === symbol).map((item) => {
      return {
        symbol: item.symbol,
        open: item.average_open_price,
        balance: item.position_qty,
        closablePosition: 9999,
        // @ts-ignore
        unrealPnl: item.unrealized_pnl ?? 0,
        interest: 0,
        unrealPnlDecimal: 2,
        basePriceDecimal: 4
      };
    });
    renderer?.renderPositions(positionList);
  }, [renderer, positions, symbol, displayControlSetting, state]);
  useEffect(() => {
    if (!displayControlSetting || !displayControlSetting.buySell) {
      renderer?.renderFilledOrders([], 6);
      return;
    }
    const currentSymbolFillOrders = fillOrders?.filter(
      (item) => item.symbol === symbol
    );
    renderer?.renderFilledOrders(currentSymbolFillOrders ?? [], quote_dp ?? 6);
  }, [renderer, fillOrders, symbol, quote_dp, displayControlSetting]);
  useEffect(() => {
    let tpslOrder = [];
    let positionTpsl = [];
    let limitOrder = [];
    let stopOrder = [];
    let bracketOrder = [];
    let trailingStopOrder = [];
    if (state.status < AccountStatusEnum.EnableTrading && state.status !== AccountStatusEnum.EnableTradingWithoutConnected) {
      renderer?.renderPendingOrders([]);
      return;
    }
    const symbolPosition = (positions ?? []).find(
      (item) => item.symbol === symbol
    );
    pendingOrders?.forEach((order) => {
      if (symbol !== order.symbol) {
        return;
      }
      if (!order.algo_order_id) {
        limitOrder.push(order);
      } else if (order.algo_order_id) {
        if (order.algo_type === "POSITIONAL_TP_SL" /* POSITIONAL_TP_SL */) {
          for (const child_order of order.child_orders) {
            child_order.root_algo_order_algo_type = order.algo_type;
            if (child_order.trigger_price && child_order.status !== OrderStatus.FILLED) {
              positionTpsl.push(child_order);
            }
          }
        } else if (order.algo_type === "TP_SL" /* TP_SL */) {
          if (symbolPosition) {
            for (const child_order of order.child_orders) {
              child_order.root_algo_order_algo_type = order.algo_type;
              child_order.position_qty = symbolPosition.position_qty;
              if (child_order.trigger_price && child_order.status !== OrderStatus.FILLED) {
                tpslOrder.push(child_order);
              }
            }
          }
        } else if (order.algo_type === "STOP_LOSS" /* STOP_LOSS */ || order.algo_type === "TAKE_PROFIT" /* TAKE_PROFIT */) {
          stopOrder.push(order);
        } else if (order.algo_type === "BRACKET" /* BRACKET */) {
          bracketOrder.push(order);
        } else if (order.algo_type === "TRAILING_STOP" /* TRAILING_STOP */) {
          if (order.is_activated && order.extreme_price) {
            trailingStopOrder.push(order);
          }
        }
      }
    });
    if (displayControlSetting) {
      if (!displayControlSetting.positionTpsl) {
        positionTpsl = [];
      }
      if (!displayControlSetting.tpsl) {
        tpslOrder = [];
      }
      if (!displayControlSetting.limitOrders) {
        limitOrder = [];
        bracketOrder = [];
      }
      if (!displayControlSetting.stopOrders) {
        stopOrder = [];
      }
      if (!displayControlSetting.trailingStop) {
        trailingStopOrder = [];
      }
    }
    renderer?.renderPendingOrders(
      tpslOrder.concat(positionTpsl).concat(limitOrder).concat(stopOrder).concat(bracketOrder).concat(trailingStopOrder)
    );
  }, [
    renderer,
    pendingOrders,
    symbol,
    displayControlSetting,
    positions,
    state.status
  ]);
  return [createRenderer.current, removeRenderer.current];
}

// src/tradingviewAdapter/renderer/brokerHostHandler.ts
function preventDefaultRenderHack(host) {
  host.setBrokerConnectionAdapter = function(adapter) {
    const delegate = {
      subscribe: () => {
      },
      unsubscribe: () => {
      },
      unsubscribeAll: () => {
      }
    };
    Object.defineProperty(adapter, "_ordersCache", {
      get: function() {
        return {
          start: () => {
          },
          stop: () => {
          },
          update: () => {
          },
          partialUpdate: () => {
          },
          fullUpdate: () => {
          },
          getObjects: async () => [],
          updateDelegate: delegate,
          partialUpdateDelegate: delegate
        };
      },
      set: () => {
      }
    });
    adapter._waitForOrderModification = async () => true;
    this._adapter = adapter;
  };
}
function forceSilentOrdersPlacement(instance, host) {
  instance.onChartReady(() => {
    host.silentOrdersPlacement().subscribe((val) => {
      if (!val) {
        host.silentOrdersPlacement().setValue(true);
        if (instance) {
          instance._iFrame.contentDocument.querySelector(
            ".wrapper-3X2QgaDd"
          ).className = "wrapper-3X2QgaDd highButtons-3X2QgaDd";
        }
        host.sellBuyButtonsVisibility()?.setValue(false);
      }
    });
  });
}
function brokerHostHandler(instance, host) {
  preventDefaultRenderHack(host);
  forceSilentOrdersPlacement(instance, host);
}

// src/tradingviewAdapter/widget/util.ts
var waitForElm = (iframeDocument, selector) => {
  return new Promise((resolve) => {
    const initialIframe = iframeDocument.querySelector(selector);
    if (initialIframe) {
      resolve(initialIframe);
    }
    const observer = new MutationObserver(() => {
      const iframe = iframeDocument.querySelector(selector);
      if (iframe) {
        resolve(iframe);
        observer.disconnect();
      }
    });
    observer.observe(iframeDocument, {
      childList: true,
      subtree: true
    });
  });
};

// src/tradingviewAdapter/widget/chart_hack.ts
var ChartHack = class {
  constructor({ iframeDocument }) {
    this.iframeDocument = iframeDocument;
  }
  defaultHack() {
    this.showFavoriteStarByDefault();
  }
  showFavoriteStarByDefault() {
    waitForElm(this.iframeDocument, ".dropdown-2R6OKuTS").then(() => {
      const items = this.iframeDocument.querySelectorAll(
        ".toolbox-2IihgTnv.showOnHover-2IihgTnv"
      );
      items.forEach((item) => {
        item.style.opacity = "1";
      });
    });
  }
};

// src/tradingviewAdapter/widget/option.ts
var getDisabledFeatures = (mode) => {
  let disabledFeatures = [
    "header_symbol_search",
    "volume_force_overlay",
    "trading_account_manager",
    "drawing_templates",
    "open_account_manager",
    "right_toolbar",
    "support_multicharts",
    "header_layouttoggle",
    "order_panel",
    "order_info",
    "trading_notifications",
    "display_market_status",
    "broker_button",
    "add_to_watchlist",
    "chart_crosshair_menu",
    "header_fullscreen_button",
    "header_widget"
  ];
  if (mode === 3 /* MOBILE */) {
    disabledFeatures = [
      ...disabledFeatures,
      "left_toolbar",
      "timeframes_toolbar",
      "go_to_date",
      "timezone_menu",
      // add volume back
      "create_volume_indicator_by_default",
      "buy_sell_buttons"
    ];
  }
  if (mode === 0 /* BASIC */) {
    disabledFeatures = [
      ...disabledFeatures,
      "header_widget",
      "left_toolbar",
      "timeframes_toolbar",
      "buy_sell_buttons"
    ];
  } else if (mode === 1 /* ADVANCED */) {
    disabledFeatures = [
      ...disabledFeatures,
      "left_toolbar",
      "timeframes_toolbar",
      "buy_sell_buttons"
    ];
  }
  return disabledFeatures;
};
function getOptions(options, mode, externalEnabledFeatures, externalDisabledFeatures) {
  const defaultEnabledFeatures = [
    "hide_left_toolbar_by_default",
    "order_panel_close_button",
    "iframe_loading_compatibility_mode"
  ];
  const mergedEnabledFeatures = Array.from(
    /* @__PURE__ */ new Set([...defaultEnabledFeatures, ...externalEnabledFeatures || []])
  );
  const defaultDisabledFeatures = getDisabledFeatures(mode);
  const mergedDisabledFeatures = Array.from(
    /* @__PURE__ */ new Set([...defaultDisabledFeatures, ...externalDisabledFeatures || []])
  );
  return {
    ...options,
    disabled_features: mergedDisabledFeatures,
    enabled_features: mergedEnabledFeatures,
    auto_save_delay: 0.1,
    broker_config: {
      configFlags: {
        supportStopLimitOrders: true,
        supportReversePosition: false
      }
    }
  };
}

// src/tradingviewAdapter/widget/persistUtils.ts
var getChartSettingAdapterKey = (chartKey) => `${chartKey}_adapter`;
var defaultSettings = {
  "trading.chart.proterty": JSON.stringify({
    showSellBuyButtons: 0,
    noConfirmEnabled: 1,
    qweqrq: 0,
    showPricesWithZeroVolume: 1,
    showSpread: 1,
    orderExecutedSoundParams: '{"enabled":0,"name":"alert/alarm_clock"}'
  }),
  "hint.startFocusedZoom": "true"
};
var chartCacheMap = /* @__PURE__ */ new Map();
var parseAndCacheChartData = (chartKey, savedDataString, adapterSettingString) => {
  const settingAdapterKey = getChartSettingAdapterKey(chartKey);
  try {
    const transformedSavedDataString = savedDataString;
    const savedData = transformedSavedDataString ? JSON.parse(transformedSavedDataString) : void 0;
    const adapterSetting = adapterSettingString ? JSON.parse(adapterSettingString) : defaultSettings;
    if (transformedSavedDataString) {
      chartCacheMap.set(chartKey, transformedSavedDataString);
    }
    if (adapterSettingString) {
      chartCacheMap.set(settingAdapterKey, adapterSettingString);
    }
    return { savedData, adapterSetting };
  } catch (e) {
  }
  return { savedData: void 0, adapterSetting: defaultSettings };
};
var getChartData = async (chartKey, isLoggedIn) => {
  const settingAdapterKey = getChartSettingAdapterKey(chartKey);
  const localStorageSavedData = localStorage.getItem(chartKey) || "";
  const localStorageAdapterSetting = localStorage.getItem(settingAdapterKey) || "";
  try {
    if (chartCacheMap.has(chartKey) && chartCacheMap.has(settingAdapterKey)) {
      return {
        savedData: JSON.parse(chartCacheMap.get(chartKey)),
        adapterSetting: JSON.parse(chartCacheMap.get(settingAdapterKey))
      };
    }
  } catch (e) {
  }
  return parseAndCacheChartData(
    chartKey,
    localStorageSavedData,
    localStorageAdapterSetting
  );
};
var saveChartData = async (chartKey, setting, isLoggedIn) => {
  if (setting) {
    localStorage.setItem(chartKey, setting);
    chartCacheMap.set(chartKey, setting);
  }
};
var saveChartAdapterSetting = async (chartKey, setting, isLoggedIn) => {
  const settingAdapterKey = getChartSettingAdapterKey(chartKey);
  if (setting) {
    localStorage.setItem(settingAdapterKey, setting);
    chartCacheMap.set(settingAdapterKey, setting);
  }
};

// src/tradingviewAdapter/widget/index.ts
var debounce = (func, delay) => {
  let timer = null;
  const debounced = (...args) => {
    timer && window.clearTimeout(timer);
    timer = setTimeout(() => {
      func(...args);
    }, delay);
  };
  debounced.cancel = () => {
    timer && window.clearTimeout(timer);
    timer = null;
  };
  return debounced;
};
var DEFAULT_SETTINGS_KEY = "chartProp_default";
var DEBOUNCE_SAVE_TIME = 300;
var Widget = class {
  constructor(props) {
    this._instance = null;
    this._onClick = null;
    this._datafeed = null;
    this._chartKey = DEFAULT_SETTINGS_KEY;
    this._adapterSetting = defaultSettings;
    this._savedData = null;
    this._isLoggedIn = false;
    this.debounceSaveChart = debounce(() => {
      try {
        this._instance?.save((chartProps) => {
          if (!Object.is(this._savedData, chartProps)) {
            this._savedData = chartProps;
            saveChartData(
              this._chartKey,
              JSON.stringify(chartProps),
              this._isLoggedIn
            );
          }
        });
      } catch (e) {
      }
    }, DEBOUNCE_SAVE_TIME * 2);
    this.debounceSaveChartAdapterSetting = debounce(() => {
      saveChartAdapterSetting(
        this._chartKey,
        JSON.stringify(this._adapterSetting),
        this._isLoggedIn
      );
    }, DEBOUNCE_SAVE_TIME);
    this._create(props);
  }
  remove() {
    this.unsubscribeClick();
    this._datafeed?.remove();
    this._broker?.remove();
    this._instance?.remove();
    this.debounceSaveChart.cancel();
    this.debounceSaveChartAdapterSetting.cancel();
  }
  updateOverrides(overrides) {
    if (!this.instance) {
      return;
    }
    this.instance.applyOverrides(overrides);
  }
  setSymbol(symbol, interval, callback) {
    try {
      this._instance?.onChartReady(() => {
        let currentInterval = interval ?? this._instance?.symbolInterval()?.interval;
        if (!currentInterval) {
          currentInterval = 1;
        }
        this._instance?.setSymbol(symbol, currentInterval, callback);
      });
    } catch (e) {
    }
  }
  setResolution(resolution) {
    if (this._instance) {
      this._instance.activeChart().setResolution(resolution);
    }
  }
  executeActionById(actionId) {
    try {
      this._instance?.onChartReady(() => {
        this._instance?.activeChart().executeActionById(actionId);
      });
    } catch (e) {
    }
  }
  changeLineType(lineType) {
    try {
      this._instance?.onChartReady(() => {
        this._instance?.activeChart().setChartType(lineType);
      });
    } catch (e) {
    }
  }
  subscribeClick(onClick) {
    this._onClick = onClick;
    this._instance?.onChartReady(() => {
      this._instance?._iFrame.contentDocument?.addEventListener(
        "click",
        this._onClick
      );
    });
  }
  unsubscribeClick() {
    this._instance?._iFrame.contentDocument?.removeEventListener(
      "click",
      this._onClick
    );
  }
  get instance() {
    return this._instance;
  }
  chartHack() {
    this._instance?.onChartReady(() => {
      const iframeDocument = this._instance._iFrame.contentWindow.document;
      new ChartHack({ iframeDocument }).defaultHack();
    });
  }
  subscribeAutoSave() {
    this._instance?.onChartReady(() => {
      this._instance?.subscribe("onAutoSaveNeeded", () => {
        this.debounceSaveChart();
      });
      this._instance?.activeChart().onVisibleRangeChanged().subscribe(null, () => {
        this.debounceSaveChart();
      });
    });
  }
  async _create({
    options,
    chartKey,
    mode,
    onClick,
    enabled_features,
    disabled_features
  }) {
    const getBroker = options.getBroker;
    const widgetOptions = {
      fullscreen: options.fullscreen ?? true,
      autosize: options.autosize ?? false,
      timezone: options.timezone,
      symbol: options.symbol,
      library_path: options.libraryPath,
      interval: options.interval ?? "1",
      custom_css_url: options.customCssUrl,
      custom_font_family: options.customFontFamily,
      datafeed: options.datafeed,
      studies_overrides: options.studiesOverrides,
      locale: options.locale,
      theme: options.theme,
      loading_screen: options.loadingScreen,
      overrides: options.overrides,
      container: options.container,
      favorites: {
        intervals: [
          "1",
          "3",
          "5",
          "15",
          "30",
          "60",
          "240",
          "1d",
          "1W",
          "1M"
        ],
        chartTypes: ["Area", "Line"]
      },
      broker_factory: getBroker ? (host) => {
        if (this._broker) {
          this._broker.remove();
        }
        this._broker = getBroker(
          this._instance,
          host
        );
        return this._broker;
      } : void 0,
      custom_indicators_getter: options.customIndicatorsGetter
    };
    this._datafeed = options.datafeed;
    if (chartKey) {
      this._chartKey = chartKey;
    }
    const { savedData, adapterSetting } = await getChartData(
      this._chartKey,
      this._isLoggedIn
    );
    this._adapterSetting = adapterSetting;
    this._savedData = savedData;
    const mergedOptions = getOptions(
      widgetOptions,
      mode,
      enabled_features,
      disabled_features
    );
    this._instance = new TradingView.widget({
      ...mergedOptions,
      ...options.customIndicatorsGetter != null && {
        custom_indicators_getter: options.customIndicatorsGetter
      },
      interval: adapterSetting["chart.lastUsedTimeBasedResolution"] ?? widgetOptions.interval,
      saved_data: savedData,
      settings_adapter: {
        initialSettings: adapterSetting,
        setValue: (key, value) => {
          this._adapterSetting = { ...this._adapterSetting, [key]: value };
          this.debounceSaveChartAdapterSetting();
        },
        removeValue: () => {
        }
      }
    });
    this._instance.onChartReady(() => {
      if (options.symbol && this._instance?.activeChart().symbol() !== withoutExchangePrefix(options.symbol)) {
        this.setSymbol(options.symbol);
      }
    });
    this.subscribeAutoSave();
    this.subscribeClick(onClick);
    this.chartHack();
  }
};

// src/components/tradingview.script.ts
var CHART_KEY = "SDK_Tradingview";
var MOBILE_CHART_KEY = "SDK_Moblie_Tradingview";
var getChartKey = (isMobile) => {
  return isMobile ? MOBILE_CHART_KEY : CHART_KEY;
};
var defaultLocale = (localeCode) => {
  return localeCode === "id" ? "id_ID" : localeCode === "tc" ? "zh_TW" : localeCode;
};
function useTradingviewScript(props) {
  const {
    scriptSRC: tradingViewScriptSrc,
    libraryPath,
    customCssUrl: tradingViewCustomCssUrl,
    overrides: customerOverrides,
    studiesOverrides: customerStudiesOverrides,
    symbol,
    theme,
    loadingScreen: customerLoadingScreen,
    mode,
    colorConfig: customerColorConfig,
    locale = defaultLocale,
    classNames,
    enabled_features,
    disabled_features,
    customIndicatorsGetter
  } = props;
  const localeCode = useLocaleCode();
  const chart = useRef(null);
  const apiBaseUrl = useConfig("apiBaseUrl");
  const { state: accountState } = useAccount();
  const [side, setSide] = useState(OrderSide.SELL);
  const symbolsInfo = useSymbolsInfo();
  const [fullscreen, setFullscreen] = useLocalStorage(
    TradingviewFullscreenKey,
    false
  );
  const { onSubmit, submitting } = useOrderEntry_deprecated(
    {
      symbol: symbol ?? "",
      side,
      order_type: OrderType.MARKET
    },
    {
      watchOrderbook: true
    }
  );
  const [displayControlState, setDisplayControlState] = useState(() => {
    const displaySettingInfo = localStorage.getItem(
      TradingViewSDKLocalstorageKey.displayControlSetting
    );
    if (displaySettingInfo) {
      return JSON.parse(displaySettingInfo);
    }
    return {
      position: true,
      buySell: true,
      limitOrders: true,
      stopOrders: true,
      tpsl: true,
      positionTpsl: true,
      trailingStop: true
    };
  });
  const [interval, setInterval] = useState(() => {
    const lastUsedInterval = localStorage.getItem(
      TradingViewSDKLocalstorageKey.interval
    );
    if (!lastUsedInterval) {
      return "15";
    }
    return lastUsedInterval;
  });
  const [lineType, setLineType] = useState(() => {
    const lastUsedLineType = localStorage.getItem(
      TradingViewSDKLocalstorageKey.lineType
    );
    if (!lastUsedLineType) {
      return "1";
    }
    return lastUsedLineType;
  });
  const isMobile = useMediaQuery(MEDIA_TABLET);
  const colorConfig = useMemo(
    () => Object.assign({}, defaultColorConfig, customerColorConfig ?? {}),
    [customerColorConfig]
  );
  const loadingScreen = useMemo(() => {
    if (typeof customerLoadingScreen === "object") {
      return customerLoadingScreen;
    }
    return {
      backgroundColor: chartBG
    };
  }, [customerLoadingScreen]);
  const ws = useWS();
  const [chartingLibrarySciprtReady, setChartingLibrarySciprtReady] = useState(false);
  const closePositionConfirmCallback = (data) => {
    const symbolInfo = symbolsInfo[symbol];
    if (!symbolInfo) {
      return;
    }
    const side2 = new Decimal(data.balance).greaterThan(0) ? OrderSide.SELL : OrderSide.BUY;
    const order = {
      //   order_price: undefined,
      order_quantity: new Decimal(data.balance).abs().toNumber(),
      symbol,
      order_type: OrderType.MARKET,
      side: side2,
      reduce_only: true
    };
    setSide(side2);
    modal.show("MarketCloseConfirmID", {
      base: symbolInfo("base"),
      quantity: data.balance,
      onConfirm: async () => {
        return onSubmit(order).catch((error) => {
          if (typeof error === "string") {
            toast.error(error);
          } else {
            toast.error(error.message);
          }
        });
      },
      submitting
    });
  };
  const chartRef = useRef(null);
  const isLoggedIn = useMemo(() => {
    if (accountState.status < AccountStatusEnum.EnableTrading && accountState.status !== AccountStatusEnum.EnableTradingWithoutConnected) {
      return false;
    }
    return true;
  }, [accountState]);
  const broker = useBroker_default({
    closeConfirm: closePositionConfirmCallback,
    colorConfig,
    onToast: toast,
    symbol: symbol ?? "",
    mode
  });
  const [createRenderer, removeRenderer] = useCreateRenderer(
    symbol,
    displayControlState
  );
  const onFullScreenChange = () => {
    if (fullscreen) {
      setFullscreen(false);
    } else {
      setFullscreen(true);
    }
    props.onFullScreenChange?.(!fullscreen);
  };
  const changeInterval = (newInterval) => {
    if (!chart.current) {
      return;
    }
    localStorage.setItem(TradingViewSDKLocalstorageKey.interval, newInterval);
    setInterval(newInterval);
    chart.current?.setResolution(newInterval);
  };
  const changeLineType = (newLineType) => {
    if (!chart.current) {
      return;
    }
    localStorage.setItem(TradingViewSDKLocalstorageKey.lineType, newLineType);
    setLineType(newLineType);
    chart.current?.changeLineType(Number(newLineType));
  };
  const changeDisplaySetting = (newSetting) => {
    localStorage.setItem(
      TradingViewSDKLocalstorageKey.displayControlSetting,
      JSON.stringify(newSetting)
    );
    setDisplayControlState(newSetting);
  };
  const openChartSetting = () => {
    if (!chart.current) {
      return;
    }
    chart.current.executeActionById("chartProperties");
  };
  const openChartIndicators = () => {
    if (!chart.current) {
      return;
    }
    chart.current.executeActionById("insertIndicator");
  };
  useEffect(() => {
    if (!tradingViewScriptSrc) {
      return;
    }
    if (!chartRef.current) {
      return;
    }
    if (!chartingLibrarySciprtReady) {
      const script = document.createElement("script");
      script.setAttribute("data-nscript", "afterInteractive");
      script.src = tradingViewScriptSrc;
      script.async = true;
      script.type = "text/javascript";
      script.onload = () => {
        setChartingLibrarySciprtReady(true);
      };
      script.onerror = () => {
      };
      chartRef.current.appendChild(script);
    }
  }, [chartRef, chartingLibrarySciprtReady, tradingViewScriptSrc]);
  useEffect(() => {
    if (!symbol) {
      return;
    }
    if (!chartingLibrarySciprtReady || !tradingViewScriptSrc) {
      return;
    }
    const defaultOverrides = getOveriides(colorConfig, isMobile);
    const overrides = customerOverrides ? Object.assign({}, defaultOverrides.overrides, customerOverrides) : defaultOverrides.overrides;
    const studiesOverrides = customerStudiesOverrides ? Object.assign(
      {},
      defaultOverrides.studiesOverrides,
      customerStudiesOverrides
    ) : defaultOverrides.studiesOverrides;
    if (chartRef.current) {
      const options = {
        // fullscreen: fullscreen ?? false,
        fullscreen: false,
        autosize: true,
        symbol: withExchangePrefix(symbol),
        locale: typeof locale === "function" ? locale(localeCode) : locale,
        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
        container: chartRef.current,
        libraryPath,
        customCssUrl: tradingViewCustomCssUrl,
        interval: interval ?? "1",
        theme: theme ?? "dark",
        loadingScreen: loadingScreen ?? {},
        overrides,
        studiesOverrides,
        datafeed: new Datafeed(apiBaseUrl, ws),
        contextMenu: {
          items_processor: async (defaultItems) => {
            return defaultItems;
          }
        },
        // todo broker effect sell/buy
        getBroker: (instance, host) => {
          brokerHostHandler(instance, host);
          return getBrokerAdapter_default(host, broker);
        },
        customIndicatorsGetter
      };
      const chartProps = {
        options,
        chartKey: getChartKey(isMobile),
        mode,
        onClick: () => {
        },
        enabled_features,
        disabled_features
      };
      chart.current = new Widget(chartProps);
    }
    return () => {
      chart.current?.remove();
    };
  }, [
    chartingLibrarySciprtReady,
    isMobile,
    mode,
    chart,
    chartRef,
    chartingLibrarySciprtReady,
    tradingViewScriptSrc,
    colorConfig,
    locale,
    localeCode,
    customIndicatorsGetter
  ]);
  useEffect(() => {
    ws.on(
      "status:change",
      (message) => {
        if (!message.isPrivate && message.isReconnect) {
          if (typeof window.onResetCacheNeededCallback === "function") {
            window.onResetCacheNeededCallback();
            if (chart.current?.instance) {
              chart.current?.instance.activeChart()?.resetData();
            }
          }
        }
      },
      "tradingview"
    );
  }, [ws]);
  useEffect(() => {
    if (chart.current && chart.current?.instance) {
      chart.current?.instance?.onChartReady(() => {
        if (isLoggedIn && chart.current?.instance) {
          createRenderer(
            chart.current.instance,
            void 0,
            broker,
            chartRef.current
          );
        }
      });
    }
    return () => {
      removeRenderer();
    };
  }, [chart.current, isLoggedIn]);
  useEffect(() => {
    if (!symbol || !chart.current) {
      return;
    }
    chart.current?.setSymbol(symbol);
    const service = new WebsocketService(ws);
    service.subscribeSymbol(symbol);
    return () => {
      service.unsubscribeKline(symbol);
    };
  }, [symbol]);
  return {
    tradingViewScriptSrc,
    chartRef,
    changeDisplaySetting,
    displayControlState,
    interval,
    changeInterval,
    lineType,
    changeLineType,
    openChartSetting,
    openChartIndicators,
    symbol,
    onFullScreenChange,
    classNames,
    fullscreen
  };
}

// src/components/tradingview.ui.tsx
init_icons();
var Link = (props) => {
  const { url, children } = props;
  return /* @__PURE__ */ jsx(
    "span",
    {
      onClick: () => window.open(url),
      className: "oui-cursor-pointer oui-px-0.5 oui-text-primary-light oui-underline",
      children
    }
  );
};
var NoTradingview = () => {
  const { t } = useTranslation();
  return /* @__PURE__ */ jsx("div", { className: "oui-absolute oui-inset-0 oui-z-0 oui-flex oui-flex-col oui-items-center oui-justify-start oui-p-2 oui-text-base-contrast-80 md:oui-justify-center md:oui-p-10", children: /* @__PURE__ */ jsxs("div", { children: [
    /* @__PURE__ */ jsx("p", { className: "oui-mb-6 oui-text-xs", children: t("tradingView.noScriptSrc") }),
    /* @__PURE__ */ jsx("p", { className: "oui-mb-3 oui-pl-0 oui-text-2xs oui-text-base-contrast-54 md:oui-pl-2 md:oui-text-base md:oui-text-base-contrast-80", children: /* @__PURE__ */ jsx(
      Trans,
      {
        i18nKey: "tradingView.noScriptSrc.1",
        components: [
          /* @__PURE__ */ jsx(
            Link,
            {
              url: "https://www.tradingview.com/advanced-charts"
            },
            "tradingview-advanced-charts"
          )
        ]
      }
    ) }),
    /* @__PURE__ */ jsx("p", { className: "oui-pl-0 oui-text-2xs oui-text-base-contrast-54 md:oui-pl-2 md:oui-text-base md:oui-text-base-contrast-80", children: /* @__PURE__ */ jsx(
      Trans,
      {
        i18nKey: "tradingView.noScriptSrc.2",
        components: [
          /* @__PURE__ */ jsx(
            Link,
            {
              url: "https://orderly.network/docs/sdks/react/components/trading#tradingviewconfig"
            },
            "tradingview-config"
          )
        ]
      }
    ) })
  ] }) });
};
var TopBar = (props) => {
  return /* @__PURE__ */ jsx("div", { className: "top-toolbar oui-flex oui-h-[44px] oui-justify-between md:oui-justify-start oui-items-center oui-p-2 md:oui-px-3 md:oui-pt-3 md:oui-pb-[14px]", children: props.children });
};
var topBar_default = TopBar;
var LazyLineType = React3.lazy(() => Promise.resolve().then(() => (init_lineType(), lineType_exports)));
var LazyTimeInterval = React3.lazy(
  () => Promise.resolve().then(() => (init_timeInterval(), timeInterval_exports)).then((mod) => ({ default: mod.TimeInterval }))
);
var LazyMobileDisplayControl = React3.lazy(
  () => Promise.resolve().then(() => (init_displayControl(), displayControl_exports)).then((mod) => ({
    default: mod.MobileDisplayControl
  }))
);
var LazyDesktopDisplayControl = React3.lazy(
  () => Promise.resolve().then(() => (init_displayControl(), displayControl_exports)).then((mod) => ({
    default: mod.DesktopDisplayControl
  }))
);
var OperateButton = ({ children, onClick }) => {
  return /* @__PURE__ */ jsx(
    Box,
    {
      onClick,
      className: "oui-cursor-pointer oui-w-[18px] oui-h-[18px] oui-text-base-contrast-36  hover:oui-text-base-contrast-80",
      children
    }
  );
};
var ZoomOutIcon = (props) => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "18",
      height: "18",
      viewBox: "0 0 18 18",
      fill: "currentColor",
      ...props,
      children: /* @__PURE__ */ jsx("path", { d: "M15.0008 2.24304C14.8088 2.24304 14.6085 2.30755 14.4615 2.4538L11.2508 5.66455V3.74304H9.75079V7.49304C9.75079 7.90704 10.0868 8.24304 10.5008 8.24304H14.2508V6.74304H12.3285L15.54 3.53229C15.8325 3.23904 15.8325 2.74705 15.54 2.4538C15.393 2.30755 15.1928 2.24304 15.0008 2.24304ZM3.7508 9.74303V11.243H5.67231L2.46156 14.4538C2.16906 14.747 2.16906 15.239 2.46156 15.5323C2.75481 15.8248 3.2468 15.8248 3.54005 15.5323L6.7508 12.3215V14.243H8.25079V10.493C8.25079 10.079 7.9148 9.74303 7.5008 9.74303H3.7508Z" })
    }
  );
};
var ZoomInIcon = (props) => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: "18",
      height: "18",
      viewBox: "0 0 18 18",
      fill: "currentColor",
      ...props,
      children: /* @__PURE__ */ jsx("path", { d: "M7.49219 9.74304C7.30026 9.74304 7.09964 9.80755 6.95309 9.9538L3.74219 13.1646V11.243H2.24219V14.993C2.24219 15.407 2.57796 15.743 2.99219 15.743H6.74219V14.243H4.82031L8.03121 11.0323C8.32416 10.739 8.32416 10.247 8.03121 9.9538C7.88481 9.80755 7.68404 9.74304 7.49219 9.74304ZM11.2509 2.24304V3.74304H13.1728L9.96186 6.9538C9.66899 7.24705 9.66899 7.73904 9.96186 8.03229C10.2547 8.32479 10.7471 8.32479 11.04 8.03229L14.2509 4.82153V6.74304H15.7509V2.99304C15.7509 2.57904 15.4151 2.24304 15.0009 2.24304H11.2509Z" })
    }
  );
};
var TradingviewUI = forwardRef((props, ref) => {
  const {
    chartRef,
    interval,
    changeDisplaySetting,
    displayControlState,
    tradingViewScriptSrc,
    changeInterval,
    lineType,
    changeLineType,
    openChartSetting,
    openChartIndicators,
    onFullScreenChange
  } = props;
  const isMobile = useMediaQuery(MEDIA_TABLET);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: cn("oui-relative oui-size-full", props.classNames?.root),
      children: !tradingViewScriptSrc ? /* @__PURE__ */ jsx(NoTradingview, {}) : /* @__PURE__ */ jsxs(
        "div",
        {
          className: cn(
            "oui-absolute oui-inset-0 oui-z-[1] oui-flex oui-flex-col",
            props.classNames?.content
          ),
          children: [
            /* @__PURE__ */ jsx(topBar_default, { children: isMobile ? /* @__PURE__ */ jsxs(
              Flex,
              {
                gapX: 2,
                width: "100%",
                justify: "between",
                className: "oui-hide-scrollbar oui-overflow-x-scroll",
                children: [
                  /* @__PURE__ */ jsx(React3.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
                    LazyTimeInterval,
                    {
                      interval: interval ?? "15",
                      changeInterval
                    }
                  ) }),
                  /* @__PURE__ */ jsx(OperateButton, { onClick: openChartIndicators, children: /* @__PURE__ */ jsx(IndicatorsIcon, {}) }),
                  /* @__PURE__ */ jsx(React3.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
                    LazyMobileDisplayControl,
                    {
                      displayControlState,
                      changeDisplayControlState: changeDisplaySetting
                    }
                  ) })
                ]
              }
            ) : /* @__PURE__ */ jsxs(Flex, { justify: "between", itemAlign: "center", width: "100%", children: [
              /* @__PURE__ */ jsxs(Flex, { children: [
                /* @__PURE__ */ jsx(React3.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
                  LazyTimeInterval,
                  {
                    interval: interval ?? "1",
                    changeInterval
                  }
                ) }),
                /* @__PURE__ */ jsx(
                  Divider,
                  {
                    direction: "vertical",
                    className: "oui-h-4",
                    mx: 2,
                    intensity: 8
                  }
                ),
                /* @__PURE__ */ jsxs(Flex, { justify: "start", itemAlign: "center", gap: 2, children: [
                  /* @__PURE__ */ jsx(React3.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
                    LazyDesktopDisplayControl,
                    {
                      displayControlState,
                      changeDisplayControlState: changeDisplaySetting
                    }
                  ) }),
                  /* @__PURE__ */ jsx(OperateButton, { onClick: openChartIndicators, children: /* @__PURE__ */ jsx(IndicatorsIcon, {}) }),
                  /* @__PURE__ */ jsx(React3.Suspense, { fallback: null, children: /* @__PURE__ */ jsx(
                    LazyLineType,
                    {
                      lineType,
                      changeLineType
                    }
                  ) }),
                  /* @__PURE__ */ jsx(OperateButton, { onClick: openChartSetting, children: /* @__PURE__ */ jsx(SettingIcon, {}) })
                ] })
              ] }),
              /* @__PURE__ */ jsx(Flex, { children: props.fullscreen ? /* @__PURE__ */ jsx(
                ZoomOutIcon,
                {
                  className: "oui-text-base-contrast-54 hover:oui-text-base-contrast oui-cursor-pointer",
                  onClick: onFullScreenChange
                }
              ) : /* @__PURE__ */ jsx(
                ZoomInIcon,
                {
                  className: "oui-text-base-contrast-54 hover:oui-text-base-contrast oui-cursor-pointer",
                  onClick: onFullScreenChange
                }
              ) })
            ] }) }),
            /* @__PURE__ */ jsx("div", { ref: chartRef, className: "oui-size-full oui-overflow-hidden" })
          ]
        }
      )
    }
  );
});
var TradingviewWidget = forwardRef((props, ref) => {
  const state = useTradingviewScript(props);
  return /* @__PURE__ */ jsx(TradingviewUI, { ...state, ref });
});

export { TradingviewUI, TradingviewWidget, useTradingviewScript };
