// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/pricefeed/instructions/setPrice.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Option,
  OptionOrNullable,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  option,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import {
  PriceParams,
  PriceParamsArgs,
  getPriceParamsSerializer,
} from '../types';

// Accounts.
export type SetPriceInstructionAccounts = {
  updater: Signer;
  priceFeed: PublicKey | Pda;
};

// Data.
export type SetPriceInstructionData = {
  discriminator: Uint8Array;
  dstEid: number;
  params: Option<PriceParams>;
};

export type SetPriceInstructionDataArgs = {
  dstEid: number;
  params: OptionOrNullable<PriceParamsArgs>;
};

export function getSetPriceInstructionDataSerializer(): Serializer<
  SetPriceInstructionDataArgs,
  SetPriceInstructionData
> {
  return mapSerializer<
    SetPriceInstructionDataArgs,
    any,
    SetPriceInstructionData
  >(
    struct<SetPriceInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['dstEid', u32()],
        ['params', option(getPriceParamsSerializer())],
      ],
      { description: 'SetPriceInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([16, 19, 182, 8, 149, 83, 72, 181]),
    })
  ) as Serializer<SetPriceInstructionDataArgs, SetPriceInstructionData>;
}

// Args.
export type SetPriceInstructionArgs = SetPriceInstructionDataArgs;

// Instruction.
export function setPrice(
  context: Pick<Context, 'programs'>,
  input: SetPriceInstructionAccounts & SetPriceInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'pricefeed',
    '8ahPGPjEbpgGaZx2NV1iG5Shj7TDwvsjkEDcGWjt94TP'
  );

  // Accounts.
  const resolvedAccounts = {
    updater: {
      index: 0,
      isWritable: false as boolean,
      value: input.updater ?? null,
    },
    priceFeed: {
      index: 1,
      isWritable: true as boolean,
      value: input.priceFeed ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: SetPriceInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getSetPriceInstructionDataSerializer().serialize(
    resolvedArgs as SetPriceInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
