// @orderly.network/ui-positions/dist/index.js
'use strict';

var ui = require('@orderly.network/ui');
var React2 = require('react');
var i18n = require('@orderly.network/i18n');
var types = require('@orderly.network/types');
var utils = require('@orderly.network/utils');
var jsxRuntime = require('react/jsx-runtime');
var hooks = require('@orderly.network/hooks');
var reactApp = require('@orderly.network/react-app');
var perp = require('@orderly.network/perp');
var uiConnector = require('@orderly.network/ui-connector');
var uiLeverage = require('@orderly.network/ui-leverage');
var uiShare = require('@orderly.network/ui-share');
var uiTpsl = require('@orderly.network/ui-tpsl');
var dateFns = require('date-fns');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React2__default = /*#__PURE__*/_interopDefault(React2);

// src/index.ts
var ConfirmHeader = (props) => {
  const { hideCloseIcon = false } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-relative oui-w-full oui-border-b oui-border-line-4 oui-pb-3", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "base", children: props.title }),
    !hideCloseIcon && /* @__PURE__ */ jsxRuntime.jsx(
      "button",
      {
        onClick: props.onClose,
        className: "oui-absolute oui-right-0 oui-top-0 oui-p-2 oui-text-base-contrast-54 hover:oui-text-base-contrast-80",
        children: /* @__PURE__ */ jsxRuntime.jsx(ui.CloseIcon, { size: 18, color: "white" })
      }
    )
  ] });
};
var ConfirmFooter = (props) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      id: "oui-positions-confirm-footer",
      gap: 2,
      width: "100%",
      className: "oui-mt-3 oui-pb-1",
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Button,
          {
            id: "oui-positions-confirm-footer-cancel-button",
            color: "secondary",
            fullWidth: true,
            onClick: props.onCancel,
            size: "md",
            children: t("common.cancel")
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.ThrottledButton,
          {
            id: "oui-positions-confirm-footer-confirm-button",
            onClick: props.onConfirm,
            fullWidth: true,
            loading: props.submitting,
            disabled: props.disabled,
            size: "md",
            children: t("common.confirm")
          }
        )
      ]
    }
  );
};
var OrderDetail = (props) => {
  const { quantity, price, quoteDp, side } = props;
  const { t } = i18n.useTranslation();
  const total = React2.useMemo(() => {
    if (price && quantity) {
      return new utils.Decimal(price).mul(quantity).toFixed(quoteDp, utils.Decimal.ROUND_DOWN);
    }
    return "--";
  }, [price, quantity]);
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      gap: 1,
      width: "100%",
      className: "oui-text-sm oui-text-base-contrast-54",
      py: 5,
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 1, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("common.qty") }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: side === types.OrderSide.BUY ? "success" : "danger", children: quantity })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 1, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("common.price") }),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text.formatted,
            {
              intensity: 98,
              suffix: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: "USDC" }),
              children: price
            }
          )
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 1, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("common.notional") }),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text.formatted,
            {
              intensity: 98,
              suffix: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: "USDC" }),
              children: total
            }
          )
        ] })
      ]
    }
  );
};
var MarketCloseConfirm = (props) => {
  const { t } = i18n.useTranslation();
  const onCancel = () => {
    const func = props?.onClose ?? props.close;
    func?.();
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", className: props.classNames?.root, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ConfirmHeader,
      {
        onClose: onCancel,
        title: t("positions.marketClose"),
        hideCloseIcon: props.hideCloseIcon
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, size: "sm", className: "oui-my-5", children: t("positions.marketClose.description", {
      quantity: utils.commifyOptional(props.quantity),
      base: props.base
    }) }),
    /* @__PURE__ */ jsxRuntime.jsx(
      ConfirmFooter,
      {
        onCancel,
        onConfirm: async () => {
          await props.onConfirm?.();
          onCancel();
        },
        submitting: props.submitting
      }
    )
  ] });
};
var LimitConfirmDialog = (props) => {
  const { order, quoteDp, quantity, price } = props;
  const { side } = order;
  const { t } = i18n.useTranslation();
  const onCancel = () => {
    props.onClose?.();
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ConfirmHeader,
      {
        onClose: onCancel,
        title: t("positions.limitClose"),
        hideCloseIcon: props.hideCloseIcon
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { mt: 5, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, size: "sm", children: t("positions.limitClose.description", {
      quantity: utils.commify(props.quantity),
      base: props.base
    }) }) }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, mb: 4, mt: 5, justify: "between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.formatted,
        {
          rule: "symbol",
          formatString: "base-type",
          size: "base",
          showIcon: true,
          children: order.symbol
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "neutral", size: "xs", children: t("orderEntry.orderType.limit") }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Badge,
          {
            color: side === types.OrderSide.BUY ? "success" : "danger",
            size: "xs",
            children: side === types.OrderSide.BUY ? t("common.buy") : t("common.sell")
          }
        )
      ] })
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { className: "oui-w-full" }),
    /* @__PURE__ */ jsxRuntime.jsx(
      OrderDetail,
      {
        className: "oui-text-sm",
        price,
        quantity,
        side: order.side,
        quoteDp: quoteDp ?? 2
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      ConfirmFooter,
      {
        onCancel,
        onConfirm: props.onConfirm,
        submitting: props.submitting
      }
    )
  ] });
};
var PositionsRowContext = React2.createContext(
  {}
);
var usePositionsRowContext = () => {
  return React2.useContext(PositionsRowContext);
};
function useEndReached(sentinelRef, onEndReached) {
  const observer = React2.useRef();
  const cb = React2.useRef(onEndReached);
  cb.current = onEndReached;
  React2.useEffect(() => {
    const options = {
      root: null,
      rootMargin: "0px",
      threshold: 0
    };
    const handleObserver = (entries) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          cb.current?.();
        }
      });
    };
    observer.current = new IntersectionObserver(handleObserver, options);
    return () => {
      observer.current?.disconnect();
    };
  }, []);
  React2.useEffect(() => {
    observer.current?.observe(sentinelRef.current);
  }, []);
}
var EndReachedBox = (props) => {
  const sentinelRef = React2.useRef(null);
  const { onEndReached } = props;
  useEndReached(sentinelRef, () => {
    onEndReached?.();
  });
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    props.children,
    /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        ref: sentinelRef,
        className: "oui-relative oui-invisible oui-h-[25px] oui-bg-red-400 oui-top-[-300px]"
      }
    )
  ] });
};
var PAGE_SIZE = 60;
var FundingFeeHistoryUI = ({ total, symbol, start_t, end_t }) => {
  const { t } = i18n.useTranslation();
  const { isMobile } = ui.useScreen();
  const { isLoading, data, setSize } = hooks.usePrivateInfiniteQuery(
    (pageIndex, previousPageData) => {
      if ((!previousPageData || (previousPageData.length ?? 0) < PAGE_SIZE) && pageIndex > 0)
        return null;
      return `/v1/funding_fee/history?page=${pageIndex + 1}&size=${PAGE_SIZE}&symbol=${symbol}&start_t=${start_t}&end_t=${end_t}`;
    },
    {
      revalidateFirstPage: false
    }
  );
  const loadMore = React2.useCallback(() => {
    setSize((prev) => {
      return prev + 1;
    });
  }, [setSize]);
  const flattenData = React2.useMemo(() => {
    if (!Array.isArray(data)) return [];
    return data.flat().map((item) => {
      return {
        ...item,
        funding_fee: -item.funding_fee
      };
    });
  }, [data]);
  const listView = React2.useMemo(() => {
    if (isMobile) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        HistoryDataListViewSimple,
        {
          data: flattenData ?? types.EMPTY_LIST,
          isLoading,
          loadMore
        }
      );
    }
    return /* @__PURE__ */ jsxRuntime.jsx(
      HistoryDataListView,
      {
        data: flattenData ?? types.EMPTY_LIST,
        isLoading,
        loadMore
      }
    );
  }, [isMobile, flattenData, isLoading]);
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
    /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Grid,
      {
        cols: 2,
        gapX: 3,
        className: "oui-sticky oui-top-0 oui-z-10 oui-bg-base-8 oui-py-4",
        children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-rounded-lg oui-border oui-border-line-6 oui-bg-base-9 oui-p-3", children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", gap: 1, itemAlign: "start", children: [
            /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-text-2xs oui-text-base-contrast-36", children: t("common.symbol") }),
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.formatted,
              {
                rule: "symbol",
                className: "oui-font-semibold",
                intensity: 98,
                children: symbol
              }
            )
          ] }) }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-rounded-lg oui-border oui-border-line-6 oui-bg-base-9 oui-p-3", children: /* @__PURE__ */ jsxRuntime.jsx(
            ui.Statistic,
            {
              label: isMobile ? /* @__PURE__ */ jsxRuntime.jsx(
                FundingFeeLabelButton,
                {
                  label: `${t("funding.fundingFee")} (USDC)`,
                  tooltip: t("positions.fundingFee.tooltip"),
                  size: 14
                }
              ) : /* @__PURE__ */ jsxRuntime.jsx(
                FundingFeeLabel,
                {
                  label: `${t("funding.fundingFee")} (USDC)`,
                  tooltip: t("positions.fundingFee.tooltip"),
                  size: 14
                }
              ),
              valueProps: {
                ignoreDP: true,
                coloring: true,
                showIdentifier: true
              },
              children: total
            }
          ) })
        ]
      }
    ),
    listView
  ] });
};
var FundingFeeLabelButton = ({ label, tooltip, size }) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1", children: [
    /* @__PURE__ */ jsxRuntime.jsx("span", { children: label }),
    tooltip && /* @__PURE__ */ jsxRuntime.jsx(
      "button",
      {
        className: "oui-flex oui-items-center",
        onClick: () => {
          ui.modal.alert({
            message: tooltip,
            title: t("positions.fundingFee.title")
          });
        },
        children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.ExclamationFillIcon,
          {
            className: "oui-cursor-pointer oui-text-base-contrast-54",
            size
          }
        )
      }
    )
  ] });
};
var FundingFeeLabel = ({ label, tooltip, size }) => {
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1", children: [
    /* @__PURE__ */ jsxRuntime.jsx("span", { children: label }),
    tooltip && /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { content: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-w-64", children: tooltip }), children: /* @__PURE__ */ jsxRuntime.jsx(
      ui.ExclamationFillIcon,
      {
        className: "oui-cursor-pointer oui-text-base-contrast-54",
        size
      }
    ) })
  ] });
};
var HistoryDataListView = ({ isLoading, data, loadMore }) => {
  const { t } = i18n.useTranslation();
  const columns = React2.useMemo(() => {
    return [
      {
        title: t("common.time"),
        dataIndex: "created_time",
        width: 120,
        render: (value) => {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.formatted, { rule: "date", children: value });
        }
      },
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(FundingFeeLabel, { label: t("funding.fundingRate"), size: 12 }),
        dataIndex: "funding_rate",
        formatter: (value) => new utils.Decimal(value).mul(100).toString(),
        render: (value) => {
          return /* @__PURE__ */ jsxRuntime.jsx("span", { children: `${value}%` });
        }
      },
      {
        title: t("funding.paymentType"),
        dataIndex: "payment_type",
        formatter: (value) => {
          return value === "Pay" ? t("funding.paymentType.paid") : t("funding.paymentType.received");
        },
        render: (value) => /* @__PURE__ */ jsxRuntime.jsx("span", { children: value })
      },
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(
          FundingFeeLabel,
          {
            label: t("common.amount"),
            tooltip: t("positions.fundingRate.tooltip"),
            size: 12
          }
        ),
        dataIndex: "funding_fee",
        render: (value) => {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { rule: "price", coloring: true, showIdentifier: true, ignoreDP: true, children: value });
        }
      }
    ];
  }, [t]);
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-[calc(80vh_-_132px_-_8px)] oui-overflow-y-auto", children: /* @__PURE__ */ jsxRuntime.jsx(EndReachedBox, { onEndReached: loadMore, children: /* @__PURE__ */ jsxRuntime.jsx(
    ui.DataTable,
    {
      classNames: {
        root: ui.cn("oui-h-auto oui-bg-base-8 oui-text-sm")
      },
      columns,
      dataSource: data ?? types.EMPTY_LIST,
      loading: isLoading
    }
  ) }) });
};
var HistoryDataListViewSimple = ({
  data,
  isLoading,
  loadMore
}) => {
  const renderItem = React2.useCallback((item) => {
    return /* @__PURE__ */ jsxRuntime.jsx(FundingFeeItem, { item });
  }, []);
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-[calc(80vh_-_104px)] oui-overflow-y-auto", children: /* @__PURE__ */ jsxRuntime.jsx(
    ui.ListView,
    {
      dataSource: data,
      renderItem,
      isLoading,
      contentClassName: "oui-space-y-0",
      loadMore
    }
  ) });
};
var FundingFeeItem = ({ item }) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-space-y-2 oui-border-t oui-border-line-6 oui-py-2", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Statistic,
        {
          label: /* @__PURE__ */ jsxRuntime.jsx(FundingFeeLabelButton, { label: t("funding.fundingRate"), size: 12 }),
          classNames: {
            label: "oui-text-2xs"
          },
          valueProps: {
            ignoreDP: true,
            as: "div",
            className: "oui-text-xs",
            suffix: "%"
          },
          children: new utils.Decimal(item.funding_rate).mul(100).toNumber()
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Statistic,
        {
          label: /* @__PURE__ */ jsxRuntime.jsx(
            FundingFeeLabelButton,
            {
              label: t("common.amount"),
              tooltip: t("positions.fundingRate.tooltip"),
              size: 12
            }
          ),
          className: "oui-items-end",
          classNames: {
            label: "oui-text-2xs"
          },
          valueProps: {
            ignoreDP: true,
            coloring: true,
            as: "div",
            className: "oui-text-xs",
            showIdentifier: true
          },
          children: item.funding_fee
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.formatted,
        {
          rule: "date",
          className: "oui-text-base-contrast-36",
          size: "2xs",
          children: item.created_time
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 80, children: item.payment_type === "Pay" ? t("funding.paymentType.paid") : t("funding.paymentType.received") })
    ] })
  ] });
};
var FundingFeeButton = ({ fee, symbol, start_t, end_t }) => {
  const { t } = i18n.useTranslation();
  const [isOpen, { setTrue, setFalse }] = hooks.useBoolean(false);
  const { isMobile } = ui.useScreen();
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx("button", { onClick: setTrue, children: /* @__PURE__ */ jsxRuntime.jsx(
      ui.Text.numeral,
      {
        rule: "price",
        coloring: true,
        showIdentifier: true,
        ignoreDP: true,
        className: "oui-border-b oui-border-line-16 oui-border-dashed oui-py-0.5",
        children: fee
      }
    ) }),
    isMobile ? /* @__PURE__ */ jsxRuntime.jsx(
      ui.SimpleSheet,
      {
        open: isOpen,
        onOpenChange: setFalse,
        title: t("funding.fundingFee"),
        classNames: {
          body: "oui-max-h-[80vh] oui-py-0"
        },
        children: /* @__PURE__ */ jsxRuntime.jsx(
          FundingFeeHistoryUI,
          {
            total: fee,
            symbol,
            start_t,
            end_t
          }
        )
      }
    ) : /* @__PURE__ */ jsxRuntime.jsx(
      ui.SimpleDialog,
      {
        open: isOpen,
        onOpenChange: setFalse,
        title: t("funding.fundingFee"),
        classNames: {
          content: "lg:oui-max-w-[640px]",
          body: "oui-max-h-[80vh] oui-bg-base-8 lg:oui-py-0"
        },
        children: /* @__PURE__ */ jsxRuntime.jsx(
          FundingFeeHistoryUI,
          {
            total: fee,
            symbol,
            start_t,
            end_t
          }
        )
      }
    )
  ] });
};
var calculatePositions = (positions2, symbolsInfo, accountInfo, tpslOrders) => {
  return positions2.map((item) => {
    const info = symbolsInfo[item.symbol];
    const notional = perp.positions.notional(item.position_qty, item.mark_price);
    const account2 = accountInfo.find(
      (acc) => acc.account_id === item.account_id
    );
    const baseMMR = info?.("base_mmr");
    const baseIMR = info?.("base_imr");
    if (!baseMMR || !baseIMR) {
      return item;
    }
    const MMR = perp.positions.MMR({
      baseMMR,
      baseIMR,
      IMRFactor: account2?.imr_factor[item.symbol] ?? 0,
      positionNotional: notional,
      IMR_factor_power: 4 / 5
    });
    const mm = perp.positions.maintenanceMargin({
      positionQty: item.position_qty,
      markPrice: item.mark_price,
      MMR
    });
    const unrealPnl = perp.positions.unrealizedPnL({
      qty: item.position_qty,
      openPrice: item?.average_open_price,
      markPrice: item.mark_price
    });
    const maxLeverage = item.leverage || 1;
    const imr = perp.account.IMR({
      maxLeverage,
      baseIMR,
      IMR_Factor: account2?.imr_factor[item.symbol] ?? 0,
      positionNotional: notional,
      ordersNotional: 0,
      IMR_factor_power: 4 / 5
    });
    const unrealPnlROI = perp.positions.unrealizedPnLROI({
      positionQty: item.position_qty,
      openPrice: item.average_open_price,
      IMR: imr,
      unrealizedPnL: unrealPnl
    });
    let unrealPnl_index = 0;
    let unrealPnlROI_index = 0;
    if (item.index_price) {
      unrealPnl_index = perp.positions.unrealizedPnL({
        qty: item.position_qty,
        openPrice: item?.average_open_price,
        markPrice: item.index_price
      });
      unrealPnlROI_index = perp.positions.unrealizedPnLROI({
        positionQty: item.position_qty,
        openPrice: item.average_open_price,
        IMR: imr,
        unrealizedPnL: unrealPnl_index
      });
    }
    const filteredTPSLOrders = tpslOrders.filter(
      (tpslOrder) => tpslOrder.account_id === item.account_id
    );
    const tpsl = formatTPSL(filteredTPSLOrders, item.symbol);
    return {
      ...item,
      ...tpsl,
      mmr: MMR,
      mm,
      notional,
      unrealized_pnl: unrealPnl,
      unrealized_pnl_ROI: unrealPnlROI,
      unrealized_pnl_ROI_index: unrealPnlROI_index
    };
  });
};
function formatTPSL(tpslOrders, symbol) {
  if (Array.isArray(tpslOrders) && tpslOrders.length) {
    const { fullPositionOrder, partialPositionOrders } = hooks.findPositionTPSLFromOrders(tpslOrders, symbol);
    const full_tp_sl = fullPositionOrder ? hooks.findTPSLFromOrder(fullPositionOrder) : void 0;
    const partialPossitionOrder = partialPositionOrders && partialPositionOrders.length ? partialPositionOrders[0] : void 0;
    const partial_tp_sl = partialPossitionOrder ? hooks.findTPSLFromOrder(partialPossitionOrder) : void 0;
    return {
      full_tp_sl: {
        tp_trigger_price: full_tp_sl?.tp_trigger_price,
        sl_trigger_price: full_tp_sl?.sl_trigger_price,
        algo_order: fullPositionOrder
      },
      partial_tp_sl: {
        order_num: partialPositionOrders?.length ?? 0,
        tp_trigger_price: partial_tp_sl?.tp_trigger_price,
        sl_trigger_price: partial_tp_sl?.sl_trigger_price,
        algo_order: partialPossitionOrder
      }
    };
  }
}
var signatureMiddleware = (account2, accountId) => {
  const apiBaseUrl = hooks.useConfig("apiBaseUrl");
  return (useSWRNext) => {
    return (key, fetcher2, config) => {
      try {
        const extendedFetcher = async (args) => {
          const url = Array.isArray(args) ? args[0] : args;
          const fullUrl = `${apiBaseUrl}${url}`;
          const signer = account2.signer;
          const payload = { method: "GET", url };
          const signature = await signer.sign(payload, utils.getTimestamp());
          const ids = Array.isArray(accountId) ? accountId : [accountId];
          return Promise.all(
            ids.map((id) => {
              return fetcher2(fullUrl, {
                headers: {
                  ...signature,
                  "orderly-account-id": id
                }
              });
            })
          );
        };
        return useSWRNext(key, extendedFetcher, config);
      } catch (e) {
        throw e;
      }
    };
  };
};
var useSubAccountQuery = (query, options) => {
  const { formatter, accountId, ...swrOptions } = options || {};
  const { state, account: account2 } = hooks.useAccount();
  const middleware = Array.isArray(options?.use) ? options?.use ?? [] : [];
  const ids = Array.isArray(accountId) ? accountId : [accountId];
  const shouldFetch = ids.filter(Boolean).length && (state.status >= types.AccountStatusEnum.EnableTrading || state.status === types.AccountStatusEnum.EnableTradingWithoutConnected);
  return hooks.useSWR(
    () => shouldFetch ? [query, ids] : null,
    (url, init) => {
      return hooks.fetcher(url, init, { formatter });
    },
    {
      ...swrOptions,
      use: [...middleware, signatureMiddleware(account2, ids)],
      onError: () => {
      }
    }
  );
};
function useSubAccountTPSL(subAccountIds) {
  const ee = hooks.useEventEmitter();
  const { data: algoOrdersResponse, mutate: mutateTPSLOrders } = useSubAccountQuery(
    `/v1/algo/orders?size=100&page=1&status=${types.OrderStatus.INCOMPLETE}`,
    {
      accountId: subAccountIds,
      formatter: (data) => data,
      revalidateOnFocus: false
    }
  );
  const tpslOrders = React2.useMemo(() => {
    if (!Array.isArray(algoOrdersResponse)) {
      return [];
    }
    const algoOrders = algoOrdersResponse?.map(
      (item, index) => item.rows.map((order) => ({
        ...order,
        account_id: subAccountIds[index]
      }))
    )?.flat();
    return algoOrders?.filter(
      (order) => [types.AlgoOrderRootType.POSITIONAL_TP_SL, types.AlgoOrderRootType.TP_SL].includes(
        order.algo_type
      )
    );
  }, [algoOrdersResponse, subAccountIds]);
  const refresh = hooks.useDebouncedCallback(() => {
    mutateTPSLOrders();
  }, 200);
  React2.useEffect(() => {
    const handler = (position) => {
      if (position.account_id) {
        refresh();
      }
    };
    ee.on("tpsl:updateOrder", handler);
    return () => {
      ee.off("tpsl:updateOrder", handler);
    };
  }, [ee]);
  return { tpslOrders, mutateTPSLOrders };
}

// src/components/positions/combinePositions.script.ts
var useCombinePositionsScript = (props) => {
  const {
    symbol,
    calcMode,
    includedPendingOrder,
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    onSymbolChange,
    selectedAccount
  } = props;
  const { pagination, setPage } = ui.usePagination({ pageSize: 50 });
  const symbolsInfo = hooks.useSymbolsInfo();
  const { state } = hooks.useAccount();
  const [mainAccountPositions, , { isLoading }] = hooks.usePositionStream(symbol, {
    calcMode,
    includedPendingOrder
  });
  const {
    data: newPositions = [],
    isLoading: isPositionLoading,
    mutate: mutatePositions
  } = hooks.usePrivateQuery("/v1/client/aggregate/positions", {
    errorRetryCount: 3
  });
  const { allAccountIds, subAccountIds } = React2.useMemo(() => {
    const uniqueIds = new Set(
      newPositions.filter((item) => item.account_id).map((item) => item.account_id).filter(Boolean)
    );
    const allAccountIds2 = Array.from(uniqueIds);
    const subAccountIds2 = allAccountIds2.filter(
      (item) => item !== state.mainAccountId
    );
    return { allAccountIds: allAccountIds2, subAccountIds: subAccountIds2 };
  }, [newPositions, state.mainAccountId]);
  const { data: accountInfo = [], isLoading: isAccountInfoLoading } = useSubAccountQuery("/v1/client/info", {
    accountId: allAccountIds,
    revalidateOnFocus: false
  });
  const { tpslOrders, mutateTPSLOrders } = useSubAccountTPSL(subAccountIds);
  const subAccountPositions = React2.useMemo(() => {
    return calculatePositions(
      newPositions.filter((item) => item.account_id !== state.mainAccountId),
      symbolsInfo,
      accountInfo,
      tpslOrders
    );
  }, [newPositions, symbolsInfo, accountInfo, state.mainAccountId, tpslOrders]);
  const allPositions = React2.useMemo(() => {
    return [...mainAccountPositions?.rows, ...subAccountPositions].filter(
      (item) => item.position_qty !== 0
    );
  }, [mainAccountPositions, subAccountPositions]);
  const dataSource = reactApp.useDataTap(allPositions) ?? [];
  const filtered = React2.useMemo(() => {
    if (!selectedAccount || selectedAccount === "All accounts" /* ALL */) {
      return dataSource;
    }
    return dataSource.filter((item) => {
      if (selectedAccount === "Main accounts" /* MAIN */) {
        return item.account_id === state.mainAccountId || !item.account_id;
      } else {
        return item.account_id === selectedAccount;
      }
    });
  }, [dataSource, selectedAccount, state.mainAccountId]);
  const groupDataSource = React2.useMemo(() => {
    return groupDataByAccount(filtered, {
      mainAccountId: state.mainAccountId,
      subAccounts: state.subAccounts
    });
  }, [filtered, state.mainAccountId, state.subAccounts]);
  const loading = isLoading || isPositionLoading || isAccountInfoLoading;
  React2.useEffect(() => {
    setPage(1);
  }, [symbol]);
  const mutateList = React2.useCallback(() => {
    mutatePositions();
    mutateTPSLOrders();
  }, []);
  return {
    tableData: groupDataSource,
    isLoading: loading,
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    symbol,
    onSymbolChange,
    pagination,
    mutatePositions: mutateList
  };
};
var groupDataByAccount = (data, options) => {
  const { mainAccountId = "", subAccounts = [] } = options;
  const map = /* @__PURE__ */ new Map();
  for (const item of data) {
    const accountId = item.account_id || mainAccountId;
    const findSubAccount = subAccounts.find((item2) => item2.id === accountId);
    if (map.has(accountId)) {
      map.get(accountId)?.children?.push(item);
    } else {
      map.set(accountId, {
        account_id: accountId,
        description: accountId === mainAccountId ? i18n.i18n.t("common.mainAccount") : findSubAccount?.description || ui.formatAddress(findSubAccount?.id || ""),
        children: [item]
      });
    }
  }
  return {
    expanded: Array.from(map.keys()),
    dataSource: Array.from(map.values())
  };
};

// src/constants.ts
var TRADING_POSITIONS_SORT_STORAGE_KEY = "orderly_trading_positions_sort";
var LIQ_DISTANCE_THRESHOLD = 10;
var compareValues = (aValue, bValue) => {
  if (aValue == null && bValue == null) return 0;
  if (aValue == null) return 1;
  if (bValue == null) return -1;
  const aStr = String(aValue).trim();
  const bStr = String(bValue).trim();
  const aNum = Number(aStr);
  const bNum = Number(bStr);
  const aIsNumber = !isNaN(aNum) && isFinite(aNum) && /^-?\d*\.?\d+([eE][+-]?\d+)?$/.test(aStr);
  const bIsNumber = !isNaN(bNum) && isFinite(bNum) && /^-?\d*\.?\d+([eE][+-]?\d+)?$/.test(bStr);
  if (aIsNumber && bIsNumber) {
    return aNum - bNum;
  }
  const aIsDate = /^\d{4}-\d{2}-\d{2}/.test(aStr) || /^\d{13}$/.test(aStr);
  const bIsDate = /^\d{4}-\d{2}-\d{2}/.test(bStr) || /^\d{13}$/.test(bStr);
  if (aIsDate && bIsDate) {
    const aDate = new Date(aValue);
    const bDate = new Date(bValue);
    if (!isNaN(aDate.getTime()) && !isNaN(bDate.getTime())) {
      return aDate.getTime() - bDate.getTime();
    }
  }
  return aStr.localeCompare(bStr, void 0, {
    sensitivity: "base",
    numeric: false
  });
};
function sortList(list, sort) {
  const { sortKey, sortOrder } = sort || {};
  const sortedList = [...list || []];
  if (sortKey && sortOrder) {
    sortedList.sort((a, b) => {
      const comparison = compareValues(a[sortKey], b[sortKey]);
      return sortOrder === "desc" ? -comparison : comparison;
    });
  }
  return sortedList;
}
function useSort(initialSort, onSortChange) {
  const [sort, setSort] = React2.useState(initialSort);
  const onSort = React2.useCallback((options) => {
    const nextSort = options ? {
      sortKey: options.sortKey,
      sortOrder: options.sort
    } : void 0;
    setSort(nextSort);
    onSortChange?.(nextSort);
  }, []);
  const getSortedList = React2.useCallback(
    (list) => sortList(list, sort),
    [sort]
  );
  return {
    sort,
    onSort,
    getSortedList
  };
}
var OrderInfoCard = ({
  title,
  side,
  leverage,
  qty,
  baseDp,
  estLiqPrice,
  estPnL,
  pnlNotionalDecimalPrecision = 2,
  className
}) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      gap: 3,
      itemAlign: "start",
      className: `oui-bg-base-6 oui-rounded-lg oui-p-3 oui-w-full oui-font-weight-semibold ${className || ""}`,
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", itemAlign: "center", gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "semibold", intensity: 98, children: title }),
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "center", gap: 1, children: [
            /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: side === types.OrderSide.SELL ? "sell" : "buy", size: "xs", children: side === types.OrderSide.SELL ? t("common.sell") : t("common.buy") }),
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Badge, { color: "neutral", size: "xs", children: [
              leverage,
              "X"
            ] })
          ] })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(
          ui.Flex,
          {
            direction: "column",
            justify: "between",
            itemAlign: "center",
            width: "100%",
            gap: 1,
            children: [
              /* @__PURE__ */ jsxRuntime.jsxs(
                ui.Flex,
                {
                  direction: "row",
                  justify: "between",
                  itemAlign: "center",
                  width: "100%",
                  gap: 1,
                  children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 54, children: t("common.qty") }),
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: baseDp, size: "sm", color: "danger", padding: false, children: qty })
                  ]
                }
              ),
              /* @__PURE__ */ jsxRuntime.jsxs(
                ui.Flex,
                {
                  direction: "row",
                  justify: "between",
                  itemAlign: "center",
                  width: "100%",
                  gap: 1,
                  children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 54, children: t("common.price") }),
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "semibold", children: t("common.market") })
                  ]
                }
              ),
              estPnL && /* @__PURE__ */ jsxRuntime.jsxs(
                ui.Flex,
                {
                  direction: "row",
                  justify: "between",
                  itemAlign: "center",
                  width: "100%",
                  gap: 1,
                  children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 54, children: t("orderEntry.estPnL") }),
                    /* @__PURE__ */ jsxRuntime.jsx(
                      ui.Text.pnl,
                      {
                        dp: pnlNotionalDecimalPrecision,
                        coloring: true,
                        size: "sm",
                        weight: "semibold",
                        children: estPnL
                      }
                    )
                  ]
                }
              ),
              estLiqPrice && /* @__PURE__ */ jsxRuntime.jsxs(
                ui.Flex,
                {
                  direction: "row",
                  justify: "between",
                  itemAlign: "center",
                  width: "100%",
                  gap: 1,
                  children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 54, children: t("orderEntry.estLiqPrice") }),
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", weight: "semibold", children: estLiqPrice })
                  ]
                }
              )
            ]
          }
        )
      ]
    }
  );
};
var ReversePosition = (props) => {
  const {
    displayInfo,
    validationError,
    className,
    style,
    onConfirm,
    onCancel
  } = props;
  const { t } = i18n.useTranslation();
  if (!displayInfo) {
    return null;
  }
  const {
    symbol,
    base,
    quote,
    baseDp,
    quoteDp,
    positionQty,
    reverseQty,
    markPrice,
    leverage,
    isLong,
    unrealizedPnL,
    pnlNotionalDecimalPrecision
  } = displayInfo;
  const closeSide = !isLong ? types.OrderSide.SELL : types.OrderSide.BUY;
  const openSide = isLong ? types.OrderSide.SELL : types.OrderSide.BUY;
  const closeAction = isLong ? t("positions.reverse.marketCloseLong") : t("positions.reverse.marketCloseShort");
  const openAction = openSide === types.OrderSide.BUY ? t("positions.reverse.marketOpenLong") : t("positions.reverse.marketOpenShort");
  const reverseTo = isLong ? t("positions.reverse.reverseToShort") : t("positions.reverse.reverseToLong");
  const reverseToIcon = isLong ? /* @__PURE__ */ jsxRuntime.jsx(ui.ArrowDownShortIcon, { size: 16, color: "danger", opacity: 1 }) : /* @__PURE__ */ jsxRuntime.jsx(ui.ArrowUpShortIcon, { size: 16, color: "success", opacity: 1 });
  const priceLabel = /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { itemAlign: "center", gap: 1, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: quoteDp, size: "sm", intensity: 80, children: markPrice }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 36, children: quote })
  ] });
  const showBelowMinError = validationError === "belowMin";
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      className,
      style,
      gap: 4,
      width: "100%",
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", gap: 2, width: "100%", children: [
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", itemAlign: "center", width: "100%", children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.formatted,
              {
                size: "base",
                weight: "semibold",
                rule: "symbol",
                formatString: "base-type",
                intensity: 98,
                showIcon: true,
                children: symbol
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: isLong ? "sell" : "buy", size: "xs", children: reverseTo })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", itemAlign: "center", width: "100%", children: [
            /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", intensity: 54, children: t("common.markPrice") }),
            priceLabel
          ] })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { intensity: 4, className: "oui-w-full" }),
        /* @__PURE__ */ jsxRuntime.jsx(
          OrderInfoCard,
          {
            title: closeAction,
            side: closeSide,
            leverage,
            qty: positionQty,
            baseDp,
            estPnL: unrealizedPnL?.toString(),
            pnlNotionalDecimalPrecision
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "row", itemAlign: "center", width: "100%", children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { intensity: 8, className: "oui-w-full" }),
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { className: "oui-px-4 oui-py-[3px] oui-border oui-border-base-contrast-12 oui-rounded-full oui-shrink-0", children: [
            reverseToIcon,
            /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", color: isLong ? "danger" : "success", children: reverseTo })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { intensity: 8, className: "oui-w-full" })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(
          OrderInfoCard,
          {
            title: openAction,
            side: openSide,
            leverage,
            qty: reverseQty,
            baseDp
          }
        ),
        showBelowMinError ? /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", color: "danger", weight: "semibold", children: t("positions.reverse.error.belowMin") }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", color: "warning", weight: "semibold", children: t("positions.reverse.description") })
      ]
    }
  );
};
var MAX_BATCH_ORDER_SIZE = 20;
var useReversePositionEnabled = () => {
  const { isMobile, isDesktop } = ui.useScreen();
  const [desktopEnabled, setDesktopEnabled] = hooks.useLocalStorage(
    "orderly_reverse_position_enabled_desktop",
    true
  );
  const [mobileEnabled, setMobileEnabled] = hooks.useLocalStorage(
    "orderly_reverse_position_enabled_mobile",
    false
  );
  const isEnabled = React2.useMemo(() => {
    if (isMobile) {
      return mobileEnabled;
    }
    if (isDesktop) {
      return desktopEnabled;
    }
    return false;
  }, [isMobile, isDesktop, desktopEnabled, mobileEnabled]);
  const setEnabled = React2.useCallback(
    (enabled) => {
      if (isMobile) {
        setMobileEnabled(enabled);
      } else if (isDesktop) {
        setDesktopEnabled(enabled);
      }
    },
    [isMobile, isDesktop, setMobileEnabled, setDesktopEnabled]
  );
  return {
    isEnabled,
    setEnabled
  };
};
var useReversePositionScript = (options) => {
  const { position, onSuccess, onError } = options || {};
  const { isEnabled, setEnabled } = useReversePositionEnabled();
  const [pnlNotionalDecimalPrecision] = hooks.useLocalStorage(
    "pnlNotionalDecimalPrecision",
    2
  );
  const symbolsInfo = hooks.useSymbolsInfo();
  const symbol = position?.symbol || "";
  const { data: symbolMarketPrice } = hooks.useMarkPrice(symbol);
  const symbolInfo = symbolsInfo?.[symbol];
  const calcMode = options?.calcMode || "markPrice";
  const [positionData] = hooks.usePositionStream(symbol, {
    calcMode,
    includedPendingOrder: false
  });
  const rawPositionRows = reactApp.useDataTap(positionData?.rows, { fallbackData: [] });
  const unrealizedPnL = React2.useMemo(() => {
    if (!rawPositionRows || !symbol) return position?.unrealized_pnl;
    const currentPosition = rawPositionRows.find(
      (p) => p.symbol === symbol && p.position_qty !== 0
    );
    return currentPosition?.unrealized_pnl ?? position?.unrealized_pnl;
  }, [rawPositionRows, symbol]);
  const baseMin = React2.useMemo(() => {
    if (!symbolInfo) return 0;
    return symbolInfo("base_min") || 0;
  }, [symbolInfo]);
  const baseMax = React2.useMemo(() => {
    if (!symbolInfo) return 0;
    return symbolInfo("base_max") || 0;
  }, [symbolInfo]);
  const baseDp = React2.useMemo(() => {
    if (!symbolInfo) return 6;
    return symbolInfo("base_dp") || 6;
  }, [symbolInfo]);
  const positionQty = React2.useMemo(() => {
    if (!position) return 0;
    return Math.abs(position.position_qty);
  }, [position]);
  const isLong = React2.useMemo(() => {
    if (!position) return false;
    return position.position_qty > 0;
  }, [position]);
  const reverseQty = positionQty;
  const validationError = React2.useMemo(() => {
    if (!position || !symbolInfo) return null;
    if (baseMin > 0 && reverseQty < baseMin) {
      return "belowMin";
    }
    return null;
  }, [position, symbolInfo, reverseQty, baseMin]);
  const splitOrders = React2.useMemo(() => {
    if (!position || !symbolInfo || baseMax <= 0 || reverseQty <= 0) {
      return { needsSplit: false, orders: [] };
    }
    const buildOrder = (qty, side, reduceOnly) => {
      return {
        symbol: position.symbol,
        order_type: types.OrderType.MARKET,
        side,
        order_quantity: new utils.Decimal(qty).todp(baseDp).toString(),
        reduce_only: reduceOnly
      };
    };
    const closeSide = isLong ? types.OrderSide.SELL : types.OrderSide.BUY;
    const openSide = isLong ? types.OrderSide.SELL : types.OrderSide.BUY;
    if (reverseQty <= baseMax) {
      return {
        needsSplit: false,
        orders: [
          buildOrder(reverseQty, closeSide, true),
          buildOrder(reverseQty, openSide, false)
        ]
      };
    }
    const orders = [];
    const perOrderQty = baseMax;
    const numOrders = Math.ceil(reverseQty / baseMax);
    for (let i = 0; i < numOrders - 1; i++) {
      orders.push(buildOrder(perOrderQty, closeSide, true));
    }
    orders.push(
      buildOrder(reverseQty - perOrderQty * (numOrders - 1), closeSide, true)
    );
    for (let i = 0; i < numOrders - 1; i++) {
      orders.push(buildOrder(perOrderQty, openSide, false));
    }
    orders.push(
      buildOrder(reverseQty - perOrderQty * (numOrders - 1), openSide, false)
    );
    return {
      needsSplit: true,
      orders
    };
  }, [position, symbolInfo, reverseQty, baseMax, baseDp]);
  const [isReversing, setIsReversing] = React2.useState(false);
  const [doBatchCreateOrder] = hooks.useSubAccountMutation(
    "/v1/batch-order",
    "POST",
    {
      accountId: position?.account_id
    }
  );
  const reversePosition = React2.useCallback(async () => {
    if (!position || positionQty === 0) {
      return false;
    }
    if (validationError) {
      return false;
    }
    setIsReversing(true);
    try {
      const ordersArray = splitOrders.orders;
      if (ordersArray.length > MAX_BATCH_ORDER_SIZE) {
        for (let i = 0; i < ordersArray.length; i += MAX_BATCH_ORDER_SIZE) {
          const batch = ordersArray.slice(i, i + MAX_BATCH_ORDER_SIZE);
          const result = await doBatchCreateOrder({
            orders: batch,
            symbol: position.symbol
          });
          await new Promise(
            (resolve) => setTimeout(resolve, batch.length * 110)
          );
          if (!result || result.error) {
            throw result?.error || new Error("Batch order failed");
          }
        }
      } else {
        const result = await doBatchCreateOrder({
          orders: ordersArray,
          symbol: position.symbol
        });
        if (!result || result.error) {
          throw result?.error || new Error("Batch order failed");
        }
      }
      onSuccess?.();
      return true;
    } catch (error) {
      onError?.(error);
      return false;
    } finally {
      setIsReversing(false);
    }
  }, [
    position,
    positionQty,
    reverseQty,
    isLong,
    doBatchCreateOrder,
    splitOrders,
    symbolInfo,
    validationError,
    onSuccess,
    onError
  ]);
  const displayInfo = React2.useMemo(() => {
    if (!position || !symbolInfo) {
      return null;
    }
    const base = symbolInfo("base");
    const quote = symbolInfo("quote");
    const baseDp2 = symbolInfo("base_dp");
    const quoteDp = symbolInfo("quote_dp");
    const leverage = position.leverage || 1;
    return {
      symbol: position.symbol,
      base,
      quote,
      baseDp: baseDp2,
      quoteDp,
      positionQty: new utils.Decimal(positionQty).todp(baseDp2).toString(),
      reverseQty: new utils.Decimal(reverseQty).todp(baseDp2).toString(),
      markPrice: symbolMarketPrice ? new utils.Decimal(symbolMarketPrice).todp(quoteDp).toString() : "--",
      leverage,
      isLong,
      unrealizedPnL,
      pnlNotionalDecimalPrecision
    };
  }, [
    position,
    symbolMarketPrice,
    symbolInfo,
    positionQty,
    reverseQty,
    isLong,
    unrealizedPnL,
    pnlNotionalDecimalPrecision
  ]);
  return {
    isEnabled,
    setEnabled,
    reversePosition,
    isReversing,
    displayInfo,
    positionQty,
    reverseQty,
    isLong,
    validationError,
    splitOrders
  };
};
var ReversePositionWidget = (props) => {
  const { position, close, resolve, reject } = props;
  const { t } = i18n.useTranslation();
  const { visible, hide, onOpenChange } = ui.useModal();
  const state = useReversePositionScript({
    position,
    onSuccess: () => {
      resolve?.(true);
      hide();
    },
    onError: (error) => {
      reject?.(error);
      hide();
    }
  });
  const actions = React2.useMemo(() => {
    const hasValidationError = !!state.validationError;
    return {
      primary: {
        label: t("common.confirm"),
        onClick: async () => {
          try {
            const result = await state.reversePosition();
            if (result) {
              resolve?.(true);
              hide();
              return true;
            } else {
              reject?.(false);
              return false;
            }
          } catch (error) {
            reject?.(error);
            throw error;
          }
        },
        loading: state.isReversing,
        disabled: state.isReversing || !state.displayInfo || hasValidationError
      },
      secondary: {
        label: t("common.cancel"),
        disabled: state.isReversing,
        onClick: async () => {
          reject?.("cancel");
          hide();
          return false;
        }
      }
    };
  }, [state, t, resolve, reject, hide]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.SimpleDialog,
    {
      open: visible,
      onOpenChange,
      size: "sm",
      title: i18n.i18n.t("positions.reverse.title"),
      classNames: {
        content: "oui-border oui-border-line-6"
      },
      actions,
      closable: !state.isReversing,
      children: /* @__PURE__ */ jsxRuntime.jsx(ReversePosition, { ...state })
    }
  );
};
var ReversePositionDialogId = "ReversePositionDialogId";
ui.registerSimpleDialog(
  ReversePositionDialogId,
  ReversePositionWidget,
  {
    size: "sm",
    classNames: {
      content: "oui-border oui-border-line-6"
    },
    title: () => i18n.i18n.t("positions.reverse.title")
  }
);
var PositionsTabName = /* @__PURE__ */ ((PositionsTabName2) => {
  PositionsTabName2["Positions"] = "positions";
  PositionsTabName2["PositionHistory"] = "positionHistory";
  return PositionsTabName2;
})(PositionsTabName || {});
function useTabSort(options) {
  const [tabSort, setTabSort] = hooks.useSessionStorage(options.storageKey, {
    ["positions" /* Positions */]: {
      sortKey: "unrealized_pnl",
      sortOrder: "desc"
    },
    ["positionHistory" /* PositionHistory */]: {
      sortKey: "close_timestamp",
      sortOrder: "desc"
    }
  });
  const onTabSort = React2.useCallback(
    (type) => (sort) => {
      setTabSort({ ...tabSort, [type]: sort });
    },
    [tabSort, setTabSort]
  );
  return {
    tabSort,
    onTabSort
  };
}

// src/components/positions/positions.script.ts
var usePositionsScript = (props) => {
  const {
    symbol,
    calcMode,
    includedPendingOrder,
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    onSymbolChange,
    enableSortingStorage = true
    // Default to true for backward compatibility
  } = props;
  const { pagination, setPage } = ui.usePagination({ pageSize: 50 });
  const { isEnabled: positionReverse } = useReversePositionEnabled();
  const { tabSort, onTabSort } = useTabSort({
    storageKey: TRADING_POSITIONS_SORT_STORAGE_KEY
  });
  const { onSort, getSortedList, sort } = useSort(
    enableSortingStorage ? tabSort?.["positions" /* Positions */] : void 0,
    enableSortingStorage ? onTabSort("positions" /* Positions */) : void 0
  );
  React2__default.default.useEffect(() => {
    setPage(1);
  }, [symbol]);
  const [data, , { isLoading }] = hooks.usePositionStream(symbol, {
    calcMode,
    includedPendingOrder
  });
  const rawDataSource = reactApp.useDataTap(data?.rows, { fallbackData: [] }) ?? void 0;
  const dataSource = getSortedList(rawDataSource || []);
  return {
    dataSource,
    isLoading,
    // will be use isLoading when usePositionStream support
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    symbol,
    onSymbolChange,
    pagination,
    onSort,
    positionReverse,
    initialSort: enableSortingStorage ? tabSort?.["positions" /* Positions */] : void 0
  };
};
var SymbolContext = React2.createContext(
  {}
);
var useSymbolContext = () => {
  return React2.useContext(SymbolContext);
};
var SymbolProvider = (props) => {
  const { symbol, children } = props;
  const symbolInfo = hooks.useSymbolsInfo()[symbol];
  const memoizedValue = React2.useMemo(() => {
    return {
      symbol,
      base_dp: symbolInfo("base_dp"),
      quote_dp: symbolInfo("quote_dp"),
      base_tick: symbolInfo("base_tick"),
      quote_tick: symbolInfo("quote_tick"),
      base: symbolInfo("base"),
      quote: symbolInfo("quote"),
      origin: symbolInfo(),
      quote_max: symbolInfo("quote_max"),
      quote_min: symbolInfo("quote_min")
    };
  }, [symbol, symbolInfo]);
  return /* @__PURE__ */ jsxRuntime.jsx(SymbolContext.Provider, { value: memoizedValue, children });
};
var RwaStatusTag = ({ symbol }) => {
  const { isRwa, open } = hooks.useGetRwaSymbolOpenStatus(symbol);
  const { t } = i18n.useTranslation();
  if (!isRwa) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Flex,
    {
      r: "base",
      px: 2,
      height: 18,
      className: ui.cn(
        "oui-shrink-0",
        open ? "oui-bg-success/15" : "oui-bg-danger/15"
      ),
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", color: open ? "success" : "danger", children: open ? t("trading.rwa.marketHours") : t("trading.rwa.outsideMarketHours") })
    }
  );
};
var QuantitySlider = (props) => {
  const { t } = i18n.useTranslation();
  const [sliderValue, setSliderValue] = React2.useState(props.value);
  React2.useEffect(() => {
    setSliderValue(props.value);
  }, [props.value]);
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [25, 50, 75, 100].map((e, index) => /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        variant: "outlined",
        size: "xs",
        color: "secondary",
        onClick: () => {
          props.onValueChange(e);
          setSliderValue(e);
        },
        className: ui.cn(
          "oui-w-1/4",
          sliderValue === e ? "oui-border-primary oui-text-primary" : ""
        ),
        children: `${e}%`
      },
      index
    )) }),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Slider,
      {
        showTip: true,
        markCount: 4,
        min: 0,
        max: 100,
        value: [props.value],
        color: "primary",
        onValueChange: (value) => {
          props.onValueChange(value[0]);
          setSliderValue(value[0]);
        }
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", justify: "between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "primary", size: "2xs", children: `${props.value}%` }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, onClick: props.onMax, className: "oui-cursor-pointer", children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", color: "primary", children: t("common.max") }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.numeral,
          {
            intensity: 54,
            size: "2xs",
            dp: props.base_dp,
            padding: false,
            children: props.max
          }
        )
      ] })
    ] })
  ] });
};
var MobileClosePosition = (props) => {
  const {
    base,
    base_dp,
    quote,
    quote_dp,
    sheetOpen,
    setSheetOpen,
    dialogOpen,
    setDialogOpen,
    quantity,
    price,
    submitting,
    priceErrorMsg,
    quantityErrorMsg,
    disabled,
    isMarketClose,
    position,
    maxQty,
    isBuy
  } = props;
  const { t } = i18n.useTranslation();
  const title = isMarketClose ? t("positions.marketClose") : t("positions.limitClose");
  const orderType = isMarketClose ? t("orderEntry.orderType.market") : t("orderEntry.orderType.limit");
  const orderSide = isBuy ? /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "success", size: "xs", children: t("common.buy") }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "danger", size: "xs", children: t("common.sell") });
  const header = /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", justify: "between", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text.formatted, { rule: "symbol", showIcon: true, children: position.symbol }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "neutral", size: "xs", children: orderType }),
      orderSide
    ] })
  ] });
  const lastPrice = /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", justify: "between", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("common.lastPrice") }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: quote_dp, suffix: /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { intensity: 36, children: [
      " ",
      quote
    ] }), children: position.mark_price })
  ] });
  const inputForm = !isMarketClose && /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, { children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", direction: "column", gap: 2, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Input.tooltip,
      {
        prefix: t("common.price"),
        suffix: quote,
        align: "right",
        fullWidth: true,
        autoComplete: "off",
        formatters: [
          ui.inputFormatter.numberFormatter,
          ui.inputFormatter.dpFormatter(quote_dp)
        ],
        triggerClassName: "oui-w-full",
        tooltip: priceErrorMsg,
        color: priceErrorMsg ? "danger" : void 0,
        value: price,
        onValueChange: props.updatePriceChange,
        classNames: {
          prefix: "oui-text-base-contrast-54",
          suffix: "oui-text-base-contrast-54",
          root: ui.cn(
            "oui-w-full oui-outline-line-12",
            priceErrorMsg ? "oui-outline-danger" : void 0
          )
        }
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Input.tooltip,
      {
        prefix: t("common.quantity"),
        suffix: base,
        align: "right",
        fullWidth: true,
        autoComplete: "off",
        formatters: [
          ui.inputFormatter.numberFormatter,
          ui.inputFormatter.dpFormatter(base_dp)
        ],
        triggerClassName: "oui-w-full",
        tooltip: quantityErrorMsg,
        color: quantityErrorMsg ? "danger" : void 0,
        value: quantity,
        onBlur: () => props.formatQuantityToBaseTick(quantity),
        onValueChange: props.updateQuantity,
        classNames: {
          prefix: "oui-text-base-contrast-54",
          suffix: "oui-text-base-contrast-54",
          root: ui.cn(
            "oui-w-full oui-outline-line-12",
            quantityErrorMsg ? "oui-outline-danger" : void 0
          )
        }
      }
    )
  ] }) });
  const slider = /* @__PURE__ */ jsxRuntime.jsx(
    QuantitySlider,
    {
      value: props.sliderValue,
      onValueChange: props.onSliderValueChange,
      base_dp,
      max: maxQty,
      onMax: props.onMax
    }
  );
  const footer = /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", gap: 3, mt: 2, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { fullWidth: true, color: "secondary", onClick: props.onCloseSheet, children: t("common.cancel") }),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.ThrottledButton,
      {
        fullWidth: true,
        disabled: submitting || disabled,
        loading: submitting,
        onClick: (e) => {
          e.stopPropagation();
          e.preventDefault();
          props.onDoubleConfirm();
        },
        children: t("common.confirm")
      }
    )
  ] });
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        variant: "outlined",
        color: "secondary",
        size: "sm",
        className: "oui-border-base-contrast-36",
        onClick: (e) => {
          e.stopPropagation();
          e.preventDefault();
          setSheetOpen(true);
        },
        children: title
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.SimpleSheet, { title, open: sheetOpen, onOpenChange: setSheetOpen, children: /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Flex,
      {
        direction: "column",
        gap: 3,
        width: "100%",
        itemAlign: "start",
        className: "oui-text-sm",
        children: [
          header,
          /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { className: "oui-w-full" }),
          lastPrice,
          inputForm,
          slider,
          footer
        ]
      }
    ) }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.SimpleDialog, { open: dialogOpen, onOpenChange: setDialogOpen, size: "xs", children: isMarketClose ? /* @__PURE__ */ jsxRuntime.jsx(
      MarketCloseConfirm,
      {
        base,
        quantity,
        submitting,
        onClose: props.onCloseDialog,
        onConfirm: props.onConfirm,
        hideCloseIcon: true
      }
    ) : /* @__PURE__ */ jsxRuntime.jsx(
      LimitConfirmDialog,
      {
        base,
        quoteDp: quote_dp,
        quantity,
        price,
        submitting,
        order: props.closeOrderData,
        onClose: props.onCloseDialog,
        onConfirm: props.onConfirm,
        hideCloseIcon: true
      }
    ) })
  ] });
};
var DesktopClosePosition = (props) => {
  const {
    position,
    base,
    base_dp,
    quote_dp,
    price,
    quantity,
    submitting,
    dialogOpen,
    setDialogOpen,
    isMarketClose,
    popoverOpen,
    setPopoverOpen,
    maxQty,
    quantityErrorMsg,
    priceErrorMsg,
    isEntirePosition
  } = props;
  const { t } = i18n.useTranslation();
  const quantityForm = /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", gap: 2, mb: 1, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Input.tooltip,
      {
        ref: props.quantityInputRef,
        prefix: t("common.quantity"),
        suffix: isEntirePosition ? /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text,
          {
            size: "2xs",
            intensity: 54,
            className: "oui-cursor-pointer oui-px-3",
            onClick: props.onEntirePosition,
            children: t("tpsl.entirePosition")
          }
        ) : base,
        align: "right",
        size: "md",
        fullWidth: true,
        autoComplete: "off",
        formatters: [
          ui.inputFormatter.numberFormatter,
          ui.inputFormatter.dpFormatter(base_dp)
        ],
        triggerClassName: "oui-w-full",
        tooltip: quantityErrorMsg,
        color: quantityErrorMsg ? "danger" : void 0,
        value: isEntirePosition ? "" : quantity,
        onBlur: () => props.formatQuantityToBaseTick(quantity),
        onValueChange: props.updateQuantity,
        classNames: {
          prefix: "oui-text-base-contrast-54",
          suffix: "oui-text-base-contrast-54",
          root: ui.cn(
            "oui-w-full oui-outline-line-12",
            quantityErrorMsg ? "oui-outline-danger" : void 0
          )
        }
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        onClick: props.onMax,
        size: "md",
        variant: "outlined",
        className: ui.cn(
          "oui-text-2xs",
          isEntirePosition ? "oui-border-primary-light oui-text-primary-light hover:oui-bg-primary-light/20" : "oui-border-line-12 oui-bg-base-6 oui-text-base-contrast-54 hover:oui-bg-base-5"
        ),
        children: t("common.position")
      }
    )
  ] });
  const suffix = /* @__PURE__ */ jsxRuntime.jsx(
    ui.Select.options,
    {
      variant: "text",
      size: "md",
      options: [
        {
          label: t("orderEntry.orderType.limit"),
          value: types.OrderType.LIMIT
        },
        {
          label: t("orderEntry.orderType.market"),
          value: types.OrderType.MARKET
        }
      ],
      classNames: {
        // set the width of the trigger to the width of the content
        trigger: "oui-w-[--radix-select-content-available-width]"
      },
      value: props.type,
      onValueChange: (value) => {
        props.updateOrderType(value);
      },
      contentProps: {
        align: "end",
        className: "oui-border oui-border-line-6"
      }
    }
  );
  const disabledInput = isMarketClose;
  const priceForm = /* @__PURE__ */ jsxRuntime.jsx(
    ui.Input.tooltip,
    {
      suffix,
      size: "md",
      fullWidth: true,
      autoComplete: "off",
      formatters: [
        ui.inputFormatter.numberFormatter,
        ui.inputFormatter.dpFormatter(quote_dp)
      ],
      triggerClassName: "oui-w-full",
      tooltip: priceErrorMsg,
      color: priceErrorMsg ? "danger" : void 0,
      value: price,
      onValueChange: props.updatePriceChange,
      placeholder: disabledInput ? "--" : "",
      disabled: disabledInput,
      classNames: {
        prefix: "oui-text-base-contrast-54",
        suffix: "oui-text-base-contrast-54",
        root: ui.cn(
          "oui-w-full oui-outline-line-12",
          priceErrorMsg ? "oui-outline-danger" : void 0,
          // when disabled, override the outline color
          disabledInput ? "focus-within:oui-outline-line-12" : void 0
        )
      }
    }
  );
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.PopoverRoot, { open: popoverOpen, onOpenChange: setPopoverOpen, children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.PopoverTrigger,
        {
          asChild: true,
          onClick: () => {
            setPopoverOpen(true);
          },
          children: /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { variant: "outlined", size: "sm", color: "secondary", children: t("positions.column.close") })
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.PopoverContent,
        {
          className: ui.cn(
            "oui-w-[360px] oui-p-5",
            popoverOpen ? "oui-visible" : "oui-invisible"
          ),
          align: "end",
          side: "top",
          onOpenAutoFocus: (e) => e.preventDefault(),
          children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gapY: 2, direction: "column", children: [
            quantityForm,
            /* @__PURE__ */ jsxRuntime.jsx(
              QuantitySlider,
              {
                value: props.sliderValue,
                onValueChange: props.onSliderValueChange,
                base_dp,
                max: maxQty,
                onMax: props.onMax
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { my: 2, intensity: 8, className: "oui-w-full" }),
            priceForm,
            /* @__PURE__ */ jsxRuntime.jsx(
              ConfirmFooter,
              {
                onConfirm: props.onDoubleConfirm,
                onCancel: props.onClosePopover,
                submitting,
                disabled: props.disabled
              }
            )
          ] })
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.SimpleDialog, { open: dialogOpen, onOpenChange: setDialogOpen, size: "sm", children: isMarketClose ? /* @__PURE__ */ jsxRuntime.jsx(
      MarketCloseConfirm,
      {
        base,
        quantity,
        submitting,
        onClose: props.onCloseDialog,
        onConfirm: props.onConfirm,
        classNames: {
          root: "oui-items-start"
        },
        hideCloseIcon: true
      }
    ) : /* @__PURE__ */ jsxRuntime.jsx(
      LimitConfirmDialog,
      {
        base,
        quoteDp: quote_dp,
        quantity,
        price,
        submitting,
        order: props.closeOrderData,
        onClose: props.onCloseDialog,
        onConfirm: props.onConfirm,
        hideCloseIcon: true
      }
    ) })
  ] });
};
var useClosePositionScript = (props) => {
  const [sheetOpen, setSheetOpen] = React2.useState(false);
  const [dialogOpen, setDialogOpen] = React2.useState(false);
  const [popoverOpen, setPopoverOpen] = React2.useState(false);
  const [sliderValue, setSliderValue] = React2.useState(100);
  const quantityInputRef = React2.useRef(null);
  const [orderConfirm] = hooks.useLocalStorage("orderly_order_confirm", true);
  const {
    position,
    onSubmit,
    price,
    quantity,
    closeOrderData,
    submitting,
    updatePriceChange,
    updateOrderType,
    updateQuantity,
    errors,
    baseTick,
    type
  } = usePositionsRowContext();
  const { quote_dp, base_dp, base, quote } = useSymbolContext();
  const { getErrorMsg } = reactApp.useOrderEntryFormErrorMsg(errors);
  const isBuy = position.position_qty > 0;
  const maxQty = Math.abs(position.position_qty);
  const closeType = props.type || type;
  const isMarketClose = closeType === types.OrderType.MARKET;
  const isEntirePosition = maxQty?.toString() === quantity.toString();
  React2.useEffect(() => {
    if (sheetOpen) {
      updateOrderType(props.type);
    }
  }, [props.type, sheetOpen]);
  const { priceErrorMsg, quantityErrorMsg } = React2.useMemo(() => {
    return {
      priceErrorMsg: getErrorMsg("order_price"),
      quantityErrorMsg: getErrorMsg("order_quantity")
    };
  }, [errors]);
  const disabled = !!(priceErrorMsg || quantityErrorMsg);
  const formatQuantityToBaseTick = React2.useCallback(
    (value) => {
      if (baseTick && baseTick > 0) {
        const formatQty = hooks.utils.formatNumber(value, baseTick) ?? value;
        updateQuantity(formatQty);
      }
    },
    [baseTick, updateQuantity]
  );
  const onSliderValueChange = React2.useCallback(
    (value) => {
      setSliderValue(value);
      const qty = new utils.Decimal(value).div(100).mul(maxQty).toFixed(base_dp, utils.Decimal.ROUND_DOWN);
      formatQuantityToBaseTick(qty);
    },
    [maxQty, base_dp, formatQuantityToBaseTick]
  );
  React2.useEffect(() => {
    const qty = Math.min(Number(quantity || 0), maxQty);
    const slider = new utils.Decimal(qty).div(maxQty).mul(100).toDecimalPlaces(2, utils.Decimal.ROUND_DOWN).toNumber();
    setSliderValue(slider);
  }, [quantity]);
  const onMax = React2.useCallback(() => {
    updateQuantity(maxQty?.toString());
  }, [maxQty, updateQuantity]);
  const onConfirm = React2.useCallback(() => {
    return onSubmit().then((res) => {
      setSheetOpen(false);
      setDialogOpen(false);
      setPopoverOpen(false);
    });
  }, [onSubmit]);
  const onDoubleConfirm = React2.useCallback(() => {
    if (!orderConfirm) {
      return onConfirm();
    }
    setDialogOpen(true);
    return Promise.resolve();
  }, [onConfirm, orderConfirm]);
  const onCloseSheet = React2.useCallback(() => {
    setSheetOpen(false);
  }, []);
  const onCloseDialog = React2.useCallback(() => {
    setDialogOpen(false);
  }, []);
  const onClosePopover = React2.useCallback(() => {
    setPopoverOpen(false);
  }, []);
  const onEntirePosition = React2.useCallback(() => {
    updateQuantity("0");
    quantityInputRef.current?.focus();
    setTimeout(() => {
      quantityInputRef.current?.setSelectionRange(0, 1);
    }, 0);
  }, [updateQuantity]);
  return {
    position,
    closeOrderData,
    quote_dp,
    base_dp,
    base,
    quote,
    price,
    quantity,
    submitting,
    onConfirm,
    onDoubleConfirm,
    updatePriceChange,
    updateQuantity,
    formatQuantityToBaseTick,
    updateOrderType,
    sliderValue,
    onSliderValueChange,
    priceErrorMsg,
    quantityErrorMsg,
    disabled,
    isMarketClose,
    maxQty,
    onMax,
    isBuy,
    isEntirePosition,
    type: closeType,
    quantityInputRef,
    onEntirePosition,
    sheetOpen,
    onCloseSheet,
    setSheetOpen,
    dialogOpen,
    setDialogOpen,
    onCloseDialog,
    popoverOpen,
    setPopoverOpen,
    onClosePopover
  };
};
var ClosePositionWidget = (props) => {
  const state = useClosePositionScript(props);
  const { isMobile } = ui.useScreen();
  if (isMobile) {
    return /* @__PURE__ */ jsxRuntime.jsx(MobileClosePosition, { ...state });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(DesktopClosePosition, { ...state });
};
var TPSLEditIcon = () => {
  const { position, baseDp, quoteDp, tpslOrder } = usePositionsRowContext();
  const { isMobile } = ui.useScreen();
  const onEdit = () => {
    const dialogId = isMobile ? uiTpsl.TPSLDetailSheetId : uiTpsl.TPSLDetailDialogId;
    ui.modal.show(dialogId, {
      order: tpslOrder,
      position,
      baseDP: baseDp,
      quoteDP: quoteDp
    });
  };
  return (
    // <PositionTPSLPopover
    //   position={position}
    //   order={tpslOrder}
    //   baseDP={baseDp}
    //   quoteDP={quoteDp}
    //   isEditing
    // >
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.EditIcon,
      {
        onClick: onEdit,
        opacity: 1,
        className: "oui-cursor-pointer oui-text-base-contrast-54",
        size: 16
      }
    )
  );
};
var AddIcon = (props) => {
  const { position, baseDp, quoteDp, tpslOrder } = usePositionsRowContext();
  const { t } = i18n.useTranslation();
  const { isMobile } = ui.useScreen();
  const onAdd = () => {
    const dialogId = isMobile ? uiTpsl.TPSLSheetId : uiTpsl.TPSLDialogId;
    const modalParams = {
      position,
      symbol: position.symbol,
      baseDP: baseDp,
      quoteDP: quoteDp,
      isEditing: false,
      positionType: props.positionType
    };
    ui.modal.show(dialogId, modalParams);
  };
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Text,
    {
      className: ui.cn(
        "oui-cursor-pointer oui-text-base-contrast",
        isMobile && "oui-text-base-contrast-80"
      ),
      onClick: onAdd,
      children: t("tpsl.add")
    }
  );
};
var LeverageBadge = (props) => {
  const { symbol, leverage } = props;
  const showModal = () => {
    ui.modal.show(props.modalId, {
      symbol,
      curLeverage: Number(leverage)
    });
  };
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: ui.cn(
        "oui-flex oui-h-[18px] oui-items-center oui-gap-1",
        "oui-cursor-pointer oui-rounded oui-bg-line-6 oui-px-2",
        "oui-text-2xs oui-font-semibold oui-text-base-contrast-36"
      ),
      onClick: showModal,
      children: leverage ? /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 0, rm: utils.Decimal.ROUND_DOWN, size: "2xs", unit: "X", children: leverage }) : /* @__PURE__ */ jsxRuntime.jsx(LeverageDisplay, { symbol })
    }
  );
};
var LeverageDisplay = ({ symbol }) => {
  const leverage = hooks.useLeverageBySymbol(symbol);
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 0, rm: utils.Decimal.ROUND_DOWN, size: "2xs", unit: "X", children: leverage || 1 });
};
var renderQuantity = (value) => {
  const symbolInfo = useSymbolContext();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Text.numeral,
    {
      dp: symbolInfo.base_dp,
      rm: utils.Decimal.ROUND_DOWN,
      padding: false,
      coloring: true,
      children: value
    }
  );
};
var NumeralWithCtx = (props) => {
  const { tick = "quote_dp", children, ...rest } = props;
  const symbolInfo = useSymbolContext();
  if (!symbolInfo) {
    throw new Error("NumeralWithCtx must be used inside SymbolProvider");
  }
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { as: "span", dp: symbolInfo[tick], ...rest, children });
};
var PartialTPSL = (props) => {
  const {
    orderNum,
    tpTriggerPrice,
    slTriggerPrice,
    direction = "column"
  } = props;
  const {
    partialTPSLOrder: order,
    quoteDp,
    baseDp,
    position
  } = usePositionsRowContext();
  const symbolInfo = hooks.useSymbolsInfo();
  const { t } = i18n.useTranslation();
  const side = position.position_qty > 0 ? types.OrderSide.BUY : types.OrderSide.SELL;
  const slPriceError = hooks.useTpslPriceChecker({
    slPrice: slTriggerPrice?.toString() ?? void 0,
    liqPrice: position.est_liq_price ?? null,
    side
  });
  const child = React2.useMemo(() => {
    const children = [];
    if (!order?.symbol) return /* @__PURE__ */ jsxRuntime.jsx(AddIcon, { positionType: types.PositionType.PARTIAL });
    if (tpTriggerPrice) {
      children.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.formatted,
          {
            className: ui.cn(
              "oui-text-trade-profit oui-gap-0 oui-decoration-white/20"
            ),
            rule: "price",
            dp: symbolInfo[order.symbol]("quote_dp", 2),
            children: tpTriggerPrice,
            prefix: !slTriggerPrice || direction === "column" ? /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: `${t("tpsl.tp")} - ` }) : ""
          },
          "tp"
        )
      );
    }
    if (slTriggerPrice) {
      children.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.formatted,
          {
            className: ui.cn(
              "oui-text-trade-loss oui-gap-0 oui-decoration-white/20 "
            ),
            rule: "price",
            dp: symbolInfo[order.symbol]("quote_dp", 2),
            prefix: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: `${t("tpsl.sl")} - ` }),
            suffix: /* @__PURE__ */ jsxRuntime.jsx(uiTpsl.CloseToLiqPriceIcon, { slPriceError }),
            children: slTriggerPrice
          },
          "sl"
        )
      );
    }
    if (children.length === 0)
      return /* @__PURE__ */ jsxRuntime.jsx(AddIcon, { positionType: types.PositionType.PARTIAL });
    if (children.length === 2 && direction === "row") {
      children.splice(1, 0, /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: "/" }, "split"));
    }
    return children;
  }, [tpTriggerPrice, slTriggerPrice, order?.symbol, t, slPriceError]);
  const hasTPSL = Array.isArray(child) ? !!child.length : !child;
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { className: "oui-gap-[6px]", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: ui.cn(
          "oui-inline-flex oui-text-base-contrast-36",
          direction === "column" ? "oui-flex-col" : "oui-flex-row oui-gap-1"
        ),
        children: child
      }
    ),
    hasTPSL && /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { className: "oui-text-base-contrast-54", children: [
        "(",
        orderNum,
        ")"
      ] }),
      /* @__PURE__ */ jsxRuntime.jsx(TPSLEditIcon, {})
    ] })
  ] });
};
var ReversePositionButton = (props) => {
  const { position } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: "",
      style: { transform: "rotate(90deg)" },
      onClick: () => {
        ui.modal.show(ReversePositionDialogId, {
          position
        });
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(
        ui.ArrowLeftRightIcon,
        {
          size: 15,
          opacity: 1,
          className: "oui-cursor-pointer oui-text-base-contrast-54 hover:oui-text-base-contrast-80"
        }
      )
    }
  );
};
var ShareButton = (props) => {
  if (!props.sharePnLConfig) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    "button",
    {
      type: "button",
      onClick: (e) => {
        e.stopPropagation();
        props.showModal();
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.ShareIcon, { color: "white", opacity: 0.54, size: props.iconSize ?? 16 })
    }
  );
};
var useShareButtonScript = (props) => {
  const { position, sharePnLConfig, iconSize } = props;
  const { getFirstRefCode } = hooks.useReferralInfo();
  const symbolsInfo = hooks.useSymbolsInfo();
  const { data: accountInfo } = hooks.useAccountInfo();
  const refCode = React2.useMemo(() => {
    return getFirstRefCode()?.code;
  }, [getFirstRefCode]);
  const symbolLeverage = hooks.useLeverageBySymbol(position.symbol);
  const getHistoryEntity = () => {
    const netPnL = position.netPnL || 0;
    const openPrice = Math.abs(position.avg_open_price);
    const quantity = Math.abs(position.closed_position_qty);
    let roi;
    const symbolInfo = symbolsInfo[position.symbol];
    const baseIMR = symbolInfo("base_imr");
    const IMR_Factor = symbolInfo("imr_factor");
    if (netPnL !== 0 && quantity !== 0 && openPrice !== 0 && accountInfo?.max_leverage && baseIMR && // IMR_Factor is possible to be 0
    typeof IMR_Factor !== "undefined") {
      const notional = perp.positions.notional(quantity, openPrice);
      const maxLeverage = position.leverage ? position.leverage : accountInfo.max_leverage;
      const imr = perp.account.IMR({
        maxLeverage,
        baseIMR,
        IMR_Factor,
        positionNotional: notional,
        ordersNotional: 0,
        IMR_factor_power: 4 / 5
      });
      const unrealizedPnLROI = perp.positions.unrealizedPnLROI({
        positionQty: quantity,
        openPrice,
        IMR: imr,
        unrealizedPnL: netPnL
      });
      roi = utils.formatNum.roi(unrealizedPnLROI * 100)?.toString();
    }
    return {
      side: position.side,
      pnl: netPnL,
      roi,
      openPrice,
      closePrice: Math.abs(position.avg_close_price),
      openTime: position.open_timestamp,
      closeTime: position.close_timestamp,
      quantity: position.closed_position_qty
    };
  };
  const showModal = () => {
    const entity = props.isPositionHistory ? getHistoryEntity() : {
      side: position.position_qty > 0 ? "LONG" : "SHORT",
      pnl: utils.formatNum.pnl(position.unrealized_pnl)?.toNumber() ?? position.unrealized_pnl,
      roi: utils.formatNum.roi(position.unrealized_pnl_ROI)?.mul(100).toNumber() ?? position.unrealized_pnl_ROI,
      openPrice: Math.abs(position.average_open_price),
      markPrice: position.mark_price,
      openTime: position.timestamp,
      quantity: position.position_qty
    };
    ui.modal.show(props.modalId, {
      pnl: {
        entity: {
          symbol: position.symbol,
          // when position.leverage is empty, use leverage from useSymbolLeverage
          leverage: position.leverage || symbolLeverage,
          ...entity
        },
        refCode,
        ...sharePnLConfig
      }
    });
  };
  return {
    iconSize,
    sharePnLConfig,
    showModal
  };
};
var ShareButtonWidget = (props) => {
  const state = useShareButtonScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(ShareButton, { ...state });
};
var TriggerPrice = (props) => {
  const { stopLossPrice, takeProfitPrice } = props;
  const { tpslOrder, position } = usePositionsRowContext();
  const side = position.position_qty > 0 ? types.OrderSide.BUY : types.OrderSide.SELL;
  const slPriceError = hooks.useTpslPriceChecker({
    slPrice: stopLossPrice?.toString() ?? void 0,
    liqPrice: position.est_liq_price ?? null,
    side
  });
  return /* @__PURE__ */ jsxRuntime.jsx(
    TPSLTriggerPrice,
    {
      stopLossPrice,
      takeProfitPrice,
      direction: "column",
      order: tpslOrder,
      position,
      slPriceError,
      tooltip: true
    }
  );
};
var TriggerPriceItem = (props) => {
  const { qty, price, entryPrice, orderSide, orderType, symbolInfo } = props;
  const pnl = utils.formatNum.pnl(
    hooks.utils.priceToPnl(
      {
        qty,
        price,
        entryPrice,
        orderSide,
        orderType
      },
      {
        symbol: symbolInfo
      }
    )
  ).toNumber();
  const type = orderType === types.AlgoOrderType.TAKE_PROFIT ? "TP" : "SL";
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, className: "oui-mr-1", children: `${type} PnL:` }),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Text.formatted,
      {
        rule: "price",
        className: pnl === 0 ? "oui-text-base-contrast-36" : pnl > 0 ? "oui-text-trade-profit oui-gap-0" : "oui-text-trade-loss oui-gap-0",
        prefix: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: pnl === 0 ? "" : pnl > 0 ? "+" : "-" }),
        suffix: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 36, className: "oui-ml-1", children: "USDC" }),
        children: `${Math.abs(pnl)}`
      }
    )
  ] });
};
var TPSLTriggerPrice = (props) => {
  const { direction = "row", order, position } = props;
  const symbolInfo = hooks.useSymbolsInfo();
  const { t } = i18n.useTranslation();
  const pnl = React2.useMemo(() => {
    const msgs = [];
    if (!props.tooltip || !order || !position) return;
    let quantity = order.quantity;
    if (quantity === 0) {
      if (order.child_orders[0].type === "CLOSE_POSITION") {
        quantity = position.position_qty;
      }
    }
    if (!!props.takeProfitPrice) {
      msgs.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          TriggerPriceItem,
          {
            qty: quantity,
            price: props.takeProfitPrice,
            entryPrice: position.average_open_price,
            orderSide: order.side,
            orderType: types.AlgoOrderType.TAKE_PROFIT,
            symbolInfo: symbolInfo[order.symbol]()
          },
          "tp"
        )
      );
    }
    if (!!props.stopLossPrice) {
      msgs.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          TriggerPriceItem,
          {
            qty: quantity,
            price: props.stopLossPrice,
            entryPrice: position.average_open_price,
            orderSide: order.side,
            orderType: types.AlgoOrderType.STOP_LOSS,
            symbolInfo: symbolInfo[order.symbol]()
          },
          "sl"
        )
      );
    }
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: msgs });
  }, [
    props.takeProfitPrice,
    props.stopLossPrice,
    position?.average_open_price,
    order?.side,
    order?.quantity,
    order?.algo_type
  ]);
  const child = React2.useMemo(() => {
    const children = [];
    if (!order?.symbol) return /* @__PURE__ */ jsxRuntime.jsx(AddIcon, { positionType: types.PositionType.FULL });
    if (props.takeProfitPrice) {
      children.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.formatted,
          {
            className: ui.cn(
              "oui-text-trade-profit oui-gap-0 oui-decoration-white/20"
            ),
            rule: "price",
            dp: symbolInfo[order.symbol]("quote_dp", 2),
            children: props.takeProfitPrice,
            prefix: !props.stopLossPrice || direction === "column" ? /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: `${t("tpsl.tp")} - ` }) : ""
          },
          "tp"
        )
      );
    }
    if (props.stopLossPrice) {
      children.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.formatted,
          {
            className: ui.cn(
              "oui-text-trade-loss oui-gap-0 oui-decoration-white/20 "
            ),
            rule: "price",
            dp: symbolInfo[order.symbol]("quote_dp", 2),
            children: props.stopLossPrice,
            prefix: !props.takeProfitPrice || direction === "column" ? /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: `${t("tpsl.sl")} - ` }) : "",
            suffix: /* @__PURE__ */ jsxRuntime.jsx(uiTpsl.CloseToLiqPriceIcon, { slPriceError: props.slPriceError })
          },
          "sl"
        )
      );
    }
    if (children.length === 0)
      return /* @__PURE__ */ jsxRuntime.jsx(AddIcon, { positionType: types.PositionType.FULL });
    if (children.length === 2 && direction === "row") {
      children.splice(1, 0, /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: "/" }, "split"));
    }
    return children;
  }, [
    props.takeProfitPrice,
    props.stopLossPrice,
    order?.symbol,
    t,
    props.slPriceError
  ]);
  const content = /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: ui.cn(
        "oui-inline-flex oui-text-base-contrast-36",
        props.direction === "column" ? "oui-flex-col" : "oui-flex-row oui-gap-1",
        props.className
      ),
      children: child
    }
  );
  const showIcon = Array.isArray(child) ? !!child.length : !child;
  if (props.tooltip) {
    return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gapX: 2, children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { content: pnl, children: content }),
      showIcon && /* @__PURE__ */ jsxRuntime.jsx(TPSLEditIcon, {})
    ] });
  }
  return content;
};
var UnrealizedPnLPopoverCard = (props) => {
  const [unPnlPriceBasis, setUnPnlPriceBasic] = hooks.useLocalStorage(
    "unPnlPriceBasis",
    "markPrice"
  );
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      gap: 3,
      className: "oui-text-base-contrast-54 oui-items-start",
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.column.unrealPnl.tooltip") }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { className: "oui-w-full" }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mb-0", children: t("positions.column.unrealPnl.priceBasis") }),
        /* @__PURE__ */ jsxRuntime.jsx(
          UnPnlPriceBasisCheckBox,
          {
            value: unPnlPriceBasis,
            onValueChange: setUnPnlPriceBasic
          }
        )
      ]
    }
  );
};
var UnPnlPriceBasisCheckBox = (props) => {
  const { value, onValueChange } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      RadioButton,
      {
        sel: value === "markPrice",
        label: t("common.markPrice"),
        value: "markPrice",
        onCheckChange: onValueChange
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      RadioButton,
      {
        sel: value === "lastPrice",
        label: t("common.lastPrice"),
        value: "lastPrice",
        onCheckChange: onValueChange
      }
    )
  ] });
};
var RadioButton = (props) => {
  const { sel, label, value, onCheckChange } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      onClick: (e) => {
        onCheckChange(value);
        e.stopPropagation();
      },
      gap: 1,
      className: "oui-cursor-pointer",
      children: [
        sel ? /* @__PURE__ */ jsxRuntime.jsx(SelIcon, {}) : /* @__PURE__ */ jsxRuntime.jsx(UnselIcon, {}),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "xs", intensity: sel ? 98 : 54, children: label })
      ]
    }
  );
};
var SelIcon = () => {
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "svg",
    {
      width: "16",
      height: "16",
      viewBox: "0 0 16 16",
      fill: "currentColor",
      xmlns: "http://www.w3.org/2000/svg",
      className: "oui-fill-white",
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          "path",
          {
            d: "M8.01 1.333a6.667 6.667 0 1 0 0 13.333 6.667 6.667 0 0 0 0-13.333m0 1.333a5.334 5.334 0 1 1-.001 10.667 5.334 5.334 0 0 1 0-10.667",
            fill: "#fff",
            fillOpacity: ".36"
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx("circle", { cx: "8", cy: "8", r: "3.333" })
      ]
    }
  );
};
var UnselIcon = () => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    "svg",
    {
      width: "16",
      height: "16",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: /* @__PURE__ */ jsxRuntime.jsx(
        "path",
        {
          d: "M8.01 1.333a6.667 6.667 0 1 0 0 13.333 6.667 6.667 0 0 0 0-13.333m0 1.333a5.334 5.334 0 1 1-.001 10.667 5.334 5.334 0 0 1 0-10.667",
          fill: "#fff",
          fillOpacity: ".54"
        }
      )
    }
  );
};
var useColumn = (config) => {
  const {
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    onSymbolChange,
    positionReverse
  } = config;
  const { t } = i18n.useTranslation();
  React2.useRef(Date.now().toString());
  const column = React2.useMemo(
    () => [
      {
        title: t("common.symbol"),
        dataIndex: "symbol",
        fixed: "left",
        width: 200,
        onSort: (r1, r2) => {
          return r1.symbol?.localeCompare(r2.symbol || "");
        },
        render: (value, record) => /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Box,
            {
              width: 4,
              height: 20,
              className: ui.cn(
                "oui-h-[38px] oui-rounded-[1px]",
                record.position_qty > 0 ? "oui-bg-trade-profit" : "oui-bg-trade-loss"
              )
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: "start", children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.formatted,
              {
                formatString: "base-type",
                className: "oui-cursor-pointer",
                onClick: (e) => {
                  onSymbolChange?.({ symbol: value });
                  e.stopPropagation();
                  e.preventDefault();
                },
                children: `${value.split("_")[1]}-PERP`
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, children: [
              /* @__PURE__ */ jsxRuntime.jsx(
                LeverageBadge,
                {
                  symbol: value,
                  leverage: record.leverage,
                  modalId: uiLeverage.SymbolLeverageDialogId
                }
              ),
              /* @__PURE__ */ jsxRuntime.jsx(RwaStatusTag, { symbol: value })
            ] })
          ] })
        ] })
      },
      {
        title: t("common.quantity"),
        dataIndex: "position_qty",
        // onSort: true,
        onSort: (a, b) => {
          return a.position_qty - b.position_qty;
        },
        width: 120,
        className: "oui-pl-6",
        // rule: "price",
        // numeralProps: {
        //   coloring: true,
        //   // tick: "base_dp",
        // },
        render: renderQuantity
        // render: (value: string) => (
        //   <NumeralWithCtx
        //     coloring
        //     className="oui-font-semibold"
        //     tick="base_dp"
        //   >
        //     {value}
        //   </NumeralWithCtx>
        // ),
      },
      {
        title: t("common.avgOpen"),
        className: "oui-h-[48px]",
        width: 120,
        onSort: true,
        dataIndex: "average_open_price",
        render: (value) => {
          return /* @__PURE__ */ jsxRuntime.jsx(NumeralWithCtx, { rm: utils.Decimal.ROUND_DOWN, children: value });
        }
      },
      {
        title: t("common.markPrice"),
        dataIndex: "mark_price",
        width: 120,
        onSort: true,
        className: "oui-h-[48px]",
        render: (value) => {
          return /* @__PURE__ */ jsxRuntime.jsx(NumeralWithCtx, { rm: utils.Decimal.ROUND_DOWN, children: value });
        }
      },
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(
          ui.Tooltip,
          {
            className: "oui-max-w-[280px] oui-bg-base-8 oui-p-3 oui-text-2xs oui-text-base-contrast-54",
            content: /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-pretty", children: t("common.liquidationPrice.tooltip") }),
              /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(
                "a",
                {
                  href: "https://orderly.network/docs/introduction/trade-on-orderly/perpetual-futures/liquidations",
                  target: "_blank",
                  rel: "noopener noreferrer",
                  className: "oui-text-primary oui-underline",
                  children: t("common.liquidationPrice.tooltip.learnMore")
                }
              ) })
            ] }),
            children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { className: "oui-underline oui-decoration-dotted", children: t("positions.column.liqPrice") })
          }
        ),
        width: 100,
        onSort: true,
        dataIndex: "est_liq_price",
        render: (value, record) => {
          const liqPrice = Number(value);
          const markPrice = record.mark_price;
          if (liqPrice === 0 || !markPrice || markPrice === 0 || Math.abs(liqPrice - markPrice) / markPrice >= LIQ_DISTANCE_THRESHOLD) {
            return "--";
          }
          return /* @__PURE__ */ jsxRuntime.jsx(
            NumeralWithCtx,
            {
              rm: utils.Decimal.ROUND_DOWN,
              className: liqPrice > 0 ? "oui-text-warning-light" : "",
              children: value ?? "--"
            }
          );
        }
      },
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(
          ui.HoverCard,
          {
            content: /* @__PURE__ */ jsxRuntime.jsx(UnrealizedPnLPopoverCard, {}),
            side: "top",
            align: "center",
            className: "oui-max-w-[280px] oui-text-2xs",
            children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { className: "oui-underline oui-decoration-dotted", children: t("common.unrealizedPnl") })
          }
        ),
        dataIndex: "unrealized_pnl",
        width: 180,
        onSort: true,
        rule: "price",
        numeralProps: {
          coloring: true
          // tick: "base_dp",
        },
        // hint: <UnrealizedPnLPopoverCard />,
        render: (value, record) => {
          return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, children: [
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.pnl,
                {
                  dp: pnlNotionalDecimalPrecision,
                  coloring: true,
                  className: "oui-font-semibold",
                  children: value
                }
              ),
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.roi,
                {
                  rule: "percentages",
                  dp: pnlNotionalDecimalPrecision,
                  coloring: true,
                  className: "oui-font-semibold",
                  prefix: "(",
                  suffix: ")",
                  children: record.unrealized_pnl_ROI
                }
              )
            ] }),
            /* @__PURE__ */ jsxRuntime.jsx(
              ShareButtonWidget,
              {
                position: record,
                sharePnLConfig,
                modalId: uiShare.SharePnLDialogId
              }
            )
          ] });
        }
      },
      {
        title: t("common.fullTPSL"),
        dataIndex: "full_tpsl",
        width: 150,
        render: (_, record) => {
          return /* @__PURE__ */ jsxRuntime.jsx(
            TriggerPrice,
            {
              stopLossPrice: record.full_tp_sl?.sl_trigger_price,
              takeProfitPrice: record.full_tp_sl?.tp_trigger_price
            }
          );
        }
      },
      {
        title: t("common.partialTPSL"),
        dataIndex: "partial_tpsl",
        width: 150,
        render: (_, record) => /* @__PURE__ */ jsxRuntime.jsx(
          PartialTPSL,
          {
            orderNum: record.partial_tp_sl?.order_num,
            tpTriggerPrice: record.partial_tp_sl?.tp_trigger_price,
            slTriggerPrice: record.partial_tp_sl?.sl_trigger_price
          }
        )
      },
      {
        title: t("common.notional"),
        dataIndex: "notional",
        className: "oui-h-[48px]",
        width: 100,
        onSort: true,
        render: (value) => /* @__PURE__ */ jsxRuntime.jsx(ui.Text.notional, { dp: pnlNotionalDecimalPrecision, children: value })
      },
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(
          ui.Tooltip,
          {
            className: "oui-max-w-[280px] oui-bg-base-8 oui-p-3 oui-text-2xs oui-text-base-contrast-54",
            content: /* @__PURE__ */ jsxRuntime.jsxs(
              ui.Flex,
              {
                direction: "column",
                gap: 3,
                className: "oui-rounded-sm oui-bg-base-8 oui-text-base-contrast-54",
                children: [
                  /* @__PURE__ */ jsxRuntime.jsx("span", { children: t("positions.column.margin.tooltip") }),
                  /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { className: "oui-w-full" }),
                  /* @__PURE__ */ jsxRuntime.jsx("span", { children: t("positions.column.margin.formula") })
                ]
              }
            ),
            children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { className: "oui-underline oui-decoration-dotted", children: t("positions.column.margin") })
          }
        ),
        dataIndex: "mm",
        onSort: true,
        width: 140,
        rule: "price",
        render: (value) => /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { children: value })
      },
      // {
      //   title: t("funding.fundingFee"),
      //   dataIndex: "fundingFee",
      //   width: 100,
      //   render: (value, record) => (
      //     <FundingFeeButton
      //       fee={value}
      //       symbol={record.symbol}
      //       start_t={record.timestamp.toString()}
      //       end_t={fundingFeeEndTime.current}
      //     />
      //   ),
      // },
      // {
      //   title: t("common.qty"),
      //   dataIndex: "close_qty",
      //   width: 100,
      //   fixed: "right",
      //   render: renderQuantityInput,
      // },
      // {
      //   title: t("common.price"),
      //   dataIndex: "close_price",
      //   width: 100,
      //   fixed: "right",
      //   render: renderPriceInput,
      //   // render: (value: string) => <PriceInput />,
      // },
      {
        title: null,
        dataIndex: "close_position",
        align: "right",
        width: positionReverse ? 100 : 70,
        fixed: "right",
        render(_, record) {
          return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gapX: 2, justify: "end", children: [
            /* @__PURE__ */ jsxRuntime.jsx(ClosePositionWidget, {}),
            positionReverse && /* @__PURE__ */ jsxRuntime.jsx(ReversePositionButton, { position: record })
          ] });
        }
      }
    ],
    [pnlNotionalDecimalPrecision, sharePnLConfig, t, positionReverse]
  );
  return column;
};
var SymbolToken = (props) => {
  const { item } = props;
  const isBuy = item.position_qty > 0;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Text.formatted,
    {
      rule: "symbol",
      formatString: "base-type",
      size: "2xs",
      suffix: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1", children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: isBuy ? "success" : "danger", size: "xs", children: isBuy ? t("common.long") : t("common.short") }),
        /* @__PURE__ */ jsxRuntime.jsx(
          LeverageBadge,
          {
            symbol: item.symbol,
            leverage: item.leverage,
            modalId: uiLeverage.SymbolLeverageSheetId
          }
        )
      ] }),
      showIcon: true,
      onClick: () => {
        props.onSymbolChange?.({ symbol: item.symbol });
      },
      children: item.symbol
    }
  );
};
var UnrealPnL = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const unrealPnlLabel = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-text-right", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 36, className: "oui-underline oui-decoration-dotted", children: t("common.unrealizedPnl") }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 20, className: "oui-ml-0.5", children: "(USDC)" })
  ] });
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 3, children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", className: "oui-text-2xs", itemAlign: "end", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Tips,
        {
          trigger: unrealPnlLabel,
          content: /* @__PURE__ */ jsxRuntime.jsx(UnrealizedPnLPopoverCard, {}),
          title: t("common.tips")
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.pnl,
        {
          size: "xs",
          dp: props.pnlNotionalDecimalPrecision,
          coloring: true,
          className: "orderly-font-semibold",
          suffix: /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text.roi,
            {
              rule: "percentages",
              dp: props.pnlNotionalDecimalPrecision,
              prefix: "(",
              suffix: ")",
              className: ui.cn(
                "oui-ml-1",
                item.unrealized_pnl_ROI > 0 ? "oui-text-success-darken" : "oui-text-danger-darken"
              ),
              children: item.unrealized_pnl_ROI
            }
          ),
          children: item.unrealized_pnl
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(
      ShareButtonWidget,
      {
        position: item,
        sharePnLConfig: props.sharePnLConfig,
        modalId: uiShare.SharePnLBottomSheetId,
        iconSize: 12
      }
    )
  ] });
};
var Qty = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: t("common.qty"),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.base_dp, padding: false, coloring: true, children: item.position_qty })
    }
  );
};
var Margin = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const marginTipsContent = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-text-2xs oui-text-base-contrast-80", children: [
    /* @__PURE__ */ jsxRuntime.jsx("div", { children: t("positions.column.margin.tooltip") }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-my-2 oui-h-px oui-w-full oui-bg-base-8" }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { children: t("positions.column.margin.formula") })
  ] });
  const marginLabel = /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { className: "oui-underline oui-decoration-dotted", children: t("positions.column.margin") });
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Tips,
        {
          trigger: marginLabel,
          content: marginTipsContent,
          title: t("common.tips")
        }
      ),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 2, intensity: 80, children: item.mm })
    }
  );
};
var Notional = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      align: "end",
      label: /* @__PURE__ */ jsxRuntime.jsxs("span", { children: [
        t("common.notional"),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 20, children: "(USDC)" })
      ] }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 2, intensity: 80, children: item.notional })
    }
  );
};
var AvgOpen = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: t("common.avgOpen"),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.quote_dp, rm: utils.Decimal.ROUND_DOWN, intensity: 80, children: item.average_open_price })
    }
  );
};
var MarkPrice = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: t("common.markPrice"),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.quote_dp, rm: utils.Decimal.ROUND_DOWN, intensity: 80, children: item.mark_price })
    }
  );
};
var LiqPrice = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const liqPrice = React2.useMemo(() => {
    const estLiqPrice = item.est_liq_price;
    const markPrice = item.mark_price;
    if (!estLiqPrice || estLiqPrice <= 0) return "-";
    if (!markPrice || markPrice === 0) return "-";
    if (Math.abs(estLiqPrice - markPrice) / markPrice >= LIQ_DISTANCE_THRESHOLD) {
      return "-";
    }
    return estLiqPrice;
  }, [item.est_liq_price, item.mark_price]);
  const liqTipsContent = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-text-2xs oui-text-base-contrast-80", children: [
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-pretty", children: t("common.liquidationPrice.tooltip") }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mt-2", children: /* @__PURE__ */ jsxRuntime.jsx(
      "a",
      {
        href: "https://orderly.network/docs/introduction/trade-on-orderly/perpetual-futures/liquidations",
        target: "_blank",
        rel: "noopener noreferrer",
        className: "oui-text-primary oui-underline",
        children: t("common.liquidationPrice.tooltip.learnMore")
      }
    ) })
  ] });
  const liqLabel = /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { className: "oui-underline oui-decoration-dotted", children: t("positions.column.liqPrice") });
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Tips,
        {
          trigger: liqLabel,
          content: liqTipsContent,
          title: t("common.tips")
        }
      ),
      align: "end",
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.quote_dp, rm: utils.Decimal.ROUND_DOWN, color: "warning", children: liqPrice })
    }
  );
};
var TPSLPrice = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const slPriceError = hooks.useTpslPriceChecker({
    slPrice: item.full_tp_sl?.sl_trigger_price?.toString() ?? void 0,
    liqPrice: item.est_liq_price ?? null,
    side: item.position_qty > 0 ? types.OrderSide.BUY : types.OrderSide.SELL
  });
  const partialSlPriceError = hooks.useTpslPriceChecker({
    slPrice: item.partial_tp_sl?.sl_trigger_price?.toString() ?? void 0,
    liqPrice: item.est_liq_price ?? null,
    side: item.position_qty > 0 ? types.OrderSide.BUY : types.OrderSide.SELL
  });
  const fullTPSL = React2.useMemo(() => {
    if (item.full_tp_sl?.tp_trigger_price == null && item.full_tp_sl?.sl_trigger_price == null)
      return /* @__PURE__ */ jsxRuntime.jsx(AddIcon, { positionType: types.PositionType.FULL });
    return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { className: "oui-gap-[2px]", children: [
      item.full_tp_sl?.tp_trigger_price && /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { color: "buy", children: item.full_tp_sl.tp_trigger_price }),
      item.full_tp_sl?.sl_trigger_price && "/",
      item.full_tp_sl?.sl_trigger_price && /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.numeral,
        {
          as: "div",
          color: "sell",
          suffix: /* @__PURE__ */ jsxRuntime.jsx(uiTpsl.CloseToLiqPriceIcon, { slPriceError }),
          className: "oui-flex oui-items-center",
          children: item.full_tp_sl.sl_trigger_price
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(TPSLEditIcon, {})
    ] });
  }, [item.full_tp_sl]);
  const partialTPSL = React2.useMemo(() => {
    if (item.partial_tp_sl?.tp_trigger_price == null && item.partial_tp_sl?.sl_trigger_price == null)
      return /* @__PURE__ */ jsxRuntime.jsx(AddIcon, { positionType: types.PositionType.PARTIAL });
    return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { className: "oui-gap-[2px]", itemAlign: "center", children: [
      item.partial_tp_sl?.tp_trigger_price && /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { color: "buy", children: item.partial_tp_sl.tp_trigger_price }),
      item.partial_tp_sl?.sl_trigger_price && "/",
      item.partial_tp_sl?.sl_trigger_price && /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.numeral,
        {
          as: "div",
          color: "sell",
          suffix: /* @__PURE__ */ jsxRuntime.jsx(uiTpsl.CloseToLiqPriceIcon, { slPriceError: partialSlPriceError }),
          className: "oui-flex oui-items-center",
          children: item.partial_tp_sl.sl_trigger_price
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: `(${item.partial_tp_sl?.order_num})` }),
      /* @__PURE__ */ jsxRuntime.jsx(TPSLEditIcon, {})
    ] });
  }, [item.partial_tp_sl]);
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Grid, { cols: 2, rows: 1, gap: 2, width: "100%", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Flex,
      {
        className: "oui-text-2xs oui-text-base-contrast-36",
        direction: "column",
        itemAlign: "start",
        children: [
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { children: [
            t("common.fullTPSL"),
            ": "
          ] }),
          fullTPSL
        ]
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Flex,
      {
        className: "oui-text-2xs oui-grid-cols-end oui-text-base-contrast-36",
        direction: "column",
        itemAlign: "end",
        children: [
          /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { children: [
            t("common.partialTPSL"),
            ": "
          ] }),
          partialTPSL
        ]
      }
    )
  ] });
};
var PositionCell = (props) => {
  const { className, ...rest } = props;
  const header = /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", children: [
    /* @__PURE__ */ jsxRuntime.jsx(SymbolToken, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(UnrealPnL, { ...props })
  ] });
  const body = /* @__PURE__ */ jsxRuntime.jsxs(ui.Grid, { cols: 3, rows: 2, gap: 2, width: "100%", children: [
    /* @__PURE__ */ jsxRuntime.jsx(Qty, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(Margin, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(Notional, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(AvgOpen, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(MarkPrice, { ...props }),
    /* @__PURE__ */ jsxRuntime.jsx(LiqPrice, { ...props })
  ] });
  const buttons = /* @__PURE__ */ jsxRuntime.jsxs(ui.Grid, { width: "100%", gap: 2, cols: 2, rows: 1, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ClosePositionWidget, { type: types.OrderType.LIMIT }),
    /* @__PURE__ */ jsxRuntime.jsx(ClosePositionWidget, { type: types.OrderType.MARKET })
  ] });
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      width: "100%",
      gap: 2,
      p: 2,
      r: "xl",
      itemAlign: "start",
      className: ui.cn(className, "oui-bg-base-9"),
      children: [
        header,
        /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { intensity: 6, className: "oui-w-full" }),
        body,
        /* @__PURE__ */ jsxRuntime.jsx(TPSLPrice, { ...rest }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
          buttons,
          props.positionReverse && /* @__PURE__ */ jsxRuntime.jsx(ReversePositionButton, { position: props.item })
        ] })
      ]
    }
  );
};

// src/components/positions/mobile/positionCell/positionCell.script.tsx
var usePositionCellScript = (props) => {
  const symbolInfo = useSymbolContext();
  return {
    ...props,
    ...symbolInfo
  };
};
var PositionCellWidget = (props) => {
  const state = usePositionCellScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(PositionCell, { ...state, className: props.className });
};
var PositionsRowProvider = (props) => {
  const { position, children, mutatePositions } = props;
  const { t } = i18n.useTranslation();
  const [quantity, setQuantity] = React2.useState(
    Math.abs(position.position_qty).toString()
  );
  const [price, setPrice] = React2.useState("");
  const [type, setType] = React2.useState(types.OrderType.MARKET);
  React2.useEffect(() => {
    setQuantity(Math.abs(position.position_qty).toString());
  }, [position.position_qty]);
  const symbol = position.symbol;
  const symbolsInfo = hooks.useSymbolsInfo();
  const symbolInfo = symbolsInfo?.[symbol];
  const quoteDp = symbolInfo("quote_dp");
  const baseDp = symbolInfo("base_dp");
  const baseTick = symbolInfo("base_tick");
  const {
    side,
    closeOrderData,
    submit,
    isMutating: submitting,
    errors,
    calculate
  } = hooks.usePositionClose({
    position,
    order: {
      type,
      quantity,
      price
    }
  });
  const updateQuantity = React2.useCallback(
    (value) => {
      const newValues = calculate(
        {},
        "order_quantity",
        value,
        position.mark_price,
        symbolInfo()
      );
      setQuantity(newValues.order_quantity);
    },
    [calculate, symbolsInfo, position.mark_price]
  );
  const updatePriceChange = React2.useCallback(
    (value) => {
      const newValues = calculate(
        {},
        "order_price",
        value,
        position.mark_price,
        symbolInfo()
      );
      setPrice(newValues.order_price);
    },
    [calculate, symbolsInfo, position.mark_price]
  );
  const updateOrderType = React2.useCallback(
    (type2) => {
      setType(type2);
      setPrice(type2 === types.OrderType.LIMIT ? position.mark_price?.toString() : "");
    },
    [position.mark_price]
  );
  const onSubmit = React2.useCallback(async () => {
    if (Number(quantity) > symbolInfo("base_max")) {
      ui.toast.error(
        /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
          t("positions.limitClose.errors.exceed.title"),
          /* @__PURE__ */ jsxRuntime.jsx("br", {}),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-break-normal", children: t("positions.limitClose.errors.exceed.description", {
            quantity,
            symbol: transSymbolformString(symbol, "base"),
            maxQuantity: symbolInfo("base_max")
          }) })
        ] })
      );
      return;
    }
    return submit().then((res) => {
      if (res.success) {
        mutatePositions?.();
        return res;
      }
      if (res.message) {
        ui.toast.error(res.message);
      }
      throw true;
    }).catch((err) => {
      if (err.message) {
        ui.toast.error(err.message);
      }
      return false;
    });
  }, [submit]);
  const memoizedValue = React2.useMemo(() => {
    return {
      quantity,
      price,
      type,
      side,
      position,
      updatePriceChange,
      updateQuantity,
      updateOrderType,
      tpslOrder: position.full_tp_sl?.algo_order,
      partialTPSLOrder: position.partial_tp_sl?.algo_order,
      onSubmit,
      submitting,
      closeOrderData,
      quoteDp,
      baseDp,
      baseTick,
      errors
    };
  }, [
    quantity,
    price,
    type,
    side,
    position,
    updatePriceChange,
    updateQuantity,
    updateOrderType,
    onSubmit,
    submitting,
    closeOrderData,
    quoteDp,
    baseDp,
    baseTick,
    errors
  ]);
  return /* @__PURE__ */ jsxRuntime.jsx(PositionsRowContext.Provider, { value: memoizedValue, children });
};
function transSymbolformString(symbol, formatString = "base") {
  const arr = symbol.split("_");
  const type = arr[0];
  const base = arr[1];
  const quote = arr[2];
  return (formatString ?? "base-quote").replace("type", type).replace("base", base).replace("quote", quote);
}
var Positions = (props) => {
  const {
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    pagination,
    isLoading,
    dataSource,
    onSymbolChange,
    positionReverse
  } = props;
  const columns = useColumn({
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    onSymbolChange,
    positionReverse
  });
  return /* @__PURE__ */ jsxRuntime.jsx(
    uiConnector.AuthGuardDataTable,
    {
      loading: isLoading,
      id: "oui-desktop-positions-content",
      columns,
      bordered: true,
      dataSource,
      generatedRowKey: (record) => record.symbol,
      renderRowContainer: (record, index, children) => {
        return /* @__PURE__ */ jsxRuntime.jsx(SymbolProvider, { symbol: record.symbol, children: /* @__PURE__ */ jsxRuntime.jsx(PositionsRowProvider, { position: record, children }) });
      },
      manualPagination: false,
      pagination,
      manualSorting: true,
      onSort: props.onSort,
      initialSort: props.initialSort ? {
        sortKey: props.initialSort.sortKey,
        sort: props.initialSort.sortOrder
      } : void 0,
      testIds: {
        body: "oui-testid-dataList-position-tab-body"
      }
    }
  );
};
var MobilePositions = (props) => {
  const {
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    dataSource,
    onSymbolChange,
    positionReverse
  } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.ListView,
    {
      className: "oui-hide-scrollbar oui-w-full oui-space-y-0 oui-overflow-y-hidden",
      contentClassName: "!oui-space-y-1",
      dataSource,
      renderItem: (item, index) => /* @__PURE__ */ jsxRuntime.jsx(SymbolProvider, { symbol: item.symbol, children: /* @__PURE__ */ jsxRuntime.jsx(PositionsRowProvider, { position: item, children: /* @__PURE__ */ jsxRuntime.jsx(
        PositionCellWidget,
        {
          item,
          index,
          pnlNotionalDecimalPrecision,
          sharePnLConfig,
          onSymbolChange,
          positionReverse
        }
      ) }) })
    }
  );
};
var CombinePositions = (props) => {
  const {
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    pagination,
    isLoading,
    tableData,
    onSymbolChange,
    mutatePositions
  } = props;
  const columns = useColumn({
    pnlNotionalDecimalPrecision,
    sharePnLConfig,
    onSymbolChange
  });
  const { dataSource = [] } = tableData;
  return /* @__PURE__ */ jsxRuntime.jsx(
    uiConnector.AuthGuardDataTable,
    {
      bordered: true,
      loading: isLoading,
      id: "oui-desktop-positions-content",
      columns,
      dataSource,
      expanded: true,
      getSubRows: (row) => row.children,
      generatedRowKey: (record) => `${record.account_id}${record.symbol || ""}`,
      onCell: (column, record) => {
        const isGroup = (record.children ?? []).length > 0;
        if (isGroup) {
          return {
            children: column.id === "symbol" ? /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "neutral", size: "xs", children: record?.description || ui.formatAddress(record?.id) }) : null
          };
        }
      },
      renderRowContainer: (record, index, children) => {
        if (record.symbol) {
          return /* @__PURE__ */ jsxRuntime.jsx(SymbolProvider, { symbol: record.symbol, children: /* @__PURE__ */ jsxRuntime.jsx(
            PositionsRowProvider,
            {
              position: record,
              mutatePositions,
              children
            }
          ) });
        }
        return children;
      },
      manualPagination: false,
      pagination,
      classNames: {
        scroll: "oui-pb-10"
      },
      testIds: {
        body: "oui-testid-dataList-position-tab-body"
      }
    }
  );
};
var PositionsWidget = (props) => {
  const state = usePositionsScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(Positions, { ...state });
};
var MobilePositionsWidget = (props) => {
  const state = usePositionsScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(MobilePositions, { ...state });
};
var CombinePositionsWidget = (props) => {
  const state = useCombinePositionsScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(CombinePositions, { ...state });
};
var usePositionHistoryColumn = (props) => {
  const { onSymbolChange, pnlNotionalDecimalPrecision } = props;
  const { t } = i18n.useTranslation();
  const column = React2.useMemo(
    () => [
      // instrument
      {
        title: t("common.symbol"),
        dataIndex: "symbol",
        fixed: "left",
        width: 200,
        onSort: (r1, r2) => {
          return r1.symbol?.localeCompare(r2.symbol || "");
        },
        render: (value, record) => /* @__PURE__ */ jsxRuntime.jsx(SymbolInfo, { record, onSymbolChange })
      },
      // quantity
      {
        title: t("positions.history.column.closed&maxClosed"),
        dataIndex: "close_maxClose",
        width: 200,
        render: (value, record) => /* @__PURE__ */ jsxRuntime.jsx(Quantity, { record })
      },
      // net pnl
      {
        title: t("positions.history.column.netPnl"),
        dataIndex: "netPnL",
        width: 140,
        onSort: (a, b) => {
          if (a.netPnL == null || b.netPnL == null) return -1;
          return (a.netPnL ?? 0) - (b.netPnL ?? 0);
        },
        render: (_, record) => /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gapX: 1, children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            NetPnL,
            {
              record,
              pnlNotionalDecimalPrecision
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            ShareButtonWidget,
            {
              position: record,
              sharePnLConfig: props.sharePnLConfig,
              modalId: uiShare.SharePnLDialogId,
              isPositionHistory: true
            }
          )
        ] })
      },
      // avg open
      {
        title: t("common.avgPrice"),
        dataIndex: "avg_open",
        width: 140,
        render: (_, record) => {
          const avgOpen = record.avg_open_price != null ? Math.abs(record.avg_open_price) : "--";
          const { quote_dp } = useSymbolContext();
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: quote_dp, padding: false, children: avgOpen });
        }
      },
      // avg close
      {
        title: t("common.avgClose"),
        dataIndex: "avg_close",
        width: 175,
        render: (_, record) => {
          const avgClose = record.avg_close_price != null ? Math.abs(record.avg_close_price) : "--";
          const { quote_dp } = useSymbolContext();
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: quote_dp, padding: false, children: avgClose });
        }
      },
      {
        title: t("funding.fundingFee"),
        dataIndex: "accumulated_funding_fee",
        render: (value, record) => {
          return /* @__PURE__ */ jsxRuntime.jsx(
            FundingFeeButton,
            {
              fee: -value,
              symbol: record.symbol,
              start_t: record.open_timestamp.toString(),
              end_t: (record.close_timestamp || record.last_update_time)?.toString()
            }
          );
        }
      },
      // time opened
      {
        title: t("positions.history.column.timeOpened"),
        dataIndex: "open_timestamp",
        width: 175,
        onSort: true,
        render: (_, record) => /* @__PURE__ */ jsxRuntime.jsx(ui.Text.formatted, { rule: "date", formatString: "yyyy-MM-dd HH:mm:ss", children: record.open_timestamp })
      },
      // time close
      {
        title: t("positions.history.column.timeClosed"),
        dataIndex: "close_timestamp",
        width: 175,
        onSort: true,
        render: (_, record) => {
          if (record.position_status == "closed" && record.close_timestamp) {
            return /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.formatted,
              {
                rule: "date",
                formatString: "yyyy-MM-dd HH:mm:ss",
                children: record.close_timestamp ?? "--"
              }
            );
          }
          return "--";
        }
      },
      // updated time
      {
        title: t("positions.history.column.updatedTime"),
        dataIndex: "last_update_time",
        width: 175,
        onSort: true,
        render: (_, record) => /* @__PURE__ */ jsxRuntime.jsx(ui.Text.formatted, { rule: "date", formatString: "yyyy-MM-dd HH:mm:ss", children: record.last_update_time })
      }
    ],
    [pnlNotionalDecimalPrecision, t]
  );
  return column;
};
var SymbolInfo = (props) => {
  const { record, onSymbolChange } = props;
  const { t } = i18n.useTranslation();
  const tags = React2.useMemo(() => {
    const list = [];
    const status = record.position_status;
    const renderStatus = () => {
      if (status === "closed") {
        return t("positions.history.status.closed");
      } else if (status === "partial_closed") {
        return t("positions.history.status.partialClosed");
      } else {
        return ui.capitalizeFirstLetter(status.replace("_", " "));
      }
    };
    list.push(
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Badge,
        {
          color: status !== "closed" ? "primaryLight" : "neutral",
          size: "xs",
          className: "oui-whitespace-nowrap oui-break-normal",
          children: renderStatus()
        },
        `status-${status}`
      )
    );
    if (record.type === "adl") {
      list.push(
        /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "danger", size: "xs", children: t("positions.history.type.adl") }, `type-${record.type}`)
      );
    } else if (record.type === "liquidated") {
      list.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Tooltip,
          {
            className: "oui-min-w-[204px] oui-bg-base-5",
            arrow: { className: "oui-fill-base-5" },
            content: /* @__PURE__ */ jsxRuntime.jsxs(
              ui.Flex,
              {
                direction: "column",
                itemAlign: "start",
                className: "oui-text-2xs",
                children: [
                  record.liquidation_id != null && /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("positions.history.liquidated.liquidationId") }),
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 98, children: record.liquidation_id })
                  ] }),
                  /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("positions.history.liquidated.liquidatorFee") }),
                    /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { color: "lose", children: [
                      record.liquidator_fee > 0 && "-",
                      utils.commifyOptional(record.liquidator_fee)
                    ] })
                  ] }),
                  /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
                    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("positions.history.liquidated.insFundFee") }),
                    /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { color: "lose", children: [
                      record.insurance_fund_fee > 0 && "-",
                      utils.commifyOptional(record.insurance_fund_fee)
                    ] })
                  ] })
                ]
              }
            ),
            children: /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { size: "xs", color: "danger", className: "oui-cursor-pointer", children: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-underline oui-decoration-dashed oui-decoration-1", children: t("positions.history.type.liquidated") }) }) })
          },
          `type-${record.type}`
        )
      );
    }
    list.push(
      /* @__PURE__ */ jsxRuntime.jsx(
        LeverageBadge2,
        {
          symbol: record.symbol,
          leverage: record.leverage
        },
        `leverage-${record.symbol}`
      )
    );
    return list;
  }, [record, t]);
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, height: 48, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Box,
      {
        width: 4,
        height: 38,
        className: ui.cn(
          "oui-shrink-0 oui-rounded-[1px]",
          record.side === "LONG" ? "oui-bg-trade-profit" : "oui-bg-trade-loss"
        )
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: "start", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.formatted,
        {
          formatString: "base-type",
          className: "oui-cursor-pointer",
          onClick: (e) => {
            onSymbolChange?.({ symbol: record.symbol });
            e.stopPropagation();
            e.preventDefault();
          },
          children: `${record.symbol.split("_")[1]}-PERP`
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { gap: 1, children: tags })
    ] })
  ] });
};
var Quantity = (props) => {
  const { record } = props;
  const { base_dp } = useSymbolContext();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      gap: 1,
      direction: "column",
      itemAlign: "start",
      className: "oui-truncate",
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: base_dp, padding: false, children: Math.abs(record.closed_position_qty) }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: base_dp, padding: false, className: "oui-truncate", children: Math.abs(record.max_position_qty) })
      ]
    }
  );
};
var NetPnL = (props) => {
  const { record, pnlNotionalDecimalPrecision } = props;
  const { t } = i18n.useTranslation();
  const netPnl = record.netPnL != null ? record.netPnL : void 0;
  const text = () => /* @__PURE__ */ jsxRuntime.jsx(
    ui.Text.pnl,
    {
      dp: pnlNotionalDecimalPrecision,
      color: record.netPnL != null ? record.netPnL > 0 ? "profit" : "lose" : void 0,
      className: netPnl == null ? "" : "oui-cursor-pointer oui-border-b oui-border-dashed oui-border-line-12",
      children: netPnl ?? "--"
    }
  );
  if (record.netPnL == null) return text();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Tooltip,
    {
      delayDuration: 200,
      content: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: "start", className: "oui-text-2xs", children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 80, children: t("positions.history.column.netPnl") }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("common.realizedPnl") }),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              color: record.realized_pnl >= 0 ? "profit" : "lose",
              className: "oui-cursor-pointer",
              children: utils.commifyOptional(utils.formatNum.pnl(record.realized_pnl)?.toString())
            }
          )
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("funding.fundingFee") }),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              color: record.accumulated_funding_fee > 0 ? "lose" : "profit",
              className: "oui-cursor-pointer",
              children: utils.commifyOptional(-record.accumulated_funding_fee)
            }
          )
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("positions.history.netPnl.tradingFee") }),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              color: record.trading_fee > 0 ? "lose" : "profit",
              className: "oui-cursor-pointer",
              children: utils.commifyOptional(-record.trading_fee)
            }
          )
        ] })
      ] }),
      className: "oui-min-w-[204px] oui-bg-base-5",
      arrow: {
        className: "oui-fill-base-5"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx("div", { children: text() })
    }
  );
};
var LeverageBadge2 = (props) => {
  const { symbol, leverage } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "neutral", size: "xs", children: leverage ? /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 0, size: "2xs", unit: "X", children: leverage }) : /* @__PURE__ */ jsxRuntime.jsx(LeverageDisplay2, { symbol }) });
};
var LeverageDisplay2 = ({ symbol }) => {
  const leverage = hooks.useMaxLeverage(symbol);
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: 0, size: "2xs", unit: "X", children: leverage });
};
function offsetStartOfDay(date) {
  if (date == null) return date;
  const newDate = new Date(date);
  newDate.setHours(0, 0, 0, 0);
  return newDate;
}
function offsetEndOfDay(date) {
  if (date == null) return date;
  const newDate = new Date(date);
  newDate.setHours(23, 59, 59, 999);
  return newDate;
}
var formatDatePickerRange = (option) => ({
  from: offsetStartOfDay(option.from),
  to: offsetEndOfDay(option.to ?? option.from)
});
var areDatesEqual = (date1, date2) => {
  return date1.getTime() === date2.getTime();
};

// src/components/positionHistory/positionHistory.script.tsx
var usePositionHistoryScript = (props) => {
  const {
    onSymbolChange,
    symbol,
    pnlNotionalDecimalPrecision,
    enableSortingStorage = true
  } = props;
  const { state } = hooks.useAccount();
  const { tabSort, onTabSort } = useTabSort({
    storageKey: TRADING_POSITIONS_SORT_STORAGE_KEY
  });
  const { onSort, getSortedList } = useSort(
    enableSortingStorage ? tabSort?.["positionHistory" /* PositionHistory */] : void 0,
    enableSortingStorage ? onTabSort("positionHistory" /* PositionHistory */) : void 0
  );
  const { data, isLoading } = hooks.usePrivateQuery(
    symbol ? `/v1/position_history?symbol=${symbol}&limit=1000` : "/v1/position_history?limit=1000",
    {
      formatter(data2) {
        return (data2.rows ?? null)?.map(
          (item) => {
            if (item.realized_pnl != null && item.accumulated_funding_fee != null && item.trading_fee != null) {
              const netPnL = item.realized_pnl - item.accumulated_funding_fee - item.trading_fee;
              return {
                ...item,
                netPnL
              };
            }
            return item;
          }
        );
      },
      revalidateOnFocus: true
    }
  );
  const { pagination, setPage } = ui.usePagination({
    pageSize: 10
  });
  const {
    status,
    side,
    dateRange,
    filterDays,
    updateFilterDays,
    filterItems,
    onFilter
  } = useFilter();
  React2.useEffect(() => {
    setPage(1);
  }, [status, side, dateRange, filterDays, symbol]);
  const filterData = React2.useMemo(() => {
    if (data == null) return data;
    return data.filter((item) => {
      const sideFilter = side === "all" /* all */ ? true : item.side.toLowerCase() === side.toLowerCase();
      const statusFilter = status === "all" /* all */ ? true : item.position_status.toLowerCase() === status.toLowerCase();
      const time = item?.last_update_time ?? item.open_timestamp;
      const dateFilter = dateRange.from && dateRange.to ? time >= dateRange.from.getTime() && time <= dateRange.to.getTime() : true;
      const symbolFilter = symbol ? item.symbol == symbol : true;
      return sideFilter && statusFilter && dateFilter && symbolFilter;
    });
  }, [status, side, dateRange, data, symbol]);
  const rawDataSource = reactApp.useDataTap(filterData, {
    accountStatus: state.status === types.AccountStatusEnum.EnableTradingWithoutConnected ? types.AccountStatusEnum.EnableTradingWithoutConnected : types.AccountStatusEnum.EnableTrading,
    fallbackData: []
  });
  const dataSource = getSortedList(rawDataSource || []);
  return {
    dataSource,
    isLoading,
    onSymbolChange,
    pagination,
    filterItems,
    onFilter,
    symbol,
    filterDays,
    updateFilterDays,
    pnlNotionalDecimalPrecision,
    onSort,
    initialSort: enableSortingStorage ? tabSort?.["positionHistory" /* PositionHistory */] : void 0
  };
};
var useFilter = () => {
  const { t } = i18n.useTranslation();
  const [status, setStatus] = React2.useState(
    "all" /* all */
  );
  const [side, setSide] = React2.useState(
    "all" /* all */
  );
  const defaultRange = formatDatePickerRange({
    to: offsetEndOfDay(/* @__PURE__ */ new Date()),
    from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 89))
  });
  const [filterDays, setFilterDays] = React2.useState(90);
  const [dateRange, setDateRange] = React2.useState(defaultRange);
  const updateFilterDays = (days) => {
    setFilterDays(days);
    setDateRange({
      from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), days - 1)),
      to: offsetEndOfDay(/* @__PURE__ */ new Date())
    });
  };
  const onFilter = (filter) => {
    if (filter.name === "side") {
      setSide(filter.value);
    }
    if (filter.name === "status") {
      setStatus(filter.value);
    }
    if (filter.name === "dateRange") {
      const newDateRange = formatDatePickerRange(filter.value);
      setDateRange(newDateRange);
      if (newDateRange.from && newDateRange.to) {
        const diffDays = Math.abs(dateFns.differenceInDays(newDateRange.from, newDateRange.to)) + 1;
        const dateRangeMap = {
          1: {
            from: offsetStartOfDay(/* @__PURE__ */ new Date()),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          },
          7: {
            from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 6)),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          },
          30: {
            from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 29)),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          },
          90: {
            from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 89)),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          }
        };
        const dateRange2 = dateRangeMap[diffDays];
        if (dateRange2 && areDatesEqual(dateRange2.from, newDateRange.from) && areDatesEqual(dateRange2.to, newDateRange.to)) {
          setFilterDays(diffDays);
        } else {
          setFilterDays(null);
        }
      }
    }
  };
  const { isMobile } = ui.useScreen();
  const filterItems = React2.useMemo(() => {
    const sideFilter = {
      type: "select",
      name: "side",
      options: [
        {
          label: t("common.side.all"),
          value: "all"
        },
        {
          label: t("common.buy"),
          value: "LONG" /* buy */
        },
        {
          label: t("common.sell"),
          value: "SHORT" /* sell */
        }
      ],
      value: side
    };
    const dateRangeFilter = {
      type: "range",
      name: "dateRange",
      value: dateRange,
      fromDate: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 89)),
      toDate: offsetEndOfDay(/* @__PURE__ */ new Date())
    };
    const statusFilter = {
      type: "select",
      name: "status",
      options: [
        {
          label: t("common.status.all"),
          value: "all"
        },
        {
          label: t("positions.history.status.closed"),
          value: "closed" /* closed */
        },
        {
          label: t("positions.history.status.partialClosed"),
          value: "partial_closed" /* partial_closed */
        }
      ],
      value: status
    };
    if (isMobile) {
      return [sideFilter, statusFilter];
    }
    return [sideFilter, statusFilter, dateRangeFilter];
  }, [side, status, dateRange]);
  return {
    filterItems,
    onFilter,
    side,
    dateRange,
    status,
    filterDays,
    updateFilterDays
  };
};
var SymbolToken2 = (props) => {
  const { side, symbol } = props.item;
  const { t } = i18n.useTranslation();
  const isBuy = side === "LONG" /* buy */;
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Text.formatted,
    {
      intensity: 80,
      rule: "symbol",
      formatString: "base-type",
      size: "sm",
      prefix: /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: isBuy ? "success" : "danger", size: "xs", children: isBuy ? t("common.buy") : t("common.sell") }),
      onClick: () => {
        props.onSymbolChange?.({ symbol });
      },
      children: symbol
    }
  );
};
var PositionHistoryType = (props) => {
  const { item: record } = props;
  const { t } = i18n.useTranslation();
  const showAlert = () => {
    ui.modal.alert({
      title: t("positions.liquidation"),
      message: /* @__PURE__ */ jsxRuntime.jsxs(
        ui.Flex,
        {
          direction: "column",
          width: "100%",
          gap: 2,
          className: "oui-text-2xs oui-text-base-contrast-54",
          children: [
            record.liquidation_id != null && /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.liquidated.liquidationId") }),
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 98, children: record.liquidation_id })
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.liquidated.liquidatorFee") }),
              /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { color: "lose", children: [
                record.liquidator_fee > 0 && "-",
                utils.commifyOptional(record.liquidator_fee)
              ] })
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.liquidated.insFundFee") }),
              /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { color: "lose", children: [
                record.insurance_fund_fee > 0 && "-",
                utils.commifyOptional(record.insurance_fund_fee)
              ] })
            ] })
          ]
        }
      )
    });
  };
  const tags = React2.useMemo(() => {
    const list = [];
    const status = record.position_status;
    const renderStatus = () => {
      if (status === "closed") {
        return t("positions.history.status.closed");
      } else if (status === "partial_closed") {
        return t("positions.history.status.partialClosed");
      } else {
        return ui.capitalizeFirstLetter(status.replace("_", " "));
      }
    };
    list.push(
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Badge,
        {
          color: status !== "closed" ? "primaryLight" : "neutral",
          size: "xs",
          children: renderStatus()
        },
        1
      )
    );
    if (record.type === "adl") {
      list.push(
        /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { color: "danger", size: "xs", children: t("positions.history.type.adl") }, 2)
      );
    } else if (record.type === "liquidated") {
      list.push(
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Badge,
          {
            size: "xs",
            color: "danger",
            className: "oui-cursor-pointer",
            onClick: showAlert,
            children: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-underline oui-decoration-dashed oui-decoration-1", children: t("positions.history.type.liquidated") })
          },
          3
        )
      );
    }
    return list;
  }, [record]);
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { gap: 1, children: tags });
};
var ClosedQty = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const closedQty = item.closed_position_qty != null ? Math.abs(item.closed_position_qty) : "--";
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.column.closed") }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.base_dp, padding: false, coloring: true, intensity: 80, children: closedQty })
    }
  );
};
var MaxClosedQty = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const maxClosedQty = item.max_position_qty != null ? Math.abs(item.max_position_qty) : "--";
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.column.maxClosed") }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.base_dp, padding: false, coloring: true, intensity: 80, children: maxClosedQty })
    }
  );
};
var AvgOpen2 = (props) => {
  const { item } = props;
  const avgOpen = item.avg_open_price != null ? Math.abs(item.avg_open_price) : "--";
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, children: [
        t("common.avgOpen"),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 20, children: "(USDC)" })
      ] }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.quote_dp, padding: false, coloring: true, intensity: 80, children: avgOpen })
    }
  );
};
var AvgClosed = (props) => {
  const { item } = props;
  const avgClose = item.avg_close_price != null ? Math.abs(item.avg_close_price) : "--";
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, children: [
        t("common.avgClose"),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 20, children: "(USDC)" })
      ] }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { dp: props.quote_dp, padding: false, coloring: true, intensity: 80, children: avgClose })
    }
  );
};
var OpenTime = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.column.timeOpened") }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      align: "end",
      children: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.formatted,
        {
          intensity: 80,
          formatString: "yyyy-MM-dd HH:mm:ss",
          rule: "date",
          children: item.open_timestamp
        }
      )
    }
  );
};
var ClosedTime = (props) => {
  const { item } = props;
  const { t } = i18n.useTranslation();
  const child = item.position_status == "closed" && item.close_timestamp ? /* @__PURE__ */ jsxRuntime.jsx(
    ui.Text.formatted,
    {
      intensity: 80,
      formatString: "yyyy-MM-dd HH:mm:ss",
      rule: "date",
      children: item.close_timestamp
    }
  ) : "--";
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Statistic,
    {
      label: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: t("positions.history.column.timeClosed") }),
      classNames: {
        root: "oui-text-xs",
        label: "oui-text-2xs"
      },
      align: "end",
      children: child
    }
  );
};
var FundingFee2 = (props) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      justify: "end",
      className: "oui-text-2xs oui-w-full oui-py-2 oui-gap-1 oui-gap-1",
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { intensity: 36, children: [
          t("funding.fundingFee"),
          ": "
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(
          FundingFeeButton,
          {
            fee: -props.item.accumulated_funding_fee,
            symbol: props.item.symbol,
            start_t: props.item.open_timestamp.toString(),
            end_t: (props.item.close_timestamp || props.item.last_update_time)?.toString()
          }
        )
      ]
    }
  );
};
var PositionHistoryCell = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      width: "100%",
      gap: 2,
      itemAlign: "start",
      className: props.classNames?.root,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(Header, { ...props }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, { intensity: 6, className: "oui-w-full" }),
        /* @__PURE__ */ jsxRuntime.jsx(Body, { ...props })
      ]
    }
  );
};
var Header = (props) => {
  const { t } = i18n.useTranslation();
  const showAlert = () => {
    const { item: record } = props;
    ui.modal.alert({
      title: t("positions.history.column.netPnl"),
      message: /* @__PURE__ */ jsxRuntime.jsxs(
        ui.Flex,
        {
          direction: "column",
          width: "100%",
          gap: 2,
          className: "oui-text-2xs oui-text-base-contrast-54",
          children: [
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("common.realizedPnl") }),
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: record.realized_pnl >= 0 ? "profit" : "lose", children: utils.commifyOptional(record.realized_pnl) })
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("funding.fundingFee") }),
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text,
                {
                  color: record.accumulated_funding_fee > 0 ? "lose" : "profit",
                  children: utils.commifyOptional(-record.accumulated_funding_fee)
                }
              )
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "between", width: "100%", gap: 2, children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { intensity: 54, children: t("positions.history.netPnl.tradingFee") }),
              /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: record.trading_fee > 0 ? "lose" : "profit", children: utils.commifyOptional(-record.trading_fee) })
            ] })
          ]
        }
      )
    });
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 1, width: "100%", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Flex,
      {
        direction: "column",
        itemAlign: "start",
        className: "oui-flex-1",
        gap: 1,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(SymbolToken2, { ...props }),
          /* @__PURE__ */ jsxRuntime.jsx(PositionHistoryType, { ...props })
        ]
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: "end", className: "oui-flex-1", children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 36, children: t("positions.history.column.netPnl") }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gapX: 1, children: [
        /* @__PURE__ */ jsxRuntime.jsx("button", { onClick: showAlert, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { size: "xs", coloring: true, children: props.item.netPnL ?? "--" }) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ShareButtonWidget,
          {
            position: props.item,
            sharePnLConfig: props.sharePnLConfig,
            modalId: uiShare.SharePnLBottomSheetId,
            iconSize: 12,
            isPositionHistory: true
          }
        )
      ] })
    ] })
  ] });
};
var Body = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-w-full", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(
      ui.Grid,
      {
        cols: 3,
        rows: 2,
        width: "100%",
        gap: 0,
        className: "oui-grid-cols-[1fr,1fr,135px]",
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(ClosedQty, { ...props }),
          /* @__PURE__ */ jsxRuntime.jsx(MaxClosedQty, { ...props }),
          /* @__PURE__ */ jsxRuntime.jsx(OpenTime, { ...props }),
          /* @__PURE__ */ jsxRuntime.jsx(AvgOpen2, { ...props }),
          /* @__PURE__ */ jsxRuntime.jsx(AvgClosed, { ...props }),
          /* @__PURE__ */ jsxRuntime.jsx(ClosedTime, { ...props })
        ]
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(FundingFee2, { ...props })
  ] });
};

// src/components/positionHistory/mobile/positionHistoryCell.script.tsx
var usePositionHistoryCellScript = (props) => {
  const symbolInfo = useSymbolContext();
  return {
    ...props,
    ...symbolInfo
  };
};
var PositionHistoryCellWidget = (props) => {
  const state = usePositionHistoryCellScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(PositionHistoryCell, { ...state });
};
var PositionHistory = (props) => {
  const { onSymbolChange, pagination, pnlNotionalDecimalPrecision } = props;
  const column = usePositionHistoryColumn({
    onSymbolChange,
    pnlNotionalDecimalPrecision,
    sharePnLConfig: props.sharePnLConfig
  });
  const { t } = i18n.useTranslation();
  const dayLabel = React2.useMemo(() => {
    return {
      1: t("common.select.1d"),
      7: t("common.select.7d"),
      30: t("common.select.30d"),
      90: t("common.select.90d")
    };
  }, [t]);
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", width: "100%", height: "100%", itemAlign: "start", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 3, children: [
      props.filterItems.length > 0 && /* @__PURE__ */ jsxRuntime.jsx(
        ui.DataFilter,
        {
          items: props.filterItems,
          onFilter: (value) => {
            props.onFilter(value);
          }
        }
      ),
      [1, 7, 30, 90].map((value) => {
        return /* @__PURE__ */ jsxRuntime.jsxs(
          "button",
          {
            className: "oui-relative oui-px-2 oui-py-[2px] oui-text-sm",
            children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.gradient,
                {
                  color: props.filterDays === value ? "brand" : void 0,
                  className: ui.cn(
                    "oui-whitespace-nowrap oui-break-normal",
                    props.filterDays !== value ? "oui-text-base-contrast-54" : ""
                  ),
                  children: dayLabel[value] || `${value}D`
                }
              ) }),
              /* @__PURE__ */ jsxRuntime.jsx(
                "div",
                {
                  className: "oui-absolute oui-inset-0 oui-rounded oui-opacity-[.12] oui-gradient-primary",
                  onClick: () => {
                    props.updateFilterDays(value);
                  }
                }
              )
            ]
          },
          `item-${value}`
        );
      })
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(
      uiConnector.AuthGuardDataTable,
      {
        loading: props.isLoading,
        id: "oui-desktop-positions-content",
        columns: column,
        bordered: true,
        dataSource: props.dataSource,
        generatedRowKey: (record) => `${record.symbol}_${record.position_id}`,
        renderRowContainer: (record, index, children) => /* @__PURE__ */ jsxRuntime.jsx(SymbolProvider, { symbol: record.symbol, children }),
        manualPagination: false,
        pagination,
        manualSorting: true,
        onSort: props.onSort,
        initialSort: props.initialSort ? {
          sortKey: props.initialSort.sortKey,
          sort: props.initialSort.sortOrder
        } : void 0,
        testIds: {
          body: "oui-testid-dataList-positionHistory-tab-body"
        },
        classNames: {
          root: "!oui-h-[calc(100%_-_49px)]"
        }
      }
    )
  ] });
};
var MobilePositionHistory = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Grid,
    {
      cols: 1,
      rows: 2,
      className: "oui-w-full oui-grid-rows-[auto,1fr]",
      gap: 1,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { gap: 2, p: 2, className: "oui-rounded-b-xl oui-bg-base-9", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.DataFilter,
          {
            className: "oui-border-none oui-py-0",
            items: props.filterItems,
            onFilter: (value) => {
              props.onFilter(value);
            }
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.ListView,
          {
            className: ui.cn(
              "oui-hide-scrollbar oui-w-full oui-space-y-0 oui-overflow-y-hidden",
              props.classNames?.root
            ),
            contentClassName: ui.cn("!oui-space-y-1", props.classNames?.content),
            dataSource: props.dataSource,
            renderItem: (item, index) => /* @__PURE__ */ jsxRuntime.jsx(SymbolProvider, { symbol: item.symbol, children: /* @__PURE__ */ jsxRuntime.jsx(
              PositionHistoryCellWidget,
              {
                item,
                index,
                onSymbolChange: props.onSymbolChange,
                classNames: {
                  root: props.classNames?.cell
                },
                sharePnLConfig: props.sharePnLConfig
              }
            ) })
          }
        )
      ]
    }
  );
};
var PositionHistoryWidget = (props) => {
  const state = usePositionHistoryScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(PositionHistory, { ...state, sharePnLConfig: props.sharePnLConfig });
};
var MobilePositionHistoryWidget = (props) => {
  const { classNames, ...rest } = props;
  const state = usePositionHistoryScript(rest);
  return /* @__PURE__ */ jsxRuntime.jsx(
    MobilePositionHistory,
    {
      ...state,
      classNames,
      sharePnLConfig: props.sharePnLConfig
    }
  );
};
var TooltipButton = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Tooltip,
    {
      className: "oui-w-[275px] oui-bg-base-6",
      content: props.tooltip,
      arrow: {
        className: "oui-fill-base-6"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx("button", { className: "oui-border-b oui-border-dashed oui-border-line-12", children: props.label })
    }
  );
};
var useLiquidationColumn = (props) => {
  const { t } = i18n.useTranslation();
  const column = React2.useMemo(
    () => [
      // Time
      {
        title: t("common.time"),
        dataIndex: "timestamp",
        width: 202,
        render: (value) => /* @__PURE__ */ jsxRuntime.jsx(ui.Text.formatted, { rule: "date", formatString: "yyyy-MM-dd HH:mm:ss", children: value })
      },
      // Liquidation id
      {
        title: t("positions.Liquidation.column.liquidationId"),
        dataIndex: "liquidation_id",
        width: 120,
        render: (value) => /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: value })
      },
      // Symbol
      {
        title: t("common.symbol"),
        dataIndex: "Symbol",
        // width: 202,
        render: (_, record) => /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { direction: "column", itemAlign: "start", children: record.positions_by_perp?.map((item) => /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.formatted,
          {
            rule: "symbol",
            formatString: "base-quote",
            children: item.symbol
          },
          item.symbol
        )) })
      },
      // Price (USDC)
      {
        title: t("positions.Liquidation.column.markPrice"),
        dataIndex: "Price_(USDC)",
        // width: 202,
        render: (_, record) => {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { direction: "column", itemAlign: "start", children: record.positions_by_perp?.map((item) => (
            // <SymbolProvider symbol={item.symbol}>
            // </SymbolProvider>
            /* @__PURE__ */ jsxRuntime.jsx(
              FormattedText,
              {
                value: item.mark_price,
                type: "quote"
              },
              item.symbol
            )
          )) });
        }
      },
      // Quantity
      {
        title: t("common.quantity"),
        dataIndex: "Quantity",
        // width: 202,
        render: (_, record) => {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { direction: "column", itemAlign: "start", children: record.positions_by_perp?.map((item) => (
            // <SymbolProvider symbol={item.symbol}>
            // </SymbolProvider>
            /* @__PURE__ */ jsxRuntime.jsx(
              FormattedText,
              {
                value: item.position_qty,
                type: "base"
              },
              item.symbol
            )
          )) });
        }
      },
      // net pnl
      // {
      //   title: t("positions.Liquidation.column.insFundTransfer"),
      //   dataIndex: "transfer_amount_to_insurance_fund",
      //   width: 202,
      //   render: (value) => {
      //     return <Text>{commifyOptional(value)}</Text>;
      //   },
      // },
      // Liquidation fee rate
      // TODO: use i18
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(
          TooltipButton,
          {
            tooltip: t("positions.Liquidation.col.tooltip.feeRate"),
            label: t("positions.Liquidation.column.liquidationFeeRate")
          }
        ),
        dataIndex: "liquidationFeeRate",
        // width: 202,
        render: (value) => {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { rule: "percentages", children: value });
        }
      },
      // Liquidation Fee
      {
        title: /* @__PURE__ */ jsxRuntime.jsx(
          TooltipButton,
          {
            tooltip: t("positions.Liquidation.col.tooltip.fee"),
            label: t("positions.Liquidation.column.liquidationFee")
          }
        ),
        dataIndex: "abs_liquidation_fee",
        width: 100,
        render: (abs_liquidation_fee, record) => {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { direction: "column", itemAlign: "start", children: record.positions_by_perp?.map((item) => (
            // <SymbolProvider symbol={item.symbol}>
            // </SymbolProvider>
            /* @__PURE__ */ jsxRuntime.jsx(
              FormattedText,
              {
                value: item.abs_liquidation_fee,
                type: "quote"
              },
              item.symbol
            )
          )) });
        }
      },
      {
        title: "",
        dataIndex: "actions",
        align: "right",
        width: 40,
        render: (_, record, index, ctx) => {
          const isExpanded = ctx.row.getIsExpanded();
          return /* @__PURE__ */ jsxRuntime.jsx(
            "button",
            {
              className: "oui-p-2",
              onClick: () => {
                ctx.row.getToggleExpandedHandler()();
              },
              children: /* @__PURE__ */ jsxRuntime.jsx(ArrowIcon, { className: isExpanded ? "oui-rotate-180" : "" })
            }
          );
        }
      }
    ],
    [t]
  );
  return column;
};
var FormattedText = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: utils.commifyOptional(props.value) });
};
var ArrowIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    "svg",
    {
      width: "16",
      height: "16",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      className: props.className,
      children: /* @__PURE__ */ jsxRuntime.jsx(
        "path",
        {
          d: "M3.884 6.02a.67.67 0 0 0-.436.27.68.68 0 0 0 .187.933L7.63 9.88a.68.68 0 0 0 .749 0l3.994-2.657a.68.68 0 0 0 .187-.934.68.68 0 0 0-.936-.186L8.003 8.51l-3.62-2.407a.65.65 0 0 0-.499-.084",
          fill: "#fff",
          fillOpacity: ".54"
        }
      )
    }
  );
};
var LiquidationCell = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Flex,
    {
      direction: "column",
      width: "100%",
      gap: 2,
      itemAlign: "start",
      className: ui.cn(
        "oui-rounded-xl oui-bg-base-9 oui-p-2",
        props.classNames?.root
      ),
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(Header2, { ...props }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Divider, {}),
        /* @__PURE__ */ jsxRuntime.jsx(Body2, { ...props })
      ]
    },
    props.item.timestamp
  );
};
var Header2 = (props) => {
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { width: "100%", justify: "between", itemAlign: "start", children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { direction: "column", itemAlign: "start", gap: 1, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { gap: 2, itemAlign: "center", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { direction: "column", gap: 1, itemAlign: "start", children: /* @__PURE__ */ jsxRuntime.jsx(
      ui.Text.formatted,
      {
        rule: "symbol",
        formatString: "base-quote",
        size: "xs",
        intensity: 80,
        children: props.item.positions_by_perp?.[0]?.symbol || ""
      }
    ) }) }) }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: "end", children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 36, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.formatted, { rule: "date", formatString: "yyyy-MM-dd HH:mm:ss", children: props.item.timestamp }) }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { size: "2xs", intensity: 36, children: [
        `${t("positions.Liquidation.column.liquidationId")}: `,
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { as: "span", intensity: 80, children: props.item.liquidation_id })
      ] })
    ] })
  ] });
};
var Body2 = (props) => {
  const position = props.item.positions_by_perp?.[0];
  const { t } = i18n.useTranslation();
  const mr = React2.useMemo(() => {
    if (isNaN(props.item?.margin_ratio)) return "--";
    return `${new utils.Decimal(props.item.margin_ratio).mul(100).todp(2, utils.Decimal.ROUND_DOWN).toNumber()}%`;
  }, [props.item?.margin_ratio]);
  if (!position) return null;
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", width: "100%", gap: 1, children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, width: "100%", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.column.markPrice"),
          value: utils.commifyOptional(position.transfer_price),
          className: "oui-flex-1"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("common.quantity"),
          value: utils.commifyOptional(position.position_qty),
          className: "oui-flex-1"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.column.liquidationFeeRate"),
          value: props.item.liquidationFeeRate,
          className: "oui-flex-1",
          align: "end"
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, width: "100%", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.column.liquidationFee"),
          value: utils.commifyOptional(position.abs_liquidation_fee),
          className: "oui-flex-1"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.expand.label.mr"),
          value: mr,
          className: "oui-flex-1"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.expand.label.mmr"),
          value: props.item.formatted_account_mmr ? `${props.item.formatted_account_mmr}%` : "--",
          className: "oui-flex-1",
          align: "end"
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 2, width: "100%", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.expand.label.collateral"),
          value: utils.commifyOptional(props.item.collateral_value),
          className: "oui-flex-1"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        DataItem,
        {
          label: t("positions.Liquidation.expand.label.notional"),
          value: utils.commifyOptional(props.item.position_notional),
          className: "oui-flex-1"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex-1 oui-opacity-0" }),
      " "
    ] })
  ] });
};
var DataItem = ({ label, value, className, align = "start" }) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", itemAlign: align, className, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 36, children: label }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "xs", intensity: 80, children: value })
  ] });
};

// src/components/liquidation/mobile/liquidationCell.script.tsx
var useLiquidationCellScript = (props) => {
  return {
    ...props
  };
};
var LiquidationCellWidget = (props) => {
  const { classNames, ...rest } = props;
  const state = useLiquidationCellScript(rest);
  return /* @__PURE__ */ jsxRuntime.jsx(LiquidationCell, { classNames, ...state });
};
var Liquidation = (props) => {
  const [expanded, setExpanded] = React2.useState({});
  const column = useLiquidationColumn();
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", width: "100%", height: "100%", itemAlign: "start", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { gap: 3, children: [
      props.filterItems.length > 0 && /* @__PURE__ */ jsxRuntime.jsx(
        ui.DataFilter,
        {
          items: props.filterItems,
          onFilter: (value) => {
            props.onFilter(value);
          }
        }
      ),
      [1, 7, 30, 90].map((value) => {
        return /* @__PURE__ */ jsxRuntime.jsxs(
          "button",
          {
            className: "oui-relative oui-px-2 oui-py-[2px] oui-text-sm",
            children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.gradient,
                {
                  color: props.filterDays === value ? "brand" : void 0,
                  className: props.filterDays !== value ? "oui-text-base-contrast-54" : "",
                  children: `${value}D`
                }
              ) }),
              /* @__PURE__ */ jsxRuntime.jsx(
                "div",
                {
                  className: "oui-absolute oui-inset-0 oui-rounded oui-opacity-[.12] oui-gradient-primary",
                  onClick: () => {
                    props.updateFilterDays(value);
                  }
                }
              )
            ]
          },
          `item-${value}`
        );
      })
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(
      uiConnector.AuthGuardDataTable,
      {
        loading: props.isLoading,
        id: "oui-desktop-liquidation-content",
        columns: column,
        bordered: true,
        dataSource: props.dataSource,
        generatedRowKey: (record) => `${record.liquidation_id}`,
        pagination: props.pagination,
        testIds: {
          body: "oui-testid-dataList-liquidation-tab-body"
        },
        classNames: {
          root: "!oui-h-[calc(100%_-_49px)]"
        },
        expanded,
        onExpandedChange: setExpanded,
        getRowCanExpand: () => true,
        expandRowRender: (record) => {
          return /* @__PURE__ */ jsxRuntime.jsx(LiquidationDetails, { record: record.original });
        }
      }
    )
  ] });
};
var TooltipHeaderCell = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx("th", { className: "oui-h-10 oui-border-b oui-border-line oui-text-left oui-text-base-contrast-36", children: /* @__PURE__ */ jsxRuntime.jsx(
    ui.Tooltip,
    {
      content: props.tooltip,
      className: "oui-w-[275px] oui-bg-base-6",
      side: props.side,
      arrow: {
        className: "oui-fill-base-6"
      },
      children: /* @__PURE__ */ jsxRuntime.jsx("button", { className: "oui-border-b oui-border-dashed oui-border-line-12", children: props.label })
    }
  ) });
};
var LiquidationDetails = (props) => {
  const { margin_ratio, account_mmr, collateral_value, position_notional } = props.record;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-w-full oui-bg-base-8 oui-px-6", children: /* @__PURE__ */ jsxRuntime.jsxs("table", { className: "oui-w-full", children: [
    /* @__PURE__ */ jsxRuntime.jsx("thead", { children: /* @__PURE__ */ jsxRuntime.jsxs("tr", { children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        TooltipHeaderCell,
        {
          tooltip: t("positions.Liquidation.expand.tooltip.mr"),
          label: t("positions.Liquidation.expand.label.mr"),
          side: "left"
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        TooltipHeaderCell,
        {
          tooltip: t("positions.Liquidation.expand.tooltip.mmr"),
          label: t("positions.Liquidation.expand.label.mmr")
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        TooltipHeaderCell,
        {
          tooltip: t("positions.Liquidation.expand.tooltip.collateral"),
          label: t("positions.Liquidation.expand.label.collateral")
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        TooltipHeaderCell,
        {
          tooltip: t("positions.Liquidation.expand.tooltip.notional"),
          label: t("positions.Liquidation.expand.label.notional")
        }
      )
    ] }) }),
    /* @__PURE__ */ jsxRuntime.jsx("tbody", { children: /* @__PURE__ */ jsxRuntime.jsxs("tr", { children: [
      /* @__PURE__ */ jsxRuntime.jsx("td", { className: "oui-h-10", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { rule: "percentages", children: margin_ratio }) }),
      /* @__PURE__ */ jsxRuntime.jsx("td", { className: "oui-h-10", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { rule: "percentages", children: account_mmr }) }),
      /* @__PURE__ */ jsxRuntime.jsx("td", { className: "oui-h-10", children: utils.commifyOptional(collateral_value) }),
      /* @__PURE__ */ jsxRuntime.jsx("td", { className: "oui-h-10", children: utils.commifyOptional(position_notional) })
    ] }) })
  ] }) });
};
var MobileLiquidation = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ui.Grid,
    {
      cols: 1,
      rows: 2,
      className: "oui-w-full oui-grid-rows-[auto,1fr]",
      gap: 1,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { gap: 2, p: 2, className: "oui-rounded-b-xl oui-bg-base-9", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.DataFilter,
          {
            className: "oui-border-none oui-py-0",
            items: props.filterItems,
            onFilter: (value) => {
              props.onFilter(value);
            }
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.ListView,
          {
            className: ui.cn(
              "oui-hide-scrollbar oui-w-full oui-space-y-0 oui-overflow-y-hidden",
              props.classNames?.root
            ),
            contentClassName: ui.cn("!oui-space-y-1", props.classNames?.content),
            dataSource: props.dataSource,
            loadMore: props.loadMore,
            renderItem: (item, index) => /* @__PURE__ */ jsxRuntime.jsx(
              LiquidationCellWidget,
              {
                item,
                index,
                classNames: {
                  root: props.classNames?.cell
                }
              }
            )
          }
        )
      ]
    }
  );
};
var useLiquidationScript = (props) => {
  const { symbol, enableLoadMore } = props;
  const { page, pageSize, setPage, pagination, parsePagination } = ui.usePagination({
    pageSize: 10
  });
  const { dateRange, filterDays, updateFilterDays, filterItems, onFilter } = useFilter2();
  React2.useEffect(() => {
    setPage(1);
  }, [symbol, dateRange, filterDays]);
  const [data, { meta, isLoading, loadMore }] = useLiquidation({
    page: enableLoadMore ? void 0 : page,
    size: pageSize,
    symbol,
    start_t: dateRange.from != null ? dateRange.from.getTime() : void 0,
    end_t: dateRange.to != null ? dateRange.to.getTime() : void 0
  });
  const dataSource = reactApp.useDataTap(data, {
    fallbackData: []
  });
  return {
    dataSource,
    isLoading,
    loadMore,
    pagination: parsePagination(meta),
    // filter
    dateRange,
    filterDays,
    updateFilterDays,
    filterItems,
    onFilter
  };
};
var useLiquidation = (props) => {
  const ordersResponse = hooks.usePrivateInfiniteQuery(generateKeyFun(props), {
    initialSize: 1,
    formatter: (data2) => data2,
    revalidateOnFocus: true
  });
  const meta = React2.useMemo(() => {
    return ordersResponse.data?.[0]?.meta;
  }, [ordersResponse.data?.[0]]);
  const data = React2.useMemo(() => {
    return ordersResponse.data?.map(
      (item) => (
        // @ts-ignore
        item.rows?.map((item2) => {
          let liquidationFeeRate = null;
          const firstPosition = item2.positions_by_perp[0];
          if (firstPosition) {
            liquidationFeeRate = new utils.Decimal(firstPosition.liquidator_fee).add(firstPosition.insurance_fund_fee).toNumber();
          }
          return {
            ...item2,
            formatted_account_mmr: isNaN(item2.account_mmr) ? null : new utils.Decimal(item2.account_mmr).mul(100).toFixed(2).toString(),
            liquidationFeeRate
          };
        })
      )
    )?.flat();
  }, [ordersResponse.data]);
  const isLoading = ordersResponse.isLoading;
  const loadMore = () => {
    ordersResponse.setSize(ordersResponse.size + 1);
  };
  return [data, { meta, isLoading, loadMore }];
};
var useFilter2 = () => {
  const defaultRange = formatDatePickerRange({
    to: offsetEndOfDay(/* @__PURE__ */ new Date()),
    from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 89))
  });
  const [filterDays, setFilterDays] = React2.useState(90);
  const [dateRange, setDateRange] = React2.useState(defaultRange);
  const updateFilterDays = (days) => {
    setFilterDays(days);
    setDateRange({
      from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), days - 1)),
      to: offsetEndOfDay(/* @__PURE__ */ new Date())
    });
  };
  const onFilter = (filter) => {
    if (filter.name === "dateRange") {
      const newDateRange = formatDatePickerRange(filter.value);
      setDateRange(newDateRange);
      if (newDateRange.from && newDateRange.to) {
        const diffDays = Math.abs(dateFns.differenceInDays(newDateRange.from, newDateRange.to)) + 1;
        const dateRangeMap = {
          1: {
            from: offsetStartOfDay(/* @__PURE__ */ new Date()),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          },
          7: {
            from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 6)),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          },
          30: {
            from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 29)),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          },
          90: {
            from: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 89)),
            to: offsetEndOfDay(/* @__PURE__ */ new Date())
          }
        };
        const dateRange2 = dateRangeMap[diffDays];
        if (dateRange2 && areDatesEqual(dateRange2.from, newDateRange.from) && areDatesEqual(dateRange2.to, newDateRange.to)) {
          setFilterDays(diffDays);
        } else {
          setFilterDays(null);
        }
      }
    }
  };
  const { isMobile } = ui.useScreen();
  const filterItems = React2.useMemo(() => {
    const dateRangeFilter = {
      type: "range",
      name: "dateRange",
      value: dateRange,
      fromDate: offsetStartOfDay(dateFns.subDays(/* @__PURE__ */ new Date(), 89)),
      toDate: offsetEndOfDay(/* @__PURE__ */ new Date())
    };
    if (isMobile) {
      return [dateRangeFilter];
    }
    return [dateRangeFilter];
  }, [dateRange]);
  return {
    filterItems,
    onFilter,
    dateRange,
    filterDays,
    updateFilterDays
  };
};
var generateKeyFun = (args) => (pageIndex, previousPageData) => {
  if (previousPageData && !previousPageData.rows?.length) return null;
  const { symbol, size = 10, page, end_t, start_t } = args;
  const search = new URLSearchParams([
    ["size", size.toString()],
    ["page", `${pageIndex + 1}`]
  ]);
  if (page) {
    search.set("page", `${page}`);
  }
  if (symbol) {
    search.set(`symbol`, symbol);
  }
  if (end_t) {
    search.set(`end_t`, `${end_t}`);
  }
  if (start_t) {
    search.set(`start_t`, `${start_t}`);
  }
  return `/v1/liquidations?${search.toString()}`;
};
var LiquidationWidget = (props) => {
  const state = useLiquidationScript(props);
  return /* @__PURE__ */ jsxRuntime.jsx(Liquidation, { ...state });
};
var MobileLiquidationWidget = (props) => {
  const { classNames, ...rest } = props;
  const state = useLiquidationScript(rest);
  return /* @__PURE__ */ jsxRuntime.jsx(MobileLiquidation, { classNames, ...state });
};
var CloseAllPositions = (props) => {
  const { onCloseAll, hasOpenPositions, isClosing, className, style, symbol } = props;
  const { t } = i18n.useTranslation();
  if (symbol !== void 0) {
    return /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, {});
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Button,
    {
      onClick: onCloseAll,
      disabled: !hasOpenPositions || isClosing,
      loading: isClosing,
      variant: "outlined",
      color: "secondary",
      size: "xs",
      className: ui.cn("disabled:oui-bg-transport", className),
      style,
      children: t("positions.closeAll")
    }
  );
};
var useCloseAllPositionsScript = (options) => {
  const { symbol, onSuccess } = options || {};
  const { t } = i18n.useTranslation();
  const { tracking } = hooks.useTrack();
  const [positionsData] = hooks.usePositionStream(symbol);
  const openPositions = React2.useMemo(() => {
    if (!positionsData?.rows) return [];
    return positionsData.rows.filter(
      (position) => position.position_qty !== 0 && position.position_qty !== void 0
    );
  }, [positionsData]);
  const hasOpenPositions = React2.useMemo(() => {
    return openPositions.length > 0;
  }, [openPositions]);
  const [doCreateOrder, { isMutating }] = hooks.useSubAccountMutation(
    "/v1/order",
    "POST"
  );
  const closeAllPositions = React2.useCallback(async () => {
    if (!hasOpenPositions) return;
    try {
      const closePromises = openPositions.map(
        (position) => {
          const side = position.position_qty > 0 ? types.OrderSide.SELL : types.OrderSide.BUY;
          const quantity = Math.abs(position.position_qty);
          return doCreateOrder({
            symbol: position.symbol,
            order_type: types.OrderType.MARKET,
            side,
            order_quantity: quantity,
            reduce_only: true
          });
        }
      );
      await Promise.all(closePromises);
      ui.toast.success(t("positions.closeAll.success"));
      onSuccess?.();
      return true;
    } catch (error) {
      if (error?.message !== void 0) {
        ui.toast.error(error.message);
      }
      return false;
    }
  }, [openPositions, hasOpenPositions, doCreateOrder, t, onSuccess]);
  const onCloseAll = React2.useCallback(() => {
    ui.modal.confirm({
      title: t("positions.closeAll"),
      content: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "sm", children: t("positions.closeAll.description") }),
      onCancel: async () => {
      },
      onOk: async () => {
        try {
          tracking(types.TrackerEventName.confirmCloseAllPositions, {
            positions_count: openPositions.length
          });
          const result = await closeAllPositions();
          return Promise.resolve(result);
        } catch (error) {
          return Promise.resolve(false);
        }
      }
    });
  }, [t, tracking, openPositions, closeAllPositions]);
  return {
    onCloseAll,
    hasOpenPositions,
    isClosing: isMutating,
    openPositionsCount: openPositions.length,
    symbol
  };
};
var CloseAllPositionsWidget = (props) => {
  const { className, style, symbol, onSuccess } = props;
  const state = useCloseAllPositionsScript({ symbol, onSuccess });
  return /* @__PURE__ */ jsxRuntime.jsx(CloseAllPositions, { ...state, className, style });
};

// src/index.ts
var MarketCloseConfirmID = "MarketCloseConfirmID";
ui.registerSimpleDialog(MarketCloseConfirmID, MarketCloseConfirm, {
  size: "md",
  closable: false
});

exports.CloseAllPositions = CloseAllPositions;
exports.CloseAllPositionsWidget = CloseAllPositionsWidget;
exports.CombinePositionsWidget = CombinePositionsWidget;
exports.FundingFeeButton = FundingFeeButton;
exports.FundingFeeHistoryUI = FundingFeeHistoryUI;
exports.Liquidation = Liquidation;
exports.LiquidationWidget = LiquidationWidget;
exports.MarketCloseConfirmID = MarketCloseConfirmID;
exports.MobileLiquidation = MobileLiquidation;
exports.MobileLiquidationWidget = MobileLiquidationWidget;
exports.MobilePositionHistory = MobilePositionHistory;
exports.MobilePositionHistoryWidget = MobilePositionHistoryWidget;
exports.MobilePositionsWidget = MobilePositionsWidget;
exports.OrderInfoCard = OrderInfoCard;
exports.PositionHistory = PositionHistory;
exports.PositionHistoryWidget = PositionHistoryWidget;
exports.PositionsTabName = PositionsTabName;
exports.PositionsWidget = PositionsWidget;
exports.ReversePosition = ReversePosition;
exports.ReversePositionDialogId = ReversePositionDialogId;
exports.ReversePositionWidget = ReversePositionWidget;
exports.sortList = sortList;
exports.useCloseAllPositionsScript = useCloseAllPositionsScript;
exports.useLiquidationScript = useLiquidationScript;
exports.usePositionHistoryScript = usePositionHistoryScript;
exports.usePositionsRowContext = usePositionsRowContext;
exports.useReversePositionEnabled = useReversePositionEnabled;
exports.useReversePositionScript = useReversePositionScript;
exports.useSort = useSort;
exports.useTabSort = useTabSort;
