// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/events/composeDeliveredEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u16,
} from '@metaplex-foundation/umi/serializers';

export type ComposeDeliveredEvent = {
  from: PublicKey;
  to: PublicKey;
  guid: Uint8Array;
  index: number;
};

export type ComposeDeliveredEventArgs = ComposeDeliveredEvent;

export function getComposeDeliveredEventSerializer(): Serializer<
  ComposeDeliveredEventArgs,
  ComposeDeliveredEvent
> {
  return struct<ComposeDeliveredEvent>(
    [
      ['from', publicKeySerializer()],
      ['to', publicKeySerializer()],
      ['guid', bytes({ size: 32 })],
      ['index', u16()],
    ],
    { description: 'ComposeDeliveredEvent' }
  ) as Serializer<ComposeDeliveredEventArgs, ComposeDeliveredEvent>;
}
