// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/events/inboundNonceSkippedEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';

export type InboundNonceSkippedEvent = {
  srcEid: number;
  sender: Uint8Array;
  receiver: PublicKey;
  nonce: bigint;
};

export type InboundNonceSkippedEventArgs = {
  srcEid: number;
  sender: Uint8Array;
  receiver: PublicKey;
  nonce: number | bigint;
};

export function getInboundNonceSkippedEventSerializer(): Serializer<
  InboundNonceSkippedEventArgs,
  InboundNonceSkippedEvent
> {
  return struct<InboundNonceSkippedEvent>(
    [
      ['srcEid', u32()],
      ['sender', bytes({ size: 32 })],
      ['receiver', publicKeySerializer()],
      ['nonce', u64()],
    ],
    { description: 'InboundNonceSkippedEvent' }
  ) as Serializer<InboundNonceSkippedEventArgs, InboundNonceSkippedEvent>;
}
