// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/events/lzComposeAlertEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u16,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';

export type LzComposeAlertEvent = {
  executor: PublicKey;
  from: PublicKey;
  to: PublicKey;
  guid: Uint8Array;
  index: number;
  computeUnits: bigint;
  value: bigint;
  message: Uint8Array;
  extraData: Uint8Array;
  reason: Uint8Array;
};

export type LzComposeAlertEventArgs = {
  executor: PublicKey;
  from: PublicKey;
  to: PublicKey;
  guid: Uint8Array;
  index: number;
  computeUnits: number | bigint;
  value: number | bigint;
  message: Uint8Array;
  extraData: Uint8Array;
  reason: Uint8Array;
};

export function getLzComposeAlertEventSerializer(): Serializer<
  LzComposeAlertEventArgs,
  LzComposeAlertEvent
> {
  return struct<LzComposeAlertEvent>(
    [
      ['executor', publicKeySerializer()],
      ['from', publicKeySerializer()],
      ['to', publicKeySerializer()],
      ['guid', bytes({ size: 32 })],
      ['index', u16()],
      ['computeUnits', u64()],
      ['value', u64()],
      ['message', bytes({ size: u32() })],
      ['extraData', bytes({ size: u32() })],
      ['reason', bytes({ size: u32() })],
    ],
    { description: 'LzComposeAlertEvent' }
  ) as Serializer<LzComposeAlertEventArgs, LzComposeAlertEvent>;
}
