// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/events/lzReceiveAlertEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';

export type LzReceiveAlertEvent = {
  receiver: PublicKey;
  executor: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: bigint;
  guid: Uint8Array;
  computeUnits: bigint;
  value: bigint;
  message: Uint8Array;
  extraData: Uint8Array;
  reason: Uint8Array;
};

export type LzReceiveAlertEventArgs = {
  receiver: PublicKey;
  executor: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: number | bigint;
  guid: Uint8Array;
  computeUnits: number | bigint;
  value: number | bigint;
  message: Uint8Array;
  extraData: Uint8Array;
  reason: Uint8Array;
};

export function getLzReceiveAlertEventSerializer(): Serializer<
  LzReceiveAlertEventArgs,
  LzReceiveAlertEvent
> {
  return struct<LzReceiveAlertEvent>(
    [
      ['receiver', publicKeySerializer()],
      ['executor', publicKeySerializer()],
      ['srcEid', u32()],
      ['sender', bytes({ size: 32 })],
      ['nonce', u64()],
      ['guid', bytes({ size: 32 })],
      ['computeUnits', u64()],
      ['value', u64()],
      ['message', bytes({ size: u32() })],
      ['extraData', bytes({ size: u32() })],
      ['reason', bytes({ size: u32() })],
    ],
    { description: 'LzReceiveAlertEvent' }
  ) as Serializer<LzReceiveAlertEventArgs, LzReceiveAlertEvent>;
}
