// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/events/packetVerifiedEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';

export type PacketVerifiedEvent = {
  srcEid: number;
  sender: Uint8Array;
  receiver: PublicKey;
  nonce: bigint;
  payloadHash: Uint8Array;
};

export type PacketVerifiedEventArgs = {
  srcEid: number;
  sender: Uint8Array;
  receiver: PublicKey;
  nonce: number | bigint;
  payloadHash: Uint8Array;
};

export function getPacketVerifiedEventSerializer(): Serializer<
  PacketVerifiedEventArgs,
  PacketVerifiedEvent
> {
  return struct<PacketVerifiedEvent>(
    [
      ['srcEid', u32()],
      ['sender', bytes({ size: 32 })],
      ['receiver', publicKeySerializer()],
      ['nonce', u64()],
      ['payloadHash', bytes({ size: 32 })],
    ],
    { description: 'PacketVerifiedEvent' }
  ) as Serializer<PacketVerifiedEventArgs, PacketVerifiedEvent>;
}
