// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/events/receiveLibraryTimeoutSetEvent.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable, PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  option,
  publicKey as publicKeySerializer,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  ReceiveLibraryTimeout,
  ReceiveLibraryTimeoutArgs,
  getReceiveLibraryTimeoutSerializer,
} from '../types';

export type ReceiveLibraryTimeoutSetEvent = {
  receiver: PublicKey;
  eid: number;
  timeout: Option<ReceiveLibraryTimeout>;
};

export type ReceiveLibraryTimeoutSetEventArgs = {
  receiver: PublicKey;
  eid: number;
  timeout: OptionOrNullable<ReceiveLibraryTimeoutArgs>;
};

export function getReceiveLibraryTimeoutSetEventSerializer(): Serializer<
  ReceiveLibraryTimeoutSetEventArgs,
  ReceiveLibraryTimeoutSetEvent
> {
  return struct<ReceiveLibraryTimeoutSetEvent>(
    [
      ['receiver', publicKeySerializer()],
      ['eid', u32()],
      ['timeout', option(getReceiveLibraryTimeoutSerializer())],
    ],
    { description: 'ReceiveLibraryTimeoutSetEvent' }
  ) as Serializer<
    ReceiveLibraryTimeoutSetEventArgs,
    ReceiveLibraryTimeoutSetEvent
  >;
}
