// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/simple_messagelib/accounts/sendConfigStore.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Account,
  Context,
  Pda,
  PublicKey,
  RpcAccount,
  RpcGetAccountOptions,
  RpcGetAccountsOptions,
  assertAccountExists,
  deserializeAccount,
  gpaBuilder,
  publicKey as toPublicKey,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u32,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type SendConfigStore = Account<SendConfigStoreAccountData>;

export type SendConfigStoreAccountData = {
  discriminator: Uint8Array;
  bump: number;
  data: Uint8Array;
};

export type SendConfigStoreAccountDataArgs = { bump: number; data: Uint8Array };

export function getSendConfigStoreAccountDataSerializer(): Serializer<
  SendConfigStoreAccountDataArgs,
  SendConfigStoreAccountData
> {
  return mapSerializer<
    SendConfigStoreAccountDataArgs,
    any,
    SendConfigStoreAccountData
  >(
    struct<SendConfigStoreAccountData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['bump', u8()],
        ['data', bytes({ size: u32() })],
      ],
      { description: 'SendConfigStoreAccountData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([246, 55, 104, 200, 90, 164, 50, 196]),
    })
  ) as Serializer<SendConfigStoreAccountDataArgs, SendConfigStoreAccountData>;
}

export function deserializeSendConfigStore(
  rawAccount: RpcAccount
): SendConfigStore {
  return deserializeAccount(
    rawAccount,
    getSendConfigStoreAccountDataSerializer()
  );
}

export async function fetchSendConfigStore(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<SendConfigStore> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  assertAccountExists(maybeAccount, 'SendConfigStore');
  return deserializeSendConfigStore(maybeAccount);
}

export async function safeFetchSendConfigStore(
  context: Pick<Context, 'rpc'>,
  publicKey: PublicKey | Pda,
  options?: RpcGetAccountOptions
): Promise<SendConfigStore | null> {
  const maybeAccount = await context.rpc.getAccount(
    toPublicKey(publicKey, false),
    options
  );
  return maybeAccount.exists ? deserializeSendConfigStore(maybeAccount) : null;
}

export async function fetchAllSendConfigStore(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<SendConfigStore[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts.map((maybeAccount) => {
    assertAccountExists(maybeAccount, 'SendConfigStore');
    return deserializeSendConfigStore(maybeAccount);
  });
}

export async function safeFetchAllSendConfigStore(
  context: Pick<Context, 'rpc'>,
  publicKeys: Array<PublicKey | Pda>,
  options?: RpcGetAccountsOptions
): Promise<SendConfigStore[]> {
  const maybeAccounts = await context.rpc.getAccounts(
    publicKeys.map((key) => toPublicKey(key, false)),
    options
  );
  return maybeAccounts
    .filter((maybeAccount) => maybeAccount.exists)
    .map((maybeAccount) =>
      deserializeSendConfigStore(maybeAccount as RpcAccount)
    );
}

export function getSendConfigStoreGpaBuilder(
  context: Pick<Context, 'rpc' | 'programs'>
) {
  const programId = context.programs.getPublicKey(
    'simpleMessagelib',
    '6GsmxMTHAAiFKfemuM4zBjumTjNSX5CAiw4xSSXM2Toy'
  );
  return gpaBuilder(context, programId)
    .registerFields<{
      discriminator: Uint8Array;
      bump: number;
      data: Uint8Array;
    }>({
      discriminator: [0, bytes({ size: 8 })],
      bump: [8, u8()],
      data: [9, bytes({ size: u32() })],
    })
    .deserializeUsing<SendConfigStore>((account) =>
      deserializeSendConfigStore(account)
    )
    .whereField(
      'discriminator',
      new Uint8Array([246, 55, 104, 200, 90, 164, 50, 196])
    );
}
