// @orderly.network/web3-provider-ethers/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.ts
var index_exports = {};
__export(index_exports, {
  EthersProvider: () => EthersProvider
});
module.exports = __toCommonJS(index_exports);

// src/ethersProvider.ts
var import_ethers = require("ethers");
var import_types = require("@orderly.network/types");

// src/parseError.ts
var import_ethers_decode_error = require("ethers-decode-error");
var errorDecoder = import_ethers_decode_error.ErrorDecoder.create();
function parseError(rawError) {
  return __async(this, null, function* () {
    var _a;
    const error = yield errorDecoder.decode(rawError);
    const reason = (_a = error.reason) != null ? _a : "";
    console.error("parsedError", error);
    return __spreadProps(__spreadValues({}, error), {
      message: replacePrefix(reason)
    });
  });
}
function replacePrefix(reason) {
  const prefixes = ["ethers-user-denied: ", "ethers-unsupported: "];
  if (typeof reason !== "string") {
    return reason;
  }
  for (const prefix of prefixes) {
    if (reason.startsWith(prefix)) {
      return reason.replace(prefix, "");
    }
  }
  return reason;
}

// src/ethersProvider.ts
var EthersProvider = class {
  constructor() {
    __publicField(this, "_provider");
    __publicField(this, "_originalProvider");
  }
  parseUnits(amount, decimals) {
    return import_ethers.ethers.parseUnits(amount, decimals).toString();
  }
  formatUnits(amount, decimals) {
    return import_ethers.ethers.formatUnits(amount, decimals);
  }
  set provider(provider) {
    this._provider = new import_ethers.BrowserProvider(provider, "any");
    this._originalProvider = provider;
  }
  get browserProvider() {
    if (!this._provider) {
      throw new Error("browserProvider is not initialized");
    }
    return this._provider;
  }
  signTypedData(address, toSignatureMessage) {
    return __async(this, null, function* () {
      var _a;
      return yield (_a = this.browserProvider) == null ? void 0 : _a.send("eth_signTypedData_v4", [
        address,
        toSignatureMessage
      ]);
    });
  }
  call(address, method, params, options) {
    return __async(this, null, function* () {
      const writeMethod = ["approve"];
      if (this._originalProvider.agwWallet && writeMethod.includes(method)) {
        try {
          const transactionHash = yield this._originalProvider.writeContract({
            abi: options.abi,
            address,
            functionName: method,
            args: params
          });
          return {
            hash: transactionHash
          };
        } catch (error) {
          const parsedEthersError = yield parseError(error);
          throw parsedEthersError;
        }
      }
      const singer = yield this.browserProvider.getSigner();
      const contract = new import_ethers.ethers.Contract(address, options.abi, singer);
      return contract[method].apply(null, params).catch((error) => __async(this, null, function* () {
        const parsedEthersError = yield parseError(error);
        throw parsedEthersError;
      }));
    });
  }
  send(method, params) {
    return this.browserProvider.send(method, params);
  }
  sendTransaction(contractAddress, method, payload, options) {
    return __async(this, null, function* () {
      var _a;
      const singer = yield (_a = this.browserProvider) == null ? void 0 : _a.getSigner();
      if (!singer) {
        throw new Error("singer is not exist");
      }
      const contract = new import_ethers.ethers.Contract(
        contractAddress,
        options.abi,
        this.browserProvider
      );
      const encodeFunctionData = contract.interface.encodeFunctionData(
        method,
        payload.data
      );
      const tx = {
        from: payload.from,
        to: payload.to,
        data: encodeFunctionData,
        value: payload.value
      };
      try {
        if (method === "depositTo") {
          return yield this._originalProvider.sendTransaction({
            to: contractAddress,
            data: encodeFunctionData,
            value: payload.value
          });
        }
        return yield singer.sendTransaction(tx);
      } catch (error) {
        const parsedEthersError = yield parseError(error);
        throw parsedEthersError;
      }
    });
  }
  pollTransactionReceiptWithBackoff(txHash, baseInterval = 1e3, maxInterval = 6e3, maxRetries = 30) {
    return __async(this, null, function* () {
      let interval = baseInterval;
      let retries = 0;
      while (retries < maxRetries) {
        try {
          const receipt = yield this.browserProvider.getTransactionReceipt(txHash);
          if (receipt) {
            return receipt;
          }
        } catch (error) {
        }
        yield new Promise((resolve) => setTimeout(resolve, interval));
        interval = Math.min(interval * 2, maxInterval);
        retries++;
      }
      throw new Error("Transaction did not complete after maximum retries.");
    });
  }
  callOnChain(chain, address, method, params, options) {
    const provider = new import_ethers.ethers.JsonRpcProvider(chain.public_rpc_url);
    const contract = new import_ethers.ethers.Contract(address, options.abi, provider);
    return contract[method].apply(null, params).catch((error) => __async(null, null, function* () {
      const parsedEthersError = yield parseError(error);
      throw parsedEthersError;
    }));
  }
  getBalance(userAddress) {
    return this.browserProvider.getBalance(userAddress).then((res) => {
      return res;
    });
  }
  getBalances(addresses) {
    return __async(this, null, function* () {
      if (!addresses || addresses.length === 0) {
        return [];
      }
      const MULTICALL3_ADDRESS = "0xcA11bde05977b3631167028862bE2a173976CA11";
      const MULTICALL_ABI = [
        "function tryAggregate(bool requireSuccess, tuple(address target, bytes callData)[] calls) public view returns (tuple(bool success, bytes returnData)[] returnData)",
        "function getEthBalance(address addr) view returns (uint256 balance)"
        // Multicall3 自带的方法
      ];
      const ERC20_ABI = [
        "function balanceOf(address account) view returns (uint256)"
      ];
      try {
        const signer = yield this.browserProvider.getSigner();
        const userAddress = yield signer.getAddress();
        const erc20Interface = new import_ethers.ethers.Interface(ERC20_ABI);
        const multicallInterface = new import_ethers.ethers.Interface(MULTICALL_ABI);
        const multicallContract = new import_ethers.ethers.Contract(
          MULTICALL3_ADDRESS,
          MULTICALL_ABI,
          this.browserProvider
        );
        const calls = addresses.map((tokenAddr) => {
          if ((0, import_types.isNativeTokenChecker)(tokenAddr)) {
            return {
              target: MULTICALL3_ADDRESS,
              callData: multicallInterface.encodeFunctionData("getEthBalance", [
                userAddress
              ])
            };
          }
          return {
            target: tokenAddr,
            callData: erc20Interface.encodeFunctionData("balanceOf", [
              userAddress
            ])
          };
        });
        const results = yield multicallContract.tryAggregate.staticCall(
          false,
          calls
        );
        const balances = results.map(
          (result, index) => {
            var _a;
            const [success, returnData] = result;
            if (success && returnData !== "0x") {
              const tokenAddr = addresses[index];
              const isNative = (0, import_types.isNativeTokenChecker)(tokenAddr);
              try {
                let decoded;
                if (isNative) {
                  decoded = multicallInterface.decodeFunctionResult(
                    "getEthBalance",
                    returnData
                  );
                } else {
                  decoded = erc20Interface.decodeFunctionResult(
                    "balanceOf",
                    returnData
                  );
                }
                return ((_a = decoded == null ? void 0 : decoded[0]) == null ? void 0 : _a.toString()) || "0";
              } catch (error) {
                console.warn("Failed to decode balanceOf result:", error);
                return "0";
              }
            } else {
              return "0";
            }
          }
        );
        return balances;
      } catch (error) {
        console.error("getBalances error", error);
        const parsedEthersError = yield parseError(error);
        throw parsedEthersError;
      }
    });
  }
  estimateGasFee(contractAddress, method, payload, options) {
    return __async(this, null, function* () {
      var _a, _b;
      const contract = new import_ethers.ethers.Contract(
        contractAddress,
        options.abi,
        this.browserProvider
      );
      const encodeFunctionData = contract.interface.encodeFunctionData(
        method,
        payload.data
      );
      const tx = {
        from: payload.from,
        to: payload.to,
        data: encodeFunctionData,
        value: payload.value
      };
      let gas = BigInt(0);
      const defaultGas = BigInt(8e5);
      try {
        gas = yield (_a = this.browserProvider) == null ? void 0 : _a.estimateGas(tx);
      } catch (error) {
        gas = defaultGas;
        console.error("estimateGasFee error", error);
      }
      const feeData = yield this.browserProvider.getFeeData();
      const gasPrice = (_b = feeData.gasPrice) != null ? _b : feeData.maxFeePerGas;
      if (!gasPrice) {
        throw new Error("gasPrice is not available");
      }
      const costInWei = gas * gasPrice;
      return costInWei;
    });
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EthersProvider
});
