// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/endpoint/programs/endpoint.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  ClusterFilter,
  Context,
  Program,
  PublicKey,
} from '@metaplex-foundation/umi';
import { getEndpointErrorFromCode, getEndpointErrorFromName } from '../errors';

export const ENDPOINT_PROGRAM_ID =
  '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6' as PublicKey<'76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'>;

export function createEndpointProgram(): Program {
  return {
    name: 'endpoint',
    publicKey: ENDPOINT_PROGRAM_ID,
    getErrorFromCode(code: number, cause?: Error) {
      return getEndpointErrorFromCode(code, this, cause);
    },
    getErrorFromName(name: string, cause?: Error) {
      return getEndpointErrorFromName(name, this, cause);
    },
    isOnCluster() {
      return true;
    },
  };
}

export function getEndpointProgram<T extends Program = Program>(
  context: Pick<Context, 'programs'>,
  clusterFilter?: ClusterFilter
): T {
  return context.programs.get<T>('endpoint', clusterFilter);
}

export function getEndpointProgramId(
  context: Pick<Context, 'programs'>,
  clusterFilter?: ClusterFilter
): PublicKey {
  return context.programs.getPublicKey(
    'endpoint',
    ENDPOINT_PROGRAM_ID,
    clusterFilter
  );
}
