// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/uln/errors/index.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * Unauthorized: ''
 *
 * @category Errors
 * @category generated
 */
export class UnauthorizedError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'Unauthorized'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, UnauthorizedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new UnauthorizedError())
createErrorFromNameLookup.set('Unauthorized', () => new UnauthorizedError())

/**
 * InvalidAmount: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidAmountError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'InvalidAmount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidAmountError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new InvalidAmountError())
createErrorFromNameLookup.set('InvalidAmount', () => new InvalidAmountError())

/**
 * InvalidExecutor: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidExecutorError extends Error {
  readonly code: number = 0x1772
  readonly name: string = 'InvalidExecutor'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidExecutorError)
    }
  }
}

createErrorFromCodeLookup.set(0x1772, () => new InvalidExecutorError())
createErrorFromNameLookup.set(
  'InvalidExecutor',
  () => new InvalidExecutorError()
)

/**
 * ZeroMessageSize: ''
 *
 * @category Errors
 * @category generated
 */
export class ZeroMessageSizeError extends Error {
  readonly code: number = 0x1773
  readonly name: string = 'ZeroMessageSize'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ZeroMessageSizeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1773, () => new ZeroMessageSizeError())
createErrorFromNameLookup.set(
  'ZeroMessageSize',
  () => new ZeroMessageSizeError()
)

/**
 * InvalidRequiredDVNCount: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidRequiredDVNCountError extends Error {
  readonly code: number = 0x1774
  readonly name: string = 'InvalidRequiredDVNCount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidRequiredDVNCountError)
    }
  }
}

createErrorFromCodeLookup.set(0x1774, () => new InvalidRequiredDVNCountError())
createErrorFromNameLookup.set(
  'InvalidRequiredDVNCount',
  () => new InvalidRequiredDVNCountError()
)

/**
 * InvalidOptionalDVNCount: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidOptionalDVNCountError extends Error {
  readonly code: number = 0x1775
  readonly name: string = 'InvalidOptionalDVNCount'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidOptionalDVNCountError)
    }
  }
}

createErrorFromCodeLookup.set(0x1775, () => new InvalidOptionalDVNCountError())
createErrorFromNameLookup.set(
  'InvalidOptionalDVNCount',
  () => new InvalidOptionalDVNCountError()
)

/**
 * InvalidOptionalDVNThreshold: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidOptionalDVNThresholdError extends Error {
  readonly code: number = 0x1776
  readonly name: string = 'InvalidOptionalDVNThreshold'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidOptionalDVNThresholdError)
    }
  }
}

createErrorFromCodeLookup.set(
  0x1776,
  () => new InvalidOptionalDVNThresholdError()
)
createErrorFromNameLookup.set(
  'InvalidOptionalDVNThreshold',
  () => new InvalidOptionalDVNThresholdError()
)

/**
 * InvalidConfirmations: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidConfirmationsError extends Error {
  readonly code: number = 0x1777
  readonly name: string = 'InvalidConfirmations'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidConfirmationsError)
    }
  }
}

createErrorFromCodeLookup.set(0x1777, () => new InvalidConfirmationsError())
createErrorFromNameLookup.set(
  'InvalidConfirmations',
  () => new InvalidConfirmationsError()
)

/**
 * AtLeastOneDVN: ''
 *
 * @category Errors
 * @category generated
 */
export class AtLeastOneDVNError extends Error {
  readonly code: number = 0x1778
  readonly name: string = 'AtLeastOneDVN'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, AtLeastOneDVNError)
    }
  }
}

createErrorFromCodeLookup.set(0x1778, () => new AtLeastOneDVNError())
createErrorFromNameLookup.set('AtLeastOneDVN', () => new AtLeastOneDVNError())

/**
 * Unsorted: ''
 *
 * @category Errors
 * @category generated
 */
export class UnsortedError extends Error {
  readonly code: number = 0x1779
  readonly name: string = 'Unsorted'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, UnsortedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1779, () => new UnsortedError())
createErrorFromNameLookup.set('Unsorted', () => new UnsortedError())

/**
 * InvalidTreasuryFeeCap: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidTreasuryFeeCapError extends Error {
  readonly code: number = 0x177a
  readonly name: string = 'InvalidTreasuryFeeCap'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidTreasuryFeeCapError)
    }
  }
}

createErrorFromCodeLookup.set(0x177a, () => new InvalidTreasuryFeeCapError())
createErrorFromNameLookup.set(
  'InvalidTreasuryFeeCap',
  () => new InvalidTreasuryFeeCapError()
)

/**
 * InvalidPacketVersion: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidPacketVersionError extends Error {
  readonly code: number = 0x177b
  readonly name: string = 'InvalidPacketVersion'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidPacketVersionError)
    }
  }
}

createErrorFromCodeLookup.set(0x177b, () => new InvalidPacketVersionError())
createErrorFromNameLookup.set(
  'InvalidPacketVersion',
  () => new InvalidPacketVersionError()
)

/**
 * InvalidEid: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidEidError extends Error {
  readonly code: number = 0x177c
  readonly name: string = 'InvalidEid'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidEidError)
    }
  }
}

createErrorFromCodeLookup.set(0x177c, () => new InvalidEidError())
createErrorFromNameLookup.set('InvalidEid', () => new InvalidEidError())

/**
 * Verifying: ''
 *
 * @category Errors
 * @category generated
 */
export class VerifyingError extends Error {
  readonly code: number = 0x177d
  readonly name: string = 'Verifying'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, VerifyingError)
    }
  }
}

createErrorFromCodeLookup.set(0x177d, () => new VerifyingError())
createErrorFromNameLookup.set('Verifying', () => new VerifyingError())

/**
 * InvalidWorkerId: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidWorkerIdError extends Error {
  readonly code: number = 0x177e
  readonly name: string = 'InvalidWorkerId'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidWorkerIdError)
    }
  }
}

createErrorFromCodeLookup.set(0x177e, () => new InvalidWorkerIdError())
createErrorFromNameLookup.set(
  'InvalidWorkerId',
  () => new InvalidWorkerIdError()
)

/**
 * InvalidOptionType: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidOptionTypeError extends Error {
  readonly code: number = 0x177f
  readonly name: string = 'InvalidOptionType'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidOptionTypeError)
    }
  }
}

createErrorFromCodeLookup.set(0x177f, () => new InvalidOptionTypeError())
createErrorFromNameLookup.set(
  'InvalidOptionType',
  () => new InvalidOptionTypeError()
)

/**
 * InvalidBps: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidBpsError extends Error {
  readonly code: number = 0x1780
  readonly name: string = 'InvalidBps'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidBpsError)
    }
  }
}

createErrorFromCodeLookup.set(0x1780, () => new InvalidBpsError())
createErrorFromNameLookup.set('InvalidBps', () => new InvalidBpsError())

/**
 * ExceededMaxMessageSize: ''
 *
 * @category Errors
 * @category generated
 */
export class ExceededMaxMessageSizeError extends Error {
  readonly code: number = 0x1781
  readonly name: string = 'ExceededMaxMessageSize'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ExceededMaxMessageSizeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1781, () => new ExceededMaxMessageSizeError())
createErrorFromNameLookup.set(
  'ExceededMaxMessageSize',
  () => new ExceededMaxMessageSizeError()
)

/**
 * InvalidExecutorProgram: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidExecutorProgramError extends Error {
  readonly code: number = 0x1782
  readonly name: string = 'InvalidExecutorProgram'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidExecutorProgramError)
    }
  }
}

createErrorFromCodeLookup.set(0x1782, () => new InvalidExecutorProgramError())
createErrorFromNameLookup.set(
  'InvalidExecutorProgram',
  () => new InvalidExecutorProgramError()
)

/**
 * InvalidAccountLength: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidAccountLengthError extends Error {
  readonly code: number = 0x1783
  readonly name: string = 'InvalidAccountLength'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidAccountLengthError)
    }
  }
}

createErrorFromCodeLookup.set(0x1783, () => new InvalidAccountLengthError())
createErrorFromNameLookup.set(
  'InvalidAccountLength',
  () => new InvalidAccountLengthError()
)

/**
 * InvalidDvnProgram: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidDvnProgramError extends Error {
  readonly code: number = 0x1784
  readonly name: string = 'InvalidDvnProgram'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidDvnProgramError)
    }
  }
}

createErrorFromCodeLookup.set(0x1784, () => new InvalidDvnProgramError())
createErrorFromNameLookup.set(
  'InvalidDvnProgram',
  () => new InvalidDvnProgramError()
)

/**
 * InvalidDvn: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidDvnError extends Error {
  readonly code: number = 0x1785
  readonly name: string = 'InvalidDvn'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidDvnError)
    }
  }
}

createErrorFromCodeLookup.set(0x1785, () => new InvalidDvnError())
createErrorFromNameLookup.set('InvalidDvn', () => new InvalidDvnError())

/**
 * LzTokenUnavailable: ''
 *
 * @category Errors
 * @category generated
 */
export class LzTokenUnavailableError extends Error {
  readonly code: number = 0x1786
  readonly name: string = 'LzTokenUnavailable'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, LzTokenUnavailableError)
    }
  }
}

createErrorFromCodeLookup.set(0x1786, () => new LzTokenUnavailableError())
createErrorFromNameLookup.set(
  'LzTokenUnavailable',
  () => new LzTokenUnavailableError()
)

/**
 * InsufficientFee: ''
 *
 * @category Errors
 * @category generated
 */
export class InsufficientFeeError extends Error {
  readonly code: number = 0x1787
  readonly name: string = 'InsufficientFee'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InsufficientFeeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1787, () => new InsufficientFeeError())
createErrorFromNameLookup.set(
  'InsufficientFee',
  () => new InsufficientFeeError()
)

/**
 * InvalidTreasury: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidTreasuryError extends Error {
  readonly code: number = 0x1788
  readonly name: string = 'InvalidTreasury'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidTreasuryError)
    }
  }
}

createErrorFromCodeLookup.set(0x1788, () => new InvalidTreasuryError())
createErrorFromNameLookup.set(
  'InvalidTreasury',
  () => new InvalidTreasuryError()
)

/**
 * InvalidLzTokenMint: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidLzTokenMintError extends Error {
  readonly code: number = 0x1789
  readonly name: string = 'InvalidLzTokenMint'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidLzTokenMintError)
    }
  }
}

createErrorFromCodeLookup.set(0x1789, () => new InvalidLzTokenMintError())
createErrorFromNameLookup.set(
  'InvalidLzTokenMint',
  () => new InvalidLzTokenMintError()
)

/**
 * InvalidConfigType: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidConfigTypeError extends Error {
  readonly code: number = 0x178a
  readonly name: string = 'InvalidConfigType'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidConfigTypeError)
    }
  }
}

createErrorFromCodeLookup.set(0x178a, () => new InvalidConfigTypeError())
createErrorFromNameLookup.set(
  'InvalidConfigType',
  () => new InvalidConfigTypeError()
)

/**
 * InvalidConfirmation: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidConfirmationError extends Error {
  readonly code: number = 0x178b
  readonly name: string = 'InvalidConfirmation'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidConfirmationError)
    }
  }
}

createErrorFromCodeLookup.set(0x178b, () => new InvalidConfirmationError())
createErrorFromNameLookup.set(
  'InvalidConfirmation',
  () => new InvalidConfirmationError()
)

/**
 * InvalidType1Size: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidType1SizeError extends Error {
  readonly code: number = 0x178c
  readonly name: string = 'InvalidType1Size'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidType1SizeError)
    }
  }
}

createErrorFromCodeLookup.set(0x178c, () => new InvalidType1SizeError())
createErrorFromNameLookup.set(
  'InvalidType1Size',
  () => new InvalidType1SizeError()
)

/**
 * InvalidType2Size: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidType2SizeError extends Error {
  readonly code: number = 0x178d
  readonly name: string = 'InvalidType2Size'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidType2SizeError)
    }
  }
}

createErrorFromCodeLookup.set(0x178d, () => new InvalidType2SizeError())
createErrorFromNameLookup.set(
  'InvalidType2Size',
  () => new InvalidType2SizeError()
)

/**
 * ExceededU128: ''
 *
 * @category Errors
 * @category generated
 */
export class ExceededU128Error extends Error {
  readonly code: number = 0x178e
  readonly name: string = 'ExceededU128'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ExceededU128Error)
    }
  }
}

createErrorFromCodeLookup.set(0x178e, () => new ExceededU128Error())
createErrorFromNameLookup.set('ExceededU128', () => new ExceededU128Error())

/**
 * NonSigner: ''
 *
 * @category Errors
 * @category generated
 */
export class NonSignerError extends Error {
  readonly code: number = 0x178f
  readonly name: string = 'NonSigner'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, NonSignerError)
    }
  }
}

createErrorFromCodeLookup.set(0x178f, () => new NonSignerError())
createErrorFromNameLookup.set('NonSigner', () => new NonSignerError())

/**
 * InvalidPayer: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidPayerError extends Error {
  readonly code: number = 0x1790
  readonly name: string = 'InvalidPayer'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidPayerError)
    }
  }
}

createErrorFromCodeLookup.set(0x1790, () => new InvalidPayerError())
createErrorFromNameLookup.set('InvalidPayer', () => new InvalidPayerError())

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
