// @layerzerolabs/lz-foundation/dist/index.cjs
'use strict';

var blake2b$1 = require('@noble/hashes/blake2b');
var sha256 = require('@noble/hashes/sha256');
var sha3 = require('@noble/hashes/sha3');
var ed25519 = require('@noble/ed25519');
var secp256k1 = require('@noble/secp256k1');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var ed25519__namespace = /*#__PURE__*/_interopNamespace(ed25519);
var secp256k1__namespace = /*#__PURE__*/_interopNamespace(secp256k1);

var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
function keccak_256(message) {
  return keccak_256();
}
function blake2b(message) {
  return blake2b$1.blake2b(message);
}
function sha3_256(message) {
  return sha3.sha3_256(message);
}
function sha2_256(message) {
  return sha256.sha256(message);
}

// src/sign/constant.ts
var SignAlgorithm = /* @__PURE__ */ ((SignAlgorithm2) => {
  SignAlgorithm2[SignAlgorithm2["NATIVE"] = 0] = "NATIVE";
  SignAlgorithm2[SignAlgorithm2["SECP256K1"] = 1] = "SECP256K1";
  SignAlgorithm2[SignAlgorithm2["ED25519"] = 2] = "ED25519";
  return SignAlgorithm2;
})(SignAlgorithm || {});

// src/sign/ed25519.ts
var ed25519_exports = {};
__export(ed25519_exports, {
  getPublicKey: () => getPublicKey2,
  signHash: () => signHash
});

// src/sign/internal.ts
function normalizePrivateKey(privateKey) {
  if (typeof privateKey === "string") {
    return privateKey.replace(/^0x/, "");
  } else if (typeof privateKey === "bigint") {
    return privateKey.toString(16).padStart(64, "0");
  } else {
    return privateKey;
  }
}

// src/sign/ed25519.ts
async function signHash(hash, privateKey) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  privateKey = normalizePrivateKey(privateKey);
  const pubKey = await ed25519__namespace.getPublicKey(privateKey);
  const signature = await ed25519__namespace.sign(hash, privateKey);
  const isValid = await ed25519__namespace.verify(signature, hash, pubKey);
  if (!isValid) {
    throw new Error("Invalid signature");
  }
  return signature;
}
async function getPublicKey2(privateKey) {
  privateKey = normalizePrivateKey(privateKey);
  return ed25519__namespace.getPublicKey(privateKey);
}

// src/sign/secp256k1.ts
var secp256k1_exports = {};
__export(secp256k1_exports, {
  getPublicKey: () => getPublicKey4,
  signHash: () => signHash2
});
async function signHash2(hash, privateKey) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  privateKey = normalizePrivateKey(privateKey);
  const pubKey = secp256k1__namespace.getPublicKey(privateKey, false);
  const [signature, recoveryId] = await secp256k1__namespace.sign(hash, privateKey, {
    // recovered - true indicates the recovered bit should be included in the result
    recovered: true,
    // canonical - true makes signatures compatible with libsecp256k1, false makes signatures compatible with openssl
    canonical: true,
    // der - whether the returned signature should be in DER format. If false, it would be in Compact format (32-byte r + 32-byte s)
    der: false,
    // extraEntropy - When true, it would automatically be filled with 32 bytes of cryptographically secure entropy. Strongly recommended to pass true to improve security:
    extraEntropy: void 0
  });
  const isValid = secp256k1__namespace.verify(signature, hash, pubKey, { strict: true });
  if (!isValid) {
    throw new Error("Invalid signature");
  }
  return Uint8Array.from([...signature, recoveryId]);
}
async function getPublicKey4(privateKey) {
  privateKey = normalizePrivateKey(privateKey);
  return Promise.resolve(secp256k1__namespace.getPublicKey(privateKey, false));
}

// src/sign/index.ts
async function signHash3(hash, privateKey, algorithm) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  const privKey = typeof privateKey === "string" ? privateKey.replace(/^0x/, "") : Buffer.from(privateKey).toString("hex");
  switch (algorithm) {
    case 1 /* SECP256K1 */:
      return signHash2(hash, privKey);
    case 2 /* ED25519 */:
      return signHash(hash, privKey);
    default:
      throw new Error(`Unsupported algorithm: ${algorithm}`);
  }
}

exports.SignAlgorithm = SignAlgorithm;
exports.blake2b = blake2b;
exports.ed25519 = ed25519_exports;
exports.keccak_256 = keccak_256;
exports.secp256k1 = secp256k1_exports;
exports.sha2_256 = sha2_256;
exports.sha3_256 = sha3_256;
exports.signHash = signHash3;
