// @layerzerolabs/lz-foundation/dist/index.mjs
import { blake2b as blake2b$1 } from '@noble/hashes/blake2b';
import { sha256 } from '@noble/hashes/sha256';
import { sha3_256 as sha3_256$1 } from '@noble/hashes/sha3';
import * as ed25519 from '@noble/ed25519';
import * as secp256k1 from '@noble/secp256k1';

var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
function keccak_256(message) {
  return keccak_256();
}
function blake2b(message) {
  return blake2b$1(message);
}
function sha3_256(message) {
  return sha3_256$1(message);
}
function sha2_256(message) {
  return sha256(message);
}

// src/sign/constant.ts
var SignAlgorithm = /* @__PURE__ */ ((SignAlgorithm2) => {
  SignAlgorithm2[SignAlgorithm2["NATIVE"] = 0] = "NATIVE";
  SignAlgorithm2[SignAlgorithm2["SECP256K1"] = 1] = "SECP256K1";
  SignAlgorithm2[SignAlgorithm2["ED25519"] = 2] = "ED25519";
  return SignAlgorithm2;
})(SignAlgorithm || {});

// src/sign/ed25519.ts
var ed25519_exports = {};
__export(ed25519_exports, {
  getPublicKey: () => getPublicKey2,
  signHash: () => signHash
});

// src/sign/internal.ts
function normalizePrivateKey(privateKey) {
  if (typeof privateKey === "string") {
    return privateKey.replace(/^0x/, "");
  } else if (typeof privateKey === "bigint") {
    return privateKey.toString(16).padStart(64, "0");
  } else {
    return privateKey;
  }
}

// src/sign/ed25519.ts
async function signHash(hash, privateKey) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  privateKey = normalizePrivateKey(privateKey);
  const pubKey = await ed25519.getPublicKey(privateKey);
  const signature = await ed25519.sign(hash, privateKey);
  const isValid = await ed25519.verify(signature, hash, pubKey);
  if (!isValid) {
    throw new Error("Invalid signature");
  }
  return signature;
}
async function getPublicKey2(privateKey) {
  privateKey = normalizePrivateKey(privateKey);
  return ed25519.getPublicKey(privateKey);
}

// src/sign/secp256k1.ts
var secp256k1_exports = {};
__export(secp256k1_exports, {
  getPublicKey: () => getPublicKey4,
  signHash: () => signHash2
});
async function signHash2(hash, privateKey) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  privateKey = normalizePrivateKey(privateKey);
  const pubKey = secp256k1.getPublicKey(privateKey, false);
  const [signature, recoveryId] = await secp256k1.sign(hash, privateKey, {
    // recovered - true indicates the recovered bit should be included in the result
    recovered: true,
    // canonical - true makes signatures compatible with libsecp256k1, false makes signatures compatible with openssl
    canonical: true,
    // der - whether the returned signature should be in DER format. If false, it would be in Compact format (32-byte r + 32-byte s)
    der: false,
    // extraEntropy - When true, it would automatically be filled with 32 bytes of cryptographically secure entropy. Strongly recommended to pass true to improve security:
    extraEntropy: void 0
  });
  const isValid = secp256k1.verify(signature, hash, pubKey, { strict: true });
  if (!isValid) {
    throw new Error("Invalid signature");
  }
  return Uint8Array.from([...signature, recoveryId]);
}
async function getPublicKey4(privateKey) {
  privateKey = normalizePrivateKey(privateKey);
  return Promise.resolve(secp256k1.getPublicKey(privateKey, false));
}

// src/sign/index.ts
async function signHash3(hash, privateKey, algorithm) {
  if (hash.length !== 32) {
    throw new Error("Hash must be 32 bytes long");
  }
  const privKey = typeof privateKey === "string" ? privateKey.replace(/^0x/, "") : Buffer.from(privateKey).toString("hex");
  switch (algorithm) {
    case 1 /* SECP256K1 */:
      return signHash2(hash, privKey);
    case 2 /* ED25519 */:
      return signHash(hash, privKey);
    default:
      throw new Error(`Unsupported algorithm: ${algorithm}`);
  }
}

export { SignAlgorithm, blake2b, ed25519_exports as ed25519, keccak_256, secp256k1_exports as secp256k1, sha2_256, sha3_256, signHash3 as signHash };
