// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/errors/dvn.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Program, ProgramError } from '@metaplex-foundation/umi';

type ProgramErrorConstructor = new (
  program: Program,
  cause?: Error
) => ProgramError;
const codeToErrorMap: Map<number, ProgramErrorConstructor> = new Map();
const nameToErrorMap: Map<string, ProgramErrorConstructor> = new Map();

/** InvalidSignatureLen */
export class InvalidSignatureLenError extends ProgramError {
  override readonly name: string = 'InvalidSignatureLen';

  readonly code: number = 0x1770; // 6000

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1770, InvalidSignatureLenError);
nameToErrorMap.set('InvalidSignatureLen', InvalidSignatureLenError);

/** NotAdmin */
export class NotAdminError extends ProgramError {
  override readonly name: string = 'NotAdmin';

  readonly code: number = 0x1771; // 6001

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1771, NotAdminError);
nameToErrorMap.set('NotAdmin', NotAdminError);

/** MsgLibNotAllowed */
export class MsgLibNotAllowedError extends ProgramError {
  override readonly name: string = 'MsgLibNotAllowed';

  readonly code: number = 0x1772; // 6002

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1772, MsgLibNotAllowedError);
nameToErrorMap.set('MsgLibNotAllowed', MsgLibNotAllowedError);

/** InvalidQuorum */
export class InvalidQuorumError extends ProgramError {
  override readonly name: string = 'InvalidQuorum';

  readonly code: number = 0x1773; // 6003

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1773, InvalidQuorumError);
nameToErrorMap.set('InvalidQuorum', InvalidQuorumError);

/** InvalidSignersLen */
export class InvalidSignersLenError extends ProgramError {
  override readonly name: string = 'InvalidSignersLen';

  readonly code: number = 0x1774; // 6004

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1774, InvalidSignersLenError);
nameToErrorMap.set('InvalidSignersLen', InvalidSignersLenError);

/** UniqueOwners */
export class UniqueOwnersError extends ProgramError {
  override readonly name: string = 'UniqueOwners';

  readonly code: number = 0x1775; // 6005

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1775, UniqueOwnersError);
nameToErrorMap.set('UniqueOwners', UniqueOwnersError);

/** SignatureError */
export class SignatureErrorError extends ProgramError {
  override readonly name: string = 'SignatureError';

  readonly code: number = 0x1776; // 6006

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1776, SignatureErrorError);
nameToErrorMap.set('SignatureError', SignatureErrorError);

/** SignerNotInCommittee */
export class SignerNotInCommitteeError extends ProgramError {
  override readonly name: string = 'SignerNotInCommittee';

  readonly code: number = 0x1777; // 6007

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1777, SignerNotInCommitteeError);
nameToErrorMap.set('SignerNotInCommittee', SignerNotInCommitteeError);

/** TooManyAdmins */
export class TooManyAdminsError extends ProgramError {
  override readonly name: string = 'TooManyAdmins';

  readonly code: number = 0x1778; // 6008

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1778, TooManyAdminsError);
nameToErrorMap.set('TooManyAdmins', TooManyAdminsError);

/** TooManyOptionTypes */
export class TooManyOptionTypesError extends ProgramError {
  override readonly name: string = 'TooManyOptionTypes';

  readonly code: number = 0x1779; // 6009

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1779, TooManyOptionTypesError);
nameToErrorMap.set('TooManyOptionTypes', TooManyOptionTypesError);

/** DuplicateSignature */
export class DuplicateSignatureError extends ProgramError {
  override readonly name: string = 'DuplicateSignature';

  readonly code: number = 0x177a; // 6010

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177a, DuplicateSignatureError);
nameToErrorMap.set('DuplicateSignature', DuplicateSignatureError);

/** Expired */
export class ExpiredError extends ProgramError {
  override readonly name: string = 'Expired';

  readonly code: number = 0x177b; // 6011

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177b, ExpiredError);
nameToErrorMap.set('Expired', ExpiredError);

/** InvalidVid */
export class InvalidVidError extends ProgramError {
  override readonly name: string = 'InvalidVid';

  readonly code: number = 0x177c; // 6012

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177c, InvalidVidError);
nameToErrorMap.set('InvalidVid', InvalidVidError);

/** Paused */
export class PausedError extends ProgramError {
  override readonly name: string = 'Paused';

  readonly code: number = 0x177d; // 6013

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177d, PausedError);
nameToErrorMap.set('Paused', PausedError);

/** UnexpiredExecuteHash */
export class UnexpiredExecuteHashError extends ProgramError {
  override readonly name: string = 'UnexpiredExecuteHash';

  readonly code: number = 0x177e; // 6014

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177e, UnexpiredExecuteHashError);
nameToErrorMap.set('UnexpiredExecuteHash', UnexpiredExecuteHashError);

/** InvalidAmount */
export class InvalidAmountError extends ProgramError {
  override readonly name: string = 'InvalidAmount';

  readonly code: number = 0x177f; // 6015

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177f, InvalidAmountError);
nameToErrorMap.set('InvalidAmount', InvalidAmountError);

/** EidNotSupported */
export class EidNotSupportedError extends ProgramError {
  override readonly name: string = 'EidNotSupported';

  readonly code: number = 0x1780; // 6016

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1780, EidNotSupportedError);
nameToErrorMap.set('EidNotSupported', EidNotSupportedError);

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
export function getDvnErrorFromCode(
  code: number,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
export function getDvnErrorFromName(
  name: string,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}
