// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/burn.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { BurnParams, burnParamsBeet } from '../types/BurnParams'

/**
 * @category Instructions
 * @category Burn
 * @category generated
 */
export type BurnInstructionArgs = {
  params: BurnParams
}
/**
 * @category Instructions
 * @category Burn
 * @category generated
 */
export const burnStruct = new beet.BeetArgsStruct<
  BurnInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', burnParamsBeet],
  ],
  'BurnInstructionArgs'
)
/**
 * Accounts required by the _burn_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [] nonce
 * @property [_writable_] payloadHash
 * @property [_writable_] endpoint
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Burn
 * @category generated
 */
export type BurnInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  nonce: web3.PublicKey
  payloadHash: web3.PublicKey
  endpoint: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const burnInstructionDiscriminator = [116, 110, 29, 56, 107, 219, 42, 93]

/**
 * Creates a _Burn_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Burn
 * @category generated
 */
export function createBurnInstruction(
  accounts: BurnInstructionAccounts,
  args: BurnInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = burnStruct.serialize({
    instructionDiscriminator: burnInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Burn_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Burn
 * @category generated
 */
export function createBurnInstructionAccounts(
  accounts: BurnInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
