// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/clear.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { ClearParams, clearParamsBeet } from '../types/ClearParams'

/**
 * @category Instructions
 * @category Clear
 * @category generated
 */
export type ClearInstructionArgs = {
  params: ClearParams
}
/**
 * @category Instructions
 * @category Clear
 * @category generated
 */
export const clearStruct = new beet.FixableBeetArgsStruct<
  ClearInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', clearParamsBeet],
  ],
  'ClearInstructionArgs'
)
/**
 * Accounts required by the _clear_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [] nonce
 * @property [_writable_] payloadHash
 * @property [_writable_] endpoint
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Clear
 * @category generated
 */
export type ClearInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  nonce: web3.PublicKey
  payloadHash: web3.PublicKey
  endpoint: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const clearInstructionDiscriminator = [
  250, 39, 28, 213, 123, 163, 133, 5,
]

/**
 * Creates a _Clear_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Clear
 * @category generated
 */
export function createClearInstruction(
  accounts: ClearInstructionAccounts,
  args: ClearInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = clearStruct.serialize({
    instructionDiscriminator: clearInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Clear_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Clear
 * @category generated
 */
export function createClearInstructionAccounts(
  accounts: ClearInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
