// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/initReceiveLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitReceiveLibraryParams,
  initReceiveLibraryParamsBeet,
} from '../types/InitReceiveLibraryParams'

/**
 * @category Instructions
 * @category InitReceiveLibrary
 * @category generated
 */
export type InitReceiveLibraryInstructionArgs = {
  params: InitReceiveLibraryParams
}
/**
 * @category Instructions
 * @category InitReceiveLibrary
 * @category generated
 */
export const initReceiveLibraryStruct = new beet.BeetArgsStruct<
  InitReceiveLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initReceiveLibraryParamsBeet],
  ],
  'InitReceiveLibraryInstructionArgs'
)
/**
 * Accounts required by the _initReceiveLibrary_ instruction
 *
 * @property [_writable_, **signer**] delegate
 * @property [] oappRegistry
 * @property [_writable_] receiveLibraryConfig
 * @category Instructions
 * @category InitReceiveLibrary
 * @category generated
 */
export type InitReceiveLibraryInstructionAccounts = {
  delegate: web3.PublicKey
  oappRegistry: web3.PublicKey
  receiveLibraryConfig: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initReceiveLibraryInstructionDiscriminator = [
  197, 114, 81, 100, 45, 233, 36, 230,
]

/**
 * Creates a _InitReceiveLibrary_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitReceiveLibrary
 * @category generated
 */
export function createInitReceiveLibraryInstruction(
  accounts: InitReceiveLibraryInstructionAccounts,
  args: InitReceiveLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initReceiveLibraryStruct.serialize({
    instructionDiscriminator: initReceiveLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.delegate,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitReceiveLibrary_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitReceiveLibrary
 * @category generated
 */
export function createInitReceiveLibraryInstructionAccounts(
  accounts: InitReceiveLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.delegate,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
