// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/initEndpoint.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitEndpointParams,
  initEndpointParamsBeet,
} from '../types/InitEndpointParams'

/**
 * @category Instructions
 * @category InitEndpoint
 * @category generated
 */
export type InitEndpointInstructionArgs = {
  params: InitEndpointParams
}
/**
 * @category Instructions
 * @category InitEndpoint
 * @category generated
 */
export const initEndpointStruct = new beet.BeetArgsStruct<
  InitEndpointInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initEndpointParamsBeet],
  ],
  'InitEndpointInstructionArgs'
)
/**
 * Accounts required by the _initEndpoint_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [_writable_] endpoint
 * @category Instructions
 * @category InitEndpoint
 * @category generated
 */
export type InitEndpointInstructionAccounts = {
  payer: web3.PublicKey
  endpoint: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initEndpointInstructionDiscriminator = [
  178, 30, 29, 207, 120, 225, 246, 134,
]

/**
 * Creates a _InitEndpoint_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitEndpoint
 * @category generated
 */
export function createInitEndpointInstruction(
  accounts: InitEndpointInstructionAccounts,
  args: InitEndpointInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initEndpointStruct.serialize({
    instructionDiscriminator: initEndpointInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitEndpoint_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitEndpoint
 * @category generated
 */
export function createInitEndpointInstructionAccounts(
  accounts: InitEndpointInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
