// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/initDefaultReceiveLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitDefaultReceiveLibraryParams,
  initDefaultReceiveLibraryParamsBeet,
} from '../types/InitDefaultReceiveLibraryParams'

/**
 * @category Instructions
 * @category InitDefaultReceiveLibrary
 * @category generated
 */
export type InitDefaultReceiveLibraryInstructionArgs = {
  params: InitDefaultReceiveLibraryParams
}
/**
 * @category Instructions
 * @category InitDefaultReceiveLibrary
 * @category generated
 */
export const initDefaultReceiveLibraryStruct = new beet.BeetArgsStruct<
  InitDefaultReceiveLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initDefaultReceiveLibraryParamsBeet],
  ],
  'InitDefaultReceiveLibraryInstructionArgs'
)
/**
 * Accounts required by the _initDefaultReceiveLibrary_ instruction
 *
 * @property [_writable_, **signer**] admin
 * @property [] endpoint
 * @property [_writable_] defaultReceiveLibraryConfig
 * @property [] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category InitDefaultReceiveLibrary
 * @category generated
 */
export type InitDefaultReceiveLibraryInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  defaultReceiveLibraryConfig: web3.PublicKey
  messageLibInfo: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initDefaultReceiveLibraryInstructionDiscriminator = [
  32, 202, 76, 22, 42, 249, 227, 109,
]

/**
 * Creates a _InitDefaultReceiveLibrary_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitDefaultReceiveLibrary
 * @category generated
 */
export function createInitDefaultReceiveLibraryInstruction(
  accounts: InitDefaultReceiveLibraryInstructionAccounts,
  args: InitDefaultReceiveLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initDefaultReceiveLibraryStruct.serialize({
    instructionDiscriminator: initDefaultReceiveLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitDefaultReceiveLibrary_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitDefaultReceiveLibrary
 * @category generated
 */
export function createInitDefaultReceiveLibraryInstructionAccounts(
  accounts: InitDefaultReceiveLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
