// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/initDefaultSendLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitDefaultSendLibraryParams,
  initDefaultSendLibraryParamsBeet,
} from '../types/InitDefaultSendLibraryParams'

/**
 * @category Instructions
 * @category InitDefaultSendLibrary
 * @category generated
 */
export type InitDefaultSendLibraryInstructionArgs = {
  params: InitDefaultSendLibraryParams
}
/**
 * @category Instructions
 * @category InitDefaultSendLibrary
 * @category generated
 */
export const initDefaultSendLibraryStruct = new beet.BeetArgsStruct<
  InitDefaultSendLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initDefaultSendLibraryParamsBeet],
  ],
  'InitDefaultSendLibraryInstructionArgs'
)
/**
 * Accounts required by the _initDefaultSendLibrary_ instruction
 *
 * @property [_writable_, **signer**] admin
 * @property [] endpoint
 * @property [_writable_] defaultSendLibraryConfig
 * @property [] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category InitDefaultSendLibrary
 * @category generated
 */
export type InitDefaultSendLibraryInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  defaultSendLibraryConfig: web3.PublicKey
  messageLibInfo: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initDefaultSendLibraryInstructionDiscriminator = [
  120, 187, 15, 31, 174, 97, 138, 58,
]

/**
 * Creates a _InitDefaultSendLibrary_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitDefaultSendLibrary
 * @category generated
 */
export function createInitDefaultSendLibraryInstruction(
  accounts: InitDefaultSendLibraryInstructionAccounts,
  args: InitDefaultSendLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initDefaultSendLibraryStruct.serialize({
    instructionDiscriminator: initDefaultSendLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultSendLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitDefaultSendLibrary_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitDefaultSendLibrary
 * @category generated
 */
export function createInitDefaultSendLibraryInstructionAccounts(
  accounts: InitDefaultSendLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultSendLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
