// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/lzComposeAlert.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  LzComposeAlertParams,
  lzComposeAlertParamsBeet,
} from '../types/LzComposeAlertParams'

/**
 * @category Instructions
 * @category LzComposeAlert
 * @category generated
 */
export type LzComposeAlertInstructionArgs = {
  params: LzComposeAlertParams
}
/**
 * @category Instructions
 * @category LzComposeAlert
 * @category generated
 */
export const lzComposeAlertStruct = new beet.FixableBeetArgsStruct<
  LzComposeAlertInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', lzComposeAlertParamsBeet],
  ],
  'LzComposeAlertInstructionArgs'
)
/**
 * Accounts required by the _lzComposeAlert_ instruction
 *
 * @property [**signer**] executor
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category LzComposeAlert
 * @category generated
 */
export type LzComposeAlertInstructionAccounts = {
  executor: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const lzComposeAlertInstructionDiscriminator = [
  27, 91, 198, 77, 66, 92, 122, 167,
]

/**
 * Creates a _LzComposeAlert_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category LzComposeAlert
 * @category generated
 */
export function createLzComposeAlertInstruction(
  accounts: LzComposeAlertInstructionAccounts,
  args: LzComposeAlertInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = lzComposeAlertStruct.serialize({
    instructionDiscriminator: lzComposeAlertInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _LzComposeAlert_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category LzComposeAlert
 * @category generated
 */
export function createLzComposeAlertInstructionAccounts(
  accounts: LzComposeAlertInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
