// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/lzReceiveAlert.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  LzReceiveAlertParams,
  lzReceiveAlertParamsBeet,
} from '../types/LzReceiveAlertParams'

/**
 * @category Instructions
 * @category LzReceiveAlert
 * @category generated
 */
export type LzReceiveAlertInstructionArgs = {
  params: LzReceiveAlertParams
}
/**
 * @category Instructions
 * @category LzReceiveAlert
 * @category generated
 */
export const lzReceiveAlertStruct = new beet.FixableBeetArgsStruct<
  LzReceiveAlertInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', lzReceiveAlertParamsBeet],
  ],
  'LzReceiveAlertInstructionArgs'
)
/**
 * Accounts required by the _lzReceiveAlert_ instruction
 *
 * @property [**signer**] executor
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category LzReceiveAlert
 * @category generated
 */
export type LzReceiveAlertInstructionAccounts = {
  executor: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const lzReceiveAlertInstructionDiscriminator = [
  131, 141, 48, 222, 15, 235, 141, 160,
]

/**
 * Creates a _LzReceiveAlert_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category LzReceiveAlert
 * @category generated
 */
export function createLzReceiveAlertInstruction(
  accounts: LzReceiveAlertInstructionAccounts,
  args: LzReceiveAlertInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = lzReceiveAlertStruct.serialize({
    instructionDiscriminator: lzReceiveAlertInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _LzReceiveAlert_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category LzReceiveAlert
 * @category generated
 */
export function createLzReceiveAlertInstructionAccounts(
  accounts: LzReceiveAlertInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
