// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/nilify.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { NilifyParams, nilifyParamsBeet } from '../types/NilifyParams'

/**
 * @category Instructions
 * @category Nilify
 * @category generated
 */
export type NilifyInstructionArgs = {
  params: NilifyParams
}
/**
 * @category Instructions
 * @category Nilify
 * @category generated
 */
export const nilifyStruct = new beet.BeetArgsStruct<
  NilifyInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', nilifyParamsBeet],
  ],
  'NilifyInstructionArgs'
)
/**
 * Accounts required by the _nilify_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [_writable_] nonce
 * @property [_writable_] pendingInboundNonce
 * @property [_writable_] payloadHash
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Nilify
 * @category generated
 */
export type NilifyInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  nonce: web3.PublicKey
  pendingInboundNonce: web3.PublicKey
  payloadHash: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const nilifyInstructionDiscriminator = [
  143, 136, 129, 199, 36, 35, 160, 85,
]

/**
 * Creates a _Nilify_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Nilify
 * @category generated
 */
export function createNilifyInstruction(
  accounts: NilifyInstructionAccounts,
  args: NilifyInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = nilifyStruct.serialize({
    instructionDiscriminator: nilifyInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.pendingInboundNonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Nilify_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Nilify
 * @category generated
 */
export function createNilifyInstructionAccounts(
  accounts: NilifyInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.pendingInboundNonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
