// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/registerOapp.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  RegisterOAppParams,
  registerOAppParamsBeet,
} from '../types/RegisterOAppParams'

/**
 * @category Instructions
 * @category RegisterOapp
 * @category generated
 */
export type RegisterOappInstructionArgs = {
  params: RegisterOAppParams
}
/**
 * @category Instructions
 * @category RegisterOapp
 * @category generated
 */
export const registerOappStruct = new beet.BeetArgsStruct<
  RegisterOappInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', registerOAppParamsBeet],
  ],
  'RegisterOappInstructionArgs'
)
/**
 * Accounts required by the _registerOapp_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [**signer**] oapp
 * @property [_writable_] oappRegistry
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category RegisterOapp
 * @category generated
 */
export type RegisterOappInstructionAccounts = {
  payer: web3.PublicKey
  oapp: web3.PublicKey
  oappRegistry: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const registerOappInstructionDiscriminator = [
  129, 89, 71, 68, 11, 82, 210, 125,
]

/**
 * Creates a _RegisterOapp_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category RegisterOapp
 * @category generated
 */
export function createRegisterOappInstruction(
  accounts: RegisterOappInstructionAccounts,
  args: RegisterOappInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = registerOappStruct.serialize({
    instructionDiscriminator: registerOappInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.oapp,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _RegisterOapp_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category RegisterOapp
 * @category generated
 */
export function createRegisterOappInstructionAccounts(
  accounts: RegisterOappInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.oapp,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
