// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/registerLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  RegisterLibraryParams,
  registerLibraryParamsBeet,
} from '../types/RegisterLibraryParams'

/**
 * @category Instructions
 * @category RegisterLibrary
 * @category generated
 */
export type RegisterLibraryInstructionArgs = {
  params: RegisterLibraryParams
}
/**
 * @category Instructions
 * @category RegisterLibrary
 * @category generated
 */
export const registerLibraryStruct = new beet.BeetArgsStruct<
  RegisterLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', registerLibraryParamsBeet],
  ],
  'RegisterLibraryInstructionArgs'
)
/**
 * Accounts required by the _registerLibrary_ instruction
 *
 * @property [_writable_, **signer**] admin
 * @property [] endpoint
 * @property [_writable_] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category RegisterLibrary
 * @category generated
 */
export type RegisterLibraryInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  messageLibInfo: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const registerLibraryInstructionDiscriminator = [
  23, 171, 28, 116, 111, 193, 238, 142,
]

/**
 * Creates a _RegisterLibrary_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category RegisterLibrary
 * @category generated
 */
export function createRegisterLibraryInstruction(
  accounts: RegisterLibraryInstructionAccounts,
  args: RegisterLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = registerLibraryStruct.serialize({
    instructionDiscriminator: registerLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _RegisterLibrary_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category RegisterLibrary
 * @category generated
 */
export function createRegisterLibraryInstructionAccounts(
  accounts: RegisterLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
