// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/sendCompose.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SendComposeParams,
  sendComposeParamsBeet,
} from '../types/SendComposeParams'

/**
 * @category Instructions
 * @category SendCompose
 * @category generated
 */
export type SendComposeInstructionArgs = {
  params: SendComposeParams
}
/**
 * @category Instructions
 * @category SendCompose
 * @category generated
 */
export const sendComposeStruct = new beet.FixableBeetArgsStruct<
  SendComposeInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', sendComposeParamsBeet],
  ],
  'SendComposeInstructionArgs'
)
/**
 * Accounts required by the _sendCompose_ instruction
 *
 * @property [**signer**] from
 * @property [_writable_, **signer**] payer
 * @property [_writable_] composeMessage
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SendCompose
 * @category generated
 */
export type SendComposeInstructionAccounts = {
  from: web3.PublicKey
  payer: web3.PublicKey
  composeMessage: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const sendComposeInstructionDiscriminator = [
  75, 38, 228, 168, 43, 39, 238, 229,
]

/**
 * Creates a _SendCompose_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SendCompose
 * @category generated
 */
export function createSendComposeInstruction(
  accounts: SendComposeInstructionAccounts,
  args: SendComposeInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = sendComposeStruct.serialize({
    instructionDiscriminator: sendComposeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.from,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.composeMessage,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SendCompose_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SendCompose
 * @category generated
 */
export function createSendComposeInstructionAccounts(
  accounts: SendComposeInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.from,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.composeMessage,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
