// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setDefaultSendLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetDefaultSendLibraryParams,
  setDefaultSendLibraryParamsBeet,
} from '../types/SetDefaultSendLibraryParams'

/**
 * @category Instructions
 * @category SetDefaultSendLibrary
 * @category generated
 */
export type SetDefaultSendLibraryInstructionArgs = {
  params: SetDefaultSendLibraryParams
}
/**
 * @category Instructions
 * @category SetDefaultSendLibrary
 * @category generated
 */
export const setDefaultSendLibraryStruct = new beet.BeetArgsStruct<
  SetDefaultSendLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setDefaultSendLibraryParamsBeet],
  ],
  'SetDefaultSendLibraryInstructionArgs'
)
/**
 * Accounts required by the _setDefaultSendLibrary_ instruction
 *
 * @property [**signer**] admin
 * @property [] endpoint
 * @property [_writable_] defaultSendLibraryConfig
 * @property [] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetDefaultSendLibrary
 * @category generated
 */
export type SetDefaultSendLibraryInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  defaultSendLibraryConfig: web3.PublicKey
  messageLibInfo: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setDefaultSendLibraryInstructionDiscriminator = [
  220, 215, 110, 127, 237, 178, 215, 170,
]

/**
 * Creates a _SetDefaultSendLibrary_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetDefaultSendLibrary
 * @category generated
 */
export function createSetDefaultSendLibraryInstruction(
  accounts: SetDefaultSendLibraryInstructionAccounts,
  args: SetDefaultSendLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setDefaultSendLibraryStruct.serialize({
    instructionDiscriminator: setDefaultSendLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultSendLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetDefaultSendLibrary_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetDefaultSendLibrary
 * @category generated
 */
export function createSetDefaultSendLibraryInstructionAccounts(
  accounts: SetDefaultSendLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultSendLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
