// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setDefaultReceiveLibraryTimeout.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetDefaultReceiveLibraryTimeoutParams,
  setDefaultReceiveLibraryTimeoutParamsBeet,
} from '../types/SetDefaultReceiveLibraryTimeoutParams'

/**
 * @category Instructions
 * @category SetDefaultReceiveLibraryTimeout
 * @category generated
 */
export type SetDefaultReceiveLibraryTimeoutInstructionArgs = {
  params: SetDefaultReceiveLibraryTimeoutParams
}
/**
 * @category Instructions
 * @category SetDefaultReceiveLibraryTimeout
 * @category generated
 */
export const setDefaultReceiveLibraryTimeoutStruct = new beet.BeetArgsStruct<
  SetDefaultReceiveLibraryTimeoutInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setDefaultReceiveLibraryTimeoutParamsBeet],
  ],
  'SetDefaultReceiveLibraryTimeoutInstructionArgs'
)
/**
 * Accounts required by the _setDefaultReceiveLibraryTimeout_ instruction
 *
 * @property [**signer**] admin
 * @property [] endpoint
 * @property [_writable_] defaultReceiveLibraryConfig
 * @property [] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetDefaultReceiveLibraryTimeout
 * @category generated
 */
export type SetDefaultReceiveLibraryTimeoutInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  defaultReceiveLibraryConfig: web3.PublicKey
  messageLibInfo: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setDefaultReceiveLibraryTimeoutInstructionDiscriminator = [
  118, 32, 94, 51, 25, 247, 16, 232,
]

/**
 * Creates a _SetDefaultReceiveLibraryTimeout_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetDefaultReceiveLibraryTimeout
 * @category generated
 */
export function createSetDefaultReceiveLibraryTimeoutInstruction(
  accounts: SetDefaultReceiveLibraryTimeoutInstructionAccounts,
  args: SetDefaultReceiveLibraryTimeoutInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setDefaultReceiveLibraryTimeoutStruct.serialize({
    instructionDiscriminator:
      setDefaultReceiveLibraryTimeoutInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetDefaultReceiveLibraryTimeout_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetDefaultReceiveLibraryTimeout
 * @category generated
 */
export function createSetDefaultReceiveLibraryTimeoutInstructionAccounts(
  accounts: SetDefaultReceiveLibraryTimeoutInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
