// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setDefaultReceiveLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetDefaultReceiveLibraryParams,
  setDefaultReceiveLibraryParamsBeet,
} from '../types/SetDefaultReceiveLibraryParams'

/**
 * @category Instructions
 * @category SetDefaultReceiveLibrary
 * @category generated
 */
export type SetDefaultReceiveLibraryInstructionArgs = {
  params: SetDefaultReceiveLibraryParams
}
/**
 * @category Instructions
 * @category SetDefaultReceiveLibrary
 * @category generated
 */
export const setDefaultReceiveLibraryStruct = new beet.BeetArgsStruct<
  SetDefaultReceiveLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setDefaultReceiveLibraryParamsBeet],
  ],
  'SetDefaultReceiveLibraryInstructionArgs'
)
/**
 * Accounts required by the _setDefaultReceiveLibrary_ instruction
 *
 * @property [**signer**] admin
 * @property [] endpoint
 * @property [_writable_] defaultReceiveLibraryConfig
 * @property [] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetDefaultReceiveLibrary
 * @category generated
 */
export type SetDefaultReceiveLibraryInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  defaultReceiveLibraryConfig: web3.PublicKey
  messageLibInfo: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setDefaultReceiveLibraryInstructionDiscriminator = [
  14, 162, 167, 212, 13, 20, 151, 129,
]

/**
 * Creates a _SetDefaultReceiveLibrary_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetDefaultReceiveLibrary
 * @category generated
 */
export function createSetDefaultReceiveLibraryInstruction(
  accounts: SetDefaultReceiveLibraryInstructionAccounts,
  args: SetDefaultReceiveLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setDefaultReceiveLibraryStruct.serialize({
    instructionDiscriminator: setDefaultReceiveLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetDefaultReceiveLibrary_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetDefaultReceiveLibrary
 * @category generated
 */
export function createSetDefaultReceiveLibraryInstructionAccounts(
  accounts: SetDefaultReceiveLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.defaultReceiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
