// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setDelegate.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetDelegateParams,
  setDelegateParamsBeet,
} from '../types/SetDelegateParams'

/**
 * @category Instructions
 * @category SetDelegate
 * @category generated
 */
export type SetDelegateInstructionArgs = {
  params: SetDelegateParams
}
/**
 * @category Instructions
 * @category SetDelegate
 * @category generated
 */
export const setDelegateStruct = new beet.BeetArgsStruct<
  SetDelegateInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setDelegateParamsBeet],
  ],
  'SetDelegateInstructionArgs'
)
/**
 * Accounts required by the _setDelegate_ instruction
 *
 * @property [**signer**] oapp
 * @property [_writable_] oappRegistry
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetDelegate
 * @category generated
 */
export type SetDelegateInstructionAccounts = {
  oapp: web3.PublicKey
  oappRegistry: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setDelegateInstructionDiscriminator = [
  242, 30, 46, 76, 108, 235, 128, 181,
]

/**
 * Creates a _SetDelegate_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetDelegate
 * @category generated
 */
export function createSetDelegateInstruction(
  accounts: SetDelegateInstructionAccounts,
  args: SetDelegateInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setDelegateStruct.serialize({
    instructionDiscriminator: setDelegateInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.oapp,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetDelegate_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetDelegate
 * @category generated
 */
export function createSetDelegateInstructionAccounts(
  accounts: SetDelegateInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.oapp,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
