// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setLzToken.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetLzTokenParams,
  setLzTokenParamsBeet,
} from '../types/SetLzTokenParams'

/**
 * @category Instructions
 * @category SetLzToken
 * @category generated
 */
export type SetLzTokenInstructionArgs = {
  params: SetLzTokenParams
}
/**
 * @category Instructions
 * @category SetLzToken
 * @category generated
 */
export const setLzTokenStruct = new beet.FixableBeetArgsStruct<
  SetLzTokenInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setLzTokenParamsBeet],
  ],
  'SetLzTokenInstructionArgs'
)
/**
 * Accounts required by the _setLzToken_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] endpoint
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetLzToken
 * @category generated
 */
export type SetLzTokenInstructionAccounts = {
  admin: web3.PublicKey
  endpoint: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setLzTokenInstructionDiscriminator = [
  22, 151, 112, 174, 213, 225, 223, 72,
]

/**
 * Creates a _SetLzToken_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetLzToken
 * @category generated
 */
export function createSetLzTokenInstruction(
  accounts: SetLzTokenInstructionAccounts,
  args: SetLzTokenInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setLzTokenStruct.serialize({
    instructionDiscriminator: setLzTokenInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetLzToken_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetLzToken
 * @category generated
 */
export function createSetLzTokenInstructionAccounts(
  accounts: SetLzTokenInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
