// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setReceiveLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetReceiveLibraryParams,
  setReceiveLibraryParamsBeet,
} from '../types/SetReceiveLibraryParams'

/**
 * @category Instructions
 * @category SetReceiveLibrary
 * @category generated
 */
export type SetReceiveLibraryInstructionArgs = {
  params: SetReceiveLibraryParams
}
/**
 * @category Instructions
 * @category SetReceiveLibrary
 * @category generated
 */
export const setReceiveLibraryStruct = new beet.BeetArgsStruct<
  SetReceiveLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setReceiveLibraryParamsBeet],
  ],
  'SetReceiveLibraryInstructionArgs'
)
/**
 * Accounts required by the _setReceiveLibrary_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [_writable_] receiveLibraryConfig
 * @property [] messageLibInfo (optional)
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetReceiveLibrary
 * @category generated
 */
export type SetReceiveLibraryInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  receiveLibraryConfig: web3.PublicKey
  messageLibInfo?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setReceiveLibraryInstructionDiscriminator = [
  223, 172, 180, 105, 165, 161, 147, 228,
]

/**
 * Creates a _SetReceiveLibrary_ instruction.
 *
 * Optional accounts that are not provided default to the program ID since
 * this was indicated in the IDL from which this instruction was generated.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetReceiveLibrary
 * @category generated
 */
export function createSetReceiveLibraryInstruction(
  accounts: SetReceiveLibraryInstructionAccounts,
  args: SetReceiveLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setReceiveLibraryStruct.serialize({
    instructionDiscriminator: setReceiveLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo ?? programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetReceiveLibrary_ instructionAccounts.
 *
 * Optional accounts that are not provided default to the program ID since
 * this was indicated in the IDL from which this instruction was generated.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetReceiveLibrary
 * @category generated
 */
export function createSetReceiveLibraryInstructionAccounts(
  accounts: SetReceiveLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo ?? programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
