// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setReceiveLibraryTimeout.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetReceiveLibraryTimeoutParams,
  setReceiveLibraryTimeoutParamsBeet,
} from '../types/SetReceiveLibraryTimeoutParams'

/**
 * @category Instructions
 * @category SetReceiveLibraryTimeout
 * @category generated
 */
export type SetReceiveLibraryTimeoutInstructionArgs = {
  params: SetReceiveLibraryTimeoutParams
}
/**
 * @category Instructions
 * @category SetReceiveLibraryTimeout
 * @category generated
 */
export const setReceiveLibraryTimeoutStruct = new beet.BeetArgsStruct<
  SetReceiveLibraryTimeoutInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setReceiveLibraryTimeoutParamsBeet],
  ],
  'SetReceiveLibraryTimeoutInstructionArgs'
)
/**
 * Accounts required by the _setReceiveLibraryTimeout_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [_writable_] receiveLibraryConfig
 * @property [] messageLibInfo
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetReceiveLibraryTimeout
 * @category generated
 */
export type SetReceiveLibraryTimeoutInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  receiveLibraryConfig: web3.PublicKey
  messageLibInfo: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setReceiveLibraryTimeoutInstructionDiscriminator = [
  33, 159, 29, 53, 218, 156, 41, 192,
]

/**
 * Creates a _SetReceiveLibraryTimeout_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetReceiveLibraryTimeout
 * @category generated
 */
export function createSetReceiveLibraryTimeoutInstruction(
  accounts: SetReceiveLibraryTimeoutInstructionAccounts,
  args: SetReceiveLibraryTimeoutInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setReceiveLibraryTimeoutStruct.serialize({
    instructionDiscriminator: setReceiveLibraryTimeoutInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetReceiveLibraryTimeout_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetReceiveLibraryTimeout
 * @category generated
 */
export function createSetReceiveLibraryTimeoutInstructionAccounts(
  accounts: SetReceiveLibraryTimeoutInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.receiveLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
