// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/setSendLibrary.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  SetSendLibraryParams,
  setSendLibraryParamsBeet,
} from '../types/SetSendLibraryParams'

/**
 * @category Instructions
 * @category SetSendLibrary
 * @category generated
 */
export type SetSendLibraryInstructionArgs = {
  params: SetSendLibraryParams
}
/**
 * @category Instructions
 * @category SetSendLibrary
 * @category generated
 */
export const setSendLibraryStruct = new beet.BeetArgsStruct<
  SetSendLibraryInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setSendLibraryParamsBeet],
  ],
  'SetSendLibraryInstructionArgs'
)
/**
 * Accounts required by the _setSendLibrary_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [_writable_] sendLibraryConfig
 * @property [] messageLibInfo (optional)
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetSendLibrary
 * @category generated
 */
export type SetSendLibraryInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  sendLibraryConfig: web3.PublicKey
  messageLibInfo?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setSendLibraryInstructionDiscriminator = [
  251, 118, 78, 158, 134, 149, 129, 5,
]

/**
 * Creates a _SetSendLibrary_ instruction.
 *
 * Optional accounts that are not provided default to the program ID since
 * this was indicated in the IDL from which this instruction was generated.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetSendLibrary
 * @category generated
 */
export function createSetSendLibraryInstruction(
  accounts: SetSendLibraryInstructionAccounts,
  args: SetSendLibraryInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setSendLibraryStruct.serialize({
    instructionDiscriminator: setSendLibraryInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo ?? programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetSendLibrary_ instructionAccounts.
 *
 * Optional accounts that are not provided default to the program ID since
 * this was indicated in the IDL from which this instruction was generated.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetSendLibrary
 * @category generated
 */
export function createSetSendLibraryInstructionAccounts(
  accounts: SetSendLibraryInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.sendLibraryConfig,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.messageLibInfo ?? programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
