// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/instructions/skip.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { SkipParams, skipParamsBeet } from '../types/SkipParams'

/**
 * @category Instructions
 * @category Skip
 * @category generated
 */
export type SkipInstructionArgs = {
  params: SkipParams
}
/**
 * @category Instructions
 * @category Skip
 * @category generated
 */
export const skipStruct = new beet.BeetArgsStruct<
  SkipInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', skipParamsBeet],
  ],
  'SkipInstructionArgs'
)
/**
 * Accounts required by the _skip_ instruction
 *
 * @property [**signer**] signer
 * @property [] oappRegistry
 * @property [_writable_] nonce
 * @property [_writable_] pendingInboundNonce
 * @property [_writable_] payloadHash
 * @property [_writable_] endpoint
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Skip
 * @category generated
 */
export type SkipInstructionAccounts = {
  signer: web3.PublicKey
  oappRegistry: web3.PublicKey
  nonce: web3.PublicKey
  pendingInboundNonce: web3.PublicKey
  payloadHash: web3.PublicKey
  endpoint: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const skipInstructionDiscriminator = [154, 63, 181, 53, 19, 26, 117, 45]

/**
 * Creates a _Skip_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Skip
 * @category generated
 */
export function createSkipInstruction(
  accounts: SkipInstructionAccounts,
  args: SkipInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = skipStruct.serialize({
    instructionDiscriminator: skipInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.pendingInboundNonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Skip_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Skip
 * @category generated
 */
export function createSkipInstructionAccounts(
  accounts: SkipInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.oappRegistry,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.nonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.pendingInboundNonce,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.payloadHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.endpoint,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
