// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/simple_messagelib/errors/simpleMessagelib.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Program, ProgramError } from '@metaplex-foundation/umi';

type ProgramErrorConstructor = new (
  program: Program,
  cause?: Error
) => ProgramError;
const codeToErrorMap: Map<number, ProgramErrorConstructor> = new Map();
const nameToErrorMap: Map<string, ProgramErrorConstructor> = new Map();

/** OnlyWhitelistedCaller */
export class OnlyWhitelistedCallerError extends ProgramError {
  override readonly name: string = 'OnlyWhitelistedCaller';

  readonly code: number = 0x1770; // 6000

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1770, OnlyWhitelistedCallerError);
nameToErrorMap.set('OnlyWhitelistedCaller', OnlyWhitelistedCallerError);

/** InsufficientFee */
export class InsufficientFeeError extends ProgramError {
  override readonly name: string = 'InsufficientFee';

  readonly code: number = 0x1771; // 6001

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1771, InsufficientFeeError);
nameToErrorMap.set('InsufficientFee', InsufficientFeeError);

/** InvalidAmount */
export class InvalidAmountError extends ProgramError {
  override readonly name: string = 'InvalidAmount';

  readonly code: number = 0x1772; // 6002

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1772, InvalidAmountError);
nameToErrorMap.set('InvalidAmount', InvalidAmountError);

/** InvalidConfigType */
export class InvalidConfigTypeError extends ProgramError {
  override readonly name: string = 'InvalidConfigType';

  readonly code: number = 0x1773; // 6003

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1773, InvalidConfigTypeError);
nameToErrorMap.set('InvalidConfigType', InvalidConfigTypeError);

/** InvalidLzTokenMint */
export class InvalidLzTokenMintError extends ProgramError {
  override readonly name: string = 'InvalidLzTokenMint';

  readonly code: number = 0x1774; // 6004

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1774, InvalidLzTokenMintError);
nameToErrorMap.set('InvalidLzTokenMint', InvalidLzTokenMintError);

/** LzTokenUnavailable */
export class LzTokenUnavailableError extends ProgramError {
  override readonly name: string = 'LzTokenUnavailable';

  readonly code: number = 0x1775; // 6005

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1775, LzTokenUnavailableError);
nameToErrorMap.set('LzTokenUnavailable', LzTokenUnavailableError);

/** SendReentrancy */
export class SendReentrancyError extends ProgramError {
  override readonly name: string = 'SendReentrancy';

  readonly code: number = 0x1776; // 6006

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1776, SendReentrancyError);
nameToErrorMap.set('SendReentrancy', SendReentrancyError);

/** OnlyRevert */
export class OnlyRevertError extends ProgramError {
  override readonly name: string = 'OnlyRevert';

  readonly code: number = 0x1777; // 6007

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1777, OnlyRevertError);
nameToErrorMap.set('OnlyRevert', OnlyRevertError);

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
export function getSimpleMessagelibErrorFromCode(
  code: number,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
export function getSimpleMessagelibErrorFromName(
  name: string,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}
