// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/accounts/MessageLibInfo.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { MessageLibType, messageLibTypeBeet } from '../types/MessageLibType'

/**
 * Arguments used to create {@link MessageLibInfo}
 * @category Accounts
 * @category generated
 */
export type MessageLibInfoArgs = {
  messageLibType: MessageLibType
  bump: number
  messageLibBump: number
}

export const messageLibInfoDiscriminator = [103, 102, 218, 28, 204, 135, 71, 14]
/**
 * Holds the data for the {@link MessageLibInfo} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class MessageLibInfo implements MessageLibInfoArgs {
  private constructor(
    readonly messageLibType: MessageLibType,
    readonly bump: number,
    readonly messageLibBump: number
  ) {}

  /**
   * Creates a {@link MessageLibInfo} instance from the provided args.
   */
  static fromArgs(args: MessageLibInfoArgs) {
    return new MessageLibInfo(
      args.messageLibType,
      args.bump,
      args.messageLibBump
    )
  }

  /**
   * Deserializes the {@link MessageLibInfo} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [MessageLibInfo, number] {
    return MessageLibInfo.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link MessageLibInfo} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<MessageLibInfo> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find MessageLibInfo account at ${address}`)
    }
    return MessageLibInfo.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, messageLibInfoBeet)
  }

  /**
   * Deserializes the {@link MessageLibInfo} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [MessageLibInfo, number] {
    return messageLibInfoBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link MessageLibInfo} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return messageLibInfoBeet.serialize({
      accountDiscriminator: messageLibInfoDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link MessageLibInfo}
   */
  static get byteSize() {
    return messageLibInfoBeet.byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link MessageLibInfo} data from rent
   *
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      MessageLibInfo.byteSize,
      commitment
    )
  }

  /**
   * Determines if the provided {@link Buffer} has the correct byte size to
   * hold {@link MessageLibInfo} data.
   */
  static hasCorrectByteSize(buf: Buffer, offset = 0) {
    return buf.byteLength - offset === MessageLibInfo.byteSize
  }

  /**
   * Returns a readable version of {@link MessageLibInfo} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      messageLibType: 'MessageLibType.' + MessageLibType[this.messageLibType],
      bump: this.bump,
      messageLibBump: this.messageLibBump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const messageLibInfoBeet = new beet.BeetStruct<
  MessageLibInfo,
  MessageLibInfoArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['messageLibType', messageLibTypeBeet],
    ['bump', beet.u8],
    ['messageLibBump', beet.u8],
  ],
  MessageLibInfo.fromArgs,
  'MessageLibInfo'
)
