// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/accounts/OAppRegistry.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import * as beet from '@metaplex-foundation/beet'

/**
 * Arguments used to create {@link OAppRegistry}
 * @category Accounts
 * @category generated
 */
export type OAppRegistryArgs = {
  delegate: web3.PublicKey
  bump: number
}

export const oAppRegistryDiscriminator = [6, 152, 199, 30, 217, 50, 69, 149]
/**
 * Holds the data for the {@link OAppRegistry} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class OAppRegistry implements OAppRegistryArgs {
  private constructor(
    readonly delegate: web3.PublicKey,
    readonly bump: number
  ) {}

  /**
   * Creates a {@link OAppRegistry} instance from the provided args.
   */
  static fromArgs(args: OAppRegistryArgs) {
    return new OAppRegistry(args.delegate, args.bump)
  }

  /**
   * Deserializes the {@link OAppRegistry} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [OAppRegistry, number] {
    return OAppRegistry.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link OAppRegistry} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<OAppRegistry> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find OAppRegistry account at ${address}`)
    }
    return OAppRegistry.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, oAppRegistryBeet)
  }

  /**
   * Deserializes the {@link OAppRegistry} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [OAppRegistry, number] {
    return oAppRegistryBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link OAppRegistry} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return oAppRegistryBeet.serialize({
      accountDiscriminator: oAppRegistryDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link OAppRegistry}
   */
  static get byteSize() {
    return oAppRegistryBeet.byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link OAppRegistry} data from rent
   *
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      OAppRegistry.byteSize,
      commitment
    )
  }

  /**
   * Determines if the provided {@link Buffer} has the correct byte size to
   * hold {@link OAppRegistry} data.
   */
  static hasCorrectByteSize(buf: Buffer, offset = 0) {
    return buf.byteLength - offset === OAppRegistry.byteSize
  }

  /**
   * Returns a readable version of {@link OAppRegistry} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      delegate: this.delegate.toBase58(),
      bump: this.bump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const oAppRegistryBeet = new beet.BeetStruct<
  OAppRegistry,
  OAppRegistryArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['delegate', beetSolana.publicKey],
    ['bump', beet.u8],
  ],
  OAppRegistry.fromArgs,
  'OAppRegistry'
)
