// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/accounts/ReceiveLibraryConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import {
  ReceiveLibraryTimeout,
  receiveLibraryTimeoutBeet,
} from '../types/ReceiveLibraryTimeout'

/**
 * Arguments used to create {@link ReceiveLibraryConfig}
 * @category Accounts
 * @category generated
 */
export type ReceiveLibraryConfigArgs = {
  messageLib: web3.PublicKey
  timeout: beet.COption<ReceiveLibraryTimeout>
  bump: number
}

export const receiveLibraryConfigDiscriminator = [
  142, 226, 251, 138, 1, 206, 91, 193,
]
/**
 * Holds the data for the {@link ReceiveLibraryConfig} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class ReceiveLibraryConfig implements ReceiveLibraryConfigArgs {
  private constructor(
    readonly messageLib: web3.PublicKey,
    readonly timeout: beet.COption<ReceiveLibraryTimeout>,
    readonly bump: number
  ) {}

  /**
   * Creates a {@link ReceiveLibraryConfig} instance from the provided args.
   */
  static fromArgs(args: ReceiveLibraryConfigArgs) {
    return new ReceiveLibraryConfig(args.messageLib, args.timeout, args.bump)
  }

  /**
   * Deserializes the {@link ReceiveLibraryConfig} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [ReceiveLibraryConfig, number] {
    return ReceiveLibraryConfig.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link ReceiveLibraryConfig} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<ReceiveLibraryConfig> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(
        `Unable to find ReceiveLibraryConfig account at ${address}`
      )
    }
    return ReceiveLibraryConfig.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, receiveLibraryConfigBeet)
  }

  /**
   * Deserializes the {@link ReceiveLibraryConfig} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [ReceiveLibraryConfig, number] {
    return receiveLibraryConfigBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link ReceiveLibraryConfig} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return receiveLibraryConfigBeet.serialize({
      accountDiscriminator: receiveLibraryConfigDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link ReceiveLibraryConfig} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: ReceiveLibraryConfigArgs) {
    const instance = ReceiveLibraryConfig.fromArgs(args)
    return receiveLibraryConfigBeet.toFixedFromValue({
      accountDiscriminator: receiveLibraryConfigDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link ReceiveLibraryConfig} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: ReceiveLibraryConfigArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      ReceiveLibraryConfig.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link ReceiveLibraryConfig} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      messageLib: this.messageLib.toBase58(),
      timeout: this.timeout,
      bump: this.bump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const receiveLibraryConfigBeet = new beet.FixableBeetStruct<
  ReceiveLibraryConfig,
  ReceiveLibraryConfigArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['messageLib', beetSolana.publicKey],
    ['timeout', beet.coption(receiveLibraryTimeoutBeet)],
    ['bump', beet.u8],
  ],
  ReceiveLibraryConfig.fromArgs,
  'ReceiveLibraryConfig'
)
