// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/accounts/SendLibraryConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import * as beet from '@metaplex-foundation/beet'

/**
 * Arguments used to create {@link SendLibraryConfig}
 * @category Accounts
 * @category generated
 */
export type SendLibraryConfigArgs = {
  messageLib: web3.PublicKey
  bump: number
}

export const sendLibraryConfigDiscriminator = [
  61, 238, 31, 72, 251, 117, 66, 176,
]
/**
 * Holds the data for the {@link SendLibraryConfig} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class SendLibraryConfig implements SendLibraryConfigArgs {
  private constructor(
    readonly messageLib: web3.PublicKey,
    readonly bump: number
  ) {}

  /**
   * Creates a {@link SendLibraryConfig} instance from the provided args.
   */
  static fromArgs(args: SendLibraryConfigArgs) {
    return new SendLibraryConfig(args.messageLib, args.bump)
  }

  /**
   * Deserializes the {@link SendLibraryConfig} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [SendLibraryConfig, number] {
    return SendLibraryConfig.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link SendLibraryConfig} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<SendLibraryConfig> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find SendLibraryConfig account at ${address}`)
    }
    return SendLibraryConfig.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '76y77prsiCMvXMjuoZ5VRrhG5qYBrUMYTE5WgHqgjEn6'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, sendLibraryConfigBeet)
  }

  /**
   * Deserializes the {@link SendLibraryConfig} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [SendLibraryConfig, number] {
    return sendLibraryConfigBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link SendLibraryConfig} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return sendLibraryConfigBeet.serialize({
      accountDiscriminator: sendLibraryConfigDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link SendLibraryConfig}
   */
  static get byteSize() {
    return sendLibraryConfigBeet.byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link SendLibraryConfig} data from rent
   *
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      SendLibraryConfig.byteSize,
      commitment
    )
  }

  /**
   * Determines if the provided {@link Buffer} has the correct byte size to
   * hold {@link SendLibraryConfig} data.
   */
  static hasCorrectByteSize(buf: Buffer, offset = 0) {
    return buf.byteLength - offset === SendLibraryConfig.byteSize
  }

  /**
   * Returns a readable version of {@link SendLibraryConfig} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      messageLib: this.messageLib.toBase58(),
      bump: this.bump,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const sendLibraryConfigBeet = new beet.BeetStruct<
  SendLibraryConfig,
  SendLibraryConfigArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['messageLib', beetSolana.publicKey],
    ['bump', beet.u8],
  ],
  SendLibraryConfig.fromArgs,
  'SendLibraryConfig'
)
