// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/instructions/adminSetConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  AdminSetConfigParams,
  adminSetConfigParamsBeet,
} from '../types/AdminSetConfigParams'

/**
 * @category Instructions
 * @category AdminSetConfig
 * @category generated
 */
export type AdminSetConfigInstructionArgs = {
  params: AdminSetConfigParams
}
/**
 * @category Instructions
 * @category AdminSetConfig
 * @category generated
 */
export const adminSetConfigStruct = new beet.FixableBeetArgsStruct<
  AdminSetConfigInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', adminSetConfigParamsBeet],
  ],
  'AdminSetConfigInstructionArgs'
)
/**
 * Accounts required by the _adminSetConfig_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] config
 * @category Instructions
 * @category AdminSetConfig
 * @category generated
 */
export type AdminSetConfigInstructionAccounts = {
  admin: web3.PublicKey
  config: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const adminSetConfigInstructionDiscriminator = [
  35, 56, 160, 84, 132, 156, 27, 79,
]

/**
 * Creates a _AdminSetConfig_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category AdminSetConfig
 * @category generated
 */
export function createAdminSetConfigInstruction(
  accounts: AdminSetConfigInstructionAccounts,
  args: AdminSetConfigInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = adminSetConfigStruct.serialize({
    instructionDiscriminator: adminSetConfigInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _AdminSetConfig_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category AdminSetConfig
 * @category generated
 */
export function createAdminSetConfigInstructionAccounts(
  accounts: AdminSetConfigInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
