// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/instructions/compose.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { ComposeParams, composeParamsBeet } from '../types/ComposeParams'

/**
 * @category Instructions
 * @category Compose
 * @category generated
 */
export type ComposeInstructionArgs = {
  params: ComposeParams
}
/**
 * @category Instructions
 * @category Compose
 * @category generated
 */
export const composeStruct = new beet.FixableBeetArgsStruct<
  ComposeInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', composeParamsBeet],
  ],
  'ComposeInstructionArgs'
)
/**
 * Accounts required by the _compose_ instruction
 *
 * @property [_writable_, **signer**] executor
 * @property [] config
 * @property [] endpointProgram
 * @property [] endpointEventAuthority
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Compose
 * @category generated
 */
export type ComposeInstructionAccounts = {
  executor: web3.PublicKey
  config: web3.PublicKey
  endpointProgram: web3.PublicKey
  endpointEventAuthority: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const composeInstructionDiscriminator = [
  106, 64, 131, 142, 7, 159, 42, 15,
]

/**
 * Creates a _Compose_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Compose
 * @category generated
 */
export function createComposeInstruction(
  accounts: ComposeInstructionAccounts,
  args: ComposeInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = composeStruct.serialize({
    instructionDiscriminator: composeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointEventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Compose_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Compose
 * @category generated
 */
export function createComposeInstructionAccounts(
  accounts: ComposeInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointEventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
