// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/instructions/execute.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { ExecuteParams, executeParamsBeet } from '../types/ExecuteParams'

/**
 * @category Instructions
 * @category Execute
 * @category generated
 */
export type ExecuteInstructionArgs = {
  params: ExecuteParams
}
/**
 * @category Instructions
 * @category Execute
 * @category generated
 */
export const executeStruct = new beet.FixableBeetArgsStruct<
  ExecuteInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', executeParamsBeet],
  ],
  'ExecuteInstructionArgs'
)
/**
 * Accounts required by the _execute_ instruction
 *
 * @property [_writable_, **signer**] executor
 * @property [] config
 * @property [] endpointProgram
 * @property [] endpointEventAuthority
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Execute
 * @category generated
 */
export type ExecuteInstructionAccounts = {
  executor: web3.PublicKey
  config: web3.PublicKey
  endpointProgram: web3.PublicKey
  endpointEventAuthority: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const executeInstructionDiscriminator = [
  130, 221, 242, 154, 13, 193, 189, 29,
]

/**
 * Creates a _Execute_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Execute
 * @category generated
 */
export function createExecuteInstruction(
  accounts: ExecuteInstructionAccounts,
  args: ExecuteInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = executeStruct.serialize({
    instructionDiscriminator: executeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointEventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Execute_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Execute
 * @category generated
 */
export function createExecuteInstructionAccounts(
  accounts: ExecuteInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.endpointEventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
