// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/instructions/initExecutor.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  InitExecutorParams,
  initExecutorParamsBeet,
} from '../types/InitExecutorParams'

/**
 * @category Instructions
 * @category InitExecutor
 * @category generated
 */
export type InitExecutorInstructionArgs = {
  params: InitExecutorParams
}
/**
 * @category Instructions
 * @category InitExecutor
 * @category generated
 */
export const initExecutorStruct = new beet.FixableBeetArgsStruct<
  InitExecutorInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initExecutorParamsBeet],
  ],
  'InitExecutorInstructionArgs'
)
/**
 * Accounts required by the _initExecutor_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [_writable_] config
 * @category Instructions
 * @category InitExecutor
 * @category generated
 */
export type InitExecutorInstructionAccounts = {
  payer: web3.PublicKey
  config: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initExecutorInstructionDiscriminator = [
  195, 68, 116, 173, 55, 173, 159, 31,
]

/**
 * Creates a _InitExecutor_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitExecutor
 * @category generated
 */
export function createInitExecutorInstruction(
  accounts: InitExecutorInstructionAccounts,
  args: InitExecutorInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initExecutorStruct.serialize({
    instructionDiscriminator: initExecutorInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitExecutor_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitExecutor
 * @category generated
 */
export function createInitExecutorInstructionAccounts(
  accounts: InitExecutorInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
