// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/instructions/nativeDrop.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  NativeDropParams,
  nativeDropParamsBeet,
} from '../types/NativeDropParams'

/**
 * @category Instructions
 * @category NativeDrop
 * @category generated
 */
export type NativeDropInstructionArgs = {
  params: NativeDropParams
}
/**
 * @category Instructions
 * @category NativeDrop
 * @category generated
 */
export const nativeDropStruct = new beet.FixableBeetArgsStruct<
  NativeDropInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', nativeDropParamsBeet],
  ],
  'NativeDropInstructionArgs'
)
/**
 * Accounts required by the _nativeDrop_ instruction
 *
 * @property [_writable_, **signer**] executor
 * @property [_writable_] config
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category NativeDrop
 * @category generated
 */
export type NativeDropInstructionAccounts = {
  executor: web3.PublicKey
  config: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const nativeDropInstructionDiscriminator = [
  105, 30, 218, 121, 155, 234, 85, 12,
]

/**
 * Creates a _NativeDrop_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category NativeDrop
 * @category generated
 */
export function createNativeDropInstruction(
  accounts: NativeDropInstructionAccounts,
  args: NativeDropInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = nativeDropStruct.serialize({
    instructionDiscriminator: nativeDropInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _NativeDrop_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category NativeDrop
 * @category generated
 */
export function createNativeDropInstructionAccounts(
  accounts: NativeDropInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.executor,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
