// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/types/config.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  GetDataEnumKind,
  GetDataEnumKindContent,
  Serializer,
  dataEnum,
  struct,
  tuple,
} from '@metaplex-foundation/umi/serializers';
import {
  ExecutorConfig,
  ExecutorConfigArgs,
  UlnConfig,
  UlnConfigArgs,
  getExecutorConfigSerializer,
  getUlnConfigSerializer,
} from '.';

export type Config =
  | { __kind: 'SendUln'; fields: [UlnConfig] }
  | { __kind: 'ReceiveUln'; fields: [UlnConfig] }
  | { __kind: 'Executor'; fields: [ExecutorConfig] };

export type ConfigArgs =
  | { __kind: 'SendUln'; fields: [UlnConfigArgs] }
  | { __kind: 'ReceiveUln'; fields: [UlnConfigArgs] }
  | { __kind: 'Executor'; fields: [ExecutorConfigArgs] };

export function getConfigSerializer(): Serializer<ConfigArgs, Config> {
  return dataEnum<Config>(
    [
      [
        'SendUln',
        struct<GetDataEnumKindContent<Config, 'SendUln'>>([
          ['fields', tuple([getUlnConfigSerializer()])],
        ]),
      ],
      [
        'ReceiveUln',
        struct<GetDataEnumKindContent<Config, 'ReceiveUln'>>([
          ['fields', tuple([getUlnConfigSerializer()])],
        ]),
      ],
      [
        'Executor',
        struct<GetDataEnumKindContent<Config, 'Executor'>>([
          ['fields', tuple([getExecutorConfigSerializer()])],
        ]),
      ],
    ],
    { description: 'Config' }
  ) as Serializer<ConfigArgs, Config>;
}

// Data Enum Helpers.
export function config(
  kind: 'SendUln',
  data: GetDataEnumKindContent<ConfigArgs, 'SendUln'>['fields']
): GetDataEnumKind<ConfigArgs, 'SendUln'>;
export function config(
  kind: 'ReceiveUln',
  data: GetDataEnumKindContent<ConfigArgs, 'ReceiveUln'>['fields']
): GetDataEnumKind<ConfigArgs, 'ReceiveUln'>;
export function config(
  kind: 'Executor',
  data: GetDataEnumKindContent<ConfigArgs, 'Executor'>['fields']
): GetDataEnumKind<ConfigArgs, 'Executor'>;
export function config<K extends ConfigArgs['__kind']>(
  kind: K,
  data?: any
): Extract<ConfigArgs, { __kind: K }> {
  return Array.isArray(data)
    ? { __kind: kind, fields: data }
    : { __kind: kind, ...(data ?? {}) };
}
export function isConfig<K extends Config['__kind']>(
  kind: K,
  value: Config
): value is Config & { __kind: K } {
  return value.__kind === kind;
}
