// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/types/packet.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';

export type Packet = {
  nonce: bigint;
  srcEid: number;
  sender: PublicKey;
  dstEid: number;
  receiver: Uint8Array;
  guid: Uint8Array;
  message: Uint8Array;
};

export type PacketArgs = {
  nonce: number | bigint;
  srcEid: number;
  sender: PublicKey;
  dstEid: number;
  receiver: Uint8Array;
  guid: Uint8Array;
  message: Uint8Array;
};

export function getPacketSerializer(): Serializer<PacketArgs, Packet> {
  return struct<Packet>(
    [
      ['nonce', u64()],
      ['srcEid', u32()],
      ['sender', publicKeySerializer()],
      ['dstEid', u32()],
      ['receiver', bytes({ size: 32 })],
      ['guid', bytes({ size: 32 })],
      ['message', bytes({ size: u32() })],
    ],
    { description: 'Packet' }
  ) as Serializer<PacketArgs, Packet>;
}
