// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/uln/types/treasury.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable, PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  option,
  publicKey as publicKeySerializer,
  struct,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  LzTokenTreasury,
  LzTokenTreasuryArgs,
  getLzTokenTreasurySerializer,
} from '.';

export type Treasury = {
  admin: Option<PublicKey>;
  nativeReceiver: PublicKey;
  nativeFeeBps: bigint;
  lzToken: Option<LzTokenTreasury>;
};

export type TreasuryArgs = {
  admin: OptionOrNullable<PublicKey>;
  nativeReceiver: PublicKey;
  nativeFeeBps: number | bigint;
  lzToken: OptionOrNullable<LzTokenTreasuryArgs>;
};

export function getTreasurySerializer(): Serializer<TreasuryArgs, Treasury> {
  return struct<Treasury>(
    [
      ['admin', option(publicKeySerializer())],
      ['nativeReceiver', publicKeySerializer()],
      ['nativeFeeBps', u64()],
      ['lzToken', option(getLzTokenTreasurySerializer())],
    ],
    { description: 'Treasury' }
  ) as Serializer<TreasuryArgs, Treasury>;
}
