// @orderly.network/chart/dist/index.js
'use strict';

var react = require('react');
var recharts = require('recharts');
var i18n = require('@orderly.network/i18n');
var ui = require('@orderly.network/ui');
var utils = require('@orderly.network/utils');
var jsxRuntime = require('react/jsx-runtime');

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// ../../node_modules/.pnpm/tailwindcss@3.4.17_ts-node@10.9.2_@swc+core@1.13.2_@swc+helpers@0.5.17__@types+node@24.3.1_typescript@5.8.3_/node_modules/tailwindcss/lib/util/createPlugin.js
var require_createPlugin = __commonJS({
  "../../node_modules/.pnpm/tailwindcss@3.4.17_ts-node@10.9.2_@swc+core@1.13.2_@swc+helpers@0.5.17__@types+node@24.3.1_typescript@5.8.3_/node_modules/tailwindcss/lib/util/createPlugin.js"(exports) {
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    Object.defineProperty(exports, "default", {
      enumerable: true,
      get: function() {
        return _default;
      }
    });
    function createPlugin(plugin2, config) {
      return {
        handler: plugin2,
        config
      };
    }
    createPlugin.withOptions = function(pluginFunction, configFunction = () => ({})) {
      const optionsFunction = function(options) {
        return {
          __options: options,
          handler: pluginFunction(options),
          config: configFunction(options)
        };
      };
      optionsFunction.__isOptionsFunction = true;
      optionsFunction.__pluginFunction = pluginFunction;
      optionsFunction.__configFunction = configFunction;
      return optionsFunction;
    };
    var _default = createPlugin;
  }
});

// ../../node_modules/.pnpm/tailwindcss@3.4.17_ts-node@10.9.2_@swc+core@1.13.2_@swc+helpers@0.5.17__@types+node@24.3.1_typescript@5.8.3_/node_modules/tailwindcss/lib/public/create-plugin.js
var require_create_plugin = __commonJS({
  "../../node_modules/.pnpm/tailwindcss@3.4.17_ts-node@10.9.2_@swc+core@1.13.2_@swc+helpers@0.5.17__@types+node@24.3.1_typescript@5.8.3_/node_modules/tailwindcss/lib/public/create-plugin.js"(exports) {
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    Object.defineProperty(exports, "default", {
      enumerable: true,
      get: function() {
        return _default;
      }
    });
    var _createPlugin = /* @__PURE__ */ _interop_require_default(require_createPlugin());
    function _interop_require_default(obj) {
      return obj && obj.__esModule ? obj : {
        default: obj
      };
    }
    var _default = _createPlugin.default;
  }
});

// ../../node_modules/.pnpm/tailwindcss@3.4.17_ts-node@10.9.2_@swc+core@1.13.2_@swc+helpers@0.5.17__@types+node@24.3.1_typescript@5.8.3_/node_modules/tailwindcss/plugin.js
var require_plugin = __commonJS({
  "../../node_modules/.pnpm/tailwindcss@3.4.17_ts-node@10.9.2_@swc+core@1.13.2_@swc+helpers@0.5.17__@types+node@24.3.1_typescript@5.8.3_/node_modules/tailwindcss/plugin.js"(exports, module) {
    var createPlugin = require_create_plugin();
    module.exports = (createPlugin.__esModule ? createPlugin : { default: createPlugin }).default;
  }
});
var tickFormatter = (value) => {
  const absValue = Math.abs(value);
  const dp = absValue === 0 ? 0 : absValue <= 10 ? 2 : absValue <= 100 ? 1 : 0;
  const formatted = utils.numberToHumanStyle(absValue, dp);
  return value < 0 ? `-${formatted}` : formatted;
};
var OrderlyChartTooltip = (props) => {
  const {
    label,
    value,
    prefix,
    unit = "USDC",
    coloring = false,
    dp,
    rm
  } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { intensity: 600, p: 3, r: "md", children: [
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "row", className: props.titleClassName, children: [
      prefix,
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Text.numeral,
        {
          unit,
          as: "div",
          size: "sm",
          coloring,
          showIdentifier: coloring,
          unitClassName: "oui-text-base-contrast-54 oui-ml-1",
          weight: "semibold",
          rm,
          dp,
          children: value
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: "2xs", intensity: 54, weight: "semibold", children: label })
  ] });
};

// src/utils/theme.ts
var getThemeColors = () => {
  const root = document.documentElement;
  const computedStyle = getComputedStyle(root);
  const colors = {
    primary: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-primary")
    ),
    primaryLight: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-primary-light")
    ),
    secondary: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-secondary")
    ),
    success: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-success")
    ),
    warning: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-warning")
    ),
    danger: convertToRGB(computedStyle.getPropertyValue("--oui-color-danger")),
    info: convertToRGB(computedStyle.getPropertyValue("--oui-color-info")),
    loss: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-trading-loss")
    ),
    profit: convertToRGB(
      computedStyle.getPropertyValue("--oui-color-trading-profit")
    )
  };
  return colors;
};
var convertToRGB = (color) => {
  return `rgb(${color.split(" ").join(",")})`;
};

// src/orderly/useColors.ts
var useColors = (colors) => {
  const _colors = react.useMemo(() => {
    const themeColors = getThemeColors();
    return {
      profit: colors?.profit || themeColors.profit,
      loss: colors?.loss || themeColors.loss,
      primary: themeColors.primary,
      primaryLight: themeColors.primaryLight
    };
  }, [colors]);
  return _colors;
};
var RoundedRectangle = (props) => {
  const { fill, x, y, width, height } = props;
  const absHeight = Math.abs(height);
  return /* @__PURE__ */ jsxRuntime.jsx(
    "rect",
    {
      rx: 2,
      x,
      y: height > 0 ? y : y + height,
      width,
      height: absHeight,
      stroke: "none",
      fill
    }
  );
};
var XAxisLabel = (props) => {
  const { x, y, stroke, payload, index, width, containerWidth } = props;
  const { t } = i18n.useTranslation();
  const _x = index === 0 ? 48 : containerWidth > 0 ? containerWidth - 10 : width + payload.offset;
  return /* @__PURE__ */ jsxRuntime.jsx("g", { transform: `translate(${_x},${y - 6})`, children: /* @__PURE__ */ jsxRuntime.jsx(
    "text",
    {
      x: 0,
      y: 0,
      dy: 16,
      textAnchor: index === 0 ? "start" : "end",
      fontSize: 10,
      fill: "rgba(255,255,255,0.54)",
      children: index === 0 ? payload.value : t("chart.now")
    }
  ) });
};
var CustomizedCross = (props) => {
  const { width, height, stroke, fill } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(
    recharts.Cross,
    {
      x: props.x + props.width / 2,
      top: props.top,
      height,
      width: 1,
      stroke: "rgba(255,255,255,0.16)",
      strokeDasharray: "3 2",
      fill: "none"
    }
  );
};
var CustomTooltip = (props) => {
  const { active, payload, label } = props;
  const todayStr = react.useRef((/* @__PURE__ */ new Date()).toISOString().split("T")[0]);
  const { t } = i18n.useTranslation();
  if (active && payload && payload.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      OrderlyChartTooltip,
      {
        label: label === todayStr.current ? t("chart.now") : label,
        value: payload[0].value,
        coloring: true
      }
    );
  }
  return null;
};
var PnLBarChart = (props) => {
  const { invisible, data, responsiveContainerProps } = props;
  const colors = useColors(props.colors);
  const widthRef = react.useRef(0);
  return /* @__PURE__ */ jsxRuntime.jsx(
    recharts.ResponsiveContainer,
    {
      className: ui.cn(invisible && "chart-invisible"),
      onResize: (width) => {
        widthRef.current = width;
      },
      ...responsiveContainerProps,
      children: /* @__PURE__ */ jsxRuntime.jsxs(
        recharts.BarChart,
        {
          data,
          margin: { left: -10, top: 10, right: 10, bottom: 30 },
          children: [
            !invisible && /* @__PURE__ */ jsxRuntime.jsx(
              recharts.Tooltip,
              {
                cursor: /* @__PURE__ */ jsxRuntime.jsx(CustomizedCross, {}),
                content: /* @__PURE__ */ jsxRuntime.jsx(CustomTooltip, {})
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(recharts.CartesianGrid, { vertical: false, stroke: "#FFFFFF", strokeOpacity: 0.04 }),
            /* @__PURE__ */ jsxRuntime.jsx(recharts.ReferenceLine, { y: 0, stroke: "rgba(0,0,0,0.04)" }),
            !invisible && /* @__PURE__ */ jsxRuntime.jsx(recharts.Bar, { dataKey: "pnl", shape: /* @__PURE__ */ jsxRuntime.jsx(RoundedRectangle, {}), children: data.map((entry, index) => {
              return /* @__PURE__ */ jsxRuntime.jsx(
                recharts.Cell,
                {
                  fill: entry.pnl > 0 ? colors.profit : colors.loss
                },
                `cell-${index}`
              );
            }) }),
            /* @__PURE__ */ jsxRuntime.jsx(
              recharts.YAxis,
              {
                tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
                tickFormatter: (value) => tickFormatter(value),
                tickLine: false,
                axisLine: false,
                dataKey: "pnl"
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              recharts.XAxis,
              {
                dataKey: "date",
                tickLine: false,
                interval: data.length - 2,
                height: 1,
                tick: /* @__PURE__ */ jsxRuntime.jsx(XAxisLabel, { containerWidth: widthRef.current }),
                stroke: "#FFFFFF",
                strokeOpacity: 0.04
              }
            )
          ]
        }
      )
    }
  );
};
var XAxisLabel2 = (props) => {
  const { x, y, stroke, payload, index } = props;
  const { t } = i18n.useTranslation();
  return /* @__PURE__ */ jsxRuntime.jsx("g", { transform: `translate(${x},${y - 6})`, children: /* @__PURE__ */ jsxRuntime.jsx(
    "text",
    {
      x: 0,
      y: 0,
      dy: 16,
      textAnchor: "end",
      fontSize: 10,
      fill: "rgba(255,255,255,0.54)",
      children: index === 0 ? payload.value : t("chart.now")
    }
  ) });
};
var CustomTooltip2 = (props) => {
  const { active, payload, label } = props;
  const todayStr = react.useRef((/* @__PURE__ */ new Date()).toISOString().split("T")[0]);
  const { t } = i18n.useTranslation();
  if (active && payload && payload.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      OrderlyChartTooltip,
      {
        label: label === todayStr.current ? t("chart.now") : label,
        value: payload[0].value,
        coloring: true
      }
    );
  }
  return null;
};
var dataTransfer = (data) => {
  const series = [];
  data?.reduce((acc, item) => {
    acc += item.pnl;
    series.push({ ...item, pnl: acc, _pnl: item.pnl });
    return acc;
  }, 0);
  return series;
};
var PnlLineChart = (props) => {
  const { responsiveContainerProps } = props;
  const colors = useColors(props.colors);
  const { isMobile } = ui.useScreen();
  const data = react.useMemo(() => dataTransfer(props.data), [props.data]);
  const chartComponent = /* @__PURE__ */ jsxRuntime.jsxs(
    recharts.LineChart,
    {
      data,
      margin: { top: 20, right: 10, left: -10, bottom: 0 },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(recharts.CartesianGrid, { vertical: false, stroke: "#FFFFFF", strokeOpacity: 0.04 }),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.XAxis,
          {
            dataKey: "date",
            interval: props.data.length - 2,
            tick: /* @__PURE__ */ jsxRuntime.jsx(XAxisLabel2, {}),
            stroke: "#FFFFFF",
            strokeOpacity: 0.04
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.YAxis,
          {
            dataKey: "pnl",
            tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
            tickLine: false,
            axisLine: false,
            tickFormatter
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Tooltip,
          {
            cursor: { strokeDasharray: "3 2", strokeOpacity: 0.16 },
            content: /* @__PURE__ */ jsxRuntime.jsx(CustomTooltip2, {})
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Line,
          {
            type: "natural",
            dataKey: "pnl",
            stroke: colors.primary,
            strokeWidth: isMobile ? 1.5 : 2,
            dot: false,
            isAnimationActive: false
          }
        )
      ]
    }
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    recharts.ResponsiveContainer,
    {
      className: props.invisible ? "chart-invisible" : void 0,
      ...responsiveContainerProps,
      children: chartComponent
    }
  );
};
var CustomTooltip3 = (props) => {
  const { active, payload, label } = props;
  const todayStr = react.useRef((/* @__PURE__ */ new Date()).toISOString().split("T")[0]);
  const { t } = i18n.useTranslation();
  if (active && payload && payload.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      OrderlyChartTooltip,
      {
        label: label === todayStr.current ? t("chart.now") : label,
        value: payload[0].value,
        coloring: true
      }
    );
  }
  return null;
};
var dataTransfer2 = (data) => {
  const series = [];
  data?.reduce((acc, item) => {
    acc += item.pnl;
    series.push({ ...item, pnl: acc, _pnl: item.pnl });
    return acc;
  }, 0);
  return series;
};
var PnlAreaChart = (props) => {
  const { responsiveContainerProps } = props;
  const colors = useColors(props.colors);
  const { isMobile } = ui.useScreen();
  const colorId = react.useId();
  const data = react.useMemo(() => dataTransfer2(props.data), [props.data]);
  const baseValue = react.useMemo(
    () => data.map((item) => item.pnl).sort((a, b) => a - b)[0] || 0,
    [data]
  );
  const chartComponent = /* @__PURE__ */ jsxRuntime.jsxs(
    recharts.AreaChart,
    {
      data,
      margin: { top: 20, right: 10, left: -10, bottom: 0 },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(recharts.CartesianGrid, { vertical: false, stroke: "#FFFFFF", strokeOpacity: 0.04 }),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.XAxis,
          {
            dataKey: "date",
            interval: props.data.length - 2,
            tick: /* @__PURE__ */ jsxRuntime.jsx(XAxisLabel2, {}),
            stroke: "#FFFFFF",
            strokeOpacity: 0.04
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.YAxis,
          {
            dataKey: "pnl",
            tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
            tickLine: false,
            axisLine: false,
            tickFormatter
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Tooltip,
          {
            cursor: { strokeDasharray: "3 2", strokeOpacity: 0.16 },
            content: /* @__PURE__ */ jsxRuntime.jsx(CustomTooltip3, {})
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          /* @__PURE__ */ jsxRuntime.jsx("defs", { children: /* @__PURE__ */ jsxRuntime.jsxs("linearGradient", { id: colorId, x1: "0", y1: "0", x2: "0", y2: "1", children: [
            /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: colors.primary, offset: "0%", stopOpacity: 0.5 }),
            /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: colors.primary, offset: "100%", stopOpacity: 0 })
          ] }) }),
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.Area,
            {
              type: "natural",
              dataKey: "pnl",
              stroke: colors.primary,
              strokeWidth: isMobile ? 1.5 : 2,
              dot: false,
              isAnimationActive: false,
              fill: `url(#${colorId})`,
              baseValue: baseValue || 0
            }
          )
        ] })
      ]
    }
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    recharts.ResponsiveContainer,
    {
      className: props.invisible ? "chart-invisible" : void 0,
      ...responsiveContainerProps,
      children: chartComponent
    }
  );
};
var CustomTooltip4 = (props) => {
  const { active, payload, label } = props;
  const todayStr = react.useRef((/* @__PURE__ */ new Date()).toISOString().split("T")[0]);
  const { t } = i18n.useTranslation();
  if (active && payload && payload.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      OrderlyChartTooltip,
      {
        label: label === todayStr.current ? t("chart.now") : label,
        value: payload[0].value
      }
    );
  }
  return null;
};
var AssetLineChart = (props) => {
  const { responsiveContainerProps } = props;
  const colors = useColors(props.colors);
  const colorId = react.useId();
  const { isMobile } = ui.useScreen();
  const chartComponent = isMobile ? /* @__PURE__ */ jsxRuntime.jsxs(
    recharts.AreaChart,
    {
      width: 530,
      height: 180,
      data: props.data,
      margin: { top: 20, right: 10, left: -20, bottom: -10 },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(recharts.CartesianGrid, { vertical: false, stroke: "#FFFFFF", strokeOpacity: 0.04 }),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.XAxis,
          {
            dataKey: "date",
            interval: props.data.length - 2,
            tick: /* @__PURE__ */ jsxRuntime.jsx(XAxisLabel2, {}),
            stroke: "#FFFFFF",
            strokeOpacity: 0.04
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.YAxis,
          {
            dataKey: "account_value",
            tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
            tickLine: false,
            axisLine: false,
            tickFormatter: (value) => tickFormatter(value)
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          /* @__PURE__ */ jsxRuntime.jsx("defs", { children: /* @__PURE__ */ jsxRuntime.jsxs("linearGradient", { id: colorId, x1: "0", y1: "0", x2: "0", y2: "1", children: [
            /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: "#00B49E", offset: "0%", stopOpacity: 0.5 }),
            /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: "#00B49E", offset: "100%", stopOpacity: 0 })
          ] }) }),
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.Area,
            {
              type: "natural",
              dataKey: "account_value",
              stroke: colors.profit,
              strokeWidth: isMobile ? 1.5 : 2,
              dot: false,
              isAnimationActive: false,
              fill: `url(#${colorId})`
            }
          )
        ] })
      ]
    }
  ) : /* @__PURE__ */ jsxRuntime.jsxs(
    recharts.LineChart,
    {
      width: 530,
      height: 180,
      data: props.data,
      margin: { top: 20, right: 10, left: -20, bottom: -10 },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(recharts.CartesianGrid, { vertical: false, stroke: "#FFFFFF", strokeOpacity: 0.04 }),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.XAxis,
          {
            dataKey: "date",
            interval: props.data.length - 2,
            tick: /* @__PURE__ */ jsxRuntime.jsx(XAxisLabel2, {}),
            stroke: "#FFFFFF",
            strokeOpacity: 0.04
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.YAxis,
          {
            dataKey: "account_value",
            tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
            tickLine: false,
            axisLine: false,
            tickFormatter
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Tooltip,
          {
            cursor: { strokeDasharray: "3 2", strokeOpacity: 0.16 },
            content: /* @__PURE__ */ jsxRuntime.jsx(CustomTooltip4, {})
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Line,
          {
            type: "natural",
            dataKey: "account_value",
            stroke: colors.profit,
            strokeWidth: isMobile ? 1.5 : 2,
            dot: false,
            isAnimationActive: false
          }
        )
      ]
    }
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    recharts.ResponsiveContainer,
    {
      className: props.invisible ? "chart-invisible" : void 0,
      ...responsiveContainerProps,
      children: chartComponent
    }
  );
};
var CustomTooltip5 = (props) => {
  const { active, payload, label } = props;
  const todayStr = react.useRef((/* @__PURE__ */ new Date()).toISOString().split("T")[0]);
  const { t } = i18n.useTranslation();
  if (active && payload && payload.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      OrderlyChartTooltip,
      {
        label: label === todayStr.current ? t("chart.now") : label,
        value: payload[0].value
      }
    );
  }
  return null;
};
var AssetAreaChart = (props) => {
  const { responsiveContainerProps } = props;
  const colors = useColors(props.colors);
  const colorId = react.useId();
  const { isMobile } = ui.useScreen();
  const chartComponent = /* @__PURE__ */ jsxRuntime.jsxs(
    recharts.AreaChart,
    {
      width: 530,
      height: 180,
      data: props.data,
      margin: { top: 20, right: 10, left: -20, bottom: -10 },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(recharts.CartesianGrid, { vertical: false, stroke: "#FFFFFF", strokeOpacity: 0.04 }),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.XAxis,
          {
            dataKey: "date",
            interval: props.data.length - 2,
            tick: /* @__PURE__ */ jsxRuntime.jsx(XAxisLabel2, {}),
            stroke: "#FFFFFF",
            strokeOpacity: 0.04
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.YAxis,
          {
            dataKey: "account_value",
            tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
            tickLine: false,
            axisLine: false,
            tickFormatter
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Tooltip,
          {
            cursor: { strokeDasharray: "3 2", strokeOpacity: 0.16 },
            content: /* @__PURE__ */ jsxRuntime.jsx(CustomTooltip5, {})
          }
        ),
        !props.invisible && /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          /* @__PURE__ */ jsxRuntime.jsx("defs", { children: /* @__PURE__ */ jsxRuntime.jsxs("linearGradient", { id: colorId, x1: "0", y1: "0", x2: "0", y2: "1", children: [
            /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: "#00B49E", offset: "0%", stopOpacity: 0.5 }),
            /* @__PURE__ */ jsxRuntime.jsx("stop", { stopColor: "#00B49E", offset: "100%", stopOpacity: 0 })
          ] }) }),
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.Area,
            {
              type: "natural",
              dataKey: "account_value",
              stroke: colors.profit,
              strokeWidth: isMobile ? 1.5 : 2,
              dot: false,
              isAnimationActive: false,
              fill: `url(#${colorId})`
            }
          )
        ] })
      ]
    }
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    recharts.ResponsiveContainer,
    {
      className: props.invisible ? "chart-invisible" : void 0,
      ...responsiveContainerProps,
      children: chartComponent
    }
  );
};
var RoundedRectangle2 = (props) => {
  const { fill, x, y, width, height, opacity } = props;
  const absHeight = Math.abs(height);
  return /* @__PURE__ */ jsxRuntime.jsx(
    "rect",
    {
      rx: 2,
      x,
      y: height > 0 ? y : y + height,
      width,
      height: absHeight,
      stroke: "none",
      fill,
      opacity
    }
  );
};
var CustomizedCross2 = (props) => {
  const { width, height, payload, stroke, fill } = props;
  if (payload?.[0]?.value === 0) {
    return null;
  }
  return (
    // @ts-ignore
    /* @__PURE__ */ jsxRuntime.jsx(
      recharts.Cross,
      {
        x: props.x + props.width / 2,
        top: props.top,
        height,
        width: 1,
        stroke: "rgba(255,255,255,0.16)",
        strokeDasharray: "3 2",
        fill: "none"
      }
    )
  );
};
var CustomTooltip6 = (props) => {
  const { active, payload, label, tooltip } = props;
  if (payload?.[0]?.value === 0) {
    return null;
  }
  if (active && payload && payload.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      OrderlyChartTooltip,
      {
        label,
        value: payload[0].value,
        titleClassName: "oui-gap-4",
        rm: tooltip?.rm,
        dp: tooltip?.dp
      }
    );
  }
  return null;
};
var VolBarChart = (props) => {
  const colors = useColors(
    props.colors?.fill ? { profit: props.colors?.fill, loss: props.colors?.fill } : void 0
  );
  const isEmpty = props.data?.reduce((pre, cur) => pre + cur.volume, 0) === 0;
  const maxVolume = props.data?.reduce(
    (pre, cur) => pre > cur.volume ? pre : cur.volume,
    0
  );
  const decimal = maxVolume <= 10 ? 2 : maxVolume <= 100 ? 1 : 0;
  return (
    // @ts-ignore
    /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { className: ui.cn(props.className), children: /* @__PURE__ */ jsxRuntime.jsx(recharts.ResponsiveContainer, { children: /* @__PURE__ */ jsxRuntime.jsxs(
      recharts.BarChart,
      {
        data: props.data,
        margin: { left: -0, top: 6, right: 0, bottom: 20 },
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.Tooltip,
            {
              cursor: /* @__PURE__ */ jsxRuntime.jsx(CustomizedCross2, {}),
              content: /* @__PURE__ */ jsxRuntime.jsx(CustomTooltip6, { tooltip: props.tooltip })
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.CartesianGrid,
            {
              vertical: false,
              stroke: "#FFFFFF",
              strokeOpacity: 0.08,
              repeatCount: 6
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(recharts.ReferenceLine, { y: 0, stroke: "#000" }),
          /* @__PURE__ */ jsxRuntime.jsx(recharts.Bar, { dataKey: "volume", shape: /* @__PURE__ */ jsxRuntime.jsx(RoundedRectangle2, {}), minPointSize: 1, children: props.data.map((entry, index) => {
            return (
              // @ts-ignore
              /* @__PURE__ */ jsxRuntime.jsx(
                recharts.Cell,
                {
                  fill: entry.volume > 0 ? colors.profit : colors.loss,
                  opacity: entry.opacity
                },
                `cell-${index}`
              )
            );
          }) }),
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.YAxis,
            {
              tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
              tickLine: false,
              axisLine: false,
              dataKey: "volume",
              tickFormatter: (value, index) => {
                if (isEmpty) return `${index * 100}`;
                return numberToHumanStyle2(value, decimal);
              },
              width: 45
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            recharts.XAxis,
            {
              dataKey: "date",
              tickLine: false,
              interval: props.data.length - 2,
              height: 1,
              tick: { fontSize: 10, fill: "rgba(255,255,255,0.54)" },
              stroke: "rgb(229, 231, 235)",
              strokeOpacity: 0.2
            }
          )
        ]
      }
    ) }) })
  );
};
function numberToHumanStyle2(number, decimalPlaces = 0) {
  const abbreviations = ["", "K", "M", "B", "T"];
  let index = 0;
  while (number >= 1e3 && index < abbreviations.length - 1) {
    number /= 1e3;
    index++;
  }
  const roundedNumber = toFixedWithoutRounding(number, decimalPlaces);
  return `${roundedNumber}${abbreviations[index]}`;
}
function toFixedWithoutRounding(num, fix) {
  const numStr = num.toString();
  const decimalIndex = numStr.indexOf(".");
  if (decimalIndex === -1 || fix === 0) {
    return numStr.split(".")[0];
  }
  const cutoffIndex = decimalIndex + fix + 1;
  return numStr.slice(0, cutoffIndex);
}

// src/tailwindcss/theme.ts
var import_plugin = __toESM(require_plugin());
var chartPlugin = () => (0, import_plugin.default)(function({ addComponents, addBase }) {
  addComponents(
    {
      ".xAxis": {
        ".recharts-cartesian-axis-tick:first-child text": {
          "text-anchor": "start"
        },
        ".recharts-cartesian-axis-tick:last-child text": {
          "text-anchor": "end"
        }
      }
    },
    {
      respectPrefix: false
    }
  );
});

Object.defineProperty(exports, "Area", {
  enumerable: true,
  get: function () { return recharts.Area; }
});
Object.defineProperty(exports, "AreaChart", {
  enumerable: true,
  get: function () { return recharts.AreaChart; }
});
Object.defineProperty(exports, "Bar", {
  enumerable: true,
  get: function () { return recharts.Bar; }
});
Object.defineProperty(exports, "BarChart", {
  enumerable: true,
  get: function () { return recharts.BarChart; }
});
Object.defineProperty(exports, "Cell", {
  enumerable: true,
  get: function () { return recharts.Cell; }
});
Object.defineProperty(exports, "Line", {
  enumerable: true,
  get: function () { return recharts.Line; }
});
Object.defineProperty(exports, "LineChart", {
  enumerable: true,
  get: function () { return recharts.LineChart; }
});
exports.AssetAreaChart = AssetAreaChart;
exports.AssetLineChart = AssetLineChart;
exports.PnLBarChart = PnLBarChart;
exports.PnlAreaChart = PnlAreaChart;
exports.PnlLineChart = PnlLineChart;
exports.VolBarChart = VolBarChart;
exports.chartPlugin = chartPlugin;
