// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/instructions/executable.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type ExecutableInstructionAccounts = {
  nonce: PublicKey | Pda;
  payloadHash: PublicKey | Pda;
};

// Data.
export type ExecutableInstructionData = {
  discriminator: Uint8Array;
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: bigint;
};

export type ExecutableInstructionDataArgs = {
  receiver: PublicKey;
  srcEid: number;
  sender: Uint8Array;
  nonce: number | bigint;
};

export function getExecutableInstructionDataSerializer(): Serializer<
  ExecutableInstructionDataArgs,
  ExecutableInstructionData
> {
  return mapSerializer<
    ExecutableInstructionDataArgs,
    any,
    ExecutableInstructionData
  >(
    struct<ExecutableInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['receiver', publicKeySerializer()],
        ['srcEid', u32()],
        ['sender', bytes({ size: 32 })],
        ['nonce', u64()],
      ],
      { description: 'ExecutableInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([57, 165, 171, 234, 157, 191, 156, 25]),
    })
  ) as Serializer<ExecutableInstructionDataArgs, ExecutableInstructionData>;
}

// Args.
export type ExecutableInstructionArgs = ExecutableInstructionDataArgs;

// Instruction.
export function executable(
  context: Pick<Context, 'programs'>,
  accounts: ExecutableInstructionAccounts,
  args: ExecutableInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'executor',
    '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
  );

  // Accounts.
  const resolvedAccounts = {
    nonce: {
      index: 0,
      isWritable: false as boolean,
      value: accounts.nonce ?? null,
    },
    payloadHash: {
      index: 1,
      isWritable: false as boolean,
      value: accounts.payloadHash ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: ExecutableInstructionArgs = { ...args };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getExecutableInstructionDataSerializer().serialize(
    resolvedArgs as ExecutableInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
