// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/instructions/execute.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
  u64,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';
import {
  LzReceiveParams,
  LzReceiveParamsArgs,
  getLzReceiveParamsSerializer,
} from '../types';

// Accounts.
export type ExecuteInstructionAccounts = {
  executor: Signer;
  config: PublicKey | Pda;
  endpointProgram: PublicKey | Pda;
  /** The authority for the endpoint program to emit events */
  endpointEventAuthority: PublicKey | Pda;
  eventAuthority: PublicKey | Pda;
  program: PublicKey | Pda;
};

// Data.
export type ExecuteInstructionData = {
  discriminator: Uint8Array;
  receiver: PublicKey;
  lzReceive: LzReceiveParams;
  value: bigint;
  computeUnits: bigint;
};

export type ExecuteInstructionDataArgs = {
  receiver: PublicKey;
  lzReceive: LzReceiveParamsArgs;
  value: number | bigint;
  computeUnits: number | bigint;
};

export function getExecuteInstructionDataSerializer(): Serializer<
  ExecuteInstructionDataArgs,
  ExecuteInstructionData
> {
  return mapSerializer<ExecuteInstructionDataArgs, any, ExecuteInstructionData>(
    struct<ExecuteInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['receiver', publicKeySerializer()],
        ['lzReceive', getLzReceiveParamsSerializer()],
        ['value', u64()],
        ['computeUnits', u64()],
      ],
      { description: 'ExecuteInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([130, 221, 242, 154, 13, 193, 189, 29]),
    })
  ) as Serializer<ExecuteInstructionDataArgs, ExecuteInstructionData>;
}

// Args.
export type ExecuteInstructionArgs = ExecuteInstructionDataArgs;

// Instruction.
export function execute(
  context: Pick<Context, 'programs'>,
  input: ExecuteInstructionAccounts & ExecuteInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'executor',
    '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
  );

  // Accounts.
  const resolvedAccounts = {
    executor: {
      index: 0,
      isWritable: true as boolean,
      value: input.executor ?? null,
    },
    config: {
      index: 1,
      isWritable: false as boolean,
      value: input.config ?? null,
    },
    endpointProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.endpointProgram ?? null,
    },
    endpointEventAuthority: {
      index: 3,
      isWritable: false as boolean,
      value: input.endpointEventAuthority ?? null,
    },
    eventAuthority: {
      index: 4,
      isWritable: false as boolean,
      value: input.eventAuthority ?? null,
    },
    program: {
      index: 5,
      isWritable: false as boolean,
      value: input.program ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: ExecuteInstructionArgs = { ...input };

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getExecuteInstructionDataSerializer().serialize(
    resolvedArgs as ExecuteInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
