// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/instructions/extendExecutorConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type ExtendExecutorConfigInstructionAccounts = {
  admin: Signer;
  config: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
};

// Data.
export type ExtendExecutorConfigInstructionData = { discriminator: Uint8Array };

export type ExtendExecutorConfigInstructionDataArgs = {};

export function getExtendExecutorConfigInstructionDataSerializer(): Serializer<
  ExtendExecutorConfigInstructionDataArgs,
  ExtendExecutorConfigInstructionData
> {
  return mapSerializer<
    ExtendExecutorConfigInstructionDataArgs,
    any,
    ExtendExecutorConfigInstructionData
  >(
    struct<ExtendExecutorConfigInstructionData>(
      [['discriminator', bytes({ size: 8 })]],
      { description: 'ExtendExecutorConfigInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([238, 223, 196, 220, 61, 71, 72, 85]),
    })
  ) as Serializer<
    ExtendExecutorConfigInstructionDataArgs,
    ExtendExecutorConfigInstructionData
  >;
}

// Instruction.
export function extendExecutorConfig(
  context: Pick<Context, 'programs'>,
  input: ExtendExecutorConfigInstructionAccounts
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'executor',
    '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
  );

  // Accounts.
  const resolvedAccounts = {
    admin: {
      index: 0,
      isWritable: true as boolean,
      value: input.admin ?? null,
    },
    config: {
      index: 1,
      isWritable: true as boolean,
      value: input.config ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Default values.
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getExtendExecutorConfigInstructionDataSerializer().serialize({});

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
