// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/instructions/initExecutor.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  mapSerializer,
  publicKey as publicKeySerializer,
  struct,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type InitExecutorInstructionAccounts = {
  payer?: Signer;
  config: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
};

// Data.
export type InitExecutorInstructionData = {
  discriminator: Uint8Array;
  owner: PublicKey;
  admins: Array<PublicKey>;
  executors: Array<PublicKey>;
  msglibs: Array<PublicKey>;
  priceFeed: PublicKey;
};

export type InitExecutorInstructionDataArgs = {
  owner: PublicKey;
  admins: Array<PublicKey>;
  executors: Array<PublicKey>;
  msglibs: Array<PublicKey>;
  priceFeed: PublicKey;
};

export function getInitExecutorInstructionDataSerializer(): Serializer<
  InitExecutorInstructionDataArgs,
  InitExecutorInstructionData
> {
  return mapSerializer<
    InitExecutorInstructionDataArgs,
    any,
    InitExecutorInstructionData
  >(
    struct<InitExecutorInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['owner', publicKeySerializer()],
        ['admins', array(publicKeySerializer())],
        ['executors', array(publicKeySerializer())],
        ['msglibs', array(publicKeySerializer())],
        ['priceFeed', publicKeySerializer()],
      ],
      { description: 'InitExecutorInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([195, 68, 116, 173, 55, 173, 159, 31]),
    })
  ) as Serializer<InitExecutorInstructionDataArgs, InitExecutorInstructionData>;
}

// Args.
export type InitExecutorInstructionArgs = InitExecutorInstructionDataArgs;

// Instruction.
export function initExecutor(
  context: Pick<Context, 'payer' | 'programs'>,
  input: InitExecutorInstructionAccounts & InitExecutorInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'executor',
    '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
  );

  // Accounts.
  const resolvedAccounts = {
    payer: {
      index: 0,
      isWritable: true as boolean,
      value: input.payer ?? null,
    },
    config: {
      index: 1,
      isWritable: true as boolean,
      value: input.config ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: InitExecutorInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.payer.value) {
    resolvedAccounts.payer.value = context.payer;
  }
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getInitExecutorInstructionDataSerializer().serialize(
    resolvedArgs as InitExecutorInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
