// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/instructions/postExecute.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  publicKey,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type PostExecuteInstructionAccounts = {
  executor: Signer;
  /** Execution context account containing state from pre-execution */
  context: PublicKey | Pda;
  /** Instruction sysvar account for introspection */
  instructionSysvar?: PublicKey | Pda;
};

// Data.
export type PostExecuteInstructionData = {
  discriminator: Uint8Array;
  /** Version of the execution context to use */
  contextVersion: number;
};

export type PostExecuteInstructionDataArgs = {
  /** Version of the execution context to use */
  contextVersion: number;
};

export function getPostExecuteInstructionDataSerializer(): Serializer<
  PostExecuteInstructionDataArgs,
  PostExecuteInstructionData
> {
  return mapSerializer<
    PostExecuteInstructionDataArgs,
    any,
    PostExecuteInstructionData
  >(
    struct<PostExecuteInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['contextVersion', u8()],
      ],
      { description: 'PostExecuteInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([74, 168, 167, 167, 76, 220, 95, 56]),
    })
  ) as Serializer<PostExecuteInstructionDataArgs, PostExecuteInstructionData>;
}

// Args.
export type PostExecuteInstructionArgs = PostExecuteInstructionDataArgs;

// Instruction.
export function postExecute(
  context: Pick<Context, 'programs'>,
  input: PostExecuteInstructionAccounts & PostExecuteInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'executor',
    '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
  );

  // Accounts.
  const resolvedAccounts = {
    executor: {
      index: 0,
      isWritable: false as boolean,
      value: input.executor ?? null,
    },
    context: {
      index: 1,
      isWritable: true as boolean,
      value: input.context ?? null,
    },
    instructionSysvar: {
      index: 2,
      isWritable: false as boolean,
      value: input.instructionSysvar ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: PostExecuteInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.instructionSysvar.value) {
    resolvedAccounts.instructionSysvar.value = publicKey(
      'Sysvar1nstructions1111111111111111111111111'
    );
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getPostExecuteInstructionDataSerializer().serialize(
    resolvedArgs as PostExecuteInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
