// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/instructions/preExecute.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  Context,
  Pda,
  PublicKey,
  Signer,
  TransactionBuilder,
  publicKey,
  transactionBuilder,
} from '@metaplex-foundation/umi';
import {
  Serializer,
  bytes,
  mapSerializer,
  struct,
  u64,
  u8,
} from '@metaplex-foundation/umi/serializers';
import {
  ResolvedAccount,
  ResolvedAccountsWithIndices,
  getAccountMetasAndSigners,
} from '../shared';

// Accounts.
export type PreExecuteInstructionAccounts = {
  executor: Signer;
  /**
   * Execution context account that stores state between pre and post execution.
   * This account is versioned to support future protocol upgrades.
   */

  context: PublicKey | Pda;
  systemProgram?: PublicKey | Pda;
  /** Instruction sysvar account for introspection */
  instructionSysvar?: PublicKey | Pda;
};

// Data.
export type PreExecuteInstructionData = {
  discriminator: Uint8Array;
  /** Version of the execution context to use */
  contextVersion: number;
  /** Maximum fee that can be charged for execution */
  feeLimit: bigint;
};

export type PreExecuteInstructionDataArgs = {
  /** Version of the execution context to use */
  contextVersion: number;
  /** Maximum fee that can be charged for execution */
  feeLimit: number | bigint;
};

export function getPreExecuteInstructionDataSerializer(): Serializer<
  PreExecuteInstructionDataArgs,
  PreExecuteInstructionData
> {
  return mapSerializer<
    PreExecuteInstructionDataArgs,
    any,
    PreExecuteInstructionData
  >(
    struct<PreExecuteInstructionData>(
      [
        ['discriminator', bytes({ size: 8 })],
        ['contextVersion', u8()],
        ['feeLimit', u64()],
      ],
      { description: 'PreExecuteInstructionData' }
    ),
    (value) => ({
      ...value,
      discriminator: new Uint8Array([207, 125, 18, 148, 72, 192, 198, 83]),
    })
  ) as Serializer<PreExecuteInstructionDataArgs, PreExecuteInstructionData>;
}

// Args.
export type PreExecuteInstructionArgs = PreExecuteInstructionDataArgs;

// Instruction.
export function preExecute(
  context: Pick<Context, 'programs'>,
  input: PreExecuteInstructionAccounts & PreExecuteInstructionArgs
): TransactionBuilder {
  // Program ID.
  const programId = context.programs.getPublicKey(
    'executor',
    '6doghB248px58JSSwG4qejQ46kFMW4AMj7vzJnWZHNZn'
  );

  // Accounts.
  const resolvedAccounts = {
    executor: {
      index: 0,
      isWritable: true as boolean,
      value: input.executor ?? null,
    },
    context: {
      index: 1,
      isWritable: true as boolean,
      value: input.context ?? null,
    },
    systemProgram: {
      index: 2,
      isWritable: false as boolean,
      value: input.systemProgram ?? null,
    },
    instructionSysvar: {
      index: 3,
      isWritable: false as boolean,
      value: input.instructionSysvar ?? null,
    },
  } satisfies ResolvedAccountsWithIndices;

  // Arguments.
  const resolvedArgs: PreExecuteInstructionArgs = { ...input };

  // Default values.
  if (!resolvedAccounts.systemProgram.value) {
    resolvedAccounts.systemProgram.value = context.programs.getPublicKey(
      'splSystem',
      '11111111111111111111111111111111'
    );
    resolvedAccounts.systemProgram.isWritable = false;
  }
  if (!resolvedAccounts.instructionSysvar.value) {
    resolvedAccounts.instructionSysvar.value = publicKey(
      'Sysvar1nstructions1111111111111111111111111'
    );
  }

  // Accounts in order.
  const orderedAccounts: ResolvedAccount[] = Object.values(
    resolvedAccounts
  ).sort((a, b) => a.index - b.index);

  // Keys and Signers.
  const [keys, signers] = getAccountMetasAndSigners(
    orderedAccounts,
    'programId',
    programId
  );

  // Data.
  const data = getPreExecuteInstructionDataSerializer().serialize(
    resolvedArgs as PreExecuteInstructionDataArgs
  );

  // Bytes Created On Chain.
  const bytesCreatedOnChain = 0;

  return transactionBuilder([
    { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
  ]);
}
