// @orderly.network/vaults/dist/index.js
'use strict';

var React = require('react');
var hooks = require('@orderly.network/hooks');
var zustand = require('zustand');
var i18n = require('@orderly.network/i18n');
var jsxRuntime = require('react/jsx-runtime');
var ramda = require('ramda');
var ui = require('@orderly.network/ui');
var types = require('@orderly.network/types');
var utils = require('@orderly.network/utils');
var uiTransfer = require('@orderly.network/ui-transfer');
var uiConnector = require('@orderly.network/ui-connector');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/api/env.ts
var VAULTS_API_URLS, VAULTS_WEBSITE_URLS;
var init_env = __esm({
  "src/api/env.ts"() {
    VAULTS_API_URLS = {
      prod: "https://api-sv.orderly.org",
      staging: "https://testnet-api-sv.orderly.org",
      qa: "https://qa-api-sv-aliyun.orderly.org",
      dev: "https://dev-api-sv.orderly.org"
    };
    VAULTS_WEBSITE_URLS = {
      prod: "https://app.orderly.network",
      staging: "https://staging-app.orderly.network",
      qa: "https://qa-app.orderly.network",
      dev: "https://dev-app.orderly.network"
    };
  }
});
function useSVApiUrl() {
  const env = hooks.useGetEnv();
  const apiUrl = React.useMemo(() => {
    return VAULTS_API_URLS[env];
  }, [env]);
  return apiUrl;
}
var init_useSVAPIUrl = __esm({
  "src/hooks/useSVAPIUrl.ts"() {
    init_env();
  }
});

// src/api/request.ts
function buildQueryString(params) {
  const filteredParams = filterUndefinedParams(params);
  const searchParams = new URLSearchParams();
  Object.entries(filteredParams).forEach(([key, value]) => {
    if (value !== null && value !== void 0) {
      searchParams.append(key, String(value));
    }
  });
  return searchParams.toString();
}
function filterUndefinedParams(params) {
  const filtered = {};
  Object.entries(params).forEach(([key, value]) => {
    if (value !== void 0 && value !== null) {
      if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
        filtered[key] = value;
      } else if (typeof value === "object" && value !== null) {
        const objEntries = Object.entries(value);
        objEntries.forEach(([objKey, objValue]) => {
          if (objValue !== void 0 && objValue !== null && (typeof objValue === "string" || typeof objValue === "number" || typeof objValue === "boolean")) {
            filtered[objKey] = objValue;
          }
        });
      }
    }
  });
  return filtered;
}
function appendQueryParams(url, params) {
  if (!params || Object.keys(params).length === 0) {
    return url;
  }
  const queryString = buildQueryString(params);
  if (!queryString) {
    return url;
  }
  const separator = url.includes("?") ? "&" : "?";
  return `${url}${separator}${queryString}`;
}
var VaultsApiError, RequestClient, defaultClient, request_default;
var init_request = __esm({
  "src/api/request.ts"() {
    VaultsApiError = class extends Error {
      constructor(message, code, status, response) {
        super(message);
        this.code = code;
        this.status = status;
        this.response = response;
        this.name = "VaultsApiError";
      }
    };
    RequestClient = class {
      constructor() {
        this.requestInterceptors = [];
        this.responseInterceptors = [];
        this.defaultConfig = {
          timeout: 1e4,
          retry: 3,
          retryDelay: 1e3,
          validateStatus: (status) => status >= 200 && status < 300,
          // mode: 'cors',
          // credentials: 'include',
          headers: {
            "Content-Type": "application/json"
          }
        };
      }
      // Add request interceptor
      addRequestInterceptor(interceptor) {
        this.requestInterceptors.push(interceptor);
      }
      // Add response interceptor
      addResponseInterceptor(interceptor) {
        this.responseInterceptors.push(interceptor);
      }
      // Create timeout controller
      createTimeoutController(timeout) {
        const controller = new AbortController();
        setTimeout(() => controller.abort(), timeout);
        return controller;
      }
      // Merge headers
      mergeHeaders(defaultHeaders = {}, customHeaders = {}) {
        return {
          ...defaultHeaders,
          ...customHeaders
        };
      }
      // Sleep utility for retry delay
      sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
      }
      // Main request method
      async request(url, config = {}) {
        const finalConfig = {
          ...this.defaultConfig,
          ...config,
          headers: this.mergeHeaders(this.defaultConfig.headers, config.headers)
        };
        let processedConfig = finalConfig;
        for (const interceptor of this.requestInterceptors) {
          processedConfig = await interceptor(processedConfig);
        }
        let fullUrl = url;
        if (typeof url === "string" && !url.startsWith("http") && processedConfig.baseURL) {
          fullUrl = `${processedConfig.baseURL}${url.startsWith("/") ? "" : "/"}${url}`;
        }
        if (processedConfig.params) {
          fullUrl = typeof fullUrl === "string" ? appendQueryParams(fullUrl, processedConfig.params) : fullUrl;
        }
        let timeoutController;
        if (processedConfig.timeout && processedConfig.timeout > 0) {
          timeoutController = this.createTimeoutController(processedConfig.timeout);
          processedConfig.signal = timeoutController.signal;
        }
        const {
          retry = 0,
          retryDelay = 1e3,
          validateStatus,
          params,
          baseURL,
          timeout,
          data,
          method,
          ...fetchOptions
        } = processedConfig;
        const finalFetchOptions = {
          ...fetchOptions,
          method
        };
        if (data !== void 0) {
          finalFetchOptions.body = typeof data === "string" ? data : JSON.stringify(data);
        }
        let lastError;
        for (let attempt = 0; attempt <= retry; attempt++) {
          try {
            const response = await fetch(fullUrl, finalFetchOptions);
            if (!validateStatus(response.status)) {
              let errorMessage = response.statusText;
              let errorCode;
              try {
                const errorData = await response.json();
                errorMessage = errorData.message || errorData.code || errorMessage;
                errorCode = errorData.code;
              } catch {
              }
              throw new VaultsApiError(
                errorMessage,
                errorCode,
                response.status,
                response
              );
            }
            let responseData;
            try {
              responseData = await response.json();
            } catch {
              throw new VaultsApiError(
                "Invalid JSON response",
                "PARSE_ERROR",
                response.status,
                response
              );
            }
            let processedData = responseData;
            for (const interceptor of this.responseInterceptors) {
              processedData = await interceptor(response, processedData);
            }
            if (typeof processedData === "object" && processedData !== null) {
              if ("success" in processedData) {
                const apiResponse = processedData;
                if (apiResponse.success) {
                  return apiResponse.data;
                } else {
                  throw new VaultsApiError(
                    apiResponse.message || "API request failed",
                    apiResponse.code,
                    response.status,
                    response
                  );
                }
              }
            }
            return processedData;
          } catch (error) {
            lastError = error instanceof Error ? error : new Error(String(error));
            if (error instanceof VaultsApiError || error?.name === "AbortError" || attempt === retry) {
              throw lastError;
            }
            if (attempt < retry) {
              await this.sleep(retryDelay);
            }
          } finally {
            timeoutController?.abort();
          }
        }
        throw lastError;
      }
      // Simplified HTTP methods - all accept a single config object
      async get(url, config = {}) {
        return this.request(url, { ...config, method: "GET" });
      }
      async post(url, config = {}) {
        return this.request(url, { ...config, method: "POST" });
      }
      async put(url, config = {}) {
        return this.request(url, { ...config, method: "PUT" });
      }
      async delete(url, config = {}) {
        return this.request(url, { ...config, method: "DELETE" });
      }
      async patch(url, config = {}) {
        return this.request(url, { ...config, method: "PATCH" });
      }
    };
    defaultClient = new RequestClient();
    request_default = defaultClient;
  }
});

// src/api/api.ts
async function getVaultInfo(baseUrl, params) {
  return request_default.get(
    "/v1/public/strategy_vault/vault/info",
    {
      params,
      baseURL: baseUrl
    }
  );
}
async function getVaultLpPerformance(baseUrl, params) {
  return request_default.get(
    "/v1/public/strategy_vault/vault/performance",
    {
      params,
      baseURL: baseUrl
    }
  );
}
async function getVaultLpInfo(baseUrl, params) {
  return request_default.get(
    "/v1/public/strategy_vault/lp/info",
    {
      params,
      baseURL: baseUrl
    }
  );
}
async function getVaultOverallInfo(baseUrl, params) {
  return request_default.get(
    "/v1/public/strategy_vault/vault/overall_info",
    {
      params,
      baseURL: baseUrl
    }
  );
}
var init_api = __esm({
  "src/api/api.ts"() {
    init_request();
  }
});
exports.useVaultsStore = void 0; exports.useVaultInfoState = void 0; exports.useVaultLpPerformanceState = void 0; exports.useVaultLpInfoState = void 0; exports.useVaultOverallInfoState = void 0; exports.useVaultInfoActions = void 0; exports.useVaultLpPerformanceActions = void 0; exports.useVaultLpInfoActions = void 0; exports.useVaultOverallInfoActions = void 0; exports.useVaultLpPerformanceById = void 0; exports.useVaultLpInfoById = void 0; exports.useVaultLpPerformanceIds = void 0; exports.useVaultLpInfoIds = void 0; exports.useVaultLpPerformanceArray = void 0; exports.useVaultLpInfoArray = void 0;
var init_vaultsStore = __esm({
  "src/store/vaultsStore.ts"() {
    init_api();
    exports.useVaultsStore = zustand.create((set, get) => ({
      // Initial state
      baseUrl: "",
      vaultInfo: {
        data: [],
        loading: false,
        error: null,
        lastUpdated: null,
        params: null
      },
      vaultLpPerformance: {
        data: {},
        loading: false,
        error: null,
        lastUpdated: null,
        params: null
      },
      vaultLpInfo: {
        data: {},
        loading: false,
        error: null,
        lastUpdated: null,
        params: null
      },
      vaultOverallInfo: {
        data: null,
        loading: false,
        error: null,
        lastUpdated: null,
        params: null
      },
      vaultsPageConfig: null,
      // Set base URL
      setBaseUrl: (baseUrl) => {
        set({ baseUrl });
      },
      // Vault info actions
      fetchVaultInfo: async (params, baseUrl) => {
        const state = get();
        const url = baseUrl || state.baseUrl;
        if (!url) {
          set((state2) => ({
            vaultInfo: {
              ...state2.vaultInfo,
              error: "Base URL is required"
            }
          }));
          return;
        }
        set((state2) => ({
          baseUrl: baseUrl || state2.baseUrl,
          vaultInfo: {
            ...state2.vaultInfo,
            loading: true,
            error: null,
            params: params || null
          }
        }));
        try {
          const response = await getVaultInfo(url, params);
          set((state2) => ({
            vaultInfo: {
              ...state2.vaultInfo,
              data: response.rows,
              loading: false,
              error: null,
              lastUpdated: Date.now()
            }
          }));
        } catch (error) {
          set((state2) => ({
            vaultInfo: {
              ...state2.vaultInfo,
              loading: false,
              error: error instanceof Error ? error.message : "Failed to fetch vault info"
            }
          }));
        }
      },
      refreshVaultInfo: async () => {
        const state = get();
        if (state.vaultInfo.lastUpdated && state.baseUrl) {
          await state.fetchVaultInfo(state.vaultInfo.params || void 0);
        }
      },
      // Vault LP performance actions
      fetchVaultLpPerformance: async (params, baseUrl) => {
        const state = get();
        const url = baseUrl || state.baseUrl;
        if (!url) {
          set((state2) => ({
            vaultLpPerformance: {
              ...state2.vaultLpPerformance,
              error: "Base URL is required"
            }
          }));
          return;
        }
        set((state2) => ({
          baseUrl: baseUrl || state2.baseUrl,
          vaultLpPerformance: {
            ...state2.vaultLpPerformance,
            loading: true,
            error: null,
            params
          }
        }));
        try {
          const response = await getVaultLpPerformance(
            url,
            params
          );
          set((state2) => ({
            vaultLpPerformance: {
              ...state2.vaultLpPerformance,
              data: {
                ...state2.vaultLpPerformance.data,
                [params.vault_id]: response.rows
              },
              loading: false,
              error: null,
              lastUpdated: Date.now()
            }
          }));
        } catch (error) {
          set((state2) => ({
            vaultLpPerformance: {
              ...state2.vaultLpPerformance,
              loading: false,
              error: error instanceof Error ? error.message : "Failed to fetch vault LP performance"
            }
          }));
        }
      },
      refreshVaultLpPerformance: async () => {
        const state = get();
        if (state.vaultLpPerformance.lastUpdated && state.vaultLpPerformance.params && state.baseUrl) {
          await state.fetchVaultLpPerformance(state.vaultLpPerformance.params);
        }
      },
      // Vault LP info actions
      fetchVaultLpInfo: async (params, baseUrl) => {
        const state = get();
        const url = baseUrl || state.baseUrl;
        if (!url) {
          set((state2) => ({
            vaultLpInfo: {
              ...state2.vaultLpInfo,
              error: "Base URL is required"
            }
          }));
          return;
        }
        set((state2) => ({
          baseUrl: baseUrl || state2.baseUrl,
          vaultLpInfo: {
            ...state2.vaultLpInfo,
            loading: true,
            error: null,
            params
          }
        }));
        try {
          const response = await getVaultLpInfo(url, params);
          set((state2) => ({
            vaultLpInfo: {
              ...state2.vaultLpInfo,
              data: {
                ...state2.vaultLpInfo.data,
                [params.vault_id]: response.rows
              },
              loading: false,
              error: null,
              lastUpdated: Date.now()
            }
          }));
        } catch (error) {
          set((state2) => ({
            vaultLpInfo: {
              ...state2.vaultLpInfo,
              loading: false,
              error: error instanceof Error ? error.message : "Failed to fetch vault LP info"
            }
          }));
        }
      },
      refreshVaultLpInfo: async () => {
        const state = get();
        if (state.vaultLpInfo.lastUpdated && state.vaultLpInfo.params && state.baseUrl) {
          await state.fetchVaultLpInfo(state.vaultLpInfo.params);
        }
      },
      // Vault overall info actions
      fetchVaultOverallInfo: async (params, baseUrl) => {
        const state = get();
        const url = baseUrl || state.baseUrl;
        if (!url) {
          set((state2) => ({
            vaultOverallInfo: {
              ...state2.vaultOverallInfo,
              error: "Base URL is required"
            }
          }));
          return;
        }
        set((state2) => ({
          baseUrl: baseUrl || state2.baseUrl,
          vaultOverallInfo: {
            ...state2.vaultOverallInfo,
            loading: true,
            error: null,
            params: params || null
          }
        }));
        try {
          const response = await getVaultOverallInfo(
            url,
            params
          );
          set((state2) => ({
            vaultOverallInfo: {
              ...state2.vaultOverallInfo,
              data: response,
              loading: false,
              error: null,
              lastUpdated: Date.now()
            }
          }));
        } catch (error) {
          set((state2) => ({
            vaultOverallInfo: {
              ...state2.vaultOverallInfo,
              loading: false,
              error: error instanceof Error ? error.message : "Failed to fetch vault overall info"
            }
          }));
        }
      },
      refreshVaultOverallInfo: async () => {
        const state = get();
        if (state.vaultOverallInfo.lastUpdated && state.baseUrl) {
          await state.fetchVaultOverallInfo(
            state.vaultOverallInfo.params || void 0
          );
        }
      },
      setVaultsPageConfig: (config) => {
        set({ vaultsPageConfig: config });
      }
    }));
    exports.useVaultInfoState = () => exports.useVaultsStore((state) => state.vaultInfo);
    exports.useVaultLpPerformanceState = () => exports.useVaultsStore((state) => state.vaultLpPerformance);
    exports.useVaultLpInfoState = () => exports.useVaultsStore((state) => state.vaultLpInfo);
    exports.useVaultOverallInfoState = () => exports.useVaultsStore((state) => state.vaultOverallInfo);
    exports.useVaultInfoActions = () => exports.useVaultsStore((state) => ({
      fetchVaultInfo: state.fetchVaultInfo,
      refreshVaultInfo: state.refreshVaultInfo
    }));
    exports.useVaultLpPerformanceActions = () => exports.useVaultsStore((state) => ({
      fetchVaultLpPerformance: state.fetchVaultLpPerformance,
      refreshVaultLpPerformance: state.refreshVaultLpPerformance
    }));
    exports.useVaultLpInfoActions = () => exports.useVaultsStore((state) => ({
      fetchVaultLpInfo: state.fetchVaultLpInfo,
      refreshVaultLpInfo: state.refreshVaultLpInfo
    }));
    exports.useVaultOverallInfoActions = () => exports.useVaultsStore((state) => ({
      fetchVaultOverallInfo: state.fetchVaultOverallInfo,
      refreshVaultOverallInfo: state.refreshVaultOverallInfo
    }));
    exports.useVaultLpPerformanceById = (vaultId) => exports.useVaultsStore((state) => state.vaultLpPerformance.data[vaultId]);
    exports.useVaultLpInfoById = (vaultId) => exports.useVaultsStore((state) => state.vaultLpInfo.data[vaultId]);
    exports.useVaultLpPerformanceIds = () => exports.useVaultsStore((state) => Object.keys(state.vaultLpPerformance.data));
    exports.useVaultLpInfoIds = () => exports.useVaultsStore((state) => Object.keys(state.vaultLpInfo.data));
    exports.useVaultLpPerformanceArray = () => exports.useVaultsStore(
      (state) => Object.values(state.vaultLpPerformance.data).flat()
    );
    exports.useVaultLpInfoArray = () => exports.useVaultsStore((state) => Object.values(state.vaultLpInfo.data).flat());
  }
});

// src/store/index.ts
var init_store = __esm({
  "src/store/index.ts"() {
    init_vaultsStore();
  }
});
exports.VaultsHeaderDesktop = void 0;
var init_vaults_header_desktop = __esm({
  "src/components/vaults-header/vaults-header.desktop.tsx"() {
    exports.VaultsHeaderDesktop = (props) => {
      const { supportVaults, headerImage, brokerName } = props;
      const { t } = i18n.useTranslation();
      const supportVaultsList = React.useMemo(() => {
        return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-items-center", children: supportVaults.map((chain, index) => /* @__PURE__ */ jsxRuntime.jsx(
          "img",
          {
            src: `https://oss.orderly.network/static/network_logo/${chain.chain_id}.png`,
            alt: chain.chain_id,
            className: "oui-relative oui-size-5",
            style: {
              marginLeft: index > 0 ? "-4px" : "0",
              zIndex: supportVaults.length - index
            }
          },
          chain.chain_id
        )) });
      }, [supportVaults]);
      const headerImageElement = React.useMemo(() => {
        if (typeof headerImage === "string") {
          return /* @__PURE__ */ jsxRuntime.jsx(
            "img",
            {
              src: headerImage,
              alt: "header",
              className: "oui-bg-contain oui-bg-center oui-bg-no-repeat"
            }
          );
        }
        return headerImage;
      }, [headerImage]);
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-justify-between", children: [
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-max-w-[726px] oui-flex-col", children: [
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1 oui-text-[18px] oui-font-normal oui-text-base-contrast-54", children: [
            t("vaults.availableOn"),
            " ",
            supportVaultsList
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mb-6 oui-mt-8 oui-text-5xl oui-font-bold oui-leading-[44px] oui-text-base-contrast", children: t("vaults.header.title") }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-xl oui-font-normal oui-text-base-contrast-54", children: t("vaults.header.description", { brokerName }) })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-[238px] oui-w-[360px]", children: headerImageElement })
      ] });
    };
  }
});
var VaultsHeaderMobile;
var init_vaults_header_mobile = __esm({
  "src/components/vaults-header/vaults-header.mobile.tsx"() {
    VaultsHeaderMobile = (props) => {
      const { supportVaults, headerImage } = props;
      const { t } = i18n.useTranslation();
      const brokerName = hooks.useConfig("brokerName");
      const supportVaultsList = React.useMemo(() => {
        return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-items-center", children: supportVaults.map((chain, index) => /* @__PURE__ */ jsxRuntime.jsx(
          "img",
          {
            src: `https://oss.orderly.network/static/network_logo/${chain.chain_id}.png`,
            alt: chain.chain_id,
            className: "oui-relative oui-size-[18px]",
            style: {
              marginLeft: index > 0 ? "-4px" : "0",
              zIndex: supportVaults.length - index
            }
          },
          chain.chain_id
        )) });
      }, [supportVaults]);
      const headerImageElement = React.useMemo(() => {
        if (typeof headerImage === "string") {
          return /* @__PURE__ */ jsxRuntime.jsx(
            "img",
            {
              src: headerImage,
              alt: "header",
              className: "oui-bg-contain oui-bg-center oui-bg-no-repeat"
            }
          );
        }
        return headerImage;
      }, [headerImage]);
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-relative", children: [
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col", children: [
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1 oui-text-2xs oui-font-semibold oui-text-base-contrast-54", children: [
            t("vaults.availableOn"),
            " ",
            supportVaultsList
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mb-4 oui-mt-6 oui-min-h-[56px] oui-max-w-[calc(100%-187px)] oui-text-xl oui-font-bold oui-text-base-contrast", children: t("vaults.header.title") }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-max-w-[450px] oui-text-2xs oui-font-normal oui-text-base-contrast-54", children: t("vaults.header.description", { brokerName }) })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-absolute -oui-right-3 -oui-top-4 oui-h-[124px] oui-w-[187px]", children: headerImageElement })
      ] });
    };
  }
});
exports.useVaultsHeaderScript = void 0;
var init_vaults_header_script = __esm({
  "src/components/vaults-header/vaults-header.script.ts"() {
    init_store();
    exports.useVaultsHeaderScript = () => {
      const { vaultInfo, vaultsPageConfig } = exports.useVaultsStore();
      const brokerName = hooks.useConfig("brokerName");
      const supportVaults = React.useMemo(() => {
        const array = [];
        vaultInfo.data.forEach((vault) => {
          if (Array.isArray(vault?.supported_chains)) {
            array.push(...vault.supported_chains);
          }
        });
        return ramda.uniqBy((item) => item.chain_id, array);
      }, [vaultInfo.data]);
      return {
        supportVaults,
        headerImage: vaultsPageConfig?.headerImage,
        brokerName
      };
    };
  }
});
exports.VaultsHeaderWidget = void 0;
var init_vaults_header_widget = __esm({
  "src/components/vaults-header/vaults-header.widget.tsx"() {
    init_vaults_header_desktop();
    init_vaults_header_mobile();
    init_vaults_header_script();
    exports.VaultsHeaderWidget = () => {
      const state = exports.useVaultsHeaderScript();
      const { isMobile } = ui.useScreen();
      return isMobile ? /* @__PURE__ */ jsxRuntime.jsx(VaultsHeaderMobile, { ...state }) : /* @__PURE__ */ jsxRuntime.jsx(exports.VaultsHeaderDesktop, { ...state });
    };
  }
});

// src/components/vaults-header/index.ts
var vaults_header_exports = {};
__export(vaults_header_exports, {
  VaultsHeaderDesktop: () => exports.VaultsHeaderDesktop,
  VaultsHeaderWidget: () => exports.VaultsHeaderWidget,
  useVaultsHeaderScript: () => exports.useVaultsHeaderScript
});
var init_vaults_header = __esm({
  "src/components/vaults-header/index.ts"() {
    init_vaults_header_widget();
    init_vaults_header_desktop();
    init_vaults_header_script();
  }
});
var VaultsIntroductionDesktop, VaultsIntroductionItem;
var init_vaults_introduction_desktop = __esm({
  "src/components/vaults-introduction/vaults-introduction.desktop.tsx"() {
    VaultsIntroductionDesktop = (props) => {
      const { t } = i18n.useTranslation();
      const { isMobile } = ui.useScreen();
      const { vaultsInfo } = props;
      return /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          className: ui.cn(
            "oui-flex oui-items-center  ",
            "oui-bg-gradient-to-b oui-from-[#26292E] oui-to-[#1B1D22]",
            isMobile ? "oui-gap-3 oui-rounded-xl oui-p-3" : "oui-gap-6 oui-rounded-2xl oui-p-6"
          ),
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              VaultsIntroductionItem,
              {
                title: t("vaults.introduction.tvl"),
                value: vaultsInfo.tvl,
                textProps: { currency: "$" },
                isMobile
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Divider,
              {
                direction: "vertical",
                className: ui.cn(
                  "oui-bg-white/[0.12]",
                  isMobile ? "oui-h-[44px]" : "oui-h-[64px]"
                )
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              VaultsIntroductionItem,
              {
                title: t("common.vaults"),
                value: vaultsInfo.vaultsCount,
                isMobile
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Divider,
              {
                direction: "vertical",
                className: ui.cn(
                  "oui-bg-white/[0.12]",
                  isMobile ? "oui-h-[44px]" : "oui-h-[64px]"
                )
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              VaultsIntroductionItem,
              {
                title: t("vaults.introduction.depositors"),
                value: vaultsInfo.lpCount,
                isMobile
              }
            )
          ]
        }
      );
    };
    VaultsIntroductionItem = (props) => {
      const { title, value, textProps, isMobile } = props;
      return /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          className: ui.cn(
            "oui-flex oui-flex-1 oui-flex-col oui-items-center ",
            isMobile ? "oui-gap-0.5" : "oui-gap-3"
          ),
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              "div",
              {
                className: ui.cn(
                  "oui-font-normal oui-text-base-contrast-54",
                  isMobile ? "oui-text-2xs" : "oui-text-base"
                ),
                children: title
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.numeral,
              {
                className: ui.cn(
                  "oui-font-semibold",
                  isMobile ? "oui-text-base" : "oui-text-xl"
                ),
                dp: 0,
                ...textProps,
                children: value
              }
            )
          ]
        }
      );
    };
  }
});
var useVaultsIntroductionScript;
var init_vaults_introduction_script = __esm({
  "src/components/vaults-introduction/vaults-introduction.script.ts"() {
    init_store();
    useVaultsIntroductionScript = () => {
      const { data: overallInfo, loading, error } = exports.useVaultOverallInfoState();
      const vaultsInfo = React.useMemo(() => {
        if (!overallInfo) {
          return {
            tvl: 0,
            lpCount: 0,
            vaultsCount: 0
          };
        }
        return {
          tvl: overallInfo.strategy_vaults_tvl,
          lpCount: overallInfo.strategy_vaults_lp_count,
          vaultsCount: overallInfo.strategy_vaults_count
        };
      }, [overallInfo]);
      return {
        vaultsInfo,
        loading,
        error
      };
    };
  }
});
exports.VaultsIntroductionWidget = void 0;
var init_vaults_introduction_widget = __esm({
  "src/components/vaults-introduction/vaults-introduction.widget.tsx"() {
    init_vaults_introduction_desktop();
    init_vaults_introduction_script();
    exports.VaultsIntroductionWidget = () => {
      const state = useVaultsIntroductionScript();
      return /* @__PURE__ */ jsxRuntime.jsx(VaultsIntroductionDesktop, { ...state });
    };
  }
});

// src/components/vaults-introduction/index.ts
var vaults_introduction_exports = {};
__export(vaults_introduction_exports, {
  VaultsIntroductionWidget: () => exports.VaultsIntroductionWidget
});
var init_vaults_introduction = __esm({
  "src/components/vaults-introduction/index.ts"() {
    init_vaults_introduction_widget();
  }
});

// src/types/vault.ts
exports.RoleType = void 0; exports.OperationType = void 0;
var init_vault = __esm({
  "src/types/vault.ts"() {
    exports.RoleType = /* @__PURE__ */ ((RoleType2) => {
      RoleType2["LP"] = "lp";
      RoleType2["SP"] = "sp";
      return RoleType2;
    })(exports.RoleType || {});
    exports.OperationType = /* @__PURE__ */ ((OperationType2) => {
      OperationType2["DEPOSIT"] = "deposit";
      OperationType2["WITHDRAWAL"] = "withdrawal";
      return OperationType2;
    })(exports.OperationType || {});
  }
});

// src/contract/index.ts
var VAULTS_CONTRACT_ADDRESSES;
var init_contract = __esm({
  "src/contract/index.ts"() {
    VAULTS_CONTRACT_ADDRESSES = {
      prod: {
        vaultProtocol: "0x70Fe7d65Ac7c1a1732f64d2E6fC0E33622D0C991",
        vaultCrossChainManager: "0x58c9747ccAAE56182C7d9c814F5eaca395D8c93B",
        vaultPvLedger: "0xB7E792f0da9104A27288421583748215AefFFd78",
        vaultId: "0xa3426a1cef4052c056fced18099be899d93f1427d13b9a1df1806b91fad3d0c2",
        spAddress: "0x8bAA309D93FFFB54A64444FD98E10d92D4d9Eb22"
      },
      staging: {
        vaultProtocol: "0x6B6059259B4096ea6420Eb5e08a22214d2303aE9",
        vaultCrossChainManager: "0x510dD61a988797114A9a51b0d228E894037BD9cb",
        vaultPvLedger: "0x20AFe57C75D1C548A9Da265fBFC5416c43783589",
        vaultId: "0x95514fb145354f07bb889f711e856481b5ed52fce52200148aa834b3b29544c8",
        spAddress: "0xbddfd22ef902a4898147a1ca5b985d03c62a8c41"
      },
      qa: {
        vaultProtocol: "0xF5b12d5F1db6DAB8C7c0561152b5e4bb8fD5eb38",
        vaultCrossChainManager: "0xB8a3245407571804b4db5219cBe8D9F2EA1828DE",
        vaultPvLedger: "0x15c9B5705CaCB4eb140dfd7a8c651E18027e5146",
        vaultId: "0x4812cbb88f4025372a3e2acd10d02b5f680d7d1fe78091f6cfde80122c861099",
        spAddress: "0xbddfd22ef902a4898147a1ca5b985d03c62a8c41"
      },
      dev: {
        vaultProtocol: "0xA292E1126703F804FBD5671F034c7226538C54C7",
        vaultCrossChainManager: "0x7568ACC147Af12b02713143C3177D7e89C28A9A6",
        vaultPvLedger: "0x2a5909498d85650744dD5CF964F2136962e5AE6E",
        vaultId: "0x4812cbb88f4025372a3e2acd10d02b5f680d7d1fe78091f6cfde80122c861099",
        spAddress: "0xbddfd22ef902a4898147a1ca5b985d03c62a8c41"
      }
    };
  }
});

// src/utils/operationPayload.ts
var getToAccountPayloadType;
var init_operationPayload = __esm({
  "src/utils/operationPayload.ts"() {
    init_vault();
    getToAccountPayloadType = (type, role) => {
      if (type === "deposit" /* DEPOSIT */ && role === "lp" /* LP */) {
        return 0;
      }
      if (type === "withdrawal" /* WITHDRAWAL */ && role === "lp" /* LP */) {
        return 1;
      }
      if (type === "deposit" /* DEPOSIT */ && role === "sp" /* SP */) {
        return 2;
      }
      if (type === "withdrawal" /* WITHDRAWAL */ && role === "sp" /* SP */) {
        return 3;
      }
      return 1;
    };
  }
});
var useLatestOperationScript;
var init_latest_script = __esm({
  "src/components/vault-operation/depositAndWithdraw/latest.script.ts"() {
    useLatestOperationScript = (props) => {
      const { type, vaultId } = props;
      const { data: operationHistory, mutate: mutateOperationHistory } = hooks.usePrivateQuery(
        `/v1/account_sv_transaction_history?type=${type}&vault_id=${vaultId}&size=1`,
        {
          formatter: (response) => {
            return response?.rows || types.EMPTY_LIST;
          },
          revalidateOnFocus: false
        }
      );
      const latestOperation = React.useMemo(() => {
        return operationHistory?.[0];
      }, [operationHistory]);
      return {
        latestOperation,
        refetch: mutateOperationHistory
      };
    };
  }
});
var useOperationScript;
var init_operation_script = __esm({
  "src/components/vault-operation/depositAndWithdraw/operation.script.ts"() {
    init_contract();
    init_vault();
    init_operationPayload();
    init_latest_script();
    useOperationScript = (props) => {
      const { account, state } = hooks.useAccount();
      const env = hooks.useGetEnv();
      const { track } = hooks.useTrack();
      const verifyingContract = React.useMemo(() => {
        return VAULTS_CONTRACT_ADDRESSES[env].vaultPvLedger;
      }, [env]);
      const [postVaultOperation] = hooks.useMutation("/v1/sv_operation_request");
      const { refetch: refetchOperationHistory } = useLatestOperationScript({
        type: props.type,
        vaultId: props.vaultId
      });
      const disabledOperation = React.useMemo(() => {
        return state?.accountId !== state?.mainAccountId;
      }, [state?.accountId, state?.mainAccountId]);
      const handleOperation = async ({
        amount,
        vaultId
      }) => {
        if (state.accountId !== state.mainAccountId) {
          ui.toast.error(
            "Please switch to the main account to perform this operation."
          );
          return;
        }
        if (props.type === "deposit" /* DEPOSIT */) {
          if (new utils.Decimal(amount).lt(10)) {
            ui.toast.error("Deposit amount is less than the minimum 10 USDC.");
            return;
          }
        }
        const payloadType = getToAccountPayloadType(props.type, "lp" /* LP */);
        const requestParams = {
          payloadType,
          amount: new utils.Decimal(Number(amount)).mul(10 ** 6).toString(),
          vaultId,
          token: "USDC",
          domain: {
            name: "Orderly",
            version: "1",
            chainId: state.connectWallet?.chainId,
            verifyingContract
          }
        };
        const { message, signatured } = await account.generateDexRequest(requestParams);
        try {
          const res = await postVaultOperation({
            message: {
              ...message,
              chainId: state.connectWallet?.chainId
            },
            signature: signatured,
            userAddress: state.address,
            verifyingContract
          });
          if (!res.success) {
            if (props.type === "deposit" /* DEPOSIT */) {
              track(types.TrackerEventName.vaultDepositFailed, {
                msg: res.message
              });
            } else {
              track(types.TrackerEventName.vaultWithdrawFailed, {
                msg: res.message
              });
            }
            ui.toast.error(res.message);
            return;
          }
          await refetchOperationHistory();
          if (props.type === "deposit" /* DEPOSIT */) {
            track(types.TrackerEventName.vaultDepositSuccess, {
              quantity: amount,
              vaultId
            });
          } else {
            track(types.TrackerEventName.vaultWithdrawSuccess, {
              quantity: amount,
              vaultId
            });
          }
          ui.toast.success(`${props.type} successful`);
        } catch (error) {
          const errorMessage = error instanceof Error ? error.message : String(error);
          if (props.type === "deposit" /* DEPOSIT */) {
            track(types.TrackerEventName.vaultDepositFailed, {
              msg: errorMessage
            });
          } else {
            track(types.TrackerEventName.vaultWithdrawFailed, {
              msg: errorMessage
            });
          }
          ui.toast.error(errorMessage);
        }
      };
      return {
        handleOperation,
        disabledOperation
      };
    };
  }
});
var useVaultDepositFormScript;
var init_vault_deposit_form_script = __esm({
  "src/components/vault-operation/deposit/vault-deposit-form.script.ts"() {
    init_vaultsStore();
    init_vault();
    init_operation_script();
    useVaultDepositFormScript = (props) => {
      const { vaultId } = props;
      const [quantity, setQuantity] = React.useState("");
      const { vaultInfo } = exports.useVaultsStore();
      const { handleOperation, disabledOperation } = useOperationScript({
        type: "deposit" /* DEPOSIT */,
        vaultId
      });
      const { holding } = hooks.useCollateral();
      const { t } = i18n.useTranslation();
      const currentVault = React.useMemo(() => {
        return vaultInfo.data.find((v) => v.vault_id === vaultId);
      }, [vaultInfo.data, vaultId]);
      const minDepositAmount = currentVault?.min_deposit_amount || 0;
      const maxWithdrawalAmount = hooks.useMaxWithdrawal("USDC");
      const availableBalance = React.useMemo(() => {
        return holding?.find((h) => h.token === "USDC")?.holding || 0;
      }, [holding]);
      const maxQuantity = React.useMemo(() => {
        return Math.min(maxWithdrawalAmount, availableBalance);
      }, [maxWithdrawalAmount, availableBalance]);
      const sharePrice = currentVault?.est_main_share_price;
      const shares = React.useMemo(() => {
        if (!sharePrice || !quantity) {
          return "-";
        }
        return new utils.Decimal(quantity).div(sharePrice).toString();
      }, [quantity, sharePrice]);
      const handleDeposit = async () => {
        if (!shares || shares === "-") {
          return;
        }
        await handleOperation({
          amount: quantity,
          vaultId
        });
        setQuantity("");
      };
      const token = hooks.useTokenInfo("USDC");
      const sourceToken = React.useMemo(() => {
        return {
          ...token,
          display_name: token?.token,
          symbol: token?.token
        };
      }, [token]);
      const onQuantityChange = (value) => {
        if (value && new utils.Decimal(value).gt(maxQuantity)) {
          setQuantity(maxQuantity.toString());
          return;
        }
        setQuantity(value);
      };
      const disabledDeposit = React.useMemo(() => {
        return !quantity || quantity === "0" || disabledOperation || !!quantity && new utils.Decimal(quantity).lt(minDepositAmount);
      }, [quantity, disabledOperation, minDepositAmount]);
      const inputHint = React.useMemo(() => {
        if (quantity && new utils.Decimal(quantity).lt(minDepositAmount)) {
          return {
            hintMessage: t("vaults.operation.error.minDeposit", {
              amount: minDepositAmount
            }),
            status: "error"
          };
        }
        return {
          hintMessage: "",
          status: ""
        };
      }, [quantity, minDepositAmount, t]);
      return {
        quantity,
        onQuantityChange,
        sourceToken,
        maxQuantity,
        sharePrice,
        shares,
        handleDeposit,
        vaultId,
        disabledDeposit,
        disabledOperation,
        inputHint
      };
    };
  }
});

// src/components/vault-operation/latest-deposit/latest-deposit.script.ts
var useLatestDepositScript;
var init_latest_deposit_script = __esm({
  "src/components/vault-operation/latest-deposit/latest-deposit.script.ts"() {
    init_vault();
    init_latest_script();
    useLatestDepositScript = ({ vaultId }) => {
      const { latestOperation } = useLatestOperationScript({
        type: "deposit" /* DEPOSIT */,
        vaultId
      });
      return {
        latestOperation
      };
    };
  }
});

// src/utils/getOperationStatusColor.ts
var getOperationStatusColor;
var init_getOperationStatusColor = __esm({
  "src/utils/getOperationStatusColor.ts"() {
    getOperationStatusColor = (status) => {
      if (status === "completed") {
        return "success";
      }
      if (status === "rejected" || status === "failed") {
        return "danger";
      }
      return "primary";
    };
  }
});
var LatestDepositUI;
var init_latest_deposit_ui = __esm({
  "src/components/vault-operation/latest-deposit/latest-deposit.ui.tsx"() {
    init_getOperationStatusColor();
    LatestDepositUI = (props) => {
      const { latestOperation } = props;
      const { t } = i18n.useTranslation();
      if (!latestOperation) {
        return null;
      }
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-mt-3 oui-flex oui-flex-col oui-gap-2", children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-xs oui-font-normal oui-text-base-contrast-54", children: t("vaults.deposit.latestDeposit") }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-[44px] oui-rounded-lg oui-border oui-border-white/[0.12] oui-p-3", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center", children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: ui.cn(
                "oui-mr-1 oui-size-1 oui-rounded-full",
                latestOperation.status === "completed" && "oui-bg-success",
                latestOperation.status === "rejected" && "oui-bg-danger",
                latestOperation.status === "failed" && "oui-bg-danger",
                (latestOperation.status === "pending" || latestOperation.status === "new") && "oui-bg-primary"
              )
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              color: getOperationStatusColor(latestOperation.status),
              className: "oui-text-xs oui-font-normal",
              children: latestOperation.status.slice(0, 1).toUpperCase() + latestOperation.status.slice(1)
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(ui.TokenIcon, { name: "USDC", className: "oui-ml-auto oui-mr-1 oui-size-4" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { className: "oui-text-sm oui-font-semibold oui-text-base-contrast-80", children: latestOperation.amount_change })
        ] }) })
      ] });
    };
  }
});
exports.LatestDepositWidget = void 0;
var init_latest_deposit_widget = __esm({
  "src/components/vault-operation/latest-deposit/latest-deposit.widget.tsx"() {
    init_latest_deposit_script();
    init_latest_deposit_ui();
    exports.LatestDepositWidget = (props) => {
      const { vaultId } = props;
      const state = useLatestDepositScript({ vaultId });
      return /* @__PURE__ */ jsxRuntime.jsx(LatestDepositUI, { ...state });
    };
  }
});
var VaultDepositForm;
var init_vault_deposit_form_ui = __esm({
  "src/components/vault-operation/deposit/vault-deposit-form.ui.tsx"() {
    init_latest_deposit_widget();
    VaultDepositForm = (props) => {
      const {
        quantity,
        onQuantityChange,
        sourceToken,
        maxQuantity,
        shares,
        handleDeposit,
        vaultId,
        disabledDeposit,
        disabledOperation,
        inputHint
      } = props;
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
        /* @__PURE__ */ jsxRuntime.jsx(uiTransfer.BrokerWallet, {}),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { mt: 3, mb: 1, children: /* @__PURE__ */ jsxRuntime.jsx(
          uiTransfer.QuantityInput,
          {
            value: quantity,
            onValueChange: onQuantityChange,
            token: { ...sourceToken, precision: 6 },
            testId: "oui-testid-vault-deposit-dialog-quantity-input",
            hintMessage: inputHint.hintMessage,
            status: inputHint.status
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          uiTransfer.AvailableQuantity,
          {
            maxQuantity: maxQuantity?.toString() || "0",
            onClick: () => {
              onQuantityChange(maxQuantity?.toString() || "0");
            }
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsxs(
          ui.Box,
          {
            mt: 5,
            mb: 1,
            className: "oui-flex oui-items-center oui-justify-between oui-text-sm oui-font-semibold oui-text-base-contrast-54",
            children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { children: t("vaults.deposit.estShares") }),
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.numeral,
                {
                  dp: 6,
                  padding: false,
                  suffix: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-ml-1 oui-text-base-contrast-54", children: t("vaults.deposit.shares") }),
                  children: shares
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsxs(
          ui.Box,
          {
            mb: 5,
            className: "oui-flex oui-items-center oui-justify-between oui-text-sm oui-font-semibold oui-text-base-contrast-54",
            children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { children: t("vaults.deposit.lockupDuration") }),
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.Text.numeral,
                {
                  suffix: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-ml-1 oui-text-base-contrast-54", children: "hrs" }),
                  children: "48"
                }
              )
            ]
          }
        ),
        disabledOperation && /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mt-3 oui-text-center", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "warning", className: "oui-text-sm oui-font-semibold", children: t("vaults.operation.error.switchAccount") }) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Button,
          {
            fullWidth: true,
            color: "primary",
            disabled: disabledDeposit,
            className: "oui-mt-3",
            onClick: handleDeposit,
            children: t("common.deposit")
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(exports.LatestDepositWidget, { vaultId })
      ] });
    };
  }
});
exports.VaultDepositWidget = void 0;
var init_vault_deposit_form_widget = __esm({
  "src/components/vault-operation/deposit/vault-deposit-form.widget.tsx"() {
    init_vault_deposit_form_script();
    init_vault_deposit_form_ui();
    exports.VaultDepositWidget = (props) => {
      const { vaultId } = props;
      const state = useVaultDepositFormScript({ vaultId });
      return /* @__PURE__ */ jsxRuntime.jsx(VaultDepositForm, { ...state });
    };
  }
});

// src/components/vault-operation/deposit/index.ts
var init_deposit = __esm({
  "src/components/vault-operation/deposit/index.ts"() {
    init_vault_deposit_form_widget();
  }
});
var useVaultWithdrawFormScript;
var init_vault_withdraw_form_script = __esm({
  "src/components/vault-operation/withdraw/vault-withdraw-form.script.ts"() {
    init_store();
    init_vault();
    init_operation_script();
    useVaultWithdrawFormScript = (props) => {
      const { vaultId } = props;
      const [quantity, setQuantity] = React.useState("");
      const { handleOperation, disabledOperation } = useOperationScript({
        type: "withdrawal" /* WITHDRAWAL */,
        vaultId
      });
      const vaultLpInfo = exports.useVaultLpInfoById(vaultId);
      const { vaultInfo } = exports.useVaultsStore();
      const { t } = i18n.useTranslation();
      const currentVault = React.useMemo(() => {
        return vaultInfo.data.find((v) => v.vault_id === vaultId);
      }, [vaultInfo.data, vaultId]);
      const sharePrice = currentVault?.est_main_share_price || 0;
      const minWithdrawalAmount = currentVault?.min_withdrawal_amount || 0;
      const maxQuantity = vaultLpInfo?.[0]?.available_main_shares || 0;
      const receivingAmount = React.useMemo(() => {
        if (!quantity || !sharePrice) {
          return "-";
        }
        return new utils.Decimal(quantity).mul(sharePrice).toString();
      }, [quantity, sharePrice]);
      const handleWithdraw = async () => {
        await handleOperation({
          amount: quantity,
          vaultId
        });
      };
      const onQuantityChange = (value) => {
        if (value && new utils.Decimal(value).gt(maxQuantity)) {
          setQuantity(maxQuantity.toString());
        } else {
          setQuantity(value);
        }
      };
      const isMinAmount = React.useMemo(() => {
        if (!quantity || !sharePrice || !minWithdrawalAmount) return false;
        const isAll = quantity === maxQuantity.toString();
        if (isAll) {
          return false;
        }
        const receiving = new utils.Decimal(quantity).mul(sharePrice);
        return receiving.lt(minWithdrawalAmount);
      }, [quantity, sharePrice, minWithdrawalAmount, maxQuantity]);
      const disabledWithdraw = React.useMemo(() => {
        return !quantity || quantity === "0" || disabledOperation || !!quantity && isMinAmount;
      }, [quantity, disabledOperation, isMinAmount]);
      const inputHint = React.useMemo(() => {
        if (quantity && isMinAmount) {
          return {
            hintMessage: t("vaults.operation.error.minWithdrawal", {
              amount: minWithdrawalAmount
            }),
            status: "error"
          };
        }
        return {
          hintMessage: "",
          status: ""
        };
      }, [quantity, isMinAmount, minWithdrawalAmount, t]);
      return {
        quantity,
        onQuantityChange,
        maxQuantity,
        handleWithdraw,
        vaultId,
        sharePrice,
        receivingAmount,
        disabledWithdraw,
        disabledOperation,
        inputHint
      };
    };
  }
});

// src/components/vault-operation/latest-withdraw/latest-withdraw-script.ts
var useLatestWithdrawScript;
var init_latest_withdraw_script = __esm({
  "src/components/vault-operation/latest-withdraw/latest-withdraw-script.ts"() {
    init_vault();
    init_latest_script();
    useLatestWithdrawScript = ({ vaultId }) => {
      const { latestOperation } = useLatestOperationScript({
        type: "withdrawal" /* WITHDRAWAL */,
        vaultId
      });
      return {
        latestOperation
      };
    };
  }
});
var ProcessDot, ProcessLine, ProcessText, WithdrawProcess, WithdrawProcessWidget;
var init_withdraw_process_ui = __esm({
  "src/components/vault-operation/withdraw/withdraw-process.ui.tsx"() {
    ProcessDot = () => {
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-size-2 oui-rounded-full oui-bg-white/[0.54]" });
    };
    ProcessLine = () => {
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-px oui-flex-1 oui-bg-white/[0.2]" });
    };
    ProcessText = ({
      children,
      className
    }) => {
      return /* @__PURE__ */ jsxRuntime.jsx(
        "div",
        {
          className: ui.cn(
            "oui-flex-1 oui-text-2xs oui-font-normal oui-text-base-contrast-80",
            className
          ),
          children
        }
      );
    };
    WithdrawProcess = () => {
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center", children: [
          /* @__PURE__ */ jsxRuntime.jsx(ProcessDot, {}),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex-1", children: /* @__PURE__ */ jsxRuntime.jsx(ProcessLine, {}) }),
          /* @__PURE__ */ jsxRuntime.jsx(ProcessDot, {}),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex-1", children: /* @__PURE__ */ jsxRuntime.jsx(ProcessLine, {}) }),
          /* @__PURE__ */ jsxRuntime.jsx(ProcessDot, {})
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-mt-1 oui-flex oui-justify-between", children: [
          /* @__PURE__ */ jsxRuntime.jsx(ProcessText, { children: t("vaults.withdraw.process.initiate") }),
          /* @__PURE__ */ jsxRuntime.jsx(ProcessText, { className: "oui-text-center", children: t("vaults.withdraw.process.vaultProcess") }),
          /* @__PURE__ */ jsxRuntime.jsx(ProcessText, { className: "oui-text-right", children: t("vaults.withdraw.process.transferred") })
        ] })
      ] });
    };
    WithdrawProcessWidget = () => {
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-2 oui-rounded-xl oui-border oui-border-white/[0.12] oui-p-3", children: [
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-justify-between oui-text-sm oui-font-semibold oui-text-base-contrast-54", children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { children: t("vaults.withdraw.process.withdrawProcess") }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "primary", children: t("vaults.withdraw.process.upTo") })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(WithdrawProcess, {})
      ] });
    };
  }
});
var LatestWithdrawUI;
var init_latest_withdraw_ui = __esm({
  "src/components/vault-operation/latest-withdraw/latest-withdraw-ui.tsx"() {
    init_getOperationStatusColor();
    init_withdraw_process_ui();
    LatestWithdrawUI = (props) => {
      const { latestOperation } = props;
      const { t } = i18n.useTranslation();
      if (!latestOperation) {
        return /* @__PURE__ */ jsxRuntime.jsx(WithdrawProcessWidget, {});
      }
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-mt-3 oui-flex oui-flex-col oui-gap-2", children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-xs oui-font-normal oui-text-base-contrast-54", children: t("vaults.withdraw.latestWithdraw") }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-[44px] oui-rounded-lg oui-border oui-border-white/[0.12] oui-p-3", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center", children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: ui.cn(
                "oui-mr-1 oui-size-1 oui-rounded-full",
                latestOperation.status === "completed" && "oui-bg-success",
                latestOperation.status === "rejected" && "oui-bg-danger",
                latestOperation.status === "failed" && "oui-bg-danger",
                (latestOperation.status === "pending" || latestOperation.status === "new") && "oui-bg-primary"
              )
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Text,
            {
              color: getOperationStatusColor(latestOperation.status),
              className: "oui-text-xs oui-font-normal",
              children: latestOperation.status.slice(0, 1).toUpperCase() + latestOperation.status.slice(1)
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(ui.TokenIcon, { name: "USDC", className: "oui-ml-auto oui-mr-1 oui-size-4" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text.numeral, { className: "oui-text-sm oui-font-semibold oui-text-base-contrast-80", children: latestOperation.amount_change || "-" })
        ] }) })
      ] });
    };
  }
});
exports.LatestWithdrawWidget = void 0;
var init_latest_withdraw_widget = __esm({
  "src/components/vault-operation/latest-withdraw/latest-withdraw-widget.tsx"() {
    init_latest_withdraw_script();
    init_latest_withdraw_ui();
    exports.LatestWithdrawWidget = (props) => {
      const { vaultId } = props;
      const state = useLatestWithdrawScript({ vaultId });
      return /* @__PURE__ */ jsxRuntime.jsx(LatestWithdrawUI, { ...state });
    };
  }
});

// src/components/vault-operation/latest-withdraw/index.ts
var init_latest_withdraw = __esm({
  "src/components/vault-operation/latest-withdraw/index.ts"() {
    init_latest_withdraw_widget();
  }
});
var VaultWithdrawForm, WithdrawKVItem, WithdrawInitialContent, WithdrawInitialDialog, WithdrawInitialSheet;
var init_vault_withdraw_form_ui = __esm({
  "src/components/vault-operation/withdraw/vault-withdraw-form.ui.tsx"() {
    init_latest_withdraw();
    init_withdraw_process_ui();
    VaultWithdrawForm = (props) => {
      const {
        quantity,
        onQuantityChange,
        maxQuantity,
        handleWithdraw,
        vaultId,
        sharePrice,
        receivingAmount,
        disabledWithdraw,
        disabledOperation,
        inputHint
      } = props;
      const { isMobile } = ui.useScreen();
      const handleInitialWithdraw = () => {
        ui.modal.show(isMobile ? WithdrawInitialSheet : WithdrawInitialDialog, {
          quantity,
          receivingAmount,
          handleWithdraw
        });
      };
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
        /* @__PURE__ */ jsxRuntime.jsx(uiTransfer.BrokerWallet, {}),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { mt: 3, mb: 1, children: /* @__PURE__ */ jsxRuntime.jsx(
          uiTransfer.QuantityInput,
          {
            value: quantity,
            onValueChange: onQuantityChange,
            token: { display_name: "Shares", precision: 6 },
            testId: "oui-testid-vault-withdraw-dialog-quantity-input",
            hintMessage: inputHint.hintMessage,
            status: inputHint.status
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          uiTransfer.AvailableQuantity,
          {
            maxQuantity,
            onClick: () => {
              onQuantityChange(maxQuantity.toString());
            }
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          WithdrawKVItem,
          {
            label: t("vaults.withdraw.estPricePerShare"),
            currency: "$",
            value: sharePrice || "-",
            className: "oui-mb-1 oui-mt-5"
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          WithdrawKVItem,
          {
            label: t("vaults.withdraw.estReceivingAmount"),
            value: receivingAmount,
            suffix: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-ml-1 oui-text-base-contrast-36", children: "USDC" }),
            className: "oui-mb-5"
          }
        ),
        disabledOperation && /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mt-3 oui-text-center", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "warning", className: "oui-text-sm oui-font-semibold", children: t("vaults.operation.error.switchAccount") }) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Button,
          {
            fullWidth: true,
            color: "primary",
            disabled: disabledWithdraw,
            className: "oui-mt-3",
            onClick: handleInitialWithdraw,
            children: t("common.withdraw")
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-mt-3", children: /* @__PURE__ */ jsxRuntime.jsx(exports.LatestWithdrawWidget, { vaultId }) })
      ] });
    };
    WithdrawKVItem = ({
      label,
      value,
      suffix,
      currency,
      className
    }) => {
      return /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          className: ui.cn(
            "oui-flex oui-items-center oui-justify-between oui-text-sm oui-font-semibold oui-text-base-contrast-54",
            className
          ),
          children: [
            /* @__PURE__ */ jsxRuntime.jsx("div", { children: label }),
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.numeral,
              {
                currency,
                className: "oui-text-base-contrast",
                suffix,
                children: value
              }
            )
          ]
        }
      );
    };
    WithdrawInitialContent = (props) => {
      const { quantity, receivingAmount, handleWithdraw, hide } = props;
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col", children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          WithdrawKVItem,
          {
            label: t("vaults.withdraw.dialog.withdrawalAmount"),
            value: quantity,
            suffix: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-ml-1 oui-text-base-contrast-36", children: "Shares" })
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          WithdrawKVItem,
          {
            label: t("vaults.withdraw.dialog.estimatedReceiving"),
            value: receivingAmount,
            suffix: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-ml-1 oui-text-base-contrast-36", children: "USDC" }),
            className: "oui-mt-1"
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { color: "warning", className: "oui-my-5 oui-text-sm oui-font-semibold", children: t("vaults.withdraw.dialog.note") }),
        /* @__PURE__ */ jsxRuntime.jsx(WithdrawProcessWidget, {}),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Button,
          {
            fullWidth: true,
            color: "primary",
            className: "oui-mt-5",
            onClick: async () => {
              await handleWithdraw();
              hide?.();
            },
            children: t("vaults.withdraw.dialog.initiateWithdrawal")
          }
        )
      ] });
    };
    WithdrawInitialDialog = ui.modal.create(
      ({
        quantity,
        receivingAmount,
        handleWithdraw
      }) => {
        const { visible, hide, onOpenChange } = ui.useModal();
        const { t } = i18n.useTranslation();
        return /* @__PURE__ */ jsxRuntime.jsx(
          ui.SimpleDialog,
          {
            title: t("vaults.withdraw.dialog.title"),
            open: visible,
            onOpenChange,
            children: /* @__PURE__ */ jsxRuntime.jsx(
              WithdrawInitialContent,
              {
                quantity,
                receivingAmount,
                handleWithdraw,
                hide
              }
            )
          }
        );
      }
    );
    WithdrawInitialSheet = ui.modal.create(
      ({
        quantity,
        receivingAmount,
        handleWithdraw
      }) => {
        const { visible, hide, onOpenChange } = ui.useModal();
        const { t } = i18n.useTranslation();
        return /* @__PURE__ */ jsxRuntime.jsx(
          ui.SimpleSheet,
          {
            title: t("vaults.withdraw.dialog.title"),
            open: visible,
            onOpenChange,
            children: /* @__PURE__ */ jsxRuntime.jsx(
              WithdrawInitialContent,
              {
                quantity,
                receivingAmount,
                handleWithdraw,
                hide
              }
            )
          }
        );
      }
    );
  }
});
exports.VaultWithdrawWidget = void 0;
var init_vault_withdraw_form_widget = __esm({
  "src/components/vault-operation/withdraw/vault-withdraw-form.widget.tsx"() {
    init_vault_withdraw_form_script();
    init_vault_withdraw_form_ui();
    exports.VaultWithdrawWidget = (props) => {
      const { vaultId } = props;
      const state = useVaultWithdrawFormScript({ vaultId });
      return /* @__PURE__ */ jsxRuntime.jsx(VaultWithdrawForm, { ...state });
    };
  }
});

// src/components/vault-operation/withdraw/index.ts
var init_withdraw = __esm({
  "src/components/vault-operation/withdraw/index.ts"() {
    init_vault_withdraw_form_widget();
  }
});
exports.VaultDepositAndWithdrawWithDialogId = void 0; exports.VaultDepositAndWithdrawWithSheetId = void 0; exports.VaultDepositAndWithdraw = void 0;
var init_depositAndWithdraw = __esm({
  "src/components/vault-operation/depositAndWithdraw/index.tsx"() {
    init_deposit();
    init_withdraw();
    exports.VaultDepositAndWithdrawWithDialogId = "VaultDepositAndWithdrawWithDialogId";
    exports.VaultDepositAndWithdrawWithSheetId = "VaultDepositAndWithdrawWithSheetId";
    exports.VaultDepositAndWithdraw = (props) => {
      const [activeTab, setActiveTab] = React.useState(
        props.activeTab || "deposit"
      );
      const { vaultId } = props;
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsxs(
        ui.Tabs,
        {
          value: activeTab,
          onValueChange: setActiveTab,
          variant: "contained",
          size: "lg",
          classNames: {
            tabsList: "oui-px-0",
            tabsContent: "oui-pt-5"
          },
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.TabPanel,
              {
                title: t("common.deposit"),
                icon: /* @__PURE__ */ jsxRuntime.jsx(ui.ArrowDownSquareFillIcon, {}),
                value: "deposit",
                children: /* @__PURE__ */ jsxRuntime.jsx(exports.VaultDepositWidget, { vaultId })
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              ui.TabPanel,
              {
                title: t("common.withdraw"),
                icon: /* @__PURE__ */ jsxRuntime.jsx(ui.ArrowUpSquareFillIcon, {}),
                value: "withdraw",
                children: /* @__PURE__ */ jsxRuntime.jsx(exports.VaultWithdrawWidget, { vaultId })
              }
            )
          ]
        }
      );
    };
    ui.registerSimpleDialog(
      exports.VaultDepositAndWithdrawWithDialogId,
      exports.VaultDepositAndWithdraw,
      {
        size: "md",
        classNames: {
          content: "oui-border oui-border-line-6"
        }
      }
    );
    ui.registerSimpleSheet(
      exports.VaultDepositAndWithdrawWithSheetId,
      exports.VaultDepositAndWithdraw
    );
  }
});

// src/components/vault-card/constants.ts
exports.ORDERLY_ICON = void 0; exports.ORDERLY_VAULT_TITLE = void 0; exports.ORDERLY_VAULT_DESCRIPTION = void 0; exports.getBrokerIconUrl = void 0;
var init_constants = __esm({
  "src/components/vault-card/constants.ts"() {
    exports.ORDERLY_ICON = "https://oss.orderly.network/static/symbol_logo/ORDER.png";
    exports.ORDERLY_VAULT_TITLE = "Orderly OmniVault";
    exports.ORDERLY_VAULT_DESCRIPTION = "Earn passive yields effortlessly, no trading expertise required. OmniVault deploys market-making strategies, taking on liquidations, and accrue platform fees.";
    exports.getBrokerIconUrl = (brokerId) => {
      switch (brokerId) {
        case "orderly":
          return "https://oss.orderly.network/static/symbol_logo/ORDER.png";
        case "woofi_pro":
          return "https://oss.orderly.network/static/broker_logo/woofi_pro.png";
        case "aden":
          return "/vaults/broker/aden.png";
        case "vooi":
          return "https://oss.orderly.network/static/broker_logo/vooi.png";
        default:
          return `https://oss.orderly.network/static/broker_logo/${brokerId}.png`;
      }
    };
  }
});
exports.useVaultCardScript = void 0;
var init_vaultCard_script = __esm({
  "src/components/vault-card/vaultCard.script.ts"() {
    init_env();
    init_useSVAPIUrl();
    init_vaultsStore();
    init_depositAndWithdraw();
    init_constants();
    exports.useVaultCardScript = (vault) => {
      const { t } = i18n.useTranslation();
      const vaultLpInfo = exports.useVaultLpInfoById(vault.vault_id);
      const { fetchVaultLpInfo } = exports.useVaultsStore();
      const env = hooks.useGetEnv();
      const { isMobile } = ui.useScreen();
      const { state } = hooks.useAccount();
      const svApiUrl = useSVApiUrl();
      const { holding } = hooks.useCollateral();
      const availableBalance = React.useMemo(() => {
        return holding?.find((h) => h.token === "USDC")?.holding || 0;
      }, [holding]);
      React.useEffect(() => {
        if (!state.address || !svApiUrl || !vault.vault_id) {
          return;
        }
        fetchVaultLpInfo(
          { vault_id: vault.vault_id, wallet_address: state.address },
          svApiUrl
        );
      }, [vault.vault_id, state.address, svApiUrl]);
      const memoizedAvailableBalance = React.useMemo(() => {
        return availableBalance;
      }, [availableBalance]);
      const lpInfo = React.useMemo(() => {
        const info = vaultLpInfo?.[0];
        if (!info) {
          return {
            deposits: "--",
            earnings: "--"
          };
        }
        return {
          deposits: info.lp_tvl - info.potential_pnl,
          earnings: info.potential_pnl
        };
      }, [vaultLpInfo]);
      const isEVMConnected = React.useMemo(() => {
        return state.chainNamespace === "EVM";
      }, [state.chainNamespace]);
      const isSOLConnected = React.useMemo(() => {
        return state.chainNamespace === "SOL";
      }, [state.chainNamespace]);
      const isWrongNetwork = React.useMemo(() => {
        return state.chainNamespace !== "EVM" && state.chainNamespace !== "SOL" && state.accountId !== state.mainAccountId;
      }, [state.chainNamespace, state.accountId, state.mainAccountId]);
      const isButtonsDisabled = React.useMemo(() => {
        return vault.status !== "live";
      }, [vault.status]);
      const openDepositAndWithdraw = (activeTab) => {
        ui.modal.show(
          isMobile ? exports.VaultDepositAndWithdrawWithSheetId : exports.VaultDepositAndWithdrawWithDialogId,
          {
            activeTab,
            vaultId: vault.vault_id
          }
        );
      };
      const openVaultWebsite = () => {
        const baseUrl = VAULTS_WEBSITE_URLS[env] || VAULTS_WEBSITE_URLS.prod;
        window.open(`${baseUrl}/vaults/${vault.vault_address}`, "_blank");
      };
      return {
        title: vault.vault_name,
        description: vault.description,
        icon: exports.getBrokerIconUrl(vault.broker_id),
        vaultInfo: vault,
        lpInfo,
        isEVMConnected,
        isSOLConnected,
        openDepositAndWithdraw,
        availableBalance: memoizedAvailableBalance,
        openVaultWebsite,
        isWrongNetwork,
        isButtonsDisabled
      };
    };
  }
});
var parseMarkdownLinks;
var init_parseMarkdownLinks = __esm({
  "src/utils/parseMarkdownLinks.tsx"() {
    parseMarkdownLinks = (text) => {
      if (!text) {
        return [""];
      }
      const markdownLinkRegex = /\[([^\]]+)\]\(([^)]+)\)/g;
      const parts = [];
      let lastIndex = 0;
      let match;
      while ((match = markdownLinkRegex.exec(text)) !== null) {
        const [fullMatch, linkText, url] = match;
        const matchIndex = match.index;
        if (matchIndex > lastIndex) {
          parts.push(text.substring(lastIndex, matchIndex));
        }
        parts.push(
          /* @__PURE__ */ jsxRuntime.jsx(
            "a",
            {
              href: url,
              target: "_blank",
              rel: "noopener noreferrer",
              className: "oui-underline hover:oui-text-base-contrast-80",
              children: linkText
            },
            matchIndex
          )
        );
        lastIndex = matchIndex + fullMatch.length;
      }
      if (lastIndex < text.length) {
        parts.push(text.substring(lastIndex));
      }
      return parts.length > 0 ? parts : [text];
    };
  }
});
exports.formatAllTimeReturn = void 0; exports.VaultCard = void 0; var VaultInfoItem, LpInfoItem, VaultCardOperation;
var init_vaultCard_ui = __esm({
  "src/components/vault-card/vaultCard.ui.tsx"() {
    init_parseMarkdownLinks();
    exports.formatAllTimeReturn = (status, vaultAge, lifetimeApy) => {
      if (status === "pre_launch" || vaultAge !== null && vaultAge < 7) {
        return "--";
      }
      if (lifetimeApy > 100) {
        return ">10000%";
      }
      return new utils.Decimal(lifetimeApy).mul(100).toDecimalPlaces(2, utils.Decimal.ROUND_UP).toFixed(2) + "%";
    };
    exports.VaultCard = (props) => {
      const {
        title,
        vaultInfo,
        lpInfo,
        description,
        isEVMConnected,
        isSOLConnected,
        openDepositAndWithdraw,
        availableBalance,
        openVaultWebsite,
        icon,
        isButtonsDisabled
      } = props;
      const { t } = i18n.useTranslation();
      const { isMobile } = ui.useScreen();
      const formattedAllTimeReturn = React.useMemo(
        () => exports.formatAllTimeReturn(
          vaultInfo.status,
          vaultInfo.vault_age,
          vaultInfo.lifetime_apy
        ),
        [vaultInfo.status, vaultInfo.vault_age, vaultInfo.lifetime_apy]
      );
      const getStatusTag = () => {
        const { status } = vaultInfo;
        if (status === "live") {
          return {
            text: t("vaults.card.status.active"),
            color: "#00C076",
            bgColor: "rgba(0, 192, 118, 0.15)"
          };
        } else if (status === "pre_launch") {
          return {
            text: t("vaults.card.launchingSoon"),
            color: "#E88800",
            bgColor: "rgba(232, 136, 0, 0.15)"
          };
        } else if (status === "closing") {
          return {
            text: t("vaults.card.status.closing"),
            color: "#FF6B6B",
            bgColor: "rgba(255, 107, 107, 0.15)"
          };
        } else if (status === "closed") {
          return {
            text: t("vaults.card.status.closed"),
            color: "#999999",
            bgColor: "rgba(153, 153, 153, 0.15)"
          };
        }
        return null;
      };
      const statusTag = getStatusTag();
      React.useMemo(() => {
        const chains = Array.isArray(vaultInfo?.supported_chains) ? vaultInfo.supported_chains : [];
        return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-items-center", children: chains.map((chain, index) => /* @__PURE__ */ jsxRuntime.jsx(
          "img",
          {
            src: `https://oss.orderly.network/static/network_logo/${chain.chain_id}.png`,
            alt: chain.chain_id,
            className: ui.cn(
              "oui-relative",
              isMobile ? "oui-size-[18px]" : "oui-size-5"
            ),
            style: {
              marginLeft: index > 0 ? "-4px" : "0",
              zIndex: chains.length - index
            }
          },
          chain.chain_id
        )) });
      }, [vaultInfo.supported_chains, isMobile]);
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-h-[435px] oui-overflow-hidden oui-rounded-2xl oui-bg-base-9", children: /* @__PURE__ */ jsxRuntime.jsx(
        "div",
        {
          className: ui.cn(
            "oui-absolute oui-left-0 oui-top-0 oui-z-20 oui-size-full oui-p-6",
            "oui-overflow-y-auto oui-custom-scrollbar"
          ),
          children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-2", children: [
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1", children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-[18px] oui-font-semibold oui-text-white oui-break-words", children: title }),
              statusTag && /* @__PURE__ */ jsxRuntime.jsx(
                "div",
                {
                  className: "oui-flex oui-w-fit oui-items-center oui-gap-[10px] oui-px-2",
                  style: {
                    height: "18px",
                    borderRadius: "4px",
                    background: statusTag.bgColor,
                    color: statusTag.color,
                    fontSize: "12px",
                    fontWeight: 600,
                    lineHeight: "18px",
                    letterSpacing: "0.36px"
                  },
                  children: statusTag.text
                }
              )
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1", children: [
              /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-h-[54px]", children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-2xs oui-font-normal oui-leading-[18px] oui-text-base-contrast-54 oui-line-clamp-3", children: parseMarkdownLinks(description) }) }),
              /* @__PURE__ */ jsxRuntime.jsxs(
                "button",
                {
                  onClick: openVaultWebsite,
                  className: "oui-flex oui-w-fit oui-items-center oui-gap-1 oui-text-xs oui-font-medium",
                  style: { color: "#608CFF" },
                  children: [
                    /* @__PURE__ */ jsxRuntime.jsx("span", { children: t("vaults.card.viewMore") }),
                    /* @__PURE__ */ jsxRuntime.jsx(
                      ui.ArrowRightUpSquareFillIcon,
                      {
                        style: { color: "#608CFF" },
                        width: 16,
                        height: 16,
                        viewBox: "0 0 18 18"
                      }
                    )
                  ]
                }
              )
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-6", children: [
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-2", children: [
                /* @__PURE__ */ jsxRuntime.jsx(
                  VaultInfoItem,
                  {
                    label: t("vaults.card.tvl"),
                    value: vaultInfo.tvl,
                    textProps: {
                      currency: "$",
                      dp: 0,
                      type: "numeral"
                    }
                  }
                ),
                /* @__PURE__ */ jsxRuntime.jsx(
                  VaultInfoItem,
                  {
                    label: t("vaults.card.allTimeReturn"),
                    value: formattedAllTimeReturn,
                    textProps: {
                      color: "brand",
                      type: "gradient"
                    },
                    showTooltip: vaultInfo.status === "pre_launch" || vaultInfo.vault_age !== null && vaultInfo.vault_age < 7,
                    tooltipContent: t("vaults.card.allTimeReturnTooltip")
                  }
                )
              ] }),
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-items-center oui-gap-2 oui-rounded-lg oui-bg-white/[0.06] oui-p-3", children: [
                /* @__PURE__ */ jsxRuntime.jsx(
                  LpInfoItem,
                  {
                    label: t("vaults.card.myDeposits"),
                    value: lpInfo.deposits
                  }
                ),
                /* @__PURE__ */ jsxRuntime.jsx(
                  LpInfoItem,
                  {
                    label: t("vaults.card.myEarnings"),
                    value: lpInfo.earnings
                  }
                )
              ] })
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-4 oui-mt-2", children: [
              /* @__PURE__ */ jsxRuntime.jsx(
                LpInfoItem,
                {
                  label: t("vaults.card.accountBalance"),
                  value: availableBalance
                }
              ),
              /* @__PURE__ */ jsxRuntime.jsx(
                VaultCardOperation,
                {
                  isEVMConnected,
                  isSOLConnected,
                  openDepositAndWithdraw,
                  isButtonsDisabled
                }
              )
            ] })
          ] })
        }
      ) });
    };
    VaultInfoItem = (props) => {
      const { label, value, textProps, showTooltip, tooltipContent } = props;
      return /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          className: ui.cn(
            "oui-flex oui-flex-1 oui-flex-col oui-items-center oui-justify-center oui-px-3 oui-py-2",
            "oui-rounded-lg oui-border oui-border-solid oui-border-white/[0.12]"
          ),
          children: [
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1 oui-text-2xs oui-font-normal oui-leading-[18px] oui-text-base-contrast-54", children: [
              label,
              showTooltip && tooltipContent && /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { content: tooltipContent, delayDuration: 100, children: /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(
                ui.InfoCircleIcon,
                {
                  size: 12,
                  opacity: 0.54,
                  className: "oui-text-base-contrast"
                }
              ) }) })
            ] }),
            textProps.type === "gradient" ? /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.gradient,
              {
                className: "oui-text-base oui-font-semibold",
                ...textProps,
                children: value
              }
            ) : /* @__PURE__ */ jsxRuntime.jsx(
              ui.Text.numeral,
              {
                className: "oui-text-base oui-font-semibold",
                ...textProps,
                children: value
              }
            )
          ]
        }
      );
    };
    LpInfoItem = (props) => {
      const { label, value, textProps } = props;
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-w-full oui-items-center oui-justify-between oui-text-2xs oui-font-normal", children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-base-contrast-54", children: label }),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.numeral,
          {
            className: "oui-text-base-contrast-80",
            ...textProps,
            dp: 2,
            suffix: /* @__PURE__ */ jsxRuntime.jsx("span", { className: "oui-ml-1 oui-text-base-contrast-36", children: "USDC" }),
            children: value
          }
        )
      ] });
    };
    VaultCardOperation = (props) => {
      const {
        isEVMConnected,
        isSOLConnected,
        openDepositAndWithdraw,
        isButtonsDisabled
      } = props;
      const { t } = i18n.useTranslation();
      return /* @__PURE__ */ jsxRuntime.jsx(uiConnector.AuthGuard, { buttonProps: { size: "md", fullWidth: true }, children: isEVMConnected || isSOLConnected ? /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-2", children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Button,
          {
            className: "oui-flex-1",
            size: "md",
            disabled: isButtonsDisabled,
            onClick: () => openDepositAndWithdraw("deposit"),
            children: t("common.deposit")
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Button,
          {
            className: "oui-flex-1",
            size: "md",
            color: "secondary",
            disabled: isButtonsDisabled,
            onClick: () => openDepositAndWithdraw("withdraw"),
            children: t("common.withdraw")
          }
        )
      ] }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { size: "md", color: "warning", children: t("connector.wrongNetwork") }) });
    };
  }
});
exports.VaultCardWidget = void 0;
var init_vaultCard_widget = __esm({
  "src/components/vault-card/vaultCard.widget.tsx"() {
    init_vaultCard_script();
    init_vaultCard_ui();
    exports.VaultCardWidget = (props) => {
      const { vault } = props;
      const state = exports.useVaultCardScript(vault);
      return /* @__PURE__ */ jsxRuntime.jsx(exports.VaultCard, { ...state });
    };
  }
});

// src/components/vault-card/index.ts
var init_vault_card = __esm({
  "src/components/vault-card/index.ts"() {
    init_vaultCard_widget();
    init_vaultCard_script();
    init_vaultCard_ui();
    init_constants();
  }
});
exports.VaultsList = void 0; var VaultListRow;
var init_vaults_list = __esm({
  "src/components/all-vaults/vaults-list.tsx"() {
    init_vaultCard_script();
    init_vaultCard_ui();
    exports.VaultsList = ({ vaults }) => {
      const { t } = i18n.useTranslation();
      const [sortField, setSortField] = React.useState("apy");
      const [sortDirection, setSortDirection] = React.useState("desc");
      const shouldShowApyTooltip = React.useMemo(() => {
        return vaults.some(
          (vault) => vault.status === "pre_launch" || vault.vault_age !== null && vault.vault_age < 7
        );
      }, [vaults]);
      const handleSort = (field) => {
        if (sortField === field) {
          setSortDirection(sortDirection === "asc" ? "desc" : "asc");
        } else {
          setSortField(field);
          setSortDirection("desc");
        }
      };
      const sortedVaults = React.useMemo(() => {
        return [...vaults].sort((a, b) => {
          let aValue = 0;
          let bValue = 0;
          switch (sortField) {
            case "tvl":
              aValue = a.tvl;
              bValue = b.tvl;
              break;
            case "apy":
              aValue = a.lifetime_apy;
              bValue = b.lifetime_apy;
              break;
            // For deposits, pnl, balance - these need LP info which we'll handle in VaultListRow
            default:
              return 0;
          }
          return sortDirection === "asc" ? aValue - bValue : bValue - aValue;
        });
      }, [vaults, sortField, sortDirection]);
      const SortIcon = ({ field }) => {
        const isActive = sortField === field;
        return /* @__PURE__ */ jsxRuntime.jsxs(
          "svg",
          {
            width: "12",
            height: "12",
            viewBox: "0 0 12 12",
            fill: "none",
            xmlns: "http://www.w3.org/2000/svg",
            className: ui.cn(
              "oui-ml-1 oui-transition-colors",
              isActive ? "oui-text-base-contrast" : "oui-text-base-contrast-36"
            ),
            children: [
              /* @__PURE__ */ jsxRuntime.jsx(
                "path",
                {
                  d: "M6 2L8 4H4L6 2Z",
                  fill: "currentColor",
                  opacity: isActive && sortDirection === "asc" ? 1 : 0.3
                }
              ),
              /* @__PURE__ */ jsxRuntime.jsx(
                "path",
                {
                  d: "M6 10L4 8H8L6 10Z",
                  fill: "currentColor",
                  opacity: isActive && sortDirection === "desc" ? 1 : 0.3
                }
              )
            ]
          }
        );
      };
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-3", children: [
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-grid oui-grid-cols-[2fr_1fr_1fr_1fr_1fr_1fr_1.5fr] oui-gap-4 oui-px-4 oui-py-3", children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-2xs oui-font-normal oui-text-base-contrast-54", children: t("vaults.list.poolName") }),
          /* @__PURE__ */ jsxRuntime.jsxs(
            "button",
            {
              onClick: () => handleSort("tvl"),
              className: "oui-flex oui-items-center oui-text-2xs oui-font-normal oui-text-base-contrast-54 hover:oui-text-base-contrast",
              children: [
                "TVL",
                /* @__PURE__ */ jsxRuntime.jsx(SortIcon, { field: "tvl" })
              ]
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-gap-1", children: [
            /* @__PURE__ */ jsxRuntime.jsxs(
              "button",
              {
                onClick: () => handleSort("apy"),
                className: "oui-flex oui-items-center oui-text-2xs oui-font-normal oui-text-base-contrast-54 hover:oui-text-base-contrast",
                children: [
                  t("vaults.card.allTimeReturn"),
                  /* @__PURE__ */ jsxRuntime.jsx(SortIcon, { field: "apy" })
                ]
              }
            ),
            shouldShowApyTooltip && /* @__PURE__ */ jsxRuntime.jsx(
              ui.Tooltip,
              {
                content: t("vaults.card.allTimeReturnTooltip"),
                delayDuration: 100,
                children: /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(
                  ui.InfoCircleIcon,
                  {
                    size: 12,
                    opacity: 0.54,
                    className: "oui-text-base-contrast"
                  }
                ) })
              }
            )
          ] }),
          /* @__PURE__ */ jsxRuntime.jsxs(
            "button",
            {
              onClick: () => handleSort("deposits"),
              className: "oui-flex oui-items-center oui-text-2xs oui-font-normal oui-text-base-contrast-54 hover:oui-text-base-contrast",
              children: [
                t("vaults.list.myDeposits"),
                /* @__PURE__ */ jsxRuntime.jsx(SortIcon, { field: "deposits" })
              ]
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs(
            "button",
            {
              onClick: () => handleSort("pnl"),
              className: "oui-flex oui-items-center oui-text-2xs oui-font-normal oui-text-base-contrast-54 hover:oui-text-base-contrast",
              children: [
                t("vaults.list.allTimePnl"),
                /* @__PURE__ */ jsxRuntime.jsx(SortIcon, { field: "pnl" })
              ]
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-2xs oui-font-normal oui-text-base-contrast-54", children: t("vaults.card.accountBalance") }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-2xs oui-font-normal oui-text-base-contrast-54", children: t("vaults.list.operate") })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-flex-col oui-gap-2", children: sortedVaults.map((vault) => /* @__PURE__ */ jsxRuntime.jsx(VaultListRow, { vault }, vault.vault_id)) })
      ] });
    };
    VaultListRow = ({ vault }) => {
      const { t } = i18n.useTranslation();
      const script = exports.useVaultCardScript(vault);
      const {
        title,
        vaultInfo,
        lpInfo,
        isEVMConnected,
        isSOLConnected,
        openDepositAndWithdraw,
        availableBalance,
        openVaultWebsite,
        icon,
        isButtonsDisabled
      } = script;
      vaultInfo.status === "pre_launch";
      const { isMobile } = ui.useScreen();
      const getStatusTag = () => {
        const { status } = vaultInfo;
        if (status === "live") {
          return {
            text: t("vaults.card.status.active"),
            color: "#00C076",
            bgColor: "rgba(0, 192, 118, 0.15)"
          };
        } else if (status === "pre_launch") {
          return {
            text: t("vaults.card.launchingSoon"),
            color: "#E88800",
            bgColor: "rgba(232, 136, 0, 0.15)"
          };
        } else if (status === "closing") {
          return {
            text: t("vaults.card.status.closing"),
            color: "#FF6B6B",
            bgColor: "rgba(255, 107, 107, 0.15)"
          };
        } else if (status === "closed") {
          return {
            text: t("vaults.card.status.closed"),
            color: "#999999",
            bgColor: "rgba(153, 153, 153, 0.15)"
          };
        }
        return null;
      };
      const statusTag = getStatusTag();
      const formattedAllTimeReturn = React.useMemo(
        () => exports.formatAllTimeReturn(
          vaultInfo.status,
          vaultInfo.vault_age,
          vaultInfo.lifetime_apy
        ),
        [vaultInfo.status, vaultInfo.vault_age, vaultInfo.lifetime_apy]
      );
      React.useMemo(() => {
        const chains = Array.isArray(vaultInfo?.supported_chains) ? vaultInfo.supported_chains : [];
        const ICON_SIZE = isMobile ? 18 : 20;
        return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-items-center", children: chains.map((chain, index) => /* @__PURE__ */ jsxRuntime.jsx(
          "img",
          {
            src: `https://oss.orderly.network/static/network_logo/${chain.chain_id}.png`,
            alt: chain.chain_id,
            className: ui.cn(
              "oui-relative",
              isMobile ? "oui-size-[18px]" : "oui-size-5"
            ),
            style: {
              marginLeft: index > 0 ? "-4px" : "0",
              zIndex: chains.length - index,
              width: ICON_SIZE,
              height: ICON_SIZE
            }
          },
          chain.chain_id
        )) });
      }, [vaultInfo.supported_chains, isMobile]);
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-relative oui-grid oui-grid-cols-[2fr_1fr_1fr_1fr_1fr_1fr_1.5fr] oui-gap-4 oui-rounded-2xl oui-bg-base-9 oui-px-4 oui-py-4 oui-items-center oui-overflow-hidden", children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10 oui-flex oui-items-center oui-gap-2", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-1", children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-sm oui-font-semibold oui-text-base-contrast", children: title }),
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-wrap oui-items-center oui-gap-2", children: [
            statusTag && /* @__PURE__ */ jsxRuntime.jsx(
              "div",
              {
                className: "oui-flex oui-items-center oui-gap-[10px] oui-px-2 oui-whitespace-nowrap oui-flex-shrink-0",
                style: {
                  height: "18px",
                  borderRadius: "4px",
                  background: statusTag.bgColor,
                  color: statusTag.color,
                  fontSize: "12px",
                  fontWeight: 600,
                  lineHeight: "18px",
                  letterSpacing: "0.36px"
                },
                children: statusTag.text
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsxs(
              "button",
              {
                onClick: openVaultWebsite,
                className: "oui-flex oui-items-center oui-gap-1 oui-text-xs oui-font-medium",
                style: { color: "#608CFF" },
                children: [
                  /* @__PURE__ */ jsxRuntime.jsx("span", { children: t("vaults.card.viewMore") }),
                  /* @__PURE__ */ jsxRuntime.jsx(
                    ui.ArrowRightUpSquareFillIcon,
                    {
                      style: { color: "#608CFF" },
                      width: 16,
                      height: 16,
                      viewBox: "0 0 18 18"
                    }
                  )
                ]
              }
            )
          ] })
        ] }) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.numeral,
          {
            className: "oui-text-sm oui-font-normal oui-text-base-contrast",
            currency: "$",
            dp: 0,
            children: vaultInfo.tvl
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text.gradient, { className: "oui-text-sm oui-font-semibold", color: "brand", children: formattedAllTimeReturn }) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.numeral,
          {
            className: "oui-text-sm oui-font-normal oui-text-base-contrast",
            dp: 2,
            children: lpInfo.deposits
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.numeral,
          {
            className: "oui-text-sm oui-font-normal oui-text-base-contrast",
            dp: 2,
            children: lpInfo.earnings
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10", children: /* @__PURE__ */ jsxRuntime.jsx(
          ui.Text.numeral,
          {
            className: "oui-text-sm oui-font-normal oui-text-base-contrast",
            dp: 2,
            children: availableBalance
          }
        ) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-relative oui-z-10 oui-flex oui-items-center oui-gap-2", children: /* @__PURE__ */ jsxRuntime.jsx(uiConnector.AuthGuard, { buttonProps: { size: "sm" }, children: isEVMConnected || isSOLConnected ? /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Button,
            {
              size: "sm",
              className: "oui-flex-1",
              disabled: isButtonsDisabled,
              onClick: () => openDepositAndWithdraw("deposit"),
              children: t("common.deposit")
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            ui.Button,
            {
              size: "sm",
              color: "secondary",
              className: "oui-flex-1",
              disabled: isButtonsDisabled,
              onClick: () => openDepositAndWithdraw("withdraw"),
              children: t("common.withdraw")
            }
          )
        ] }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { size: "sm", color: "warning", fullWidth: true, children: t("connector.wrongNetwork") }) }) })
      ] });
    };
  }
});
exports.ViewModeToggle = void 0;
var init_view_mode_toggle = __esm({
  "src/components/all-vaults/view-mode-toggle.tsx"() {
    exports.ViewModeToggle = ({ mode, onChange }) => {
      const toggleMode = () => {
        onChange(mode === "grid" ? "list" : "grid");
      };
      return /* @__PURE__ */ jsxRuntime.jsx(
        "button",
        {
          onClick: toggleMode,
          className: "oui-flex oui-h-8 oui-w-8 oui-items-center oui-justify-center oui-rounded oui-transition-colors hover:oui-bg-base-6",
          "aria-label": mode === "grid" ? "Switch to list view" : "Switch to grid view",
          children: mode === "grid" ? /* @__PURE__ */ jsxRuntime.jsx(
            "svg",
            {
              width: "18",
              height: "18",
              viewBox: "0 0 18 18",
              fill: "none",
              xmlns: "http://www.w3.org/2000/svg",
              children: /* @__PURE__ */ jsxRuntime.jsx(
                "path",
                {
                  d: "M2 4h14v2H2V4zm0 4h14v2H2V8zm0 4h14v2H2v-2z",
                  fill: "currentColor",
                  opacity: "0.54"
                }
              )
            }
          ) : /* @__PURE__ */ jsxRuntime.jsx(
            "svg",
            {
              width: "18",
              height: "18",
              viewBox: "0 0 18 18",
              fill: "none",
              xmlns: "http://www.w3.org/2000/svg",
              children: /* @__PURE__ */ jsxRuntime.jsx(
                "path",
                {
                  d: "M2 2h6v6H2V2zm0 8h6v6H2v-6zm8-8h6v6h-6V2zm0 8h6v6h-6v-6z",
                  fill: "currentColor",
                  opacity: "0.54"
                }
              )
            }
          )
        }
      );
    };
  }
});
exports.AllVaultsDesktop = void 0;
var init_all_vaults_desktop = __esm({
  "src/components/all-vaults/all-vaults.desktop.tsx"() {
    init_vault_card();
    init_vaults_list();
    init_view_mode_toggle();
    exports.AllVaultsDesktop = (props) => {
      const { vaults } = props;
      const { t } = i18n.useTranslation();
      const [viewMode, setViewMode] = React.useState("grid");
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-flex-col oui-gap-6", children: [
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "oui-flex oui-items-center oui-justify-between", children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-text-xl oui-font-normal oui-text-base-contrast", children: t("vaults.allVaults") }),
          /* @__PURE__ */ jsxRuntime.jsx(exports.ViewModeToggle, { mode: viewMode, onChange: setViewMode })
        ] }),
        viewMode === "grid" ? /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-grid oui-grid-cols-3 oui-gap-4 min-[1024px]:oui-grid-cols-3", children: vaults.map((vault) => /* @__PURE__ */ jsxRuntime.jsx(exports.VaultCardWidget, { vault }, vault.vault_id)) }) : /* @__PURE__ */ jsxRuntime.jsx(exports.VaultsList, { vaults })
      ] });
    };
  }
});
var AllVaultsMobile;
var init_all_vaults_mobile = __esm({
  "src/components/all-vaults/all-vaults.mobile.tsx"() {
    init_vault_card();
    AllVaultsMobile = (props) => {
      const { vaults } = props;
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "oui-flex oui-flex-col oui-gap-3", children: vaults.map((vault) => /* @__PURE__ */ jsxRuntime.jsx(exports.VaultCardWidget, { vault }, vault.vault_id)) });
    };
  }
});
exports.AllVaultsWidget = void 0;
var init_all_vaults_widget = __esm({
  "src/components/all-vaults/all-vaults.widget.tsx"() {
    init_vaultsStore();
    init_all_vaults_desktop();
    init_all_vaults_mobile();
    exports.AllVaultsWidget = () => {
      const { data } = exports.useVaultInfoState();
      const { isMobile } = ui.useScreen();
      return isMobile ? /* @__PURE__ */ jsxRuntime.jsx(AllVaultsMobile, { vaults: data }) : /* @__PURE__ */ jsxRuntime.jsx(exports.AllVaultsDesktop, { vaults: data });
    };
  }
});

// src/components/all-vaults/index.ts
var all_vaults_exports = {};
__export(all_vaults_exports, {
  AllVaultsDesktop: () => exports.AllVaultsDesktop,
  AllVaultsWidget: () => exports.AllVaultsWidget,
  VaultsList: () => exports.VaultsList,
  ViewModeToggle: () => exports.ViewModeToggle
});
var init_all_vaults = __esm({
  "src/components/all-vaults/index.ts"() {
    init_all_vaults_widget();
    init_all_vaults_desktop();
    init_view_mode_toggle();
    init_vaults_list();
  }
});

// src/components/provider/vaults-provider.tsx
init_useSVAPIUrl();
init_store();
var VaultsProvider = (props) => {
  const svApiUrl = useSVApiUrl();
  const brokerId = hooks.useConfig("brokerId");
  const {
    fetchVaultInfo,
    fetchVaultOverallInfo,
    setVaultsPageConfig,
    vaultsPageConfig
  } = exports.useVaultsStore();
  React.useEffect(() => {
    if (props.config) {
      setVaultsPageConfig(props.config);
    }
  }, [props.config]);
  React.useEffect(() => {
    if (!svApiUrl) {
      return;
    }
    fetchVaultInfo(void 0, svApiUrl);
  }, [svApiUrl]);
  React.useEffect(() => {
    if (!svApiUrl || !brokerId) {
      return;
    }
    const brokerIds = vaultsPageConfig?.overallInfoBrokerIds ? vaultsPageConfig.overallInfoBrokerIds : `orderly,${brokerId}`;
    fetchVaultOverallInfo(
      brokerIds ? { broker_ids: brokerIds } : void 0,
      svApiUrl
    );
  }, [
    svApiUrl,
    brokerId,
    vaultsPageConfig?.overallInfoBrokerIds,
    fetchVaultOverallInfo
  ]);
  return /* @__PURE__ */ jsxRuntime.jsx("div", { children: props.children });
};
var LazyVaultsHeaderWidget = React__default.default.lazy(
  () => Promise.resolve().then(() => (init_vaults_header(), vaults_header_exports)).then((mod) => {
    return { default: mod.VaultsHeaderWidget };
  })
);
var LazyVaultsIntroductionWidget = React__default.default.lazy(
  () => Promise.resolve().then(() => (init_vaults_introduction(), vaults_introduction_exports)).then((mod) => {
    return { default: mod.VaultsIntroductionWidget };
  })
);
var LazyAllVaultsWidget = React__default.default.lazy(
  () => Promise.resolve().then(() => (init_all_vaults(), all_vaults_exports)).then((mod) => {
    return { default: mod.AllVaultsWidget };
  })
);
var VaultsPage = (props) => {
  const { isMobile } = ui.useScreen();
  return /* @__PURE__ */ jsxRuntime.jsx(VaultsProvider, { ...props, children: /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: ui.cn(
        "oui-relative oui-min-h-screen ",
        "oui-bg-base-10",
        isMobile ? "oui-px-3 oui-py-6" : "oui-px-6 oui-py-12",
        props?.className
      ),
      children: /* @__PURE__ */ jsxRuntime.jsxs(
        "div",
        {
          id: "vaults-content",
          className: ui.cn(
            "oui-mx-auto oui-flex oui-max-w-[1200px] oui-flex-col",
            isMobile ? "oui-gap-6" : "oui-gap-12"
          ),
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(React__default.default.Suspense, { fallback: null, children: /* @__PURE__ */ jsxRuntime.jsx(LazyVaultsHeaderWidget, {}) }),
            /* @__PURE__ */ jsxRuntime.jsx(React__default.default.Suspense, { fallback: null, children: /* @__PURE__ */ jsxRuntime.jsx(LazyVaultsIntroductionWidget, {}) }),
            /* @__PURE__ */ jsxRuntime.jsx(React__default.default.Suspense, { fallback: null, children: /* @__PURE__ */ jsxRuntime.jsx(LazyAllVaultsWidget, {}) })
          ]
        }
      )
    }
  ) });
};

// src/types/index.ts
init_vault();

// src/index.ts
init_api();
init_store();

// src/components/index.ts
init_all_vaults();
init_vault_card();
init_vaults_header();
init_vaults_introduction();

// src/components/vault-operation/index.ts
init_depositAndWithdraw();
init_deposit();
init_withdraw();

// src/components/vault-operation/latest-deposit/index.ts
init_latest_deposit_widget();

// src/components/vault-operation/index.ts
init_latest_withdraw();

exports.VaultsPage = VaultsPage;
exports.getVaultInfo = getVaultInfo;
exports.getVaultLpInfo = getVaultLpInfo;
exports.getVaultLpPerformance = getVaultLpPerformance;
exports.getVaultOverallInfo = getVaultOverallInfo;
