// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/events/NativeDropAppliedEvent.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'
import { NativeDropRequest, nativeDropRequestBeet } from '../types/NativeDropRequest'
export type NativeDropAppliedEvent = {
  srcEid: number
  sender: number[] /* size: 32 */
  nonce: beet.bignum
  dstEid: number
  oapp: web3.PublicKey
  nativeDropRequests: NativeDropRequest[]
  successes: boolean[]
}

/**
 * @category userTypes
 * @category generated
 */
export const nativeDropAppliedEventBeet =
  new beet.FixableBeetArgsStruct<NativeDropAppliedEvent>(
    [
      ['srcEid', beet.u32],
      ['sender', beet.uniformFixedSizeArray(beet.u8, 32)],
      ['nonce', beet.u64],
      ['dstEid', beet.u32],
      ['oapp', beetSolana.publicKey],
      ['nativeDropRequests', beet.array(nativeDropRequestBeet)],
      ['successes', beet.array(beet.bool)],
    ],
    'NativeDropAppliedEvent'
  )
