// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/pricefeed/types/modelType.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import {
  GetDataEnumKind,
  GetDataEnumKindContent,
  Serializer,
  dataEnum,
  struct,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';

export type ModelType =
  | { __kind: 'Arbitrum'; gasPerL2Tx: bigint; gasPerL1CalldataByte: number }
  | { __kind: 'Optimism'; l1Eid: number };

export type ModelTypeArgs =
  | {
      __kind: 'Arbitrum';
      gasPerL2Tx: number | bigint;
      gasPerL1CalldataByte: number;
    }
  | { __kind: 'Optimism'; l1Eid: number };

export function getModelTypeSerializer(): Serializer<ModelTypeArgs, ModelType> {
  return dataEnum<ModelType>(
    [
      [
        'Arbitrum',
        struct<GetDataEnumKindContent<ModelType, 'Arbitrum'>>([
          ['gasPerL2Tx', u64()],
          ['gasPerL1CalldataByte', u32()],
        ]),
      ],
      [
        'Optimism',
        struct<GetDataEnumKindContent<ModelType, 'Optimism'>>([
          ['l1Eid', u32()],
        ]),
      ],
    ],
    { description: 'ModelType' }
  ) as Serializer<ModelTypeArgs, ModelType>;
}

// Data Enum Helpers.
export function modelType(
  kind: 'Arbitrum',
  data: GetDataEnumKindContent<ModelTypeArgs, 'Arbitrum'>
): GetDataEnumKind<ModelTypeArgs, 'Arbitrum'>;
export function modelType(
  kind: 'Optimism',
  data: GetDataEnumKindContent<ModelTypeArgs, 'Optimism'>
): GetDataEnumKind<ModelTypeArgs, 'Optimism'>;
export function modelType<K extends ModelTypeArgs['__kind']>(
  kind: K,
  data?: any
): Extract<ModelTypeArgs, { __kind: K }> {
  return Array.isArray(data)
    ? { __kind: kind, fields: data }
    : { __kind: kind, ...(data ?? {}) };
}
export function isModelType<K extends ModelType['__kind']>(
  kind: K,
  value: ModelType
): value is ModelType & { __kind: K } {
  return value.__kind === kind;
}
