// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/pricefeed/types/price.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable } from '@metaplex-foundation/umi';
import {
  Serializer,
  option,
  struct,
  u128,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import { ModelType, ModelTypeArgs, getModelTypeSerializer } from '.';

export type Price = {
  eid: number;
  priceRatio: bigint;
  gasPriceInUnit: bigint;
  gasPerByte: number;
  modelType: Option<ModelType>;
};

export type PriceArgs = {
  eid: number;
  priceRatio: number | bigint;
  gasPriceInUnit: number | bigint;
  gasPerByte: number;
  modelType: OptionOrNullable<ModelTypeArgs>;
};

export function getPriceSerializer(): Serializer<PriceArgs, Price> {
  return struct<Price>(
    [
      ['eid', u32()],
      ['priceRatio', u128()],
      ['gasPriceInUnit', u64()],
      ['gasPerByte', u32()],
      ['modelType', option(getModelTypeSerializer())],
    ],
    { description: 'Price' }
  ) as Serializer<PriceArgs, Price>;
}
