// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/pricefeed/types/priceParams.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable } from '@metaplex-foundation/umi';
import {
  Serializer,
  option,
  struct,
  u128,
  u32,
  u64,
} from '@metaplex-foundation/umi/serializers';
import { ModelType, ModelTypeArgs, getModelTypeSerializer } from '.';

export type PriceParams = {
  priceRatio: bigint;
  gasPriceInUnit: bigint;
  gasPerByte: number;
  modelType: Option<ModelType>;
};

export type PriceParamsArgs = {
  priceRatio: number | bigint;
  gasPriceInUnit: number | bigint;
  gasPerByte: number;
  modelType: OptionOrNullable<ModelTypeArgs>;
};

export function getPriceParamsSerializer(): Serializer<
  PriceParamsArgs,
  PriceParams
> {
  return struct<PriceParams>(
    [
      ['priceRatio', u128()],
      ['gasPriceInUnit', u64()],
      ['gasPerByte', u32()],
      ['modelType', option(getModelTypeSerializer())],
    ],
    { description: 'PriceParams' }
  ) as Serializer<PriceParamsArgs, PriceParams>;
}
