// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/executor/errors/executor.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Program, ProgramError } from '@metaplex-foundation/umi';

type ProgramErrorConstructor = new (
  program: Program,
  cause?: Error
) => ProgramError;
const codeToErrorMap: Map<number, ProgramErrorConstructor> = new Map();
const nameToErrorMap: Map<string, ProgramErrorConstructor> = new Map();

/** InvalidSize */
export class InvalidSizeError extends ProgramError {
  override readonly name: string = 'InvalidSize';

  readonly code: number = 0x1770; // 6000

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1770, InvalidSizeError);
nameToErrorMap.set('InvalidSize', InvalidSizeError);

/** Paused */
export class PausedError extends ProgramError {
  override readonly name: string = 'Paused';

  readonly code: number = 0x1771; // 6001

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1771, PausedError);
nameToErrorMap.set('Paused', PausedError);

/** UnsupportedOptionType */
export class UnsupportedOptionTypeError extends ProgramError {
  override readonly name: string = 'UnsupportedOptionType';

  readonly code: number = 0x1772; // 6002

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1772, UnsupportedOptionTypeError);
nameToErrorMap.set('UnsupportedOptionType', UnsupportedOptionTypeError);

/** ZeroLzComposeGasProvided */
export class ZeroLzComposeGasProvidedError extends ProgramError {
  override readonly name: string = 'ZeroLzComposeGasProvided';

  readonly code: number = 0x1773; // 6003

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1773, ZeroLzComposeGasProvidedError);
nameToErrorMap.set('ZeroLzComposeGasProvided', ZeroLzComposeGasProvidedError);

/** ZeroLzReceiveGasProvided */
export class ZeroLzReceiveGasProvidedError extends ProgramError {
  override readonly name: string = 'ZeroLzReceiveGasProvided';

  readonly code: number = 0x1774; // 6004

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1774, ZeroLzReceiveGasProvidedError);
nameToErrorMap.set('ZeroLzReceiveGasProvided', ZeroLzReceiveGasProvidedError);

/** NativeAmountExceedsCap */
export class NativeAmountExceedsCapError extends ProgramError {
  override readonly name: string = 'NativeAmountExceedsCap';

  readonly code: number = 0x1775; // 6005

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1775, NativeAmountExceedsCapError);
nameToErrorMap.set('NativeAmountExceedsCap', NativeAmountExceedsCapError);

/** NotAdmin */
export class NotAdminError extends ProgramError {
  override readonly name: string = 'NotAdmin';

  readonly code: number = 0x1776; // 6006

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1776, NotAdminError);
nameToErrorMap.set('NotAdmin', NotAdminError);

/** NotExecutor */
export class NotExecutorError extends ProgramError {
  override readonly name: string = 'NotExecutor';

  readonly code: number = 0x1777; // 6007

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1777, NotExecutorError);
nameToErrorMap.set('NotExecutor', NotExecutorError);

/** MsgLibNotAllowed */
export class MsgLibNotAllowedError extends ProgramError {
  override readonly name: string = 'MsgLibNotAllowed';

  readonly code: number = 0x1778; // 6008

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1778, MsgLibNotAllowedError);
nameToErrorMap.set('MsgLibNotAllowed', MsgLibNotAllowedError);

/** TooManyAdmins */
export class TooManyAdminsError extends ProgramError {
  override readonly name: string = 'TooManyAdmins';

  readonly code: number = 0x1779; // 6009

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1779, TooManyAdminsError);
nameToErrorMap.set('TooManyAdmins', TooManyAdminsError);

/** TooManyExecutors */
export class TooManyExecutorsError extends ProgramError {
  override readonly name: string = 'TooManyExecutors';

  readonly code: number = 0x177a; // 6010

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177a, TooManyExecutorsError);
nameToErrorMap.set('TooManyExecutors', TooManyExecutorsError);

/** TooManyOptionTypes */
export class TooManyOptionTypesError extends ProgramError {
  override readonly name: string = 'TooManyOptionTypes';

  readonly code: number = 0x177b; // 6011

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177b, TooManyOptionTypesError);
nameToErrorMap.set('TooManyOptionTypes', TooManyOptionTypesError);

/** InvalidNativeDropRequestsLength */
export class InvalidNativeDropRequestsLengthError extends ProgramError {
  override readonly name: string = 'InvalidNativeDropRequestsLength';

  readonly code: number = 0x177c; // 6012

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177c, InvalidNativeDropRequestsLengthError);
nameToErrorMap.set(
  'InvalidNativeDropRequestsLength',
  InvalidNativeDropRequestsLengthError
);

/** InvalidNativeDropReceiver */
export class InvalidNativeDropReceiverError extends ProgramError {
  override readonly name: string = 'InvalidNativeDropReceiver';

  readonly code: number = 0x177d; // 6013

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177d, InvalidNativeDropReceiverError);
nameToErrorMap.set('InvalidNativeDropReceiver', InvalidNativeDropReceiverError);

/** InsufficientBalance */
export class InsufficientBalanceError extends ProgramError {
  override readonly name: string = 'InsufficientBalance';

  readonly code: number = 0x177e; // 6014

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177e, InsufficientBalanceError);
nameToErrorMap.set('InsufficientBalance', InsufficientBalanceError);

/** EidNotSupported */
export class EidNotSupportedError extends ProgramError {
  override readonly name: string = 'EidNotSupported';

  readonly code: number = 0x177f; // 6015

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x177f, EidNotSupportedError);
nameToErrorMap.set('EidNotSupported', EidNotSupportedError);

/** ExecutorIsAdmin */
export class ExecutorIsAdminError extends ProgramError {
  override readonly name: string = 'ExecutorIsAdmin';

  readonly code: number = 0x1780; // 6016

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1780, ExecutorIsAdminError);
nameToErrorMap.set('ExecutorIsAdmin', ExecutorIsAdminError);

/** InvalidOwner */
export class InvalidOwnerError extends ProgramError {
  override readonly name: string = 'InvalidOwner';

  readonly code: number = 0x1781; // 6017

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1781, InvalidOwnerError);
nameToErrorMap.set('InvalidOwner', InvalidOwnerError);

/** InvalidInstructionSequence */
export class InvalidInstructionSequenceError extends ProgramError {
  override readonly name: string = 'InvalidInstructionSequence';

  readonly code: number = 0x1782; // 6018

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1782, InvalidInstructionSequenceError);
nameToErrorMap.set(
  'InvalidInstructionSequence',
  InvalidInstructionSequenceError
);

/** ContextAccountAlreadyInitialized */
export class ContextAccountAlreadyInitializedError extends ProgramError {
  override readonly name: string = 'ContextAccountAlreadyInitialized';

  readonly code: number = 0x1783; // 6019

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1783, ContextAccountAlreadyInitializedError);
nameToErrorMap.set(
  'ContextAccountAlreadyInitialized',
  ContextAccountAlreadyInitializedError
);

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
export function getExecutorErrorFromCode(
  code: number,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
export function getExecutorErrorFromName(
  name: string,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}
