// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/simple_messagelib/accounts/SendConfigStore.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import * as beetSolana from '@metaplex-foundation/beet-solana'

/**
 * Arguments used to create {@link SendConfigStore}
 * @category Accounts
 * @category generated
 */
export type SendConfigStoreArgs = {
  bump: number
  data: Uint8Array
}

export const sendConfigStoreDiscriminator = [
  246, 55, 104, 200, 90, 164, 50, 196,
]
/**
 * Holds the data for the {@link SendConfigStore} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
export class SendConfigStore implements SendConfigStoreArgs {
  private constructor(readonly bump: number, readonly data: Uint8Array) {}

  /**
   * Creates a {@link SendConfigStore} instance from the provided args.
   */
  static fromArgs(args: SendConfigStoreArgs) {
    return new SendConfigStore(args.bump, args.data)
  }

  /**
   * Deserializes the {@link SendConfigStore} from the data of the provided {@link web3.AccountInfo}.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static fromAccountInfo(
    accountInfo: web3.AccountInfo<Buffer>,
    offset = 0
  ): [SendConfigStore, number] {
    return SendConfigStore.deserialize(accountInfo.data, offset)
  }

  /**
   * Retrieves the account info from the provided address and deserializes
   * the {@link SendConfigStore} from its data.
   *
   * @throws Error if no account info is found at the address or if deserialization fails
   */
  static async fromAccountAddress(
    connection: web3.Connection,
    address: web3.PublicKey,
    commitmentOrConfig?: web3.Commitment | web3.GetAccountInfoConfig
  ): Promise<SendConfigStore> {
    const accountInfo = await connection.getAccountInfo(
      address,
      commitmentOrConfig
    )
    if (accountInfo == null) {
      throw new Error(`Unable to find SendConfigStore account at ${address}`)
    }
    return SendConfigStore.fromAccountInfo(accountInfo, 0)[0]
  }

  /**
   * Provides a {@link web3.Connection.getProgramAccounts} config builder,
   * to fetch accounts matching filters that can be specified via that builder.
   *
   * @param programId - the program that owns the accounts we are filtering
   */
  static gpaBuilder(
    programId: web3.PublicKey = new web3.PublicKey(
      '6GsmxMTHAAiFKfemuM4zBjumTjNSX5CAiw4xSSXM2Toy'
    )
  ) {
    return beetSolana.GpaBuilder.fromStruct(programId, sendConfigStoreBeet)
  }

  /**
   * Deserializes the {@link SendConfigStore} from the provided data Buffer.
   * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
   */
  static deserialize(buf: Buffer, offset = 0): [SendConfigStore, number] {
    return sendConfigStoreBeet.deserialize(buf, offset)
  }

  /**
   * Serializes the {@link SendConfigStore} into a Buffer.
   * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
   */
  serialize(): [Buffer, number] {
    return sendConfigStoreBeet.serialize({
      accountDiscriminator: sendConfigStoreDiscriminator,
      ...this,
    })
  }

  /**
   * Returns the byteSize of a {@link Buffer} holding the serialized data of
   * {@link SendConfigStore} for the provided args.
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   */
  static byteSize(args: SendConfigStoreArgs) {
    const instance = SendConfigStore.fromArgs(args)
    return sendConfigStoreBeet.toFixedFromValue({
      accountDiscriminator: sendConfigStoreDiscriminator,
      ...instance,
    }).byteSize
  }

  /**
   * Fetches the minimum balance needed to exempt an account holding
   * {@link SendConfigStore} data from rent
   *
   * @param args need to be provided since the byte size for this account
   * depends on them
   * @param connection used to retrieve the rent exemption information
   */
  static async getMinimumBalanceForRentExemption(
    args: SendConfigStoreArgs,
    connection: web3.Connection,
    commitment?: web3.Commitment
  ): Promise<number> {
    return connection.getMinimumBalanceForRentExemption(
      SendConfigStore.byteSize(args),
      commitment
    )
  }

  /**
   * Returns a readable version of {@link SendConfigStore} properties
   * and can be used to convert to JSON and/or logging
   */
  pretty() {
    return {
      bump: this.bump,
      data: this.data,
    }
  }
}

/**
 * @category Accounts
 * @category generated
 */
export const sendConfigStoreBeet = new beet.FixableBeetStruct<
  SendConfigStore,
  SendConfigStoreArgs & {
    accountDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['accountDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['bump', beet.u8],
    ['data', beet.bytes],
  ],
  SendConfigStore.fromArgs,
  'SendConfigStore'
)
