// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/pricefeed/errors/pricefeed.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Program, ProgramError } from '@metaplex-foundation/umi';

type ProgramErrorConstructor = new (
  program: Program,
  cause?: Error
) => ProgramError;
const codeToErrorMap: Map<number, ProgramErrorConstructor> = new Map();
const nameToErrorMap: Map<string, ProgramErrorConstructor> = new Map();

/** TooManyUpdaters */
export class TooManyUpdatersError extends ProgramError {
  override readonly name: string = 'TooManyUpdaters';

  readonly code: number = 0x1770; // 6000

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1770, TooManyUpdatersError);
nameToErrorMap.set('TooManyUpdaters', TooManyUpdatersError);

/** InvalidUpdater */
export class InvalidUpdaterError extends ProgramError {
  override readonly name: string = 'InvalidUpdater';

  readonly code: number = 0x1771; // 6001

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1771, InvalidUpdaterError);
nameToErrorMap.set('InvalidUpdater', InvalidUpdaterError);

/** NotFound */
export class NotFoundError extends ProgramError {
  override readonly name: string = 'NotFound';

  readonly code: number = 0x1772; // 6002

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1772, NotFoundError);
nameToErrorMap.set('NotFound', NotFoundError);

/** InvalidSize */
export class InvalidSizeError extends ProgramError {
  override readonly name: string = 'InvalidSize';

  readonly code: number = 0x1773; // 6003

  constructor(program: Program, cause?: Error) {
    super('', program, cause);
  }
}
codeToErrorMap.set(0x1773, InvalidSizeError);
nameToErrorMap.set('InvalidSize', InvalidSizeError);

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
export function getPricefeedErrorFromCode(
  code: number,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = codeToErrorMap.get(code);
  return constructor ? new constructor(program, cause) : null;
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
export function getPricefeedErrorFromName(
  name: string,
  program: Program,
  cause?: Error
): ProgramError | null {
  const constructor = nameToErrorMap.get(name);
  return constructor ? new constructor(program, cause) : null;
}
