// @layerzerolabs/lz-core/dist/index.cjs
'use strict';

// src/transaction.ts
var Block = class _Block {
  constructor(block) {
    this.block = block;
    this.type = "Block";
  }
  /**
   * Creates an instance of Block from a non-promise value.
   *
   * @param {NonPromise<T>} raw - The raw value to create the Block from.
   * @returns {Block} The created Block instance.
   */
  static from(raw) {
    return new _Block(raw);
  }
};
var BlockWithTransactions = class _BlockWithTransactions {
  constructor(block) {
    this.block = block;
    this.type = "BlockWithTransactions";
  }
  /**
   * Creates an instance of BlockWithTransactions from a non-promise value.
   *
   * @param {NonPromise<T>} raw - The raw value to create the BlockWithTransactions from.
   * @returns {BlockWithTransactions} The created BlockWithTransactions instance.
   */
  static from(raw) {
    return new _BlockWithTransactions(raw);
  }
};
var TransactionRequest = class _TransactionRequest {
  constructor(request) {
    this.request = request;
    this.type = "TransactionRequest";
  }
  /**
   * Creates an instance of TransactionRequest from a non-promise value.
   *
   * @param {Exclude<NonPromise<T>, TransactionTypes>} raw - The raw value to create the TransactionRequest from.
   * @returns {TransactionRequest} The created TransactionRequest instance.
   */
  static from(raw) {
    return new _TransactionRequest(raw);
  }
};
var TransactionResponse = class _TransactionResponse {
  constructor(response) {
    this.response = response;
    this.type = "TransactionResponse";
  }
  /**
   * Creates an instance of TransactionResponse from a non-promise value.
   *
   * @param {Exclude<NonPromise<T>, TransactionTypes>} raw - The raw value to create the TransactionResponse from.
   * @returns {TransactionResponse} The created TransactionResponse instance.
   */
  static from(raw) {
    return new _TransactionResponse(raw);
  }
};
var TransactionReceipt = class _TransactionReceipt {
  constructor(receipt) {
    this.receipt = receipt;
    this.type = "TransactionReceipt";
  }
  /**
   * Creates an instance of TransactionReceipt from a non-promise value.
   *
   * @param {Exclude<NonPromise<T>, TransactionTypes>} raw - The raw value to create the TransactionReceipt from.
   * @returns {TransactionReceipt} The created TransactionReceipt instance.
   */
  static from(raw) {
    return new _TransactionReceipt(raw);
  }
};
var SignedTransaction = class _SignedTransaction {
  constructor(signed) {
    this.signed = signed;
    this.type = "SignedTransaction";
  }
  /**
   * Creates an instance of SignedTransaction from a non-promise value.
   *
   * @param {Exclude<NonPromise<T>, TransactionTypes>} raw - The raw value to create the SignedTransaction from.
   * @returns {SignedTransaction} The created SignedTransaction instance.
   */
  static from(raw) {
    return new _SignedTransaction(raw);
  }
};
var TransactionPending = class _TransactionPending {
  constructor(pending) {
    this.pending = pending;
    this.type = "TransactionPending";
  }
  /**
   * Creates an instance of TransactionPending from a non-promise value.
   *
   * @param {Exclude<NonPromise<T>, TransactionTypes>} raw - The raw value to create the TransactionPending from.
   * @returns {TransactionPending} The created TransactionPending instance.
   */
  static from(raw) {
    return new _TransactionPending(raw);
  }
};

// src/signer.ts
function isConnectable(obj) {
  if (typeof obj !== "object" || obj === null) return false;
  const fn = obj.connect;
  return typeof fn === "function" && typeof fn.length === "number" && fn.length === 1;
}

exports.Block = Block;
exports.BlockWithTransactions = BlockWithTransactions;
exports.SignedTransaction = SignedTransaction;
exports.TransactionPending = TransactionPending;
exports.TransactionReceipt = TransactionReceipt;
exports.TransactionRequest = TransactionRequest;
exports.TransactionResponse = TransactionResponse;
exports.isConnectable = isConnectable;
