// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/types/adminConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  GetDataEnumKind,
  GetDataEnumKindContent,
  Serializer,
  array,
  dataEnum,
  publicKey as publicKeySerializer,
  struct,
  tuple,
  u16,
  u32,
} from '@metaplex-foundation/umi/serializers';
import { DstConfig, DstConfigArgs, getDstConfigSerializer } from '.';

export type AdminConfig =
  | { __kind: 'Admins'; fields: [Array<PublicKey>] }
  | { __kind: 'DefaultMultiplierBps'; fields: [number] }
  | { __kind: 'DstConfigs'; fields: [Array<DstConfig>] }
  | { __kind: 'PriceFeed'; fields: [PublicKey] }
  | { __kind: 'RemoveDstConfigs'; fields: [Array<number>] };

export type AdminConfigArgs =
  | { __kind: 'Admins'; fields: [Array<PublicKey>] }
  | { __kind: 'DefaultMultiplierBps'; fields: [number] }
  | { __kind: 'DstConfigs'; fields: [Array<DstConfigArgs>] }
  | { __kind: 'PriceFeed'; fields: [PublicKey] }
  | { __kind: 'RemoveDstConfigs'; fields: [Array<number>] };

export function getAdminConfigSerializer(): Serializer<
  AdminConfigArgs,
  AdminConfig
> {
  return dataEnum<AdminConfig>(
    [
      [
        'Admins',
        struct<GetDataEnumKindContent<AdminConfig, 'Admins'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'DefaultMultiplierBps',
        struct<GetDataEnumKindContent<AdminConfig, 'DefaultMultiplierBps'>>([
          ['fields', tuple([u16()])],
        ]),
      ],
      [
        'DstConfigs',
        struct<GetDataEnumKindContent<AdminConfig, 'DstConfigs'>>([
          ['fields', tuple([array(getDstConfigSerializer())])],
        ]),
      ],
      [
        'PriceFeed',
        struct<GetDataEnumKindContent<AdminConfig, 'PriceFeed'>>([
          ['fields', tuple([publicKeySerializer()])],
        ]),
      ],
      [
        'RemoveDstConfigs',
        struct<GetDataEnumKindContent<AdminConfig, 'RemoveDstConfigs'>>([
          ['fields', tuple([array(u32())])],
        ]),
      ],
    ],
    { description: 'AdminConfig' }
  ) as Serializer<AdminConfigArgs, AdminConfig>;
}

// Data Enum Helpers.
export function adminConfig(
  kind: 'Admins',
  data: GetDataEnumKindContent<AdminConfigArgs, 'Admins'>['fields']
): GetDataEnumKind<AdminConfigArgs, 'Admins'>;
export function adminConfig(
  kind: 'DefaultMultiplierBps',
  data: GetDataEnumKindContent<
    AdminConfigArgs,
    'DefaultMultiplierBps'
  >['fields']
): GetDataEnumKind<AdminConfigArgs, 'DefaultMultiplierBps'>;
export function adminConfig(
  kind: 'DstConfigs',
  data: GetDataEnumKindContent<AdminConfigArgs, 'DstConfigs'>['fields']
): GetDataEnumKind<AdminConfigArgs, 'DstConfigs'>;
export function adminConfig(
  kind: 'PriceFeed',
  data: GetDataEnumKindContent<AdminConfigArgs, 'PriceFeed'>['fields']
): GetDataEnumKind<AdminConfigArgs, 'PriceFeed'>;
export function adminConfig(
  kind: 'RemoveDstConfigs',
  data: GetDataEnumKindContent<AdminConfigArgs, 'RemoveDstConfigs'>['fields']
): GetDataEnumKind<AdminConfigArgs, 'RemoveDstConfigs'>;
export function adminConfig<K extends AdminConfigArgs['__kind']>(
  kind: K,
  data?: any
): Extract<AdminConfigArgs, { __kind: K }> {
  return Array.isArray(data)
    ? { __kind: kind, fields: data }
    : { __kind: kind, ...(data ?? {}) };
}
export function isAdminConfig<K extends AdminConfig['__kind']>(
  kind: K,
  value: AdminConfig
): value is AdminConfig & { __kind: K } {
  return value.__kind === kind;
}
