// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/types/dstConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { Option, OptionOrNullable } from '@metaplex-foundation/umi';
import {
  Serializer,
  option,
  struct,
  u128,
  u16,
  u32,
} from '@metaplex-foundation/umi/serializers';

export type DstConfig = {
  eid: number;
  dstGas: number;
  multiplierBps: Option<number>;
  floorMarginUsd: Option<bigint>;
};

export type DstConfigArgs = {
  eid: number;
  dstGas: number;
  multiplierBps: OptionOrNullable<number>;
  floorMarginUsd: OptionOrNullable<number | bigint>;
};

export function getDstConfigSerializer(): Serializer<DstConfigArgs, DstConfig> {
  return struct<DstConfig>(
    [
      ['eid', u32()],
      ['dstGas', u32()],
      ['multiplierBps', option(u16())],
      ['floorMarginUsd', option(u128())],
    ],
    { description: 'DstConfig' }
  ) as Serializer<DstConfigArgs, DstConfig>;
}
