// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/types/executeTransactionDigest.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  bytes,
  i64,
  publicKey as publicKeySerializer,
  struct,
  u32,
} from '@metaplex-foundation/umi/serializers';
import {
  TransactionAccount,
  TransactionAccountArgs,
  getTransactionAccountSerializer,
} from '.';

export type ExecuteTransactionDigest = {
  vid: number;
  programId: PublicKey;
  accounts: Array<TransactionAccount>;
  data: Uint8Array;
  expiration: bigint;
};

export type ExecuteTransactionDigestArgs = {
  vid: number;
  programId: PublicKey;
  accounts: Array<TransactionAccountArgs>;
  data: Uint8Array;
  expiration: number | bigint;
};

export function getExecuteTransactionDigestSerializer(): Serializer<
  ExecuteTransactionDigestArgs,
  ExecuteTransactionDigest
> {
  return struct<ExecuteTransactionDigest>(
    [
      ['vid', u32()],
      ['programId', publicKeySerializer()],
      ['accounts', array(getTransactionAccountSerializer())],
      ['data', bytes({ size: u32() })],
      ['expiration', i64()],
    ],
    { description: 'ExecuteTransactionDigest' }
  ) as Serializer<ExecuteTransactionDigestArgs, ExecuteTransactionDigest>;
}
