// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/types/multisigConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  GetDataEnumKind,
  GetDataEnumKindContent,
  Serializer,
  array,
  bool,
  bytes,
  dataEnum,
  publicKey as publicKeySerializer,
  struct,
  tuple,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type MultisigConfig =
  | { __kind: 'Admins'; fields: [Array<PublicKey>] }
  | { __kind: 'Allowlist'; fields: [Array<PublicKey>] }
  | { __kind: 'Denylist'; fields: [Array<PublicKey>] }
  | { __kind: 'Msglibs'; fields: [Array<PublicKey>] }
  | { __kind: 'Paused'; fields: [boolean] }
  | { __kind: 'Quorum'; fields: [number] }
  | { __kind: 'Signers'; fields: [Array<Uint8Array>] };

export type MultisigConfigArgs = MultisigConfig;

export function getMultisigConfigSerializer(): Serializer<
  MultisigConfigArgs,
  MultisigConfig
> {
  return dataEnum<MultisigConfig>(
    [
      [
        'Admins',
        struct<GetDataEnumKindContent<MultisigConfig, 'Admins'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Allowlist',
        struct<GetDataEnumKindContent<MultisigConfig, 'Allowlist'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Denylist',
        struct<GetDataEnumKindContent<MultisigConfig, 'Denylist'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Msglibs',
        struct<GetDataEnumKindContent<MultisigConfig, 'Msglibs'>>([
          ['fields', tuple([array(publicKeySerializer())])],
        ]),
      ],
      [
        'Paused',
        struct<GetDataEnumKindContent<MultisigConfig, 'Paused'>>([
          ['fields', tuple([bool()])],
        ]),
      ],
      [
        'Quorum',
        struct<GetDataEnumKindContent<MultisigConfig, 'Quorum'>>([
          ['fields', tuple([u8()])],
        ]),
      ],
      [
        'Signers',
        struct<GetDataEnumKindContent<MultisigConfig, 'Signers'>>([
          ['fields', tuple([array(bytes({ size: 64 }))])],
        ]),
      ],
    ],
    { description: 'MultisigConfig' }
  ) as Serializer<MultisigConfigArgs, MultisigConfig>;
}

// Data Enum Helpers.
export function multisigConfig(
  kind: 'Admins',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Admins'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Admins'>;
export function multisigConfig(
  kind: 'Allowlist',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Allowlist'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Allowlist'>;
export function multisigConfig(
  kind: 'Denylist',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Denylist'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Denylist'>;
export function multisigConfig(
  kind: 'Msglibs',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Msglibs'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Msglibs'>;
export function multisigConfig(
  kind: 'Paused',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Paused'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Paused'>;
export function multisigConfig(
  kind: 'Quorum',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Quorum'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Quorum'>;
export function multisigConfig(
  kind: 'Signers',
  data: GetDataEnumKindContent<MultisigConfigArgs, 'Signers'>['fields']
): GetDataEnumKind<MultisigConfigArgs, 'Signers'>;
export function multisigConfig<K extends MultisigConfigArgs['__kind']>(
  kind: K,
  data?: any
): Extract<MultisigConfigArgs, { __kind: K }> {
  return Array.isArray(data)
    ? { __kind: kind, fields: data }
    : { __kind: kind, ...(data ?? {}) };
}
export function isMultisigConfig<K extends MultisigConfig['__kind']>(
  kind: K,
  value: MultisigConfig
): value is MultisigConfig & { __kind: K } {
  return value.__kind === kind;
}
