// @layerzerolabs/lz-solana-sdk-v2/src/generated/kinobi/dvn/types/ulnConfig.ts
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/kinobi-so/kinobi
 */

import { PublicKey } from '@metaplex-foundation/umi';
import {
  Serializer,
  array,
  publicKey as publicKeySerializer,
  struct,
  u64,
  u8,
} from '@metaplex-foundation/umi/serializers';

export type UlnConfig = {
  confirmations: bigint;
  requiredDvnCount: number;
  optionalDvnCount: number;
  optionalDvnThreshold: number;
  requiredDvns: Array<PublicKey>;
  optionalDvns: Array<PublicKey>;
};

export type UlnConfigArgs = {
  confirmations: number | bigint;
  requiredDvnCount: number;
  optionalDvnCount: number;
  optionalDvnThreshold: number;
  requiredDvns: Array<PublicKey>;
  optionalDvns: Array<PublicKey>;
};

export function getUlnConfigSerializer(): Serializer<UlnConfigArgs, UlnConfig> {
  return struct<UlnConfig>(
    [
      ['confirmations', u64()],
      ['requiredDvnCount', u8()],
      ['optionalDvnCount', u8()],
      ['optionalDvnThreshold', u8()],
      ['requiredDvns', array(publicKeySerializer())],
      ['optionalDvns', array(publicKeySerializer())],
    ],
    { description: 'UlnConfig' }
  ) as Serializer<UlnConfigArgs, UlnConfig>;
}
