// @orderly.network/ui-order-entry/dist/index.mjs
import React2, { forwardRef, useState, useImperativeHandle, useEffect, memo, createContext, useMemo, useRef, useCallback, useContext, useId } from 'react';
import { useLocalStorage, utils, usePositionStream, useEventEmitter, useDebouncedCallback, useFeeState, useRwaSymbolsInfoStore, useOrderlyContext, useMemoizedFn, ERROR_MSG_CODES, useOrderEntry, useMarginRatio, useComputedLTV, useTpslPriceChecker, useHoldingStream, useAppStore, useIndexPricesStream, useGetEstLiqPrice, useTrack, useQuery, useSymbolLeverage, useBoolean } from '@orderly.network/hooks';
import { useTranslation, i18n } from '@orderly.network/i18n';
import { useOrderEntryFormErrorMsg, useCanTrade } from '@orderly.network/react-app';
import { EMPTY_LIST, DistributionType, TrailingCallbackType, OrderSide, OrderType, PositionType, BBOOrderType, OrderLevel, TrackerEventName } from '@orderly.network/types';
import { ExclamationFillIcon, modal, Text, Tooltip, TooltipTrigger, Flex, Input, cn, inputFormatter, Box, registerSimpleDialog, SimpleDialog, Select, Grid, Checkbox, Slider, textVariants, SettingFillIcon, Badge, Divider, Button, TokenIcon, DataTable, SimpleDropdownMenu, CaretDownIcon, Switch, useScreen, toast, ThrottledButton, SimpleSheet, InfoCircleIcon, AddCircleIcon, PopoverRoot, PopoverTrigger, PopoverContent, DotStatus, EditIcon as EditIcon$1 } from '@orderly.network/ui';
import { TPSLPositionTypeWidget, TPSLAdvancedWidget } from '@orderly.network/ui-tpsl';
import { Decimal, zero, todpIfNeed, getBBOType, removeTrailingZeros } from '@orderly.network/utils';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { AuthGuard } from '@orderly.network/ui-connector';
import { account } from '@orderly.network/perp';
import { SymbolLeverageSheetId, SymbolLeverageDialogId } from '@orderly.network/ui-leverage';

// src/orderEntry.ui.tsx
var AdditionalInfo = (props) => {
  const { pinned, orderTypeExtra } = props;
  const { t } = useTranslation();
  const onTypeToggle = (type) => (checked) => {
    if (props.onValueChange) {
      props.onValueChange(
        "order_type_ext",
        checked ? type : ""
        // orderTypeExtra === type ? "" : type
      );
    }
  };
  useEffect(() => {
    props.onValueChange?.("visible_quantity", props.hidden ? 0 : 1);
  }, [props.hidden]);
  return /* @__PURE__ */ jsxs("div", { className: "oui-text-base-contrast-54", children: [
    /* @__PURE__ */ jsxs(
      Flex,
      {
        justify: pinned ? "start" : "between",
        mb: 3,
        width: pinned ? "unset" : "100%",
        className: "oui-gap-x-2 md:oui-gap-x-3",
        wrap: "wrap",
        gapY: 1,
        children: [
          /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", children: [
            /* @__PURE__ */ jsx(
              Checkbox,
              {
                "data-testid": "oui-testid-orderEntry-postOnly-checkBox",
                id: "toggle_order_post_only",
                className: "oui-peer",
                color: "white",
                variant: "radio",
                disabled: !props.showExtra,
                checked: orderTypeExtra === OrderType.POST_ONLY,
                onCheckedChange: onTypeToggle(OrderType.POST_ONLY)
              }
            ),
            /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t("orderEntry.orderType.postOnly.tooltip"),
                className: "oui-max-w-[280px] oui-text-2xs oui-text-base-contrast-80",
                children: /* @__PURE__ */ jsx(
                  "label",
                  {
                    htmlFor: "toggle_order_post_only",
                    className: cn(
                      "oui-ml-1 oui-text-2xs peer-data-[disabled]:oui-text-base-contrast-20",
                      "oui-cursor-pointer oui-whitespace-nowrap oui-break-normal oui-border-b oui-border-dashed oui-border-line-12"
                    ),
                    children: t("orderEntry.orderType.postOnly")
                  }
                )
              }
            )
          ] }),
          /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", children: [
            /* @__PURE__ */ jsx(
              Checkbox,
              {
                "data-testid": "oui-testid-orderEntry-ioc-checkBox",
                id: "toggle_order_iov",
                color: "white",
                className: "oui-peer",
                variant: "radio",
                checked: orderTypeExtra === OrderType.IOC,
                onCheckedChange: onTypeToggle(OrderType.IOC),
                disabled: !props.showExtra
              }
            ),
            /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t("orderEntry.orderType.ioc.tooltip"),
                className: "oui-max-w-[280px] oui-text-2xs oui-text-base-contrast-80",
                children: /* @__PURE__ */ jsx(
                  "label",
                  {
                    htmlFor: "toggle_order_iov",
                    className: cn(
                      "oui-ml-1 oui-text-2xs peer-data-[disabled]:oui-text-base-contrast-20",
                      "oui-cursor-pointer oui-whitespace-nowrap oui-break-normal oui-border-b oui-border-dashed oui-border-line-12"
                    ),
                    children: t("orderEntry.orderType.ioc")
                  }
                )
              }
            )
          ] }),
          /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", children: [
            /* @__PURE__ */ jsx(
              Checkbox,
              {
                "data-testid": "oui-testid-orderEntry-fox-checkBox",
                id: "toggle_order_fok",
                color: "white",
                variant: "radio",
                className: "oui-peer",
                checked: orderTypeExtra === OrderType.FOK,
                onCheckedChange: onTypeToggle(OrderType.FOK),
                disabled: !props.showExtra
              }
            ),
            /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t("orderEntry.orderType.fok.tooltip"),
                className: "oui-max-w-[280px] oui-text-2xs oui-text-base-contrast-80",
                children: /* @__PURE__ */ jsx(
                  "label",
                  {
                    htmlFor: "toggle_order_fok",
                    className: cn(
                      "oui-ml-1 oui-text-2xs peer-data-[disabled]:oui-text-base-contrast-20",
                      "oui-cursor-pointer oui-whitespace-nowrap oui-break-normal oui-border-b oui-border-dashed oui-border-line-12"
                    ),
                    children: t("orderEntry.orderType.fok")
                  }
                )
              }
            )
          ] })
        ]
      }
    ),
    /* @__PURE__ */ jsxs(Flex, { gapX: 6, children: [
      /* @__PURE__ */ jsxs(Flex, { children: [
        /* @__PURE__ */ jsx(
          Checkbox,
          {
            "data-testid": "oui-testid-orderEntry-orderConfirm-checkBox",
            id: "toggle_order_confirm",
            color: "white",
            checked: props.needConfirm,
            onCheckedChange: (checked) => {
              props.setNeedConfirm(!!checked);
            }
          }
        ),
        /* @__PURE__ */ jsx(
          "label",
          {
            htmlFor: "toggle_order_confirm",
            className: "oui-ml-1 oui-text-2xs",
            children: t("orderEntry.orderConfirm")
          }
        )
      ] }),
      /* @__PURE__ */ jsxs(Flex, { children: [
        /* @__PURE__ */ jsx(
          Checkbox,
          {
            "data-testid": "oui-testid-orderEntry-hidden-checkBox",
            id: "toggle_order_hidden",
            color: "white",
            checked: props.hidden,
            onCheckedChange: (checked) => {
              props.setHidden(checked);
            }
          }
        ),
        /* @__PURE__ */ jsx(
          Tooltip,
          {
            content: t("orderEntry.hidden.tooltip"),
            className: "oui-max-w-[280px] oui-text-2xs oui-text-base-contrast-80",
            children: /* @__PURE__ */ jsx(
              "label",
              {
                htmlFor: "toggle_order_hidden",
                className: "oui-ml-1 oui-cursor-pointer oui-border-b oui-border-dashed oui-border-line-12 oui-text-2xs",
                children: t("orderEntry.hidden")
              }
            )
          }
        )
      ] })
    ] }),
    !pinned && /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(Divider, { className: "oui-my-3" }),
      /* @__PURE__ */ jsxs(Flex, { children: [
        /* @__PURE__ */ jsx(
          Switch,
          {
            "data-testid": "oui-testid-orderEntry-additional-keepVisible-switch",
            id: "toggle_order_keep_visible",
            onCheckedChange: (checked) => {
              props.setPinned(checked);
            }
          }
        ),
        /* @__PURE__ */ jsx(
          "label",
          {
            htmlFor: "toggle_order_keep_visible",
            className: "oui-ml-1 oui-text-2xs",
            children: t("orderEntry.keepVisible")
          }
        )
      ] })
    ] })
  ] });
};
function AdditionalConfigButton(props) {
  const [open2, setOpen] = useState(false);
  return /* @__PURE__ */ jsxs(PopoverRoot, { open: open2, onOpenChange: setOpen, children: [
    /* @__PURE__ */ jsx(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsx(
      "button",
      {
        "data-testid": "oui-testid-orderEntry-additional-button",
        onClick: () => {
          setOpen(true);
        },
        children: /* @__PURE__ */ jsx(
          "svg",
          {
            width: "16",
            height: "16",
            viewBox: "0 0 16 16",
            fill: "currentColor",
            xmlns: "http://www.w3.org/2000/svg",
            className: "oui-fill-white/[.36] hover:oui-fill-white/80",
            children: /* @__PURE__ */ jsx(
              "path",
              {
                d: "M3.332 2.665a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667V3.332a.667.667 0 0 0-.667-.667zm4 0a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667V3.332a.667.667 0 0 0-.667-.667zm4 0a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667V3.332a.667.667 0 0 0-.667-.667zm-8 4a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667V7.332a.667.667 0 0 0-.667-.667zm4 0a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667V7.332a.667.667 0 0 0-.667-.667zm4 0a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667V7.332a.667.667 0 0 0-.667-.667zm-8 4a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667v-1.333a.667.667 0 0 0-.667-.667zm4 0a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667v-1.333a.667.667 0 0 0-.667-.667zm4 0a.667.667 0 0 0-.667.667v1.333c0 .368.299.667.667.667h1.333a.667.667 0 0 0 .667-.667v-1.333a.667.667 0 0 0-.667-.667z"
              }
            )
          }
        )
      }
    ) }),
    /* @__PURE__ */ jsx(PopoverContent, { side: "top", align: "end", className: "oui-w-[230px]", children: /* @__PURE__ */ jsx(AdditionalInfo, { ...props }) })
  ] });
}
var defaultPath = "M10.007 1.302a.74.74 0 0 0-.486.214c-1.033.989-1.349 1.815-.972 2.948-.88.675-1.437.84-2.536.84-1.503 0-2.484.182-3.152.85v.02a1.583 1.583 0 0 0 0 2.248l1.867 1.882-3.181 3.18c-.26.26-.28.696-.02.956.261.26.699.26.959 0l3.193-3.194 1.87 1.861a1.585 1.585 0 0 0 2.25 0h.02c.668-.667.854-1.523.854-3.144 0-1.03.212-1.758.852-2.523 1.233.361 1.95.015 2.961-.995a.68.68 0 0 0 .188-.48c0-.234-.06-.593-.209-1.04a5.34 5.34 0 0 0-1.312-2.103 5.35 5.35 0 0 0-2.104-1.312c-.448-.15-.808-.208-1.042-.208";
var PinButton = (props) => {
  const [path, setPath] = useState(defaultPath);
  return /* @__PURE__ */ jsx("button", { ...props, children: /* @__PURE__ */ jsx(
    "svg",
    {
      width: 16,
      height: 16,
      viewBox: "0 0 16 16",
      fill: "currentColor",
      xmlns: "http://www.w3.org/2000/svg",
      focusable: false,
      onMouseEnter: () => {
        setPath(
          'M10.008 1.302a.74.74 0 0 0-.486.214c-1.033.989-1.349 1.815-.972 2.948-.88.675-1.437.84-2.536.84-1.503 0-2.484.182-3.152.85v.02a1.583 1.583 0 0 0 0 2.248l1.867 1.882-3.181 3.18c-.26.26-.28.696-.02.956.261.26.699.26.959 0l3.193-3.194 1.87 1.861a1.585 1.585 0 0 0 2.25 0h.02c.668-.667.854-1.523.854-3.144 0-1.03.212-1.758.853-2.523 1.232.361 1.95.015 2.96-.995a.68.68 0 0 0 .188-.48c0-.234-.06-.593-.209-1.04a5.34 5.34 0 0 0-1.312-2.103A5.35 5.35 0 0 0 11.05 1.51c-.448-.15-.808-.208-1.042-.208m.258 1.37c.708.131 1.421.6 1.93 1.107.507.508.94 1.13 1.119 1.945-.636.61-1.026.658-1.662.323a.67.67 0 0 0-.779.117c-1.214 1.213-1.533 2.314-1.533 3.8 0 1.292-.076 1.773-.48 2.206-.113.123-.27.104-.374 0L3.799 7.486a.24.24 0 0 1-.017-.34c.239-.29.769-.515 2.226-.514 1.742.001 2.668-.448 3.812-1.52a.67.67 0 0 0 .125-.77c-.343-.686-.29-1.047.321-1.67"'
        );
      },
      onMouseLeave: () => {
        setPath(defaultPath);
      },
      className: "oui-text-primary-darken",
      children: /* @__PURE__ */ jsx("path", { d: path })
    }
  ) });
};
function AdvancedTPSLResult(props) {
  const { order: formattedOrder, symbolInfo, onEdit, onDelete, errors } = props;
  const { getErrorMsg } = useOrderEntryFormErrorMsg(errors);
  const { t } = useTranslation();
  const renderTp = () => {
    const error = getErrorMsg("tp_trigger_price");
    if (formattedOrder.tp_trigger_price || formattedOrder.tp_order_price) {
      return /* @__PURE__ */ jsxs(
        Flex,
        {
          direction: "column",
          itemAlign: "start",
          className: "oui-w-full",
          gap: 4,
          children: [
            /* @__PURE__ */ jsxs(
              Flex,
              {
                direction: "column",
                itemAlign: "start",
                justify: "between",
                gapY: 1,
                className: "oui-w-full",
                children: [
                  /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                    /* @__PURE__ */ jsx(Text, { children: t("tpsl.tpTriggerPrice") }),
                    /* @__PURE__ */ jsx(
                      Text.numeral,
                      {
                        suffix: /* @__PURE__ */ jsx(Text, { className: "oui-ml-1 oui-text-base-contrast-36", children: symbolInfo.quote }),
                        className: "oui-text-base-contrast",
                        dp: symbolInfo.quote_dp,
                        children: formattedOrder.tp_trigger_price ?? ""
                      }
                    )
                  ] }),
                  /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                    /* @__PURE__ */ jsx(Text, { children: t("tpsl.tpOrderPrice") }),
                    formattedOrder.tp_order_type === OrderType.LIMIT ? /* @__PURE__ */ jsx(
                      Text.numeral,
                      {
                        suffix: /* @__PURE__ */ jsx(Text, { className: "oui-ml-1 oui-text-base-contrast-36", children: symbolInfo.quote }),
                        className: "oui-text-base-contrast",
                        dp: symbolInfo.quote_dp,
                        children: formattedOrder.tp_order_price ?? ""
                      }
                    ) : /* @__PURE__ */ jsx(Text, { className: "oui-text-base-contrast", children: "Market" })
                  ] }),
                  /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                    /* @__PURE__ */ jsx(Text, { children: t("tpsl.totalEstTpPnl") }),
                    /* @__PURE__ */ jsx(
                      Text.numeral,
                      {
                        suffix: /* @__PURE__ */ jsx(Text, { className: "oui-ml-1 oui-text-base-contrast-36", children: symbolInfo.quote }),
                        coloring: true,
                        dp: 2,
                        children: Number(formattedOrder.tp_pnl)
                      }
                    )
                  ] })
                ]
              }
            ),
            error && /* @__PURE__ */ jsxs(
              Flex,
              {
                justify: "start",
                itemAlign: "start",
                gap: 2,
                className: "oui-w-full",
                children: [
                  /* @__PURE__ */ jsx("div", { className: "oui-relative oui-top-[7px] oui-size-1 oui-rounded-full oui-bg-danger" }),
                  /* @__PURE__ */ jsx(Text, { className: "oui-text-danger", children: error })
                ]
              }
            )
          ]
        }
      );
    }
    return null;
  };
  const renderSl = () => {
    if (formattedOrder.sl_trigger_price || formattedOrder.sl_order_price) {
      const error = getErrorMsg("sl_trigger_price");
      return /* @__PURE__ */ jsx(
        Flex,
        {
          direction: "column",
          itemAlign: "start",
          className: "oui-w-full",
          gap: 4,
          children: /* @__PURE__ */ jsxs(
            Flex,
            {
              direction: "column",
              itemAlign: "start",
              justify: "between",
              gapY: 1,
              className: "oui-w-full",
              children: [
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.slTriggerPrice") }),
                  /* @__PURE__ */ jsx(
                    Text.numeral,
                    {
                      suffix: /* @__PURE__ */ jsx(Text, { className: "oui-ml-1 oui-text-base-contrast-36", children: symbolInfo.quote }),
                      className: "oui-text-base-contrast",
                      dp: symbolInfo.quote_dp,
                      children: formattedOrder.sl_trigger_price ?? ""
                    }
                  )
                ] }),
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.slOrderPrice") }),
                  formattedOrder.sl_order_type === OrderType.LIMIT ? /* @__PURE__ */ jsx(
                    Text.numeral,
                    {
                      suffix: /* @__PURE__ */ jsx(Text, { className: "oui-ml-1 oui-text-base-contrast-36", children: symbolInfo.quote }),
                      className: "oui-text-base-contrast",
                      dp: symbolInfo.quote_dp,
                      children: formattedOrder.sl_order_price ?? ""
                    }
                  ) : /* @__PURE__ */ jsx(Text, { className: "oui-text-base-contrast", children: "Market" })
                ] }),
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.totalEstSlPnl") }),
                  /* @__PURE__ */ jsx(
                    Text.numeral,
                    {
                      coloring: true,
                      suffix: /* @__PURE__ */ jsx(Text, { className: "oui-ml-1 oui-text-base-contrast-36", children: symbolInfo.quote }),
                      dp: 2,
                      children: Number(formattedOrder.sl_pnl)
                    }
                  )
                ] }),
                error && /* @__PURE__ */ jsxs(
                  Flex,
                  {
                    justify: "start",
                    itemAlign: "start",
                    gap: 2,
                    className: "oui-w-full",
                    children: [
                      /* @__PURE__ */ jsx("div", { className: "oui-relative oui-top-[7px] oui-size-1 oui-rounded-full oui-bg-danger" }),
                      /* @__PURE__ */ jsx(Text, { className: "oui-text-danger", children: error })
                    ]
                  }
                )
              ]
            }
          )
        }
      );
    }
    return null;
  };
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      direction: "column",
      itemAlign: "start",
      className: "oui-w-full oui-text-2xs",
      gap: 4,
      children: [
        /* @__PURE__ */ jsxs(Flex, { justify: "between", itemAlign: "start", className: "oui-w-full", children: [
          /* @__PURE__ */ jsx(Text, { children: t("common.tpsl") }),
          /* @__PURE__ */ jsxs(Flex, { gap: 2, children: [
            /* @__PURE__ */ jsx(
              DeleteIcon,
              {
                size: 12,
                className: "oui-cursor-pointer oui-text-base-contrast-54 hover:oui-text-base-contrast",
                opacity: 1,
                onClick: onDelete
              }
            ),
            /* @__PURE__ */ jsx(
              EditIcon,
              {
                size: 12,
                className: "oui-cursor-pointer oui-text-base-contrast-54 hover:oui-text-base-contrast",
                onClick: onEdit
              }
            )
          ] })
        ] }),
        /* @__PURE__ */ jsxs(Flex, { justify: "between", itemAlign: "start", className: "oui-w-full", children: [
          /* @__PURE__ */ jsx(Text, { children: t("tpsl.mode") }),
          /* @__PURE__ */ jsx(Text, { className: "oui-text-base-contrast", children: formattedOrder.position_type === PositionType.FULL ? t("tpsl.fullPosition") : t("tpsl.partialPosition") })
        ] }),
        renderTp(),
        renderSl(),
        /* @__PURE__ */ jsx(Divider, { className: "oui-mb-2 oui-w-full" })
      ]
    }
  );
}
var DeleteIcon = (props) => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: props.size,
      height: props.size,
      viewBox: "0 0 12 12",
      fill: "currentColor",
      ...props,
      children: /* @__PURE__ */ jsx("path", { d: "M5.99903 0.976562C5.44653 0.976562 4.99903 1.42406 4.99903 1.97656H2.49902C2.22302 1.97656 1.99902 2.20056 1.99902 2.47656C1.99902 2.75256 2.22302 2.97656 2.49902 2.97656H9.49903C9.77503 2.97656 9.99903 2.75256 9.99903 2.47656C9.99903 2.20056 9.77503 1.97656 9.49903 1.97656H6.99903C6.99903 1.42406 6.55153 0.976562 5.99903 0.976562ZM2.49902 3.97655V8.97654C2.49902 10.0715 3.40152 10.961 4.49903 10.961L7.51453 10.9765C8.61203 10.9765 9.49903 10.074 9.49903 8.97654V3.97655H2.49902ZM4.99903 5.47655C5.27503 5.47655 5.49903 5.70055 5.49903 5.97655V8.97654C5.49903 9.25254 5.27503 9.47654 4.99903 9.47654C4.72303 9.47654 4.49903 9.25254 4.49903 8.97654V5.97655C4.49903 5.70055 4.72303 5.47655 4.99903 5.47655ZM6.99903 5.47655C7.27503 5.47655 7.49903 5.70055 7.49903 5.97655V8.97654C7.49903 9.25254 7.27503 9.47654 6.99903 9.47654C6.72303 9.47654 6.49903 9.25254 6.49903 8.97654V5.97655C6.49903 5.70055 6.72303 5.47655 6.99903 5.47655Z" })
    }
  );
};
var EditIcon = (props) => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      viewBox: "0 0 12 12",
      width: props.size,
      height: props.size,
      fill: "currentColor",
      ...props,
      children: /* @__PURE__ */ jsx("path", { d: "M8.49779 0.976562C8.36529 0.976562 8.23229 1.02357 8.13829 1.11707C7.86029 1.39507 6.85979 2.39558 6.63779 2.61808L2.13529 7.12059L1.63479 7.62059C1.56529 7.69059 1.52929 7.78958 1.50979 7.88658L1.00979 10.3881C0.939788 10.7381 1.23779 11.0361 1.58779 10.9666C1.90079 10.9036 3.77679 10.5286 4.08929 10.4661C4.18629 10.4466 4.28529 10.4106 4.35529 10.3411L4.85529 9.84059L9.35779 5.33808C9.58029 5.11608 10.5808 4.11506 10.8588 3.83756C10.9523 3.74356 10.9993 3.61056 10.9993 3.47806C10.9993 2.65956 10.7908 2.07456 10.3583 1.63306C9.92179 1.18756 9.33879 0.976562 8.49779 0.976562ZM8.69479 1.98606C9.14629 2.01256 9.43879 2.11608 9.63929 2.32108C9.84429 2.53008 9.97379 2.82008 10.0018 3.26258C9.72779 3.53608 9.32679 3.93106 8.99829 4.25956C8.60179 3.86306 8.11279 3.37407 7.71629 2.97757C8.04529 2.64907 8.42129 2.25956 8.69479 1.98606ZM6.99729 3.69657L8.27929 4.97858L4.49579 8.76207L3.21379 7.48009L6.99729 3.69657ZM2.49479 8.19908L3.77679 9.48107L3.72979 9.52809C3.39979 9.59409 2.73329 9.73359 2.11929 9.85659L2.44779 8.24608L2.49479 8.19908Z" })
    }
  );
};
var RegularFeesUI = (props) => {
  const { t } = useTranslation();
  const { taker, maker } = props;
  const originalTrailingFees = /* @__PURE__ */ jsx(Flex, { itemAlign: "center", justify: "between", width: "100%", gap: 1, children: /* @__PURE__ */ jsxs(Flex, { width: "100%", itemAlign: "center", justify: "between", children: [
    /* @__PURE__ */ jsx(Text, { className: "oui-truncate", size: "2xs", children: t("common.fees") }),
    /* @__PURE__ */ jsx(
      AuthGuard,
      {
        fallback: () => /* @__PURE__ */ jsxs(Text, { className: "oui-truncate", size: "2xs", children: [
          t("dmm.taker"),
          ": --% / ",
          t("dmm.maker"),
          ": --%"
        ] }),
        children: /* @__PURE__ */ jsxs(Flex, { gap: 1, children: [
          /* @__PURE__ */ jsxs(Text, { className: "oui-truncate", size: "2xs", children: [
            t("dmm.taker"),
            ":"
          ] }),
          /* @__PURE__ */ jsx(Text, { size: "2xs", className: "oui-text-base-contrast-80", children: taker }),
          /* @__PURE__ */ jsx(Text, { size: "2xs", children: "/" }),
          /* @__PURE__ */ jsxs(Text, { className: "oui-truncate", size: "2xs", children: [
            t("dmm.maker"),
            ":"
          ] }),
          /* @__PURE__ */ jsx(Text, { size: "2xs", className: "oui-text-base-contrast-80", children: maker })
        ] })
      }
    )
  ] }) });
  return originalTrailingFees;
};
var RegularFeesWidget = (props) => {
  return /* @__PURE__ */ jsx(RegularFeesUI, { ...props });
};
var FeesWidget = ({ symbol }) => {
  const { takerFee, makerFee, rwaTakerFee, rwaMakerFee } = useFeeState();
  const info = useRwaSymbolsInfoStore();
  const isRwa = info?.[symbol] !== void 0;
  return /* @__PURE__ */ jsx(
    RegularFeesWidget,
    {
      taker: isRwa ? rwaTakerFee : takerFee,
      maker: isRwa ? rwaMakerFee : makerFee
    }
  );
};
var options = [0.01, 0.05, 0.1];
var SlippageEditor = forwardRef((props, ref) => {
  const { t } = useTranslation();
  const [value, setValue] = useState();
  const [customValue, setCustomValue] = useState("");
  const [error, setError] = useState(void 0);
  useImperativeHandle(ref, () => ({
    getValue: () => customValue ? new Decimal(customValue)?.toNumber() : value
  }));
  useEffect(() => {
    if (props.initialValue && !options.includes(props.initialValue)) {
      setCustomValue(props.initialValue.toString());
    } else {
      setValue(props.initialValue);
    }
  }, [props.initialValue, open]);
  const onClick = (val) => {
    setValue(val);
    setCustomValue("");
    setError(void 0);
  };
  const onValueChange = (val) => {
    if (!val) {
      setCustomValue(val);
      return;
    }
    const d = new Decimal(val);
    setValue(void 0);
    if (d.gt(3)) {
      setCustomValue("3");
      setError(t("orderEntry.slippage.error.exceed"));
    } else {
      setCustomValue(val);
      setError(void 0);
    }
  };
  const toolTipButton = props.isMobile ? /* @__PURE__ */ jsx(
    "button",
    {
      onClick: () => {
        modal.alert({
          title: t("common.tips"),
          message: /* @__PURE__ */ jsx(Text, { size: "2xs", children: t("orderEntry.slippage.tips") })
        });
      },
      children: /* @__PURE__ */ jsx(ExclamationFillIcon, { className: "oui-text-base-contrast-54", size: 16 })
    }
  ) : /* @__PURE__ */ jsx(
    Tooltip,
    {
      content: /* @__PURE__ */ jsx(Text, { intensity: 80, size: "2xs", children: t("orderEntry.slippage.tips") }),
      className: "oui-w-[260px] oui-bg-base-6",
      arrow: { className: "oui-fill-base-6" },
      children: /* @__PURE__ */ jsx(TooltipTrigger, { children: /* @__PURE__ */ jsx(ExclamationFillIcon, { className: "oui-text-base-contrast-54", size: 16 }) })
    }
  );
  return /* @__PURE__ */ jsxs("div", { className: "oui-text-2xs", children: [
    /* @__PURE__ */ jsxs(Flex, { mb: 2, gapX: 1, children: [
      /* @__PURE__ */ jsx(Text, { size: "xs", children: t("orderEntry.slippage") }),
      toolTipButton
    ] }),
    /* @__PURE__ */ jsxs(Flex, { gapX: 2, children: [
      options.map((item) => {
        const isActive = value === item;
        return /* @__PURE__ */ jsx(
          SlippageItem,
          {
            value: item,
            isActive,
            onClick: () => {
              onClick(item);
            }
          },
          item
        );
      }),
      /* @__PURE__ */ jsx(
        Input,
        {
          suffix: "%",
          formatters: [
            inputFormatter.numberFormatter,
            inputFormatter.dpFormatter(2)
          ],
          value: customValue,
          onValueChange,
          classNames: {
            root: cn(
              "oui-rounded-md oui-bg-base-6",
              "oui-h-[40px] oui-w-[74px]"
            ),
            input: "oui-text-base-contrast",
            additional: "oui-pl-1"
          }
        }
      )
    ] }),
    !!error && /* @__PURE__ */ jsx(Box, { mt: 5, className: "-oui-mb-5", children: /* @__PURE__ */ jsx(Text, { size: "2xs", color: "danger", children: error }) })
  ] });
});
var SlippageItem = ({ value, isActive, onClick }) => {
  return /* @__PURE__ */ jsx(
    Flex,
    {
      intensity: 600,
      justify: "center",
      itemAlign: "center",
      r: "md",
      width: 74,
      height: 40,
      className: cn(
        "oui-cursor-pointer oui-select-none",
        isActive ? "oui-bg-primary-light oui-text-primary-contrast/80" : "oui-text-base-contrast-80"
      ),
      onClick,
      children: /* @__PURE__ */ jsxs(Text, { size: "sm", children: [
        value,
        "%"
      ] })
    }
  );
};
var SlippageCell = (props) => {
  const { t } = useTranslation();
  const [open2, { setTrue: setOpen, setFalse: setClose, toggle }] = useBoolean(false);
  const { isMobile } = useScreen();
  const slippageRef = useRef(null);
  const onConfirm = () => {
    const val = slippageRef.current?.getValue();
    props.setSlippage(!val ? "1" : val.toString());
    setClose();
    return Promise.resolve(true);
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      SimpleDialog,
      {
        open: open2,
        onOpenChange: toggle,
        title: t("common.settings"),
        contentProps: { size: isMobile ? "xs" : "sm" },
        actions: {
          primary: {
            disabled: false,
            label: t("common.save"),
            onClick: onConfirm
          },
          secondary: {
            label: t("common.cancel"),
            onClick: () => setClose()
          }
        },
        children: /* @__PURE__ */ jsx(
          SlippageEditor,
          {
            ref: slippageRef,
            isMobile,
            initialValue: props.slippage ? Number(props.slippage) : void 0
          }
        )
      }
    ),
    /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
      /* @__PURE__ */ jsx(Text, { size: "2xs", children: t("orderEntry.slippage") }),
      /* @__PURE__ */ jsx(
        AuthGuard,
        {
          fallback: () => /* @__PURE__ */ jsxs(Text, { size: "2xs", children: [
            t("orderEntry.slippage.est"),
            ": -% / ",
            t("common.max"),
            ": --%"
          ] }),
          children: /* @__PURE__ */ jsxs(Flex, { gap: 1, children: [
            /* @__PURE__ */ jsx(
              Text.numeral,
              {
                size: "2xs",
                rule: "percentages",
                prefix: `${t("orderEntry.slippage.est")}: `,
                suffix: ` / ${t("common.max")}: `,
                children: props.estSlippage ?? 0
              }
            ),
            /* @__PURE__ */ jsx("button", { className: "oui-text-2xs", onClick: () => setOpen(), children: /* @__PURE__ */ jsxs(Flex, { className: "oui-gap-0.5", as: "span", children: [
              /* @__PURE__ */ jsx(Text, { size: "2xs", className: "oui-text-primary", children: `${props.slippage || "-"}%` }),
              /* @__PURE__ */ jsx(
                EditIcon$1,
                {
                  className: "oui-text-primary oui-hidden md:oui-block",
                  size: 12,
                  opacity: 1
                }
              )
            ] }) })
          ] })
        }
      )
    ] })
  ] });
};
var SlippageUI = (props) => {
  return /* @__PURE__ */ jsx(SlippageCell, { ...props });
};
function AssetInfo(props) {
  const { canTrade, disableFeatures, orderType, symbol } = props;
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  const displayEstLiqPrice = useGetEstLiqPrice({
    estLiqPrice: props.estLiqPrice,
    symbol,
    side: props.side
  });
  return /* @__PURE__ */ jsxs("div", { className: "oui-space-y-[2px] xl:oui-space-y-1", children: [
    /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
      isMobile ? /* @__PURE__ */ jsx(Text, { size: "2xs", children: t("orderEntry.estLiqPrice") }) : /* @__PURE__ */ jsx(
        Tooltip,
        {
          content: /* @__PURE__ */ jsxs("div", { className: "oui-min-w-[204px] oui-max-w-[280px] oui-text-2xs oui-leading-normal oui-text-base-contrast-80", children: [
            /* @__PURE__ */ jsx("div", { className: "oui-text-pretty", children: t("common.liquidationPrice.tooltip") }),
            /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(
              "a",
              {
                href: "https://orderly.network/docs/introduction/trade-on-orderly/perpetual-futures/liquidations",
                target: "_blank",
                rel: "noopener noreferrer",
                className: "oui-text-primary oui-underline",
                children: t("common.liquidationPrice.tooltip.learnMore")
              }
            ) })
          ] }),
          children: /* @__PURE__ */ jsx(
            Text,
            {
              size: "2xs",
              className: "oui-cursor-pointer oui-border-b oui-border-dashed oui-border-line-12",
              children: t("orderEntry.estLiqPrice")
            }
          )
        }
      ),
      /* @__PURE__ */ jsx(
        Text.numeral,
        {
          unit: props.quote,
          size: "2xs",
          dp: props.dp,
          className: "oui-text-base-contrast-80",
          unitClassName: "oui-ml-1 oui-text-base-contrast-36",
          children: canTrade ? displayEstLiqPrice ?? "--" : "--"
        }
      )
    ] }),
    orderType === OrderType.MARKET && !disableFeatures?.includes("slippageSetting") && /* @__PURE__ */ jsx(
      SlippageUI,
      {
        slippage: props.slippage,
        setSlippage: props.setSlippage,
        estSlippage: props.estSlippage
      }
    ),
    !disableFeatures?.includes("feesInfo") && /* @__PURE__ */ jsx(FeesWidget, { symbol: props.symbol })
  ] });
}
var calculateLTVColor = (val) => {
  if (val >= 0 && val < 50) {
    return "oui-text-success";
  } else if (val >= 50 && val < 80) {
    return "oui-text-warning";
  } else if (val >= 80) {
    return "oui-text-danger";
  } else {
    return "";
  }
};
var LTVRiskTooltipUI = (props) => {
  const { t } = useTranslation();
  const {
    ltv_threshold,
    negative_usdc_threshold,
    isThresholdLoading,
    holdingData = [],
    currentLtv,
    onConvert
  } = props;
  return /* @__PURE__ */ jsxs(Flex, { gap: 1, className: "oui-w-72 oui-max-w-72", direction: "column", children: [
    /* @__PURE__ */ jsxs(Flex, { width: "100%", justify: "between", itemAlign: "center", children: [
      /* @__PURE__ */ jsx(Text, { intensity: 36, size: "xs", children: t("common.assets") }),
      /* @__PURE__ */ jsx(Text, { intensity: 36, size: "xs", children: t("transfer.deposit.collateralContribution") })
    ] }),
    holdingData.map((asset, index) => {
      return /* @__PURE__ */ jsxs(
        Flex,
        {
          width: "100%",
          justify: "between",
          itemAlign: "center",
          children: [
            /* @__PURE__ */ jsx(Text, { intensity: 80, size: "xs", children: asset.token }),
            /* @__PURE__ */ jsx(
              Text,
              {
                size: "xs",
                intensity: 80,
                className: cn(
                  Number(asset.collateralContribution) < 0 && "oui-text-warning"
                ),
                children: removeTrailingZeros(asset.collateralContribution)
              }
            )
          ]
        },
        `item-${index}`
      );
    }),
    /* @__PURE__ */ jsx(Divider, { className: "oui-w-full" }),
    /* @__PURE__ */ jsxs(Flex, { width: "100%", justify: "between", itemAlign: "center", children: [
      /* @__PURE__ */ jsx(Text, { intensity: 36, size: "xs", children: t("transfer.LTV.currentLTV") }),
      /* @__PURE__ */ jsxs(
        Text,
        {
          size: "xs",
          intensity: 36,
          className: cn("oui-select-none", calculateLTVColor(currentLtv)),
          children: [
            currentLtv,
            "%"
          ]
        }
      )
    ] }),
    /* @__PURE__ */ jsx(Text, { className: "oui-py-2", intensity: 54, size: "2xs", children: t("transfer.LTV.tooltip", {
      threshold: isThresholdLoading ? "-" : ltv_threshold,
      usdcThreshold: isThresholdLoading ? "-" : negative_usdc_threshold
    }) }),
    /* @__PURE__ */ jsx(
      Button,
      {
        fullWidth: true,
        size: "md",
        variant: "outlined",
        color: "secondary",
        onClick: onConvert,
        children: t("transfer.convert.convertAssets")
      }
    )
  ] });
};
var useConvertThreshold = () => {
  const { data, error, isLoading } = useQuery(
    "/v1/public/auto_convert_threshold",
    { errorRetryCount: 3 }
  );
  return {
    ltv_threshold: new Decimal(data?.ltv_threshold ?? 0).mul(100).toNumber(),
    negative_usdc_threshold: data?.negative_usdc_threshold,
    isLoading,
    error
  };
};
var useLTVTooltipScript = () => {
  const { data: holdingList = [], isLoading: isHoldingLoading } = useHoldingStream();
  const {
    ltv_threshold,
    negative_usdc_threshold,
    isLoading: isThresholdLoading
  } = useConvertThreshold();
  const tokensInfo = useAppStore((state) => state.tokensInfo);
  const { getIndexPrice } = useIndexPricesStream();
  const holdingData = holdingList.map((item) => {
    const tokenInfo = tokensInfo?.find(({ token }) => token === item.token);
    const indexPrice = getIndexPrice(item.token);
    const collateralRatio = tokenInfo ? account.collateralRatio({
      baseWeight: tokenInfo.base_weight ?? 0,
      discountFactor: tokenInfo.discount_factor ?? 0,
      collateralQty: item.holding,
      collateralCap: tokenInfo?.user_max_qty ?? item.holding,
      indexPrice
    }) : zero;
    const collateralContribution = account.collateralContribution({
      collateralQty: item.holding,
      collateralCap: tokenInfo?.user_max_qty ?? item.holding,
      collateralRatio: collateralRatio.toNumber(),
      indexPrice
    });
    return {
      ...item,
      collateralContribution
    };
  });
  const currentLtv = useComputedLTV();
  const onConvert = useCallback(async () => {
    return modal.show("ConvertDialogId");
  }, []);
  return {
    holdingData,
    isHoldingLoading,
    ltv_threshold,
    negative_usdc_threshold,
    isThresholdLoading,
    currentLtv,
    onConvert
  };
};
var LTVRiskTooltipWidget = () => {
  const state = useLTVTooltipScript();
  return /* @__PURE__ */ jsx(LTVRiskTooltipUI, { ...state });
};
var Available = (props) => {
  const { canTrade, currentLtv, quote, freeCollateral } = props;
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  const showLTV = useMemo(() => {
    return typeof currentLtv === "number" && !Number.isNaN(currentLtv) && currentLtv > 0;
  }, [currentLtv]);
  return /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", justify: "between", children: [
    /* @__PURE__ */ jsx(Text, { size: "2xs", children: t("common.available") }),
    /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", justify: "center", gap: 1, children: [
      showLTV && /* @__PURE__ */ jsx(
        Tooltip,
        {
          className: "oui-bg-base-6 oui-p-2",
          content: /* @__PURE__ */ jsx(LTVRiskTooltipWidget, {}),
          children: /* @__PURE__ */ jsx(
            InfoCircleIcon,
            {
              className: "oui-cursor-pointer oui-text-warning oui-opacity-80"
            }
          )
        }
      ),
      /* @__PURE__ */ jsx(
        Text.numeral,
        {
          unit: quote,
          size: "2xs",
          className: "oui-text-base-contrast-80",
          unitClassName: "oui-ml-1 oui-text-base-contrast-54",
          dp: 2,
          padding: false,
          children: canTrade ? freeCollateral : 0
        }
      ),
      /* @__PURE__ */ jsx(
        Button,
        {
          variant: "text",
          size: "xs",
          color: "secondary",
          className: "oui-p-0 hover:oui-text-base-contrast-80",
          onClick: () => {
            const handleDomId = isMobile ? "DepositAndWithdrawWithSheetId" : "DepositAndWithdrawWithDialogId";
            modal.show(handleDomId, {
              activeTab: "deposit"
            });
          },
          children: /* @__PURE__ */ jsx(AddCircleIcon, { opacity: 1 })
        }
      )
    ] })
  ] });
};
function isBBOOrder(options2) {
  const { order_type, order_type_ext } = options2;
  const isBBO = [OrderType.ASK, OrderType.BID].includes(order_type_ext);
  if (order_type) {
    return order_type === OrderType.LIMIT && isBBO;
  }
  return isBBO;
}
function getOrderTypeByBBO(value, size) {
  if ([BBOOrderType.COUNTERPARTY1, BBOOrderType.COUNTERPARTY5].includes(value)) {
    return size === OrderSide.BUY ? OrderType.ASK : OrderType.BID;
  }
  if ([BBOOrderType.QUEUE1, BBOOrderType.QUEUE5].includes(value)) {
    return size === OrderSide.BUY ? OrderType.BID : OrderType.ASK;
  }
}
function getOrderLevelByBBO(value) {
  if ([BBOOrderType.COUNTERPARTY1, BBOOrderType.QUEUE1].includes(value)) {
    return OrderLevel.ONE;
  }
  if ([BBOOrderType.COUNTERPARTY5, BBOOrderType.QUEUE5].includes(value)) {
    return OrderLevel.FIVE;
  }
}
function getScaledPlaceOrderMessage(result) {
  const rows = result?.data?.rows || [];
  if (rows.length > 0) {
    const totalCount = rows.length;
    const successCount = rows.filter((row) => row.success).length;
    if (successCount === totalCount) {
      return i18n.t("orderEntry.scaledOrder.fullySuccessful", {
        total: totalCount
      });
    }
    if (successCount === 0) {
      return i18n.t("orderEntry.scaledOrder.allFailed");
    }
    return i18n.t("orderEntry.scaledOrder.partiallySuccessful", {
      successCount,
      total: totalCount
    });
  }
}
var safeNumber = (val) => {
  return Number.isNaN(Number(val)) ? 0 : Number(val);
};
var OrderConfirmDialog = (props) => {
  const { symbolInfo, order, onConfirm, onCancel } = props;
  const { quote, quote_dp, base_dp } = symbolInfo;
  const { side, order_type, order_type_ext, level, symbol } = order;
  const { t } = useTranslation();
  const [{ rows: positions }] = usePositionStream(symbol);
  const position = positions?.[0];
  const positionQty = position?.position_qty;
  const [_, setNeedConfirm] = useLocalStorage("orderly_order_confirm", true);
  const renderPositionType = () => {
    if (order.position_type === PositionType.FULL) {
      return /* @__PURE__ */ jsx(Text, { children: t("tpsl.positionType.full") });
    }
    return /* @__PURE__ */ jsx(Text, { children: t("tpsl.positionType.partial") });
  };
  const renderPrice = () => {
    if (order_type === OrderType.MARKET || order_type === OrderType.STOP_MARKET) {
      return /* @__PURE__ */ jsx(Text, { intensity: 80, children: t("common.marketPrice") });
    }
    if (isBBOOrder({ order_type, order_type_ext })) {
      const bboType = getBBOType({
        type: order_type_ext,
        side,
        level
      });
      const label = {
        [BBOOrderType.COUNTERPARTY1]: t("orderEntry.bbo.counterparty1"),
        [BBOOrderType.COUNTERPARTY5]: t("orderEntry.bbo.counterparty5"),
        [BBOOrderType.QUEUE1]: t("orderEntry.bbo.queue1"),
        [BBOOrderType.QUEUE5]: t("orderEntry.bbo.queue5")
      }[bboType];
      return /* @__PURE__ */ jsx(Text, { intensity: 80, children: label });
    }
    return /* @__PURE__ */ jsx(
      Text.numeral,
      {
        unit: quote,
        rule: "price",
        className: "oui-text-base-contrast",
        unitClassName: "oui-text-base-contrast-36 oui-ml-1",
        dp: quote_dp,
        padding: false,
        children: order.order_price
      }
    );
  };
  const renderTPSLPrice = ({
    price,
    isOrderPrice,
    isEnable,
    colorType
  }) => {
    if (!isEnable) {
      return /* @__PURE__ */ jsx(Text, { className: "oui-text-base-contrast-36", children: "-- USDC" });
    }
    if (!price) {
      if (isOrderPrice) {
        return /* @__PURE__ */ jsx(Text, { className: "oui-text-base-contrast-36", children: t("common.marketPrice") });
      }
    }
    return /* @__PURE__ */ jsx(
      Text.numeral,
      {
        unit: "USDC",
        rule: "price",
        className: cn(
          "oui-text-base-contrast",
          colorType === "TP" ? "oui-text-trade-profit" : "oui-text-trade-loss"
        ),
        unitClassName: "oui-text-base-contrast-36 oui-ml-1",
        dp: quote_dp,
        padding: false,
        children: price
      }
    );
  };
  const renderTPSLQty = () => {
    if (!positionQty || !order.order_quantity) {
      return null;
    }
    let qty = new Decimal(order.order_quantity);
    if (order.position_type === PositionType.FULL) {
      qty = qty.plus(new Decimal(positionQty ?? 0));
    }
    return /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
      /* @__PURE__ */ jsx(Text, { children: order.position_type === PositionType.FULL ? t("common.positionQty") : t("common.orderQty") }),
      /* @__PURE__ */ jsx(
        Text.numeral,
        {
          rule: "price",
          dp: base_dp,
          padding: false,
          className: "oui-text-base-contrast",
          children: qty.toNumber()
        }
      )
    ] });
  };
  const renderPriceAndTotal = () => {
    if (order_type === OrderType.TRAILING_STOP) {
      const { activated_price, callback_value, callback_rate } = order;
      const callbackView = callback_rate ? /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
        /* @__PURE__ */ jsx(Text, { children: t("orderEntry.trailingRate") }),
        /* @__PURE__ */ jsxs(Text, { className: "oui-text-base-contrast", children: [
          callback_rate,
          "%"
        ] })
      ] }) : /* @__PURE__ */ jsx(
        OrderItem,
        {
          title: t("orderEntry.trailingValue"),
          value: callback_value,
          unit: quote,
          dp: quote_dp
        }
      );
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        activated_price && /* @__PURE__ */ jsx(
          OrderItem,
          {
            title: t("common.triggerPrice"),
            value: activated_price,
            unit: quote,
            dp: quote_dp
          }
        ),
        callbackView
      ] });
    }
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.price") }),
        renderPrice()
      ] }),
      /* @__PURE__ */ jsx(
        OrderItem,
        {
          title: t("common.estTotal"),
          value: order.total,
          unit: quote,
          dp: quote_dp
        }
      )
    ] });
  };
  const header = /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
    /* @__PURE__ */ jsx(Text.formatted, { rule: "symbol", showIcon: true, children: order.symbol }),
    /* @__PURE__ */ jsxs(Flex, { justify: "end", gapX: 1, children: [
      /* @__PURE__ */ jsx(OrderTypeTag, { type: order_type }),
      side === OrderSide.BUY ? /* @__PURE__ */ jsx(Badge, { color: "buy", size: "sm", children: t("common.buy") }) : /* @__PURE__ */ jsx(Badge, { color: "sell", size: "sm", children: t("common.sell") })
    ] })
  ] });
  const quantityItem = /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
    /* @__PURE__ */ jsx(Text, { children: t("common.orderQty") }),
    /* @__PURE__ */ jsx(
      Text.numeral,
      {
        rule: "price",
        dp: base_dp,
        padding: false,
        className: "oui-text-base-contrast",
        children: order.order_quantity
      }
    )
  ] });
  const triggerPriceItem = (order_type === OrderType.STOP_LIMIT || order_type === OrderType.STOP_MARKET && order.trigger_price) && /* @__PURE__ */ jsx(
    OrderItem,
    {
      title: t("common.trigger"),
      value: order.trigger_price,
      unit: quote,
      dp: quote_dp
    }
  );
  const tpslTriggerPrice = (order.tp_trigger_price || order.sl_trigger_price) && /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Divider, { className: "oui-my-4" }),
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: textVariants({
          size: "sm",
          intensity: 54,
          className: "oui-space-y-1 oui-w-full oui-flex oui-flex-col oui-gap-3"
        }),
        children: [
          /* @__PURE__ */ jsx(Text, { className: "oui-text-base-contrast", children: renderPositionType() }),
          renderTPSLQty(),
          /* @__PURE__ */ jsxs(
            Flex,
            {
              direction: "column",
              justify: "between",
              itemAlign: "start",
              gap: 1,
              className: "oui-w-full",
              children: [
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.tpTriggerPrice") }),
                  renderTPSLPrice({
                    price: order.tp_trigger_price ?? "",
                    isOrderPrice: false,
                    isEnable: !!order.tp_trigger_price,
                    colorType: "TP"
                  })
                ] }),
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.tpOrderPrice") }),
                  renderTPSLPrice({
                    price: order.tp_order_price ?? "",
                    isOrderPrice: true,
                    isEnable: !!order.tp_trigger_price,
                    colorType: "TP"
                  })
                ] })
              ]
            }
          ),
          /* @__PURE__ */ jsxs(
            Flex,
            {
              direction: "column",
              justify: "between",
              itemAlign: "start",
              gap: 1,
              className: "oui-w-full",
              children: [
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.slTriggerPrice") }),
                  renderTPSLPrice({
                    price: order.sl_trigger_price ?? "",
                    isOrderPrice: false,
                    isEnable: !!order.sl_trigger_price,
                    colorType: "SL"
                  })
                ] }),
                /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-w-full", children: [
                  /* @__PURE__ */ jsx(Text, { children: t("tpsl.slOrderPrice") }),
                  renderTPSLPrice({
                    price: order.sl_order_price ?? "",
                    isOrderPrice: true,
                    isEnable: !!order.sl_trigger_price,
                    colorType: "SL"
                  })
                ] })
              ]
            }
          )
        ]
      }
    )
  ] });
  const orderConfirmCheckbox = /* @__PURE__ */ jsxs(Flex, { gapX: 1, pt: 4, pb: 5, children: [
    /* @__PURE__ */ jsx(
      Checkbox,
      {
        id: "orderConfirm",
        color: "white",
        onCheckedChange: (checked) => {
          setNeedConfirm(!!!checked);
        }
      }
    ),
    /* @__PURE__ */ jsx(
      "label",
      {
        htmlFor: "orderConfirm",
        className: textVariants({
          size: "xs",
          intensity: 54
        }),
        children: t("orderEntry.disableOrderConfirm")
      }
    )
  ] });
  const buttons = /* @__PURE__ */ jsxs(Grid, { cols: 2, gapX: 3, children: [
    /* @__PURE__ */ jsx(Button, { color: "secondary", size: "md", onClick: () => onCancel(), children: t("common.cancel") }),
    /* @__PURE__ */ jsx(Button, { size: "md", onClick: () => onConfirm(), children: t("common.confirm") })
  ] });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    header,
    /* @__PURE__ */ jsx(Divider, { className: "oui-my-4" }),
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: textVariants({
          size: "sm",
          intensity: 54,
          className: "oui-space-y-1"
        }),
        children: [
          quantityItem,
          triggerPriceItem,
          renderPriceAndTotal()
        ]
      }
    ),
    tpslTriggerPrice,
    orderConfirmCheckbox,
    buttons
  ] });
};
var OrderItem = (props) => {
  const { title, value, unit, dp } = props;
  return /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
    /* @__PURE__ */ jsx(Text, { children: title }),
    /* @__PURE__ */ jsx(
      Text.numeral,
      {
        unit,
        rule: "price",
        dp,
        padding: false,
        className: "oui-text-base-contrast",
        unitClassName: "oui-text-base-contrast-36 oui-ml-1",
        children: value
      }
    )
  ] });
};
OrderConfirmDialog.displayName = "OrderConfirmDialog";
var OrderTypeTag = (props) => {
  const { t } = useTranslation();
  const typeStr = useMemo(() => {
    switch (props.type) {
      case OrderType.LIMIT:
        return t("orderEntry.orderType.limit");
      case OrderType.MARKET:
        return t("common.marketPrice");
      case OrderType.STOP_LIMIT:
        return t("orderEntry.orderType.stopLimit");
      case OrderType.STOP_MARKET:
        return t("orderEntry.orderType.stopMarket");
      case OrderType.TRAILING_STOP:
        return t("orderEntry.orderType.trailingStop");
      default:
        return "";
    }
  }, [props.type]);
  return /* @__PURE__ */ jsx(Badge, { color: "neutral", size: "sm", children: typeStr });
};
var Dialog = (props) => {
  const { close, resolve, reject, ...rest } = props;
  return /* @__PURE__ */ jsx(
    OrderConfirmDialog,
    {
      ...rest,
      onCancel: () => {
        reject();
        close();
      },
      onConfirm: () => {
        resolve();
        close();
      }
    }
  );
};
var orderConfirmDialogId = "orderConfirm";
registerSimpleDialog(orderConfirmDialogId, Dialog, {
  size: "sm",
  title: () => i18n.t("orderEntry.orderConfirm")
});
var MaxQtyConfirm = memo((props) => {
  const { t } = useTranslation();
  return /* @__PURE__ */ jsx(
    SimpleDialog,
    {
      open: props.open,
      title: t("orderEntry.orderConfirm"),
      closable: true,
      onOpenChange: props.onOpenChange,
      size: "sm",
      actions: {
        primary: {
          label: t("orderEntry.placeOrderNow"),
          className: "oui-text-sm oui-font-semibold oui-w-[100%] oui-h-8",
          onClick: () => {
            props.onConfirm();
            return Promise.resolve();
          }
        },
        secondary: {
          label: t("common.cancel"),
          className: "oui-text-sm oui-font-semibold oui-w-[100%] oui-h-8",
          onClick: () => {
            props.onOpenChange(false);
            return Promise.resolve();
          }
        }
      },
      children: /* @__PURE__ */ jsx("div", { className: "oui-text-2xs lg:oui-text-sm", children: t("orderEntry.maxQty.reminder.content", {
        maxQty: `${props.maxQty} ${props.base}`
      }) })
    }
  );
});
var ScaledOrderConfirm = (props) => {
  const { order, symbolInfo, dataSource, national, askAndBid, totalQuantity } = props;
  const { base, quote, base_dp, quote_dp } = symbolInfo;
  const { t } = useTranslation();
  const onCancel = () => {
    props.reject();
    props.close?.();
  };
  const onConfirm = () => {
    props.resolve();
    props.close?.();
  };
  const columns = useMemo(() => {
    return [
      {
        title: t("common.symbol"),
        dataIndex: "symbol",
        width: 125,
        render: (value, record) => {
          return /* @__PURE__ */ jsxs(Flex, { gap: 2, children: [
            /* @__PURE__ */ jsx(
              "div",
              {
                className: cn(
                  "oui-h-[38px] oui-w-1 oui-shrink-0 oui-rounded-[1px]",
                  record.side === OrderSide.BUY ? "oui-bg-trade-profit" : "oui-bg-trade-loss"
                )
              }
            ),
            /* @__PURE__ */ jsxs(Flex, { direction: "column", itemAlign: "start", children: [
              /* @__PURE__ */ jsxs(Flex, { gapX: 1, children: [
                /* @__PURE__ */ jsx(TokenIcon, { symbol: value, className: "oui-size-3" }),
                /* @__PURE__ */ jsx(
                  Text.formatted,
                  {
                    rule: "symbol",
                    size: "xs",
                    formatString: "base-type",
                    children: value
                  }
                )
              ] }),
              /* @__PURE__ */ jsx(Badge, { color: "neutral", size: "xs", children: t("orderEntry.orderType.limit") })
            ] })
          ] });
        }
      },
      {
        title: t("common.quantity"),
        dataIndex: "order_quantity",
        width: 100,
        render: (value, record) => {
          return /* @__PURE__ */ jsx(
            Text.numeral,
            {
              rule: "price",
              dp: base_dp,
              padding: false,
              color: record.side === OrderSide.BUY ? "buy" : "sell",
              children: value
            }
          );
        }
      },
      {
        title: t("common.orderPrice"),
        dataIndex: "order_price",
        width: 100,
        render: (value, record) => {
          const showWarning = !!(record.side === OrderSide.BUY ? askAndBid?.[0] && Number(value) >= askAndBid?.[0] : askAndBid?.[1] && Number(value) <= askAndBid?.[1]);
          return /* @__PURE__ */ jsxs(Flex, { gapX: 1, children: [
            /* @__PURE__ */ jsx(Text.numeral, { rule: "price", dp: quote_dp, children: value }),
            showWarning && /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t(
                  "orderEntry.confirmScaledOrder.orderPrice.warning"
                ),
                className: "oui-w-[240px] oui-text-2xs oui-font-semibold oui-text-base-contrast-80",
                children: /* @__PURE__ */ jsx(TooltipIcon, { className: "oui-text-warning-darken" })
              }
            )
          ] });
        }
      }
    ];
  }, [t, symbolInfo, askAndBid, base_dp, quote_dp]);
  return /* @__PURE__ */ jsxs("div", { className: "oui-font-semibold", children: [
    /* @__PURE__ */ jsx(
      DataTable,
      {
        classNames: {
          root: cn(
            "oui-bg-base-7",
            "oui-rounded-lg",
            // need to set overflow hidden because table header will avoid the border radius
            "oui-overflow-hidden",
            // "oui-text-2xs lg:oui-text-xs",
            // if orders is greater than 6, set the height to 320px to show scroll bar
            order.orders?.length >= 6 && "oui-h-[320px]"
          ),
          // set the min height of the table to show 2 rows
          scroll: "!oui-min-h-[130px]"
        },
        dataSource,
        columns,
        bordered: true,
        onRow: () => {
          return {
            className: cn("oui-h-[50px]")
          };
        }
      }
    ),
    /* @__PURE__ */ jsxs("div", { className: "oui-mb-5 oui-mt-4 oui-text-2xs", children: [
      /* @__PURE__ */ jsxs(Flex, { justify: "between", children: [
        /* @__PURE__ */ jsx(Text, { children: t("orderEntry.totalOrders") }),
        /* @__PURE__ */ jsx(Text, { intensity: 80, children: order.orders?.length })
      ] }),
      /* @__PURE__ */ jsxs(Flex, { justify: "between", mt: 2, children: [
        /* @__PURE__ */ jsx(Text, { children: t("orderEntry.totalQuantity") }),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            rule: "price",
            unit: base,
            dp: base_dp,
            padding: false,
            intensity: 80,
            unitClassName: "oui-text-base-contrast-36 oui-ml-1",
            children: totalQuantity
          }
        )
      ] }),
      /* @__PURE__ */ jsxs(Flex, { justify: "between", mt: 2, children: [
        /* @__PURE__ */ jsx(Text, { children: t("common.notional") }),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            rule: "price",
            unit: quote,
            dp: quote_dp,
            padding: false,
            intensity: 80,
            unitClassName: "oui-text-base-contrast-36 oui-ml-1",
            children: national
          }
        )
      ] })
    ] }),
    /* @__PURE__ */ jsxs(Grid, { cols: 2, gapX: 3, children: [
      /* @__PURE__ */ jsx(Button, { color: "secondary", size: "md", onClick: onCancel, children: t("common.cancel") }),
      /* @__PURE__ */ jsx(Button, { size: "md", onClick: onConfirm, children: t("common.confirm") })
    ] })
  ] });
};
var TooltipIcon = forwardRef(
  (props, ref) => {
    return /* @__PURE__ */ jsx(
      "svg",
      {
        width: "12",
        height: "13",
        viewBox: "0 0 12 13",
        fill: "currentColor",
        xmlns: "http://www.w3.org/2000/svg",
        ref,
        ...props,
        children: /* @__PURE__ */ jsx("path", { d: "M5.99951 1.50708C3.23811 1.50708 0.999512 3.74558 0.999512 6.50708C0.999512 9.26858 3.23811 11.5071 5.99951 11.5071C8.76091 11.5071 10.9995 9.26858 10.9995 6.50708C10.9995 3.74558 8.76091 1.50708 5.99951 1.50708ZM5.99951 4.00708C6.27566 4.00708 6.49951 4.23108 6.49951 4.50708C6.49951 4.78308 6.27566 5.00708 5.99951 5.00708C5.72336 5.00708 5.49951 4.78308 5.49951 4.50708C5.49951 4.23108 5.72336 4.00708 5.99951 4.00708ZM5.99951 5.50708C6.27566 5.50708 6.49951 5.73108 6.49951 6.00708V8.50708C6.49951 8.78308 6.27566 9.00708 5.99951 9.00708C5.72336 9.00708 5.49951 8.78308 5.49951 8.50708V6.00708C5.49951 5.73108 5.72336 5.50708 5.99951 5.50708Z" })
      }
    );
  }
);
function useAskAndBid() {
  const ee = useEventEmitter();
  const [askAndBid, setAskAndBid] = useState([0, 0]);
  const onOrderBookUpdate = useDebouncedCallback((data) => {
    const ask0 = data.asks?.[data.asks.length - 1]?.[0];
    const bid0 = data.bids?.[0]?.[0];
    setAskAndBid([ask0, bid0]);
  }, 200);
  useEffect(() => {
    ee.on("orderbook:update", onOrderBookUpdate);
    return () => {
      ee.off("orderbook:update", onOrderBookUpdate);
      onOrderBookUpdate.cancel();
    };
  }, [onOrderBookUpdate]);
  return askAndBid;
}

// src/components/dialog/scaledOrderConfirm/scaledOrderConfirm.script.ts
function useScaledOrderConfirmScript(options2) {
  const { order, symbolInfo } = options2;
  const orders = order.orders;
  const askAndBid = useAskAndBid();
  const national = useMemo(() => {
    const national2 = orders.reduce((acc, order2) => {
      return acc.add(new Decimal(order2.order_price).mul(order2.order_quantity));
    }, zero);
    return national2.toNumber();
  }, [orders]);
  const totalQuantity = useMemo(() => {
    const totalQuantity2 = orders.reduce((acc, order2) => {
      return acc.add(new Decimal(order2.order_quantity));
    }, zero);
    return totalQuantity2.toString();
  }, [orders, symbolInfo.base_dp]);
  return { dataSource: orders, national, askAndBid, totalQuantity };
}
var ScaledOrderConfirmWidget = (props) => {
  const state = useScaledOrderConfirmScript(props);
  return /* @__PURE__ */ jsx(ScaledOrderConfirm, { ...props, ...state });
};
var scaledOrderConfirmDialogId = "scaledOrderConfirm";
registerSimpleDialog(scaledOrderConfirmDialogId, ScaledOrderConfirmWidget, {
  size: "md",
  title: () => i18n.t("orderEntry.confirmScaledOrder"),
  contentProps: {
    // className: "oui-p-0",
  }
});
var OrderTypeSelect = (props) => {
  const { t } = useTranslation();
  const options2 = useMemo(() => {
    return [
      { label: t("orderEntry.orderType.limitOrder"), value: OrderType.LIMIT },
      { label: t("orderEntry.orderType.marketOrder"), value: OrderType.MARKET },
      {
        label: t("orderEntry.orderType.stopLimit"),
        value: OrderType.STOP_LIMIT
      },
      {
        label: t("orderEntry.orderType.stopMarket"),
        value: OrderType.STOP_MARKET
      },
      {
        label: t("orderEntry.orderType.scaledOrder"),
        value: OrderType.SCALED
      },
      {
        label: t("orderEntry.orderType.trailingStop"),
        value: OrderType.TRAILING_STOP
      }
    ];
  }, [t]);
  const displayLabelMap = useMemo(() => {
    return {
      [OrderType.LIMIT]: t("orderEntry.orderType.limit"),
      [OrderType.MARKET]: t("common.marketPrice"),
      [OrderType.STOP_LIMIT]: t("orderEntry.orderType.stopLimit"),
      [OrderType.STOP_MARKET]: t("orderEntry.orderType.stopMarket"),
      [OrderType.SCALED]: t("orderEntry.orderType.scaledOrder"),
      [OrderType.TRAILING_STOP]: t("orderEntry.orderType.trailingStop")
    };
  }, [t]);
  return /* @__PURE__ */ jsx(
    Select.options,
    {
      testid: "oui-testid-orderEntry-orderType-button",
      currentValue: props.type,
      value: props.type,
      options: options2,
      onValueChange: props.onChange,
      contentProps: {
        className: "oui-bg-base-8"
      },
      classNames: {
        trigger: "oui-bg-base-6 oui-border-line"
      },
      valueFormatter: (value, option) => {
        const item = options2.find((o) => o.value === value);
        if (!item) {
          return /* @__PURE__ */ jsx(Text, { size: "xs", children: option.placeholder });
        }
        const label = displayLabelMap[value];
        return /* @__PURE__ */ jsx(
          Text,
          {
            size: "xs",
            color: props.canTrade ? props.side === OrderSide.BUY ? "buy" : "sell" : void 0,
            children: label
          }
        );
      },
      size: "md"
    }
  );
};
var LeverageBadge = (props) => {
  const { symbol, side, symbolLeverage } = props;
  const { isMobile } = useScreen();
  const { maxLeverage } = useSymbolLeverage(symbol);
  const curLeverage = symbolLeverage || maxLeverage;
  const showModal = () => {
    const modalId = isMobile ? SymbolLeverageSheetId : SymbolLeverageDialogId;
    modal.show(modalId, {
      symbol,
      side,
      curLeverage
    });
  };
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      justify: "center",
      itemAlign: "center",
      gapX: 1,
      className: cn(
        "oui-h-8",
        "oui-rounded oui-border oui-border-line oui-bg-base-6",
        "oui-cursor-pointer oui-select-none oui-text-xs oui-font-semibold oui-text-base-contrast-54"
      ),
      onClick: showModal,
      children: [
        /* @__PURE__ */ jsx(Text, { children: "Cross" }),
        /* @__PURE__ */ jsx(Text.numeral, { dp: 0, rm: Decimal.ROUND_DOWN, unit: "X", children: curLeverage })
      ]
    }
  );
};
function OrderEntryHeader(props) {
  const { canTrade, side, order_type, setOrderValue } = props;
  const { t } = useTranslation();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: cn(
          "oui-grid oui-w-full oui-flex-1 oui-gap-x-2 lg:oui-flex lg:oui-gap-x-[6px]",
          "oui-grid-cols-2"
        ),
        children: [
          /* @__PURE__ */ jsx(
            Button,
            {
              onClick: () => {
                props.setOrderValue("side", OrderSide.BUY);
              },
              size: "md",
              fullWidth: true,
              "data-type": OrderSide.BUY,
              className: cn(
                side === OrderSide.BUY && canTrade ? "oui-bg-success-darken hover:oui-bg-success-darken/80 active:oui-bg-success-darken/80" : "oui-bg-base-7 oui-text-base-contrast-36 hover:oui-bg-base-6 active:oui-bg-base-6"
              ),
              "data-testid": "oui-testid-orderEntry-side-buy-button",
              children: t("common.buy")
            }
          ),
          /* @__PURE__ */ jsx(
            Button,
            {
              onClick: () => {
                props.setOrderValue("side", OrderSide.SELL);
              },
              "data-type": OrderSide.SELL,
              fullWidth: true,
              size: "md",
              className: cn(
                side === OrderSide.SELL && props.canTrade ? "oui-bg-danger-darken hover:oui-bg-danger-darken/80 active:oui-bg-danger-darken/80" : "oui-bg-base-7 oui-text-base-contrast-36 hover:oui-bg-base-6 active:oui-bg-base-6"
              ),
              "data-testid": "oui-testid-orderEntry-side-sell-button",
              children: t("common.sell")
            }
          )
        ]
      }
    ),
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: cn(
          "oui-grid oui-gap-x-2 lg:oui-flex lg:oui-gap-x-[6px]",
          "oui-grid-cols-2"
        ),
        children: [
          /* @__PURE__ */ jsx("div", { className: "oui-w-full", children: /* @__PURE__ */ jsx(
            OrderTypeSelect,
            {
              type: order_type,
              side,
              canTrade,
              onChange: (type) => {
                setOrderValue("order_type", type);
              }
            }
          ) }),
          /* @__PURE__ */ jsx("div", { className: "oui-w-full", children: /* @__PURE__ */ jsx(
            LeverageBadge,
            {
              symbol: props.symbol,
              side: props.side,
              symbolLeverage: props.symbolLeverage
            }
          ) })
        ]
      }
    )
  ] });
}
var OrderEntryContext = createContext(
  {}
);
var useOrderEntryContext = () => {
  return useContext(OrderEntryContext);
};
var OrderEntryProvider = (props) => {
  const {
    errorMsgVisible,
    symbolInfo,
    onFocus,
    onBlur,
    getErrorMsg,
    setOrderValue,
    setOrderValues,
    currentFocusInput,
    errors,
    priceInputRef,
    priceInputContainerRef,
    triggerPriceInputRef,
    activatedPriceInputRef,
    lastQuantityInputType,
    leverage
  } = props;
  const memoizedValue = useMemo(() => {
    return {
      errorMsgVisible,
      symbolInfo,
      onFocus,
      onBlur,
      getErrorMsg,
      setOrderValue,
      setOrderValues,
      currentFocusInput,
      errors,
      // refs
      priceInputRef,
      priceInputContainerRef,
      triggerPriceInputRef,
      activatedPriceInputRef,
      lastQuantityInputType,
      leverage
    };
  }, [
    errorMsgVisible,
    symbolInfo,
    onFocus,
    onBlur,
    getErrorMsg,
    setOrderValue,
    setOrderValues,
    currentFocusInput,
    errors,
    priceInputRef,
    priceInputContainerRef,
    triggerPriceInputRef,
    activatedPriceInputRef,
    lastQuantityInputType,
    leverage
  ]);
  return /* @__PURE__ */ jsx(OrderEntryContext.Provider, { value: memoizedValue, children: props.children });
};
var CustomInput = forwardRef(
  (props, ref) => {
    const { placeholder = "0" } = props;
    const { errorMsgVisible } = useOrderEntryContext();
    return /* @__PURE__ */ jsx(
      Input.tooltip,
      {
        ref,
        tooltip: errorMsgVisible ? props.error : void 0,
        autoComplete: "off",
        autoFocus: props.autoFocus,
        size: "lg",
        placeholder: props.readonly ? "" : placeholder,
        id: props.id,
        name: props.name,
        color: props.error ? "danger" : void 0,
        prefix: props.prefix || /* @__PURE__ */ jsx(InputLabel, { id: props.id, className: props.classNames?.prefix, children: props.label }),
        suffix: props.suffix,
        value: props.readonly ? "" : props.value || "",
        onValueChange: props.onChange,
        onFocus: (event) => {
          props.onFocus?.(event);
        },
        onBlur: (event) => {
          props.onBlur?.(event);
        },
        formatters: props.overrideFormatters || [
          ...props.formatters ?? EMPTY_LIST,
          inputFormatter.numberFormatter,
          inputFormatter.currencyFormatter,
          inputFormatter.decimalPointFormatter
        ],
        classNames: {
          root: cn(
            "orderly-order-entry oui-relative oui-h-[54px] oui-rounded oui-border oui-border-solid oui-border-line oui-px-2 oui-py-1 group-first:oui-rounded-t-xl group-last:oui-rounded-b-xl",
            props.className,
            props.classNames?.root
          ),
          input: cn("oui-mb-1 oui-mt-5 oui-h-5", props?.classNames?.input),
          // prefix: cn(props.classNames?.prefix),
          suffix: cn(
            "oui-absolute oui-right-0 oui-top-0 oui-justify-start oui-py-2 oui-text-2xs oui-text-base-contrast-36",
            props.classNames?.suffix
          )
        },
        readOnly: props.readonly
      }
    );
  }
);
CustomInput.displayName = "CustomInput";
var InputLabel = (props) => {
  return /* @__PURE__ */ jsx(
    "label",
    {
      htmlFor: props.id,
      className: cn(
        "oui-absolute oui-left-2 oui-top-[7px] oui-text-2xs oui-text-base-contrast-36",
        props.className
      ),
      children: props.children
    }
  );
};
var BBOOrderTypeSelect = (props) => {
  const { t } = useTranslation();
  const options2 = useMemo(
    () => [
      {
        label: t("orderEntry.bbo.counterparty1"),
        value: BBOOrderType.COUNTERPARTY1
      },
      {
        label: t("orderEntry.bbo.counterparty5"),
        value: BBOOrderType.COUNTERPARTY5
      },
      {
        label: t("orderEntry.bbo.queue1"),
        value: BBOOrderType.QUEUE1
      },
      {
        label: t("orderEntry.bbo.queue5"),
        value: BBOOrderType.QUEUE5
      }
    ],
    []
  );
  return /* @__PURE__ */ jsx(
    Select.options,
    {
      testid: "oui-testid-orderEntry-bbo-orderType-button",
      currentValue: props.value,
      value: props.value,
      options: options2,
      onValueChange: props.onChange,
      contentProps: {
        className: "oui-bg-base-8 oui-w-full",
        style: props.contentStyle
      },
      size: "sm",
      classNames: {
        trigger: "oui-border-none oui-bg-transparent"
      },
      valueFormatter: (value, option) => {
        const item = options2.find((item2) => item2.value === value);
        return /* @__PURE__ */ jsx(Box, { children: /* @__PURE__ */ jsx(Text, { size: "sm", children: item?.label }) });
      }
    }
  );
};
var LimitPriceSuffix = (props) => {
  const { quote, bbo, fillMiddleValue } = props;
  const { t } = useTranslation();
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      direction: "column",
      itemAlign: "end",
      className: cn("oui-order-entry-limit-price-input-suffix", "oui-text-2xs"),
      children: [
        quote,
        /* @__PURE__ */ jsxs(Flex, { justify: "end", itemAlign: "center", gap: 2, children: [
          /* @__PURE__ */ jsx(
            Flex,
            {
              px: 3,
              height: 20,
              justify: "center",
              itemAlign: "center",
              r: "base",
              className: cn(
                "oui-mt-[2px] oui-cursor-pointer oui-select-none oui-border",
                bbo.bboStatus === "on" /* ON */ ? "oui-border-primary" : "oui-border-line-12",
                bbo.bboStatus === "disabled" /* DISABLED */ && "oui-cursor-not-allowed"
              ),
              onClick: () => {
                if (bbo.bboStatus === "disabled" /* DISABLED */) {
                  modal.dialog({
                    title: t("common.tips"),
                    size: "xs",
                    content: /* @__PURE__ */ jsx(Text, { intensity: 54, children: t("orderEntry.bbo.disabled.tips") })
                  });
                } else {
                  bbo.toggleBBO();
                }
              },
              children: /* @__PURE__ */ jsx(
                Text,
                {
                  className: cn(
                    bbo.bboStatus === "on" /* ON */ && "oui-text-primary",
                    bbo.bboStatus === "off" /* OFF */ && "oui-text-base-contrast-54",
                    bbo.bboStatus === "disabled" /* DISABLED */ && "oui-text-base-contrast-20"
                  ),
                  children: t("orderEntry.bbo")
                }
              )
            }
          ),
          /* @__PURE__ */ jsx(
            Text,
            {
              className: cn(
                "oui-select-none",
                "oui-cursor-pointer oui-text-primary"
              ),
              onClick: fillMiddleValue,
              children: "Mid"
            }
          )
        ] })
      ]
    }
  );
};
var PriceInput = (props) => {
  const { bbo } = props;
  const { t } = useTranslation();
  const {
    symbolInfo,
    onFocus,
    onBlur,
    getErrorMsg,
    setOrderValue,
    priceInputRef,
    priceInputContainerRef
  } = useOrderEntryContext();
  const { quote, quote_dp } = symbolInfo;
  const readOnly = bbo.bboStatus === "on" /* ON */;
  const suffix = props.order_type === OrderType.LIMIT ? /* @__PURE__ */ jsx(
    LimitPriceSuffix,
    {
      quote,
      bbo,
      fillMiddleValue: props.fillMiddleValue
    }
  ) : quote;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref: priceInputContainerRef,
      className: "oui-group oui-relative oui-w-full",
      children: [
        /* @__PURE__ */ jsx(
          CustomInput,
          {
            id: "order_price_input",
            name: "order_price_input",
            label: t("common.price"),
            suffix,
            value: props.order_price,
            onChange: (e) => {
              setOrderValue("order_price", e);
            },
            error: getErrorMsg("order_price"),
            formatters: [inputFormatter.dpFormatter(quote_dp)],
            onFocus: onFocus(1 /* PRICE */),
            onBlur: onBlur(1 /* PRICE */),
            readonly: readOnly,
            ref: priceInputRef,
            classNames: {
              root: cn(readOnly && "focus-within:oui-outline-transparent "),
              input: cn(readOnly && "oui-cursor-auto")
            }
          }
        ),
        bbo.bboStatus === "on" /* ON */ && /* @__PURE__ */ jsx("div", { className: cn("oui-absolute oui-bottom-1 oui-left-0"), children: /* @__PURE__ */ jsx(
          BBOOrderTypeSelect,
          {
            value: bbo.bboType,
            onChange: bbo.onBBOChange,
            contentStyle: {
              width: props.priceInputContainerWidth
            }
          }
        ) })
      ]
    }
  );
};
var QuantityInput = memo((props) => {
  const { t } = useTranslation();
  const { symbolInfo, onFocus, onBlur, getErrorMsg, setOrderValue } = useOrderEntryContext();
  const { base, base_dp } = symbolInfo;
  return /* @__PURE__ */ jsx(
    CustomInput,
    {
      id: "order_quantity_input",
      name: "order_quantity_input",
      label: t("common.qty"),
      suffix: base,
      error: getErrorMsg("order_quantity"),
      value: props.order_quantity,
      onChange: (e) => {
        setOrderValue("order_quantity", e);
      },
      formatters: [inputFormatter.dpFormatter(base_dp)],
      onFocus: onFocus(3 /* QUANTITY */),
      onBlur: onBlur(3 /* QUANTITY */),
      className: "!oui-rounded-r",
      classNames: {
        suffix: "oui-justify-end"
      }
    }
  );
});
QuantityInput.displayName = "QuantityInput";
var TotalTypeSelect = memo((props) => {
  const { t } = useTranslation();
  const options2 = useMemo(() => {
    return [
      {
        label: t("orderEntry.orderSize"),
        value: "orderSize" /* OrderSize */
      },
      {
        label: t("orderEntry.initialMargin"),
        value: "initialMargin" /* InitialMargin */
      }
    ];
  }, [t]);
  return /* @__PURE__ */ jsx(
    Select.options,
    {
      size: "xs",
      value: props.value,
      valueFormatter: (value, option) => {
        const item = options2.find((o) => o.value === value);
        return item?.label + "\u2248";
      },
      classNames: {
        trigger: cn(
          "oui-w-auto oui-border-none oui-bg-transparent",
          "oui-absolute oui-left-0 oui-top-[5px] oui-text-2xs oui-text-base-contrast-36"
        )
      },
      onValueChange: props.onChange,
      options: options2
    }
  );
});
TotalTypeSelect.displayName = "TotalTypeSelect";
var TotalInput = memo((props) => {
  const { t } = useTranslation();
  const { total } = props;
  const [margin, setMargin] = useState("");
  const {
    symbolInfo,
    onFocus,
    onBlur,
    getErrorMsg,
    setOrderValue,
    leverage = 1,
    currentFocusInput
  } = useOrderEntryContext();
  const { quote } = symbolInfo;
  const [totalType, setTotalType] = useLocalStorage(
    "orderly_order_total_type",
    "orderSize" /* OrderSize */
  );
  useEffect(() => {
    if (total) {
      if (currentFocusInput !== 6 /* MARGIN */) {
        const margin2 = new Decimal(total).div(leverage).todp(2).toString();
        setMargin(margin2);
      }
    } else {
      setMargin("");
    }
  }, [total, leverage, currentFocusInput]);
  const onMarginChange = (val) => {
    const total2 = val ? new Decimal(val).mul(leverage).toString() : "";
    setOrderValue("total", total2);
    setMargin(val);
  };
  const prefix = /* @__PURE__ */ jsx(TotalTypeSelect, { value: totalType, onChange: setTotalType });
  if (totalType === "initialMargin" /* InitialMargin */) {
    return /* @__PURE__ */ jsx(
      CustomInput,
      {
        id: "order_margin_input",
        name: "order_margin_input",
        label: `${t("orderEntry.initialMargin")}\u2248`,
        prefix,
        suffix: quote,
        error: getErrorMsg("total") ? "initial margin error" : "",
        value: margin,
        onChange: onMarginChange,
        formatters: [inputFormatter.dpFormatter(2)],
        onFocus: onFocus(6 /* MARGIN */),
        onBlur: onBlur(6 /* MARGIN */),
        className: "!oui-rounded-l",
        classNames: {
          suffix: "oui-justify-end"
        }
      }
    );
  }
  return /* @__PURE__ */ jsx(
    CustomInput,
    {
      id: "order_total_input",
      name: "order_total_input",
      label: `${t("orderEntry.orderSize")}\u2248`,
      prefix,
      suffix: quote,
      error: getErrorMsg("total"),
      value: props.total,
      onChange: (val) => {
        setOrderValue("total", val);
      },
      className: "!oui-rounded-l",
      classNames: {
        suffix: "oui-justify-end"
      },
      formatters: [inputFormatter.dpFormatter(2)],
      onFocus: onFocus(5 /* TOTAL */),
      onBlur: onBlur(5 /* TOTAL */)
    }
  );
});
TotalInput.displayName = "TotalInput";
var QtyAndTotalInput = memo((props) => {
  return /* @__PURE__ */ jsxs(Grid, { cols: 2, className: "oui-group oui-space-x-1", children: [
    /* @__PURE__ */ jsx(QuantityInput, { order_quantity: props.order_quantity }),
    /* @__PURE__ */ jsx(TotalInput, { total: props.total })
  ] });
});
QtyAndTotalInput.displayName = "QtyAndTotalInput";
var QuantityDistributionInput = memo((props) => {
  const { t } = useTranslation();
  const { setOrderValue } = useOrderEntryContext();
  const showHint = () => {
    modal.dialog({
      title: t("common.tips"),
      size: "sm",
      content: /* @__PURE__ */ jsx(QuantityDistributionHint, { value: props.distribution_type })
      // classNames: {
      //   content: "oui-bg-base-6",
      // },
    });
  };
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      direction: "column",
      itemAlign: "start",
      justify: "center",
      p: 2,
      r: "base",
      width: "100%",
      intensity: 600,
      className: cn(
        "oui-t-rounded oui-text-base-contrast-36",
        "oui-border oui-border-solid oui-border-line",
        props.className
      ),
      children: [
        /* @__PURE__ */ jsx(
          Text,
          {
            size: "2xs",
            className: "oui-cursor-pointer oui-border-b oui-border-dashed oui-border-line-12",
            onClick: showHint,
            children: t("orderEntry.quantityDistribution")
          }
        ),
        /* @__PURE__ */ jsx(
          QuantityDistribution,
          {
            value: props.distribution_type,
            onValueChange: (value) => {
              setOrderValue("distribution_type", value);
            }
          }
        )
      ]
    }
  );
});
var QuantityDistributionHint = (props) => {
  const { t } = useTranslation();
  const [type, setType] = useState(DistributionType.FLAT);
  useEffect(() => {
    setType(
      [
        DistributionType.FLAT,
        DistributionType.ASCENDING,
        DistributionType.DESCENDING
      ].includes(props.value) ? props.value : DistributionType.FLAT
    );
  }, [props.value]);
  const content = useMemo(() => {
    return [
      {
        type: DistributionType.FLAT,
        title: t("orderEntry.distributionType.flat"),
        description: t("orderEntry.distributionType.flat.description"),
        formula: `${t("orderEntry.skew")} = 1`,
        quantity: /* @__PURE__ */ jsx(FlatQuantity, {})
      },
      {
        type: DistributionType.ASCENDING,
        title: t("orderEntry.distributionType.ascending"),
        description: t("orderEntry.distributionType.ascending.description"),
        formula: `${t("orderEntry.skew")} > 1`,
        quantity: /* @__PURE__ */ jsx(AscendingQuantity, {})
      },
      {
        type: DistributionType.DESCENDING,
        title: t("orderEntry.distributionType.descending"),
        description: t("orderEntry.distributionType.descending.description"),
        formula: `0 < ${t("orderEntry.skew")} < 1`,
        quantity: /* @__PURE__ */ jsx(DescendingQuantity, {})
      }
    ];
  }, []);
  const currentContent = useMemo(() => {
    return content.find((item) => item.type === type);
  }, [content, type]);
  return /* @__PURE__ */ jsxs("div", { className: "oui-text-2xs oui-font-semibold oui-text-base-contrast-54", children: [
    /* @__PURE__ */ jsxs(Flex, { direction: "column", itemAlign: "start", gapY: 1, children: [
      /* @__PURE__ */ jsx(Text, { intensity: 80, children: t("orderEntry.quantityDistribution.description") }),
      /* @__PURE__ */ jsx(Text, { children: t("orderEntry.quantityDistribution.formula") })
    ] }),
    /* @__PURE__ */ jsxs(Flex, { intensity: 600, p: 1, r: "base", mt: 3, itemAlign: "start", children: [
      /* @__PURE__ */ jsx(Flex, { direction: "column", itemAlign: "start", children: content.map((item) => /* @__PURE__ */ jsx(
        Box,
        {
          intensity: type === item.type ? 500 : 600,
          width: 78,
          p: 2,
          className: "oui-cursor-pointer oui-rounded-l",
          onClick: () => {
            setType(item.type);
          },
          children: /* @__PURE__ */ jsx(Text, { children: item.title })
        },
        item.title
      )) }),
      /* @__PURE__ */ jsxs(
        Flex,
        {
          direction: "column",
          itemAlign: "start",
          gapY: 1,
          p: 2,
          intensity: 500,
          width: "100%",
          className: cn(
            "oui-rounded-r",
            type === DistributionType.FLAT && "oui-rounded-bl",
            type === DistributionType.ASCENDING && "oui-rounded-l",
            type === DistributionType.DESCENDING && "oui-rounded-tl"
          ),
          children: [
            /* @__PURE__ */ jsx("div", { children: currentContent?.formula }),
            /* @__PURE__ */ jsx(Text, { children: currentContent?.description }),
            /* @__PURE__ */ jsx(Flex, { width: "100%", justify: "center", children: currentContent?.quantity }),
            /* @__PURE__ */ jsxs(Flex, { width: "100%", gapX: 1, justify: "center", children: [
              /* @__PURE__ */ jsx(Text, { className: "oui-text-base-1", children: t("common.price") }),
              /* @__PURE__ */ jsx(PriceChart, {})
            ] })
          ]
        }
      )
    ] })
  ] });
};
var QuantityDistribution = (props) => {
  const { value, onValueChange } = props;
  const { t } = useTranslation();
  const onChange = (value2) => (checked) => {
    onValueChange(value2);
  };
  const distributionTypeMap = useMemo(() => {
    return {
      [DistributionType.FLAT]: t("orderEntry.distributionType.flat"),
      [DistributionType.ASCENDING]: t(
        "orderEntry.distributionType.ascending.abbr"
      ),
      [DistributionType.DESCENDING]: t(
        "orderEntry.distributionType.descending.abbr"
      ),
      [DistributionType.CUSTOM]: t("orderEntry.distributionType.custom")
    };
  }, [t]);
  return /* @__PURE__ */ jsx(Flex, { className: cn("oui-gap-x-[6px] lg:oui-gap-x-2"), wrap: "wrap", children: Object.values(DistributionType).map((type) => {
    return /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", children: [
      /* @__PURE__ */ jsx(
        Checkbox,
        {
          id: `distribution-type-${type}`,
          color: "white",
          variant: "radio",
          checked: value === type,
          onCheckedChange: onChange(type)
        }
      ),
      /* @__PURE__ */ jsx(
        "label",
        {
          htmlFor: `distribution-type-${type}`,
          className: cn(
            "oui-text-2xs",
            "oui-ml-1",
            "oui-whitespace-nowrap oui-break-normal"
          ),
          children: distributionTypeMap[type]
        }
      )
    ] }, type);
  }) });
};
var FlatQuantity = () => {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      width: "218",
      height: "40",
      viewBox: "0 0 218 40",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: [
        /* @__PURE__ */ jsx(
          "rect",
          {
            x: "18",
            y: "18",
            width: "181",
            height: "4",
            fill: "rgb(var(--oui-color-base-2))"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "15",
            cy: "20",
            r: "6",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "62",
            cy: "20",
            r: "6",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "109",
            cy: "20",
            r: "6",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "156",
            cy: "20",
            r: "6",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "203",
            cy: "20",
            r: "6",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        )
      ]
    }
  );
};
var AscendingQuantity = () => {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      width: "218",
      height: "40",
      viewBox: "0 0 218 40",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: [
        /* @__PURE__ */ jsx(
          "rect",
          {
            x: "18",
            y: "18",
            width: "181",
            height: "4",
            fill: "rgb(var(--oui-color-base-2))"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "15",
            cy: "20",
            r: "4.28571",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "1.42857"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "62",
            cy: "20",
            r: "5.14286",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "1.71429"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "109",
            cy: "20",
            r: "6",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "156",
            cy: "20",
            r: "6.85714",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2.28571"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "203",
            cy: "20",
            r: "7.71429",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2.57143"
          }
        )
      ]
    }
  );
};
var DescendingQuantity = () => {
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      width: "218",
      height: "40",
      viewBox: "0 0 218 40",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: [
        /* @__PURE__ */ jsx(
          "rect",
          {
            width: "181",
            height: "4",
            transform: "matrix(-1 0 0 1 200 18)",
            fill: "rgb(var(--oui-color-base-2))"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "5",
            cy: "5",
            r: "4.28571",
            transform: "matrix(-1 0 0 1 208 15)",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "1.42857"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "6",
            cy: "6",
            r: "5.14286",
            transform: "matrix(-1 0 0 1 162 14)",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "1.71429"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "7",
            cy: "7",
            r: "6",
            transform: "matrix(-1 0 0 1 116 13)",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "8",
            cy: "8",
            r: "6.85714",
            transform: "matrix(-1 0 0 1 70 12)",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2.28571"
          }
        ),
        /* @__PURE__ */ jsx(
          "circle",
          {
            cx: "9",
            cy: "9",
            r: "7.71429",
            transform: "matrix(-1 0 0 1 24 11)",
            fill: "rgb(var(--oui-color-base-2))",
            stroke: "rgb(var(--oui-color-base-5))",
            strokeWidth: "2.57143"
          }
        )
      ]
    }
  );
};
var PriceChart = () => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      width: "176",
      height: "8",
      viewBox: "0 0 176 8",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: /* @__PURE__ */ jsx(
        "path",
        {
          d: "M175.354 4.35355C175.549 4.15829 175.549 3.84171 175.354 3.64645L172.172 0.464466C171.976 0.269204 171.66 0.269204 171.464 0.464466C171.269 0.659728 171.269 0.976311 171.464 1.17157L174.293 4L171.464 6.82843C171.269 7.02369 171.269 7.34027 171.464 7.53553C171.66 7.7308 171.976 7.7308 172.172 7.53553L175.354 4.35355ZM0 4V4.5H175V4V3.5H0V4Z",
          fill: "rgb(var(--oui-color-base-1))"
        }
      )
    }
  );
};
var ScaledPriceInput = memo((props) => {
  const { t } = useTranslation();
  const { symbolInfo, onFocus, onBlur, getErrorMsg, setOrderValue } = useOrderEntryContext();
  const { quote, quote_dp } = symbolInfo;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      CustomInput,
      {
        label: t("orderEntry.startPrice"),
        suffix: quote,
        id: "order_start_price_input",
        value: props.start_price,
        error: getErrorMsg("start_price"),
        onChange: (e) => {
          setOrderValue("start_price", e);
        },
        formatters: [inputFormatter.dpFormatter(quote_dp)],
        onFocus: onFocus(7 /* START_PRICE */),
        onBlur: onBlur(7 /* START_PRICE */),
        classNames: {
          root: "oui-rounded-t-xl"
        }
      }
    ),
    /* @__PURE__ */ jsx(
      CustomInput,
      {
        label: t("orderEntry.endPrice"),
        suffix: quote,
        id: "order_end_price_input",
        value: props.end_price,
        error: getErrorMsg("end_price"),
        onChange: (val) => {
          setOrderValue("end_price", val);
        },
        formatters: [inputFormatter.dpFormatter(quote_dp)],
        onFocus: onFocus(8 /* END_PRICE */),
        onBlur: onBlur(8 /* END_PRICE */)
      }
    )
  ] });
});
ScaledPriceInput.displayName = "ScaledPriceInput";
var valueRenderer = (value) => {
  return /* @__PURE__ */ jsx(Text, { size: "2xs", intensity: 36, children: value });
};
var ScaledQuantityUnit = (props) => {
  const { base, quote } = props;
  const options2 = useMemo(() => {
    return [{ name: quote }, { name: base }];
  }, [base, quote]);
  return /* @__PURE__ */ jsx(
    Select.tokens,
    {
      variant: "text",
      size: "xs",
      iconSize: "2xs",
      tokens: options2,
      classNames: {
        trigger: cn(
          "oui-absolute oui-right-0 oui-top-1",
          "oui-w-full oui-justify-end"
        )
      },
      value: props.value,
      onValueChange: props.onValueChange,
      valueFormatter: valueRenderer,
      contentProps: {
        align: "end",
        alignOffset: -1,
        sideOffset: -4,
        className: "oui-border oui-border-line-6"
      }
    }
  );
};
var ScaledQuantityInput = memo((props) => {
  const { t } = useTranslation();
  const { errors, symbolInfo, onFocus, onBlur, getErrorMsg, setOrderValue } = useOrderEntryContext();
  const [quantityUnit, setQuantityUnit] = useLocalStorage(
    "orderly_order_quantity_unit",
    "quote"
  );
  const { base, quote, base_dp, quote_dp } = symbolInfo;
  const isBase = quantityUnit === "base";
  const unit = isBase ? base : quote;
  const suffix = /* @__PURE__ */ jsx(
    ScaledQuantityUnit,
    {
      base,
      quote,
      value: unit,
      onValueChange: (value) => {
        setQuantityUnit(value === base ? "base" : "quote");
      }
    }
  );
  if (isBase) {
    return /* @__PURE__ */ jsx(
      CustomInput,
      {
        label: t("common.qty"),
        suffix,
        id: "order_quantity_input",
        name: "order_quantity_input",
        className: "!oui-rounded-r",
        value: props.order_quantity,
        error: getErrorMsg(
          "order_quantity",
          `${errors?.order_quantity?.value} ${base}`
        ),
        onChange: (val) => {
          setOrderValue("order_quantity", val);
        },
        formatters: [inputFormatter.dpFormatter(base_dp)],
        onFocus: onFocus(3 /* QUANTITY */),
        onBlur: onBlur(3 /* QUANTITY */)
      }
    );
  }
  return /* @__PURE__ */ jsx(
    CustomInput,
    {
      label: t("common.qty"),
      suffix,
      id: "order_total_input",
      name: "order_total_input",
      className: "!oui-rounded-r",
      value: props.total,
      error: getErrorMsg("order_quantity", `${errors?.total?.value} ${quote}`),
      onChange: (val) => {
        setOrderValue("total", val);
      },
      formatters: [inputFormatter.dpFormatter(quote_dp)],
      onFocus: onFocus(5 /* TOTAL */),
      onBlur: onBlur(5 /* TOTAL */)
    }
  );
});
var SkewInput = memo((props) => {
  const { t } = useTranslation();
  const { onFocus, onBlur, getErrorMsg, setOrderValue } = useOrderEntryContext();
  return /* @__PURE__ */ jsx(
    CustomInput,
    {
      id: "order_skew_input",
      label: t("orderEntry.skew"),
      value: props.skew,
      error: getErrorMsg("skew"),
      onChange: (val) => {
        setOrderValue("skew", val);
      },
      onFocus: onFocus(10 /* SKEW */),
      onBlur: onBlur(10 /* SKEW */),
      overrideFormatters: [
        inputFormatter.rangeFormatter({ min: 0, max: 100, dp: 2 }),
        inputFormatter.dpFormatter(2)
      ],
      classNames: {
        root: "oui-rounded-b-xl"
      }
    }
  );
});
var TotalOrdersInput = memo((props) => {
  const { t } = useTranslation();
  const { onFocus, onBlur, getErrorMsg, setOrderValue } = useOrderEntryContext();
  return /* @__PURE__ */ jsx(
    CustomInput,
    {
      label: t("orderEntry.totalOrders"),
      placeholder: "2-20",
      id: "order_total_orders_input",
      className: "!oui-rounded-l",
      value: props.total_orders,
      error: getErrorMsg("total_orders"),
      onChange: (val) => {
        setOrderValue("total_orders", val);
      },
      overrideFormatters: [
        // inputFormatter.rangeFormatter({ min: 2, max: 20 }),
        inputFormatter.numberFormatter,
        inputFormatter.dpFormatter(0)
      ],
      onFocus: onFocus(9 /* TOTAL_ORDERS */),
      onBlur: onBlur(9 /* TOTAL_ORDERS */)
    }
  );
});
var ScaledOrderInput = (props) => {
  const { values } = props;
  const showSkewInput = values.distribution_type === DistributionType.CUSTOM;
  return /* @__PURE__ */ jsxs("div", { className: "oui-space-y-1", children: [
    /* @__PURE__ */ jsx(
      ScaledPriceInput,
      {
        start_price: values.start_price,
        end_price: values.end_price
      }
    ),
    /* @__PURE__ */ jsxs(Grid, { cols: 2, className: "oui-group oui-space-x-1", children: [
      /* @__PURE__ */ jsx(
        ScaledQuantityInput,
        {
          order_quantity: values.order_quantity,
          total: values.total
        }
      ),
      /* @__PURE__ */ jsx(TotalOrdersInput, { total_orders: values.total_orders })
    ] }),
    /* @__PURE__ */ jsx(
      QuantityDistributionInput,
      {
        distribution_type: values.distribution_type,
        className: cn(!showSkewInput && "oui-rounded-b-xl")
      }
    ),
    showSkewInput && /* @__PURE__ */ jsx(SkewInput, { skew: values.skew })
  ] });
};
var TriggerPriceInput = memo((props) => {
  const { t } = useTranslation();
  const {
    symbolInfo,
    onFocus,
    onBlur,
    getErrorMsg,
    setOrderValue,
    triggerPriceInputRef
  } = useOrderEntryContext();
  return /* @__PURE__ */ jsx("div", { className: "oui-group", children: /* @__PURE__ */ jsx(
    CustomInput,
    {
      ref: triggerPriceInputRef,
      id: "order_trigger_price_input",
      name: "order_trigger_price_input",
      label: t("common.trigger"),
      suffix: symbolInfo.quote,
      value: props.trigger_price,
      onChange: (e) => {
        setOrderValue("trigger_price", e);
      },
      error: getErrorMsg("trigger_price"),
      formatters: [inputFormatter.dpFormatter(symbolInfo.quote_dp)],
      onFocus: onFocus(2 /* TRIGGER_PRICE */),
      onBlur: onBlur(2 /* TRIGGER_PRICE */)
    }
  ) });
});
TriggerPriceInput.displayName = "TriggerPriceInput";
var ActivePriceInput = memo((props) => {
  const { t } = useTranslation();
  const {
    symbolInfo,
    onFocus,
    onBlur,
    getErrorMsg,
    setOrderValue,
    activatedPriceInputRef
  } = useOrderEntryContext();
  return /* @__PURE__ */ jsx("div", { className: "oui-group", children: /* @__PURE__ */ jsx(
    CustomInput,
    {
      ref: activatedPriceInputRef,
      id: "order_activated_price_input",
      name: "order_activated_price_input",
      label: t("common.triggerPrice"),
      placeholder: `0 (${t("common.optional")})`,
      suffix: symbolInfo.quote,
      error: getErrorMsg("activated_price"),
      value: props.activated_price,
      onChange: (val) => {
        setOrderValue("activated_price", val);
      },
      formatters: [inputFormatter.dpFormatter(symbolInfo.quote_dp)],
      onFocus: onFocus(11 /* ACTIVATED_PRICE */),
      onBlur: onBlur(11 /* ACTIVATED_PRICE */)
    }
  ) });
});
ActivePriceInput.displayName = "ActivePriceInput";
var percentages = [1, 2, 3, 5];
var CallbackRatePercentages = memo(
  (props) => {
    const { setOrderValue } = useOrderEntryContext();
    return /* @__PURE__ */ jsx(Flex, { gapX: 2, className: props.className, children: percentages.map((item) => {
      const value = item.toString();
      const isActive = props.callback_rate === value;
      return /* @__PURE__ */ jsx(
        Flex,
        {
          justify: "center",
          itemAlign: "center",
          width: 46,
          height: 18,
          className: cn(
            "oui-cursor-pointer oui-select-none",
            "oui-rounded oui-border",
            isActive ? "oui-border-base-contrast-36" : "oui-border-line-12"
          ),
          onClick: () => {
            setOrderValue("callback_rate", value);
            props.onClick?.(value);
          },
          children: /* @__PURE__ */ jsxs(Text, { size: "2xs", intensity: isActive ? 80 : 36, weight: "semibold", children: [
            item,
            "%"
          ] })
        },
        item
      );
    }) });
  }
);
CallbackRatePercentages.displayName = "CallbackRatePercentages";
var TrailingCallbackSelect = memo(
  (props) => {
    const { quote } = props;
    const options2 = useMemo(() => {
      return [
        { label: quote, value: TrailingCallbackType.VALUE },
        { label: "%", value: TrailingCallbackType.RATE }
      ];
    }, [quote]);
    return /* @__PURE__ */ jsx(
      Select.options,
      {
        variant: "text",
        size: "xs",
        options: options2,
        classNames: {
          trigger: cn(
            "oui-absolute oui-right-0 oui-top-1",
            "oui-w-[124px] oui-justify-end",
            "oui-text-base-contrast-36"
          )
        },
        value: props.value,
        onValueChange: props.onValueChange,
        contentProps: {
          align: "end",
          alignOffset: -1,
          sideOffset: -4,
          className: "oui-border oui-border-line-6"
        }
      }
    );
  }
);
TrailingCallbackSelect.displayName = "trailingCallbackSelect";
var TrailingCallbackInput = memo(
  (props) => {
    const { callback_value, callback_rate } = props;
    const { t } = useTranslation();
    const {
      symbolInfo,
      onFocus,
      onBlur,
      getErrorMsg,
      setOrderValue,
      setOrderValues
    } = useOrderEntryContext();
    const { quote, quote_dp } = symbolInfo;
    const lastCallbackValueRef = useRef();
    const lastCallbackRateRef = useRef();
    const [callbackType, setCallbackType] = useLocalStorage(
      "orderly_order_trailing_callback_type",
      TrailingCallbackType.VALUE
    );
    const onCallbackTypeChange = useCallback(
      (type) => {
        setCallbackType(type);
        if (type === TrailingCallbackType.RATE) {
          lastCallbackValueRef.current = callback_value;
          setOrderValues({
            callback_value: "",
            callback_rate: lastCallbackRateRef.current
          });
        } else {
          lastCallbackRateRef.current = callback_rate;
          setOrderValues({
            callback_value: lastCallbackValueRef.current,
            callback_rate: ""
          });
        }
      },
      [callback_value, callback_rate]
    );
    const suffix = /* @__PURE__ */ jsx(
      TrailingCallbackSelect,
      {
        quote,
        value: callbackType,
        onValueChange: onCallbackTypeChange
      }
    );
    if (callbackType === TrailingCallbackType.RATE) {
      return (
        // <Tooltip
        //   content={tooltipContent}
        //   open={open && !!tooltipContent}
        //   sideOffset={-25}
        //   className="!oui-p-3"
        // >
        /* @__PURE__ */ jsxs("div", { className: "oui-relative", children: [
          /* @__PURE__ */ jsx(
            CustomInput,
            {
              id: "order_callback_rate_input",
              name: "order_callback_rate_input",
              label: t("orderEntry.trailingRate"),
              suffix,
              error: getErrorMsg("callback_rate"),
              value: callback_rate,
              onChange: (val) => {
                setOrderValue("callback_rate", val);
              },
              formatters: [inputFormatter.dpFormatter(1)],
              onFocus: (e) => {
                onFocus(13 /* CALLBACK_RATE */)(e);
              },
              onBlur: (e) => {
                onBlur(13 /* CALLBACK_RATE */)(e);
              },
              classNames: {
                root: "oui-h-[68px]",
                input: "oui-mb-5",
                prefix: "!oui-top-1"
              }
            }
          ),
          /* @__PURE__ */ jsx(
            CallbackRatePercentages,
            {
              className: "oui-absolute oui-bottom-1 oui-left-2",
              callback_rate
            }
          )
        ] })
      );
    }
    return /* @__PURE__ */ jsx(
      CustomInput,
      {
        id: "order_callback_value_input",
        name: "order_callback_value_input",
        label: t("orderEntry.trailingValue"),
        suffix,
        error: getErrorMsg("callback_value"),
        value: callback_value,
        onChange: (val) => {
          setOrderValue("callback_value", val);
        },
        formatters: [inputFormatter.dpFormatter(quote_dp)],
        onFocus: onFocus(12 /* CALLBACK_VALUE */),
        onBlur: onBlur(12 /* CALLBACK_VALUE */),
        classNames: {
          input: "!oui-mb-[6px]",
          prefix: "!oui-top-1"
        }
      }
    );
  }
);
TrailingCallbackInput.displayName = "TrailingCallbackInput";
var TrailingStopInput = (props) => {
  const { values } = props;
  return /* @__PURE__ */ jsxs("div", { className: "oui-space-y-1", children: [
    /* @__PURE__ */ jsx(ActivePriceInput, { activated_price: values.activated_price }),
    /* @__PURE__ */ jsx(
      TrailingCallbackInput,
      {
        callback_value: values.callback_value,
        callback_rate: values.callback_rate
      }
    ),
    /* @__PURE__ */ jsx(
      QtyAndTotalInput,
      {
        order_quantity: values.order_quantity,
        total: values.total
      }
    )
  ] });
};
function OrderInput(props) {
  const { values } = props;
  const type = values.order_type;
  if (type === OrderType.SCALED) {
    return /* @__PURE__ */ jsx(ScaledOrderInput, { values });
  }
  if (type === OrderType.TRAILING_STOP) {
    return /* @__PURE__ */ jsx(TrailingStopInput, { values });
  }
  const showTriggerPrice = type === OrderType.STOP_LIMIT || type === OrderType.STOP_MARKET;
  const showPrice = type === OrderType.LIMIT || type === OrderType.STOP_LIMIT;
  const triggerPriceInput = showTriggerPrice && /* @__PURE__ */ jsx(TriggerPriceInput, { trigger_price: values.trigger_price });
  const priceInput = showPrice && /* @__PURE__ */ jsx(
    PriceInput,
    {
      order_type: values.order_type,
      order_price: values.order_price,
      bbo: props.bbo,
      fillMiddleValue: props.fillMiddleValue,
      priceInputContainerWidth: props.priceInputContainerWidth
    }
  );
  return /* @__PURE__ */ jsxs("div", { className: "oui-space-y-1", children: [
    triggerPriceInput,
    priceInput,
    /* @__PURE__ */ jsx(
      QtyAndTotalInput,
      {
        order_quantity: values.order_quantity,
        total: values.total
      }
    )
  ] });
}
var SLIDER_MIN = 0;
var SLIDER_MAX = 100;
var QuantitySlider = memo((props) => {
  const { canTrade, side, order_quantity, maxQty } = props;
  const [sliderValue, setSliderValue] = useState(0);
  const { setOrderValue, symbolInfo, lastQuantityInputType } = useOrderEntryContext();
  const { base_dp, base_tick } = symbolInfo;
  const { t } = useTranslation();
  const color = useMemo(
    () => canTrade ? side === OrderSide.BUY ? "buy" : "sell" : void 0,
    [side, canTrade]
  );
  const maxLabel = useMemo(() => {
    return side === OrderSide.BUY ? t("orderEntry.maxBuy") : t("orderEntry.maxSell");
  }, [side, t]);
  const onSliderValueChange = (value) => {
    lastQuantityInputType.current = 4 /* QUANTITY_SLIDER */;
    setSliderValue(value);
  };
  const sliderToQuantity = (value) => {
    const newQty = new Decimal(value).div(SLIDER_MAX).mul(maxQty).toFixed(base_dp, Decimal.ROUND_DOWN);
    setOrderValue("order_quantity", utils.formatNumber(newQty, base_tick));
  };
  const onMax = () => {
    onSliderValueChange(SLIDER_MAX);
    if (sliderValue === SLIDER_MAX) {
      sliderToQuantity(SLIDER_MAX);
    }
  };
  useEffect(() => {
    if (lastQuantityInputType.current === 4 /* QUANTITY_SLIDER */) {
      sliderToQuantity(sliderValue);
    }
  }, [sliderValue, maxQty]);
  useEffect(() => {
    const quantityToSlider = () => {
      if (order_quantity && Number(order_quantity) !== 0 && maxQty !== 0) {
        return new Decimal(Math.min(Number(order_quantity), maxQty)).div(maxQty).mul(SLIDER_MAX).todp(2, Decimal.ROUND_DOWN).toNumber();
      }
      return 0;
    };
    if (lastQuantityInputType.current !== 4 /* QUANTITY_SLIDER */) {
      setSliderValue(quantityToSlider());
    }
  }, [order_quantity, maxQty]);
  return /* @__PURE__ */ jsxs("div", { children: [
    /* @__PURE__ */ jsx(
      Slider,
      {
        disabled: maxQty === 0 || !canTrade,
        value: [sliderValue],
        color,
        markCount: 4,
        showTip: true,
        onValueChange: (e) => {
          onSliderValueChange(e[0]);
        },
        min: SLIDER_MIN,
        max: SLIDER_MAX
      }
    ),
    /* @__PURE__ */ jsxs(Flex, { justify: "between", className: "oui-pt-1 xl:oui-pt-2", children: [
      /* @__PURE__ */ jsx(
        Text.numeral,
        {
          size: "2xs",
          color,
          dp: 2,
          padding: false,
          suffix: "%",
          children: canTrade ? sliderValue : 0
        }
      ),
      /* @__PURE__ */ jsxs(Flex, { children: [
        /* @__PURE__ */ jsx(
          "button",
          {
            className: textVariants({
              size: "2xs",
              className: "oui-mr-1"
            }),
            onClick: onMax,
            "data-testid": "oui-testid-orderEntry-maxQty-value-button",
            children: maxLabel
          }
        ),
        /* @__PURE__ */ jsx(
          Text.numeral,
          {
            size: "2xs",
            color,
            dp: base_dp,
            padding: false,
            "data-testid": "oui-testid-orderEntry-maxQty-value",
            children: canTrade ? maxQty : 0
          }
        )
      ] })
    ] })
  ] });
});
QuantitySlider.displayName = "QuantitySlider";
var ReduceOnlySwitch = ({
  checked,
  onCheckedChange,
  className,
  testId = "oui-testid-orderEntry-reduceOnly-switch"
}) => {
  const { t } = useTranslation();
  return /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", gapX: 1, className, children: [
    /* @__PURE__ */ jsx(
      Switch,
      {
        "data-testid": testId,
        className: "oui-h-[14px]",
        id: "reduceOnly",
        checked,
        onCheckedChange
      }
    ),
    /* @__PURE__ */ jsx("label", { htmlFor: "reduceOnly", className: "oui-text-xs", children: t("orderEntry.reduceOnly") })
  ] });
};
var PnlInputContext = createContext(
  {}
);
var usePnlInputContext = () => {
  return useContext(PnlInputContext);
};

// src/components/pnlInput/useBuilder.script.ts
var usePNLInputBuilder = (props) => {
  const { type, values, quote_dp } = props;
  const { t } = useTranslation();
  const [focus, setFocus] = useState(true);
  const { mode, setMode, tipsEle } = usePnlInputContext();
  const [tipVisible, setTipVisible] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  const key = useMemo(() => {
    switch (mode) {
      case "Offset" /* OFFSET */:
        return `${type.toLowerCase()}_offset`;
      case "Offset%" /* PERCENTAGE */:
        return `${type.toLowerCase()}_offset_percentage`;
      default:
        return `${type.toLowerCase()}_pnl`;
    }
  }, [mode]);
  const [innerValue, setInnerValue] = useState(
    values[mode]
  );
  useEffect(() => {
    if (isFocused) {
      return;
    }
    setInnerValue(values[mode]);
  }, [values, mode, isFocused]);
  const modes = useMemo(() => {
    return [
      {
        label: t("tpsl.pnl"),
        value: "PnL" /* PnL */,
        testId: `${"PnL" /* PnL */}_menu_item`
      },
      {
        label: t("tpsl.offset"),
        value: "Offset" /* OFFSET */,
        testId: `${"Offset" /* OFFSET */}_mneu_item`
      },
      {
        label: `${t("tpsl.offset")}%`,
        value: "Offset%" /* PERCENTAGE */,
        testId: `${"Offset%" /* PERCENTAGE */}_menu_item`
      }
    ];
  }, [t]);
  const modeLabelMap = useMemo(() => {
    return {
      ["PnL" /* PnL */]: t("tpsl.pnl"),
      ["Offset" /* OFFSET */]: t("tpsl.offset"),
      ["Offset%" /* PERCENTAGE */]: `${t("tpsl.offset")}%`
    };
  }, [t]);
  const percentageSuffix = useRef("");
  const onValueChange = (value) => {
    setInnerValue(value);
    props.onChange(key, value);
  };
  const onFocus = () => {
    setTipVisible(true);
    setIsFocused(true);
  };
  const onBlur = () => {
    setTipVisible(false);
    setIsFocused(false);
    props.onChange(key, innerValue);
  };
  const formatter = (options2) => {
    const { dp = 2 } = options2;
    return {
      onRenderBefore: (value, options3) => {
        value = `${value}`;
        if (focus) {
          if (type === "SL" && mode === "PnL" /* PnL */) {
            value = value.startsWith("-") ? value : "-" + value;
          }
        }
        if (value === "" || value === "-") return "";
        if (mode === "Offset%" /* PERCENTAGE */) {
          return `${new Decimal(
            value.replace(
              new RegExp(percentageSuffix.current.replace(".", "\\.") + "$"),
              ""
            )
          ).mul(100).todp(2, 4).toString()}${percentageSuffix.current}`;
        } else if (mode === "Offset" /* OFFSET */) {
          value = todpIfNeed(value, dp);
        } else ;
        return `${value}`;
      },
      onSendBefore: (value) => {
        if (/^\-?0{2,}$/.test(value)) {
          return "0";
        }
        if (mode === "Offset%" /* PERCENTAGE */) {
          if (value !== "") {
            value = todpIfNeed(value, 2);
            const endStr = value.match(/\.0{0,2}$/);
            if (!!endStr) {
              percentageSuffix.current = endStr[0];
            } else {
              percentageSuffix.current = "";
            }
            value = new Decimal(value).div(100).toString();
            value = `${value}${percentageSuffix.current}`;
          }
        } else if (mode === "PnL" /* PnL */ && type === "SL" && focus) {
          value = value.startsWith("-") ? value : "-" + value;
        } else {
          value = todpIfNeed(value, dp);
        }
        if (value === "" || value === "-") return "";
        return value;
      }
    };
  };
  return {
    mode,
    modes,
    modeLabelMap,
    formatter,
    onModeChange: (mode2) => {
      setMode(mode2);
    },
    onFocus,
    onBlur,
    value: innerValue,
    onValueChange,
    quote_dp,
    tips: tipVisible ? tipsEle : void 0,
    setFocus
  };
};
var PNLInput = (props) => {
  const {
    mode,
    modes,
    modeLabelMap,
    onModeChange,
    onValueChange,
    quote,
    quote_dp,
    value,
    type,
    tips,
    onFocus,
    onBlur,
    setFocus
  } = props;
  const [prefix, setPrefix] = useState(mode);
  const [placeholder, setPlaceholder] = useState(
    mode === "Offset%" /* PERCENTAGE */ ? "%" : quote
  );
  useEffect(() => {
    setPrefix(mode);
    setPlaceholder(mode === "Offset%" /* PERCENTAGE */ ? "%" : quote);
  }, [mode]);
  useEffect(() => {
    setPrefix(!!value ? "" : mode);
  }, [value]);
  const id = useMemo(() => `${type.toLowerCase()}_${mode.toLowerCase()}`, []);
  return /* @__PURE__ */ jsx(
    Input.tooltip,
    {
      prefix: modeLabelMap[prefix] || prefix,
      size: "md",
      placeholder,
      id,
      align: "right",
      value,
      tooltip: tips,
      tooltipProps: {
        content: {
          side: props.type === "TP" ? "top" : "bottom"
        }
      },
      "data-testid": props.testIds?.input,
      autoComplete: "off",
      onValueChange,
      formatters: [
        props.formatter({ dp: quote_dp, mode, type }),
        inputFormatter.currencyFormatter
        // inputFormatter.identifierFormatter(),
      ],
      classNames: {
        root: type === "TP" ? "oui-text-trade-profit" : "oui-text-trade-loss",
        additional: "oui-text-base-contrast-54",
        input: "oui-text-inherit"
      },
      onFocus: () => {
        setPrefix("");
        setPlaceholder("");
        setFocus(true);
        onFocus();
      },
      onBlur: () => {
        setPrefix(!!value ? "" : mode);
        setPlaceholder(mode === "Offset%" /* PERCENTAGE */ ? "%" : quote);
        onBlur();
      },
      suffix: /* @__PURE__ */ jsxs(Fragment, { children: [
        mode === "Offset%" /* PERCENTAGE */ && !!value && /* @__PURE__ */ jsx(Text, { size: "2xs", color: "inherit", className: "oui-ml-[2px]", children: "%" }),
        /* @__PURE__ */ jsx(
          PNLMenus,
          {
            mode,
            modes,
            onModeChange: (item) => onModeChange(item.value),
            testId: props.testIds?.dropDown
          }
        )
      ] })
    }
  );
};
var PNLMenus = (props) => {
  return /* @__PURE__ */ jsx(
    SimpleDropdownMenu,
    {
      currentValue: props.mode,
      menu: props.modes,
      align: "end",
      size: "xs",
      className: "oui-min-w-[80px]",
      onCloseAutoFocus: (event) => event.preventDefault(),
      onSelect: (item) => props.onModeChange(item),
      children: /* @__PURE__ */ jsx("button", { className: "oui-p-2", "data-testid": props.testId, children: /* @__PURE__ */ jsx(CaretDownIcon, { size: 12, color: "white" }) })
    }
  );
};
var PnlInputWidget = (props) => {
  const { testIds, quote, ...rest } = props;
  const state = usePNLInputBuilder(rest);
  return /* @__PURE__ */ jsx(PNLInput, { ...state, testIds, quote, type: props.type });
};
var PnlInputProvider = (props) => {
  const { type, values, children } = props;
  const [mode, setMode] = useLocalStorage(
    "TP/SL_Mode",
    "Offset%" /* PERCENTAGE */
  );
  const { t } = useTranslation();
  const tipsEle = useMemo(() => {
    if (!values.PnL || !values.trigger_price) {
      return null;
    }
    return /* @__PURE__ */ jsxs(Flex, { children: [
      /* @__PURE__ */ jsx("span", { className: "oui-text-xs oui-text-base-contrast-54", children: mode === "PnL" /* PnL */ ? t("orderEntry.estRoi") : t("orderEntry.estPnL") }),
      mode === "PnL" /* PnL */ ? /* @__PURE__ */ jsx(
        Text.numeral,
        {
          rule: "percentages",
          className: cn(
            "oui-ml-1 oui-text-xs",
            type === "TP" ? "oui-text-trade-profit" : "oui-text-trade-loss"
          ),
          children: values.ROI
        }
      ) : /* @__PURE__ */ jsx(
        Text.numeral,
        {
          rule: "price",
          className: cn(
            "oui-ml-1 oui-text-xs",
            type === "TP" ? "oui-text-trade-profit" : "oui-text-trade-loss"
          ),
          children: values.PnL
        }
      )
    ] });
  }, [mode, values.ROI, values.PnL, values.trigger_price]);
  const memoizedValue = useMemo(() => {
    return { mode, setMode, tipsEle };
  }, [mode, setMode, tipsEle]);
  return /* @__PURE__ */ jsx(PnlInputContext.Provider, { value: memoizedValue, children });
};
var OrderTPSL = (props) => {
  const tpslFormRef = React2.useRef(null);
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  useEffect(() => {
    if (props.orderType !== OrderType.LIMIT && props.orderType !== OrderType.MARKET) {
      props.onSwitchChanged(false);
    }
  }, [props.orderType]);
  if (props.orderType !== OrderType.LIMIT && props.orderType !== OrderType.MARKET) {
    return null;
  }
  const isSlPriceWarning = props.errors?.["sl_trigger_price"]?.["type"] === ERROR_MSG_CODES.SL_PRICE_WARNING;
  return /* @__PURE__ */ jsxs("div", { children: [
    /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", justify: "between", children: [
      /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", gapX: 1, children: [
        /* @__PURE__ */ jsx(
          Switch,
          {
            id: "order_entry_tpsl",
            className: "oui-h-[14px]",
            checked: props.switchState,
            disabled: props.orderType !== OrderType.LIMIT && props.orderType !== OrderType.MARKET,
            onCheckedChange: (checked) => {
              props.onSwitchChanged(checked);
            }
          }
        ),
        /* @__PURE__ */ jsx("label", { htmlFor: "order_entry_tpsl", className: "oui-text-xs", children: t("common.tpsl") })
      ] }),
      /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", gapX: 2, children: [
        isMobile && props.onReduceOnlyChange && /* @__PURE__ */ jsx(
          ReduceOnlySwitch,
          {
            checked: props.reduceOnlyChecked ?? false,
            onCheckedChange: props.onReduceOnlyChange
          }
        ),
        !isMobile && /* @__PURE__ */ jsx(
          TPSLAdvancedButton,
          {
            className: cn(
              "oui-group oui-invisible",
              props.switchState && "oui-visible"
            ),
            showTPSLAdvanced: props.showTPSLAdvanced
          }
        )
      ] })
    ] }),
    /* @__PURE__ */ jsx(
      "div",
      {
        className: cn(
          "oui-max-h-0 oui-overflow-hidden oui-transition-all",
          props.switchState && "oui-max-h-[120px]"
        ),
        onTransitionEnd: () => {
          tpslFormRef.current?.style.setProperty(
            "opacity",
            props.switchState ? "1" : "0"
          );
        },
        children: /* @__PURE__ */ jsx(
          TPSLInputForm,
          {
            ref: tpslFormRef,
            setOrderValue: props.setOrderValue,
            onChange: props.onChange,
            values: props.values,
            errors: props.errors,
            quote_dp: props.quote_dp,
            showTPSLAdvanced: props.showTPSLAdvanced,
            isMobile,
            isSlPriceWarning
          }
        )
      }
    ),
    isSlPriceWarning && /* @__PURE__ */ jsx(TPSLPriceWarning, { errors: props.errors })
  ] });
};
var TPSLPriceWarning = (props) => {
  const { getErrorMsg } = useOrderEntryFormErrorMsg(props.errors);
  return /* @__PURE__ */ jsx(
    DotStatus,
    {
      color: "warning",
      size: "xs",
      label: getErrorMsg("sl_trigger_price")
    }
  );
};
var TPSLInputForm = React2.forwardRef((props, ref) => {
  const { getErrorMsg } = useOrderEntryFormErrorMsg(props.errors);
  const { t } = useTranslation();
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref,
      className: "oui-space-y-1 oui-px-px oui-py-2 oui-transition-all",
      children: [
        /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", justify: "between", gapX: 2, children: [
          /* @__PURE__ */ jsx(
            TPSLPositionTypeWidget,
            {
              value: props.values.position_type,
              onChange: props.onChange
            }
          ),
          props.isMobile && /* @__PURE__ */ jsx(
            TPSLAdvancedButton,
            {
              showTPSLAdvanced: props.showTPSLAdvanced,
              isMobile: props.isMobile
            }
          )
        ] }),
        /* @__PURE__ */ jsx(PnlInputProvider, { values: props.values.tp, type: "TP", children: /* @__PURE__ */ jsx(
          TPSLInputRow,
          {
            type: "TP",
            error: getErrorMsg("tp_trigger_price"),
            onChange: props.onChange,
            values: props.values.tp,
            quote_dp: props.quote_dp,
            testIds: {
              first: "oui-testid-orderEntry-tpsl-tpPrice-input",
              second: "oui-testid-orderEntry-tpsl-tpPnl-input",
              dropDown: "oui-testid-orderEntry-tpsl-tp-dropDown-trigger-button"
            }
          }
        ) }),
        /* @__PURE__ */ jsx(PnlInputProvider, { values: props.values.sl, type: "SL", children: /* @__PURE__ */ jsx(
          TPSLInputRow,
          {
            type: "SL",
            error: props.isSlPriceWarning ? void 0 : getErrorMsg("sl_trigger_price"),
            displayErrorMessage: getErrorMsg("sl_trigger_price") !== void 0,
            onChange: props.onChange,
            values: props.values.sl,
            quote_dp: props.quote_dp,
            testIds: {
              first: "oui-testid-orderEntry-tpsl-slPrice-input",
              second: "oui-testid-orderEntry-tpsl-slPnl-input",
              dropDown: "oui-testid-orderEntry-tpsl-sl-dropDown-trigger-button"
            },
            classNames: {
              root: props.isSlPriceWarning ? "oui-outline-warning-darken focus-within:oui-outline-warning-darken" : void 0
            }
          }
        ) })
      ]
    }
  );
});
TPSLInputForm.displayName = "TPSLInputForm";
var TPSLAdvancedButton = (props) => {
  const { t } = useTranslation();
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      itemAlign: "center",
      gapX: 1,
      onClick: props.showTPSLAdvanced,
      className: cn("oui-group oui-cursor-pointer", props.className),
      children: [
        /* @__PURE__ */ jsx(
          Text,
          {
            className: cn(
              "oui-cursor-pointer group-hover:oui-text-base-contrast",
              props.isMobile ? "oui-text-2xs" : "oui-text-sm"
            ),
            children: t("tpsl.advanced")
          }
        ),
        /* @__PURE__ */ jsx(
          SettingFillIcon,
          {
            size: 12,
            className: "oui-text-base-contrast-54 group-hover:oui-text-base-contrast oui-cursor-pointer",
            opacity: 1,
            onClick: props.showTPSLAdvanced
          }
        )
      ]
    }
  );
};
var TPSLTriggerPriceInput = (props) => {
  const { t } = useTranslation();
  const { errorMsgVisible } = useOrderEntryContext();
  const { tipsEle } = usePnlInputContext();
  const [prefix, setPrefix] = useState(`${props.type} Price`);
  const [placeholder, setPlaceholder] = useState("USDC");
  const [tipVisible, setTipVisible] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  const [innerValue, setInnerValue] = useState(
    props.values.trigger_price ?? ""
  );
  useEffect(() => {
    if (isFocused) {
      return;
    }
    setInnerValue(props.values.trigger_price ?? "");
  }, [props.values.trigger_price, isFocused]);
  const triggerPriceToolTipEle = useMemo(() => {
    if (props.error && (errorMsgVisible || props.displayErrorMessage))
      return props.error;
    if (tipVisible) return tipsEle;
    return null;
  }, [
    props.error,
    errorMsgVisible,
    tipVisible,
    tipsEle,
    props.displayErrorMessage
  ]);
  const getPrefixLabel = (trigger_price) => {
    let _prefix = props.type === "TP" ? t("tpsl.tpPrice") : t("tpsl.slPrice");
    if (trigger_price) {
      _prefix = props.type === "TP" ? t("tpsl.tp") : t("tpsl.sl");
    }
    return _prefix;
  };
  const onValueChange = (value) => {
    setInnerValue(value);
    props.onChange(value);
  };
  useEffect(() => {
    setPrefix(getPrefixLabel(props.values.trigger_price));
    if (!isFocused) {
      setInnerValue(props.values.trigger_price ?? "");
    }
  }, [props.type, props.values.trigger_price]);
  const onFocus = () => {
    setPrefix(props.type === "TP" ? t("tpsl.tp") : t("tpsl.sl"));
    setPlaceholder("");
    setTipVisible(true);
    setIsFocused(true);
  };
  const onBlur = () => {
    setPrefix(getPrefixLabel(props.values.trigger_price));
    setPlaceholder("USDC");
    setTipVisible(false);
    setIsFocused(false);
    props.onChange(innerValue);
  };
  return /* @__PURE__ */ jsx(
    Input.tooltip,
    {
      "data-testid": props.testId,
      prefix,
      size: "md",
      placeholder,
      align: "right",
      onFocus,
      onBlur,
      tooltip: triggerPriceToolTipEle,
      tooltipProps: {
        content: {
          side: props.type === "TP" ? "top" : "bottom"
        }
      },
      color: props.error ? "danger" : void 0,
      autoComplete: "off",
      value: innerValue,
      classNames: {
        additional: "oui-text-base-contrast-54",
        root: cn("oui-pr-2 md:oui-pr-3", props.classNames?.root),
        prefix: cn("oui-pr-1 md:oui-pr-2", props.classNames?.prefix),
        input: cn(
          "oui-text-2xs placeholder:oui-text-2xs",
          props.classNames?.input
        )
      },
      onValueChange,
      formatters: [
        inputFormatter.numberFormatter,
        inputFormatter.dpFormatter(props.quote_dp ?? 2),
        inputFormatter.currencyFormatter
      ]
    }
  );
};
var TPSLInputRow = (props) => {
  return /* @__PURE__ */ jsxs(Grid, { cols: 2, gapX: 1, children: [
    /* @__PURE__ */ jsx(
      TPSLTriggerPriceInput,
      {
        testId: props.testIds?.first,
        type: props.type,
        error: props.error,
        displayErrorMessage: props.displayErrorMessage,
        values: props.values ?? "",
        classNames: props.classNames,
        onChange: (event) => {
          props.onChange(
            props.type === "SL" ? "sl_trigger_price" : "tp_trigger_price",
            event
          );
        },
        quote_dp: props.quote_dp
      }
    ),
    /* @__PURE__ */ jsx(
      PnlInputWidget,
      {
        testIds: {
          input: props.testIds?.second,
          dropDown: props.testIds?.dropDown
        },
        onChange: props.onChange,
        quote: "USDC",
        quote_dp: 2,
        type: props.type,
        values: {
          PnL: props.values.PnL,
          Offset: props.values.Offset,
          "Offset%": props.values["Offset%"],
          ROI: props.values.ROI
        }
      }
    )
  ] });
};
var OrderEntry = (props) => {
  const {
    side,
    formattedOrder,
    setOrderValue,
    setOrderValues,
    symbolInfo,
    maxQty,
    freeCollateral,
    helper,
    submit,
    metaState,
    bboStatus,
    bboType,
    onBBOChange,
    toggleBBO,
    disableFeatures,
    currentLtv,
    fillMiddleValue,
    soundAlert,
    setSoundAlert,
    currentFocusInput
  } = props;
  const [maxQtyConfirmOpen, setMaxQtyConfirmOpen] = useState(false);
  const { t } = useTranslation();
  const { isMobile } = useScreen();
  const [hasAdvancedTPSLResult, setHasAdvancedTPSLResult] = useState(false);
  const { errors, validated } = metaState;
  const [errorMsgVisible, setErrorMsgVisible] = useState(false);
  const [needConfirm, setNeedConfirm] = useLocalStorage(
    "orderly_order_confirm",
    true
  );
  const [pinned, setPinned] = useLocalStorage(
    "orderly-order-additional-pinned",
    true
  );
  const [showTPSLAdvanced, setShowTPSLAdvanced] = useState(false);
  const [hidden, setHidden] = useLocalStorage("orderly-order-hidden", false);
  const [slippage, setSlippage] = useLocalStorage("orderly-slippage", "1", {
    parseJSON: ((value) => {
      return !value || value === '""' ? "1" : JSON.parse(value);
    })
  });
  const { notification } = useOrderlyContext();
  const soundAlertId = useId();
  const { getErrorMsg } = useOrderEntryFormErrorMsg(validated ? errors : null);
  const buttonLabel = useMemo(() => {
    return side === OrderSide.BUY ? t("orderEntry.buyLong") : t("orderEntry.sellShort");
  }, [side, t, isMobile]);
  useEffect(() => {
    if (validated) {
      setErrorMsgVisible(true);
    }
  }, [validated]);
  useEffect(() => {
    if (disableFeatures?.includes("slippageSetting")) {
      return;
    }
    if (slippage) {
      setOrderValue("slippage", Number(slippage));
    } else {
      setOrderValue("slippage", void 0);
    }
  }, [slippage, disableFeatures]);
  useEffect(() => {
    const clickHandler = (event) => {
      const target = event.target;
      if (target.closest("#order-entry-submit-button")) {
        return;
      }
      setErrorMsgVisible((visible) => {
        if (visible) {
          return false;
        }
        return visible;
      });
    };
    if (errorMsgVisible) {
      document.addEventListener("click", clickHandler);
    } else {
      document.removeEventListener("click", clickHandler);
    }
    return () => {
      document.removeEventListener("click", clickHandler);
    };
  }, [errorMsgVisible]);
  const onSubmit = useMemoizedFn(async () => {
    const isScaledOrder = formattedOrder.order_type === OrderType.SCALED;
    const isSlPriceError = props.slPriceError?.sl_trigger_price?.type === ERROR_MSG_CODES.SL_PRICE_ERROR;
    helper.validate(isSlPriceError ? props.slPriceError : void 0).then(
      // validate success, it return the order
      // TODO: get order from other function
      (order) => {
        if (isScaledOrder) {
          return modal.show(scaledOrderConfirmDialogId, {
            order,
            symbolInfo,
            size: isMobile ? "sm" : "md"
          });
        }
        if (needConfirm) {
          return modal.show(orderConfirmDialogId, {
            order: formattedOrder,
            symbolInfo
          });
        }
        return true;
      },
      // should catch validate error first, then submit
      (errors2) => {
        if (errors2.slippage) {
          toast.error(t("orderEntry.slippage.error.max"));
        }
        setErrorMsgVisible(true);
        return Promise.reject();
      }
    ).then(() => {
      return submit({ resetOnSuccess: false }).then((result) => {
        if (!result.success && result.message) {
          toast.error(result.message);
        } else if (result.success && isScaledOrder) {
          const message = getScaledPlaceOrderMessage(result);
          if (message) {
            toast.success(message);
          }
        }
      });
    }).catch((error) => {
      if (error?.message) {
        toast.error(error.message);
      }
    });
  });
  const formattedMaxQty = useMemo(() => {
    return new Decimal(maxQty).todp(symbolInfo.base_dp, Decimal.ROUND_DOWN).toString();
  }, [maxQty, symbolInfo.base_dp]);
  const onMaxQtyConfirm = useCallback(() => {
    setOrderValue("order_quantity", formattedMaxQty);
    requestAnimationFrame(() => {
      onSubmit();
    });
    setMaxQtyConfirmOpen(false);
  }, [setOrderValue, formattedMaxQty]);
  const validateSubmit = async () => {
    if (formattedOrder.reduce_only && maxQty === 0) {
      return modal.confirm({
        title: t("orderEntry.reduceOnly.reminder"),
        content: t("orderEntry.reduceOnly.reminder.content"),
        okLabel: t("orderEntry.placeOrderNow"),
        onOk: async () => {
          setOrderValue("reduce_only", false);
          requestAnimationFrame(() => {
            props.resetMetaState();
            onSubmit();
          });
          return Promise.resolve(true);
        },
        onCancel: async () => {
          return Promise.resolve(false);
        }
      });
    } else if (maxQty > 0 && Number(formattedOrder.order_quantity) > maxQty) {
      setMaxQtyConfirmOpen(true);
    } else {
      onSubmit();
    }
  };
  const onShowTPSLAdvanced = () => {
    helper.validate().then(
      () => {
        setShowTPSLAdvanced(true);
      },
      (errors2) => {
        const tpslKey = /* @__PURE__ */ new Set(["tp_trigger_price", "sl_trigger_price"]);
        if (Object.keys(errors2).every((key) => tpslKey.has(key))) {
          setShowTPSLAdvanced(true);
        }
      }
    );
  };
  const onSubmitAdvancedTPSL = (order) => {
    if (order.side !== formattedOrder.side) {
      setOrderValue("side", order.side);
    }
    setOrderValues({
      position_type: order.position_type,
      tp_order_type: order.tp_order_type,
      tp_pnl: order.tp_pnl,
      tp_offset: order.tp_offset,
      tp_offset_percentage: order.tp_offset_percentage,
      tp_ROI: order.tp_ROI,
      tp_trigger_price: order.tp_trigger_price,
      tp_order_price: order.tp_order_price,
      sl_order_type: order.sl_order_type,
      sl_trigger_price: order.sl_trigger_price,
      sl_order_price: order.sl_order_price,
      sl_pnl: order.sl_pnl,
      sl_offset: order.sl_offset,
      sl_offset_percentage: order.sl_offset_percentage,
      sl_ROI: order.sl_ROI
    });
    setShowTPSLAdvanced(false);
    setHasAdvancedTPSLResult(true);
  };
  const onDeleteAdvancedTPSL = () => {
    setHasAdvancedTPSLResult(false);
    setOrderValues({
      tp_trigger_price: void 0,
      tp_order_price: void 0,
      tp_order_type: OrderType.MARKET,
      sl_trigger_price: void 0,
      sl_order_price: void 0,
      sl_order_type: OrderType.MARKET,
      tp_pnl: void 0,
      sl_pnl: void 0,
      position_type: PositionType.FULL
    });
  };
  useEffect(() => {
    setHasAdvancedTPSLResult(false);
  }, [props.symbol]);
  const showReduceOnlySection = isMobile && formattedOrder.order_type !== OrderType.LIMIT && formattedOrder.order_type !== OrderType.MARKET || !isMobile;
  const showSoundSection = Boolean(notification?.orderFilled?.media) && (notification?.orderFilled?.displayInOrderEntry ?? true);
  const additionalInfoProps = {
    pinned,
    setPinned,
    needConfirm,
    setNeedConfirm,
    hidden,
    setHidden,
    onValueChange: setOrderValue,
    orderTypeExtra: formattedOrder["order_type_ext"],
    showExtra: formattedOrder["order_type"] === OrderType.LIMIT && !props.tpslSwitch
  };
  const extraButton = !pinned && /* @__PURE__ */ jsx(AdditionalConfigButton, { ...additionalInfoProps });
  return /* @__PURE__ */ jsxs(
    OrderEntryProvider,
    {
      errors,
      errorMsgVisible,
      symbolInfo,
      onFocus: props.onFocus,
      onBlur: props.onBlur,
      getErrorMsg,
      setOrderValue,
      setOrderValues,
      currentFocusInput: currentFocusInput.current,
      priceInputRef: props.priceInputRef,
      priceInputContainerRef: props.priceInputContainerRef,
      triggerPriceInputRef: props.triggerPriceInputRef,
      activatedPriceInputRef: props.activatedPriceInputRef,
      lastQuantityInputType: props.lastQuantityInputType,
      leverage: props.symbolLeverage,
      children: [
        /* @__PURE__ */ jsx(
          MaxQtyConfirm,
          {
            open: maxQtyConfirmOpen,
            onOpenChange: setMaxQtyConfirmOpen,
            maxQty: formattedMaxQty,
            onConfirm: onMaxQtyConfirm,
            base: symbolInfo.base
          }
        ),
        /* @__PURE__ */ jsxs(
          "div",
          {
            className: "oui-space-y-2 oui-text-base-contrast-54 xl:oui-space-y-3",
            ref: props.containerRef,
            children: [
              /* @__PURE__ */ jsx(
                OrderEntryHeader,
                {
                  symbol: props.symbol,
                  canTrade: props.canTrade,
                  side,
                  order_type: formattedOrder.order_type,
                  setOrderValue,
                  symbolLeverage: props.symbolLeverage
                }
              ),
              /* @__PURE__ */ jsx(
                Available,
                {
                  currentLtv,
                  canTrade: props.canTrade,
                  quote: symbolInfo?.quote,
                  freeCollateral
                }
              ),
              /* @__PURE__ */ jsx(
                OrderInput,
                {
                  values: formattedOrder,
                  priceInputContainerWidth: props.priceInputContainerWidth,
                  fillMiddleValue,
                  bbo: {
                    bboStatus,
                    bboType,
                    onBBOChange,
                    toggleBBO
                  }
                }
              ),
              /* @__PURE__ */ jsx(
                QuantitySlider,
                {
                  canTrade: props.canTrade,
                  side: props.side,
                  order_quantity: formattedOrder.order_quantity,
                  maxQty
                }
              ),
              /* @__PURE__ */ jsx(
                ThrottledButton,
                {
                  fullWidth: true,
                  id: "order-entry-submit-button",
                  "data-type": OrderSide.BUY,
                  className: cn(
                    side === OrderSide.BUY ? "orderly-order-entry-submit-button-buy oui-bg-success-darken hover:oui-bg-success-darken/80 active:oui-bg-success-darken/80" : "orderly-order-entry-submit-button-sell oui-bg-danger-darken hover:oui-bg-danger-darken/80 active:oui-bg-danger-darken/80"
                  ),
                  onClick: validateSubmit,
                  loading: props.isMutating,
                  disabled: !props.canTrade,
                  children: buttonLabel
                }
              ),
              /* @__PURE__ */ jsx(
                AssetInfo,
                {
                  canTrade: props.canTrade,
                  quote: symbolInfo.quote,
                  estLiqPrice: props.estLiqPrice,
                  estLiqPriceDistance: props.estLiqPriceDistance,
                  estLeverage: props.estLeverage,
                  currentLeverage: props.currentLeverage,
                  slippage,
                  dp: symbolInfo.quote_dp,
                  setSlippage,
                  estSlippage: props.estSlippage,
                  orderType: formattedOrder.order_type,
                  disableFeatures,
                  symbol: props.symbol,
                  side
                }
              ),
              /* @__PURE__ */ jsx(Divider, { className: "oui-w-full" }),
              hasAdvancedTPSLResult ? /* @__PURE__ */ jsx(
                AdvancedTPSLResult,
                {
                  order: formattedOrder,
                  symbolInfo: props.symbolInfo,
                  errors: validated ? errors : null,
                  onEdit: () => {
                    setShowTPSLAdvanced(true);
                  },
                  onDelete: () => {
                    onDeleteAdvancedTPSL();
                  }
                }
              ) : /* @__PURE__ */ jsx(
                OrderTPSL,
                {
                  quote_dp: props.symbolInfo.quote_dp,
                  switchState: props.tpslSwitch,
                  onSwitchChanged: props.setTpslSwitch,
                  orderType: formattedOrder.order_type,
                  errors: validated || props.slPriceError !== void 0 ? { ...errors, ...props.slPriceError } : null,
                  setOrderValue,
                  reduceOnlyChecked: formattedOrder.reduce_only ?? false,
                  onReduceOnlyChange: (checked) => {
                    setOrderValue("reduce_only", checked);
                  },
                  values: {
                    position_type: formattedOrder.position_type ?? PositionType.PARTIAL,
                    tp: {
                      trigger_price: formattedOrder.tp_trigger_price ?? "",
                      PnL: formattedOrder.tp_pnl ?? "",
                      Offset: formattedOrder.tp_offset ?? "",
                      "Offset%": formattedOrder.tp_offset_percentage ?? "",
                      ROI: formattedOrder.tp_ROI ?? ""
                    },
                    sl: {
                      trigger_price: formattedOrder.sl_trigger_price ?? "",
                      PnL: formattedOrder.sl_pnl ?? "",
                      Offset: formattedOrder.sl_offset ?? "",
                      "Offset%": formattedOrder.sl_offset_percentage ?? "",
                      ROI: formattedOrder.sl_ROI ?? ""
                    }
                  },
                  showTPSLAdvanced: onShowTPSLAdvanced,
                  onChange: (key, value) => {
                    setOrderValue(key, value);
                  }
                }
              ),
              showReduceOnlySection && /* @__PURE__ */ jsxs(Flex, { justify: "between", itemAlign: "center", className: "oui-mt-2", children: [
                /* @__PURE__ */ jsx(
                  ReduceOnlySwitch,
                  {
                    checked: formattedOrder.reduce_only ?? false,
                    onCheckedChange: (checked) => {
                      setOrderValue("reduce_only", checked);
                    }
                  }
                ),
                !showSoundSection && extraButton
              ] }),
              showSoundSection && /* @__PURE__ */ jsxs(Flex, { justify: "between", itemAlign: "center", children: [
                /* @__PURE__ */ jsxs(Flex, { itemAlign: "center", gapX: 1, children: [
                  /* @__PURE__ */ jsx(
                    Switch,
                    {
                      className: "oui-h-[14px]",
                      id: soundAlertId,
                      checked: soundAlert,
                      onCheckedChange: (checked) => setSoundAlert(checked)
                    }
                  ),
                  /* @__PURE__ */ jsx("label", { htmlFor: soundAlertId, className: "oui-text-xs", children: t("portfolio.setting.soundAlerts") })
                ] }),
                extraButton
              ] }),
              !showSoundSection && isMobile && (formattedOrder.order_type == OrderType.LIMIT || formattedOrder.order_type == OrderType.MARKET) && !formattedOrder.reduce_only && !pinned && /* @__PURE__ */ jsx(Flex, { className: "oui-w-full", justify: "end", children: extraButton }),
              pinned && /* @__PURE__ */ jsxs(Box, { p: 2, r: "md", intensity: 700, position: "relative", children: [
                /* @__PURE__ */ jsx(AdditionalInfo, { ...additionalInfoProps }),
                /* @__PURE__ */ jsx(
                  PinButton,
                  {
                    onClick: () => {
                      setPinned(false);
                    },
                    className: "oui-group oui-absolute oui-right-2 oui-top-2",
                    "data-testid": "oui-testid-orderEntry-pinned-button"
                  }
                )
              ] })
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          SimpleSheet,
          {
            open: showTPSLAdvanced,
            onOpenChange: setShowTPSLAdvanced,
            classNames: {
              body: "oui-h-full oui-pb-0 oui-border-none",
              overlay: "!oui-bg-base-10/60",
              content: cn(
                "oui-rounded-[16px] oui-border-none !oui-p-0",
                isMobile ? "oui-inset-y-0 oui-right-0 oui-w-[280px]" : "!oui-bottom-[40px] oui-right-3 oui-top-[44px] !oui-h-auto oui-w-[360px]"
              )
            },
            contentProps: { side: "right", closeable: false },
            children: /* @__PURE__ */ jsx(
              TPSLAdvancedWidget,
              {
                setOrderValue,
                order: formattedOrder,
                onSubmit: onSubmitAdvancedTPSL,
                onClose: () => {
                  setShowTPSLAdvanced(false);
                },
                symbolLeverage: props.symbolLeverage
              }
            )
          }
        )
      ]
    }
  );
};
function useBBOState({
  tpslSwitch,
  order_type,
  order_type_ext,
  side,
  setOrderValues
}) {
  const [localBBOType, setLocalBBOType] = useLocalStorage("orderly_order_bbo_type", void 0);
  const lastBBOType = useRef(localBBOType);
  const { track } = useTrack();
  const bboStatus = useMemo(() => {
    if (tpslSwitch || [OrderType.POST_ONLY, OrderType.IOC, OrderType.FOK].includes(
      order_type_ext
    )) {
      return "disabled" /* DISABLED */;
    }
    return localBBOType && order_type === OrderType.LIMIT ? "on" /* ON */ : "off" /* OFF */;
  }, [tpslSwitch, order_type_ext, order_type, localBBOType]);
  const toggleBBO = () => {
    track(TrackerEventName.clickBBOButton);
    if (localBBOType) {
      setLocalBBOType(void 0);
      setOrderValues({
        order_type_ext: void 0,
        level: void 0
      });
    } else {
      setLocalBBOType(lastBBOType.current || BBOOrderType.COUNTERPARTY1);
    }
  };
  const onBBOChange = (value) => {
    setLocalBBOType(value);
    lastBBOType.current = value;
  };
  useEffect(() => {
    if (bboStatus === "disabled" /* DISABLED */) {
      setOrderValues({
        // if order_type_ext is not bbo(ask, bid), keep previous value
        order_type_ext: isBBOOrder({ order_type_ext }) ? void 0 : order_type_ext,
        level: void 0
      });
    }
  }, [bboStatus, order_type_ext]);
  useEffect(() => {
    if (bboStatus === "on" /* ON */) {
      const orderType = getOrderTypeByBBO(localBBOType, side);
      const orderLevel = getOrderLevelByBBO(localBBOType);
      setOrderValues({
        order_type_ext: orderType,
        level: orderLevel
      });
    }
  }, [localBBOType, bboStatus, side]);
  return {
    bboStatus,
    bboType: localBBOType,
    setBBOType: setLocalBBOType,
    onBBOChange,
    toggleBBO
  };
}
function useFocusAndBlur(props) {
  const { base_tick, order_type, order_quantity, setValue } = props;
  const currentFocusInput = useRef(0 /* NONE */);
  const lastScaledOrderPriceInput = useRef(8 /* END_PRICE */);
  const lastQuantityInputType = useRef(0 /* NONE */);
  const formatQty = () => {
    if (base_tick < 1 || // scaled order should not format quantity, because it is total quantity
    order_type === OrderType.SCALED || !order_quantity) {
      return;
    }
    const quantity = utils.formatNumber(
      order_quantity,
      new Decimal(base_tick || "0").toNumber()
    );
    setValue("order_quantity", quantity, {
      shouldUpdateLastChangedField: false
    });
  };
  const onFocus = (type) => (_) => {
    currentFocusInput.current = type;
    if ([7 /* START_PRICE */, 8 /* END_PRICE */].includes(
      currentFocusInput.current
    )) {
      lastScaledOrderPriceInput.current = type;
    }
    if ([3 /* QUANTITY */, 5 /* TOTAL */, 6 /* MARGIN */].includes(type)) {
      lastQuantityInputType.current = type;
    }
  };
  const onBlur = (type) => (_) => {
    setTimeout(() => {
      if (currentFocusInput.current !== type) {
        return;
      }
      currentFocusInput.current = 0 /* NONE */;
    }, 300);
    if (type === 3 /* QUANTITY */) {
      formatQty();
    }
  };
  return {
    currentFocusInput,
    lastScaledOrderPriceInput,
    lastQuantityInputType,
    onFocus: useMemoizedFn(onFocus),
    onBlur: useMemoizedFn(onBlur)
  };
}
function usePriceInputContainer({
  order_type_ext
}) {
  const [priceInputContainerWidth, setPriceInputContainerWidth] = useState(0);
  const priceInputContainerRef = useRef(null);
  useEffect(() => {
    const element = priceInputContainerRef.current;
    if (!element) {
      return;
    }
    const resizeObserver = new ResizeObserver((entries) => {
      for (const entry of entries) {
        const width = entry.contentRect.width;
        if (width) {
          setPriceInputContainerWidth(width);
        }
      }
    });
    resizeObserver.observe(element);
    return () => {
      resizeObserver.unobserve(element);
    };
  }, [priceInputContainerRef, order_type_ext]);
  return { priceInputContainerRef, priceInputContainerWidth };
}

// src/orderEntry.script.ts
var ORDERLY_ORDER_SOUND_ALERT_KEY = "orderly_order_sound_alert";
var useOrderEntryScript = (inputs) => {
  const { symbol } = inputs;
  const [localOrderType, setLocalOrderType] = useLocalStorage(
    "orderly-order-entry-order-type",
    OrderType.LIMIT
  );
  const [localOrderSide, setLocalOrderSide] = useLocalStorage(
    "orderly-order-entry-order-side",
    OrderSide.BUY
  );
  const { notification } = useOrderlyContext();
  const [soundAlert, setSoundAlert] = useLocalStorage(
    ORDERLY_ORDER_SOUND_ALERT_KEY,
    notification?.orderFilled?.defaultOpen ?? false
  );
  const canTrade = useCanTrade();
  const {
    formattedOrder,
    setValue,
    setValues: setOrderValues,
    symbolInfo,
    ...state
  } = useOrderEntry(symbol, {
    initialOrder: {
      symbol,
      order_type: localOrderType,
      position_type: PositionType.PARTIAL,
      side: localOrderSide
    }
  });
  const [tpslSwitch, setTpslSwitch] = useLocalStorage(
    "orderly-order-entry-tp_sl-switch",
    false
  );
  const { currentLeverage } = useMarginRatio();
  const ee = useEventEmitter();
  const triggerPriceInputRef = useRef(null);
  const priceInputRef = useRef(null);
  const activatedPriceInputRef = useRef(null);
  const { bboStatus, bboType, setBBOType, onBBOChange, toggleBBO } = useBBOState({
    tpslSwitch,
    order_type: formattedOrder.order_type,
    order_type_ext: formattedOrder.order_type_ext,
    side: formattedOrder.side,
    setOrderValues
  });
  const {
    currentFocusInput,
    lastScaledOrderPriceInput,
    lastQuantityInputType,
    onFocus,
    onBlur
  } = useFocusAndBlur({
    base_tick: symbolInfo?.base_tick,
    order_type: formattedOrder.order_type,
    order_quantity: formattedOrder.order_quantity,
    setValue
  });
  const cancelTP_SL = () => {
    setOrderValues({
      tp_trigger_price: "",
      sl_trigger_price: "",
      position_type: PositionType.FULL
    });
  };
  const enableTP_SL = () => {
    setOrderValues({
      order_type_ext: void 0,
      position_type: PositionType.FULL
    });
  };
  const setOrderValue = useMemoizedFn(
    (key, value, options2) => {
      if (key === "order_type") {
        setLocalOrderType(value);
      }
      if (key === "side") {
        setLocalOrderSide(value);
      }
      if (key === "reduce_only" && value || key === "order_type" && (value === OrderType.STOP_LIMIT || value === OrderType.STOP_MARKET)) {
        const data = {
          tp_trigger_price: "",
          sl_trigger_price: "",
          [key]: value
        };
        if (key === "order_type") {
          data["order_type_ext"] = "";
        }
        setOrderValues(data);
        return;
      }
      if (key === "order_type" && value !== OrderType.LIMIT) {
        const data = {
          level: void 0,
          order_type_ext: void 0,
          [key]: value
        };
        setOrderValues(data);
        return;
      }
      if (key === "order_type" && value === OrderType.SCALED) {
        setOrderValues({
          distribution_type: DistributionType.FLAT,
          [key]: value
        });
        return;
      }
      setValue(key, value, options2);
    }
  );
  const onTPSLSwitchChanged = (state2) => {
    setTpslSwitch(state2);
    if (state2) {
      enableTP_SL();
    } else {
      cancelTP_SL();
    }
  };
  useEffect(() => {
    const updateOrderPrice = (price) => {
      setValue("order_price", price);
    };
    ee.on("update:orderPrice", updateOrderPrice);
    return () => {
      ee.off("update:orderPrice", updateOrderPrice);
    };
  }, []);
  useEffect(() => {
    const focusInputElement = (target) => {
      requestAnimationFrame(() => {
        target?.focus();
      });
    };
    const orderBookItemClickHandler = (item) => {
      const price = removeTrailingZeros(item[0]);
      const { order_type, order_type_ext } = formattedOrder;
      if (currentFocusInput.current === 2 /* TRIGGER_PRICE */ && (order_type === OrderType.STOP_LIMIT || order_type === OrderType.STOP_MARKET)) {
        setValue("trigger_price", price);
        focusInputElement(triggerPriceInputRef.current);
        return;
      }
      if (isBBOOrder({ order_type, order_type_ext })) {
        setBBOType(void 0);
        setOrderValues({
          order_type_ext: void 0,
          level: void 0
        });
        requestAnimationFrame(() => {
          ee.emit("update:orderPrice", price);
        });
        focusInputElement(priceInputRef.current);
        return;
      }
      if (order_type === OrderType.STOP_LIMIT || order_type === OrderType.LIMIT) {
        setValue("order_price", price);
        focusInputElement(priceInputRef.current);
        return;
      }
      if (order_type === OrderType.STOP_MARKET) {
        setValue("trigger_price", price);
        focusInputElement(triggerPriceInputRef.current);
        return;
      }
      if (order_type === OrderType.MARKET) {
        setBBOType(void 0);
        setOrderValues({
          order_type: OrderType.LIMIT,
          order_price: price
        });
        focusInputElement(priceInputRef.current);
        return;
      }
      if (order_type === OrderType.SCALED && lastScaledOrderPriceInput.current) {
        const field = lastScaledOrderPriceInput.current === 7 /* START_PRICE */ ? "start_price" : "end_price";
        setValue(field, price);
        focusInputElement(priceInputRef.current);
        return;
      }
      if (order_type === OrderType.TRAILING_STOP) {
        setValue("activated_price", price);
        focusInputElement(activatedPriceInputRef.current);
        return;
      }
      setValue("order_price", price);
      focusInputElement(priceInputRef.current);
    };
    ee.on("orderbook:item:click", orderBookItemClickHandler);
    return () => {
      ee.off("orderbook:item:click", orderBookItemClickHandler);
    };
  }, [formattedOrder, symbolInfo]);
  useEffect(() => {
    state.reset();
    state.resetMetaState();
    lastQuantityInputType.current = 0 /* NONE */;
  }, [symbol]);
  useEffect(() => {
    if (formattedOrder.order_type === OrderType.SCALED && !formattedOrder.distribution_type) {
      setValue("distribution_type", DistributionType.FLAT);
    }
  }, [formattedOrder.order_type, formattedOrder.distribution_type]);
  const currentLtv = useComputedLTV();
  const askAndBid = useAskAndBid();
  const fillMiddleValue = () => {
    if (bboStatus === "on" /* ON */) {
      toggleBBO();
    }
    if (formattedOrder.order_type === OrderType.LIMIT) {
      const [bestAsk = 0, bestBid = 0] = askAndBid;
      const midPrice = new Decimal(safeNumber(bestAsk)).add(safeNumber(bestBid)).div(2).toNumber();
      requestAnimationFrame(() => {
        ee.emit("update:orderPrice", midPrice);
      });
    }
  };
  const { priceInputContainerRef, priceInputContainerWidth } = usePriceInputContainer({
    order_type_ext: formattedOrder.order_type_ext
  });
  const slPriceError = useTpslPriceChecker({
    slPrice: formattedOrder.sl_trigger_price,
    liqPrice: state.estLiqPrice,
    side: formattedOrder.side,
    currentPosition: state.currentPosition,
    orderQuantity: Number(formattedOrder.order_quantity)
  });
  useEffect(() => {
    if (formattedOrder.reduce_only) {
      setTpslSwitch(false);
    }
  }, [formattedOrder.reduce_only]);
  useEffect(() => {
    if (tpslSwitch) {
      setOrderValue("reduce_only", false);
    }
  }, [tpslSwitch]);
  return {
    ...state,
    slPriceError: slPriceError ?? void 0,
    side: formattedOrder.side,
    type: formattedOrder.order_type,
    level: formattedOrder.level,
    formattedOrder,
    setOrderValue,
    setOrderValues,
    currentLeverage,
    // cancelTP_SL,
    // enableTP_SL,
    tpslSwitch,
    setTpslSwitch: onTPSLSwitchChanged,
    symbolInfo,
    onFocus,
    onBlur,
    priceInputRef,
    priceInputContainerRef,
    priceInputContainerWidth,
    triggerPriceInputRef,
    activatedPriceInputRef,
    lastQuantityInputType,
    canTrade,
    bboStatus,
    bboType,
    onBBOChange,
    toggleBBO,
    currentLtv,
    fillMiddleValue,
    symbol,
    soundAlert,
    setSoundAlert,
    currentFocusInput
  };
};
var OrderEntryWidget = (props) => {
  const state = useOrderEntryScript(props);
  return /* @__PURE__ */ jsx(
    OrderEntry,
    {
      ...state,
      containerRef: props.containerRef,
      disableFeatures: props.disableFeatures
    }
  );
};

export { AdditionalInfo, FeesWidget, LTVRiskTooltipWidget, OrderConfirmDialog, OrderEntry, OrderEntryWidget, useOrderEntryScript };
