// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/instructions/initDvn.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { InitDvnParams, initDvnParamsBeet } from '../types/InitDvnParams'

/**
 * @category Instructions
 * @category InitDvn
 * @category generated
 */
export type InitDvnInstructionArgs = {
  params: InitDvnParams
}
/**
 * @category Instructions
 * @category InitDvn
 * @category generated
 */
export const initDvnStruct = new beet.FixableBeetArgsStruct<
  InitDvnInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', initDvnParamsBeet],
  ],
  'InitDvnInstructionArgs'
)
/**
 * Accounts required by the _initDvn_ instruction
 *
 * @property [_writable_, **signer**] payer
 * @property [_writable_] config
 * @category Instructions
 * @category InitDvn
 * @category generated
 */
export type InitDvnInstructionAccounts = {
  payer: web3.PublicKey
  config: web3.PublicKey
  systemProgram?: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const initDvnInstructionDiscriminator = [
  58, 108, 132, 12, 20, 10, 215, 210,
]

/**
 * Creates a _InitDvn_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitDvn
 * @category generated
 */
export function createInitDvnInstruction(
  accounts: InitDvnInstructionAccounts,
  args: InitDvnInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = initDvnStruct.serialize({
    instructionDiscriminator: initDvnInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _InitDvn_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category InitDvn
 * @category generated
 */
export function createInitDvnInstructionAccounts(
  accounts: InitDvnInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
