// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/instructions/invoke.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { InvokeParams, invokeParamsBeet } from '../types/InvokeParams'

/**
 * @category Instructions
 * @category Invoke
 * @category generated
 */
export type InvokeInstructionArgs = {
  params: InvokeParams
}
/**
 * @category Instructions
 * @category Invoke
 * @category generated
 */
export const invokeStruct = new beet.FixableBeetArgsStruct<
  InvokeInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', invokeParamsBeet],
  ],
  'InvokeInstructionArgs'
)
/**
 * Accounts required by the _invoke_ instruction
 *
 * @property [_writable_, **signer**] signer
 * @property [_writable_] config
 * @property [_writable_] executeHash
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category Invoke
 * @category generated
 */
export type InvokeInstructionAccounts = {
  signer: web3.PublicKey
  config: web3.PublicKey
  executeHash: web3.PublicKey
  systemProgram?: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const invokeInstructionDiscriminator = [26, 16, 169, 7, 21, 202, 242, 25]

/**
 * Creates a _Invoke_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Invoke
 * @category generated
 */
export function createInvokeInstruction(
  accounts: InvokeInstructionAccounts,
  args: InvokeInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = invokeStruct.serialize({
    instructionDiscriminator: invokeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.executeHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _Invoke_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Invoke
 * @category generated
 */
export function createInvokeInstructionAccounts(
  accounts: InvokeInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.signer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.executeHash,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
