// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/instructions/quoteDvn.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { QuoteDvnParams, quoteDvnParamsBeet } from '../types/QuoteDvnParams'

/**
 * @category Instructions
 * @category QuoteDvn
 * @category generated
 */
export type QuoteDvnInstructionArgs = {
  params: QuoteDvnParams
}
/**
 * @category Instructions
 * @category QuoteDvn
 * @category generated
 */
export const quoteDvnStruct = new beet.FixableBeetArgsStruct<
  QuoteDvnInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', quoteDvnParamsBeet],
  ],
  'QuoteDvnInstructionArgs'
)
/**
 * Accounts required by the _quoteDvn_ instruction
 *
 * @property [] dvnConfig
 * @property [] priceFeedProgram
 * @property [] priceFeedConfig
 * @category Instructions
 * @category QuoteDvn
 * @category generated
 */
export type QuoteDvnInstructionAccounts = {
  dvnConfig: web3.PublicKey
  priceFeedProgram: web3.PublicKey
  priceFeedConfig: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const quoteDvnInstructionDiscriminator = [
  223, 98, 51, 200, 123, 133, 132, 194,
]

/**
 * Creates a _QuoteDvn_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category QuoteDvn
 * @category generated
 */
export function createQuoteDvnInstruction(
  accounts: QuoteDvnInstructionAccounts,
  args: QuoteDvnInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = quoteDvnStruct.serialize({
    instructionDiscriminator: quoteDvnInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.dvnConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.priceFeedProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.priceFeedConfig,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _QuoteDvn_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category QuoteDvn
 * @category generated
 */
export function createQuoteDvnInstructionAccounts(
  accounts: QuoteDvnInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.dvnConfig,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.priceFeedProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.priceFeedConfig,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
