// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/instructions/setConfig.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import { SetConfigParams, setConfigParamsBeet } from '../types/SetConfigParams'

/**
 * @category Instructions
 * @category SetConfig
 * @category generated
 */
export type SetConfigInstructionArgs = {
  params: SetConfigParams
}
/**
 * @category Instructions
 * @category SetConfig
 * @category generated
 */
export const setConfigStruct = new beet.FixableBeetArgsStruct<
  SetConfigInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', setConfigParamsBeet],
  ],
  'SetConfigInstructionArgs'
)
/**
 * Accounts required by the _setConfig_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] config
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category SetConfig
 * @category generated
 */
export type SetConfigInstructionAccounts = {
  admin: web3.PublicKey
  config: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const setConfigInstructionDiscriminator = [
  108, 158, 154, 175, 212, 98, 52, 66,
]

/**
 * Creates a _SetConfig_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category SetConfig
 * @category generated
 */
export function createSetConfigInstruction(
  accounts: SetConfigInstructionAccounts,
  args: SetConfigInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = setConfigStruct.serialize({
    instructionDiscriminator: setConfigInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _SetConfig_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetConfig
 * @category generated
 */
export function createSetConfigInstructionAccounts(
  accounts: SetConfigInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
