// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/dvn/instructions/withdrawFee.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as beet from '@metaplex-foundation/beet'
import * as web3 from '@solana/web3.js'
import {
  WithdrawFeeParams,
  withdrawFeeParamsBeet,
} from '../types/WithdrawFeeParams'

/**
 * @category Instructions
 * @category WithdrawFee
 * @category generated
 */
export type WithdrawFeeInstructionArgs = {
  params: WithdrawFeeParams
}
/**
 * @category Instructions
 * @category WithdrawFee
 * @category generated
 */
export const withdrawFeeStruct = new beet.BeetArgsStruct<
  WithdrawFeeInstructionArgs & {
    instructionDiscriminator: number[] /* size: 8 */
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['params', withdrawFeeParamsBeet],
  ],
  'WithdrawFeeInstructionArgs'
)
/**
 * Accounts required by the _withdrawFee_ instruction
 *
 * @property [**signer**] admin
 * @property [_writable_] config
 * @property [_writable_] receiver
 * @property [] eventAuthority
 * @property [] program
 * @category Instructions
 * @category WithdrawFee
 * @category generated
 */
export type WithdrawFeeInstructionAccounts = {
  admin: web3.PublicKey
  config: web3.PublicKey
  receiver: web3.PublicKey
  eventAuthority: web3.PublicKey
  program: web3.PublicKey
  anchorRemainingAccounts?: web3.AccountMeta[]
}

export const withdrawFeeInstructionDiscriminator = [
  14, 122, 231, 218, 31, 238, 223, 150,
]

/**
 * Creates a _WithdrawFee_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category WithdrawFee
 * @category generated
 */
export function createWithdrawFeeInstruction(
  accounts: WithdrawFeeInstructionAccounts,
  args: WithdrawFeeInstructionArgs,
  programId: web3.PublicKey
) {
  const [data] = withdrawFeeStruct.serialize({
    instructionDiscriminator: withdrawFeeInstructionDiscriminator,
    ...args,
  })
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiver,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  const ix = new web3.TransactionInstruction({
    programId,
    keys,
    data,
  })
  return ix
}

/**
 * Creates a _WithdrawFee_ instructionAccounts.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category WithdrawFee
 * @category generated
 */
export function createWithdrawFeeInstructionAccounts(
  accounts: WithdrawFeeInstructionAccounts,
  programId: web3.PublicKey
) {
  const keys: web3.AccountMeta[] = [
    {
      pubkey: accounts.admin,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: accounts.config,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.receiver,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: accounts.eventAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: accounts.program,
      isWritable: false,
      isSigner: false,
    },
  ]

  if (accounts.anchorRemainingAccounts != null) {
    for (const acc of accounts.anchorRemainingAccounts) {
      keys.push(acc)
    }
  }

  return keys
}
