// @layerzerolabs/tron-utilities/dist/index.mjs
import { ethers } from 'ethers';
import TronWeb from 'tronweb';
export { default as TronWeb } from 'tronweb';
import { getLogger, sleep } from '@layerzerolabs/lz-utilities';

// src/tron.ts

// src/time.ts
var MILLISECOND = 1;
var SECOND = 1e3 * MILLISECOND;
var MINUTE = 60 * SECOND;
var HOUR = 60 * MINUTE;
var DAY = 24 * HOUR;
var Time = {
  /**
   * Millisecond unit.
   */
  MILLISECOND,
  /**
   * Second unit.
   */
  SECOND,
  /**
   * Minute unit.
   */
  MINUTE,
  /**
   * Hour unit.
   */
  HOUR,
  /**
   * Day unit.
   */
  DAY,
  /**
   * Gets the current time in milliseconds since the Unix epoch.
   * @returns {number} The current time in milliseconds.
   */
  get NOW() {
    return Date.now();
  }
};

// src/url.ts
var DefaultProviderConversionStrategy = class {
  /**
   * Converts a URL to its REST endpoint equivalent.
   *
   * @param {string} url - The URL to convert.
   * @returns {string} The converted REST endpoint URL.
   */
  asREST(url) {
    return url.replace(/\/jsonrpc$/, "");
  }
  /**
   * Converts a URL to its JSON-RPC endpoint equivalent.
   *
   * @param {string} url - The URL to convert.
   * @returns {string} The converted JSON-RPC endpoint URL.
   */
  asJSONRPC(url) {
    return url.trim().replace(/\/?$/, "/jsonrpc");
  }
};
var AnkrProviderConversionStrategy = class {
  /**
   * Converts a URL to its REST endpoint equivalent.
   *
   * @param {string} url - The URL to convert.
   * @returns {string} The converted REST endpoint URL.
   */
  asREST(url) {
    if (/tron\/[0-9a-fA-F]{10,}/.test(url)) {
      return url.replace(/\/tron_jsonrpc$/, "/premium-http/tron");
    }
    return url.replace(/\/tron_jsonrpc$/, "/http/tron");
  }
  /**
   * Converts a URL to its JSON-RPC endpoint equivalent.
   *
   * @param {string} url - The URL to convert.
   * @returns {string} The converted JSON-RPC endpoint URL.
   */
  asJSONRPC(url) {
    return url.trim().replace(/\/(premium-http|http)\/tron/, "/tron_jsonrpc");
  }
};
var QuiknodeProviderConversionStrategy = class {
  /**
   * Converts a URL to its REST endpoint equivalent.
   *
   * @param {string} url - The URL to convert.
   * @returns {string} The converted REST endpoint URL.
   */
  asREST(url) {
    return url.replace(/\/jsonrpc$/, "");
  }
  /**
   * Converts a URL to its JSON-RPC endpoint equivalent.
   *
   * @param {string} url - The URL to convert.
   * @returns {string} The converted JSON-RPC endpoint URL.
   */
  asJSONRPC(url) {
    return url.trim().replace(/\/?$/, "/jsonrpc");
  }
};
var _StrategyFactory = class _StrategyFactory {
  /**
   * Gets the appropriate provider conversion strategy for a given URL.
   *
   * @param {string} url - The URL to get the conversion strategy for.
   * @returns {ProviderConversionStrategy} The provider conversion strategy.
   */
  static get(url) {
    for (const [pattern, strategy] of this.strategies) {
      if (pattern.test(url)) {
        return strategy;
      }
    }
    return new DefaultProviderConversionStrategy();
  }
};
/**
 * A map of URL patterns to provider conversion strategies.
 */
_StrategyFactory.strategies = /* @__PURE__ */ new Map();
_StrategyFactory.strategies.set(/api.trongrid.io/, new DefaultProviderConversionStrategy());
_StrategyFactory.strategies.set(/ankr.com/, new AnkrProviderConversionStrategy());
_StrategyFactory.strategies.set(/quiknode.pro/, new QuiknodeProviderConversionStrategy());
var StrategyFactory = _StrategyFactory;

// src/tron.ts
var TronTransactionType = /* @__PURE__ */ ((TronTransactionType2) => {
  TronTransactionType2["TRIGGER_SMART_CONTRACT"] = "TriggerSmartContract";
  return TronTransactionType2;
})(TronTransactionType || {});
getLogger();
function createTronWeb(signer) {
  if (!(signer instanceof TronWeb)) {
    const fullHostUrl = typeof signer.rpcSetting === "string" ? signer.rpcSetting : signer.rpcSetting.url;
    const client = new TronWeb({
      fullHost: StrategyFactory.get(fullHostUrl).asREST(fullHostUrl),
      headers: {
        "TRON-PRO-API-KEY": typeof signer.rpcSetting === "string" ? "" : signer.rpcSetting.apiKey ?? ""
      },
      privateKey: signer.privateKey.replace(/^0x/, "")
    });
    return client;
  }
  return signer;
}
async function signAndSendTronTransaction(txn, signer, permissionId) {
  const signedTxn = await signTronTx(txn, signer, permissionId);
  const client = createTronWeb(signer);
  return sendTronTransaction(signedTxn, client);
}
async function sendTronTransaction(singedTxn, client) {
  const broadcastResponse = await client.trx.sendRawTransaction(singedTxn);
  if (broadcastResponse.code !== void 0) {
    const err = {
      error: broadcastResponse.code,
      message: TronWeb.toUtf8(broadcastResponse.message ?? "")
    };
    throw new Error(JSON.stringify(err, null, 2));
  }
  return {
    txId: singedTxn.txID,
    result: broadcastResponse,
    wait: async () => {
      return checkResult(client, singedTxn.txID);
    }
  };
}
async function signTronTx(txn, _signer, permissionId) {
  const signer = createTronWeb(_signer);
  if (typeof signer.defaultAddress.hex !== "string") {
    throw new Error(`signer.defaultAddress.hex is not a valid hex string`);
  }
  const partialTxn = {
    visible: false,
    txID: "",
    raw_data_hex: "",
    raw_data: {
      contract: [
        {
          parameter: {
            value: {
              data: txn.data.replace(/^0x/, ""),
              contract_address: TronWeb.address.toHex(txn.to),
              owner_address: signer.defaultAddress.hex
            },
            type_url: `type.googleapis.com/protocol.${"TriggerSmartContract" /* TRIGGER_SMART_CONTRACT */}`
          },
          type: "TriggerSmartContract" /* TRIGGER_SMART_CONTRACT */,
          Permission_id: permissionId
        }
      ]
    }
  };
  const fullTxn = await fullFillTxn(signer, partialTxn, {
    gasLimit: txn.gasLimit,
    gasPrice: txn.gasPrice,
    energyFactor: txn.energyFactor,
    value: txn.value
  });
  const signedTransaction = await signer.trx.sign(fullTxn);
  if (signedTransaction.signature === void 0) {
    throw new Error("Invalid private key provided");
  }
  return signedTransaction;
}
var gasPriceCache = { time: 0, value: "" };
var energyFactorsCache = /* @__PURE__ */ new Map();
async function fullFillTxn(signer, txn, overrides = {}) {
  const apiKey = signer.fullNode.headers["TRON-PRO-API-KEY"];
  const etherProvider = new ethers.providers.StaticJsonRpcProvider({
    url: StrategyFactory.get(signer.fullNode.host).asJSONRPC(signer.fullNode.host),
    headers: {
      ...typeof apiKey !== "undefined" ? { "TRON-PRO-API-KEY": apiKey } : {}
    }
  });
  async function getGasPrice() {
    const { time, value } = gasPriceCache;
    if (time > Time.NOW - 15 * Time.SECOND && value != "") return value;
    const newGaspPrice = await etherProvider.getGasPrice().then((x) => x.toString());
    gasPriceCache = { time: Time.NOW, value: newGaspPrice };
    return newGaspPrice;
  }
  async function getGasLimit(raw) {
    const gasLimit = await etherProvider.estimateGas({
      to: raw.raw_data.contract[0].parameter.value.contract_address.replace(/^41/, "0x"),
      // need to slice off the "41" prefix that Tron appends to addresses
      data: "0x" + raw.raw_data.contract[0].parameter.value.data,
      from: raw.raw_data.contract[0].parameter.value.owner_address.replace(/^41/, "0x"),
      value: overrides.value ?? 0
    });
    return gasLimit.toNumber();
  }
  async function getEnergyFactor(raw) {
    const addr = raw.raw_data.contract[0].parameter.value.contract_address;
    const cached = energyFactorsCache.get(addr);
    if (cached && cached.time > Time.NOW - 10 * Time.MINUTE) return cached.value;
    const MAX_ENERGY_FACTOR = 1.2;
    const response = await signer.fullNode.request(
      "wallet/getcontractinfo",
      { value: addr, visible: false },
      "post"
    );
    const contractInfo = response;
    const energy_factor = contractInfo.contract_state.energy_factor ?? MAX_ENERGY_FACTOR;
    energyFactorsCache.set(addr, { time: Time.NOW, value: energy_factor });
    return energy_factor;
  }
  async function getFeeLimit(raw, overrides2 = {}) {
    const MAX_FEE_LIMIT = 15e9;
    const energy_factor = overrides2.energyFactor ?? await getEnergyFactor(raw);
    const gasLimit = overrides2.gasLimit ?? await getGasLimit(raw);
    const energyPrice = overrides2.gasPrice ?? Number(await getGasPrice());
    const factor = 1 + energy_factor;
    const feeLimit = gasLimit * factor * energyPrice;
    const feeLimit_adjusted = Math.ceil(feeLimit) * 1.2;
    return Math.floor(Math.min(feeLimit_adjusted, MAX_FEE_LIMIT));
  }
  const metadata = await signer.fullNode.request("wallet/getblock", { detail: false }, "post");
  const ref_block_bytes = metadata.block_header.raw_data.number.toString(16).slice(-4).padStart(4, "0");
  const ref_block_hash = metadata.blockID.slice(16, 32);
  const expiration = metadata.block_header.raw_data.timestamp + 60 * 1e3;
  const { timestamp } = metadata.block_header.raw_data;
  const tx = {
    ...txn,
    raw_data: {
      contract: [
        {
          parameter: {
            value: {
              ...txn.raw_data.contract[0].parameter.value,
              call_value: overrides.value ?? 0
            },
            type_url: txn.raw_data.contract[0].parameter.type_url
          },
          type: txn.raw_data.contract[0].type
        }
      ],
      ref_block_bytes,
      ref_block_hash,
      expiration,
      fee_limit: await getFeeLimit(txn, overrides),
      timestamp
    }
  };
  const pb = signer.utils.transaction.txJsonToPb(tx);
  tx.txID = signer.utils.transaction.txPbToTxID(pb).replace(/^0x/, "");
  tx.raw_data_hex = signer.utils.transaction.txPbToRawDataHex(pb).toLowerCase();
  return tx;
}
async function checkResult(tronWeb, txId) {
  const startTime = Date.now();
  const waitTimeout = 1e3 * 60 * 5;
  const wait = async function(txId2) {
    let tries = 0;
    while (tries++ < 20) {
      const output2 = await tronWeb.trx.getTransactionInfo(txId2);
      if (Object.keys(output2).length > 0) {
        return output2;
      }
      if (Date.now() - startTime > waitTimeout) {
        throw new Error("Timeout waiting for transaction to be processed");
      }
      await sleep(1200);
    }
    throw new Error(`Failed to get transactionInfo for txId: ${txId2}`);
  };
  const output = await wait(txId);
  if (output.result !== void 0 && output.result === "FAILED") {
    throw new Error(TronWeb.toUtf8(output.resMessage ?? ""));
  }
  if (!Object.prototype.hasOwnProperty.call(output, "contractResult")) {
    throw new Error("Failed to execute: " + JSON.stringify(output, null, 2));
  }
  return output;
}

export { StrategyFactory, TronTransactionType, checkResult, createTronWeb, fullFillTxn, sendTronTransaction, signAndSendTronTransaction, signTronTx };
