// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/endpoint/types/LzComposeAlertParams.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

import * as web3 from '@solana/web3.js'
import * as beet from '@metaplex-foundation/beet'
import * as beetSolana from '@metaplex-foundation/beet-solana'
export type LzComposeAlertParams = {
  from: web3.PublicKey
  to: web3.PublicKey
  guid: number[] /* size: 32 */
  index: number
  computeUnits: beet.bignum
  value: beet.bignum
  message: Uint8Array
  extraData: Uint8Array
  reason: Uint8Array
}

/**
 * @category userTypes
 * @category generated
 */
export const lzComposeAlertParamsBeet =
  new beet.FixableBeetArgsStruct<LzComposeAlertParams>(
    [
      ['from', beetSolana.publicKey],
      ['to', beetSolana.publicKey],
      ['guid', beet.uniformFixedSizeArray(beet.u8, 32)],
      ['index', beet.u16],
      ['computeUnits', beet.u64],
      ['value', beet.u64],
      ['message', beet.bytes],
      ['extraData', beet.bytes],
      ['reason', beet.bytes],
    ],
    'LzComposeAlertParams'
  )
