// @layerzerolabs/lz-solana-sdk-v2/src/generated/solita/executor/errors/index.ts
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * InvalidSize: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidSizeError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'InvalidSize'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidSizeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new InvalidSizeError())
createErrorFromNameLookup.set('InvalidSize', () => new InvalidSizeError())

/**
 * Paused: ''
 *
 * @category Errors
 * @category generated
 */
export class PausedError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'Paused'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, PausedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new PausedError())
createErrorFromNameLookup.set('Paused', () => new PausedError())

/**
 * UnsupportedOptionType: ''
 *
 * @category Errors
 * @category generated
 */
export class UnsupportedOptionTypeError extends Error {
  readonly code: number = 0x1772
  readonly name: string = 'UnsupportedOptionType'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, UnsupportedOptionTypeError)
    }
  }
}

createErrorFromCodeLookup.set(0x1772, () => new UnsupportedOptionTypeError())
createErrorFromNameLookup.set(
  'UnsupportedOptionType',
  () => new UnsupportedOptionTypeError()
)

/**
 * ZeroLzComposeGasProvided: ''
 *
 * @category Errors
 * @category generated
 */
export class ZeroLzComposeGasProvidedError extends Error {
  readonly code: number = 0x1773
  readonly name: string = 'ZeroLzComposeGasProvided'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ZeroLzComposeGasProvidedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1773, () => new ZeroLzComposeGasProvidedError())
createErrorFromNameLookup.set(
  'ZeroLzComposeGasProvided',
  () => new ZeroLzComposeGasProvidedError()
)

/**
 * ZeroLzReceiveGasProvided: ''
 *
 * @category Errors
 * @category generated
 */
export class ZeroLzReceiveGasProvidedError extends Error {
  readonly code: number = 0x1774
  readonly name: string = 'ZeroLzReceiveGasProvided'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ZeroLzReceiveGasProvidedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1774, () => new ZeroLzReceiveGasProvidedError())
createErrorFromNameLookup.set(
  'ZeroLzReceiveGasProvided',
  () => new ZeroLzReceiveGasProvidedError()
)

/**
 * NativeAmountExceedsCap: ''
 *
 * @category Errors
 * @category generated
 */
export class NativeAmountExceedsCapError extends Error {
  readonly code: number = 0x1775
  readonly name: string = 'NativeAmountExceedsCap'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, NativeAmountExceedsCapError)
    }
  }
}

createErrorFromCodeLookup.set(0x1775, () => new NativeAmountExceedsCapError())
createErrorFromNameLookup.set(
  'NativeAmountExceedsCap',
  () => new NativeAmountExceedsCapError()
)

/**
 * NotAdmin: ''
 *
 * @category Errors
 * @category generated
 */
export class NotAdminError extends Error {
  readonly code: number = 0x1776
  readonly name: string = 'NotAdmin'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, NotAdminError)
    }
  }
}

createErrorFromCodeLookup.set(0x1776, () => new NotAdminError())
createErrorFromNameLookup.set('NotAdmin', () => new NotAdminError())

/**
 * NotExecutor: ''
 *
 * @category Errors
 * @category generated
 */
export class NotExecutorError extends Error {
  readonly code: number = 0x1777
  readonly name: string = 'NotExecutor'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, NotExecutorError)
    }
  }
}

createErrorFromCodeLookup.set(0x1777, () => new NotExecutorError())
createErrorFromNameLookup.set('NotExecutor', () => new NotExecutorError())

/**
 * MsgLibNotAllowed: ''
 *
 * @category Errors
 * @category generated
 */
export class MsgLibNotAllowedError extends Error {
  readonly code: number = 0x1778
  readonly name: string = 'MsgLibNotAllowed'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, MsgLibNotAllowedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1778, () => new MsgLibNotAllowedError())
createErrorFromNameLookup.set(
  'MsgLibNotAllowed',
  () => new MsgLibNotAllowedError()
)

/**
 * TooManyAdmins: ''
 *
 * @category Errors
 * @category generated
 */
export class TooManyAdminsError extends Error {
  readonly code: number = 0x1779
  readonly name: string = 'TooManyAdmins'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, TooManyAdminsError)
    }
  }
}

createErrorFromCodeLookup.set(0x1779, () => new TooManyAdminsError())
createErrorFromNameLookup.set('TooManyAdmins', () => new TooManyAdminsError())

/**
 * TooManyExecutors: ''
 *
 * @category Errors
 * @category generated
 */
export class TooManyExecutorsError extends Error {
  readonly code: number = 0x177a
  readonly name: string = 'TooManyExecutors'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, TooManyExecutorsError)
    }
  }
}

createErrorFromCodeLookup.set(0x177a, () => new TooManyExecutorsError())
createErrorFromNameLookup.set(
  'TooManyExecutors',
  () => new TooManyExecutorsError()
)

/**
 * TooManyOptionTypes: ''
 *
 * @category Errors
 * @category generated
 */
export class TooManyOptionTypesError extends Error {
  readonly code: number = 0x177b
  readonly name: string = 'TooManyOptionTypes'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, TooManyOptionTypesError)
    }
  }
}

createErrorFromCodeLookup.set(0x177b, () => new TooManyOptionTypesError())
createErrorFromNameLookup.set(
  'TooManyOptionTypes',
  () => new TooManyOptionTypesError()
)

/**
 * InvalidNativeDropRequestsLength: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidNativeDropRequestsLengthError extends Error {
  readonly code: number = 0x177c
  readonly name: string = 'InvalidNativeDropRequestsLength'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidNativeDropRequestsLengthError)
    }
  }
}

createErrorFromCodeLookup.set(
  0x177c,
  () => new InvalidNativeDropRequestsLengthError()
)
createErrorFromNameLookup.set(
  'InvalidNativeDropRequestsLength',
  () => new InvalidNativeDropRequestsLengthError()
)

/**
 * InvalidNativeDropReceiver: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidNativeDropReceiverError extends Error {
  readonly code: number = 0x177d
  readonly name: string = 'InvalidNativeDropReceiver'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidNativeDropReceiverError)
    }
  }
}

createErrorFromCodeLookup.set(
  0x177d,
  () => new InvalidNativeDropReceiverError()
)
createErrorFromNameLookup.set(
  'InvalidNativeDropReceiver',
  () => new InvalidNativeDropReceiverError()
)

/**
 * InsufficientBalance: ''
 *
 * @category Errors
 * @category generated
 */
export class InsufficientBalanceError extends Error {
  readonly code: number = 0x177e
  readonly name: string = 'InsufficientBalance'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InsufficientBalanceError)
    }
  }
}

createErrorFromCodeLookup.set(0x177e, () => new InsufficientBalanceError())
createErrorFromNameLookup.set(
  'InsufficientBalance',
  () => new InsufficientBalanceError()
)

/**
 * EidNotSupported: ''
 *
 * @category Errors
 * @category generated
 */
export class EidNotSupportedError extends Error {
  readonly code: number = 0x177f
  readonly name: string = 'EidNotSupported'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, EidNotSupportedError)
    }
  }
}

createErrorFromCodeLookup.set(0x177f, () => new EidNotSupportedError())
createErrorFromNameLookup.set(
  'EidNotSupported',
  () => new EidNotSupportedError()
)

/**
 * ExecutorIsAdmin: ''
 *
 * @category Errors
 * @category generated
 */
export class ExecutorIsAdminError extends Error {
  readonly code: number = 0x1780
  readonly name: string = 'ExecutorIsAdmin'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ExecutorIsAdminError)
    }
  }
}

createErrorFromCodeLookup.set(0x1780, () => new ExecutorIsAdminError())
createErrorFromNameLookup.set(
  'ExecutorIsAdmin',
  () => new ExecutorIsAdminError()
)

/**
 * InvalidOwner: ''
 *
 * @category Errors
 * @category generated
 */
export class InvalidOwnerError extends Error {
  readonly code: number = 0x1781
  readonly name: string = 'InvalidOwner'
  constructor() {
    super('')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InvalidOwnerError)
    }
  }
}

createErrorFromCodeLookup.set(0x1781, () => new InvalidOwnerError())
createErrorFromNameLookup.set('InvalidOwner', () => new InvalidOwnerError())

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
